# Parliament2

**NOTE: This fork of the DNAnexus version of Parliament2 (slzarate/parliament2) is currently under construction. Please continue to use the DNAnexus version (installation details below) until further notice.**

Parliament2 identifies structural variants in a given sample relative to a reference genome. These structural variants cover large deletion events that are called as Deletions of a region, Insertions of a sequence into a region, Duplications of a region, Inversions of a region, or Translocations between two regions in the genome.

Parliament2 runs a combination of tools to generate structural variant calls on whole-genome sequencing data. It can run the following callers: Breakdancer, Breakseq2, CNVnator, Delly2, Manta, and Lumpy. Because of synergies in how the programs use computational resources, these are all run in parallel. Parliament2 will produce the outputs of each of the tools for subsequent investigation.

If the option to genotype candidates is selected, this tool will run SVTyper to genotype the events and will merge these tools with the program SURVIVOR. SV events that receive genotype calls have significantly higher specificity.

If the option to visualize candidates is selected, a `.tar.gz` file containing PDF images for each SV call will be produced in order to summarize the supporting information behind the SV.

## Getting Started

### Prerequisites

You must have [Docker](https://www.docker.com/) installed in order to run Parliament2. No other dependencies are required. Please see the Docker documentation for installation information.

### Licensing
This software uses an Apache 2.0 license, which is contained within this Github repo.

### Installing

Parliament2 is meant to be run as a Docker image and is [available on DockerHub](https://hub.docker.com/r/dnanexus/parliament2/). The simplest way to install Parliament2 on your machine is to run the command `docker pull dnanexus/parliament2:<TAG>`, where the `<TAG>` is the version of Parliament2 you wish to use. You can see the full list of Parliament2 versions available on DockerHub at the [Parliament2 DockerHub repository](https://hub.docker.com/r/dnanexus/parliament2/tags/).

You can verify the successful installation using the command `docker images`, which will list every Docker image stored locally on your machine, or by running `docker run dnanexus/parliament2:<TAG> -h`, which will print the help string (printed in full at the bottom of this README).

#### Building from source

If you wish to build Parliament2 from source, you can find the most up-to-date production version of Parliament2 on our [release page](https://github.com/dnanexus/parliament2/releases) or pull from this repository (warning: this version may be unstable; the official releases will be more stable).

Once you have made your desired modifications (if applicable), run:

```
cd parliament2
tar -czvf resources.tar.gz resources/
docker build . -t parliament2
```

There should now be a Docker image named `parliament2` on your computer.

Again, you can verify the successful installation using the command `docker images`, which will list every Docker image stored locally on your machine, or by running `docker run parliament2 -h`, which will print the help string (printed in full at the bottom of this README).

## Running

### Required inputs

To run this Docker image, you must have:

* A local file with Illumina reads on which you want to call structural variants (`*.bam`)
* A reference genome file to which the BAM file was mapped (`*.fa.gz` or `*.fasta.gz`)

You may also have:

* The index for the BAM file (`*.bai`)
* The index for the reference genome file (`*.fai`)

This app is intended for whole-genome sequencing. Providing exome or panel sequencing will not produce good results. It is intended to be run on a single germline sample.

### Command-line call

To run Parliament2, use the following command-line call:

```
docker run -v <LOCAL_DIR_WITH_INPUTS>:/home/dnanexus/in -v <LOCAL_DIR_FOR_OUTPUTS>:/home/dnanexus/out dnanexus/parliament2:<TAG> --bam <BAM_NAME> --bai <INDEX_NAME> --fai <REFERENCE_INDEX> -r <REFERENCE_NAME> <OPTIONAL_ARGUMENTS>
```

You must mount two local volumes -- one that contains the inputs (`*.bam` and `*.fa.gz`) and one in which you wish to place the files generated by Parliament2. The `--bai` and `--fai` flags, along with the BAM and FASTA index files, are optional, but including either or both will speed up runtime considerably. At least one optional argument must be specified; see [below](#help) for more information on the optional arguments.

For example, if your current working directory is named `/home/dnanexus/` and looks like this:

```
.
├── input.bam
├── input.bai
├── ref_genome.fa.gz
└── outputs
```

where `outputs` is the directory you wish Parliament2 to place its results files, the command you would use to run Parliament2 could be:

```
docker run -v /home/dnanexus/:/home/dnanexus/in -v /home/dnanexus/outputs/:/home/dnanexus/out dnanexus/parliament2<TAG> --bam input.bam --bai input.bai -r ref_genome.fa.gz --breakdancer --cnvnator --manta --genotype
```

### Files generated

This app will output a number of files, representing the outputs of each of the structural variant callers. If the option to run a given step is unselected, then those outputs will not be provided.

- `lumpy.vcf`: representing the structural variant calls from Lumpy in VCF format.
- `lumpy.discordant.bam`: representing reads that Lumpy identified as discordant and used as evidence for its calls.
- `lumpy.splitters.bam`: representing reads that Lumpy identified as split-read mapped and used as evidence for its calls.
- `manta.diploidSV.vcf`: representing genotyped structural variant calls from Manta in VCF format.
- `manta.alignmentStats.txt`: representing statistics about the alignment from Manta.
- `breakdancer.ctx`: representing structural variant calls in Breakdancer's format.
- `cnvnator.output`: representing structural variant calls in CNVnator's format.
- `cnvnator.vcf`: representing structural variant calls from CNVnator in VCF format (this represents the conversion of CNVnator output to VCF format).
- `breakseq.gff`: representing structural variant calls from Breakseq2 in GFF format.
- `breakseq.vcf`: representing structural variant calls from Breakseq2 in VCF format.
- `breakseq.bam`: representing the reads mapping used as evidence for the calls generated by Breakseq2.
- `delly.deletion.vcf`: representing deletion calls made by Delly2.
- `delly.inversion.vcf`: representing inversion calls made by Delly2.
- `delly.duplication.vcf`: representing duplication calls made by Delly2.
- `delly.insertion.vcf`: representing insertion calls made by Delly2.
- `delly.translocation.vcf`: representing translocation calls made by Delly2.

- `<caller>.svtyped.vcf`: If the option to genotype candidates is selected, a genotype VCF produced by SVTyper will be generated for each caller output.
- `combined.genotyped.vcf`: If the option to genotype candidates is selected, a merged VCF file of all of the callers will be produced by SURVIVOR.

- `svviz_outputs.tar.gz`: If the option to visualize events is selected, a tarball containing a set of PDFs documenting the genomic regions for calls will be generated

## Running Parliament2 on DNAnexus

Parliament2 is available as an app on DNAnexus at https://platform.dnanexus.com/app/parliament2 (note: a DNAnexus account is required to access this link; you can create one at https://platform.dnanexus.com/login). The documentation for the app is included both on the DNAnexus platform and in the `dx_app_code` directory of this repository. A DNAnexus account is required to access the platform.

To run Parliament2 on DNAnexus, your input BAM file must be already on the DNAnexus platform. To run Parliament2 using the graphic interface, simply click the "Run" button from the app page and select your inputs. To run Parliament2 using the command-line interface, run the command `dx run parliament2 -h` and follow the guide generated. For more information on running executables on DNAnexus, see the [guide to running apps and applets](https://wiki.dnanexus.com/Command-Line-Client/Running-Apps-and-Applets).

General information on using DNAnexus can be found in the [official documentation](https://wiki.dnanexus.com/Home).

### Building Parliament2 on DNAnexus

To build Parliament2 on your own on DNAnexus, you will have to have built the Docker image locally (see [Installing](#installing)). Then:

* Run `dx-docker create-asset parliament2`. This will take approximately 45 minutes and will generate a string that you can copy-paste into the `dxapp.json` file found in the `dx_app_code/parliament2` directory under the "Regional Options" section for your region.

* Run `dx build parliament2` to build the applet from within the `dx_app_code` directory.

For more information about using DNAnexus, see the following links:

* [Tutorial for using Docker images on DNAnexus](https://wiki.dnanexus.com/Developer-Tutorials/Using-Docker-Images)
* [Tutorial for building apps on DNAnexus](https://wiki.dnanexus.com/Developer-Tutorials/Intro-to-Building-Apps) 

General information on using DNAnexus can be found in the [official documentation](https://wiki.dnanexus.com/Home).

#### Modifying Parliament2 on DNAnexus

To modify Parliament2 and run it on DNAnexus, please see the developer README in the `dx_app_code/parliament2` directory of this repository.

## More Information

### FAQ

#### Samtools error

> The tool runs but I am getting a samtools error:
```
samtools view: writing to standard output failed: Broken pipe
samtools view: error closing standard output: -1
```
> What's going on?

This is a known error message caused by splitting the BAM file. It doesn't affect the results in any way.

#### Pysam warning

> The tool runs but I am getting the following warning:
```
breakseq2 -2.2- has requirement pysam==0.7.7, but you'll have pysam 0.15.1 which is incompatible
```
> What's going on?

This is a known error message caused by how we currently manage the conflicting pysam versions required for BreakSeq and SVTyper. This issue should be resolved in v0.1.10.

#### Out of memory error

> The tool fails or runs for a long time and I see an error message in the log:
```
Out of memory: Kill process XXXX (svviz) score 112 or sacrifice child
```
> What's going on?

The svviz tool can at times consume a large amount of memory, causing the application to be killed. If you see this message in the log, you should kill the job if it is still running and retry on a machine with more memory or without enabling svviz. This issue should be resolved in v0.1.11.

### Tool versions

#### Structural variant callers
* Breakdancer: [v1.4.3](https://github.com/genome/breakdancer/releases/tag/v1.4.3)
* BreakSeq2: [v2.2](http://bioinform.github.io/breakseq2/)
* CNVnator: [v0.3.3](https://github.com/abyzovlab/CNVnator/commit/de012f2bccfd4e11e84cf685b19fc138115f2d0d)
* Delly: [v0.7.2](https://github.com/dellytools/delly/releases/tag/v0.7.2)
* Lumpy: [v0.2.13](https://github.com/arq5x/lumpy-sv/commit/f466f61e02680796192b055e4c084fbb23dcc692)
* Manta: [v1.4.0](https://anaconda.org/bioconda/manta)

#### Other tools
* SVTyper: [v0.7.0](https://github.com/hall-lab/svtyper/commit/5fc30763fd3025793ee712a563de800c010f6bea)
* svviz: [v1.5.2](https://github.com/svviz/svviz/commit/84acefa13bf0d4ad6e7e0f1d058aed6f16681142)
* SURVIVOR: [v1.0.3](https://github.com/fritzsedlazeck/SURVIVOR/commit/7c7731d71fa1cba017f470895fb3ef55f2812067)

### Additional notes

Because the field of structural variation is relatively new and complex, we viewed placing a dependency on all individual tools completing successfully as a requirement for a successful run to be too strict. In other words, if one of these tools fails while the others succeed, the app will output the results of the tools that completed and will not itself fail.

Breakseq2 may only be able to work when using the 1000 Genomes reference genome (hs37d5). For other reference genomes, you may not get Breakseq2 results.

### Supporting information

For additional information, please see the following papers:

- Lumpy: Ryan M Layer, Colby Chiang, Aaron R Quinlan, and Ira M Hall. 2014. "LUMPY: a Probabilistic Framework for Structural Variant Discovery." Genome Biology 15 (6): R84. doi:10.1186/gb-2014-15-6-r84
- Manta: Chen X, Schulz-Trieglaff O, Shaw R, Barnes B, Schlesinger F, Källberg M, Cox AJ, Kruglyak S, Saunders CT. 2015 "Manta: rapid detection of structural variants and indels for germline and cancer sequencing applications." Bioinformatics. doi: 10.1093/bioinformatics/btv710
- Breakdancer: Chen K, Wallis JW, McLellan MD, Larson DE, Kalicki JM, Pohl CS, McGrath SD, Wendl MC, Zhang Q, Locke DP, Shi X, Fulton RS, Ley TJ, Wilson RK, Ding L, Mardis ER. 2009. "BreakDancer: an algorithm for high-resolution mapping of genomic structural variation". Nature Methods 6. doi:10.1038/nmeth.1363
- Breakseq2: Abyzov A, Li S, Kim DR, Mohiyuddin M, Stütz AM, Parrish NF, Mu XJ, Clark W, Chen K, Hurles M, Korbel JO, Lam HYK, Lee C, Gerstein MB. 2015. "Analysis of deletion breakpoints from 1,092 humans reveals details of mutation mechanisms". Nature Communications 6. doi:10.1038/ncomms8256
- Delly2: Tobias Rausch, Thomas Zichner, Andreas Schlattl, Adrian M. Stuetz, Vladimir Benes, Jan O. Korbel. 2012. Delly: structural variant discovery by integrated paired-end and split-read analysis. Bioinformatics 28:i333-i339. doi: 10.1093/bioinformatics/bts378
- CNVnator: Abyzov A, Urban AE, Snyder M, Gerstein M. 2011. "CNVnator: an approach to discover, genotype, and characterize typical and atypical CNVs from family and population genome sequencing". Genome Research (6):974. doi: 10.1101/gr.114876.110
- Parliament: Adam C English, William J Salerno, Oliver A Hampton, Claudia Gonzaga-Jauregui, Shruthi Ambreth, Deborah I Ritter, Christine R Beck, Caleb F Davis, Mahmoud Dahdouli, Singer Ma, Andrew Carroll, Narayanan Veeraraghavan, Jeremy Bruestle, Becky Drees, Alex Hastie, Ernest T Lam, Simon White, Pamela Mishra, Min Wang, Yi Han, Feng Zhang, Pawel Stankiewicz, David A Wheeler, Jeffrey G Reid, Donna M Muzny, Jeffrey Rogers, Aniko Sabo, Kim C Worley, James R Lupski, Eric Boerwinkle and Richard A Gibbs. Assessing structural variation in a personal genome—towards a human reference diploid genome. BMC Genomics 2015, 16:286  doi:10.1186/s12864-015-1479-3.

## Help

```
usage: parliament2.py [-h] --bam BAM [--bai BAI] -r REF_GENOME [--fai FAI]
                      [--prefix PREFIX] [--filter_short_contigs]
                      [--breakdancer] [--breakseq] [--manta] [--cnvnator]
                      [--lumpy] [--delly_deletion] [--delly_insertion]
                      [--delly_inversion] [--delly_duplication] [--genotype]
                      [--svviz] [--svviz_only_validated_candidates]

Parliament2

optional arguments:
  -h, --help            show this help message and exit
  --bam BAM             The name of the Illumina BAM file for which to call
                        structural variants containing mapped reads.
  --bai BAI             (Optional) The name of the corresponding index for the
                        Illumina BAM file.
  -r REF_GENOME, --ref_genome REF_GENOME
                        The name of the reference file that matches the
                        reference used to map the Illumina inputs.
  --fai FAI             (Optional) The name of the corresponding index for the
                        reference genome file.
  --prefix PREFIX       (Optional) If provided, all output files will start
                        with this. If absent, the base of the BAM file name
                        will be used.
  --filter_short_contigs
                        If selected, SV calls will not be generated on contigs
                        shorter than 1 MB.
  --breakdancer         If selected, the program Breakdancer will be one of
                        the SV callers run.
  --breakseq            If selected, the program BreakSeq2 will be one of the
                        SV callers run.
  --manta               If selected, the program Manta will be one of the SV
                        callers run.
  --cnvnator            If selected, the program CNVnator will be one of the
                        SV callers run.
  --lumpy               If selected, the program Lumpy will be one of the SV
                        callers run.
  --delly_deletion      If selected, the deletion module of the program Delly2
                        will be one of the SV callers run.
  --delly_insertion     If selected, the insertion module of the program
                        Delly2 will be one of the SV callers run.
  --delly_inversion     If selected, the inversion module of the program
                        Delly2 will be one of the SV callers run.
  --delly_duplication   If selected, the duplication module of the program
                        Delly2 will be one of the SV callers run.
  --genotype            If selected, candidate events determined from the
                        individual callers will be genotyped and merged to
                        create a consensus output.
  --svviz               If selected, visualizations of genotyped SV events
                        will be produced with SVVIZ, one screenshot of support
                        per event. For this option to take effect, Genotype
                        must be selected.
  --svviz_only_validated_candidates
                        Run SVVIZ only on validated candidates? For this
                        option to be relevant, SVVIZ must be selected. NOT
                        selecting this will make the SVVIZ component run
                        longer.
```
# Parliament2 Developer Readme

## Adding a structural variant caller

Parliament2 is currently optimized to run efficiently at scale with a number of structural variant callers. As a result, some SV callers have been excluded due to their longer runtimes. However, if you are not running Parliament2 at scale, you can add additional SV callers.

There are four main actions required in order to add an SV caller:

* Adding the SV caller and its dependencies to the Docker image
* Running the SV caller
* Converting the SV caller's outputs into VCF format
* Running SVTyper on the SV caller's outputs

### Adding the SV caller and its dependencies

If the SV caller can be installed using conda, pip, or another package installer, simply add a line reading `RUN <package_installer> install <CALLER>` to the Dockerfile. Otherwise, you will have to compile the SV caller code for an Ubuntu 14.04 machine and add it to the `resources.tar.gz` archive when building the Docker image. You can compile the SV caller code for an Ubuntu machine using a DNAnexus cloud workstation (available at https://platform.dnanexus.com/app/cloud_workstation with a DNAnexus account).

You will need to rebuild the Docker image, which can take up to 15 minutes and requires at least 5 GB of memory on your local machine, and create a DNAnexus asset from the Docker image, which can take up to an hour.

Steps involved:

* Download the Parliament2 files by running `git clone https://github.com/dnanexus/parliament2.git`. You will have to run `tar -czvf resources.tar.gz resources/` in the directory `parliament2` in order to build the Docker image. **Optional**: Add the requisite files for your SV caller into the `resources/` directory *or* modify the Dockerfile accordingly.

* Run `docker build . -t parliament2` from within the `parliament2` directory created by GitHub. There should now be a Docker image named `parliament2` on your computer. You can verify the successful installation using the command `docker images`, which will list every Docker image stored locally on your machine, or by running `docker run parliament2 -h`, which will print the help string (printed in full at the bottom of this README).

* Run `dx-docker create-asset parliament2`. This will take approximately 45 minutes and will generate a string that you can copy-paste into the `dxapp.json` file under the "Regional Options" section for your region.

* Run `dx build parliament2` to build the applet.

### Running the SV caller (`run_sv_callers.sh`)

In order to do this, you must first determine whether you can or can't parallelize the caller's execution by contig.

#### Parallelizing callers by contig

Splitting the BAM file by contig has already been taken care of; lines 103-167 comprise the block of code dedicated to running these callers.

You will need to modify the following sections:

* The `if` statement under the section "Process management for launching jobs"
* The "thread-checking" section

You will need to add the following lines:

* A check to see if the caller is being run (`if [[ "$run_CALLER" == "true" ]]`)
* Running the caller on a single contig with a timeout of 6h (`timeout 6h CALLER`)
* Defining a command to merge later (`CALLER_concat="$CALLER_concat CALLER.$count.vcf"`)

The best thing to do is to emulate how other callers are being run.

#### Running other callers

If the caller has a built-in method for parallelization, you can run it outside of the code block that splits the BAM file by contig.

You will need to add the following lines:

* A check to see if the caller is being run (`if [[ "$run_CALLER" == "true" ]]`)
* Running the caller

Examples of running other callers are demonstrated in the lines of code running BreakSeq and Manta.

### Converting the SV caller's outputs into VCF format (if applicable) (`convert_to_vcf.sh`)

In order to run SVTyper on the SV caller's outputs, they must first be converted to VCF format. This varies depending on whether the caller has been parallelized by contig or not.

#### Parallelized by contig

You will need to add the following lines:

* A check to see if the caller is being run (`if [[ "$run_CALLER" == "true" ]]`)
* A concatenation of the various files produced by the caller (`python /convertHeader.py "$prefix" "$CALLER_concat" | vcf-sort -c | uniq > CALLER.vcf`)
* Make a copy of the VCF file to add the prefix for upload (`cp CALLER.vcf "$prefix".CALLER.vcf`)
* Upload the VCF file

#### Callers run normally

If the caller's output files have not been split by contig, and the caller already produces outputs in VCF format, there is no need to do anything.

However, if the caller produces outputs in non-VCF format, you will have to write your own conversion script to convert the outputs to VCF format.

### Running SVTyper on the SV caller's outputs (`run_svtyper.sh`)

You will need to add the following lines:

* A check to see if the caller is being run (`if [[ "$run_CALLER" == "true" ]]`)
* A `mkdir` to capture SVTyper's outputs (`mkdir svtype_CALLER`)
* Running SVTyper itself (`timeout -k 500 60m ./parallelize_svtyper.sh /home/dnanexus/CALLER.vcf /home/dnanexus/svtype_CALLER /home/dnanexus/CALLER.svtyped.vcf input.bam`)

## Additional Information

If you require assistance with Parliament2, please contact <support@dnanexus.com>.<!-- dx-header -->

## What does this app do?

This app runs a combination of tools to generate structural variant calls on whole-genome sequencing data. This tool runs the following callers: Breakdancer, Breakseq2, CNVnator, Delly2, Manta, and Lumpy. Because of synergies in how the programs use computational resources, these are all run in parallel. This app will produce the outputs of each of the tools for subsequent investigation.

If the option to genotype candidates is selected, this app will run SVTyper to genotype the events and will merge these tools with the program SURVIVOR. SV events that receive genotype calls have significantly higher specificity.

If the option to visualize candidates is selected, a tar file containing a PDF image for each SV call will be produced in order to summarize the supporting information behind the SV.

## What are typical use cases for this app?

This app is used to identify structural variants in a given sample relative to a reference genome. These structural variants cover large deletion events that are called as Deletions of a region, Insertions of a sequence into a region, Duplications of a region, Inversions of a region, or Translocations between two regions in the genome.

## What data are required for this app to run?

This app requires a BAM file containing the mapped reads of an individual relative to the reference genome. This app is intended for whole-genome sequencing. Providing exome or panel sequencing will not produce good results. It is intended to be run on a single germline sample.

In addition, the app requires the reference genome in gzipped FASTA format (`fa.gz` or `fasta.gz`). The reference genome should match the version used to generate the BAM file.

## What does this app output?

This app will output a number of files, representing the outputs of each of the structural variant callers. If the option to run a given step is unselected, then those outputs will not be provided.

- `lumpy.vcf`: representing the structural variant calls from Lumpy in VCF format.
- `lumpy.discordant.bam`: representing reads that Lumpy identified as discordant and used as evidence for its calls.
- `lumpy.splitters.bam`: representing reads that Lumpy identified as split-read mapped and used as evidence for its calls.
- `manta.diploidSV.vcf`: representing genotyped structural variant calls from Manta in VCF format.
- `manta.alignmentStats.txt`: representing statistics about the alignment from Manta.
- `breakdancer.ctx`: representing structural variant calls in Breakdancer's format.
- `cnvnator.output`: representing structural variant calls in CNVnator's format.
- `cnvnator.vcf`: representing structural variant calls from CNVnator in VCF format (this represents the conversion of CNVnator output to VCF format).
- `breakseq.gff`: representing structural variant calls from Breakseq2 in GFF format.
- `breakseq.vcf`: representing structural variant calls from Breakseq2 in VCF format.
- `breakseq.bam`: representing the reads mapping used as evidence for the calls generated by Breakseq2.
- `delly.deletion.vcf`: representing deletion calls made by Delly2.
- `delly.inversion.vcf`: representing inversion calls made by Delly2.
- `delly.duplication.vcf`: representing duplication calls made by Delly2.
- `delly.insertion.vcf`: representing insertion calls made by Delly2.
- `delly.translocation.vcf`: representing translocation calls made by Delly2.

- `<caller>.svtyped.vcf`: If the option to genotype candidates is selected, a genotype VCF produced by svtyper will be generated for each caller output.
- `combined.genotyped.vcf`: If the option to genotype candidates is selected, a merged VCF file of all of the callers will be produced by SURVIVOR.

- `svviz_outputs.tar.gz`: If the option to visualize events is selected, a tarball containing a set of PDFs documenting the genomic regions for calls will be generated

## Additional Notes

Because the field of structural variation is relatively new and complex, we viewed placing a dependency on all individual tools completing successfully as a requirement for a successful run to be too strict. In other words, if one of these tools fails while the others succeed, the app will output the results of the tools that completed and will not itself fail.

Breakseq2 may only be able to work when using the 1000 Genomes reference genome (hs37d5). For other reference genomes, you may not get Breakseq2 results.

The svviz tool can at times consume a large amount of memory, causing the application to be killed. If you see 
a message like:
```
Out of memory: Kill process XXXX (svviz) score 112 or sacrifice child
```
in the log, you should kill the job if it is still running and retry on a machine with more memory or without enabling svviz.


Parliament2 is open-source and can be found here: <https://github.com/dnanexus/parliament2>. The Docker container can be found here: <https://hub.docker.com/r/dnanexus/parliament2/>.

## Supporting information

For additional information, please see the following papers:

- Lumpy: Ryan M Layer, Colby Chiang, Aaron R Quinlan, and Ira M Hall. 2014. "LUMPY: a Probabilistic Framework for Structural Variant Discovery." Genome Biology 15 (6): R84. doi:10.1186/gb-2014-15-6-r84
- Manta: Chen X, Schulz-Trieglaff O, Shaw R, Barnes B, Schlesinger F, Källberg M, Cox AJ, Kruglyak S, Saunders CT. 2015 "Manta: rapid detection of structural variants and indels for germline and cancer sequencing applications." Bioinformatics. doi: 10.1093/bioinformatics/btv710
- Breakdancer: Chen K, Wallis JW, McLellan MD, Larson DE, Kalicki JM, Pohl CS, McGrath SD, Wendl MC, Zhang Q, Locke DP, Shi X, Fulton RS, Ley TJ, Wilson RK, Ding L, Mardis ER. 2009. "BreakDancer: an algorithm for high-resolution mapping of genomic structural variation". Nature Methods 6. doi:10.1038/nmeth.1363
- Breakseq2: Abyzov A, Li S, Kim DR, Mohiyuddin M, Stütz AM, Parrish NF, Mu XJ, Clark W, Chen K, Hurles M, Korbel JO, Lam HYK, Lee C, Gerstein MB. 2015. "Analysis of deletion breakpoints from 1,092 humans reveals details of mutation mechanisms". Nature Communications 6. doi:10.1038/ncomms8256
- Delly2: Tobias Rausch, Thomas Zichner, Andreas Schlattl, Adrian M. Stuetz, Vladimir Benes, Jan O. Korbel. 2012. Delly: structural variant discovery by integrated paired-end and split-read analysis. Bioinformatics 28:i333-i339. doi: 10.1093/bioinformatics/bts378
- CNVnator: Abyzov A, Urban AE, Snyder M, Gerstein M. 2011. "CNVnator: an approach to discover, genotype, and characterize typical and atypical CNVs from family and population genome sequencing". Genome Research (6):974. doi: 10.1101/gr.114876.110
- Parliament: Adam C English, William J Salerno, Oliver A Hampton, Claudia Gonzaga-Jauregui, Shruthi Ambreth, Deborah I Ritter, Christine R Beck, Caleb F Davis, Mahmoud Dahdouli, Singer Ma, Andrew Carroll, Narayanan Veeraraghavan, Jeremy Bruestle, Becky Drees, Alex Hastie, Ernest T Lam, Simon White, Pamela Mishra, Min Wang, Yi Han, Feng Zhang, Pawel Stankiewicz, David A Wheeler, Jeffrey G Reid, Donna M Muzny, Jeffrey Rogers, Aniko Sabo, Kim C Worley, James R Lupski, Eric Boerwinkle and Richard A Gibbs. Assessing structural variation in a personal genome—towards a human reference diploid genome. BMC Genomics 2015, 16:286  doi:10.1186/s12864-015-1479-3.
*NOTE: This is LUMPY 0.2.13 with additional changes to allow lumpyexpress to function when the main file is a CRAM, not a BAM. The splitters and discordants must still be BAM files as LUMPY itself doesn't yet support CRAM as an input. This requires the hexdump command be available.*

For questions and discussion about LUMPY please visit the forum at: 

https://groups.google.com/forum/#!forum/lumpy-discuss

[![Build Status](https://travis-ci.org/arq5x/lumpy-sv.svg?branch=master)](https://travis-ci.org/arq5x/lumpy-sv)

LUMPY
=====

A probabilistic framework for structural variant discovery.

Ryan M Layer, Colby Chiang, Aaron R Quinlan, and Ira M Hall. 2014.
"LUMPY: a Probabilistic Framework for Structural Variant Discovery."
Genome Biology 15 (6): R84.
[doi:10.1186/gb-2014-15-6-r84](http://dx.doi.org/10.1186/gb-2014-15-6-r84).

## Table of Contents
1. [Quick start](#quick-start)
2. [Installation](#installation)
3. [LUMPY Express usage](#lumpy-express-usage): Automated breakpoint detection for standard analyses.
4. [LUMPY (traditional) usage](#lumpy-traditional-usage): Flexible and customizable breakpoint detection for advanced users.
5. [Example workflows](#example-workflows)
6. [Test data](#test-data)
7. [Troubleshooting](#troubleshooting)

## Quick start

Download and install
```
git clone --recursive https://github.com/arq5x/lumpy-sv.git
cd lumpy-sv
make
cp bin/* /usr/local/bin/.
```

Run LUMPY Express
```
lumpyexpress \
    -B my.bam \
    -S my.splitters.bam \
    -D my.discordants.bam \
    -o output.vcf
```

## Installation

##### Requirements
- LUMPY
    * g++ compiler
    * CMake
- LUMPY Express (optional)
    * Samtools (0.1.18+) ([htslib.org/](http://www.htslib.org/))
    * SAMBLASTER (0.1.19+) ([github repo](https://github.com/GregoryFaust/samblaster))
    * Python 2.7 ([python.org/](https://www.python.org/)) with pysam (0.8.3+) and NumPy (1.8.1+)
    * sambamba ([gihub repo](https://github.com/lomereiter/sambamba))
    * gawk ([GNU project](https://www.gnu.org/software/gawk/))

##### Install

Default method to install:

```
git clone --recursive git@github.com:arq5x/lumpy-sv.git
cd lumpy-sv
make
cp bin/* /usr/local/bin/.
```

Installing with costom zlib (gzopen64 compile error):

```
git clone --recursive git@github.com:arq5x/lumpy-sv.git
cd lumpy-sv
export ZLIB_PATH="/usr/lib/x86_64-linux-gnu/"; #when /usr/lib/x86_64-linux-gnu/libz.so
make
cp bin/* /usr/local/bin/.
```


## LUMPY Express usage
Automated breakpoint detection for standard analyses.

```
usage:   lumpyexpress [options]
```

**Required arguments**
```
     -B FILE  coordinate-sorted BAM file(s) (comma separated)
     -S FILE  split reads BAM file(s) (comma separated)
     -D FILE  discordant reads BAM files(s) (comma separated)

```

**Optional arguments**
```
-o STR    output [fullBam.bam.vcf]
-x FILE   BED file to exclude
-P        output probability curves for each variant
-m INT    minimum sample weight for a call [4]
-r FLOAT  trim threshold [0]
-T DIR    temp directory [./output_prefix.XXXXXXXXXXXX]
-k        keep temporary files
-K FILE   path to lumpyexpress.config file
            (default: same directory as lumpyexpress)
-v        verbose
-h        show this message
```

#### Configuration
LUMPY Express runs several external program whose paths are specified in
[scripts/lumpyexpress.config](scripts/lumpyexpress.config). This config
must reside in the same directory as lumpyexpress, or be specified explicitly
with the -K flag.

The installation Makefile auto-generates a lumpyexpress.config file
and places it in the "bin" directory.

#### Input
LUMPY Express expects BWA-MEM aligned BAM files as input.
It automatically parses sample, library, and read group information using the @RG
tags in the BAM header.
Each BAM file is expected to contain exactly one sample.

The minimum input is a coordinate-sorted BAM file (-B), from which LUMPY Express
extracts splitters and discordants using SAMBLASTER before running LUMPY.
Optionally, users may supply coordinate-sorted splitter (-S) and discordant (-D)
BAM files which will bypass SAMBLASTER extraction for faster analysis.

#### Output
LUMPY Express produces a VCF file according to [VCF spec 4.2](https://samtools.github.io/hts-specs/VCFv4.2.pdf).

## LUMPY (traditional) usage
Flexible and customizable breakpoint detection for advanced users.

```
usage:    lumpy [options]
```

**Options**
```
-g       Genome file (defines chromosome order)
-e       Show evidence for each call
-w       File read windows size (default 1000000)
-mw      minimum weight across all samples for a call
-msw     minimum per-sample weight for a call
-tt      trim threshold
-x       exclude file bed file
-t       temp file prefix, must be to a writeable directory
-P       output probability curve for each variant
-b       output as BEDPE instead of VCF

-sr      bam_file:<file name>,
         id:<sample name>,
       	 back_distance:<distance>,
         min_mapping_threshold:<mapping quality>,
         weight:<sample weight>,
         min_clip:<minimum clip length>,
         read_group:<string>

-pe      bam_file:<file name>,
         id:<sample name>,
         histo_file:<file name>,
         mean:<value>,
         stdev:<value>,
         read_length:<length>,
         min_non_overlap:<length>,
         discordant_z:<z value>,
         back_distance:<distance>,
         min_mapping_threshold:<mapping quality>,
         weight:<sample weight>,
         read_group:<string>

-bedpe   bedpe_file:<bedpe file>,
         id:<sample name>,
         weight:<sample weight>
```

## Example workflows

#### Pre-processing
We recommend aligning data with [SpeedSeq](https://github.com/cc2qe/speedseq), which
performs BWA-MEM alignment, marks duplicates and extracts split and discordant
read-pairs.
```
speedseq align -R "@RG\tID:id\tSM:sample\tLB:lib" \
    human_g1k_v37.fasta \
    sample.1.fq \
    sample.2.fq
```

Otherwise, data may be aligned with BWA-MEM.

```
# Align the data
bwa mem -R "@RG\tID:id\tSM:sample\tLB:lib" human_g1k_v37.fasta sample.1.fq sample.2.fq \
    | samblaster --excludeDups --addMateTags --maxSplitCount 2 --minNonOverlap 20 \
    | samtools view -S -b - \
    > sample.bam

# Extract the discordant paired-end alignments.
samtools view -b -F 1294 sample.bam > sample.discordants.unsorted.bam

# Extract the split-read alignments
samtools view -h sample.bam \
    | scripts/extractSplitReads_BwaMem -i stdin \
    | samtools view -Sb - \
    > sample.splitters.unsorted.bam

# Sort both alignments
samtools sort sample.discordants.unsorted.bam sample.discordants
samtools sort sample.splitters.unsorted.bam sample.splitters
```

#### Running LUMPY
LUMPY has two distinct execution alternatives. LUMPY Express is a simplified wrapper for standard analyses.
LUMPY (traditional) is more customizable, for advanced users and specialized experiments.

##### LUMPY Express
- Run LUMPY Express on a single sample with pre-extracted splitters and discordants
    ```
    lumpyexpress \
        -B sample.bam \
        -S sample.splitters.bam \
        -D sample.discordants.bam \
        -o sample.vcf
    ```

- Run LUMPY Express jointly on multiple samples with pre-extracted splitters and discordants
    ```
    lumpyexpress \
        -B sample1.bam,sample2.bam,sample3.bam \
        -S sample1.splitters.bam,sample2.splitters.bam,sample3.splitters.bam \
        -D sample1.discordants.bam,sample2.discordants.bam,sample3.discordants.bam \
        -o multi_sample.vcf
    ```

- Run LUMPY Express on a tumor-normal pair
    ```
    lumpyexpress \
        -B tumor.bam,normal.bam \
        -S tumor.splitters.bam,normal.splitters.bam \
        -D tumor.discordants.bam,normal.discordants.bam \
        -o tumor_normal.vcf
    ```

##### LUMPY (traditional)
First, generate empirical insert size statistics on each library in the BAM file
```
samtools view -r readgroup1 sample.bam \
    | tail -n+100000 \
    | scripts/pairend_distro.py \
    -r 101 \
    -X 4 \
    -N 10000 \
    -o sample.lib1.histo
```
The above script (scripts/pairend_distro.py) will display mean and stdev to screen.
For these examples we will assume the mean is 500 and the stdev is 50.

- Run LUMPY with paired-end and split-reads.
    ```
    lumpy \
        -mw 4 \
        -tt 0 \
        -pe id:sample,bam_file:sample.discordants.bam,histo_file:sample.lib1.histo,mean:500,stdev:50,read_length:101,min_non_overlap:101,discordant_z:5,back_distance:10,weight:1,min_mapping_threshold:20 \
        -sr id:sample,bam_file:sample.splitters.bam,back_distance:10,weight:1,min_mapping_threshold:20 \
        > sample.vcf
    ```

- Run LUMPY on a BAM file with multiple libraries.
    ```
    lumpy \
        -mw 4 \
        -tt 0 \
        -pe id:sample,read_group:rg1,bam_file:sample.discordants.bam,histo_file:sample.lib1.histo,mean:500,stdev:50,read_length:101,min_non_overlap:101,discordant_z:5,back_distance:10,weight:1,min_mapping_threshold:20 \
        -pe id:sample,read_group:rg2,bam_file:sample.discordants.bam,histo_file:sample.lib2.histo,mean:500,stdev:50,read_length:101,min_non_overlap:101,discordant_z:5,back_distance:10,weight:1,min_mapping_threshold:20 \
        -sr id:sample,bam_file:sample.splitters.bam,back_distance:10,weight:1,min_mapping_threshold:20 \
        > sample.vcf
    ```

- Run LUMPY on multiple samples with multiple libraries.
    ```
    lumpy \
        -mw 4 \
        -tt 0 \
        -pe id:sample1,bam_file:sample1.discordants.bam,read_group:rg1,read_group:rg2,histo_file:sample1.lib1.histo,mean:500,stdev:50,read_length:101,min_non_overlap:101,discordant_z:5,back_distance:10,weight:1,min_mapping_threshold:20 \
        -pe id:sample1,bam_file:sample1.discordants.bam,read_group:rg3,histo_file:sample1.lib2.histo,mean:500,stdev:50,read_length:101,min_non_overlap:101,discordant_z:5,back_distance:10,weight:1,min_mapping_threshold:20 \
        -pe id:sample2,bam_file:sample2.discordants.bam,read_group:rg4,histo_file:sample2.lib1.histo,mean:500,stdev:50,read_length:101,min_non_overlap:101,discordant_z:5,back_distance:10,weight:1,min_mapping_threshold:20 \
        -sr id:sample1,bam_file:sample1.splitters.bam,back_distance:10,weight:1,min_mapping_threshold:20 \
        -sr id:sample2,bam_file:sample2.splitters.bam,back_distance:10,weight:1,min_mapping_threshold:20 \
        > multi_sample.vcf
    ```

- Run LUMPY with low complexity regions excluded.  
    Heng Li provides a set of low complexity regions in the supplementary information of his paper, "Toward better
    understanding of artifacts in variant calling from high-coverage samples" at
    https://doi.org/10.1093/bioinformatics/btu356. 
    ```
    unzip btu356_Supplementary_Data.zip
    unzip btu356-suppl_data.zip
    lumpy \
        -mw 4 \
        -tt 0.0 \
        -x btu356_LCR-hs37d5.bed/btu356_LCR-hs37d5.bed \
        -pe bam_file:sample.discordants.bam,histo_file:sample.pe.histo,mean:500,stdev:50,read_length:101,min_non_overlap:101,discordant_z:5,back_distance:10,weight:1,id:sample,min_mapping_threshold:1 \
        -sr bam_file:sample.sr.sort.bam,back_distance:10,weight:1,id:sample,min_mapping_threshold:1 \
        > sample.exclude.vcf
    ```
- Run LUMPY with regions of very high coverage excluded.  
    We can direct lumpy to ignore certain regions by using the
    exclude region option. In this example we find and then
    exclude regions that have very high coverage. First we
    use the get_coverages.py script to find the min, max, and
    mean coverages of the the sr and pe bam files, and to
    create coverage profiles for both files.

    ```
    python ../scripts/get_coverages.py \
        sample.pe.sort.bam \
	sample.sr.sort.bam
    # sample.pe.sort.bam.coverage  min:1   max:14  mean(non-zero):2.35557521272
    # sample.sr.sort.bam.coverage  min:1   max:7   mean(non-zero):1.08945936729
    ```
    
    From this output, we will choose to exclude regions that
    have more than 10x coverage.  To create the exclude file
    we will use the get_exclude_regions.py script to create
    the exclude.bed file
    ```
    python ../scripts/get_exclude_regions.py \
        10 \
	exclude.bed \
	sample.pe.sort.bam \
	sample.sr.sort.bam
    ```

    We now rerun lumpy with the exclude (-x) option
    ```
    lumpy \
        -mw 4 \
        -tt 0.0 \
        -x exclude.bed \
        -pe bam_file:sample.discordants.bam,histo_file:sample.pe.histo,mean:500,stdev:50,read_length:101,min_non_overlap:101,discordant_z:5,back_distance:10,weight:1,id:sample,min_mapping_threshold:1 \
        -sr bam_file:sample.sr.sort.bam,back_distance:10,weight:1,id:sample,min_mapping_threshold:1 \
        > sample.exclude.vcf
    ```

#### Post-processing
[SVTyper](https://github.com/cc2qe/svtyper) can call genotypes on LUMPY output VCF files
using a Bayesian maximum likelihood algorithm.
```
svtyper \      
    -B sample.bam \
    -S sample.splitters.bam \
    -i sample.vcf
    > sample.gt.vcf
```

## Test data
The `test/test.sh` script executes lumpy against several simulated data sets
and compares the results to the known correct result.  The sample data sets can
be found at http://layerlab.org/lumpy/data.tar.gz.  This tar ball should be
extracted into the top-level lumpy directory.  The script `test/test.sh` checks
for the the existence of this directory before running LUMPY.

## Troubleshooting
All of the bam files that lumpy processes must be position sorted. To check if your bams are sorted correctly, use the check_sorting.py script
```
python ../scripts/check_sorting.py \
    pe.pos_sorted.bam \
    sr.pos_sorted.bam \
    pe.name_sorted.bam
# pe.pos_sorted.bam
# in order
# sr.pos_sorted.bam
# in order
# pe.name_sorted.bam
# out of order:   chr10   102292476   occurred after   chr10   102292893
```
Thread pool tests
=================
The thread_pool.c file has a built-in test program which is enabled when compiling with TEST_MAIN defined. The test program can be run in four different modes by giving a command-line parameter: unordered, ordered1, ordered2, and pipe. The modes and their expected outputs are described below.

unordered
---------
Dispatches TASK_SIZE (=1000) jobs to the thread pool and waits for them to finish. The job index (0..TASK_SIZE-1) is passed as a parameter. The job function is doit_square_u, which sleeps for a while and then prints the square of its input parameter to stdout.

Expected output when n = 1:
```
RESULT: 0
...
RESULT: 998001
```

Expected output when n > 1: same, but in jumbled up order.

ordered1
--------
Dispatches TASK_SIZE (=1000) jobs to the thread pool in non-blocking mode. Results are returned on the result queue and are pulled in order. The job index (0..TASK_SIZE-1) is passed as a parameter. The job function is doit_square, which sleeps for a while and then returns the square of its input parameter as a result. Some of the jobs take way longer than the others to finish.

The expected output is the results printed in order, regardless of n.

ordered2
--------
Starts a dispatcher thread which dispatches jobs to the thread pool. After all regular jobs have been dispatched, a sentinel job follows where the input parameter is set to -1, which receives special handling in doit_square to return the -1 as the result.

Results are consumed on the main thread using hts_tpool_next_result_wait, until the end-of-job marker is found.

The expected output is the results printed in order, regardless of n.

pipe
----
This program uses one thread pool (hts_tpool) and three queues (hts_tpool_process) shared across threads using a pipe_opt struct. There are four threads: input, stage1to2, stage2to3, and output.

The input thread (pipe_input_thread procedure) dispatches jobs to the thread pool with the job number (1..TASK_SIZE) and an end-of-job flag as parameters. The jobs are executed by the pipe_stage1 procedure, which multiplies by 256 and sleeps for a short while.

The stage1to2 thread (pipe_stage1to2 procedure) pulls results from the first queue (q1) and passes them to new jobs in the thread pool. These jobs are executed by the pipe_stage2 procedure, which does the same as pipe_stage1, only slower.

The stage2to3 thread is similar to the stage1to2 thread. It pulls from the second queue and dispatches new jobs to be executed by the pipe_stage3 procedure. pipe_stage3 is similar to pipe_stage1.

The output thread pulls from the third queue.

Expected output:
```
I 00000001
1 00000100
2 00010000
O 01000000
...
I 000003e8
1 0003e800
2 03e80000
O e8000000
```
...but not in order, because the input queues might be served in any order.

However, if only the lines from the output thread are printed, they should be in order regardless of the number of threads.
# bamkit
Tools for common BAM file manipulations
Thread pool tests
=================
The thread_pool.c file has a built-in test program which is enabled when compiling with TEST_MAIN defined. The test program can be run in four different modes by giving a command-line parameter: unordered, ordered1, ordered2, and pipe. The modes and their expected outputs are described below.

unordered
---------
Dispatches TASK_SIZE (=1000) jobs to the thread pool and waits for them to finish. The job index (0..TASK_SIZE-1) is passed as a parameter. The job function is doit_square_u, which sleeps for a while and then prints the square of its input parameter to stdout.

Expected output when n = 1:
```
RESULT: 0
...
RESULT: 998001
```

Expected output when n > 1: same, but in jumbled up order.

ordered1
--------
Dispatches TASK_SIZE (=1000) jobs to the thread pool in non-blocking mode. Results are returned on the result queue and are pulled in order. The job index (0..TASK_SIZE-1) is passed as a parameter. The job function is doit_square, which sleeps for a while and then returns the square of its input parameter as a result. Some of the jobs take way longer than the others to finish.

The expected output is the results printed in order, regardless of n.

ordered2
--------
Starts a dispatcher thread which dispatches jobs to the thread pool. After all regular jobs have been dispatched, a sentinel job follows where the input parameter is set to -1, which receives special handling in doit_square to return the -1 as the result.

Results are consumed on the main thread using hts_tpool_next_result_wait, until the end-of-job marker is found.

The expected output is the results printed in order, regardless of n.

pipe
----
This program uses one thread pool (hts_tpool) and three queues (hts_tpool_process) shared across threads using a pipe_opt struct. There are four threads: input, stage1to2, stage2to3, and output.

The input thread (pipe_input_thread procedure) dispatches jobs to the thread pool with the job number (1..TASK_SIZE) and an end-of-job flag as parameters. The jobs are executed by the pipe_stage1 procedure, which multiplies by 256 and sleeps for a short while.

The stage1to2 thread (pipe_stage1to2 procedure) pulls results from the first queue (q1) and passes them to new jobs in the thread pool. These jobs are executed by the pipe_stage2 procedure, which does the same as pipe_stage1, only slower.

The stage2to3 thread is similar to the stage1to2 thread. It pulls from the second queue and dispatches new jobs to be executed by the pipe_stage3 procedure. pipe_stage3 is similar to pipe_stage1.

The output thread pulls from the third queue.

Expected output:
```
I 00000001
1 00000100
2 00010000
O 01000000
...
I 000003e8
1 0003e800
2 03e80000
O e8000000
```
...but not in order, because the input queues might be served in any order.

However, if only the lines from the output thread are printed, they should be in order regardless of the number of threads.
# JupyROOT
A software layer to integrate Jupyter notebooks and ROOT.

## Installation
1. [Install ROOT6](https://root.cern.ch/building-root) (> 6.05)
2. Install dependencies: pip install jupyter metakernel

## Start using ROOTbooks
Set up the ROOT environment (`. $ROOTSYS/bin/thisroot.[c]sh`) and type in your
shell:
```
root --notebook
```
This will start a ROOT-flavoured notebook server in your computer.

Alternatively, if you would like to use the Jupyter command directly, you 
can do on Linux:
```
cp -r $ROOTSYS/etc/notebook/kernels/root ~/.local/share/jupyter/kernels
jupyter notebook
```
and on OSx:
```
cp -r $ROOTSYS/etc/notebook/kernels/root /Users/$USER/Library/Jupyter/kernels/
jupyter notebook
```

Once the server is up, you can use ROOT with two kernels:

1. ROOT C++: new kernel provided by ROOT
2. Python 2: already provided by Jupyter

##  C++ ROOTbook
ROOT offers a C++ kernel that transforms the notebook in a ROOT prompt.
Embedded graphics, syntax highlighting and tab completion are among
the features provided by this kernel.

An example of how you would plot a histogram in a C++ ROOTbook is:
```cpp
TCanvas c;
TH1F h("h","ROOT Histo;X;Y",64,-4,4);
h.FillRandom("gaus");
h.Draw();
c.Draw();
``` 

## Python ROOTbook
If you prefer to use Python, you can create a new Python 2 kernel and
import the ROOT libraries:
```python
import ROOT
```
And then you could write something like:
```python
c = ROOT.TCanvas("c")
h = ROOT.TH1F("h","ROOT Histo;X;Y",64,-4,4)
```
Additionally, you can mix Python and C++ in the same notebook
by using the **%%cpp** magic:
```cpp
%%cpp
h->FillRandom("gaus");
h->Draw();
c->Draw();
```
<b>BreakSeq2: Ultrafast and accurate nucleotide-resolution analysis of structural variants</b>

See http://bioinform.github.io/breakseq2/ for help and downloads.
# JSROOT changelog

## Changes in 5.1.2
1. Fix - support newest TFormula in TF1 (#127)
2. Fix - ignore NaN value in saved TF1 buffer
3. Fix - correctly treat transparency in geo painter
4. Fix - disable useFontCache for SVG mathjax output
5. Fix - produce PNG image for objects with special symbols in names


## Changes in 5.1.1
1. Fix - invoke callback in JSROOT.draw() at proper time
2. Fix - support TGeoHMatrix, produced after GDML conversion
3. Fix - support also TGeoScale and TGeoGenTrans matrixes
4. Fix - update histograms with all provided functions (#125)


## Changes in 5.1.0
1. New 'float' browser kind overlays with objects drawing
2. Browser can be enabled after drawing with 'nobrowser' mode
3. One can hide browser or switch browser kind at any time
4. New 'horizontal' and 'vertical' layouts for object display.
   One could configure several frames, each divided on sub-frames.
   Like display=horiz231 will create three horizontal frames, 
   divided on 2,3 and 1 sub-frames.  
5. One could enable status line where current tooltip info will be shown  
6. Improve enlarge functionality - now works with all layouts
7. Do not display all canvas tool buttons by default - provide toggle button instead 
8. Let move TAxis title, its position now similar to ROOT graphics
9. Support 'col0' option for TH2Poly class to suppress empty bins
10. Implement for TH3 'box2', 'box3', 'glbox2', 'glcol' draw options
11. Support more superscript/subscript letters in normal text output
12. Correctly handle unzoom with logx/logy scales
13. Let disable stamp parameter in file url with "-" sign at the end of file name
14. Let use quotes in the URL parameters to protect complex arguments with special symbols
15. Introduce direct streamers - like TBasket or TRef
    Benefit - one can add custom streamers of such kind or reuse existing
16. Handle TMatrixTSym classes in I/O  
17. Correctly count TH3 statistic in TTree::Draw
18. Recognize bower installation when "bower_components/jsroot/scripts" string
    appears in the script path (#120)  


## Changes in 5.0.3
1. Fix - prevent exception when discover HTML element position (#121)
2. Fix - prevent I/O failure when server automatically gzip response (#119)
3. Fix - lego drawing for stacked TH1 histograms
4. Fix - when change global tooltips settings, also change for each sub-pad


## Changes in 5.0.2
1. Fix - read branch entries as arrays
2. Fix - command submission to THttpServer
3. Fix - let refill statbox also for empty histogram
4. Fix - problem with online TTree::Draw and ROOT6 


## Changes in 5.0.1
1. Support older ROOT files, created before 2010
2. Support TBranchObject - appears in old files
3. Correctly set TBasket buffer position for the entry
4. Fix - problem with empty STL containers
5. Fix - empty baskets at the end of branch store
6. Fix - problem with zooming in THStack


## Changes in 5.0.0
1. Reading TTree data
    - all kinds of branches, including split STL containers
    - branches with several elementary leaves
    - branches from different ROOT files
    - JSROOT.TSelector class to access TTree data
    - simple access to branch data with "dump" draw option
2. TTree::Draw support
    - simple 1D/2D/3D histograms
    - simple cut conditions
    - configurable histogram like "px:py>>hist(50,-5,5,50,-5,5)"
    - strings support
    - iterate over arrays indexes, let use another branch as index values
    - support "Entry$" and "Entries$" variables in expressions
    - bits histogram like "event.fTracks.fBits>>bits(16)"
    - special handling of TBits
    - arbitrary math function from JavaScript Math class, some TMath:: function from ROOT
    - if branch is object, one could use methods "TMath::Abs(lep1_p4.X()+lep1_p4.Y())"
    - interactive player to configure and execute draw expression
3. Full support of Float16_t and Double32_t types in I/O
4. Drawing of RooPlot objects, I/O support for RooFit classes
5. Many improvements in object inspector
    - support of large lists; only first part is shown
    - support of large arrays; values group in decades
    - allow to call draw function for sub-elements in inspector
6. Canvas or selected sub-pad can be enlarged when double-clicked outside frame (#116)
   Complete drawing will be expanded to the visible space.
   Not available for flex, tabs and collapsible layouts.
7. Support reading of local ROOT files with HTML5 FileReader.
   Files can be selected only with interactive dialog.
8. Combine "Ctrl" and "Shift" keys with mouse click on the items:
     - with Shift key typically object inspector will be activated
     - with Ctrl key alternative draw options will be used (like colz for TH2)
9. Update libraries
      - d3.js     - 4.4.4
      - three.js  - 84
      - jquery    - 3.3.1
      - jquery-ui - 1.12.1 


## Changes in 4.8.2
1. Support compressed array, produced with newest TBufferJSON
   - $arr field identify such array and contains data type
   - native arrays are used when decoding such array
   - zero values are not stored
   - many similar values stored as one with repetition factor
   - position stored only when differ from produced with previous block
   - array [3,3,3,3,3,3,1,2,2,2,2,2,2,2] compressed as {$arr:"Int",len:14,v:3,n:6,v1:1,v2:2,n2:7}


## Changes in 4.8.1
1. Support new JSON format, produced with newest TBufferJSON
   - object references stored as {"$ref":12}
   - pair objects for std::map marked with "$pair" : "pair<type1,type2>" data member
   - old JSON format will be recognized automatically and supported as well
2. Fix - better selection of Y range for log scale
3. Provide JSROOT.parse_multi function to correctly parse response of multi.json request,
   support it in the JSROOT.NewHttpRequest method as well.
4. Fix - correctly calculate integral for TH1
5. Partially support new TFormula with complex C code inside


## Changes in 4.8.0
1. Many improvements in the I/O part 
   - support most of STL containers
   - support TMap and TClonesArray containers
   - all kind of multidimensional arrays
   - correct treatment of foreign classes
   - supports different versions of class in the same file
   - support members like ClassName* fField; //[fCnt]
   - support const char*
   - support fixed-size array of TString, TObject and TNamed 
2. Many new draw options for different classes are supported:
    - TGraph  - 'z', 'x', '||', '[]', '>', '|>', '5', 'X+', 'Y+'
    - TH1     - '*', 'L', 'LF2', 'B', 'B1', 'TEXT', 'E0', 'E3', 'E4', 'EX0', 'X+', 'Y+'
    - TH2     - 'E', 'col1', 'box', 'box1', 'surf3', 'surf7', 'base0'
    - TH2     - 'same' with 'box', 'col', 'cont', 'lego', 'surf'
    - TH3     - 'scat', use by default
    - TF1/TF2 - 'nosave' to ignore saved buffer
    - TCanvas - logx/y/z, gridx/y, tickx/y 
    - THStack - 'lego' and other 3D draw options
3. Implement drawing of TProfile2D, TF2, TGraph2D, TGraph2DErrors and TMarker
4. Fix - correctly place TGAxis relative to frame (when exists) 
5. When superimpose items, one can specify individual options
     ...&item=histo1+histo2&opt=hist+e1
     ...&item=[histo1,histo2]&opt=[hist,e1]
6. Support loading of TStyle object, providing in URL
     ...&style=item_name  or ...&style=json_file_name 
   All values are copied directly to JSROOT.gStyle object.  
7. Add callback argument into JSROOT.draw() function. 
   Function will be called after drawing of object is completed.
   Painter for drawn object will be provided as first argument (or null in case of error). 
8. Improve cleanup of JSROOT objects 


## Changes in 4.7.1
1. Workaround for MathJax output - scaling not always works in Firefox
2. Fix - bin scaling for box draw option for TH2 and TH3 histograms 
3. Fix - increase points limits for contour plots
4. Fix - position of 3D canvas in WebKit browsers
5. Fix - use abs bin content in RMS calculations
6. Fix - support char star* and object arrays in I/O
7. Fix - correct decoding of TAxis time offset
8. Fix - checksum reading for foreign classes


## Changes in 4.7.0
1. Implement simple TTree::Draw over single leaf (#80)
   Support basic types, fixed-size arrays and several vector types 
2. Display of TEveTrack(s) and TEvePointSet(s) over drawn geometry (drag and drop)
   Also browsing, toggling, highlight of tracks and hits are done.
3. Let set default geo colors as TGeoManager::DefaultColors() does  
4. Let use original ROOT macros to configure visibility of geometry volumes. Like:
     &file=files/alice2.root&item=Geometry;1&opt=macro:macros/geomAlice.C   
   One can set default colors or colors/transperency for selected volumes.
   Also volume, selected for drawing in the macro, will be used in the JSROOT
5. Support drawing of TH2Poly class with 'col' and 'lego' options
6. Implement 'CONT', 'ARR' and 'SURF' draw options for TH2 class 
7. Support basic drawing of TPolyLine class 
8. Interactive axis zooming in 3D with mouse, very much like to 2D
9. Zooming and tool buttons via keyboards  


## Changes in 4.6.0
1. Improvements in TGeo drawings
   - support of large (~10M volumes) models, only most significant volumes are shown
   - one could activate several clip planes (only with WebGL)
   - interaction with object browser to change visibility flags or focus on selected volume
   - support of floating browser for TGeo objects 
   - intensive use of HTML Worker to offload computation tasks and keep interactivity   
   - enable more details when changing camera position/zoom
   - better and faster build of composite shapes
2. Improvements in histograms 3D drawing
   - all lego options: lego1..lego4, combined with 'fb', 'bb', '0' or 'z'
   - support axis labels on lego plots
   - support lego plots for TH1 
3. Improvements in all 3D graphics
   - upgrade three.js to r79
   - use of THREE.BufferGeometry for all components
   - significant (up to factor 10) performance improvement
4. Implement box and hbox draw options for TH1 class
5. Implement drawing of axes ticks on opposite side (when fTickx/y specified)  
6. Preliminary support of candle plot (many options to be implemented)
7. Update draw attributes (fill/line/position) when monitor objects  


## Changes in 4.5.3
1. Fix - position of TFrame in canvas/pad
2. Fix - use histogram fMinimum/fMaximum when creating color palette
3. Fix - correctly draw empty th2 bins when zmin<0 is specified  
4. Fix - limit th2 text output size 
5. Fix - use histogram fMinimum/fMaximum when drawing z axis in lego plot 
6. Fix - error in TGeoCtub shape creation
7. Fix - error in pcon/pgon shapes when Rmin===0 


## Changes in 4.5.1
1. Fix - correctly handle ^2..^9 in TFormula equations
2. Fix - support TMath::Gaus in TFormula 
3. Fix - correctly display ^2 and ^3 in SVG text output 
4. Fix - do not show tooltips for empty TProfile bins
5. Fix - statbox toggling was not working on subpads
6. Fix - positioning of 3D objects in Webkit browsers in complex layouts
7. Fix - difference in TF1 between ROOT5/6 (#54)  
 

## Changes in 4.5.0
1. Zooming with mouse wheel
2. Context menus for many different objects attributes are provided
3. Context menu for every drawn object can be activated via toolbar button 
4. Support browsing of TTask and derived classes (#40)
5. Apply user range for drawing TH1/TH2 histograms, also when superimposed (#44) 
6. Implement scaling factor like x10^3 on the vertical axes 
7. Provide shortcut buttons for each subpad
8. Implement simple drawing for TBox, TWbox, TSliderBox classes


## Changes in 4.4.4
1. Fix - toggling of statbox was not working in all situations 
2. Fix - for mouse rect zooming use only left mouse button
3. Fix - correctly draw TH2 with lego option, when histogram has negative bin content
4. Fix - log axis drawing with no visible ticks 


## Changes in 4.4.3
1. Fix - wrong selection of TH1 Y axis range when errors are displayed (#44) 
2. Fix - apply user range for TH1 X-axis zooming (#44)   
3. Fix - protect against pathological case of 1-bin histogram
4. Fix - use error plot by default in TH1 only when positive sumw2 entry exists
5. Fix - for TH2 box draw option draw at least 1px rect for non-empty bin
6. Fix - support transparency (alpha) in TColor (#45)
7. Fix - correct tooltip handling for graphs with lines and markers
8. Fix - interactive zooming in TH2 when doing histogram update  


## Changes in 4.4.2
1. Fix - statistic collection for TH2
2. Fix - correct handling of empty TList in browser/inspector
3. Fix - support TFolder in browser/inspector (#40)


## Changes in 4.4.1
1. Fix - colz palette resize when drawing histogram second time
2. Fix - use embedded in TCanvas color for background color of canvas itself
3. Fix - rotate too long X axis text labels 
4. Fix - draw histogram bins on frame boundary
5. Fix - use alternative color for shapes with default black color  
6. Fix - correctly handle pcon/pgon shape with rmin==rmax on top or bottom side
  

## Changes in 4.4
1. Fix faces orientation for all TGeo shapes. 
2. Improve TGeoTorus creation - handle all parameters combinations
3. Implement TGeoCompositeShape, using ThreeCSG.js
4. Fix problem with color palette when switch to 3D mode (#28)
5. Use nested CSS classes to avoid conflicts with other libraries (#29)
6. Let move and resize TFrame
7. Improve TH1/TH2 drawings
   - draw all histograms points in the range (no any skipped bins)
   - minimize SVG code for drawing (up to factor 100)
   - gives significant speedup in drawings
8. SVG code improvement for TGraph, TF1, TAxis drawings
9. Provide new tooltip kind
   - created only when needed (minimizing SVG code)
   - tooltip can be drawn for every object in the frame
   - touch devices are supported 
10. Fix - let draw same object on the canvas with different options
11. Create cached list of known class methods. It can be extended by users. 
12. Use of cached methods improves binary I/O performance by 20%
13. Support TGaxis
14. Project now can be obtained via 'bower install jsroot' 
15. Support 'scat' and 'text' draw options for TH2 
16. Support in binary I/O zipped buffer bigger than 16M
17. Correctly handle in binary I/O pointer on TArray object (like in THnSparseArrayChunk)


## Changes in 4.3
1. Implement TGeoCtub, TGeoParaboloid and TGeoHype shapes
2. Support TGeoTube with Rmin==0 
3. Exclude empty faces in TGeoArb8  
4. Improve TGeoSphere creation - handle all parameters combinations  
5. Introduce JSROOT.cleanup() function to safely clear all drawn objects 
6. Fix wrong resize method in 'tabs' and 'collapsible' layouts
7. Fix canvas resize problem (issue #27)
8. Fix zero-height canvas when draw TGeo in collapsible layout
9. Fix problem of simultaneous move TGeo drawings and canvas in flexible layout
 

## Changes in 4.2
1. Significant performance improvements in 3D drawings - TGeo/TH2/TH3 
2. Implement TGeoPara, TGeoGtra, TGeoXtru and TGeoEltu shapes
3. Optimize (reduce vertices number) for others TGeo shapes 
4. Correct rotation/translation/scaling of TGeo nodes 
5. Workaround for axis reflection (not directly supported in three.js) 
6. Support array of objects in I/O (like in TAxis3D)
7. Correct reading of multi-dim arrays like Double_t fXY[8][2];
8. Provide canvas toolbar for actions like savepng or unzoom  
9. Implement JSROOT.resize() function to let resize drawing after changes in page layout 
10. Fix error with title display/update 
 

## Changes in 4.1
1. Introduce object inspector - one could browse object members of any class
2. Let draw sub-items from TCanvas list of primitives like sub-pad or TLatex
3. Provide possibility to save drawn SVG canvas as PNG 
4. TGraph drawing optimization - limit number of drawn points   
5. Implement painter for TPolyMarker3D
6. Improve drawing and update of TMultiGraph 
7. Reorganize 3D drawing of TH2/TH3 histograms, allow to mix 2D and 3D display together  
8. Support overlay of 3D graphic over SVG canvas (used for IE)
9. Fix problems and improve flex(ible) layout


## Changes in 4.0
1. New TGeo classes support:
   - browsing  through volumes hierarchy
   - changing visibility flags
   - drawing of selected volumes
2. New 'flex' layout:
   - create frames like in Multi Document Interface
   - one could move/resize/minimize/maximize such frames
3. Significant (factor 4) I/O performance improvement:
   - use ArrayBuffer class in HTTP requests instead of String
   - use native arrays (like Int32Array) for array data members
   - highly optimize streamer infos handling 
4. TH2 drawing optimization:
   - if there are too many non-empty bins, combine them together
   - when zoom-in, all original bins will be displayed separately
   - let draw big TH2 histogram faster than in 1 sec
   - optimization can be disabled by providing '&optimize=0' in URL             
5. TF1 drawing optimization:
   - function 'compiled' only once
6. Reorganize scripts structure:
   - move all math functions to JSRootMath.js
   - TH2, TF1, THStack and TMultiGraph painters moved into JSRootPainter.more.js script
   - reduce size of scripts required for default functionality
7. Update all basic libraries:
    - d3.js - v3.5.9, 
    - jquery.js - v2.1.4, 
    - jquery-ui.js - v1.11.4, 
    - three.js - r73  
8. Implement ROOT6-like color palettes:
    - all palettes in range 51...112 are implemented 
    - by default palette 57 is used
    - one could change default palette with '&palette=111' in URL
    - or palette can be specified in draw option like '&opt=colz,pal77'


## Changes in 3.9
1. Support non-equidistant bins for TH1/TH2 objects.
2. Display entries count from histo.fEntries member, only when not set use computed value  
3. Support italic and bold text when used with MathJax
4. Improve TF1 drawing - support exp function in TFormula, fix errors with logx scale, enable zoom-in, (re)calculate function points when zooming
5. Support several columns in TLegend
6. Introduce context menus for x/y axis, add some items similar to native ROOT menus
7. Introduce context menu for TPaveStats, let switch single elements in the box   
8. Enable usage of all context menus on touch devices 
9. Implement JSROOT.Math.Prob function, provides probability value in stat box 
10. Introduce context menu for color palette (z axis)
11. Implement col0 and col0z draw option for TH2 histograms, similar to ROOT6


## Changes in 3.8
1. Let use HTML element pointer in JSROOT.draw function like:
       JSROOT.draw(document.getElementsByTagName("div")[0], obj, "hist");
   Normally unique identifier was used before, which is not required any longer.
   Of course, old functionality with element identifier will work as well. 
2. TreePlayer can also be used for trees, which not yet read from the file.
   Requires appropriate changes in TRootSniffer class. 
3. Fix error in I/O with members like:   `Double_t *fArr; //[fN]`  
4. Introduce JSROOT.OpenFile function. It loads I/O functionality automatically,
   therefore can be used directly after loading JSRootCore.js script
5. Same is done with JSROOT.draw function. It is defined in the JSRootCore.js
   and can be used directly. Makes usage of JSROOT easier    
6. Introduce JSRootPainter.more.js script, where painters for auxiliary classes
   will be implemented.
7. Implement painter for TEllipse, TLine, TArrow classes     
8. Fix several problems with markers drawing; implement plus, asterisk, mult symbols. 
9. Implement custom layout, which allows to configure user-defined layout for displayed objects
10. Fix errors with scaling of axis labels.     
11. Support also Y axis with custom labels like: http://jsroot.gsi.de/dev/?nobrowser&file=../files/atlas.root&item=LEDShapeHeightCorr_Gain0;1&opt=col   


## Changes in 3.7
1. Support of X axis with custom labels like: http://jsroot.gsi.de/dev/?nobrowser&json=../files/hist_xlabels.json
2. Extend functionality of JSROOT.addDrawFunc() function. One could register type-specific
   `make_request` and `after_request` functions; `icon`, `prereq`, `script`, `monitor` properties.
   This let add more custom elements to the generic gui, implemented with JSROOT.HierarchyPainter   
3. Provide full support of require.js. One could load now JSRootCore.js script like:

      <script type="text/javascript" src="require.js" data-main="scripts/JSRootCore.js"></script>
      
   After this several modules are defined and can be used with syntax like:
   
      require(['JSRootPainter'], function(jsroot) { /*any user code*/});
      
   Also inside JSROOT require.js used to load all dependencies. 


## Changes in 3.6
1. Try to provide workaround for websites where require.js already loaded.
   This makes problem by direct loading of jquery and jquery-ui 
2. Provide workaround for older version of jquery-ui 
3. Prompt for input of command arguments
4. After command execution one could automatically reload hierarchy (_hreload property) or
   update view of displayed object (_update_item property)    
5. Use HiearchyPainter for implementing draw.htm. This let us handle
   all different kinds of extra attributes in central place 
6. Fix problem in tabs layout - new tab should be add to direct child
7. When drawing several tabs, activate frame before drawing - only then
   real frame size will be set
8. Fix problem with GetBBox - it only can be used for visible elements in mozilla.    
9. Support drawing of fit parameters in stat box, use (as far as possible) stat and
   fit format for statistic display 
10. Implement 'g' formatting kind for stat box output - one need to checks 
    significant digits when producing output.  
11. Support new draw options for TGraph: 'C', 'B1', '0', '2', '3', '4', '[]'
12. Primary support for STL containers in IO part. Allows to read ROOT6 TF1.
13. Full support of TGraphBentErrors
14. Support objects drawing from JSON files in default user interface, including
    monitoring. One could open file from link like: https://root.cern.ch/js/dev/?json=demo/canvas_tf1.json 
15. Introduce JSROOT.FFormat function to convert numeric values into string according
    format like 6.4g or 5.7e. Used for statistic display.


## Changes in 3.5
1. Fix error in vertical text alignment
2. Many improvements in TPaletteAxis drawing - draw label, avoid too large ticks.
3. Fix error with col drawing - bin with maximum value got wrong color  
4. Test for existing jquery.js, jquery-ui.js and d3.js libraries, reuse when provided 
5. Fix several I/O problems; now one could read files, produced in Geant4
6. Implement 'e2' drawing option for TH1 class, 
   use by default 'e' option when TH1 has non-empty fSumw2
7. Reuse statistic from histogram itself, when no axis selection done 
8. Support log/lin z scale for color drawing
9. Implement interactive z-scale selection on TPaletteAxis 
10. Allow to redraw item with other draw options (before one should clear drawings)
11. Several improvements in THttpServer user interface - repair hierarchy reload,
    hide unsupported context menu entries, status line update 


## Changes in 3.4
1. Support usage of minimized versions of .js and .css files.
   Minimized scripts used by default on web servers.
2. Implement JSROOT.extend instead of jQuery.extend, reduce
   usage of jquery.js in core JSROOT classes
3. Implement main graphics without jquery at all,
   such mode used in `nobrowser` mode.
4. Provide optional latex drawing with MathJax SVG.
   TMathText always drawn with MathJax,
   other classes require `mathjax` option in URL
5. Improve drawing of different text classes, correctly handle
   their alignment and scaling, special handling for IE
6. Fix error with time axes - time offset was not correctly interpreted


## Changes in 3.3
1. Use d3.time.scale for display of time scales
2. Within JSRootCore.js script URL one could specify JSROOT
   functionality to be loaded: '2d', '3d', 'io', 'load', 'onload'.
   Old method with JSROOT.AssertPrerequisites will also work.
3. With THttpServer JSROOT now provides simple control functionality.
   One could publish commands and execute them from the browser
4. One could open several ROOT files simultaneously
5. Add 'simple' layout - drawing uses full space on the right side
6. Allow to open ROOT files in online session (via url parameter)
7. One could monitor simultaneously objects from server and root files
8. Implement 'autocol' draw option  - when superimposing histograms,
   their line colors will be automatically assigned
9. Implement 'nostat' draw option - disabled stat drawing
10. Using '_same_' identifier in item name, one can easily draw or superimpose
    similar items from different files. Could be used in URL like:
      `...&files=[file1.root,file2.root]&items=[file1.root/hpx, file2.root/_same_]`
      `...&files=[file1.root,file2.root]&item=file1.root/hpx+file2.root/_same_`
    Main limitation - file names should have similar length.
11. When 'autozoom' specified in draw options, histogram zoomed into
    non-empty content. Same command available via context menu.
12. Item of 'Text' kind can be created. It is displayed as
    plain text in the browser. If property 'mathjax' specified,
    MathJax.js library will be loaded and used for rendering.
    See httpcontrol.C macro for example.
13. When using foreignObject, provide workaround for absolute positioning
    problem in Chrome/Safari, see <http://bit.ly/1wjqCQ9>


## Changes in 3.2
1. Support JSON objects embedding in html pages, produced by THttpServer
2. For small resize of canvas use autoscale functionality of SVG. Only when
   relative changes too large, redraw complete canvas again.
3. Use touch-punch.min.js to process touch events with jquery-ui
4. Even when several TH1/TGraph/TF1 objects with fill attribute overlap each other,
   one able to get tooltip for underlying objects
5. Use jquery-ui menu for context menu
6. From context menu one could select several options for drawing
7. Provide user interface for executing TTree::Draw on THttpServer
8. 3D graphic (three.js) works only with IE11


## Changes in 3.1
1. Correctly show tooltips in case of overlapped objects
2. Implement JSROOT.Create() method to create supported
   in JavaScript ROOT classes like TH1 or TGraph
3. Fix problem with JSROOT.draw in HTML element with zero width (display:none)
4. Provide possibility to load user scripts with JSROOT.BuildSimpleGUI
   and JSROOT.AssertPrerequisites, also with main index.htm
5. Support of TCutG drawing
6. Implement hierarchy display (former dtree) with jQuery
7. Fix several problems in drawing optimization
8. Implement dragging objects from hierarchy browser into existing canvas
   to superimpose several objects
9. Implement col2 and col3 draw options, using html5 canvas
10. Support 'p' and 'p0' draw options for TH1 class


## Development of version 3.0

### November 2014
1. Better font size and position in pave stats
2. Resize/move of element only inside correspondent pad
3. Adjust of frame size when Y-axis exceed pad limits
4. Correct values in tooltip for THStack
5. Exclude drawing of markers from TGraph outside visible range
6. Drawing of canvas without TFrame object
7. Many other small bug fixes and improvements, thanks to Maximilian Dietrich

### October 2014
1.  Add "shortcut icon"
2.  Add demo of online THttpServer - shell script copies data from
    running httpserver.C macro on Apache webserver
3.  Evaluate 'monitoring' parameter for online server like:
      <http://localhost:8080/?monitoring=1000>
    Parameter defines how often displayed objects should be updated.
4.  Implement 'opt' and 'opts' URL parameters for main page.
5.  Show progress with scripts loading in the browser window
6.  When one appends "+" to the filename, its content read completely with first I/O operation.
7.  Implement JS custom streamer for TCanvas, restore aspect ratio when drawing
8.  Major redesign of drawing classes. Resize and update of TCanvas are implemented.
    All major draw functions working with HTML element id as first argument.
9.  Extract 3D drawings into separate JSRoot3DPainter.js script
10. Use newest three.min.js (r68) for 3D drawings, solves problem with Firefox.
11. Introduce generic list of draw functions for all supported classes.
12. Add possibility to 'expand' normal objects in the hierarchy browser.
    For instance, this gives access to single elements of canvas,
    when whole canvas cannot be drawn.
13. Correct usage of colors map, provided with TCanvas.
14. Introduce JSROOT.redraw() function which is capable to create or update object drawing.
15. In main index.htm page browser can be disabled (nobrowser parameter) and
    page can be used to display only specified items from the file
16. Add support of TPolyMarker3D in binary I/O

### September 2014
1. First try to handle resize of the browser,
   for the moment works only with collapsible layout
2. Also first try to interactively move separation line between
   browser and drawing field.
3. Small fix of minor ticks drawing on the axis
4. Introduce display class for MDI drawing. Provide two implementations -
   'collapsible' for old kind and 'tabs' for new kinds.
5. Adjust size of color palette drawing when labels would take more place as provided.
6. Add correct filling of statistic for TProfile,
   fix small problem with underflow/overflow bins.
7. Provide way to select display kind ('collapsible', 'tabs') in the simple GUI.
8. Implement 'grid' display, one could specify any number of division like
   'grid 3x3' or 'grid 4x2'.
9. MDI display object created at the moment when first draw is performed.
10. Introduce painter class for TCanvas, support resize and update of canvas drawing
11. Resize almost works for all layouts and all objects kinds.
12. Implement JSROOT.GetUrlOption to extract options from document URL.
13. Provide example fileitem.htm how read and display item from ROOT file.
14. In default index.htm page one could specify 'file', 'layout',
    'item' and 'items' parameters like:
      <http://root.cern.ch/js/3.0/index.htm?file=../files/hsimple.root&layout=grid3x2&item=hpx;1>
15. Support direct reading of objects from sub-sub-directories.
16. Introduce demo.htm, which demonstrates online usage of JSROOT.
17. One could use demo.htm directly with THttpServer providing address like:
     <http://localhost:8080/jsrootsys/demo/demo.htm?addr=../../Files/job1.root/hpx/root.json.gz&layout=3x3>
18. Also for online server process url options like 'item', 'items', 'layout'
19. Possibility to generate URL, which reproduces opened page with layout and drawn items

### August 2014
1. All communication between server and browser done with JSON format.
2. Fix small error in dtree.js - one should always set
   last sibling (_ls) property while tree can be dynamically changed.
3. In JSRootCore.js provide central function, which handles different kinds
   of XMLHttpRequest.  Use only async requests, also when getting file header.
4. Fully reorganize data management in file/tree/directory/collection hierarchical
   display. Now complete description collected in HPainter class and decoupled from
   visualization, performed with dTree.js.
5. Remove all global variables from the code.
6. Automatic scripts/style loading handled via JSROOT.loadScript() function.
   One can specify arbitrary scripts list, which asynchronously loaded by browser.
7. Method to build simple GUI changed and more simplified :). The example in index.htm.
   While loadScript and AssertPrerequisites functions moved to JSROOT, one
   can easily build many different kinds of GUIs, reusing provided JSRootCore.js functions.
8. In example.htm also use AssertPrerequisites to load necessary scripts.
   This helps to keep code up-to-date even by big changes in JavaScript code.
9. Provide monitoring of online THttpServer with similar interface as for ROOT files.
10. Fix several errors in TKey Streamer, use member names as in ROOT itself.
11. Keep the only version identifier JSROOT.version for JS code
12. One can specify in JSROOT.AssertPrerequisites functionality which is required.
    One could specify '2d', 'io' (default) or '3d'.
13. Use new AssertPrerequisites functionality to load only required functionality.
14. When displaying single element, one could specify draw options and monitor property like:
        <http://localhost:8080/Files/job1.root/hpxpy/draw.htm?opt=col&monitor=2000>
     Such link is best possibility to integrate display into different HTML pages,
     using `<iframe/>` tag like:
        `<iframe src="http://localhost:8080/Files/job1.root/hpx/draw.htm"` 
          `style="width: 800px; height:600px"></iframe>`
15. Remove 'JSROOTIO.' prefix from _typename. Now real class name is used.
16. Use in all scripts JSROOT as central 'namespace'
17. Introduce context menu in 3D, use it for switch between 2D/3D modes
18. Use own code to generate hierarchical structure in HTML, replace dtree.js which is
    extremely slow for complex hierarchies. Dramatically improve performance for
    structures with large (~1000) number of items.
19. Deliver to the server title of the objects, display it as hint in the browser.
20. Better handling of special characters in the hierarchies - allows to display
    symbols like ' or " in the file structure.

### July 2014
1. Migration to d3.v3.js and jQuery v2.1.1
2. Fix errors in filling of histogram statbox
3. Possibility of move and resize of statbox, title, color palete
4. Remove many (not all) global variables
5. Example with direct usage of JSRootIO graphics
6. Example of inserting ROOT graphics from THttpServer into `<iframe></iframe>`

### May 2014
1. This JSRootIO code together with THttpServer class included
   in ROOT repository

### March 2014
1. Introduce JSROOT.TBuffer class, which plays similar role
   as TBuffer in native ROOT I/O. Simplifies I/O logic,
   reduce duplication of code in many places, fix errors.
   Main advantage - one could try to keep code synchronous with C++.
2. Avoid objects cloning when object referenced several times.
3. Treat special cases (collection, arrays) in one place.
   This is major advantage, while any new classes need to be implemented only once.
4. Object representation, produced by JSRootIO is similar to
   objects, produced by TBufferJSON class. By this one can exchange
   I/O engine and use same JavaSctript graphic for display.
5. More clear functions to display different elements of the file.
   In the future functions should be fully separated from I/O part
   and organized in similar way as online part.
6. Eliminate usage of gFile pointer in the I/O part.
7. Provide TBufferJSON::JsonWriteMember method. It allows to stream any
   selected data member of the class. Supported are:
   basic data types, arrays of basic data types, TString, TArray classes.
   Also any object as data member can be streamed.
8. TRootSniffer do not creates sublevels for base classes
# JupyROOT
A software layer to integrate Jupyter notebooks and ROOT.

## Installation
1. [Install ROOT6](https://root.cern.ch/building-root) (> 6.05)
2. Install dependencies: pip install jupyter metakernel

## Start using ROOTbooks
Set up the ROOT environment (`. $ROOTSYS/bin/thisroot.[c]sh`) and type in your
shell:
```
root --notebook
```
This will start a ROOT-flavoured notebook server in your computer.

Alternatively, if you would like to use the Jupyter command directly, you 
can do on Linux:
```
cp -r $ROOTSYS/etc/notebook/kernels/root ~/.local/share/jupyter/kernels
jupyter notebook
```
and on OSx:
```
cp -r $ROOTSYS/etc/notebook/kernels/root /Users/$USER/Library/Jupyter/kernels/
jupyter notebook
```

Once the server is up, you can use ROOT with two kernels:

1. ROOT C++: new kernel provided by ROOT
2. Python 2: already provided by Jupyter

##  C++ ROOTbook
ROOT offers a C++ kernel that transforms the notebook in a ROOT prompt.
Embedded graphics, syntax highlighting and tab completion are among
the features provided by this kernel.

An example of how you would plot a histogram in a C++ ROOTbook is:
```cpp
TCanvas c;
TH1F h("h","ROOT Histo;X;Y",64,-4,4);
h.FillRandom("gaus");
h.Draw();
c.Draw();
``` 

## Python ROOTbook
If you prefer to use Python, you can create a new Python 2 kernel and
import the ROOT libraries:
```python
import ROOT
```
And then you could write something like:
```python
c = ROOT.TCanvas("c")
h = ROOT.TH1F("h","ROOT Histo;X;Y",64,-4,4)
```
Additionally, you can mix Python and C++ in the same notebook
by using the **%%cpp** magic:
```cpp
%%cpp
h->FillRandom("gaus");
h->Draw();
c->Draw();
```
The structure of this directory is as follows:

```
.
├── 1000_genomes
│   ├── 1000_genomes # output VCFs from 1KGBP data
│   └── 1000_genomes_pass # output VCFs from 1KBP data that had a quality value above the PASS threshold
├── dx_job_logs # logs from the DNAnexus jobs run for the benchmarking
└── hg002_benchmarks
    ├── HG002-NA24385-50x.70_percent.markdup.realigned.combined.genotyped.vcf # the output VCF from the HG002 benchmark
    └── sv_caller_outputs # individual outputs from SV callers used in the HG002 benchmark
```
