# oce

## 1.7.0

* Add `read.ctd.aml()`.
* Change `read.met()` to handle a new `encoding` argument.
* Fix `drawPalette()` to obey the `at` and `labels` arguments.
* Fix `plot,ctd-method()` to recognize `which=13` for spice.

## 1.6.1 (on CRAN)

* Remove a journal link that has become unavailable.
* Fix "length_1" error (thanks to B. Ripley for the heads-up).

## 1.6.0

* Fix image-size declarations in 3 man pages (required by CRAN).
* Improve sub-class vignette.

## 1.5.0 (on CRAN)

* Add `[["?"]]` facility to all objects.
* Add `showNA` argument to `vectorShow()`.
* Add `adpConvertRawToNumeric()`.
* Add `adpFlagPastBoundary()`.
* Add `ctdRepair()`.
* Add `labelWithUnit()`.
* Add `swSR()`.
* Add `swSstar()`.
* Change `as.ctd()` to drop `other` parameter, deprecated 4y ago.
* Change `read.odf()` to handle more CODE and UNIT values.
* Change `oce.plot.ts()` by adding `simplify` argument.
* Change `plot,section-method()`, correcting temperature label to "T" and
  adding many more plot types.
* Change `pwelch()`, improving low-frequency results.
* Fix `[[` to yield ITS90 temperature for all classes, not just `ctd`.
* Fix `as.ctd()` handling of temperature scale of first argument.
* Fix `as.section()` handling of list of `argo` objects as first argument.
* Fix `colormap()` handling of `name` argument.
* Mark `byteToBinary()` as defunct (slated for removal in 1.6.0).
* Mark `renameData()` as defunct (slated for removal in 1.6.0).

## 1.4.0

* `colormap()` uses `oceColorsViridis` as default colour scheme.
* `download.topo()` uses new NOAA server.
* `mapPlot()` docs warn about problems with version 0.9-8 of sf package.
* `plotProfile()` handles units of added columns.
* `read.argo()` reads more metadata.

## 1.3.0

* Remove a broken link (OK locally but failing in CRAN tests).
* Fix an error in the example for `setFlags,ctd-method`.
* Add `addSpine()` for defining section spines.
* Add `angle2hms()`.
* Add `argoJuldToTime()`.
* Add `data(amsr)` and improve `subset,amsr-method()`.
* Add `data(tidalCurrent)` dataset of tidal currents from Foreman.
* Add `oceAxis()`.
* Add `preferAdjusted()` for `argo-class` data.
* Add `read.ctd.odv()`.
* Add `snakeToCamel()` to convert variable names in e.g. Argo netcdf files.
* Add `timeToArgoJuld()`.
* Change `drawPalette()` default colour palette to `oce.colorsViridis()`.
* Change `imagep()` default colour palette to `oce.colorsViridis()`.
* Change `plot,amsr-method()` default colour palette to `oce.colorsViridis()`.
* Change `plot,section-method()` default colour palette to `oce.colorsViridis()`.
* Change `plotTS()` to trim isopycnals to realistic salinities and temperatures.
* Change `read.argo()` and `data(argo)` to use camelCase in all metadata.
* Change user-oriented github website generator to `pkgdown`.
* Extend `[[,argo-method` to accept `"ID"` as an alternative to `"id"`.
* Extend `as.cm()` to accept `adp` and `adv` objects.
* Extend `oce.plot.ts()`, adding argument `logStyle`.
* Extend `read.adp.rdi()` to handle 38kHz RDI adp files.
* Extend `read.argo()` and `data(argo)` by adding three more 'CYCLE' variables.
* Extend `read.ctd.sbe()` to handle 4 new (Beckman) oxygen variables.
* Extend `read.ctd.sbe()` to handle `.btl` files.
* Fix `plot,coastline-method` error in box drawing.
* Fix `plotTS` error in auto-scaling if S and T have mixed NA status.
* Fix `subset,argo-method()` error in `flags`, `location`, and `*QC` in `metadata`.
* Remove dependence on `rgdal` package, using `sf` for map projections.

## 1.2.0

* Address argument doc/code warning that blocked CRAN acceptance.
* Accommodate new ocedata (needed to meet new CRAN dependency rule).
* Add `oceRenameData()` and `oceRenamemetadata()`.
* Deprecate `renameData()`.
* Improve axis control for `mapPlot()` and `plot,coastline-method`.
* Remove `addColumn()`, `ctdAddColumn()`, `ctdUpdateHeader()`,
  `findInOrdered()`, `mapMeridians()`, `mapZones()`, and
  `oce.as.POSIXlt()`, all of which have been marked as
  "Deprecated" for the past two CRAN releases.
* Extend `read.met()` to read a new Environment Canada data format.
* Handle some upcoming rgdal/R-devel changes.
* Extend `sectionSmooth()` to handle grids beyond data range.
* Extend `read.adp.rdi()` to trim `time` if it longer than other data.
* Improve `geodDist()` to handle second location correctly when `alongPath=FALSE`.
* Improve vignettes.
* Fix itemized-list documentation problem pointed out by Kurt Hornik.

## 1.1-1

* Address "link-time-optimization" issues for CRAN submission.

## 1.1-0

* fix argument error in a .C() call
* Extend `read.odf()` to store ODF header as a list in `metadata`.
* Add `presentTime()`.
* Add vignette on subclassing oce.
* Extend `sectionSmooth()` to handle output grid better, and offer 2 new methods.
* Extend `read.ctd.sbe()` to handle more column names.
* Extend mapping code to support i386/windows.
* Extend `imagep()` to handle combined `flipy` and `ylim` arguments better.

## 1.0-1

* Renamed 0.9-24, released with OAR book publication.

## 0.9-24

* Extend `read.odf()` to store ODF header in `metadata`.
* Remove `makeSection()`, after its deprecate/defunct period ended.
* Make defunct: `addColumn()`, `byteToBinary()`, ` ctdAddColumn()`,
  `ctdUpdateHeader()`, `findInOrdered(), `mapZones()`, `mapMeridians()`,
`oce.as.POSIXlt()`, and `oce.magic()`
* Add `as.tidem()`.
* Extend `read.odf()` to handle multiple `NULL_VALUE`s.
* Improve `handleFlags()`.
* Add `rotateAboutZ()`.
* Add `concatenate()`.
* Add `read.woa()`.
* Extend `colormap()` to include `colfunction` in return value.

## 0.9-23

* tidem() permits inferred constituents
* tidem() follows T_TIDE phase convention
* adorn argument removed from some plot functions (after being defunct 1+ year)
* deprecate findInOrdered()

## 0.9-22

* plot,section-method() can use external bathymetry
* lowpass() added
* ctdTrim() can isolate upcasts
* deprecate byteToBinary()
* address a CRAN NOTE about UTF-8 strings in data

## 0.9-21

* adpEnsembleAverage() added
* mapCoordinateSystem() added
* woce-exchange missing-value inference
* woce-exchange autodetection, plus read all data types
* download.amsr() updated for data-server change
* download.met() added
* as.ctd() can specify columns within argo data
* read.adp.rdi() handles more file types and larger RDI files
* read.odf() handles contradictory metadata better
* oce.as.POSIXlt() deprecated
* oceSetData() handles units better
* address R_registerRoutines build-time warning
* fix memory-usage error in coastlineCut()

## 0.9-20

* remove observatory object type (changed format, seldom-used type)
* remove pangaea object type (undocumented format, seldom-used type)
* improvements for BioArgo data
* Spanish translation by Pablo Vald√©s
* make read.gps() handle more files
* deprecate ctdAddColumn() and ctdUpdateHeader()
* download.topo() added and read.topo() updated to handle new formats
* download.amsr() added
* composite() added
* addColumn() marked as deprecated
* read.cm.s4() reads all data in file
* as.cm() added
* read.odf() handles Aanderaa current meter data
* mapPlot() and mapAxis() get cex.axis parameter
* subset(amsr) added; various other amsr improvements

## 0.9-19

* deprecate 'adorn' in plot functions
* make read.ctd() obey missingValue (renamed)
* improve map projections
* add renameData()
* improve support for SBE files
* remove is.beam(), is.xyz(), is.enu() and coordinate(), since
  x[["coordinate"]] now permits these simply
* replace several function-style accessors with [[ style
* fix test-suite error relating to an rgdal change
* add handleFlags() as generic plus specifics for ctd and argo classes
* use S4 documentation, e.g. ?'plot,ctd-method' instead of ?plot.ctd
* add subset(argo, "adjusted")
* make read.argo() read all documented data and metadata fields
* add trimString()

## 0.9-18

* improve plot.coastline() and mapPlot()
* add support for G1SST satellite
* all objects now have metadata items for units and flags
* ctdTrim() method renamed: old A and B are new A; old C is new B
* support more channels and features of rsk files
* as.adp() added
* convert argo objects to sections
* makeSection() deprecated; use as.section() instead
* read.adp.rdi() handles Teledyne/RDI vsn 23.19 bottom-track data
* geodXyInverse() added; geod functions now spell out longitude etc
* read.odf() speeded up by a factor of about 30
* add colour palettes from Kristen Thyng's cmocean Python package
* as.oce() added
* rename 'drifter' class as 'argo' to recognize what it actually handles
* add oceColorsViridis()
* interpBarnes() has new argument 'pregrid'
* binMean2D() has new argument 'flatten'
* data(topoWorld) now has longitude from -179.5 to 180
* ODF2oce() added
* read.odf() handles more data types
* read.adp.rdi() reads more VmDas (navigational) data
* ITS-90 is now the default temperature unit
* ctd objects can have vector longitude and latitude
* logger class renamed to rsk
* bremen class added
* coastlineCut() added
* rgdal package used instead of local PROJ.4 source code
* mapproj-style map projections eliminated

## 0.9-17

* remove an exit() call in a C function

## 0.9-16

* cite PROJ.4 contributors in DESCRIPTION
* address package-build warnings; improve DESCRIPTION (thanks, BR)

## 0.9-15

* plot.echosounder() gets new argument 'drawPalette'
* data(landsat) taken from ocedata (and shrunk)
* data(nao) and data(soi) moved to ocedata
* mapTissot() added
* read.logger() with ctd-type data infers salinity if necessary
* ctdTrim() can have "method" as a function
* as.topo() can convert "bathy" objects from the marmap package
* "+.section" renamed sectionAddStation()
* oceApprox() renamed oce.approx(), with an alias for compatibility
* oce.grid() added
* "tdr" class renamed "logger"
* swCSTp() added
* swConductivity() now swThermalConductivity(); using Caldwell (1974).
* many sw() functions handle eos="gsw"
* library(gsw) replaces teos()
* curl() added
* handle new NOAA ascii topography dataset type
* mapGrid() added
* data(soi) moved to ocedata package
* lonlat2map() added
* map2lonlat() made more accurate
* permit proj4 style map projections
* landsat-7 (and possibly 4 and 5) support
* landsat decimation/plotting improvements
* drifterGrid() added
* mapDirectionField() added
* mapArrows() added
* read.index() added
* utm2lonlat() and lonlat2utm() added
* data(met) changed to time of Hurricane Juan

## 0.9-14

* fix compile-time warning
* colormap() added, and functionality added to imagep() and drawPalette()
* imagep() and drawPalette() new arg. 'axisPalette' (suggested C. Richards)
* drawPalette() has new args: plot, pos, levels, and cex.axis
* ctdDecimate() permits user-supplied method function
* data(nao) added; data(soi) updated, and names improved in latter
* landsat support added
* plotProfile() and plotTS() get new argument 'pt.bg'
* plot.section() grids the data if needed
* rename 'sealevelHalifax' dataset as 'sealevel'
* translate some axis names (Spanish, French, German and Mandarin)
* plotProfile() now has ytype="depth"
* mapImage() and friends now demand a map exists first
* runlm() added
* mapScalebar() added
* subset.topo() added
* read.topo() now handles GEBCO netcdf files
* decimate() now handles topo objects
* reverse oce.colorsGebco colours for water
* drawIsopycnals() and plotTS(): improve isopycnal labels
* ctdDecimate() handles new method 'unesco'
* coastline improvements (now 3 resolutions)
* webtide() improvements
* read.observatory() added
* read.ctd.itp() added
* mapImage() speeded up (60X in 1deg SST example)
* binApply1D() and binApply2D() added
* binCount1D(), binMean1D(), binCount2D(), and binMean2D() added
* numberAsHMS() added
* gps data type added (provisionally)
* various functions: reorder arguments putting longitude before latitude
* magneticField() replaces magneticDeclination()
* ungrid() added
* mapPlot(): improve zone and meridian aesthetics
* detrend() returns list with detrend vector and coefs
* decodeTime() improvements
* interpBarnes() has new arg 'trim'
* standardDepths() added
* mapLongitudeLatitudeXY() added
* read.ctd.woce() now accepts a format used once by Arctic scientists
* swRrho() added
* plotProfile() now accepts type="o"
* read.adv.nortek(): read salinity
* ctdDecimate(): fix bug for NULL columns [issue 327]
* plot.section() now chooses best coastline file
* swPressure() added
* ctdFindProfiles() added
* as.ctd(): new arguments filename etc
* remove oceBisect(), which merely reproduces uniroot()
* move large/specialized datasets to library(ocedata)

## 0.9-13

* prepare for an upcoming release of R-devel

## 0.9-12

* move vignette source directory

## 0.9-11

* numberAsPOSIXct() now handles types "sas" and "spss"
* data(turbulence) added
* plot.echosounder(): use white for below-threshold values
* plot.echosounder(): add arg 'beam' to e.g. display Sv
* read.echosounder(): handle dual-beam and split-beam data
* read.echosounder(): decode bottom-pick
* swSoundAbsorption()
* read.adp: improve doc regarding variable orientation of RDI files

## 0.9-10

* improve line breaks in some help pages (thanks, B. Ripley)

## 0.9-9

* mapPlot() uses ... graphical parameters (e.g. cex) in labelling axes
* fix recent bug in reading CNV files (issue 318)
* plotTS(..., mar=NULL) now avoids setting par(mar)

## 0.9-8

* fix memory error in teos handler (thanks, B. Ripley)

## 0.9-7

* make ctdTrim() work in R 3.0
* support for adp vmdas nav data (coded by Clark Richards)
* approx3d(): added
* drawPalette(): new arg 'fullpage'
* plot.lobo(): add 'which' argument
* read.coastline.openstreetmap() added
* data(levitus) added
* plot.coastline(): new arguments 'projection' etc to use mapPlot()
* mapContour(): permit first argument to be a 'topo' object
* mapPlot(): add arguments axes, bg, fill, and drawBox

## 0.9-6

* address compilation error on Windows-64 platform
* make some long-running operations more interruptible
* oceApprox(): permit equal x values [issue 279]

## 0.9-5

* halve package check time

## 0.9-4

* as.section() added
* add data(endeavour)
* add map projections, with provisional functions mapPlot() etc
* plot.ctd(): add argument 'add'
* add crude mapping support
* interpBarnes(): add args 'xgl' and 'ygl'
* read.section(): add 'directory' argument
* sectionSmooth(): add 'barnes' method
* oce.plot.ts() and oce.axis.POSIXct(): add tformat argument
* read.aquadopp() and cousins added [issues 253 and 258]
* read.ctd() guesses waterDepth if not supplied [issue 253]
* plotTS(): extend isopycnal range [issue 252]
* plot.section() contouring improved
* data(oceans) added
* data(papa) added
* fillGap() works on matrices as well as vectors
* read.lobo() is more flexible on column names
* imagep(..., axes=FALSE): no longer plots axis box [issue 249]
* oce.plot.ts(): alter time auto-ranging method
* oceBisect(): increase generality of function whose root is sought
* plotTS(): clean axes after isopycnals drawn
* teos10: update test values (with help from PB)
* read.adp.nortek(): handle missing 'to' argument as documented
* beamUnattenuateAdp() renamed beamUnspreadAdp()
* plotTS() and plotProfile(): permit type='n'
* plotProfile(ytype='z'): fix bug in y label
* improve error message if TEOS-10 is missing

## 0.9-3

* example(riley): remove error in R-devel
* plot.ctd(..., which=7) improvement
* swSCTp(): accept conductivity in mS/cm and S/m
* as.ctd(): accept TEOS-10 values SA and CT
* make TEOS routines more resistant to unphysical values

## 0.9-2

* remove unwanted files during build process

## 0.9-1

* pwelch(): fix bug in frequency
* tidem(): add argument 'regress'
* rename oce.plot.sticks() as plotSticks(); delete stickplot()
* use S4 style for all plot() methods
* data(giss) added
* data(topo2) added
* plotPolar() added
* oce.magic() new name for magic(); renamed oceMagic() in 0.9-15
* errorBars() added
* numberAsPOSIXct() add a leap second for July 2012

## 0.8-10

* plot.topo(): automatic cut-point shift
* add support for Nortek aquadopp-profiler instruments
* oce.contour() added
* lisst class added
* plotTaylor() added
* oce.as.raw() added
* binmap() added

## 0.8-9

* make the newly-added TEOS-10 functions compile on Windows

## 0.8-8

* data(cm) added
* plot.coastline(): improve argument 'geographical'
* add prettyPosition()
* add grad()
* add (preliminary) support for TEOS-10, installed separately
* add integrateTrapezoid()
* plotTS(): add argument teos
* add teos() as an interface to the TEOS-10 library
* add moonAngle() plus other astronomy functions, e.g. julianDay()

## 0.8-7

* plotInset(): remove 'bg' and 'fg' arguments
* plotTS(): add 'bg' argument
* rename 'dt' object and functions to 'tdr'
* plotTS(): add argument 'inset'
* read.adp.rdi(): fix a bug in bottom-tracking; add soundSpeed vector
* data(echosounder) created
* formatPosition() created
* plot(echosounder): new arguments atTop and labelsTop
* adp objects now also hold 'percent good' for RDI instruments
* plot.drifter(): improve multi-panel plots
* imagep(): add argument 'missingColor'
* imagep() and drawPalette(): add arguments 'labels' and 'at'

## 0.8-6

* imagep(): permit POSIXt type for zlim

## 0.8-5

* reconstruct data(ctd) so plot will focus maps better
* plotInset() added
* grid() works for oce-based time axes
* imagep(..., filledContours): obey ylab argument
* subset(): permit subsetting ADP by pressure
* oce.plot.ts(): obey arguments xaxt and yaxt
* improve support for shapefile coastlines

## 0.8-4

* replace 'center' argument for plotting topography and coastline with
  clatitude and clongitude
* add 'echosounder' class
* fullFilename(): handle URLs properly
* ADV objects: add access to slow variables (e.g. headingSlow)

## 0.8-3

* add 'met' class
* subset.oce() handles section objects
* draw.section(): permit xtype="latitude" or "longitude"
* imagep() has drawPalette="space" option

## 0.8-2

* fix bug in read.coastline(), reversing lon and lat [issue162]
* fix bug in S4 adp validity checker
* geodDist(): add argument 'alongPath'
* remove compilation warning about doc for summary.ctd()

## 0.8-1

* switch to S4 classes, and add new accessors and 'show' functions.
* add support for RBR 'rsk' files
* plot.section() now uses 'col' if supplied

## 0.7-1

* improve oce.colorsJet()
* improve plotProfile() vertical range with missing data
* add data(drag)
* plotProfile() and plot.ctd(): add xtype="spice"
* add data(geosecs235)
* plotProfile() and plot.ctd(): add argument 'keepNA'
* as.ctd(): add arguments 'other' and 'missingValue'
* read.lobo(): tolerate more formats, including missing velocities
* add data(schmitt)
* add accessor spice()
* handle O2 and nutrients in CTD data and sections
* swSCTp(): use surface pressure as a default
* plotTS(): add arguments inSitu and referencePressure
* plotTS(): handle section objects
* fix bugs reported in issues 150, 151, 153, 155, 156
* imagep(): change default drawContours to FALSE

## 0.6-1

* add 'adv' dataset
* generalize swN2() by adding new arg 'derivs'
* oce.plot.ts() gets new arg 'axes'
* add accessor distance()
* rename to plotTS() and plotScan() to avoid S3 check warnings
* switch makeFilter's argument asKernel to TRUE

## 0.5-1

* add support for ARGO drifters
* latitude() and longitude(): new argument to repeat byDepth
* addCtdColumn(): replace if column already exists
* add CTD accessors for lat, lon, pressure, salinity, temperature

## 0.4-1

* plot.sealevel(): only show 3 panels, for clarity
* plot.ctd(): add argument 'type'
* imagep(): handle x, y, and z as image() does
* oce.plot.ts(): inferred ylim matches data range within provided xlim
* read.adv.nortek(): handle extra analog data
* add CTD accessors longitude(), latitude(), time(), elevation(), etc

## 0.3-1

* add read.oce.odf()
* add findInOrdered
* generalize tidem() argument list
* read.adv.nortek(): handle burst-mode data
* add datasets wilson, redfieldNP, redfieldNC, redfieldPlankton, and riley
* change data(ctd) to be derived from data(ctdRaw)
* change to monitor=FALSE for all reading functions
* add accessor functions heading(), latitude(), longitude(), pitch(),
  pressure(), processingLog(), roll(), salinity(), temperature(),
  time(), and velocity()
* rename history as processingLog
* flatten data objects to be 1 level thick (for future matlab exports)
* add threenum() added, and use it in all summaries (for speed)

## 0.2-3

* head.adp() added
* tail.adp() added
* extract() added

## 0.2-2

* velocityStatistics() added
* new names: oceApprox() integerToAscii() rangeLimit() ctdRaw
* topoInterpolate() added
* numberAsPOSIXct(): add type 'argos'
* rename beamAttenuateAdp() as beamUnattenuateAdp()

## 0.2-1

* camel-case function and argument names (and the NEWS file)

## 0.1-83

* add binAverage()
* fillGap(): add 'rule' argument (analogous to same for approx())
* add rescale()
* read.pt(): fix timing error; use as.pt() to create return value
* add as.pt()
* read.adp.rdi() uses checksums
* number.as.POSIXct(type="gps"): account for leap seconds

## 0.1-82

* remove warning that occurs for R 2.13.0 (alpha)
* number.as.POSIXct(): accept GPS times
* rename e.g. adv.2enu() as to.enu.adv()
* speed up conversion of ADP and ADV to enu by more than a factor of 10
* change RDI coordinate handling
* rename match.bytes() as matchBytes()
* add drawPalette()
* remove matlab2POSIXt(), now a sub-case of number.as.POSIXct()
* number.as.POSIXct(): added
* oceBiset(): added
* despike(): add argument 'action'
* detrend(): added
* read.adp.sontek(): handle PCADP type
* read.adp.sontek.serial(): added
* read.section(): handle WOCE quality flags for salinity
* retime(): added
* read.adv.sontek.serial(): correct error in class of 'a' and 'c'
* despike(): use string value for argument 'method'
* logger.toc(): fix bug in discovering files ranging over days
* as.ctd(): make temperature and salinity into vectors, if not already
* plot.TS(): add argument 'use.smoothScatter'
* oce.debug(): flush the console after printing a message
* sw.theta(): rename the method possibilities, to lower case
* in some sw functions, rename 'pref' as 'reference.pressure'
* read.ctd(): fix bug in getting start.time for some time formats
* pwelch() added
* plot.sealevel(): make y axis obey range in data subset, if xlim is given
* adp.2enu() added
* coastline.world: improve resolution by a factor of 4 or so
* read.ctd.woce(): infer water depth as max(pressure) if not in header
* section.smooth(): handle misordered stations; handle missing values better
* plot.section(): allow strings for 'which'; improve contouring
* plot.ctd() and plot.profile(): add 'use.smoothScatter' argument
* coriolis(): improve omega value

## 0.1-81

* add despike()
* add range.limit()
* add unabbreviate.time()
* add support for bottom-tracking RDI ADCPs
* add support for interocean 's4' current meters
* add unwrap.angle()
* read.adv.nortek(): detect the velocity range
* add processing.log.add(), an alternative to processing.log.append()
* plot.ctd(): add more plot types; which=9 for salinity, not density
* as.ctd(): improve flexibilty
* predict.tidem(): add 'newdata' argument
* adv.2enu() added
* adp.xyz2enu(): make it correct for heading.bias
* make.filter(): can now produce tskernel type
* improve (but temporarily limit) fill.gap
* plot.adp(): add 'use.layout' argument
* window.oce() added
* make objects remember full filename, not just local filename (issue8)
* plot.tidem(): remove argument 'plot.type', using 'which' instead
* read.pt(): add arguments from, by, and to
* fix issue 57 [read.adp.rdi() read the heading incorrectly]
* add "rr" method to ctdDecimate()
* oceApprox() added
* topoWorld dataset added
* plot.ctd(): by default, make coastlines extend to box boundaries
* reformulate sun.angle in R
* fix google-code issue 56 [plot.topo() should accept land.z=NULL]
* fix google-code issue 55 [plot.topo() could go past poles]
* fix google-code issue 54 [plot.topo() was resetting par() on exit]
* fix google-code issue 53 [interp.barnes() gives poor error msg for
  mismatched x and y]
* fix google-code issue 52 [plot.section() draws incorrect bottom shape]

## 0.1-80

* add sun.angle [not tested yet; needs R reformulation]
* fix google-code issue 51 [summary.section() does not report water depth]
* fix google-code issue 50 [non-bug relating to sun elevation]
* fix google-code issue 49 [demo(TS) broken]

## 0.1-79

* fix google-code issue 48 [oce.plot.sticks() ignored page ratio]
* fix google-code issue 47 [cannot read new MEDS sealevel format]

## 0.1-78

* add data(RRprofile)
* to plot.TS(), add arguments lwd.rho and lty.rho

## 0.1-77

* refactor adp code wrt transformation matrices
* add matrix.smooth()
* improve labels for oce.axis.POSIXct()
* set default for 'debug' argument to getOption("oceDebug") in all functions
* plot.profile(): add argument ytype; change 'type' to 'xtype'
* add swZ()
* add oce.smooth()
* add read.adv(), etc., supporting Nortek and Sontek devices
* add read.adp(), etc., supporting Nortek, RDI and Sontek devices
* add oce.plot.ts()
* add imagep()
* add bcd2integer()
* add matlab2POSIXt()
* to most plot functions, add 'mgp' and 'mar' arguments
* plot.pt(): add plim and Tlim arguments
* gravity(): give default for latitude argument
* plot.sealevel(): remove argument focus.time and add argument which
* rename tdr functions as pt, to reflect oral convention
* add geod.xy()
* add argument 'mgp' to all plotting functions
* use abbreviations for axis names if space is tight
* add argument 'adorn' to all plot() functions
* add oce.colorsTwo(), oce.colorsJet(), and oce.colorsPalette()
* add byte2binary()
* add "lty.grid" to all CTD plotting functions
* add "+.section", a more convenient way to build sections from stations
* rework summary() for all existing objects
* as.sealevel(): rename sampling.interval as deltat

## 0.1-76

* sw.N2(): make it tolerate NAs; adjust df to make result smoother
* add makeFilter()
* add decimate(), which may eventually replace ctd.decimate()
* read.tdr(): handle 5-column data files; add tz argument
* plot.profile(): add types sigma+dpdt and sigma+time
* ctdTrim(): add pmin parameter
* plot.ctd(): add Slim, Tlim, plim, lonlim, latlim args; add maps
* add coastline.sle dataset
* plot.TS(): fix isopycnal labels for fresh water; add args Slim, Tlim
* read.ctd(): make it understand another SBE format
* add parseLatlon()
* oce.edit(): add argument 'action'
* add oce.write.table()
* add fillGap()

## 0.1-75

* add addColumn()
* add undriftTime()
* add tdrPatm()
* make readTdr() gather serial no. info; show this on plot.pt()
* switch to recommended version-number format
* remove section.subset(), replaced by subset()
* improve log items in functions that read and assemble oce objects
    http://code.google.com/p/r-oce/issues/detail?id=38
* add subset.oce()
* add header()
* add argument 'which' to read.rbrdtr()
* make read.rbrdtr() understand headers better
* add arguments 'xtype' and 'ytype' to plot.section()
* improve accuracy of bottom drawn by plot.section()
* add section.smooth()
* fix bug making section.smooth() fail if <4 good data at a level
    http://code.google.com/p/r-oce/issues/detail?id=36
* fix bug making summary.ctd() fail for results of section.smooth()
    http://code.google.com/p/r-oce/issues/detail?id=34
* fix bug making sw.dynamic.height() choke on empty stations
    http://code.google.com/p/r-oce/issues/detail?id=33
* fix bug making plot.section() die if x is unordered
    http://code.google.com/p/r-oce/issues/detail?id=36
* rename rbrpt functions to tdr functions, improving generality

## 0.1.74

* add argument "src" to as.ctd()
  http://code.google.com/p/r-oce/issues/detail?id=22
* plot.section() RHS axis needs tics
  http://code.google.com/p/r-oce/issues/detail?id=32
* trim axis whitespace in plot.profile() and plot.TS()
  http://code.google.com/p/r-oce/issues/detail?id=31
* make read.oce() understand WOCE section type
  http://code.google.com/p/r-oce/issues/detail?id=30
* speed up swSpice()
  http://code.google.com/p/r-oce/issues/detail?id=29
* add read.pt(), summary.pt(), plot.pt(), and ptTrim()
* make plot.TS() isopycnal label size be same on top and right sides
  http://code.google.com/p/r-oce/issues/detail?id=26
* add lwd argument to plot.profile()
* give plot.section() ability to control contour levels and labels
* give plot.section() tics for station-location

## 0.1.73

* plot.topo() narrows autoscale to xlim-ylim region, if provided
* add plot.ctd() 'textpanel' option, and put the profiles in the same row
* fix bug in plot.profile() to let it take Slim,..., as args
  http://code.google.com/p/r-oce/issues/detail?id=25
* fix bug in plot.profile() font size
  http://code.google.com/p/r-oce/issues/detail?id=24
* fix bug in plot.profile(type="S_T") positioning
  http://code.google.com/p/r-oce/issues/detail?id=23
* fix bug in section plot (bottom was missing)
  http://code.google.com/p/r-oce/issues/detail?id=21
* fix bug in size of salinity axis label for plot.ctd()
* use pch=21 for TS plots (so data density is more visible)

## 0.1.72

* add interp.barnes()

## 0.1.71

* fix bug in makeSection (ignored the list, if a list provided)
  http://code.google.com/p/r-oce/issues/detail?id=18&q=label:Type-Defect
* add oce.edit(), later renamed oceEdit()
* add topoMaritimes dataset
* add read.topo(), plot.topo(), and summary.topo()
* add gebcoColors() ... renamed oce.colorsGebco() in version 0.1.77
* make plot.section() check that pressures coincide

## 0.1.70

* plot.TS(): make isopycnal list work better for nearly-fresh water
* trimCtd(): improve equilibration-phase deletion
* read.ctd(): handle cases without cruise information or scan column
* permit setting xlab and ylab in plot.TS()
* make processing log timestamps be in GMT
* add as.windrose() and plot.windrose()
* add sealevel.tuk sea-level dataset
* add tidem(), predict.tide(), summary.tide(), plot.tide(), data(tidedata)
* establish a uniform form for objects created by "read" and "as"
* let swRho() and similar functions take a ctd object as a single argument
* add swDynamicHeight()
* make section.grid() and ctd.decimate() extrapolate to surface
* add summary.oce()
* make all objects inherit from a new class "oce"
* fix google-code issue 12: read.sealevel() had a hard-wired filename

## 0.1.69

* modify documentation slightly
* start a migration to more uniform object structures
* rename as.CTD() to as.ctd(), to make it consistent with similar functions
* add read.section()
* add sectionGrid(), which only grids in p at the moment
* add a03 dataset (renamed 'section' in 0.9-13)
* fix SF bug 1833719: warnings from read.ctd()
* fix google-code issue 8: read.coastline() not producing data
* improve ability of trimCtd() to ignore spurious initial data
* add "connectPoints" option to plot.TS()
* add "station" item to "ctd" object
* change as.CTD() to produce sigma.theta instead of sigma

## 0.1.68

* add coastline.world; increase resolution of coastline.maritimes
* plot.TS() no longer rotates the RH margin isopycnal labels

## 0.1.67

* allow swRho() and similar take matrices as arguments for S, T, etc

## 0.1.66

* fix bug in plot.TS() that prevented it from showing some isopycnals

## 0.1.65

* switch to UNESCO formulation for sw potential temperature by default
* prevent warnings with C90 compiler on some machines

## 0.1.64

* add read.oce(), a generic function for reading several oceanographic files
* make read.ctd() understand WOCE-exchange files
* make read.sealevel() understand comma-separated data from MEDS
* make plot.sealevel() skip spectral graphs if timeseries has NA values
* improve aesthetics of plot.ctd()
* extend plot.sealevel() by adding focus.time argument
* add ctdDecimate()
* add oce.as.POSIXlt()
* add latlonFormat(), latFormat(), and lonFormat()
* add as.coastline()
* add make.section()
* add summary.section()
* add plot.section()
* add data(section)

## 0.1.63

* add swSoundSpeed()

## 0.1.62

* add historyAppend() (later named processingLogAppend()) and related code
* improve ctdTrim()

## 0.1.61

* add rudimentary read/summary/plot functions for lobo datasets

## 0.1.60

* add vignette; address some build-stage warning messages

## 0.1.59

* set up data files to permit latin1 encoding of some content

## 0.1.58

* rename all seawater-related functions to e.g. swTheta(), to allow
  for atmospheric analogs

## 0.1.57

* make as.CTD() accept length(p)=1, e.g. for surface plots

## 0.1.56

* make read.coastline() handle S files

## 0.1.55

* keep swTheta() from complaining if length(p) = 1

## 0.1.54

* add swAlpha() and swBeta()

## 0.1.53

* add swAlphaOverBeta()

## 0.1.52

* document the use of df in swN2(), plot.ctd(), etc.

## 0.1.51

* permit swRho() (and similar) to take NAs in args (bug B32)

## 0.1.50

* permit swRho() and descendants to take scalar pressure.

## 0.1.49

* add as.CTD()

## 0.1.48

* add "UNESCO1983" formulation to swTheta() as an alternative

## 0.1.47

* speed up swTheta() by moving vector work from R to C

## 0.1.46

* make geodDist() accept a vector for location 1 and a scalar for location 2

## 0.1.45

* make read.coastline() understand matlab and Splus formats

## 0.1.44

* make plot.coastline() scale lat and lon correctly

## 0.1.43

* add read.coastline(), summary.coastline(), and plot.coastline()

## 0.1.42

* add as.sealevel()

## 0.1.41

* add read.sealevel(), summary.sealevel(), and plot.sealevel()

## 0.1.40

* add name option to plot.ctd.scan()

## 0.1.39

* add lapse.rate()

## 0.1.38

* fix B27 (header not updated for columns added during read.ctd)

## 0.1.37

* rename as ctdTrim()
* rename to ctdWrite()

## 0.1.36

* add write.ctd()
* add ctdAddColumn()
* add ctdUpdateHeader()

## 0.1.35

* add plot.ctd.scan()

## 0.1.34

* make read.ctd() calculate S if missing but C, T, and p are given
* add swSCTp()

## 0.1.33

* add swConductivity()

## 0.1.32

* rename to swRho()

## 0.1.31

* use namespace

## 0.1.30

* add oce.viscosity()

## 0.1.29

* fix B23 (not done yet)
* fix B22 (not done within oce; done at system level)

## 0.1.28

* add pressure method to trimCtd()

## 0.1.27

* vectorize swRho() since it was too slow on large datasets

## 0.1.26

* make demo(oce.ctd) print data

## 0.1.25

* fix some documentation errors that the 2.0.0 CHECK noticed

## 0.1.24

* profiles have correct y-axis
* OSX port, but must first do e.g. export PKG_LIBS="-L/sw/lib"

## 0.1.23

* OSX port (you must first do export PKG_LIBS="-L/sw/lib")

## 0.1.22

* add trim.ctd()

## 0.1.21

* permit lat/lon with the "N"-type indicator at start

## 0.1.20

* fix bug B16 (label the isopycnals on TS diagram)

## 0.1.19

* make read.ctd handle more file types

## 0.1.18

* add oceCoriolis, oceDepth, oceGravity, oceTFreeze

## 0.1.17

* add oceSpecificHeat oce.N2; fix bug B18

## 0.1.16

* get data(ctd) working

## 0.1.15

* add oceSTRho

## 0.1.14

* fix bug B15 (maybe)

## 0.1.13

* add oceTSRho

## 0.1.12

* fix bugs B10, B11, and B13

## 0.1.11

* fix bugs B6, B7, and B12

## 0.1.10

* add geodDist()
* add plot.ctd(), which is rudimentary for now
* fix bug B8

## 0.1.9

* add spice() function

## 0.1.8

* permit all eos functions to handle missing values
* fix bug B4

## 0.1.7

* add summary(ctd) method.
* make read.ctd() return an object of type ctd.
* get tests/ working.

## 0.1.6

* create tests/oce.test1.R

## 0.1.5

* add oceSigma(), oceSigmaT(), oceSigmaTheta()

## 0.1.4

* permit various EOS functions to handle lists
* fix bug B3

## 0.1.3

* add ocerho()

## 0.1.2

* fix file-path bug
* fix bug B2

## 0.1.1

* add theta()
* fix bug B1

## 0.1.0

* add read.ctd()

# Submission of 1.6-1

This submission addresses two problems of which I learned since a submission of
1.6-0 yesterday:

1. Use a doi{} rather than a url{} for a journal article, preventing reliance
   on what seems to be an unstable website.
2. Fix two "length_1" errors, kindly pointed out by B. Ripley in an email dated
   2022-02-18.  (This involves using `&&` or `||` where `&` and `|` was
   proper.)


# Tests

## Local Tests

Local MacOS-12.2(beta) R-4.1.2 CMD (BUILD, INSTALL, CHECK): no ERRORs, no
WARNINGs, and 2 NOTEs, one naming the author the other reporting that
sub-directories were of size of 1MB or more: R 3.0Mb, data 1.0Mb, doc 2.0Mb,
and help 3.7Mb.

## Github R-CMD-check Action Tests

Passes on all tested platforms:

* windows-latest (release)
* macOS-latest (release)
* ubuntu-20.04 (release)
* ubuntu-20.04 (devel)

R-CMD-check github action reports no problems on
* windows-latest (release)
* macOS-latest (release)
* ubuntu-20.04 (release)
* ubuntu-20.04 (devel)


## Remote Windows Checks

Using
```R
devtools:::build_win_release()
devtools:::build_win_oldrelease()
devtools:::build_win_devel()
```
identified no errors.

# Reverse Dependency Checks

Using
```
# devtools::install_github("r-lib/revdepcheck")
revdepcheck::revdep_check(quiet=FALSE)
```
identified no errors and no warnings.

# Contributor Covenant Code of Conduct

## Our Pledge

In the interest of fostering an open and welcoming environment, we as contributors and maintainers pledge to making participation in our project and our community a harassment-free experience for everyone, regardless of age, body size, disability, ethnicity, gender identity and expression, level of experience, nationality, personal appearance, race, religion, or sexual identity and orientation.

## Our Standards

Examples of behavior that contributes to creating a positive environment include:

* Using welcoming and inclusive language
* Being respectful of differing viewpoints and experiences
* Gracefully accepting constructive criticism
* Focusing on what is best for the community
* Showing empathy towards other community members

Examples of unacceptable behavior by participants include:

* The use of sexualized language or imagery and unwelcome sexual attention or advances
* Trolling, insulting/derogatory comments, and personal or political attacks
* Public or private harassment
* Publishing others' private information, such as a physical or electronic address, without explicit permission
* Other conduct which could reasonably be considered inappropriate in a professional setting

## Our Responsibilities

Project maintainers are responsible for clarifying the standards of acceptable behavior and are expected to take appropriate and fair corrective action in response to any instances of unacceptable behavior.

Project maintainers have the right and responsibility to remove, edit, or reject comments, commits, code, wiki edits, issues, and other contributions that are not aligned to this Code of Conduct, or to ban temporarily or permanently any contributor for other behaviors that they deem inappropriate, threatening, offensive, or harmful.

## Scope

This Code of Conduct applies both within project spaces and in public spaces when an individual is representing the project or its community. Examples of representing a project or community include using an official project e-mail address, posting via an official social media account, or acting as an appointed representative at an online or offline event. Representation of a project may be further defined and clarified by project maintainers.

## Enforcement

Instances of abusive, harassing, or otherwise unacceptable behavior may be reported by contacting the project team at kelley.dan@gmail.com. The project team will review and investigate all complaints, and will respond in a way that it deems appropriate to the circumstances. The project team is obligated to maintain confidentiality with regard to the reporter of an incident. Further details of specific enforcement policies may be posted separately.

Project maintainers who do not follow or enforce the Code of Conduct in good faith may face temporary or permanent repercussions as determined by other members of the project's leadership.

## Attribution

This Code of Conduct is adapted from the [Contributor Covenant][homepage], version 1.4, available at [http://contributor-covenant.org/version/1/4][version]

[homepage]: http://contributor-covenant.org
[version]: http://contributor-covenant.org/version/1/4/
There are several ways to contribute to oce. These range from making bug
reports and suggesting new features, through proposing code/documentation
changes, all the way to providing patches or pull requests for new
code/documentation. It is important to note that contributions nearer the start
of this spectrum are more common than those nearer the end, and that they can
be just as valuable. Oce is meant to be a tool to help oceanographers in their
work, and so the oce developers are always interested to hear about areas for
improvement. Not all ideas will be taken up, of course, since practicality is
at the heart of oce. The developers weigh benefits against costs, and they are
busy with their own science, so they tend to focus their time on important
things that can be handled easily.

In the history of oce, very few users have suggested gratuitous or bad code
changes, so there seems to be little reason to describe here the usual sort of
suggestions, beyond the obvious: (a) communicate with the developers before
suggesting a large code change; (b) use github resources such as pull requests
to submit suggested changes; (c) test any suggested code not just on your data
but also by running the 'check' procedures in R; and (d) pattern your code on
that used in oce, with respect to variable names, spacing, etc.

The oce developers are scientists, not programmers, and this means that they
are quite aware of the need to keep data private until publication. For this
reason, users who report bugs or request the support of a new type can be sure
that emailing (or otherwise providing) data to the oce developers will not put
their data in jeopardy. An issue reporter should upload data to github *only*
if they are sure that the data are public, for once something goes into github,
it is likely open forever.

**Preface.** This file is for developers only. It is in the ``.Rbuildignore`` file, so it
does not get included in the R package.

# How to build oce through a Makefile

The following works well. It can help to have a target (called ``ocequick``
perhaps) that skips the ``CHECK`` action.

    OCEVSN=$(shell awk '/Version/{print($$2)}' oce/DESCRIPTION)
    oce: force
        cd oce ; Rscript -e "roxygen2::roxygenise()"
        R CMD BUILD oce
        R CMD CHECK --as-cran oce_${OCEVSN}.tar.gz
        R CMD INSTALL oce_${OCEVSN}.tar.gz
	
# Diary




# oce <img src="https://raw.githubusercontent.com/dankelley/oce/develop/oce-logo-3.png" align="right" height="95" />

<!-- badges: start -->

[![status](https://joss.theoj.org/papers/10.21105/joss.03594/status.svg)](https://joss.theoj.org/papers/10.21105/joss.03594)
[![R-CMD-check](https://github.com/dankelley/oce/workflows/R-CMD-check/badge.svg)](https://github.com/dankelley/oce/actions)
[![Codecov test
coverage](https://codecov.io/gh/dankelley/oce/branch/develop/graph/badge.svg)](https://app.codecov.io/gh/dankelley/oce?branch=develop)
[![CRAN\_Status\_Badge](https://www.r-pkg.org/badges/version/oce)](https://cran.r-project.org/package=oce)
![RStudio CRAN mirror
downloads](https://cranlogs.r-pkg.org/badges/last-month/oce) ![RStudio
CRAN mirror downloads](https://cranlogs.r-pkg.org/badges/last-week/oce)
![RStudio CRAN mirror
downloads](https://cranlogs.r-pkg.org/badges/last-day/oce)
<!-- badges: end -->

## Why use R for oceanographic analysis?

The R language is popular in many branches of science, and Oceanography is no
exception. With its broad statistical support, R is a natural choice for
oceanographers in the biological, chemical and geological sub-disciplines.
However, some physical oceanographers have remained attached to Matlab, which
was widely adopted during the 1990s. Lately, this has been changing, as
oceanographers turn to open-source systems such as Python and R. A particular
strength of R is its provision of many powerful and well-vetted packages for
handling specialized calculations. The oce package is a prime example.

## About oce

The oce package handles a wide variety of tasks that come up in the analysis of
Oceanographic data. In addition to the present README file, a brief sketch of
the package has been written by the core developers (Kelley Dan E., Clark
Richards and Chantelle Layton, 2022. [oce: an R package for Oceanographic
Analysis](https://doi.org/10.21105/joss.03594). Journal of Open Source
Software, 7(71), 3594), and the primary developer uses the package extensively
in his book about the place of R in oceanographic analysis (Kelley, Dan E.,
2018. [Oceanographic Analysis with
R](https://link.springer.com/us/book/9781493988426). New York. Springer-Verlag
ISBN 978-1-4939-8844-0).  Details of oce functions are provided within the
R help system, and in the package [webpage](https://dankelley.github.io/oce/).

## Installing oce

Stable versions of oce are available from CRAN, and may be installed
from within R, in the same way as other packages. However, the CRAN
version is only updated a few times a year (pursuant to policy), so many
users install the `"develop"` branch instead. This branch may be updated
several times per day, as the authors fix bugs or add features that are
motivated by day-to-day usage. This is the branch favoured by users who
need new features or who would wish to contribute to Oce development.

The easy way to install the `"develop"` branch is to execute the
following commands in R.

    remotes::install_github("dankelley/oce", ref="develop")

and most readers should also install Ocedata, with

    remotes::install_github("dankelley/ocedata", ref="main")

## Evolution of oce

Oce is emphatically an open-source system, and so the participation of
users is very important. This is why Git is used for version control of
the Oce source code, and why GitHub is the host for that code. Users are
invited to take part in the development process, by suggesting features,
by reporting bugs, or just by watching as others do such things.
Oceanography is a collaborative discipline, so it makes sense that the
evolution of Oce be similarly collaborative.

## Examples using built-in datasets

### CTD

    library(oce)
    data(ctd)
    plot(ctd, which=c(1,2,3,5), type="l", span=150)

![Sample CTD plot.](https://raw.githubusercontent.com/dankelley/oce/develop/oce-demo-1.png)


### Acoustic Doppler profiler

    library(oce)
    data(adp)
    plot(adp)

![Sample adp plot.](https://raw.githubusercontent.com/dankelley/oce/develop/oce-demo-2.png)

### Sealevel and tides

    library(oce)
    data(sealevel)
    m <- tidem(sealevel)
    par(mfrow=c(2, 1))
    plot(sealevel, which=1)
    plot(m)

![Sample sealevel plot.](https://raw.githubusercontent.com/dankelley/oce/develop/oce-demo-3.png)

### Echosounder

    library(oce)
    data(echosounder)
    plot(echosounder, which=2, drawTimeRange=TRUE, drawBottom=TRUE)

![Sample echosounder plot.](https://raw.githubusercontent.com/dankelley/oce/develop/oce-demo-4.png)

### Map

    library(oce)
    par(mar=rep(0.5, 4))
    data(endeavour, package="ocedata")
    data(coastlineWorld, package="oce")
    mapPlot(coastlineWorld, col="gray")
    mapPoints(endeavour$longitude, endeavour$latitude, pch=20, col="red")

![Sample map plot.](https://raw.githubusercontent.com/dankelley/oce/develop/oce-demo-5.png)

### Landsat image

    library(ocedata)
    library(oce)
    data(landsat)
    plot(landsat)

![Sample landsat image plot.](https://raw.githubusercontent.com/dankelley/oce/develop/oce-demo-6.png)

Instructions: please remove these sentences, replacing them with (1) a sentence or two describing the issue, (2) a pared-down code example, and (3) the output from `sessionInfo()`. An ideal way to prepare your code is to use the `reprex` function from the `reprex` package. If your code needs data, please attach relevant files to your comment, or, if the data are private, email the files to the developers (kelley.dan@gmail.com and clark.richards@gmail.com), who will respect your privacy.

**Contents of directory.** This document explains how the files in this
directory were created.  In many cases, the data are truncated using `dd`,
because the original files are too large to be stored on github.

**Instructions for adding to this directory.** 

1. Place entries in alphabetical order. Change the numbering of the headers, as
   required. Try to use file names that will have some meaning to a reader
doing a file listing. 

2. Put an entry in a `.R` file in the parent directory to read the file and
   test the resultant data in some useful way. The file name should be start
with `test_local`.  Some tests may involve comparison against a second source,
e.g. information provided by manufacturer's software or inferred by inspection
of the data file with an editor or other software.  It also very helpful to
have consistency checks to guard against changes to oce; these can be
constructed simply by reading in the data once, and creating a test that the
same results are obtained by rereading.

**How these file are used.** Simply adding a file here does nothing useful.  

# 1. ADP: RDI, Nortek, and Sontek

## 1.1 RDI
```
dd if=/data/archive/sleiwex/2008/moorings/m09/adp/rdi_2615/raw/adp_rdi_2615.000 \
   of=rdi bs=40000 count=1
```

## 1.2 Nortek Aquadopp
```
dd if=/data/archive/sleiwex/2008/moorings/m05/adp/nortek_aqd2843/raw/adp_nortek_aqd2843.prf \
   of=adp_nortek_aquadopp bs=40000 count=1
```

## 1.3 Sontek
```
dd if=/data/archive/sleiwex/2008/moorings/m07/adp/sontek_h53/raw/adp_sontek_h53.adp \
   of=adp_sontek bs=40000 count=1
```



# 2. ADV

## 2.1 Nortek

```
dd if=/data/archive/sleiwex/2008/moorings/m05/adv/nortek_1943/raw/adv_nortek_1943.vec \
   of=adv_nortek_vector bs=40000 count=1
```

## 2.w Sontek

```
dd if=/data/archive/sleiwex/2008/moorings/m03/adv/sontek_b373h/raw/adv_sontek_b373h.adr \
   of=adv_sontek bs=40000 count=1
```



# 3. Argo

```
cp ~/git/oce/create_data/argo/6900388_prof.nc .
mv ~/Dropbox/BR5904179_001.nc . # from a server; this is a BIOargo with new fields
```


# 4. AMSR satellite

```
cp ~/data/amsr/f34_20160808v7.2.gz .
```


# 5. Bremen

```
cp /data/flemishCap/msm27_ladcp/msm27_003.ladcp .
```


# 6. cm

```
cp /data/archive/sleiwex/2008/moorings/m11/cm/interocean_0811786/manufacturer/cm_interocean_0811786.s4a.tab .
```

# 7. CTD

18HU2010014_00003_00001_ct1.csv is from https://cchdo.ucsd.edu/data/9837/18HU20100513_ct1.zip


```
curl https://www.nodc.noaa.gov/woce/woce_v3/wocedata_1/whp/exchange/example_ct1.csv > example_ct1.csv
```


# Echosounder

```
cp /data/archive/sleiwex/2008/fielddata/2008-07-04/Merlu/Biosonics/20080704_112452.dt4 echosounder.dt4
```

# 8. GPS

```
cp ~/src/rgdal/branches/rgdal_iconv/inst/vectors/test_trk2.gpx .
```

# 9. Ice-tethered profiler
The original source was ftp://ftp.whoi.edu/whoinet/itpdata/itp99grddata.zip
```
cp  ~/Downloads/itp99grddata/itp99grd0000.dat  .
```

# 10. Index

```
curl http://www.cgd.ucar.edu/cas/catalog/climind/SOI.signal.ascii > SOI.signal.ascii
```

# 11. Lobo

```
cp ../../../create_data/lobo/lobo.dat  .
```

# 12. a broken ctd file

`CTD_98911_1P_1_DN.txt` is from an ODF file that CR had. It has p, temperature,
and theta, but no salinity or conductivity. It is worth testing because it's
odd, basically.

# 13. Nortek Signature adp file in ad2cp format

`ad2cp_01.ad2cp` is the first 20,000 bytes of a source file generously provided
by github user @krillthor, who started issue 1676 and provided tests that
helped greatly during the code improvements motivated by that issue.  It was
created with

```
head -c 50000 X.ad2cp > ad2cp_01.ad2cp
```

where `X.ad2cp` is a stand-in for the original file name.

# sample01.txt

Sample file from an oce user.

See https://github.com/dankelley/oce/issues/1696 for details.

# sample02.txt

First 100 lines of a CTD sample file named `ODV_vertical_profiles_ctd.txt`,
acquired by clicking the "EXAMPLES OF SEADATANET ODV FILES" button at
https://www.seadatanet.org/Standards/Data-Transport-Formats

# sample03.txt

First 100 lines of a CTD sample file named
`ODV_vertical_profiles_ctd_with_instruments.txt`, acquired by clicking the
"EXAMPLES OF SEADATANET ODV FILES" button at
https://www.seadatanet.org/Standards/Data-Transport-Formats

# sample04.txt

First 100 lines of a CTD sample file named `ODV_timeseries_chemical_data.txt`,
acquired by clicking the "EXAMPLES OF SEADATANET ODV FILES" button at
https://www.seadatanet.org/Standards/Data-Transport-Formats


This directory is used by the developers, to explore ideas for oce enhancements
in full view of users. However, nothing stored here enters the oce package,
since this directory is named in the `.Rbuildignore` file.

Directories herein are usually named for github issues, so that e.g. the
directory named `1490` stores work related to issue 1490 (see
https://github.com/dankelley/oce/issues/1490).

---
title: Updating oce data(topoWorld)
author: Dan Kelley
date: 2019 May 11
---

To get an up-to-date idea of the URL for topo data I visited

https://maps.ngdc.noaa.gov/viewers/wcs-client/

and used a dialog box to set lonW=0 lonE=10 latS=0 latN=10, which yielded the following URL at the 'download' link.

https://gis.ngdc.noaa.gov/mapviewer-support/wcs-proxy/wcs.groovy?filename=etopo1_bedrock.nc&request=getcoverage&version=1.0.0&service=wcs&coverage=etopo1_bedrock&CRS=EPSG:4326&format=netcdf&resx=0.016666666666666667&resy=0.016666666666666667&bbox=0,0,9.999999999999831,9.999999999999837

For comparison, oce create_data/topo/create_topoWorld_new.R uses

```
topoFile <- download.topo(west=-179.5, east=180, south=-89.5, north=90,
                          resolution=30, format="netcdf", destdir=".", debug=3)
```
uses the URL

https://gis.ngdc.noaa.gov/cgi-bin/public/wcs/etopo1.xyz?filename=etopo1.nc&request=getcoverage&version=1.0.0&service=wcs&coverage=etopo1&CRS=EPSG:4326&format=netcdf&resx=0.500000&resy=0.500000&bbox=-179.500000,-89.500000,180.000000,90.000000

and the data read from that have lon=-179.25 -178.75 ... 179.75 (length 719)

By contrast, the data(topoWorld) as of yesterday has lon=-179.5 -179.0 ... 180 (length 720)

So what gives?? Why doesn't the new one start at -179.5?  Let's try with full limits (call this test2):

https://gis.ngdc.noaa.gov/mapviewer-support/wcs-proxy/wcs.groovy?filename=etopo1.nc&request=getcoverage&version=1.0.0&service=wcs&coverage=etopo1&CRS=EPSG:4326&format=netcdf&resx=0.500000&resy=0.500000&bbox=-180,-90,180,90

```
f <- nc_open("~/Downloads/test2.nc")
vectorShow(ncvar_get(f, "lon"))
[1] "ncvar_get(f, \"lon\"): -180.0, -179.5, ..., 179.5, 180.0 (length 721)\n"
```

So, now try with (call that test3)

https://gis.ngdc.noaa.gov/mapviewer-support/wcs-proxy/wcs.groovy?filename=etopo1.nc&request=getcoverage&version=1.0.0&service=wcs&coverage=etopo1&CRS=EPSG:4326&format=netcdf&resx=0.500000&resy=0.500000&bbox=-179.5,-89.5,180,90


```
f <- nc_open("~/Downloads/test3.nc")
vectorShow(ncvar_get(f, "lon"))
[1] "ncvar_get(f, \"lon\"): -179.5, -179.0, ..., 179.5, 180.0 (length 720)\n"
```

so that at least is going to give us the same lon and lat. What about z, though? The built-in dataset has
```
topoWorld[["z"]][1:3, 1:3]
     [,1] [,2] [,3]
[1,] 2895 2987 3078
[2,] 2895 2987 3078
[3,] 2895 2987 3078
```
and the new one has
```
     [,1] [,2] [,3]
[1,] 2869 2985 3080
[2,] 2868 2983 3080
[3,] 2866 2981 3079
```

so not the same. Oh my. I'll try some broader tests in 1551a.R
I give up.  I tried maybe 20 tweaks (each with a git-push and 5 to 10 minute
wait) to get good breaks in the text.  I will instead now work with the .md
file that is created.  That means changing the workflow.  The old workflow and
old Rmd are here. Note that the latter is *not* the final product.  It is just
here for posterity.
# CRAN 1.3.0 submission: trial 1

**Rejection email follows**

CAUTION: The Sender of this email is not from within Dalhousie.

Dear maintainer,

package oce_1.3-0.tar.gz does not pass the incoming checks automatically, please see the following pre-tests:
Windows: <https://win-builder.r-project.org/incoming_pretest/oce_1.3-0_20210127_175118/Windows/00check.log>
Status: 1 NOTE
Debian: <https://win-builder.r-project.org/incoming_pretest/oce_1.3-0_20210127_175118/Debian/00check.log>
Status: 1 NOTE

Last released version's CRAN status: NOTE: 8, OK: 4
See: <https://CRAN.R-project.org/web/checks/check_results_oce.html>

CRAN Web: <https://cran.r-project.org/package=oce>

Please fix all problems and resubmit a fixed version via the webform.
If you are not sure how to fix the problems shown, please ask for help on the R-package-devel mailing list:
<https://stat.ethz.ch/mailman/listinfo/r-package-devel>
If you are fairly certain the rejection is a false positive, please reply-all to this message and explain.

More details are given in the directory:
<https://win-builder.r-project.org/incoming_pretest/oce_1.3-0_20210127_175118/>
The files will be removed after roughly 7 days.

*** Strong rev. depends ***: dendroTools morphomap seacarb soundecology SWMPr vprr

Best regards,
CRAN teams' auto-check service
Flavor: r-devel-windows-ix86+x86_64
Check: CRAN incoming feasibility, Result: NA
 Maintainer: 'Dan Kelley <Dan.Kelley@Dal.Ca>'

 Found the following (possibly) invalid URLs:
   URL: https://www.whoi.edu/page.do?pid=23096
     From: man/read.ctd.itp.Rd
     Status: Error
     Message: libcurl error code 7:
       	Failed to connect to www.whoi.edu port 443: Timed out

 Size of tarball: 5499705 bytes

Flavor: r-devel-windows-ix86+x86_64
Check: Overall checktime, Result: NOTE
 Overall checktime 18 min > 10 min

Flavor: r-devel-linux-x86_64-debian-gcc
Check: CRAN incoming feasibility, Result: NOTE
 Maintainer: 'Dan Kelley <Dan.Kelley@Dal.Ca>'

 Size of tarball: 5499705 bytes
 
# CRAN 1.3.0 submission: trial 2

**Rejection email follows**

CAUTION: The Sender of this email is not from within Dalhousie.

Dear maintainer,

package oce_1.3-0.tar.gz does not pass the incoming checks automatically, please see the following pre-tests:
Windows: <https://win-builder.r-project.org/incoming_pretest/oce_1.3-0_20210127_211327/Windows/00check.log>
Status: 1 NOTE
Debian: <https://win-builder.r-project.org/incoming_pretest/oce_1.3-0_20210127_211327/Debian/00check.log>
Status: 1 NOTE

Last released version's CRAN status: NOTE: 8, OK: 4
See: <https://CRAN.R-project.org/web/checks/check_results_oce.html>

CRAN Web: <https://cran.r-project.org/package=oce>

Please fix all problems and resubmit a fixed version via the webform.
If you are not sure how to fix the problems shown, please ask for help on the R-package-devel mailing list:
<https://stat.ethz.ch/mailman/listinfo/r-package-devel>
If you are fairly certain the rejection is a false positive, please reply-all to this message and explain.

More details are given in the directory:
<https://win-builder.r-project.org/incoming_pretest/oce_1.3-0_20210127_211327/>
The files will be removed after roughly 7 days.

*** Strong rev. depends ***: dendroTools morphomap seacarb soundecology SWMPr vprr

Best regards,
CRAN teams' auto-check service
Flavor: r-devel-windows-ix86+x86_64
Check: CRAN incoming feasibility, Result: NA
 Maintainer: 'Dan Kelley <Dan.Kelley@Dal.Ca>'

 Size of tarball: 5500268 bytes

Flavor: r-devel-windows-ix86+x86_64
Check: Overall checktime, Result: NOTE
 Overall checktime 16 min > 10 min

Flavor: r-devel-linux-x86_64-debian-gcc
Check: CRAN incoming feasibility, Result: NOTE
 Maintainer: 'Dan Kelley <Dan.Kelley@Dal.Ca>'

 Size of tarball: 5500268 bytes
 

Thoughts on submission.

Maybe the size is just too large.  Below are some thoughts on that.

# 1.2.0 on CRAN
```
   sub-directories of 1Mb or more:
      R      2.4Mb
      doc    2.3Mb
      help   2.7Mb
      libs   4.1Mb
```


# 1.3.0 trial 1 submitted 2020-01-27

tarball size 5.2M
```
$ ls -lh oce_1.3-0.tar.gz
-rw-r--r--  1 kelley  staff   5.2M 27 Jan 19:14 oce_1.3-0.tar.gz
```

# 1.3.0 trial 2 submitted 2020-01-27

I didn't record the tarball size.

This has (CRAN build)

# 1.3.0 trial 3A (not submitted) with figures moved out of package

I can drop the tarball size to 4.5M (is this under a limit??) by moving those
figures elsewhere.
```
ls -lh oce_1.3-0.tar.gz
-rw-r--r--  1 kelley  staff   4.5M 27 Jan 19:20 oce_1.3-0.tar.gz
```

This has (local build)
```
N  checking installed package size ...
     installed size is 10.6Mb
     sub-directories of 1Mb or more:
       R      3.0Mb
       data   1.2Mb
       doc    2.3Mb
       help   3.0Mb
```
NOTE: I may be able to shrink the data to under 1Mb, getting it below the NOTE
threshold.


# 1.3.0 trial 3B (not submitted) with topoNS.rda removed and AMSR.rda trimmed

Since the main purpose is for testing the code, I don't worry too much about
going for a smaller area.  Doing that, I get tarball size
```
ls -lh ../oce_1.3-0.tar.gz
-rw-r--r--  1 kelley  staff   4.3M 27 Jan 20:17 ../oce_1.3-0.tar.gz
```
and the report during local checking is
```
     installed size is 10.4Mb
     sub-directories of 1Mb or more:
       R      3.0Mb
       data   1.0Mb
       doc    2.3Mb
       help   3.0Mb
```
so not quite below the limit.  `amsr.rda` is only 9.8Kb now, so further trimming doesn't
seem worthwhile.

The top 4 files now are
* topoWorld.rda 349KB
* coastlineWorld.rda 106 KB
* landsat.rda 100 KB
and the first two cannot really be trimmed easily.  That leaves landsat.rda;
I'll try halving its size.

# 1.3.0 trial 3C (not submitted) with vignette image size reduced

Previously, the figure size was set to 5in by 5in.  Switching to 4 by 4 trims
the source tarball to 4.1M (cf previous value 4.3M).  The details with 4 by 4
are:
```
N  checking installed package size ...
     installed size is 10.0Mb
     sub-directories of 1Mb or more:
       R      3.0Mb
       data   1.0Mb
       doc    2.1Mb
       help   2.8Mb
```


# Notes on CRAN 1.3.0 submission

* [trial 1](trial_1.md)
* [trial 2](trial_2.md)
* [trial 3](trial_3.md)
* 
Two things are being done in this directory (see
https://github.com/dankelley/oce/discussions/1756 for discussion).

1. Create a high-resolution dataset for the Nova Scotia region.
2. Solve a problem with plotting topo with GEBCO colours.  Originally, before
   2020-12-14, shallow waters were being plotted with a reddish (land-like)
   hue.  This was basically because the palette is constructed by blending
   colours, and the water colours are bluish, while the land ones are reddish,
   but mixing blue and red does not yield white.  I solved this by adding a
   white band, although I found that band to be too wide in tests, so I first
   expanded the 9-element water and land bands to be longer.  See the above-named website for details.

NOTE:  the new datafile, called `topoNS.rda` for now, is 87K (was 163K before
compression).

The files test schemes for subsetting coastline files.

# 01.R

Delete 'island' portions that have no intersection with a chosen polygon.

# 02.R

First test with cutting polygons on a vertical line (here, to the east of a
point clicked on the plot). This is a slow operation, but (a) a person will
only need to use it in constructing coastline subsets to speed up individual
projects, so a 1-minute operation is not a real pain and (b) it is purposely
written in an indexing/looping style, so it can be made into C easily.

# 03.R

Trim four sides (but wrong on step 3)

# 04.R

Still broken (in a new way)

# 05.R

Trial solution using raster package functions.

# Implementing the "RBR" profile detection method

* `detectprofiles.m` the matlab code from RSK tools that implements the algorithm
The `*.R` files are some tests done to try out ideas for labeling lines drawn
by `mapContour()`. There is no intention to reproduce the full features of the
basic R function `contour()`, but the hope is to provide at least crude
functionality.

If things seem promising, they may be incorporated into `R/map.R`, perhaps with
the detailed work being done in a C function, for speed.

The numbers in the filenames show the evolution of my tests.

-- Dan Kelley, Feb 2019.


This information comes from a python script based on code provided at https://stackoverflow.com/questions/47019616/gdal-doesnt-recognize-bipolar-conic-western-hemisphere-projection

162 total projections, 46 supported, 116 not supported

Supported: aea, aeqd, bonne, cass, cea, eck1, eck2, eck3, eck4, eck5, eck6, eqc, eqdc, etmerc, gall, geos, gnom, goode, gstmerc, igh, krovak, laea, lcc, merc, mill, moll, nzmg, omerc, ortho, poly, robin, sinu, somerc, stere, sterea, tmerc, tpeqd, utm, vandg, wag1, wag2, wag3, wag4, wag5, wag6, wag7

Not supported: affine, airy, aitoff, alsk, apian, august, axisswap, bacon, bertin1953, bipc, boggs, calcofi, cart, cc, ccon, chamb, collg, comill, crast, deformation, denoy, eqearth, euler, fahey, fouc, fouc_s, geoc, geogoffset, gins8, gn_sinu, gs48, gs50, hammer, hatano, healpix, helmert, hgridshift, horner, imw_p, isea, kav5, kav7, labrd, lagrng, larr, lask, latlon, lcca, leac, lee_os, lonlat, loxim, lsat, mbt_fps, mbt_s, mbtfpp, mbtfpq, mbtfps, mil_os, misrsom, molobadekas, molodensky, murd1, murd2, murd3, natearth, natearth2, nell, nell_h, nicol, noop, nsper, ob_tran, ocea, oea, ortel, patterson, pconic, pipeline, pop, push, putp1, putp2, putp3, putp3p, putp4p, putp5, putp5p, putp6, putp6p, qsc, qua_aut, rhealpix, rouss, rpoly, sch, tcc, tcea, times, tissot, tobmerc, tpers, unitconvert, ups, urm5, urmfps, vandg2, vandg3, vandg4, vgridshift, vitk1, webmerc, weren, wink1, wink2, wintri
# Instructions

1. All datasets should be created here. 
2. Subdirectories should be named to match those datasets.
3. Add an item for the new subdirectory in the first line of the Makefile in the present directory.
4. Copy an existing `Makefile` to newly-added directories, changing the first line as appropriate.
5. Add any new dataset to the various `check*.R` files found here. These check things like
  a. `check_units.R` -- ensure the existence of `units` within `metadata` (becomes moot once it works, late 2015)
  b. `check_summary.R` -- check output of `summary()` is good on all datasets (e.g. are units good?)


Manual steps:
 1. visit http://www.naturalearthdata.com/downloads/110m-cultural-vectors
 2. click the link named 'Download countries' in the section called 'Admin 0 - Countries'
 3. type 'make clean ; make' here
 4. if you like the results, type 'make install' here

Running `make` here creates `tideCurrent.rda` here, and running `make install`
copies it to `../data`.  The work is done with `create_tideCurrent.R`, and the
data files `tide8.dat` and `tide9.dat` are from a zipfile [2] downloaded on
2020 aug 5.  It should be noted that [2] also contains the full text of [1].

The data format is format is described on page 5 (PDF page 12) of [1], and
`create_tidalCurrent.R` uses this as the first step in reading the data, after
which it does some housekeeping tasks to put the data into a usable format as
described in the documentation provided by `?tidalCurrent`.

## References

1. Foreman, M. G. G. ‚ÄúManual for Tidal Currents Analysis and Prediction.‚Äù
   Pacific Marine Science Report. British Columbia, Canada: Institute of Ocean
   Sciences, Patricia Bay, 1978.
2. https://www.dfo-mpo.gc.ca/science/documents/data-donnees/tidal-marees/tidpack.zip

It is difficult to download these data since MEDS seems to supply only a GUI system.

At one time, this worked:

http://www.meds-sdmm.dfo-mpo.gc.ca/isdm-gdsi/twl-mne/inventory-inventaire/download-telecharger.asp?File=E:\iusr_tmpfiles\TWL\490-1-JAN-2003_slev.csv\&Name=490-1-JAN-2003_slev.csv\&Size=159411

but it failed when I tried it on 2018-03-13. I can download the data by
clicking the link, but if I save that link in Safari I get

http://www.isdm-gdsi.gc.ca/isdm-gdsi/twl-mne/inventory-inventaire/download-telecharger.asp?File=E:\iusr_tmpfiles\TWL\490-01-JAN-2003_slev.csv&Name=490-01-JAN-2003_slev.csv&Size=159411

which doesn't seem to be a link I can use permanently. Thanks, MEDS, way to go, buddy.

# Data source

ftp ftp://ftp.ifremer.fr/ifremer/argo/dac/bodc/6900388/6900388_prof.nc

downloaded 2017 July 7 0908h

# Standardized citation

Argo (2017). Argo float data and metadata from Global Data Assembly Centre
(Argo GDAC) - Snapshot of Argo GDAC of July, 8st 2017. SEANOE.
http://doi.org/10.17882/42182#50865


This file is not installed within oce, but the method of generating it is shown
in the docs for read.odf(). The reason for not including the file is basically
that read.odf() is a stop-gap measure provided to help people whilst we wait
for BIO to release an ODF package.
These files were downloaded 2015-09-29 by Dan Kelley from the ``cmocean/rgb``
directory of the [cmocean](https://github.com/kthyng/cmocean) python
application, and are used in oce (to be distributed under GPL>=2) according to
permission granted by the ``cmocean`` author, Kristen Thyng, as stated in an
email sent to Dan Kelley and Clark Richards September 29, 2015.
This directory is used by the developers to do perform tests that are in
addition to those in the `tests` directory.  These additional tests rely on
private files that are not included in the CRAN version of oce, owing to
reasons of file size and privacy.

Some of the data files used here are stored in a Dropbox folder shared by
developers DK and CR, and others are stored in local files. The scheme is to
loop over the results of `list.files()`, and that means that the tests should
still work even files are missing.

In some cases, e.g. for adp/RDI files, the tests merely see whether data can be
read without error. In other cases, `summary` and `plot` are used, on the
expectation that the person running tests will scan the files visually to look
for problems (e.g. missing units for some SBE data types).

Files:

* README.md this file
* Makefile use this to run tests
* adp_rdi_01.R tests some adp datasets of the RDI variety
* ctd_beaufort_sea.R tests CTD data from Beaufort Sea (private data in DK possession)
* ctd_flemish_cap.R tests some CTD data from Flemish Cap (private data in DK possession)
* ctd_sleiwex_coriolissII.R tests some CTD data from SLEIWEX experiment (private data in DK and CR possession)
* odf_01.R tests various ODF datasets (private data in DK possession)

The .gitignore in this directory is set up to not track any files (except this
README.md file). This means that the developers can put any data files or .R
files here that they like, without compromising data privacy.
---
title: "oce: an R package for Oceanographic Analysis"
tags:
  - R
  - oceanography
  - measurement
  - data analysis
authors:
  - name: Dan E. Kelley^[Corresponding author.]
    orcid: 0000-0001-7808-5911
    affiliation: 1
  - name: Clark Richards
    orcid: 0000-0002-7833-206X
    affiliation: 2
  - name: Chantelle Layton
    orcid: 0000-0002-3199-5763
    affiliation: 3
affiliations:
 - name: Dan E. Kelley, Professor, Dalhousie University
   index: 1
 - name: Clark Richards, Research Scientist, Bedford Institute of Oceanography, Department of Fisheries and Oceans, Canada; also Adjunct Professor, Dalhousie University
   index: 2
 - name: Chantelle Layton, Physical Scientist, Bedford Institute of Oceanography, Department of Fisheries and Oceans, Canada
   index: 3
date: 26 May 2021
year: 2021
bibliography: paper.bib
output: rticles::joss_article
csl: apa.csl
journal: JOSS
---

# Statement of Need

Oceanographic field experiments often employ a suite of instrument types, each
reporting data in a different format. Many of these formats are complex and
difficult to decode. Manufacturers usually provide software for accessing data
produced by their instruments, but it is usually proprietary and closed-source,
making it difficult for researchers to analyse their data in novel ways or to
combine data from multiple instruments. The `oce` package
[@kelley__aut_oce_2021] addresses such issues in the R language^[The authors are not aware of similarly comprehensive tools in other open-source languages, although there are tools for individual tasks, as noted at https://pyoceans.github.io/sea-py/, for Python tools.] with functions that handle dozens
of data formats. It also has facilities for the specialized calculations and
data displays that are particular to oceanography.  Since `oce` is written in
the R language [@ihaka_r_1996;@r_core_team_introduction_2021], it forms a link
to a vast array of general tools that oceanographers use in their work
[@kelley_oceanographic_2018].

# Overview

The `oce` package has been hosted since 2009 on
CRAN^[CRAN is at https://cran.r-project.org.], the Comprehensive R Archive Network
[@hornik_comprehensive_2012].  The CRAN version, which is updated once or twice
a year, may be installed by typing `install.packages("oce")` in an R console.
Users who need newer features may use
`remotes::install_github("dankelley/oce",ref="develop")` to download and build
the development branch^[Since some components of oce are written in lower-level languages such as Fortran, C and C++, building the package from source requires compilers for those languages. These are freely available, and an oce wiki page (https://github.com/dankelley/oce/wiki/Configuring-a-Linux-Mint-20-system-to-be-able-to-build-oce-from-source) explains how they and other things may be set up, on a variety of computer types.]. Those wishing to view or participate in the development
process are welcome to do so, at \url{https://github.com/dankelley/oce}.

The package has functions for decoding many data formats.  These functions
return S4 objects with slots holding (a) the data, (b) related metadata, and
(c) a log of `oce` functions that made the object.  This is illustrated by
executing the following in an R session, for a built-in object creating by
reading a profiling instrument called a CTD.


```r
library(oce)                           # load library
data(ctd)                              # load a built-in sample file
slotNames(ctd)                         # see 'slot' names
```

![CTD analysis example. Top left: depth variation of temperature and salinity, with pressure as a surrogate for depth (1 decibar corresponds to approximately 1 metre of depth). Top right: depth variation of a potential density anomaly and the square of the buoyancy frequency, two measures indicating the gravitational stability of the water column. Bottom left: co-variation of salinity and temperature, with contours of potential data anomaly. Bottom-right: map of the study region, gray indicating land.](figure_1.png)

In practice, `slotNames()` is seldom used. Rather, the second step after
loading an object, or reading it from a data file, is often to get a textual
overview with `summary(ctd)`, or a graphical overview. For example,
```r
plot(ctd)
```
produces Figure 1, which shows the basic information needed in oceanographic
studies.

It is also common to exert fine-grained control of graphical representations,
e.g.
```r
plot(ctd, which="temperature")         # results not shown
```
plots just the temperature variation with depth. The variations of other
properties may be shown by setting `which` appropriately, and this argument may
also be used to specify other types of plots, in addition to the
depth-variation form.

Besides this `"ctd"` subclass, `oce` supports dozens of other subclasses that
cover a wide range of oceanographic instrumentation.  In every case, the same
`summary()` and `plot()` function calls provide textual and graphical
representations of the data.  This specialization of these two generic
functions simplifies analysis considerably.  For example, if `PATTERN` is a
regular expression that specifies a set of data files, whether of a single
instrument type or multiple instrument types, then

```r
for (file in list.files(PATTERN)) {
    d <- read.oce(file)
    summary(d)
    plot(d)
}
```
will provide information about each data file of interest, forming a good
first stage of analysis.

Oce also provides other generic functions, including `subset()` for focusing on
subsets of data, `handleFlags()` for processing data-quality flags, and `[[`
for accessing data.  The last of these is particularly worthy of note, for two
reasons.

1. `[[` finds information regardless of where it is stored in the object. For
   example, a CTD does not measure longitude and latitude, but if these things
are known, they will be stored in the `metadata` slot, not the `data` slot.
Other objects might have longitude in the `data` slot. This detail is
immaterial to users, because `[[` looks in both slots.  Therefore, code written
for one object type will often work for another type.

2. `[[` can access not just information stored within the object, but also
   things that can be calculated from that information. For example, CTD files
typically hold information from which seawater density may be computed
[@millero_history_2010;@mcdougall_getting_2011], and so `[[` is set up to
compute it, with e.g. `ctd[["density"]]`. This same scheme works for other
computable elements.

The `[[` function acts as a sort of bridge from the oceanographic realm to the
general R realm, with its thousands of useful and well-vetted packages.  This
is one of the ways in which `oce` reduces the need to create new tools, letting
analysts focus on oceanography, not coding.

# Example: Tidal Analysis

A more detailed example may help to solidify some of the key aspects of `oce`.
Many readers will have an interest in tides, so we will work with a record of
sea level in Halifax Harbour.

Consider the code given below, which produces Figure 2.  A built-in sealevel
file is used, to make a reproducible example, but replacing the `data()` call
with a `read.sealevel()` call will handle data files in standard formats.  Note
that the `tidem()` function is fairly sophisticated, with over 500 lines of R
code being used to apply the specialized procedures of tidal analysis
[@godin_analysis_1972;@pawlowicz_classical_2002;@foreman_versatile_2009].

```r
library(oce)                           # load library
data(sealevel)                         # use built-in example dataset
t <- sealevel[["time"]]                # extract time
eta <- sealevel[["elevation"]]         # extract sea level
m <- tidem(sealevel)                   # fit tidal model
etaDetided <- eta - predict(m)         # de-tide observations
par(mfrow=c(2, 1))                     # set up a two-panel plot
oce.plot.ts(t, eta, xaxs="i",          # top: observed sea level
    grid=TRUE, ylab="Sea level [m]")
oce.plot.ts(t, etaDetided, xaxs="i",   # bottom: de-tided sea level
    grid=TRUE, ylab="De-tided sea level [m]")
```

![Tidal analysis example. Top: sea level variation in Halifax Harbour during the year 2003. Bottom: sea level after removing a fit to the tides.](figure_2.png)


A comparison of the panels in Figure 2 reveals that tides explain much of the
sea level variation in Halifax Harbour.  The lower panel illustrates an
increase of detided variance during the winter months, as expected at a
northern mid-latitude.  More surprising is the large spike towards the end of
September.  This is a result of Hurricane Juan, which swept over Halifax at
that time, causing a storm surge of approximately 1.5m that, along with high
waves, caused major damage in the harbour [@xu_extreme_2012].  (Readers might
find it informative to supply an `xlim` argument to the plot calls, to narrow
in on the event and ponder the significance of its timing with respect to the
tide.)

# Conclusions

The `oce` package provides for many aspects of oceanographic analysis, having
evolved in an open-source environment for more than a decade. For much of that
time, the hosting has been on github.com/dankelley/oce, where users can see the
details of over 9000 git commits and 1700 closed issues.  The developers have
benefited from a supportive user community, members of which have contributed
insightful bug reports and suggestions for improvements.  New features are
added continually, to handle new instrument types, new data repositories, and
new methods. Physical oceanography is a major focus of the package, but we hope
this paper will generate interest in other communities, ranging from
climatologists to those in marine disciplines such as chemistry and biology.
Our other goal is to encourage the development of new R packages, such as
`argoFloats` [@kelley_argofloats_2021], that build upon `oce`.

# Acknowledgments

We thank the many users who have helped with `oce` development over the years,
by finding bugs, suggesting new features, and providing us with manufacturers'
documentation about data formats for various oceanographic instruments. In
addition, we thank the editor and reviewers of this manuscript, for their
insights and also for their patience during the submission process.

# References

---
title: "oce: an R package for Oceanographic Analysis"
tags:
  - R
  - oceanography
  - measurement
  - data analysis
authors:
  - name: Dan E. Kelley^[Corresponding author.]
    orcid: 0000-0001-7808-5911
    affiliation: 1
  - name: Clark Richards
    orcid: 0000-0002-7833-206X
    affiliation: 2
  - name: Chantelle Layton
    orcid: 0000-0002-3199-5763
    affiliation: 3
affiliations:
 - name: Dan E. Kelley, Professor, Dalhousie University
   index: 1
 - name: Clark Richards, Research Scientist, Bedford Institute of Oceanography, Department of Fisheries and Oceans, Canada; also Adjunct Professor, Dalhousie University
   index: 2
 - name: Chantelle Layton, Physical Scientist, Bedford Institute of Oceanography, Department of Fisheries and Oceans, Canada
   index: 3
date: 26 May 2021
year: 2021
bibliography: paper.bib
output: rticles::joss_article
csl: apa.csl
journal: JOSS
---

# Statement of Need

Oceanographic field experiments often employ a suite of instrument types, each
reporting data in a different format. Many of these formats are complex and
difficult to decode. Although manufacturers usually provide software for
accessing data produced by their instruments, it tends to be proprietary and
closed-source. This is a problem for researchers seeking to analyse their data
in novel ways, or combine data from multiple instruments. The `oce` package
[@kelley__aut_oce_2021] addresses such issues by providing functions that
handle dozens of data formats. In addition, it facilitates specialized
calculations and data displays that are particular to the discipline.  Since
`oce` is written in the R language
[@ihaka_r_1996;@r_core_team_introduction_2021], it forms a link to an array of
more general tools that oceanographers may need in their work
[@kelley_oceanographic_2018].

# Overview

The `oce` package has been hosted on CRAN [@noauthor_comprehensive_2021] since
the year 2009. The CRAN version, which is updated once or twice a year, may be
installed by typing `install.packages("oce")` in an R console. Users who need
newer features may use `remotes::install_github("dankelley/oce",ref="develop")`
to download and build the development branch. Those wishing to view or
participate in the development process are welcome to do so, at
\url{https://github.com/dankelley/oce}.

The package has functions for decoding many data formats.  These functions
return S4 objects with slots holding (a) the data, (b) related metadata, and
(c) a log of `oce` functions that made the object.  This is illustrated by
executing the following in an R session, for a built-in object creating by
reading a profiling instrument called a CTD.
```{r setup, include=FALSE}
library(oce)
options(width=70)
knitr::opts_chunk$set(fig.path="", dev="png", dpi=300, pointsize=8)
data(ctd)
```
```{r eval=FALSE}
library(oce)                           # load library
data(ctd)                              # load a built-in sample file
slotNames(ctd)                         # see 'slot' names
```

The next step after loading an object, or reading it from a data file, is often
to get a textual overview with `summary(ctd)`, or a graphical overview, e.g.
with `plot(ctd)` producing Figure 1.  It is also common to exert fine-grained
control of graphical representations, with e.g.  `plot(ctd,
which="temperature")` to plot just the temperature variation with depth
(results not shown here). The variations of other properties may be shown by
setting `which` appropriately, and this argument can also be used to specify
other types of plots, in addition to the depth-variation form.

```{r echo=FALSE,results="hide", fig.cap="CTD analysis example. Top left: depth variation of temperature and salinity. This uses pressure as a surrogate for depth, with 1 decibar corresponding to approximately 1 metre of depth. Top right: variation of density excess over 1000 kg/m$^3$, and $N^2$, a measure of the gravitational stability of the water column. Bottom left: co-variation of salinity and temperature. Bottom-right: map of the study region, with gray being land in a coarse representation of the coastline.", dev.args=list(pointsize=10)}
plot(ctd)
```

Besides this `"ctd"` subclass, `oce` supports dozens of other subclasses that
cover a wide range of oceanographic instrumentation.  In every case, the same
`"summary()"` and `"plot()"` function calls provide textual and graphical
representations of the data.  This specialization of these two generic
functions simplifies analysis considerably.  For example, if `PATTERN` is a
regular expression that specifies a set of data files, whether of a single
instrument type or multiple instrument types, then
```{r eval=FALSE}
for (file in list.files(PATTERN)) {
    d <- read.oce(file)
    summary(d)
    plot(d)
}
```
will provide information about each data file of interest, forming a good
first stage of analysis.

Oce also provides other generic functions, including `subset()` for focusing on
subsets of data, `handleFlags()` for processing data-quality flags, and `[[`
for accessing data.  The last of these is particularly worthy of note, for two
reasons.

1. `[[` finds information regardless of where it is stored in the object. For
   example, a CTD does not measure longitude and latitude, but if these things
   are known, they are stored in the `metadata` slot, not the `data` slot.
   Other objects might have longitude in the `data` slot. This detail is
   immaterial to users, because `[[` looks in both slots.  Therefore, code
   written for one object type will often work for another type.

2. `[[` can access not just information stored within the object, but also
   things that can be calculated from that information. For example, CTD files
   typically hold information which seawater density may be computed
   [@millero_history_2010;@mcdougall_getting_2011], and so `[[` is set up to
   compute it, if requested. This same scheme works for other computable
   elements.

The `[[` function acts as a sort of bridge from the oceanographic realm to the
general R realm, with its thousands of useful and well-vetted packages.  This
reduces the need to create new tools, letting analysts focus on oceanography,
not coding.

# Example: Tidal Analysis

A more detailed example may help to solidify some of the key aspects of `oce`.
Many readers will have an interest in tides, so we will work with a year-long
record of sea level, $\eta=\eta(t)$ in Halifax Harbour, in the year 2003,
during which the city was struck by Hurricane Juan.

Consider the code given below, which produces Figure 2.  A built-in sealevel
file is used, to make a reproducible example, but replacing the `data()` call
with a `read.sealevel()` call will handle data files in standard formats.  Note
that the `tidem()` function is fairly sophisticated with over 500 lines of R
code being used to apply the specialized procedures of tidal analysis
[@godin_analysis_1972;@pawlowicz_classical_2002;@foreman_versatile_2009].
Readers who see that the function evokes the `lm()` function for linear models,
may not be surprised that `oce` provides a function named `predict()`, for
generating tidal predictions.

```{r results="hide", fig.cap="Tidal analysis example. Top: a year of sea level variation in Halifax Harbour. Bottom: sea level after removing a fit to the tides.", dev.args=list(pointsize=10)}
library(oce)                           # load library
data(sealevel)                         # use built-in example dataset
t <- sealevel[["time"]]                # extract time
eta <- sealevel[["elevation"]]         # extract sea level
m <- tidem(sealevel)                   # fit tidal model
etaDetided <- eta - predict(m)         # de-tide observations
par(mfrow=c(2, 1))                     # set up a two-panel plot
oce.plot.ts(t, eta, xaxs="i",          # top: observed sea level
    grid=TRUE, ylab="Sea level [m]")
oce.plot.ts(t, etaDetided, xaxs="i",   # bottom: de-tided sea level
    grid=TRUE, ylab="De-tided sea level [m]")
```

A comparison of the panels in Figure 2 reveals that tides explain much of the
sea level variation in Halifax Harbour.  The lower panel illustrates an
increase of detided variance during the winter months, as expected at a
northern mid-latitude.  More surprising is the large spike towards the end of
September.  This is a result of Hurricane Juan, which swept over Halifax at
that time, causing a storm surge of approximately 1.5m that, along with high
waves, caused major damage in the harbour [@xu_extreme_2012].  (Readers might
find it informative to supply an `xlim` argument to the plot calls, to narrow
in on the event.)

# Conclusions

The `oce` package provides for many aspects of oceanographic analysis, having
evolved in an open-source environment for more than a decade.  The developers
have benefited from a supportive user community, members of which have
contributed insightful bug reports and suggestions for improvements.  New
features are added continually, to handle new instrument types, new data
repositories, and new methods. Physical oceanography is a major focus of the
package, but we hope this paper will generate interest in other communities,
ranging from climatologists to those in marine disciplines such as chemistry
and biology.  Our other goal is to encourage the development of new R packages,
such as `argoFloats` [@kelley_argofloats_2021], that build upon `oce`.

# References

---
title: An idea for coastline clipping
author: Dan Kelley
date: 2019-02-22
---

**Abstract.** CL and I have been talking about subsetting coastline files to
those segments that intersect a plotting zone. (R does this automatically for
xy plots, but for mapPlot(), the lon-lat to x-y step that preceeds that is a
waste of time.) The reason is basically speed: especially for shiny apps, it
can be annoying to wait for a long time for a world coastline to be drawn over
a region that is only a few 100km wide. This document proposes a scheme -- I am
looking for feedback on whether I'm mixed up.

It is quite simple to ignore regions that lie entirely outside a plotting zone,
as illustrated by oce/sandbox/coastlineSubset  in the ``develop'' branch of
oce.

Regions that partly intersect a plotting region are trickier to solve, because
quite a lot of time can be wasted on points that go far outside the box.  For
example, a box that covers the Bay of Fundy and surrounding coastline might
bring in a long trace for the main body of continental Canada, and another one
for the USA. It would be sensible to boil all those points down to just a few
that would be sufficient to make the new coastline run *just outside* the plot
box, so that either a line or a polygon fill would represent the coastline,
seen through the plot window.

A possible method is as follows. Consider the diagram given below. I didn't
label the subplots, but call them Figures 1A, 1B, 1C and 1D, in reading order.
In each, the curvy line is intended to be a coastline. Suppose that this is
using a convention that the coastline goes clockwise around the land. (All of
what I say can be reversed for the opposite convention. Note that I am not --
yet -- thinking about lakes within land!)

Consider Fig 1A. In my proposed method, we trace along from i=1, and find at
the point labelled (1) (well, circled 1, actually) that the line has crossed
into the box. We keep following the line until it exits the box. Let's denote
that exit point as (2). I propose we walk along along the plot-region sides,
starting at (2) until we get back to (1). Now, imagine taking all the coastline
points that lie to the left of box side 2 and below side 3 as being at the
proper ``y'' but with an ``x'' that is just to the left of side 2. (I am using
R notation to name sides.) Similarly, we move the points that are above the top
side (side 3) to have the correct ``x'' and a ``y'' that is just above side 3.
The points that are in the corner region at top-left can be remapped to a point
outside the top-left corner.) We continue this, down along side 4 (the
right-hand side). We have warped the coastline to run just outside the box, in
a *straight* line. Now, we don't actually need all those points on the line. We
can put a single point to the left of (2), and connect it to this
outside-corner point at top left, then to a similar corner at top right, and
finally to a point just to the right of (1). We can thus remove possibly
thousands of points, replacing them with 3 points.

I think that will work on other sides as well -- it's just a matter of keeping
track.


Now move to Fig 1B. This is like Fig 1A except that we have a second entry at
point (3). But I think a similar scheme of outside points will work.

The other diagrams are a bit more complicated. But I think a general scheme can
be devised that will handle all such cases. The basic idea is to look for entry
and leaving points, and to create fake outside points as needed. I suppose a
simple way is to do literally what I said above: create all those outside
points, and then find sequences points that are along horizontal and vertial
lines, and boil them down to just two points each. (I do a scheme like that in
the ``gri'' language, because the output from gri is postscript, and when
postscript was invented, there was a limit of 1500 points in a connected line!)

This would be way easier to talk about in person, of course.

**Question** does this scheme seem possible, in principle?

![a](coastline_clip_fig1.jpg)

---
title: "3. Analysis of acoustic-Doppler data"
author: "Dan Kelley (https://orcid.org/0000-0001-7808-5911)"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_vignette:
    toc: true
    number_sections: true
    fig_caption: yes
    fig_width: 3.5
    fig_height: 3.5
    dpi: 72
    dev.args: list(pointsize=11)
vignette: >
  %\VignetteIndexEntry{3. Analysis of acoustic-Doppler data}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

<!-- HOW TO BUILD THE VIGNETTE. -->
<!-- Look in oce.Rmd for instructions. -->

```{r, echo = FALSE}
knitr::opts_chunk$set(collapse = TRUE, comment = "#>")
```

**Abstract.** This vignette explains the basics of using the oce package to
handle measurements made by both mooring and ship based acoustic Doppler
profilers.  As in the related vignette about working with [data-quality
flags](E_flags.html), the material is organized to reflect typical workflow.  See
[the main oce vignette](A_oce.html) for general information about oce.  The
sample code provided here relies on the oce package, loaded as follows.
```{r}
library(oce)
```

# Reading adp files

Acoustic Doppler profiling instruments are called ADCPs by some manufacturers
and ADPs by others.  The class name `adp` is used in oce, to handle both types.
It is important to realize that there are *many* data formats for data of this
type, and that the files are normally in a dense binary format that cannot be
read with a text editor. It is necessary to use software to work with such
files. Although oce can handle many file formats, and although its authors do
their best to keep up with changes to file formats (at least for popular
instruments), users will sometimes be forced to use manufacturer software to
read the data, or at least convert to a format that can be manually read into
R, and then converted to an `adp` object using the `as.adp()` function.

Suppose `f` is a character string naming a data file. In many cases,
```{r eval=FALSE}
d <- read.oce(f)
```
will read the file, because `read.oce` will look inside the file to try to
guess the format. If this does not work, one ought to try e.g.
```{r eval=FALSE}
d <- read.adp(f)
```
which, again, will try to determine which subtype of adp file has been
provided.  If that fails, it might help to try a specialized function, such as
`read.adp.rdi()` for an RDI file, `read.adp.nortek()` for a Nortek file, or
`read.adp.sontek()` for a Sontek file. (Actually, each of these can handle a
variety of sub-formats).

If any of these `read.` functions is called with just a single argument naming
the file, then the entire dataset will be read in. This can be slow for large
files (e.g. taking of order 0.1s per Mb), and so it can help to use the `by`
argument, e.g. for a 138Mb file on the author's computer,
```{r eval=FALSE}
f <- "/data/archive/sleiwex/2008/moorings/m09/adp/rdi_2615/raw/adp_rdi_2615.000"
dall  <- read.oce(f)
```
takes 15.2s of user time, but
```{r eval=FALSE}
d100 <- read.oce(f, by=100)
```
takes 0.2s. It is also possible to window the data, using the `from` and `to`
arguments, which (like `by`) can be integers that count ensembles (or "profiles")
or a POSIX time (see `?read.adp`). For example, the
adp dataset provided by oce was read with
```{r eval=FALSE}
read.oce(f, from=as.POSIXct("2008-06-26", tz="UTC")`,
            to=as.POSIXct("2008-06-27", tz="UTC"),
            by="60:00"
            latitude=47.88126, longitude=-69.73433)
```
with `f` as defined above. This also illustrates that it is possibly to specify
the location of an instrument; the further processing that was done to create
`data(adp)` will be explained in due course.

+**Note**: Users who are not familiar with ADP data may noticed that there is a
combination of many file types when dealing with ADP data. Some of these
include: .ENR, which are raw data only received from the ADP, .ENS, similar to
.ENR but also contain data merged from other file types without processing,
.ENX, merged data that are processed including calculations to transformed data
including east, north, and vertical velocities, .STA, which are a short term
average of measurements, and .LTA, which are a long term average of
measurements.


# Summarizing adp data

The generic function `summary` provides a useful overview of an adp data
object.  For example,
```{r results="hide"}
data(adp)
summary(adp)
```
(results not shown \ldots the goal is to encourage the user to try this!)
shows the sort of information normally present in an RDI file. Other
manufacturer types will have somewhat different summaries, because both the
metadata and the data differ from instrument to instrument. In all cases, there
will be a vector holding time (named `time`), another holding the distance from
instrument to the centre of a measurement bin, measured along a line
equidistant between the beams (named `distance`), along with an array
holding velocity (named `v`). (Other arrays may include, as in this RDI case,
`q` for data quality, `a` for backscatter amplitude, and `g` for percentage
goodness.)

Note that the summary reveals that instrument had been set up to record in
`beam` coordinates, but that processing had been done to convert to `enu`
(east-north-up) coordinates.

**Exercise 1.** Determine the name of the `metadata` item that holds the
coordinate system of an adp object.

After completing Exercise 1, the user will notice that both `latitude` and
`longitude` are available in the metadata for this built-in, moored, adp
dataset. It should be noted that if this dataset was a ship-based adp dataset,
then `latitude` and `longitude` would not be included. Instead, the data would
include `firstLatitude`, `firstLongitude`, `lastLatitude`, and `lastLongitude`,
which provide an indication of the ship drift during sampling.

# Changing coordinate systems

It is critical for analysts to be aware of the coordinate system of an adp
dataset, simply because those systems are so different. In what oce calls the
`beam` system, velocities in each of the acoustic beams are measured along
the slant angle of the beam. These data may be combined (using a
transformation matrix which is part of the metadata displayed with the
`summary()` function) into what oce calls a `xyz` system that is Cartesian and
fixed to the instrument. Since instruments record heading, pitch, and roll, the
`xyz` velocities can be converted easily to `enu` velocities.

In some applications, the scientific focus is primarily on the enu data, but it
is a good idea to calculate and study the data at all stages of processing.
Thus, although the `toEnu()` function will convert data measured in any
coordinate system into an enu variant, many analysts will do something along
the lines of
```{r eval=FALSE}
beam <- read.oce(f)
xyx <- beamToXyz(beam)
enu <- xyzToEnu(xyz, declination=-18.1)
```

where the declination illustrated is the value used in creating the `data(adp)`
dataset. (Compensating for declination is necessary because compasses are
calibrated at a particular place and time, and compasses have no way of
"knowing" the local angle between true north and magnetic north.)

# Plotting

The generic `plot()` function is specialized to handle adp objects in a variety
of simple but helpful ways. The type of plot is determined by the `which`
argument. The default for `which` is to create a multi-panel image plot that
shows time-space dependence for each velocity component. The number of panels
depends on the number of beams used by the instrument, and each one has a label
indicating the component name. For example,
```{r fig.height=7}
plot(adp)
```

has velocity panels named `east`, `north` and `up`, along with one
named `error`, which is an estimate of the velocity-inference error.

Options to this plot variant can be used to control how distance from
instrument is shown (which can be useful in visualizing data from upward- and
downward-aligned instruments), what colours are used to represent the signals,
etc.

Dozens of other plot types are also provided; see `?"plot,adp-method"` for
details -- but be warned that a complex plot requires many arguments!

When dealing with ship-based adp data, in order to get the true `east`,
`north`, and `up` velocity, the user must remove the speed of the ship. To do
this, use the `subtractBottomVelocity()` function, which subtracts the bottom
tracking velocities from an adp object.

**Exercise 2.** Create a u-v scatterplot.

**Exercise 3.** Write the steps to plot the `east`, `north`, and `up` velocity
without the speed of the ship for a file named
`COR2019002_20190818T064815_007_000000.ENS`.

# Subsetting

Adp data can be subsetted in a wide variety of ways, e.g.
```{r fig.height=7}
plot(subset(adp, distance<20))
```

plots only data within 20m of the instrument.  See `?"subset,adp-method"` for
more on subsetting.

**Exercise 4.** Plot the adp data for the first half of the sampling interval.

# Accessing data within adp objects

All oce objects share a common framework, with so-called "slots" for
metadata, data, and processing log. It is possible to examine, and modify, the
elements of each of these slots using base R constructs, but it is usually
better to use oce accessor functions.  The details are provided with
`?"[[,adp-method"` and `?"[[<-,adp-method"`, but a few examples should suffice
to sketch the general pattern.

For example,
```{r}
time <- adp[["time"]]
distance <- adp[["distance"]]
```
extracts the times and distances of the profiles within `data(adp)`, while the
array holding the velocities is extract with
```{r}
v <- adp[["v"]]
```

Altering values uses the `[[` to the left of the assignment operator, but this
is not commonly required in simple work, so it is best for readers to focus
first on accessing data. The exercises may help with this.

Some of the fields contained in adp objects, particularly the amplitude,
correlation, and percent good fields (`a`, `q`, and `g`) are stored as
`raw`-type. This is done primarly to save memory space, as expanded all of
those fields into `numeric` types can be slow and they get large quickly.
However, to actually work with the numbers, users can extract a "numeric"
version by supplying the second argument to the `[[` function, e.g.
```{r}
a <- adp[["a", "numeric"]]
```

This has the advantage of returning an object with the same matrix dimensions
as the original (note that functions like `as.numeric()` typically discard
dimension information, thereby converting an array to a vector).

**Exercise 5.** Plot a time series of mid-depth and depth-averaged eastward velocities.

**Exercise 6.** Perform an eigen analysis of eastward and northward velocity.

**Exercise 7.** Plot a time series of pressure variations, to reveal tidal
height.


# Solutions to exercises

**Exercise 1.** Determine the name of the `metadata` item that holds the
coordinate system of an adp object.

**Solution.**

With adp as loaded in the text, we may the names in the metadata with:
```{r results="hide"}
sort(names(adp[["metadata"]]))
```
(results not shown), after which a keen eye will notice that two elements
relate to coordinates:
```{r}
adp[["originalCoordinate"]]
adp[["oceCoordinate"]]
```
and comparison with the output of `summary(adp)` reveals that the former is the
coordinate system in which the data were originally measured, while the
latter is the coordinate system of the transformed data. Using
```{r}
processingLogShow(adp)
```
reveals the steps in the data transformation.


**Exercise 2.** Create a u-v scatterplot.

**Solution.**

```{r}
plot(adp, which="uv")
```

**Exercise 3.** Write the steps to plot the `east`, `north`, and `up` velocity
without the speed of the ship for a file named
`COR2019002_20190818T064815_007_000000.ENS`.

**Solution.**

```{r, eval=FALSE, message=FALSE, warning=FALSE, error=FALSE}
library(oce)
adcp <- read.adp("COR2019002_20190818T064815_007_000000.ENS")
enu <- toEnu(adcp)
removeShipSpeed <- subtractBottomVelocity(enu)
plot(removeShipSpeed, which=1:3)
```


**Exercise 4.** Plot the adp data for the first half of the sampling interval.

**Solution.**

```{r fig.height=7}
plot(subset(adp, time < median(adp[["time"]])))
```


**Exercise 5.** Plot a time series of mid-depth and depth-averaged eastward velocities.

**Solution.**
```{r}
time <- adp[["time"]]
v <- adp[["v"]]
# The second index is for bin number, the third for beam number
midIndex <- dim(v)[2]/2
eastMid <- v[, midIndex, 1] # third index is beam
distance <- adp[["distance"]][midIndex]
oce.plot.ts(time, eastMid, ylab="Eastward velocity [m/s]")
## Depth mean; note that na.rm, is passed by apply() to mean()
eastMean <- apply(v[,,1], 1, mean, na.rm=TRUE)
lines(time, eastMean, col=2)
legend("top", lwd=1, col=1:2,
       legend=c(paste("At", distance, "m"), "Depth Avg"))
```

Here, `oce.plot.ts()` is used instead of the basic R function `plot()`, because
it provides a margin note on the detailed time interval.


**Exercise 6.** Perform an eigen analysis of eastward and northward velocity.

**Solution.**
```{r}
u <- adp[["v"]][,,1]
v <- adp[["v"]][,,2]
ok <- is.finite(u) & is.finite(v) # remove NA values
u <- u[ok]
v <- v[ok]
eigen(cov(data.frame(u, v)))
```

Note that this action is equivalent to principal component analysis, and it
might be wiser to use
```{r}
pr <- prcomp(data.frame(u,v))
```

to do the analysis, which yields the same principal axes
```{r}
pr
```

but has the advantage of having associated methods, including for plotting.



**Exercise 7.** Plot a time series of pressure variations, to reveal tidal
height.

**Solution.**
```{r}
time <- adp[["time"]]
pressure <- adp[["pressure"]]
oce.plot.ts(time, pressure)
```

As a matter of interest, a tidal analysis may be done with
```{r}
m <- tidem(as.sealevel(pressure, time))
```

where the list of undetermined tidal constituents is long, because this tidal
record is so short (see `?tidem`). The fitted constituents are as follows
```{r}
summary(m)
```

(Note that it fitted for M2, but not S2, because the Rayleigh criterion
prevents inferring both, and the conventional tidal analysis favours M2 over S2
if the time series is too short to fit for more than one semi-diurnal
constituent.)

A useful step might be to draw a tidal fit along with the data. It makes sense
to construct a finer time grid, to get a smooth curve, as in the following (see
`?predict.tidem`).
```{r}
oce.plot.ts(time, pressure, type="p", col="blue")
timePredict <- seq(min(time), max(time), length.out=200)
pressurePredict <- predict(m, timePredict)
lines(timePredict, pressurePredict, col="red")
```

---
title: "5. Dealing with data-quality flags"
author: "Dan Kelley (https://orcid.org/0000-0001-7808-5911)"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_vignette:
    toc: true
    number_sections: true
    fig_caption: yes
    fig_width: 3.5
    fig_height: 3.5
    dpi: 72
    dev.args: list(pointsize=11)
vignette: >
  %\VignetteIndexEntry{5. Dealing with data-quality flags}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

<!-- HOW TO BUILD THE VIGNETTE. -->
<!-- Look in oce.Rmd for instructions. -->


```{r, echo = FALSE}
knitr::opts_chunk$set(collapse = TRUE, comment = "#>")
```

**Abstract.** This vignette explains the basics of working with data-quality
flags in oce. As in the vignette about working with [adp data](C_adp.html), the
material is organized to reflect typical workflow.  See the main oce
[vignette](A_oce.html) for general information about oce.  The sample code
provided here relies on the oce package, loaded as follows.
```{r}
library(oce)
```

# Overview

## Flag context

It is common for oceanographers to link data with "flags" that indicate data
quality.  This scheme permits retention of the original measurements, on the
assumption that they still hold some value, while at the same time providing an
indication of problems.  Using flags is superior to simply deleting suspicious
data, or to changing recorded values to `NA`, because it lets analysts try to
recover from the difficulty, perhaps by switching to another sensor, by
reducing the weights of individual data during averaging operations, etc.

Another advantage of using flags over simply discarding bad data is that flags
can be nuanced; e.g. distinguishing between unchecked data, data thought to be
good, data that are suspected of being bad, data that are sure to be faulty,
etc.

Flags are sometimes determined by automated analysis of data (e.g. range checks
on individual fields, departures from gravitational stability, loops in TS
diagrams, etc.). It is common for an analyst to use such tests as a starting
point for further processing that incorporates the study of graphical displays,
comparison with related measurements, etc.

## Flag conventions

A frustrating aspect of oceanographic analysis is that different data archiving
agencies employ different systems for flags. For example, the World
Hydrographic Programme designates good bottle/CTD data with a flag value of 2,
whereas the World Ocean Database (which might be available at
\code{https://www.ncei.noaa.gov/data/oceans/woa/WOD/CODES/TXT/Definition_of_Quality_Flags.txt},
although that link failed when tested on 2021-06-26) uses 0 for good data.

Analysts must be aware of the scheme used for a particular dataset.  It is also
good to be on the lookout for particular values in data, because oceanographers
sometimes set suspicious values to non-physical values that are easy to
recognize, e.g. -999, -99.99, or similar, and sometimes these coded values may
contradict formal flags, when both are present.


## Flag storage

Data-quality flags are stored in the `flags` entry in the `metadata` slot of
`oce` objects. This entry is a list, containing items with names matching the
to names of data elements that are stored in the `data` slot. Oce functions
assume that the flags and the data are in one-to-one correspondence, so the
flags must have the same dimensionality as the data.

## Flag inspection

It is possible to inspect flags using e.g.
\code{a[["metadata"]]\$flags\$salinity} to see flags corresponding to
\code{a[["data"]]\$salinity}, but a simpler notation is also provided, with
e.g. \code{a[["salinityFlag"]]} returning the flags for the salinity data.
(Note the \code{Flag} suffix in this notation; without that, the salinity data
would be returned.)

## Flag creation

Although it is possible to set and alter flags directly, it is much better to
use oce functions to do so, because (a) they will not create dimensionality or
name conflicts with the data and (b) they record their actions in the object's
`processingLog` slot.

Creating flags is usually a three-step process:

1. Use `initializeFlagScheme` to establish a
   mapping between flag numerical value and flag meaning.  This step is not
   strictly necessary, but it is useful because it stores the flag scheme
   within the object's `metadata` slot, as a form of documentation. This step
   may be done only once for a given object, because changing a flag scheme in
   the middle of processing invites errors.

2. Use `initializeFlags` to set up space for flag
   storage, and to initialize values (normally to a code indicating either
   unchecked data or acceptable data). This may be done only once per data
   item.

3. Use `setFlags` to assign individual flag values to
   individual data. This may be used repeatedly for a given object. Indeed, as
   illustrated later in this document, it is common to call this function
   within a loop that displays a graph, invites the user to click on points to
   be flagged, then displays the new graph, etc.

Note that some reading operations set up flags automatically, if the data file
already holds flags. For example, `read.argo` can do this,
because Argo data have a fixed flag scheme, and archived Argo data often come
with flag values already inserted.

## Handling flags

As with setting flags, analysts are free to extract flag values and use them
for any purpose that comes to mind. Simple cases may be handled with the oce
function named `handleFlags`, which is set up to respond
to particular flag values with particular actions.  It has reasonable defaults
for different data types, and it can detect a flag scheme that has been set up
with `setFlagScheme`. The default action of
setting the data that have been flagged as bad to the `NA` value may suffice in
many plotting or analysis situations.

# Sample working procedure

## CTD range checks

(This section is an expansion of Example 1 shown by `?"setFlags,ctd-method"`.)

The oce package provides a dataset
```{r}
library(oce)
data(ctdRaw)
```

that contains some clearly anomalous values that are revealed clearly
in a summary plot (Figure 1):
```{r fig.cap="Figure 1. Summary plot for raw CTD profile."}
plot(ctdRaw)
```

As Figure 1 shows, `ctdTrim` removes most of the anomalous
data by examining the variation of the pressure signal over time, it may also
be of interest to see how well simple range checks can perform in cleaning up
the data. Salinity certainly cannot be negative, but in an oceanographic
setting it is common to relax that criterion somewhat, perhaps insisting that
Absolute Salinity $S_A$ exceed $25$g/kg. This value might work in other
situations as well, and the same could be said of an upper limit of $40$g/kg.
Similarly, it might make sense to bound temperature between, say, $-2^\circ$C
and $40^\circ$C for application throughout much of the world ocean.

These criteria can be supplied to `setFlags` in various
ways, but the simplest is to create logical vectors, e.g.
```{r}
badS <- with(ctdRaw[["data"]], salinity < 25 | 40 < salinity)
badT <- with(ctdRaw[["data"]], temperature < -2 | 40 < temperature)
```

In the above, `with` has been used to avoid inserting `salinity` and
`temperature` in the namespace, but it would also be common to use e.g.
```{r}
S <- ctdRaw[["salinity"]]
T <- ctdRaw[["temperature"]]
bad <- (S < 25 | 40 < S) | (T < -2 | 40 < T)
```

Since the goal here is to illustrate setting multiple flags, the `badS` and
`badT` values will be used. The first step is to copy the original data, so
that the flag operations will not alter `ctdRaw`:
```{r}
qc <- ctdRaw
```

Work flow is best documented if a flag scheme is established, and the "WHP CTD
exchange" scheme is a reasonable choice; using
```{r}
qc <- initializeFlagScheme(qc, "WHP CTD")
```

stores a note on the scheme in the `metadata` slot of `qc`. Importantly,
however, it does *not* store any flag values. The next step is to initialize
flag values. For example, to set up flag storage for salinity and
temperature, use e.g.
```{r}
qc <- initializeFlags(qc, "salinity", 2)
qc <- initializeFlags(qc, "temperature", 2)
```

to create storage, and initialize all entries to the `"acceptable"` value
(which, for this flag scheme, is the number `2`).  Once storage has been
initialized for a given variable, further calls to
`initializeFlags` will have no effect, apart from a
warning.

Note that it is entirely possible to use
`initializeFlags` with a numerical value for its
third argument.  One advantage of using
`initializeFlagScheme` is that it clarifies
code, but the bigger advantage is that it embeds the scheme within the object,
so that a second analyst could examine it later and be clear on the meanings of
the numerical codes. This is very important, because there is not much
agreement within the oceanographic community on which flag scheme to use, for
many data types (Argo being an exception).

At this stage, individual data can be flagged with
`setFlags`. This function can be called any number of
times. Continuing along with our example, we may mark bad salinities with
```{r}
qc <- setFlags(qc, "salinity", badS, value="bad")
```

We can see that the flag got inserted by using `summary(qc)`, but for brevity here
another method is:
```{r}
names(qc[["flags"]])
```

Now, temperature flags may be inserted with
```{r}
qc <- setFlags(qc, "temperature", badT, value="bad")
```

Readers ought to use `summary(qc)` to get more details of how flags were
handled, and how many bad salinities and temperatures were flagged.

If `qc` is plotted with `plot(qc)`, the results will match those of
`plot(ctdRaw)`. This is because setting flags has no effect on plots, because
it alters *flags* but not *data*. One more step is required to test whether
this procedure has cleaned up the data significantly: we must "handle" the
flags, using
```{r}
qch <- handleFlags(qc, flags=list(c(1, 3:9)))
```

Comparing Figure 1 with the summary plot for `qch` (Figure 2), constructed with
```{r fig.cap="Figure 2. Summary plot for range-checked CTD profile."}
plot(qch)
```

shows significant improvements. The downcast and the upcast can be seen quite
clearly now, although there appears to be an issue of low salinity at the
turnaround point.  Setting a flag for pressure increase with time will isolate
the downcast somewhat, although some smoothing will be required.  Another issue
related to the path of the instrument is that it may have been held below the
surface for a while to equilibrate. Again, a flag could be set up to remove
such data. However, it ought to be noted that `ctdTrim` can
be used to address issues relating to instrument movement.

## Interactive editing of CTD profiles

(This section is an expansion of Example 2 shown by `?"setFlags,ctd-method"`.)

The `ctd` dataset provided by the `oce` package is similar to `ctdRaw`, except
that only downcast data are provided. Even so, there are still some points that
might be considered suspicious. A common way to find such points is to plot TS
diagrams, looking for decreases in density with depth.

Running the following code in an interactive session will demonstrate a simple
way to use a TS diagram to identify suspicious data. Pasting this into an R
console will show a plot with lines between measurements made at successive
depths. Clicking on any point will flag it, and so the point will then
disappear on the plot.  Clicking to the right of the plot frame will exit the
procedure, after which `qc` is a ctd object with flags as set, and data set
to `NA` where these flags indicate bad data.

```{r eval=FALSE}
options(eos="gsw")
data(ctd)
qc <- ctd
qc <- initializeFlagScheme(qc, "WHP CTD")
qc <- initializeFlags(qc, "salinity", 2)
Sspan <- diff(range(qc[["SA"]]))
Tspan <- diff(range(qc[["CT"]]))
n <- length(qc[["SA"]])
par(mfrow=c(1, 1))
plotTS(qc, type="o")
message("Click on bad points; quit by clicking to right of plot")
for (i in seq_len(n)) {
    xy <- locator(1)
    if (xy$x > par("usr")[2])
        break
    i <- which.min(abs(qc[["SA"]] - xy$x)/Sspan + abs(qc[["CT"]] - xy$y)/Tspan)
    qc <- setFlags(qc, "salinity", i=i, value="bad")
    qc <- handleFlags(qc)
    plotTS(qc, type="o")
}
```

It would be a simple matter to extend this simple example to a shiny
application that displays other data. For example, there could be panels for
profiles, as well as the TS plot. The system could track clicks in any of the
panels, taking appropriate actions. It would be sensible to have a staged
procedure, in which clicking (or brushing) in one panel would cause a replot of
all panels, with the selected data indicated in some way, so that the analyst
could then choose whether to go to the next stage, of clicking a button to
indicate bad data. Another button might be provided to undo such operations, or
to show the original data for comparison. The point is that a wide variety of
flag operations are handled very easily in R, with oce.

## Hydrographic sections

The flag-handling functions work for a variety of oce objects. For example,
sections, which are built up from a sequence of ctd profiles, are handled
easily with functions of the *same names* as for the ctd case.

As a simple example, the following shows how to clean up the "A03" Atlantic
section that is provided with oce.

```{r}
data(section)
s <- handleFlags(section, flags=list(c(1, 3:9)))
par(mfrow=c(2, 1))
plotTS(section)
plotTS(s)
```

Note, in the above, that the archiving agency had evidently flagged not just
wild data (e.g. the salinity near $26$g/kg) but also data that were anomalous
in more subtle ways (e.g. cleaning up several points that stood out from the
data cloud, below $10^\circ$C). In fact, the flags in this data set, as in most
archived hydrographic data sets, are the result of a multifaceted inspection
scheme that is more demanding and useful than simple range checks.


# Acoustic-Doppler profiler data

By now, the reader should be able to understand the following use of
data-quality flags for the `adp` dataset. Note that there is only a small
difference (at 8h and 20h) between the (gray) bad data as identified by the
instrument and those identified by the scheme illustrated here; look near 8h
and 20h. However, altering the values for `G` and `V4` will reveal that the
schemes do differ.

```{r}
data(adp)
v <- adp[["v"]]
i2 <- array(FALSE, dim=dim(v)) # construct array to match 'v'
g <- adp[["g", "numeric"]]
G <- 25                  # for percent-good field, named 'g'
V4 <- 0.45               # for error velocity field, in beam 4
for (k in 1:3)
    i2[,,k] <- ((g[,,k]+g[,,4]) < G) | (v[,,4] > V4)
adpQC2 <- initializeFlags(adp, "v", 2)
adpQC2 <- setFlags(adpQC2, "v", i2, 3)
adpClean2 <- handleFlags(adpQC2, flags=list(3), actions=list("NA"))
par(mfrow=c(2,1))
plot(adp, which="u1")                  # original
plot(adpClean2, which="u1")            # altered
```

Note, in the above, that `initializeFlagScheme`
has not been used, because the author is unaware of any common notation for
such data. The values used for `G` and `V4` were provided by colleagues at the
Bedford Institute of Oceanography.


---
title: "4. Using map projections"
author: "Dan Kelley (https://orcid.org/0000-0001-7808-5911)"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_vignette:
    toc: true
    number_sections: true
    fig_caption: yes
    fig_width: 3.5
    fig_height: 3.5
    dpi: 72
    dev.args: list(pointsize=11)
vignette: >
  %\VignetteIndexEntry{4. Using map projections}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

<!-- HOW TO BUILD THE VIGNETTE. -->
<!-- Look in oce.Rmd for instructions. -->

```{r, echo = FALSE}
knitr::opts_chunk$set(collapse = TRUE, comment = "#>")
```

**Abstract.** An overview of the handling of map projections in oce is
presented, along with a few examples of common projections.

# Introduction

Map projections provide methods for representing the three-dimensional surface
of the earth as two-dimensional plots.  Although most oceanographers are likely
to be familiar with the basic ideas of map projection, they may find it helpful
to consult the wide literature on this topic, whether to learn about the
details of individual projections and to get advice on the best choice of
projection for a particular task. Snyder (1987), Snyder (1993) and Snyder and
Voxland (1994) are all excellent sources on these topics, and the first and
last of these are available for free online.

Oce handles map projections by calling the `sf_project` function of the
`sf` R package, and so the notation for representing the projection is
borrowed from `sf`.  This system will be familiar to many readers,
because it is used in other R packages, and more widely in the software called
PROJ, which has interfaces in python and other programming languages (PROJ
contributors, 2020). Since oce uses inverse projections in some of its
graphical work, only PROJ projections that have inverses are incorporated into
oce.  Furthermore, some projections are omitted from oce because they have been
witnessed to cause problems on the oce developer's computers, including
infinite loops and core dumps. See the help for the oce function `mapPlot` for
a list of the available projections, and some advice on choosing them.

You must ensure that the `sf` package is installed for the examples provided
in this vignette to work.  If this package is not installed, it may be installed
with
```{r eval=FALSE}
install.packages("sf")
```
(Of course, this step need not be repeated to use `oce`.)


There are far too many projections to illustrate here. See
<https://dankelley.github.io/r/2015/04/03/oce-proj.html> for a blog item that
provides examples of the available projections. Note that some have a problem
with spurious horizontal lines. This can result from coastline segments that
cross the edge of the plotting area, and getting rid of this is tricky enough
to be the heart of the longest-lived bug in the `oce` issue list, i.e.
<https://github.com/dankelley/oce/issues/388>. In some instances the function
`coastlineCut()` can help, but it is provisional and subject to change.


# Avoiding projections for small regions

For map views that span only a few tens or hundreds of kilometers, it may be
sufficient to plot directly, in rectilinear longitude-latitude space, but with
an appropriate aspect ratio so that circular islands will appear circular in
the plot.

To see the need for setting the aspect ratio, consider the following view of
eastern North America.

```{r}
library(oce)
data(coastlineWorld)
lon <- coastlineWorld[["longitude"]]
lat <- coastlineWorld[["latitude"]]
par(mar=c(4, 4, 0.5, 0.5))
plot(lon, lat, type="l",
     xlim=c(-130, -50), ylim=c(40, 50))
```

Readers familiar with this region will notice that the coastline
shapes are distorted.  The solution to this problem is to set
the `asp` argument of `plot()` to a value that will compensate
for meridional convergence at the poles, as below.

```{r}
par(mar=c(4, 4, 0.5, 0.5))
plot(lon, lat, type="l",
     xlim=c(-130, -50), ylim=c(40, 50), asp=1/cos(45*pi/180))
```

Although the above approach is not exactly taxing, the effort of setting the
aspect ratio and setting line-type plots can be spared by using the generic
`plot()` function for `coastline` objects, as follows.

```{r}
plot(coastlineWorld, clongitude=-90, clatitude=45, span=7000)
```

Note that this `plot` is being transferred to the specialized method for a
`coastline` object (defined by the `oce` package) and that this means that
there are some arguments in addition to those of the normal `plot` function.
In this example, the second and third arguments specify the central spot of
interest, and the fourth is a suggested diagonal span, in kilometres.  (The
span is not always obeyed exactly, and the results also depend on aspect ratio
of the plot device.)

The graphs shown above share a common strength: the axes for longitude and
latitude are orthogonal, and the scale along each axis is linear. This makes it
easy for readers to identify the location of features of the diagrams. However,
using linear axes on large-scale views leads to distortions of coastline shapes
and relative feature sizes, and motivates the use of map projections. (Readers
who question the previous sentence are again encouraged to consult some of the
references in the bibliography.)

# Choosing a map projection

## World views

It makes sense to switch to a coarser coastline file for the examples to
follow, because plotting a detailed coastline in a world view leads to a
scribbling effect that obscures the large-scale coastline shapes.

```{r}
data(coastlineWorld)
```

The function used to produce maps with projections is
`mapPlot`, and with default arguments it produces

```{r fig.cap="World coastline with default (Mollweide) projection.", fig.width=5, fig.height=2.7, dpi=72, dev.args=list(pointsize=10)}
par(mar=c(1.5, 1.5, 0.5, 0.5))
mapPlot(coastlineWorld, col="lightgray")
```

Note that the land is coloured in this example, but that sometimes doing that
will yield odd-looking plots, because of a bug in how some projections are
handled in oce.

This default plot uses the Mollweide projection (see any of the Snyder
references in the bibliography), which is set by the default value
`"+proj=moll"` for `mapPlot`.
Another popular world view is the Robinson projection, which has been used by
Rand McNally, the National Geographic Society, and other groups.

**Exercise 1.** Plot the world with a Robinson projection, using gray shading for land.

**Exercise 2.** Plot an image of world topography with the Mollweide projection.

## Polar views

In both these views (and in most world-spanning projections) there is
substantial distortion at high latitudes. The Stereographic projection offers a
solution to this problem, e.g. in the following (which employs a trick on the
latitude limit, specifying an image point on the other side of the planet, with
the pole in between).

```{r fig.cap="Polar view with stereographic projection.", fig.width=5, fig.height=5, dpi=72, dev.args=list(pointsize=10)}
par(mar=c(1.5, 1.5, 0.5, 0.5))
mapPlot(coastlineWorld, longitudelim=c(-130, 50), latitudelim=c(60, 120),
        projection="+proj=stere +lat_0=90", col='gray')
```

Here, the region of interest is defined with `longitudelim` and `latitudelim`,
instead of with `clongitude`, `clatitude` and `span`. Note that the latitude
limit is set to extend past the pole, which is a way of making the plot include
the pole.  See the table in the documentation for `mapPlot` for a listing of
arguments, and for citations to external resources that explain what they mean.

**Exercise 3.** Draw a view of Antarctica and the Southern Ocean.


## Mid-latitude views

The Lambert Conformal view is often used in maps that span wide
longitudinal ranges, e.g. a map of Canada may be produced as follows.

```{r}
par(mar=c(1.5, 1.5, 0.5, 0.5))
mapPlot(coastlineWorld,
    longitudelim=c(-130,-55), latitudelim=c(45, 70),
    projection="+proj=lcc +lat_1=50 +lat_2=65 +lon_0=-100", col='gray')
```

Here, `lat_1` and `lat_2`, the latitudes where the projection cone intersects
the earth, are set to span the populated region of Canada and the northern USA,
and `lon_0`, the meridian that will be vertical on the plot, is chosen to lie
near the middle of the continent.

**Exercise 4.** Plot the eastern North Atlantic using a Mercator projection.

**Exercise 5.** Plot the eastern North Atlantic using an Albers Equal-Area
Conic projection.



# Adding to map plots

Several functions are provided by oce to draw additional features on maps.
These include
`mapText` for adding text,
`mapPoints` for adding points,
`mapLines` for adding lines,
`mapContour` for adding contours, and
`mapImage` for adding images.
Each of these requires a map to have been drawn first with
`mapPlot`. The interfaces ought to be similar
enough to base-R functions that readers can decide how to accomplish
common tasks.

**Exercise 6.** Plot the world with the Goode projection, superimposing sea surface temperature as contours.



# Solutions to exercises

**Exercise 1.** Plot the world with a Robinson projection, using gray shading for land.

```{r fig.cap="World coastline with Robinson projection.", fig.width=5, fig.height=2.7, dpi=72, dev.args=list(pointsize=10)}
par(mar=rep(0.5, 4))
mapPlot(coastlineWorld, col="lightgray", projection="+proj=robin")
```

**Exercise 2.** Plot an image of world topography with the Mollweide projection.

Note that some adjustment of the margins is required to fit the colorbar.

```{r fig.cap="World topography with Mollweide projection.", fig.width=5, fig.height=2.7, dpi=72, dev.args=list(pointsize=10)}
par(mar=c(1.5, 1, 1.5, 1))
data(topoWorld)
topo <- decimate(topoWorld, 2) # coarsen grid: 4X faster plot
lon <- topo[["longitude"]]
lat <- topo[["latitude"]]
z <- topo[["z"]]
cm <- colormap(name="gmt_globe")
drawPalette(colormap=cm)
mapPlot(coastlineWorld, projection="+proj=moll", grid=FALSE, col="lightgray")
mapImage(lon, lat, z, colormap=cm)
```

**Exercise 3.** Draw a view of Antarctica and the Southern Ocean.

```{r}
par(mar=c(1.5, 1.5, 0.5, 0.5))
mapPlot(coastlineWorld, longitudelim=c(-180, 180), latitudelim=c(-130, -50),
       projection="+proj=stere +lat_0=-90", col='gray', grid=15)
```


**Exercise 4.** Plot the eastern North Atlantic using a Mercator projection.

```{r fig.cap="Eastern North Atlantic with Mercator projection.", fig.width=5, fig.height=5, dpi=72, dev.args=list(pointsize=10)}
par(mar=c(1.5, 1.5, 0.5, 0.5))
mapPlot(coastlineWorld, col="lightgray",
        projection="+proj=merc",
        longitudelim=c(-80, -40), latitudelim=c(20, 60))
```

**Exercise 5.** Plot the eastern North Atlantic using an Albers Equal-Area
Conic projection.

In the solution given below, note how the `lon_0` argument is set to be at the
central value of `longitudelim`. This makes the $60^\circ$W line be vertical on
the plot. The settings for `lat_1` and `lat_2` are the locations at which the
cone intersects the earth, so distortion is minimized at those latitudes. Note
how the lines of constant longitude are changed, compared with those in the
Mercator view. It may be informative to compare the relative ratio of the area
of Nova Scotia (at $45^\circ$N) and Ungava Bay (at $60^\circ$N) in this Albers
Equal-Area projection, and the Mercator projection shown in the previous
diagram.



```{r fig.cap="Eastern North Atlantic with Albers equal-area projection.", fig.width=5, fig.height=5, dpi=72, dev.args=list(pointsize=10)}
par(mar=c(1.5, 1.5, 0.5, 0.5))
mapPlot(coastlineWorld, col="lightgray",
        projection="+proj=aea +lat_1=30 +lat_2=50 +lon_0=-60",
        longitudelim=c(-80, -40), latitudelim=c(20, 60))
```

**Exercise 6.** Plot the world with the Goode projection, superimposing sea surface temperature as contours.

Note that the temperature is provided by the `ocedata` package.

```{r fig.cap="SST contours with the Goode projection", fig.width=5, fig.height=2.7, dpi=72, dev.args=list(pointsize=10)}
par(mar=rep(0.5, 4))
mapPlot(coastlineWorld, projection="+proj=goode", col="lightgray")
if (requireNamespace("ocedata", quietly=TRUE)) {
    data(levitus, package="ocedata")
    mapContour(levitus[['longitude']], levitus[['latitude']], levitus[['SST']])
}
```


# References

PROJ contributors. "PROJ Coordinate Transformation Software Library." Open
Source Geospatial Foundation, 2020.
<https://proj.org/>

Snyder, John P. "Map Projections-a Working Manual." Washington: U.S.
Geological survey professional paper 1395, 1987.
<https://pubs.usgs.gov/pp/1395/report.pdf>

Snyder, John Parr. "Flattening the Earth: Two Thousand Years of Map
Projections." Chicago, IL: University of Chicago Press, 1993.
<https://press.uchicago.edu/ucp/books/book/chicago/F/bo3632853.html>

Snyder, John P., and Philip M. Voxland. "An Album of Map Projections."
U. S.  Geological Survey Professional Paper. Washington, DC, USA: U. S.
Department of the Interior: U. S. Geological Survey, 1994.
<https://pubs.er.usgs.gov/publication/pp1453>


---
title: "2. Analysis of CTD data"
author: "Dan Kelley (https://orcid.org/0000-0001-7808-5911) and Clark Richards (https://orcid.org/0000-0002-7833-206X)"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_vignette:
    toc: true
    number_sections: true
    fig_caption: yes
    fig_width: 3.5
    fig_height: 3.5
    dpi: 72
    dev.args: list(pointsize=11)
vignette: >
  %\VignetteIndexEntry{2. Analysis of CTD data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

<!-- HOW TO BUILD THE VIGNETTE. -->
<!-- 1. edit this file in vignettes/ctd.Rmd -->
<!-- 2. in an R console, use setwd() or otherwise move to the oce -->
<!--    directory, and then execute -->
<!--      devtools::build_vignettes() -->
<!-- 3. build oce. In the shell, you might use -->
<!--      cd oce ; echo "devtools::document(roclets=c('rd','collate','vignette'))" | R --no-save -->
<!--      CMD build --compact-vignettes="qpdf" oce -->
<!--      R CMD check oce_1.1-0.tar.gz -->
<!--      R CMD INSTALL oce_1.1-0.tar.gz -->
<!--    where the version number might need updating, or in an R console you might use -->
<!--      devtools::install(build_vignettes=TRUE) -->


```{r, echo = FALSE}
knitr::opts_chunk$set(collapse = TRUE, comment = "#>")
```

**Abstract.** The basics of working with CTD data are outlined in this
vignette, as a supplement to a shorter entry in the [main vignette](A_oce.html).



# Example with pre-trimmed data

Many of the object types supported by `oce` come with built-in data.  For an
example, `data(ctd)` yields a CTD profile that has been trimmed to just the
downcast portion of the sampling.  (See the next section to learn how to do
this trimming.)  A summary and plot (Figure 2) are created as follows.

```{r fig.cap="**Figure 2.** An overview of a ctd dataset.", fig.width=6, fig.height=6, dpi=72, dev.args=list(pointsize=14), message=FALSE}
library(oce)
data(ctd)
summary(ctd)
plot(ctd)
```

Accessing the data within this `ctd` object can be done directly, e.g.
`ctd@data$pressure` holds the pressure record, but it is usually better to use
an accessor function that is provided with `oce`. This function is named `[[`,
and it takes a character string as an argument, e.g. `ctd[["pressure"]]` yields
the pressure column. The accessor notation is preferable to direct access
because it is simpler for the user. For example, several `oce` objects store
the data in single-byte or two-byte chunks, to match the raw format used by the
instruments, and the accessor function takes care of translating these values
to what are sometimes called "science" units.

**Exercise 1.** Plot a profile of $\sigma_\theta$ and $N^2$ within the
pycnocline in the built-in `ctd` data set.  (Hint: use `subset()`.)

# Example with raw data

Practicing Oceanographers may be wondering how the CTD cast used in the
preceding section was trimmed of equilibration-phase and upcast-phase data.
Spurious data from these phases must be trimmed as a first step in processing.
For example, consider the following code.
```{r fig.cap="**Figure 3.** Scanwise plot of the `ctdRaw` sample data set.  Note the spike at the start, the equilibration phase before the downcast, and the spurious freshening signal near the start of the upcast.", fig.width=5, fig.height=3, dpi=72, dev.args=list(pointsize=12)}
data(ctdRaw)
plotScan(ctdRaw)
```

This produces a two-panel plot (Figure 3) of the data as a time-series,
revealing not just the desired downcast, but also an earlier equilibration
phase and a later upcast.  The x-axis in Figure 3 is the scan number, which is
a convenient index for extraction of the downcast portion of the profile by an
essentially manual method, e.g. proceeding with a sequence of commands such as

```{r eval=FALSE}
plotScan(ctdTrim(ctdRaw, "range",
                 parameters=list(item="scan", from=140, to=250)))
plotScan(ctdTrim(ctdRaw, "range",
                 parameters=list(item="scan", from=150, to=250)))
```

This method of making decisions based on plotted information is probably the
most robust method of trimming data.  However, for quick work, users may be
satisfied with the results of automatic downcast detection, e.g.
```{r eval=FALSE}
ctdTrimmed <- ctdTrim(ctdRaw)
```

It should be noted that `ctdTrim()` inserts entries into the object's log file,
so that the details of how the trimming was done are recorded together with the
data.

Once the profile has been trimmed, one may wish to use `ctd.decimate()` to smooth
the data and interpolate the smoothed results to uniformly-spaced pressure
values.

Taking these things together, a quick visual examination of a CTD file takes
just one line of code:
```{r eval=FALSE}
plot(ctdDecimate(ctdTrim(read.ctd("stn123.cnv"))))
```

# Example with WOCE archive data

The package has a harder time scanning the headers of data files in the WOCE
archive format than it does in the Seabird format illustrated in the previous
examples. This is mainly because front-line researchers tend to work in the
Seabird format, and partly because the WOCE format is odd. For example, the
first line of a WOCE file is of the form `CTD,20060609WHPOSIODAM` (or
`BOTTLE,...`).  Scanning the item to the left of the comma is not difficult
(although there are variants to the two shown, e.g. `CTDO` sometimes occurs).
The part to the right of the comma is more difficult.  The first part is a date
(`yyyymmdd`) so that is no problem.  But then things start to get tricky.  In
the example provided, this text contains the division of the institute (WHPO),
the institute itself (SIO), and initial of the investigator (DAM). The problem
is that no dividers separate these items, and that there seem to be no
standards for the item lengths. Rather than spend a great deal of time coding
special cases (e.g. scanning to see if the string `WHOI` occurs in the header
line), the approach taken with `oce` is to ignore such issues relating to
quirky headers, on the assumption that users can scan human-written headers
with high skill.

Quite commonly, CTD profiles taken during a cruise are collected together in a
sequence of files in a given directory.  For a real-world example, one might
visit the website mentioned in the code provided below, download and expand the
zip file, enter the directory thus formed, and run the code to get an overall
TS plot for all the CTD stations of this cruise. (Caution: the link seems to
change from time to time.)

```{r eval=FALSE}
library(oce)
# http://cchdo.ucsd.edu/data/7971/ar18_58JH19941029_ct1.zip
# setwd("~/Downloads/ar18_58JH19941029_ct1")
files <- system("ls *.csv", intern=TRUE)
for (i in 1:length(files)) {
    x <- read.ctd(files[i])
    if (i == 1) {
        plotTS(x, Slim=c(31, 35.5), Tlim=c(-2, 10), type='o')
    } else {
        points(x[["salinity"]], x[["potential temperature"]])
        lines(x[["salinity"]], x[["potential temperature"]])
    }
}
```
The `[[` notation is explained in Section 3, but this example conveys the gist,
that it permits accessing data, or derived data, from within an object.

In the above, lines connect the points within a given profile. This can be a
useful method for a quick scan looking for outliers. Another is to colour-code
the profiles, although this gets confusing with large datasets, in which case
the method of the following exercise might be useful.

**Exercise 2. (advanced)** Make a multi-file plot summarizing the TS
relationship in a collection of CTD datasets, with each plot file showing the
overall relationship in gray and the individual profile in black.


# Solutions to exercises

**Exercise 1.** Plot a profile of $\sigma_\theta$ and $N^2$ within the
pycnocline in the built-in `ctd` data set.  (Hint: use `subset()`.)

Although one may argue as to the limits of the pycnocline, for illustration let
us say it is in 5 dbar to 12 dbar range.  One way to do this is

```{r fig.width=5, fig.height=5, fig.keep="none"}
library(oce)
data(ctd)
pycnocline <- ctdTrim(ctd, "range",
                      parameters=list(item="pressure", from=5, to=12))
plotProfile(pycnocline, which="density+N2")
```

Another is
```{r fig.width=5, fig.height=5, fig.keep="none"}
library(oce)
data(ctd)
pycnocline <- subset(ctd, 5 <= pressure & pressure <= 12)
plotProfile(pycnocline, which="density+N2")
```

**Exercise 2. (advanced)** Make a multi-file plot summarizing the TS
relationship in a collection of CTD datasets, with each plot file showing the
overall relationship in gray and the individual profile in black.

The code provided below creates 91 PNG files, with names `ar18_01.png`,
`ar18_02.png`, etc. Loading these in a view that permits quick paging through
this file list is an easy way to spot suspicious data, since each plot has the
station number at the top.  (Users trying this example should bear in mind that
this is a fairly large dataset, so the processing will take up to a minute.)

```{r eval=FALSE}
library(oce)
# http://cchdo.ucsd.edu/data/7971/ar18_58JH19941029_ct1.zip
# setwd("~/Downloads/ar18_58JH19941029_ct1")
files <- system("ls *.csv", intern=TRUE)
n <- length(files)
ctds <- vector("list", n) # to hold the CTD objects
station <- vector("list", n)
for (i in 1:n) {
    ctds[[i]] <- read.ctd(files[i])
    station[[i]] <- ctds[[i]][["station"]]
}
S <- unlist(lapply(1:n, function(i) ctds[[i]][["salinity"]]))
T <- unlist(lapply(1:n, function(i) ctds[[i]][["temperature"]]))
p <- unlist(lapply(1:n, function(i) ctds[[i]][["pressure"]]))
overall <- as.ctd(S, T, p)
png("ar18_%02d.png")
for (i in 1:n) {
    plotTS(overall, col='gray')
    lines(ctds[[i]][["salinity"]], ctds[[i]][["potential temperature"]])
    mtext(station[i], side=3, line=0)
}
dev.off()
```


---
title: "1. Introduction to Oce"
author: "Dan Kelley (https://orcid.org/0000-0001-7808-5911) and Clark Richards (https://orcid.org/0000-0002-7833-206X)"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_vignette:
    toc: true
    number_sections: true
    fig_caption: yes
    fig_width: 3.5
    fig_height: 3.5
    dpi: 72
    dev.args: list(pointsize=11)
vignette: >
  %\VignetteIndexEntry{1. Introduction to Oce}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options:
  markdown:
    wrap: 72
---

<!-- HOW TO BUILD AND VIEW THE OCE VIGNETTES. -->

<!-- 1. Edit this file in vignettes/oce.Rmd -->

```{=html}
<!-- 2. Do not bother trying to build vignettes in RStudio. Instead, ->
<!--    in a unix shell, type the following: -->
```
<!--        R CMD build --compact-vignettes="qpdf" oce -->

<!--        R CMD install oce_1.1-2.tar.gz -->

<!--    where the version number may need adjustment in the second command. -->

<!-- 3. To see the results, type the following in an R console: -->

<!--       browseVignettes("oce") -->

```{r, echo = FALSE}
knitr::opts_chunk$set(collapse = TRUE, comment = "#>")
```

**Abstract.** The `oce` package makes it easy to read, summarize and
plot data from a variety of Oceanographic instruments, isolating the
researcher from the quirky data formats that are common in this field.
It also provides functions for working with basic seawater properties
such as the equation of state, and with derived quantities such as the
buoyancy frequency. Although simple enough to be used in a teaching
context, `oce` is powerful enough for a research setting. These things
are illustrated here in a general context; see also the vignettes on
[using data-quality flags](E_flags.html) and processing data from
[acoustic Doppler profilers](C_adp.html).

# Introduction

Oceanographers must deal with measurements made by a wide variety of
instruments, a task that is complicated by a tendency of instrument
manufacturers to invent new data formats. Although manufacturers often
provide software for scanning data files and producing overview plots,
this software is of limited use to researchers who work with several
instrument types at the same time, and who need to move beyond
engineering plots to scientific plots and statistical analysis.

# Architecture of oce objects

![**Figure 1.** Basic layout of a CTD object. All `oce` objects contain
slots named `data`, `metadata`, and `processingLog`, with the contents
depending on the type of data.](ctd-object.png)

The need to scan diverse data files was one motivation for the creation
of `oce`, but an equal goal was to make it easy to work with the data
once they are in the system. This was accomplished partly by the
provision of specialized and generic (overloaded) functions to work with
the data, and partly by providing accessor methods that make it
convenient to reach inside the data objects (see next section).

As illustrated in Figure 1, each `oce` object contains three slots:

-   `data`, a list containing the actual data, e.g., for a CTD object
    (see also the [vignette about CTD data](B_ctd.html)), this will
    contain `pressure`, `temperature`, etc.
-   `metadata`, a list containing information about the data, such as
    units, data-quality flags, sampling locations, etc.
-   `processingLog`, a list that documents how the object was created
    and possibly changed thereafter.

# Accessing data in oce objects

For detailed analysis, users may want to access data *within* oce
objects. While it is possible to descend through the object using the
"slot" and "list" notation (e.g. `d@data$salinity` yields the salinity
within an oce object named `d`), this approach is not recommended. It is
better to use the `[[` notation, which derives from the generic
"Extract" method for accessing parts of an object. A general
introduction is provided by

```{r eval=FALSE}
?`[[,oce-method`
```

and details are provided for individual object classes with e.g.

```{r eval=FALSE}
?`[[,ctd-method`
```

for the `ctd` class. The notation is very simple. For example, suppose
that `d` is an object that stores `salinity` in either its `data` slot
or `metadata` slot. Then,

```{r eval=FALSE}
S <- d[['salinity']]
```

will extract the salinity.

The `[[` method first looks in the `metadata` slot, but if the item is
not found there, it proceeds to look in the `data` slot. This two-step
scheme is helpful because it frees the user from having to know where
data are stored, e.g. in a `ctd` object, `latitude` might be stored in
the `metadata` (for a conventional CTD cast) or in the `data` slot (for
the slantwise sampling of a glider).

In addition to accessing data within an object, the `[[` notation also
permits the extraction of *derived* information. There are two reasons
for this.

1.  It can give valuable performance enhancements in some cases, e.g.
    the data in the `landsat` class are stored in two byte-level arrays,
    which yields a marked improvement over the 8-byte arrays that R
    generally uses on a 64-bit machine. The `[[` assembles these
    byte-scale chunks into a conventional R numerical matrix, which can
    be quite convenient for users who wish to operate on the data
    without learning how to assemble the bytes.
2.  It provides a uniformity of notation that can be helpful to users.
    In `oce`, objects derived from data files tend to hold just the
    information in those files, not derived information. For example,
    For example, CTD datasets often provide in-situ temperature but not
    potential temperature (since the latter is not measured). The
    in-situ temperature is found with e.g. `d[["temperature"]]`, and so
    it seems natural to write `d[["theta"]]` to get the potential
    temperature. If this is done, then the `ctd` version of `[[` first
    looks in the data slot, and returns `theta` if it is found there, as
    may sometimes be the case (depending on the choice of the person who
    created the dataset). However, if it is not found, then `[[` calls
    `swTheta()` to calculate the value, and returns the result.

Finally, it is also possible to extract the entirety of either the
`metadata` or `data` slot, e.g.

```{r eval=FALSE}
data <- d[['data']]
```

yields the full data slot, which is a list with elements that can be
accessed in the conventional way, e.g. for a `ctd` object,

```{r eval=FALSE}
data$temperature
```

retrieves the temperature. For obvious reasons, the method of
derived-quantity access (e.g. the `theta` of the example above) will not
work.

# Modifying data in oce objects

There are several schemes for modifying the data within `oce` objects.
By analogy with the `[[` notation of the previous section, e.g. the
following

```{r eval=FALSE}
data(ctd)
ctd[["temperatureAboveFreezing"]] <- ctd[["temperature"]] - swTFreeze(ctd)
```

will store the excess over freezing temperature into the `ctd` object.

Further information on this notation is provided by

```{r eval=FALSE}
?"[[<-,oce-method"
```

The above works only within the `data` slot. To store within the
`metadata` slot, consider using e.g.

```{r extractExample, eval=FALSE}
ctd[["metadata"]]$scientist <- "Dalhousie Oceanography 4120/5120 Class of 2003"
```

sets the "scientist".

For archival work, it is important to store reasons for changing things
in an object. Two functions are provided for this purpose: `oceSetData`
and `oceSetMetadata`. For example, a better way to change the scientist
might be to write

```{r eval=FALSE}
ctd <- oceSetMetadata(ctd, name="scientist",
                      value="Dalhousie Oceanography 4120/5120 Class of 2003",
                      note="give credit where it's due")
```

and a better way to store `temperatureAboveFreezing` would be

```{r eval=FALSE}
ctd <- oceSetData(ctd, name="temperatureAboveFreezing",
                      value=ctd[["temperature"]] - swTFreeze(ctd),
                      unit=list(unit=expression(degree*C), scale="ITS-90"),
                      originalName="-",
                      note="add temperatureAboveFreezing, for ice-related calculations")
```

which illustrates that this notation, as opposed to the `[[<-` notation,
permits the specification of a unit and an `originalName`, both of
which, together with the `note`, are displayed by `summary(ctd)`.

# Oce functions

The uniformity of the various `oce` objects helps users build skill in
examining and modifying objects. Fine-scale control is provided
throughout `oce`, but the best way to learn is to start with simplest
tools and their defaults. For example, the following will read a CTD
file named `"station1.cnv"`, summarize the contents, and plot an
overview of the data, with profiles, a TS diagram, and a map (Figure 2).

```{r eval=FALSE}
library(oce)
d <- read.oce("station1.cnv")
summary(d)
plot(d)
```

The reader should stop now and try this on a file of their own. The
pattern will work with a fairly wide variety of file types, because
`read.oce()` examines the file name and contents to try to discover what
it is. For an example, if `read.oce()` is given the name of a file
created by an Acoustic Doppler Profiler, it will return an object
inheriting from class `"adp"`, so the `summary()` and `plot()` calls
will be tailored to that type, e.g. the graph will show images of
time-distance variation in each of the measured velocity components.

**Notes on oce function names.**

1.  As just illustrated, the general function to read a dataset ends in
    `.oce`, and the name is a signal that the returned object is of
    class `oce`. Depending on the file contents, `d` will also have an
    additional class, e.g. if the file contains CTD data, then the
    object would inherit from two classes, `oce` and `ctd`, with the
    second being used to tailor the graphics by passing control to the
    `ctd` variant of the generic `plot` function (use
    `help("plot,ctd-method")` to learn more).

2.  Generally, `oce` functions employ a "camel case" naming convention,
    in which a function that is described by several words is named by
    stringing the words together, capitalizing the first letter of
    second and subsequent words. For example, `ctdFindProfiles()`
    locates individual profiles within a `ctd` object that holds
    data acquired during a cyclic raising and lowering of the
    CTD instrument.

3.  Function names begin with `oce` in cases where a more natural name
    would be in conflict with a function in the base R system or a
    package commonly used by Oceanographers. For example, `oceContour()`
    is a function that provides an alternative to `contour()` in the
    `graphics` package.

# Calculation of seawater properties

The `oce` package provides many functions for dealing with seawater
properties. Perhaps the most used is `swRho(S,T,p)`, which computes
seawater density $\rho=\rho(S, T, p)$, where $S$ is salinity, $T$ is
*in-situ* temperature in $^\circ$C (on the ITS-90 scale), and $p$ is
seawater pressure, i.e. the excess over atmospheric pressure, in dbar.
(This and similar functions starts with the letters `sw` to designate
that they relate to seawater properties.) The result is a number in the
order of $1000$kg/m$^3$. For many purposes, Oceanographers prefer to use
the density anomaly $\sigma=\rho-1000$kg/m$^3$, provided with
`swSigma(salinity,temperature,pressure)`, or its adiabatic cousin
$\sigma_\theta$, provided with `swSigmaTheta()`.

Most of the functions can use either the UNESCO or GSW (TEOS-10)
formulation of seawater properties, with the choice set by an argument
called `eos`. It should be noted that `oce` uses the `gsw` package for
GSW calculations. **Caution:** the results obtained with `eos="gsw"` in
`oce` functions may differ from the results obtained when using the
`gsw` functions directly, due to unit conventions. For example,
`swCSTp(..., eos="gsw")` reports conductivity ratio for consistency with
the UNESCO formulation, however the underlying `gsw` function
`gsw::gsw_C_from_SP()` reports conductivity in mS/cm.

A partial list of seawater functions is as follows:

-   `swAlpha()` for thermal expansion coefficient
    $\alpha=-\rho_0^{-1}\partial\rho/\partial T$
-   `swAlphaOverBeta()` for $\alpha/\beta$
-   `swBeta()` for haline compression coefficient
    $\beta=\rho_0^{-1}\partial\rho/\partial S$
-   `swConductivity()` for conductivity from $S$, $T$ and $p$
-   `swDepth()` for depth from $p$ and latitude
-   `swDynamicHeight()` for dynamic height
-   `swLapseRate()` for adiabatic lapse rate
-   `swN2()` for buoyancy frequency
-   `swRho()` for density $\rho$ from $S$, $T$ and $p$
-   `swSCTp()` for salinity from conductivity, temperature and pressure
-   `swSTrho()` for salinity from temperature and density
-   `swSigma()` for $\rho-1000$,kg/m$^3$
-   `swSigmaT()` for $\sigma$ with $p$ set to zero and temperature
    unaltered
-   `swSigmaTheta()` for$\sigma$ with $p$ set to zero and temperature
    altered adiabatically
-   `swSoundSpeed()` for speed of sound
-   `swSpecificHeat()` for specific heat
-   `swSpice()` for a quantity used in double-diffusive research
-   `swTFreeze()` for freezing temperature
-   `swTSrho()` for temperature from salinity and density
-   `swTheta()` for potential temperature $\theta$
-   `swViscosity()` for viscosity

Details and examples are provided in the documentation of these
functions.

The following exercise may be of help to readers who prefer to learn by
doing. (Answers are provided at the end of this document.)

**Exercise 1.** a. What is the density of a seawater parcel at pressure
100 dbar, with salinity 34 PSU and temperature 10$^\circ$C? b. What
temperature would the parcel have if raised adiabatically to the
surface? c. What density would it have if raised adiabatically to the
surface? d. What density would it have if lowered about 100m, increasing
the pressure to 200 dbar? e. Draw a blank $T$-$S$ diagram with $S$ from
30 to 40 PSU and $T$ from -2 to 20$^\circ$C.

# CTD data

The `read.oce` function recognizes a wide variety of CTD data formats,
and the associated `plot` function can produce many types of graphical
display. In addition, there are several functions that aid in the
analysis of such data. See the [ctd vignette](B_ctd.html) for more on
dealing with CTD data.

# Section plots

The commands

```{r eval=FALSE}
data(section)
plot(section, which=c(1, 2, 3, 99))
```

will plot a summary diagram containing sections of $T$, $S$, and
$\sigma_\theta$, along with a chart indicating station locations. In
addition to such overview diagrams, the `section` variant of the generic
`plot` function can also create individual plots of individual
properties (use `help("plot,section-method")` to learn more).

Some section datasets are supplied in a pre-gridded format, but it is
also common to have different pressure levels at each station. For such
cases, `sectionGrid()` may be used, e.g. the following produces Figure
4. The ship was travelling westward from the Mediterranean towards North
America, taking 124 stations in total; the `stationId` value selects the
last few stations of the section, during which the ship headed toward
the northwest, crossing isobaths (and perhaps, the Gulf Stream) at
nearly right angles.

```{r fig.cap="**Figure 4.** Portion of the CTD section designated A03, showing the Gulf Sream region.  The square on the cruise track corresponds to zero distance on the section.", fig.width=4, fig.height=4, dpi=72, dev.args=list(pointsize=12)}
library(oce)
data(section)
GS <- subset(section, 102 <= stationId & stationId <= 124)
GSg <- sectionGrid(GS, p=seq(0, 1600, 25))
plot(GSg, which=c(1,99), map.xlim=c(-85,-(64+13/60)))
```

**Exercise 2.** Draw a $T$-$S$ diagram for the section data, using black
symbols to the east of 30W and gray symbols to the west, thus
highlighting Mediterranean-derived waters. Use `handleFlags()` (see
[using data-quality flags](E_flags.html)) to discard questionable data,
and use the accessor function `[[`.

**Exercise 3.** Plot dynamic height and geostrophic velocity across the
Gulf Stream. (Hint: use the `swDynamicHeight()` function.)

# Maps

Oce has several functions that facilitate the drawing of maps. A variety
of projections are provided, with the mathematics of projection being
handled behind the scenes with the `sf` package. An introduction to
drawing maps is provided with `?mapPlot`, and the [map projection
vignette](D_map_projections.html) provides much more detail.

# Sea-level data

The following code graphs a built-in dataset of sea-level time series
(Figure 9). The top panel provides an overview of the entire data set.
The second panel is narrowed to the most recent month, which should
reveal spring-neap cycles if the tide is mixed. The third panel is a log
spectrum, with a few tidal constituents indicated. The `section` variant
of the generic `plot` function provides other possibilities for
plotting, including a cumulative spectrum that can be quite informative
(use `help("plot,sealevel-method")` to learn more).

```{r fig.cap="**Figure 9.** Sea-level timeseries measured in 2003 in Halifax Harbour.", fig.width=5, fig.height=4, dpi=72, dev.args=list(pointsize=16)}
library(oce)
data(sealevel)
plot(sealevel)
```

**Exercise 4.** Illustrate Halifax sea-level variations during Hurricane
Juan, near the end of September, 2003.

A preliminary version of tidal analysis is provided by the `tidem`
function provided in this version of the package, but readers are
cautioned that the results are certain to change in a future version.
(The problems involve phase and the inference of satellite nodes.)

**Exercise 5.** Plot the de-tided Halifax sea level during Autumn 2003,
to see whether Hurricane Juan is visible. (Hint: use `predict` with the
results from `tidem`.)

# Acoustic Doppler Profiler data

The following commands produce Figure 10, showing one velocity component
of currents measured in the St Lawrence Estuary Internal Wave
Experiment. This plot type is just one of many provided by the `adp`
variant of the generic `plot` function (see `?"plot,adp-method"`). See
[the adp vignette](C_adp.html) for much more on acoustic Doppler profiler
data.

```{r fig.cap="**Figure 10.** Measurements made with a bottom-mounted ADP in the St Lawrence Estuary. The line near the surface indicates pressure measured by the ADP.", fig.width=4, fig.height=2, dpi=72}
library(oce)
data(adp)
plot(adp, which=1)
lines(adp[['time']], adp[['pressure']], lwd=2)
```

# Handling data-quality flags

Archives of CTD/bottle and Argo drifter measurements commonly supply
data-quality flags that provide an indication of the trust to be put in
individual data points. Oce has a flexible scheme for dealing with such
flags, and also for inserting flags into these or any data type; see the
[using data-quality flags vignette](E_flags.html) for more..

# Working in non-English languages

Many of the `oce` plotting functions produce axis labels that can be
displayed in languages other than English. At the time of writing,
French, German, Spanish, and Mandarin are supported in at least a
rudimentary way. Setting the language can be done at the general system
level, or within R, as indicated below (results not shown).

```{r fig.keep="none"}
library(oce)
Sys.setenv(LANGUAGE="fr")
data(ctd)
plot(ctd)
```

```{r echo=FALSE}
Sys.setenv(LANGUAGE="en")
```

Most of the translated items were found by online dictionaries, and so
they may be incorrect for oceanographic usage. Readers can help out in
the translation effort, if they have knowledge of how nautical words
such as *Pitch* and *Roll* and technical terms such as *Absolute
Salinity* and *Potential Temperature* should be written in various
languages.

# Extending oce

The oce object structure can be used as a basis for new object types.
This has the advantage the basic operations of oce will carry over to
the new types. For example, the accessing operator `[[` and `summary`
function will work as expected, and so will such aspects as the handling
of data-quality flags and units. More details on setting up classes that
inherit from oce are provided in [the subclassing
vignette](F_subclassing.html).

# The future of oce

The present version of `oce` can only handle data types that the authors
have been using in their research. New data types will be added as the
need arises in that work, but the authors would also be happy to add
other data types that are likely to prove useful to the Oceanographic
community. (The data types need not be restricted to Physical
Oceanography, but the authors will need some help in dealing with other
types of data, given their research focus.)

Two principles will guide the addition of data types and functions: (a)
the need, as perceived by the authors or by other contributors and (b)
the ease with which the additions can be made. One might call this
development-by-triage, by analogy to the scheme used in Emergency Rooms
to organize medical effort efficiently.

# Development website

The site <https://github.com/dankelley/oce> provides a window on the
development that goes on between the CRAN releases of the package.
Readers are requested to visit the site to report bugs, to suggest new
features, or just to see how `oce` development is coming along. Note
that the `development` branch is used by the authors in their work, and
is updated so frequently that it must be considered unstable, at least
in those spots being changed on a given day. Official CRAN releases
derive from the `master` branch, and are done when the code is
considered to be of reasonable stability and quality. This is all in a
common pattern for open-source software.

# Solutions to exercises

**Exercise 1.** Seawater properties. In the UNESCO system we may write

```{r fig.width=5, fig.height=5, fig.keep="none"}
library(oce)
swRho(34, 10, 100, eos="unesco")
swTheta(34, 10, 100, eos="unesco")
swRho(34, swTheta(34, 10, 100, eos="unesco"), 0, eos="unesco")
swRho(34, swTheta(34, 10, 100, 200, eos="unesco"), 200, eos="unesco")
plotTS(as.ctd(c(30,40),c(-2,20),rep(0,2)), eos="unesco", grid=TRUE, col="white")
```

and in the Gibbs SeaWater system, we use `eos="gws"` and must supply
`longitude` and `latitude` arguments to the `sw*()` calls and also to
the `as.ctd()` call.

**Exercise 2.** Draw a $T$-$S$ diagram for the section data, using black
symbols to the east of 30W and gray symbols to the west, thus
highlighting Mediterranean-derived waters. Use `handleFlags()` (see
[using data-quality flags](E_flags.html)) to discard questionable data,
and use the accessor function `[[`.

We will use the Gibbs SeaWater system, so `as.ctd()` needs location
information.

```{r fig.width=5, fig.height=5, fig.keep="none"}
library(oce)
data(section)
s <- handleFlags(section, flags=list(c(1, 3:9)))
ctd <- as.ctd(s[["salinity"]], s[["temperature"]], s[["pressure"]],
              longitude=s[["longitude"]], latitude=s[["latitude"]])
col <- ifelse(s[["longitude"]] > -30, "black", "gray")
plotTS(ctd, col=col, eos="gsw")
```

**Exercise 3.** Plot dynamic height and geostrophic velocity across the
Gulf Stream. (Hint: use the `swDynamicHeight` function.)

(Try `?swDynamicHeight` for hints on smoothing.)

```{r fig.width=5, fig.height=3, fig.keep="none"}
library(oce)
data(section)
GS <- subset(section, 102 <= stationId & stationId <= 124)
dh <- swDynamicHeight(GS)
par(mfrow=c(2,1), mar=c(3, 3, 1, 1), mgp=c(2, 0.7, 0))
plot(dh$distance, dh$height, type="l", xlab="", ylab="Dyn. Height [m]")
grid()
# 1e3 metres per kilometre
latMean <- mean(GS[["latitude"]])
f <- coriolis(latMean)
g <- gravity(latMean)
v <- diff(dh$height)/diff(dh$distance) * g / f / 1e3
plot(dh$distance[-1], v, type="l", xlab="Distance [km]", ylab="Velocity [m/s]")
grid()
abline(h=0, col='red')
```

**Exercise 4.** Halifax sea-level during Hurricane Juan, near the end of
September, 2003.

A web search will tell you that Hurricane Juan hit about midnight,
2003-sep-28. The first author can verify that the strongest winds
occurred a bit after midnight -- that was the time he moved to a room
without windows, in fear of flying glass.

```{r fig.width=7, fig.height=3, fig.keep="none"}
library(oce)
data(sealevel)
# Focus on 2003-Sep-28 to 29th, the time when Hurricane Juan caused flooding
plot(sealevel,which=1,xlim=as.POSIXct(c("2003-09-24","2003-10-05"), tz="UTC"))
abline(v=as.POSIXct("2003-09-29 04:00:00", tz="UTC"), col="red")
mtext("Juan", at=as.POSIXct("2003-09-29 04:00:00", tz="UTC"), col="red")
```

**Exercise 5.** Plot the de-tided Halifax sea level during Autumn 2003,
to see whether Hurricane Juan is visible. (Hint: use `predict` with the
results from `tidem`.)

```{r results="hide", fig.keep="none"}
library(oce)
data(sealevel)
m <- tidem(sealevel)
oce.plot.ts(sealevel[['time']], sealevel[['elevation']] - predict(m),
            ylab="Detided sealevel [m]",
            xlim=c(as.POSIXct("2003-09-20"), as.POSIXct("2003-10-08")))
```

The spike reveals a surge of about 1.5m, on the 29th of September, 2003.
---
title: "6. Subclassing oce objects"
author: "Dan Kelley (https://orcid.org/0000-0001-7808-5911)"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_vignette:
    toc: true
    number_sections: true
    fig_caption: yes
    fig_width: 4
    fig_height: 4
    dpi: 72
    dev.args: list(pointsize=11)
vignette: >
  %\VignetteIndexEntry{6. Subclassing oce objects}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

<!-- HOW TO BUILD THE VIGNETTE. -->
<!-- Look in oce.Rmd for instructions. -->


```{r, echo=FALSE}
knitr::opts_chunk$set(collapse=TRUE, comment="#>")
```

**Abstract.** This vignette explains how to create new object classes, using
the low-level `oce` class as a foundation. This scheme is beneficial, because
the newly-formed objects automatically gain the key properties of oce objects,
in terms of operators such as `[[` and `[[<-`, functions such as `subset()` and
`summary()`, schemes for handling units, etc. The `setMethod()` function is
central to the process of creating new classes, as will be illustrated here in
a practical way, using drifting-buoy data for example.


# A drifting buoy

Data from buoy number 4201703 were downloaded from
http://www.coriolis.eu.org/Data-Products/Data-Delivery/Data-selection on
2022-01-09, by deselecting all data types except for "Drifting buoy", selecting
a data window south of Nova Scotia and of area similar to that province,
expanding the tarball created by the website and then focussing on the
downloaded file named `"drifting-buoys-4201703.csv`".

There are many data in the source data file, but only a subset is used here:
starting with 300 samples of just five of the data fields: time (called `t` for
brevity), longitude (`lon`), latitude (`lat`), temperature (`Tatm`) and
atmospheric pressure (`Patm`). This set was further trimmed by removing data
ensembles in which temperature was not reported.  The result was 146
observations of 5 variables. (Note that they are *not* spaced uniformly in
time, a feature that is of importance only at the end of this document.)

The data were combined into a data frame named `db` and packaged in an RDA
file, an overview of which is provided as follows.
```{r}
library(oce)
load(system.file("extdata", "drifter.rda", package="oce"))
str(db)
```

Before showing how to create a new subclass, it will first be shown that such
data could also be stored in the foundational class of the package.

# Using the base oce class

The base class is named `"oce"`, and so a new object of that type can be created with
```{r}
o <- new("oce")
```

At this point, the `summary()` method does not reveal much:
```{r}
summary(o)
```
but if we populate the object with some data and metadata
```{r}
o <- oceSetData(o, "time", db$t)
o <- oceSetData(o, "longitude", db$lon)
o <- oceSetData(o, "latitude", db$lat)
o <- oceSetMetadata(o, "ID", 4201703)
```
then the summary is more informative
```{r}
summary(o)
```

Note, however, that the `ID` is not listed by `summary()`.  This is because the
general form of that function is set up to focus on the contents of the `data`
slot, not the `metadata` slot. (Extending `summary()` will described later.)

Objects of `oce-class` automatically have other methods, besides
`summary()`.  For example, the accessing operator `[[` works for the
object that was just created:
```{r}
str(o[["latitude"]])
```

We can also use the default `plot()` function to show the interdependence of
entries in the `data` slot.
```{r fig.cap="**Figure 1.** Demonstration of base-level plot()."}
plot(o)
```

Although this plot provides a useful overview of the data, it is not something
an analyst would consider publishing.  To get more useful plot types, we must
create a new subclass that has a specialized `plot()` method.

# Creating a new subclass

A reasonable name for our new subclass might be `"drifter"`.  With this
choice, the class can be defined in a single line of code:
```{r}
drifter <- setClass(Class="drifter", contains="oce")
```

Once this has been executed, the user can create a new object of this subclass
with
```{r}
d <- new("drifter")
```

Although it is simple to add data and metadata as in the previous section, it
is more convenient to supply this information when `new()` is used.  To
accomplish this, we must set up an initialization function, as follows.

```{r}
setMethod(f="initialize",
  signature="drifter",
  definition=function(.Object, time, longitude, latitude, ID="unknown") {
      if (missing(time))
          stop("In new(drifter) : must provide 'time'", call.=FALSE)
      if (missing(longitude))
          stop("In new(drifter) : must provide 'longitude'", call.=FALSE)
      if (missing(latitude))
          stop("In new(drifter) : must provide 'latitude'", call.=FALSE)
      .Object@data$time <- time
      .Object@data$longitude <- longitude
      .Object@data$latitude <- latitude
      .Object@metadata$ID <- ID
      .Object@processingLog$time <- presentTime()
      .Object@processingLog$value <- "create 'drifter' object"
      return(.Object)
})
```

This method takes away the choice of where to put things (in `data` or in
`metadata`), saving users from having to make that decision, possibly breaking
other code.  Also, note that some error checking has been used here, so that an
attempt to use `new("drifter")` without providing time, etc., produces an error
message, as shown below.
```
d <- new("drifter")
Error: In new(drifter) : must provide 'time'
```


With this setup, we may now create an object that we can use for the rest of
this vignette:
```{r}
d <- new("drifter", time=db$t, longitude=db$lon, latitude=db$lat, ID=4201703)
```
and this will be used in the next three subsections, showing how to specialize
the `plot()` method, the `summary()` method, and the `[[` accessor. Note the
use of `setMethod()` in each instance.

## Specializing the plot() function

The `f`, `signature` and `definition` arguments follow the same pattern as in
the previous example, and readers ought to focus on last of these. Here, three
plot types are provided, as selected by an argument called `which`. The first
two give time-series plots of drifter longitude and latitude, and the third
shows the trajectory.

```{r}
setMethod(f="plot",
    signature=signature("drifter"),
    definition=function(x, which=1, ...) {
        lonlab <- expression("Longitude ["*degree*"E]")
        latlab <- expression("Latitude ["*degree*"N]")
        if (which == 1) {
            oce.plot.ts(x[["time"]], x[["longitude"]],
                ylab=lonlab, ...)
        } else if (which == 2) {
            oce.plot.ts(x[["time"]], x[["latitude"]],
                ylab=latlab,...)
        } else if (which == 3) {
            asp <- 1 / cos(mean(range(x[["latitude"]]) * pi / 180))
            plot(x[["longitude"]], x[["latitude"]],
                asp=asp, xlab=lonlab, ylab=latlab, ...)
        } else {
            stop("In plot,drifter-method : try which=1, 2 or 3", call.=FALSE)
        }
    })
```

The three plot types are fairly rudimentary, but the `...` argument can be used
for customization, as in the following example.

```{r fig.cap="**Figure 2.** Demonstration of specialized plot()."}
par(mar=c(3.3,3.3,1,1), mgp=c(2,0.7,0))
layout(matrix(c(1,3,2,3),nrow=2,byrow=TRUE))
plot(d, which=1, drawTimeRange=FALSE)
plot(d, which=2, drawTimeRange=FALSE)
plot(d, which=3)
```


## Specializing the summary() function

Here, our sole purpose is to add an indication of the ID of the drifter. The
rest of the information provided by the base-level `summary()` function will
suffice to summarize the rest.

So, with
```{r}
setMethod(f="summary",
    signature="drifter",
    definition=function(object, ...) {
        cat("CTD Summary\n-----------\n\n")
        cat("* ID:          ", object[["ID"]], "\n", sep="")
        invisible(callNextMethod())
    })
```

An example follows.
```{r}
summary(d)
```



## Specializing the `[[` accessor

The variation of drifter position provides information on ocean velocity, and
this is of sufficient interest that it might be worth adding to `[[`.  This may
be done as follows.

```{r}
setMethod(f="[[",
    signature(x="drifter", i="ANY", j="ANY"),
    definition=function(x, i, j, ...) {
        if (i == "velocity") {
            D <- function(x) {
                dx <- diff(x)
                c(dx[1], dx)
            }
            dist <- geodDist(x[["longitude"]], x[["latitude"]])
            lat <- x[["latitude"]]
            lon <- x[["longitude"]]
            dt <- D(as.numeric(x[["time"]])) # seconds
            scalex <- 111.12e3         # m per degree latitude
            scaley <- scalex * cos(lat*pi/180)
            u <- scalex * D(lon) / dt
            v <- scaley * D(lat) / dt
            list(u=u, v=v)
        } else {
            callNextMethod()
        }
    })
```

As a test, we can plot the velocity components.
```{r fig.cap="**Figure 3.** Velocities inferred from drifter motion."}
uv <- d[["velocity"]]
par(mfrow=c(2,1))
oce.plot.ts(d[["time"]], uv$u, ylab="Eastward velocity [m/s]", grid=TRUE)
oce.plot.ts(d[["time"]], uv$v, ylab="Northward velocity [m/s]", grid=TRUE)
```

The `acf()` function, which computes lagged autocorrelation, can shed some
light on the periodicity of the velocity signals (but see Exercise 3), as
follows.

```{r fig.cap="**Figure 4.** Autocorrelation analysis of drifter velocities, showing also the M2 period."}
data(tidedata)
M2period <- 1 / with(tidedata$const, freq[[which(name=="M2")]])
uv <- d[["velocity"]]
par(mfrow=c(2,1), mar=c(3,3,1,1), mgp=c(2,0.7,0))
acf(uv$u, main="", ylab="u ACF")
abline(v=M2period, col=2)
acf(uv$v, main="", ylab="v ACF")
abline(v=M2period, col=2)
```


# Summary

As has been illustrated above, the `setMethod()` function is the key to
altering generic functions such as `plot()`, `summary()` and `[[`. The details
of how this works are explained in numerous manuals, books, and websites, but
impatient readers can save some time by simply copying and expanding on the
code provided in the previous sections.

# Appendix: exercises for the reader

1. Add atmospheric temperature and pressure (both contained in the source RDA
   file) to the `drifter` object initialization function.  Be sure to check
   that they were provided or, if you choose, let them be optional.

2. Think of other useful plot types, and add these to the `plot()` method.

3. For the final lagged-autocorrelation function, use `approx()` to get new
   time-series with a constant sampling rate.

4. Again for the lagged-autocorrelation function, add a line for local Coriolis
   period.

5. As an alternative to the lagged-autocorrelation method, test for periodicity
   of velocity by using `nls()`, supplied with a sinusoidal function that has
   adjustable amplitude, phase, and frequency.

% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{labelWithUnit}
\alias{labelWithUnit}
\title{Create label with unit}
\usage{
labelWithUnit(name, unit = NULL)
}
\arguments{
\item{name}{character value naming a quantity.}

\item{unit}{a list containing items \code{unit} and (optionally) \code{scale}, only the
first of which, an \code{\link[=expression]{expression()}}, is used.  If \code{unit} is not provided,
then a default will be used (see \dQuote{Details}).}
}
\value{
\code{labelWithUnit} returns a language object, created with \code{\link[=bquote]{bquote()}},
that that may supplied as a text string to \code{\link[=legend]{legend()}}, \code{\link[=mtext]{mtext()}}, \code{\link[=text]{text()}},
etc.
}
\description{
\code{labelWithUnit} creates a label with a unit, for graphical
display, e.g. by \code{\link{plot,section-method}}.
The unit is enclosed in square brackets, although setting
\code{options(oceUnitBracket="(")} will cause parentheses to be
used, instead.
}
\details{
If \code{name} is in a standard list, then alterations are made as appropriate,
e.g. \code{"SA"} or \code{"Absolute Salinity"} yields an S with subscript A; \code{"CT"} or
\code{"Conservative Temperature"} yields an upper-case Theta, \code{sigmaTheta}
yields a sigma with subscript theta, \code{sigma0} yields
sigma with subscript 0 (with similar for 1 through 4), \code{"N2"} yields "N" with
superscript 2, and \code{"pressure"} yields "p".
These basic hydrographic quantities have default units that will
be used if \code{unit} is not supplied (see \dQuote{Examples}).

In addition to the above, several chemical names are recognized,
but no unit is guessed for them, because the oceanographic
community lacks agreed-upon standards.

If \code{name} is not recognized, then it is simply repeated in the
return value.
}
\examples{
library(oce)
# 1. temperature has a predefined unit, but this can be overruled
labelWithUnit("temperature")
labelWithUnit("temperature",
    list(unit=expression(m/s), scale="erroneous"))
# 2. phosphate lacks a predefined unit
labelWithUnit("phosphate")
data(section)
labelWithUnit("phosphate",
    section[["station",1]][["phosphateUnit"]])

}
\seealso{
Other functions that create labels: 
\code{\link{resizableLabel}()}
}
\author{
Dan Kelley
}
\concept{functions that create labels}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/tides.R
\name{[[,tidem-method}
\alias{[[,tidem-method}
\title{Extract Something From a tidem Object}
\usage{
\S4method{[[}{tidem}(x, i, j, ...)
}
\arguments{
\item{x}{a \linkS4class{tidem} object.}

\item{i}{character value indicating the name of an item to extract.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{ignored.}
}
\description{
Generally, the \code{[[} method lets users extract information from \code{oce}
objects, without having to know the details of the internal storage.  For
many \code{oce} sub-classes, \code{[[} can also return quantities that are computed
from the object's contents.
}
\details{
A two-step process is used to try to find the requested information. First, a
class-specific function is used (see \dQuote{Details of the Specialized
Method}).  If this yields nothing, then a general method is used (see
\dQuote{Details of the General Method}).  If both methods fail, then \code{[[}
returns NULL.

Some understanding of the subclass is required to know what can be retrieved
with \code{[[}. When dealing with an unfamiliar subclass, it can be useful to
first use \code{x[["?"]]} to get a listing of the retrievable items. See
\dQuote{Details of the Specialized Method} for more information.
}
\section{Details of the Specialized Method}{

\itemize{
\item If \code{i} is \code{"?"}, then the return value is a list
containing four items, each of which is a character vector
holding the names of things that can be accessed with \code{[[}.
The \code{data} and \code{metadata} items hold the names of
entries in the object's data and metadata
slots, respectively. The
\code{metadataDerived} holds only \code{""}, because
no derived metadata values are defined for \code{cm} objects.
\item If \code{i} is \code{"frequency"} or \code{"freq"}, then a vector of
constituent frequencies (stored
as \code{freq} in the data slot) is returned.
\item If \code{i} is \code{"amplitude"} then a vector of constituent amplitudes
is returned.
\item If \code{i} is \code{"phase"} then a vector of constituent phases
is returned.
}
}

\section{Details of the General Method}{


Note: the text of this section is identical for all \code{oce} subclasses, and so
some of what you read here may not be relevant to the class being described
in this help page.

If the specialized method produces no matches, the following generalized
method is applied. As with the specialized method, the procedure hinges first
on the values of \code{i} and, optionally, \code{j}. The work proceeds in steps, by
testing a sequence of possible conditions in sequence.
\enumerate{
\item A check is made as to whether \code{i} names one of the standard \code{oce} slots.
If so, \code{[[} returns the slot contents of that slot.  Thus, \code{x[["metadata"]]}
will retrieve the \code{metadata} slot, while \code{x[["data"]]} and
\code{x[["processingLog"]]} return those slots.
\item If \code{i} is a string ending in the \code{"Unit"}, then the characters preceding
that string are taken to be the name of an item in the data object, and a
list containing the unit is returned (or \code{NULL} if there is no such unit).
This list consists of an item named \code{unit}, which is an \code{\link[=expression]{expression()}}, and
an item named \code{scale}, which is a string describing the measurement scale.
If the string ends in \code{" unit"}, e.g.  \code{x[["temperature unit"]]} (note the
space), then just the expression is returned, and if it ends in \code{" scale"},
then just the scale is returned.
\item If \code{i} is a string ending in \code{"Flag"}, then the corresponding data-quality
flag is returned (or \code{NULL} if there is no such flag).
\item If the object holds hydrographic information (pressure, salinity,
temperature, longitude and latitude) then another set of possibilities
arises.  If \code{i} is \code{"sigmaTheta"}, then the value of \code{\link[=swSigmaTheta]{swSigmaTheta()}} is
called with \code{x} as the sole argument, and the results are returned.
Similarly, \code{\link[=swSigma0]{swSigma0()}} is used if \code{i="sigma0"}, and \code{\link[=swSpice]{swSpice()}} is used if
\code{i="spice"}. Of course, these actions only make sense for objects that
contain the relevant items within their \code{data} slot.
\item After these possibilities are eliminated, the action depends on whether
\code{j} has been provided.  If \code{j} is not provided, or is the string \code{""}, then
\code{i} is sought in the \code{metadata} slot, and then in the \code{data} slot, returning
whichever is found first.  In other words, if \code{j} is not provided, the
\code{metadata} slot takes preference over the \code{data} slot. However, if \code{j} is
provided, then it must be either the string \code{"metadata"} or \code{"data"}, and it
directs where to look.
\item If none of the above-listed conditions holds, then \code{NULL} is returned.
}
}

\seealso{
Other functions that extract parts of oce objects: 
\code{\link{[[,adp-method}},
\code{\link{[[,adv-method}},
\code{\link{[[,amsr-method}},
\code{\link{[[,argo-method}},
\code{\link{[[,bremen-method}},
\code{\link{[[,cm-method}},
\code{\link{[[,coastline-method}},
\code{\link{[[,ctd-method}},
\code{\link{[[,echosounder-method}},
\code{\link{[[,g1sst-method}},
\code{\link{[[,gps-method}},
\code{\link{[[,ladp-method}},
\code{\link{[[,landsat-method}},
\code{\link{[[,lisst-method}},
\code{\link{[[,lobo-method}},
\code{\link{[[,met-method}},
\code{\link{[[,oce-method}},
\code{\link{[[,odf-method}},
\code{\link{[[,rsk-method}},
\code{\link{[[,sealevel-method}},
\code{\link{[[,section-method}},
\code{\link{[[,topo-method}},
\code{\link{[[,windrose-method}},
\code{\link{[[,xbt-method}},
\code{\link{[[<-,adv-method}}

Other things related to tides: 
\code{\link{[[<-,tidem-method}},
\code{\link{as.tidem}()},
\code{\link{plot,tidem-method}},
\code{\link{predict.tidem}()},
\code{\link{summary,tidem-method}},
\code{\link{tidalCurrent}},
\code{\link{tidedata}},
\code{\link{tidem-class}},
\code{\link{tidemAstron}()},
\code{\link{tidemVuf}()},
\code{\link{tidem}},
\code{\link{webtide}()}
}
\concept{functions that extract parts of oce objects}
\concept{things related to tides}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/argo.R
\name{argoJuldToTime}
\alias{argoJuldToTime}
\title{Convert Argo Julian Day (juld) to time}
\usage{
argoJuldToTime(jday)
}
\arguments{
\item{jday}{A numerical value indicating the julian day in the Argo convention,
with day=0 at 1950-01-01.}
}
\description{
Convert Argo Julian Day (juld) to time
}
\examples{
argoJuldToTime(25749)

}
\author{
Jaimie Harbin and Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/odf.R
\name{ODF2oce}
\alias{ODF2oce}
\title{Create ODF object from the output of \code{ODF::read_ODF()}.}
\usage{
ODF2oce(ODF, coerce = TRUE, debug = getOption("oceDebug"))
}
\arguments{
\item{ODF}{A list as returned by \code{read_ODF} in the \code{ODF} package}

\item{coerce}{A logical value indicating whether to coerce the return value
to an appropriate object type, if possible.}

\item{debug}{a flag that turns on debugging.  Set to 1 to get a moderate
amount of debugging information, or to 2 to get more.}
}
\value{
An oce object, possibly coerced to a subtype.
}
\description{
As of August 11, 2015, \code{ODF::read_ODF} returns a list with 9 elements,
one named \code{DATA}, which is a \code{\link[=data.frame]{data.frame()}} containing the
columnar data, the others being headers of various sorts.  The present
function constructs an oce object from such data, facilitating processing
and plotting with the general oce functions.
This involves storing the 8 headers verbatim in the
\code{odfHeaders} in the \code{metadata} slot, and also
copying some of the header
information into more standard names (e.g.  \code{metadata@longitude} is a
copy of \code{metadata@odfHeader$EVENT_HEADER$INITIAL_LATITUDE}).  As for
the \code{DATA}, they are stored in the \code{data} slot, after renaming
from ODF to oce convention using \code{\link[=ODFNames2oceNames]{ODFNames2oceNames()}}.
}
\section{Caution}{
 This function may change as the \code{ODF} package
changes.  Since \code{ODF} has not been released yet, this should not affect
any users except those involved in the development of \code{oce} and
\code{ODF}.
}

\seealso{
Other things related to odf data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{ODFListFromHeader}()},
\code{\link{ODFNames2oceNames}()},
\code{\link{[[,odf-method}},
\code{\link{[[<-,odf-method}},
\code{\link{odf-class}},
\code{\link{plot,odf-method}},
\code{\link{read.ctd.odf}()},
\code{\link{read.odf}()},
\code{\link{subset,odf-method}},
\code{\link{summary,odf-method}}
}
\author{
Dan Kelley
}
\concept{things related to odf data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/rsk.R
\name{[[<-,rsk-method}
\alias{[[<-,rsk-method}
\title{Replace Parts of a Rsk Object}
\usage{
\S4method{[[}{rsk}(x, i, j, ...) <- value
}
\arguments{
\item{x}{an \linkS4class{rsk} object.}

\item{i}{character value naming the item to replace.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{optional additional information (ignored).}

\item{value}{The value to be placed into \code{x}, somewhere.}
}
\description{
The \code{[[<-} method works for all \linkS4class{oce} objects.
The purpose, as with the related extraction method, \code{[[},
is to insulate users from the internal details of \linkS4class{oce}
objects, by looking for items within the various storage
slots of the object. Items not actually stored can also be
replaced, including units and data-quality
flags.
}
\details{
As with \code{[[} method, the procedure works in steps.

First, the \code{metadata} slot of \code{x} is checked to
see whether it contains something named with \code{i}.
If so, then the named item is replaced with \code{value}.

Otherwise, if the string value of \code{i} ends in \code{Unit}, then the
characters preceding that are taken as the name of a variable, and
the \code{metadata} slot of \code{x} is updated to store that unit, e.g.\preformatted{x[["temperatureUnits"]] <- list(unit=expression(degree*F),scale="")
}

Similarly, if \code{i} ends in \code{Flag}, then quality-control
flags are set up as defined by \code{result}, e.g.\preformatted{o[["temperatureFlags"]] <- c(2,4,2,2)
}

Otherwise, \code{\link[=pmatch]{pmatch()}} is used for a partial-string match with
the names of the items that are in the \code{data} slot of \code{x}.
The first item found (if any) is then updated to hold the value \code{result}.

If none of these conditions is met, a warning is issued.
}
\seealso{
Other functions that replace parts of oce objects: 
\code{\link{[[<-,adp-method}},
\code{\link{[[<-,amsr-method}},
\code{\link{[[<-,argo-method}},
\code{\link{[[<-,bremen-method}},
\code{\link{[[<-,cm-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{[[<-,echosounder-method}},
\code{\link{[[<-,g1sst-method}},
\code{\link{[[<-,gps-method}},
\code{\link{[[<-,ladp-method}},
\code{\link{[[<-,landsat-method}},
\code{\link{[[<-,lisst-method}},
\code{\link{[[<-,lobo-method}},
\code{\link{[[<-,met-method}},
\code{\link{[[<-,oce-method}},
\code{\link{[[<-,odf-method}},
\code{\link{[[<-,sealevel-method}},
\code{\link{[[<-,section-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{[[<-,topo-method}},
\code{\link{[[<-,windrose-method}},
\code{\link{[[<-,xbt-method}}

Other things related to rsk data: 
\code{\link{[[,rsk-method}},
\code{\link{as.rsk}()},
\code{\link{plot,rsk-method}},
\code{\link{read.rsk}()},
\code{\link{rsk-class}},
\code{\link{rskPatm}()},
\code{\link{rskToc}()},
\code{\link{rsk}},
\code{\link{subset,rsk-method}},
\code{\link{summary,rsk-method}}
}
\concept{functions that replace parts of oce objects}
\concept{things related to rsk data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{oceColorsTurbidity}
\alias{oceColorsTurbidity}
\alias{oce.colorsTurbidity}
\title{Create colors suitable for turbidity fields}
\usage{
oceColorsTurbidity(n)
}
\arguments{
\item{n}{number of colors to create.}
}
\value{
A vector of color specifications.
}
\description{
Create a set of colors for displaying turbidity values,
based on the scheme devised by Thyng et al. (2016) and
presented in a python package by Thyng (2019).
The color specifications were transliterated from
python to R on 2015-09-29, but have not been adjusted
since, even though the python source has changed.  This
is to prevent breaking old \code{oce} code.  To get the
latest versions of these colours or other colours,
use the \CRANpkg{cmocean} R package
(Thyng, Richards, and Krylov, 2019) directly, as is
illustrated (with the "matter" scheme) in Example 2.
Note that the \CRANpkg{cmocean} core functions provide a way
to select between various versions of the colour schemes.
It is also worth considering the palettes provided by the
\CRANpkg{viridis} package, as illustrated (with the
"inferno" scheme) in Example 3.
}
\examples{
library(oce)

# Example 1
imagep(volcano, col=oceColorsTurbidity(128),
       zlab="oceColorsTurbidity")
\dontrun{
# Example 2 (requires the cmocean package)
imagep(volcano, col=cmocean::cmocean("matter"),
       zlab='cmocean::cmocean("matter")')}

\dontrun{
# Example 3 (requires the viridis package)
imagep(volcano, col=viridis::inferno,
       zlab='viridis::inferno')}

}
\references{
\itemize{
\item Thyng, Kristen, Chad Greene, Robert Hetland, Heather Zimmerle, and Steven DiMarco.
\dQuote{True Colors of Oceanography: Guidelines for Effective and Accurate Colormap Selection.}
Oceanography 29, no. 3 (September 1, 2016): 9‚Äì13.
\doi{10.5670/oceanog.2016.66}
\item Thyng, Kristen. Kthyng/Cmocean. Python, 2019.
\url{https://github.com/kthyng/cmocean}.
\item Thyng, Kristen, Clark Richards, and Ivan Krylov.
Cmocean: Beautiful Colour Maps for Oceanography (version 0.2), 2019.
\url{https://CRAN.R-project.org/package=cmocean}.
}

The following references provide information on choosing
colour schemes, that are suitable for viewers who have
colour deficiencies.

Light, Adam, and Patrick J. Bartlein. "The End of the Rainbow? Color
Schemes for Improved Data Graphics." \emph{Eos, Transactions American Geophysical
Union} 85, no. 40 (2004): 385.
DOI: 10.1029/2004EO400002

Stephenson, David B. "Comment on 'Color Schemes for Improved Data
Graphics', by A Light and P.J. Bartlein." \emph{Eos, Transactions American
Geophysical Union} 86, no.  20 (2005): 196.
DOI: 10.1029/2005EO200005

Light, Adam, and Patrick J. Bartlein. "Reply to 'Comment on Color
Schemes for Improved Data Graphics,' by A. Light and P.J.  Bartlein'."
\emph{Eos, Transactions American Geophysical Union} 86, no. 20 (2005): 196‚Äì196.
DOI: 10.1029/2005EO200006
}
\seealso{
Other things related to colors: 
\code{\link{colormapGMT}()},
\code{\link{colormap}()},
\code{\link{oceColors9B}()},
\code{\link{oceColorsCDOM}()},
\code{\link{oceColorsChlorophyll}()},
\code{\link{oceColorsClosure}()},
\code{\link{oceColorsDensity}()},
\code{\link{oceColorsFreesurface}()},
\code{\link{oceColorsGebco}()},
\code{\link{oceColorsJet}()},
\code{\link{oceColorsOxygen}()},
\code{\link{oceColorsPAR}()},
\code{\link{oceColorsPalette}()},
\code{\link{oceColorsPhase}()},
\code{\link{oceColorsSalinity}()},
\code{\link{oceColorsTemperature}()},
\code{\link{oceColorsTurbo}()},
\code{\link{oceColorsTwo}()},
\code{\link{oceColorsVelocity}()},
\code{\link{oceColorsViridis}()},
\code{\link{oceColorsVorticity}()},
\code{\link{ocecolors}}
}
\author{
Krysten M. Thyng (Python version), Dan Kelley (R transliteration)
}
\concept{things related to colors}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.R
\name{as.adp}
\alias{as.adp}
\title{Create an ADP Object}
\usage{
as.adp(
  time,
  distance,
  v,
  a = NULL,
  q = NULL,
  orientation = "upward",
  coordinate = "enu"
)
}
\arguments{
\item{time}{of observations in POSIXct format}

\item{distance}{to centre of bins}

\item{v}{array of velocities, with first index for time, second for bin number, and third for beam number}

\item{a}{amplitude, a \code{\link[=raw]{raw()}} array with dimensions matching \code{u}}

\item{q}{quality, a \code{\link[=raw]{raw()}} array with dimensions matching \code{u}}

\item{orientation}{a string indicating sensor orientation, e.g. \code{"upward"} and \code{"downward"}}

\item{coordinate}{a string indicating the coordinate system, \code{"enu"}, \code{"beam"}, \code{"xy"}, or \code{"other"}}
}
\value{
An \linkS4class{adp} object.
}
\description{
Create an ADP Object
}
\details{
Construct an \linkS4class{adp} object.  Only a basic
subset of the typical \code{data} slot is represented in the arguments
to this function, on the assumption that typical usage in reading data
is to set up a nearly-blank \linkS4class{adp} object, the \code{data}
slot of which is then inserted.  However, in some testing situations it
can be useful to set up artificial \code{adp} objects, so the other
arguments may be useful.
}
\examples{
data(adp)
t <- adp[["time"]]
d <- adp[["distance"]]
v <- adp[["v"]]
a <- as.adp(time=t, distance=d, v=v)
\donttest{
plot(a)
}

}
\seealso{
Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley
}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ladp.R
\name{plot,ladp-method}
\alias{plot,ladp-method}
\alias{plot.ladp}
\title{Plot an ladp Object}
\usage{
\S4method{plot}{ladp}(x, which = c("u", "v"), ...)
}
\arguments{
\item{x}{an \linkS4class{ladp} object.}

\item{which}{a character vector storing names of items to be plotted.}

\item{...}{Other arguments, passed to plotting functions.}
}
\description{
Uses \code{\link[=plotProfile]{plotProfile()}} to create panels of depth variation of easterly
and northerly velocity components.
}
\seealso{
Other things related to ladp data: 
\code{\link{[[,ladp-method}},
\code{\link{[[<-,ladp-method}},
\code{\link{as.ladp}()},
\code{\link{ladp-class}},
\code{\link{summary,ladp-method}}

Other functions that plot oce data: 
\code{\link{download.amsr}()},
\code{\link{plot,adp-method}},
\code{\link{plot,adv-method}},
\code{\link{plot,amsr-method}},
\code{\link{plot,argo-method}},
\code{\link{plot,bremen-method}},
\code{\link{plot,cm-method}},
\code{\link{plot,coastline-method}},
\code{\link{plot,ctd-method}},
\code{\link{plot,gps-method}},
\code{\link{plot,landsat-method}},
\code{\link{plot,lisst-method}},
\code{\link{plot,lobo-method}},
\code{\link{plot,met-method}},
\code{\link{plot,odf-method}},
\code{\link{plot,rsk-method}},
\code{\link{plot,satellite-method}},
\code{\link{plot,sealevel-method}},
\code{\link{plot,section-method}},
\code{\link{plot,tidem-method}},
\code{\link{plot,topo-method}},
\code{\link{plot,windrose-method}},
\code{\link{plot,xbt-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{tidem-class}}
}
\author{
Dan Kelley
}
\concept{functions that plot oce data}
\concept{things related to ladp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/tides.R
\docType{class}
\name{tidem-class}
\alias{tidem-class}
\title{Class to Store Tidal Models}
\description{
This class stores tidal-constituent models.
}
\section{Slots}{

\describe{
\item{\code{data}}{As with all \code{oce} objects, the \code{data} slot
for \code{tidem} objects is a
\link{list} containing the main data for the object.
{}}

\item{\code{metadata}}{As with all \code{oce} objects, the \code{metadata} slot
for \code{tidem} objects is a \link{list} containing
information about the \code{data} or about the object itself.
{}}

\item{\code{processingLog}}{As with all \code{oce} objects, the \code{processingLog} slot
for \code{tidem} objects is a
\link{list} with entries describing the creation and evolution
of the object. The contents are updated by various \code{oce} functions to
keep a record of processing steps.  Object summaries and
\code{\link[=processingLogShow]{processingLogShow()}} both display the log.}
}}

\section{Modifying slot contents}{


Although the \code{[[<-} operator may permit modification of the contents
of \linkS4class{tidem} objects (see \code{\link{[[<-,tidem-method}}),
it is better to use \code{\link[=oceSetData]{oceSetData()}} and \code{\link[=oceSetMetadata]{oceSetMetadata()}},
because those functions save an entry in the \code{processingLog}
that describes the change.
}

\section{Retrieving slot contents}{


The full contents of the \code{data} and \code{metadata} slots of a \linkS4class{tidem}
object may be retrieved in the standard R way using \code{\link[=slot]{slot()}}. For
example \code{slot(o,"data")} returns the \code{data} slot of an object named \code{o},
and similarly \code{slot(o,"metadata")} returns
the \code{metadata} slot.

The slots may also be obtained with the \code{\link{[[,tidem-method}}
operator, as e.g. \code{o[["data"]]} and \code{o[["metadata"]]}, respectively.

The \code{\link{[[,tidem-method}} operator can also
be used to retrieve items from within the \code{data} and \code{metadata} slots.
For example, \code{o[["temperature"]]} can be used to retrieve temperature
from an object containing that quantity. The rule is that a named
quantity is sought first within the object's \code{metadata} slot,
with the \code{data} slot being checked only if \code{metadata} does not
contain the item. This \code{[[} method can also be used to get
certain derived quantities, if the object contains sufficient
information to calculate them. For example, an object that holds
(practical) salinity, temperature and pressure, along with
longitude and latitude, has sufficient information to compute
Absolute Salinity, and so \code{o[["SA"]]} will yield the
calculated Absolute Salinity.

It is also possible to find items more directly, using \code{\link[=oceGetData]{oceGetData()}} and
\code{\link[=oceGetMetadata]{oceGetMetadata()}}, but neither of these functions can
retrieve derived items.
}

\seealso{
Other functions that plot oce data: 
\code{\link{download.amsr}()},
\code{\link{plot,adp-method}},
\code{\link{plot,adv-method}},
\code{\link{plot,amsr-method}},
\code{\link{plot,argo-method}},
\code{\link{plot,bremen-method}},
\code{\link{plot,cm-method}},
\code{\link{plot,coastline-method}},
\code{\link{plot,ctd-method}},
\code{\link{plot,gps-method}},
\code{\link{plot,ladp-method}},
\code{\link{plot,landsat-method}},
\code{\link{plot,lisst-method}},
\code{\link{plot,lobo-method}},
\code{\link{plot,met-method}},
\code{\link{plot,odf-method}},
\code{\link{plot,rsk-method}},
\code{\link{plot,satellite-method}},
\code{\link{plot,sealevel-method}},
\code{\link{plot,section-method}},
\code{\link{plot,tidem-method}},
\code{\link{plot,topo-method}},
\code{\link{plot,windrose-method}},
\code{\link{plot,xbt-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()}

Other things related to tides: 
\code{\link{[[,tidem-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{as.tidem}()},
\code{\link{plot,tidem-method}},
\code{\link{predict.tidem}()},
\code{\link{summary,tidem-method}},
\code{\link{tidalCurrent}},
\code{\link{tidedata}},
\code{\link{tidemAstron}()},
\code{\link{tidemVuf}()},
\code{\link{tidem}},
\code{\link{webtide}()}
}
\author{
Dan Kelley
}
\concept{functions that plot oce data}
\concept{things related to tides}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swRrho}
\alias{swRrho}
\title{Density ratio}
\usage{
swRrho(
  ctd,
  sense = c("diffusive", "finger"),
  smoothingLength = 10,
  df,
  eos = getOption("oceEOS", default = "gsw")
)
}
\arguments{
\item{ctd}{an \linkS4class{oce} object that holds \code{salinity}, \code{temperature}, and
\code{pressure}.  If \code{eos} is \code{"gsw"}, then it must also hold \code{longitude} and
\code{latitude}.}

\item{sense}{an indication of the sense of double diffusion under study and
therefore of the definition of Rrho; see \sQuote{Details}}

\item{smoothingLength}{ignored if \code{df} supplied, but otherwise the
latter is calculated as the number of data points, divided by the number
within a depth interval of \code{smoothingLength} metres.}

\item{df}{if given, this is provided to \code{\link[=smooth.spline]{smooth.spline()}}.}

\item{eos}{equation of state, either \code{"unesco"} or \code{"gsw"}.}
}
\value{
Density ratio defined in either the \code{"diffusive"} or
\code{"finger"} sense.
}
\description{
Compute density ratio
}
\details{
This computes Rrho (density ratio) from a \code{ctd} object.

If \code{eos="unesco"}, this is done by calculating salinity and
potential-temperature derivatives from smoothing splines whose properties
are governed by \code{smoothingLength} or \code{df}.  If
\code{sense="diffusive"} the definition is
\eqn{(beta*dS/dz)/(alpha*d(theta)/dz)}{(beta*dS/dz)/(alpha*d(theta)/dz)} and
the reciprocal for \code{"finger"}.

If \code{eos="gsw"}, this is done by extracting absolute salinity and
conservative temperature, smoothing with a smoothing spline as in the
\code{"unesco"} case, and then calling \code{\link[gsw:gsw_Turner_Rsubrho]{gsw::gsw_Turner_Rsubrho()}}
on these smoothed fields. Since the gsw function works on mid-point
pressures, \code{\link[=approx]{approx()}} is used to interpolate back to the original
pressures.

If the default arguments are acceptable, \code{ctd[["Rrho"]]} may be used
instead of \code{swRrho(ctd)}.
}
\examples{
library(oce)
data(ctd)
u <- swRrho(ctd, eos="unesco")
g <- swRrho(ctd, eos="gsw")
p <- ctd[["p"]]
plot(u, p, ylim=rev(range(p)), type='l', xlab=expression(R[rho]))
lines(g, p, lty=2, col='red')
legend("topright", lty=1:2, legend=c("unesco", "gsw"), col=c("black", "red"))

}
\seealso{
Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley and Chantelle Layton
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/AllClass.R
\name{handleFlags,oce-method}
\alias{handleFlags,oce-method}
\title{Handle flags in oce objects}
\usage{
\S4method{handleFlags}{oce}(
  object,
  flags = NULL,
  actions = NULL,
  where = NULL,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{object}{an \linkS4class{oce} object.}

\item{flags}{A \link{list} specifying flag values upon which
actions will be taken. This can take two forms.
\itemize{
\item In the first form, the
list has named elements each containing a vector of integers. For example,
salinities flagged with values of 1 or 3:9 would be specified
by \code{flags=list(salinity=c(1,3:9))}. Several data items can be specified,
e.g. \code{flags=list(salinity=c(1,3:9), temperature=c(1,3:9))} indicates
that the actions are to take place for both salinity and temperature.
\item In the second form, \code{flags} is a list holding a single \emph{unnamed} vector, and
this means to apply the actions to \emph{all} the data entries.  For example,
\code{flags=list(c(1,3:9))} means to apply not just to salinity and temperature,
but to everything within the \code{data} slot.
}

If \code{flags} is not provided, then \code{\link[=defaultFlags]{defaultFlags()}} is called, to try to
determine a reasonable default.}

\item{actions}{an optional \link{list} that contains items with
names that match those in the \code{flags} argument.  If \code{actions}
is not supplied, the default will be to set all values identified by
\code{flags} to \code{NA}; this can also be specified by
specifying \code{actions=list("NA")}. It is also possible to specify
functions that calculate replacement values. These are provided
with \code{object} as the single argument, and must return a
replacement for the data item in question.
See \dQuote{Details} for the default that is used if \code{actions} is not supplied.}

\item{where}{an optional character value that permits the function to work with
objects that store flags in e.g. \code{object@metadata$flags$where}
instead of in \code{object@metadata$flags}, and data within
\code{object@data$where} instead of within \code{object@data}. The
default value of \code{NULL} means to look withing \code{object@metadata}
itself, and this is the default within \code{oce}.  (The purpose of \code{where}
is to make \code{oce} extensible by other packages, which may choose to store
data two levels deep in the \code{data} slot.)}

\item{debug}{An optional integer specifying the degree of debugging, with
value 0 meaning to skip debugging and 1 or higher meaning to print some
information about the arguments and the data. It is usually a good idea to set
this to 1 for initial work with a dataset, to see which flags are being
handled for each data item. If not supplied, this defaults to the value of
\code{\link{getOption}}\code{("oceDebug")}.}
}
\description{
Data-quality flags are stored in the \code{metadata}
slot of \linkS4class{oce} objects in a
\link{list} named \code{flags}.
The present function (a generic that has specialized versions
for various data classes) provides a way to
manipulate the contents of the \code{data} slot, based on
such data-quality flags. For example, a common operation is to replace
erroneous data with \code{NA}.

If the \code{flags} within \code{object}'s \code{metadata} slot is empty,
then \code{object} is returned, unaltered.
Otherwise, \code{handleFlags} examines \code{object@metadata$flags}
in the context of the \code{flags} argument, and then
carries out actions that are specified by the \code{actions} argument.
By default, this sets the returned \code{data} entries to \code{NA},
wherever the corresponding \code{metadata$flag} values
signal unreliable data. To maintain a hint as to why
\code{data} were changed, \code{metadata$flags} in the
returned value is a direct copy of the corresponding
entry in \code{object}.
}
\details{
Base-level handling of flags.
}
\seealso{
Other functions relating to data-quality flags: 
\code{\link{defaultFlags}()},
\code{\link{handleFlags,adp-method}},
\code{\link{handleFlags,argo-method}},
\code{\link{handleFlags,ctd-method}},
\code{\link{handleFlags,section-method}},
\code{\link{handleFlags}()},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{initializeFlagScheme,oce-method}},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{initializeFlagSchemeInternal}()},
\code{\link{initializeFlagScheme}()},
\code{\link{initializeFlags,adp-method}},
\code{\link{initializeFlags,oce-method}},
\code{\link{initializeFlagsInternal}()},
\code{\link{initializeFlags}()},
\code{\link{setFlags,adp-method}},
\code{\link{setFlags,ctd-method}},
\code{\link{setFlags,oce-method}},
\code{\link{setFlags}()}
}
\concept{functions relating to data-quality flags}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/odf.R
\name{plot,odf-method}
\alias{plot,odf-method}
\alias{plot.odf}
\title{Plot an odf Object}
\usage{
\S4method{plot}{odf}(x, blanks = TRUE, debug = getOption("oceDebug"))
}
\arguments{
\item{x}{an \linkS4class{odf} object.}

\item{blanks}{A logical value that indicates whether to include dummy
plots for data items that lack any finite values.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\description{
Plot data contained within an ODF object,
using \code{\link[=oce.plot.ts]{oce.plot.ts()}} to create panels of time-series plots for
all the columns contained in the \code{odf} object (or just those that
contain at least one finite value, if \code{blanks} is \code{FALSE}).
If the object's \code{data} slot does not contain \code{time}, then
\code{\link[=pairs]{pairs()}} is used to plot all the elements in the \code{data}
slot that contain at least one finite value.
These actions are both
crude and there are no arguments to control the behaviour, but this
function is really just a stop-gap measure, since in practical work
\linkS4class{odf} objects are usually cast to other types, and those types
tend to have more useful plots.
}
\seealso{
Other functions that plot oce data: 
\code{\link{download.amsr}()},
\code{\link{plot,adp-method}},
\code{\link{plot,adv-method}},
\code{\link{plot,amsr-method}},
\code{\link{plot,argo-method}},
\code{\link{plot,bremen-method}},
\code{\link{plot,cm-method}},
\code{\link{plot,coastline-method}},
\code{\link{plot,ctd-method}},
\code{\link{plot,gps-method}},
\code{\link{plot,ladp-method}},
\code{\link{plot,landsat-method}},
\code{\link{plot,lisst-method}},
\code{\link{plot,lobo-method}},
\code{\link{plot,met-method}},
\code{\link{plot,rsk-method}},
\code{\link{plot,satellite-method}},
\code{\link{plot,sealevel-method}},
\code{\link{plot,section-method}},
\code{\link{plot,tidem-method}},
\code{\link{plot,topo-method}},
\code{\link{plot,windrose-method}},
\code{\link{plot,xbt-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{tidem-class}}

Other things related to odf data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{ODF2oce}()},
\code{\link{ODFListFromHeader}()},
\code{\link{ODFNames2oceNames}()},
\code{\link{[[,odf-method}},
\code{\link{[[<-,odf-method}},
\code{\link{odf-class}},
\code{\link{read.ctd.odf}()},
\code{\link{read.odf}()},
\code{\link{subset,odf-method}},
\code{\link{summary,odf-method}}
}
\author{
Dan Kelley
}
\concept{functions that plot oce data}
\concept{things related to odf data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swBeta}
\alias{swBeta}
\title{Seawater haline contraction coefficient}
\usage{
swBeta(
  salinity,
  temperature = NULL,
  pressure = 0,
  longitude = NULL,
  latitude = NULL,
  eos = getOption("oceEOS", default = "gsw")
)
}
\arguments{
\item{salinity}{either practical salinity (in which case \code{temperature}
and \code{pressure} must be provided) \emph{or} an \code{oce} object (in
which case \code{salinity}, etc. are inferred from the object).}

\item{temperature}{\emph{in-situ} temperature (\eqn{^\circ}{deg}C), defined
on the ITS-90 scale; see \dQuote{Temperature units} in the documentation for
\code{\link[=swRho]{swRho()}}.}

\item{pressure}{seawater pressure (dbar)}

\item{longitude}{longitude of observation (only used if \code{eos="gsw"};
see \sQuote{Details}).}

\item{latitude}{latitude of observation (only used if \code{eos="gsw"}; see
\sQuote{Details}).}

\item{eos}{equation of state, either \code{"unesco"} or \code{"gsw"}.}
}
\value{
Value in 1/psu.
}
\description{
Compute \eqn{\beta}{beta}, the haline contraction coefficient for seawater.
}
\references{
The \code{eos="unesco"} formulae are based on the UNESCO
equation of state, but are formulated empirically by Trevor J. McDougall,
1987, Neutral Surfaces, Journal of Physical Oceanography, volume 17, pages
1950-1964. The \code{eos="gsw"} formulae come from GSW; see references in
the \code{\link[=swRho]{swRho()}} documentation.
}
\seealso{
Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{integrateTrapezoid}
\alias{integrateTrapezoid}
\title{Trapezoidal Integration}
\usage{
integrateTrapezoid(x, y, type = c("A", "dA", "cA"), xmin, xmax)
}
\arguments{
\item{x, y}{vectors of x and y values. In the normal case, these
vectors are both supplied, and of equal length. There are also two
special cases. First, if \code{y} is missing, then
\code{x} is taken to be \code{y}, and a new \code{x} is constructed
as \code{\link{seq_along}}\verb{(y)1. Second, if }length(x)\verb{is 1 and}length(y)\verb{exceeds 1, then}x\verb{is replaced by}x*\code{[}seq_along\verb{]}(y)`.}

\item{type}{Flag indicating the desired return value (see \dQuote{Value}).}

\item{xmin, xmax}{Optional numbers indicating the range of the integration.
These values may be used to restrict the range of integration, or to
extend it; in either case, \code{\link[=approx]{approx()}} with \code{rule=2}
is used to create new x and y vectors.}
}
\value{
If \code{type="A"} (the default), a single value is returned,
containing the estimate of the integral of \code{y=y(x)}.  If
\code{type="dA"}, a numeric vector of the same length as \code{x}, of which
the first element is zero, the second element is the integral between
\code{x[1]} and \code{x[2]}, etc.  If \code{type="cA"}, the result is the
cumulative sum (as in \code{\link[=cumsum]{cumsum()}}) of the values that would be
returned for \code{type="dA"}.  See \dQuote{Examples}.
}
\description{
Estimate the integral of one-dimensional function using the trapezoidal
rule.
}
\section{Bugs}{
 There is no handling of \code{NA} values.
}

\examples{
x <- seq(0, 1, length.out=10) # try larger length.out to see if area approaches 2
y <- 2*x + 3*x^2
A <- integrateTrapezoid(x, y)
dA <- integrateTrapezoid(x, y, "dA")
cA <- integrateTrapezoid(x, y, "cA")
print(A)
print(sum(dA))
print(tail(cA, 1))
print(integrateTrapezoid(diff(x[1:2]), y))
print(integrateTrapezoid(y))
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/xbt.R
\name{as.xbt}
\alias{as.xbt}
\title{Create an xbt object}
\usage{
as.xbt(
  z,
  temperature,
  longitude = NA,
  latitude = NA,
  filename = "",
  sequenceNumber = NA,
  serialNumber = ""
)
}
\arguments{
\item{z}{numeric vector giving vertical coordinates of measurements. This is the negative of
depth, i.e. \code{z} is 0 at the air-sea interface, and negative within the water column.}

\item{temperature}{numeric vector giving in-situ temperatures at the \code{z} values.}

\item{longitude, latitude}{location in degE and degN.}

\item{filename}{character value naming source file.}

\item{sequenceNumber}{numerical value of the sequence number of the XBT drop.}

\item{serialNumber}{character value holding the serial number of the XBT.}
}
\value{
An \linkS4class{xbt} object.
}
\description{
Create an xbt object
}
\seealso{
Other things related to xbt data: 
\code{\link{[[,xbt-method}},
\code{\link{[[<-,xbt-method}},
\code{\link{plot,xbt-method}},
\code{\link{read.xbt.noaa1}()},
\code{\link{read.xbt}()},
\code{\link{subset,xbt-method}},
\code{\link{summary,xbt-method}},
\code{\link{xbt-class}},
\code{\link{xbt.edf}},
\code{\link{xbt}}
}
\author{
Dan Kelley
}
\concept{things related to xbt data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/colors.R
\docType{data}
\name{ocecolors}
\alias{ocecolors}
\title{Data that define some color palettes}
\source{
The data come from the matplotlib site
\url{https://github.com/matplotlib/matplotlib}.
}
\description{
The \code{ocecolors} dataset is a list containing color-schemes, used
by \code{\link[=oceColorsClosure]{oceColorsClosure()}} to create functions such as
\code{\link[=oceColorsViridis]{oceColorsViridis()}}.
}
\references{
The following references provide information on choosing
colour schemes, that are suitable for viewers who have
colour deficiencies.

Light, Adam, and Patrick J. Bartlein. "The End of the Rainbow? Color
Schemes for Improved Data Graphics." \emph{Eos, Transactions American Geophysical
Union} 85, no. 40 (2004): 385.
DOI: 10.1029/2004EO400002

Stephenson, David B. "Comment on 'Color Schemes for Improved Data
Graphics', by A Light and P.J. Bartlein." \emph{Eos, Transactions American
Geophysical Union} 86, no.  20 (2005): 196.
DOI: 10.1029/2005EO200005

Light, Adam, and Patrick J. Bartlein. "Reply to 'Comment on Color
Schemes for Improved Data Graphics,' by A. Light and P.J.  Bartlein'."
\emph{Eos, Transactions American Geophysical Union} 86, no. 20 (2005): 196‚Äì196.
DOI: 10.1029/2005EO200006

The following references provide information on choosing
colour schemes, that are suitable for viewers who have
colour deficiencies.

Light, Adam, and Patrick J. Bartlein. "The End of the Rainbow? Color
Schemes for Improved Data Graphics." \emph{Eos, Transactions American Geophysical
Union} 85, no. 40 (2004): 385.
DOI: 10.1029/2004EO400002

Stephenson, David B. "Comment on 'Color Schemes for Improved Data
Graphics', by A Light and P.J. Bartlein." \emph{Eos, Transactions American
Geophysical Union} 86, no.  20 (2005): 196.
DOI: 10.1029/2005EO200005

Light, Adam, and Patrick J. Bartlein. "Reply to 'Comment on Color
Schemes for Improved Data Graphics,' by A. Light and P.J.  Bartlein'."
\emph{Eos, Transactions American Geophysical Union} 86, no. 20 (2005): 196‚Äì196.
DOI: 10.1029/2005EO200006
}
\seealso{
Other datasets provided with oce: 
\code{\link{adp}},
\code{\link{adv}},
\code{\link{amsr}},
\code{\link{argo}},
\code{\link{cm}},
\code{\link{coastlineWorld}},
\code{\link{ctdRaw}},
\code{\link{ctd}},
\code{\link{echosounder}},
\code{\link{landsat}},
\code{\link{lisst}},
\code{\link{lobo}},
\code{\link{met}},
\code{\link{rsk}},
\code{\link{sealevelTuktoyaktuk}},
\code{\link{sealevel}},
\code{\link{section}},
\code{\link{topoWorld}},
\code{\link{wind}},
\code{\link{xbt}}

Other things related to colors: 
\code{\link{colormapGMT}()},
\code{\link{colormap}()},
\code{\link{oceColors9B}()},
\code{\link{oceColorsCDOM}()},
\code{\link{oceColorsChlorophyll}()},
\code{\link{oceColorsClosure}()},
\code{\link{oceColorsDensity}()},
\code{\link{oceColorsFreesurface}()},
\code{\link{oceColorsGebco}()},
\code{\link{oceColorsJet}()},
\code{\link{oceColorsOxygen}()},
\code{\link{oceColorsPAR}()},
\code{\link{oceColorsPalette}()},
\code{\link{oceColorsPhase}()},
\code{\link{oceColorsSalinity}()},
\code{\link{oceColorsTemperature}()},
\code{\link{oceColorsTurbidity}()},
\code{\link{oceColorsTurbo}()},
\code{\link{oceColorsTwo}()},
\code{\link{oceColorsVelocity}()},
\code{\link{oceColorsViridis}()},
\code{\link{oceColorsVorticity}()}
}
\author{
Authored by matplotlib contributers, packaged (with license permission) in oce by Dan Kelley
}
\concept{datasets provided with oce}
\concept{things related to colors}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{oceSmooth}
\alias{oceSmooth}
\alias{oce.smooth}
\title{Smooth an Oce Object}
\usage{
oceSmooth(x, ...)
}
\arguments{
\item{x}{an \linkS4class{oce} object.}

\item{\dots}{parameters to be supplied to \code{\link[=smooth]{smooth()}}, which does
the actual work.}
}
\value{
An \linkS4class{oce} object that has been smoothed appropriately.
}
\description{
Each data element is smoothed as a timeseries. For ADP data, this is done
along time, not distance.  Time vectors, if any, are not smoothed.  A good
use of \code{oce.smooth} is for despiking noisy data.
}
\examples{
library(oce)
data(ctd)
d <- oce.smooth(ctd)
plot(d)
}
\seealso{
The work is done with \code{\link[=smooth]{smooth()}}, and the \code{...}
arguments are handed to it directly by \code{oce.smooth}.
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.R
\name{ctdTrim}
\alias{ctdTrim}
\title{Trim Beginning and Ending of a CTD cast}
\usage{
ctdTrim(
  x,
  method,
  removeDepthInversions = FALSE,
  parameters = NULL,
  indices = FALSE,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{x}{a \linkS4class{ctd} object.}

\item{method}{A string (or a vector of two strings) specifying the trimming method, or a function to
be used to determine data indices to keep.  If \code{method} is not provided, \code{"downcast"} is
assumed. See \dQuote{Details}.}

\item{removeDepthInversions}{Logical value indicating whether to remove any levels at which depth is
less than, or equal to, a depth above.  (This is needed if the object is to be assembled into a
section, unless \code{\link[=ctdDecimate]{ctdDecimate()}} will be used, which will remove the inversions.}

\item{parameters}{A list whose elements depend on the method; see \dQuote{Details}.}

\item{indices}{Logical value indicating what to return. If \code{indices=FALSE} (the default),
then the return value is a subsetted \linkS4class{ctd} object. If \code{indices=TRUE},
then the return value is a logical vector that could be used to subset the data
with \code{\link[=subset,ctd-method]{subset,ctd-method()}} or to set data-quality flags.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\value{
Either a \linkS4class{ctd} object of or a logical vector of length
matching the data. The first option is the default. The second option,
achieved by setting \code{indices=FALSE}, may be useful in constructing
data flags to be inserted into the object.
}
\description{
Often in CTD profiling, the goal is to isolate only the downcast, discarding measurements made in
the air, in an equilibration phase in which the device is held below the water surface, and then the
upcast phase that follows the downcast.  This is handled reasonably well by \code{ctdTrim} with
\code{method="downcast"}, although it is almost always best to use \code{\link[=plotScan]{plotScan()}} to
investigate the data, and then use the \code{method="index"} or \code{method="scan"} method based on
visual inspection of the data.
}
\details{
\code{ctdTrim} begins by examining the pressure differences between subsequent samples. If
these are all of the same value, then the input \code{ctd} object is returned, unaltered.
This handles the case of pressure-binned data. However, if the pressure difference
varies, a variety of approaches are taken to trimming the dataset.
\itemize{
\item If \code{method[1]} is \code{"downcast"} then an attempt is made to keep only data for
which the CTD is descending.  This is done in stages, with variants based on \code{method[2]}, if
supplied.
\enumerate{
\item The pressure data are despiked with a smooth() filter with method "3R".
This removes wild spikes that arise from poor instrument connections, etc.
\item \emph{Step 2.} If no \code{parameters} are given, then any data with negative pressures
are deleted.  If there is a parameter named \code{pmin}, then that pressure (in decibars)
is used instead as the lower limit. This is a commonly-used setup, e.g.
\code{ctdTrim(ctd, parameters=list(pmin=1))} removes the top decibar (roughly 1m) from
the data.  Specifying \code{pmin} is a simple way to remove near-surface
data, such as a shallow equilibration phase, and if specified will cause \code{ctdTrim}
to skip step 4 below.
\item The maximum pressure is determined, and data acquired subsequent to
that point are deleted.  This removes the upcast and any subsequent data.
\item If the \code{pmin} parameter is not specified, an attempt is made to remove an initial
equilibrium phase by a regression of pressure on scan number.  There are three
variants to this, depending on the value of the second \code{method} element.
If \code{method} is \code{"A"} (or not given), the procedure is to
call \code{\link[=nls]{nls()}} to fit a piecewise linear model of pressure as a function of scan,
in which pressure is constant for scan less than a critical value, and then
linearly varying for with scan. This is  meant to handle the common situation
in which the CTD is held at roughly constant depth (typically
a metre or so) to equilibrate, before it is lowered through the water column.
If \code{method} is \code{"B"}, the procedure is similar, except that the pressure
in the surface region is taken to be zero (this does not make
much sense, but it might help in some cases). Note that, prior to early 2016,
method \code{"B"} was called method \code{"C"}; the old \code{"B"} method was judged useless
and so it was removed.
}
\item If \code{method="upcast"}, a sort of reverse of \code{"downcast"} is used. This
was added in late April 2017 and has not been well tested yet.
\item If \code{method="sbe"}, a method similar to that described
in the SBE Data Processing manual is used to remove the "soak"
period at the beginning of a cast (see Section 6 under subsection
"Loop Edit"). The method is based on the soak procedure whereby
the instrument sits at a fixed depth for a period of time, after
which it is raised toward the surface before beginning the actual
downcast. This enables equilibration of the sensors while still
permitting reasonably good near-surface data. Parameters for the
method can be passed using the \code{parameters} argument, which
include \code{minSoak} (the minimum depth for the soak) and
\code{maxSoak} the maximum depth of the soak. The method finds
the minimum pressure prior to the \code{maxSoak} value being
passed, each of which occurring after the scan in which the
\code{minSoak} value was reached. For the method to work, the
pre-cast pressure minimum must be less than the \code{minSoak}
value. The default values of \code{minSoak} and \code{maxSoak}
are 1 and 20 dbar, respectively.
\item If \code{method="index"} or \code{"scan"}, then each column of data is subsetted according to the
value of \code{parameters}. If the latter is a logical vector of length matching data column
length, then it is used directly for subsetting. If \code{parameters} is a numerical vector with
two elements, then the index or scan values that lie between \code{parameters[1]}
and \code{parameters[2]} (inclusive) are used for subsetting.  The
two-element method is probably the most useful, with the values being determined by visual
inspection of the results of \code{\link[=plotScan]{plotScan()}}. While this may take a minute or two, the
analyst should bear in mind that a deep-water CTD profile might take 6 hours, corresponding to
ship-time costs exceeding a week of salary.
\item If \code{method="range"} then data are selected based on the value of the column named
\code{parameters$item}.  This may be by range or by critical value.  By range: select values
between \code{parameters$from} (the lower limit) and \code{parameters$to} (the upper limit) By
critical value: select if the named column exceeds the value.  For example, \code{ctd2 <-
    ctdTrim(ctd, "range", parameters=list(item="scan", from=5))} starts at scan number 5 and
continues to the end, while
\code{ctdTrim(ctd,"range",parameters=list(item="scan",from=5,to=100))} also starts at scan 5,
but extends only to scan 100.
\item If \code{method} is a function, then it must return a vector of \code{\link[=logical]{logical()}}
values, computed based on two arguments: \code{data} (a
\code{\link[=list]{list()}}), and \code{parameters} as supplied to \code{ctdTrim}.  Both
\code{inferWaterDepth} and \code{removeInversions} are ignored in the function case. See
\dQuote{Examples}.
}
}
\examples{
\dontrun{
library(oce)
data(ctdRaw)
plot(ctdRaw) # barely recognizable, due to pre- and post-cast junk
plot(ctdTrim(ctdRaw)) # looks like a real profile ...
plot(ctdDecimate(ctdTrim(ctdRaw),method="boxcar")) # ... smoothed
# Demonstrate use of a function. The scan limits were chosen
# by using locator(2) on a graph made by plotScan(ctdRaw).
trimByIndex<-function(data, parameters) {
  parameters[1] < data$scan & data$scan < parameters[2]
}
trimmed <- ctdTrim(ctdRaw, trimByIndex, parameters=c(130, 380))
plot(trimmed)
}

}
\references{
The Seabird CTD instrument is described at
\verb{http://www.seabird.com/products/spec_sheets/19plusdata.htm}.

Seasoft V2: SBE Data Processing, SeaBird Scientific, 05/26/2016
}
\seealso{
Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}
}
\author{
Dan Kelley and Clark Richards
}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/coastline.R
\name{read.coastline.shapefile}
\alias{read.coastline.shapefile}
\title{Read a Coastline File in Shapefile Format}
\usage{
read.coastline.shapefile(
  file,
  lonlim = c(-180, 180),
  latlim = c(-90, 90),
  debug = getOption("oceDebug"),
  monitor = FALSE,
  processingLog
)
}
\arguments{
\item{file}{name of file containing coastline data (a file ending in \code{.shp})
or a zipfile that contains such a file, with a corresponding name.
The second scheme is useful for files downloaded from the NaturalEarth
website (see reference 2).}

\item{lonlim, latlim}{numerical vectors specifying the
west and east edges (and south and north edges) of a focus window.
Coastline polygons that do not intersect the defined box are
skipped, which can be useful in narrowing high-resolution world-scale
data to a local application.}

\item{debug}{set to TRUE to print information about the header, etc.}

\item{monitor}{Logical indicating whether to print an indication of progress through
the file.}

\item{processingLog}{if provided, the action item to be stored in the log.
(Typically only provided for internal calls; the default that it provides is
better for normal calls by a user.)}
}
\value{
x a \linkS4class{coastline} object.
}
\description{
Read coastline data stored in the shapefile format (see reference 1).
}
\section{A hack for depth contours}{
 The following demonstrates that this
code is getting close to working with depth contours.  This should be
handled more internally, and a new object for depth contours should be
constructed, of which coastlines could be a subset.
}

\references{
\enumerate{
\item The ``shapefile'' format is described in
\emph{ESRI Shapefile Technical Description}, March 1998, available at
\url{https://www.esri.com/content/dam/esrisites/sitecore-archive/Files/Pdfs/library/whitepapers/pdfs/shapefile.pdf}
(last checked 2021-03-24).
\item The NaturalEarth website \url{https://www.naturalearthdata.com/downloads/}
provides coastline datasets in three resolutions, along with similar files
lakes and rivers, for borders, etc. It is highly recommended.
}
}
\seealso{
Other things related to coastline data: 
\code{\link{[[,coastline-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{as.coastline}()},
\code{\link{coastline-class}},
\code{\link{coastlineBest}()},
\code{\link{coastlineCut}()},
\code{\link{coastlineWorld}},
\code{\link{download.coastline}()},
\code{\link{plot,coastline-method}},
\code{\link{read.coastline.openstreetmap}()},
\code{\link{subset,coastline-method}},
\code{\link{summary,coastline-method}}
}
\author{
Dan Kelley
}
\concept{things related to coastline data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.odv.R
\name{read.ctd.odv}
\alias{read.ctd.odv}
\title{Read an ODV-type CTD File}
\usage{
read.ctd.odv(
  file,
  columns = NULL,
  station = NULL,
  missingValue,
  deploymentType,
  monitor = FALSE,
  debug = getOption("oceDebug"),
  processingLog,
  ...
)
}
\arguments{
\item{file}{a connection or a character string giving the name of the file to
load.  For \code{\link[=read.ctd.sbe]{read.ctd.sbe()}} and \code{\link[=read.ctd.woce]{read.ctd.woce()}}, this may be a
wildcard (e.g. \code{"*.cnv"} or \code{"*.csv"}) in which case the return
value is a vector containing CTD objects created by reading the files from
\code{\link[=list.files]{list.files()}} with \code{pattern} set to the specified wildcard
pattern.}

\item{columns}{an optional \link{list} that can be used to convert unrecognized
data names to resultant variable names.  This is used only by
\code{\link[=read.ctd.sbe]{read.ctd.sbe()}} and \code{\link[=read.ctd.odf]{read.ctd.odf()}}. For example,
if a data file named salinity as \code{"SAL"}, then using\preformatted{d <- read.ctd(f, columns=list(
    salinity=list(name="SAL",
                  unit=list(unit=expression(),
                  scale="PSS-78"))))
}

would assign the \code{"SAL"} column to the \code{salinity} entry in the data
slot of the CTD object returned by the \verb{read.*} function.}

\item{station}{optional character string containing an identifying name or
number for the station. This can be useful if the routine cannot determine the
name automatically, or if another name is preferred.}

\item{missingValue}{optional missing-value flag; data matching this value will
be set to \code{NA} upon reading. If this is provided, then it overrules any
missing-value flag found in the data. For Seabird (\code{.cnv}) files, there is
usually no need to set \code{missingValue}, because it can be inferred from the
header (typically as -9.990e-29). Set \code{missingValue=NULL} to turn off
missing-value detection, even in \code{.cnv} files that contain missing-value
codes in their headers. If \code{missingValue} is not specified,
then an attempt is made to infer such a value from the data, by testing
whether salinity and/or temperature has a minimum that is under -8 in value;
this should catch common values in files, without false positives. A warning
will be issued in this case, and a note inserted in the processing log of
the return value.}

\item{deploymentType}{character string indicating the type of deployment. Use
\code{"unknown"} if this is not known, \code{"profile"} for a profile (in
which the data were acquired during a downcast, while the device was lowered
into the water column, perhaps also including an upcast; \code{"moored"} if
the device is installed on a fixed mooring, \code{"thermosalinograph"} (or
\code{"tsg"}) if the device is mounted on a moving vessel, to record
near-surface properties, or \code{"towyo"} if the device is repeatedly
lowered and raised.}

\item{monitor}{boolean, set to \code{TRUE} to provide an indication of
progress.  This is useful if \code{filename} is a wildcard.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed.}

\item{processingLog}{if provided, the action item to be stored in the log.
This is typically only provided for internal calls; the default that it provides is
better for normal calls by a user.}

\item{...}{additional arguments, passed to called routines.}
}
\value{
a \linkS4class{ctd} object. The details of the contents
depend on the source file. The \code{metadata} slot is particularly
variable across data formats, because the meta-information provided
in those formats varies so widely.
}
\description{
Read an ODV-type CTD File
}
\details{
\code{\link[=read.ctd.odv]{read.ctd.odv()}} attempts to read files stored in ODV format,
used by some European data providers. This works only crudely, as of
2020-05-17. In particular, the translation from ODV parameter names to oce
names is \emph{very} limited.  For example, only one of the dozens of possibilities
for variants of phosphate is handled at the moment, and that is because
this was the variant supplied in a test file sent to the author on
2020-05-16.  It is unlikely that full support of ODV files will
become available in \code{\link[=read.ctd.odv]{read.ctd.odv()}}, given the lack of a comprehensive source
listing ODV variable names and their meanings, and low user
interest.
}
\references{
\enumerate{
\item \url{https://www.bodc.ac.uk/resources/delivery_formats/odv_format/} describes
the \code{ODV} format.
\item \url{https://vocab.nerc.ac.uk/collection/P07/current/} may be
worth consulting for variable names.
}
}
\seealso{
Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}
}
\author{
Dan Kelley
}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{showMetadataItem}
\alias{showMetadataItem}
\title{Show metadata item}
\usage{
showMetadataItem(
  object,
  name,
  label = "",
  postlabel = "",
  isdate = FALSE,
  quote = FALSE
)
}
\arguments{
\item{object}{an \linkS4class{oce} object.}

\item{name}{name of item}

\item{label}{label to print before item}

\item{postlabel}{label to print after item}

\item{isdate}{boolean indicating whether the item is a time}

\item{quote}{boolean indicating whether to enclose the item in quotes}
}
\description{
This is a helper function for various \code{summary} functions.
}
\examples{
library(oce)
data(ctd)
showMetadataItem(ctd, "ship", "ship")
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.R
\name{toEnuAdp}
\alias{toEnuAdp}
\title{Convert an ADP Object to ENU Coordinates}
\usage{
toEnuAdp(x, declination = 0, debug = getOption("oceDebug"))
}
\arguments{
\item{x}{an \linkS4class{adp} object.}

\item{declination}{magnetic declination to be added to the heading, to get
ENU with N as "true" north.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\description{
Convert an ADP Object to ENU Coordinates
}
\references{
\enumerate{
\item @template nortekCoordTemplate
}
}
\seealso{
See \code{\link[=read.adp]{read.adp()}} for notes on functions relating to
\code{"adp"} objects.  Also, see \code{\link[=beamToXyzAdp]{beamToXyzAdp()}} and
\code{\link[=xyzToEnuAdp]{xyzToEnuAdp()}}.

Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley
}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swSigmaTheta}
\alias{swSigmaTheta}
\title{Seawater potential density anomaly}
\usage{
swSigmaTheta(
  salinity,
  temperature = NULL,
  pressure = NULL,
  referencePressure = 0,
  longitude = NULL,
  latitude = NULL,
  eos = getOption("oceEOS", default = "gsw")
)
}
\arguments{
\item{salinity}{either practical salinity (in which case \code{temperature}
and \code{pressure} must be provided) \emph{or} an \code{oce} object, in
which case \code{salinity}, \code{temperature} (in the ITS-90 scale; see
next item), etc. are inferred from the object, ignoring the
other parameters, if they are supplied.}

\item{temperature}{\emph{in-situ} temperature (\eqn{^\circ}{deg}C), defined
on the ITS-90 scale.  This scale is used by GSW-style calculation (as
requested by setting \code{eos="gsw"}), and is the value contained within
\code{ctd} objects (and probably most other objects created with data
acquired in the past decade or two). Since the UNESCO-style calculation is
based on IPTS-68, the temperature is converted within the present function,
using \code{\link[=T68fromT90]{T68fromT90()}}.}

\item{pressure}{pressure (dbar)}

\item{referencePressure}{The reference pressure, in dbar.}

\item{longitude}{longitude of observation (only used if \code{eos="gsw"};
see \sQuote{Details}).}

\item{latitude}{latitude of observation (only used if \code{eos="gsw"}; see
\sQuote{Details}).}

\item{eos}{equation of state, either \code{"unesco"} (references 1 and 2)
or \code{"gsw"} (references 3 and 4).}
}
\value{
Potential density anomaly (kg/m\eqn{^3}{^3}), defined as
\eqn{\sigma_\theta=\rho(S,\theta(S,t,p),0}{sigma_theta=rho(S,theta(S,t,p),0)}
\itemize{
\item 1000 kg/m\eqn{^3}{^3}.
}
}
\description{
Compute the potential density (minus 1000 kg/m^3) that seawater would have if raised
adiabatically to the surface.  In the UNESCO system, this quantity is
is denoted \eqn{\sigma_\theta}{sigma-theta} (hence the function name), but in
the GSW system, it is denoted \code{sigma0}.
}
\details{
If the first argument is an \code{oce} object, then salinity, etc., are
extracted from it, and used for the calculation instead of any values
provided in the other arguments.
}
\examples{
stopifnot(abs(26.4212790994 - swSigmaTheta(35, 13, 1000, eos="unesco")) < 1e-7)

}
\references{
See citations provided in the \code{\link[=swRho]{swRho()}} documentation.
}
\seealso{
Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/rsk.R
\name{plot,rsk-method}
\alias{plot,rsk-method}
\alias{plot.rsk}
\title{Plot a rsk Object}
\usage{
\S4method{plot}{rsk}(
  x,
  which = "timeseries",
  tlim,
  ylim,
  xlab,
  ylab,
  tformat,
  drawTimeRange = getOption("oceDrawTimeRange"),
  abbreviateTimeRange = getOption("oceAbbreviateTimeRange"),
  useSmoothScatter = FALSE,
  mgp = getOption("oceMgp"),
  mar = c(mgp[1] + 1.5, mgp[1] + 1.5, 1.5, 1.5),
  main = "",
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{x}{an \linkS4class{rsk} object.}

\item{which}{character indicating desired plot types.  These are
graphed in panels running down from the top of the page.  See
\dQuote{Details} for the meanings of various values of
\code{which}.}

\item{tlim}{optional limits for time axis.  If not provided, the value will be
inferred from the data.}

\item{ylim}{optional limits for the y axis.  If not provided, the
value will be inferred from the data.  (It is helpful to
specify this, if the auto-scaled value will be inappropriate,
e.g. if more lines are to be added later). Note that this is
ignored, unless \code{length(which) == 1} and \code{which}
corresponds to one of the data fields. If a multipanel plot of
a specific subset of the data fields is desired with
\code{ylim} control, it should be done panel by panel (see
Examples).}

\item{xlab}{optional label for x axis.}

\item{ylab}{optional label for y axis.}

\item{tformat}{optional argument passed to \code{\link[=oce.plot.ts]{oce.plot.ts()}}, for plot
types that call that function.  (See \code{\link[=strptime]{strptime()}} for the format
used.)}

\item{drawTimeRange}{boolean that applies to panels with time as the horizontal
axis, indicating whether to draw the time range in the top-left margin of the
plot.}

\item{abbreviateTimeRange}{boolean that applies to panels with time as the
horizontal axis, indicating whether to abbreviate the second time in the time
range (e.g. skipping the year, month, day, etc. if it's the same as the start
time).}

\item{useSmoothScatter}{a boolean to cause \code{\link[=smoothScatter]{smoothScatter()}} to be
used for profile plots, instead of \code{\link[=plot]{plot()}}.}

\item{mgp}{3-element numerical vector to use for \code{\link{par}}\code{("mgp")}, and
also for \code{par(mar)}, computed from this.  The default is tighter than the
R default, in order to use more space for the data and less for the axes.}

\item{mar}{value to be used with \code{\link{par}}\code{("mar")}.}

\item{main}{main title for plot, used just on the top panel, if there are several panels.}

\item{debug}{a flag that turns on debugging, if it exceeds 0.}

\item{...}{optional arguments passed to plotting functions.}
}
\description{
Rsk data may be in many forms, and it is not easy to devise a general plotting
strategy for all of them. The present function is quite crude, on the
assumption that users will understand their own datasets, and that they can
devise plots that are best-suited to their applications.  Sometimes, the
sensible scheme is to coerce the object into another form, e.g. using
\code{plot(as.ctd(rsk))} if the object contains CTD-like data.  Other times,
users should extract data from the \code{rsk} object and construct plots
themselves. The idea is to use the present function mainly to get an overview,
and for that reason, the default plot type (set by \code{which}) is a set of
time-series plots, because the one thing that is definitely known about
\code{rsk} objects is that they contain a \code{time} vector in their
\code{data} slot.
}
\details{
Plots produced are time series plots of the data in the
object. The default, \code{which="timeseries"} plots all data
fields, and over-rides any other specification. Specific fields
can be plotted by naming the field,
e.g. \code{which="temperature"} to plot a time series of just
the temperature field.
}
\examples{
library(oce)
data(rsk)
plot(rsk) # default timeseries plot of all data fields

## A multipanel plot of just pressure and temperature with ylim
par(mfrow=c(2, 1))
plot(rsk, which="pressure", ylim=c(10, 30))
plot(rsk, which="temperature", ylim=c(2, 4))

}
\seealso{
The documentation for \linkS4class{rsk} explains the structure of
\code{rsk} objects, and also outlines the other functions dealing with them.

Other functions that plot oce data: 
\code{\link{download.amsr}()},
\code{\link{plot,adp-method}},
\code{\link{plot,adv-method}},
\code{\link{plot,amsr-method}},
\code{\link{plot,argo-method}},
\code{\link{plot,bremen-method}},
\code{\link{plot,cm-method}},
\code{\link{plot,coastline-method}},
\code{\link{plot,ctd-method}},
\code{\link{plot,gps-method}},
\code{\link{plot,ladp-method}},
\code{\link{plot,landsat-method}},
\code{\link{plot,lisst-method}},
\code{\link{plot,lobo-method}},
\code{\link{plot,met-method}},
\code{\link{plot,odf-method}},
\code{\link{plot,satellite-method}},
\code{\link{plot,sealevel-method}},
\code{\link{plot,section-method}},
\code{\link{plot,tidem-method}},
\code{\link{plot,topo-method}},
\code{\link{plot,windrose-method}},
\code{\link{plot,xbt-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{tidem-class}}

Other things related to rsk data: 
\code{\link{[[,rsk-method}},
\code{\link{[[<-,rsk-method}},
\code{\link{as.rsk}()},
\code{\link{read.rsk}()},
\code{\link{rsk-class}},
\code{\link{rskPatm}()},
\code{\link{rskToc}()},
\code{\link{rsk}},
\code{\link{subset,rsk-method}},
\code{\link{summary,rsk-method}}
}
\author{
Dan Kelley and Clark Richards
}
\concept{functions that plot oce data}
\concept{things related to rsk data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/AllClass.R
\name{handleFlags,vector-method}
\alias{handleFlags,vector-method}
\title{Signal erroneous application to non-oce objects}
\usage{
\S4method{handleFlags}{vector}(
  object,
  flags = list(),
  actions = list(),
  where = list(),
  debug = getOption("oceDebug")
)
}
\arguments{
\item{object}{A vector, which cannot be the case for \code{oce} objects.}

\item{flags}{Ignored.}

\item{actions}{Ignored.}

\item{where}{Ignored.}

\item{debug}{Ignored.}
}
\description{
Signal erroneous application to non-oce objects
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/argo.R
\name{preferAdjusted}
\alias{preferAdjusted}
\title{Set Preference for Adjusted Values}
\usage{
preferAdjusted(argo, which = "all", fallback = TRUE)
}
\arguments{
\item{argo}{An \linkS4class{argo} object.}

\item{which}{A character vector naming the items for which
(depending also on the value of \code{fallback}) adjusted values
are to be sought by future calls to \code{\link{[[,argo-method}}.
The short names are used, e.g. \code{which="oxygen"} means that
adjusted oxygen is to be returned in future calls
such as \code{argo[["oxygen"]]}.  The default,
\code{"all"}, means to  use adjusted values for any item in \code{argo}
that has adjusted values.}

\item{fallback}{A logical value indicating whether to fall back
to unadjusted values for any data field in which the
adjusted values are all \code{NA}.  The default value, \code{TRUE},
avoids a problem with biogeochemical fields, where adjustment
of any one field may lead to insertion of "adjusted" values for
other fields that consist of nothing more than \code{NA}s.}
}
\value{
An \linkS4class{argo} object its \code{metadata} slot altered
(in its \code{adjustedWhich} and \code{adjustedFallback} elements)
as a signal for how \code{\link{[[,argo-method}} should
function on the object.
}
\description{
\linkS4class{argo} data can contain "adjusted" forms of data items,
which may be more trustworthy than the original
data, and \code{preferAdjusted} lets the user express a preference
for such adjusted data.  This means that using
\code{\link{[[,argo-method}} on the results returned by \code{preferAdjusted}
will (if possible) return adjusted data, and also use those adjusted
data in computations of derived quantities such as Absolute Salinity.
The preference applies also to units and to data-quality flags,
both of which can be returned by \code{\link{[[,argo-method}}, as
discussed in \dQuote{Details}.
}
\details{
\code{preferAdjusted()} merely sets two items in the \code{metadata} slot of the
returned \linkS4class{argo} object. The real action is carried out by
\code{\link{[[,argo-method}} but, for convenience, the details are explained here.

Consider salinity, for example.
If \code{which} equals \code{"all"}, or if it is a character
vector containing \code{"salinity"}, then using
\code{\link{[[,argo-method}} on the returned object
will yield the adjusted forms of the salinity data,
its associated flags, or its units.  Thus, in the salinity
case,
\itemize{
\item \code{argo[["salinity"]]} will attempt to return \code{argo@data$salinityAdjusted}
instead of returning \code{argo@data$salinity}, although if the adjusted values
are all \code{NA} then, depending on the value of \code{fallback}, the
unadjusted values may be returned; similarly
\item \code{argo[["salinityFlags"]]} will attempt to return
\code{argo@metadata$flags$salinityAdjusted}
instead of \code{argo@metadata$flags$salinity}, and
\item \code{argo[["salinityUnits"]]} will attempt to return
\code{argo@metadata$units$salinityAdjusted}
instead of \code{argo@metadata$units$salinity}.
}

The default value, \code{which="all"}, indicates that this
preference for adjusted values will apply to all the
elements of the \code{data} slot of the returned object, along
with associated flags and units. This can be handy for quick
work, but analysts may also choose to restrict their use of
adjusted values to a subset of variables, based on their own
decisions about data quality or accuracy.

The default value \code{fallback=TRUE} indicates that later calls to
\code{\link{[[,argo-method}} should return unadjusted values for any
data items that have \code{NA} for all the adjusted values.  This
condition is rare for core variables (salinity, temperature and
pressure) but is annoyingly common for biogeochemical variables; see
e.g. Section 2.2.5 of Reference 1 for a discussion of
the conditions under which Argo netcdf files contain
adjusted values. Setting \code{fallback=FALSE} means that adjusted
values (if they exist) will always be returned, even if they
are a useless collection of \code{NA} values.

Error fields, such as \code{salinityAdjustedError}, are returned
as-is by \code{\link{[[,argo-method}}, regardless of whether
the object was created by \code{preferAdjusted}.

It should be noted that, regardless of whether \code{preferAdjusted}
has been used, the analyst can always access either unadjusted
or adjusted data directly, using the original variable names stored
in the source netcdf file.  For example, \code{argo[["PSAL"]]}
yields unadjusted salinity values, and
\code{argo[["PSAL_ADJUSTED"]]} yields adjusted values (if they exist, or
\code{NULL} if they do not).
Similarly, adjusted value can always be obtained by using a form
like \code{argo[["salinityAdjusted"]]}.
}
\examples{
library(oce)
data(argo)
argoAdjusted <- preferAdjusted(argo)
all.equal(argo[["salinityAdjusted"]], argoAdjusted[["salinity"]])
all.equal(argo[["salinityFlagsAdjusted"]], argoAdjusted[["salinityFlags"]])
all.equal(argo[["salinityUnitsAdjusted"]], argoAdjusted[["salinityUnits"]])

}
\references{
\enumerate{
\item Argo Data Management Team. "Argo User's Manual V3.3." Ifremer,
November 28, 2019.
\doi{10.13155/29825}
}
}
\author{
Dan Kelley, based on discussions with Jaimie Harbin (with
respect to the \code{\link{[[,argo-method}} interface) and Clark Richards
(with respect to storing the preference in the \code{metadata} slot).
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/AllClass.R
\name{subset,oce-method}
\alias{subset,oce-method}
\title{Subset an oce Object}
\usage{
\S4method{subset}{oce}(x, subset, ...)
}
\arguments{
\item{x}{an \linkS4class{oce} object.}

\item{subset}{a logical expression indicating how to take the subset; the form depends on the sub-class.}

\item{...}{optional arguments, used in some specialized methods, e.g. \code{\link[=subset,section-method]{subset,section-method()}}.}
}
\value{
An oce object.
}
\description{
This is a basic class for general oce objects.  It has specialised
versions for most sub-classes, e.g. \code{\link[=subset,ctd-method]{subset,ctd-method()}}
for \code{ctd} objects.
}
\examples{
library(oce)
data(ctd)
# Select just the top 10 metres (pressure less than 10 dbar)
top10 <- subset(ctd, pressure < 10)
par(mfrow=c(1, 2))
plotProfile(ctd)
plotProfile(top10)
}
\seealso{
Other functions that subset oce objects: 
\code{\link{subset,adp-method}},
\code{\link{subset,adv-method}},
\code{\link{subset,amsr-method}},
\code{\link{subset,argo-method}},
\code{\link{subset,cm-method}},
\code{\link{subset,coastline-method}},
\code{\link{subset,ctd-method}},
\code{\link{subset,echosounder-method}},
\code{\link{subset,lobo-method}},
\code{\link{subset,met-method}},
\code{\link{subset,odf-method}},
\code{\link{subset,rsk-method}},
\code{\link{subset,sealevel-method}},
\code{\link{subset,section-method}},
\code{\link{subset,topo-method}},
\code{\link{subset,xbt-method}}
}
\concept{functions that subset oce objects}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adv.R
\name{beamToXyzAdv}
\alias{beamToXyzAdv}
\title{Convert ADV from Beam to XYZ Coordinates}
\usage{
beamToXyzAdv(x, debug = getOption("oceDebug"))
}
\arguments{
\item{x}{an \linkS4class{adv} object.}

\item{debug}{a flag that, if non-zero, turns on debugging.  Higher values
yield more extensive debugging.}
}
\description{
Convert ADV velocity components from a beam-based coordinate system to a
xyz-based coordinate system.
}
\details{
The coordinate transformation is done using the transformation matrix
contained in \code{transformation.matrix} in the
\code{metadata} slot, which is normally
inferred from the header in the binary file.  If there is no such matrix
(e.g. if the data were streamed through a data logger that did not capture
the header), \code{beamToXyzAdv} the user will need to store one in
\code{x}, e.g. by doing something like the following:\preformatted{x[["transformation.matrix"]] <- rbind(c(11100, -5771, -5321),
                                      c( #' 291, 9716, -10002),
                                      c( 1409, 1409, 1409)) / 4096
}
}
\references{
\enumerate{
\item @template nortekCoordTemplate
}
}
\seealso{
See \code{\link[=read.adv]{read.adv()}} for notes on functions relating to
\code{"adv"} objects.

Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley
}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{oceApprox}
\alias{oceApprox}
\alias{oce.approx}
\title{Interpolate 1D Data with UNESCO or Reiniger-Ross Algorithm}
\usage{
oceApprox(x, y, xout, method = c("rr", "unesco"))
}
\arguments{
\item{x}{the independent variable (z or p, usually).}

\item{y}{the dependent variable.}

\item{xout}{the values of the independent variable at which interpolation is
to be done.}

\item{method}{method to use.  See \dQuote{Details}.}
}
\value{
A vector of interpolated values, corresponding to the \code{xout}
values and equal in number.
}
\description{
Interpolate one-dimensional data using schemes that permit curvature but
tends minimize extrema that are not well-indicated by the data.
}
\details{
Setting \code{method="rr"} yields the weighted-parabola algorithm of
Reiniger and Ross (1968).  For procedure is as follows.  First, the
interpolant for any \code{xout} value that is outside the range of \code{x}
is set to NA.  Next, linear interpolation is used for any \code{xout} value
that has only one smaller neighboring \code{x} value, or one larger
neighboring value.  For all other values of \code{xout}, the 4 neighboring
points \code{x} are sought, two smaller and two larger.  Then two parabolas
are determined, one from the two smaller points plus the nearest larger
point, and the other from the nearest smaller point and the two larger
points.  A weighted sum of these two parabolas provides the interpolated
value.  Note that, in the notation of Reiniger and Ross (1968), this
algorithm uses \code{m}=2 and \code{n}=1.  (A future version of this routine
might provide the ability to modify these values.)

Setting \code{method="unesco"} yields the method that is used by the U.S.
National Oceanographic Data Center. It is described in pages 48-50 of
reference 2; reference 3 presumably contains the same information but it is
not as easily accessible.  The method works as follows.
\itemize{
\item If there are data above 5m depth, then the surface value is taken to
equal to the shallowest recorded value.
\item Distance bounds are put on the four neighboring points, and the
Reiniger-Ross method is used for interpolated points with sufficiently four
close neighbors.  The bounds are described in table 15 of reference 2 only
for so-called standard depths; in the present instance they are transformed
to the following rules.  Inner neighbors must be within 5m for data above
10m, 50m above 250m 100m above 900m, 200m above 2000m, or within 1000m
otherwise.  Outer neighbors must be within 200m above 500m, 400m above
1300m, or 1000m otherwise.  If two or more points meet these criteria,
Lagrangian interpolation is used.  If not, \code{NA} is used as the
interpolant.
}

After these rules are applied, the interpolated value is compared with the
values immediately above and below it, and if it is outside the range,
simple linear interpolation is used.
}
\examples{
library(oce)
if (require(ocedata)) {
    data(RRprofile)
    zz <- seq(0, 2000, 2)
    plot(RRprofile$temperature, RRprofile$depth, ylim=c(500, 0), xlim=c(2, 11))
    # Contrast two methods
    a1 <- oce.approx(RRprofile$depth, RRprofile$temperature, zz, "rr")
    a2 <- oce.approx(RRprofile$depth, RRprofile$temperature, zz, "unesco")
    lines(a1, zz)
    lines(a2, zz, col='red')
    legend("bottomright",lwd=1,col=1:2, legend=c("rr","unesco"),cex=3/4)
}
}
\references{
\enumerate{
\item R.F. Reiniger and C.K. Ross, 1968.  A method of interpolation with
application to oceanographic data.  \emph{Deep Sea Research}, \strong{15},
185-193.
\item Daphne R. Johnson, Tim P. Boyer, Hernan E. Garcia, Ricardo A.
Locarnini, Olga K. Baranova, and Melissa M. Zweng, 2011. World Ocean
Database 2009 Documentation.  NODC Internal report 20.  Ocean Climate
Laboratory, National Oceanographic Data Center.  Silver Spring, Maryland.
\item UNESCO, 1991. Processing of oceanographic station data, 138 pp.,
Imprimerie des Presses Universitaires de France, United Nations Educational,
Scientific and Cultural Organization, France.
}
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/echosounder.R
\name{[[,echosounder-method}
\alias{[[,echosounder-method}
\title{Extract Something from an Echosounder Object}
\usage{
\S4method{[[}{echosounder}(x, i, j, ...)
}
\arguments{
\item{x}{an \linkS4class{echosounder} object.}

\item{i}{character value indicating the name of an item to extract.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{ignored.}
}
\description{
Generally, the \code{[[} method lets users extract information from \code{oce}
objects, without having to know the details of the internal storage.  For
many \code{oce} sub-classes, \code{[[} can also return quantities that are computed
from the object's contents.
}
\details{
A two-step process is used to try to find the requested information. First, a
class-specific function is used (see \dQuote{Details of the Specialized
Method}).  If this yields nothing, then a general method is used (see
\dQuote{Details of the General Method}).  If both methods fail, then \code{[[}
returns NULL.

Some understanding of the subclass is required to know what can be retrieved
with \code{[[}. When dealing with an unfamiliar subclass, it can be useful to
first use \code{x[["?"]]} to get a listing of the retrievable items. See
\dQuote{Details of the Specialized Method} for more information.
}
\section{Details of the Specialized Method}{

\itemize{
\item If \code{i} is \code{"?"}, then the return value is a list
containing four items, each of which is a character vector
holding the names of things that can be accessed with \code{[[}.
The \code{data} and \code{metadata} items hold the names of
entries in the object's data and metadata
slots, respectively. The \code{metadataDerived} item
is NULL, while the \code{dataDerived} item holds
\code{"Sv"} and \code{"TS"} (see next).
\item If \code{i} is \code{"Sv"}, then the following is returned.
\preformatted{
20*log10(a) -
  (x@metadata$sourceLevel+x@metadata$receiverSensitivity+x@metadata$transmitPower) +
  20*log10(r) +
  2*absorption*r -
  x@metadata$correction +
  10*log10(soundSpeed*x@metadata$pulseDuration/1e6*psi/2)
}
\item If \code{i} is \code{"TS"}, then the following is returned.
\preformatted{
20*log10(a) -
  (x@metadata$sourceLevel+x@metadata$receiverSensitivity+x@metadata$transmitPower) +
  40*log10(r) +
  2*absorption*r +
  x@metadata$correction
}
}
}

\section{Details of the General Method}{


Note: the text of this section is identical for all \code{oce} subclasses, and so
some of what you read here may not be relevant to the class being described
in this help page.

If the specialized method produces no matches, the following generalized
method is applied. As with the specialized method, the procedure hinges first
on the values of \code{i} and, optionally, \code{j}. The work proceeds in steps, by
testing a sequence of possible conditions in sequence.
\enumerate{
\item A check is made as to whether \code{i} names one of the standard \code{oce} slots.
If so, \code{[[} returns the slot contents of that slot.  Thus, \code{x[["metadata"]]}
will retrieve the \code{metadata} slot, while \code{x[["data"]]} and
\code{x[["processingLog"]]} return those slots.
\item If \code{i} is a string ending in the \code{"Unit"}, then the characters preceding
that string are taken to be the name of an item in the data object, and a
list containing the unit is returned (or \code{NULL} if there is no such unit).
This list consists of an item named \code{unit}, which is an \code{\link[=expression]{expression()}}, and
an item named \code{scale}, which is a string describing the measurement scale.
If the string ends in \code{" unit"}, e.g.  \code{x[["temperature unit"]]} (note the
space), then just the expression is returned, and if it ends in \code{" scale"},
then just the scale is returned.
\item If \code{i} is a string ending in \code{"Flag"}, then the corresponding data-quality
flag is returned (or \code{NULL} if there is no such flag).
\item If the object holds hydrographic information (pressure, salinity,
temperature, longitude and latitude) then another set of possibilities
arises.  If \code{i} is \code{"sigmaTheta"}, then the value of \code{\link[=swSigmaTheta]{swSigmaTheta()}} is
called with \code{x} as the sole argument, and the results are returned.
Similarly, \code{\link[=swSigma0]{swSigma0()}} is used if \code{i="sigma0"}, and \code{\link[=swSpice]{swSpice()}} is used if
\code{i="spice"}. Of course, these actions only make sense for objects that
contain the relevant items within their \code{data} slot.
\item After these possibilities are eliminated, the action depends on whether
\code{j} has been provided.  If \code{j} is not provided, or is the string \code{""}, then
\code{i} is sought in the \code{metadata} slot, and then in the \code{data} slot, returning
whichever is found first.  In other words, if \code{j} is not provided, the
\code{metadata} slot takes preference over the \code{data} slot. However, if \code{j} is
provided, then it must be either the string \code{"metadata"} or \code{"data"}, and it
directs where to look.
\item If none of the above-listed conditions holds, then \code{NULL} is returned.
}
}

\seealso{
Other functions that extract parts of oce objects: 
\code{\link{[[,adp-method}},
\code{\link{[[,adv-method}},
\code{\link{[[,amsr-method}},
\code{\link{[[,argo-method}},
\code{\link{[[,bremen-method}},
\code{\link{[[,cm-method}},
\code{\link{[[,coastline-method}},
\code{\link{[[,ctd-method}},
\code{\link{[[,g1sst-method}},
\code{\link{[[,gps-method}},
\code{\link{[[,ladp-method}},
\code{\link{[[,landsat-method}},
\code{\link{[[,lisst-method}},
\code{\link{[[,lobo-method}},
\code{\link{[[,met-method}},
\code{\link{[[,oce-method}},
\code{\link{[[,odf-method}},
\code{\link{[[,rsk-method}},
\code{\link{[[,sealevel-method}},
\code{\link{[[,section-method}},
\code{\link{[[,tidem-method}},
\code{\link{[[,topo-method}},
\code{\link{[[,windrose-method}},
\code{\link{[[,xbt-method}},
\code{\link{[[<-,adv-method}}

Other things related to echosounder data: 
\code{\link{[[<-,echosounder-method}},
\code{\link{as.echosounder}()},
\code{\link{echosounder-class}},
\code{\link{echosounder}},
\code{\link{findBottom}()},
\code{\link{plot,echosounder-method}},
\code{\link{read.echosounder}()},
\code{\link{subset,echosounder-method}},
\code{\link{summary,echosounder-method}}
}
\author{
Dan Kelley
}
\concept{functions that extract parts of oce objects}
\concept{things related to echosounder data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.R
\name{handleFlags,ctd-method}
\alias{handleFlags,ctd-method}
\title{Handle Flags in ctd Objects}
\usage{
\S4method{handleFlags}{ctd}(
  object,
  flags = NULL,
  actions = NULL,
  where = NULL,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{object}{a \linkS4class{ctd} object.}

\item{flags}{A \link{list} specifying flag values upon which
actions will be taken. This can take two forms.
\itemize{
\item In the first form, the
list has named elements each containing a vector of integers. For example,
salinities flagged with values of 1 or 3:9 would be specified
by \code{flags=list(salinity=c(1,3:9))}. Several data items can be specified,
e.g. \code{flags=list(salinity=c(1,3:9), temperature=c(1,3:9))} indicates
that the actions are to take place for both salinity and temperature.
\item In the second form, \code{flags} is a list holding a single \emph{unnamed} vector, and
this means to apply the actions to \emph{all} the data entries.  For example,
\code{flags=list(c(1,3:9))} means to apply not just to salinity and temperature,
but to everything within the \code{data} slot.
}

If \code{flags} is not provided, then \code{\link[=defaultFlags]{defaultFlags()}} is called, to try to
determine a reasonable default.}

\item{actions}{an optional \link{list} that contains items with
names that match those in the \code{flags} argument.  If \code{actions}
is not supplied, the default will be to set all values identified by
\code{flags} to \code{NA}; this can also be specified by
specifying \code{actions=list("NA")}. It is also possible to specify
functions that calculate replacement values. These are provided
with \code{object} as the single argument, and must return a
replacement for the data item in question.
See \dQuote{Details} for the default that is used if \code{actions} is not supplied.}

\item{where}{an optional character value that permits the function to work with
objects that store flags in e.g. \code{object@metadata$flags$where}
instead of in \code{object@metadata$flags}, and data within
\code{object@data$where} instead of within \code{object@data}. The
default value of \code{NULL} means to look withing \code{object@metadata}
itself, and this is the default within \code{oce}.  (The purpose of \code{where}
is to make \code{oce} extensible by other packages, which may choose to store
data two levels deep in the \code{data} slot.)}

\item{debug}{An optional integer specifying the degree of debugging, with
value 0 meaning to skip debugging and 1 or higher meaning to print some
information about the arguments and the data. It is usually a good idea to set
this to 1 for initial work with a dataset, to see which flags are being
handled for each data item. If not supplied, this defaults to the value of
\code{\link{getOption}}\code{("oceDebug")}.}
}
\description{
Data-quality flags are stored in the \code{metadata}
slot of \linkS4class{oce} objects in a
\link{list} named \code{flags}.
The present function (a generic that has specialized versions
for various data classes) provides a way to
manipulate the contents of the \code{data} slot, based on
such data-quality flags. For example, a common operation is to replace
erroneous data with \code{NA}.

If the \code{flags} within \code{object}'s \code{metadata} slot is empty,
then \code{object} is returned, unaltered.
Otherwise, \code{handleFlags} examines \code{object@metadata$flags}
in the context of the \code{flags} argument, and then
carries out actions that are specified by the \code{actions} argument.
By default, this sets the returned \code{data} entries to \code{NA},
wherever the corresponding \code{metadata$flag} values
signal unreliable data. To maintain a hint as to why
\code{data} were changed, \code{metadata$flags} in the
returned value is a direct copy of the corresponding
entry in \code{object}.
}
\examples{
library(oce)
data(section)
stn <- section[["station", 100]]
# 1. Default: anything not flagged as 2 is set to NA, to focus
# solely on 'good', in the World Hydrographic Program scheme.
STN1 <- handleFlags(stn, flags=list(c(1, 3:9)))
data.frame(old=stn[["salinity"]], new=STN1[["salinity"]], salinityFlag=stn[["salinityFlag"]])

# 2. Use bottle salinity, if it is good and ctd is bad
replace <- 2 == stn[["salinityBottleFlag"]] & 2 != stn[["salinityFlag"]]
S <- ifelse(replace, stn[["salinityBottle"]], stn[["salinity"]])
STN2 <- oceSetData(stn, "salinity", S)

# 3. Use smoothed TS relationship to nudge questionable data.
f <- function(x) {
  S <- x[["salinity"]]
  T <- x[["temperature"]]
  df <- 0.5 * length(S) # smooths a bit
  sp <- smooth.spline(T, S, df=df)
  0.5 * (S + predict(sp, T)$y)
}
par(mfrow=c(1,2))
STN3 <- handleFlags(stn, flags=list(salinity=c(1,3:9)), action=list(salinity=f))
plotProfile(stn, "salinity", mar=c(3, 3, 3, 1))
p <- stn[["pressure"]]
par(mar=c(3, 3, 3, 1))
plot(STN3[["salinity"]] - stn[["salinity"]], p, ylim=rev(range(p)))

# 4. Single-variable flags (vector specification)
data(section)
# Multiple-flag scheme: one per data item
A <- section[["station", 100]]
deep <- A[["pressure"]] > 1500
flag <- ifelse(deep, 7, 2)
for (flagName in names(A[["flags"]]))
    A[[paste(flagName, "Flag", sep="")]] <- flag
Af <- handleFlags(A)
stopifnot(all.equal(is.na(Af[["salinity"]]), deep))

# 5. Single-variable flags (list specification)
B <- section[["station", 100]]
B[["flags"]] <- list(flag)
Bf <- handleFlags(B)
stopifnot(all.equal(is.na(Bf[["salinity"]]), deep))

}
\references{
The following link used to work, but failed as of December 2020.
\enumerate{
\item \verb{https://www.nodc.noaa.gov/woce/woce_v3/wocedata_1/whp/exchange/exchange_format_desc.htm}
}
}
\seealso{
Other functions relating to data-quality flags: 
\code{\link{defaultFlags}()},
\code{\link{handleFlags,adp-method}},
\code{\link{handleFlags,argo-method}},
\code{\link{handleFlags,oce-method}},
\code{\link{handleFlags,section-method}},
\code{\link{handleFlags}()},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{initializeFlagScheme,oce-method}},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{initializeFlagSchemeInternal}()},
\code{\link{initializeFlagScheme}()},
\code{\link{initializeFlags,adp-method}},
\code{\link{initializeFlags,oce-method}},
\code{\link{initializeFlagsInternal}()},
\code{\link{initializeFlags}()},
\code{\link{setFlags,adp-method}},
\code{\link{setFlags,ctd-method}},
\code{\link{setFlags,oce-method}},
\code{\link{setFlags}()}

Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}
}
\concept{functions relating to data-quality flags}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.R
\name{binmapAdp}
\alias{binmapAdp}
\title{Bin-map an ADP object}
\usage{
binmapAdp(x, debug = getOption("oceDebug"))
}
\arguments{
\item{x}{an \linkS4class{adp} object.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\value{
An \linkS4class{adp} object.
}
\description{
Bin-map an ADP object, by interpolating velocities, backscatter amplitudes,
etc., to uniform depth bins, thus compensating for the pitch and roll of the
instrument.  This only makes sense for ADP objects that are in beam
coordinates.
}
\section{Bugs}{
 This only works for 4-beam RDI ADP objects.
}

\examples{
\dontrun{
library(oce)
beam <- read.oce("/data/archive/sleiwex/2008/moorings/m09/adp/rdi_2615/raw/adp_rdi_2615.000",
                 from=as.POSIXct("2008-06-26", tz="UTC"),
                 to=as.POSIXct("2008-06-26 00:10:00", tz="UTC"),
                 longitude=-69.73433, latitude=47.88126)
beam2 <- binmapAdp(beam)
plot(enuToOther(toEnu(beam), heading=-31.5))
plot(enuToOther(toEnu(beam2), heading=-31.5))
plot(beam, which=5:8) # backscatter amplitude
plot(beam2, which=5:8)
}

}
\references{
The method was devised by Clark Richards for use in his PhD work
at Department of Oceanography at Dalhousie University.
}
\seealso{
See \linkS4class{adp} for a discussion of \code{adp} objects
and notes on the many functions dealing with them.

Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley and Clark Richards
}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/section.R
\name{sectionAddStation}
\alias{sectionAddStation}
\alias{sectionAddCtd}
\title{Add a CTD Station to a Section}
\usage{
sectionAddStation(section, station)
}
\arguments{
\item{section}{A section to which a station is to be added.}

\item{station}{A ctd object holding data for the station to be added.}
}
\value{
A \linkS4class{section} object.
}
\description{
Add a CTD profile to an existing section.
}
\section{Historical note}{

Until March 2015, this operation was carried out with the \code{+} operator,
but at that time, the syntax was flagged by the development version of R, so it
was changed to the present form.
}

\examples{
library(oce)
data(ctd)
ctd2 <- ctd
ctd2[["temperature"]] <- ctd2[["temperature"]] + 0.5
ctd2[["latitude"]] <- ctd2[["latitude"]] + 0.1
section <- as.section(c("ctd", "ctd2"))
ctd3 <- ctd
ctd3[["temperature"]] <- ctd[["temperature"]] + 1
ctd3[["latitude"]] <- ctd[["latitude"]] + 0.1
ctd3[["station"]] <- "Stn 3"
sectionAddStation(section, ctd3)

}
\seealso{
Other things related to section data: 
\code{\link{[[,section-method}},
\code{\link{[[<-,section-method}},
\code{\link{as.section}()},
\code{\link{handleFlags,section-method}},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{plot,section-method}},
\code{\link{read.section}()},
\code{\link{section-class}},
\code{\link{sectionGrid}()},
\code{\link{sectionSmooth}()},
\code{\link{sectionSort}()},
\code{\link{section}},
\code{\link{subset,section-method}},
\code{\link{summary,section-method}}
}
\author{
Dan Kelley
}
\concept{things related to section data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.itp.R
\name{read.ctd.itp}
\alias{read.ctd.itp}
\title{Read an ITP-type CTD File}
\usage{
read.ctd.itp(
  file,
  columns = NULL,
  station = NULL,
  missingValue,
  deploymentType = "unknown",
  monitor = FALSE,
  debug = getOption("oceDebug"),
  processingLog,
  ...
)
}
\arguments{
\item{file}{a connection or a character string giving the name of the file to
load.  For \code{\link[=read.ctd.sbe]{read.ctd.sbe()}} and \code{\link[=read.ctd.woce]{read.ctd.woce()}}, this may be a
wildcard (e.g. \code{"*.cnv"} or \code{"*.csv"}) in which case the return
value is a vector containing CTD objects created by reading the files from
\code{\link[=list.files]{list.files()}} with \code{pattern} set to the specified wildcard
pattern.}

\item{columns}{an optional \link{list} that can be used to convert unrecognized
data names to resultant variable names.  This is used only by
\code{\link[=read.ctd.sbe]{read.ctd.sbe()}} and \code{\link[=read.ctd.odf]{read.ctd.odf()}}. For example,
if a data file named salinity as \code{"SAL"}, then using\preformatted{d <- read.ctd(f, columns=list(
    salinity=list(name="SAL",
                  unit=list(unit=expression(),
                  scale="PSS-78"))))
}

would assign the \code{"SAL"} column to the \code{salinity} entry in the data
slot of the CTD object returned by the \verb{read.*} function.}

\item{station}{optional character string containing an identifying name or
number for the station. This can be useful if the routine cannot determine the
name automatically, or if another name is preferred.}

\item{missingValue}{optional missing-value flag; data matching this value will
be set to \code{NA} upon reading. If this is provided, then it overrules any
missing-value flag found in the data. For Seabird (\code{.cnv}) files, there is
usually no need to set \code{missingValue}, because it can be inferred from the
header (typically as -9.990e-29). Set \code{missingValue=NULL} to turn off
missing-value detection, even in \code{.cnv} files that contain missing-value
codes in their headers. If \code{missingValue} is not specified,
then an attempt is made to infer such a value from the data, by testing
whether salinity and/or temperature has a minimum that is under -8 in value;
this should catch common values in files, without false positives. A warning
will be issued in this case, and a note inserted in the processing log of
the return value.}

\item{deploymentType}{character string indicating the type of deployment. Use
\code{"unknown"} if this is not known, \code{"profile"} for a profile (in
which the data were acquired during a downcast, while the device was lowered
into the water column, perhaps also including an upcast; \code{"moored"} if
the device is installed on a fixed mooring, \code{"thermosalinograph"} (or
\code{"tsg"}) if the device is mounted on a moving vessel, to record
near-surface properties, or \code{"towyo"} if the device is repeatedly
lowered and raised.}

\item{monitor}{boolean, set to \code{TRUE} to provide an indication of
progress.  This is useful if \code{filename} is a wildcard.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed.}

\item{processingLog}{if provided, the action item to be stored in the log.
This is typically only provided for internal calls; the default that it provides is
better for normal calls by a user.}

\item{...}{additional arguments, passed to called routines.}
}
\value{
a \linkS4class{ctd} object. The details of the contents
depend on the source file. The \code{metadata} slot is particularly
variable across data formats, because the meta-information provided
in those formats varies so widely.
}
\description{
Read an ITP-type CTD File
}
\references{
Information about ice-tethered profile data is provided at
\code{https://www.whoi.edu/page.do?pid=23096}, which also provides a link for
downloading data.  Note that the present version only handles data in
profiler-mode, not fixed-depth mode.
}
\seealso{
Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}

Other functions that read ctd data: 
\code{\link{read.ctd.aml}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()}
}
\author{
Dan Kelley

\code{read.ctd.itp} reads ice-tethered-profiler data that are stored
in a format files used by WHOI servers as of 2016-2017. Lacking
documentation on the format, the author constructed this function
to work with some files that were on-hand. Whether the function will
prove robust is an open question.
}
\concept{functions that read ctd data}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/amsr.R
\name{[[,amsr-method}
\alias{[[,amsr-method}
\title{Extract Something From an amsr Object}
\usage{
\S4method{[[}{amsr}(x, i, j, ...)
}
\arguments{
\item{x}{an \linkS4class{amsr} object.}

\item{i}{character value indicating the name of an item to extract.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{ignored.}
}
\value{
In all cases, the returned value is a matrix with
with \code{NA} values inserted at locations where
the raw data equal \code{as.raw(251:255)}, as explained
in \dQuote{Details}.
}
\description{
Generally, the \code{[[} method lets users extract information from \code{oce}
objects, without having to know the details of the internal storage.  For
many \code{oce} sub-classes, \code{[[} can also return quantities that are computed
from the object's contents.
}
\details{
A two-step process is used to try to find the requested information. First, a
class-specific function is used (see \dQuote{Details of the Specialized
Method}).  If this yields nothing, then a general method is used (see
\dQuote{Details of the General Method}).  If both methods fail, then \code{[[}
returns NULL.

Some understanding of the subclass is required to know what can be retrieved
with \code{[[}. When dealing with an unfamiliar subclass, it can be useful to
first use \code{x[["?"]]} to get a listing of the retrievable items. See
\dQuote{Details of the Specialized Method} for more information.
}
\section{Details of the Specialized Method}{


If \code{i} is \code{"?"}, then the return value is a list
containing four items, each of which is a character vector
holding the names of things that can be accessed with \code{[[}.
The \code{data} and \code{metadata} items hold the names of
entries in the object's data and metadata
slots, respectively. The \code{dataDerived}
and \code{metadataDerived} items are each NULL, because
no derived values are defined by \linkS4class{cm} objects.

Data within the \code{data} slot may be found directly, e.g.
\code{i="SSTDay"} will yield sea-surface temperature in the daytime
satellite, and \code{i="SSTNight"} is used to access the nighttime data. In
addition, \code{i="SST"} yields a computed average of the night and day values
(using just one of these, if the other is missing). This scheme of
providing computed averages works for
all the data stored in \code{amsr} objects, namely:
\code{time}, \code{SST}, \code{LFwind}, \code{MFwind},
\code{vapor}, \code{cloud} and \code{rain}.  In each case, the default
is to calculate values in scientific units, unless \code{j="raw"}, in
which case the raw data are returned.

The conversion from raw to scientific units is done with formulae
found at \verb{http://www.remss.com/missions/amsre}, e.g. SST is
computed by converting the raw value to an integer (between 0 and 255),
multiplying by 0.15C, and subtracting 3C.

The \code{"raw"} mode can be useful
in decoding the various types of missing value that are used by \code{amsr}
data, namely \code{as.raw(255)} for land, \code{as.raw(254)} for
a missing observation, \code{as.raw(253)} for a bad observation,
\code{as.raw(252)} for sea ice, or \code{as.raw(251)} for missing SST
due to rain or missing water vapour due to heavy rain. Note that
something special has to be done for e.g. \code{d[["SST","raw"]]}
because the idea is that this syntax (as opposed to specifying
\code{"SSTDay"}) is a request to try to find good
data by looking at both the Day and Night measurements. The scheme
employed is quite detailed. Denote by "A" the raw value of the desired field
in the daytime pass, and by "B" the corresponding value in the
nighttime pass. If either A or B is 255, the code for land, then the
result will be 255. If A is 254 (i.e. there is no observation),
then B is returned, and the reverse holds also. Similarly, if either
A or B equals 253 (bad observation), then the other is returned.
The same is done for code 252 (ice) and code 251 (rain).
}

\section{Details of the General Method}{


Note: the text of this section is identical for all \code{oce} subclasses, and so
some of what you read here may not be relevant to the class being described
in this help page.

If the specialized method produces no matches, the following generalized
method is applied. As with the specialized method, the procedure hinges first
on the values of \code{i} and, optionally, \code{j}. The work proceeds in steps, by
testing a sequence of possible conditions in sequence.
\enumerate{
\item A check is made as to whether \code{i} names one of the standard \code{oce} slots.
If so, \code{[[} returns the slot contents of that slot.  Thus, \code{x[["metadata"]]}
will retrieve the \code{metadata} slot, while \code{x[["data"]]} and
\code{x[["processingLog"]]} return those slots.
\item If \code{i} is a string ending in the \code{"Unit"}, then the characters preceding
that string are taken to be the name of an item in the data object, and a
list containing the unit is returned (or \code{NULL} if there is no such unit).
This list consists of an item named \code{unit}, which is an \code{\link[=expression]{expression()}}, and
an item named \code{scale}, which is a string describing the measurement scale.
If the string ends in \code{" unit"}, e.g.  \code{x[["temperature unit"]]} (note the
space), then just the expression is returned, and if it ends in \code{" scale"},
then just the scale is returned.
\item If \code{i} is a string ending in \code{"Flag"}, then the corresponding data-quality
flag is returned (or \code{NULL} if there is no such flag).
\item If the object holds hydrographic information (pressure, salinity,
temperature, longitude and latitude) then another set of possibilities
arises.  If \code{i} is \code{"sigmaTheta"}, then the value of \code{\link[=swSigmaTheta]{swSigmaTheta()}} is
called with \code{x} as the sole argument, and the results are returned.
Similarly, \code{\link[=swSigma0]{swSigma0()}} is used if \code{i="sigma0"}, and \code{\link[=swSpice]{swSpice()}} is used if
\code{i="spice"}. Of course, these actions only make sense for objects that
contain the relevant items within their \code{data} slot.
\item After these possibilities are eliminated, the action depends on whether
\code{j} has been provided.  If \code{j} is not provided, or is the string \code{""}, then
\code{i} is sought in the \code{metadata} slot, and then in the \code{data} slot, returning
whichever is found first.  In other words, if \code{j} is not provided, the
\code{metadata} slot takes preference over the \code{data} slot. However, if \code{j} is
provided, then it must be either the string \code{"metadata"} or \code{"data"}, and it
directs where to look.
\item If none of the above-listed conditions holds, then \code{NULL} is returned.
}
}

\examples{
# Histogram of SST values
library(oce)
data(amsr)
hist(amsr[["SST"]])

}
\seealso{
Other functions that extract parts of oce objects: 
\code{\link{[[,adp-method}},
\code{\link{[[,adv-method}},
\code{\link{[[,argo-method}},
\code{\link{[[,bremen-method}},
\code{\link{[[,cm-method}},
\code{\link{[[,coastline-method}},
\code{\link{[[,ctd-method}},
\code{\link{[[,echosounder-method}},
\code{\link{[[,g1sst-method}},
\code{\link{[[,gps-method}},
\code{\link{[[,ladp-method}},
\code{\link{[[,landsat-method}},
\code{\link{[[,lisst-method}},
\code{\link{[[,lobo-method}},
\code{\link{[[,met-method}},
\code{\link{[[,oce-method}},
\code{\link{[[,odf-method}},
\code{\link{[[,rsk-method}},
\code{\link{[[,sealevel-method}},
\code{\link{[[,section-method}},
\code{\link{[[,tidem-method}},
\code{\link{[[,topo-method}},
\code{\link{[[,windrose-method}},
\code{\link{[[,xbt-method}},
\code{\link{[[<-,adv-method}}

Other things related to amsr data: 
\code{\link{[[<-,amsr-method}},
\code{\link{amsr-class}},
\code{\link{amsr}},
\code{\link{composite,amsr-method}},
\code{\link{download.amsr}()},
\code{\link{plot,amsr-method}},
\code{\link{read.amsr}()},
\code{\link{subset,amsr-method}},
\code{\link{summary,amsr-method}}
}
\author{
Dan Kelley
}
\concept{functions that extract parts of oce objects}
\concept{things related to amsr data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/lobo.R
\name{[[,lobo-method}
\alias{[[,lobo-method}
\title{Extract Something From a LOBO Object}
\usage{
\S4method{[[}{lobo}(x, i, j, ...)
}
\arguments{
\item{x}{a \linkS4class{lobo} object.}

\item{i}{character value indicating the name of an item to extract.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{ignored.}
}
\description{
Generally, the \code{[[} method lets users extract information from \code{oce}
objects, without having to know the details of the internal storage.  For
many \code{oce} sub-classes, \code{[[} can also return quantities that are computed
from the object's contents.
}
\details{
A two-step process is used to try to find the requested information. First, a
class-specific function is used (see \dQuote{Details of the Specialized
Method}).  If this yields nothing, then a general method is used (see
\dQuote{Details of the General Method}).  If both methods fail, then \code{[[}
returns NULL.

Some understanding of the subclass is required to know what can be retrieved
with \code{[[}. When dealing with an unfamiliar subclass, it can be useful to
first use \code{x[["?"]]} to get a listing of the retrievable items. See
\dQuote{Details of the Specialized Method} for more information.
}
\section{Details of the Specialized Method}{

\itemize{
\item If \code{i} is \code{"?"}, then the return value is a list
containing four items, each of which is a character vector
holding the names of things that can be accessed with \code{[[}.
The \code{data} and \code{metadata} items hold the names of
entries in the object's data and metadata
slots, respectively. The \code{dataDerived}
and \code{metadataDerived} items are each NULL, because
no derived values are defined by \code{cm} objects.
}
}

\section{Details of the General Method}{


Note: the text of this section is identical for all \code{oce} subclasses, and so
some of what you read here may not be relevant to the class being described
in this help page.

If the specialized method produces no matches, the following generalized
method is applied. As with the specialized method, the procedure hinges first
on the values of \code{i} and, optionally, \code{j}. The work proceeds in steps, by
testing a sequence of possible conditions in sequence.
\enumerate{
\item A check is made as to whether \code{i} names one of the standard \code{oce} slots.
If so, \code{[[} returns the slot contents of that slot.  Thus, \code{x[["metadata"]]}
will retrieve the \code{metadata} slot, while \code{x[["data"]]} and
\code{x[["processingLog"]]} return those slots.
\item If \code{i} is a string ending in the \code{"Unit"}, then the characters preceding
that string are taken to be the name of an item in the data object, and a
list containing the unit is returned (or \code{NULL} if there is no such unit).
This list consists of an item named \code{unit}, which is an \code{\link[=expression]{expression()}}, and
an item named \code{scale}, which is a string describing the measurement scale.
If the string ends in \code{" unit"}, e.g.  \code{x[["temperature unit"]]} (note the
space), then just the expression is returned, and if it ends in \code{" scale"},
then just the scale is returned.
\item If \code{i} is a string ending in \code{"Flag"}, then the corresponding data-quality
flag is returned (or \code{NULL} if there is no such flag).
\item If the object holds hydrographic information (pressure, salinity,
temperature, longitude and latitude) then another set of possibilities
arises.  If \code{i} is \code{"sigmaTheta"}, then the value of \code{\link[=swSigmaTheta]{swSigmaTheta()}} is
called with \code{x} as the sole argument, and the results are returned.
Similarly, \code{\link[=swSigma0]{swSigma0()}} is used if \code{i="sigma0"}, and \code{\link[=swSpice]{swSpice()}} is used if
\code{i="spice"}. Of course, these actions only make sense for objects that
contain the relevant items within their \code{data} slot.
\item After these possibilities are eliminated, the action depends on whether
\code{j} has been provided.  If \code{j} is not provided, or is the string \code{""}, then
\code{i} is sought in the \code{metadata} slot, and then in the \code{data} slot, returning
whichever is found first.  In other words, if \code{j} is not provided, the
\code{metadata} slot takes preference over the \code{data} slot. However, if \code{j} is
provided, then it must be either the string \code{"metadata"} or \code{"data"}, and it
directs where to look.
\item If none of the above-listed conditions holds, then \code{NULL} is returned.
}
}

\seealso{
Other functions that extract parts of oce objects: 
\code{\link{[[,adp-method}},
\code{\link{[[,adv-method}},
\code{\link{[[,amsr-method}},
\code{\link{[[,argo-method}},
\code{\link{[[,bremen-method}},
\code{\link{[[,cm-method}},
\code{\link{[[,coastline-method}},
\code{\link{[[,ctd-method}},
\code{\link{[[,echosounder-method}},
\code{\link{[[,g1sst-method}},
\code{\link{[[,gps-method}},
\code{\link{[[,ladp-method}},
\code{\link{[[,landsat-method}},
\code{\link{[[,lisst-method}},
\code{\link{[[,met-method}},
\code{\link{[[,oce-method}},
\code{\link{[[,odf-method}},
\code{\link{[[,rsk-method}},
\code{\link{[[,sealevel-method}},
\code{\link{[[,section-method}},
\code{\link{[[,tidem-method}},
\code{\link{[[,topo-method}},
\code{\link{[[,windrose-method}},
\code{\link{[[,xbt-method}},
\code{\link{[[<-,adv-method}}

Other things related to lobo data: 
\code{\link{[[<-,lobo-method}},
\code{\link{as.lobo}()},
\code{\link{lobo-class}},
\code{\link{lobo}},
\code{\link{plot,lobo-method}},
\code{\link{read.lobo}()},
\code{\link{subset,lobo-method}},
\code{\link{summary,lobo-method}}
}
\author{
Dan Kelley
}
\concept{functions that extract parts of oce objects}
\concept{things related to lobo data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adv.R
\name{read.adv}
\alias{read.adv}
\title{Read an ADV data file}
\usage{
read.adv(
  file,
  from = 1,
  to,
  by = 1,
  tz = getOption("oceTz"),
  type = c("nortek", "sontek", "sontek.adr", "sontek.text"),
  header = TRUE,
  longitude = NA,
  latitude = NA,
  start = NULL,
  deltat = NA,
  debug = getOption("oceDebug"),
  monitor = FALSE,
  processingLog = NULL
)
}
\arguments{
\item{file}{a connection or a character string giving the name of the file
to load.  It is also possible to give \code{file} as a vector of filenames,
to handle the case of data split into files by a data logger.  In the
multi-file case, \code{header} must be \code{FALSE}, \code{start} must be a
vector of times, and \code{deltat} must be provided.}

\item{from}{index number of the first profile to be read, or the time of
that profile, as created with \code{\link[=as.POSIXct]{as.POSIXct()}} (hint: use
\code{tz="UTC"}).  This argument is ignored if \code{header==FALSE}. See
\dQuote{Examples}.}

\item{to}{indication of the last profile to read, in a format matching that
of \code{from}.  This is ignored if \code{header==FALSE}.}

\item{by}{an indication of the stride length to use while walking through
the file.  This is ignored if \code{header==FALSE}.  Otherwise, if this is
an integer, then \code{by-1} profiles are skipped between each pair of
profiles that is read. This may not make much sense, if the data are not
equi-spaced in time.  If \code{by} is a string representing a time interval,
in colon-separated format, then this interval is divided by the sampling
interval, to get the stride length. \emph{BUG:} \code{by} only partially
works; see \dQuote{Bugs}.}

\item{tz}{character string indicating time zone to be assumed in the data.}

\item{type}{character string indicating type of file, and used by
\code{read.adv} to dispatch to one of the speciality functions.}

\item{header}{A logical value indicating whether the file starts with a header.
(This will not be the case for files that are created by data loggers that
chop the raw data up into a series of sub-files, e.g. once per hour.)}

\item{longitude}{optional signed number indicating the longitude in degrees
East.}

\item{latitude}{optional signed number indicating the latitude in degrees
North.}

\item{start}{the time of the first sample, typically created with
\code{\link[=as.POSIXct]{as.POSIXct()}}.  This may be a vector of times,
if \code{filename} is a vector of file names.}

\item{deltat}{the time between samples. (This is mandatory if
\code{header=FALSE}.)}

\item{debug}{a flag that turns on debugging.  The value indicates the depth
within the call stack to which debugging applies.  For example,
\code{read.adv.nortek()} calls \code{read.header.nortek()}, so that
\code{read.adv.nortek(...,debug=2)} provides information about not just the
main body of the data file, but also the details of the header.}

\item{monitor}{boolean value indicating whether to indicate the progress
of reading the file, by using \code{\link[=txtProgressBar]{txtProgressBar()}} or otherwise.  The value
of \code{monitor} is changed to \code{FALSE} automatically, for non-interactive
sessions.}

\item{processingLog}{if provided, the action item to be stored in the log.
This parameter is typically only provided for internal calls; the default
that it provides is better for normal calls by a user.}
}
\value{
An \linkS4class{adv} object that contains measurements made with an ADV device.

The \code{metadata} contains information as given in the following table.
The \verb{Nortek name'' is the name used in the Nortek System Integrator Guide (reference 1) and the }Sontek name'' is the name used in the relevant
Sontek documentation.  References are given in square brackets.

\tabular{llll}{

\strong{\code{metadata} name}\tab \strong{Nortek name} \tab \strong{Sontek name} \tab \strong{Meaning}\cr

\code{manufacturer}\tab - \tab - \tab Either \code{"nortek"} or \code{"sontek"}\cr

\code{instrumentType}\tab  - \tab - \tab Either \code{"vector"} or \code{"adv"}\cr

\code{filename}\tab  - \tab - \tab Name of data file(s)\cr

\code{latitude}\tab - \tab - \tab  Latitude of mooring (if applicable)\cr

\code{longitude}\tab - \tab - \tab Longitude of mooring (if applicable)\cr

\code{numberOfSamples}\tab - \tab - \tab Number of data samples in file\cr

\code{numberOfBeams}\tab NBeams (reference 1, p18) \tab - \tab Number of beams (always 3)\cr

\code{numberOfBeamSequencesPerBurst}\tab NPings \tab - \tab number of beam sequences per burst\cr

\code{measurementInterval}\tab MeasInterval (reference 1 p31) \tab - \tab \cr

\code{samplingRate}\tab 512/(AvgInterval) (reference 1 p30; reference 4) \tab - #' \tab \cr

}

The \code{data} list contains items with names corresponding to \code{adp}
objects, with an exception for Nortek data.  Nortek instruments report some
things at a time interval that is longer than the velocity sampling, and
these are stored in \code{data} as \code{timeSlow}, \code{headingSlow},
\code{pitchSlow}, \code{rollSlow}, and \code{temperatureSlow}; if burst
sampling was used, there will also be items \code{recordsBurst} and
\code{timeBurst}.

The \code{processingLog} is in the standard format.
}
\description{
Read an ADV data file, producing an object of type \code{adv}. This
function works by transferring control to a more specialized function,
e.g. \code{\link[=read.adp.nortek]{read.adp.nortek()}} and \code{\link[=read.adp.sontek]{read.adp.sontek()}},
and in many cases users will find it preferable to either use these or
the several even more specialized functions, if the file type is
known.
}
\details{
Files \emph{without} headers may be created in experiments in which a data
logger was set up to monitor the serial data stream from an instrument.  The
lack of header information places a burden on the user, who must supply such
basic information as the times of observations, the instrument orientation,
the instrument coordinate system, etc.  Example 3 below shows how to deal
with such files.  Three things should be noted.
\enumerate{
\item The user must choose the appropriate \code{read.adv} variant
corresponding to the instrument in question.  (This is necessary because
\code{\link[=oceMagic]{oceMagic()}}, which is used by the generic \code{\link[=read.oce]{read.oce()}}
routine, cannot determine the type of instrument by examining a file that
lacks a header.)
\item The call to the \code{read} function must include a start time
(\code{start}) and the number of seconds between data (\code{deltat}),
again, because the instrument data stream may lack those things when the
device is set to a serial mode.  Also, of course, it is necessary to set
\code{header=FALSE} in the function call.
\item Once the file has been read in, the user will be obliged to specify
other information, for the object to be well-formed.  For example, the
\code{read} function will have no way of knowing the instrument orientation,
the coordinate system being used, the transformation matrix to go from
\code{"beam"} to \code{"xyz"} coordinates, or the instrument heading, pitch,
and roll, to go from \code{"xyz"} coordinates to \code{"enu"} coordinates.
Such things are illustrated in example 3 below.
}

In ADV data files, velocities are coded to signed 2-byte integers, with a
scale factor being used to convert to velocity in metres per second.  These
two facts control the maximum recordable velocity and the velocity
resolution, values that may be retrieved for an ADV object name \code{d}
with \code{d[["velocityMaximum"]]} and \code{d[["velocityResolution"]]}.
}
\section{Nortek files}{


\strong{Sampling-rate and similar issues}

The data format is inferred from the System Integrator Guide (reference 1A) and System
Integrator Manual (reference 1B).  These document lacks clarity in spots, and so
\code{read.adv.nortek} contains some assumptions that are noted here, so
that users will be aware of possible problems.

A prominent example is the specification of the sampling rate, stored in
\code{metadata$sampingRate} in the return value.  Repeated examination of
the System Integrator Guide (reference 1) failed to indicate where this value is
stored in the various headers contained in Vector datasets.  After some
experimentation with a few data files, \code{read.adv.nortek} was set up to
calculate \code{metadata$samplingRate} as \code{512/AvgInterval} where
\code{AvgInterval} is a part of the \verb{User Configuration'' header (reference 1 p30), where the explanation is }average interval in seconds'').  This formula was
developed through trial and error, but it was confirmed later on the Nortek
discussion group, and it should appear in upcoming versions of (reference 1).

Another basic issue is the determination of whether an instrument had
recorded in continuous mode or burst mode.  One might infer that
\code{TimCtrlReg} in the \verb{User Configuration'' header (reference 1 p30) determines this, in bits 1 and 2.  However, this was the case in test files available to the author.  For this reason, `read.adv.nortek` infers the mode by reverse engineering of data files of known configuration.  The present version of `read.adv.nortek` determines the sampling mode from the ```NRecords`'' item of the }Vector Velocity Data'' header, which seems
to be 0 for data collected continuously, and non-zero for data collected in
bursts.

Taking these things together, we come upon the issue of how to infer
sampling times for Nortek instruments.  There do not seem to be definitive
documents on this, and so \code{read.adv.nortek} is based partly on
information (of unknown quality) found on Nortek discussion boards.  The
present version of \code{read.adv.nortek} infers the times of velocity
observations differently, depending on whether the instrument was set to
record in burst mode or continuous mode.  For burst mode, times stated in
the burst headers are used, but for continuous mode, times stated in the
``vector system data'' are used.  On the advice found on a Nortek discussion
board, the burst-mode times are offset by 2 seconds to allow for the
instrument warm-up period.

\strong{Handling IMU (inertial measurement unit) data}

Starting in March
2016, \code{read.adv.nortek} has offered some support for handling IMU
(inertial measurement unit) data incorporated into Nortek binary files. This
is not described in the Nortek document named \verb{System Integrator Guide'' (reference 1A) but it appeared in }System Integrator Manual'' (reference 1B;
reference 1C). Confusingly, 1B described 3 varieties of data, whereas 1C does not
describe any of these, but describes instead a fourth variety. As of March
2016, \code{read.adv.nortek} handles all 4 varieties, because files in the
various schemes appear to exist. In \code{oce}, the varieties are named
after the byte code that flags them.  (Variety \code{c3} is the one
described in (reference 1C); the others were described in (reference 1B).)
The variety is stored in the \code{metadata} slot of the returned object as
a string named \code{IMUtype}.

For each variety, the reader is cautioned that strong tests have not been
performed on the code.  One way to test the code is to compare with textual
data files produced by the Nortek software.  In March 2016, an \code{oce}
user shared a dataset of the \code{c3} variety, and this permitted detailed
comparison between the text file and the values inferred by
\code{read.adv.nortek}. The test suggested agreement (to within the
resolution printed in the text file) for velocity (\code{v} in the
\code{data} slot), signal amplitude (\code{a}), correlation (\code{q}),
pressure (\code{p}), the three components of IMU delta angle
(\code{IMUdeltaAngleX} etc), and all components of the rotation matrix
(\code{IMUrotation}).  However, the delta velocity signals did not match,
with \code{IMUdeltaVelocityX} disagreeing in the second decimal place,
\code{IMUdeltaVelocityY} component disagreeing in the first, and
\code{IMUdeltaVelocityZ} being out by a factor of about 10. This is github
issue 893 (\url{https://github.com/dankelley/oce/issues/893}).
\itemize{
\item Variety \code{c3} (signalled by byte 5 of a sequence being
\code{0xc3}) provides information on what Nortek calls DeltaAngle,
DeltaVelocity and Orientation Matrix. (Apart from the orientation matrix,
Nortek provides no documentation on what these quantities mean.) In the
object returned by \code{read.adv.nortek}, these are stored in the
\code{data} slot as \code{IMUdeltaAngleX}, \code{IMUdeltaAngleY},
\code{IMUdeltaAngleZ}, \code{IMUdeltaVelocityX}, \code{IMUdeltaVelocityY},
\code{IMUdeltaVelocityZ}, and \code{IMUrotation}, all vectors except the
last, which is a 3D array. In addition to these, \code{IMUtimestamp} is a
time-stamp, which is not defined in the Nortek documents but seems, from IMU
documents (reference 5), to be defined based on a clock that ticks once per 16
microseconds. Caution may be required in dealing with this timestamp, since
it seemed sensible in one test case (variety \code{d3}) but kept reseting to
zero in another (variety \code{c3}). The lack of Nortek documentation on
most of these quantities is a roadblock to implementing \code{oce} functions
dealing with IMU-enabled datasets
\item Variety \code{cc} (signalled by byte 5 of a sequence being
\code{0xcc}) provides information on acceleration, angular rotation rate,
magnetic vector and orientation matrix.  Each is a timeseries. Acceleration
is stored in the \code{data} slot as \code{IMUaccelX}, \code{IMUaccelY},
\code{IMUaccelz}. The angular rotation components are \code{IMUngrtX},
\code{IMUngrtY} and \code{IMUngrtz}. The magnetic data are in
\code{IMUmagrtx}, \code{IMUmagrty} and \code{IMUmagrtz}. Finally,
\code{IMUmatrix} is a rotation matrix made up from elements named
\code{M11}, \code{M12}, etc in the Nortek documentation.  In addition to all
of these, \code{IMUtime} stores time in seconds, with an origin whose
definition is not stated in reference 1B.
\item Variety \code{d2} (signalled by byte 5 being \code{0xd2}) provides
information on gyro-stabilized acceleration, angular rate and magnetometer
vectors.  The data stored \code{MUaccelX}, \code{IMUangrtX},
\code{IMUmagrtX}, with similar for \code{Y} and \code{Z}.  Again, time is in
\code{IMUtime}. This data type has not been tested as of mid-March 2016,
because the developers do not have a test file with which to test.
\item Variety \code{d3} (signalled by byte 5 being \code{0xd3}) provides
information on DeltaAngle, DeltaVelocity and magnetometer vectors, stored in
\code{IMUdeltaAngleX}, \code{IMUdeltaVelocityX}, and
\code{IMUdeltaMagVectorX}, with similar for \code{Y} and \code{Z}. Again,
time is in \code{IMUtime}. This data type has not been tested as of
mid-March 2016, because the developers do not have a test file with which to
test.
}
}

\examples{
\dontrun{
library(oce)
# A nortek Vector file
d <- read.oce("/data/archive/sleiwex/2008/moorings/m05/adv/nortek_1943/raw/adv_nortek_1943.vec",
              from=as.POSIXct("2008-06-26 00:00:00", tz="UTC"),
              to=as.POSIXct("2008-06-26 00:00:10", tz="UTC"))
plot(d, which=c(1:3,15))
}
}
\references{
1A. Nortek AS.  System Integrator Guide (paradopp family of products). June
2008.  (Doc No: PSI00-0101-0608).  (Users may find it helpful to also
examine newer versions of the guide.)

1B. Nortek AS.  System Integrator Manual. Dec 2014.
(\code{system-integrator-manual_Dec2014_jan.pdf})

1C. Nortek AS.  System Integrator Manual. March 2016.
(\code{system-integrator-manual_Mar2016.pdf})
\enumerate{
\item SonTek/YSI ADVField/Hydra Acoustic Doppler Velocimeter (Field) Technical
Documentation (Sept 1, 2001).
\item Appendix 2.2.3 of the Sontek ADV operation Manual, Firmware Version 4.0
(Oct 1997).
\item Nortek Knowledge Center (http://www.nortekusa.com/en/knowledge-center)
\item A document describing an IMU unit that seems to be close to the one named
in (references 1B and C) as being an adjunct to Nortek Vector systems is at
\verb{http://files.microstrain.com/3DM-GX3-35-Data-Communications-Protocol.pdf}
}
}
\seealso{
Other things related to adv data: 
\code{\link{[[,adv-method}},
\code{\link{[[<-,adv-method}},
\code{\link{adv-class}},
\code{\link{adv}},
\code{\link{beamName}()},
\code{\link{beamToXyz}()},
\code{\link{enuToOtherAdv}()},
\code{\link{enuToOther}()},
\code{\link{plot,adv-method}},
\code{\link{read.adv.nortek}()},
\code{\link{read.adv.sontek.adr}()},
\code{\link{read.adv.sontek.serial}()},
\code{\link{read.adv.sontek.text}()},
\code{\link{rotateAboutZ}()},
\code{\link{subset,adv-method}},
\code{\link{summary,adv-method}},
\code{\link{toEnuAdv}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdv}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley
}
\concept{things related to adv data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{plotPolar}
\alias{plotPolar}
\title{Draw a Polar Plot}
\usage{
plotPolar(r, theta, debug = getOption("oceDebug"), ...)
}
\arguments{
\item{r}{radii of points to plot.}

\item{theta}{angles of points to plot, in degrees.}

\item{debug}{a flag that turns on debugging.  Set to 1 to get a moderate
amount of debugging information, or to 2 to get more.}

\item{\dots}{optional arguments passed to the lower-level plotting
functions.}
}
\description{
Creates a crude polar plot.
}
\examples{
library(oce)
r <- rnorm(50, mean=2, sd=0.1)
theta <- runif(50, 0, 360)
plotPolar(r, theta)
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/echosounder.R
\docType{class}
\name{echosounder-class}
\alias{echosounder-class}
\title{Class to Store Echosounder Data}
\description{
This class stores echosounder data. Echosounder objects may be
read with \code{\link[=read.echosounder]{read.echosounder()}},
summarized with \code{\link[=summary,echosounder-method]{summary,echosounder-method()}},
and plotted with \code{\link[=plot,echosounder-method]{plot,echosounder-method()}}.
The \code{\link[=findBottom]{findBottom()}}
function infers the ocean bottom from tracing the strongest reflector from
ping to ping.
}
\details{
\itemize{
\item An infrequently updated record of the instrument position, in
\code{timeSlow}, \code{longitudeSlow} and \code{latitudeSlow}.  These are
used in plotting maps with \code{\link[=plot,echosounder-method]{plot,echosounder-method()}}.
\item An interpolated record of the instrument position, in \code{time},
\code{longitude}, and \code{latitude}.  Linear interpolation is used to
infer the longitude and latitude from the variables listed above.
\item \code{depth}, vector of depths of echo samples (measured positive
downwards in the water column).  This is calculated from the inter-sample
time interval and the sound speed provided as the \code{soundSpeed} argument
to \code{\link[=read.echosounder]{read.echosounder()}}, so altering the value of the latter will
alter the echosounder plots provided by \code{\link[=plot,echosounder-method]{plot,echosounder-method()}}.
\item The echosounder signal amplitude \code{a}, a matrix whose number of
rows matches the length of \code{time}, etc., and number of columns equal to
the length of \code{depth}.  Thus, for example, \code{a[100,]} represents
the depth-dependent amplitude at the time of the 100th ping.
\item A matrix named \code{b} exists for dual-beam and split-beam cases.
For dual-beam data, this is the wide-beam data, whereas \code{a} is the
narrow-beam data.  For split-beam data, this is the x-angle data.
\item A matrix named \code{c} exists for split-beam data, containing the
y-angle data.
\item In addition to these matrices, ad-hoc calculated matrices named
\code{Sv} and \code{TS} may be accessed as explained in the next section.
}
}
\section{Slots}{

\describe{
\item{\code{data}}{As with all \code{oce} objects, the \code{data} slot
for \code{echosounder} objects is a
\link{list} containing the main data for the object.
{}}

\item{\code{metadata}}{As with all \code{oce} objects, the \code{metadata} slot
for \code{echosounder} objects is a \link{list} containing
information about the \code{data} or about the object itself.
{}}

\item{\code{processingLog}}{As with all \code{oce} objects, the \code{processingLog} slot
for \code{echosounder} objects is a
\link{list} with entries describing the creation and evolution
of the object. The contents are updated by various \code{oce} functions to
keep a record of processing steps.  Object summaries and
\code{\link[=processingLogShow]{processingLogShow()}} both display the log.}
}}

\section{Modifying slot contents}{


Although the \code{[[<-} operator may permit modification of the contents
of \linkS4class{echosounder} objects (see \code{\link{[[<-,echosounder-method}}),
it is better to use \code{\link[=oceSetData]{oceSetData()}} and \code{\link[=oceSetMetadata]{oceSetMetadata()}},
because those functions save an entry in the \code{processingLog}
that describes the change.
}

\section{Retrieving slot contents}{


The full contents of the \code{data} and \code{metadata} slots of a \linkS4class{echosounder}
object may be retrieved in the standard R way using \code{\link[=slot]{slot()}}. For
example \code{slot(o,"data")} returns the \code{data} slot of an object named \code{o},
and similarly \code{slot(o,"metadata")} returns
the \code{metadata} slot.

The slots may also be obtained with the \code{\link{[[,echosounder-method}}
operator, as e.g. \code{o[["data"]]} and \code{o[["metadata"]]}, respectively.

The \code{\link{[[,echosounder-method}} operator can also
be used to retrieve items from within the \code{data} and \code{metadata} slots.
For example, \code{o[["temperature"]]} can be used to retrieve temperature
from an object containing that quantity. The rule is that a named
quantity is sought first within the object's \code{metadata} slot,
with the \code{data} slot being checked only if \code{metadata} does not
contain the item. This \code{[[} method can also be used to get
certain derived quantities, if the object contains sufficient
information to calculate them. For example, an object that holds
(practical) salinity, temperature and pressure, along with
longitude and latitude, has sufficient information to compute
Absolute Salinity, and so \code{o[["SA"]]} will yield the
calculated Absolute Salinity.

It is also possible to find items more directly, using \code{\link[=oceGetData]{oceGetData()}} and
\code{\link[=oceGetMetadata]{oceGetMetadata()}}, but neither of these functions can
retrieve derived items.
}

\seealso{
Other things related to echosounder data: 
\code{\link{[[,echosounder-method}},
\code{\link{[[<-,echosounder-method}},
\code{\link{as.echosounder}()},
\code{\link{echosounder}},
\code{\link{findBottom}()},
\code{\link{plot,echosounder-method}},
\code{\link{read.echosounder}()},
\code{\link{subset,echosounder-method}},
\code{\link{summary,echosounder-method}}
}
\author{
Dan Kelley
}
\concept{things related to echosounder data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/map.R
\name{mapLines}
\alias{mapLines}
\title{Add Lines to a Map}
\usage{
mapLines(longitude, latitude, greatCircle = FALSE, ...)
}
\arguments{
\item{longitude}{numeric vector of longitudes of points to be plotted, or an
object from which longitude and latitude can be inferred (e.g. a coastline
file, or the return value from \code{\link[=mapLocator]{mapLocator()}}), in which case the
following two arguments are ignored.}

\item{latitude}{vector of latitudes of points to be plotted.}

\item{greatCircle}{a logical value indicating whether to render line
segments as great circles.  (Ignored.)}

\item{\dots}{optional arguments passed to \code{\link[=lines]{lines()}}.}
}
\description{
Plot lines on an existing map, by analogy to \code{\link[=lines]{lines()}}.
}
\examples{
\donttest{
if (utils::packageVersion("sf") != "0.9.8") {
    # sf version 0.9-8 has a problem with this projection
    library(oce)
    data(coastlineWorld)
    mapPlot(coastlineWorld, type='l',
            longitudelim=c(-80, 10), latitudelim=c(0, 120),
            projection="+proj=ortho +lon_0=-40")
    lon <- c(-63.5744, 0.1062)             # Halifax CA to London UK
    lat <- c(44.6479, 51.5171)
    mapPoints(lon, lat, col='red')
    mapLines(lon, lat, col='red')
}
}

}
\seealso{
A map must first have been created with \code{\link[=mapPlot]{mapPlot()}}.

Other functions related to maps: 
\code{\link{formatPosition}()},
\code{\link{lonlat2map}()},
\code{\link{lonlat2utm}()},
\code{\link{map2lonlat}()},
\code{\link{mapArrows}()},
\code{\link{mapAxis}()},
\code{\link{mapContour}()},
\code{\link{mapCoordinateSystem}()},
\code{\link{mapDirectionField}()},
\code{\link{mapGrid}()},
\code{\link{mapImage}()},
\code{\link{mapLocator}()},
\code{\link{mapLongitudeLatitudeXY}()},
\code{\link{mapPlot}()},
\code{\link{mapPoints}()},
\code{\link{mapPolygon}()},
\code{\link{mapScalebar}()},
\code{\link{mapText}()},
\code{\link{mapTissot}()},
\code{\link{oceCRS}()},
\code{\link{shiftLongitude}()},
\code{\link{usrLonLat}()},
\code{\link{utm2lonlat}()}
}
\author{
Dan Kelley
}
\concept{functions related to maps}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/argo.R
\name{[[<-,argo-method}
\alias{[[<-,argo-method}
\title{Replace Parts of an Argo Object}
\usage{
\S4method{[[}{argo}(x, i, j, ...) <- value
}
\arguments{
\item{x}{an \linkS4class{argo} object.}

\item{i}{character value naming the item to replace.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{optional additional information (ignored).}

\item{value}{The value to be placed into \code{x}, somewhere.}
}
\description{
The \code{[[<-} method works for all \linkS4class{oce} objects.
The purpose, as with the related extraction method, \code{[[},
is to insulate users from the internal details of \linkS4class{oce}
objects, by looking for items within the various storage
slots of the object. Items not actually stored can also be
replaced, including units and data-quality
flags.
}
\details{
As with \code{[[} method, the procedure works in steps.

First, the \code{metadata} slot of \code{x} is checked to
see whether it contains something named with \code{i}.
If so, then the named item is replaced with \code{value}.

Otherwise, if the string value of \code{i} ends in \code{Unit}, then the
characters preceding that are taken as the name of a variable, and
the \code{metadata} slot of \code{x} is updated to store that unit, e.g.\preformatted{x[["temperatureUnits"]] <- list(unit=expression(degree*F),scale="")
}

Similarly, if \code{i} ends in \code{Flag}, then quality-control
flags are set up as defined by \code{result}, e.g.\preformatted{o[["temperatureFlags"]] <- c(2,4,2,2)
}

Otherwise, \code{\link[=pmatch]{pmatch()}} is used for a partial-string match with
the names of the items that are in the \code{data} slot of \code{x}.
The first item found (if any) is then updated to hold the value \code{result}.

If none of these conditions is met, a warning is issued.
}
\seealso{
Other functions that replace parts of oce objects: 
\code{\link{[[<-,adp-method}},
\code{\link{[[<-,amsr-method}},
\code{\link{[[<-,bremen-method}},
\code{\link{[[<-,cm-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{[[<-,echosounder-method}},
\code{\link{[[<-,g1sst-method}},
\code{\link{[[<-,gps-method}},
\code{\link{[[<-,ladp-method}},
\code{\link{[[<-,landsat-method}},
\code{\link{[[<-,lisst-method}},
\code{\link{[[<-,lobo-method}},
\code{\link{[[<-,met-method}},
\code{\link{[[<-,oce-method}},
\code{\link{[[<-,odf-method}},
\code{\link{[[<-,rsk-method}},
\code{\link{[[<-,sealevel-method}},
\code{\link{[[<-,section-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{[[<-,topo-method}},
\code{\link{[[<-,windrose-method}},
\code{\link{[[<-,xbt-method}}

Other things related to argo data: 
\code{\link{[[,argo-method}},
\code{\link{argo-class}},
\code{\link{argoGrid}()},
\code{\link{argoNames2oceNames}()},
\code{\link{argo}},
\code{\link{as.argo}()},
\code{\link{handleFlags,argo-method}},
\code{\link{plot,argo-method}},
\code{\link{read.argo}()},
\code{\link{subset,argo-method}},
\code{\link{summary,argo-method}}
}
\concept{functions that replace parts of oce objects}
\concept{things related to argo data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/argo.R
\name{argoGrid}
\alias{argoGrid}
\title{Grid Argo float data}
\usage{
argoGrid(argo, p, debug = getOption("oceDebug"), ...)
}
\arguments{
\item{argo}{A \code{argo} object to be gridded.}

\item{p}{Optional indication of the pressure levels to which interpolation
should be done.  If this is not supplied, the pressure levels will be
calculated based on the existing values, using medians. If \code{p="levitus"},
then pressures will be set to be those of the Levitus atlas, given by
\code{\link[=standardDepths]{standardDepths()}}, trimmed to the maximum pressure in \code{argo}.
If \code{p} is a single numerical value, it is taken as the number of
subdivisions to use in a call to \code{\link[=seq]{seq()}} that has range from 0 to the
maximum pressure in \code{argo}.  Finally, if a vector numerical values is
provided, then it is used as is.}

\item{debug}{A flag that turns on debugging.  Higher values provide deeper
debugging.}

\item{...}{Optional arguments to \code{\link[=approx]{approx()}}, which is used to do the
gridding.}
}
\value{
x an \linkS4class{argo} object.
}
\description{
Grid an Argo float, by interpolating to fixed pressure levels.
The gridding is done with \code{\link[=approx]{approx()}}.  If there is
sufficient user demand, other methods may be added, by analogy to
\code{\link[=sectionGrid]{sectionGrid()}}.
}
\section{A note about flags}{


Data-quality flags contained within the original object are ignored by this
function, and the returned value contains no such flags.  This is because such
flags represent an assessment of the original data, not of quantities derived
from those data.  This function produces a warning to this effect. The
recommended practice is to use \code{\link[=handleFlags]{handleFlags()}} or some other means to
deal with flags before calling the present function.
}

\examples{
library(oce)
data(argo)
g <- argoGrid(argo, p=seq(0, 100, 1))
par(mfrow=c(2,1))
t <- g[["time"]]
z <- -g[["pressure"]][,1]
## Set zlim because of spurious temperatures.
imagep(t, z, t(g[['temperature']]), ylim=c(-100,0), zlim=c(0,20))
imagep(t, z, t(g[['salinity']]), ylim=c(-100,0))

}
\seealso{
Other things related to argo data: 
\code{\link{[[,argo-method}},
\code{\link{[[<-,argo-method}},
\code{\link{argo-class}},
\code{\link{argoNames2oceNames}()},
\code{\link{argo}},
\code{\link{as.argo}()},
\code{\link{handleFlags,argo-method}},
\code{\link{plot,argo-method}},
\code{\link{read.argo}()},
\code{\link{subset,argo-method}},
\code{\link{summary,argo-method}}
}
\author{
Dan Kelley and Clark Richards
}
\concept{things related to argo data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/topo.R
\name{subset,topo-method}
\alias{subset,topo-method}
\title{Subset a Topo Object}
\usage{
\S4method{subset}{topo}(x, subset, ...)
}
\arguments{
\item{x}{a \linkS4class{topo} object.}

\item{subset}{A condition to be applied to the \code{data} portion of \code{x}.
See \sQuote{Details}.}

\item{...}{Ignored.}
}
\value{
A new \linkS4class{topo} object.
}
\description{
This function is somewhat analogous to \code{\link[=subset.data.frame]{subset.data.frame()}}.
Subsetting can be by \code{time} or \code{distance}, but these may not be
combined; use a sequence of calls to subset by both.
}
\examples{
## northern hemisphere
library(oce)
data(topoWorld)
plot(subset(topoWorld, latitude > 0))

}
\seealso{
Other things related to topo data: 
\code{\link{[[,topo-method}},
\code{\link{[[<-,topo-method}},
\code{\link{as.topo}()},
\code{\link{download.topo}()},
\code{\link{plot,topo-method}},
\code{\link{read.topo}()},
\code{\link{summary,topo-method}},
\code{\link{topo-class}},
\code{\link{topoInterpolate}()},
\code{\link{topoWorld}}

Other functions that subset oce objects: 
\code{\link{subset,adp-method}},
\code{\link{subset,adv-method}},
\code{\link{subset,amsr-method}},
\code{\link{subset,argo-method}},
\code{\link{subset,cm-method}},
\code{\link{subset,coastline-method}},
\code{\link{subset,ctd-method}},
\code{\link{subset,echosounder-method}},
\code{\link{subset,lobo-method}},
\code{\link{subset,met-method}},
\code{\link{subset,oce-method}},
\code{\link{subset,odf-method}},
\code{\link{subset,rsk-method}},
\code{\link{subset,sealevel-method}},
\code{\link{subset,section-method}},
\code{\link{subset,xbt-method}}
}
\author{
Dan Kelley
}
\concept{functions that subset oce objects}
\concept{things related to topo data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ad.R
\name{velocityStatistics}
\alias{velocityStatistics}
\title{Report Statistics of adp or adv Velocities}
\usage{
velocityStatistics(x, control, ...)
}
\arguments{
\item{x}{an \linkS4class{adp} or \linkS4class{adv} object.}

\item{control}{An optional \link{list} used to specify more information.
This is presently ignored for \code{adv} objects.  For \code{adp} objects, if
\code{control$bin} is an integer, it is taken as the bin to be selected
(otherwise, an average across bins is used).}

\item{\dots}{additional arguments that are used in the call to
\code{\link[=mean]{mean()}}.}
}
\value{
A list containing items the major and minor axes of the covariance
ellipse (\code{ellipseMajor} and \code{ellipseMinor}), the angle of the
major axis anticlockwise of the horizontal axis (\code{ellipseAngle}), and
the x and y components of the mean velocity (\code{uMean} and \code{vMean}).
}
\description{
Report statistics of ADP or ADV velocities, such as means and variance
ellipses.
}
\examples{
library(oce)
data(adp)
a <- velocityStatistics(adp)
print(a)
t <- seq(0, 2*pi, length.out=100)
theta <- a$ellipseAngle * pi / 180
y <- a$ellipseMajor * cos(t) * sin(theta) + a$ellipseMinor * sin(t) * cos(theta)
x <- a$ellipseMajor * cos(t) * cos(theta) - a$ellipseMinor * sin(t) * sin(theta)
plot(adp, which="uv+ellipse+arrow")
lines(x, y, col='blue', lty="dashed", lwd=5)
arrows(0, 0, a$uMean, a$vMean, lwd=5, length=1/10, col='blue', lty="dashed")

}
\seealso{
Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}

Other things related to adv data: 
\code{\link{[[,adv-method}},
\code{\link{[[<-,adv-method}},
\code{\link{adv-class}},
\code{\link{adv}},
\code{\link{beamName}()},
\code{\link{beamToXyz}()},
\code{\link{enuToOtherAdv}()},
\code{\link{enuToOther}()},
\code{\link{plot,adv-method}},
\code{\link{read.adv.nortek}()},
\code{\link{read.adv.sontek.adr}()},
\code{\link{read.adv.sontek.serial}()},
\code{\link{read.adv.sontek.text}()},
\code{\link{read.adv}()},
\code{\link{rotateAboutZ}()},
\code{\link{subset,adv-method}},
\code{\link{summary,adv-method}},
\code{\link{toEnuAdv}()},
\code{\link{toEnu}()},
\code{\link{xyzToEnuAdv}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley
}
\concept{things related to adp data}
\concept{things related to adv data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{formatCI}
\alias{formatCI}
\title{Confidence interval in parenthetic notation}
\usage{
formatCI(ci, style = c("+/-", "parentheses"), model, digits = NULL)
}
\arguments{
\item{ci}{optional vector of length 2 or 3.}

\item{style}{string indicating notation to be used.}

\item{model}{optional regression model, e.g. returned by \code{\link[=lm]{lm()}} or
\code{\link[=nls]{nls()}}.}

\item{digits}{optional number of digits to use; if not supplied,
[\code{getOption}[\code{("digits")} is used.}
}
\value{
If \code{ci} is given, the result is a character string with the
estimate and its uncertainty, in plus/minus or parenthetic notation.  If
\code{model} is given, the result is a 1-column matrix holding character
strings, with row names corresponding to the parameters of the model.
}
\description{
Format a confidence interval in parenthetic notation.
}
\details{
If a \code{model} is given, then \code{ci} is ignored, and a confidence
interval is calculated using \code{\link[=confint]{confint()}} with \code{level} set to
0.6914619.  This \code{level} corresponds to a range of plus or minus one
standard deviation, for the t distribution and a large number of degrees of
freedom (since \code{qt(0.6914619, 100000)} is 0.5).

If \code{model} is missing, \code{ci} must be provided.  If it contains 3
elements, then first and third elements are taken as the range of the
confidence interval (which by convention should use the \code{level} stated
in the previous paragraph), and the second element is taken as the central
value.  Alternatively, if \code{ci} has 2 elements, they are taken to be
bounds of the confidence interval and their mean is taken to be the central
value.

In the \verb{+/-} notation, e.g. \eqn{a \pm b}{a +/- b} means that the true
value lies between \eqn{a-b}{a-b} and \eqn{a+b}{a+b} with a high degree of
certainty.  Mills et al. (1993, section 4.1 on page 83) suggest that
\eqn{b}{b} should be set equal to 2 times the standard uncertainty or
standard deviation.  JCGM (2008, section 7.2.2 on pages 25 and 26), however,
suggest that \eqn{b}{b} should be set to the standard uncertainty, while
also recommending that the \eqn{\pm}{+/-} notation be avoided altogether.

The \code{parentheses} notation is often called the compact notation.  In
it, the digits in parentheses indicate the uncertainty in the corresponding
digits to their left, e.g. 12.34(3) means that the last digit (4) has an
uncertainty of 3.  However, as with the \eqn{\pm}{+/-} notation, different
authorities offer different advice on defining this uncertainty; Mills et
al. (1993, section 4.1 on page 83) provide an example in which the
parenthetic notation has the same value as the \eqn{\pm}{+/-} notation,
while JCM (2008, section 7.2.2 on pages 25 and 26) suggest halving the
number put in parentheses.

The \code{foramtci} function is based on the JCM (2008) notation, i.e.
\code{formatCI(ci=c(8,12), style="+/-")} yields \code{"10+-2"}, and
\code{formatCI(ci=c(8,12), style="parentheses")} yields \code{"10(2)"}.

\strong{Note:} if the confidence range exceeds the value, the
\code{parentheses} format reverts to \verb{+/-} format.
}
\examples{
x <- seq(0, 1, length.out=300)
y <- rnorm(n=300, mean=10, sd=1) * x
m <- lm(y~x)
print(formatCI(model=m))
}
\references{
JCGM, 2008. \emph{Evaluation of measurement data - Guide to the
expression of uncertainty in measurement (JCGM 100:2008)}, published by the
Joint Committee for Guides in Metrology,
http://www.bipm.org/en/publications/guides/gum.html (see section
7.2.2 for a summary of notation, which shows equal values to the right of a
\verb{+-} sign and in parentheses.

I. Mills, T. Cvitas, K. Homann, N. Kallay, and K. Kuchitsu, 1993.
\emph{Quantities, Units and Symbols in Physical Chemistry}, published
Blackwell Science for the International Union of Pure and Applied Chemistry.
(See section 4.1, page 83, for a summary of notation, which shows that a
value to the right of a \verb{+-} sign is to be halved if put in
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{ocePmatch}
\alias{ocePmatch}
\alias{oce.pmatch}
\title{Partial matching of strings or numbers}
\usage{
ocePmatch(x, table, nomatch = NA_integer_, duplicates.ok = FALSE)
}
\arguments{
\item{x}{a code, or vector of codes.  This may be numeric, in which case it
is simply returned without further analysis of the other arguments, or it
may be string-based, in which case \code{\link[=pmatch]{pmatch()}} is used to find
numeric matches.}

\item{table}{a list that maps strings to numbers; \code{\link[=pmatch]{pmatch()}} is
used on \code{names(table)}.  If the name contains characters that are
normally not permitted in a variable name, use quotes, e.g.
\code{list(salinity=1, temperature=2, "salinity+temperature"=3)}.}

\item{nomatch}{value to be returned for cases of no match (passed to
\code{\link[=pmatch]{pmatch()}}.}

\item{duplicates.ok}{code for the handling of duplicates (passed to
\code{\link[=pmatch]{pmatch()}}).}
}
\value{
A number, or vector of numbers, corresponding to the matches.
Non-matches are indicated with \code{NA} values, or whatever value is given
by the \code{NA} argument.
}
\description{
An extended version of \code{\link[=pmatch]{pmatch()}} that allows \code{x} to be
numeric or string-based.  As with \code{\link[=pmatch]{pmatch()}}, partial string
matches are handled.
This is a wrapper that is useful mainly for \code{which} arguments to
plotting functions.
}
\examples{
library(oce)
oce.pmatch(c("s", "at", "te"), list(salinity=1, temperature=3.1))
}
\seealso{
Since \code{\link[=pmatch]{pmatch()}} is used for the actual matching, its
documentation should be consulted.
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/coastline.R
\docType{class}
\name{coastline-class}
\alias{coastline-class}
\title{Class to Store Coastline Data}
\description{
This class stores coastline data.
}
\section{Slots}{

\describe{
\item{\code{data}}{As with all \code{oce} objects, the \code{data} slot
for \code{coastline} objects is a
\link{list} containing the main data for the object.
The key items stored in this slot are \code{longitude} and \code{latitude}.}

\item{\code{metadata}}{As with all \code{oce} objects, the \code{metadata} slot
for \code{coastline} objects is a \link{list} containing
information about the \code{data} or about the object itself.
{}}

\item{\code{processingLog}}{As with all \code{oce} objects, the \code{processingLog} slot
for \code{coastline} objects is a
\link{list} with entries describing the creation and evolution
of the object. The contents are updated by various \code{oce} functions to
keep a record of processing steps.  Object summaries and
\code{\link[=processingLogShow]{processingLogShow()}} both display the log.}
}}

\section{Modifying slot contents}{


Although the \code{[[<-} operator may permit modification of the contents
of \linkS4class{coastline} objects (see \code{\link{[[<-,coastline-method}}),
it is better to use \code{\link[=oceSetData]{oceSetData()}} and \code{\link[=oceSetMetadata]{oceSetMetadata()}},
because those functions save an entry in the \code{processingLog}
that describes the change.
}

\section{Retrieving slot contents}{


The full contents of the \code{data} and \code{metadata} slots of a \linkS4class{coastline}
object may be retrieved in the standard R way using \code{\link[=slot]{slot()}}. For
example \code{slot(o,"data")} returns the \code{data} slot of an object named \code{o},
and similarly \code{slot(o,"metadata")} returns
the \code{metadata} slot.

The slots may also be obtained with the \code{\link{[[,coastline-method}}
operator, as e.g. \code{o[["data"]]} and \code{o[["metadata"]]}, respectively.

The \code{\link{[[,coastline-method}} operator can also
be used to retrieve items from within the \code{data} and \code{metadata} slots.
For example, \code{o[["temperature"]]} can be used to retrieve temperature
from an object containing that quantity. The rule is that a named
quantity is sought first within the object's \code{metadata} slot,
with the \code{data} slot being checked only if \code{metadata} does not
contain the item. This \code{[[} method can also be used to get
certain derived quantities, if the object contains sufficient
information to calculate them. For example, an object that holds
(practical) salinity, temperature and pressure, along with
longitude and latitude, has sufficient information to compute
Absolute Salinity, and so \code{o[["SA"]]} will yield the
calculated Absolute Salinity.

It is also possible to find items more directly, using \code{\link[=oceGetData]{oceGetData()}} and
\code{\link[=oceGetMetadata]{oceGetMetadata()}}, but neither of these functions can
retrieve derived items.
}

\seealso{
Other classes provided by oce: 
\code{\link{adp-class}},
\code{\link{adv-class}},
\code{\link{argo-class}},
\code{\link{bremen-class}},
\code{\link{cm-class}},
\code{\link{ctd-class}},
\code{\link{lisst-class}},
\code{\link{lobo-class}},
\code{\link{met-class}},
\code{\link{oce-class}},
\code{\link{odf-class}},
\code{\link{rsk-class}},
\code{\link{sealevel-class}},
\code{\link{section-class}},
\code{\link{topo-class}},
\code{\link{windrose-class}},
\code{\link{xbt-class}}

Other things related to coastline data: 
\code{\link{[[,coastline-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{as.coastline}()},
\code{\link{coastlineBest}()},
\code{\link{coastlineCut}()},
\code{\link{coastlineWorld}},
\code{\link{download.coastline}()},
\code{\link{plot,coastline-method}},
\code{\link{read.coastline.openstreetmap}()},
\code{\link{read.coastline.shapefile}()},
\code{\link{subset,coastline-method}},
\code{\link{summary,coastline-method}}
}
\author{
Dan Kelley
}
\concept{classes provided by oce}
\concept{things related to coastline data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{resizableLabel}
\alias{resizableLabel}
\title{Provide axis names in adjustable sizes}
\usage{
resizableLabel(
  item,
  axis = "x",
  sep,
  unit = NULL,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{item}{code for the label. The following common values are recognized:
\code{"absolute salinity"}, \code{"along-spine distance km"}, \code{"along-track distance km"},
\code{"C"}, \code{"conductivity mS/cm"}, \code{"conductivity S/m"}, \code{"conservative temperature"},
\code{"CT"}, \code{"depth"}, \code{"direction"}, \code{"distance"}, \code{"distance km"}, \code{"eastward"},
\code{"elevation"}, \code{"fluorescence"}, \code{"frequency cph"}, \code{"heading"}, \code{"latitude"},
\code{"longitude"}, \code{"N2"}, \code{"nitrate"}, \code{"nitrite"}, \code{"northward"}, \code{"oxygen"},
\code{"oxygen mL/L"}, \code{"oxygen saturation"}, \code{"oxygen umol/kg"}, \code{"oxygen umol/L"},
\code{"p"}, \code{"phosphate"}, \code{"pitch"}, \code{"roll"}, \code{"S"}, \code{"SA"},
\code{"sigma0"}, \code{"sigma1"}, \code{"sigma2"}, \code{"sigma3"}, \code{"sigma4"},
\code{"sigmaTheta"},
\code{"silicate"}, \code{"sound speed"}, \code{"spectral density m2/cph"}, \code{"speed"},
\code{"spice"}, \code{"T"}, \code{"theta"}, \code{"tritium"}, \code{"u"}, \code{"v"}, \code{"w"}, or \code{"z"}.
Other values may also be recognized, and if an unrecognized item is
given, then it is returned, unaltered.}

\item{axis}{a string indicating which axis to use; must be \code{x} or
\code{y}.}

\item{sep}{optional character string inserted between the unit and the
parentheses or brackets that enclose it. If not provided, then
\code{\link{getOption}}\code{("oceUnitSep")} is checked. If that exists, then it is
used as the separator; if not, no separator is used.}

\item{unit}{optional unit to use, if the default is not satisfactory. This
might be the case if for example temperature was not measured in Celcius.}

\item{debug}{optional debugging flag. Setting to 0 turns off debugging,
while setting to 1 causes some debugging information to be printed.}
}
\value{
A character string or expression, in either a long or a shorter
format, for use in the indicated axis at the present plot size.  Whether the
unit is enclosed in parentheses or square brackets is determined by the
value of \code{getOption("oceUnitBracket")}, which may be \code{"["} or
\code{"("}.  Whether spaces are used between the unit and these deliminators
is set by \code{psep} or \code{\link{getOption}}\code{("oceUnitSep")}.
}
\description{
Provide axis names in adjustable sizes, e.g. using T instead of Temperature,
and including units as appropriate.
Used by e.g. \code{\link[=plot,ctd-method]{plot,ctd-method()}}.
}
\seealso{
Other functions that create labels: 
\code{\link{labelWithUnit}()}
}
\author{
Dan Kelley
}
\concept{functions that create labels}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.R
\name{subtractBottomVelocity}
\alias{subtractBottomVelocity}
\title{Subtract Bottom Velocity from ADP}
\usage{
subtractBottomVelocity(x, despike = FALSE, debug = getOption("oceDebug"))
}
\arguments{
\item{x}{an \linkS4class{adp} object that contains bottom-tracking velocities.}

\item{despike}{either a logical value or a univariate function. This
controls whether the bottom velocity (\code{bv}) values should be altered before they are
subtracted from the beam velocities. If it is \code{TRUE} then the \code{bv} values are despiked
first by calling \code{\link[=despike]{despike()}}. If it is a function, then that function is used instead of
\code{\link[=despike]{despike()}}, e.g. \code{function(x) despike(x, reference="smooth")} would change the reference
function for despiking from its default of \code{"median"}.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\description{
Subtracts bottom tracking velocities from an \code{"adp"} object. Works for
all coordinate systems (\code{beam}, \code{xyz}, and \code{enu}).
}
\seealso{
See \code{\link[=read.adp]{read.adp()}} for notes on functions relating to
\code{"adp"} objects, and \linkS4class{adp} for notes on the ADP
object class.

Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley and Clark Richards
}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{binCount2D}
\alias{binCount2D}
\title{Bin-count matrix data}
\usage{
binCount2D(x, y, xbreaks, ybreaks, flatten = FALSE)
}
\arguments{
\item{x}{vector of numerical values.}

\item{y}{vector of numerical values.}

\item{xbreaks}{Vector of values of \code{x} at the boundaries between bins, calculated using
[\verb{pretty``]}(x)` if not supplied.}

\item{ybreaks}{Vector of values of \code{y} at the boundaries between bins, calculated using
\code{\link{pretty}}\code{(y)} if not supplied.}

\item{flatten}{A logical value indicating whether
the return value also contains equilength
vectors \code{x}, \code{y}, \code{z} and \code{n}, a flattened
representation of \code{xmids}, \code{ymids}, \code{result} and
\code{number}.}
}
\value{
A list with the following elements: the breaks (\code{xbreaks}
and \code{ybreaks}), the midpoints (\code{xmids} and \code{ymids})
between those breaks, and
the count (\code{number}) of \code{f} values in the boxes defined
between successive breaks.
}
\description{
Count the number of elements of a given matrix z=z(x,y) that fall within
successive pairs of breaks in x and y.
}
\seealso{
Other bin-related functions: 
\code{\link{binApply1D}()},
\code{\link{binApply2D}()},
\code{\link{binAverage}()},
\code{\link{binCount1D}()},
\code{\link{binMean1D}()},
\code{\link{binMean2D}()}
}
\author{
Dan Kelley
}
\concept{bin-related functions}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/topo.R
\name{topoInterpolate}
\alias{topoInterpolate}
\title{Interpolate Within a Topo Object}
\usage{
topoInterpolate(longitude, latitude, topo)
}
\arguments{
\item{longitude}{Vector of longitudes (in the same sign convention as used in
\code{topo}).}

\item{latitude}{Vector of latitudes (in the same sign convention as used in
\code{topo}).}

\item{topo}{A \linkS4class{topo} object.}
}
\value{
Vector of heights giving the elevation of the earth above means sea
level at the indicated location on the earth.
}
\description{
Bilinear interpolation is used so that values will vary smoothly within a
longitude-latitude grid cell. Note that the sign convention for
\code{longitude} and \code{latitude} must match that in \code{topo}.
}
\examples{
library(oce)
data(topoWorld)
# "The Gully", approx. 400m deep, connects Gulf of St Lawrence with North Atlantic
topoInterpolate(45, -57, topoWorld)

}
\seealso{
Other things related to topo data: 
\code{\link{[[,topo-method}},
\code{\link{[[<-,topo-method}},
\code{\link{as.topo}()},
\code{\link{download.topo}()},
\code{\link{plot,topo-method}},
\code{\link{read.topo}()},
\code{\link{subset,topo-method}},
\code{\link{summary,topo-method}},
\code{\link{topo-class}},
\code{\link{topoWorld}}
}
\author{
Dan Kelley
}
\concept{things related to topo data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/processingLog.R
\name{processingLogAppend}
\alias{processingLogAppend}
\title{Append an item to a processing log}
\usage{
processingLogAppend(h, value = "")
}
\arguments{
\item{h}{either the \code{processingLog} slot of an object, or
an \code{oce} object from which the processingLog will be extracted}

\item{value}{A string indicating the text of the log entry.}
}
\value{
An \code{\link[=list]{list()}} containing items named
\code{time} and \code{value}, i.e. the times of entries
and the text notations of those entries..
}
\description{
Append an item to a processing log
}
\seealso{
Other things related to processing logs: 
\code{\link{processingLog<-}()},
\code{\link{processingLogItem}()},
\code{\link{processingLogShow}()}
}
\concept{things related to processing logs}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/map.R
\name{formatPosition}
\alias{formatPosition}
\title{Format Geographical Position in Degrees and Minutes}
\usage{
formatPosition(
  latlon,
  isLat = TRUE,
  type = c("list", "string", "expression"),
  showHemi = TRUE
)
}
\arguments{
\item{latlon}{a vector of latitudes or longitudes}

\item{isLat}{a boolean that indicates whether the quantity is latitude or
longitude}

\item{type}{a string indicating the type of return value (see below)}

\item{showHemi}{a boolean that indicates whether to indicate the hemisphere}
}
\value{
A list containing \code{degrees}, \code{minutes}, \code{seconds},
and \code{hemispheres}, or a vector of strings or (broken) a vector of
expressions.
}
\description{
Format geographical positions to degrees, minutes, and hemispheres
}
\examples{
library(oce)
formatPosition(10+1:10/60+2.8/3600)
formatPosition(10+1:10/60+2.8/3600, type="string")

}
\seealso{
Other functions related to maps: 
\code{\link{lonlat2map}()},
\code{\link{lonlat2utm}()},
\code{\link{map2lonlat}()},
\code{\link{mapArrows}()},
\code{\link{mapAxis}()},
\code{\link{mapContour}()},
\code{\link{mapCoordinateSystem}()},
\code{\link{mapDirectionField}()},
\code{\link{mapGrid}()},
\code{\link{mapImage}()},
\code{\link{mapLines}()},
\code{\link{mapLocator}()},
\code{\link{mapLongitudeLatitudeXY}()},
\code{\link{mapPlot}()},
\code{\link{mapPoints}()},
\code{\link{mapPolygon}()},
\code{\link{mapScalebar}()},
\code{\link{mapText}()},
\code{\link{mapTissot}()},
\code{\link{oceCRS}()},
\code{\link{shiftLongitude}()},
\code{\link{usrLonLat}()},
\code{\link{utm2lonlat}()}
}
\author{
Dan Kelley
}
\concept{functions related to maps}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/topo.R
\docType{class}
\name{topo-class}
\alias{topo-class}
\title{Class to Store Topographic Data}
\description{
This class stores topographic data, as read with
\code{\link[=read.topo]{read.topo()}} or assembled with \code{\link[=as.topo]{as.topo()}}.
Plotting is handled with \code{\link[=plot,topo-method]{plot,topo-method()}}
and summaries with \code{\link[=summary,topo-method]{summary,topo-method()}}.
}
\section{Slots}{

\describe{
\item{\code{data}}{As with all \code{oce} objects, the \code{data} slot
for \code{topo} objects is a
\link{list} containing the main data for the object.
The key items stored in this slot are: \code{longititude}, \code{latitude}, and \code{z}.}

\item{\code{metadata}}{As with all \code{oce} objects, the \code{metadata} slot
for \code{topo} objects is a \link{list} containing
information about the \code{data} or about the object itself.
{}}

\item{\code{processingLog}}{As with all \code{oce} objects, the \code{processingLog} slot
for \code{topo} objects is a
\link{list} with entries describing the creation and evolution
of the object. The contents are updated by various \code{oce} functions to
keep a record of processing steps.  Object summaries and
\code{\link[=processingLogShow]{processingLogShow()}} both display the log.}
}}

\section{Modifying slot contents}{


Although the \code{[[<-} operator may permit modification of the contents
of \linkS4class{topo} objects (see \code{\link{[[<-,topo-method}}),
it is better to use \code{\link[=oceSetData]{oceSetData()}} and \code{\link[=oceSetMetadata]{oceSetMetadata()}},
because those functions save an entry in the \code{processingLog}
that describes the change.
}

\section{Retrieving slot contents}{


The full contents of the \code{data} and \code{metadata} slots of a \linkS4class{topo}
object may be retrieved in the standard R way using \code{\link[=slot]{slot()}}. For
example \code{slot(o,"data")} returns the \code{data} slot of an object named \code{o},
and similarly \code{slot(o,"metadata")} returns
the \code{metadata} slot.

The slots may also be obtained with the \code{\link{[[,topo-method}}
operator, as e.g. \code{o[["data"]]} and \code{o[["metadata"]]}, respectively.

The \code{\link{[[,topo-method}} operator can also
be used to retrieve items from within the \code{data} and \code{metadata} slots.
For example, \code{o[["temperature"]]} can be used to retrieve temperature
from an object containing that quantity. The rule is that a named
quantity is sought first within the object's \code{metadata} slot,
with the \code{data} slot being checked only if \code{metadata} does not
contain the item. This \code{[[} method can also be used to get
certain derived quantities, if the object contains sufficient
information to calculate them. For example, an object that holds
(practical) salinity, temperature and pressure, along with
longitude and latitude, has sufficient information to compute
Absolute Salinity, and so \code{o[["SA"]]} will yield the
calculated Absolute Salinity.

It is also possible to find items more directly, using \code{\link[=oceGetData]{oceGetData()}} and
\code{\link[=oceGetMetadata]{oceGetMetadata()}}, but neither of these functions can
retrieve derived items.
}

\seealso{
Other classes provided by oce: 
\code{\link{adp-class}},
\code{\link{adv-class}},
\code{\link{argo-class}},
\code{\link{bremen-class}},
\code{\link{cm-class}},
\code{\link{coastline-class}},
\code{\link{ctd-class}},
\code{\link{lisst-class}},
\code{\link{lobo-class}},
\code{\link{met-class}},
\code{\link{oce-class}},
\code{\link{odf-class}},
\code{\link{rsk-class}},
\code{\link{sealevel-class}},
\code{\link{section-class}},
\code{\link{windrose-class}},
\code{\link{xbt-class}}

Other things related to topo data: 
\code{\link{[[,topo-method}},
\code{\link{[[<-,topo-method}},
\code{\link{as.topo}()},
\code{\link{download.topo}()},
\code{\link{plot,topo-method}},
\code{\link{read.topo}()},
\code{\link{subset,topo-method}},
\code{\link{summary,topo-method}},
\code{\link{topoInterpolate}()},
\code{\link{topoWorld}}
}
\author{
Dan Kelley
}
\concept{classes provided by oce}
\concept{things related to topo data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/RcppExports.R
\name{bilinearInterp}
\alias{bilinearInterp}
\title{Bilinear Interpolation Within a Grid}
\usage{
bilinearInterp(x, y, gx, gy, g)
}
\arguments{
\item{x}{vector of x values at which to interpolate}

\item{y}{vector of y values at which to interpolate}

\item{gx}{vector of x values for the grid}

\item{gy}{vector of y values for the grid}

\item{g}{matrix of the grid values}
}
\value{
vector of interpolated values
}
\description{
This is used by \code{\link{topoInterpolate}}.
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sealevel.R
\docType{data}
\name{sealevelTuktoyaktuk}
\alias{sealevelTuktoyaktuk}
\title{Sea-level data set acquired in 1975 at Tuktoyaktuk}
\source{
The data were based on the \code{T_TIDE} dataset, which in turn
seems to be based on Appendix 7.2 of Foreman (1977).  Minor editing was on
file format, and then the \code{sealevelTuktoyaktuk} object was created
using \code{\link[=as.sealevel]{as.sealevel()}}.
}
\description{
This sea-level dataset is provided with in Appendix 7.2 of Foreman (1977)
and also with the \code{T_TIDE} package (Pawlowicz et al., 2002). It results
from measurements made in 1975 at Tuktoyaktuk, Northwest Territories,
Canada.
}
\details{
The data set contains 1584 points, some of which have NA for sea-level
height.

Although Foreman's Appendix 7.2 states that times are in Mountain standard
time, the timezone is set to \code{UTC} in the present case, so that the
results will be similar to those he provides in his Appendix 7.3.
}
\section{Historical note}{

Until Jan 6, 2018, the time in this dataset had been increased
by 7 hours. However, this alteration was removed on this date,
to make for simpler comparison of amplitude and phase output with
the results obtained by Foreman (1977) and Pawlowicz et al. (2002).
}

\examples{
\donttest{
library(oce)
data(sealevelTuktoyaktuk)
time <- sealevelTuktoyaktuk[["time"]]
elevation <- sealevelTuktoyaktuk[["elevation"]]
oce.plot.ts(time, elevation, type='l', ylab="Height [m]", ylim=c(-2, 6))
legend("topleft", legend=c("Tuktoyaktuk (1975)","Detided"),
       col=c("black","red"),lwd=1)
tide <- tidem(sealevelTuktoyaktuk)
detided <- elevation - predict(tide)
lines(time, detided, col="red")
}

}
\references{
Foreman, M. G. G., 1977.  Manual for tidal heights analysis and
prediction.  Pacific Marine Science Report 77-10, Institute of Ocean
Sciences, Patricia Bay, Sidney, BC, 58pp.

Pawlowicz, Rich, Bob Beardsley, and Steve Lentz, 2002.  Classical tidal
harmonic analysis including error estimates in MATLAB using \code{T_TIDE}.
Computers and Geosciences, 28, 929-937.
}
\seealso{
Other datasets provided with oce: 
\code{\link{adp}},
\code{\link{adv}},
\code{\link{amsr}},
\code{\link{argo}},
\code{\link{cm}},
\code{\link{coastlineWorld}},
\code{\link{ctdRaw}},
\code{\link{ctd}},
\code{\link{echosounder}},
\code{\link{landsat}},
\code{\link{lisst}},
\code{\link{lobo}},
\code{\link{met}},
\code{\link{ocecolors}},
\code{\link{rsk}},
\code{\link{sealevel}},
\code{\link{section}},
\code{\link{topoWorld}},
\code{\link{wind}},
\code{\link{xbt}}

Other things related to sealevel data: 
\code{\link{[[,sealevel-method}},
\code{\link{[[<-,sealevel-method}},
\code{\link{as.sealevel}()},
\code{\link{plot,sealevel-method}},
\code{\link{read.sealevel}()},
\code{\link{sealevel-class}},
\code{\link{sealevel}},
\code{\link{subset,sealevel-method}},
\code{\link{summary,sealevel-method}}
}
\concept{datasets provided with oce}
\concept{things related to sealevel data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/map.R
\name{map2lonlat}
\alias{map2lonlat}
\title{Convert X and Y to Longitude and Latitude}
\usage{
map2lonlat(x, y, init = NULL, debug = getOption("oceDebug"))
}
\arguments{
\item{x}{vector containing the x component of points in the projected space, or
a list containing items named \code{x} and \code{y}, in which case the next
argument is ignored.}

\item{y}{vector containing the y coordinate of points in the projected space
(ignored if \code{x} is a list, as described above).}

\item{init}{vector containing the initial guesses for longitude and latitude,
presently ignored.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\value{
A list containing \code{longitude} and \code{latitude}, with \code{NA}
values indicating points that are off the globe as displayed.
}
\description{
Convert from x-y coordinates to longitude and latitude. This is normally called
internally within oce; see \sQuote{Bugs}.
A projection must already have been set up, by a call to \code{\link[=mapPlot]{mapPlot()}}
or \code{\link[=lonlat2map]{lonlat2map()}}. It should be noted that not all projections are
handled well; see \sQuote{Bugs}.
}
\section{Bugs}{

\code{oce} uses the \code{\link[sf:sf_project]{sf::sf_project()}} function to handle projections.
Only those projections that
have inverses are permitted within \code{oce}, and of that subset, some are omitted
because the \code{oce} developers have experienced problems with them.
}

\examples{
library(oce)
## Cape Split, in the Minas Basin of the Bay of Fundy
cs <- list(longitude=-64.49657, latitude=45.33462)
xy <- lonlat2map(cs, projection="+proj=merc")
map2lonlat(xy)

}
\seealso{
\code{\link[=lonlat2map]{lonlat2map()}} does the inverse operation.

A map must first have been created with \code{\link[=mapPlot]{mapPlot()}}.

Other functions related to maps: 
\code{\link{formatPosition}()},
\code{\link{lonlat2map}()},
\code{\link{lonlat2utm}()},
\code{\link{mapArrows}()},
\code{\link{mapAxis}()},
\code{\link{mapContour}()},
\code{\link{mapCoordinateSystem}()},
\code{\link{mapDirectionField}()},
\code{\link{mapGrid}()},
\code{\link{mapImage}()},
\code{\link{mapLines}()},
\code{\link{mapLocator}()},
\code{\link{mapLongitudeLatitudeXY}()},
\code{\link{mapPlot}()},
\code{\link{mapPoints}()},
\code{\link{mapPolygon}()},
\code{\link{mapScalebar}()},
\code{\link{mapText}()},
\code{\link{mapTissot}()},
\code{\link{oceCRS}()},
\code{\link{shiftLongitude}()},
\code{\link{usrLonLat}()},
\code{\link{utm2lonlat}()}
}
\author{
Dan Kelley
}
\concept{functions related to maps}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/accessors.R
\name{oceRenameMetadata}
\alias{oceRenameMetadata}
\title{Rename Something in an oce metadata Slot}
\usage{
oceRenameMetadata(object, old, new, note = "")
}
\arguments{
\item{object}{an \linkS4class{oce} object.}

\item{old}{character value that matches the name of an item in \code{object}'s \code{metadata} slot.}

\item{new}{character value to be used as the new name that matches the name of an item in
\code{object}'s \code{metadata} slot. Thus must not be the name of something that is already in the \code{metadata}
slot. If \code{new} is the  same as \code{old}, then the object is returned unaltered.}

\item{note}{character value that holds an explanation of the reason for the change. If this
is a string of non-zero length, then this is inserted in the processing log of the returned
value. If it is \code{NULL}, then no entry is added to the processing log.  Otherwise, the processing
log gets a new item that is constructed from the function call.}
}
\description{
Rename an item within the \code{metadata} slot of an \linkS4class{oce} object.
}
\seealso{
Other things related to the metadata slot: 
\code{\link{oceDeleteMetadata}()},
\code{\link{oceGetMetadata}()},
\code{\link{oceSetMetadata}()}
}
\author{
Dan Kelley
}
\concept{things related to the metadata slot}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/topo.R
\name{as.topo}
\alias{as.topo}
\title{Coerce Data into Topo Object}
\usage{
as.topo(longitude, latitude, z, filename = "")
}
\arguments{
\item{longitude}{Either a vector of longitudes (in degrees east, and bounded by
-180 and 180), or a \code{bathy} object created by \code{getNOAA.bathy()} from
the \code{marmap} package; in the second case, all other arguments are ignored.}

\item{latitude}{A vector of latitudes.}

\item{z}{A matrix of heights (positive over land).}

\item{filename}{Name of data (used when called by \code{\link[=read.topo]{read.topo()}}.}
}
\value{
A \linkS4class{topo} object.
}
\description{
Coerce Data into Topo Object
}
\seealso{
Other things related to topo data: 
\code{\link{[[,topo-method}},
\code{\link{[[<-,topo-method}},
\code{\link{download.topo}()},
\code{\link{plot,topo-method}},
\code{\link{read.topo}()},
\code{\link{subset,topo-method}},
\code{\link{summary,topo-method}},
\code{\link{topo-class}},
\code{\link{topoInterpolate}()},
\code{\link{topoWorld}}
}
\author{
Dan Kelley
}
\concept{things related to topo data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/xbt.R
\name{read.xbt.noaa1}
\alias{read.xbt.noaa1}
\title{Read a NOAA format for AXBTs}
\usage{
read.xbt.noaa1(
  file,
  debug = getOption("oceDebug"),
  missingValue = -9.99,
  processingLog
)
}
\arguments{
\item{file}{character value naming a file, or a file connection, containing the data.}

\item{debug}{a flag that turns on debugging.  The value indicates the depth
within the call stack to which debugging applies.}

\item{missingValue}{numerical value that is to be interpreted as \code{NA}}

\item{processingLog}{if provided, the action item to be stored in the log.  This
parameter is typically only provided for internal calls; the default that it
provides is better for normal calls by a user.}
}
\value{
An \linkS4class{xbt} object.
}
\description{
This file format, described at \code{https://www.aoml.noaa.gov/phod/dhos/axbt.php}, contains a header
line, followed by data lines.  For example, a particular file at this site has first
three lines as follows.\preformatted{181.589 20100709 140820  -85.336  25.290 N42RF GL10 14    2010-190-15:49:18
  -0.0 27.52 -9.99
  -1.5 27.52 -9.99
}

where the items on the header line are (1) a year-day (ignored here), (2)
YYYYMMDD, (3) HHMMSS, (4) longitude, (5) latitude, (6) aircraft wing
number, (7) project name, (8) AXBT channel and (9) AXBT ID.  The other lines hold vertical
coordinate in metres, temperature and temperature error; -9.99 is a missing-value
code.  (This formatting information is extracted from a file named \code{readme.axbt} that
is provided with the data.)
}
\seealso{
Other things related to xbt data: 
\code{\link{[[,xbt-method}},
\code{\link{[[<-,xbt-method}},
\code{\link{as.xbt}()},
\code{\link{plot,xbt-method}},
\code{\link{read.xbt}()},
\code{\link{subset,xbt-method}},
\code{\link{summary,xbt-method}},
\code{\link{xbt-class}},
\code{\link{xbt.edf}},
\code{\link{xbt}}
}
\author{
Dan Kelley
}
\concept{things related to xbt data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{bcdToInteger}
\alias{bcdToInteger}
\title{Decode BCD to integer}
\usage{
bcdToInteger(x, endian = c("little", "big"))
}
\arguments{
\item{x}{a raw value, or vector of raw values, coded in binary-coded
decimal.}

\item{endian}{character string indicating the endian-ness ("big" or
"little").  The PC/intel convention is to use "little", and so most data
files are in that format.}
}
\value{
An integer, or list of integers.
}
\description{
Decode BCD to integer
}
\examples{
library(oce)
twenty.five <- bcdToInteger(as.raw(0x25))
thirty.seven <- as.integer(as.raw(0x25))
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.woce.R
\name{oceUnits2whpUnits}
\alias{oceUnits2whpUnits}
\title{Translate oce unit to WHP unit}
\usage{
oceUnits2whpUnits(units, scales)
}
\arguments{
\item{units}{vector of expressions for units in oce notation.}

\item{scales}{vector of strings for scales in oce notation.}
}
\value{
vector of strings holding WOCE-style names.
}
\description{
Translate oce units to WHP-style strings,
to match patterns.
}
\references{
Several online sources list WOCE names. An example is
\url{https://cchdo.github.io/hdo-assets/documentation/manuals/pdf/90_1/chap4.pdf}
}
\seealso{
Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}

Other functions that interpret variable names and units from headers: 
\code{\link{ODFNames2oceNames}()},
\code{\link{cnvName2oceName}()},
\code{\link{oceNames2whpNames}()},
\code{\link{unitFromStringRsk}()},
\code{\link{unitFromString}()},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()}
}
\author{
Dan Kelley
}
\concept{functions that interpret variable names and units from headers}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/rsk.R
\name{rskPatm}
\alias{rskPatm}
\title{Estimate Atmospheric Pressure in Rsk Object}
\usage{
rskPatm(x, dp = 0.5)
}
\arguments{
\item{x}{an \linkS4class{rsk} object.}

\item{dp}{Half-width of pressure window to be examined (in decibars).}
}
\value{
A list of four estimates: \code{sap}, the median, the mean, and the weighted mean.
}
\description{
Estimate atmospheric pressure in rsk record.
}
\details{
Pressures must be in decibars for this to work.  First, a subset of pressures is
created, in which the range is \code{sap-dp} to \code{sap+dp}.  Here,
\code{sap}=10.1325 dbar is standard sealevel atmospheric pressure.  Within this
window, three measures of central tendency are calculated: the median, the mean,
and a weighted mean that has weight given by \eqn{exp(-2*((p
    - sap) / dp)^2)}{exp(-2*((p - sap) / dp)^2)}.
}
\examples{
library(oce)
data(rsk)
print(rskPatm(rsk))

}
\seealso{
The documentation for \linkS4class{rsk} explains the structure of
\code{rsk} objects, and also outlines the other functions dealing with them.

Other things related to rsk data: 
\code{\link{[[,rsk-method}},
\code{\link{[[<-,rsk-method}},
\code{\link{as.rsk}()},
\code{\link{plot,rsk-method}},
\code{\link{read.rsk}()},
\code{\link{rsk-class}},
\code{\link{rskToc}()},
\code{\link{rsk}},
\code{\link{subset,rsk-method}},
\code{\link{summary,rsk-method}}
}
\author{
Dan Kelley
}
\concept{things related to rsk data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{undriftTime}
\alias{undriftTime}
\title{Correct for drift in instrument clock}
\usage{
undriftTime(x, slowEnd = 0, tname = "time")
}
\arguments{
\item{x}{an \linkS4class{oce} object.}

\item{slowEnd}{number of seconds to add to final instrument time, to get the
correct time of the final sample.  This will be a positive number, for a
"slow" instrument clock.}

\item{tname}{Character string indicating the name of the time column in the
\code{data} slot of \code{x}.}
}
\value{
An object of the same class as \code{x}, with the \code{data} slot
adjusted appropriately.
}
\description{
It is assumed that the instrument clock matches the real time at the start
of the sampling, and that the clock drifts linearly (i.e. is uniformly fast
or slow) over the sampling interval.  Linear interpolation is used to infer
the values of all variables in the \code{data} slot.  The data length is
altered in this process, e.g. a slow instrument clock (positive
\code{slowEnd}) takes too few samples in a given time interval, so
\code{undriftTime} will increase the number of data.
}
\examples{
\dontrun{
library(oce)
rbr011855 <- read.oce(
 "/data/archive/sleiwex/2008/moorings/m08/pt/rbr_011855/raw/pt_rbr_011855.dat")
d <- subset(rbr011855, time < as.POSIXct("2008-06-25 10:05:00"))
x <- undriftTime(d, 1)   # clock lost 1 second over whole experiment
summary(d)
summary(x)
}
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/topo.R
\docType{data}
\name{topoWorld}
\alias{topoWorld}
\title{Global Topographic Dataset at Half-degree Resolution}
\source{
This is created with \code{\link[=read.topo]{read.topo()}}, using a file downloaded with
\preformatted{
topoFile <- download.topo(west=-180, east=180, south=-90, north=90,
                         resolution=30, destdir=".")
}
}
\usage{
data(topoWorld)
}
\description{
Global topographic dataset at half-degree resolution, downloaded from
a NOAA server on May 18, 2019.  Longitude, accessible as
\code{topoWorld[["longitude"]]}, ranges from -179.75 to 129.75 degrees north.
Latitude (\code{topoWorld[["latitude"]]}) ranges from -89.75 to 89.75 degrees east.
Height (\code{topoWorld[["z"]]}) is measured in metres above nominal sea level.

The coarse resolution can be a problem in plotting depth contours along with
coastlines in regions of steep topography. For example, near the southeast
corner of Newfoundland, a 200m contour will overlap a coastline drawn with
\code{coastlineWorldFine} from the \CRANpkg{ocedata} package. The solution in such cases is to
download a higher-resolution topography file, perhaps using
\code{\link[=download.topo]{download.topo()}}, and then use \code{\link[=read.topo]{read.topo()}}
to create another \code{topo} object.  (With other data
sources, \code{\link[=as.topo]{as.topo()}} may be helpful.)
}
\section{Historical note}{

From late 2009 until May 18, 2019, the \code{topoWorld} dataset was created
with a fairly complicated code that read a binary file downloaded from NOAA
(\samp{http://www.ngdc.noaa.gov/mgg/global/relief/ETOPO5/TOPO/ETOPO5}),
decoded, decimated from 1/12th degree resolution to 1/2 degree resolution, and
passed through \code{\link[=matrixShiftLongitude]{matrixShiftLongitude()}} to put longitude
between -180 and 180 degrees. The new scheme for creating the dataset,
(see \dQuote{Source}) is much simpler, and also a much better model
of how users are likely to deal with topography files in the more
modern netCDF format. Note that the new version differs from the old one
in longitude and latitude being shifted by 1/4 degree,
and by a mean  elevation difference of under 10m. The old and new
versions appear identical when plotted at the global scale that is
the recommended for such a coarse topographic file.
}

\examples{
\dontrun{
library(oce)
data(topoWorld)
par(mfrow=c(2, 1))
plot(topoWorld, location=NULL)
imagep(topoWorld)
}

}
\seealso{
Other datasets provided with oce: 
\code{\link{adp}},
\code{\link{adv}},
\code{\link{amsr}},
\code{\link{argo}},
\code{\link{cm}},
\code{\link{coastlineWorld}},
\code{\link{ctdRaw}},
\code{\link{ctd}},
\code{\link{echosounder}},
\code{\link{landsat}},
\code{\link{lisst}},
\code{\link{lobo}},
\code{\link{met}},
\code{\link{ocecolors}},
\code{\link{rsk}},
\code{\link{sealevelTuktoyaktuk}},
\code{\link{sealevel}},
\code{\link{section}},
\code{\link{wind}},
\code{\link{xbt}}

Other things related to topo data: 
\code{\link{[[,topo-method}},
\code{\link{[[<-,topo-method}},
\code{\link{as.topo}()},
\code{\link{download.topo}()},
\code{\link{plot,topo-method}},
\code{\link{read.topo}()},
\code{\link{subset,topo-method}},
\code{\link{summary,topo-method}},
\code{\link{topo-class}},
\code{\link{topoInterpolate}()}
}
\concept{datasets provided with oce}
\concept{things related to topo data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/coastline.R
\docType{data}
\name{coastlineWorld}
\alias{coastlineWorld}
\title{World Coastline}
\source{
Downloaded from \url{https://www.naturalearthdata.com}, in
\code{ne_110m_admin_0_countries.shp} in July 2015, with an
update on December 16, 2017.
}
\description{
This is a coarse resolution coastline at scale 1:110M, with 10,696 points,
suitable for world-scale plots plotted at a small size, e.g. inset diagrams.
Finer resolution coastline files are provided in the
\code{ocedata} package.
}
\section{Installing your own datasets}{
 Follow the procedure along the lines
described in \dQuote{Details}, where of course your source file will differ.
Also, you should change the name of the coastline object from
\code{coastlineWorld}, to avoid conflicts with the built-in dataset. Save
the \code{.rda} file to some directory of your choosing, e.g. perhaps
\verb{/data/coastlines} or \verb{~/data/coastlines} on a unix-type machine.
Then, whenever you need the file, use \code{\link[=load]{load()}} to load it.  Most
users find it convenient to do the loading in an \code{\link[=Rprofile]{Rprofile()}}
startup file.
}

\seealso{
Other datasets provided with oce: 
\code{\link{adp}},
\code{\link{adv}},
\code{\link{amsr}},
\code{\link{argo}},
\code{\link{cm}},
\code{\link{ctdRaw}},
\code{\link{ctd}},
\code{\link{echosounder}},
\code{\link{landsat}},
\code{\link{lisst}},
\code{\link{lobo}},
\code{\link{met}},
\code{\link{ocecolors}},
\code{\link{rsk}},
\code{\link{sealevelTuktoyaktuk}},
\code{\link{sealevel}},
\code{\link{section}},
\code{\link{topoWorld}},
\code{\link{wind}},
\code{\link{xbt}}

Other things related to coastline data: 
\code{\link{[[,coastline-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{as.coastline}()},
\code{\link{coastline-class}},
\code{\link{coastlineBest}()},
\code{\link{coastlineCut}()},
\code{\link{download.coastline}()},
\code{\link{plot,coastline-method}},
\code{\link{read.coastline.openstreetmap}()},
\code{\link{read.coastline.shapefile}()},
\code{\link{subset,coastline-method}},
\code{\link{summary,coastline-method}}
}
\concept{datasets provided with oce}
\concept{things related to coastline data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{read.oce}
\alias{read.oce}
\title{Read an Oceanographic Data File}
\usage{
read.oce(file, ...)
}
\arguments{
\item{file}{a connection or a character string giving the name of the file
to load.}

\item{...}{arguments to be handed to whichever instrument-specific reading
function is selected, based on the header.}
}
\value{
An \linkS4class{oce} object of that is
specialized to the data type, e.g. \linkS4class{ctd},
if the data file contains \code{ctd} data.
}
\description{
Read an oceanographic data file, auto-discovering the file type from the
first line of the file.
This function tries to infer the file type from the first line, using
\code{\link[=oceMagic]{oceMagic()}}.  If it can be discovered, then an
instrument-specific file reading function is called, with the \code{file}
and with any additional arguments being supplied.
}
\examples{
library(oce)
x <- read.oce(system.file("extdata", "ctd.cnv", package="oce"))
plot(x) # summary with TS and profiles
plotTS(x) # just the TS
}
\seealso{
The file type is determined by \code{\link[=oceMagic]{oceMagic()}}.  If the file
type can be determined, then one of the following is called:
\code{\link[=read.ctd]{read.ctd()}}, \code{\link[=read.coastline]{read.coastline()}}
\code{\link[=read.lobo]{read.lobo()}}, \code{\link[=read.rsk]{read.rsk()}},
\code{\link[=read.sealevel]{read.sealevel()}}, etc.
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.woce.R
\name{read.ctd.woce}
\alias{read.ctd.woce}
\title{Read a WOCE-exchange CTD File}
\usage{
read.ctd.woce(
  file,
  columns = NULL,
  station = NULL,
  missingValue,
  deploymentType = "unknown",
  monitor = FALSE,
  debug = getOption("oceDebug"),
  processingLog,
  ...
)
}
\arguments{
\item{file}{a connection or a character string giving the name of the file to
load.  For \code{\link[=read.ctd.sbe]{read.ctd.sbe()}} and \code{\link[=read.ctd.woce]{read.ctd.woce()}}, this may be a
wildcard (e.g. \code{"*.cnv"} or \code{"*.csv"}) in which case the return
value is a vector containing CTD objects created by reading the files from
\code{\link[=list.files]{list.files()}} with \code{pattern} set to the specified wildcard
pattern.}

\item{columns}{an optional \link{list} that can be used to convert unrecognized
data names to resultant variable names.  This is used only by
\code{\link[=read.ctd.sbe]{read.ctd.sbe()}} and \code{\link[=read.ctd.odf]{read.ctd.odf()}}. For example,
if a data file named salinity as \code{"SAL"}, then using\preformatted{d <- read.ctd(f, columns=list(
    salinity=list(name="SAL",
                  unit=list(unit=expression(),
                  scale="PSS-78"))))
}

would assign the \code{"SAL"} column to the \code{salinity} entry in the data
slot of the CTD object returned by the \verb{read.*} function.}

\item{station}{optional character string containing an identifying name or
number for the station. This can be useful if the routine cannot determine the
name automatically, or if another name is preferred.}

\item{missingValue}{optional missing-value flag; data matching this value will
be set to \code{NA} upon reading. If this is provided, then it overrules any
missing-value flag found in the data. For Seabird (\code{.cnv}) files, there is
usually no need to set \code{missingValue}, because it can be inferred from the
header (typically as -9.990e-29). Set \code{missingValue=NULL} to turn off
missing-value detection, even in \code{.cnv} files that contain missing-value
codes in their headers. If \code{missingValue} is not specified,
then an attempt is made to infer such a value from the data, by testing
whether salinity and/or temperature has a minimum that is under -8 in value;
this should catch common values in files, without false positives. A warning
will be issued in this case, and a note inserted in the processing log of
the return value.}

\item{deploymentType}{character string indicating the type of deployment. Use
\code{"unknown"} if this is not known, \code{"profile"} for a profile (in
which the data were acquired during a downcast, while the device was lowered
into the water column, perhaps also including an upcast; \code{"moored"} if
the device is installed on a fixed mooring, \code{"thermosalinograph"} (or
\code{"tsg"}) if the device is mounted on a moving vessel, to record
near-surface properties, or \code{"towyo"} if the device is repeatedly
lowered and raised.}

\item{monitor}{boolean, set to \code{TRUE} to provide an indication of
progress.  This is useful if \code{filename} is a wildcard.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed.}

\item{processingLog}{if provided, the action item to be stored in the log.
This is typically only provided for internal calls; the default that it provides is
better for normal calls by a user.}

\item{...}{additional arguments, passed to called routines.}
}
\value{
a \linkS4class{ctd} object. The details of the contents
depend on the source file. The \code{metadata} slot is particularly
variable across data formats, because the meta-information provided
in those formats varies so widely.
}
\description{
This reads WOCE exchange files that start with the string \code{"CTD"}.
There are two variants: one in which the first 4 characters are
\code{"CTD,"} and the other in which the first 3 characters are
again \code{"CTD"} but no other non-whitespace characters occur on
the line.
}
\references{
The WOCE-exchange format was once described at
\verb{http://woce.nodc.noaa.gov/woce_v3/wocedata_1/whp/exchange/exchange_format_desc.htm}
although that link is no longer working as of December 2020.
}
\seealso{
Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}

Other functions that read ctd data: 
\code{\link{read.ctd.aml}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd}()}
}
\author{
Dan Kelley
}
\concept{functions that read ctd data}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{angleRemap}
\alias{angleRemap}
\title{Convert angles from 0:360 to -180:180}
\usage{
angleRemap(theta)
}
\arguments{
\item{theta}{an angle (in degrees) that is in the range from 0 to 360
degrees}
}
\value{
A vector of angles, in the range -180 to 180.
}
\description{
This is mostly used for instrument heading angles, in cases where the
instrument is aligned nearly northward, so that small variations in heading
(e.g. due to mooring motion) can yield values that swing from small angles
to large angles, because of the modulo-360 cut point.
The method is to use the cosine and sine of the angle in order to find "x"
and "y" values on a unit circle, and then to use \code{\link[=atan2]{atan2()}} to
infer the angles.
}
\examples{

library(oce)
## fake some heading data that lie near due-north (0 degrees)
n <- 20
heading <- 360 + rnorm(n, sd=10)
heading <- ifelse(heading > 360, heading - 360, heading)
x <- 1:n
plot(x, heading, ylim=c(-10, 360), type='l', col='lightgray', lwd=10)
lines(x, angleRemap(heading))
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{curl}
\alias{curl}
\title{Curl of 2D vector field}
\usage{
curl(u, v, x, y, geographical = FALSE, method = 1)
}
\arguments{
\item{u}{matrix containing the 'x' component of a vector field}

\item{v}{matrix containing the 'y' component of a vector field}

\item{x}{the x values for the matrices, a vector of length equal to the
number of rows in \code{u} and \code{v}.}

\item{y}{the y values for the matrices, a vector of length equal to the
number of cols in \code{u} and \code{v}.}

\item{geographical}{logical value indicating whether \code{x} and \code{y}
are longitude and latitude, in which case spherical trigonometry is used.}

\item{method}{A number indicating the method to be used to calculate the
first-difference approximations to the derivatives.  See \dQuote{Details}.}
}
\value{
A list containing vectors \code{x} and \code{y}, along with matrix
\code{curl}.  See \dQuote{Details} for the lengths and dimensions, for
various values of \code{method}.
}
\description{
Calculate the z component of the curl of an x-y vector field.
}
\details{
The computed component of the curl is defined by \eqn{\partial }{dv/dx -
du/dy}\eqn{ v/\partial x - \partial u/\partial y}{dv/dx - du/dy} and the
estimate is made using first-difference approximations to the derivatives.
Two methods are provided, selected by the value of \code{method}.
\itemize{
\item For \code{method=1}, a centred-difference, 5-point stencil is used in
the interior of the domain.  For example, \eqn{\partial v/\partial x}{dv/dx}
is given by the ratio of \eqn{v_{i+1,j}-v_{i-1,j}}{v[i+1,j]-v[i-1,j]} to the
x extent of the grid cell at index \eqn{j}{j}. (The cell extents depend on
the value of \code{geographical}.)  Then, the edges are filled in with
nearest-neighbour values. Finally, the corners are filled in with the
adjacent value along a diagonal.  If \code{geographical=TRUE}, then \code{x}
and \code{y} are taken to be longitude and latitude in degrees, and the
earth shape is approximated as a sphere with radius 6371km.  The resultant
\code{x} and \code{y} are identical to the provided values, and the
resultant \code{curl} is a matrix with dimension identical to that of
\code{u}.
\item For \code{method=2}, each interior cell in the grid is considered
individually, with derivatives calculated at the cell center. For example,
\eqn{\partial v/\partial x}{dv/dx} is given by the ratio of
\eqn{0.5*(v_{i+1,j}+v_{i+1,j+1}) -
0.5*(v_{i,j}+v_{i,j+1})}{0.5*(v[i+1,j]+v[i+1,j+1]) - 0.5*(v[i,j]+v[i,j+1])}
to the average of the x extent of the grid cell at indices \eqn{j}{j} and
\eqn{j+1}{j+1}. (The cell extents depend on the value of
\code{geographical}.)  The returned \code{x} and \code{y} values are the
mid-points of the supplied values. Thus, the returned \code{x} and \code{y}
are shorter than the supplied values by 1 item, and the returned \code{curl}
matrix dimensions are similarly reduced compared with the dimensions of
\code{u} and \code{v}.
}
}
\section{Development status.}{
 This function is under active development as
of December 2014 and is unlikely to be stabilized until February 2015.
}

\examples{
library(oce)
## 1. Shear flow with uniform curl.
x <- 1:4
y <- 1:10
u <- outer(x, y, function(x, y) y/2)
v <- outer(x, y, function(x, y) -x/2)
C <- curl(u, v, x, y, FALSE)

## 2. Rankine vortex: constant curl inside circle, zero outside
rankine <- function(x, y)
{
    r <- sqrt(x^2 + y^2)
    theta <- atan2(y, x)
    speed <- ifelse(r < 1, 0.5*r, 0.5/r)
    list(u=-speed*sin(theta), v=speed*cos(theta))
}
x <- seq(-2, 2, length.out=100)
y <- seq(-2, 2, length.out=50)
u <- outer(x, y, function(x, y) rankine(x, y)$u)
v <- outer(x, y, function(x, y) rankine(x, y)$v)
C <- curl(u, v, x, y, FALSE)
## plot results
par(mfrow=c(2, 2))
imagep(x, y, u, zlab="u", asp=1)
imagep(x, y, v, zlab="v", asp=1)
imagep(x, y, C$curl, zlab="curl", asp=1)
hist(C$curl, breaks=100)
}
\seealso{
Other things relating to vector calculus: 
\code{\link{grad}()}
}
\author{
Dan Kelley and Chantelle Layton
}
\concept{things relating to vector calculus}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/AllClass.R
\name{presentTime}
\alias{presentTime}
\title{Get the present time, in a stated timezone}
\usage{
presentTime(tz = "UTC")
}
\arguments{
\item{tz}{String indicating the desired timezone. The default is
to use UTC, which is used very commonly in oceanographic work. To
get the local time, use \code{tz=""} or \code{tz=NULL},}
}
\value{
A \code{\link[=POSIXct]{POSIXct()}}-style object holding the present time, in the
indicated timezone.
}
\description{
Get the present time, in a stated timezone
}
\examples{
presentTime() # UTC
presentTime("") # the local timezone

}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sealevel.R
\name{summary,sealevel-method}
\alias{summary,sealevel-method}
\title{Summarize a Sealevel Object}
\usage{
\S4method{summary}{sealevel}(object, ...)
}
\arguments{
\item{object}{A \linkS4class{sealevel} object.}

\item{\dots}{further arguments passed to or from other methods.}
}
\value{
A matrix containing statistics of the elements of the \code{data}
slot.
}
\description{
Summarizes some of the data in a sealevel object.
}
\examples{
library(oce)
data(sealevel)
summary(sealevel)

}
\seealso{
Other things related to sealevel data: 
\code{\link{[[,sealevel-method}},
\code{\link{[[<-,sealevel-method}},
\code{\link{as.sealevel}()},
\code{\link{plot,sealevel-method}},
\code{\link{read.sealevel}()},
\code{\link{sealevel-class}},
\code{\link{sealevelTuktoyaktuk}},
\code{\link{sealevel}},
\code{\link{subset,sealevel-method}}
}
\author{
Dan Kelley
}
\concept{things related to sealevel data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adv.R
\name{subset,adv-method}
\alias{subset,adv-method}
\title{Subset an ADV Object}
\usage{
\S4method{subset}{adv}(x, subset, ...)
}
\arguments{
\item{x}{an \linkS4class{adv} object.}

\item{subset}{a condition to be applied to the \code{data} portion of \code{x}.
See \sQuote{Details}.}

\item{\dots}{ignored.}
}
\value{
A new \linkS4class{adv} object.
}
\description{
Subset an adv (acoustic Doppler profile) object.  This function is somewhat
analogous to \code{\link[=subset.data.frame]{subset.data.frame()}}, except that subsets can only be
specified in terms of \code{time}.
}
\examples{
library(oce)
data(adv)
plot(adv)
plot(subset(adv, time < mean(range(adv[['time']]))))

}
\seealso{
Other things related to adv data: 
\code{\link{[[,adv-method}},
\code{\link{[[<-,adv-method}},
\code{\link{adv-class}},
\code{\link{adv}},
\code{\link{beamName}()},
\code{\link{beamToXyz}()},
\code{\link{enuToOtherAdv}()},
\code{\link{enuToOther}()},
\code{\link{plot,adv-method}},
\code{\link{read.adv.nortek}()},
\code{\link{read.adv.sontek.adr}()},
\code{\link{read.adv.sontek.serial}()},
\code{\link{read.adv.sontek.text}()},
\code{\link{read.adv}()},
\code{\link{rotateAboutZ}()},
\code{\link{summary,adv-method}},
\code{\link{toEnuAdv}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdv}()},
\code{\link{xyzToEnu}()}

Other functions that subset oce objects: 
\code{\link{subset,adp-method}},
\code{\link{subset,amsr-method}},
\code{\link{subset,argo-method}},
\code{\link{subset,cm-method}},
\code{\link{subset,coastline-method}},
\code{\link{subset,ctd-method}},
\code{\link{subset,echosounder-method}},
\code{\link{subset,lobo-method}},
\code{\link{subset,met-method}},
\code{\link{subset,oce-method}},
\code{\link{subset,odf-method}},
\code{\link{subset,rsk-method}},
\code{\link{subset,sealevel-method}},
\code{\link{subset,section-method}},
\code{\link{subset,topo-method}},
\code{\link{subset,xbt-method}}
}
\author{
Dan Kelley
}
\concept{functions that subset oce objects}
\concept{things related to adv data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/landsat.R
\docType{data}
\name{landsat}
\alias{landsat}
\title{Sample landsat Dataset}
\description{
This is a subset of the Landsat-8 image designated LC80080292014065LGN00, an
image from March 2014 that covers Nova Scotia and portions of the Bay of
Fundy and the Scotian Shelf. The image is decimated to reduce the memory
requirements of this package, yielding a spatial resolution of about 2km.
}
\details{
The original data were downloaded from the USGS earthexplorer website, although
other sites can also be used to uncover it by name.  The original
data were decimated by a factor of 100 in longitude and latitude,
to reduce the file size from 1G to 100K.
}
\seealso{
Other satellite datasets provided with oce: 
\code{\link{amsr}}

Other datasets provided with oce: 
\code{\link{adp}},
\code{\link{adv}},
\code{\link{amsr}},
\code{\link{argo}},
\code{\link{cm}},
\code{\link{coastlineWorld}},
\code{\link{ctdRaw}},
\code{\link{ctd}},
\code{\link{echosounder}},
\code{\link{lisst}},
\code{\link{lobo}},
\code{\link{met}},
\code{\link{ocecolors}},
\code{\link{rsk}},
\code{\link{sealevelTuktoyaktuk}},
\code{\link{sealevel}},
\code{\link{section}},
\code{\link{topoWorld}},
\code{\link{wind}},
\code{\link{xbt}}

Other things related to landsat data: 
\code{\link{[[,landsat-method}},
\code{\link{[[<-,landsat-method}},
\code{\link{landsat-class}},
\code{\link{landsatAdd}()},
\code{\link{landsatTrim}()},
\code{\link{plot,landsat-method}},
\code{\link{read.landsat}()},
\code{\link{summary,landsat-method}}
}
\concept{datasets provided with oce}
\concept{satellite datasets provided with oce}
\concept{things related to landsat data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swSigma3}
\alias{swSigma3}
\title{Seawater potential density anomaly referenced to 3000db pressure}
\usage{
swSigma3(
  salinity,
  temperature = NULL,
  pressure = NULL,
  longitude = NULL,
  latitude = NULL,
  eos = getOption("oceEOS", default = "gsw")
)
}
\arguments{
\item{salinity}{either practical salinity (in which case \code{temperature}
and \code{pressure} must be provided) \emph{or} an \code{oce} object, in
which case \code{salinity}, \code{temperature} (in the ITS-90 scale; see
next item), etc. are inferred from the object, ignoring the
other parameters, if they are supplied.}

\item{temperature}{\emph{in-situ} temperature (\eqn{^\circ}{deg}C), defined
on the ITS-90 scale.  This scale is used by GSW-style calculation (as
requested by setting \code{eos="gsw"}), and is the value contained within
\code{ctd} objects (and probably most other objects created with data
acquired in the past decade or two). Since the UNESCO-style calculation is
based on IPTS-68, the temperature is converted within the present function,
using \code{\link[=T68fromT90]{T68fromT90()}}.}

\item{pressure}{pressure (dbar)}

\item{longitude}{longitude of observation (only used if \code{eos="gsw"};
see \sQuote{Details}).}

\item{latitude}{latitude of observation (only used if \code{eos="gsw"}; see
\sQuote{Details}).}

\item{eos}{equation of state, either \code{"unesco"} (references 1 and 2)
or \code{"gsw"} (references 3 and 4).}
}
\value{
Potential density anomaly (kg/m\eqn{^3}{^3}).
}
\description{
Compute \eqn{\sigma_\theta}{sigma}, the potential density of seawater (minus
1000 kg/m\eqn{^3}{^3}), referenced to 3000db pressure.
}
\details{
Definition:
\eqn{\sigma_1=\sigma_\theta=\rho(S,\theta(S,t,p),3000}{sigma_1=sigma_theta=rho(S,theta(S,t,p),3000)}
\itemize{
\item 1000 kg/m\eqn{^3}{^3}.
}
}
\references{
See citations provided in the \code{\link[=swRho]{swRho()}} documentation.
}
\seealso{
Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/landsat.R
\name{landsatAdd}
\alias{landsatAdd}
\title{Add a Band to a landsat Object}
\usage{
landsatAdd(x, data, name, debug = getOption("oceDebug"))
}
\arguments{
\item{x}{a \linkS4class{landsat} object.}

\item{data}{A matrix of data, with dimensions matching that of entries already in \code{x}.}

\item{name}{The name to be used for the data, i.e. the data can later be
accessed with \code{d[[name]]} where \code{d} is the name of the return value
from the present function.}

\item{debug}{A flag that turns on debugging.  Set to 1 to get a moderate amount of debugging
information, or a higher value for more debugging.}
}
\value{
A \linkS4class{landsat} object, with a new data band.
}
\description{
Add a band to a \linkS4class{landsat} object. Note that
it will be stored in numeric form, not raw form, and therefore
it will require much more storage than data read with
\code{\link[=read.landsat]{read.landsat()}}.
}
\seealso{
The documentation for the \linkS4class{landsat} class explains the
structure of landsat objects, and also outlines the other functions dealing
with them.

Other things related to landsat data: 
\code{\link{[[,landsat-method}},
\code{\link{[[<-,landsat-method}},
\code{\link{landsat-class}},
\code{\link{landsatTrim}()},
\code{\link{landsat}},
\code{\link{plot,landsat-method}},
\code{\link{read.landsat}()},
\code{\link{summary,landsat-method}}
}
\author{
Dan Kelley
}
\concept{things related to landsat data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swCSTp}
\alias{swCSTp}
\title{Electrical conductivity ratio from salinity, temperature and pressure}
\usage{
swCSTp(
  salinity,
  temperature = 15,
  pressure = 0,
  eos = getOption("oceEOS", default = "gsw")
)
}
\arguments{
\item{salinity}{practical salinity, or a CTD object (in which case its
temperature and pressure are used, and the next two arguments are ignored)}

\item{temperature}{\emph{in-situ} temperature (\eqn{^\circ}{deg}C), defined
on the ITS-90 scale; see the examples, as well as the
\dQuote{Temperature units} section in the documentation for \code{\link[=swRho]{swRho()}}.}

\item{pressure}{pressure (dbar)}

\item{eos}{equation of state, either \code{"unesco"} or \code{"gsw"}.}
}
\value{
Conductivity ratio (unitless), i.e. the ratio of conductivity to the
conductivity at salinity=35, temperature=15 (IPTS-68 scale) and pressure=0,
which has numerical value 42.9140 mS/cm = 4.29140 S/m (see
Culkin and Smith, 1980, in the regression result cited at the bottom of
the left-hand column on page 23).
}
\description{
Compute electrical conductivity ratio based on salinity, temperature, and
pressure (relative to the conductivity of seawater with salinity=35,
temperature68=15, and pressure=0).
}
\details{
If \code{eos="unesco"}, the calculation is done by a bisection root search
on the UNESCO formula relating salinity to conductivity, temperature, and
pressure (see \code{\link[=swSCTp]{swSCTp()}}).  If it is \code{"gsw"} then the
Gibbs-SeaWater formulation is used, via \code{\link[=gsw_C_from_SP]{gsw_C_from_SP()}}.
}
\examples{
stopifnot(abs(1.0 - swCSTp(35, T90fromT68(15), 0, eos="unesco")) < 1e-7)
stopifnot(abs(1.0 - swCSTp(34.25045, T90fromT68(15), 2000, eos="unesco")) < 1e-7)
stopifnot(abs(1.0 - swCSTp(34.25045, T90fromT68(15), 2000, eos="gsw")) < 1e-7)

}
\references{
\enumerate{
\item Fofonoff, P. and R. C. Millard Jr, 1983. Algorithms for
computation of fundamental properties of seawater. \emph{Unesco Technical
Papers in Marine Science}, \emph{44}, 53 pp.
\item Culkin, F., and Norman D. Smith, 1980. Determination of the concentration of
potassium chloride solution having the same electrical conductivity, at 15 C
and infinite frequency, as standard seawater of salinity 35.0000 ppt
(Chlorinity 19.37394 ppt). \emph{IEEE Journal of Oceanic Engineering},
\emph{5}, pp 22-23.
}
}
\seealso{
For thermal (as opposed to electrical) conductivity, see
\code{\link[=swThermalConductivity]{swThermalConductivity()}}. For computation of salinity from
electrical conductivity, see \code{\link[=swSCTp]{swSCTp()}}.

Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/odf.R
\name{subset,odf-method}
\alias{subset,odf-method}
\title{Subset an ODF object}
\usage{
\S4method{subset}{odf}(x, subset, ...)
}
\arguments{
\item{x}{an \linkS4class{odf} object.}

\item{subset}{a condition to be applied to the \code{data} portion of
\code{x}.  See \sQuote{Details}.}

\item{\dots}{ignored.}
}
\value{
An \linkS4class{odf} object.
}
\description{
This function is somewhat analogous to \code{\link[=subset.data.frame]{subset.data.frame()}}.
}
\details{
It seems likely that users will first convert the odf object into
another class (e.g. ctd) and use the subset method of that class;
note that some of those methods interpret the \dots argument.
}
\seealso{
Other things related to odf data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{ODF2oce}()},
\code{\link{ODFListFromHeader}()},
\code{\link{ODFNames2oceNames}()},
\code{\link{[[,odf-method}},
\code{\link{[[<-,odf-method}},
\code{\link{odf-class}},
\code{\link{plot,odf-method}},
\code{\link{read.ctd.odf}()},
\code{\link{read.odf}()},
\code{\link{summary,odf-method}}

Other functions that subset oce objects: 
\code{\link{subset,adp-method}},
\code{\link{subset,adv-method}},
\code{\link{subset,amsr-method}},
\code{\link{subset,argo-method}},
\code{\link{subset,cm-method}},
\code{\link{subset,coastline-method}},
\code{\link{subset,ctd-method}},
\code{\link{subset,echosounder-method}},
\code{\link{subset,lobo-method}},
\code{\link{subset,met-method}},
\code{\link{subset,oce-method}},
\code{\link{subset,rsk-method}},
\code{\link{subset,sealevel-method}},
\code{\link{subset,section-method}},
\code{\link{subset,topo-method}},
\code{\link{subset,xbt-method}}
}
\author{
Dan Kelley
}
\concept{functions that subset oce objects}
\concept{things related to odf data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/argo.R
\name{plot,argo-method}
\alias{plot,argo-method}
\alias{plot.argo}
\title{Plot an argo Object}
\usage{
\S4method{plot}{argo}(
  x,
  which = 1,
  level,
  coastline = c("best", "coastlineWorld", "coastlineWorldMedium", "coastlineWorldFine",
    "none"),
  cex = 1,
  pch = 1,
  type = "p",
  col,
  fill = FALSE,
  projection = NULL,
  mgp = getOption("oceMgp"),
  mar = c(mgp[1] + 1.5, mgp[1] + 1.5, 1.5, 1.5),
  tformat,
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{x}{an \linkS4class{argo} object.}

\item{which}{list of desired plot types, one of the following. Note
that \code{\link[=oce.pmatch]{oce.pmatch()}} is used to try to complete partial
character matches, and that an error will occur if the match is
not complete (e.g. \code{"salinity"} matches to both
\code{"salinity ts"} and \code{"salinity profile"}.).
\itemize{
\item \code{which=1}, \code{which="trajectory"} or \code{which="map"} gives a
plot of the argo trajectory, with the coastline, if one is provided.
\item \code{which=2} or \code{"salinity ts"} gives a time series of
salinity at the indicated level(s)
\item \code{which=3} or \code{"temperature ts"} gives a time series
of temperature at the indicated level(s)
\item \code{which=4} or \code{"TS"} gives a TS diagram at the
indicated level(s)
\item \code{which=5} or \code{"salinity profile"} gives a salinity
profile of all the data (with S and p trimmed to the 1 and 99
percentiles)
\item \code{which=6} or \code{"temperature profile"} gives a
temperature profile (with T and p trimmed to the 1 and 99
percentiles)
}}

\item{level}{depth pseudo-level to plot, for \code{which=2} and higher.  May be an
integer, in which case it refers to an index of depth (1 being the top)
or it may be the string "all" which means to plot all data.}

\item{coastline}{character string giving the coastline to be used in an Argo-location
map, or \code{"best"} to pick the one with highest resolution, or
\code{"none"} to avoid drawing the coastline.}

\item{cex}{size of plotting symbols to be used if \code{type='p'}.}

\item{pch}{type of plotting symbols to be used if \code{type='p'}.}

\item{type}{plot type, either \code{"l"} or \code{"p"}.}

\item{col}{optional list of colors for plotting.}

\item{fill}{Either a logical, indicating whether to fill the land with
light-gray, or a color name.  Owing to problems with some projections, the
default is not to fill.}

\item{projection}{indication of the projection to be used
in trajectory maps. If this is \code{NULL}, no projection is used, although
the plot aspect ratio will be set to yield zero shape distortion at the
mean float latitude.  If \code{projection="automatic"}, then one
of two projections is used: stereopolar (i.e. \code{"+proj=stere +lon_0=X"}
where \code{X} is the mean longitude), or Mercator (i.e. \code{"+proj=merc"})
otherwise.  Otherwise, \code{projection} must be a character string specifying
a projection in the notation used by \code{\link[=oceProject]{oceProject()}} and \code{\link[=mapPlot]{mapPlot()}}.}

\item{mgp}{3-element numerical vector to use for \code{par(mgp)}, and also for
\code{par(mar)}, computed from this.  The default is tighter than the R
default, in order to use more space for the data and less for the axes.}

\item{mar}{value to be used with \code{par('mar')}.}

\item{tformat}{optional argument passed to \code{\link[=oce.plot.ts]{oce.plot.ts()}}, for plot
types that call that function.  (See \code{\link[=strptime]{strptime()}} for the format
used.)}

\item{debug}{debugging flag.}

\item{...}{optional arguments passed to plotting functions.}
}
\value{
None.
}
\description{
Plot a summary diagram for argo data.
}
\examples{
library(oce)
data(argo)
tc <- cut(argo[["time"]], "year")
# Example 1: plot map, which reveals float trajectory.
plot(argo, pch=as.integer(tc))
year <- substr(levels(tc), 1, 4)
data(topoWorld)
contour(topoWorld[['longitude']], topoWorld[['latitude']],
        topoWorld[['z']], add=TRUE)
legend("bottomleft", pch=seq_along(year), legend=year, bg="white", cex=3/4)

# Example 2: plot map, TS, T(z) and S(z). Note the use
# of handleFlags(), to skip over questionable data.
plot(handleFlags(argo), which=c(1, 4, 6, 5))

}
\seealso{
Other things related to argo data: 
\code{\link{[[,argo-method}},
\code{\link{[[<-,argo-method}},
\code{\link{argo-class}},
\code{\link{argoGrid}()},
\code{\link{argoNames2oceNames}()},
\code{\link{argo}},
\code{\link{as.argo}()},
\code{\link{handleFlags,argo-method}},
\code{\link{read.argo}()},
\code{\link{subset,argo-method}},
\code{\link{summary,argo-method}}

Other functions that plot oce data: 
\code{\link{download.amsr}()},
\code{\link{plot,adp-method}},
\code{\link{plot,adv-method}},
\code{\link{plot,amsr-method}},
\code{\link{plot,bremen-method}},
\code{\link{plot,cm-method}},
\code{\link{plot,coastline-method}},
\code{\link{plot,ctd-method}},
\code{\link{plot,gps-method}},
\code{\link{plot,ladp-method}},
\code{\link{plot,landsat-method}},
\code{\link{plot,lisst-method}},
\code{\link{plot,lobo-method}},
\code{\link{plot,met-method}},
\code{\link{plot,odf-method}},
\code{\link{plot,rsk-method}},
\code{\link{plot,satellite-method}},
\code{\link{plot,sealevel-method}},
\code{\link{plot,section-method}},
\code{\link{plot,tidem-method}},
\code{\link{plot,topo-method}},
\code{\link{plot,windrose-method}},
\code{\link{plot,xbt-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{tidem-class}}
}
\author{
Dan Kelley
}
\concept{functions that plot oce data}
\concept{things related to argo data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{byteToBinary}
\alias{byteToBinary}
\title{Format bytes as binary \link{defunct}}
\usage{
byteToBinary(x, endian = "big")
}
\arguments{
\item{x}{an integer to be interpreted as a byte.}

\item{endian}{character string indicating the endian-ness ("big" or
"little"). \strong{WARNING:} This argument will be removed soon.}
}
\value{
A character string representing the bit strings for the elements of
\code{x}, in order of significance for the \code{endian="big"} case.
(The nibbles, or 4-bit sequences, are interchanged in the now-deprecated
\code{"little"} case.)
See \dQuote{Examples} for how this relates to the output from
\link{rawToBits}.
}
\description{
\strong{WARNING:} The \code{endian} argument will soon be removed
from this function; see \link{oce-defunct}.
This is because the actions for \code{endian="little"} made
no sense in practical work. The default value for \code{endian}
was changed to \code{"big"} on 2017 May 6.
}
\examples{
library(oce)
## Note comparison with rawToBits():
a <- as.raw(0x0a)
byteToBinary(a, "big")        # "00001010"
as.integer(rev(rawToBits(a))) # 0 0 0 0 1 0 1 0
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adv.sontek.R
\name{read.adv.sontek.serial}
\alias{read.adv.sontek.serial}
\title{Read an ADV data file}
\usage{
read.adv.sontek.serial(
  file,
  from = 1,
  to,
  by = 1,
  tz = getOption("oceTz"),
  longitude = NA,
  latitude = NA,
  start = NULL,
  deltat = NULL,
  debug = getOption("oceDebug"),
  monitor = FALSE,
  processingLog = NULL
)
}
\arguments{
\item{file}{a connection or a character string giving the name of the file
to load.  It is also possible to give \code{file} as a vector of filenames,
to handle the case of data split into files by a data logger.  In the
multi-file case, \code{header} must be \code{FALSE}, \code{start} must be a
vector of times, and \code{deltat} must be provided.}

\item{from}{index number of the first profile to be read, or the time of
that profile, as created with \code{\link[=as.POSIXct]{as.POSIXct()}} (hint: use
\code{tz="UTC"}).  This argument is ignored if \code{header==FALSE}. See
\dQuote{Examples}.}

\item{to}{indication of the last profile to read, in a format matching that
of \code{from}.  This is ignored if \code{header==FALSE}.}

\item{by}{an indication of the stride length to use while walking through
the file.  This is ignored if \code{header==FALSE}.  Otherwise, if this is
an integer, then \code{by-1} profiles are skipped between each pair of
profiles that is read. This may not make much sense, if the data are not
equi-spaced in time.  If \code{by} is a string representing a time interval,
in colon-separated format, then this interval is divided by the sampling
interval, to get the stride length. \emph{BUG:} \code{by} only partially
works; see \dQuote{Bugs}.}

\item{tz}{character string indicating time zone to be assumed in the data.}

\item{longitude}{optional signed number indicating the longitude in degrees
East.}

\item{latitude}{optional signed number indicating the latitude in degrees
North.}

\item{start}{the time of the first sample, typically created with
\code{\link[=as.POSIXct]{as.POSIXct()}}.  This may be a vector of times,
if \code{filename} is a vector of file names.}

\item{deltat}{the time between samples.}

\item{debug}{a flag that turns on debugging.  The value indicates the depth
within the call stack to which debugging applies.  For example,
\code{read.adv.nortek()} calls \code{read.header.nortek()}, so that
\code{read.adv.nortek(...,debug=2)} provides information about not just the
main body of the data file, but also the details of the header.}

\item{monitor}{boolean value indicating whether to indicate the progress
of reading the file, by using \code{\link[=txtProgressBar]{txtProgressBar()}} or otherwise.  The value
of \code{monitor} is changed to \code{FALSE} automatically, for non-interactive
sessions.}

\item{processingLog}{if provided, the action item to be stored in the log.
This parameter is typically only provided for internal calls; the default
that it provides is better for normal calls by a user.}
}
\value{
An \linkS4class{adv} object that contains measurements made with an ADV device.

The \code{metadata} contains information as given in the following table.
The \verb{Nortek name'' is the name used in the Nortek System Integrator Guide (reference 1) and the }Sontek name'' is the name used in the relevant
Sontek documentation.  References are given in square brackets.

\tabular{llll}{

\strong{\code{metadata} name}\tab \strong{Nortek name} \tab \strong{Sontek name} \tab \strong{Meaning}\cr

\code{manufacturer}\tab - \tab - \tab Either \code{"nortek"} or \code{"sontek"}\cr

\code{instrumentType}\tab  - \tab - \tab Either \code{"vector"} or \code{"adv"}\cr

\code{filename}\tab  - \tab - \tab Name of data file(s)\cr

\code{latitude}\tab - \tab - \tab  Latitude of mooring (if applicable)\cr

\code{longitude}\tab - \tab - \tab Longitude of mooring (if applicable)\cr

\code{numberOfSamples}\tab - \tab - \tab Number of data samples in file\cr

\code{numberOfBeams}\tab NBeams (reference 1, p18) \tab - \tab Number of beams (always 3)\cr

\code{numberOfBeamSequencesPerBurst}\tab NPings \tab - \tab number of beam sequences per burst\cr

\code{measurementInterval}\tab MeasInterval (reference 1 p31) \tab - \tab \cr

\code{samplingRate}\tab 512/(AvgInterval) (reference 1 p30; reference 4) \tab - #' \tab \cr

}

The \code{data} list contains items with names corresponding to \code{adp}
objects, with an exception for Nortek data.  Nortek instruments report some
things at a time interval that is longer than the velocity sampling, and
these are stored in \code{data} as \code{timeSlow}, \code{headingSlow},
\code{pitchSlow}, \code{rollSlow}, and \code{temperatureSlow}; if burst
sampling was used, there will also be items \code{recordsBurst} and
\code{timeBurst}.

The \code{processingLog} is in the standard format.
}
\description{
Read an ADV data file, producing an object of type \code{adv}. This
function works by transferring control to a more specialized function,
e.g. \code{\link[=read.adp.nortek]{read.adp.nortek()}} and \code{\link[=read.adp.sontek]{read.adp.sontek()}},
and in many cases users will find it preferable to either use these or
the several even more specialized functions, if the file type is
known.
}
\details{
Files \emph{without} headers may be created in experiments in which a data
logger was set up to monitor the serial data stream from an instrument.  The
lack of header information places a burden on the user, who must supply such
basic information as the times of observations, the instrument orientation,
the instrument coordinate system, etc.  Example 3 below shows how to deal
with such files.  Three things should be noted.
\enumerate{
\item The user must choose the appropriate \code{read.adv} variant
corresponding to the instrument in question.  (This is necessary because
\code{\link[=oceMagic]{oceMagic()}}, which is used by the generic \code{\link[=read.oce]{read.oce()}}
routine, cannot determine the type of instrument by examining a file that
lacks a header.)
\item The call to the \code{read} function must include a start time
(\code{start}) and the number of seconds between data (\code{deltat}),
again, because the instrument data stream may lack those things when the
device is set to a serial mode.  Also, of course, it is necessary to set
\code{header=FALSE} in the function call.
\item Once the file has been read in, the user will be obliged to specify
other information, for the object to be well-formed.  For example, the
\code{read} function will have no way of knowing the instrument orientation,
the coordinate system being used, the transformation matrix to go from
\code{"beam"} to \code{"xyz"} coordinates, or the instrument heading, pitch,
and roll, to go from \code{"xyz"} coordinates to \code{"enu"} coordinates.
Such things are illustrated in example 3 below.
}

In ADV data files, velocities are coded to signed 2-byte integers, with a
scale factor being used to convert to velocity in metres per second.  These
two facts control the maximum recordable velocity and the velocity
resolution, values that may be retrieved for an ADV object name \code{d}
with \code{d[["velocityMaximum"]]} and \code{d[["velocityResolution"]]}.
}
\section{Nortek files}{


\strong{Sampling-rate and similar issues}

The data format is inferred from the System Integrator Guide (reference 1A) and System
Integrator Manual (reference 1B).  These document lacks clarity in spots, and so
\code{read.adv.nortek} contains some assumptions that are noted here, so
that users will be aware of possible problems.

A prominent example is the specification of the sampling rate, stored in
\code{metadata$sampingRate} in the return value.  Repeated examination of
the System Integrator Guide (reference 1) failed to indicate where this value is
stored in the various headers contained in Vector datasets.  After some
experimentation with a few data files, \code{read.adv.nortek} was set up to
calculate \code{metadata$samplingRate} as \code{512/AvgInterval} where
\code{AvgInterval} is a part of the \verb{User Configuration'' header (reference 1 p30), where the explanation is }average interval in seconds'').  This formula was
developed through trial and error, but it was confirmed later on the Nortek
discussion group, and it should appear in upcoming versions of (reference 1).

Another basic issue is the determination of whether an instrument had
recorded in continuous mode or burst mode.  One might infer that
\code{TimCtrlReg} in the \verb{User Configuration'' header (reference 1 p30) determines this, in bits 1 and 2.  However, this was the case in test files available to the author.  For this reason, `read.adv.nortek` infers the mode by reverse engineering of data files of known configuration.  The present version of `read.adv.nortek` determines the sampling mode from the ```NRecords`'' item of the }Vector Velocity Data'' header, which seems
to be 0 for data collected continuously, and non-zero for data collected in
bursts.

Taking these things together, we come upon the issue of how to infer
sampling times for Nortek instruments.  There do not seem to be definitive
documents on this, and so \code{read.adv.nortek} is based partly on
information (of unknown quality) found on Nortek discussion boards.  The
present version of \code{read.adv.nortek} infers the times of velocity
observations differently, depending on whether the instrument was set to
record in burst mode or continuous mode.  For burst mode, times stated in
the burst headers are used, but for continuous mode, times stated in the
``vector system data'' are used.  On the advice found on a Nortek discussion
board, the burst-mode times are offset by 2 seconds to allow for the
instrument warm-up period.

\strong{Handling IMU (inertial measurement unit) data}

Starting in March
2016, \code{read.adv.nortek} has offered some support for handling IMU
(inertial measurement unit) data incorporated into Nortek binary files. This
is not described in the Nortek document named \verb{System Integrator Guide'' (reference 1A) but it appeared in }System Integrator Manual'' (reference 1B;
reference 1C). Confusingly, 1B described 3 varieties of data, whereas 1C does not
describe any of these, but describes instead a fourth variety. As of March
2016, \code{read.adv.nortek} handles all 4 varieties, because files in the
various schemes appear to exist. In \code{oce}, the varieties are named
after the byte code that flags them.  (Variety \code{c3} is the one
described in (reference 1C); the others were described in (reference 1B).)
The variety is stored in the \code{metadata} slot of the returned object as
a string named \code{IMUtype}.

For each variety, the reader is cautioned that strong tests have not been
performed on the code.  One way to test the code is to compare with textual
data files produced by the Nortek software.  In March 2016, an \code{oce}
user shared a dataset of the \code{c3} variety, and this permitted detailed
comparison between the text file and the values inferred by
\code{read.adv.nortek}. The test suggested agreement (to within the
resolution printed in the text file) for velocity (\code{v} in the
\code{data} slot), signal amplitude (\code{a}), correlation (\code{q}),
pressure (\code{p}), the three components of IMU delta angle
(\code{IMUdeltaAngleX} etc), and all components of the rotation matrix
(\code{IMUrotation}).  However, the delta velocity signals did not match,
with \code{IMUdeltaVelocityX} disagreeing in the second decimal place,
\code{IMUdeltaVelocityY} component disagreeing in the first, and
\code{IMUdeltaVelocityZ} being out by a factor of about 10. This is github
issue 893 (\url{https://github.com/dankelley/oce/issues/893}).
\itemize{
\item Variety \code{c3} (signalled by byte 5 of a sequence being
\code{0xc3}) provides information on what Nortek calls DeltaAngle,
DeltaVelocity and Orientation Matrix. (Apart from the orientation matrix,
Nortek provides no documentation on what these quantities mean.) In the
object returned by \code{read.adv.nortek}, these are stored in the
\code{data} slot as \code{IMUdeltaAngleX}, \code{IMUdeltaAngleY},
\code{IMUdeltaAngleZ}, \code{IMUdeltaVelocityX}, \code{IMUdeltaVelocityY},
\code{IMUdeltaVelocityZ}, and \code{IMUrotation}, all vectors except the
last, which is a 3D array. In addition to these, \code{IMUtimestamp} is a
time-stamp, which is not defined in the Nortek documents but seems, from IMU
documents (reference 5), to be defined based on a clock that ticks once per 16
microseconds. Caution may be required in dealing with this timestamp, since
it seemed sensible in one test case (variety \code{d3}) but kept reseting to
zero in another (variety \code{c3}). The lack of Nortek documentation on
most of these quantities is a roadblock to implementing \code{oce} functions
dealing with IMU-enabled datasets
\item Variety \code{cc} (signalled by byte 5 of a sequence being
\code{0xcc}) provides information on acceleration, angular rotation rate,
magnetic vector and orientation matrix.  Each is a timeseries. Acceleration
is stored in the \code{data} slot as \code{IMUaccelX}, \code{IMUaccelY},
\code{IMUaccelz}. The angular rotation components are \code{IMUngrtX},
\code{IMUngrtY} and \code{IMUngrtz}. The magnetic data are in
\code{IMUmagrtx}, \code{IMUmagrty} and \code{IMUmagrtz}. Finally,
\code{IMUmatrix} is a rotation matrix made up from elements named
\code{M11}, \code{M12}, etc in the Nortek documentation.  In addition to all
of these, \code{IMUtime} stores time in seconds, with an origin whose
definition is not stated in reference 1B.
\item Variety \code{d2} (signalled by byte 5 being \code{0xd2}) provides
information on gyro-stabilized acceleration, angular rate and magnetometer
vectors.  The data stored \code{MUaccelX}, \code{IMUangrtX},
\code{IMUmagrtX}, with similar for \code{Y} and \code{Z}.  Again, time is in
\code{IMUtime}. This data type has not been tested as of mid-March 2016,
because the developers do not have a test file with which to test.
\item Variety \code{d3} (signalled by byte 5 being \code{0xd3}) provides
information on DeltaAngle, DeltaVelocity and magnetometer vectors, stored in
\code{IMUdeltaAngleX}, \code{IMUdeltaVelocityX}, and
\code{IMUdeltaMagVectorX}, with similar for \code{Y} and \code{Z}. Again,
time is in \code{IMUtime}. This data type has not been tested as of
mid-March 2016, because the developers do not have a test file with which to
test.
}
}

\examples{
\dontrun{
library(oce)
# A nortek Vector file
d <- read.oce("/data/archive/sleiwex/2008/moorings/m05/adv/nortek_1943/raw/adv_nortek_1943.vec",
              from=as.POSIXct("2008-06-26 00:00:00", tz="UTC"),
              to=as.POSIXct("2008-06-26 00:00:10", tz="UTC"))
plot(d, which=c(1:3,15))
}
}
\references{
1A. Nortek AS.  System Integrator Guide (paradopp family of products). June
2008.  (Doc No: PSI00-0101-0608).  (Users may find it helpful to also
examine newer versions of the guide.)

1B. Nortek AS.  System Integrator Manual. Dec 2014.
(\code{system-integrator-manual_Dec2014_jan.pdf})

1C. Nortek AS.  System Integrator Manual. March 2016.
(\code{system-integrator-manual_Mar2016.pdf})
\enumerate{
\item SonTek/YSI ADVField/Hydra Acoustic Doppler Velocimeter (Field) Technical
Documentation (Sept 1, 2001).
\item Appendix 2.2.3 of the Sontek ADV operation Manual, Firmware Version 4.0
(Oct 1997).
\item Nortek Knowledge Center (http://www.nortekusa.com/en/knowledge-center)
\item A document describing an IMU unit that seems to be close to the one named
in (references 1B and C) as being an adjunct to Nortek Vector systems is at
\verb{http://files.microstrain.com/3DM-GX3-35-Data-Communications-Protocol.pdf}
}
}
\seealso{
Other things related to adv data: 
\code{\link{[[,adv-method}},
\code{\link{[[<-,adv-method}},
\code{\link{adv-class}},
\code{\link{adv}},
\code{\link{beamName}()},
\code{\link{beamToXyz}()},
\code{\link{enuToOtherAdv}()},
\code{\link{enuToOther}()},
\code{\link{plot,adv-method}},
\code{\link{read.adv.nortek}()},
\code{\link{read.adv.sontek.adr}()},
\code{\link{read.adv.sontek.text}()},
\code{\link{read.adv}()},
\code{\link{rotateAboutZ}()},
\code{\link{subset,adv-method}},
\code{\link{summary,adv-method}},
\code{\link{toEnuAdv}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdv}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley
}
\concept{things related to adv data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.R
\name{beamToXyzAdpAD2CP}
\alias{beamToXyzAdpAD2CP}
\title{Convert AD2CP-style adp data From Beam to XYZ Coordinates}
\usage{
beamToXyzAdpAD2CP(x, debug = getOption("oceDebug"))
}
\arguments{
\item{x}{an \linkS4class{adp} object.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\description{
This looks at all the items in the \code{data} slot of \code{x}, to
see if they contain an array named \code{v} that holds velocity.
If that velocity has 4 components, and if \code{oceCoordinate} for
the item is \code{"beam"}, then
along-beam velocity components \eqn{B_1}{B1}
\eqn{B_2}{B1}, \eqn{B_3}{B3}, and \eqn{B_4}{B4}
are converted to instrument-oriented Cartesian velocity components \eqn{u}{u}
\eqn{v}{v} and \eqn{w}{w}
using the convex-geometry formulae from section 5.5 of reference 1,
viz.
\eqn{u=ca(B_1-B_2)}{u=a*(B1-B2)}, \eqn{v=ca(B_4-B_3)}{v=a*(B4-B3)},
\eqn{w=-b(B_1+B_2+B_3+B_4)}{w=-b*(B1+B2+B3+B4)}. In addition to these,
an estimate of the
error in velocity is computed as
\eqn{e=d(B_1+B_2-B_3-B_4)}{e=d*(B1+B2-B3-B4)}.
The geometrical factors in these formulae are:
\eqn{a=1/(2\sin\theta)}{a=1/(2*sin(theta))}
where \eqn{\theta}{theta} is the angle the beams make to the axial direction
(which is available as \code{x[["beamAngle"]]}),
\eqn{b=1/(4\cos\theta)}{b=1/(4*cos(theta))}, and
\eqn{d=a/\sqrt{2}}{d=a/sqrt(2)}.
}
\references{
\enumerate{
\item Teledyne RD Instruments.
\dQuote{ADCP Coordinate Transformation: Formulas and Calculations,}
January 2010. P/N 951-6079-00.
}
}
\seealso{
Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/map.R
\name{mapContour}
\alias{mapContour}
\title{Add Contours on a Existing map}
\usage{
mapContour(
  longitude,
  latitude,
  z,
  nlevels = 10,
  levels = pretty(range(z, na.rm = TRUE), nlevels),
  labcex = 0.6,
  drawlabels = TRUE,
  underlay = "erase",
  col = par("fg"),
  lty = par("lty"),
  lwd = par("lwd"),
  debug = getOption("oceDebug")
)
}
\arguments{
\item{longitude}{numeric vector of longitudes of points to be plotted, or an object of
class \code{topo} (see \linkS4class{topo}), in which case
\code{longitude}, \code{latitude} and \code{z} are inferred from that object.}

\item{latitude}{numeric vector of latitudes of points to be plotted.}

\item{z}{matrix to be contoured. The number of rows and columns in \code{z}
must equal the lengths of \code{longitude} and \code{latitude}, respectively.}

\item{nlevels}{number of contour levels, if and only if \code{levels} is not supplied.}

\item{levels}{vector of contour levels.}

\item{labcex}{\code{cex} value used for contour labelling. As with
\code{\link[=contour]{contour()}}, this is an absolute size, not a multiple of
\code{\link{par}}\code{("cex")}.}

\item{drawlabels}{logical value or vector indicating whether to draw contour
labels.  If the length of \code{drawlabels} is less than the number of
levels specified, then \code{\link[=rep]{rep()}} is used to increase the length,
providing a value for each contour line. For those levels that are thus
indicated, labels are added, at a spot where the contour line is
closest to horizontal on the page. First, though, the region underneath
the label is filled with the colour given by \code{\link{par}}\code{("bg")}.
See \dQuote{Limitations} for notes on the status of contour
labelling, and its limitations.}

\item{underlay}{character value relating to handling labels. If
this equals \code{"erase"} (which is the default), then the contour line
is drawn  first, then the area under the label is erased (filled with
white 'ink'), and then the label is drawn. This can be useful
in drawing coarsely-spaced labelled contours on top of finely-spaced
unlabelled contours. On the other hand, if \code{underlay} equals
\code{"interrupt"}, then the contour line is interrupted in the
region of the label, which is closer to the scheme used by the
base \code{\link[=contour]{contour()}} function.}

\item{col}{colour of the contour line, as for \code{\link{par}}\code{("col")},
except here \code{col} gets lengthened by calling \code{\link[=rep]{rep()}},
so that individual contours can be coloured distinctly.}

\item{lty}{type of the contour line, as for \code{\link{par}}\code{("lty")},
except for lengthening, as described for \code{col}.}

\item{lwd}{width of the contour line, as for \code{\link{par}}\code{("lwd")},
except for lengthening, as described for \code{col} and \code{lty}.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\description{
Plot contours on an existing map.
}
\details{
Adds contour lines to an existing map, using \code{\link[=mapLines]{mapLines()}}.

The ability to label the contours was added in February, 2019, and
how this works may change through the summer months of that year.
Note that label placement in \code{mapContour} is handled differently
than in \code{\link[=contour]{contour()}}.
}
\examples{
\dontrun{
library(oce)
data(coastlineWorld)
if (requireNamespace("ocedata", quietly=TRUE)) {
    data(levitus, package="ocedata")
    par(mar=rep(1, 4))
    mapPlot(coastlineWorld, projection="+proj=robin", col="lightgray")
    mapContour(levitus[['longitude']], levitus[['latitude']], levitus[['SST']])
}
}

}
\seealso{
A map must first have been created with \code{\link[=mapPlot]{mapPlot()}}.

Other functions related to maps: 
\code{\link{formatPosition}()},
\code{\link{lonlat2map}()},
\code{\link{lonlat2utm}()},
\code{\link{map2lonlat}()},
\code{\link{mapArrows}()},
\code{\link{mapAxis}()},
\code{\link{mapCoordinateSystem}()},
\code{\link{mapDirectionField}()},
\code{\link{mapGrid}()},
\code{\link{mapImage}()},
\code{\link{mapLines}()},
\code{\link{mapLocator}()},
\code{\link{mapLongitudeLatitudeXY}()},
\code{\link{mapPlot}()},
\code{\link{mapPoints}()},
\code{\link{mapPolygon}()},
\code{\link{mapScalebar}()},
\code{\link{mapText}()},
\code{\link{mapTissot}()},
\code{\link{oceCRS}()},
\code{\link{shiftLongitude}()},
\code{\link{usrLonLat}()},
\code{\link{utm2lonlat}()}
}
\author{
Dan Kelley
}
\concept{functions related to maps}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/section.R
\docType{class}
\name{section-class}
\alias{section-class}
\title{Class to Store Hydrographic Section Data}
\description{
This class stores data from oceanographic section surveys.
}
\details{
Sections can be read with \code{\link[=read.section]{read.section()}} or created with
\code{\link[=read.section]{read.section()}} or created from CTD objects by using
\code{\link[=as.section]{as.section()}} or by adding a ctd station to an existing section with
\code{\link[=sectionAddStation]{sectionAddStation()}}.

Sections may be sorted with \code{\link[=sectionSort]{sectionSort()}}, subsetted with
\code{\link[=subset,section-method]{subset,section-method()}}, smoothed with \code{\link[=sectionSmooth]{sectionSmooth()}}, and
gridded with \code{\link[=sectionGrid]{sectionGrid()}}.  A "spine" may be added to a section
with \code{\link[=addSpine]{addSpine()}}.  Sections may be summarized with
\code{\link[=summary,section-method]{summary,section-method()}} and plotted
with \code{\link[=plot,section-method]{plot,section-method()}}.

The sample dataset \code{\link[=section]{section()}} contains data along WOCE line A03.
}
\section{Slots}{

\describe{
\item{\code{data}}{As with all \code{oce} objects, the \code{data} slot
for \code{section} objects is a
\link{list} containing the main data for the object.
{}}

\item{\code{metadata}}{As with all \code{oce} objects, the \code{metadata} slot
for \code{section} objects is a \link{list} containing
information about the \code{data} or about the object itself.
Examples that are of common interest include \code{stationId}, \code{longitude}, \code{latitude} and \code{time}.}

\item{\code{processingLog}}{As with all \code{oce} objects, the \code{processingLog} slot
for \code{section} objects is a
\link{list} with entries describing the creation and evolution
of the object. The contents are updated by various \code{oce} functions to
keep a record of processing steps.  Object summaries and
\code{\link[=processingLogShow]{processingLogShow()}} both display the log.}
}}

\section{Modifying slot contents}{


Although the \code{[[<-} operator may permit modification of the contents
of \linkS4class{section} objects (see \code{\link{[[<-,section-method}}),
it is better to use \code{\link[=oceSetData]{oceSetData()}} and \code{\link[=oceSetMetadata]{oceSetMetadata()}},
because those functions save an entry in the \code{processingLog}
that describes the change.
}

\section{Retrieving slot contents}{


The full contents of the \code{data} and \code{metadata} slots of a \linkS4class{section}
object may be retrieved in the standard R way using \code{\link[=slot]{slot()}}. For
example \code{slot(o,"data")} returns the \code{data} slot of an object named \code{o},
and similarly \code{slot(o,"metadata")} returns
the \code{metadata} slot.

The slots may also be obtained with the \code{\link{[[,section-method}}
operator, as e.g. \code{o[["data"]]} and \code{o[["metadata"]]}, respectively.

The \code{\link{[[,section-method}} operator can also
be used to retrieve items from within the \code{data} and \code{metadata} slots.
For example, \code{o[["temperature"]]} can be used to retrieve temperature
from an object containing that quantity. The rule is that a named
quantity is sought first within the object's \code{metadata} slot,
with the \code{data} slot being checked only if \code{metadata} does not
contain the item. This \code{[[} method can also be used to get
certain derived quantities, if the object contains sufficient
information to calculate them. For example, an object that holds
(practical) salinity, temperature and pressure, along with
longitude and latitude, has sufficient information to compute
Absolute Salinity, and so \code{o[["SA"]]} will yield the
calculated Absolute Salinity.

It is also possible to find items more directly, using \code{\link[=oceGetData]{oceGetData()}} and
\code{\link[=oceGetMetadata]{oceGetMetadata()}}, but neither of these functions can
retrieve derived items.
}

\examples{
library(oce)
data(section)
plot(section[['station', 1]])
pairs(cbind(z=-section[["pressure"]],T=section[["temperature"]],S=section[["salinity"]]))
## T profiles for first few stations in section, at common scale
par(mfrow=c(3,3))
Tlim <- range(section[["temperature"]])
ylim <- rev(range(section[["pressure"]]))
for (stn in section[["station", 1:9]])
    plotProfile(stn, xtype='potential temperature', ylim=ylim, Tlim=Tlim)

}
\seealso{
Other classes provided by oce: 
\code{\link{adp-class}},
\code{\link{adv-class}},
\code{\link{argo-class}},
\code{\link{bremen-class}},
\code{\link{cm-class}},
\code{\link{coastline-class}},
\code{\link{ctd-class}},
\code{\link{lisst-class}},
\code{\link{lobo-class}},
\code{\link{met-class}},
\code{\link{oce-class}},
\code{\link{odf-class}},
\code{\link{rsk-class}},
\code{\link{sealevel-class}},
\code{\link{topo-class}},
\code{\link{windrose-class}},
\code{\link{xbt-class}}

Other things related to section data: 
\code{\link{[[,section-method}},
\code{\link{[[<-,section-method}},
\code{\link{as.section}()},
\code{\link{handleFlags,section-method}},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{plot,section-method}},
\code{\link{read.section}()},
\code{\link{sectionAddStation}()},
\code{\link{sectionGrid}()},
\code{\link{sectionSmooth}()},
\code{\link{sectionSort}()},
\code{\link{section}},
\code{\link{subset,section-method}},
\code{\link{summary,section-method}}
}
\author{
Dan Kelley
}
\concept{classes provided by oce}
\concept{things related to section data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/rsk.R
\name{rskToc}
\alias{rskToc}
\title{Decode table-of-contents File from a Rsk File}
\usage{
rskToc(dir, from, to, debug = getOption("oceDebug"))
}
\arguments{
\item{dir}{name of a directory containing a single table-of-contents file, with
\code{.TBL} at the end of its file name.}

\item{from}{optional \code{\link[=POSIXct]{POSIXct()}} time, indicating the beginning of a
data interval of interest.  This must have timezone \code{"UTC"}.}

\item{to}{optional \code{\link[=POSIXct]{POSIXct()}} time, indicating the end of a data
interval of interest.  This must have timezone \code{"UTC"}.}

\item{debug}{optional integer to control debugging, with positive values
indicating to print information about the processing.}
}
\value{
A list with two elements: \code{filename}, a vector of file names, and
\code{startTime}, a vector of \code{\link[=POSIXct]{POSIXct()}} times indicating the (real)
times of the first datum in the corresponding files.
}
\description{
Decode table-of-contents file from a rsk file, of the sort used by some
researchers at Dalhousie University.
}
\details{
It is assumed that the \code{.TBL} file contains lines of the form \code{"File
  \\day179\\SL08A179.023 started at Fri Jun 27 22:00:00 2008"} The first step is
to parse these lines to get day and hour information, i.e.  179 and 023 in the
line above.  Then, recognizing that it is common to change the names of such
files, the rest of the file-name information in the line is ignored, and instead
a new file name is constructed based on the data files that are found in the
directory.  (In other words, the \code{"\\\\day179\\\\SL08A"} portion of the line is
replaced.)  Once the file list is complete, with all times put into R format,
then (optionally) the list is trimmed to the time interval indicated by
\code{from} and \code{to}.  It is important that \code{from} and \code{to} be in
the \code{UTC} time zone, because that time zone is used in decoding the lines
in the \code{.TBL} file.
}
\examples{
\dontrun{
table <- rskToc("/data/archive/sleiwex/2008/moorings/m05/adv/sontek_202h/raw",
from=as.POSIXct("2008-07-01 00:00:00", tz="UTC"),
    to=as.POSIXct("2008-07-01 12:00:00", tz="UTC"))
print(table)
}

}
\seealso{
Other things related to rsk data: 
\code{\link{[[,rsk-method}},
\code{\link{[[<-,rsk-method}},
\code{\link{as.rsk}()},
\code{\link{plot,rsk-method}},
\code{\link{read.rsk}()},
\code{\link{rsk-class}},
\code{\link{rskPatm}()},
\code{\link{rsk}},
\code{\link{subset,rsk-method}},
\code{\link{summary,rsk-method}}
}
\author{
Dan Kelley
}
\concept{things related to rsk data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/lisst.R
\docType{class}
\name{lisst-class}
\alias{lisst-class}
\title{Class to Store LISST Data}
\description{
This class stores LISST (Laser in-situ scattering and transmissometry) data.
}
\section{Slots}{

\describe{
\item{\code{data}}{As with all \code{oce} objects, the \code{data} slot
for \code{lisst} objects is a
\link{list} containing the main data for the object.
{}}

\item{\code{metadata}}{As with all \code{oce} objects, the \code{metadata} slot
for \code{lisst} objects is a \link{list} containing
information about the \code{data} or about the object itself.
{}}

\item{\code{processingLog}}{As with all \code{oce} objects, the \code{processingLog} slot
for \code{lisst} objects is a
\link{list} with entries describing the creation and evolution
of the object. The contents are updated by various \code{oce} functions to
keep a record of processing steps.  Object summaries and
\code{\link[=processingLogShow]{processingLogShow()}} both display the log.}
}}

\section{Modifying slot contents}{


Although the \code{[[<-} operator may permit modification of the contents
of \linkS4class{lisst} objects (see \code{\link{[[<-,lisst-method}}),
it is better to use \code{\link[=oceSetData]{oceSetData()}} and \code{\link[=oceSetMetadata]{oceSetMetadata()}},
because those functions save an entry in the \code{processingLog}
that describes the change.
}

\section{Retrieving slot contents}{


The full contents of the \code{data} and \code{metadata} slots of a \linkS4class{lisst}
object may be retrieved in the standard R way using \code{\link[=slot]{slot()}}. For
example \code{slot(o,"data")} returns the \code{data} slot of an object named \code{o},
and similarly \code{slot(o,"metadata")} returns
the \code{metadata} slot.

The slots may also be obtained with the \code{\link{[[,lisst-method}}
operator, as e.g. \code{o[["data"]]} and \code{o[["metadata"]]}, respectively.

The \code{\link{[[,lisst-method}} operator can also
be used to retrieve items from within the \code{data} and \code{metadata} slots.
For example, \code{o[["temperature"]]} can be used to retrieve temperature
from an object containing that quantity. The rule is that a named
quantity is sought first within the object's \code{metadata} slot,
with the \code{data} slot being checked only if \code{metadata} does not
contain the item. This \code{[[} method can also be used to get
certain derived quantities, if the object contains sufficient
information to calculate them. For example, an object that holds
(practical) salinity, temperature and pressure, along with
longitude and latitude, has sufficient information to compute
Absolute Salinity, and so \code{o[["SA"]]} will yield the
calculated Absolute Salinity.

It is also possible to find items more directly, using \code{\link[=oceGetData]{oceGetData()}} and
\code{\link[=oceGetMetadata]{oceGetMetadata()}}, but neither of these functions can
retrieve derived items.
}

\references{
Information about LISST instruments is provided at
the manufacturer's website, \verb{https://www.sequoiasci.com}.
}
\seealso{
Other classes provided by oce: 
\code{\link{adp-class}},
\code{\link{adv-class}},
\code{\link{argo-class}},
\code{\link{bremen-class}},
\code{\link{cm-class}},
\code{\link{coastline-class}},
\code{\link{ctd-class}},
\code{\link{lobo-class}},
\code{\link{met-class}},
\code{\link{oce-class}},
\code{\link{odf-class}},
\code{\link{rsk-class}},
\code{\link{sealevel-class}},
\code{\link{section-class}},
\code{\link{topo-class}},
\code{\link{windrose-class}},
\code{\link{xbt-class}}

Other things related to lisst data: 
\code{\link{[[,lisst-method}},
\code{\link{[[<-,lisst-method}},
\code{\link{as.lisst}()},
\code{\link{plot,lisst-method}},
\code{\link{read.lisst}()},
\code{\link{summary,lisst-method}}
}
\author{
Dan Kelley
}
\concept{classes provided by oce}
\concept{things related to lisst data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{binMean2D}
\alias{binMean2D}
\title{Bin-average f=f(x,y)}
\usage{
binMean2D(
  x,
  y,
  f,
  xbreaks,
  ybreaks,
  flatten = FALSE,
  fill = FALSE,
  fillgap = -1
)
}
\arguments{
\item{x}{vector of numerical values.}

\item{y}{vector of numerical values.}

\item{f}{Matrix of numerical values, a matrix f=f(x,y).}

\item{xbreaks}{Vector of values of \code{x} at the boundaries between bins, calculated using
\code{\link{pretty}}\code{(x)} if not supplied.}

\item{ybreaks}{Vector of values of \code{y} at the boundaries between bins, calculated using
\code{\link{pretty}}\code{(y)} if not supplied.}

\item{flatten}{A logical value indicating whether
the return value also contains equilength
vectors \code{x}, \code{y}, \code{z} and \code{n}, a flattened
representation of \code{xmids}, \code{ymids}, \code{result} and
\code{number}.}

\item{fill}{Logical value indicating whether to fill \code{NA}-value gaps in
the matrix. Gaps will be filled as the average of linear interpolations
across rows and columns. See \code{fillgap}, which works together with this.}

\item{fillgap}{Integer controlling the size of gap that can be filled
across. If this is negative (as in the default), gaps will be filled
regardless of their size. If it is positive, then gaps exceeding this
number of indices will not be filled.}
}
\value{
A list with the following elements: the midpoints (renamed as
\code{x} and \code{y}), the count (\code{number}) of \code{f(x,y)} values
for \code{x} and \code{y} values that lie between corresponding breaks,
and the resultant average (\code{f}) of \code{f(x,y)}, classified by the
\code{x} and \code{y} breaks.
}
\description{
Average the values of a vector \code{f(x,y)} in bins defined on
vectors \code{x} and \code{y}. A common example might be averaging
spatial data into location bins.
}
\examples{
library(oce)
x <- runif(500)
y <- runif(500)
f <- x + y
xb <- seq(0, 1, 0.1)
yb <- seq(0, 1, 0.2)
m <- binMean2D(x, y, f, xb, yb)
plot(x, y)
contour(m$xmids, m$ymids, m$result, add=TRUE, levels=seq(0, 2, 0.5), labcex=1)

}
\seealso{
Other bin-related functions: 
\code{\link{binApply1D}()},
\code{\link{binApply2D}()},
\code{\link{binAverage}()},
\code{\link{binCount1D}()},
\code{\link{binCount2D}()},
\code{\link{binMean1D}()}
}
\author{
Dan Kelley
}
\concept{bin-related functions}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/landsat.R
\name{[[<-,landsat-method}
\alias{[[<-,landsat-method}
\title{Replace Parts of a landsat Object}
\usage{
\S4method{[[}{landsat}(x, i, j, ...) <- value
}
\arguments{
\item{x}{a \linkS4class{landsat} object.}

\item{i}{character value naming the item to replace.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{optional additional information (ignored).}

\item{value}{The value to be placed into \code{x}, somewhere.}
}
\description{
The \code{[[<-} method works for all \linkS4class{oce} objects.
The purpose, as with the related extraction method, \code{[[},
is to insulate users from the internal details of \linkS4class{oce}
objects, by looking for items within the various storage
slots of the object. Items not actually stored can also be
replaced, including units and data-quality
flags.
}
\details{
As with \code{[[} method, the procedure works in steps.

First, the \code{metadata} slot of \code{x} is checked to
see whether it contains something named with \code{i}.
If so, then the named item is replaced with \code{value}.

Otherwise, if the string value of \code{i} ends in \code{Unit}, then the
characters preceding that are taken as the name of a variable, and
the \code{metadata} slot of \code{x} is updated to store that unit, e.g.\preformatted{x[["temperatureUnits"]] <- list(unit=expression(degree*F),scale="")
}

Similarly, if \code{i} ends in \code{Flag}, then quality-control
flags are set up as defined by \code{result}, e.g.\preformatted{o[["temperatureFlags"]] <- c(2,4,2,2)
}

Otherwise, \code{\link[=pmatch]{pmatch()}} is used for a partial-string match with
the names of the items that are in the \code{data} slot of \code{x}.
The first item found (if any) is then updated to hold the value \code{result}.

If none of these conditions is met, a warning is issued.
}
\seealso{
Other functions that replace parts of oce objects: 
\code{\link{[[<-,adp-method}},
\code{\link{[[<-,amsr-method}},
\code{\link{[[<-,argo-method}},
\code{\link{[[<-,bremen-method}},
\code{\link{[[<-,cm-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{[[<-,echosounder-method}},
\code{\link{[[<-,g1sst-method}},
\code{\link{[[<-,gps-method}},
\code{\link{[[<-,ladp-method}},
\code{\link{[[<-,lisst-method}},
\code{\link{[[<-,lobo-method}},
\code{\link{[[<-,met-method}},
\code{\link{[[<-,oce-method}},
\code{\link{[[<-,odf-method}},
\code{\link{[[<-,rsk-method}},
\code{\link{[[<-,sealevel-method}},
\code{\link{[[<-,section-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{[[<-,topo-method}},
\code{\link{[[<-,windrose-method}},
\code{\link{[[<-,xbt-method}}

Other things related to landsat data: 
\code{\link{[[,landsat-method}},
\code{\link{landsat-class}},
\code{\link{landsatAdd}()},
\code{\link{landsatTrim}()},
\code{\link{landsat}},
\code{\link{plot,landsat-method}},
\code{\link{read.landsat}()},
\code{\link{summary,landsat-method}}
}
\concept{functions that replace parts of oce objects}
\concept{things related to landsat data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.sontek.R
\name{read.adp.sontek}
\alias{read.adp.sontek}
\title{Read a Sontek ADP File}
\usage{
read.adp.sontek(
  file,
  from = 1,
  to,
  by = 1,
  tz = getOption("oceTz"),
  longitude = NA,
  latitude = NA,
  type = c("adp", "pcadp"),
  monitor = FALSE,
  despike = FALSE,
  processingLog,
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{file}{a connection or a character string giving the name of the file
to load.  (For \code{read.adp.sontek.serial}, this is generally a list of
files, which will be concatenated.)}

\item{from}{indication of the first profile to read.  This can be an
integer, the sequence number of the first profile to read, or a POSIXt time
before which profiles should be skipped, or a character string that converts
to a POSIXt time (assuming UTC timezone).  See \dQuote{Examples}, and make
careful note of the use of the \code{tz} argument. If \code{from} is not
supplied, it defaults to 1.}

\item{to}{an optional indication of the last profile to read, in a
format as described for \code{from}.  As a special case, \code{to=0} means
to read the file to the end.  If \code{to} is not supplied, then it defaults
to 0.}

\item{by}{an optional indication of the stride length to use while walking through
the file.  If this is an integer, then \code{by-1} profiles are skipped
between each pair of profiles that is read, e.g. the default \code{by=1}
means to read all the data.  (For RDI files \emph{only}, there are some
extra features to avoid running out of memory; see \dQuote{Memory considerations}.)}

\item{tz}{character string indicating time zone to be assumed in the data.}

\item{longitude}{optional signed number indicating the longitude in degrees
East.}

\item{latitude}{optional signed number indicating the latitude in degrees
North.}

\item{type}{A character string indicating the type of instrument.}

\item{monitor}{boolean value indicating whether to indicate the progress
of reading the file, by using \code{\link[=txtProgressBar]{txtProgressBar()}} or otherwise.  The value
of \code{monitor} is changed to \code{FALSE} automatically, for non-interactive
sessions.}

\item{despike}{if \code{TRUE}, \code{\link[=despike]{despike()}} will be used to clean
anomalous spikes in heading, etc.}

\item{processingLog}{if provided, the action item to be stored in the log.
(Typically only provided for internal calls; the default that it provides is
better for normal calls by a user.)}

\item{debug}{a flag that turns on debugging.  Set to 1 to get a moderate
amount of debugging information, or to 2 to get more.}

\item{\dots}{optional additional arguments that some (but not all)
\verb{read.adp.*()} functions pass to lower-level functions.}
}
\value{
An \linkS4class{adp} object.
The contents of that object make sense for the particular instrument
type under study, e.g. if the data file contains
NMEA strings, then navigational data will be stored in an item
called \code{nmea} in the \code{data} slot).
}
\description{
Read a Sontek acoustic-Doppler profiler file (see reference 1).
}
\references{
\enumerate{
\item Information about Sontek profilers is available at https://www.sontek.com.
}
}
\seealso{
Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley and Clark Richards
}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.R
\docType{class}
\name{ctd-class}
\alias{ctd-class}
\title{Class to Store CTD (or general hydrographic) Data}
\description{
This class stores hydrographic data such as measured with a CTD (conductivity,
temperature, depth) instrument, or with other systems that produce
similar data.  Data repositories may store conductivity, temperature
and depth, as in the instrument name, but it is also common to store
salinity, temperature and pressure instead (or in addition). For this
reason, \code{ctd} objects are required to hold \code{salinity},
\code{temperature} and \code{pressure} in their \code{data} slot,
with other data being optional. Formulae are available for converting
between variants of these data triplets, e.g. \code{\link[=swSCTp]{swSCTp()}}
can calculate \code{salinity} given \code{conductivity}, \code{temperature}
and \code{pressure}, and these are used by the main functions that
create \code{ctd} objects. For example, if \code{\link[=read.ctd.sbe]{read.ctd.sbe()}}
is used to read a Seabird file that contains only conductivity, temperature
and pressure, then that function will automatically append a data
item to hold salinity.  Since \code{\link[=as.ctd]{as.ctd()}} does the same with
salinity, the result this is that all \code{ctd} objects hold \code{salinity},
\code{temperature} and \code{pressure}, which are henceforth called
the three basic quantities.
}
\details{
Different units and scales are permitted for the three basic quantities, and
most \code{oce} functions check those units and scales before
doing calculations (e.g. of seawater density), because those calculations
demand certain units and scales. The way this is handled is that the
accessor function \code{\link{[[,ctd-method}}] returns values in standardized
form. For example, a \code{ctd} object might hold temperature defined on the
IPTS-68 scale, but e.g. \code{ctd[["temperature"]]} returns a value on the ITS-90
scale. (The conversion is done with \code{\link[=T90fromT68]{T90fromT68()}}.)  Similarly,
pressure may be stored in either dbars or PSI, but e.g. \code{ctd[["pressure"]]}
returns a value in dbars, after dividing by 0.689476 if the value is
stored in PSI. Luckily, there is (as of early 2016) only one salinity scale in
common use in data files, namely PSS-78.
}
\section{Slots}{

\describe{
\item{\code{data}}{As with all \code{oce} objects, the \code{data} slot
for \code{ctd} objects is a
\link{list} containing the main data for the object.
The key items stored in this slot are: \code{salinity}, \code{temperature}, and \code{pressure}, although in many instances there are quite a few additional items.}

\item{\code{metadata}}{As with all \code{oce} objects, the \code{metadata} slot
for \code{ctd} objects is a \link{list} containing
information about the \code{data} or about the object itself.
An example of the former might be the location at which a \code{ctd} measurement was made, stored in \code{longitude} and \code{latitude}, and of the latter might be \code{filename}, the name of the data source.}

\item{\code{processingLog}}{As with all \code{oce} objects, the \code{processingLog} slot
for \code{ctd} objects is a
\link{list} with entries describing the creation and evolution
of the object. The contents are updated by various \code{oce} functions to
keep a record of processing steps.  Object summaries and
\code{\link[=processingLogShow]{processingLogShow()}} both display the log.}
}}

\section{Modifying slot contents}{


Although the \code{[[<-} operator may permit modification of the contents
of \linkS4class{ctd} objects (see \code{\link{[[<-,ctd-method}}),
it is better to use \code{\link[=oceSetData]{oceSetData()}} and \code{\link[=oceSetMetadata]{oceSetMetadata()}},
because those functions save an entry in the \code{processingLog}
that describes the change.
}

\section{Retrieving slot contents}{


The full contents of the \code{data} and \code{metadata} slots of a \linkS4class{ctd}
object may be retrieved in the standard R way using \code{\link[=slot]{slot()}}. For
example \code{slot(o,"data")} returns the \code{data} slot of an object named \code{o},
and similarly \code{slot(o,"metadata")} returns
the \code{metadata} slot.

The slots may also be obtained with the \code{\link{[[,ctd-method}}
operator, as e.g. \code{o[["data"]]} and \code{o[["metadata"]]}, respectively.

The \code{\link{[[,ctd-method}} operator can also
be used to retrieve items from within the \code{data} and \code{metadata} slots.
For example, \code{o[["temperature"]]} can be used to retrieve temperature
from an object containing that quantity. The rule is that a named
quantity is sought first within the object's \code{metadata} slot,
with the \code{data} slot being checked only if \code{metadata} does not
contain the item. This \code{[[} method can also be used to get
certain derived quantities, if the object contains sufficient
information to calculate them. For example, an object that holds
(practical) salinity, temperature and pressure, along with
longitude and latitude, has sufficient information to compute
Absolute Salinity, and so \code{o[["SA"]]} will yield the
calculated Absolute Salinity.

It is also possible to find items more directly, using \code{\link[=oceGetData]{oceGetData()}} and
\code{\link[=oceGetMetadata]{oceGetMetadata()}}, but neither of these functions can
retrieve derived items.
}

\section{Reading/creating \code{ctd} objects}{

A file containing CTD profile data may be read with
\code{\link[=read.ctd]{read.ctd()}}, and a CTD object can also be created with
\code{\link[=as.ctd]{as.ctd()}}.  See \code{\link[=read.ctd]{read.ctd()}} for references on data
formats used in CTD files. Data can also be assembled into
\code{ctd} objects with \code{\link[=as.ctd]{as.ctd()}}.

Statistical summaries are provided by \code{\link[=summary,ctd-method]{summary,ctd-method()}}, while
\code{\link[=show]{show()}} displays an overview.

CTD objects may be plotted with \code{\link[=plot,ctd-method]{plot,ctd-method()}}, which does much of its
work by calling \code{\link[=plotProfile]{plotProfile()}} or \code{\link[=plotTS]{plotTS()}}, both of
which can also be called by the user, to get fine control over the plots.

A CTD profile can be isolated from a larger record with \code{\link[=ctdTrim]{ctdTrim()}},
a task made easier when \code{\link[=plotScan]{plotScan()}} is used to examine the results.
Towyow data can be split up into sets of profiles (ascending or descending)
with \code{\link[=ctdFindProfiles]{ctdFindProfiles()}}.  CTD data may be smoothed and/or cast onto
specified pressure levels with \code{\link[=ctdDecimate]{ctdDecimate()}}.

As with all oce objects, low-level manipulation may be done with
\code{\link[=oceSetData]{oceSetData()}} and \code{\link[=oceSetMetadata]{oceSetMetadata()}}. Additionally,
many of the contents of CTD objects may be altered with the
\code{\link{[[<-,ctd-method}} scheme,
and sufficiently skilled users may even manipulate the contents directly.
}

\section{Data sources}{


Archived CTD (and other) data may be found on servers such as
\enumerate{
\item \url{https://cchdo.ucsd.edu/}
}
}

\examples{

# 1. Create a ctd object with fake data.
a <- as.ctd(salinity=35+1:3/10, temperature=10-1:3/10, pressure=1:3)
summary(a)

# 2. Fix a typo in a station latitude (fake! it's actually okay)
data(ctd)
ctd <- oceSetMetadata(ctd, "latitude", ctd[["latitude"]]-0.001,
                     "fix latitude typo in log book")

}
\seealso{
Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}

Other classes provided by oce: 
\code{\link{adp-class}},
\code{\link{adv-class}},
\code{\link{argo-class}},
\code{\link{bremen-class}},
\code{\link{cm-class}},
\code{\link{coastline-class}},
\code{\link{lisst-class}},
\code{\link{lobo-class}},
\code{\link{met-class}},
\code{\link{oce-class}},
\code{\link{odf-class}},
\code{\link{rsk-class}},
\code{\link{sealevel-class}},
\code{\link{section-class}},
\code{\link{topo-class}},
\code{\link{windrose-class}},
\code{\link{xbt-class}}
}
\author{
Dan Kelley
}
\concept{classes provided by oce}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/gps.R
\name{[[,gps-method}
\alias{[[,gps-method}
\title{Extract Something From a GPS Object}
\usage{
\S4method{[[}{gps}(x, i, j, ...)
}
\arguments{
\item{x}{a \linkS4class{gps} object.}

\item{i}{character value indicating the name of an item to extract.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{ignored.}
}
\description{
Generally, the \code{[[} method lets users extract information from \code{oce}
objects, without having to know the details of the internal storage.  For
many \code{oce} sub-classes, \code{[[} can also return quantities that are computed
from the object's contents.
}
\details{
A two-step process is used to try to find the requested information. First, a
class-specific function is used (see \dQuote{Details of the Specialized
Method}).  If this yields nothing, then a general method is used (see
\dQuote{Details of the General Method}).  If both methods fail, then \code{[[}
returns NULL.

Some understanding of the subclass is required to know what can be retrieved
with \code{[[}. When dealing with an unfamiliar subclass, it can be useful to
first use \code{x[["?"]]} to get a listing of the retrievable items. See
\dQuote{Details of the Specialized Method} for more information.
}
\section{Details of the Specialized Method}{

\itemize{
\item If \code{i} is \code{"?"}, then the return value is a list
containing four items, each of which is a character vector
holding the names of things that can be accessed with \code{[[}.
The \code{data} and \code{metadata} items hold the names of
entries in the object's data and metadata
slots, respectively. The \code{dataDerived}
and \code{metadataDerived} items are each NULL, because
no derived values are defined by \code{gps} objects.
\item If \code{i} is \code{"longitude"} or \code{"latitude"}, then the corresponding
vector is returned.
\item If \code{i} is \code{"filename"} then a filename is returned, if
known (i.e. if the object was created with \code{\link[=read.gps]{read.gps()}} or
with \code{\link[=as.gps]{as.gps()}} with the \code{filename} argument specified).
}
}

\section{Details of the General Method}{


Note: the text of this section is identical for all \code{oce} subclasses, and so
some of what you read here may not be relevant to the class being described
in this help page.

If the specialized method produces no matches, the following generalized
method is applied. As with the specialized method, the procedure hinges first
on the values of \code{i} and, optionally, \code{j}. The work proceeds in steps, by
testing a sequence of possible conditions in sequence.
\enumerate{
\item A check is made as to whether \code{i} names one of the standard \code{oce} slots.
If so, \code{[[} returns the slot contents of that slot.  Thus, \code{x[["metadata"]]}
will retrieve the \code{metadata} slot, while \code{x[["data"]]} and
\code{x[["processingLog"]]} return those slots.
\item If \code{i} is a string ending in the \code{"Unit"}, then the characters preceding
that string are taken to be the name of an item in the data object, and a
list containing the unit is returned (or \code{NULL} if there is no such unit).
This list consists of an item named \code{unit}, which is an \code{\link[=expression]{expression()}}, and
an item named \code{scale}, which is a string describing the measurement scale.
If the string ends in \code{" unit"}, e.g.  \code{x[["temperature unit"]]} (note the
space), then just the expression is returned, and if it ends in \code{" scale"},
then just the scale is returned.
\item If \code{i} is a string ending in \code{"Flag"}, then the corresponding data-quality
flag is returned (or \code{NULL} if there is no such flag).
\item If the object holds hydrographic information (pressure, salinity,
temperature, longitude and latitude) then another set of possibilities
arises.  If \code{i} is \code{"sigmaTheta"}, then the value of \code{\link[=swSigmaTheta]{swSigmaTheta()}} is
called with \code{x} as the sole argument, and the results are returned.
Similarly, \code{\link[=swSigma0]{swSigma0()}} is used if \code{i="sigma0"}, and \code{\link[=swSpice]{swSpice()}} is used if
\code{i="spice"}. Of course, these actions only make sense for objects that
contain the relevant items within their \code{data} slot.
\item After these possibilities are eliminated, the action depends on whether
\code{j} has been provided.  If \code{j} is not provided, or is the string \code{""}, then
\code{i} is sought in the \code{metadata} slot, and then in the \code{data} slot, returning
whichever is found first.  In other words, if \code{j} is not provided, the
\code{metadata} slot takes preference over the \code{data} slot. However, if \code{j} is
provided, then it must be either the string \code{"metadata"} or \code{"data"}, and it
directs where to look.
\item If none of the above-listed conditions holds, then \code{NULL} is returned.
}
}

\seealso{
Other functions that extract parts of oce objects: 
\code{\link{[[,adp-method}},
\code{\link{[[,adv-method}},
\code{\link{[[,amsr-method}},
\code{\link{[[,argo-method}},
\code{\link{[[,bremen-method}},
\code{\link{[[,cm-method}},
\code{\link{[[,coastline-method}},
\code{\link{[[,ctd-method}},
\code{\link{[[,echosounder-method}},
\code{\link{[[,g1sst-method}},
\code{\link{[[,ladp-method}},
\code{\link{[[,landsat-method}},
\code{\link{[[,lisst-method}},
\code{\link{[[,lobo-method}},
\code{\link{[[,met-method}},
\code{\link{[[,oce-method}},
\code{\link{[[,odf-method}},
\code{\link{[[,rsk-method}},
\code{\link{[[,sealevel-method}},
\code{\link{[[,section-method}},
\code{\link{[[,tidem-method}},
\code{\link{[[,topo-method}},
\code{\link{[[,windrose-method}},
\code{\link{[[,xbt-method}},
\code{\link{[[<-,adv-method}}

Other things related to gps data: 
\code{\link{[[<-,gps-method}},
\code{\link{as.gps}()},
\code{\link{gps-class}},
\code{\link{plot,gps-method}},
\code{\link{read.gps}()},
\code{\link{summary,gps-method}}
}
\author{
Dan Kelley
}
\concept{functions that extract parts of oce objects}
\concept{things related to gps data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sun.R
\name{sunAngle}
\alias{sunAngle}
\title{Solar Angle as Function of Space and Time}
\usage{
sunAngle(t, longitude = 0, latitude = 0, useRefraction = FALSE)
}
\arguments{
\item{t}{time, a POSIXt object (converted to timezone \code{"UTC"},
if it is not already in that timezone), a character or numeric value that
corresponds to such a time.}

\item{longitude}{observer longitude in degrees east.}

\item{latitude}{observer latitude in degrees north.}

\item{useRefraction}{boolean, set to \code{TRUE} to apply a correction for
atmospheric refraction.}
}
\value{
A list containing the following:
\itemize{
\item \code{time} the time
\item \code{azimuth}, in degrees eastward of north, from 0 to 360.
\item \code{altitude}, in degrees above the horizon,  ranging from -90 to 90.
\item \code{diameter}, solar diameter, in degrees.
\item \code{distance} to sun, in astronomical units.
\item \code{declination} angle in degrees, computed with \code{\link[=sunDeclinationRightAscension]{sunDeclinationRightAscension()}}.
\item \code{rightAscension} angle in degrees, computed with \code{\link[=sunDeclinationRightAscension]{sunDeclinationRightAscension()}}.
\if{html}{\figure{starCoords.png}{options: width="400"}}
}
}
\description{
This calculates solar angle, based on a NASA-provided Fortran
program, which (according to comments in the code) is in turn
based on "The Astronomical Almanac".
}
\examples{

rise <- as.POSIXct("2011-03-03 06:49:00", tz="UTC") + 4*3600
set <- as.POSIXct("2011-03-03 18:04:00", tz="UTC") + 4*3600
mismatch <- function(lonlat)
{
    sunAngle(rise, lonlat[1], lonlat[2])$altitude^2 + sunAngle(set, lonlat[1], lonlat[2])$altitude^2
}
result <- optim(c(1,1), mismatch)
lon.hfx <- (-63.55274)
lat.hfx <- 44.65
dist <- geodDist(result$par[1], result$par[2], lon.hfx, lat.hfx)
cat(sprintf("Infer Halifax latitude \%.2f and longitude \%.2f; distance mismatch \%.0f km",
            result$par[2], result$par[1], dist))

}
\references{
Regarding \code{declination} and \code{rightAscension}, see
references in the documentation for \code{\link[=sunDeclinationRightAscension]{sunDeclinationRightAscension()}}.
The other items are based on Fortran code retrieved from
the file \code{sunae.f}, downloaded from the ftp site
\code{climate1.gsfc.nasa.gov/wiscombe/Solar_Rad/SunAngles}
on 2009-11-1.  Comments in that code list as references:

Michalsky, J., 1988: The Astronomical Almanac's algorithm for approximate
solar position (1950-2050), Solar Energy 40, 227-235

The Astronomical Almanac, U.S. Gov't Printing Office, Washington, D.C.
(published every year).

The code comments suggest that the appendix in Michalsky (1988) contains
errors, and declares the use of the following formulae in the 1995 version
the Almanac:
\itemize{
\item p. A12: approximation to sunrise/set times
\item p. B61: solar altitude (AKA elevation) and azimuth
\item p. B62: refraction correction
\item p. C24: mean longitude, mean anomaly, ecliptic
longitude, obliquity of ecliptic, right ascension, declination, Earth-Sun
distance, angular diameter of Sun
\item p. L2: Greenwich mean sidereal time (ignoring T^2, T^3 terms)
}

The code lists authors as Dr. Joe Michalsky and Dr. Lee Harrison (State
University of New York), with modifications by Dr. Warren Wiscombe (NASA
Goddard Space Flight Center).
}
\seealso{
The corresponding function for the moon is \code{\link[=moonAngle]{moonAngle()}}.

Other things related to astronomy: 
\code{\link{angle2hms}()},
\code{\link{eclipticalToEquatorial}()},
\code{\link{equatorialToLocalHorizontal}()},
\code{\link{julianCenturyAnomaly}()},
\code{\link{julianDay}()},
\code{\link{moonAngle}()},
\code{\link{siderealTime}()},
\code{\link{sunDeclinationRightAscension}()}
}
\author{
Dan Kelley
}
\concept{things related to astronomy}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{unabbreviateYear}
\alias{unabbreviateYear}
\title{Determine year from various abbreviations}
\usage{
unabbreviateYear(year)
}
\arguments{
\item{year}{a year, or vector of years, possibly abbreviated}
}
\description{
Various data files may contain various abbreviations for years.  For
example, 99 refers to 1999, and 8 refers to 2008.  Sometimes, even 108
refers to 2008 (the idea being that the "zero" year was 1900).  This
function deals with the three cases mentioned.  It will fail if someone
supplies 60, meaning year 2060 as opposed to 1960.
}
\examples{
fullYear <- unabbreviateYear(c(99, 8, 108))
}
\seealso{
Other things related to time: 
\code{\link{ctimeToSeconds}()},
\code{\link{julianCenturyAnomaly}()},
\code{\link{julianDay}()},
\code{\link{numberAsHMS}()},
\code{\link{numberAsPOSIXct}()},
\code{\link{secondsToCtime}()}
}
\author{
Dan Kelley
}
\concept{things related to time}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/AllClass.R
\name{initializeFlagSchemeInternal}
\alias{initializeFlagSchemeInternal}
\title{Establish a data-quality scheme for a oce object}
\usage{
initializeFlagSchemeInternal(
  object,
  name = NULL,
  mapping = NULL,
  default = NULL,
  update = NULL,
  debug = 0
)
}
\arguments{
\item{object}{An oce object.}

\item{name}{a character value naming the scheme. If this refers
to a pre-defined scheme, then \code{mapping} must not be provided,
because doing so would contradict the pre-defined scheme, defeating
its purpose of providing concreteness and clarity.}

\item{mapping}{a list of named items describing the mapping from
flag meaning to flag numerical value, e.g \code{list(good=1, bad=2)}
might be used for a hypothetical class.}

\item{default}{an integer vector of flag values that are not considered
to be good. If this is not provided, but if \code{name} is \code{"argo"},
\code{"BODC"},
\code{"DFO"},
\code{"WHP bottle"}, or
\code{"WHP CTD"}, then a conservative value will be set automatically,
equal to the list of flag values that designate bad or questionable data.
For example, for \code{name="WHP CTD"}, the setting will be
\code{c(1,3,4,5,6,7,9)}, leaving only value \code{2}, which corresponds
with "acceptable" in the notation used for that flag scheme.}

\item{update}{a logical value indicating whether the scheme provided is
to update an existing scheme.  The default value, \code{FALSE}, prevents such
an attempt to alter an existing flag scheme, if one is already embedded
in \code{object}.}

\item{debug}{an integer set to 0 for quiet action or to 1 for some debugging.}
}
\value{
An object with the \code{metadata} slot containing \code{flagScheme}.
}
\description{
This function stores add an item named \code{flagScheme}
to the \code{metadata} slot of an object inheriting from
\linkS4class{oce}. This is a list containing two
items: \code{name} and \code{mapping}, as provided in the
function arguments.
The purpose is both to document a flag scheme
and to make it so that \code{\link[=initializeFlags]{initializeFlags()}},
\code{\link[=setFlags]{setFlags()}} and \code{\link[=handleFlags]{handleFlags()}}
can specify flags by
name, as opposed to number. This is a generic function,
that may be specialized to the class of \code{object}
(see \dQuote{Details}).
}
\details{
The following pre-defined schemes are available (note that the
names are simplified from the phrases used in defining
documentation):
\itemize{
\item \code{name="argo"} defaults \code{mapping} to
OLD (prior to June 10, 2020)
}\preformatted{list(not_assessed=0, passed_all_tests=1, probably_good=2,
     probably_bad=3, bad=4, averaged=7,
     interpolated=8, missing=9)
}

NEW (after June 10, 2020)\preformatted{list(not_assessed=0, passed_all_tests=1, probably_good=2,
     probably_bad=3, bad=4, changed=5, not_used_6=6, not_used_7=7,
     estimated=8, missing=9)
}

See reference 1 for a deeper explanation of the meanings of these codes.
\itemize{
\item \code{name="BODC"} defaults \code{mapping} to
}\preformatted{list(no_quality_control=0, good=1, probably_good=2,
     probably_bad=3, bad=4, changed=5,
     below_detection=6, in_excess=7, interpolated=8,
     missing=9)
}

See reference 2 for a deeper explanation of the meanings of these codes,
and note that codes \code{A} and \code{Q} are not provided in
oce.
\itemize{
\item \code{name="DFO"} defaults \code{mapping} to
}\preformatted{list(no_quality_control=0, appears_correct=1, appears_inconsistent=2,
     doubtful=3, erroneous=4, changed=5,
     qc_by_originator=8, missing=9)
}

See reference 3 for a deeper explanation of the meanings of these codes.
\itemize{
\item \code{name="WHP bottle"} defaults \code{mapping} to
}\preformatted{list(no_information=1, no_problems_noted=2, leaking=3,
     did_not_trip=4, not_reported=5, discrepency=6,
     unknown_problem=7, did_not_trip=8, no_sample=9)
}

See reference 4 for a deeper explanation of the meanings of these codes.
\itemize{
\item \code{name="WHP CTD"} defaults \code{mapping} to
}\preformatted{list(not_calibrated=1, acceptable=2, questionable=3,
    bad=4, not_reported=5, interpolated=6,
    despiked=7, missing=9)
}

See reference 4 for a deeper explanation of the meanings of these codes.
}
\references{
\enumerate{
\item The codes for \code{"argo"} are derived from information in Table 4.1
of Wong, Annie, Robert Keeley, Thierry Carval, and Argo Data Management Team
(8 January 2020), "Argo Quality Control Manual for CTD and Trajectory Data, Version 3.3,"
available at \url{https://archimer.ifremer.fr/doc/00228/33951/}
as of June 2020.
\item The codes for \code{"BODC"} are defined at
http://seadatanet.maris2.nl/v_bodc_vocab_v2/browse.asp?order=conceptid&formname=search&screen=0&lib=l20
\item The codes for \code{"DFO"} are defined at
http://www.dfo-mpo.gc.ca/science/data-donnees/code/list/014-eng.html
\item The codes for \code{"WHP CTD"} and \code{"WHP bottle"} are defined at
https://www.nodc.noaa.gov/woce/woce_v3/wocedata_1/whp/exchange/exchange_format_desc.htm
}
}
\seealso{
Other functions relating to data-quality flags: 
\code{\link{defaultFlags}()},
\code{\link{handleFlags,adp-method}},
\code{\link{handleFlags,argo-method}},
\code{\link{handleFlags,ctd-method}},
\code{\link{handleFlags,oce-method}},
\code{\link{handleFlags,section-method}},
\code{\link{handleFlags}()},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{initializeFlagScheme,oce-method}},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{initializeFlagScheme}()},
\code{\link{initializeFlags,adp-method}},
\code{\link{initializeFlags,oce-method}},
\code{\link{initializeFlagsInternal}()},
\code{\link{initializeFlags}()},
\code{\link{setFlags,adp-method}},
\code{\link{setFlags,ctd-method}},
\code{\link{setFlags,oce-method}},
\code{\link{setFlags}()}

Other things related to oce data: 
\code{\link{initializeFlagScheme,oce-method}},
\code{\link{initializeFlagScheme}()}
}
\concept{functions relating to data-quality flags}
\concept{things related to oce data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{oceConvolve}
\alias{oceConvolve}
\alias{oce.convolve}
\title{Convolve two time series}
\usage{
oceConvolve(x, f, end = 2)
}
\arguments{
\item{x}{a numerical vector of observations.}

\item{f}{a numerical vector of filter coefficients.}

\item{end}{a flag that controls how to handle the points of the \code{x}
series that have indices less than the length of \code{f}.  If \code{end=0},
the values are set to 0.  If \code{end=1}, the original x values are used
there.  If \code{end=2}, that fraction of the \code{f} values that overlap
with \code{x} are used.}
}
\value{
A vector of the convolution output.
}
\description{
Convolve two time series, using a backward-looking method.
This function provides a straightforward convolution, which may be useful to
those who prefer not to use \code{\link[=convolve]{convolve()}} and \code{filter} in the
\code{stats} package.
}
\examples{
library(oce)
t <- 0:1027
n <- length(t)
signal <- ifelse(sin(t * 2 * pi / 128) > 0, 1, 0)
tau <- 10
filter <- exp(-seq(5*tau, 0) / tau)
filter <- filter / sum(filter)
observation <- oce.convolve(signal, filter)
plot(t, signal, type='l')
lines(t, observation, lty='dotted')
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.R
\name{setFlags,ctd-method}
\alias{setFlags,ctd-method}
\title{Set data-quality flags within a ctd object}
\usage{
\S4method{setFlags}{ctd}(
  object,
  name = NULL,
  i = NULL,
  value = NULL,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{object}{An oce object.}

\item{name}{Character string indicating the name of the variable to be flagged. If
this variable is not contained in the object's \code{data} slot, an error is reported.}

\item{i}{Indication of where to insert the flags; see \dQuote{Description} for
general rules and \dQuote{Details} for rules for \linkS4class{ctd}
objects.}

\item{value}{The value to be inserted in the flag.}

\item{debug}{Integer set to 0 for quiet action or to 1 for some debugging.}
}
\value{
An object with flags set as indicated.
}
\description{
This function changes specified entries in the data-quality
flags of a ctd object, which are stored within
a list named \code{flags} that resides in the \code{metadata}
slot. If the object already has a flag set up for \code{name},
then only the specified entries are altered. If not, the flag
entry is first created and its entries set to \code{default},
after which the entries specified by \code{i}
are changed to \code{value}.

The specification is made with \code{i}, the form of which
is determined by the data item in question. Generally,
the rules are as follows:
\enumerate{
\item If the data item is a vector, then \code{i} must be (a)
an integer vector specifying indices to be set to \code{value},
(b) a logical vector of length matching the data item, with
\code{TRUE} meaning to set the flag to \code{value}, or (c)
a function that takes an \code{oce} object as its single
argument, and returns a vector in either of the forms
just described.
\item If the data item is an array, then \code{i} must be
(a) a data frame of integers whose rows specify spots to change
(where the number of columns matches the number of dimensions
of the data item), (b) a logical array that has dimension equal to
that of the data item, or (c) a function that takes an \code{oce}
object as its single input and returns such a data frame or array.
}

See \dQuote{Details} for the particular case of \linkS4class{ctd} objects.
}
\details{
Since all the entries in the \code{data} slot of ctd objects are vectors, \code{i} must be a vector (either logical as in Example 1 or integer as in Example 2), or a function taking a \code{ctd} object and returning such a vector (see \dQuote{Indexing rules}).
}
\examples{
library(oce)
# Example 1: Range-check salinity
data(ctdRaw)
# Salinity and temperature range checks
qc <- ctdRaw
# Initialize flags to 2, meaning good data in the default
# scheme for handleFlags(ctd).
qc <- initializeFlags(qc, "salinity", 2)
qc <- initializeFlags(qc, "temperature", 2)
# Flag bad salinities as 4
oddS <- with(qc[["data"]], salinity < 25 | 40 < salinity)
qc <- setFlags(qc, name="salinity", i=oddS, value=4)
# Flag bad temperatures as 4
oddT <- with(qc[["data"]], temperature < -2 | 40 < temperature)
qc <- setFlags(qc, name="temperature", i=oddT, value=4)
# Compare results in TS space
par(mfrow=c(2, 1))
plotTS(ctdRaw)
plotTS(handleFlags(qc, flags=c(1, 3:9)))

# Example 2: Interactive flag assignment based on TS plot, using
# WHP scheme to define 'acceptable' and 'bad' codes
\dontrun{
options(eos="gsw")
data(ctd)
qc <- ctd
qc <- initializeFlagScheme(qc, "WHP CTD")
qc <- initializeFlags(qc, "salinity", 2)
Sspan <- diff(range(qc[["SA"]]))
Tspan <- diff(range(qc[["CT"]]))
n <- length(qc[["SA"]])
par(mfrow=c(1, 1))
plotTS(qc, type="o")
message("Click on bad points; quit by clicking to right of plot")
for (i in seq_len(n)) {
    xy <- locator(1)
    if (xy$x > par("usr")[2])
        break
    i <- which.min(abs(qc[["SA"]] - xy$x)/Sspan + abs(qc[["CT"]] - xy$y)/Tspan)
    qc <- setFlags(qc, "salinity", i=i, value=4)
    qc <- handleFlags(qc, flags=list(salinity=4))
    plotTS(qc, type="o")
}
}

}
\seealso{
Other functions relating to data-quality flags: 
\code{\link{defaultFlags}()},
\code{\link{handleFlags,adp-method}},
\code{\link{handleFlags,argo-method}},
\code{\link{handleFlags,ctd-method}},
\code{\link{handleFlags,oce-method}},
\code{\link{handleFlags,section-method}},
\code{\link{handleFlags}()},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{initializeFlagScheme,oce-method}},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{initializeFlagSchemeInternal}()},
\code{\link{initializeFlagScheme}()},
\code{\link{initializeFlags,adp-method}},
\code{\link{initializeFlags,oce-method}},
\code{\link{initializeFlagsInternal}()},
\code{\link{initializeFlags}()},
\code{\link{setFlags,adp-method}},
\code{\link{setFlags,oce-method}},
\code{\link{setFlags}()}

Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}
}
\concept{functions relating to data-quality flags}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{ctimeToSeconds}
\alias{ctimeToSeconds}
\title{Interpret a character string as a time interval}
\usage{
ctimeToSeconds(ctime)
}
\arguments{
\item{ctime}{a character string (see \sQuote{Details}.}
}
\value{
A numeric value, the number of seconds represented by the string.
}
\description{
Interpret a character string as a time interval
Strings are of the form MM:SS or HH:MM:SS.
}
\examples{
library(oce)
cat("10      = ", ctimeToSeconds("10"), "s\n", sep="")
cat("01:04   = ", ctimeToSeconds("01:04"), "s\n", sep="")
cat("1:00:00 = ", ctimeToSeconds("1:00:00"), "s\n", sep="")
}
\seealso{
See \code{\link[=secondsToCtime]{secondsToCtime()}}, the inverse of this.

Other things related to time: 
\code{\link{julianCenturyAnomaly}()},
\code{\link{julianDay}()},
\code{\link{numberAsHMS}()},
\code{\link{numberAsPOSIXct}()},
\code{\link{secondsToCtime}()},
\code{\link{unabbreviateYear}()}
}
\author{
Dan Kelley
}
\concept{things related to time}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{read.woa}
\alias{read.woa}
\title{Read a World Ocean Atlas NetCDF File}
\usage{
read.woa(file, name, positive = FALSE)
}
\arguments{
\item{file}{character string naming the file}

\item{name}{of variable to extract. If not provided, an
error message is issued that lists the names of data in the file.}

\item{positive}{logical value indicating whether \code{longitude} should
be converted to be in the range from 0 to 360, with \code{name}
being shuffled accordingly. This is set to \code{FALSE} by default,
because the usual oce convention is for longitude to range between -180
to +180.}
}
\value{
A list containing vectors \code{longitude}, \code{latitude},
\code{depth}, and an array with the specified name. If \code{positive}
is true, then \code{longitude} will be converted to range from 0
to 360, and the array will be shuffled accordingly.
}
\description{
Read a World Ocean Atlas NetCDF File
}
\examples{
\dontrun{
## Mean SST at 5-degree spatial resolution
tmn <- read.woa("/data/woa13/woa13_decav_t00_5dv2.nc", "t_mn")
imagep(tmn$longitude, tmn$latitude, tmn$t_mn[,,1], zlab="SST")
}
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/amsr.R
\name{summary,amsr-method}
\alias{summary,amsr-method}
\title{Summarize an amsr Object}
\usage{
\S4method{summary}{amsr}(object, ...)
}
\arguments{
\item{object}{an \linkS4class{amsr} object.}

\item{...}{ignored.}
}
\description{
Although the data are stored in \code{\link[=raw]{raw()}} form, the summary
presents results in physical units.
}
\seealso{
Other things related to amsr data: 
\code{\link{[[,amsr-method}},
\code{\link{[[<-,amsr-method}},
\code{\link{amsr-class}},
\code{\link{amsr}},
\code{\link{composite,amsr-method}},
\code{\link{download.amsr}()},
\code{\link{plot,amsr-method}},
\code{\link{read.amsr}()},
\code{\link{subset,amsr-method}}
}
\author{
Dan Kelley
}
\concept{things related to amsr data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{binApply2D}
\alias{binApply2D}
\title{Apply a function to matrix data}
\usage{
binApply2D(x, y, f, xbreaks, ybreaks, FUN, ...)
}
\arguments{
\item{x}{a vector of numerical values.}

\item{y}{a vector of numerical values.}

\item{f}{a vector of data to which the elements of \code{FUN} may be
supplied}

\item{xbreaks}{values of \code{x} at the boundaries between the
bins; calculated using \code{\link[=pretty]{pretty()}} if not supplied.}

\item{ybreaks}{as \code{xbreaks}, but for \code{y}.}

\item{FUN}{univariate function that is applied to the \code{f} data within
any given bin}

\item{\dots}{arguments to pass to the function \code{FUN}}
}
\value{
A list with the following elements: the breaks in \code{x} and
\code{y} (i.e. \code{xbreaks} and \code{ybreaks}), the break mid-points
(i.e. \code{xmids} and \code{ymids}), and a matrix containing the
result of applying \code{FUN()} to the \code{f} values, as
subsetted by these breaks.
}
\description{
The function \code{FUN} is applied to \code{f} in bins specified by
\code{xbreaks} and \code{ybreaks}.  (If \code{FUN} is \code{\link[=mean]{mean()}},
consider using \code{\link[=binMean2D]{binMean2D()}} instead, since it should be faster.)
}
\examples{
library(oce)
\donttest{
## secchi depths in lat and lon bins
if (requireNamespace("ocedata", quietly=TRUE)) {
    data(secchi, package="ocedata")
    ## Note that zlim is provided to the colormap(), to prevent a few
    ## points from setting a very wide scale.
    cm <- colormap(z=secchi$depth, col=oceColorsViridis, zlim=c(0, 15))
    par(mar=c(2, 2, 2, 2))
    drawPalette(colormap=cm, zlab="Secchi Depth")
    data(coastlineWorld)
    mapPlot(coastlineWorld, longitudelim=c(-5, 20), latitudelim=c(50, 66),
      grid=5, col='gray', projection="+proj=lcc +lat_1=50 +lat_2=65")
    bc <- binApply2D(secchi$longitude, secchi$latitude,
                     pretty(secchi$longitude, 80),
                     pretty(secchi$latitude, 40),
                     f=secchi$depth, FUN=mean)
    mapImage(bc$xmids, bc$ymids, bc$result, zlim=cm$zlim, col=cm$zcol)
    mapPolygon(coastlineWorld, col="gray")
}
}

}
\seealso{
Other bin-related functions: 
\code{\link{binApply1D}()},
\code{\link{binAverage}()},
\code{\link{binCount1D}()},
\code{\link{binCount2D}()},
\code{\link{binMean1D}()},
\code{\link{binMean2D}()}
}
\author{
Dan Kelley
}
\concept{bin-related functions}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/coastline.R
\name{download.coastline}
\alias{download.coastline}
\title{Download a coastline File}
\usage{
download.coastline(
  resolution,
  item = "coastline",
  destdir = ".",
  destfile,
  server = "naturalearth",
  debug = getOption("oceDebug")
)
}
\arguments{
\item{resolution}{A character value specifying the desired resolution. The permitted
choices are \code{"10m"} (for 1:10M resolution, the most detailed),
\code{"50m"} (for 1:50M resolution)
and \code{"110m"} (for 1:110M resolution). If \code{resolution} is not supplied,
\code{"50m"} will be used.}

\item{item}{A character value indicating the quantity to be downloaded.
This is normally one of \code{"coastline"}, \code{"land"}, \code{"ocean"},
\code{"rivers_lakes_centerlines"}, or \code{"lakes"}, but the NaturalEarth
server has other types, and advanced users can discover their names by inspecting
the URLs of links on the NaturalEarth site, and use them for \code{item}.
If \code{item} is not supplied, it defaults to \code{"coastline"}.}

\item{destdir}{Optional string indicating the directory in which to store
downloaded files. If not supplied, \code{"."} is used, i.e. the data file
is stored in the present working directory.}

\item{destfile}{Optional string indicating the name of the file.
If not supplied, the file name is constructed from the other
parameters of the function call, so subsequent calls with the same
parameters will yield the same result, thus providing the key
to the caching scheme.}

\item{server}{A character value specifying the server that is to supply
the data. At the moment, the only permitted value is \code{"naturalearth"},
which is the default if \code{server} is not supplied.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\value{
A character value indicating the filename of the result; if
there is a problem of any kind, the result will be the empty
string.
}
\description{
Constructs a query to the NaturalEarth server (see reference 1) to download coastline
data (or lake data, river data, etc) in any of three resolutions.
}
\examples{
\dontrun{
library(oce)
# User must create directory ~/data/coastline first.
# As of September 2016, the downloaded file, named
# "ne_50m_coastline.zip", occupies 443K bytes.
filename <- download.coastline(destdir="~/data/coastline")
coastline <- read.coastline(filename)
plot(coastline)
}

}
\references{
\enumerate{
\item The NaturalEarth server is at \url{https://www.naturalearthdata.com}
}
}
\seealso{
The work is done with \code{\link[utils:download.file]{utils::download.file()}}.

Other functions that download files: 
\code{\link{download.amsr}()},
\code{\link{download.met}()},
\code{\link{download.topo}()}

Other things related to coastline data: 
\code{\link{[[,coastline-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{as.coastline}()},
\code{\link{coastline-class}},
\code{\link{coastlineBest}()},
\code{\link{coastlineCut}()},
\code{\link{coastlineWorld}},
\code{\link{plot,coastline-method}},
\code{\link{read.coastline.openstreetmap}()},
\code{\link{read.coastline.shapefile}()},
\code{\link{subset,coastline-method}},
\code{\link{summary,coastline-method}}
}
\author{
Dan Kelley
}
\concept{functions that download files}
\concept{things related to coastline data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/processingLog.R
\name{processingLogShow}
\alias{processingLogShow}
\title{Show the processing log of an oce object}
\usage{
processingLogShow(x)
}
\arguments{
\item{x}{an \linkS4class{oce} object.}
}
\description{
Show the processing log of an oce object
}
\seealso{
Other things related to processing logs: 
\code{\link{processingLog<-}()},
\code{\link{processingLogAppend}()},
\code{\link{processingLogItem}()}
}
\concept{things related to processing logs}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swSigmaT}
\alias{swSigmaT}
\title{Seawater quasi-potential density anomaly}
\usage{
swSigmaT(
  salinity,
  temperature = NULL,
  pressure = NULL,
  longitude = NULL,
  latitude = NULL,
  eos = getOption("oceEOS", default = "gsw")
)
}
\arguments{
\item{salinity}{either practical salinity (in which case \code{temperature}
and \code{pressure} must be provided) \emph{or} an \code{oce} object, in
which case \code{salinity}, \code{temperature} (in the ITS-90 scale; see
next item), etc. are inferred from the object, ignoring the
other parameters, if they are supplied.}

\item{temperature}{\emph{in-situ} temperature (\eqn{^\circ}{deg}C), defined
on the ITS-90 scale.  This scale is used by GSW-style calculation (as
requested by setting \code{eos="gsw"}), and is the value contained within
\code{ctd} objects (and probably most other objects created with data
acquired in the past decade or two). Since the UNESCO-style calculation is
based on IPTS-68, the temperature is converted within the present function,
using \code{\link[=T68fromT90]{T68fromT90()}}.}

\item{pressure}{pressure (dbar)}

\item{longitude}{longitude of observation (only used if \code{eos="gsw"};
see \sQuote{Details}).}

\item{latitude}{latitude of observation (only used if \code{eos="gsw"}; see
\sQuote{Details}).}

\item{eos}{equation of state, either \code{"unesco"} (references 1 and 2)
or \code{"gsw"} (references 3 and 4).}
}
\value{
Quasi-potential density anomaly (kg/m\eqn{^3}{^3}), defined as the
density calculated with pressure set to zero.
}
\description{
Compute \eqn{\sigma_t}{sigma-t}, a rough estimate of potential density of
seawater, minus 1000 kg/m\eqn{^3}{^3}.
}
\details{
If the first argument is an \linkS4class{oce} object, then salinity, etc., are
extracted from it, and used for the calculation.
}
\examples{
swSigmaT(35, 13, 1000, longitude=300, latitude=30, eos="gsw") # 26.39623
swSigmaT(35, T90fromT68(13), 1000, eos="unesco") # 26.39354

}
\references{
See citations provided in the \code{\link[=swRho]{swRho()}} documentation.
}
\seealso{
Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/bremen.R
\name{plot,bremen-method}
\alias{plot,bremen-method}
\alias{plot.bremen}
\title{Plot a bremen Object}
\usage{
\S4method{plot}{bremen}(x, type, ...)
}
\arguments{
\item{x}{a \linkS4class{bremen} object.}

\item{type}{Optional string indicating the type to which \code{x} should be
coerced before plotting. The choices are \code{ctd} and \code{ladp}.}

\item{...}{Other arguments, passed to plotting functions.}
}
\description{
Plot a \linkS4class{bremen} object.
If the first argument seems to be a CTD dataset, this uses \code{\link[=plot,ctd-method]{plot,ctd-method()}};
otherwise, that argument is assumed to be a \linkS4class{ladp} object, and a two-panel
plot is created with \code{\link[=plot,ladp-method]{plot,ladp-method()}} to show velocity variation with
pressure.
}
\seealso{
Other functions that plot oce data: 
\code{\link{download.amsr}()},
\code{\link{plot,adp-method}},
\code{\link{plot,adv-method}},
\code{\link{plot,amsr-method}},
\code{\link{plot,argo-method}},
\code{\link{plot,cm-method}},
\code{\link{plot,coastline-method}},
\code{\link{plot,ctd-method}},
\code{\link{plot,gps-method}},
\code{\link{plot,ladp-method}},
\code{\link{plot,landsat-method}},
\code{\link{plot,lisst-method}},
\code{\link{plot,lobo-method}},
\code{\link{plot,met-method}},
\code{\link{plot,odf-method}},
\code{\link{plot,rsk-method}},
\code{\link{plot,satellite-method}},
\code{\link{plot,sealevel-method}},
\code{\link{plot,section-method}},
\code{\link{plot,tidem-method}},
\code{\link{plot,topo-method}},
\code{\link{plot,windrose-method}},
\code{\link{plot,xbt-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{tidem-class}}

Other things related to bremen data: 
\code{\link{[[,bremen-method}},
\code{\link{[[<-,bremen-method}},
\code{\link{bremen-class}},
\code{\link{read.bremen}()},
\code{\link{summary,bremen-method}}
}
\author{
Dan Kelley
}
\concept{functions that plot oce data}
\concept{things related to bremen data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sealevel.R
\name{plot,sealevel-method}
\alias{plot,sealevel-method}
\alias{plot.sealevel}
\title{Plot a sealevel Object}
\usage{
\S4method{plot}{sealevel}(
  x,
  which = 1:3,
  drawTimeRange = getOption("oceDrawTimeRange"),
  mgp = getOption("oceMgp"),
  mar = c(mgp[1] + 0.5, mgp[1] + 1.5, mgp[2] + 1, mgp[2] + 3/4),
  marginsAsImage = FALSE,
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{x}{a \linkS4class{sealevel} object.}

\item{which}{a numerical or string vector indicating desired plot types,
with possibilities 1 or \code{"all"} for a time-series of all the elevations, 2 or
\code{"month"} for a time-series of just the first month, 3 or
\code{"spectrum"} for a power spectrum (truncated to frequencies below 0.1
cycles per hour, or 4 or \code{"cumulativespectrum"} for a cumulative
integral of the power spectrum.}

\item{drawTimeRange}{boolean that applies to panels with time as the
horizontal axis, indicating whether to draw the time range in the top-left
margin of the plot.}

\item{mgp}{3-element numerical vector to use for \code{\link{par}}\code{("mgp")}, and also
for \code{\link{par}}\code{("mar")}, computed from this.  The default is tighter than the R
default, in order to use more space for the data and less for the axes.}

\item{mar}{value to be used with \code{\link{par}}\code{("mar")}.}

\item{marginsAsImage}{boolean, \code{TRUE} to put a wide margin to the right
of time-series plots, matching the space used up by a palette in an
\code{\link[=imagep]{imagep()}} plot.}

\item{debug}{a flag that turns on debugging, if it exceeds 0.}

\item{\dots}{optional arguments passed to plotting functions.}
}
\value{
None.
}
\description{
Creates a plot for a sea-level dataset, in one of two varieties.  Depending
on the length of \code{which}, either a single-panel or multi-panel plot is
drawn.  If there is just one panel, then the value of \code{par} used in
\verb{plot,sealevel-method} is retained upon exit, making it convenient to add to
the plot.  For multi-panel plots, \code{par} is returned to the value it had
before the call.
}
\section{Historical Note}{

Until 2020-02-06, sea-level plots had the mean value removed, and indicated
with a tick mark and margin note on the right-hand side of the plot.
This behaviour was confusing. The change did not go through the usual
deprecation process, because the margin-note behaviour had not
been documented.
}

\examples{
library(oce)
data(sealevel)
## local Halifax time is UTC + 4h
juan <- as.POSIXct("2003-09-29 00:15:00", tz="UTC")+4*3600
plot(sealevel, which=1, xlim=juan+86400*c(-7, 7))
abline(v=juan, col='red')

}
\references{
The example refers to Hurricane Juan, which caused a great deal
of damage to Halifax in 2003.  Since this was in the era of the digital
photo, a casual web search will uncover some spectacular images of damage,
from both wind and storm surge.
Landfall, within 30km of this sealevel
gauge, was between 00:10 and 00:20 Halifax local time on Monday, Sept 29,
2003.
}
\seealso{
The documentation for the \linkS4class{sealevel} class explains the
structure of sealevel objects, and also outlines the other functions dealing
with them.

Other functions that plot oce data: 
\code{\link{download.amsr}()},
\code{\link{plot,adp-method}},
\code{\link{plot,adv-method}},
\code{\link{plot,amsr-method}},
\code{\link{plot,argo-method}},
\code{\link{plot,bremen-method}},
\code{\link{plot,cm-method}},
\code{\link{plot,coastline-method}},
\code{\link{plot,ctd-method}},
\code{\link{plot,gps-method}},
\code{\link{plot,ladp-method}},
\code{\link{plot,landsat-method}},
\code{\link{plot,lisst-method}},
\code{\link{plot,lobo-method}},
\code{\link{plot,met-method}},
\code{\link{plot,odf-method}},
\code{\link{plot,rsk-method}},
\code{\link{plot,satellite-method}},
\code{\link{plot,section-method}},
\code{\link{plot,tidem-method}},
\code{\link{plot,topo-method}},
\code{\link{plot,windrose-method}},
\code{\link{plot,xbt-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{tidem-class}}

Other things related to sealevel data: 
\code{\link{[[,sealevel-method}},
\code{\link{[[<-,sealevel-method}},
\code{\link{as.sealevel}()},
\code{\link{read.sealevel}()},
\code{\link{sealevel-class}},
\code{\link{sealevelTuktoyaktuk}},
\code{\link{sealevel}},
\code{\link{subset,sealevel-method}},
\code{\link{summary,sealevel-method}}
}
\author{
Dan Kelley
}
\concept{functions that plot oce data}
\concept{things related to sealevel data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{oceEdit}
\alias{oceEdit}
\alias{oce.edit}
\title{Edit an Oce Object}
\usage{
oceEdit(
  x,
  item,
  value,
  action,
  reason = "",
  person = "",
  debug = getOption("oceDebug")
)
}
\arguments{
\item{x}{an \linkS4class{oce} object.  The exact action of \code{\link[=oceEdit]{oceEdit()}} depends
on the sub-class of \code{x}.}

\item{item}{if supplied, a character string naming an item in the object's
\code{metadata} or \code{data} slot, the former being checked first.
An exception is if \code{item} starts with \code{"data@"} or
\code{"metadata@"}, in which case the named slot is updated with a changed
value of the contents of \code{item} after the \code{@} character.}

\item{value}{new value for \code{item}, if both supplied.}

\item{action}{optional character string containing R code to carry out some
action on the object.}

\item{reason}{character string giving the reason for the change.}

\item{person}{character string giving the name of person making the change.}

\item{debug}{an integer that specifies a level of debugging, with 0 or less
indicating no debugging, and 1 or more indicating debugging.}
}
\value{
A \linkS4class{oce} object, altered
appropriately, and with a log item indicating the nature of the alteration.
}
\description{
Edit an element of an oce object, inserting a note in the processing
log of the returned object.
}
\details{
There are several ways to use this function.
\enumerate{
\item If both an \code{item} and \code{value} are supplied, then
either the object's metadata or data slot may be altered. There are
two ways in which this can be done.
\itemize{
\item Case 1A. If the \code{item} string does not contain an
\code{@} character, then the \code{metadata} slot is examined
for an entry named \code{item}, and that is modified if so.
Alternatively, if \code{item} is found in \code{metadata}, then
that value is modified. However, if \code{item} is not found in
either \code{metadata} or \code{data}, then an error is reported
(see 1B for how to add something that does not yet exist).
\item Case 1B. If the \code{item} string contains
the \code{@} character, then the text to the left of that character
must be either \code{"metadata"} or \code{"data"}, and it names the slot
in which the change is done. In contrast with case 1A, this will
\emph{create} a new item, if it is not already in existence.
}
\item If \code{item} and \code{value} are not supplied, then \code{action} must
be supplied.  This is a character string specifying some action to be
performed on the object, e.g. a manipulation of a column.  The action must
refer to the object as \code{x}; see Examples.
}

In any case, a log entry is stored in the object, to document the change.
Indeed, this is the main benefit to using this function, instead of altering
the object directly.  The log entry will be most useful if it contains a
brief note on the \code{reason} for the change, and the name of the
\code{person} doing the work.
}
\examples{
library(oce)
data(ctd)
ctd2 <- oceEdit(ctd, item="latitude", value=47.8879,
               reason="illustration", person="Dan Kelley")
ctd3 <- oceEdit(ctd,action="x@data$pressure<-x@data$pressure-1")
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/map.R
\name{mapPolygon}
\alias{mapPolygon}
\title{Add a Polygon to a Map}
\usage{
mapPolygon(
  longitude,
  latitude,
  density = NULL,
  angle = 45,
  border = NULL,
  col = NA,
  lty = par("lty"),
  ...,
  fillOddEven = FALSE
)
}
\arguments{
\item{longitude}{numeric vector of longitudes of points defining the polygon,
to be plotted, or an object from
which both longitude and latitude can be inferred (e.g. a coastline file, or
the return value from \code{\link[=mapLocator]{mapLocator()}}), in which case the \code{latitude}
argument are ignored.}

\item{latitude}{numeric vector of latitudes of points to be plotted (ignored
if both longitude and latitude can be determined from the first argument).}

\item{density, angle, border, col, lty, ..., fillOddEven}{handled as
\code{\link[=polygon]{polygon()}} handles the same arguments.}
}
\description{
\code{mapPolygon} adds a polygon to an existing map.
}
\examples{
\donttest{
library(oce)
data(coastlineWorld)
data(topoWorld)

## Bathymetry near southeastern Canada
par(mfrow=c(1,1), mar=c(2,2,1,1))
cm <- colormap(zlim=c(-5000, 0), col=oceColorsGebco)
drawPalette(colormap=cm)
lonlim<- c(-60,-50)
latlim<- c(40,60)
mapPlot(coastlineWorld, longitudelim=lonlim,
latitudelim=latlim, projection="+proj=merc", grid=FALSE)
mapImage(topoWorld, colormap=cm)
mapPolygon(coastlineWorld[['longitude']], coastlineWorld[['latitude']], col="lightgray")
}

}
\seealso{
A map must first have been created with \code{\link[=mapPlot]{mapPlot()}}.

Other functions related to maps: 
\code{\link{formatPosition}()},
\code{\link{lonlat2map}()},
\code{\link{lonlat2utm}()},
\code{\link{map2lonlat}()},
\code{\link{mapArrows}()},
\code{\link{mapAxis}()},
\code{\link{mapContour}()},
\code{\link{mapCoordinateSystem}()},
\code{\link{mapDirectionField}()},
\code{\link{mapGrid}()},
\code{\link{mapImage}()},
\code{\link{mapLines}()},
\code{\link{mapLocator}()},
\code{\link{mapLongitudeLatitudeXY}()},
\code{\link{mapPlot}()},
\code{\link{mapPoints}()},
\code{\link{mapScalebar}()},
\code{\link{mapText}()},
\code{\link{mapTissot}()},
\code{\link{oceCRS}()},
\code{\link{shiftLongitude}()},
\code{\link{usrLonLat}()},
\code{\link{utm2lonlat}()}
}
\author{
Dan Kelley
}
\concept{functions related to maps}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{binAverage}
\alias{binAverage}
\title{Bin-average a vector y, based on x values}
\usage{
binAverage(x, y, xmin, xmax, xinc)
}
\arguments{
\item{x}{a vector of numerical values.}

\item{y}{a vector of numerical values.}

\item{xmin}{x value at the lower limit of first bin; the minimum \code{x}
will be used if this is not provided.}

\item{xmax}{x value at the upper limit of last bin; the maximum \code{x}
will be used if this is not provided.}

\item{xinc}{width of bins, in terms of x value; 1/10th of \code{xmax-xmin}
will be used if this is not provided.}
}
\value{
A list with two elements: \code{x}, the mid-points of the bins, and
\code{y}, the average \code{y} value in the bins.
}
\description{
The \code{y} vector is averaged in bins defined for \code{x}.  Missing
values in \code{y} are ignored.
}
\examples{
library(oce)
## A. fake linear data
x <- seq(0, 100, 1)
y <- 1 + 2 * x
plot(x, y, pch=1)
ba <- binAverage(x, y)
points(ba$x, ba$y, pch=3, col='red', cex=3)

## B. fake quadratic data
y <- 1 + x ^2
plot(x, y, pch=1)
ba <- binAverage(x, y)
points(ba$x, ba$y, pch=3, col='red', cex=3)

## C. natural data
data(co2)
plot(co2)
avg <- binAverage(time(co2), co2, 1950, 2000, 2)
points(avg$x, avg$y, col='red')

}
\seealso{
Other bin-related functions: 
\code{\link{binApply1D}()},
\code{\link{binApply2D}()},
\code{\link{binCount1D}()},
\code{\link{binCount2D}()},
\code{\link{binMean1D}()},
\code{\link{binMean2D}()}
}
\author{
Dan Kelley
}
\concept{bin-related functions}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/rsk.R
\docType{class}
\name{rsk-class}
\alias{rsk-class}
\title{Class to Store Rsk Data}
\description{
This class stores Ruskin data, from RBR (see reference 1).
}
\details{
A \linkS4class{rsk} object may be read with \code{\link[=read.rsk]{read.rsk()}} or created with
\code{\link[=as.rsk]{as.rsk()}}.  Plots can be made with \code{\link[=plot,rsk-method]{plot,rsk-method()}}, while
\code{\link[=summary,rsk-method]{summary,rsk-method()}} produces statistical summaries and \code{show}
produces overviews.   If atmospheric pressure has not been removed from the
data, the functions \code{\link[=rskPatm]{rskPatm()}} may provide guidance as to its value;
however, this last function is no equal to decent record-keeping at sea.
}
\section{Slots}{

\describe{
\item{\code{data}}{As with all \code{oce} objects, the \code{data} slot
for \code{rsk} objects is a
\link{list} containing the main data for the object.
{}}

\item{\code{metadata}}{As with all \code{oce} objects, the \code{metadata} slot
for \code{rsk} objects is a \link{list} containing
information about the \code{data} or about the object itself.
{}}

\item{\code{processingLog}}{As with all \code{oce} objects, the \code{processingLog} slot
for \code{rsk} objects is a
\link{list} with entries describing the creation and evolution
of the object. The contents are updated by various \code{oce} functions to
keep a record of processing steps.  Object summaries and
\code{\link[=processingLogShow]{processingLogShow()}} both display the log.}
}}

\section{Modifying slot contents}{


Although the \code{[[<-} operator may permit modification of the contents
of \linkS4class{rsk} objects (see \code{\link{[[<-,rsk-method}}),
it is better to use \code{\link[=oceSetData]{oceSetData()}} and \code{\link[=oceSetMetadata]{oceSetMetadata()}},
because those functions save an entry in the \code{processingLog}
that describes the change.
}

\section{Retrieving slot contents}{


The full contents of the \code{data} and \code{metadata} slots of a \linkS4class{rsk}
object may be retrieved in the standard R way using \code{\link[=slot]{slot()}}. For
example \code{slot(o,"data")} returns the \code{data} slot of an object named \code{o},
and similarly \code{slot(o,"metadata")} returns
the \code{metadata} slot.

The slots may also be obtained with the \code{\link{[[,rsk-method}}
operator, as e.g. \code{o[["data"]]} and \code{o[["metadata"]]}, respectively.

The \code{\link{[[,rsk-method}} operator can also
be used to retrieve items from within the \code{data} and \code{metadata} slots.
For example, \code{o[["temperature"]]} can be used to retrieve temperature
from an object containing that quantity. The rule is that a named
quantity is sought first within the object's \code{metadata} slot,
with the \code{data} slot being checked only if \code{metadata} does not
contain the item. This \code{[[} method can also be used to get
certain derived quantities, if the object contains sufficient
information to calculate them. For example, an object that holds
(practical) salinity, temperature and pressure, along with
longitude and latitude, has sufficient information to compute
Absolute Salinity, and so \code{o[["SA"]]} will yield the
calculated Absolute Salinity.

It is also possible to find items more directly, using \code{\link[=oceGetData]{oceGetData()}} and
\code{\link[=oceGetMetadata]{oceGetMetadata()}}, but neither of these functions can
retrieve derived items.
}

\references{
\enumerate{
\item RBR website (https://www.rbr-global.com/products)
}
}
\seealso{
Other classes provided by oce: 
\code{\link{adp-class}},
\code{\link{adv-class}},
\code{\link{argo-class}},
\code{\link{bremen-class}},
\code{\link{cm-class}},
\code{\link{coastline-class}},
\code{\link{ctd-class}},
\code{\link{lisst-class}},
\code{\link{lobo-class}},
\code{\link{met-class}},
\code{\link{oce-class}},
\code{\link{odf-class}},
\code{\link{sealevel-class}},
\code{\link{section-class}},
\code{\link{topo-class}},
\code{\link{windrose-class}},
\code{\link{xbt-class}}

Other things related to rsk data: 
\code{\link{[[,rsk-method}},
\code{\link{[[<-,rsk-method}},
\code{\link{as.rsk}()},
\code{\link{plot,rsk-method}},
\code{\link{read.rsk}()},
\code{\link{rskPatm}()},
\code{\link{rskToc}()},
\code{\link{rsk}},
\code{\link{subset,rsk-method}},
\code{\link{summary,rsk-method}}
}
\author{
Dan Kelley and Clark Richards
}
\concept{classes provided by oce}
\concept{things related to rsk data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/tides.R
\name{as.tidem}
\alias{as.tidem}
\title{Create tidem object from fitted harmonic data}
\usage{
as.tidem(tRef, latitude, name, amplitude, phase, debug = getOption("oceDebug"))
}
\arguments{
\item{tRef}{a POSIXt value indicating the mean time of the
observations used to develop the harmonic model. This is rounded
to the nearest hour in \code{\link[=as.tidem]{as.tidem()}}, to match \code{\link[=tidem]{tidem()}}.}

\item{latitude}{Numerical value indicating the latitude of the
observations that were used to create the harmonic model. This
is needed for nodal-correction procedures carried out
by \code{\link[=tidemVuf]{tidemVuf()}}.}

\item{name}{character vector holding names of constituents.}

\item{amplitude}{Numeric vector of constituent amplitudes.}

\item{phase}{Numeric vector of constituent Greenwich phases.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\value{
An object of \linkS4class{tidem}, with only minimal
contents.
}
\description{
This function is intended to provide a bridge to
\code{\link[=predict.tidem]{predict.tidem()}}, enabling tidal predictions based
on published tables of harmonic fits.
}
\details{
Note that only constituent names known to \code{\link[=tidem]{tidem()}} are handled.
The permitted names are those listed in Foreman (1978), and
tabulated with
\preformatted{
data(tidedata)
data.frame(name=tidedata$const$name, freq=tidedata$const$freq)
}
Warnings are issued for any constituent name that is not in this list; as
of the late summer of 2019, the only example seen in practice is
\code{M1}, which according to Wikipedia (2019) has frequency 0.0402557, which
is very close to that of \code{NO1}, i.e. 0.04026859, perhaps explaining
why Foreman (1978) did not handle this constituent. A warning is
issued if this or any other unhandled constituent is provided
in the \code{name} argument to \code{as.tidem()}.
}
\section{Known issues}{

There are two known differences between \code{\link[=tidem]{tidem()}} and the Matlab
\code{T_TIDE} package, as listed in references 3 and 4. Work on these
issues is planned for the summer of 2020.
}

\examples{
# Simulate a tide table with output from tidem().
data(sealevelTuktoyaktuk)
# 'm0' is model fitted by tidem()
m0 <- tidem(sealevelTuktoyaktuk)
p0 <- predict(m0, sealevelTuktoyaktuk[["time"]])
m1 <- as.tidem(mean(sealevelTuktoyaktuk[["time"]]), sealevelTuktoyaktuk[["latitude"]],
               m0[["name"]], m0[["amplitude"]], m0[["phase"]])
# Test agreement with tidem() result, by comparing predicted sealevels.
p1 <- predict(m1, sealevelTuktoyaktuk[["time"]])
stopifnot(max(abs(p1 - p0), na.rm=TRUE) < 1e-10)
# Simplified harmonic model, using large constituents
# > m0[["name"]][which(m[["amplitude"]]>0.05)]
# [1] "Z0"  "MM"  "MSF" "O1"  "K1"  "OO1" "N2"  "M2"  "S2"
h <- "
name  amplitude      phase
  Z0 1.98061875   0.000000
  MM 0.21213065 263.344739
 MSF 0.15605629 133.795004
  O1 0.07641438  74.233130
  K1 0.13473817  81.093134
 OO1 0.05309911 235.749693
  N2 0.08377108  44.521462
  M2 0.49041340  77.703594
  S2 0.22023705 137.475767"
coef <- read.table(text=h, header=TRUE)
m2 <- as.tidem(mean(sealevelTuktoyaktuk[["time"]]),
               sealevelTuktoyaktuk[["latitude"]],
               coef$name, coef$amplitude, coef$phase)
p2 <- predict(m2, sealevelTuktoyaktuk[["time"]])
stopifnot(max(abs(p2 - p0), na.rm=TRUE) < 1)
par(mfrow=c(3, 1))
oce.plot.ts(sealevelTuktoyaktuk[["time"]], p0)
ylim <- par("usr")[3:4] # to match scales in other panels
oce.plot.ts(sealevelTuktoyaktuk[["time"]], p1, ylim=ylim)
oce.plot.ts(sealevelTuktoyaktuk[["time"]], p2, ylim=ylim)

}
\references{
\enumerate{
\item Foreman, M. G. G., 1978. Manual for Tidal Currents Analysis and Prediction.
Pacific Marine Science Report. British Columbia, Canada: Institute of Ocean
Sciences, Patricia Bay.
\item Wikipedia, "Theory of Tides." https://en.wikipedia.org/wiki/Theory_of_tides
Downloaded Aug 17, 2019.
\item Github issue 1653: tidem() and t_tide do not produce identical results
https://github.com/dankelley/oce/issues/1653
\item Github issue 1654: predict(tidem()) uses all constituents, unlike T_TIDE
https://github.com/dankelley/oce/issues/1654
}
}
\seealso{
Other things related to tides: 
\code{\link{[[,tidem-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{plot,tidem-method}},
\code{\link{predict.tidem}()},
\code{\link{summary,tidem-method}},
\code{\link{tidalCurrent}},
\code{\link{tidedata}},
\code{\link{tidem-class}},
\code{\link{tidemAstron}()},
\code{\link{tidemVuf}()},
\code{\link{tidem}},
\code{\link{webtide}()}
}
\concept{things related to tides}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{ungrid}
\alias{ungrid}
\title{Extract (x, y, z) from (x, y, grid)}
\usage{
ungrid(x, y, grid)
}
\arguments{
\item{x}{a vector holding the x coordinates of grid.}

\item{y}{a vector holding the y coordinates of grid.}

\item{grid}{a matrix holding the grid.}
}
\value{
A list containing three vectors: \code{x}, the grid x values,
\code{y}, the grid y values, and \code{grid}, the grid values.
}
\description{
Extract the grid points from a grid, returning columns.
This is useful for e.g. gridding large datasets, in which the first step
might be to use \code{\link[=binMean2D]{binMean2D()}}, followed by
\code{\link[=interpBarnes]{interpBarnes()}}.
}
\examples{
library(oce)
data(wind)
u <- interpBarnes(wind$x, wind$y, wind$z)
contour(u$xg, u$yg, u$zg)
U <- ungrid(u$xg, u$yg, u$zg)
points(U$x, U$y, col=oce.colorsViridis(100)[rescale(U$grid, rlow=1, rhigh=100)], pch=20)
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/section.R
\name{[[,section-method}
\alias{[[,section-method}
\title{Extract Something From a Section Object}
\usage{
\S4method{[[}{section}(x, i, j, ...)
}
\arguments{
\item{x}{a \linkS4class{section} object.}

\item{i}{character value indicating the name of an item to extract.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{ignored.}
}
\description{
Generally, the \code{[[} method lets users extract information from \code{oce}
objects, without having to know the details of the internal storage.  For
many \code{oce} sub-classes, \code{[[} can also return quantities that are computed
from the object's contents.
}
\details{
A two-step process is used to try to find the requested information. First, a
class-specific function is used (see \dQuote{Details of the Specialized
Method}).  If this yields nothing, then a general method is used (see
\dQuote{Details of the General Method}).  If both methods fail, then \code{[[}
returns NULL.

Some understanding of the subclass is required to know what can be retrieved
with \code{[[}. When dealing with an unfamiliar subclass, it can be useful to
first use \code{x[["?"]]} to get a listing of the retrievable items. See
\dQuote{Details of the Specialized Method} for more information.
}
\section{Details of the Specialized Method}{


There are several possibilities, depending on the nature of \code{i}.
\itemize{
\item If \code{i} is \code{"?"}, then the return value is a list containing four items,
each of which is a character vector holding the names of things that can be
accessed with \code{[[}. This list is compiled by examining all the stations in
the object, and reporting an entry if it is found in any one of them. The
\code{data} and \code{metadata} items hold the names of entries in the object's data
and metadata slots, respectively. The \code{dataDerived} and \code{metadataDerived}
items hold data-like and metadata-like things that can be derived from these.
\item If \code{i} is \code{"station"}, then \code{[[} will return a \code{\link[=list]{list()}} of \linkS4class{ctd}
objects holding the station data. If \code{j} is also given, it specifies a
station (or set of stations) to be returned. if \code{j} contains just a single
value, then that station is returned, but otherwise a list is returned. If
\code{j} is an integer, then the stations are specified by index, but if it is
character, then stations are specified by the names stored within their
metadata. (Missing stations yield \code{NULL} in the return value.)
\item If \code{i} is \code{"station ID"}, then the IDs of the stations in the
section are returned.
\item If \code{i} is \code{"dynamic height"}, then an estimate of dynamic
height is returned, as calculated with \code{\link{swDynamicHeight}}\code{(x)}.
\item If \code{i} is \code{"distance"}, then the distance along the section is
returned, using \code{\link[=geodDist]{geodDist()}}.
\item If \code{i} is \code{"depth"}, then a vector containing the depths
of the stations is returned.
\item If \code{i} is \code{"z"}, then a vector containing the z
coordinates is returned.
\item If \code{i} is \code{"theta"} or \code{"potential temperature"}, then
the potential temperatures of all the stations are returned in one
vector.  Similarly, \code{"spice"} returns the property known
as spice, using \code{\link[=swSpice]{swSpice()}}.
\item If \code{i} is a string ending with \code{"Flag"}, then the characters
prior to that ending are taken to be the name of a variable contained
within the stations in the section. If this flag is available in
the first station of the section, then the flag values are looked
up for every station.
}

If \code{j} is \code{"byStation"}, then a list is returned, with
one (unnamed) item per station.
}

\section{Details of the General Method}{


Note: the text of this section is identical for all \code{oce} subclasses, and so
some of what you read here may not be relevant to the class being described
in this help page.

If the specialized method produces no matches, the following generalized
method is applied. As with the specialized method, the procedure hinges first
on the values of \code{i} and, optionally, \code{j}. The work proceeds in steps, by
testing a sequence of possible conditions in sequence.
\enumerate{
\item A check is made as to whether \code{i} names one of the standard \code{oce} slots.
If so, \code{[[} returns the slot contents of that slot.  Thus, \code{x[["metadata"]]}
will retrieve the \code{metadata} slot, while \code{x[["data"]]} and
\code{x[["processingLog"]]} return those slots.
\item If \code{i} is a string ending in the \code{"Unit"}, then the characters preceding
that string are taken to be the name of an item in the data object, and a
list containing the unit is returned (or \code{NULL} if there is no such unit).
This list consists of an item named \code{unit}, which is an \code{\link[=expression]{expression()}}, and
an item named \code{scale}, which is a string describing the measurement scale.
If the string ends in \code{" unit"}, e.g.  \code{x[["temperature unit"]]} (note the
space), then just the expression is returned, and if it ends in \code{" scale"},
then just the scale is returned.
\item If \code{i} is a string ending in \code{"Flag"}, then the corresponding data-quality
flag is returned (or \code{NULL} if there is no such flag).
\item If the object holds hydrographic information (pressure, salinity,
temperature, longitude and latitude) then another set of possibilities
arises.  If \code{i} is \code{"sigmaTheta"}, then the value of \code{\link[=swSigmaTheta]{swSigmaTheta()}} is
called with \code{x} as the sole argument, and the results are returned.
Similarly, \code{\link[=swSigma0]{swSigma0()}} is used if \code{i="sigma0"}, and \code{\link[=swSpice]{swSpice()}} is used if
\code{i="spice"}. Of course, these actions only make sense for objects that
contain the relevant items within their \code{data} slot.
\item After these possibilities are eliminated, the action depends on whether
\code{j} has been provided.  If \code{j} is not provided, or is the string \code{""}, then
\code{i} is sought in the \code{metadata} slot, and then in the \code{data} slot, returning
whichever is found first.  In other words, if \code{j} is not provided, the
\code{metadata} slot takes preference over the \code{data} slot. However, if \code{j} is
provided, then it must be either the string \code{"metadata"} or \code{"data"}, and it
directs where to look.
\item If none of the above-listed conditions holds, then \code{NULL} is returned.
}
}

\examples{
data(section)
length(section[["latitude"]])
length(section[["latitude", "byStation"]])
# Vector of all salinities, for all stations
Sv <- section[["salinity"]]
# List of salinities, grouped by station
Sl <- section[["salinity", "byStation"]]
# First station salinities
Sl[[1]]

}
\seealso{
Other functions that extract parts of oce objects: 
\code{\link{[[,adp-method}},
\code{\link{[[,adv-method}},
\code{\link{[[,amsr-method}},
\code{\link{[[,argo-method}},
\code{\link{[[,bremen-method}},
\code{\link{[[,cm-method}},
\code{\link{[[,coastline-method}},
\code{\link{[[,ctd-method}},
\code{\link{[[,echosounder-method}},
\code{\link{[[,g1sst-method}},
\code{\link{[[,gps-method}},
\code{\link{[[,ladp-method}},
\code{\link{[[,landsat-method}},
\code{\link{[[,lisst-method}},
\code{\link{[[,lobo-method}},
\code{\link{[[,met-method}},
\code{\link{[[,oce-method}},
\code{\link{[[,odf-method}},
\code{\link{[[,rsk-method}},
\code{\link{[[,sealevel-method}},
\code{\link{[[,tidem-method}},
\code{\link{[[,topo-method}},
\code{\link{[[,windrose-method}},
\code{\link{[[,xbt-method}},
\code{\link{[[<-,adv-method}}

Other things related to section data: 
\code{\link{[[<-,section-method}},
\code{\link{as.section}()},
\code{\link{handleFlags,section-method}},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{plot,section-method}},
\code{\link{read.section}()},
\code{\link{section-class}},
\code{\link{sectionAddStation}()},
\code{\link{sectionGrid}()},
\code{\link{sectionSmooth}()},
\code{\link{sectionSort}()},
\code{\link{section}},
\code{\link{subset,section-method}},
\code{\link{summary,section-method}}
}
\author{
Dan Kelley
}
\concept{functions that extract parts of oce objects}
\concept{things related to section data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/tides.R
\name{tidem}
\alias{tidem}
\title{Fit a Tidal Model to a Timeseries}
\usage{
tidem(
  t,
  x,
  constituents,
  infer = NULL,
  latitude = NULL,
  rc = 1,
  regress = lm,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{t}{A \code{sealevel} object created with
\code{\link[=read.sealevel]{read.sealevel()}} or \code{\link[=as.sealevel]{as.sealevel()}}, or a vector of
times. In the former case, time is part of the object, so \code{t} may not
be given here.  In the latter case, \code{tidem} needs a way to determine
time, so \code{t} must be given.}

\item{x}{an optional numerical vector holding something that varies with
time.  This is ignored if \code{t} is a \linkS4class{sealevel} object,
in which case it is inferred as \code{t[["elevation"]]}.}

\item{constituents}{an optional character vector holding the names
of tidal constituents to which the fit is done (see \dQuote{Details}
and \dQuote{Constituent Naming Convention}.)}

\item{infer}{a list of constituents to be inferred from
fitted constituents according to the method outlined
in Section 2.3.4 of Foreman (1978).
If \code{infer} is \code{NULL}, the default, then
no such inferences are made. Otherwise, some constituents
are computed based on other constituents, instead of being
determined by regression at the proper frequency.
If provided, \code{infer} must be a list containing
four elements:
\code{name}, a vector of strings naming the constituents to be
inferred; \code{from}, a vector of strings naming the fitted
constituents used as the sources for those inferences (these
source constituents are added to the regression list, if they
are not already there);
\code{amp}, a numerical vector of factors to be applied to the
source amplitudes; and \code{phase}, a numerical vector of angles,
in degrees, to be subtracted from the source phases. For example,
Following Foreman (1998), if any of the \code{name} items
have already been computed, then the suggested inference is ignored,
and the already-computed values are used.
\preformatted{
infer=list(name=c("P1","K2"),
           from=c("K1", "S2"),
           amp=c(0.33093, 0.27215),
           phase=c(-7.07, -22.4)
}
means that the amplitude of \code{P1} will be set as 0.33093 times the calculated amplitude
of \code{K1}, and that the \code{P1} phase will be set to the \code{K1} phase,
minus an offset of \code{-7.07} degrees.
(This example is used in the Foreman (1978) discussion of a
Fortran analysis code and also in Pawlowicz et al. (2002) discussion
of the T_TIDE Matlab code.
Rounded to the 0.1mm resolution of values reported in Foreman (1978)
and Pawlowicz et al. (2002),
the \code{tidem} results have root-mean-square amplitude difference
to Foreman's (1978) Appendix 7.3 of 0.06mm; by comparison,
the results in Table 1 of Pawlowicz et al. (2002) agree with Foreman's
results to RMS difference 0.04mm.)}

\item{latitude}{if provided, the latitude of the observations.  If not
provided, \code{tidem} will try to infer this from \code{sl}.}

\item{rc}{the value of the coefficient in the Rayleigh criterion.}

\item{regress}{function to be used for regression, by default
\code{\link[=lm]{lm()}}, but could be for example \code{rlm} from the
\code{MASS} package.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\value{
An object of \linkS4class{tidem}, consisting of
\item{const}{constituent number, e.g. 1 for \code{Z0}, 1 for \code{SA},
etc.} \item{model}{the regression model} \item{name}{a vector of constituent
names, in non-subscript format, e.g. "\code{M2}".} \item{frequency}{a vector
of constituent frequencies, in inverse hours.} \item{amplitude}{a vector of
fitted constituent amplitudes, in metres.} \item{phase}{a vector of fitted
constituent phase.  NOTE: The definition of phase is likely to change as
this function evolves.  For now, it is phase with respect to the first data
sample.} \item{p}{a vector containing a sort of p value for each
constituent.  This is calculated as the average of the p values for the
sine() and cosine() portions used in fitting; whether it makes any sense is
an open question.}
}
\description{
The fit is done in terms of sine and cosine components at the indicated
tidal frequencies, with the amplitude and phase being calculated from the
resultant coefficients on the sine and cosine terms.
}
\details{
The tidal constituents to be used in the analysis are specified as follows;
see \dQuote{Constituent Naming Convention}.
\enumerate{
\item If \code{constituents} is not provided, then the constituent
list will be made up of the 69 constituents designated by Foreman as "standard".
These include astronomical frequencies and some shallow-water frequencies,
and are as follows: \code{c("Z0", "SA", "SSA", "MSM", "MM", "MSF", "MF", "ALP1", "2Q1", "SIG1", "Q1", "RHO1", "O1", "TAU1", "BET1", "NO1", "CHI1", "PI1", "P1", "S1", "K1", "PSI1", "PHI1", "THE1", "J1", "SO1", "OO1", "UPS1", "OQ2", "EPS2", "2N2", "MU2", "N2", "NU2", "GAM2", "H1", "M2", "H2", "MKS2", "LDA2", "L2", "T2", "S2", "R2", "K2", "MSN2", "ETA2", "MO3", "M3", "SO3", "MK3", "SK3", "MN4", "M4", "SN4", "MS4", "MK4", "S4", "SK4", "2MK5", "2SK5", "2MN6", "M6", "2MS6", "2MK6", "2SM6", "MSK6", "3MK7", "M8")}.
\item If the first item in \code{constituents} is the string
\code{"standard"}, then a provisional list is set up as in Case 1, and then
the (optional) rest of the elements of \code{constituents} are examined, in
order.  Each of these constituents is based on the name of a tidal
constituent in the Foreman (1978) notation.  (To get the list, execute
\code{data(tidedata)} and then execute \code{cat(tideData$name)}.)  Each
named constituent is added to the existing list, if it is not already there.
But, if the constituent is preceded by a minus sign, then it is removed
from the list (if it is already there).  Thus, for example,
\code{constituents=c("standard", "-M2", "ST32")} would remove the M2
constituent and add the ST32 constituent.
\item If the first item is not \code{"standard"}, then the list of
constituents is processed as in Case 2, but without starting with the
standard list. As an example, \code{constituents=c("K1", "M2")} would fit
for just the K1 and M2 components. (It would be strange to use a minus sign
to remove items from the list, but the function allows that.)
}

In each of the above cases, the list is reordered in frequency prior to the
analysis, so that the results of \code{\link[=summary,tidem-method]{summary,tidem-method()}} will be in a
familiar form.

Once the constituent list is determined, \code{tidem} prunes the elements of
the list by using the Rayleigh criterion, according to which two
constituents of frequencies \eqn{f_1}{f1} and \eqn{f_2}{f2} cannot be
resolved unless the time series spans a time interval of at least
\eqn{rc/(f_1-f_2)}{rc/(f1-f2)}.

Finally, \code{tidem} looks in the remaining constituent list to check
that the application of the Rayleigh criterion has not removed any of the
constituents specified directly in the \code{constituents} argument. If
any are found to have been removed, then they are added back. This last
step was added on 2017-12-27, to make \code{tidem} behave the same
way as the Foreman (1978) code, as illustrated in his
Appendices 7.2 and 7.3. (As an aside, his Appendix 7.3 has some errors,
e.g. the frequency for the 2SK5 constituent is listed there (p58) as
0.20844743, but it is listed as 0.2084474129 in his Appendix 7.1 (p41).
For this reason, the frequency comparison is relaxed to a \code{tol}
value of \code{1e-7} in a portion of the oce test suite
(see \code{tests/testthat/test_tidem.R} in the source).

A specific example may be of help in understanding the removal of unresolvable
constituents. For example, the \code{data(sealevel)} dataset is of length
6718 hours, and this is too short to resolve the full list of constituents,
with the conventional (and, really, necessary) limit of \code{rc=1}.
From Table 1 of Foreman (1978), this timeseries is too short to resolve the
\code{SA} constituent, so that \code{SA} will not be in the resultant.
Similarly, Table 2 of Foreman (1978) dictates the removal of
\code{PI1}, \code{S1} and \code{PSI1} from the list. And, finally,
Table 3 of Foreman (1978) dictates the removal of
\code{H1}, \code{H2}, \code{T2} and \code{R2}, and since that document
suggests that \code{GAM2} be subsumed into \code{H1},
then if \code{H1} is already being deleted, then
\code{GAM2} will also be deleted.

A summary of constituents may be found with:
\preformatted{
data(tidedata)
print(tidedata$const)
}
}
\section{Bugs}{


\enumerate{
\item This function is not fully developed yet, and both the
form of the call and the results of the calculation may change.

\item The reported \code{p} value may make no sense at all, and it might be
removed in a future version of this function. Perhaps a significance level
should be presented, as in the software developed by both Foreman and
Pawlowicz.

}
}

\section{Constituent Naming Convention}{


\code{tidem} uses constituent names that follow the convention
set by Foreman (1978). This convention is slightly different
from that used in the T-TIDE package of Pawlowicz et al.
(2002), with Foreman's \code{UPS1} and \code{M8} becoming
\code{UPSI} and \code{MS} in T-TIDE. To permit the use of either notation,
\code{\link[=tidem]{tidem()}} uses \code{\link[=tidemConstituentNameFix]{tidemConstituentNameFix()}} to
convert from T-TIDE names to the
Foreman names, issuing warnings when doing so.
}

\section{Agreement with \code{T_TIDE} results}{


The \code{tidem} amplitude and phase results, obtained with
\preformatted{
tidem(sealevelTuktoyaktuk, constituents=c("standard", "M10"),
     infer=list(name=c("P1", "K2"),
                from=c("K1", "S2"),
                amp=c(0.33093, 0.27215),
                phase=c(-7.07, -22.40))),
}
are identical the \code{T_TIDE} values listed in
Table 1 of Pawlowicz et al. (2002),
after rounding amplitude and phase to 4 and 2 digits past
the decimal place, to match the format of the table.
}

\examples{
library(oce)
# The demonstration time series from Foreman (1978),
# also used in T_TIDE (Pawlowicz, 2002).
data(sealevelTuktoyaktuk)
tide <- tidem(sealevelTuktoyaktuk)
summary(tide)

# AIC analysis
extractAIC(tide[["model"]])

# Fake data at M2
library(oce)
data("tidedata")
M2 <- with(tidedata$const, freq[name=="M2"])
t <- seq(0, 10*86400, 3600)
eta <- sin(M2 * t * 2 * pi / 3600)
sl <- as.sealevel(eta)
m <- tidem(sl)
summary(m)

}
\references{
Foreman, M. G. G., 1978. Manual for Tidal Currents Analysis and Prediction.
Pacific Marine Science Report. British Columbia, Canada: Institute of Ocean
Sciences, Patricia Bay.

Foreman, M. G. G., Neufeld, E. T., 1991.  Harmonic tidal analyses of long
time series.  International Hydrographic Review, 68 (1), 95-108.

Leffler, K. E. and D. A. Jay, 2009.  Enhancing tidal harmonic analysis:
Robust (hybrid) solutions.  Continental Shelf Research, 29(1):78-88.

Pawlowicz, Rich, Bob Beardsley, and Steve Lentz, 2002.  Classical tidal
harmonic analysis including error estimates in MATLAB using \code{T_TIDE}.
Computers and Geosciences, 28, 929-937.
}
\seealso{
Other things related to tides: 
\code{\link{[[,tidem-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{as.tidem}()},
\code{\link{plot,tidem-method}},
\code{\link{predict.tidem}()},
\code{\link{summary,tidem-method}},
\code{\link{tidalCurrent}},
\code{\link{tidedata}},
\code{\link{tidem-class}},
\code{\link{tidemAstron}()},
\code{\link{tidemVuf}()},
\code{\link{webtide}()}
}
\author{
Dan Kelley
}
\concept{things related to tides}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/rsk.R
\docType{data}
\name{rsk}
\alias{rsk}
\title{Sample Rsk Dataset}
\description{
A sample \code{rsk} object derived from a Concerto CTD manufactured by RBR Ltd.
}
\details{
The data were obtained September 2015, off the west coast
of Greenland, by Matt Rutherford and Nicole Trenholm of the
Ocean Research Project, in collaboration with RBR and with the
NASA Oceans Melting Greenland project.
}
\examples{
library(oce)
data(rsk)
## The object doesn't "know" it is CTD until told so
plot(rsk)
plot(as.ctd(rsk))

}
\references{
\url{https://rbr-global.com/}
}
\seealso{
Other datasets provided with oce: 
\code{\link{adp}},
\code{\link{adv}},
\code{\link{amsr}},
\code{\link{argo}},
\code{\link{cm}},
\code{\link{coastlineWorld}},
\code{\link{ctdRaw}},
\code{\link{ctd}},
\code{\link{echosounder}},
\code{\link{landsat}},
\code{\link{lisst}},
\code{\link{lobo}},
\code{\link{met}},
\code{\link{ocecolors}},
\code{\link{sealevelTuktoyaktuk}},
\code{\link{sealevel}},
\code{\link{section}},
\code{\link{topoWorld}},
\code{\link{wind}},
\code{\link{xbt}}

Other things related to rsk data: 
\code{\link{[[,rsk-method}},
\code{\link{[[<-,rsk-method}},
\code{\link{as.rsk}()},
\code{\link{plot,rsk-method}},
\code{\link{read.rsk}()},
\code{\link{rsk-class}},
\code{\link{rskPatm}()},
\code{\link{rskToc}()},
\code{\link{subset,rsk-method}},
\code{\link{summary,rsk-method}}
}
\concept{datasets provided with oce}
\concept{things related to rsk data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/section.R
\name{sectionGrid}
\alias{sectionGrid}
\title{Grid a Section in Pressure Space}
\usage{
sectionGrid(
  section,
  p,
  method = "approx",
  trim = TRUE,
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{section}{A \code{section} object containing the section to be gridded.}

\item{p}{Optional indication of the pressure levels to which interpolation
should be done.  If this is not supplied, the pressure levels will be
calculated based on the typical spacing in the ctd profiles stored within
\code{section}.  If \code{p="levitus"}, then pressures will be set to be those
of the Levitus atlas, given by \code{\link[=standardDepths]{standardDepths()}}.
If \code{p} is a single numerical value,
it is taken as the number of subdivisions to use in a call to \code{\link[=seq]{seq()}}
that has range from 0 to the maximum pressure in \code{section}.  Finally, if a
vector numerical values is provided, perhaps. constructed with \code{\link[=seq]{seq()}}
or \code{\link{standardDepths}}\code{(5)} (as in the examples),
then it is used as is, after trimming any values that exceed the maximum
pressure in the station data stored within \code{section}.}

\item{method}{The method to use to decimate data within the stations; see
\code{\link[=ctdDecimate]{ctdDecimate()}}, which is used for the decimation.}

\item{trim}{Logical value indicating whether to trim gridded pressures
to the range of the data in \code{section}.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}

\item{...}{Optional arguments to be supplied to \code{\link[=ctdDecimate]{ctdDecimate()}},
e.g. \code{rule} controls extrapolation beyond the observed pressure range,
in the case where \code{method} equals \code{"approx"}.}
}
\value{
A \linkS4class{section} object that contains stations whose
pressure values match identically, and that has all flags set to \code{NA}.
}
\description{
Grid a section, by interpolating to fixed pressure levels.  The
\code{"approx"}, \code{"boxcar"} and \code{"lm"} methods are described in the
documentation for \code{\link[=ctdDecimate]{ctdDecimate()}}, which is used to do this
processing.
}
\details{
The default \code{"approx"} method is best for bottle data, the
\code{"boxcar"} is best for ctd data, and the \code{"lm"} method is probably
too slow to recommend for exploratory work, in which it is common to do trials
with a variety of \code{"p"} values.

The stations in the returned value have flags with names that match those
of the corresponding stations in the original \code{section}, but the values
of these flags are all set to \code{NA}. This recognizes that it makes
no sense to grid flag values, but that there is merit in initializing
a flag system, for possible use in later processing steps.
}
\examples{
# Gulf Stream
library(oce)
data(section)
GS <- subset(section, 109<=stationId&stationId<=129)
GSg <- sectionGrid(GS, p=seq(0, 5000, 100))
plot(GSg, map.xlim=c(-80,-60))
# Show effects of various depth schemes
par(mfrow=c(3, 1))
default <- sectionGrid(GS)
approxML <- sectionGrid(GS, method="approxML")
standardDepths5 <- sectionGrid(GS, p=standardDepths(5))
plot(default, which="temperature", ztype="image", ylim=c(200,0))
mtext("default sectionGrid()")
plot(approxML, which="temperature", ztype="image", ylim=c(200,0))
mtext("sectionGrid(..., method=\"approxML\")")
plot(standardDepths5, which="temperature", ztype="image", ylim=c(200,0))
mtext("sectionGrid(..., p=standardDepths(5))")

}
\seealso{
Other things related to section data: 
\code{\link{[[,section-method}},
\code{\link{[[<-,section-method}},
\code{\link{as.section}()},
\code{\link{handleFlags,section-method}},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{plot,section-method}},
\code{\link{read.section}()},
\code{\link{section-class}},
\code{\link{sectionAddStation}()},
\code{\link{sectionSmooth}()},
\code{\link{sectionSort}()},
\code{\link{section}},
\code{\link{subset,section-method}},
\code{\link{summary,section-method}}
}
\author{
Dan Kelley
}
\concept{things related to section data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/argo.R
\name{summary,argo-method}
\alias{summary,argo-method}
\alias{summary.argo}
\title{Summarize an Argo Object}
\usage{
\S4method{summary}{argo}(object, ...)
}
\arguments{
\item{...}{Further arguments passed to or from other methods.}

\item{object}{an}{object of class \code{"argo"}, usually, a result of a
call to \code{\link[=read.argo]{read.argo()}}.}
}
\value{
A matrix containing statistics of the elements of the \code{data} slot.
}
\description{
Summarizes some of the data in an \code{argo} object.
}
\details{
Pertinent summary information is presented.
}
\examples{
library(oce)
data(argo)
summary(argo)

}
\seealso{
Other things related to argo data: 
\code{\link{[[,argo-method}},
\code{\link{[[<-,argo-method}},
\code{\link{argo-class}},
\code{\link{argoGrid}()},
\code{\link{argoNames2oceNames}()},
\code{\link{argo}},
\code{\link{as.argo}()},
\code{\link{handleFlags,argo-method}},
\code{\link{plot,argo-method}},
\code{\link{read.argo}()},
\code{\link{subset,argo-method}}
}
\author{
Dan Kelley
}
\concept{things related to argo data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{rescale}
\alias{rescale}
\title{Rescale values to lie in a given range}
\usage{
rescale(x, xlow, xhigh, rlow = 0, rhigh = 1, clip = TRUE)
}
\arguments{
\item{x}{a numeric vector.}

\item{xlow}{\code{x} value to correspond to \code{rlow}.  If not given, it
will be calculated as the minimum value of \code{x}}

\item{xhigh}{\code{x} value to correspond to \code{rhigh}.  If not given, it
will be calculated as the maximum value of \code{x}}

\item{rlow}{value of the result corresponding to \code{x} equal to
\code{xlow}.}

\item{rhigh}{value of the result corresponding to \code{x} equal to
\code{xhigh}.}

\item{clip}{logical, set to \code{TRUE} to clip the result to the range
spanned by \code{rlow} and \code{rhigh}.}
}
\value{
A new vector, which has minimum \code{lim[1]} and maximum \code{lim[2]}.
}
\description{
This is helpful in e.g. developing a color scale for an image plot.  It is
not necessary that \code{rlow} be less than \code{rhigh}, and in fact
reversing them is a good way to get a reversed color scale for a plot.
}
\examples{
library(oce)
# Fake tow-yow data
t <- seq(0, 600, 5)
x <- 0.5 * t
z <- 50 * (-1 + sin(2 * pi * t / 360))
T <- 5 + 10 * exp(z / 100)
palette <- oce.colorsViridis(100)
zlim <- range(T)
drawPalette(zlim=zlim, col=palette)
plot(x, z, type='p', pch=20, cex=3,
     col=palette[rescale(T, xlow=zlim[1], xhigh=zlim[2], rlow=1, rhigh=100)])
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.R
\name{[[,adp-method}
\alias{[[,adp-method}
\title{Extract Something from an adp Object}
\usage{
\S4method{[[}{adp}(x, i, j, ...)
}
\arguments{
\item{x}{an \linkS4class{adp} object.}

\item{i}{character value indicating the name of an item to extract.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{ignored.}
}
\description{
Generally, the \code{[[} method lets users extract information from \code{oce}
objects, without having to know the details of the internal storage.  For
many \code{oce} sub-classes, \code{[[} can also return quantities that are computed
from the object's contents.
}
\details{
A two-step process is used to try to find the requested information. First, a
class-specific function is used (see \dQuote{Details of the Specialized
Method}).  If this yields nothing, then a general method is used (see
\dQuote{Details of the General Method}).  If both methods fail, then \code{[[}
returns NULL.

Some understanding of the subclass is required to know what can be retrieved
with \code{[[}. When dealing with an unfamiliar subclass, it can be useful to
first use \code{x[["?"]]} to get a listing of the retrievable items. See
\dQuote{Details of the Specialized Method} for more information.
}
\section{Details of the Specialized Method}{


Note that the entries within \linkS4class{adp} objects vary greatly, from
instrument to instrument, and so are only sketched here, and in the output
from \verb{[["?"]]}.
\itemize{
\item If \code{i} is \code{"?"}, then the return value is a list
containing four items, each of which is a character vector
holding the names of things that can be accessed with \code{[[}.
The \code{data} and \code{metadata} items hold the names of
entries in the object's data and metadata
slots, respectively. The \code{dataDerived}
and \code{metadataDerived} items are \emph{not} authoritative, because
information provided by different instruments is so varied.
\item If \code{i} is \code{"u1"} then the return value is \code{v[,1]}. The
same holds for 2, etc., depending on the number of beams in
the instrument.
\item If \code{i} is \code{"a1"} then signal amplitude is returned, and similarly
for other digits. The results can be in \code{\link[=raw]{raw()}} or numeric form,
as shown in the examples.
\item If \code{i} is \code{"q1"} then signal quality is returned, and similarly
for other digits.  As with amplitude, the result can be in \code{\link[=raw]{raw()}}
or numeric form.
\item If \code{i} is \code{"coordinate"}, then the coordinate system is
retrieved.
}
}

\section{Details of the General Method}{


Note: the text of this section is identical for all \code{oce} subclasses, and so
some of what you read here may not be relevant to the class being described
in this help page.

If the specialized method produces no matches, the following generalized
method is applied. As with the specialized method, the procedure hinges first
on the values of \code{i} and, optionally, \code{j}. The work proceeds in steps, by
testing a sequence of possible conditions in sequence.
\enumerate{
\item A check is made as to whether \code{i} names one of the standard \code{oce} slots.
If so, \code{[[} returns the slot contents of that slot.  Thus, \code{x[["metadata"]]}
will retrieve the \code{metadata} slot, while \code{x[["data"]]} and
\code{x[["processingLog"]]} return those slots.
\item If \code{i} is a string ending in the \code{"Unit"}, then the characters preceding
that string are taken to be the name of an item in the data object, and a
list containing the unit is returned (or \code{NULL} if there is no such unit).
This list consists of an item named \code{unit}, which is an \code{\link[=expression]{expression()}}, and
an item named \code{scale}, which is a string describing the measurement scale.
If the string ends in \code{" unit"}, e.g.  \code{x[["temperature unit"]]} (note the
space), then just the expression is returned, and if it ends in \code{" scale"},
then just the scale is returned.
\item If \code{i} is a string ending in \code{"Flag"}, then the corresponding data-quality
flag is returned (or \code{NULL} if there is no such flag).
\item If the object holds hydrographic information (pressure, salinity,
temperature, longitude and latitude) then another set of possibilities
arises.  If \code{i} is \code{"sigmaTheta"}, then the value of \code{\link[=swSigmaTheta]{swSigmaTheta()}} is
called with \code{x} as the sole argument, and the results are returned.
Similarly, \code{\link[=swSigma0]{swSigma0()}} is used if \code{i="sigma0"}, and \code{\link[=swSpice]{swSpice()}} is used if
\code{i="spice"}. Of course, these actions only make sense for objects that
contain the relevant items within their \code{data} slot.
\item After these possibilities are eliminated, the action depends on whether
\code{j} has been provided.  If \code{j} is not provided, or is the string \code{""}, then
\code{i} is sought in the \code{metadata} slot, and then in the \code{data} slot, returning
whichever is found first.  In other words, if \code{j} is not provided, the
\code{metadata} slot takes preference over the \code{data} slot. However, if \code{j} is
provided, then it must be either the string \code{"metadata"} or \code{"data"}, and it
directs where to look.
\item If none of the above-listed conditions holds, then \code{NULL} is returned.
}
}

\examples{
data(adp)
# Tests for beam 1, distance bin 1, first 5 observation times
adp[["v"]][1:5,1,1]
adp[["a"]][1:5,1,1]
adp[["a", "numeric"]][1:5,1,1]
as.numeric(adp[["a"]][1:5,1,1]) # same as above

}
\seealso{
Other functions that extract parts of oce objects: 
\code{\link{[[,adv-method}},
\code{\link{[[,amsr-method}},
\code{\link{[[,argo-method}},
\code{\link{[[,bremen-method}},
\code{\link{[[,cm-method}},
\code{\link{[[,coastline-method}},
\code{\link{[[,ctd-method}},
\code{\link{[[,echosounder-method}},
\code{\link{[[,g1sst-method}},
\code{\link{[[,gps-method}},
\code{\link{[[,ladp-method}},
\code{\link{[[,landsat-method}},
\code{\link{[[,lisst-method}},
\code{\link{[[,lobo-method}},
\code{\link{[[,met-method}},
\code{\link{[[,oce-method}},
\code{\link{[[,odf-method}},
\code{\link{[[,rsk-method}},
\code{\link{[[,sealevel-method}},
\code{\link{[[,section-method}},
\code{\link{[[,tidem-method}},
\code{\link{[[,topo-method}},
\code{\link{[[,windrose-method}},
\code{\link{[[,xbt-method}},
\code{\link{[[<-,adv-method}}

Other things related to adp data: 
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley
}
\concept{functions that extract parts of oce objects}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.R
\docType{data}
\name{ctdRaw}
\alias{ctdRaw}
\title{Seawater CTD Profile, Without Trimming of Extraneous Data}
\usage{
data(ctdRaw)
}
\description{
This is sample CTD profile provided for testing.  It includes not just the
(useful) portion of the dataset during which the instrument was being lowered,
but also data from the upcast and from time spent near the surface.  Spikes are
also clearly evident in the pressure record.  With such real-world wrinkles,
this dataset provides a good example of data that need trimming with
\code{\link[=ctdTrim]{ctdTrim()}}.
}
\details{
This station was sampled by students enrolled in the Dan Kelley's
Physical Oceanography class at Dalhousie University.
The data were acquired near the centre of the Bedford Basin of the
Halifax Harbour, during an October 2003 field trip of Dalhousie University's
Oceanography 4120/5120 class. The original \code{.cnv} data file had
temperature in the IPTS-68 scale, but this was converted to the more modern
scale using \code{\link[=T90fromT68]{T90fromT68()}}.
}
\seealso{
A similar dataset (trimmed to the downcast) is available as
\code{data(ctd)}.

Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}

Other datasets provided with oce: 
\code{\link{adp}},
\code{\link{adv}},
\code{\link{amsr}},
\code{\link{argo}},
\code{\link{cm}},
\code{\link{coastlineWorld}},
\code{\link{ctd}},
\code{\link{echosounder}},
\code{\link{landsat}},
\code{\link{lisst}},
\code{\link{lobo}},
\code{\link{met}},
\code{\link{ocecolors}},
\code{\link{rsk}},
\code{\link{sealevelTuktoyaktuk}},
\code{\link{sealevel}},
\code{\link{section}},
\code{\link{topoWorld}},
\code{\link{wind}},
\code{\link{xbt}}
}
\concept{datasets provided with oce}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{trimString}
\alias{trimString}
\title{Remove leading and trailing whitespace from strings}
\usage{
trimString(s)
}
\arguments{
\item{s}{vector of character strings}
}
\value{
a new vector formed by trimming leading and trailing whitespace
from the elements of \code{s}.
}
\description{
Remove leading and trailing whitespace from strings
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{standardizeLongitude}
\alias{standardizeLongitude}
\title{Put longitude in the range from -180 to 180}
\usage{
standardizeLongitude(longitude)
}
\arguments{
\item{longitude}{in degrees East, possibly exceeding 180}
}
\value{
longitude in signed degrees East
}
\description{
Put longitude in the range from -180 to 180
}
\seealso{
\code{\link[=matrixShiftLongitude]{matrixShiftLongitude()}} and \code{\link[=shiftLongitude]{shiftLongitude()}} are more
powerful relatives to \code{standardizeLongitude}.
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/extdata.R
\docType{data}
\encoding{UTF-8}
\name{d200321-001.ctd}
\alias{d200321-001.ctd}
\title{Sample ctd dataset in .ctd format}
\description{
Sample ctd dataset in .ctd format
}
\examples{
\dontrun{
read.oce(system.file("extdata", "d200321-001.ctd", package="oce"))
}

}
\seealso{
Other raw datasets: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{adp_rdi.000}},
\code{\link{ctd.cnv}},
\code{\link{d201211_0011.cnv}},
\code{\link{test_met_csv1.csv}},
\code{\link{test_met_csv2.csv}},
\code{\link{test_met_xml2.xml}},
\code{\link{xbt.edf}}

Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}
}
\concept{raw datasets}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/rsk.R
\name{rsk2ctd}
\alias{rsk2ctd}
\title{Create a ctd Object from an rsk Object}
\usage{
rsk2ctd(
  x,
  pressureAtmospheric = 0,
  longitude = NULL,
  latitude = NULL,
  ship = NULL,
  cruise = NULL,
  station = NULL,
  deploymentType = NULL,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{x}{an \linkS4class{rsk} object.}

\item{pressureAtmospheric}{A numerical value (a constant or a vector),
that is subtracted from the pressure in \code{object} before storing it in the return value.}

\item{longitude}{numerical value of longitude, in degrees East.}

\item{latitude}{numerical value of latitude, in degrees North.}

\item{ship}{optional string containing the ship from which the observations were made.}

\item{cruise}{optional string containing a cruise identifier.}

\item{station}{optional string containing a station identifier.}

\item{deploymentType}{character string indicating the type of deployment (see
\code{\link[=as.ctd]{as.ctd()}}).}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\description{
A new \code{ctd} object is assembled from the contents of the \code{rsk} object.
The data and metadata are mostly unchanged, with an important exception: the
\code{pressure} item in the \code{data} slot may altered, because \code{rsk}
instruments measure total pressure, not sea pressure; see \dQuote{Details}.
}
\details{
The \code{pressureType} element of the
\code{metadata} of \code{rsk} objects defines the pressure type, and this controls
how pressure is set up in the returned object. If \code{object@metadata$pressureType}
is \code{"absolute"} (or \code{NULL}) then the resultant pressure will be adjusted
to make it into \code{"sea"} pressure. To do this, the value of
\code{object@metadata$pressureAtmospheric} is inspected. If this is present, then
it is subtracted from \code{pressure}. If this is missing, then
standard pressure (10.1325 dbar) will be subtracted. At this stage, the
pressure should be near zero at the ocean surface, but some additional adjustment
might be necessary, and this may be indicated by setting the argument \code{pressureAtmospheric} to
a non-zero value to be subtracted from pressure.
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/met.R
\name{read.met}
\alias{read.met}
\title{Read a met File}
\usage{
read.met(
  file,
  type = NULL,
  skip = NULL,
  tz = getOption("oceTz"),
  encoding = "UTF-8-BOM",
  debug = getOption("oceDebug")
)
}
\arguments{
\item{file}{a character string naming a file that holds met data.}

\item{type}{if \code{NULL}, which is the default, then an attempt is
made to infer the type from the file contents. If this fails, it
will be necessary for the user to provide a value for the \code{type}
argument.  The permitted choices are: (a) \code{"csv"} or \code{"csv1"} for an
old CSV format no longer provided as of October 2019,
(b) \code{"csv2"} for a CSV format noticed on the Environment Canada
website in October 2019 (but note that the paired metadata file
is ignored), (c) \code{"csv3"} for a CSV format noticed on the
Environment Canada website in January 2020,
and (d) \code{"xml2"} for an XML format that was noticed
on the Environment Canada website in October 2019.}

\item{skip}{number of lines of header that occur before the actual
data.  This is ignored unless \code{type} is \code{"csv"} or \code{"csv1"}, in which case
a non-\code{NULL} value of \code{skip} is taken as the number of lines preceding
the columnar data ... and this is only needed if \code{\link[=read.met]{read.met()}} cannot
find a line starting with \code{"Date/Time"} (for csv2 format)
or \code{"Date/Time (LTC)"} (for csv3 format).}

\item{tz}{timezone assumed for time data.  This defaults to
\code{getOption("oceTz")}, which is very likely to be wrong.  In
a scientific context, where UTC is typically used for oceanographic
measurement, it makes sense to set \code{tz="UTC"}.  Note that these
data files do not contain timezone information, instead giving
data in Local Standard Time (LST).  Since LST differs from city
to city, users must make corrections to the time, as
shown in the \dQuote{Examples}, which use data for
Halifax Nova Scotia, where LST is UTC-4.}

\item{encoding}{a character value indicating the encoding to use.  This
parameter was added on 2022-03-03, when CRAN tests
for R-devel/linux-debian-clang were found to report an error for
a test file that began with a BOM (byte order
marker) character. The default, \code{"UTF-8-BOM"} (changed from the
old value of \code{"UTF-8"}) may handle
such files, according to the documentation for \code{\link[=connections]{connections()}}.}

\item{debug}{a flag that turns on debugging.  Set to 1 to get a moderate
amount of debugging information, or to 2 to get more.}
}
\value{
A \linkS4class{met} object.
}
\description{
Reads some meteorological file formats used by the Environment
Canada (see reference 1).  Since the agency does not publish the
data formats, this function has to be adjusted every few years,
when a user finds that the format has changed.
}
\examples{
# Example 1: "csv1" Environment Canada format (found to be obsolete as of Oct 2019)
csv1 <- read.met(system.file("extdata", "test_met_vsn1.csv", package="oce"))
csv1 <- oceSetData(csv1, "time", csv1[["time"]]+4*3600,
    note="add 4h to local time to get UTC time")

# Example 2: "csv2" Environment Canada format (found to be obsolete as of Jan 2022)
csv2 <- read.met(system.file("extdata", "test_met_vsn2.csv", package="oce"))
csv2 <- oceSetData(csv2, "time", csv2[["time"]]+4*3600,
    note="add 4h to local time to get UTC time")

# Example 3: "csv3" Environment Canada format. Note timezone correction
csv3 <- read.met(system.file("extdata", "test_met_vsn3.csv", package="oce"))
csv3 <- oceSetData(csv3, "time", csv3[["time"]]+4*3600,
    note="add 4h to local time to get UTC time")

# Example 4: "xml2" format. (Uncertain timezone, so not corrected.)
if (requireNamespace("XML", quietly=TRUE))
    xml2 <- read.met(system.file("extdata", "test_met_xml2.xml", package="oce"))

# Example 5: download and plot
\dontrun{
library(oce)
# Recreate data(met) and plot u(t) and v(t)
metFile <- download.met(id=6358, year=2003, month=9, destdir=".")
met <- read.met(metFile)
met <- oceSetData(met, "time", met[["time"]]+4*3600,
    note="add 4h to local time to get UTC time")
plot(met)
}

}
\references{
\enumerate{
\item Environment Canada website for Historical Climate Data
\verb{https://climate.weather.gc.ca/index_e.html}
}
}
\seealso{
Other things related to met data: 
\code{\link{[[,met-method}},
\code{\link{[[<-,met-method}},
\code{\link{as.met}()},
\code{\link{download.met}()},
\code{\link{met-class}},
\code{\link{met}},
\code{\link{plot,met-method}},
\code{\link{subset,met-method}},
\code{\link{summary,met-method}},
\code{\link{test_met_csv1.csv}},
\code{\link{test_met_csv2.csv}},
\code{\link{test_met_xml2.xml}}
}
\author{
Dan Kelley
}
\concept{things related to met data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.nortek.R
\name{is.ad2cp}
\alias{is.ad2cp}
\title{Test whether object is an AD2CP type}
\usage{
is.ad2cp(x)
}
\arguments{
\item{x}{character value naming an item.}
}
\value{
Logical value indicating whether \code{x} is an \linkS4class{adp} object,
with \code{fileType} in its \code{metadata} slot equal to \code{"AD2CP"}.
}
\description{
Test whether object is an AD2CP type
}
\seealso{
Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{oceColorsCDOM}
\alias{oceColorsCDOM}
\alias{oce.colorsCDOM}
\title{Create colors suitable for CDOM fields}
\usage{
oceColorsCDOM(n)
}
\arguments{
\item{n}{number of colors to create.}
}
\value{
A vector of color specifications.
}
\description{
Create a set of colors for displaying CDOM values,
based on the scheme devised by Thyng et al. (2016) and
presented in a python package by Thyng (2019).
The color specifications were transliterated from
python to R on 2015-09-29, but have not been adjusted
since, even though the python source has changed.  This
is to prevent breaking old \code{oce} code.  To get the
latest versions of these colours or other colours,
use the \CRANpkg{cmocean} R package
(Thyng, Richards, and Krylov, 2019) directly, as is
illustrated (with the "matter" scheme) in Example 2.
Note that the \CRANpkg{cmocean} core functions provide a way
to select between various versions of the colour schemes.
It is also worth considering the palettes provided by the
\CRANpkg{viridis} package, as illustrated (with the
"inferno" scheme) in Example 3.
}
\examples{
library(oce)

# Example 1
imagep(volcano, col=oceColorsCDOM(128),
       zlab="oceColorsCDOM")
\dontrun{
# Example 2 (requires the cmocean package)
imagep(volcano, col=cmocean::cmocean("matter"),
       zlab='cmocean::cmocean("matter")')}

\dontrun{
# Example 3 (requires the viridis package)
imagep(volcano, col=viridis::inferno,
       zlab='viridis::inferno')}

}
\references{
\itemize{
\item Thyng, Kristen, Chad Greene, Robert Hetland, Heather Zimmerle, and Steven DiMarco.
\dQuote{True Colors of Oceanography: Guidelines for Effective and Accurate Colormap Selection.}
Oceanography 29, no. 3 (September 1, 2016): 9‚Äì13.
\doi{10.5670/oceanog.2016.66}
\item Thyng, Kristen. Kthyng/Cmocean. Python, 2019.
\url{https://github.com/kthyng/cmocean}.
\item Thyng, Kristen, Clark Richards, and Ivan Krylov.
Cmocean: Beautiful Colour Maps for Oceanography (version 0.2), 2019.
\url{https://CRAN.R-project.org/package=cmocean}.
}

The following references provide information on choosing
colour schemes, that are suitable for viewers who have
colour deficiencies.

Light, Adam, and Patrick J. Bartlein. "The End of the Rainbow? Color
Schemes for Improved Data Graphics." \emph{Eos, Transactions American Geophysical
Union} 85, no. 40 (2004): 385.
DOI: 10.1029/2004EO400002

Stephenson, David B. "Comment on 'Color Schemes for Improved Data
Graphics', by A Light and P.J. Bartlein." \emph{Eos, Transactions American
Geophysical Union} 86, no.  20 (2005): 196.
DOI: 10.1029/2005EO200005

Light, Adam, and Patrick J. Bartlein. "Reply to 'Comment on Color
Schemes for Improved Data Graphics,' by A. Light and P.J.  Bartlein'."
\emph{Eos, Transactions American Geophysical Union} 86, no. 20 (2005): 196‚Äì196.
DOI: 10.1029/2005EO200006
}
\seealso{
Other things related to colors: 
\code{\link{colormapGMT}()},
\code{\link{colormap}()},
\code{\link{oceColors9B}()},
\code{\link{oceColorsChlorophyll}()},
\code{\link{oceColorsClosure}()},
\code{\link{oceColorsDensity}()},
\code{\link{oceColorsFreesurface}()},
\code{\link{oceColorsGebco}()},
\code{\link{oceColorsJet}()},
\code{\link{oceColorsOxygen}()},
\code{\link{oceColorsPAR}()},
\code{\link{oceColorsPalette}()},
\code{\link{oceColorsPhase}()},
\code{\link{oceColorsSalinity}()},
\code{\link{oceColorsTemperature}()},
\code{\link{oceColorsTurbidity}()},
\code{\link{oceColorsTurbo}()},
\code{\link{oceColorsTwo}()},
\code{\link{oceColorsVelocity}()},
\code{\link{oceColorsViridis}()},
\code{\link{oceColorsVorticity}()},
\code{\link{ocecolors}}
}
\author{
Krysten M. Thyng (Python version), Dan Kelley (R transliteration)
}
\concept{things related to colors}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{T90fromT48}
\alias{T90fromT48}
\title{Convert from ITS-48 to ITS-90 temperature}
\usage{
T90fromT48(temperature)
}
\arguments{
\item{temperature}{Vector of temperatures in \eqn{^\circ}{deg}C on the IPTS-48 scale.}
}
\value{
Corresponding temperatures in \eqn{^\circ}{deg}C on the ITS-90 scale.
}
\description{
Today's instruments typically record in the ITS-90 scale, but some old
datasets will be in the IPTS-68 scale. \code{T90fromT68()} converts from the
IPTS-68 to the ITS-90 scale, using Saunders' (1990) formula, while
\code{T68fromT90()} does the reverse. The difference between IPTS-68 and
ITS-90 values is typically a few millidegrees (see \sQuote{Examples}), which
is seldom visible on a typical temperature profile, but may be of interest
in some precise work.  Mostly for historical interest, \code{T90fromT48()}
is provided to convert from the ITS-48 system to ITS-90.
}
\examples{
library(oce)
T68 <- seq(3, 20, 1)
T90 <- T90fromT68(T68)
sqrt(mean((T68-T90)^2))

}
\references{
P. M. Saunders, 1990. The international temperature scale of
1990, ITS-90.  WOCE Newsletter, volume 10, September 1990, page 10.
http://www.nodc.noaa.gov/woce/wdiu/wocedocs/newsltr/news10/contents.htm
}
\seealso{
Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.R
\name{plotScan}
\alias{plotScan}
\title{Plot CTD data in a Low-Level Fashion}
\usage{
plotScan(
  x,
  which = 1,
  xtype = "scan",
  flipy = FALSE,
  type = "l",
  mgp = getOption("oceMgp"),
  xlim = NULL,
  ylim = NULL,
  mar = c(mgp[1] + 1.5, mgp[1] + 1.5, mgp[1], mgp[1]),
  ...,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{x}{a \linkS4class{ctd} object.}

\item{which}{Numerical vector numerical codes specifying the panels to draw: 1
for pressure vs scan, 2 for \code{diff(pressure)} vs scan, 3 for temperature vs
scan, and 4 for salinity vs scan.}

\item{xtype}{Character string indicating variable for the x axis. May be
\code{"scan"} (the default) or \code{"time"}. In the former case, a
\code{scan} variable will be created using \code{\link[=seq_along]{seq_along()}},
if necessary. In the latter case, an error results if the \code{data}
slot of \code{x} lacks a variable called \code{time}.}

\item{flipy}{Logical value, ignored unless \code{which} is 1. If \code{flipy}
is \code{TRUE}, then a pressure plot will have high pressures at the bottom
of the axis.}

\item{type}{Character indicating the line type, as for \code{\link[=plot.default]{plot.default()}}. The default
is \code{"l"}, meaning to connect data with line segments. Another good choice is
\code{"o"}, to add points at the data.}

\item{mgp}{Three-element numerical vector to use for \link{par}\code{(mgp)}, and also
for \link{par}\code{(mar)}, computed from this.  The default is tighter than the R
default, in order to use more space for the data and less for the axes.}

\item{xlim}{Limits on the x value. The default, \code{NULL}, is to select this
from the data.}

\item{ylim}{Limits on the y value. The default, \code{NULL}, is to select this
from the data.}

\item{mar}{Four-element vector be used with \link{par}\code{("mar")}.  If set
to \code{NULL}, then \link{par}\code{("mar")} is used.  A good choice for a TS diagram
with a palette to the right is \verb{mar=par("mar")+c(0, 0, 0, 1))}.}

\item{...}{Optional arguments passed to plotting functions.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\description{
Plot CTD data as time-series against scan number, to help with trimming
extraneous data from a CTD cast.
}
\examples{
library(oce)
data(ctdRaw)
plotScan(ctdRaw)
abline(v=c(130, 350), col='red') # useful for ctdTrim()

}
\seealso{
Other functions that plot oce data: 
\code{\link{download.amsr}()},
\code{\link{plot,adp-method}},
\code{\link{plot,adv-method}},
\code{\link{plot,amsr-method}},
\code{\link{plot,argo-method}},
\code{\link{plot,bremen-method}},
\code{\link{plot,cm-method}},
\code{\link{plot,coastline-method}},
\code{\link{plot,ctd-method}},
\code{\link{plot,gps-method}},
\code{\link{plot,ladp-method}},
\code{\link{plot,landsat-method}},
\code{\link{plot,lisst-method}},
\code{\link{plot,lobo-method}},
\code{\link{plot,met-method}},
\code{\link{plot,odf-method}},
\code{\link{plot,rsk-method}},
\code{\link{plot,satellite-method}},
\code{\link{plot,sealevel-method}},
\code{\link{plot,section-method}},
\code{\link{plot,tidem-method}},
\code{\link{plot,topo-method}},
\code{\link{plot,windrose-method}},
\code{\link{plot,xbt-method}},
\code{\link{plotProfile}()},
\code{\link{plotTS}()},
\code{\link{tidem-class}}

Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}
}
\author{
Dan Kelley
}
\concept{functions that plot oce data}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{oceColorsDensity}
\alias{oceColorsDensity}
\alias{oce.colorsDensity}
\title{Create colors suitable for density fields}
\usage{
oceColorsDensity(n)
}
\arguments{
\item{n}{number of colors to create.}
}
\value{
A vector of color specifications.
}
\description{
Create a set of colors for displaying density values,
based on the scheme devised by Thyng et al. (2016) and
presented in a python package by Thyng (2019).
The color specifications were transliterated from
python to R on 2015-09-29, but have not been adjusted
since, even though the python source has changed.  This
is to prevent breaking old \code{oce} code.  To get the
latest versions of these colours or other colours,
use the \CRANpkg{cmocean} R package
(Thyng, Richards, and Krylov, 2019) directly, as is
illustrated (with the "matter" scheme) in Example 2.
Note that the \CRANpkg{cmocean} core functions provide a way
to select between various versions of the colour schemes.
It is also worth considering the palettes provided by the
\CRANpkg{viridis} package, as illustrated (with the
"inferno" scheme) in Example 3.
}
\examples{
library(oce)

# Example 1
imagep(volcano, col=oceColorsDensity(128),
       zlab="oceColorsDensity")
\dontrun{
# Example 2 (requires the cmocean package)
imagep(volcano, col=cmocean::cmocean("matter"),
       zlab='cmocean::cmocean("matter")')}

\dontrun{
# Example 3 (requires the viridis package)
imagep(volcano, col=viridis::inferno,
       zlab='viridis::inferno')}

}
\references{
\itemize{
\item Thyng, Kristen, Chad Greene, Robert Hetland, Heather Zimmerle, and Steven DiMarco.
\dQuote{True Colors of Oceanography: Guidelines for Effective and Accurate Colormap Selection.}
Oceanography 29, no. 3 (September 1, 2016): 9‚Äì13.
\doi{10.5670/oceanog.2016.66}
\item Thyng, Kristen. Kthyng/Cmocean. Python, 2019.
\url{https://github.com/kthyng/cmocean}.
\item Thyng, Kristen, Clark Richards, and Ivan Krylov.
Cmocean: Beautiful Colour Maps for Oceanography (version 0.2), 2019.
\url{https://CRAN.R-project.org/package=cmocean}.
}

The following references provide information on choosing
colour schemes, that are suitable for viewers who have
colour deficiencies.

Light, Adam, and Patrick J. Bartlein. "The End of the Rainbow? Color
Schemes for Improved Data Graphics." \emph{Eos, Transactions American Geophysical
Union} 85, no. 40 (2004): 385.
DOI: 10.1029/2004EO400002

Stephenson, David B. "Comment on 'Color Schemes for Improved Data
Graphics', by A Light and P.J. Bartlein." \emph{Eos, Transactions American
Geophysical Union} 86, no.  20 (2005): 196.
DOI: 10.1029/2005EO200005

Light, Adam, and Patrick J. Bartlein. "Reply to 'Comment on Color
Schemes for Improved Data Graphics,' by A. Light and P.J.  Bartlein'."
\emph{Eos, Transactions American Geophysical Union} 86, no. 20 (2005): 196‚Äì196.
DOI: 10.1029/2005EO200006
}
\seealso{
Other things related to colors: 
\code{\link{colormapGMT}()},
\code{\link{colormap}()},
\code{\link{oceColors9B}()},
\code{\link{oceColorsCDOM}()},
\code{\link{oceColorsChlorophyll}()},
\code{\link{oceColorsClosure}()},
\code{\link{oceColorsFreesurface}()},
\code{\link{oceColorsGebco}()},
\code{\link{oceColorsJet}()},
\code{\link{oceColorsOxygen}()},
\code{\link{oceColorsPAR}()},
\code{\link{oceColorsPalette}()},
\code{\link{oceColorsPhase}()},
\code{\link{oceColorsSalinity}()},
\code{\link{oceColorsTemperature}()},
\code{\link{oceColorsTurbidity}()},
\code{\link{oceColorsTurbo}()},
\code{\link{oceColorsTwo}()},
\code{\link{oceColorsVelocity}()},
\code{\link{oceColorsViridis}()},
\code{\link{oceColorsVorticity}()},
\code{\link{ocecolors}}
}
\author{
Krysten M. Thyng (Python version), Dan Kelley (R transliteration)
}
\concept{things related to colors}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/section.R
\name{addSpine}
\alias{addSpine}
\title{Add a spine to a section object}
\usage{
addSpine(section, spine, debug = getOption("oceDebug"))
}
\arguments{
\item{section}{a \linkS4class{section} object.}

\item{spine}{either a list or a data frame, containing numeric items named
\code{longitude} and \code{latitude}, defining a path along the spine.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\value{
A \linkS4class{section} object with a spine added.
}
\description{
The purpose of this is to permit plotting with \code{xtype="spine"}, so that
the section plot will display the distance of stations projected
onto the spine.
}
\examples{
library(oce)
data(section)
sectionWest <- subset(section, longitude < -60)
spine <- list(longitude=c(-74.5, -69.2, -55), latitude=c(38.6, 36.25, 36.25))
sectionWithSpine <- addSpine(sectionWest, spine)
plot(sectionWithSpine, which="map")
plot(sectionWithSpine, xtype="distance", which="temperature")
plot(sectionWithSpine, xtype="spine", which="temperature")

}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/odf.R
\name{summary,odf-method}
\alias{summary,odf-method}
\title{Summarize an ODF Object}
\usage{
\S4method{summary}{odf}(object, ...)
}
\arguments{
\item{object}{an \linkS4class{odf} object.}

\item{\dots}{further arguments passed to or from other methods.}
}
\value{
A matrix containing statistics of the elements of the \code{data}
slot.
}
\description{
Pertinent summary information is presented, including the station name,
sampling location, data ranges, etc.
}
\seealso{
Other things related to odf data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{ODF2oce}()},
\code{\link{ODFListFromHeader}()},
\code{\link{ODFNames2oceNames}()},
\code{\link{[[,odf-method}},
\code{\link{[[<-,odf-method}},
\code{\link{odf-class}},
\code{\link{plot,odf-method}},
\code{\link{read.ctd.odf}()},
\code{\link{read.odf}()},
\code{\link{subset,odf-method}}
}
\author{
Dan Kelley
}
\concept{things related to odf data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/map.R
\name{mapCoordinateSystem}
\alias{mapCoordinateSystem}
\title{Draw a coordinate system}
\usage{
mapCoordinateSystem(longitude, latitude, L = 100, phi = 0, ...)
}
\arguments{
\item{longitude}{numeric vector of longitudes in degrees.}

\item{latitude}{numeric vector of latitudes in degrees.}

\item{L}{axis length in km.}

\item{phi}{angle, in degrees counterclockwise, that the "x" axis makes to a line of latitude.}

\item{...}{plotting arguments, passed to \code{\link[=mapArrows]{mapArrows()}};
see \dQuote{Examples} for how to control the arrow-head size.}
}
\description{
Draws arrows on a map to indicate a coordinate system, e.g. for an
to indicate a coordinate system set up so that one axis is parallel
to a coastline.
}
\details{
This is a preliminary version of this function. It only
works if the lines of constant latitude are horizontal on the plot.
}
\examples{
\dontrun{
library(oce)
if (requireNamespace("ocedata", quietly=TRUE)) {
    data(coastlineWorldFine, package='ocedata')
    HfxLon <- -63.5752
    HfxLat <- 44.6488
    mapPlot(coastlineWorldFine, proj='+proj=merc',
            longitudelim=HfxLon+c(-2,2), latitudelim=HfxLat+c(-2,2),
            col='lightgrey')
    mapCoordinateSystem(HfxLon, HfxLat, phi=45, length=0.05)
   }
}

}
\seealso{
Other functions related to maps: 
\code{\link{formatPosition}()},
\code{\link{lonlat2map}()},
\code{\link{lonlat2utm}()},
\code{\link{map2lonlat}()},
\code{\link{mapArrows}()},
\code{\link{mapAxis}()},
\code{\link{mapContour}()},
\code{\link{mapDirectionField}()},
\code{\link{mapGrid}()},
\code{\link{mapImage}()},
\code{\link{mapLines}()},
\code{\link{mapLocator}()},
\code{\link{mapLongitudeLatitudeXY}()},
\code{\link{mapPlot}()},
\code{\link{mapPoints}()},
\code{\link{mapPolygon}()},
\code{\link{mapScalebar}()},
\code{\link{mapText}()},
\code{\link{mapTissot}()},
\code{\link{oceCRS}()},
\code{\link{shiftLongitude}()},
\code{\link{usrLonLat}()},
\code{\link{utm2lonlat}()}
}
\author{
Chantelle Layton
}
\concept{functions related to maps}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ladp.R
\name{[[<-,ladp-method}
\alias{[[<-,ladp-method}
\title{title Replace Parts of a ladp Object}
\usage{
\S4method{[[}{ladp}(x, i, j, ...) <- value
}
\arguments{
\item{x}{an \linkS4class{ladp} object.}

\item{i}{character value naming the item to replace.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{optional additional information (ignored).}

\item{value}{The value to be placed into \code{x}, somewhere.}
}
\description{
The \code{[[<-} method works for all \linkS4class{oce} objects.
The purpose, as with the related extraction method, \code{[[},
is to insulate users from the internal details of \linkS4class{oce}
objects, by looking for items within the various storage
slots of the object. Items not actually stored can also be
replaced, including units and data-quality
flags.
}
\details{
As with \code{[[} method, the procedure works in steps.

First, the \code{metadata} slot of \code{x} is checked to
see whether it contains something named with \code{i}.
If so, then the named item is replaced with \code{value}.

Otherwise, if the string value of \code{i} ends in \code{Unit}, then the
characters preceding that are taken as the name of a variable, and
the \code{metadata} slot of \code{x} is updated to store that unit, e.g.\preformatted{x[["temperatureUnits"]] <- list(unit=expression(degree*F),scale="")
}

Similarly, if \code{i} ends in \code{Flag}, then quality-control
flags are set up as defined by \code{result}, e.g.\preformatted{o[["temperatureFlags"]] <- c(2,4,2,2)
}

Otherwise, \code{\link[=pmatch]{pmatch()}} is used for a partial-string match with
the names of the items that are in the \code{data} slot of \code{x}.
The first item found (if any) is then updated to hold the value \code{result}.

If none of these conditions is met, a warning is issued.
}
\seealso{
Other functions that replace parts of oce objects: 
\code{\link{[[<-,adp-method}},
\code{\link{[[<-,amsr-method}},
\code{\link{[[<-,argo-method}},
\code{\link{[[<-,bremen-method}},
\code{\link{[[<-,cm-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{[[<-,echosounder-method}},
\code{\link{[[<-,g1sst-method}},
\code{\link{[[<-,gps-method}},
\code{\link{[[<-,landsat-method}},
\code{\link{[[<-,lisst-method}},
\code{\link{[[<-,lobo-method}},
\code{\link{[[<-,met-method}},
\code{\link{[[<-,oce-method}},
\code{\link{[[<-,odf-method}},
\code{\link{[[<-,rsk-method}},
\code{\link{[[<-,sealevel-method}},
\code{\link{[[<-,section-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{[[<-,topo-method}},
\code{\link{[[<-,windrose-method}},
\code{\link{[[<-,xbt-method}}

Other things related to ladp data: 
\code{\link{[[,ladp-method}},
\code{\link{as.ladp}()},
\code{\link{ladp-class}},
\code{\link{plot,ladp-method}},
\code{\link{summary,ladp-method}}
}
\concept{functions that replace parts of oce objects}
\concept{things related to ladp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/lisst.R
\name{summary,lisst-method}
\alias{summary,lisst-method}
\title{Summarize a LISST Object}
\usage{
\S4method{summary}{lisst}(object, ...)
}
\arguments{
\item{object}{a \linkS4class{lisst} object.}

\item{...}{Ignored.}
}
\description{
Summarizes some of the data in a \code{lisst} object, presenting such information
as the station name, sampling location, data ranges, etc.
}
\examples{
library(oce)
data(lisst)
summary(lisst)

}
\seealso{
Other things related to lisst data: 
\code{\link{[[,lisst-method}},
\code{\link{[[<-,lisst-method}},
\code{\link{as.lisst}()},
\code{\link{lisst-class}},
\code{\link{plot,lisst-method}},
\code{\link{read.lisst}()}
}
\author{
Dan Kelley
}
\concept{things related to lisst data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/tides.R
\name{plot,tidem-method}
\alias{plot,tidem-method}
\alias{plot.tidem}
\title{Plot a tidem Object}
\usage{
\S4method{plot}{tidem}(
  x,
  which = 1,
  constituents = c("SA", "O1", "K1", "M2", "S2", "M4"),
  sides = NULL,
  col = "blue",
  log = "",
  mgp = getOption("oceMgp"),
  mar = c(mgp[1] + 1, mgp[1] + 1, mgp[2] + 0.25, mgp[2] + 1),
  ...
)
}
\arguments{
\item{x}{a \linkS4class{tidem} object.}

\item{which}{integer flag indicating plot type, 1 for stair-case spectral, 2
for spike spectral.}

\item{constituents}{character vector holding the names of constituents that are
to be drawn and labelled. If \code{NULL}, then no constituents will be shown.}

\item{sides}{an integer vector of length equal to that of \code{constituents},
designating the side on which the constituent labels are to be drawn. As in
all R graphics, the value \code{1} indicates the bottom of the plot, and
\code{3} indicates the top. If \code{sides=NULL}, the default, then all labels
are drawn at the top. Any value of \code{sides} that is not either 1 or 3
is converted to 3.}

\item{col}{a character vector naming colors to be used for \code{constituents}.
Ignored if \code{sides=3}. Repeated to be of the same length as
\code{constituents}, otherwise.}

\item{log}{if set to "\code{x}", the frequency axis will be logarithmic.}

\item{mgp}{3-element numerical vector to use for \code{par(mgp)}, and also
for \code{par(mar)}, computed from this.  The default is tighter than the R
default, in order to use more space for the data and less for the axes.}

\item{mar}{value to be used with \code{[}par\verb{]("mar")}.}

\item{\dots}{optional arguments passed to plotting functions.}
}
\description{
Plot a summary diagram for a tidal fit.
}
\section{Historical note}{

An argument named \code{labelIf} was removed in July 2016,
because it was discovered never to have worked as documented, and
because the more useful argument \code{constituents} had been added.
}

\examples{
\dontrun{
library(oce)
data(sealevel)
tide <- tidem(sealevel)
plot(tide)
}

}
\seealso{
Other functions that plot oce data: 
\code{\link{download.amsr}()},
\code{\link{plot,adp-method}},
\code{\link{plot,adv-method}},
\code{\link{plot,amsr-method}},
\code{\link{plot,argo-method}},
\code{\link{plot,bremen-method}},
\code{\link{plot,cm-method}},
\code{\link{plot,coastline-method}},
\code{\link{plot,ctd-method}},
\code{\link{plot,gps-method}},
\code{\link{plot,ladp-method}},
\code{\link{plot,landsat-method}},
\code{\link{plot,lisst-method}},
\code{\link{plot,lobo-method}},
\code{\link{plot,met-method}},
\code{\link{plot,odf-method}},
\code{\link{plot,rsk-method}},
\code{\link{plot,satellite-method}},
\code{\link{plot,sealevel-method}},
\code{\link{plot,section-method}},
\code{\link{plot,topo-method}},
\code{\link{plot,windrose-method}},
\code{\link{plot,xbt-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{tidem-class}}

Other things related to tides: 
\code{\link{[[,tidem-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{as.tidem}()},
\code{\link{predict.tidem}()},
\code{\link{summary,tidem-method}},
\code{\link{tidalCurrent}},
\code{\link{tidedata}},
\code{\link{tidem-class}},
\code{\link{tidemAstron}()},
\code{\link{tidemVuf}()},
\code{\link{tidem}},
\code{\link{webtide}()}
}
\author{
Dan Kelley
}
\concept{functions that plot oce data}
\concept{things related to tides}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.sontek.R
\name{read.adp.sontek.serial}
\alias{read.adp.sontek.serial}
\title{Read a serial Sontek ADP file}
\usage{
read.adp.sontek.serial(
  file,
  from = 1,
  to,
  by = 1,
  tz = getOption("oceTz"),
  longitude = NA,
  latitude = NA,
  type = c("adp", "pcadp"),
  beamAngle = 25,
  orientation,
  monitor = FALSE,
  processingLog,
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{file}{a connection or a character string giving the name of the file
to load.  (For \code{read.adp.sontek.serial}, this is generally a list of
files, which will be concatenated.)}

\item{from}{indication of the first profile to read.  This can be an
integer, the sequence number of the first profile to read, or a POSIXt time
before which profiles should be skipped, or a character string that converts
to a POSIXt time (assuming UTC timezone).  See \dQuote{Examples}, and make
careful note of the use of the \code{tz} argument. If \code{from} is not
supplied, it defaults to 1.}

\item{to}{an optional indication of the last profile to read, in a
format as described for \code{from}.  As a special case, \code{to=0} means
to read the file to the end.  If \code{to} is not supplied, then it defaults
to 0.}

\item{by}{an optional indication of the stride length to use while walking through
the file.  If this is an integer, then \code{by-1} profiles are skipped
between each pair of profiles that is read, e.g. the default \code{by=1}
means to read all the data.  (For RDI files \emph{only}, there are some
extra features to avoid running out of memory; see \dQuote{Memory considerations}.)}

\item{tz}{character string indicating time zone to be assumed in the data.}

\item{longitude}{optional signed number indicating the longitude in degrees
East.}

\item{latitude}{optional signed number indicating the latitude in degrees
North.}

\item{type}{a character string indicating the type of instrument.}

\item{beamAngle}{angle between instrument axis and beams, in degrees.}

\item{orientation}{optional character string specifying the orientation of the
sensor, provided for those cases in which it cannot be inferred from the
data file.  The valid choices are \code{"upward"}, \code{"downward"}, and
\code{"sideward"}.}

\item{monitor}{boolean value indicating whether to indicate the progress
of reading the file, by using \code{\link[=txtProgressBar]{txtProgressBar()}} or otherwise.  The value
of \code{monitor} is changed to \code{FALSE} automatically, for non-interactive
sessions.}

\item{processingLog}{if provided, the action item to be stored in the log.
(Typically only provided for internal calls; the default that it provides is
better for normal calls by a user.)}

\item{debug}{a flag that turns on debugging.  Set to 1 to get a moderate
amount of debugging information, or to 2 to get more.}

\item{\dots}{optional additional arguments that some (but not all)
\verb{read.adp.*()} functions pass to lower-level functions.}
}
\value{
An \linkS4class{adp} object.
The contents of that object make sense for the particular instrument
type under study, e.g. if the data file contains
NMEA strings, then navigational data will be stored in an item
called \code{nmea} in the \code{data} slot).
}
\description{
Read a Sontek acoustic-Doppler profiler file, in a serial form that
is possibly unique to Dalhousie University.
}
\seealso{
Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley and Clark Richards
}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/rsk.R
\name{read.rsk}
\alias{read.rsk}
\title{Read a Rsk file}
\usage{
read.rsk(
  file,
  from = 1,
  to,
  by = 1,
  type,
  tz = getOption("oceTz", default = "UTC"),
  patm = FALSE,
  processingLog,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{file}{a connection or a character string giving the name of the file to
load. Note that \code{file} must be a character string, because connections are
not used in that case, which is instead handled with database calls.}

\item{from}{indication of the first datum to read.  This can a positive integer
to indicate sequence number, the POSIX time of the first datum, or a character
string that can be converted to a POSIX time.  (For POSIX times, be careful
about the \code{tz} argument.)}

\item{to}{an indication of the last datum to be read, in the same format as
\code{from}.  If \code{to} is missing, data will be read to the end of the file.}

\item{by}{an indication of the stride length to use while walking through the
file.  If this is an integer, then \code{by-1} samples are skipped between each
pair of samples that is read.  If this is a string representing a time interval,
in colon-separated format (HH:MM:SS or MM:SS), then this interval is divided by
the sampling interval, to get the stride length.}

\item{type}{optional file type, presently can be \code{rsk} or \code{txt} (for a
text export of an RBR rsk or hex file). If this argument is not provided, an
attempt will be made to infer the type from the file name and contents.}

\item{tz}{time zone.  The value \code{oceTz} is set at package setup.}

\item{patm}{controls the handling of atmospheric pressure, an important issue
for RBR instruments that record absolute pressure; see \dQuote{Details}.}

\item{processingLog}{if provided, the action item to be stored in the log.
This is typically only provided for internal calls; the default that it provides
is better for normal calls by a user.}

\item{debug}{a flag that can be set to \code{TRUE} to turn on debugging.}
}
\value{
An \linkS4class{rsk} object.
}
\description{
Read an RBR rsk or txt file, e.g. as produced by an RBR logger, producing an
object of class \code{rsk}.
}
\details{
This can read files produced by several RBR instruments.  At the moment, five
styles are understood: (1) text file produced as an export of an RBR \code{hex}
or \code{rsk} file; (2) text file with columns for temperature and pressure
(with sampling times indicated in the header); (3) text file with four columns,
in which the date the time of day are given in the first two columns, followed
by the temperature, and pressure; (4) text file with five columns, in which
depth in the water column is given after the pressure; (5) an SQLite-based
database format. The first four options are provided mainly for historical
reasons, since RBR instruments at the date of writing commonly use the SQLite
format, though the first option is common for all instruments that produce a
\code{hex} file that can be read using Ruskin.

Options 2-4 are mostly obsolete, and will be removed from future versions.

\emph{A note on conductivity.} RBR devices record conductivity in mS/cm, and it
is this value that is stored in the object returned by \code{read.rsk}. This can
be converted to conductivity ratio (which is what many other instruments report)
by dividing by 42.914 (see Culkin and Smith, 1980) which will be necessary in
any seawater-related function that takes conductivity ratio as an argument (see
\dQuote{Examples}).

\emph{A note on pressure.} RBR devices tend to record absolute pressure (i.e.
sea pressure plus atmospheric pressure), unlike most oceanographic instruments
that record sea pressure (or an estimate thereof).  The handling of pressure
is controlled with the \code{patm} argument, for which there are three
possibilities.  (1) If \code{patm} is \code{FALSE} (the default), then
pressure read from the data file is stored in the \code{data} slot of return
value, and the \code{metadata} item \code{pressureType} is set to the string
\code{"absolute"}.  (2) If \code{patm} is \code{TRUE}, then an estimate of
atmospheric pressure is subtracted from the raw data. For data files in the
SQLite format (i.e.  \verb{*.rsk} files), this estimate will be the value read
from the file, or the ``standard atmosphere'' value 10.1325 dbar, if the file
lacks this information.  (3) If \code{patm} is a numerical value (or list of
values, one for each sampling time), then `patm` is subtracted from the
raw data.  In cases 2 and 3, an additional column named
`pressureOriginal` is added to the `data` slot to store the value
contained in the data file, and `pressureType` is set to a string
starting with `"sea"`.  See \code{\link[=as.ctd]{as.ctd()}} for details of how this
setup facilitates the conversion of \linkS4class{rsk} objects to
\linkS4class{ctd} objects.
}
\references{
Culkin, F., and Norman D. Smith, 1980. Determination of the concentration of
potassium chloride solution having the same electrical conductivity, at 15 C and
infinite frequency, as standard seawater of salinity 35.0000 ppt (Chlorinity
19.37394 ppt). \emph{IEEE Journal of Oceanic Engineering}, \strong{5}, pp 22-23.
}
\seealso{
The documentation for \linkS4class{rsk} explains the structure of
\code{rsk} objects, and also outlines other functions dealing with them.  Since
RBR has a wide variety of instruments, \code{rsk} datasets can be quite general,
and it is common to coerce \code{rsk} objects to other forms for specialized
work, e.g. \code{\link[=as.ctd]{as.ctd()}} can be used to create CTD object, so that the
generic plot obeys the CTD format.

Other things related to rsk data: 
\code{\link{[[,rsk-method}},
\code{\link{[[<-,rsk-method}},
\code{\link{as.rsk}()},
\code{\link{plot,rsk-method}},
\code{\link{rsk-class}},
\code{\link{rskPatm}()},
\code{\link{rskToc}()},
\code{\link{rsk}},
\code{\link{subset,rsk-method}},
\code{\link{summary,rsk-method}}
}
\author{
Dan Kelley and Clark Richards
}
\concept{things related to rsk data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.R
\name{parseLatLon}
\alias{parseLatLon}
\title{Parse a Latitude or Longitude String}
\usage{
parseLatLon(line, debug = getOption("oceDebug"))
}
\arguments{
\item{line}{a character string containing an indication of latitude or
longitude.}

\item{debug}{a flag that turns on debugging.  Set to 1 to get a moderate
amount of debugging information, or to 2 to get more.}
}
\value{
A numerical value of latitude or longitude.
}
\description{
Parse a latitude or longitude string, e.g. as in the header of a CTD file
The following formats are understood (for, e.g. latitude):\preformatted{** NMEA Latitude = 47 54.760 N
** Latitude: 47 53.27 N
}

Note that \code{\link[=iconv]{iconv()}} is called to convert the string to ASCII before
decoding, to change any degree (or other non-ASCII) symbols to blanks.
}
\seealso{
Used by \code{\link[=read.ctd]{read.ctd()}}.
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{errorbars}
\alias{errorbars}
\title{Draw error bars on an existing xy diagram}
\usage{
errorbars(x, y, xe, ye, percent = FALSE, style = 0, length = 0.025, ...)
}
\arguments{
\item{x, y}{coordinates of points on the existing plot.}

\item{xe, ye}{errors on x and y coordinates of points on the existing plot,
each either a single number or a vector of length identical to that of
the corresponding coordinate.}

\item{percent}{boolean flag indicating whether \code{xe} and \code{ye} are
in terms of percent of the corresponding \code{x} and \code{y} values.}

\item{style}{indication of the style of error bar.  Using \code{style=0}
yields simple line segments (drawn with \code{\link[=segments]{segments()}}) and
\code{style=1} yields line segments with short perpendicular endcaps.}

\item{length}{length of endcaps, for \code{style=1} only; it is passed to
\code{\link[=arrows]{arrows()}}, which is used to draw that style of error bars.}

\item{\dots}{graphical parameters passed to the code that produces the error
bars, e.g. to \code{\link[=segments]{segments()}} for \code{style=0}.}
}
\description{
Draw error bars on an existing xy diagram
}
\examples{
library(oce)
data(ctd)
S <- ctd[["salinity"]]
T <- ctd[["temperature"]]
plot(S, T)
errorbars(S, T, 0.05, 0.5)
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swDynamicHeight}
\alias{swDynamicHeight}
\title{Dynamic height of seawater profile}
\usage{
swDynamicHeight(
  x,
  referencePressure = 2000,
  subdivisions = 500,
  rel.tol = .Machine$double.eps^0.25,
  eos = getOption("oceEOS", default = "gsw")
)
}
\arguments{
\item{x}{a \linkS4class{section} object.}

\item{referencePressure}{reference pressure (dbar). If this exceeds the
highest pressure supplied to \code{\link[=swDynamicHeight]{swDynamicHeight()}}, then that highest
pressure is used, instead of the supplied value of
\code{referencePressure}.}

\item{subdivisions}{number of subdivisions for call to
\code{\link[=integrate]{integrate()}}.  (The default value is considerably larger than the
default for \code{\link[=integrate]{integrate()}}, because otherwise some test profiles
failed to integrate.}

\item{rel.tol}{absolute tolerance for call to \code{\link[=integrate]{integrate()}}.  Note
that this call is made in scaled coordinates, i.e. pressure is divided by
its maximum value, and dz/dp is also divided by its maximum.}

\item{eos}{equation of state, either \code{"unesco"} or \code{"gsw"}.}
}
\value{
In the first form, a list containing \code{distance}, the distance
(km( from the first station in the section and \code{height}, the dynamic
height (m). In the second form, a single value, containing the
dynamic height (m).
}
\description{
Compute the dynamic height of a column of seawater.
}
\details{
If the first argument is a \code{section}, then dynamic height is calculated
for each station within a section, and returns a list containing distance
along the section along with dynamic height.

If the first argument is a \code{ctd}, then this returns just a single
value, the dynamic height.

If \code{eos="unesco"}, processing is as follows.  First, a piecewise-linear
model of the density variation with pressure is developed using
\code{\link[stats:approxfun]{stats::approxfun()}}. (The option \code{rule=2} is used to
extrapolate the uppermost density up to the surface, preventing a possible a
bias for bottle data, in which the first depth may be a few metres below the
surface.) A second function is constructed as the density of water with
salinity 35PSU, temperature of 0\eqn{^\circ}{deg}C, and pressure as in the
\code{ctd}. The difference of the reciprocals of these densities, is then
integrated with \code{\link[stats:integrate]{stats::integrate()}} with pressure limits \code{0}
to \code{referencePressure}.  (For improved numerical results, the variables
are scaled before the integration, making both independent and dependent
variables be of order one.)

If \code{eos="gsw"}, \code{\link[gsw:gsw_geo_strf_dyn_height]{gsw::gsw_geo_strf_dyn_height()}} is used
to calculate a result in m^2/s^2, and this is divided by
9.7963\eqn{m/s^2}{m/s^2}.
If pressures are out of order, the data are sorted. If any pressure
is repeated, only the first level is used.
If there are under 4 remaining distinct
pressures, \code{NA} is returned, with a warning.
}
\examples{
\dontrun{
library(oce)
data(section)

# Dynamic height and geostrophy
par(mfcol=c(2,2))
par(mar=c(4.5,4.5,2,1))

# Left-hand column: whole section
# (The smoothing lowers Gulf Stream speed greatly)
westToEast <- subset(section, 1<=stationId&stationId<=123)
dh <- swDynamicHeight(westToEast)
plot(dh$distance, dh$height, type='p', xlab="", ylab="dyn. height [m]")
ok <- !is.na(dh$height)
smu <- supsmu(dh$distance, dh$height)
lines(smu, col="blue")
f <- coriolis(section[["station", 1]][["latitude"]])
g <- gravity(section[["station", 1]][["latitude"]])
v <- diff(smu$y)/diff(smu$x) * g / f / 1e3 # 1e3 converts to m
plot(smu$x[-1], v, type='l', col="blue", xlab="distance [km]", ylab="velocity (m/s)")

# right-hand column: gulf stream region, unsmoothed
gs <- subset(section, 102<=stationId&stationId<=124)
dh.gs <- swDynamicHeight(gs)
plot(dh.gs$distance, dh.gs$height, type='b', xlab="", ylab="dyn. height [m]")
v <- diff(dh.gs$height)/diff(dh.gs$distance) * g / f / 1e3
plot(dh.gs$distance[-1], v, type='l', col="blue",
  xlab="distance [km]", ylab="velocity (m/s)")
}

}
\references{
Gill, A.E., 1982. \emph{Atmosphere-ocean Dynamics}, Academic
Press, New York, 662 pp.
}
\seealso{
Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/odf.R
\name{ODFNames2oceNames}
\alias{ODFNames2oceNames}
\title{Translate ODF CODE strings to Oce Variable Names}
\usage{
ODFNames2oceNames(
  ODFnames,
  columns = NULL,
  PARAMETER_HEADER = NULL,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{ODFnames}{vector of character values that hold ODF names.}

\item{columns}{Optional list containing name correspondances, as described for
\code{\link[=read.ctd.odf]{read.ctd.odf()}}.}

\item{PARAMETER_HEADER}{Optional list containing information on the data variables.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\value{
A list relating ODF names to oce names (see \dQuote{Examples}).
}
\description{
Translate ODF CODE strings to oce variable names. This
is done differently for data names and quality-control
(QC) names.
}
\details{
The following table gives the recognized ODF code names for variables,
along with the translated names as used in oce objects. Note that the
code names are appended with strings such as \code{"_01"}, \code{"_02"}, etc,
for repeats. The converted name for an \code{"_01"} item is as shown below,
and for e.g. \code{"_02"} a suffix 2 is added to the oce name, etc.

QC items (which get stored as \code{flags} in object's
\code{metadata} slots) are assigned names that match those of the
parameters to which they refer.  In parsing ODF files, it is assumed
that QC items refer to the data items that precede them.  This pattern
does not seem to be documented, but it has held in all the files
examined by the author, and a similar assumption is made in other
software systems.  QC items have \code{CODE} values that are
either start with \code{"QQQQ"} or equal \code{"Q<CODE>"},
where \verb{<CODE>} matches the corresponding data item.

\tabular{lll}{
\strong{ODF Code} \tab \strong{Oce Name}      \tab \strong{Notes}                                    \cr
\code{ABSH} \tab \code{humidityAbsolute}      \tab                                              \cr
\code{ACO2} \tab \code{CO2Atmosphere}         \tab                                              \cr
\code{ALKW} \tab \code{alkalinity}            \tab                                              \cr
\code{ALKY} \tab \code{alkalinityTotal}       \tab                                              \cr
\code{ALP0} \tab \code{apha0}                 \tab                                              \cr
\code{ALTB} \tab \code{altimeter}             \tab                                              \cr
\code{ALTS} \tab \code{altitude}              \tab                                              \cr
\code{AMON} \tab \code{ammonium}              \tab                                              \cr
\code{ATMP} \tab \code{pressureAtmosphere}    \tab                                              \cr
\code{ATMS} \tab \code{pressureAtmosphereSealevel} \tab                                         \cr
\code{ATRK} \tab \code{alongTrackDisplacement} \tab                                             \cr
\code{ATTU} \tab \code{attenuation}           \tab                                              \cr
\code{AUTH} \tab \code{authority}             \tab                                              \cr
\code{BATH} \tab \code{barometricDepth}       \tab                                              \cr
\code{BATT} \tab \code{batteryVoltage}        \tab                                              \cr
\code{BEAM} \tab \code{a}                     \tab                                              \cr
\code{BNO7} \tab \code{bestNODC7Number}       \tab That is an "oh" letter, not a zero           \cr
\code{CALK} \tab \code{carbonateAlkalinity}   \tab                                              \cr
\code{CHLR} \tab \code{chlorinity}            \tab                                              \cr
\code{CHLS} \tab \code{chlorosity}            \tab                                              \cr
\code{CNDC} \tab \code{conductivity}          \tab                                              \cr
\code{CNTR} \tab \code{scan}                  \tab                                              \cr
\code{COND} \tab \code{conductivity}          \tab                                              \cr
\code{CORG} \tab \code{carbonOrganic}         \tab                                              \cr
\code{CPHL} \tab \code{chlorophyll}           \tab                                              \cr
\code{CRAT} \tab \code{conductivity}          \tab Conductivity ratio (may have spurious unit)  \cr
\code{CMNT} \tab \code{comment}               \tab                                              \cr
\code{CNDC} \tab \code{conductivity}          \tab                                              \cr
\code{COND} \tab \code{conductivity}          \tab                                              \cr
\code{CTOT} \tab \code{carbonTotal}           \tab                                              \cr
\code{DCHG} \tab \code{discharge}             \tab                                              \cr
\code{DENS} \tab \code{density}               \tab                                              \cr
\code{DEPH} \tab \code{pressure}              \tab                                              \cr
\code{DEWT} \tab \code{temperatureDewpoint}   \tab                                              \cr
\code{DOC_} \tab \code{carbonOrganicDissolved} \tab                                             \cr
\code{DON_} \tab \code{nitrogenOrganicDissolved} \tab                                           \cr
\code{DOXY} \tab \code{oxygen}                \tab                                              \cr
\code{DPDT} \tab \code{dpdt}                  \tab                                              \cr
\code{DRDP} \tab \code{drogueDepth}           \tab                                              \cr
\code{DPWT} \tab \code{dryWeight}             \tab                                              \cr
\code{DRYT} \tab \code{temperatureDryBulb}    \tab                                              \cr
\code{DYNH} \tab \code{dynamicHeight}         \tab                                              \cr
\code{ERRV} \tab \code{errorVelocity}         \tab                                              \cr
\code{EWCM} \tab \code{uMagnetic}             \tab                                              \cr
\code{EWCT} \tab \code{u}                     \tab                                              \cr
\code{FFFF} \tab \code{overall(FFFF)}         \tab Archaic overall flag, replaced by \code{QCFF}     \cr
\code{FLOR} \tab \code{fluorometer}           \tab                                              \cr
\code{GDIR} \tab \code{windDirectionGust}     \tab                                              \cr
\code{GEOP} \tab \code{geopotential}          \tab                                              \cr
\code{GSPD} \tab \code{windSpeedGust}         \tab                                              \cr
\code{HCDM} \tab \code{directionMagnetic}     \tab                                              \cr
\code{HCDT} \tab \code{directionTrue}         \tab                                              \cr
\code{HCSP} \tab \code{speedHorizontal}       \tab                                              \cr
\code{HEAD} \tab \code{heading}               \tab                                              \cr
\code{HSUL} \tab \code{hydrogenSulphide}      \tab                                              \cr
\code{IDEN} \tab \code{sampleNumber}          \tab                                              \cr
\code{LABT} \tab \code{temperatureLaboratory} \tab                                              \cr
\code{LATD} \tab \code{latitude}              \tab                                              \cr
\code{LHIS} \tab \code{lifeHistory}           \tab                                              \cr
\code{LOND} \tab \code{longitude}             \tab                                              \cr
\code{LPHT} \tab \code{pHLaboratory}          \tab                                              \cr
\code{MNSV} \tab \code{retentionFilterSize}   \tab                                              \cr
\code{MNSZ} \tab \code{organismSizeMinimum}   \tab                                              \cr
\code{MODF} \tab \code{additionalTaxonomicInformation} \tab                                     \cr
\code{MXSZ} \tab \code{organismSizeMaximum}   \tab                                              \cr
\code{NETR} \tab \code{netSolarRadiation}     \tab                                              \cr
\code{NONE} \tab \code{noWMOcode}             \tab                                              \cr
\code{NORG} \tab \code{nitrogenOrganic}       \tab                                              \cr
\code{NSCM} \tab \code{vMagnetic}             \tab                                              \cr
\code{NSCT} \tab \code{v}                     \tab                                              \cr
\code{NTOT} \tab \code{nitrogenTotal}         \tab                                              \cr
\code{NTRA} \tab \code{nitrate}               \tab                                              \cr
\code{NTRI} \tab \code{nitrite}               \tab                                              \cr
\code{NTRZ} \tab \code{nitrite+nitrate}       \tab                                              \cr
\code{NUM_} \tab \code{scansPerAverage}       \tab                                              \cr
\code{OBKS} \tab \code{turbidity}             \tab                                              \cr
\code{OCUR} \tab \code{oxygenCurrent}         \tab                                              \cr
\code{OPPR} \tab \code{oxygenPartialPressure} \tab                                              \cr
\code{OSAT} \tab \code{oxygenSaturation}      \tab                                              \cr
\code{OTMP} \tab \code{oxygenTemperature}     \tab                                              \cr
\code{OXYG} \tab \code{oxygenDissolved}       \tab                                              \cr
\code{OXYM} \tab \code{oxygenDissolved}       \tab                                              \cr
\code{OXYV} \tab \code{oxygenVoltage}         \tab                                              \cr
\code{OXV_} \tab \code{oxygenVoltageRaw}      \tab                                              \cr
\code{PCO2} \tab \code{CO2}                   \tab                                              \cr
\code{PHA_} \tab \code{phaeopigment}          \tab                                              \cr
\code{PHOS} \tab \code{phosphate}             \tab                                              \cr
\code{PHPH} \tab \code{pH}                    \tab                                              \cr
\code{PHT_} \tab \code{pHTotal}               \tab                                              \cr
\code{PIM_} \tab \code{particulateInorganicMatter} \tab                                         \cr
\code{PHY_} \tab \code{phytoplanktonCount}    \tab                                              \cr
\code{POC_} \tab \code{particulateOrganicCarbon} \tab                                           \cr
\code{POM_} \tab \code{particulateOrganicMatter} \tab                                           \cr
\code{PON_} \tab \code{particulateOrganicNitrogen} \tab                                         \cr
\code{POTM} \tab \code{theta}                 \tab                                              \cr
\code{PRES} \tab \code{pressure}              \tab                                              \cr
\code{PSAL} \tab \code{salinity}              \tab                                              \cr
\code{PSAR} \tab \code{PSAR}                  \tab                                              \cr
\code{PTCH} \tab \code{pitch}                 \tab                                              \cr
\code{QCFF} \tab \code{overall(QCFF)}         \tab Overall flag (see also archaic FFFF)         \cr
\code{RANG} \tab \code{range}                 \tab                                              \cr
\code{REFR} \tab \code{reference}             \tab                                              \cr
\code{RELH} \tab \code{humidityRelative}      \tab                                              \cr
\code{RELP} \tab \code{relativeTotalPressure} \tab                                              \cr
\code{ROLL} \tab \code{roll}                  \tab                                              \cr
\code{SDEV} \tab \code{standardDeviation}     \tab                                              \cr
\code{SECC} \tab \code{SecchiDepth}           \tab                                              \cr
\code{SEX_} \tab \code{sex}                   \tab                                              \cr
\code{SIG0} \tab \code{sigma0}                \tab                                              \cr
\code{SIGP} \tab \code{sigmaTheta}            \tab                                              \cr
\code{SIGT} \tab \code{sigmat}                \tab                                              \cr
\code{SLCA} \tab \code{silicate}              \tab                                              \cr
\code{SNCN} \tab \code{scanCounter}           \tab                                              \cr
\code{SPAR} \tab \code{SPAR}                  \tab                                              \cr
\code{SPEH} \tab \code{humiditySpecific}      \tab                                              \cr
\code{SPFR} \tab \code{sampleFraction}        \tab                                              \cr
\code{SPVO} \tab \code{specificVolume}        \tab                                              \cr
\code{SPVA} \tab \code{specificVolumeAnomaly} \tab                                              \cr
\code{STRA} \tab \code{stressAmplitude}       \tab                                              \cr
\code{STRD} \tab \code{stressDirection}       \tab                                              \cr
\code{STRU} \tab \code{stressU}               \tab                                              \cr
\code{STRV} \tab \code{stressV}               \tab                                              \cr
\code{SSAL} \tab \code{salinity}              \tab                                              \cr
\code{SVEL} \tab \code{soundVelocity}         \tab                                              \cr
\code{SYTM} \tab \code{time}                  \tab                                              \cr
\code{TAXN} \tab \code{taxonomicName}         \tab                                              \cr
\code{TE90} \tab \code{temperature}           \tab                                              \cr
\code{TEMP} \tab \code{temperature}           \tab                                              \cr
\code{TEXZT} \tab \code{text}                 \tab                                              \cr
\code{TICW} \tab \code{totalInorganicCarbon}  \tab                                              \cr
\code{TILT} \tab \code{tilt}                  \tab                                              \cr
\code{TOTP} \tab \code{pressureAbsolute}      \tab                                              \cr
\code{TPHS} \tab \code{phosphorousTotal}      \tab                                              \cr
\code{TRAN} \tab \code{lightTransmission}     \tab                                              \cr
\code{TRB_} \tab \code{turbidity}             \tab                                              \cr
\code{TRBH} \tab \code{trophicDescriptor}     \tab                                              \cr
\code{TSM_} \tab \code{suspendedMatterTotal}  \tab                                              \cr
\code{TSN_} \tab \code{taxonomicSerialNumber} \tab                                              \cr
\code{TURB} \tab \code{turbidity}             \tab                                              \cr
\code{UNKN} \tab \code{-}                     \tab                                              \cr
\code{UREA} \tab \code{urea}                  \tab                                              \cr
\code{VAIS} \tab \code{BVFrequency}           \tab                                              \cr
\code{VCSP} \tab \code{w}                     \tab                                              \cr
\code{VMXL} \tab \code{waveHeightMaximum}     \tab                                              \cr
\code{VRMS} \tab \code{waveHeightMean}        \tab                                              \cr
\code{VTCA} \tab \code{wavePeriod}            \tab                                              \cr
\code{WDIR} \tab \code{windDirection}         \tab                                              \cr
\code{WETT} \tab \code{temperatureWetBulb}    \tab                                              \cr
\code{WSPD} \tab \code{windSpeed}             \tab                                              \cr
\code{WTWT} \tab \code{wetWeight}             \tab                                              \cr
\code{ZOO_} \tab \code{zooplanktonCount}      \tab                                              \cr
}
Any code not shown in the list is transferred to the oce object without renaming, apart from
the adjustment of suffix numbers. The following code have been seen in data files from
the Bedford Institute of Oceanography: \code{ALTB}, \code{PHPH} and \code{QCFF}.
}
\examples{
ODFNames2oceNames("TEMP_01")$names # "temperature"

}
\references{
For sources that describe the ODF format, see the documentation
for the \linkS4class{odf}.
}
\seealso{
Other functions that interpret variable names and units from headers: 
\code{\link{cnvName2oceName}()},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{unitFromStringRsk}()},
\code{\link{unitFromString}()},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()}

Other things related to odf data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{ODF2oce}()},
\code{\link{ODFListFromHeader}()},
\code{\link{[[,odf-method}},
\code{\link{[[<-,odf-method}},
\code{\link{odf-class}},
\code{\link{plot,odf-method}},
\code{\link{read.ctd.odf}()},
\code{\link{read.odf}()},
\code{\link{subset,odf-method}},
\code{\link{summary,odf-method}}
}
\author{
Dan Kelley
}
\concept{functions that interpret variable names and units from headers}
\concept{things related to odf data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/coastline.R
\name{subset,coastline-method}
\alias{subset,coastline-method}
\alias{subset.coastline}
\title{Subset a Coastline Object}
\usage{
\S4method{subset}{coastline}(x, subset, ...)
}
\arguments{
\item{x}{a \linkS4class{coastline} object.}

\item{subset}{An expression indicating how to subset \code{x}. See \dQuote{Details}.}

\item{...}{optional additional arguments, the only one of which is considered
is one named \code{debug}, an integer that controls the level of debugging. If
this is not supplied, \code{debug} is assumed to be 0, meaning no debugging. If
it is 1, the steps of determining the bounding box are shown. If it is 2 or larger,
then additional processing steps are shown, including the extraction of every
polygon involved in the final result.}
}
\value{
A \code{coastline} object.
}
\description{
Subsets a coastline object according to limiting values for longitude
and latitude. This uses functions in the \CRANpkg{raster} package
for some calculations, and so it will fail unless that package is
installed.
}
\details{
As illustrated in the \dQuote{Examples}, \code{subset} must be an
expression that indicates limits on \code{latitude} and
\code{longitude}. The individual elements are provided in R notation,
not mathematical notation (i.e. \verb{30<latitude<60} would not work).
Ampersands must be used to combine the limits.  The simplest way
to understand this is to copy the example directly, and then modify
the stated limits. Note that \code{>} comparison is not permitted,
and that \code{<} is converted to \code{<=} in the calculation.
Similarly, \code{&&} is converted to \code{&}. Spaces in the
expression are ignored. For convenience, \code{longitude} and
and \code{latitude} may be abbreviated as \code{lon} and \code{lat},
as in the \dQuote{Examples}.
}
\examples{
library(oce)
data(coastlineWorld)
## Subset to a box centred on Nova Scotia, Canada
if (requireNamespace("sf")) {
    cl <- subset(coastlineWorld, -80<lon & lon<-50 & 30<lat & lat<60)
    ## The plot demonstrates that the trimming is as requested.
    plot(cl, clon=-65, clat=45, span=6000)
    rect(-80, 30, -50, 60, bg="transparent", border="red")
}
}
\seealso{
Other things related to coastline data: 
\code{\link{[[,coastline-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{as.coastline}()},
\code{\link{coastline-class}},
\code{\link{coastlineBest}()},
\code{\link{coastlineCut}()},
\code{\link{coastlineWorld}},
\code{\link{download.coastline}()},
\code{\link{plot,coastline-method}},
\code{\link{read.coastline.openstreetmap}()},
\code{\link{read.coastline.shapefile}()},
\code{\link{summary,coastline-method}}

Other functions that subset oce objects: 
\code{\link{subset,adp-method}},
\code{\link{subset,adv-method}},
\code{\link{subset,amsr-method}},
\code{\link{subset,argo-method}},
\code{\link{subset,cm-method}},
\code{\link{subset,ctd-method}},
\code{\link{subset,echosounder-method}},
\code{\link{subset,lobo-method}},
\code{\link{subset,met-method}},
\code{\link{subset,oce-method}},
\code{\link{subset,odf-method}},
\code{\link{subset,rsk-method}},
\code{\link{subset,sealevel-method}},
\code{\link{subset,section-method}},
\code{\link{subset,topo-method}},
\code{\link{subset,xbt-method}}
}
\author{
Dan Kelley
}
\concept{functions that subset oce objects}
\concept{things related to coastline data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/topo.R
\name{read.topo}
\alias{read.topo}
\title{Read a Topo File}
\usage{
read.topo(file, debug = getOption("oceDebug"))
}
\arguments{
\item{file}{Name of a file containing an ETOPO-format dataset. Three
types are permitted; see \dQuote{Details}.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\value{
A \linkS4class{topo} object that.
}
\description{
Read a file that contains topographic data in the ETOPO dataset, as was once provided by
the NOAA website (see \code{\link[=download.topo]{download.topo()}} for a good server for such
files. (As of May, 2020, there does not seem to be a way to download these
}
\details{
The three permitted file types are as follows.
\enumerate{
\item An ascii type
in which line 1 holds a label (which is ignored), whitespace, and then
the number of columns in the matrix (i.e. the number of longitude values),
line 2 is similar but for latitude, line 3 is similar but for the westernmost
longitude, line 4 is similar but for southernmost latitude, line 5
is similar but for cell size, and lines after that hold the grid.
\item A NetCDF format that was once described by NOAA as "GMT NetCDF".
\item A NetCDF format that was once described by NOAA as "NetCDF".
}
}
\examples{
\dontrun{
library(oce)
topoMaritimes <- read.topo("topoMaritimes.asc")
plot(topographyMaritimes)
}

}
\seealso{
Other things related to topo data: 
\code{\link{[[,topo-method}},
\code{\link{[[<-,topo-method}},
\code{\link{as.topo}()},
\code{\link{download.topo}()},
\code{\link{plot,topo-method}},
\code{\link{subset,topo-method}},
\code{\link{summary,topo-method}},
\code{\link{topo-class}},
\code{\link{topoInterpolate}()},
\code{\link{topoWorld}}
}
\author{
Dan Kelley
}
\concept{things related to topo data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{oce-deprecated}
\alias{oce-deprecated}
\alias{oce-defunct}
\title{Deprecated and Defunct Elements of the oce package}
\description{
Certain functions and function arguments are still provided for
compatibility with older versions of \CRANpkg{oce}, but will be removed soon.
The \CRANpkg{oce} scheme for removing functions is similar to that used by
\code{Bioconductor}: items are marked as "deprecated" in one release, marked as
"defunct" in the next, and removed in the next after that. This goal is to provide a
gentle migration path for users who keep their packages reasonably
up-to-date.
}
\details{
The following are marked "deprecated" in the present CRAN release of \CRANpkg{oce}.
Please use the replacement functions as listed below. The upcoming CRAN
release of \CRANpkg{oce} will mark these as "defunct", which is the
last step before outright removal.

\tabular{lllll}{
\strong{Deprecated}             \tab \strong{Replacement}   \tab \strong{Deprecated}  \tab \strong{Defunct} \tab \strong{Removed} \cr
\code{renameData()}             \tab \code{\link[=oceRenameData]{oceRenameData()}} \tab 1.1-2           \tab 1.1-3       \tab 1.1-4       \cr
}

The following are marked "defunct", so calling them in the
the present version produces an error message that hints at a replacement
function. Once a function is marked "defunct" on one CRAN release, it will
be slated for outright deletion in a subsequent release.

\tabular{lll}{
\strong{Defunct}         \tab \strong{Replacement}                \tab \strong{Version}\cr
\code{renameData()}      \tab \code{\link[=oceRenameData]{oceRenameData()}}              \tab 1.5        \cr
\code{byteToBinary()}    \tab \code{\link[=rawToBits]{rawToBits()}}                  \tab 1.5        \cr
}

The following functions were removed after having been marked as "deprecated"
in at least one CRAN release, and possibly as "defunct" in at least
one CRAN release.  (The version number in the table is the first
version to lack the named function.)

\tabular{lll}{
\strong{Function}        \tab \strong{Replacement}                \tab \strong{Version}\cr
\code{addColumn()}       \tab \code{\link[=oceSetData]{oceSetData()}}                 \tab 1.1-2      \cr
\code{ctdAddColumn()}    \tab \code{\link[=oceSetData]{oceSetData()}}                 \tab 1.1-2      \cr
\code{ctdUpdateHeader()} \tab \code{\link[=oceSetMetadata]{oceSetMetadata()}}             \tab 1.1-2      \cr
\code{findInOrdered()}   \tab \code{\link[=findInterval]{findInterval()}}               \tab 1.1-2      \cr
\code{makeSection()}     \tab \code{\link[=as.section]{as.section()}}                 \tab 0.9.24     \cr
\code{mapMeridians()}    \tab \code{\link[=mapGrid]{mapGrid()}}                    \tab 1.1-2      \cr
\code{mapZones()}        \tab \code{\link[=mapGrid]{mapGrid()}}                    \tab 1.1-2      \cr
\code{oce.as.POSIXlt()}  \tab \code{\link[lubridate:parse_date_time]{lubridate::parse_date_time()}} \tab 1.1-2      \cr
}

Several \CRANpkg{oce} function arguments are considered "deprecated", which
means they will be marked "defunct" in the next CRAN release. These are normally
listed in the help page for the function in question. A few that may be
of general interest are also listed below.
\itemize{
\item The \code{adorn} argument was still being checked for (in the dots argument)
until 2020 August 11.
\item The \code{eos} argument of \code{\link[=swN2]{swN2()}} was removed on 2019
April 11; for details, see the \dQuote{Deprecation Notation} section
of the documentation for \code{\link[=swN2]{swN2()}}.
\item The \code{endian} argument of \code{\link[=byteToBinary]{byteToBinary()}} will be removed sometime
in the year 2017, and should be set to \code{"big"} in the meantime.
\item The \code{parameters} argument of \code{\link[=plot,ctd-method]{plot,ctd-method()}}
was deprecated on 2016-12-30.  It was once used by
\code{\link[=plot,coastline-method]{plot,coastline-method()}} but has been ignored by that
function since February 2016.
\item The \code{orientation} argument of \code{\link[=plot,ctd-method]{plot,ctd-method()}}
was deprecated on 2016-12-30.  It was once used by
\code{\link[=plot,coastline-method]{plot,coastline-method()}} but has been ignored by that
function since February 2016.
}

Several \sQuote{oce} function arguments are considered "defunct", which
means they will be removed in the next CRAN release. They are as follows.
\itemize{
\item The \code{fill} argument of \code{\link[=mapPlot]{mapPlot()}} was confusing
to users, so it was designated as deprecated in June 2016.
(The confusion stemmed from subtle differences between
\code{\link[=plot]{plot()}} and \code{\link[=polygon]{polygon()}}, and the problem is that
\code{\link[=mapPlot]{mapPlot()}} can use either of these functions, according
to whether coastlines are to be filled.)
The functionality is preserved, in the \code{col} argument.
}
}
\seealso{
The \sQuote{Bioconductor} scheme for removing functions is
described at
\url{https://www.bioconductor.org/developers/how-to/deprecation/} and it is
extended here to function arguments.
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{decodeTime}
\alias{decodeTime}
\title{Oce Version of as.POSIXct}
\usage{
decodeTime(time, timeFormats, tz = "UTC")
}
\arguments{
\item{time}{Character string with an indication of the time.}

\item{timeFormats}{Optional vector of time formats to use, as for \code{\link[=as.POSIXct]{as.POSIXct()}}.}

\item{tz}{Time zone.}
}
\value{
A time as returned by \code{\link[=as.POSIXct]{as.POSIXct()}}.
}
\description{
Each format in \code{timeFormats} is used in turn as the \code{format}
argument to \code{\link[=as.POSIXct]{as.POSIXct()}}, and the first that produces a
non-\code{NA} result is used.  If \code{timeFormats} is missing, the
following formats are tried, in the stated order:
}
\details{
\itemize{
\item \code{"\\\%b \\\%d \\\%Y \\\%H:\\\%M:\\\%S"} (e.g. \code{"Jul 1 2013 01:02:03"})
\item \code{"\\\%b \\\%d \\\%Y"} (e.g. \code{"Jul 1 2013"})
\item \code{"\\\%B \\\%d \\\%Y \\\%H:\\\%M:\\\%S"} (e.g. \code{"July 1 2013 01:02:03"})
\item \code{"\\\%B \\\%d \\\%Y"} (e.g. \code{"July 1 2013"})
\item \code{"\\\%d \\\%b \\\%Y \\\%H:\\\%M:\\\%S"} (e.g. \code{"1 Jul 2013 01:02:03"})
\item \code{"\\\%d \\\%b \\\%Y"} (e.g. \code{"1 Jul 2013"})
\item \code{"\\\%d \\\%B \\\%Y \\\%H:\\\%M:\\\%S"} (e.g. \code{"1 July 2013 01:02:03"})
\item \code{"\\\%d \\\%B \\\%Y"} (e.g. \code{"1 July 2013"})
\item \code{"\\\%Y-\\\%m-\\\%d \\\%H:\\\%M:\\\%S"} (e.g.  \code{"2013-07-01 01:02:03"})
\item \code{"\\\%Y-\\\%m-\\\%d"} (e.g. \code{"2013-07-01"})
\item \code{"\\\%Y-\\\%b-\\\%d \\\%H:\\\%M:\\\%S"} (e.g.  \code{"2013-July-01 01:02:03"})
\item \code{"\\\%Y-\\\%b-\\\%d"} (e.g.  \code{"2013-Jul-01"})
\item \code{"\\\%Y-\\\%B-\\\%d \\\%H:\\\%M:\\\%S"} (e.g. \code{"2013-July-01 01:02:03"})
\item \code{"\\\%Y-\\\%B-\\\%d"} (e.g. \code{"2013-July-01"})
\item \code{"\\\%d-\\\%b-\\\%Y \\\%H:\\\%M:\\\%S"} (e.g.  \code{"01-Jul-2013 01:02:03"})
\item \code{"\\\%d-\\\%b-\\\%Y"} (e.g. \code{"01-Jul-2013"})
\item \code{"\\\%d-\\\%B-\\\%Y \\\%H:\\\%M:\\\%S"} (e.g. \code{"01-July-2013 01:02:03"})
\item \code{"\\\%d-\\\%B-\\\%Y"} (e.g. \code{"01-July-2013"})
\item \code{"\\\%Y/\\\%b/\\\%d \\\%H:\\\%M:\\\%S"} (e.g. \code{"2013/Jul/01 01:02:03"})
\item \code{"\\\%Y/\\\%b/\\\%d"} (e.g. \code{"2013/Jul/01"})
\item \code{"\\\%Y/\\\%B/\\\%d \\\%H:\\\%M:\\\%S"} (e.g. \code{"2013/July/01 01:02:03"})
\item \code{"\\\%Y/\\\%B/\\\%d"} (e.g. \code{"2013/July/01"})
\item \code{"\\\%Y/\\\%m/\\\%d \\\%H:\\\%M:\\\%S"} (e.g. \code{"2013/07/01 01:02:03"})
\item \code{"\\\%Y/\\\%m/\\\%d"} (e.g. \code{"2013/07/01"})
}
}
\examples{
decodeTime("July 1 2013 01:02:03")
decodeTime("Jul 1 2013 01:02:03")
decodeTime("1 July 2013 01:02:03")
decodeTime("1 Jul 2013 01:02:03")
decodeTime("2013-07-01 01:02:03")
decodeTime("2013/07/01 01:02:03")
decodeTime("2013/07/01")

}
\author{
Dan Kelley
}
\concept{functions relating to time}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/met.R
\name{[[,met-method}
\alias{[[,met-method}
\title{Extract Something From a met Object}
\usage{
\S4method{[[}{met}(x, i, j, ...)
}
\arguments{
\item{x}{a \linkS4class{met} object.}

\item{i}{character value indicating the name of an item to extract.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{ignored.}
}
\description{
Generally, the \code{[[} method lets users extract information from \code{oce}
objects, without having to know the details of the internal storage.  For
many \code{oce} sub-classes, \code{[[} can also return quantities that are computed
from the object's contents.
}
\details{
A two-step process is used to try to find the requested information. First, a
class-specific function is used (see \dQuote{Details of the Specialized
Method}).  If this yields nothing, then a general method is used (see
\dQuote{Details of the General Method}).  If both methods fail, then \code{[[}
returns NULL.

Some understanding of the subclass is required to know what can be retrieved
with \code{[[}. When dealing with an unfamiliar subclass, it can be useful to
first use \code{x[["?"]]} to get a listing of the retrievable items. See
\dQuote{Details of the Specialized Method} for more information.
}
\section{Details of the Specialized Method}{

\itemize{
\item If \code{i} is \code{"?"}, then the return value is a list
containing four items, each of which is a character vector
holding the names of things that can be accessed with \code{[[}.
The \code{data} and \code{metadata} items hold the names of
entries in the object's data and metadata
slots, respectively. The \code{dataDerived}
and \code{metadataDerived} items are each NULL, because
no derived values are defined by \linkS4class{met} objects.
}
}

\section{Details of the General Method}{


Note: the text of this section is identical for all \code{oce} subclasses, and so
some of what you read here may not be relevant to the class being described
in this help page.

If the specialized method produces no matches, the following generalized
method is applied. As with the specialized method, the procedure hinges first
on the values of \code{i} and, optionally, \code{j}. The work proceeds in steps, by
testing a sequence of possible conditions in sequence.
\enumerate{
\item A check is made as to whether \code{i} names one of the standard \code{oce} slots.
If so, \code{[[} returns the slot contents of that slot.  Thus, \code{x[["metadata"]]}
will retrieve the \code{metadata} slot, while \code{x[["data"]]} and
\code{x[["processingLog"]]} return those slots.
\item If \code{i} is a string ending in the \code{"Unit"}, then the characters preceding
that string are taken to be the name of an item in the data object, and a
list containing the unit is returned (or \code{NULL} if there is no such unit).
This list consists of an item named \code{unit}, which is an \code{\link[=expression]{expression()}}, and
an item named \code{scale}, which is a string describing the measurement scale.
If the string ends in \code{" unit"}, e.g.  \code{x[["temperature unit"]]} (note the
space), then just the expression is returned, and if it ends in \code{" scale"},
then just the scale is returned.
\item If \code{i} is a string ending in \code{"Flag"}, then the corresponding data-quality
flag is returned (or \code{NULL} if there is no such flag).
\item If the object holds hydrographic information (pressure, salinity,
temperature, longitude and latitude) then another set of possibilities
arises.  If \code{i} is \code{"sigmaTheta"}, then the value of \code{\link[=swSigmaTheta]{swSigmaTheta()}} is
called with \code{x} as the sole argument, and the results are returned.
Similarly, \code{\link[=swSigma0]{swSigma0()}} is used if \code{i="sigma0"}, and \code{\link[=swSpice]{swSpice()}} is used if
\code{i="spice"}. Of course, these actions only make sense for objects that
contain the relevant items within their \code{data} slot.
\item After these possibilities are eliminated, the action depends on whether
\code{j} has been provided.  If \code{j} is not provided, or is the string \code{""}, then
\code{i} is sought in the \code{metadata} slot, and then in the \code{data} slot, returning
whichever is found first.  In other words, if \code{j} is not provided, the
\code{metadata} slot takes preference over the \code{data} slot. However, if \code{j} is
provided, then it must be either the string \code{"metadata"} or \code{"data"}, and it
directs where to look.
\item If none of the above-listed conditions holds, then \code{NULL} is returned.
}
}

\seealso{
Other functions that extract parts of oce objects: 
\code{\link{[[,adp-method}},
\code{\link{[[,adv-method}},
\code{\link{[[,amsr-method}},
\code{\link{[[,argo-method}},
\code{\link{[[,bremen-method}},
\code{\link{[[,cm-method}},
\code{\link{[[,coastline-method}},
\code{\link{[[,ctd-method}},
\code{\link{[[,echosounder-method}},
\code{\link{[[,g1sst-method}},
\code{\link{[[,gps-method}},
\code{\link{[[,ladp-method}},
\code{\link{[[,landsat-method}},
\code{\link{[[,lisst-method}},
\code{\link{[[,lobo-method}},
\code{\link{[[,oce-method}},
\code{\link{[[,odf-method}},
\code{\link{[[,rsk-method}},
\code{\link{[[,sealevel-method}},
\code{\link{[[,section-method}},
\code{\link{[[,tidem-method}},
\code{\link{[[,topo-method}},
\code{\link{[[,windrose-method}},
\code{\link{[[,xbt-method}},
\code{\link{[[<-,adv-method}}

Other things related to met data: 
\code{\link{[[<-,met-method}},
\code{\link{as.met}()},
\code{\link{download.met}()},
\code{\link{met-class}},
\code{\link{met}},
\code{\link{plot,met-method}},
\code{\link{read.met}()},
\code{\link{subset,met-method}},
\code{\link{summary,met-method}},
\code{\link{test_met_csv1.csv}},
\code{\link{test_met_csv2.csv}},
\code{\link{test_met_xml2.xml}}
}
\author{
Dan Kelley
}
\concept{functions that extract parts of oce objects}
\concept{things related to met data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{matrixSmooth}
\alias{matrixSmooth}
\title{Smooth a Matrix}
\usage{
matrixSmooth(m, passes = 1)
}
\arguments{
\item{m}{a matrix to be smoothed.}

\item{passes}{an integer specifying the number of times the smoothing is to
be applied.}
}
\value{
A smoothed matrix.
}
\description{
The values on the edge of the matrix are unaltered.  For interior points,
the result is defined in terms in terms of the original as follows.
\eqn{r_{i,j} = (2 m_{i,j} + m_{i-1,j} + m_{i+1,j} + m_{i,j-1} +
m_{i,j+1})/6}{r_[i,j] = (2 m_[i,j] + m_[i-1,j] + m_[i+1,j] + m_[i,j-1] +
m_[i,j+1])/6}.  Note that missing values propagate to neighbours.
}
\examples{
library(oce)
opar <- par(no.readonly = TRUE)
m <- matrix(rep(seq(0, 1, length.out=5), 5), nrow=5, byrow=TRUE)
m[3, 3] <- 2
m1 <- matrixSmooth(m)
m2 <- matrixSmooth(m1)
m3 <- matrixSmooth(m2)
par(mfrow=c(2, 2))
image(m,  col=rainbow(100), zlim=c(0, 4), main="original image")
image(m1, col=rainbow(100), zlim=c(0, 4), main="smoothed 1 time")
image(m2, col=rainbow(100), zlim=c(0, 4), main="smoothed 2 times")
image(m3, col=rainbow(100), zlim=c(0, 4), main="smoothed 3 times")
par(opar)
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/section.R
\name{read.section}
\alias{read.section}
\title{Read a Section File}
\usage{
read.section(
  file,
  directory,
  sectionId = "",
  flags,
  ship = "",
  scientist = "",
  institute = "",
  missingValue = -999,
  debug = getOption("oceDebug"),
  processingLog
)
}
\arguments{
\item{file}{A file containing a set of CTD observations.  At present, only the
\emph{exchange BOT} format is accepted (see \sQuote{Details}).}

\item{directory}{A character string indicating the name of a  directory that
contains a set of CTD files that hold individual stations in the section.}

\item{sectionId}{Optional string indicating the name for the section.  If not
provided, the section ID is determined by examination of the file header.}

\item{flags}{Ignored, and deprecated (will be disallowed in a future version).}

\item{ship}{Name of the ship carrying out the sampling.}

\item{scientist}{Name of chief scientist aboard ship.}

\item{institute}{Name of chief scientist's institute.}

\item{missingValue}{Numerical value used to indicate missing data.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}

\item{processingLog}{If provided, the action item to be stored in the log.  This
is typically only provided for internal calls; the default that it provides is
better for normal calls by a user.}
}
\value{
A \linkS4class{section} object.
}
\description{
Read a file that contains a series of \code{ctd} profiles that make up an
oceanographic section.
Only \emph{exchange BOT} comma-separated value format is permitted at this time,
but other formats may be added later.  It should also be noted that the parsing
scheme was developed after inspection of the A03 data set (see Examples). This
may cause problems if the format is not universal. For example, the header must
name the salinity column "\code{CTDSAL}"; if not, salinity values will not be
read from the file.
}
\section{Disambiguating salinity}{

WOCE datasets commonly have a column named \code{CTDSAL} for salinity inferred
from a CTD and \code{SALNTY} (not a typo) for salinity derived from bottle data.
If only one of these is present in the data file, the data will be called
\code{salinity} in the \code{data} slot of the return value. However, if both
are present, then \code{CTDSAL} is stored as \code{salinity} and \code{SALNTY}
is stored as \code{salinityBottle}.
}

\references{
Several repository sites provide section data. A reasonably stable example is
\url{https://cchdo.ucsd.edu}, but a search on \code{"WOCE bottle data"} should
turn up other sites, if this ceases to exist. Only
the so-called \emph{exchange BOT} data format can be processed by \code{\link[=read.section]{read.section()}}
at this time. Data names are inferred from column headings using
\code{\link[=woceNames2oceNames]{woceNames2oceNames()}}.
}
\seealso{
Other things related to section data: 
\code{\link{[[,section-method}},
\code{\link{[[<-,section-method}},
\code{\link{as.section}()},
\code{\link{handleFlags,section-method}},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{plot,section-method}},
\code{\link{section-class}},
\code{\link{sectionAddStation}()},
\code{\link{sectionGrid}()},
\code{\link{sectionSmooth}()},
\code{\link{sectionSort}()},
\code{\link{section}},
\code{\link{subset,section-method}},
\code{\link{summary,section-method}}
}
\author{
Dan Kelley
}
\concept{things related to section data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/met.R
\name{plot,met-method}
\alias{plot,met-method}
\alias{plot.met}
\title{Plot a met Object}
\usage{
\S4method{plot}{met}(x, which = 1:4, mgp, mar, tformat, debug = getOption("oceDebug"))
}
\arguments{
\item{x}{a \linkS4class{met} object.}

\item{which}{list of desired plot types.
\itemize{
\item \code{which=1} gives a time-series plot of temperature
\item \code{which=2} gives a time-series plot of pressure
\item \code{which=3} gives a time-series plot of the x (eastward) component of velocity
\item \code{which=4} gives a time-series plot of the y (northward) component of velocity
\item \code{which=5} gives a time-series plot of speed
\item \code{which=6} gives a time-series plot of direction (degrees clockwise from north;
note that the values returned by \code{met[["direction"]]} must be multiplied by 10
to get the direction plotted)
}}

\item{mgp}{A 3-element numerical vector used with \code{\link{par}}\code{("mgp")}
to control the spacing of axis elements.  The default is tighter than the R default.}

\item{mar}{A 4-element numerical vector used with \code{\link{par}}\code{("mar")}
to control the plot margins. The default is tighter than the R default.}

\item{tformat}{optional argument passed to \code{\link[=oce.plot.ts]{oce.plot.ts()}}, for
plot types that call that function.  (See \code{\link[=strptime]{strptime()}} for the
format used.)}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\description{
Creates a multi-panel summary plot of data measured in a meteorological data
set.  cast. The panels are controlled by the \code{which} argument.
}
\details{
If more than one panel is drawn, then on exit from \code{plot.met}, the
value of \code{par} will be reset to the value it had before the function
call.  However, if only one panel is drawn, the adjustments to \code{par}
made within \code{plot.met} are left in place, so that further additions may
be made to the plot.
}
\examples{
library(oce)
data(met)
plot(met, which=3:4)

## Wind speed and direction during Hurricane Juan
## Compare with the final figure in a white paper by Chris Fogarty
## (available at http://www.novaweather.net/Hurricane_Juan_files/McNabs_plot.pdf
## downloaded 2017-01-02).
library(oce)
data(met)
t0 <- as.POSIXct("2003-09-29 04:00:00", tz="UTC")
dt <- 12 * 3600
juan <- subset(met, t0 - dt <= time & time <= t0 + dt)
par(mfrow=c(2,1))
plot(juan, which=5)
abline(v=t0)
plot(juan, which=6)
abline(v=t0)

}
\seealso{
Other functions that plot oce data: 
\code{\link{download.amsr}()},
\code{\link{plot,adp-method}},
\code{\link{plot,adv-method}},
\code{\link{plot,amsr-method}},
\code{\link{plot,argo-method}},
\code{\link{plot,bremen-method}},
\code{\link{plot,cm-method}},
\code{\link{plot,coastline-method}},
\code{\link{plot,ctd-method}},
\code{\link{plot,gps-method}},
\code{\link{plot,ladp-method}},
\code{\link{plot,landsat-method}},
\code{\link{plot,lisst-method}},
\code{\link{plot,lobo-method}},
\code{\link{plot,odf-method}},
\code{\link{plot,rsk-method}},
\code{\link{plot,satellite-method}},
\code{\link{plot,sealevel-method}},
\code{\link{plot,section-method}},
\code{\link{plot,tidem-method}},
\code{\link{plot,topo-method}},
\code{\link{plot,windrose-method}},
\code{\link{plot,xbt-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{tidem-class}}

Other things related to met data: 
\code{\link{[[,met-method}},
\code{\link{[[<-,met-method}},
\code{\link{as.met}()},
\code{\link{download.met}()},
\code{\link{met-class}},
\code{\link{met}},
\code{\link{read.met}()},
\code{\link{subset,met-method}},
\code{\link{summary,met-method}},
\code{\link{test_met_csv1.csv}},
\code{\link{test_met_csv2.csv}},
\code{\link{test_met_xml2.xml}}
}
\author{
Dan Kelley
}
\concept{functions that plot oce data}
\concept{things related to met data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{lon360}
\alias{lon360}
\title{Alter longitudes from -180:180 to 0:360 convention}
\usage{
lon360(x)
}
\arguments{
\item{x}{either a numeric vector or array, or an \linkS4class{oce} object.}
}
\description{
For numerical input, including vectors, matrices and arrays,
\code{\link[=lon360]{lon360()}} simply calls \code{\link[=ifelse]{ifelse()}} to add 360 to any negative values. For
\linkS4class{section} objects, it changes \code{longitude} in the \code{metadata} slot
and then calls itself to handle the \linkS4class{ctd} objects stored as
as the entries in \code{station} within the \code{data} slot.
For this \linkS4class{ctd} object, and indeed for all non-\linkS4class{section}
objects, \code{\link[=lon360]{lon360()}} changes \code{longitude} values in the
\code{metadata} slot (if present) and also in the \code{data} slot (again, if
present). This function is not useful for dealing with coastline
data; see \code{\link[=coastlineCut]{coastlineCut()}} for such data.
}
\examples{
lon360(c(179, -179))
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{plotSticks}
\alias{plotSticks}
\title{Draw a Stick Plot}
\usage{
plotSticks(
  x,
  y,
  u,
  v,
  yscale = 1,
  add = FALSE,
  length = 1/20,
  mgp = getOption("oceMgp"),
  mar = c(mgp[1] + 1, mgp[1] + 1, 1, 1 + par("cex")),
  xlab = "",
  ylab = "",
  col = 1,
  ...
)
}
\arguments{
\item{x}{x coordinates of stick origins.}

\item{y}{y coordinates of stick origins.  If not supplied, 0 will be used;
if length is less than that of x, the first number is repeated and the rest
are ignored.}

\item{u}{x component of stick length.}

\item{v}{y component of stick length.}

\item{yscale}{scale from u and v to y (see \dQuote{Description}).}

\item{add}{boolean, set \code{TRUE} to add to an existing plot.}

\item{length}{value to be provided to \code{\link[=arrows]{arrows()}}; here, we set a
default that is smaller than normally used, because these plots tend to be
crowded in oceanographic applications.}

\item{mgp}{3-element numerical vector to use for \code{\link{par}}\code{("mgp")}. Note
that the default \code{mar} is computed from the \code{mgp} value.
The default is tighter than the R
default, in order to use more space for the data and less for the axes.}

\item{mar}{value to be used with \code{\link{par}}\code{("mar")}.}

\item{xlab, ylab}{labels for the plot axes. The default is not to label them.}

\item{col}{color of sticks, in either numerical or character format. This is
made to have length matching that of \code{x} by a call to \code{\link[=rep]{rep()}},
which can be handy in e.g. colorizing a velocity field by day.}

\item{\dots}{graphical parameters passed down to \code{\link[=arrows]{arrows()}}.  It
is common, for example, to use smaller arrow heads than \code{\link[=arrows]{arrows()}}
uses; see \dQuote{Examples}.}
}
\description{
The arrows are drawn with directions on the graph that match the directions
indicated by the \code{u} and \code{v} components. The arrow size is set
relative to the units of the \code{y} axis, according to the value of
\code{yscale}, which has the unit of \code{v} divided by the unit of
\code{y}.
The interpretation of diagrams produced by \code{plotSticks} can be
difficult, owing to overlap in the arrows.  For this reason, it is often
a good idea to smooth \code{u} and \code{v} before using this function.
}
\examples{
library(oce)

# Flow from a point source
n <- 16
x <- rep(0, n)
y <- rep(0, n)
theta <- seq(0, 2*pi, length.out=n)
u <- sin(theta)
v <- cos(theta)
plotSticks(x, y, u, v, xlim=c(-2, 2), ylim=c(-2, 2))
rm(n, x, y, theta, u, v)

# Oceanographic example
data(met)
t <- met[["time"]]
u <- met[["u"]]
v <- met[["v"]]
p <- met[["pressure"]]
oce.plot.ts(t, p)
plotSticks(t, 99, u, v, yscale=25, add=TRUE)
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/met.R
\docType{class}
\name{met-class}
\alias{met-class}
\title{Class to Store Meteorological Data}
\description{
This class stores meteorological data. For objects created with
\code{\link[=read.met]{read.met()}}, the \code{data} slot will contain all the columns
within the original file (with some guesses as to units) in addition to
several calculated quantities such as \code{u} and \code{v}, which are
velocities in m/s (not the km/h stored in typical data files), and which
obey the oceanographic convention that \code{u>0} is a wind towards the
east.
}
\section{Slots}{

\describe{
\item{\code{data}}{As with all \code{oce} objects, the \code{data} slot
for \code{met} objects is a
\link{list} containing the main data for the object.
{}}

\item{\code{metadata}}{As with all \code{oce} objects, the \code{metadata} slot
for \code{met} objects is a \link{list} containing
information about the \code{data} or about the object itself.
{}}

\item{\code{processingLog}}{As with all \code{oce} objects, the \code{processingLog} slot
for \code{met} objects is a
\link{list} with entries describing the creation and evolution
of the object. The contents are updated by various \code{oce} functions to
keep a record of processing steps.  Object summaries and
\code{\link[=processingLogShow]{processingLogShow()}} both display the log.}
}}

\section{Modifying slot contents}{


Although the \code{[[<-} operator may permit modification of the contents
of \linkS4class{met} objects (see \code{\link{[[<-,met-method}}),
it is better to use \code{\link[=oceSetData]{oceSetData()}} and \code{\link[=oceSetMetadata]{oceSetMetadata()}},
because those functions save an entry in the \code{processingLog}
that describes the change.
}

\section{Retrieving slot contents}{


The full contents of the \code{data} and \code{metadata} slots of a \linkS4class{met}
object may be retrieved in the standard R way using \code{\link[=slot]{slot()}}. For
example \code{slot(o,"data")} returns the \code{data} slot of an object named \code{o},
and similarly \code{slot(o,"metadata")} returns
the \code{metadata} slot.

The slots may also be obtained with the \code{\link{[[,met-method}}
operator, as e.g. \code{o[["data"]]} and \code{o[["metadata"]]}, respectively.

The \code{\link{[[,met-method}} operator can also
be used to retrieve items from within the \code{data} and \code{metadata} slots.
For example, \code{o[["temperature"]]} can be used to retrieve temperature
from an object containing that quantity. The rule is that a named
quantity is sought first within the object's \code{metadata} slot,
with the \code{data} slot being checked only if \code{metadata} does not
contain the item. This \code{[[} method can also be used to get
certain derived quantities, if the object contains sufficient
information to calculate them. For example, an object that holds
(practical) salinity, temperature and pressure, along with
longitude and latitude, has sufficient information to compute
Absolute Salinity, and so \code{o[["SA"]]} will yield the
calculated Absolute Salinity.

It is also possible to find items more directly, using \code{\link[=oceGetData]{oceGetData()}} and
\code{\link[=oceGetMetadata]{oceGetMetadata()}}, but neither of these functions can
retrieve derived items.
}

\seealso{
Other classes provided by oce: 
\code{\link{adp-class}},
\code{\link{adv-class}},
\code{\link{argo-class}},
\code{\link{bremen-class}},
\code{\link{cm-class}},
\code{\link{coastline-class}},
\code{\link{ctd-class}},
\code{\link{lisst-class}},
\code{\link{lobo-class}},
\code{\link{oce-class}},
\code{\link{odf-class}},
\code{\link{rsk-class}},
\code{\link{sealevel-class}},
\code{\link{section-class}},
\code{\link{topo-class}},
\code{\link{windrose-class}},
\code{\link{xbt-class}}

Other things related to met data: 
\code{\link{[[,met-method}},
\code{\link{[[<-,met-method}},
\code{\link{as.met}()},
\code{\link{download.met}()},
\code{\link{met}},
\code{\link{plot,met-method}},
\code{\link{read.met}()},
\code{\link{subset,met-method}},
\code{\link{summary,met-method}},
\code{\link{test_met_csv1.csv}},
\code{\link{test_met_csv2.csv}},
\code{\link{test_met_xml2.xml}}
}
\author{
Dan Kelley
}
\concept{classes provided by oce}
\concept{things related to met data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/astronomy.R
\name{julianDay}
\alias{julianDay}
\title{Convert a time to a Julian day}
\usage{
julianDay(
  t,
  year = NA,
  month = NA,
  day = NA,
  hour = NA,
  min = NA,
  sec = NA,
  tz = "UTC"
)
}
\arguments{
\item{t}{a time, in POSIXt format, e.g. as created by
\code{\link[=as.POSIXct]{as.POSIXct()}}, \code{\link[=as.POSIXlt]{as.POSIXlt()}}, or
\code{\link[=numberAsPOSIXct]{numberAsPOSIXct()}}, or a character string that can be
converted to a time using \code{\link[=as.POSIXct]{as.POSIXct()}}.  If \code{t} is provided,
the other arguments are ignored.}

\item{year}{year, to be provided along with \code{month}, etc., if \code{t}
is not provided.}

\item{month}{numerical value for the month, with January being 1.
(This is required if \code{t} is not provided.)}

\item{day}{numerical value for day in month, starting at 1.
(This is required if \code{t} is not provided.)}

\item{hour}{numerical value for hour of day, in range 0 to 24.
(This is required if \code{t} is not provided.)}

\item{min}{numerical value of the minute of the hour.
(This is required if \code{t} is not provided.)}

\item{sec}{numerical value for the second of the minute.
(This is required if \code{t} is not provided.)}

\item{tz}{timezone}
}
\value{
A Julian-Day number, in astronomical convention as explained in
Meeus.
}
\description{
Convert a POSIXt time (given as either the \code{t} argument
or as the \code{year}, \code{month}, and other arguments) to a Julian day,
using the method provided in
Chapter 3 of Meeus (1982).  It should be noted that
Meeus and other astronomical treatments use fractional days, whereas the
present code follows the R convention of specifying days in whole numbers,
with hours, minutes, and seconds also provided as necessary.  Conversion is
simple, as illustrated in the example for 1977 April 26.4, for which Meeus
calculates julian day 2443259.9.  Note that the R documentation for
\code{\link[=julian]{julian()}} suggests another formula, but the point of the present
function is to match the other Meeus formulae, so that suggestion is ignored
here.
}
\examples{
library(oce)
## example from Meeus
t <- ISOdatetime(1977, 4, 26, hour=0, min=0, sec=0, tz="UTC")+0.4*86400
stopifnot(all.equal(julianDay(t), 2443259.9))

}
\references{
\itemize{
\item Meeus, Jean. Astronomical Formulas for Calculators. Second Edition.
Richmond, Virginia, USA: Willmann-Bell, 1982.
}
}
\seealso{
Other things related to astronomy: 
\code{\link{angle2hms}()},
\code{\link{eclipticalToEquatorial}()},
\code{\link{equatorialToLocalHorizontal}()},
\code{\link{julianCenturyAnomaly}()},
\code{\link{moonAngle}()},
\code{\link{siderealTime}()},
\code{\link{sunAngle}()},
\code{\link{sunDeclinationRightAscension}()}

Other things related to time: 
\code{\link{ctimeToSeconds}()},
\code{\link{julianCenturyAnomaly}()},
\code{\link{numberAsHMS}()},
\code{\link{numberAsPOSIXct}()},
\code{\link{secondsToCtime}()},
\code{\link{unabbreviateYear}()}
}
\author{
Dan Kelley
}
\concept{things related to astronomy}
\concept{things related to time}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/g1sst.R
\docType{class}
\name{g1sst-class}
\alias{g1sst-class}
\title{Class to Store G1SST Satellite-model Data}
\description{
This class stores G1SST model-satellite products.
}
\details{
G1SST is an acronym for global 1-km sea surface temperature, a product
that combines satellite data with the model output. It is provided by
the JPO ROMS (Regional Ocean Modelling System) modelling group.
See the JPL website (reference 1) to learn more about the data, and see
the \code{\link[=read.g1sst]{read.g1sst()}} documentation for an example
of downloading and plotting.

It is important not to regard G1SST data in the same category as,
say, \linkS4class{amsr} data, because the two products
differ greatly with respect to cloud cover. The satellite used by
\linkS4class{amsr} has the ability to sense water temperature
even if there is cloud cover, whereas \code{g1sst} fills in cloud
gaps with model simulations.  It can be helpful to consult
reference 1 for a given time, clicking and then unclicking the radio button
that turns off the model-based filling of cloud gaps.
}
\section{Slots}{

\describe{
\item{\code{data}}{As with all \code{oce} objects, the \code{data} slot
for \code{g1sst} objects is a
\link{list} containing the main data for the object.
{}}

\item{\code{metadata}}{As with all \code{oce} objects, the \code{metadata} slot
for \code{g1sst} objects is a \link{list} containing
information about the \code{data} or about the object itself.
{}}

\item{\code{processingLog}}{As with all \code{oce} objects, the \code{processingLog} slot
for \code{g1sst} objects is a
\link{list} with entries describing the creation and evolution
of the object. The contents are updated by various \code{oce} functions to
keep a record of processing steps.  Object summaries and
\code{\link[=processingLogShow]{processingLogShow()}} both display the log.}
}}

\section{Modifying slot contents}{


Although the \code{[[<-} operator may permit modification of the contents
of \linkS4class{g1sst} objects (see \code{\link{[[<-,g1sst-method}}),
it is better to use \code{\link[=oceSetData]{oceSetData()}} and \code{\link[=oceSetMetadata]{oceSetMetadata()}},
because those functions save an entry in the \code{processingLog}
that describes the change.
}

\section{Retrieving slot contents}{


The full contents of the \code{data} and \code{metadata} slots of a \linkS4class{g1sst}
object may be retrieved in the standard R way using \code{\link[=slot]{slot()}}. For
example \code{slot(o,"data")} returns the \code{data} slot of an object named \code{o},
and similarly \code{slot(o,"metadata")} returns
the \code{metadata} slot.

The slots may also be obtained with the \code{\link{[[,g1sst-method}}
operator, as e.g. \code{o[["data"]]} and \code{o[["metadata"]]}, respectively.

The \code{\link{[[,g1sst-method}} operator can also
be used to retrieve items from within the \code{data} and \code{metadata} slots.
For example, \code{o[["temperature"]]} can be used to retrieve temperature
from an object containing that quantity. The rule is that a named
quantity is sought first within the object's \code{metadata} slot,
with the \code{data} slot being checked only if \code{metadata} does not
contain the item. This \code{[[} method can also be used to get
certain derived quantities, if the object contains sufficient
information to calculate them. For example, an object that holds
(practical) salinity, temperature and pressure, along with
longitude and latitude, has sufficient information to compute
Absolute Salinity, and so \code{o[["SA"]]} will yield the
calculated Absolute Salinity.

It is also possible to find items more directly, using \code{\link[=oceGetData]{oceGetData()}} and
\code{\link[=oceGetMetadata]{oceGetMetadata()}}, but neither of these functions can
retrieve derived items.
}

\references{
\enumerate{
\item JPO OurOcean Portal \verb{https://ourocean.jpl.nasa.gov/SST/}
(link worked in 2016 but was seen to fail 2017 Feb 2).
}
}
\seealso{
Other classes holding satellite data: 
\code{\link{amsr-class}},
\code{\link{landsat-class}},
\code{\link{satellite-class}}

Other things related to g1sst data: 
\code{\link{[[,g1sst-method}},
\code{\link{[[<-,g1sst-method}},
\code{\link{read.g1sst}()}
}
\author{
Dan Kelley
}
\concept{classes holding satellite data}
\concept{things related to g1sst data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.sbe.R
\name{cnvName2oceName}
\alias{cnvName2oceName}
\title{Infer variable name, units and scale from a Seabird (.cnv) header line}
\usage{
cnvName2oceName(h, columns = NULL, debug = getOption("oceDebug"))
}
\arguments{
\item{h}{The header line.}

\item{columns}{Optional list containing name correspondances, as described for
\code{\link[=read.ctd.sbe]{read.ctd.sbe()}}.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\value{
\code{\link[=read.ctd.sbe]{read.ctd.sbe()}} returns a list containing \code{name} (the oce name), \code{nameOriginal} (the SBE name) and \code{unit}.
}
\description{
This function is used by \code{\link[=read.ctd.sbe]{read.ctd.sbe()}} to infer data names
and units from the coding used by Teledyne/Seabird (SBE) \code{.cnv}
files.  Lacking access to documentation on the SBE format,
the present function is based on inspection of a suite of CNV files
available to the \code{oce} developers.
}
\details{
A few sample header lines that have been encountered are:\preformatted{# name 4 = t068: temperature, IPTS-68 [deg C]
# name 3 = t090C: Temperature [ITS-90, deg C]
# name 4 = t190C: Temperature, 2 [ITS-90, deg C]
}

Examination of several CNV files suggests that it is best to
try to infer the name from the characters between the "\code{=}"
and "\code{:}" characters, because the material after the colon
seems to vary more between sample files.

The table given below indicates the translation patterns used. These are
taken from reference 1. The \code{.cnv} convention for multiple sensors is to
include optional extra digits in the name, and these are indicated
with \code{~} in the table; their decoding is done with \code{\link[=grep]{grep()}}.

It is important to note that this table is by no means complete, since there
are a great many SBE names listed in their document (reference 1), plus names
not listed there but present in data files
supplied by prominent archiving agencies. If an SBE name is not recognized,
then the oce name is set to that SBE name. This can cause problems in
some other processing steps (e.g. if \code{\link[=swRho]{swRho()}} or a similar
function is called with an \code{oce} object as first argument), and so
users are well-advised to rename the items as appropriate. The first
step in doing this is to pass the object to \code{summary()}, to discover
the SBE names in question. Then consult the SBE documentation to find
an appropriate name for the data, and either manipulate the names in the object
data slot directly or use
\code{\link[=renameData]{renameData()}} to rename the elements. Finally, please publish
an 'issue' on the oce Github site \url{https://github.com/dankelley/oce/issues}
so that the developers can add the data type in question. (To save
development time, there is no plan to add all possible data types without
a reasonable and specific expression user interest. Oxygen alone has over forty
variants.)

\tabular{llll}{
\strong{Key}       \tab \strong{Result}                     \tab \strong{Unit;scale}      \tab \strong{Notes} \cr
\code{alt}         \tab \code{altimeter}                    \tab m                   \tab    \cr
\code{altM}        \tab \code{altimeter}                    \tab m                   \tab    \cr
\code{accM}        \tab \code{acceleration}                 \tab m/s^2               \tab    \cr
\verb{bat~}        \tab \code{beamAttenuation}              \tab 1/m                 \tab    \cr
\code{C2-C1S/m}    \tab \code{conductivityDifference}       \tab S/m                 \tab    \cr
\code{C2-C1mS/cm}  \tab \code{conductivityDifference}       \tab mS/cm               \tab    \cr
\code{C2-C1uS/cm}  \tab \code{conductivityDifference}       \tab uS/cm               \tab    \cr
\code{c~mS/cm}     \tab \code{conductivity}                 \tab mS/cm               \tab    \cr
\code{cond~mS/cm}  \tab \code{conductivity}                 \tab mS/cm               \tab    \cr
\code{c~S/m}       \tab \code{conductivity}                 \tab S/m                 \tab    \cr
\code{cond~S/m}    \tab \code{conductivity}                 \tab S/m                 \tab    \cr
\code{c~uS/cm}     \tab \code{conductivity}                 \tab uS/cm               \tab    \cr
\code{cond~uS/cm}  \tab \code{conductivity}                 \tab uS/cm               \tab    \cr
\verb{CStarAt~}    \tab \code{beamAttenuation}              \tab 1/m                 \tab    \cr
\verb{CStarTr~}    \tab \code{beamTransmission}             \tab percent             \tab    \cr
\verb{density~~}   \tab \code{density}                      \tab kg/m^3              \tab    \cr
\code{depS}        \tab \code{depth}                        \tab m                   \tab    \cr
\code{depSM}       \tab \code{depth}                        \tab m                   \tab    \cr
\code{depF}        \tab \code{depth}                        \tab m                   \tab    \cr
\code{depFM}       \tab \code{depth}                        \tab m                   \tab    \cr
\code{dz/dtM}      \tab \code{descentRate}                  \tab m/s                 \tab    \cr
\verb{f~}          \tab \code{frequency}                    \tab Hz                  \tab    \cr
\verb{f~~}         \tab \code{frequency}                    \tab Hz                  \tab    \cr
\verb{flC~}        \tab \code{fluorescence}                 \tab ug/l; Chelsea Aqua 3\tab    \cr
\code{flCM}        \tab \code{fluorescence}                 \tab ug/l; Chelsea Mini Chl Con\tab\cr
\verb{flCUVA~}     \tab \code{fluorescence}                 \tab ug/l; Chelsea UV Aquatracka\tab\cr
\verb{flEC-AFL~}   \tab \code{fluorescence}                 \tab mg/m^3; WET Labs ECO-AFL/FLtab\cr
\code{flS}         \tab \code{fluorescence}                 \tab -; Seatech          \tab    \cr
\verb{flScufa~}    \tab \code{fluorescence}                 \tab -; Turner SCUFA (RFU)\tab\cr
\code{flSP}        \tab \code{fluorescence}                 \tab -; Seapoint         \tab    \cr
\code{flSPR}       \tab \code{fluorescence}                 \tab -; Seapoint, Rhodamine\tab  \cr
\code{flSPuv}      \tab \code{fluorescence}                 \tab -; Seapoint, UV      \tab   \cr
\code{flT}         \tab \code{fluorescence}                 \tab -; Turner 10-005 flT\tab   \cr
\code{gpa}         \tab \code{geopotentialAnomaly}          \tab -; J/kg              \tab   \cr
\code{latitude}    \tab \code{latitude}                     \tab degN                 \tab   \cr
\code{longitude}   \tab \code{longitude}                    \tab degE                 \tab   \cr
\code{n2satML/L}   \tab \code{nitrogenSaturation}           \tab ml/l                 \tab   \cr
\code{n2satMg/L}   \tab \code{nitrogenSaturation}           \tab mg/l                 \tab   \cr
\code{n2satumol/kg}\tab \code{nitrogenSaturation}           \tab umol/kg              \tab   \cr
\code{nbin}        \tab \code{nbin}                         \tab                      \tab   \cr
\verb{obsscufa~}   \tab \code{backscatter}                  \tab NTU; Turner SCUFA    \tab   \cr
\code{opoxMg/L}    \tab \code{oxygen}                       \tab mg/l; Optode, Aanderaa\tab   \cr
\code{opoxML/L}    \tab \code{oxygen}                       \tab ml/l; Optode, Aanderaa\tab   \cr
\code{opoxMm/L}    \tab \code{oxygen}                       \tab umol/l; Optode, Aanderaa\tab \cr
\code{opoxPS}      \tab \code{oxygen}                       \tab percent; Optode, Aanderaa\tab   \cr
\code{oxsatML/L}   \tab \code{oxygen}                       \tab ml/l; Weiss          \tab   \cr
\code{oxsatMg/L}   \tab \code{oxygen}                       \tab mg/l; Weiss          \tab   \cr
\code{oxsatMm/Kg}  \tab \code{oxygen}                       \tab umol/kg; Weiss       \tab   \cr
\code{oxsolML/L}   \tab \code{oxygen}                       \tab ml/l; Garcia-Gordon  \tab   \cr
\code{oxsolMg/L}   \tab \code{oxygen}                       \tab mg/l; Garcia-Gordon  \tab   \cr
\code{oxsolMm/Kg}  \tab \code{oxygen}                       \tab umol/kg; Garcia-Gordon\tab  \cr
\verb{par~}        \tab \code{PAR}                          \tab -; Biospherical/Licor\tab   \cr
\code{par/log}     \tab \code{PAR}                          \tab log; Satlantic       \tab   \cr
\code{ph}          \tab \code{pH}                           \tab -                    \tab   \cr
\verb{potemp~68C}  \tab \code{thetaM}                       \tab degC; IPTS-68        \tab   \cr
\verb{potemp~90C}  \tab \code{thetaM}                       \tab degC; ITS-90         \tab   \cr
\code{pr}          \tab \code{pressure}                     \tab dbar                 \tab 1 \cr
\code{prM}         \tab \code{pressure}                     \tab dbar                 \tab   \cr
\code{pr50M}       \tab \code{pressure}                     \tab dbar; SBE50          \tab   \cr
\code{prSM}        \tab \code{pressure}                     \tab dbar                 \tab   \cr
\code{prDM}        \tab \code{pressure}                     \tab dbar; digiquartz     \tab   \cr
\code{prdE}        \tab \code{pressure}                     \tab psi; strain gauge    \tab 2 \cr
\code{prDE}        \tab \code{pressure}                     \tab psi; digiquartz      \tab 2 \cr
\code{prdM}        \tab \code{pressure}                     \tab dbar; strain gauge   \tab   \cr
\code{prSM}        \tab \code{pressure}                     \tab dbar; strain gauge   \tab   \cr
\code{ptempC}      \tab \code{pressureTemperature}          \tab degC; ITS-90         \tab 3 \cr
\code{pumps}       \tab \code{pumpStatus}                   \tab                      \tab   \cr
\verb{rhodflTC~}   \tab \code{Rhodamine}                    \tab ppb; Turner Cyclops  \tab   \cr
\verb{sal~~}       \tab \code{salinity}                     \tab -, PSS-78            \tab 4 \cr
\code{sbeox~ML/L}  \tab \code{oxygen}                       \tab ml/l; SBE43          \tab   \cr
\code{sbox~ML/L}   \tab \code{oxygen}                       \tab ml/l; SBE43 (?)      \tab   \cr
\code{sbeox~Mm/Kg} \tab \code{oxygen}                       \tab umol/kg; SBE43       \tab   \cr
\code{sbox~Mm/Kg}  \tab \code{oxygen}                       \tab umol/kg; SBE43 (?)   \tab   \cr
\code{sbeox~Mm/L}  \tab \code{oxygen}                       \tab umol/l; SBE43        \tab   \cr
\code{sbox~Mm/L}   \tab \code{oxygen}                       \tab umol/l; SBE43 (?)    \tab   \cr
\code{sbeox~PS}    \tab \code{oxygen}                       \tab percent; SBE43       \tab   \cr
\code{sbox~PS}     \tab \code{oxygen}                       \tab percent; SBE43 (?)   \tab   \cr
\code{sbeox~V}     \tab \code{oxygenRaw}                    \tab V; SBE43             \tab   \cr
\code{sbox~V}      \tab \code{oxygenRaw}                    \tab V; SBE43 (?)         \tab   \cr
\code{scan}        \tab \code{scan}                         \tab -                    \tab   \cr
\verb{seaTurbMtr~} \tab \code{turbidity}                    \tab FTU; Seapoint        \tab   \cr
\code{secS-priS}   \tab \code{salinityDifference}           \tab -, PSS-78            \tab   \cr
\code{sigma-t}     \tab \code{sigmaT}                       \tab kg/m^3               \tab   \cr
\code{sigma-theta} \tab \code{sigmaTheta}                   \tab kg/m^3               \tab 5 \cr
\code{sigma-}\enc{√©}{e} \tab \code{sigmaTheta}         \tab kg/m^3               \tab 5 \cr
\code{spar}        \tab \code{spar}                         \tab -                    \tab   \cr
\code{specc}       \tab \code{conductivity}                 \tab uS/cm                \tab   \cr
\code{sva}         \tab \code{specificVolumeAnomaly}        \tab 1e-8 m^3/kg;         \tab   \cr
\verb{svCM~}       \tab \code{soundSpeed}                   \tab m/s; Chen-Millero    \tab   \cr
\verb{T2~68C}      \tab \code{temperatureDifference}        \tab degC; IPTS-68        \tab   \cr
\verb{T2~90C}      \tab \code{temperatureDifference}        \tab degC; ITS-90         \tab   \cr
\code{t~68}        \tab \code{temperature}                  \tab degC; IPTS-68        \tab   \cr
\code{t~90}        \tab \code{temperature}                  \tab degC; ITS-90         \tab   \cr
\code{t~68}        \tab \code{temperature}                  \tab degC; IPTS-68        \tab   \cr
\verb{t~68C}       \tab \code{temperature}                  \tab degC; IPTS-68        \tab   \cr
\verb{t~90C}       \tab \code{temperature}                  \tab degC; ITS-90         \tab   \cr
\code{t090Cm}      \tab \code{temperature}                  \tab degC; ITS-90         \tab   \cr
\code{t4990C}      \tab \code{temperature}                  \tab degC; ITS-90         \tab   \cr
\code{tnc90C}      \tab \code{temperature}                  \tab degC; ITS-90         \tab   \cr
\code{tsa}         \tab \code{thermostericAnomaly}          \tab 1e-8 m^3/kg          \tab   \cr
\code{tv290C}      \tab \code{temperature}                  \tab degC; ITS-90         \tab   \cr
\code{t4968C}      \tab \code{temperature}                  \tab degC; IPTS-68        \tab   \cr
\code{tnc68C}      \tab \code{temperature}                  \tab degC; IPTS-68        \tab   \cr
\code{tv268C}      \tab \code{temperature}                  \tab degC; IPTS-68        \tab   \cr
\code{t190C}       \tab \code{temperature}                  \tab degC; ITS-90         \tab   \cr
\code{tnc290C}     \tab \code{temperature}                  \tab degC; ITS-90         \tab   \cr
\code{tnc268C}     \tab \code{temperature}                  \tab degC; IPTS-68        \tab   \cr
\verb{t3890C~}     \tab \code{temperature}                  \tab degC; ITS-90         \tab   \cr
\verb{t38~90C}     \tab \code{temperature}                  \tab degC; ITS-90         \tab   \cr
\verb{t3868C~}     \tab \code{temperature}                  \tab degC; IPTS-68        \tab   \cr
\verb{t38~38C}     \tab \code{temperature}                  \tab degC; IPTS-68        \tab   \cr
\code{timeH}       \tab \code{timeH}                        \tab hour; elapsed        \tab   \cr
\code{timeJ}       \tab \code{timeJ}                        \tab julian day           \tab   \cr
\code{timeJV2}     \tab \code{timeJV2}                      \tab julian day           \tab   \cr
\code{timeK}       \tab \code{timeK}                        \tab s; since Jan 1, 2000 \tab   \cr
\code{timeM}       \tab \code{timeM}                        \tab minute; elapsed      \tab   \cr
\code{timeN}       \tab \code{timeN}                        \tab s; NMEA since Jan 1, 1970\tab\cr
\code{timeQ}       \tab \code{timeQ}                        \tab s; NMEA since Jan 1, 2000\tab\cr
\code{timeS}       \tab \code{timeS}                        \tab s; elapsed           \tab   \cr
\verb{turbflTC~}   \tab \code{turbidity}                    \tab NTU; Turner Cyclops  \tab   \cr
\code{turbflTCdiff}\tab \code{turbidityDifference}          \tab NTU; Turner Cyclops  \tab   \cr
\verb{turbWETbb~}  \tab \code{turbidity}                    \tab 1/(m\*sr); WET Labs ECO\tab   \cr
\code{turbWETbbdiff}\tab \code{turbidityDifference}         \tab 1/(m\*sr); WET Labs ECO\tab   \cr
\verb{turbWETntu~} \tab \code{turbidity}                    \tab NTU; WET Labs ECO    \tab   \cr
\code{turbWETntudiff}\tab \code{turbidityDifference}        \tab NTU; WET Labs ECO    \tab   \cr
\verb{upoly~}      \tab \code{upoly}                        \tab -                    \tab   \cr
\verb{user~}       \tab \code{user}                         \tab -                    \tab   \cr
\verb{v~~}         \tab \code{voltage}                      \tab V                    \tab   \cr
\code{wetBAttn}    \tab \code{beamAttenuation}              \tab 1/m; WET Labs AC3    \tab   \cr
\code{wetBTrans}   \tab \code{beamTransmission}             \tab percent; WET Labs AC3\tab   \cr
\verb{wetCDOM~}    \tab \code{fluorescence}                 \tab mg/m^3; WET Labs CDOM\tab   \cr
\code{wetCDOMdiff} \tab \code{fluorescenceDifference}       \tab mg/m^3; WET Labs CDOM\tab   \cr
\code{wetChAbs}    \tab \code{fluorescence}                 \tab 1/m; WET Labs AC3 absorption\tab   \cr
\verb{wetStar~}    \tab \code{fluorescence}                 \tab mg/m^3; WET Labs WETstar\tab   \cr
\code{wetStardiff} \tab \code{fluorescenceDifference}       \tab mg/m^3; WET Labs WETstar\tab   \cr
\code{xmiss}       \tab \code{beamTransmission}             \tab percent; Chelsea/Seatech\tab \cr
\verb{xmiss~}      \tab \code{beamTransmission}             \tab percent; Chelsea/Seatech\tab \cr
}
Notes:
\enumerate{
\item 'pr' is in a Dalhousie-generated data file but seems not to be in reference 1.
\item This is an odd unit, and so if \verb{sw*} functions are called on an object
containing this, a conversion will be made before performing the computation. Be
on the lookout for errors, since this is a rare situation.
\item Assume ITS-90 temperature scale, since sample \code{.cnv} file headers do not specify it.
\item Some files have PSU for this. Should we handle that? And are there other S scales to consider?
\item The 'theta' symbol (here shown accented e) may appear in different ways with different
encoding configurations, set up within R or in the operating system.
}
}
\references{
\enumerate{
\item A SBE data processing manual was once at
\verb{http://www.seabird.com/document/sbe-data-processing-manual},
but as of summer 2018, this no longer seems to be provided by SeaBird.
A web search will turn up copies of the manual that have been put
online by various research groups and data-archiving agencies.
As of 2018-07-05, the latest version was named
\code{SBEDataProcessing_7.26.4.pdf} and had release date 12/08/2017,
and this was the reference version used in coding \code{oce}.
}
}
\seealso{
Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}

Other functions that interpret variable names and units from headers: 
\code{\link{ODFNames2oceNames}()},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{unitFromStringRsk}()},
\code{\link{unitFromString}()},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()}
}
\author{
Dan Kelley
}
\concept{functions that interpret variable names and units from headers}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/AllClass.R
\name{initializeFlags}
\alias{initializeFlags}
\title{Create storage for a flag, and initialize values, for a oce object}
\usage{
initializeFlags(object, name = NULL, value = NULL, debug = 0)
}
\arguments{
\item{object}{An \linkS4class{oce} object.}

\item{name}{Character value indicating the name of a variable within the
\code{data} slot of \code{object}.}

\item{value}{Numerical or character value to be stored in the newly-created
entry within \code{flags}. (A character value will only work if
\code{\link[=initializeFlags]{initializeFlags()}} has been used first on \code{object}.)}

\item{debug}{Integer set to 0 for quiet action or to 1 for some debugging.}
}
\value{
An object with the \code{flags} item within the
\code{metadata} slot set up as indicated.
}
\description{
This function creates an item for a named variable within
the \code{flags} entry in the object's \code{metadata} slot.
The purpose is both to document a flag scheme
and to make it so that \code{\link[=initializeFlags]{initializeFlags()}}
and \code{\link[=setFlags]{setFlags()}} can specify flags by
name, in addition to number. A generic function, it is
specialized for some classes via interpretation of the
\code{scheme} argument (see \dQuote{Details}, for those
object classes that have such specializations).
}
\details{
If \code{object} already contains a \code{flags} entry with
the indicated name, then it is returned unaltered, and a warning
is issued.
}
\seealso{
Other functions relating to data-quality flags: 
\code{\link{defaultFlags}()},
\code{\link{handleFlags,adp-method}},
\code{\link{handleFlags,argo-method}},
\code{\link{handleFlags,ctd-method}},
\code{\link{handleFlags,oce-method}},
\code{\link{handleFlags,section-method}},
\code{\link{handleFlags}()},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{initializeFlagScheme,oce-method}},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{initializeFlagSchemeInternal}()},
\code{\link{initializeFlagScheme}()},
\code{\link{initializeFlags,adp-method}},
\code{\link{initializeFlags,oce-method}},
\code{\link{initializeFlagsInternal}()},
\code{\link{setFlags,adp-method}},
\code{\link{setFlags,ctd-method}},
\code{\link{setFlags,oce-method}},
\code{\link{setFlags}()}
}
\concept{functions relating to data-quality flags}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adv.R
\name{[[,adv-method}
\alias{[[,adv-method}
\title{Extract Something from an adv Object}
\usage{
\S4method{[[}{adv}(x, i, j, ...)
}
\arguments{
\item{x}{an \linkS4class{adv} object.}

\item{i}{character value indicating the name of an item to extract.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{ignored.}
}
\description{
Generally, the \code{[[} method lets users extract information from \code{oce}
objects, without having to know the details of the internal storage.  For
many \code{oce} sub-classes, \code{[[} can also return quantities that are computed
from the object's contents.
}
\details{
A two-step process is used to try to find the requested information. First, a
class-specific function is used (see \dQuote{Details of the Specialized
Method}).  If this yields nothing, then a general method is used (see
\dQuote{Details of the General Method}).  If both methods fail, then \code{[[}
returns NULL.

Some understanding of the subclass is required to know what can be retrieved
with \code{[[}. When dealing with an unfamiliar subclass, it can be useful to
first use \code{x[["?"]]} to get a listing of the retrievable items. See
\dQuote{Details of the Specialized Method} for more information.
}
\section{Details of the Specialized Method}{

\itemize{
\item If \code{i} is \code{"?"}, then the return value is a list
containing four items, each of which is a character vector
holding the names of things that can be accessed with \code{[[}.
The \code{data} and \code{metadata} items hold the names of
entries in the object's data and metadata
slots, respectively, while \code{dataDerived} and
\code{metadataDerived} hold the names of related
things that can be derived from the object's contents.
\item If \code{i} is \code{"u1"} then the return value is \code{v[,1]},
and similarly for \code{"u2"} and \code{"u3"}.
\item If \code{i} is \code{"a1"} then signal amplitude is returned, and similarly
for \code{"a2"} and \code{"a3"}. The results can be in \code{\link[=raw]{raw()}} or numeric form,
as illustrated in the \dQuote{Examples}.
\item If \code{i} is \code{"q1"} then signal quality is returned, and similarly
for \code{"q2"} and \code{"q3"}.  As with amplitude, the result can be in \code{\link[=raw]{raw()}}
or numeric form.
\item If \code{i} is \code{"heading"}, \code{"pitch"} or \code{"roll"}, then
these values are extracted from the
"slow" form in the object (e.g. in \code{headingSlow} within
the \code{data} slot). In that case, accessing by full name, e.g.
\code{x[["headingSlow"]]} retrieves the item as expected, but
\code{x[["heading"]]} interpolates to the faster timescale, using
\code{\link{approx}}\verb{(timeSlow,headingSlow,time)}.
}
}

\section{Details of the General Method}{


Note: the text of this section is identical for all \code{oce} subclasses, and so
some of what you read here may not be relevant to the class being described
in this help page.

If the specialized method produces no matches, the following generalized
method is applied. As with the specialized method, the procedure hinges first
on the values of \code{i} and, optionally, \code{j}. The work proceeds in steps, by
testing a sequence of possible conditions in sequence.
\enumerate{
\item A check is made as to whether \code{i} names one of the standard \code{oce} slots.
If so, \code{[[} returns the slot contents of that slot.  Thus, \code{x[["metadata"]]}
will retrieve the \code{metadata} slot, while \code{x[["data"]]} and
\code{x[["processingLog"]]} return those slots.
\item If \code{i} is a string ending in the \code{"Unit"}, then the characters preceding
that string are taken to be the name of an item in the data object, and a
list containing the unit is returned (or \code{NULL} if there is no such unit).
This list consists of an item named \code{unit}, which is an \code{\link[=expression]{expression()}}, and
an item named \code{scale}, which is a string describing the measurement scale.
If the string ends in \code{" unit"}, e.g.  \code{x[["temperature unit"]]} (note the
space), then just the expression is returned, and if it ends in \code{" scale"},
then just the scale is returned.
\item If \code{i} is a string ending in \code{"Flag"}, then the corresponding data-quality
flag is returned (or \code{NULL} if there is no such flag).
\item If the object holds hydrographic information (pressure, salinity,
temperature, longitude and latitude) then another set of possibilities
arises.  If \code{i} is \code{"sigmaTheta"}, then the value of \code{\link[=swSigmaTheta]{swSigmaTheta()}} is
called with \code{x} as the sole argument, and the results are returned.
Similarly, \code{\link[=swSigma0]{swSigma0()}} is used if \code{i="sigma0"}, and \code{\link[=swSpice]{swSpice()}} is used if
\code{i="spice"}. Of course, these actions only make sense for objects that
contain the relevant items within their \code{data} slot.
\item After these possibilities are eliminated, the action depends on whether
\code{j} has been provided.  If \code{j} is not provided, or is the string \code{""}, then
\code{i} is sought in the \code{metadata} slot, and then in the \code{data} slot, returning
whichever is found first.  In other words, if \code{j} is not provided, the
\code{metadata} slot takes preference over the \code{data} slot. However, if \code{j} is
provided, then it must be either the string \code{"metadata"} or \code{"data"}, and it
directs where to look.
\item If none of the above-listed conditions holds, then \code{NULL} is returned.
}
}

\examples{
data(adv)
head(adv[["q"]])            # in raw form
head(adv[["q", "numeric"]]) # in numeric form

}
\seealso{
Other functions that extract parts of oce objects: 
\code{\link{[[,adp-method}},
\code{\link{[[,amsr-method}},
\code{\link{[[,argo-method}},
\code{\link{[[,bremen-method}},
\code{\link{[[,cm-method}},
\code{\link{[[,coastline-method}},
\code{\link{[[,ctd-method}},
\code{\link{[[,echosounder-method}},
\code{\link{[[,g1sst-method}},
\code{\link{[[,gps-method}},
\code{\link{[[,ladp-method}},
\code{\link{[[,landsat-method}},
\code{\link{[[,lisst-method}},
\code{\link{[[,lobo-method}},
\code{\link{[[,met-method}},
\code{\link{[[,oce-method}},
\code{\link{[[,odf-method}},
\code{\link{[[,rsk-method}},
\code{\link{[[,sealevel-method}},
\code{\link{[[,section-method}},
\code{\link{[[,tidem-method}},
\code{\link{[[,topo-method}},
\code{\link{[[,windrose-method}},
\code{\link{[[,xbt-method}},
\code{\link{[[<-,adv-method}}

Other things related to adv data: 
\code{\link{[[<-,adv-method}},
\code{\link{adv-class}},
\code{\link{adv}},
\code{\link{beamName}()},
\code{\link{beamToXyz}()},
\code{\link{enuToOtherAdv}()},
\code{\link{enuToOther}()},
\code{\link{plot,adv-method}},
\code{\link{read.adv.nortek}()},
\code{\link{read.adv.sontek.adr}()},
\code{\link{read.adv.sontek.serial}()},
\code{\link{read.adv.sontek.text}()},
\code{\link{read.adv}()},
\code{\link{rotateAboutZ}()},
\code{\link{subset,adv-method}},
\code{\link{summary,adv-method}},
\code{\link{toEnuAdv}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdv}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley
}
\concept{functions that extract parts of oce objects}
\concept{things related to adv data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{T90fromT68}
\alias{T90fromT68}
\title{Convert from IPTS-68 to ITS-90 temperature}
\usage{
T90fromT68(temperature)
}
\arguments{
\item{temperature}{numeric vector of temperatures in \eqn{^\circ}{deg}C on the IPTS-68 scale.}
}
\value{
Corresponding temperatures in \eqn{^\circ}{deg}C on the ITS-90 scale.
}
\description{
Today's instruments typically record in the ITS-90 scale, but some old
datasets will be in the IPTS-68 scale. \code{T90fromT68()} converts from the
IPTS-68 to the ITS-90 scale, using Saunders' (1990) formula, while
\code{T68fromT90()} does the reverse. The difference between IPTS-68 and
ITS-90 values is typically a few millidegrees (see \sQuote{Examples}), which
is seldom visible on a typical temperature profile, but may be of interest
in some precise work.  Mostly for historical interest, \code{T90fromT48()}
is provided to convert from the ITS-48 system to ITS-90.
}
\examples{
library(oce)
T68 <- seq(3, 20, 1)
T90 <- T90fromT68(T68)
sqrt(mean((T68-T90)^2))

}
\references{
P. M. Saunders, 1990. The international temperature scale of
1990, ITS-90.  WOCE Newsletter, volume 10, September 1990, page 10.
http://www.nodc.noaa.gov/woce/wdiu/wocedocs/newsltr/news10/contents.htm
}
\seealso{
Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{matchBytes}
\alias{matchBytes}
\title{Locate byte sequences in a raw vector}
\usage{
matchBytes(input, b1, ...)
}
\arguments{
\item{input}{a vector of raw (byte) values.}

\item{b1}{a vector of bytes to match (must be of length 2 or 3 at present;
for 1-byte, use \code{\link[=which]{which()}}).}

\item{\dots}{additional bytes to match for (up to 2 permitted)}
}
\value{
List of the indices of \code{input} that match the start of the
\code{bytes} sequence (see example).
}
\description{
Find spots in a raw vector that match a given byte sequence.
}
\examples{
buf <- as.raw(c(0xa5, 0x11, 0xaa, 0xa5, 0x11, 0x00))
match <- matchBytes(buf, 0xa5, 0x11)
print(buf)
print(match)
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{unduplicateNames}
\alias{unduplicateNames}
\title{Rename duplicated character strings}
\usage{
unduplicateNames(strings, style = 1)
}
\arguments{
\item{strings}{Vector of character strings.}

\item{style}{An integer giving the style. If \code{style}
is \code{1}, then e.g. a triplicate of \code{"a"} yields
\code{"a"}, \code{"a1"}, and \code{"a2"}.
If \code{style} is \code{2}, then the same input yields
\code{"a_001"}, \code{"a_002"}, and \code{"a_003"}.}
}
\value{
Vector of strings with repeats distinguished by suffix.
}
\description{
Append numeric suffices to character strings, to avoid repeats.
This is used by various data
input functions, to handle the fact that several oceanographic data
formats permit the reuse of variable names within a given file.
}
\examples{
unduplicateNames(c("a", "b", "a", "c", "b"))
unduplicateNames(c("a", "b", "a", "c", "b"), style=2)
}
\seealso{
Used by \code{\link[=read.ctd.sbe]{read.ctd.sbe()}} with \code{style=1} to
rename repeated data elements (e.g. for multiple temperature sensors)
in CTD data, and by \code{\link[=read.odf]{read.odf()}} with \code{style=2} on
key-value pairs within ODF metadata.
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ad.R
\name{enuToOther}
\alias{enuToOther}
\title{Rotate acoustic-Doppler data to a new coordinate system}
\usage{
enuToOther(x, ...)
}
\arguments{
\item{x}{an \linkS4class{adp} or \linkS4class{adv} object.}

\item{\dots}{extra arguments that are passed on to \code{\link[=enuToOtherAdp]{enuToOtherAdp()}}
or \code{\link[=enuToOtherAdv]{enuToOtherAdv()}}.}
}
\value{
An object of the same class as \code{x}, but with velocities
in the rotated coordinate system
}
\description{
Rotate acoustic-Doppler data to a new coordinate system
}
\seealso{
Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}

Other things related to adv data: 
\code{\link{[[,adv-method}},
\code{\link{[[<-,adv-method}},
\code{\link{adv-class}},
\code{\link{adv}},
\code{\link{beamName}()},
\code{\link{beamToXyz}()},
\code{\link{enuToOtherAdv}()},
\code{\link{plot,adv-method}},
\code{\link{read.adv.nortek}()},
\code{\link{read.adv.sontek.adr}()},
\code{\link{read.adv.sontek.serial}()},
\code{\link{read.adv.sontek.text}()},
\code{\link{read.adv}()},
\code{\link{rotateAboutZ}()},
\code{\link{subset,adv-method}},
\code{\link{summary,adv-method}},
\code{\link{toEnuAdv}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdv}()},
\code{\link{xyzToEnu}()}
}
\concept{things related to adp data}
\concept{things related to adv data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swSigma2}
\alias{swSigma2}
\title{Seawater potential density anomaly referenced to 2000db pressure}
\usage{
swSigma2(
  salinity,
  temperature = NULL,
  pressure = NULL,
  longitude = NULL,
  latitude = NULL,
  eos = getOption("oceEOS", default = "gsw")
)
}
\arguments{
\item{salinity}{either practical salinity (in which case \code{temperature}
and \code{pressure} must be provided) \emph{or} an \code{oce} object, in
which case \code{salinity}, \code{temperature} (in the ITS-90 scale; see
next item), etc. are inferred from the object, ignoring the
other parameters, if they are supplied.}

\item{temperature}{\emph{in-situ} temperature (\eqn{^\circ}{deg}C), defined
on the ITS-90 scale.  This scale is used by GSW-style calculation (as
requested by setting \code{eos="gsw"}), and is the value contained within
\code{ctd} objects (and probably most other objects created with data
acquired in the past decade or two). Since the UNESCO-style calculation is
based on IPTS-68, the temperature is converted within the present function,
using \code{\link[=T68fromT90]{T68fromT90()}}.}

\item{pressure}{pressure (dbar)}

\item{longitude}{longitude of observation (only used if \code{eos="gsw"};
see \sQuote{Details}).}

\item{latitude}{latitude of observation (only used if \code{eos="gsw"}; see
\sQuote{Details}).}

\item{eos}{equation of state, either \code{"unesco"} (references 1 and 2)
or \code{"gsw"} (references 3 and 4).}
}
\value{
Potential density anomaly (kg/m\eqn{^3}{^3}).
}
\description{
Compute \eqn{\sigma_\theta}{sigma}, the potential density of seawater (minus
1000 kg/m\eqn{^3}{^3}), referenced to 2000db pressure.
}
\details{
Definition:
\eqn{\sigma_1=\sigma_\theta=\rho(S,\theta(S,t,p),1000}{sigma_1=sigma_theta=rho(S,theta(S,t,p),2000)}
\itemize{
\item 2000 kg/m\eqn{^3}{^3}.
}
}
\references{
See citations provided in the \code{\link[=swRho]{swRho()}} documentation.
}
\seealso{
Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ad.R
\name{beamToXyz}
\alias{beamToXyz}
\title{Change ADV or ADP coordinate systems}
\usage{
beamToXyz(x, ...)
}
\arguments{
\item{x}{an \linkS4class{adp} or \linkS4class{adv} object.}

\item{\dots}{extra arguments that are passed on to \code{\link[=beamToXyzAdp]{beamToXyzAdp()}}
or \code{\link[=beamToXyzAdv]{beamToXyzAdv()}}.}
}
\value{
An object of the same class as \code{x}, but with velocities
in xyz coordinates instead of beam coordinates.
}
\description{
Convert velocity data from an acoustic-Doppler velocimeter or
acoustic-Doppler profiler from one coordinate system to another.
}
\seealso{
Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}

Other things related to adv data: 
\code{\link{[[,adv-method}},
\code{\link{[[<-,adv-method}},
\code{\link{adv-class}},
\code{\link{adv}},
\code{\link{beamName}()},
\code{\link{enuToOtherAdv}()},
\code{\link{enuToOther}()},
\code{\link{plot,adv-method}},
\code{\link{read.adv.nortek}()},
\code{\link{read.adv.sontek.adr}()},
\code{\link{read.adv.sontek.serial}()},
\code{\link{read.adv.sontek.text}()},
\code{\link{read.adv}()},
\code{\link{rotateAboutZ}()},
\code{\link{subset,adv-method}},
\code{\link{summary,adv-method}},
\code{\link{toEnuAdv}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdv}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley
}
\concept{things related to adp data}
\concept{things related to adv data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/accessors.R
\name{oceRenameData}
\alias{oceRenameData}
\title{Rename Something in an oce data Slot}
\usage{
oceRenameData(object, old, new, note = "")
}
\arguments{
\item{object}{an \linkS4class{oce} object.}

\item{old}{character value that matches the name of an item in \code{object}'s \code{data} slot.}

\item{new}{character value to be used as the new name that matches the name of an item in
\code{object}'s \code{data} slot. Thus must not be the name of something that is already in the \code{data}
slot. If \code{new} is the  same as \code{old}, then the object is returned unaltered.}

\item{note}{character value that holds an explanation of the reason for the change. If this
is a string of non-zero length, then this is inserted in the processing log of the returned
value. If it is \code{NULL}, then no entry is added to the processing log.  Otherwise, the processing
log gets a new item that is constructed from the function call.}
}
\description{
Rename an item within the \code{data} slot of an \linkS4class{oce} object, also changing
\code{dataNamesOriginal} in the \code{metadata} slot, so that the \code{[[} accessor will
still work with the original name that was stored in the data.
}
\examples{
library(oce)
data(ctd)
CTD <- oceRenameData(ctd, "salinity", "SALT")
stopifnot(all.equal(ctd[["salinity"]], CTD[["SALT"]]))
stopifnot(all.equal(ctd[["sal00"]], CTD[["SALT"]]))

}
\seealso{
Other things related to the data slot: 
\code{\link{oceDeleteData}()},
\code{\link{oceGetData}()},
\code{\link{oceSetData}()}
}
\author{
Dan Kelley
}
\concept{things related to the data slot}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/geod.R
\name{geodGc}
\alias{geodGc}
\title{Great-circle Segments Between Points on Earth}
\usage{
geodGc(longitude, latitude, dmax)
}
\arguments{
\item{longitude}{vector of longitudes, in degrees east}

\item{latitude}{vector of latitudes, in degrees north}

\item{dmax}{maximum angular separation to tolerate between sub-segments, in
degrees.}
}
\value{
Data frame of \code{longitude} and \code{latitude}.
}
\description{
Each pair in the \code{longitude} and \code{latitude} vectors is considered
in turn.  For long vectors, this may be slow.
}
\examples{
\donttest{
library(oce)
data(coastlineWorld)
mapPlot(coastlineWorld, type="l",
        longitudelim=c(-80,10), latitudelim=c(35,80),
        projection="+proj=merc")
## Great circle from New York to Paris (Lindberg's flight)
l <- geodGc(c(-73.94,2.35), c(40.67,48.86), 1)
mapLines(l$longitude, l$latitude, col='red', lwd=2)
}

}
\references{
\enumerate{
\item \verb{http://williams.best.vwh.net/avform.htm#Intermediate}
(link worked for years but failed 2017-01-16).
}
}
\seealso{
Other functions relating to geodesy: 
\code{\link{geodDist}()},
\code{\link{geodXyInverse}()},
\code{\link{geodXy}()}
}
\author{
Dan Kelley, based on code from Clark Richards, in turn based on
formulae provided by Ed Williams (see reference 1)].
}
\concept{functions relating to geodesy}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{oceColorsGebco}
\alias{oceColorsGebco}
\alias{oce.colorsGebco}
\title{Create colors in a Gebco-like scheme}
\usage{
oceColorsGebco(
  n = 9,
  region = c("water", "land", "both"),
  type = c("fill", "line"),
  debug = getOption("oceDebug")
)
}
\arguments{
\item{n}{Number of colors to return}

\item{region}{String indicating application region,
one of \code{"water"}, \code{"land"}, or \code{"both"}.}

\item{type}{String indicating the purpose, one of \code{"fill"} or \code{"line"}.}

\item{debug}{a flag that turns on debugging.}
}
\description{
The colours were determined by examination of paper
charts printed during the GEBCO Fifth Edition era.
The hues range from dark blue to light blue, then
from light brown to dark brown.  If used to show
topography in scheme centred on z=0, this means that
near-coastal regions are light in tone, with darker
colours representing both mountains and the deep sea.
}
\examples{
library(oce)
imagep(volcano, col=oceColorsGebco(128, region="both"))

}
\seealso{
Other things related to colors: 
\code{\link{colormapGMT}()},
\code{\link{colormap}()},
\code{\link{oceColors9B}()},
\code{\link{oceColorsCDOM}()},
\code{\link{oceColorsChlorophyll}()},
\code{\link{oceColorsClosure}()},
\code{\link{oceColorsDensity}()},
\code{\link{oceColorsFreesurface}()},
\code{\link{oceColorsJet}()},
\code{\link{oceColorsOxygen}()},
\code{\link{oceColorsPAR}()},
\code{\link{oceColorsPalette}()},
\code{\link{oceColorsPhase}()},
\code{\link{oceColorsSalinity}()},
\code{\link{oceColorsTemperature}()},
\code{\link{oceColorsTurbidity}()},
\code{\link{oceColorsTurbo}()},
\code{\link{oceColorsTwo}()},
\code{\link{oceColorsVelocity}()},
\code{\link{oceColorsViridis}()},
\code{\link{oceColorsVorticity}()},
\code{\link{ocecolors}}
}
\concept{things related to colors}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{locationForGsw}
\alias{locationForGsw}
\title{Reform longitude and latitude for use in gsw computations.}
\usage{
locationForGsw(x)
}
\arguments{
\item{x}{an \linkS4class{oce} object.}
}
\value{
A list containing \code{longitude} and \code{latitude}, with
dimensionality matching \code{pressure} in the \code{data} slot
of \code{x}.  If \code{x} lacks location information (in either
its \code{metadata} or \code{data} slot) or lacks \code{pressure} in its
data slot, then the returned list will hold NULL values for
both \code{longitude} and \code{latitude}.
}
\description{
This function is mainly intended for use within \code{\link[=swAbsoluteSalinity]{swAbsoluteSalinity()}} and
similar functions that use the \code{gsw} package to compute seawater
properties in the Gibbs Seawater formulation.
}
\details{
The \code{gsw} functions
require location information to be matched up with hydrographic
information.  The scheme depends on the dimensionality of the
hydrographic variables and the location variables. For example,
the \linkS4class{ctd} stores \code{salinity} etc in vectors, an stores
just one longitude-latitude pair for each vector.  By contrast,
the \linkS4class{argo} stores \code{salinity} etc as matrices, and stores
e.g. \code{longitude} as a vector of length matching the first
dimension of \code{salinity}.

\link{locationForGsw} repeats location information as required,
returning values with dimensionality matching \code{pressure}.
}
\seealso{
Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/xbt.R
\name{[[,xbt-method}
\alias{[[,xbt-method}
\title{Extract Something From an xbt Object}
\usage{
\S4method{[[}{xbt}(x, i, j, ...)
}
\arguments{
\item{x}{an \linkS4class{xbt} object.}

\item{i}{character value indicating the name of an item to extract.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{ignored.}
}
\description{
Generally, the \code{[[} method lets users extract information from \code{oce}
objects, without having to know the details of the internal storage.  For
many \code{oce} sub-classes, \code{[[} can also return quantities that are computed
from the object's contents.
}
\details{
A two-step process is used to try to find the requested information. First, a
class-specific function is used (see \dQuote{Details of the Specialized
Method}).  If this yields nothing, then a general method is used (see
\dQuote{Details of the General Method}).  If both methods fail, then \code{[[}
returns NULL.

Some understanding of the subclass is required to know what can be retrieved
with \code{[[}. When dealing with an unfamiliar subclass, it can be useful to
first use \code{x[["?"]]} to get a listing of the retrievable items. See
\dQuote{Details of the Specialized Method} for more information.
}
\section{Details of the Specialized Method}{

\itemize{
\item If \code{i} is \code{"?"}, then the return value is a list
containing four items, each of which is a character vector
holding the names of things that can be accessed with \code{[[}.
The \code{data} and \code{metadata} items hold the names of
entries in the object's data and metadata
slots, respectively. The \code{dataDerived}
and \code{metadataDerived} items are each NULL, because
no derived values are defined by \code{cm} objects.
}
}

\section{Details of the General Method}{


Note: the text of this section is identical for all \code{oce} subclasses, and so
some of what you read here may not be relevant to the class being described
in this help page.

If the specialized method produces no matches, the following generalized
method is applied. As with the specialized method, the procedure hinges first
on the values of \code{i} and, optionally, \code{j}. The work proceeds in steps, by
testing a sequence of possible conditions in sequence.
\enumerate{
\item A check is made as to whether \code{i} names one of the standard \code{oce} slots.
If so, \code{[[} returns the slot contents of that slot.  Thus, \code{x[["metadata"]]}
will retrieve the \code{metadata} slot, while \code{x[["data"]]} and
\code{x[["processingLog"]]} return those slots.
\item If \code{i} is a string ending in the \code{"Unit"}, then the characters preceding
that string are taken to be the name of an item in the data object, and a
list containing the unit is returned (or \code{NULL} if there is no such unit).
This list consists of an item named \code{unit}, which is an \code{\link[=expression]{expression()}}, and
an item named \code{scale}, which is a string describing the measurement scale.
If the string ends in \code{" unit"}, e.g.  \code{x[["temperature unit"]]} (note the
space), then just the expression is returned, and if it ends in \code{" scale"},
then just the scale is returned.
\item If \code{i} is a string ending in \code{"Flag"}, then the corresponding data-quality
flag is returned (or \code{NULL} if there is no such flag).
\item If the object holds hydrographic information (pressure, salinity,
temperature, longitude and latitude) then another set of possibilities
arises.  If \code{i} is \code{"sigmaTheta"}, then the value of \code{\link[=swSigmaTheta]{swSigmaTheta()}} is
called with \code{x} as the sole argument, and the results are returned.
Similarly, \code{\link[=swSigma0]{swSigma0()}} is used if \code{i="sigma0"}, and \code{\link[=swSpice]{swSpice()}} is used if
\code{i="spice"}. Of course, these actions only make sense for objects that
contain the relevant items within their \code{data} slot.
\item After these possibilities are eliminated, the action depends on whether
\code{j} has been provided.  If \code{j} is not provided, or is the string \code{""}, then
\code{i} is sought in the \code{metadata} slot, and then in the \code{data} slot, returning
whichever is found first.  In other words, if \code{j} is not provided, the
\code{metadata} slot takes preference over the \code{data} slot. However, if \code{j} is
provided, then it must be either the string \code{"metadata"} or \code{"data"}, and it
directs where to look.
\item If none of the above-listed conditions holds, then \code{NULL} is returned.
}
}

\seealso{
Other functions that extract parts of oce objects: 
\code{\link{[[,adp-method}},
\code{\link{[[,adv-method}},
\code{\link{[[,amsr-method}},
\code{\link{[[,argo-method}},
\code{\link{[[,bremen-method}},
\code{\link{[[,cm-method}},
\code{\link{[[,coastline-method}},
\code{\link{[[,ctd-method}},
\code{\link{[[,echosounder-method}},
\code{\link{[[,g1sst-method}},
\code{\link{[[,gps-method}},
\code{\link{[[,ladp-method}},
\code{\link{[[,landsat-method}},
\code{\link{[[,lisst-method}},
\code{\link{[[,lobo-method}},
\code{\link{[[,met-method}},
\code{\link{[[,oce-method}},
\code{\link{[[,odf-method}},
\code{\link{[[,rsk-method}},
\code{\link{[[,sealevel-method}},
\code{\link{[[,section-method}},
\code{\link{[[,tidem-method}},
\code{\link{[[,topo-method}},
\code{\link{[[,windrose-method}},
\code{\link{[[<-,adv-method}}

Other things related to xbt data: 
\code{\link{[[<-,xbt-method}},
\code{\link{as.xbt}()},
\code{\link{plot,xbt-method}},
\code{\link{read.xbt.noaa1}()},
\code{\link{read.xbt}()},
\code{\link{subset,xbt-method}},
\code{\link{summary,xbt-method}},
\code{\link{xbt-class}},
\code{\link{xbt.edf}},
\code{\link{xbt}}
}
\author{
Dan Kelley
}
\concept{functions that extract parts of oce objects}
\concept{things related to xbt data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/map.R
\name{oceCRS}
\alias{oceCRS}
\title{Coordinate Reference System strings for some oceans}
\usage{
oceCRS(region)
}
\arguments{
\item{region}{character string indicating the region. This must be
in the following list (or a string that matches to just one entry,
with \code{\link[=pmatch]{pmatch()}}):
\code{"North Atlantic"}, \code{"South Atlantic"}, \code{"Atlantic"},
\code{"North Pacific"}, \code{"South Pacific"}, \code{"Pacific"},
\code{"Arctic"},  and \code{"Antarctic"}.}
}
\value{
string contain a CRS, which can be used as \code{projection}
in \code{\link[=mapPlot]{mapPlot()}}.
}
\description{
Create a coordinate reference string (CRS), suitable for use as a
\code{projection} argument to \code{\link[=mapPlot]{mapPlot()}} or
\code{\link[=plot,coastline-method]{plot,coastline-method()}}.
}
\section{Caution}{
 This is a preliminary version of this function,
with the results being very likely to change through the autumn of 2016,
guided by real-world usage.
}

\examples{
\donttest{
library(oce)
data(coastlineWorld)
par(mar=c(2, 2, 1, 1))
plot(coastlineWorld, projection=oceCRS("Atlantic"), span=12000)
plot(coastlineWorld, projection=oceCRS("North Atlantic"), span=8000)
plot(coastlineWorld, projection=oceCRS("South Atlantic"), span=8000)
plot(coastlineWorld, projection=oceCRS("Arctic"), span=4000)
plot(coastlineWorld, projection=oceCRS("Antarctic"), span=10000)
# Avoid ugly horizontal lines, an artifact of longitude shifting.
# Note: we cannot fill the land once we shift, either.
pacific <- coastlineCut(coastlineWorld, -180)
plot(pacific, proj=oceCRS("Pacific"), span=15000, col=NULL)
plot(pacific, proj=oceCRS("North Pacific"), span=12000, col=NULL)
plot(pacific, proj=oceCRS("South Pacific"), span=12000, col=NULL)
}
}
\seealso{
Other functions related to maps: 
\code{\link{formatPosition}()},
\code{\link{lonlat2map}()},
\code{\link{lonlat2utm}()},
\code{\link{map2lonlat}()},
\code{\link{mapArrows}()},
\code{\link{mapAxis}()},
\code{\link{mapContour}()},
\code{\link{mapCoordinateSystem}()},
\code{\link{mapDirectionField}()},
\code{\link{mapGrid}()},
\code{\link{mapImage}()},
\code{\link{mapLines}()},
\code{\link{mapLocator}()},
\code{\link{mapLongitudeLatitudeXY}()},
\code{\link{mapPlot}()},
\code{\link{mapPoints}()},
\code{\link{mapPolygon}()},
\code{\link{mapScalebar}()},
\code{\link{mapText}()},
\code{\link{mapTissot}()},
\code{\link{shiftLongitude}()},
\code{\link{usrLonLat}()},
\code{\link{utm2lonlat}()}
}
\author{
Dan Kelley
}
\concept{functions related to maps}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.sbe.R
\name{read.ctd.sbe}
\alias{read.ctd.sbe}
\title{Read a Seabird CTD File}
\usage{
read.ctd.sbe(
  file,
  columns = NULL,
  station = NULL,
  missingValue,
  deploymentType = "unknown",
  btl = FALSE,
  monitor = FALSE,
  debug = getOption("oceDebug"),
  processingLog,
  ...
)
}
\arguments{
\item{file}{a connection or a character string giving the name of the file to
load.  For \code{\link[=read.ctd.sbe]{read.ctd.sbe()}} and \code{\link[=read.ctd.woce]{read.ctd.woce()}}, this may be a
wildcard (e.g. \code{"*.cnv"} or \code{"*.csv"}) in which case the return
value is a vector containing CTD objects created by reading the files from
\code{\link[=list.files]{list.files()}} with \code{pattern} set to the specified wildcard
pattern.}

\item{columns}{an optional \link{list} that can be used to convert unrecognized
data names to resultant variable names.  This is used only by
\code{\link[=read.ctd.sbe]{read.ctd.sbe()}} and \code{\link[=read.ctd.odf]{read.ctd.odf()}}. For example,
if a data file named salinity as \code{"SAL"}, then using\preformatted{d <- read.ctd(f, columns=list(
    salinity=list(name="SAL",
                  unit=list(unit=expression(),
                  scale="PSS-78"))))
}

would assign the \code{"SAL"} column to the \code{salinity} entry in the data
slot of the CTD object returned by the \verb{read.*} function.}

\item{station}{optional character string containing an identifying name or
number for the station. This can be useful if the routine cannot determine the
name automatically, or if another name is preferred.}

\item{missingValue}{optional missing-value flag; data matching this value will
be set to \code{NA} upon reading. If this is provided, then it overrules any
missing-value flag found in the data. For Seabird (\code{.cnv}) files, there is
usually no need to set \code{missingValue}, because it can be inferred from the
header (typically as -9.990e-29). Set \code{missingValue=NULL} to turn off
missing-value detection, even in \code{.cnv} files that contain missing-value
codes in their headers. If \code{missingValue} is not specified,
then an attempt is made to infer such a value from the data, by testing
whether salinity and/or temperature has a minimum that is under -8 in value;
this should catch common values in files, without false positives. A warning
will be issued in this case, and a note inserted in the processing log of
the return value.}

\item{deploymentType}{character string indicating the type of deployment. Use
\code{"unknown"} if this is not known, \code{"profile"} for a profile (in
which the data were acquired during a downcast, while the device was lowered
into the water column, perhaps also including an upcast; \code{"moored"} if
the device is installed on a fixed mooring, \code{"thermosalinograph"} (or
\code{"tsg"}) if the device is mounted on a moving vessel, to record
near-surface properties, or \code{"towyo"} if the device is repeatedly
lowered and raised.}

\item{btl}{a logical value, with \code{TRUE} indicating that this is a \code{.BTL} file and \code{FALSE}
(the default) indicating a \code{.CNV} file.  Note that if \code{btl} is \code{TRUE}, the data column
names are taken directly from the file (without e.g. translating to \code{"Sal00"}
to \code{"salinity"}.  Also, the "avg" and "sdev" columns are blended together, with
all the latter named as in the file, but with \code{"_sdev"} appended.}

\item{monitor}{boolean, set to \code{TRUE} to provide an indication of
progress.  This is useful if \code{filename} is a wildcard.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed.}

\item{processingLog}{if provided, the action item to be stored in the log.
This is typically only provided for internal calls; the default that it provides is
better for normal calls by a user.}

\item{...}{additional arguments, passed to called routines.}
}
\value{
a \linkS4class{ctd} object. The details of the contents
depend on the source file. The \code{metadata} slot is particularly
variable across data formats, because the meta-information provided
in those formats varies so widely.
}
\description{
Read a Seabird CTD File
}
\details{
This function reads files stored in Seabird \code{.cnv} format.
Note that these files can contain multiple sensors for a given field. For example,
the file might contain a column named \code{t090C} for one
temperature sensor and \code{t190C} for a second. The first will be denoted
\code{temperature} in the \code{data} slot of the return value, and the second
will be denoted \code{temperature1}. This means that the first sensor
will be used in any future processing that accesses \code{temperature}. This
is for convenience of processing, and it does not pose a limitation, because the
data from the second sensor are also available as e.g. \code{x[["temperature1"]]},
where \code{x} is the name of the returned value.  For the details of the
mapping from \code{.cnv} names to \code{ctd} names, see \code{\link[=cnvName2oceName]{cnvName2oceName()}}.

The names of the elements in the \code{data} slot of the returned value depend on
the file type, as signalled by the \code{btl} argument.  For the default case of \code{.cnv} files,
the original data names as stored in \code{file} are stored within the \code{metadata}
slot as \code{dataNamesOriginal}, and are displayed with \code{summary} alongside the
numerical summary; see the Appendix VI of reference 2 for the meanings of these
names (in the "Short Name" column of the table spanning pages 161 through 172).
However, for the case of \code{.btl} files, the column names are as described
in the documentation entry for the \code{btl} argument.
}
\section{A note on sampling times}{

Until November of 2018,
there was a possibility for great confusion in the storage
of the time entries within the \code{data} slot, because \code{read.ctd.sbe}
renamed each of the ten variants of time (see reference 2 for a list)
as \code{"time"} in the \code{data} slot of the returned value.
For CTD profiles, this was perhaps not a great problem, but it could
lead to great confusion for moored data. Therefore, a change to \code{read.ctd.sbe} was
made, so that it would Seabird times, using the \code{start_time} entry in
the CNV file header (which is stored as \code{startTime} in the object
\code{metadata} slot), along with specific time columns as follows
(and as documented, with uneven clarity, in the
SBE Seasoft data processing manual, revision 7.26.8, Appendix VI):
\tabular{rl}{
\strong{Item} \tab \strong{Meaning}\cr
\code{timeS}   \tab seconds elapsed since \code{start_time}\cr
\code{timeM}   \tab minutes elapsed since \code{start_time}\cr
\code{timeH}   \tab hours elapsed since \code{start_time}\cr
\code{timeJ}   \tab Julian days since the start of the year of the first observation\cr
\code{timeN}   \tab NMEA-based time, in seconds past Jan 1, 1970\cr
\code{timeQ}   \tab NMEA-based time, in seconds past Jan 1, 2000\cr
\code{timeK}   \tab NMEA-based time, in seconds past Jan 1, 2000\cr
\code{timeJV2} \tab as \code{timeJ}\cr
\code{timeSCP} \tab as \code{timeJ}\cr
\code{timeY}   \tab computer time, in seconds past Jan 1, 1970\cr
}
NOTE: not all of these times have been tested properly, and so users
are asked to report incorrect times, so that \code{read.ctd.sbe} can
be improved.
}

\section{A note on scales}{

The user might encounter data files with a variety of scales for temperature and
salinity. Oce keeps track of these scales in the units it sets up for the stored
variables. For example, if \code{A} is a CTD object, then
\code{A[["temperatureUnit"]]$scale} is a character string that will indicate the scale.
Modern-day data will have \code{"ITS-90"} for that scale, and old data may have
\code{"IPTS-68"}. The point of saving the scale in this way is so that the various
formulas that deal with water properties can account for the scale, e.g. converting
from numerical values saved on the \code{"IPTS-68"} scale to the newer scale, using
\code{\link[=T90fromT68]{T90fromT68()}} before doing calculations that are expressed in
terms of the \code{"ITS-90"} scale. This is taken care of by retrieving temperatures
with the accessor function, e.g. writing \code{A[["temperature"]]} will either
retrieve the stored values (if the scale is ITS-90) or converted values (if
the scale is IPTS-68). Even though this procedure should work, users who
really care about the details of their data are well-advised to do a couple
of tests after examining the first data line of their data file in an editor.
Note that reading a file that contains IPTS-68 temperatures produces a warning.
}

\examples{
f <- system.file("extdata", "ctd.cnv", package="oce")
# Read the file in the normal way
d <- read.ctd(f)

}
\references{
\enumerate{
\item The Sea-Bird SBE 19plus profiler is described at
\verb{http://www.seabird.com/products/spec_sheets/19plusdata.htm}.  Some more
information is given in the Sea-Bird data-processing manual
(next item).
\item A SBE data processing manual was once at
\verb{http://www.seabird.com/document/sbe-data-processing-manual},
but as of summer 2018, this no longer seems to be provided by SeaBird.
A web search will turn up copies of the manual that have been put
online by various research groups and data-archiving agencies.
As of 2018-07-05, the latest version was named
\code{SBEDataProcessing_7.26.4.pdf} and had release date 12/08/2017,
and this was the reference version used in coding \code{oce}.
}
}
\seealso{
Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}

Other functions that read ctd data: 
\code{\link{read.ctd.aml}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()}
}
\author{
Dan Kelley and Clark Richards
}
\concept{functions that read ctd data}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adv.sontek.R
\name{read.adv.sontek.text}
\alias{read.adv.sontek.text}
\title{Read an ADV data file}
\usage{
read.adv.sontek.text(
  file,
  from = 1,
  to,
  by = 1,
  tz = getOption("oceTz"),
  originalCoordinate = "xyz",
  transformationMatrix,
  longitude = NA,
  latitude = NA,
  debug = getOption("oceDebug"),
  monitor = FALSE,
  processingLog = NULL
)
}
\arguments{
\item{file}{a connection or a character string giving the name of the file
to load.  It is also possible to give \code{file} as a vector of filenames,
to handle the case of data split into files by a data logger.  In the
multi-file case, \code{header} must be \code{FALSE}, \code{start} must be a
vector of times, and \code{deltat} must be provided.}

\item{from}{index number of the first profile to be read, or the time of
that profile, as created with \code{\link[=as.POSIXct]{as.POSIXct()}} (hint: use
\code{tz="UTC"}).  This argument is ignored if \code{header==FALSE}. See
\dQuote{Examples}.}

\item{to}{indication of the last profile to read, in a format matching that
of \code{from}.  This is ignored if \code{header==FALSE}.}

\item{by}{an indication of the stride length to use while walking through
the file.  This is ignored if \code{header==FALSE}.  Otherwise, if this is
an integer, then \code{by-1} profiles are skipped between each pair of
profiles that is read. This may not make much sense, if the data are not
equi-spaced in time.  If \code{by} is a string representing a time interval,
in colon-separated format, then this interval is divided by the sampling
interval, to get the stride length. \emph{BUG:} \code{by} only partially
works; see \dQuote{Bugs}.}

\item{tz}{character string indicating time zone to be assumed in the data.}

\item{originalCoordinate}{character string indicating coordinate system, one
of \code{"beam"}, \code{"xyz"}, \code{"enu"} or \code{"other"}.  (This is
needed for the case of multiple files that were created by a data logger,
because the header information is normally lost in such instances.)}

\item{transformationMatrix}{transformation matrix to use in converting beam
coordinates to xyz coordinates.  This will over-ride the matrix in the file
header, if there is one.  An example is \code{rbind(c(2.710, -1.409,
-1.299), c(0.071, 2.372, -2.442), c(0.344, 0.344, 0.344))}.}

\item{longitude}{optional signed number indicating the longitude in degrees
East.}

\item{latitude}{optional signed number indicating the latitude in degrees
North.}

\item{debug}{a flag that turns on debugging.  The value indicates the depth
within the call stack to which debugging applies.  For example,
\code{read.adv.nortek()} calls \code{read.header.nortek()}, so that
\code{read.adv.nortek(...,debug=2)} provides information about not just the
main body of the data file, but also the details of the header.}

\item{monitor}{boolean value indicating whether to indicate the progress
of reading the file, by using \code{\link[=txtProgressBar]{txtProgressBar()}} or otherwise.  The value
of \code{monitor} is changed to \code{FALSE} automatically, for non-interactive
sessions.}

\item{processingLog}{if provided, the action item to be stored in the log.
This parameter is typically only provided for internal calls; the default
that it provides is better for normal calls by a user.}
}
\value{
An \linkS4class{adv} object that contains measurements made with an ADV device.

The \code{metadata} contains information as given in the following table.
The \verb{Nortek name'' is the name used in the Nortek System Integrator Guide (reference 1) and the }Sontek name'' is the name used in the relevant
Sontek documentation.  References are given in square brackets.

\tabular{llll}{

\strong{\code{metadata} name}\tab \strong{Nortek name} \tab \strong{Sontek name} \tab \strong{Meaning}\cr

\code{manufacturer}\tab - \tab - \tab Either \code{"nortek"} or \code{"sontek"}\cr

\code{instrumentType}\tab  - \tab - \tab Either \code{"vector"} or \code{"adv"}\cr

\code{filename}\tab  - \tab - \tab Name of data file(s)\cr

\code{latitude}\tab - \tab - \tab  Latitude of mooring (if applicable)\cr

\code{longitude}\tab - \tab - \tab Longitude of mooring (if applicable)\cr

\code{numberOfSamples}\tab - \tab - \tab Number of data samples in file\cr

\code{numberOfBeams}\tab NBeams (reference 1, p18) \tab - \tab Number of beams (always 3)\cr

\code{numberOfBeamSequencesPerBurst}\tab NPings \tab - \tab number of beam sequences per burst\cr

\code{measurementInterval}\tab MeasInterval (reference 1 p31) \tab - \tab \cr

\code{samplingRate}\tab 512/(AvgInterval) (reference 1 p30; reference 4) \tab - #' \tab \cr

}

The \code{data} list contains items with names corresponding to \code{adp}
objects, with an exception for Nortek data.  Nortek instruments report some
things at a time interval that is longer than the velocity sampling, and
these are stored in \code{data} as \code{timeSlow}, \code{headingSlow},
\code{pitchSlow}, \code{rollSlow}, and \code{temperatureSlow}; if burst
sampling was used, there will also be items \code{recordsBurst} and
\code{timeBurst}.

The \code{processingLog} is in the standard format.
}
\description{
Read an ADV data file, producing an object of type \code{adv}. This
function works by transferring control to a more specialized function,
e.g. \code{\link[=read.adp.nortek]{read.adp.nortek()}} and \code{\link[=read.adp.sontek]{read.adp.sontek()}},
and in many cases users will find it preferable to either use these or
the several even more specialized functions, if the file type is
known.
}
\details{
Files \emph{without} headers may be created in experiments in which a data
logger was set up to monitor the serial data stream from an instrument.  The
lack of header information places a burden on the user, who must supply such
basic information as the times of observations, the instrument orientation,
the instrument coordinate system, etc.  Example 3 below shows how to deal
with such files.  Three things should be noted.
\enumerate{
\item The user must choose the appropriate \code{read.adv} variant
corresponding to the instrument in question.  (This is necessary because
\code{\link[=oceMagic]{oceMagic()}}, which is used by the generic \code{\link[=read.oce]{read.oce()}}
routine, cannot determine the type of instrument by examining a file that
lacks a header.)
\item The call to the \code{read} function must include a start time
(\code{start}) and the number of seconds between data (\code{deltat}),
again, because the instrument data stream may lack those things when the
device is set to a serial mode.  Also, of course, it is necessary to set
\code{header=FALSE} in the function call.
\item Once the file has been read in, the user will be obliged to specify
other information, for the object to be well-formed.  For example, the
\code{read} function will have no way of knowing the instrument orientation,
the coordinate system being used, the transformation matrix to go from
\code{"beam"} to \code{"xyz"} coordinates, or the instrument heading, pitch,
and roll, to go from \code{"xyz"} coordinates to \code{"enu"} coordinates.
Such things are illustrated in example 3 below.
}

In ADV data files, velocities are coded to signed 2-byte integers, with a
scale factor being used to convert to velocity in metres per second.  These
two facts control the maximum recordable velocity and the velocity
resolution, values that may be retrieved for an ADV object name \code{d}
with \code{d[["velocityMaximum"]]} and \code{d[["velocityResolution"]]}.
}
\section{Nortek files}{


\strong{Sampling-rate and similar issues}

The data format is inferred from the System Integrator Guide (reference 1A) and System
Integrator Manual (reference 1B).  These document lacks clarity in spots, and so
\code{read.adv.nortek} contains some assumptions that are noted here, so
that users will be aware of possible problems.

A prominent example is the specification of the sampling rate, stored in
\code{metadata$sampingRate} in the return value.  Repeated examination of
the System Integrator Guide (reference 1) failed to indicate where this value is
stored in the various headers contained in Vector datasets.  After some
experimentation with a few data files, \code{read.adv.nortek} was set up to
calculate \code{metadata$samplingRate} as \code{512/AvgInterval} where
\code{AvgInterval} is a part of the \verb{User Configuration'' header (reference 1 p30), where the explanation is }average interval in seconds'').  This formula was
developed through trial and error, but it was confirmed later on the Nortek
discussion group, and it should appear in upcoming versions of (reference 1).

Another basic issue is the determination of whether an instrument had
recorded in continuous mode or burst mode.  One might infer that
\code{TimCtrlReg} in the \verb{User Configuration'' header (reference 1 p30) determines this, in bits 1 and 2.  However, this was the case in test files available to the author.  For this reason, `read.adv.nortek` infers the mode by reverse engineering of data files of known configuration.  The present version of `read.adv.nortek` determines the sampling mode from the ```NRecords`'' item of the }Vector Velocity Data'' header, which seems
to be 0 for data collected continuously, and non-zero for data collected in
bursts.

Taking these things together, we come upon the issue of how to infer
sampling times for Nortek instruments.  There do not seem to be definitive
documents on this, and so \code{read.adv.nortek} is based partly on
information (of unknown quality) found on Nortek discussion boards.  The
present version of \code{read.adv.nortek} infers the times of velocity
observations differently, depending on whether the instrument was set to
record in burst mode or continuous mode.  For burst mode, times stated in
the burst headers are used, but for continuous mode, times stated in the
``vector system data'' are used.  On the advice found on a Nortek discussion
board, the burst-mode times are offset by 2 seconds to allow for the
instrument warm-up period.

\strong{Handling IMU (inertial measurement unit) data}

Starting in March
2016, \code{read.adv.nortek} has offered some support for handling IMU
(inertial measurement unit) data incorporated into Nortek binary files. This
is not described in the Nortek document named \verb{System Integrator Guide'' (reference 1A) but it appeared in }System Integrator Manual'' (reference 1B;
reference 1C). Confusingly, 1B described 3 varieties of data, whereas 1C does not
describe any of these, but describes instead a fourth variety. As of March
2016, \code{read.adv.nortek} handles all 4 varieties, because files in the
various schemes appear to exist. In \code{oce}, the varieties are named
after the byte code that flags them.  (Variety \code{c3} is the one
described in (reference 1C); the others were described in (reference 1B).)
The variety is stored in the \code{metadata} slot of the returned object as
a string named \code{IMUtype}.

For each variety, the reader is cautioned that strong tests have not been
performed on the code.  One way to test the code is to compare with textual
data files produced by the Nortek software.  In March 2016, an \code{oce}
user shared a dataset of the \code{c3} variety, and this permitted detailed
comparison between the text file and the values inferred by
\code{read.adv.nortek}. The test suggested agreement (to within the
resolution printed in the text file) for velocity (\code{v} in the
\code{data} slot), signal amplitude (\code{a}), correlation (\code{q}),
pressure (\code{p}), the three components of IMU delta angle
(\code{IMUdeltaAngleX} etc), and all components of the rotation matrix
(\code{IMUrotation}).  However, the delta velocity signals did not match,
with \code{IMUdeltaVelocityX} disagreeing in the second decimal place,
\code{IMUdeltaVelocityY} component disagreeing in the first, and
\code{IMUdeltaVelocityZ} being out by a factor of about 10. This is github
issue 893 (\url{https://github.com/dankelley/oce/issues/893}).
\itemize{
\item Variety \code{c3} (signalled by byte 5 of a sequence being
\code{0xc3}) provides information on what Nortek calls DeltaAngle,
DeltaVelocity and Orientation Matrix. (Apart from the orientation matrix,
Nortek provides no documentation on what these quantities mean.) In the
object returned by \code{read.adv.nortek}, these are stored in the
\code{data} slot as \code{IMUdeltaAngleX}, \code{IMUdeltaAngleY},
\code{IMUdeltaAngleZ}, \code{IMUdeltaVelocityX}, \code{IMUdeltaVelocityY},
\code{IMUdeltaVelocityZ}, and \code{IMUrotation}, all vectors except the
last, which is a 3D array. In addition to these, \code{IMUtimestamp} is a
time-stamp, which is not defined in the Nortek documents but seems, from IMU
documents (reference 5), to be defined based on a clock that ticks once per 16
microseconds. Caution may be required in dealing with this timestamp, since
it seemed sensible in one test case (variety \code{d3}) but kept reseting to
zero in another (variety \code{c3}). The lack of Nortek documentation on
most of these quantities is a roadblock to implementing \code{oce} functions
dealing with IMU-enabled datasets
\item Variety \code{cc} (signalled by byte 5 of a sequence being
\code{0xcc}) provides information on acceleration, angular rotation rate,
magnetic vector and orientation matrix.  Each is a timeseries. Acceleration
is stored in the \code{data} slot as \code{IMUaccelX}, \code{IMUaccelY},
\code{IMUaccelz}. The angular rotation components are \code{IMUngrtX},
\code{IMUngrtY} and \code{IMUngrtz}. The magnetic data are in
\code{IMUmagrtx}, \code{IMUmagrty} and \code{IMUmagrtz}. Finally,
\code{IMUmatrix} is a rotation matrix made up from elements named
\code{M11}, \code{M12}, etc in the Nortek documentation.  In addition to all
of these, \code{IMUtime} stores time in seconds, with an origin whose
definition is not stated in reference 1B.
\item Variety \code{d2} (signalled by byte 5 being \code{0xd2}) provides
information on gyro-stabilized acceleration, angular rate and magnetometer
vectors.  The data stored \code{MUaccelX}, \code{IMUangrtX},
\code{IMUmagrtX}, with similar for \code{Y} and \code{Z}.  Again, time is in
\code{IMUtime}. This data type has not been tested as of mid-March 2016,
because the developers do not have a test file with which to test.
\item Variety \code{d3} (signalled by byte 5 being \code{0xd3}) provides
information on DeltaAngle, DeltaVelocity and magnetometer vectors, stored in
\code{IMUdeltaAngleX}, \code{IMUdeltaVelocityX}, and
\code{IMUdeltaMagVectorX}, with similar for \code{Y} and \code{Z}. Again,
time is in \code{IMUtime}. This data type has not been tested as of
mid-March 2016, because the developers do not have a test file with which to
test.
}
}

\section{Note on file name}{

The \code{file} argument does not actually name a file. It names a basename
for a file. The actual file names are created by appending suffix
\code{.hd1} for one file and \code{.ts1} for another.
}

\examples{
\dontrun{
library(oce)
# A nortek Vector file
d <- read.oce("/data/archive/sleiwex/2008/moorings/m05/adv/nortek_1943/raw/adv_nortek_1943.vec",
              from=as.POSIXct("2008-06-26 00:00:00", tz="UTC"),
              to=as.POSIXct("2008-06-26 00:00:10", tz="UTC"))
plot(d, which=c(1:3,15))
}
}
\references{
1A. Nortek AS.  System Integrator Guide (paradopp family of products). June
2008.  (Doc No: PSI00-0101-0608).  (Users may find it helpful to also
examine newer versions of the guide.)

1B. Nortek AS.  System Integrator Manual. Dec 2014.
(\code{system-integrator-manual_Dec2014_jan.pdf})

1C. Nortek AS.  System Integrator Manual. March 2016.
(\code{system-integrator-manual_Mar2016.pdf})
\enumerate{
\item SonTek/YSI ADVField/Hydra Acoustic Doppler Velocimeter (Field) Technical
Documentation (Sept 1, 2001).
\item Appendix 2.2.3 of the Sontek ADV operation Manual, Firmware Version 4.0
(Oct 1997).
\item Nortek Knowledge Center (http://www.nortekusa.com/en/knowledge-center)
\item A document describing an IMU unit that seems to be close to the one named
in (references 1B and C) as being an adjunct to Nortek Vector systems is at
\verb{http://files.microstrain.com/3DM-GX3-35-Data-Communications-Protocol.pdf}
}
}
\seealso{
Other things related to adv data: 
\code{\link{[[,adv-method}},
\code{\link{[[<-,adv-method}},
\code{\link{adv-class}},
\code{\link{adv}},
\code{\link{beamName}()},
\code{\link{beamToXyz}()},
\code{\link{enuToOtherAdv}()},
\code{\link{enuToOther}()},
\code{\link{plot,adv-method}},
\code{\link{read.adv.nortek}()},
\code{\link{read.adv.sontek.adr}()},
\code{\link{read.adv.sontek.serial}()},
\code{\link{read.adv}()},
\code{\link{rotateAboutZ}()},
\code{\link{subset,adv-method}},
\code{\link{summary,adv-method}},
\code{\link{toEnuAdv}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdv}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley
}
\concept{things related to adv data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.R
\name{[[,ctd-method}
\alias{[[,ctd-method}
\title{Extract Something From a ctd Object}
\usage{
\S4method{[[}{ctd}(x, i, j, ...)
}
\arguments{
\item{x}{a \linkS4class{ctd} object.}

\item{i}{character value indicating the name of an item to extract.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{ignored.}
}
\description{
Generally, the \code{[[} method lets users extract information from \code{oce}
objects, without having to know the details of the internal storage.  For
many \code{oce} sub-classes, \code{[[} can also return quantities that are computed
from the object's contents.
}
\details{
A two-step process is used to try to find the requested information. First, a
class-specific function is used (see \dQuote{Details of the Specialized
Method}).  If this yields nothing, then a general method is used (see
\dQuote{Details of the General Method}).  If both methods fail, then \code{[[}
returns NULL.

Some understanding of the subclass is required to know what can be retrieved
with \code{[[}. When dealing with an unfamiliar subclass, it can be useful to
first use \code{x[["?"]]} to get a listing of the retrievable items. See
\dQuote{Details of the Specialized Method} for more information.
}
\section{Details of the Specialized Method}{


Some uses of \code{\link{[[,ctd-method}} involve direct retrieval of
items within the \code{data} slot of the \code{ctd} object,
while other uses involve calculations based on items in that
\code{data} slot. For example, all \code{ctd} objects
should hold an item named \code{temperature} in the \code{data}
slot, so for example \code{x[["temperature"]]} will retrieve that
item. By contrast, \code{x[["sigmaTheta"]]} is taken to be a
request to compute \eqn{\sigma_\theta}{sigma[theta]}, and so
it yields a call to \link{swTheta}\code{(x)} even if
the \code{data} slot of \code{x} might happen to contain an item
named \code{theta}. This can be confusing at first, but it tends
to lead to fewer surprises in everyday work, for otherwise the
user would be forced to check the contents of any \code{ctd}
object under analysis, to determine whether that item will be looked
up or computed. Nothing is lost in this scheme, since the data
within the object are always accessible with \code{\link[=oceGetData]{oceGetData()}}.

It should be noted that the accessor is set up to retrieve quantities
in conventional units. For example, \code{\link[=read.ctd.sbe]{read.ctd.sbe()}} is
used on a \code{.cnv} file that stores pressure in psi, it will
be stored in the same unit within the \code{ctd} object, but
\code{x[["pressure"]]} will return a value that has been converted
to decibars.  (To get pressure in PSI, use \code{x[["pressurePSI"]]}.)
Similarly, temperature is
returned in the ITS-90 scale, with a conversion having been performed with
\code{\link[=T90fromT68]{T90fromT68()}}, if the object holds temperature in
IPTS-68.  Again, temperature on the IPTS-68
scale is returned with \code{x@data$temperature}.

This preference for computed over stored quantities is accomplished
by first checking for computed quantities, and then falling
back to the general \code{[[} method if no match is found.

Some quantities are optionally computed. For example, some data files
(e.g. the one upon which the \code{\link[=section]{section()}} dataset is based)
store \code{nitrite} along with the sum of nitrite and nitrate, the
latter with name \code{NO2+NO3}. In this case, e.g. \code{x[["nitrate"]]}
will detect the setup, and subtract nitrite from the sum to yield
nitrate.

The list given below provides notes on some quantities that are
available using e.g. \code{ctd[[i]]}.
\itemize{
\item If \code{i} is \code{"?"}, then the return value is a list
containing four items, each of which is a character vector
holding the names of things that can be accessed with \code{[[}.
The \code{data} and \code{metadata} items hold the names of
entries in the object's data and metadata
slots, respectively. The \code{dataDerived}
and \code{metadataDerived} items hold the names of things
that can be inferred from the object's contents, e.g.
\code{"SA"} is named in \code{dataDerived}, indicating that
\code{argo[["SA"]]} is permitted (to compute Absolute Salinity).
\item If \code{i} is \code{"conductivity"} without a second argument (e.g. \code{a[["conductivity"]]})
then the return value is the seawater electrical conductivity (if available
or computable). However, if a second argument is given,
and it is string specifying a unit, then conversion is made to that unit. The
permitted units are: either \code{""} or \code{"ratio"} (for ratio),
\code{"uS/cm"}, \code{"mS/cm"} and \code{"S/m"}. The calculations are based on
the definition of conductivity ratio as the ratio between measured conductivity
and the standard value 4.2914 S/m.
\item If \code{i} is \code{"CT"} or \code{"Conservative Temperature"} then Conservative
Temperature, computed with \code{\link[gsw:gsw_CT_from_t]{gsw::gsw_CT_from_t()}}, is returned.
\item If \code{i} is \code{"density"} then seawater density, computed with \link{swRho}\code{(x)},
is returned. (Note that it may be better to call that function directly, to gain
control of the choice of equation of state, etc.)
\item If \code{i} is \code{"depth"} then the depth in metres below the surface, computed
with \link{swDepth}\code{(x)}, is returned.
\item If \code{i} is \code{"N2"} then the square of Brunt-Vaisala frequency, computed  with
\link{swN2}\code{(x)}, is returned.
\item If \code{i} is \code{"potential temperature"} or \code{"theta"}, then potential temperature in the
UNESCO formulation, computed with \link{swTheta}\code{(x)}, is returned.
\item If \code{i} is \code{"Rrho"} then density ratio, computed with \link{swRrho}\code{(x)}, is
returned.
\item If \code{i} is \code{"SA"} or \code{"Absolute Salinity"} then Absolute Salinity,
computed with \code{\link[gsw:gsw_SA_from_SP]{gsw::gsw_SA_from_SP()}}, is returned.
The calculation involves location as well as measured water properties.
If the object \code{x} does not containing information on the location,
then 30N and 60W is used for the calculation, and a warning is generated.
\item If \code{i} is \code{"sigmaTheta"} then a form of potential density anomaly, computed with
\link{swSigmaTheta}\code{(x)}, is returned.
\item If \code{i} is \code{"sigma0"} then potential density anomaly
referenced to a sea pressure of 0dbar (the surface), computed with \link{swSigma0}\code{(x)},
is returned.
\item If \code{i} is \code{"sigma2"} then potential density anomaly
referenced to a sea pressure of 1000dbar, computed with \link{swSigma1}\code{(x)},
is returned.
\item If \code{i} is \code{"sigma2"} then potential density anomaly
referenced to a sea pressure of 2000dbar, computed with \link{swSigma2}\code{(x)},
is returned.
\item If \code{i} is \code{"sigma3"} then potential density anomaly
referenced to a sea pressure of 3000dbar, computed with \link{swSigma3}\code{(x)},
is returned.
\item If \code{i} is \code{"sigma4"} then potential density anomaly
referenced to a sea pressure of 4000dbar, computed with \link{swSigma4}\code{(x)},
is returned.
\item If \code{i} is \code{"SP"} then salinity on the Practical Salinity Scale, which is
\code{salinity} in the \code{data} slot, is returned.
\item If \code{i} is \code{"spice"} or \code{"spiciness0"} then a variable that is in some sense
orthogonal to density, calculated with \link{swSpice}\code{(x)}, is returned.
Note that this is defined differently for \code{eos="unesco"} and
\code{eos="gsw"}.
\item If \code{i} is \code{"SR"} then Reference Salinity, computed with
\code{\link[gsw:gsw_SR_from_SP]{gsw::gsw_SR_from_SP()}}, is returned.
\item If \code{i} is \code{"Sstar"} then Preformed Salinity, computed with
\code{\link[gsw:gsw_SR_from_SP]{gsw::gsw_SR_from_SP()}}, is returned.
See \code{SA} for a note on longitude and latitude.
\item If \code{i} is \code{"time"} then either vector of times or a single
time, is returned, if available. A vector is returned if \code{time}
is present in the \code{data} slot, or if a time can be
inferred from other entries in the \code{data} slot (some of which,
such as the common \code{timeS}, also employ
\code{startTime} within the \code{metadata} slot). A single
value is returned if the dataset only has information on the start
time (which is stored as \code{startTime} within the \code{metadata}
slot. If it is impossible to determine the sampling time, then
\code{NULL} is returned. These time variants occur, in the
present version of oce, only for data read by \code{\link[=read.ctd.sbe]{read.ctd.sbe()}},
the documentation of which explains how times are computed.
\item If \code{i} is \code{"z"} then vertical coordinate in metres
above the surface, computed with \link{swZ}\code{(x)}, is returned.
}
}

\section{Details of the General Method}{


Note: the text of this section is identical for all \code{oce} subclasses, and so
some of what you read here may not be relevant to the class being described
in this help page.

If the specialized method produces no matches, the following generalized
method is applied. As with the specialized method, the procedure hinges first
on the values of \code{i} and, optionally, \code{j}. The work proceeds in steps, by
testing a sequence of possible conditions in sequence.
\enumerate{
\item A check is made as to whether \code{i} names one of the standard \code{oce} slots.
If so, \code{[[} returns the slot contents of that slot.  Thus, \code{x[["metadata"]]}
will retrieve the \code{metadata} slot, while \code{x[["data"]]} and
\code{x[["processingLog"]]} return those slots.
\item If \code{i} is a string ending in the \code{"Unit"}, then the characters preceding
that string are taken to be the name of an item in the data object, and a
list containing the unit is returned (or \code{NULL} if there is no such unit).
This list consists of an item named \code{unit}, which is an \code{\link[=expression]{expression()}}, and
an item named \code{scale}, which is a string describing the measurement scale.
If the string ends in \code{" unit"}, e.g.  \code{x[["temperature unit"]]} (note the
space), then just the expression is returned, and if it ends in \code{" scale"},
then just the scale is returned.
\item If \code{i} is a string ending in \code{"Flag"}, then the corresponding data-quality
flag is returned (or \code{NULL} if there is no such flag).
\item If the object holds hydrographic information (pressure, salinity,
temperature, longitude and latitude) then another set of possibilities
arises.  If \code{i} is \code{"sigmaTheta"}, then the value of \code{\link[=swSigmaTheta]{swSigmaTheta()}} is
called with \code{x} as the sole argument, and the results are returned.
Similarly, \code{\link[=swSigma0]{swSigma0()}} is used if \code{i="sigma0"}, and \code{\link[=swSpice]{swSpice()}} is used if
\code{i="spice"}. Of course, these actions only make sense for objects that
contain the relevant items within their \code{data} slot.
\item After these possibilities are eliminated, the action depends on whether
\code{j} has been provided.  If \code{j} is not provided, or is the string \code{""}, then
\code{i} is sought in the \code{metadata} slot, and then in the \code{data} slot, returning
whichever is found first.  In other words, if \code{j} is not provided, the
\code{metadata} slot takes preference over the \code{data} slot. However, if \code{j} is
provided, then it must be either the string \code{"metadata"} or \code{"data"}, and it
directs where to look.
\item If none of the above-listed conditions holds, then \code{NULL} is returned.
}
}

\examples{
data(ctd)
head(ctd[["temperature"]])

}
\seealso{
Other functions that extract parts of oce objects: 
\code{\link{[[,adp-method}},
\code{\link{[[,adv-method}},
\code{\link{[[,amsr-method}},
\code{\link{[[,argo-method}},
\code{\link{[[,bremen-method}},
\code{\link{[[,cm-method}},
\code{\link{[[,coastline-method}},
\code{\link{[[,echosounder-method}},
\code{\link{[[,g1sst-method}},
\code{\link{[[,gps-method}},
\code{\link{[[,ladp-method}},
\code{\link{[[,landsat-method}},
\code{\link{[[,lisst-method}},
\code{\link{[[,lobo-method}},
\code{\link{[[,met-method}},
\code{\link{[[,oce-method}},
\code{\link{[[,odf-method}},
\code{\link{[[,rsk-method}},
\code{\link{[[,sealevel-method}},
\code{\link{[[,section-method}},
\code{\link{[[,tidem-method}},
\code{\link{[[,topo-method}},
\code{\link{[[,windrose-method}},
\code{\link{[[,xbt-method}},
\code{\link{[[<-,adv-method}}

Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}
}
\author{
Dan Kelley
}
\concept{functions that extract parts of oce objects}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swConservativeTemperature}
\alias{swConservativeTemperature}
\title{Seawater conservative temperature, in GSW formulation}
\usage{
swConservativeTemperature(
  salinity,
  temperature = NULL,
  pressure = NULL,
  longitude = NULL,
  latitude = NULL
)
}
\arguments{
\item{salinity}{either practical salinity (in which case \code{temperature}
and \code{pressure} must be provided) \emph{or} an \code{oce} object (in
which case \code{salinity}, etc. are inferred from the object).}

\item{temperature}{\emph{in-situ} temperature (\eqn{^\circ}{deg}C), defined
on the ITS-90 scale; see \dQuote{Temperature units} in the documentation for
\code{\link[=swRho]{swRho()}}.}

\item{pressure}{pressure (dbar)}

\item{longitude}{longitude of observation.}

\item{latitude}{latitude of observation.}
}
\value{
Conservative temperature in degrees Celcius.
}
\description{
Compute seawater Conservative Temperature, according to the GSW/TEOS-10
formulation.
}
\details{
If the first argument is an \code{oce} object, then values for salinity,
etc., are extracted from it, and used for the calculation, and the
corresponding arguments to the present function are ignored.

The conservative temperature is calculated using the TEOS-10 function
\link[gsw:gsw_CT_from_t]{gsw::gsw_CT_from_t} from the \CRANpkg{gsw} package.
}
\examples{
swConservativeTemperature(35,10,1000,188,4) # 9.86883

}
\references{
McDougall, T.J. and P.M. Barker, 2011: Getting started with
TEOS-10 and the Gibbs Seawater (GSW) Oceanographic Toolbox, 28pp.,
SCOR/IAPSO WG127, ISBN 978-0-646-55621-5.
}
\seealso{
The related TEOS-10 quantity ``absolute salinity'' may be computed
with \code{\link[=swAbsoluteSalinity]{swAbsoluteSalinity()}}.  For a ctd object, conservative
temperature may also be recovered by indexing as e.g.
\code{ctd[["conservativeTemperature"]]} or \code{ctd[["CT"]]}.

Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/met.R
\name{as.met}
\alias{as.met}
\title{Coerce Data into met Object}
\usage{
as.met(time, temperature, pressure, u, v, filename = "(constructed from data)")
}
\arguments{
\item{time}{Either a vector of observation times (or character strings that can be
coerced into times) or the output from \code{canadaHCD::hcd_hourly} (see reference 1).}

\item{temperature}{vector of temperatures.}

\item{pressure}{vector of pressures.}

\item{u}{vector of eastward wind speed in m/s.}

\item{v}{vector of northward wind speed in m/s.}

\item{filename}{optional string indicating data source}
}
\value{
A \linkS4class{met} object.
}
\description{
Coerces a dataset into a met dataset. This fills in only a few of the typical
data fields, so the returned object is much
sparser than the output from \code{\link[=read.met]{read.met()}}. Also, almost no
metadata fields are filled in, so the resultant object does not store
station location, units of the data, data-quality flags, etc. Anyone working
with data from Environment Canada (reference 2) is advised to use \code{\link[=read.met]{read.met()}}
instead of the present function.
}
\references{
\enumerate{
\item The \code{canadaHCD} package is in development by Gavin Simpson; see
\url{https://github.com/gavinsimpson/canadaHCD} for instructions on how
to download and install from GitHub.
\item Environment Canada website for Historical Climate Data
\verb{https://climate.weather.gc.ca/index_e.html}
}
}
\seealso{
Other things related to met data: 
\code{\link{[[,met-method}},
\code{\link{[[<-,met-method}},
\code{\link{download.met}()},
\code{\link{met-class}},
\code{\link{met}},
\code{\link{plot,met-method}},
\code{\link{read.met}()},
\code{\link{subset,met-method}},
\code{\link{summary,met-method}},
\code{\link{test_met_csv1.csv}},
\code{\link{test_met_csv2.csv}},
\code{\link{test_met_xml2.xml}}
}
\author{
Dan Kelley
}
\concept{things related to met data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/map.R
\name{lonlat2map}
\alias{lonlat2map}
\title{Convert Longitude and Latitude to X and Y}
\usage{
lonlat2map(longitude, latitude, projection = "", debug = getOption("oceDebug"))
}
\arguments{
\item{longitude}{a numeric vector containing decimal longitudes, or a list
containing items named \code{longitude} and \code{latitude}, in which case
the indicated values are used, and next argument is ignored.}

\item{latitude}{a numeric vector containing decimal latitude (ignored if
\code{longitude} is a list, as described above).}

\item{projection}{optional indication of projection.  This must be character
string in the format used by the \CRANpkg{sf} package;
see \code{\link[=mapPlot]{mapPlot()}}.)}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\value{
A list containing \code{x} and \code{y}.
}
\description{
If a projection is already being used (e.g. as set by \code{\link[=mapPlot]{mapPlot()}})
then only \code{longitude} and \code{latitude} should be given, and the
other arguments will be inferred by \code{lonlat2map}.  This is important
because otherwise, if a new projection is called for, it will ruin any
additions to the existing plot.
}
\examples{
library(oce)
## Cape Split, in the Minas Basin of the Bay of Fundy
cs <- list(longitude=-64.49657, latitude=45.33462)
xy <- lonlat2map(cs, projection="+proj=merc")
map2lonlat(xy)

}
\seealso{
\code{mapLongitudeLatitudeXY} is a safer alternative, if a map has
already been drawn with \code{\link[=mapPlot]{mapPlot()}}, because that function cannot
alter an existing projection. \code{\link[=map2lonlat]{map2lonlat()}} is an inverse to
\code{map2lonlat}.

Other functions related to maps: 
\code{\link{formatPosition}()},
\code{\link{lonlat2utm}()},
\code{\link{map2lonlat}()},
\code{\link{mapArrows}()},
\code{\link{mapAxis}()},
\code{\link{mapContour}()},
\code{\link{mapCoordinateSystem}()},
\code{\link{mapDirectionField}()},
\code{\link{mapGrid}()},
\code{\link{mapImage}()},
\code{\link{mapLines}()},
\code{\link{mapLocator}()},
\code{\link{mapLongitudeLatitudeXY}()},
\code{\link{mapPlot}()},
\code{\link{mapPoints}()},
\code{\link{mapPolygon}()},
\code{\link{mapScalebar}()},
\code{\link{mapText}()},
\code{\link{mapTissot}()},
\code{\link{oceCRS}()},
\code{\link{shiftLongitude}()},
\code{\link{usrLonLat}()},
\code{\link{utm2lonlat}()}
}
\author{
Dan Kelley
}
\concept{functions related to maps}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swSTrho}
\alias{swSTrho}
\title{Seawater salinity from temperature and density}
\usage{
swSTrho(
  temperature,
  density,
  pressure,
  eos = getOption("oceEOS", default = "gsw")
)
}
\arguments{
\item{temperature}{\emph{in-situ} temperature (\eqn{^\circ}{deg}C), defined
on the ITS-90 scale; see \dQuote{Temperature units} in the documentation for
\code{\link[=swRho]{swRho()}}.}

\item{density}{\emph{in-situ} density or sigma value (\eqn{kg/m^3}{kg/m^3})}

\item{pressure}{\emph{in-situ} pressure (dbar)}

\item{eos}{equation of state, either \code{"unesco"} (see references 1 and 2)
or \code{"gsw"} (see references 3 and 4).}
}
\value{
Practical Salinity, if \code{eos="unesco"}, or Absolute Salinity, if
\code{eos="gsw"}.
}
\description{
Compute Practical or Absolute Salinity, given in-situ or Conservative
Temperature, density, and pressure. This is mainly used to draw isopycnal
lines on TS diagrams, hence the dual meanings for salinity and temperature,
depending on the value of \code{eos}.
}
\details{
For \code{eos="unesco"}, finds the practical salinity that yields the given
density, with the given in-situ temperature and pressure.  The method is a
bisection search with a salinity tolerance of 0.001.  For \code{eos="gsw"},
the function \code{\link[gsw:gsw_SA_from_rho]{gsw::gsw_SA_from_rho()}} in the \CRANpkg{gsw}
package is used
to infer Absolute Salinity from Conservative Temperature.
}
\examples{
swSTrho(10, 22, 0, eos="gsw") # 28.76285
swSTrho(10, 22, 0, eos="unesco") # 28.651625

}
\references{
\enumerate{
\item Fofonoff, P. and R. C. Millard Jr, 1983. Algorithms for computation of
fundamental properties of seawater. \emph{Unesco Technical Papers in Marine
Science}, \emph{44}, 53 pp
\item Gill, A.E., 1982. \emph{Atmosphere-ocean Dynamics}, Academic Press, New
York, 662 pp.
\item IOC, SCOR, and IAPSO (2010). The international thermodynamic equation of
seawater-2010: Calculation and use of thermodynamic properties.  Technical
Report 56, Intergovernmental Oceanographic Commission, Manuals and Guide.
\item McDougall, T.J. and P.M. Barker, 2011: Getting started with TEOS-10 and
the Gibbs Seawater (GSW) Oceanographic Toolbox, 28pp., SCOR/IAPSO WG127,
ISBN 978-0-646-55621-5.
}
}
\seealso{
\code{\link[=swTSrho]{swTSrho()}}

Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.nortek.R
\name{read.aquadoppHR}
\alias{read.aquadoppHR}
\title{Read Nortek Aquadopp-HR File}
\usage{
read.aquadoppHR(
  file,
  from = 1,
  to,
  by = 1,
  tz = getOption("oceTz"),
  longitude = NA,
  latitude = NA,
  orientation = orientation,
  distance,
  monitor = FALSE,
  despike = FALSE,
  processingLog,
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{file}{a connection or a character string giving the name of the file
to load.  (For \code{read.adp.sontek.serial}, this is generally a list of
files, which will be concatenated.)}

\item{from}{indication of the first profile to read.  This can be an
integer, the sequence number of the first profile to read, or a POSIXt time
before which profiles should be skipped, or a character string that converts
to a POSIXt time (assuming UTC timezone).  See \dQuote{Examples}, and make
careful note of the use of the \code{tz} argument. If \code{from} is not
supplied, it defaults to 1.}

\item{to}{an optional indication of the last profile to read, in a
format as described for \code{from}.  As a special case, \code{to=0} means
to read the file to the end.  If \code{to} is not supplied, then it defaults
to 0.}

\item{by}{an optional indication of the stride length to use while walking through
the file.  If this is an integer, then \code{by-1} profiles are skipped
between each pair of profiles that is read, e.g. the default \code{by=1}
means to read all the data.  (For RDI files \emph{only}, there are some
extra features to avoid running out of memory; see \dQuote{Memory considerations}.)}

\item{tz}{character string indicating time zone to be assumed in the data.}

\item{longitude}{optional signed number indicating the longitude in degrees
East.}

\item{latitude}{optional signed number indicating the latitude in degrees
North.}

\item{orientation}{Optional character string specifying the orientation of the
sensor, provided for those cases in which it cannot be inferred from the
data file.  The valid choices are \code{"upward"}, \code{"downward"}, and
\code{"sideward"}.}

\item{distance}{Optional vector holding the distances of bin centres from the
sensor.  This argument is ignored except for Nortek profilers, and need not
be given if the function determines the distances correctly from the data.
The problem is that the distance is poorly documented in the Nortek System
Integrator Guide (2008 edition, page 31), so the function must rely on
word-of-mouth formulae that do not work in all cases.}

\item{monitor}{boolean value indicating whether to indicate the progress
of reading the file, by using \code{\link[=txtProgressBar]{txtProgressBar()}} or otherwise.  The value
of \code{monitor} is changed to \code{FALSE} automatically, for non-interactive
sessions.}

\item{despike}{if \code{TRUE}, \code{\link[=despike]{despike()}} will be used to clean
anomalous spikes in heading, etc.}

\item{processingLog}{if provided, the action item to be stored in the log.
(Typically only provided for internal calls; the default that it provides is
better for normal calls by a user.)}

\item{debug}{a flag that turns on debugging.  Set to 1 to get a moderate
amount of debugging information, or to 2 to get more.}

\item{\dots}{optional additional arguments that some (but not all)
\verb{read.adp.*()} functions pass to lower-level functions.}
}
\value{
An \linkS4class{adp} object.
The contents of that object make sense for the particular instrument
type under study, e.g. if the data file contains
NMEA strings, then navigational data will be stored in an item
called \code{nmea} in the \code{data} slot).
}
\description{
The R code is based on information in
the Nortek System Integrator Guide (2008) and on postings on the Nortek
``knowledge center'' discussion board.  One might assume that the latter is
less authoritative than the former.  For example, the inference of cell size
follows advice found at
\url{https://www.nortekusa.com/en/knowledge-center/forum/hr-profilers/736804717},
which contains a typo in an early posting that is
corrected later on.
}
\references{
\enumerate{
\item Information on Nortek profilers (including the System Integrator Guide,
which explains the data format byte-by-byte) is available at
\url{https://www.nortekusa.com/}.  (One must join the site to see the
manuals.)
\item The Nortek Knowledge Center
\url{https://www.nortekusa.com/en/knowledge-center} may be of help if
problems arise in dealing with data from Nortek instruments.
}
}
\seealso{
Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley
}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swZ}
\alias{swZ}
\title{Vertical coordinate}
\usage{
swZ(pressure, latitude = 45, eos = getOption("oceEOS", default = "gsw"))
}
\arguments{
\item{pressure}{either pressure (dbar), in which case \code{lat} must also
be given, or a \code{ctd} object, in which case \code{lat} will be inferred
from the object.}

\item{latitude}{Latitude in \eqn{^\circ}{deg}N or radians north of the
equator.}

\item{eos}{indication of formulation to be used, either \code{"unesco"} or
\code{"gsw"}.}
}
\description{
Compute height above the surface. This is the negative of depth,
and so is defined simply in terms of \code{\link[=swDepth]{swDepth()}}.
}
\seealso{
Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()}
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.R
\name{ctdDecimate}
\alias{ctdDecimate}
\title{Decimate a CTD profile}
\usage{
ctdDecimate(
  x,
  p = 1,
  method = "boxcar",
  rule = 1,
  e = 1.5,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{x}{a \linkS4class{ctd} object.}

\item{p}{pressure increment, or vector of pressures.  In the first case,
pressures from 0dbar to the rounded maximum pressure are used, incrementing by
\code{p} dbars.  If a vector of pressures is given, interpolation is done to
these pressures.}

\item{method}{the method to be used for calculating decimated values.  This may
be a function or a string naming a built-in method.  The built-in methods are
as follows.
\itemize{
\item \code{"boxcar"} (based on a local average)
\item \code{"approx"} (based on linear
interpolation between neighboring points, using \code{\link[=approx]{approx()}}
with the \code{rule} argument specified here)
\item \code{"approxML"} as \code{"approx"},
except that a mixed layer is assumed to apply above the top data value; this
is done by setting the \code{yleft} argument to \code{\link[=approx]{approx()}}, and
by calling that function with \verb{rule=c(2,1))}
\item \code{"lm"} (based on local
regression, with \code{e} setting the size of the local region);
\item \code{"rr"} for the Reiniger and Ross method, carried out with \code{\link[=oce.approx]{oce.approx()}};
\item \code{"unesco"} (for the UNESCO method, carried out with \code{\link[=oce.approx]{oce.approx()}}.
}

On the other hand, if \code{method} is a function, then it must take
three arguments, the first being pressure, the second being an arbitrary
variable in another column of the data, and the third being a vector of target
pressures at which the calculation is carried out, and the return value must be
a vector.  See \dQuote{Examples}.}

\item{rule}{an integer that is passed to \code{\link[=approx]{approx()}}, in the
case where \code{method} is \code{"approx"}. Note that the default value
for \code{rule} is 1, which will inhibit extrapolation beyond the observed
pressure range. This is a change from the behaviour previous to May 8, 2017,
when a \code{rule} of 2 was used (without stating so as an argument).}

\item{e}{is an expansion coefficient used to calculate the local neighbourhoods
for the \code{"boxcar"} and \code{"lm"} methods.  If \code{e=1}, then the
neighbourhood for the i-th pressure extends from the (\code{i-1})-th pressure to
the (\code{i+1})-th pressure.  At the endpoints it is assumed that the outside
bin is of the same pressure range as the first inside bin.  For other values of
\code{e}, the neighbourhood is expanded linearly in each direction.  If the
\code{"lm"} method produces warnings about "prediction from a rank-deficient
fit", a larger value of \code{"e"} should be used.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\value{
A \linkS4class{ctd} object, with pressures that are as set by
the \code{"p"} parameter and all other properties modified appropriately.
}
\description{
Interpolate a CTD profile to specified pressure values.  This is used
by \code{\link[=sectionGrid]{sectionGrid()}}, but is also useful for dealing with individual
CTD/bottle profiles.
}
\details{
The \code{"approx"} and \code{"approxML"} methods may be best for bottle data,
in which the usual task is
to interpolate from a coarse sampling grid to a finer one. The distinction
is that \code{"approxML"} assumes a mixed-layer above the top sample value. For CTD data, the
\code{"boxcar"} method may be the preferred choice, because the task is normally
to sub-sample, and some degree of smoothing is usually desired.  (The
\code{"lm"} method can be quite slow, and its results may be quite similar to those of the
boxcar method.)

For widely-spaced data, a sort of numerical cabeling effect can result when
density is computed based on interpolated salinity and temperature.
See reference 2 for a discussion of this issue and possible solutions.
}
\section{A note about flags}{


Data-quality flags contained within the original object are ignored by this
function, and the returned value contains no such flags.  This is because such
flags represent an assessment of the original data, not of quantities derived
from those data.  This function produces a warning to this effect. The
recommended practice is to use \code{\link[=handleFlags]{handleFlags()}} or some other means to
deal with flags before calling the present function.
}

\examples{
library(oce)
data(ctd)
plotProfile(ctd, "salinity", ylim=c(10, 0))
p <- seq(0, 45, 1)
ctd2 <- ctdDecimate(ctd, p=p)
lines(ctd2[["salinity"]], ctd2[["pressure"]], col="blue")
p <- seq(0, 45, 1)
ctd3 <- ctdDecimate(ctd, p=p, method=function(x, y, xout)
                    predict(smooth.spline(x, y, df=30), xout)$y)
lines(ctd3[["salinity"]], ctd3[["pressure"]], col="red")


}
\references{
\enumerate{
\item R.F. Reiniger and C.K. Ross, 1968.  A method of interpolation with
application to oceanographic data.  \emph{Deep Sea Research}, \strong{15},
185-193.
\item Oguma, Sachiko, Toru Suzuki, Yutaka Nagata, Hidetoshi Watanabe, Hatsuyo Yamaguchi,
and Kimio Hanawa. \dQuote{Interpolation Scheme for Standard Depth Data Applicable for Areas
with a Complex Hydrographical Structure.} Journal of Atmospheric and Oceanic Technology
21, no. 4 (April 1, 2004): 704-15.
}
}
\seealso{
The documentation for \linkS4class{ctd} explains the structure of
CTD objects, and also outlines the other functions dealing with them.

Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}
}
\author{
Dan Kelley
}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/accessors.R
\name{oceDeleteData}
\alias{oceDeleteData}
\title{Delete Something in an oce data Slot}
\usage{
oceDeleteData(object, name)
}
\arguments{
\item{object}{an \linkS4class{oce} object.}

\item{name}{String indicating the name of the item to be deleted.}
}
\description{
Return a copy of the supplied object that lacks the named
element in its \code{data} slot, and that has a note
about the deletion in its processing log.
}
\seealso{
Other things related to the data slot: 
\code{\link{oceGetData}()},
\code{\link{oceRenameData}()},
\code{\link{oceSetData}()}
}
\author{
Dan Kelley
}
\concept{things related to the data slot}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/AllClass.R
\name{initializeFlags,oce-method}
\alias{initializeFlags,oce-method}
\title{Create storage for a flag, and initialize values, for a oce object}
\usage{
\S4method{initializeFlags}{oce}(
  object,
  name = NULL,
  value = NULL,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{object}{An \linkS4class{oce} object.}

\item{name}{Character value indicating the name of a variable within the
\code{data} slot of \code{object}.}

\item{value}{Numerical or character value to be stored in the newly-created
entry within \code{flags}. (A character value will only work if
\code{\link[=initializeFlags]{initializeFlags()}} has been used first on \code{object}.)}

\item{debug}{Integer set to 0 for quiet action or to 1 for some debugging.}
}
\value{
An object with the \code{flags} item within the
\code{metadata} slot set up as indicated.
}
\description{
This function creates an item for a named variable within
the \code{flags} entry in the object's \code{metadata} slot.
The purpose is both to document a flag scheme
and to make it so that \code{\link[=initializeFlags]{initializeFlags()}}
and \code{\link[=setFlags]{setFlags()}} can specify flags by
name, in addition to number. A generic function, it is
specialized for some classes via interpretation of the
\code{scheme} argument (see \dQuote{Details}, for those
object classes that have such specializations).
}
\details{
If \code{object} already contains a \code{flags} entry with
the indicated name, then it is returned unaltered, and a warning
is issued.
}
\seealso{
Other functions relating to data-quality flags: 
\code{\link{defaultFlags}()},
\code{\link{handleFlags,adp-method}},
\code{\link{handleFlags,argo-method}},
\code{\link{handleFlags,ctd-method}},
\code{\link{handleFlags,oce-method}},
\code{\link{handleFlags,section-method}},
\code{\link{handleFlags}()},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{initializeFlagScheme,oce-method}},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{initializeFlagSchemeInternal}()},
\code{\link{initializeFlagScheme}()},
\code{\link{initializeFlags,adp-method}},
\code{\link{initializeFlagsInternal}()},
\code{\link{initializeFlags}()},
\code{\link{setFlags,adp-method}},
\code{\link{setFlags,ctd-method}},
\code{\link{setFlags,oce-method}},
\code{\link{setFlags}()}
}
\concept{functions relating to data-quality flags}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/astronomy.R
\name{julianCenturyAnomaly}
\alias{julianCenturyAnomaly}
\title{Julian-Day number to Julian century}
\usage{
julianCenturyAnomaly(jd)
}
\arguments{
\item{jd}{a julian day number, e.g. as given by \code{\link[=julianDay]{julianDay()}}.}
}
\value{
Julian century since noon on January 1, 1900.
}
\description{
Convert a Julian-Day number to a time in julian centuries since noon on
January 1, 1900.
The method follows reference 1 equation 15.1.  The example reproduces the
example provided by reference 1 example 15.a, with fractional error 3e-8.
}
\examples{

t <- ISOdatetime(1978, 11, 13, 4, 35, 0, tz="UTC")
jca <- julianCenturyAnomaly(julianDay(t))
cat(format(t), "is Julian Century anomaly", format(jca, digits=8), "\n")

}
\references{
\itemize{
\item Meeus, Jean. Astronomical Formulas for Calculators. Second Edition.
Richmond, Virginia, USA: Willmann-Bell, 1982.
}
}
\seealso{
Other things related to astronomy: 
\code{\link{angle2hms}()},
\code{\link{eclipticalToEquatorial}()},
\code{\link{equatorialToLocalHorizontal}()},
\code{\link{julianDay}()},
\code{\link{moonAngle}()},
\code{\link{siderealTime}()},
\code{\link{sunAngle}()},
\code{\link{sunDeclinationRightAscension}()}

Other things related to time: 
\code{\link{ctimeToSeconds}()},
\code{\link{julianDay}()},
\code{\link{numberAsHMS}()},
\code{\link{numberAsPOSIXct}()},
\code{\link{secondsToCtime}()},
\code{\link{unabbreviateYear}()}
}
\author{
Dan Kelley
}
\concept{things related to astronomy}
\concept{things related to time}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adv.sontek.R
\name{read.adv.sontek.adr}
\alias{read.adv.sontek.adr}
\title{Read an ADV data file}
\usage{
read.adv.sontek.adr(
  file,
  from = 1,
  to,
  by = 1,
  tz = getOption("oceTz"),
  header = TRUE,
  longitude = NA,
  latitude = NA,
  debug = getOption("oceDebug"),
  monitor = FALSE,
  processingLog = NULL
)
}
\arguments{
\item{file}{a connection or a character string giving the name of the file
to load.  It is also possible to give \code{file} as a vector of filenames,
to handle the case of data split into files by a data logger.  In the
multi-file case, \code{header} must be \code{FALSE}, \code{start} must be a
vector of times, and \code{deltat} must be provided.}

\item{from}{index number of the first profile to be read, or the time of
that profile, as created with \code{\link[=as.POSIXct]{as.POSIXct()}} (hint: use
\code{tz="UTC"}).  This argument is ignored if \code{header==FALSE}. See
\dQuote{Examples}.}

\item{to}{indication of the last profile to read, in a format matching that
of \code{from}.  This is ignored if \code{header==FALSE}.}

\item{by}{an indication of the stride length to use while walking through
the file.  This is ignored if \code{header==FALSE}.  Otherwise, if this is
an integer, then \code{by-1} profiles are skipped between each pair of
profiles that is read. This may not make much sense, if the data are not
equi-spaced in time.  If \code{by} is a string representing a time interval,
in colon-separated format, then this interval is divided by the sampling
interval, to get the stride length. \emph{BUG:} \code{by} only partially
works; see \dQuote{Bugs}.}

\item{tz}{character string indicating time zone to be assumed in the data.}

\item{header}{A logical value indicating whether the file starts with a header.
(This will not be the case for files that are created by data loggers that
chop the raw data up into a series of sub-files, e.g. once per hour.)}

\item{longitude}{optional signed number indicating the longitude in degrees
East.}

\item{latitude}{optional signed number indicating the latitude in degrees
North.}

\item{debug}{a flag that turns on debugging.  The value indicates the depth
within the call stack to which debugging applies.  For example,
\code{read.adv.nortek()} calls \code{read.header.nortek()}, so that
\code{read.adv.nortek(...,debug=2)} provides information about not just the
main body of the data file, but also the details of the header.}

\item{monitor}{boolean value indicating whether to indicate the progress
of reading the file, by using \code{\link[=txtProgressBar]{txtProgressBar()}} or otherwise.  The value
of \code{monitor} is changed to \code{FALSE} automatically, for non-interactive
sessions.}

\item{processingLog}{if provided, the action item to be stored in the log.
This parameter is typically only provided for internal calls; the default
that it provides is better for normal calls by a user.}
}
\value{
An \linkS4class{adv} object that contains measurements made with an ADV device.

The \code{metadata} contains information as given in the following table.
The \verb{Nortek name'' is the name used in the Nortek System Integrator Guide (reference 1) and the }Sontek name'' is the name used in the relevant
Sontek documentation.  References are given in square brackets.

\tabular{llll}{

\strong{\code{metadata} name}\tab \strong{Nortek name} \tab \strong{Sontek name} \tab \strong{Meaning}\cr

\code{manufacturer}\tab - \tab - \tab Either \code{"nortek"} or \code{"sontek"}\cr

\code{instrumentType}\tab  - \tab - \tab Either \code{"vector"} or \code{"adv"}\cr

\code{filename}\tab  - \tab - \tab Name of data file(s)\cr

\code{latitude}\tab - \tab - \tab  Latitude of mooring (if applicable)\cr

\code{longitude}\tab - \tab - \tab Longitude of mooring (if applicable)\cr

\code{numberOfSamples}\tab - \tab - \tab Number of data samples in file\cr

\code{numberOfBeams}\tab NBeams (reference 1, p18) \tab - \tab Number of beams (always 3)\cr

\code{numberOfBeamSequencesPerBurst}\tab NPings \tab - \tab number of beam sequences per burst\cr

\code{measurementInterval}\tab MeasInterval (reference 1 p31) \tab - \tab \cr

\code{samplingRate}\tab 512/(AvgInterval) (reference 1 p30; reference 4) \tab - #' \tab \cr

}

The \code{data} list contains items with names corresponding to \code{adp}
objects, with an exception for Nortek data.  Nortek instruments report some
things at a time interval that is longer than the velocity sampling, and
these are stored in \code{data} as \code{timeSlow}, \code{headingSlow},
\code{pitchSlow}, \code{rollSlow}, and \code{temperatureSlow}; if burst
sampling was used, there will also be items \code{recordsBurst} and
\code{timeBurst}.

The \code{processingLog} is in the standard format.
}
\description{
Read an ADV data file, producing an object of type \code{adv}. This
function works by transferring control to a more specialized function,
e.g. \code{\link[=read.adp.nortek]{read.adp.nortek()}} and \code{\link[=read.adp.sontek]{read.adp.sontek()}},
and in many cases users will find it preferable to either use these or
the several even more specialized functions, if the file type is
known.
}
\details{
Files \emph{without} headers may be created in experiments in which a data
logger was set up to monitor the serial data stream from an instrument.  The
lack of header information places a burden on the user, who must supply such
basic information as the times of observations, the instrument orientation,
the instrument coordinate system, etc.  Example 3 below shows how to deal
with such files.  Three things should be noted.
\enumerate{
\item The user must choose the appropriate \code{read.adv} variant
corresponding to the instrument in question.  (This is necessary because
\code{\link[=oceMagic]{oceMagic()}}, which is used by the generic \code{\link[=read.oce]{read.oce()}}
routine, cannot determine the type of instrument by examining a file that
lacks a header.)
\item The call to the \code{read} function must include a start time
(\code{start}) and the number of seconds between data (\code{deltat}),
again, because the instrument data stream may lack those things when the
device is set to a serial mode.  Also, of course, it is necessary to set
\code{header=FALSE} in the function call.
\item Once the file has been read in, the user will be obliged to specify
other information, for the object to be well-formed.  For example, the
\code{read} function will have no way of knowing the instrument orientation,
the coordinate system being used, the transformation matrix to go from
\code{"beam"} to \code{"xyz"} coordinates, or the instrument heading, pitch,
and roll, to go from \code{"xyz"} coordinates to \code{"enu"} coordinates.
Such things are illustrated in example 3 below.
}

In ADV data files, velocities are coded to signed 2-byte integers, with a
scale factor being used to convert to velocity in metres per second.  These
two facts control the maximum recordable velocity and the velocity
resolution, values that may be retrieved for an ADV object name \code{d}
with \code{d[["velocityMaximum"]]} and \code{d[["velocityResolution"]]}.
}
\section{Nortek files}{


\strong{Sampling-rate and similar issues}

The data format is inferred from the System Integrator Guide (reference 1A) and System
Integrator Manual (reference 1B).  These document lacks clarity in spots, and so
\code{read.adv.nortek} contains some assumptions that are noted here, so
that users will be aware of possible problems.

A prominent example is the specification of the sampling rate, stored in
\code{metadata$sampingRate} in the return value.  Repeated examination of
the System Integrator Guide (reference 1) failed to indicate where this value is
stored in the various headers contained in Vector datasets.  After some
experimentation with a few data files, \code{read.adv.nortek} was set up to
calculate \code{metadata$samplingRate} as \code{512/AvgInterval} where
\code{AvgInterval} is a part of the \verb{User Configuration'' header (reference 1 p30), where the explanation is }average interval in seconds'').  This formula was
developed through trial and error, but it was confirmed later on the Nortek
discussion group, and it should appear in upcoming versions of (reference 1).

Another basic issue is the determination of whether an instrument had
recorded in continuous mode or burst mode.  One might infer that
\code{TimCtrlReg} in the \verb{User Configuration'' header (reference 1 p30) determines this, in bits 1 and 2.  However, this was the case in test files available to the author.  For this reason, `read.adv.nortek` infers the mode by reverse engineering of data files of known configuration.  The present version of `read.adv.nortek` determines the sampling mode from the ```NRecords`'' item of the }Vector Velocity Data'' header, which seems
to be 0 for data collected continuously, and non-zero for data collected in
bursts.

Taking these things together, we come upon the issue of how to infer
sampling times for Nortek instruments.  There do not seem to be definitive
documents on this, and so \code{read.adv.nortek} is based partly on
information (of unknown quality) found on Nortek discussion boards.  The
present version of \code{read.adv.nortek} infers the times of velocity
observations differently, depending on whether the instrument was set to
record in burst mode or continuous mode.  For burst mode, times stated in
the burst headers are used, but for continuous mode, times stated in the
``vector system data'' are used.  On the advice found on a Nortek discussion
board, the burst-mode times are offset by 2 seconds to allow for the
instrument warm-up period.

\strong{Handling IMU (inertial measurement unit) data}

Starting in March
2016, \code{read.adv.nortek} has offered some support for handling IMU
(inertial measurement unit) data incorporated into Nortek binary files. This
is not described in the Nortek document named \verb{System Integrator Guide'' (reference 1A) but it appeared in }System Integrator Manual'' (reference 1B;
reference 1C). Confusingly, 1B described 3 varieties of data, whereas 1C does not
describe any of these, but describes instead a fourth variety. As of March
2016, \code{read.adv.nortek} handles all 4 varieties, because files in the
various schemes appear to exist. In \code{oce}, the varieties are named
after the byte code that flags them.  (Variety \code{c3} is the one
described in (reference 1C); the others were described in (reference 1B).)
The variety is stored in the \code{metadata} slot of the returned object as
a string named \code{IMUtype}.

For each variety, the reader is cautioned that strong tests have not been
performed on the code.  One way to test the code is to compare with textual
data files produced by the Nortek software.  In March 2016, an \code{oce}
user shared a dataset of the \code{c3} variety, and this permitted detailed
comparison between the text file and the values inferred by
\code{read.adv.nortek}. The test suggested agreement (to within the
resolution printed in the text file) for velocity (\code{v} in the
\code{data} slot), signal amplitude (\code{a}), correlation (\code{q}),
pressure (\code{p}), the three components of IMU delta angle
(\code{IMUdeltaAngleX} etc), and all components of the rotation matrix
(\code{IMUrotation}).  However, the delta velocity signals did not match,
with \code{IMUdeltaVelocityX} disagreeing in the second decimal place,
\code{IMUdeltaVelocityY} component disagreeing in the first, and
\code{IMUdeltaVelocityZ} being out by a factor of about 10. This is github
issue 893 (\url{https://github.com/dankelley/oce/issues/893}).
\itemize{
\item Variety \code{c3} (signalled by byte 5 of a sequence being
\code{0xc3}) provides information on what Nortek calls DeltaAngle,
DeltaVelocity and Orientation Matrix. (Apart from the orientation matrix,
Nortek provides no documentation on what these quantities mean.) In the
object returned by \code{read.adv.nortek}, these are stored in the
\code{data} slot as \code{IMUdeltaAngleX}, \code{IMUdeltaAngleY},
\code{IMUdeltaAngleZ}, \code{IMUdeltaVelocityX}, \code{IMUdeltaVelocityY},
\code{IMUdeltaVelocityZ}, and \code{IMUrotation}, all vectors except the
last, which is a 3D array. In addition to these, \code{IMUtimestamp} is a
time-stamp, which is not defined in the Nortek documents but seems, from IMU
documents (reference 5), to be defined based on a clock that ticks once per 16
microseconds. Caution may be required in dealing with this timestamp, since
it seemed sensible in one test case (variety \code{d3}) but kept reseting to
zero in another (variety \code{c3}). The lack of Nortek documentation on
most of these quantities is a roadblock to implementing \code{oce} functions
dealing with IMU-enabled datasets
\item Variety \code{cc} (signalled by byte 5 of a sequence being
\code{0xcc}) provides information on acceleration, angular rotation rate,
magnetic vector and orientation matrix.  Each is a timeseries. Acceleration
is stored in the \code{data} slot as \code{IMUaccelX}, \code{IMUaccelY},
\code{IMUaccelz}. The angular rotation components are \code{IMUngrtX},
\code{IMUngrtY} and \code{IMUngrtz}. The magnetic data are in
\code{IMUmagrtx}, \code{IMUmagrty} and \code{IMUmagrtz}. Finally,
\code{IMUmatrix} is a rotation matrix made up from elements named
\code{M11}, \code{M12}, etc in the Nortek documentation.  In addition to all
of these, \code{IMUtime} stores time in seconds, with an origin whose
definition is not stated in reference 1B.
\item Variety \code{d2} (signalled by byte 5 being \code{0xd2}) provides
information on gyro-stabilized acceleration, angular rate and magnetometer
vectors.  The data stored \code{MUaccelX}, \code{IMUangrtX},
\code{IMUmagrtX}, with similar for \code{Y} and \code{Z}.  Again, time is in
\code{IMUtime}. This data type has not been tested as of mid-March 2016,
because the developers do not have a test file with which to test.
\item Variety \code{d3} (signalled by byte 5 being \code{0xd3}) provides
information on DeltaAngle, DeltaVelocity and magnetometer vectors, stored in
\code{IMUdeltaAngleX}, \code{IMUdeltaVelocityX}, and
\code{IMUdeltaMagVectorX}, with similar for \code{Y} and \code{Z}. Again,
time is in \code{IMUtime}. This data type has not been tested as of
mid-March 2016, because the developers do not have a test file with which to
test.
}
}

\examples{
\dontrun{
library(oce)
# A nortek Vector file
d <- read.oce("/data/archive/sleiwex/2008/moorings/m05/adv/nortek_1943/raw/adv_nortek_1943.vec",
              from=as.POSIXct("2008-06-26 00:00:00", tz="UTC"),
              to=as.POSIXct("2008-06-26 00:00:10", tz="UTC"))
plot(d, which=c(1:3,15))
}
}
\references{
1A. Nortek AS.  System Integrator Guide (paradopp family of products). June
2008.  (Doc No: PSI00-0101-0608).  (Users may find it helpful to also
examine newer versions of the guide.)

1B. Nortek AS.  System Integrator Manual. Dec 2014.
(\code{system-integrator-manual_Dec2014_jan.pdf})

1C. Nortek AS.  System Integrator Manual. March 2016.
(\code{system-integrator-manual_Mar2016.pdf})
\enumerate{
\item SonTek/YSI ADVField/Hydra Acoustic Doppler Velocimeter (Field) Technical
Documentation (Sept 1, 2001).
\item Appendix 2.2.3 of the Sontek ADV operation Manual, Firmware Version 4.0
(Oct 1997).
\item Nortek Knowledge Center (http://www.nortekusa.com/en/knowledge-center)
\item A document describing an IMU unit that seems to be close to the one named
in (references 1B and C) as being an adjunct to Nortek Vector systems is at
\verb{http://files.microstrain.com/3DM-GX3-35-Data-Communications-Protocol.pdf}
}
}
\seealso{
Other things related to adv data: 
\code{\link{[[,adv-method}},
\code{\link{[[<-,adv-method}},
\code{\link{adv-class}},
\code{\link{adv}},
\code{\link{beamName}()},
\code{\link{beamToXyz}()},
\code{\link{enuToOtherAdv}()},
\code{\link{enuToOther}()},
\code{\link{plot,adv-method}},
\code{\link{read.adv.nortek}()},
\code{\link{read.adv.sontek.serial}()},
\code{\link{read.adv.sontek.text}()},
\code{\link{read.adv}()},
\code{\link{rotateAboutZ}()},
\code{\link{subset,adv-method}},
\code{\link{summary,adv-method}},
\code{\link{toEnuAdv}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdv}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley
}
\concept{things related to adv data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/lobo.R
\docType{class}
\name{lobo-class}
\alias{lobo-class}
\title{Class to Store LOBO Data}
\description{
This class stores LOBO data.
}
\section{Slots}{

\describe{
\item{\code{data}}{As with all \code{oce} objects, the \code{data} slot
for \code{lobo} objects is a
\link{list} containing the main data for the object.
{}}

\item{\code{metadata}}{As with all \code{oce} objects, the \code{metadata} slot
for \code{lobo} objects is a \link{list} containing
information about the \code{data} or about the object itself.
{}}

\item{\code{processingLog}}{As with all \code{oce} objects, the \code{processingLog} slot
for \code{lobo} objects is a
\link{list} with entries describing the creation and evolution
of the object. The contents are updated by various \code{oce} functions to
keep a record of processing steps.  Object summaries and
\code{\link[=processingLogShow]{processingLogShow()}} both display the log.}
}}

\section{Modifying slot contents}{


Although the \code{[[<-} operator may permit modification of the contents
of \linkS4class{lobo} objects (see \code{\link{[[<-,lobo-method}}),
it is better to use \code{\link[=oceSetData]{oceSetData()}} and \code{\link[=oceSetMetadata]{oceSetMetadata()}},
because those functions save an entry in the \code{processingLog}
that describes the change.
}

\section{Retrieving slot contents}{


The full contents of the \code{data} and \code{metadata} slots of a \linkS4class{lobo}
object may be retrieved in the standard R way using \code{\link[=slot]{slot()}}. For
example \code{slot(o,"data")} returns the \code{data} slot of an object named \code{o},
and similarly \code{slot(o,"metadata")} returns
the \code{metadata} slot.

The slots may also be obtained with the \code{\link{[[,lobo-method}}
operator, as e.g. \code{o[["data"]]} and \code{o[["metadata"]]}, respectively.

The \code{\link{[[,lobo-method}} operator can also
be used to retrieve items from within the \code{data} and \code{metadata} slots.
For example, \code{o[["temperature"]]} can be used to retrieve temperature
from an object containing that quantity. The rule is that a named
quantity is sought first within the object's \code{metadata} slot,
with the \code{data} slot being checked only if \code{metadata} does not
contain the item. This \code{[[} method can also be used to get
certain derived quantities, if the object contains sufficient
information to calculate them. For example, an object that holds
(practical) salinity, temperature and pressure, along with
longitude and latitude, has sufficient information to compute
Absolute Salinity, and so \code{o[["SA"]]} will yield the
calculated Absolute Salinity.

It is also possible to find items more directly, using \code{\link[=oceGetData]{oceGetData()}} and
\code{\link[=oceGetMetadata]{oceGetMetadata()}}, but neither of these functions can
retrieve derived items.
}

\seealso{
Other classes provided by oce: 
\code{\link{adp-class}},
\code{\link{adv-class}},
\code{\link{argo-class}},
\code{\link{bremen-class}},
\code{\link{cm-class}},
\code{\link{coastline-class}},
\code{\link{ctd-class}},
\code{\link{lisst-class}},
\code{\link{met-class}},
\code{\link{oce-class}},
\code{\link{odf-class}},
\code{\link{rsk-class}},
\code{\link{sealevel-class}},
\code{\link{section-class}},
\code{\link{topo-class}},
\code{\link{windrose-class}},
\code{\link{xbt-class}}

Other things related to lobo data: 
\code{\link{[[,lobo-method}},
\code{\link{[[<-,lobo-method}},
\code{\link{as.lobo}()},
\code{\link{lobo}},
\code{\link{plot,lobo-method}},
\code{\link{read.lobo}()},
\code{\link{subset,lobo-method}},
\code{\link{summary,lobo-method}}
}
\author{
Dan Kelley
}
\concept{classes provided by oce}
\concept{things related to lobo data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.R
\name{[[<-,ctd-method}
\alias{[[<-,ctd-method}
\title{Replace Parts of a ctd Object}
\usage{
\S4method{[[}{ctd}(x, i, j, ...) <- value
}
\arguments{
\item{x}{a \linkS4class{ctd} object.}

\item{i}{character value naming the item to replace.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{optional additional information (ignored).}

\item{value}{The value to be placed into \code{x}, somewhere.}
}
\description{
The \code{[[<-} method works for all \linkS4class{oce} objects.
The purpose, as with the related extraction method, \code{[[},
is to insulate users from the internal details of \linkS4class{oce}
objects, by looking for items within the various storage
slots of the object. Items not actually stored can also be
replaced, including units and data-quality
flags.
}
\details{
As with \code{[[} method, the procedure works in steps.

First, the \code{metadata} slot of \code{x} is checked to
see whether it contains something named with \code{i}.
If so, then the named item is replaced with \code{value}.

Otherwise, if the string value of \code{i} ends in \code{Unit}, then the
characters preceding that are taken as the name of a variable, and
the \code{metadata} slot of \code{x} is updated to store that unit, e.g.\preformatted{x[["temperatureUnits"]] <- list(unit=expression(degree*F),scale="")
}

Similarly, if \code{i} ends in \code{Flag}, then quality-control
flags are set up as defined by \code{result}, e.g.\preformatted{o[["temperatureFlags"]] <- c(2,4,2,2)
}

Otherwise, \code{\link[=pmatch]{pmatch()}} is used for a partial-string match with
the names of the items that are in the \code{data} slot of \code{x}.
The first item found (if any) is then updated to hold the value \code{result}.

If none of these conditions is met, a warning is issued.
}
\examples{
data(ctd)
summary(ctd)
# Move the CTD profile a nautical mile north.
ctd[["latitude"]] <- 1/60 + ctd[["latitude"]] # acts in metadata
# Increase the salinity by 0.01.
ctd[["salinity"]] <- 0.01 + ctd[["salinity"]] # acts in data
summary(ctd)

}
\seealso{
Other functions that replace parts of oce objects: 
\code{\link{[[<-,adp-method}},
\code{\link{[[<-,amsr-method}},
\code{\link{[[<-,argo-method}},
\code{\link{[[<-,bremen-method}},
\code{\link{[[<-,cm-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{[[<-,echosounder-method}},
\code{\link{[[<-,g1sst-method}},
\code{\link{[[<-,gps-method}},
\code{\link{[[<-,ladp-method}},
\code{\link{[[<-,landsat-method}},
\code{\link{[[<-,lisst-method}},
\code{\link{[[<-,lobo-method}},
\code{\link{[[<-,met-method}},
\code{\link{[[<-,oce-method}},
\code{\link{[[<-,odf-method}},
\code{\link{[[<-,rsk-method}},
\code{\link{[[<-,sealevel-method}},
\code{\link{[[<-,section-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{[[<-,topo-method}},
\code{\link{[[<-,windrose-method}},
\code{\link{[[<-,xbt-method}}

Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}
}
\concept{functions that replace parts of oce objects}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/map.R
\name{mapLocator}
\alias{mapLocator}
\title{Locate Points on a Map}
\usage{
mapLocator(n = 512, type = "n", ...)
}
\arguments{
\item{n}{number of points to locate; see \code{\link[=locator]{locator()}}.}

\item{type}{type of connector for the points; see \code{\link[=locator]{locator()}}.}

\item{\dots}{extra arguments passed to \code{\link[=locator]{locator()}} (and either
\code{\link[=mapPoints]{mapPoints()}} or \code{\link[=mapLines]{mapLines()}}, if appropriate) if
\code{type} is not \code{'n'}.}
}
\description{
Locate points on an existing map.
This uses \code{\link[=map2lonlat]{map2lonlat()}} to infer the location in
geographical space, so it suffers the same
limitations as that function.
}
\seealso{
A map must first have been created with \code{\link[=mapPlot]{mapPlot()}}.

Other functions related to maps: 
\code{\link{formatPosition}()},
\code{\link{lonlat2map}()},
\code{\link{lonlat2utm}()},
\code{\link{map2lonlat}()},
\code{\link{mapArrows}()},
\code{\link{mapAxis}()},
\code{\link{mapContour}()},
\code{\link{mapCoordinateSystem}()},
\code{\link{mapDirectionField}()},
\code{\link{mapGrid}()},
\code{\link{mapImage}()},
\code{\link{mapLines}()},
\code{\link{mapLongitudeLatitudeXY}()},
\code{\link{mapPlot}()},
\code{\link{mapPoints}()},
\code{\link{mapPolygon}()},
\code{\link{mapScalebar}()},
\code{\link{mapText}()},
\code{\link{mapTissot}()},
\code{\link{oceCRS}()},
\code{\link{shiftLongitude}()},
\code{\link{usrLonLat}()},
\code{\link{utm2lonlat}()}
}
\author{
Dan Kelley
}
\concept{functions related to maps}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/AllClass.R
\name{handleFlagsInternal}
\alias{handleFlagsInternal}
\title{Low-level function for handling data-quality flags}
\usage{
handleFlagsInternal(object, flags, actions, where, debug = 0)
}
\arguments{
\item{object}{an \linkS4class{oce} object.}

\item{flags}{a named \link{list} of numeric values.}

\item{actions}{A character vector indicating actions to be carried out for the corresponding
\code{flags} values. This will be lengthened with \code{\link[=rep]{rep()}} if necessary, to be
of the same length as \code{flags}. A common value for \code{actions} is \code{"NA"}, which
means that data values that are flagged are replaced by \code{NA} in the returned result.}

\item{where}{An optional string that permits the function to work with
objects that store flags in e.g. \code{object@metadata$flags$where}
instead of in \code{object@metadata$flags}, and data within
\code{object@data$where} instead of within \code{object@data}. The
appropriate value for \code{where} within the oce package is
the default, \code{NULL}, which means that this extra subdirectory
is not being used.}

\item{debug}{An integer indicating the degree of debugging requested, with value \code{0}
meaning to act silently, and value \code{1} meaning to print some information about the
steps in processing.}
}
\value{
A copy of \code{object}, possibly with modifications to its
\code{data} slot, if \code{object} contains flag values that have
actions that alter the data.
}
\description{
This function is designed for internal use within the \code{oce} package.  Its
purpose is to carry out low-level processing relating to data-quality flags,
as a support for higher-level functions such \link{handleFlags,ctd-method} for
\code{ctd} objects, \link{handleFlags,adp-method} for \code{adp} objects,
etc.
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{oceColorsVelocity}
\alias{oceColorsVelocity}
\alias{oce.colorsVelocity}
\title{Create colors suitable for velocity fields}
\usage{
oceColorsVelocity(n)
}
\arguments{
\item{n}{number of colors to create.}
}
\value{
A vector of color specifications.
}
\description{
Create a set of colors for displaying velocity values,
based on the scheme devised by Thyng et al. (2016) and
presented in a python package by Thyng (2019).
The color specifications were transliterated from
python to R on 2015-09-29, but have not been adjusted
since, even though the python source has changed.  This
is to prevent breaking old \code{oce} code.  To get the
latest versions of these colours or other colours,
use the \CRANpkg{cmocean} R package
(Thyng, Richards, and Krylov, 2019) directly, as is
illustrated (with the "matter" scheme) in Example 2.
Note that the \CRANpkg{cmocean} core functions provide a way
to select between various versions of the colour schemes.
It is also worth considering the palettes provided by the
\CRANpkg{viridis} package, as illustrated (with the
"inferno" scheme) in Example 3.
}
\examples{
library(oce)

# Example 1
imagep(volcano, col=oceColorsVelocity(128),
       zlab="oceColorsVelocity")
\dontrun{
# Example 2 (requires the cmocean package)
imagep(volcano, col=cmocean::cmocean("matter"),
       zlab='cmocean::cmocean("matter")')}

\dontrun{
# Example 3 (requires the viridis package)
imagep(volcano, col=viridis::inferno,
       zlab='viridis::inferno')}

}
\references{
\itemize{
\item Thyng, Kristen, Chad Greene, Robert Hetland, Heather Zimmerle, and Steven DiMarco.
\dQuote{True Colors of Oceanography: Guidelines for Effective and Accurate Colormap Selection.}
Oceanography 29, no. 3 (September 1, 2016): 9‚Äì13.
\doi{10.5670/oceanog.2016.66}
\item Thyng, Kristen. Kthyng/Cmocean. Python, 2019.
\url{https://github.com/kthyng/cmocean}.
\item Thyng, Kristen, Clark Richards, and Ivan Krylov.
Cmocean: Beautiful Colour Maps for Oceanography (version 0.2), 2019.
\url{https://CRAN.R-project.org/package=cmocean}.
}

The following references provide information on choosing
colour schemes, that are suitable for viewers who have
colour deficiencies.

Light, Adam, and Patrick J. Bartlein. "The End of the Rainbow? Color
Schemes for Improved Data Graphics." \emph{Eos, Transactions American Geophysical
Union} 85, no. 40 (2004): 385.
DOI: 10.1029/2004EO400002

Stephenson, David B. "Comment on 'Color Schemes for Improved Data
Graphics', by A Light and P.J. Bartlein." \emph{Eos, Transactions American
Geophysical Union} 86, no.  20 (2005): 196.
DOI: 10.1029/2005EO200005

Light, Adam, and Patrick J. Bartlein. "Reply to 'Comment on Color
Schemes for Improved Data Graphics,' by A. Light and P.J.  Bartlein'."
\emph{Eos, Transactions American Geophysical Union} 86, no. 20 (2005): 196‚Äì196.
DOI: 10.1029/2005EO200006
}
\seealso{
Other things related to colors: 
\code{\link{colormapGMT}()},
\code{\link{colormap}()},
\code{\link{oceColors9B}()},
\code{\link{oceColorsCDOM}()},
\code{\link{oceColorsChlorophyll}()},
\code{\link{oceColorsClosure}()},
\code{\link{oceColorsDensity}()},
\code{\link{oceColorsFreesurface}()},
\code{\link{oceColorsGebco}()},
\code{\link{oceColorsJet}()},
\code{\link{oceColorsOxygen}()},
\code{\link{oceColorsPAR}()},
\code{\link{oceColorsPalette}()},
\code{\link{oceColorsPhase}()},
\code{\link{oceColorsSalinity}()},
\code{\link{oceColorsTemperature}()},
\code{\link{oceColorsTurbidity}()},
\code{\link{oceColorsTurbo}()},
\code{\link{oceColorsTwo}()},
\code{\link{oceColorsViridis}()},
\code{\link{oceColorsVorticity}()},
\code{\link{ocecolors}}
}
\author{
Krysten M. Thyng (Python version), Dan Kelley (R transliteration)
}
\concept{things related to colors}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/AllClass.R
\name{initializeFlagsInternal}
\alias{initializeFlagsInternal}
\title{Create storage for a flag, and initialize values, for a oce object}
\usage{
initializeFlagsInternal(
  object,
  name = NULL,
  value = NULL,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{object}{An \linkS4class{oce} object.}

\item{name}{Character value indicating the name of a variable within the
\code{data} slot of \code{object}.}

\item{value}{Numerical or character value to be stored in the newly-created
entry within \code{flags}. (A character value will only work if
\code{\link[=initializeFlags]{initializeFlags()}} has been used first on \code{object}.)}

\item{debug}{Integer set to 0 for quiet action or to 1 for some debugging.}
}
\value{
An object with the \code{flags} item within the
\code{metadata} slot set up as indicated.
}
\description{
This function creates an item for a named variable within
the \code{flags} entry in the object's \code{metadata} slot.
The purpose is both to document a flag scheme
and to make it so that \code{\link[=initializeFlags]{initializeFlags()}}
and \code{\link[=setFlags]{setFlags()}} can specify flags by
name, in addition to number. A generic function, it is
specialized for some classes via interpretation of the
\code{scheme} argument (see \dQuote{Details}, for those
object classes that have such specializations).
}
\details{
If \code{object} already contains a \code{flags} entry with
the indicated name, then it is returned unaltered, and a warning
is issued.
}
\seealso{
Other functions relating to data-quality flags: 
\code{\link{defaultFlags}()},
\code{\link{handleFlags,adp-method}},
\code{\link{handleFlags,argo-method}},
\code{\link{handleFlags,ctd-method}},
\code{\link{handleFlags,oce-method}},
\code{\link{handleFlags,section-method}},
\code{\link{handleFlags}()},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{initializeFlagScheme,oce-method}},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{initializeFlagSchemeInternal}()},
\code{\link{initializeFlagScheme}()},
\code{\link{initializeFlags,adp-method}},
\code{\link{initializeFlags,oce-method}},
\code{\link{initializeFlags}()},
\code{\link{setFlags,adp-method}},
\code{\link{setFlags,ctd-method}},
\code{\link{setFlags,oce-method}},
\code{\link{setFlags}()}
}
\concept{functions relating to data-quality flags}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/geod.R
\name{geodXyInverse}
\alias{geodXyInverse}
\title{Inverse Geodesic Calculation}
\usage{
geodXyInverse(x, y, longitudeRef, latitudeRef, debug = getOption("oceDebug"))
}
\arguments{
\item{x}{value of x in metres, as given by \code{\link[=geodXy]{geodXy()}}}

\item{y}{value of y in metres, as given by \code{\link[=geodXy]{geodXy()}}}

\item{longitudeRef}{reference longitude, as supplied to \code{\link[=geodXy]{geodXy()}}}

\item{latitudeRef}{reference latitude, as supplied to \code{\link[=geodXy]{geodXy()}}}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\value{
a data frame containing \code{longitude} and \code{latitude}
}
\description{
The calculation is done by finding a minimum value of a cost
function that is the vector difference between (\code{x},\code{y})
and the corresponding values returned by \code{\link[=geodXy]{geodXy()}}.
See \dQuote{Caution}.
}
\details{
The minimum is calculated in C for speed, using the \code{nmmin} function
that is the underpinning for the Nelder-Meade version of the R function
\code{\link[=optim]{optim()}}. If you find odd results, try setting \code{debug=1}
and rerunning, to see whether this optimizer is having difficulty
finding a minimum of the mismatch function.
}
\section{Caution}{
 This scheme is without known precedent in the literature, and
users should read the documentation carefully before deciding to use it.
}

\seealso{
Other functions relating to geodesy: 
\code{\link{geodDist}()},
\code{\link{geodGc}()},
\code{\link{geodXy}()}
}
\concept{functions relating to geodesy}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.R
\name{as.ctd}
\alias{as.ctd}
\title{Coerce data into CTD object}
\usage{
as.ctd(
  salinity,
  temperature = NULL,
  pressure = NULL,
  conductivity = NULL,
  scan = NULL,
  time = NULL,
  units = NULL,
  flags = NULL,
  missingValue = NULL,
  type = "",
  serialNumber = NULL,
  ship = NULL,
  cruise = NULL,
  station = NULL,
  startTime = NULL,
  longitude = NULL,
  latitude = NULL,
  deploymentType = "unknown",
  pressureAtmospheric = 0,
  sampleInterval = NULL,
  profile = NULL,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{salinity}{may be (1) a numeric vector holding Practical Salinity,
(2) a list or data frame holding \code{salinity} and other
hydrographic variables or (3) an \code{oce-class} object that holds
hydrographic information. If \code{salinity} is not provided,
then \code{conductivity} must be provided, so that \code{\link[=swSCTp]{swSCTp()}}
can be used to compute salinity.}

\item{temperature}{a numeric vector containing \emph{in-situ} temperature in
\eqn{^\circ}{deg}C on the ITS-90 scale; see \dQuote{Temperature units} in the
documentation for \code{\link[=swRho]{swRho()}}.}

\item{pressure}{a numeric vector containing sea pressure values, in decibars.
Typically, this vector has the same length as \code{salinity} and \code{temperature},
but it also possible to supply just one value, which will be repeated
to get the right length. Note that \code{as.ctd()} stores the
sum of \code{pressure} and \code{pressureAtmospheric} in the returned object,
although the default value for \code{pressureAtmospheric} is zero, so
in the default case, \code{pressure} is stored directly.}

\item{conductivity}{an optional numeric vector containing electrical
conductivity ratio through the water column. To convert from raw conductivity
in milliSeimens per centimeter divide by 42.914 to get conductivity ratio
(see Culkin and Smith, 1980).}

\item{scan}{optional numeric vector holding scan number.  If not provided,
this is set to \link{seq_along}\code{(salinity)}.}

\item{time}{optional vector of times of observation.}

\item{units}{an optional list containing units.  If not supplied,
defaults are set for \code{pressure}, \code{temperature}, \code{salinity},
and \code{conductivity}. Since these are simply guesses, users
are advised strongly to supply \code{units}. See \dQuote{Examples}.}

\item{flags}{if supplied, this is a \link{list} containing data-quality
flags. The elements of this list must have names that match the data
provided to the object.}

\item{missingValue}{optional missing value, indicating data that should be
taken as \code{NA}. Set to \code{NULL} to turn off this feature.}

\item{type}{optional type of CTD, e.g. "SBE"}

\item{serialNumber}{optional serial number of instrument}

\item{ship}{optional string containing the ship from which the observations were made.}

\item{cruise}{optional string containing a cruise identifier.}

\item{station}{optional string containing a station identifier.}

\item{startTime}{optional indication of the start time for the profile,
which is used in some several plotting functions.  This is best given as a
\link{POSIXt} time, but it may also be a character string
that can be converted to a time with \code{\link[=as.POSIXct]{as.POSIXct()}},
using \code{UTC} as the timezone.}

\item{longitude}{optional numerical value containing longitude in decimal
degrees, positive in the eastern hemisphere. If this is a single number,
then it is stored in the \code{metadata} slot of the returned value; if it
is a vector of numbers, then they are stored in the \code{data} slot.}

\item{latitude}{optional numerical value containing the latitude in decimal
degrees, positive in the northern hemisphere. See the note on length, for
the \code{longitude} argument.}

\item{deploymentType}{character string indicating the type of deployment. Use
\code{"unknown"} if this is not known, \code{"profile"} for a profile (in
which the data were acquired during a downcast, while the device was lowered
into the water column, perhaps also including an upcast; \code{"moored"} if
the device is installed on a fixed mooring, \code{"thermosalinograph"} (or
\code{"tsg"}) if the device is mounted on a moving vessel, to record
near-surface properties, or \code{"towyo"} if the device is repeatedly
lowered and raised.}

\item{pressureAtmospheric}{A numerical value (a constant or a vector),
that is subtracted from pressure before storing it in the return value.
(This altered pressure is also used in calculating \code{salinity}, if
that is to be computed from \code{conductivity}, etc., using
\code{\link[=swSCTp]{swSCTp()}}; see \code{salinity} above.)}

\item{sampleInterval}{optional numerical value indicating the time between
samples in the profile.}

\item{profile}{optional positive integer specifying the number of the profile
to extract from an object that has data in matrices, such as for some
\code{argo} objects. Currently the \code{profile} argument is only utilized for
\linkS4class{argo} objects.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\value{
A \linkS4class{ctd} object.
}
\description{
Assemble data into a \linkS4class{ctd} object.  This function is complicated
(spanning approximately 500 lines of code) because it tries to handle many
special cases, and tries to make sensible defaults for unspecified
parameters.  If odd results are found, users might find it helpful to call
this function with the first argument being a simple vector of Practical
Salinity values, in which case the processing of the other arguments is
relatively straightforward.
}
\section{Converting rsk objects}{

If the \code{salinity} argument is an object of \linkS4class{rsk},
then \code{as.ctd} passes it,
\code{pressureAtmospheric},
\code{longitude},
\code{latitude}
\code{ship},
\code{cruise},
\code{station} and
\code{deploymentType}
to \code{\link[=rsk2ctd]{rsk2ctd()}}, which builds the ctd object that is
returned by \code{as.ctd}. The other arguments to \code{as.ctd}
are ignored in this instance, because \code{rsk} objects already
contain their information. If required, any data or metadata
element can be added to the value returned by \code{as.ctd}
using \code{\link[=oceSetData]{oceSetData()}} or \code{\link[=oceSetMetadata]{oceSetMetadata()}},
respectively.

The returned \linkS4class{rsk} object contains pressure in a form that
may need to be adjusted, because \code{rsk} objects may contain
either absolute pressure or sea pressure. This adjustment is handled
automatically by \code{as.ctd}, by examination of the metadata item
named \code{pressureType} (described in the documentation for
\code{\link[=read.rsk]{read.rsk()}}).  Once the sea pressure is determined,
adjustments may be made with the \code{pressureAtmospheric} argument,
although in that case it is better considered a pressure adjustment
than the atmospheric pressure.

\linkS4class{rsk} objects may store sea pressure or absolute pressure (the
sum of sea pressure and atmospheric pressure), depending on how the object was
created with \code{\link[=as.rsk]{as.rsk()}} or \code{\link[=read.rsk]{read.rsk()}}.  However,
\linkS4class{ctd} objects store sea pressure, which is needed for
plotting, calculating density, etc. This poses no difficulties, however,
because \code{as.ctd} automatically converts absolute pressure to sea pressure,
if the metadata in the \linkS4class{rsk} object indicates that this is
appropriate. Further alteration of the pressure can be accomplished with the
\code{pressureAtmospheric} argument, as noted above.
}

\examples{
library(oce)
## 1. fake data, with default units
pressure <- 1:50
temperature <- 10 - tanh((pressure - 20) / 5) + 0.02*rnorm(50)
salinity <- 34 + 0.5*tanh((pressure - 20) / 5) + 0.01*rnorm(50)
ctd <- as.ctd(salinity, temperature, pressure)
# Add a new column
fluo <- 5 * exp(-pressure / 20)
ctd <- oceSetData(ctd, name="fluorescence", value=fluo,
                  unit=list(unit=expression(mg/m^3), scale=""))
summary(ctd)

## 2. fake data, with supplied units (which are the defaults, actually)
ctd <- as.ctd(salinity, temperature, pressure,
    units=list(salinity=list(unit=expression(), scale="PSS-78"),
    temperature=list(unit=expression(degree*C), scale="ITS-90"),
    pressure=list(unit=expression(dbar), scale="")))

}
\references{
Culkin, F., and Norman D. Smith, 1980. Determination of the
concentration of potassium chloride solution having the same electrical
conductivity, at 15 C and infinite frequency, as standard seawater of salinity
35.0000 ppt (Chlorinity 19.37394 ppt). \emph{IEEE Journal of Oceanic
Engineering}, volume \strong{5}, pages 22-23.
}
\seealso{
Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}
}
\author{
Dan Kelley
}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/AllClass.R
\name{[[<-,oce-method}
\alias{[[<-,oce-method}
\title{Replace Parts of an Oce Object}
\usage{
\S4method{[[}{oce}(x, i, j, ...) <- value
}
\arguments{
\item{x}{an \linkS4class{oce} object.}

\item{i}{character value naming the item to replace.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{optional additional information (ignored).}

\item{value}{The value to be placed into \code{x}, somewhere.}
}
\description{
The \code{[[<-} method works for all \linkS4class{oce} objects.
The purpose, as with the related extraction method, \code{[[},
is to insulate users from the internal details of \linkS4class{oce}
objects, by looking for items within the various storage
slots of the object. Items not actually stored can also be
replaced, including units and data-quality
flags.
}
\details{
As with \code{[[} method, the procedure works in steps.

First, the \code{metadata} slot of \code{x} is checked to
see whether it contains something named with \code{i}.
If so, then the named item is replaced with \code{value}.

Otherwise, if the string value of \code{i} ends in \code{Unit}, then the
characters preceding that are taken as the name of a variable, and
the \code{metadata} slot of \code{x} is updated to store that unit, e.g.\preformatted{x[["temperatureUnits"]] <- list(unit=expression(degree*F),scale="")
}

Similarly, if \code{i} ends in \code{Flag}, then quality-control
flags are set up as defined by \code{result}, e.g.\preformatted{o[["temperatureFlags"]] <- c(2,4,2,2)
}

Otherwise, \code{\link[=pmatch]{pmatch()}} is used for a partial-string match with
the names of the items that are in the \code{data} slot of \code{x}.
The first item found (if any) is then updated to hold the value \code{result}.

If none of these conditions is met, a warning is issued.
}
\seealso{
Other functions that replace parts of oce objects: 
\code{\link{[[<-,adp-method}},
\code{\link{[[<-,amsr-method}},
\code{\link{[[<-,argo-method}},
\code{\link{[[<-,bremen-method}},
\code{\link{[[<-,cm-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{[[<-,echosounder-method}},
\code{\link{[[<-,g1sst-method}},
\code{\link{[[<-,gps-method}},
\code{\link{[[<-,ladp-method}},
\code{\link{[[<-,landsat-method}},
\code{\link{[[<-,lisst-method}},
\code{\link{[[<-,lobo-method}},
\code{\link{[[<-,met-method}},
\code{\link{[[<-,odf-method}},
\code{\link{[[<-,rsk-method}},
\code{\link{[[<-,sealevel-method}},
\code{\link{[[<-,section-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{[[<-,topo-method}},
\code{\link{[[<-,windrose-method}},
\code{\link{[[<-,xbt-method}}
}
\author{
Dan Kelley
}
\concept{functions that replace parts of oce objects}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swAlpha}
\alias{swAlpha}
\title{Seawater thermal expansion coefficient}
\usage{
swAlpha(
  salinity,
  temperature = NULL,
  pressure = 0,
  longitude = NULL,
  latitude = NULL,
  eos = getOption("oceEOS", default = "gsw")
)
}
\arguments{
\item{salinity}{either practical salinity (in which case \code{temperature}
and \code{pressure} must be provided) \emph{or} an \code{oce} object (in
which case \code{salinity}, etc. are inferred from the object).}

\item{temperature}{\emph{in-situ} temperature (\eqn{^\circ}{deg}C), defined
on the ITS-90 scale; see \dQuote{Temperature units} in the documentation for
\code{\link[=swRho]{swRho()}}.}

\item{pressure}{pressure (dbar)}

\item{longitude}{longitude of observation (only used if \code{eos="gsw"};
see \sQuote{Details}).}

\item{latitude}{latitude of observation (only used if \code{eos="gsw"}; see
\sQuote{Details}).}

\item{eos}{equation of state, either \code{"unesco"} or \code{"gsw"}.}
}
\value{
Value in 1/degC.
}
\description{
Compute \eqn{\alpha}{alpha}, the thermal expansion coefficient for seawater.
}
\references{
The \code{eos="unesco"} formulae are based on the UNESCO
equation of state, but are formulated empirically by Trevor J. McDougall,
1987, Neutral Surfaces, Journal of Physical Oceanography, volume 17, pages
1950-1964. The \code{eos="gsw"} formulae come from GSW; see references in
the \code{\link[=swRho]{swRho()}} documentation.
}
\seealso{
Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/bremen.R
\name{summary,bremen-method}
\alias{summary,bremen-method}
\title{Summarize a Bremen Object}
\usage{
\S4method{summary}{bremen}(object, ...)
}
\arguments{
\item{object}{a \linkS4class{bremen} object.}

\item{...}{Further arguments passed to or from other methods.}
}
\description{
Pertinent summary information is presented, including the station name,
sampling location, data ranges, etc.
}
\seealso{
Other things related to bremen data: 
\code{\link{[[,bremen-method}},
\code{\link{[[<-,bremen-method}},
\code{\link{bremen-class}},
\code{\link{plot,bremen-method}},
\code{\link{read.bremen}()}
}
\author{
Dan Kelley
}
\concept{things related to bremen data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swSR}
\alias{swSR}
\title{Seawater Reference Salinity, in GSW formulation}
\usage{
swSR(salinity)
}
\arguments{
\item{salinity}{either practical salinity or an \code{oce} object
that holds salinity in its data slot.}
}
\value{
Reference Salinity, SR, in \eqn{g/kg}{g/kg}.
}
\description{
Compute seawater Reference Salinity (SR), according to the GSW/TEOS-10
formulation with \code{\link[gsw:gsw_SR_from_SP]{gsw::gsw_SR_from_SP()}} in the \CRANpkg{gsw} package.
}
\examples{
\dontrun{
SR <- swAbsoluteSalinity(35.0)
stopifnot(abs(35.16504000000  - SR) < 0.000000000000010)
}

}
\references{
McDougall, T.J. and P.M. Barker, 2011: Getting started with
TEOS-10 and the Gibbs Seawater (GSW) Oceanographic Toolbox, 28pp.,
SCOR/IAPSO WG127, ISBN 978-0-646-55621-5.
}
\seealso{
For some objects, SR may also be recovered by indexing as e.g.
\code{ctd[["SR"]]}.

Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swSigma}
\alias{swSigma}
\title{Seawater density anomaly}
\usage{
swSigma(
  salinity,
  temperature = NULL,
  pressure = NULL,
  longitude = NULL,
  latitude = NULL,
  eos = getOption("oceEOS", default = "gsw")
)
}
\arguments{
\item{salinity}{either practical salinity (in which case \code{temperature}
and \code{pressure} must be provided) \emph{or} an \code{oce} object, in
which case \code{salinity}, \code{temperature} (in the ITS-90 scale; see
next item), etc. are inferred from the object, ignoring the
other parameters, if they are supplied.}

\item{temperature}{\emph{in-situ} temperature (\eqn{^\circ}{deg}C), defined
on the ITS-90 scale.  This scale is used by GSW-style calculation (as
requested by setting \code{eos="gsw"}), and is the value contained within
\code{ctd} objects (and probably most other objects created with data
acquired in the past decade or two). Since the UNESCO-style calculation is
based on IPTS-68, the temperature is converted within the present function,
using \code{\link[=T68fromT90]{T68fromT90()}}.}

\item{pressure}{pressure (dbar)}

\item{longitude}{longitude of observation (only used if \code{eos="gsw"};
see \sQuote{Details}).}

\item{latitude}{latitude of observation (only used if \code{eos="gsw"}; see
\sQuote{Details}).}

\item{eos}{equation of state, either \code{"unesco"} (references 1 and 2)
or \code{"gsw"} (references 3 and 4).}
}
\value{
Density anomaly (kg/m\eqn{^3}{^3}), as computed with \code{\link[=swRho]{swRho()}}, minus-
1000 kg/m\eqn{^3}{^3}.
}
\description{
Compute \eqn{\sigma_\theta}{sigma}, the density of seawater, minus 1000
kg/m\eqn{^3}{^3}.
}
\examples{
library(oce)
swSigma(35, 13, 1000, longitude=300, latitude=30, eos="gsw") # 30.82374
swSigma(35, T90fromT68(13), 1000, eos="unesco")    # 30.8183

}
\references{
See citations provided in the \code{\link[=swRho]{swRho()}} documentation.
}
\seealso{
Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/AllClass.R
\name{initializeFlagScheme}
\alias{initializeFlagScheme}
\title{Establish a data-quality scheme for a oce object}
\usage{
initializeFlagScheme(
  object,
  name = NULL,
  mapping = NULL,
  default = NULL,
  update = NULL,
  debug = 0
)
}
\arguments{
\item{object}{An oce object.}

\item{name}{a character value naming the scheme. If this refers
to a pre-defined scheme, then \code{mapping} must not be provided,
because doing so would contradict the pre-defined scheme, defeating
its purpose of providing concreteness and clarity.}

\item{mapping}{a list of named items describing the mapping from
flag meaning to flag numerical value, e.g \code{list(good=1, bad=2)}
might be used for a hypothetical class.}

\item{default}{an integer vector of flag values that are not considered
to be good. If this is not provided, but if \code{name} is \code{"argo"},
\code{"BODC"},
\code{"DFO"},
\code{"WHP bottle"}, or
\code{"WHP CTD"}, then a conservative value will be set automatically,
equal to the list of flag values that designate bad or questionable data.
For example, for \code{name="WHP CTD"}, the setting will be
\code{c(1,3,4,5,6,7,9)}, leaving only value \code{2}, which corresponds
with "acceptable" in the notation used for that flag scheme.}

\item{update}{a logical value indicating whether the scheme provided is
to update an existing scheme.  The default value, \code{FALSE}, prevents such
an attempt to alter an existing flag scheme, if one is already embedded
in \code{object}.}

\item{debug}{an integer set to 0 for quiet action or to 1 for some debugging.}
}
\value{
An object with the \code{metadata} slot containing \code{flagScheme}.
}
\description{
This function stores add an item named \code{flagScheme}
to the \code{metadata} slot of an object inheriting from
\linkS4class{oce}. This is a list containing two
items: \code{name} and \code{mapping}, as provided in the
function arguments.
The purpose is both to document a flag scheme
and to make it so that \code{\link[=initializeFlags]{initializeFlags()}},
\code{\link[=setFlags]{setFlags()}} and \code{\link[=handleFlags]{handleFlags()}}
can specify flags by
name, as opposed to number. This is a generic function,
that may be specialized to the class of \code{object}
(see \dQuote{Details}).
}
\details{
The following pre-defined schemes are available (note that the
names are simplified from the phrases used in defining
documentation):
\itemize{
\item \code{name="argo"} defaults \code{mapping} to
OLD (prior to June 10, 2020)
}\preformatted{list(not_assessed=0, passed_all_tests=1, probably_good=2,
     probably_bad=3, bad=4, averaged=7,
     interpolated=8, missing=9)
}

NEW (after June 10, 2020)\preformatted{list(not_assessed=0, passed_all_tests=1, probably_good=2,
     probably_bad=3, bad=4, changed=5, not_used_6=6, not_used_7=7,
     estimated=8, missing=9)
}

See reference 1 for a deeper explanation of the meanings of these codes.
\itemize{
\item \code{name="BODC"} defaults \code{mapping} to
}\preformatted{list(no_quality_control=0, good=1, probably_good=2,
     probably_bad=3, bad=4, changed=5,
     below_detection=6, in_excess=7, interpolated=8,
     missing=9)
}

See reference 2 for a deeper explanation of the meanings of these codes,
and note that codes \code{A} and \code{Q} are not provided in
oce.
\itemize{
\item \code{name="DFO"} defaults \code{mapping} to
}\preformatted{list(no_quality_control=0, appears_correct=1, appears_inconsistent=2,
     doubtful=3, erroneous=4, changed=5,
     qc_by_originator=8, missing=9)
}

See reference 3 for a deeper explanation of the meanings of these codes.
\itemize{
\item \code{name="WHP bottle"} defaults \code{mapping} to
}\preformatted{list(no_information=1, no_problems_noted=2, leaking=3,
     did_not_trip=4, not_reported=5, discrepency=6,
     unknown_problem=7, did_not_trip=8, no_sample=9)
}

See reference 4 for a deeper explanation of the meanings of these codes.
\itemize{
\item \code{name="WHP CTD"} defaults \code{mapping} to
}\preformatted{list(not_calibrated=1, acceptable=2, questionable=3,
    bad=4, not_reported=5, interpolated=6,
    despiked=7, missing=9)
}

See reference 4 for a deeper explanation of the meanings of these codes.
}
\references{
\enumerate{
\item The codes for \code{"argo"} are derived from information in Table 4.1
of Wong, Annie, Robert Keeley, Thierry Carval, and Argo Data Management Team
(8 January 2020), "Argo Quality Control Manual for CTD and Trajectory Data, Version 3.3,"
available at \url{https://archimer.ifremer.fr/doc/00228/33951/}
as of June 2020.
\item The codes for \code{"BODC"} are defined at
http://seadatanet.maris2.nl/v_bodc_vocab_v2/browse.asp?order=conceptid&formname=search&screen=0&lib=l20
\item The codes for \code{"DFO"} are defined at
http://www.dfo-mpo.gc.ca/science/data-donnees/code/list/014-eng.html
\item The codes for \code{"WHP CTD"} and \code{"WHP bottle"} are defined at
https://www.nodc.noaa.gov/woce/woce_v3/wocedata_1/whp/exchange/exchange_format_desc.htm
}
}
\seealso{
Other functions relating to data-quality flags: 
\code{\link{defaultFlags}()},
\code{\link{handleFlags,adp-method}},
\code{\link{handleFlags,argo-method}},
\code{\link{handleFlags,ctd-method}},
\code{\link{handleFlags,oce-method}},
\code{\link{handleFlags,section-method}},
\code{\link{handleFlags}()},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{initializeFlagScheme,oce-method}},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{initializeFlagSchemeInternal}()},
\code{\link{initializeFlags,adp-method}},
\code{\link{initializeFlags,oce-method}},
\code{\link{initializeFlagsInternal}()},
\code{\link{initializeFlags}()},
\code{\link{setFlags,adp-method}},
\code{\link{setFlags,ctd-method}},
\code{\link{setFlags,oce-method}},
\code{\link{setFlags}()}

Other things related to oce data: 
\code{\link{initializeFlagScheme,oce-method}},
\code{\link{initializeFlagSchemeInternal}()}
}
\concept{functions relating to data-quality flags}
\concept{things related to oce data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/coastline.R
\name{as.coastline}
\alias{as.coastline}
\title{Coerce Data into a Coastline Object}
\usage{
as.coastline(longitude, latitude, fillable = FALSE)
}
\arguments{
\item{longitude}{the longitude in decimal degrees, positive east of
Greenwich, or a data frame with columns named \code{latitude} and
\code{longitude}, in which case these values are extracted from the data
frame and the second argument is ignored.}

\item{latitude}{the latitude in decimal degrees, positive north of the
Equator.}

\item{fillable}{boolean indicating whether the coastline can be drawn as a
filled polygon.}
}
\value{
a \linkS4class{coastline} object.
}
\description{
Coerces a sequence of longitudes and latitudes into a coastline dataset.
This may be used when \code{\link[=read.coastline]{read.coastline()}} cannot read a file, or
when the data have been manipulated.
}
\seealso{
Other things related to coastline data: 
\code{\link{[[,coastline-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{coastline-class}},
\code{\link{coastlineBest}()},
\code{\link{coastlineCut}()},
\code{\link{coastlineWorld}},
\code{\link{download.coastline}()},
\code{\link{plot,coastline-method}},
\code{\link{read.coastline.openstreetmap}()},
\code{\link{read.coastline.shapefile}()},
\code{\link{subset,coastline-method}},
\code{\link{summary,coastline-method}}
}
\author{
Dan Kelley
}
\concept{things related to coastline data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/geod.R
\name{geodXy}
\alias{geodXy}
\title{Convert From Geographical to Geodesic Coordinates}
\usage{
geodXy(
  longitude,
  latitude,
  longitudeRef,
  latitudeRef,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{longitude, latitude}{vector of longitude and latitude}

\item{longitudeRef, latitudeRef}{numeric reference location. Poor results
will be returned if these values are not close to the locations described
by \code{longitude} and \code{latitude}. A sensible approach might be
to set \code{longitudeRef} to \code{longitude[1]}, etc.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\value{
\code{geodXy} returns a data frame of \code{x} and \code{y},
geodesic distance components, measured in metres.
}
\description{
The method, which may be useful in determining coordinate systems for a
mooring array or a ship transects, calculates (x,y) from distance calculations
along geodesic curves.  See \dQuote{Caution}.
}
\details{
The calculation is as follows.
Consider the \code{i}-th point in the \code{longitude} and \code{latitude}
vectors.  To calculate \code{x[i]}, \code{\link[=geodDist]{geodDist()}} is
used is to find the distance
\emph{along a geodesic curve} connecting (\code{longitude[i]}, \code{latitude[i]}) with
(\code{longitudeRef}, \code{latitude[i]}). The resultant distance
is multiplied by -1 if \code{longitude[i]-longitudeRef} is negative,
and the result is assigned to \code{x[i]}.
A similar procedure is used for \code{y[i]}.
}
\section{Caution}{
 This scheme is without known precedent in the literature, and
users should read the documentation carefully before deciding to use it.
}

\section{Change history}{

On 2015-11-02, the names of the arguments were changed from \code{lon}, etc., to
\code{longitude}, etc., to be in keeping with other oce functions.

On 2017-04-05, four changes were made.
\enumerate{
\item Default values of \code{longitudeRef} and \code{latitudeRef} were removed,
since the old defaults were inappropriate to most work.
\item The argument called \code{rotate} was eliminated, because it only made
sense if the mean resultant x and y were zero.
\item The example was made more useful.
\item Pointers were made to \code{\link[=lonlat2utm]{lonlat2utm()}}, which may be more useful.
}
}

\examples{
\donttest{
# Develop a transect-based axis system for final data(section) stations
library(oce)
data(section)
lon <- tail(section[["longitude", "byStation"]], 26)
lat <- tail(section[["latitude", "byStation"]], 26)
lonR <- tail(lon, 1)
latR <- tail(lat, 1)
data(coastlineWorld)
mapPlot(coastlineWorld, projection="+proj=merc",
        longitudelim=c(-75,-65), latitudelim=c(35,43), col="gray")
mapPoints(lon, lat)
XY <- geodXy(lon,lat,mean(lon), mean(lat))
angle <- 180/pi*atan(coef(lm(y~x, data=XY))[2])
mapCoordinateSystem(lonR, latR, 500, angle, col=2)
# Compare UTM calculation
UTM <- lonlat2utm(lon, lat, zone=18) # we need to set the zone for this task!
angleUTM <- 180/pi*atan(coef(lm(northing~easting, data=UTM))[2])
mapCoordinateSystem(lonR, latR, 500, angleUTM, col=3)
legend("topright", lwd=1, col=2:3, bg="white", title="Axis Rotation Angle",
       legend=c(sprintf("geod: \%.1f deg", angle),
                sprintf("utm:  \%.1f deg", angleUTM)))
}
}
\seealso{
\code{\link[=geodDist]{geodDist()}}

Other functions relating to geodesy: 
\code{\link{geodDist}()},
\code{\link{geodGc}()},
\code{\link{geodXyInverse}()}
}
\author{
Dan Kelley
}
\concept{functions relating to geodesy}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{oce.plot.ts}
\alias{oce.plot.ts}
\title{Oce Variant of plot.ts}
\usage{
oce.plot.ts(
  x,
  y,
  type = "l",
  xlim,
  ylim,
  log = "",
  logStyle = "r",
  flipy = FALSE,
  xlab,
  ylab,
  drawTimeRange,
  simplify = 2560,
  fill = FALSE,
  col = par("col"),
  pch = par("pch"),
  cex = par("cex"),
  cex.axis = par("cex.axis"),
  cex.lab = par("cex.lab"),
  cex.main = par("cex.main"),
  xaxs = par("xaxs"),
  yaxs = par("yaxs"),
  mgp = getOption("oceMgp"),
  mar = c(mgp[1] + if (nchar(xlab) > 0) 1.5 else 1, mgp[1] + 1.5, mgp[2] + 1, mgp[2] +
    3/4),
  main = "",
  despike = FALSE,
  axes = TRUE,
  tformat,
  marginsAsImage = FALSE,
  grid = FALSE,
  grid.col = "lightgray",
  grid.lty = "dotted",
  grid.lwd = par("lwd"),
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{x}{the times of observations.  If this is not a \link{POSIXt} object, then an
attempt is made to convert it to one using \code{\link[=as.POSIXct]{as.POSIXct()}}.}

\item{y}{the observations.}

\item{type}{plot type, \code{"l"} for lines, \code{"p"} for points.}

\item{xlim}{optional limit for x axis.  This has an additional effect,
beyond that for conventional R functions: it effectively windows the data,
so that autoscaling will yield limits for y that make sense within the
window.}

\item{ylim}{optional limit for y axis.}

\item{log}{a character value that must be either empty (the default) for linear
\code{y} axis, or \code{"y"} for logarithmic \code{y} axis.  (Unlike
\code{\link[=plot.default]{plot.default()}} etc., \code{oce.plot.ts} does not permit
logarithmic time, or \code{x} axis.)}

\item{logStyle}{a character value that indicates how to draw the y axis, if
\code{log="y"}.  If it is \code{"r"} (the default) then the conventional R style is used,
in which a logarithmic transform connects y values to position on the "page"
of the plot device, so that tics will be nonlinearly spaced, but not
organized by integral powers of 10.  However, if it is \code{"decade"}, then
the style will be that used in the scientific literature, in which large
tick marks are used for integral powers of 10, with smaller tick marks
at integral multiples of those powers, and with labels that use exponential
format for values above 100 or below 0.01.  The value of \code{logStyle} is passed
to \code{\link[=oceAxis]{oceAxis()}}, which draws the axis.}

\item{flipy}{Logical, with \code{TRUE} indicating that the graph
should have the y axis reversed, i.e. with smaller values at
the bottom of the page.}

\item{xlab}{name for x axis; defaults to \code{""}.}

\item{ylab}{name for y axis; defaults to the plotted item.}

\item{drawTimeRange}{an optional indication of whether/how to draw a time range,
in the top-left margin of the plot; see \code{\link[=oce.axis.POSIXct]{oce.axis.POSIXct()}} for details.}

\item{simplify}{an integer value that indicates
whether to speed up \code{type="l"} plots by replacing the data
with minimum and maximum values within a subsampled time mesh.
This can speed up plots of large datasets (e.g. by factor 20 for 10^7 points),
sometimes with minor changes in appearance.
This procedure is skipped if \code{simplify} is \code{NA} or
if the number of visible data points is less than 5 times \code{simplify}.
Otherwise, \code{oce.plot.ts} creates \code{simplify} intervals ranging across
the visible time range. Intervals with under 2 finite
\code{y} data are ignored. In the rest, \code{y} values
are replaced with their range, and \code{x} values are replaced
with the repeated midpoint time. Thus, each retained sub-interval
has exactly 2 data points.
A warning is printed if this replacement is done.
The default value of \code{simplify} means that cases with
under 2560 visible points are plotted conventionally.}

\item{fill}{boolean, set \code{TRUE} to fill the curve to zero (which it
does incorrectly if there are missing values in \code{y}).}

\item{col}{The colours for points (if \code{type=="p"}) or lines (if \code{type=="l"}).
For the \code{type="p"} case,
if there are fewer \code{col} values than there are \code{x} values, then the \code{col} values
are recycled in the standard fashion.
For the \code{type="l"} case, the line is plotted in the first colour specified.}

\item{pch}{character code, used if \code{type=="p"}.
If there are fewer \code{pch} values than there are \code{x} values, then the \code{pch} values
are recycled in the standard fashion.
See \code{\link[=points]{points()}} for the possible values for \code{pch}.}

\item{cex}{numeric character expansion factor for points on plots, ignored unless
\code{type} is \code{"p"}.  This may be a single number, applied to all points, or
a vector of numbers to be applied to the points in sequence.  If there are
fewer \code{pch} values than there are \code{x} values, then the \code{pch} values are recycled
in the standard fashion. See \code{\link[=par]{par()}} for more on \code{cex}.}

\item{cex.axis, cex.lab, cex.main}{numeric character expansion factors for axis numbers,
axis names and plot titles; see \code{\link[=par]{par()}}.}

\item{xaxs}{control x axis ending; see \code{\link{par}}\code{("xaxs")}.}

\item{yaxs}{control y axis ending; see \code{\link{par}}\code{("yaxs")}.}

\item{mgp}{3-element numerical vector to use for \code{par(mgp)}, and also
for \code{par(mar)}, computed from this.  The default is tighter than the R
default, in order to use more space for the data and less for the axes.}

\item{mar}{value to be used with \code{\link{par}}\code{("mar")} to set margins.
The default value uses significantly tighter margins than is the norm in R,
which gives more space for the data.  However, in doing this, the existing
\code{par("mar")} value is ignored, which contradicts values that may have
been set by a previous call to \code{\link[=drawPalette]{drawPalette()}}.  To get plot with
a palette, first call \code{\link[=drawPalette]{drawPalette()}}, then call
\code{oce.plot.ts} with \code{mar=par("mar")}.}

\item{main}{title of plot.}

\item{despike}{boolean flag that can turn on despiking with
\code{\link[=despike]{despike()}}.}

\item{axes}{boolean, set to \code{TRUE} to get axes plotted}

\item{tformat}{optional format for labels on the time axis}

\item{marginsAsImage}{boolean indicating whether to set the right-hand
margin to the width normally taken by an image drawn with
\code{\link[=imagep]{imagep()}}.}

\item{grid}{if \code{TRUE}, a grid will be drawn for each panel.  (This
argument is needed, because calling \code{\link[=grid]{grid()}} after doing a
sequence of plots will not result in useful results for the individual
panels.}

\item{grid.col}{color of grid}

\item{grid.lty}{line type of grid}

\item{grid.lwd}{line width of grid}

\item{debug}{a flag that turns on debugging.  Set to 1 to get a moderate
amount of debugging information, or to 2 to get more.}

\item{\dots}{graphical parameters passed down to \code{\link[=plot]{plot()}}.}
}
\value{
A list is silently returned, containing \code{xat} and \code{yat},
values that can be used by \code{\link[=oce.grid]{oce.grid()}} to add a grid to the plot.
}
\description{
Plot a time-series, obeying the timezone and possibly drawing the range in
the top-left margin.
}
\details{
Depending on the version of R, the standard \code{\link[=plot]{plot()}} and
\code{\link[=plot.ts]{plot.ts()}} routines will not obey the time zone of the data.
This routine gets around that problem.  It can also plot the time range in
the top-left margin, if desired; this string includes the timezone, to
remove any possible confusion.
The time axis is drawn with \code{\link[=oce.axis.POSIXct]{oce.axis.POSIXct()}}.
}
\examples{
library(oce)
t0 <- as.POSIXct("2008-01-01", tz="UTC")
t <- seq(t0, length.out=48, by="30 min")
y <- sin(as.numeric(t - t0) * 2 * pi / (12 * 3600))
oce.plot.ts(t, y, type='l', xaxs='i')
# Show how col, pch and cex get recycled
oce.plot.ts(t, y, type='p', xaxs='i',
            col=1:3, pch=c(rep(1, 6), rep(20, 6)), cex=sqrt(1:6))
# Trimming x; note the narrowing of the y view
oce.plot.ts(t, y, type='p', xlim=c(t[6], t[12]))
# Flip the y axis
oce.plot.ts(t, y, flipy=TRUE)
}
\author{
Dan Kelley and Clark Richards
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/landsat.R
\name{[[,landsat-method}
\alias{[[,landsat-method}
\title{Extract Something From a landsat Object}
\usage{
\S4method{[[}{landsat}(x, i, j, ...)
}
\arguments{
\item{x}{a \linkS4class{landsat} object.}

\item{i}{character value indicating the name of an item to extract.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{ignored.}
}
\description{
Generally, the \code{[[} method lets users extract information from \code{oce}
objects, without having to know the details of the internal storage.  For
many \code{oce} sub-classes, \code{[[} can also return quantities that are computed
from the object's contents.
}
\details{
A two-step process is used to try to find the requested information. First, a
class-specific function is used (see \dQuote{Details of the Specialized
Method}).  If this yields nothing, then a general method is used (see
\dQuote{Details of the General Method}).  If both methods fail, then \code{[[}
returns NULL.

Some understanding of the subclass is required to know what can be retrieved
with \code{[[}. When dealing with an unfamiliar subclass, it can be useful to
first use \code{x[["?"]]} to get a listing of the retrievable items. See
\dQuote{Details of the Specialized Method} for more information.
}
\section{Details of the Specialized Method}{


If \code{i} is \code{"?"}, then the return value is a list
containing four items, each of which is a character vector
holding the names of things that can be accessed with \code{[[}.
The \code{data} and \code{metadata} items hold the names of
entries in the object's data and metadata
slots, respectively. The \code{data} entries are difficult
to deal with directly, and so users are advised to
use \code{dataDerived} instead.

\emph{Accessing band data.}  The data may be accessed with e.g.
\code{landsat[["panchromatic"]]}, for the panchromatic band.  If a new
``band'' is added with \code{\link[=landsatAdd]{landsatAdd()}}, it may be referred by
name.  In all cases, a second argument can be provided, to govern
decimation.  If this is missing, all the relevant data are returned.  If
this is present and equal to \code{TRUE}, then the data will be
automatically decimated (subsampled) to give approximately 800 elements in
the longest side of the matrix.  If this is present and numerical, then its
value governs decimation.  For example,
\code{landsat[["panchromatic",TRUE]]} will auto-decimate, typically
reducing the grid width and height from 16000 to about 800.  Similarly,
\code{landsat[["panchromatic",10]]} will reduce width and height to about
1600.  On machines with limited RAM (e.g. under about 6GB), decimation is a
good idea in almost all processing steps.  It also makes sense for
plotting, and in fact is done through the `decimate` argument of
\code{\link[=plot,landsat-method]{plot,landsat-method()}}.

\emph{Accessing derived data.}  One may retrieve several derived quantities
that are calculated from data stored in the object:
\code{landsat[["longitude"]]} and \code{landsat[["latitude"]]} give pixel
locations.  Accessing \code{landsat[["temperature"]]} creates an estimate
of ground temperature as follows (see reference 4).  First, the ``count value'' in
band 10, denoted \eqn{b_{10}}{b_10} say, is scaled with coefficients stored
in the image metadata using
\eqn{\lambda_L=b_{10}M_L+A_L}{lambda_L=b_10*M_L+A_L} where \eqn{M_L}{M_L}
and \eqn{A_L}{A_L} are values stored in the metadata (e.g.  the first in
\code{landsat@metadata$header$radiance_mult_band_10}) Then the result is
used, again with coefficients in the metadata, to compute Celcius
temperature \eqn{T=K_2/ln(\epsilon
    K_1/\lambda_L+1)-273.15}{T=K_2/ln(epsilon*K_1/\lambda_L+1)-273.15}.
The value of the emissivity \eqn{\epsilon}{epsilon} is set to unity by
\code{\link[=read.landsat]{read.landsat()}}, although it can be changed easily later, by
assigning a new value to `landsat@metadata$emissivity`. The default
emissivity value set by \code{\link[=read.landsat]{read.landsat()}} is from reference 11, and is
within the oceanic range suggested by reference 5. Adjustment is as simple as
altering `landsat@metadata$emissivity`. This value can be a single
number meant to apply for the whole image, or a matrix with dimensions
matching those of band 10.  The matrix case is probably more useful for
images of land, where one might wish to account for the different
emissivities of soil and vegetation, etc.; for example, Table 4 of
reference 9 lists 0.9668 for soil and 0.9863 for vegetation,
while Table 5 of reference 10 lists 0.971 and 0.987 for the same quantities.

\emph{Accessing metadata.} Anything in the metadata can be accessed by
name, e.g. \code{landsat[["time"]]}.  Note that some items are simply
copied over from the source data file and are not altered by e.g.
decimation.  An exception is the lat-lon box, which is altered by
\code{\link[=landsatTrim]{landsatTrim()}}.
}

\section{Details of the General Method}{


Note: the text of this section is identical for all \code{oce} subclasses, and so
some of what you read here may not be relevant to the class being described
in this help page.

If the specialized method produces no matches, the following generalized
method is applied. As with the specialized method, the procedure hinges first
on the values of \code{i} and, optionally, \code{j}. The work proceeds in steps, by
testing a sequence of possible conditions in sequence.
\enumerate{
\item A check is made as to whether \code{i} names one of the standard \code{oce} slots.
If so, \code{[[} returns the slot contents of that slot.  Thus, \code{x[["metadata"]]}
will retrieve the \code{metadata} slot, while \code{x[["data"]]} and
\code{x[["processingLog"]]} return those slots.
\item If \code{i} is a string ending in the \code{"Unit"}, then the characters preceding
that string are taken to be the name of an item in the data object, and a
list containing the unit is returned (or \code{NULL} if there is no such unit).
This list consists of an item named \code{unit}, which is an \code{\link[=expression]{expression()}}, and
an item named \code{scale}, which is a string describing the measurement scale.
If the string ends in \code{" unit"}, e.g.  \code{x[["temperature unit"]]} (note the
space), then just the expression is returned, and if it ends in \code{" scale"},
then just the scale is returned.
\item If \code{i} is a string ending in \code{"Flag"}, then the corresponding data-quality
flag is returned (or \code{NULL} if there is no such flag).
\item If the object holds hydrographic information (pressure, salinity,
temperature, longitude and latitude) then another set of possibilities
arises.  If \code{i} is \code{"sigmaTheta"}, then the value of \code{\link[=swSigmaTheta]{swSigmaTheta()}} is
called with \code{x} as the sole argument, and the results are returned.
Similarly, \code{\link[=swSigma0]{swSigma0()}} is used if \code{i="sigma0"}, and \code{\link[=swSpice]{swSpice()}} is used if
\code{i="spice"}. Of course, these actions only make sense for objects that
contain the relevant items within their \code{data} slot.
\item After these possibilities are eliminated, the action depends on whether
\code{j} has been provided.  If \code{j} is not provided, or is the string \code{""}, then
\code{i} is sought in the \code{metadata} slot, and then in the \code{data} slot, returning
whichever is found first.  In other words, if \code{j} is not provided, the
\code{metadata} slot takes preference over the \code{data} slot. However, if \code{j} is
provided, then it must be either the string \code{"metadata"} or \code{"data"}, and it
directs where to look.
\item If none of the above-listed conditions holds, then \code{NULL} is returned.
}
}

\seealso{
Other functions that extract parts of oce objects: 
\code{\link{[[,adp-method}},
\code{\link{[[,adv-method}},
\code{\link{[[,amsr-method}},
\code{\link{[[,argo-method}},
\code{\link{[[,bremen-method}},
\code{\link{[[,cm-method}},
\code{\link{[[,coastline-method}},
\code{\link{[[,ctd-method}},
\code{\link{[[,echosounder-method}},
\code{\link{[[,g1sst-method}},
\code{\link{[[,gps-method}},
\code{\link{[[,ladp-method}},
\code{\link{[[,lisst-method}},
\code{\link{[[,lobo-method}},
\code{\link{[[,met-method}},
\code{\link{[[,oce-method}},
\code{\link{[[,odf-method}},
\code{\link{[[,rsk-method}},
\code{\link{[[,sealevel-method}},
\code{\link{[[,section-method}},
\code{\link{[[,tidem-method}},
\code{\link{[[,topo-method}},
\code{\link{[[,windrose-method}},
\code{\link{[[,xbt-method}},
\code{\link{[[<-,adv-method}}

Other things related to landsat data: 
\code{\link{[[<-,landsat-method}},
\code{\link{landsat-class}},
\code{\link{landsatAdd}()},
\code{\link{landsatTrim}()},
\code{\link{landsat}},
\code{\link{plot,landsat-method}},
\code{\link{read.landsat}()},
\code{\link{summary,landsat-method}}
}
\author{
Dan Kelley
}
\concept{functions that extract parts of oce objects}
\concept{things related to landsat data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/odf.R
\name{[[<-,odf-method}
\alias{[[<-,odf-method}
\title{Replace Parts of an ODF Object}
\usage{
\S4method{[[}{odf}(x, i, j, ...) <- value
}
\arguments{
\item{x}{an \linkS4class{odf} object.}

\item{i}{character value naming the item to replace.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{optional additional information (ignored).}

\item{value}{The value to be placed into \code{x}, somewhere.}
}
\description{
The \code{[[<-} method works for all \linkS4class{oce} objects.
The purpose, as with the related extraction method, \code{[[},
is to insulate users from the internal details of \linkS4class{oce}
objects, by looking for items within the various storage
slots of the object. Items not actually stored can also be
replaced, including units and data-quality
flags.
}
\details{
As with \code{[[} method, the procedure works in steps.

First, the \code{metadata} slot of \code{x} is checked to
see whether it contains something named with \code{i}.
If so, then the named item is replaced with \code{value}.

Otherwise, if the string value of \code{i} ends in \code{Unit}, then the
characters preceding that are taken as the name of a variable, and
the \code{metadata} slot of \code{x} is updated to store that unit, e.g.\preformatted{x[["temperatureUnits"]] <- list(unit=expression(degree*F),scale="")
}

Similarly, if \code{i} ends in \code{Flag}, then quality-control
flags are set up as defined by \code{result}, e.g.\preformatted{o[["temperatureFlags"]] <- c(2,4,2,2)
}

Otherwise, \code{\link[=pmatch]{pmatch()}} is used for a partial-string match with
the names of the items that are in the \code{data} slot of \code{x}.
The first item found (if any) is then updated to hold the value \code{result}.

If none of these conditions is met, a warning is issued.
}
\seealso{
Other functions that replace parts of oce objects: 
\code{\link{[[<-,adp-method}},
\code{\link{[[<-,amsr-method}},
\code{\link{[[<-,argo-method}},
\code{\link{[[<-,bremen-method}},
\code{\link{[[<-,cm-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{[[<-,echosounder-method}},
\code{\link{[[<-,g1sst-method}},
\code{\link{[[<-,gps-method}},
\code{\link{[[<-,ladp-method}},
\code{\link{[[<-,landsat-method}},
\code{\link{[[<-,lisst-method}},
\code{\link{[[<-,lobo-method}},
\code{\link{[[<-,met-method}},
\code{\link{[[<-,oce-method}},
\code{\link{[[<-,rsk-method}},
\code{\link{[[<-,sealevel-method}},
\code{\link{[[<-,section-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{[[<-,topo-method}},
\code{\link{[[<-,windrose-method}},
\code{\link{[[<-,xbt-method}}

Other things related to odf data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{ODF2oce}()},
\code{\link{ODFListFromHeader}()},
\code{\link{ODFNames2oceNames}()},
\code{\link{[[,odf-method}},
\code{\link{odf-class}},
\code{\link{plot,odf-method}},
\code{\link{read.ctd.odf}()},
\code{\link{read.odf}()},
\code{\link{subset,odf-method}},
\code{\link{summary,odf-method}}
}
\concept{functions that replace parts of oce objects}
\concept{things related to odf data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swDepth}
\alias{swDepth}
\title{Water depth}
\usage{
swDepth(pressure, latitude = 45, eos = getOption("oceEOS", default = "gsw"))
}
\arguments{
\item{pressure}{either pressure (dbar), in which case \code{lat} must also
be given, or a \code{ctd} object, in which case \code{lat} will be inferred
from the object.}

\item{latitude}{Latitude in \eqn{^\circ}{deg}N or radians north of the
equator.}

\item{eos}{indication of formulation to be used, either \code{"unesco"} or
\code{"gsw"}.}
}
\value{
Depth below the ocean surface, in metres.
}
\description{
Compute depth below the surface (i.e. a positive number within the
water column) based on pressure and latitude. (Use \code{\link[=swZ]{swZ()}}
to get the vertical coordinate, which is negative within the water
column.)
}
\details{
If \code{eos="unesco"} then depth is calculated from pressure using Saunders
and Fofonoff's method, with the formula refitted for 1980 UNESCO equation of
state (reference 1).  If \code{eos="gsw"}, then \code{\link[gsw:gsw_z_from_p]{gsw::gsw_z_from_p()}} from
the \CRANpkg{gsw} package (references 2 and 3) is used.
}
\examples{
d <- swDepth(10, 45)

}
\references{
\enumerate{
\item Unesco 1983. Algorithms for computation of fundamental
properties of seawater, 1983. \emph{Unesco Tech. Pap. in Mar. Sci.}, No. 44,
53 pp.
\item IOC, SCOR, and IAPSO (2010). The international thermodynamic equation of
seawater-2010: Calculation and use of thermodynamic properties.  Technical
Report 56, Intergovernmental Oceanographic Commission, Manuals and Guide.
\item McDougall, T.J. and P.M. Barker, 2011: Getting started with TEOS-10 and
the Gibbs Seawater (GSW) Oceanographic Toolbox, 28pp., SCOR/IAPSO WG127,
ISBN 978-0-646-55621-5.
}
}
\seealso{
Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{rotateAboutZ}
\alias{rotateAboutZ}
\title{Rotate velocity components within an oce object}
\usage{
rotateAboutZ(x, angle)
}
\arguments{
\item{x}{an \linkS4class{adp}, \linkS4class{adv}, or \linkS4class{cm} object.}

\item{angle}{The rotation angle about the z axis, in degrees counterclockwise.}
}
\description{
Alter the horizontal components of velocities in \code{adp},
\code{adv} or \code{cm} objects, by applying a rotation about
the vertical axis.
}
\examples{
library(oce)
par(mfcol=c(2, 3))
# adp (acoustic Doppler profiler)
data(adp)
plot(adp, which="uv")
mtext("adp", side=3, line=0, adj=1, cex=0.7)
adpRotated <- rotateAboutZ(adp, 30)
plot(adpRotated, which="uv")
mtext("adp rotated 30 deg", side=3, line=0, adj=1, cex=0.7)
# adv (acoustic Doppler velocimeter)
data(adv)
plot(adv, which="uv")
mtext("adv", side=3, line=0, adj=1, cex=0.7)
advRotated <- rotateAboutZ(adv, 125)
plot(advRotated, which="uv")
mtext("adv rotated 125 deg", side=3, line=0, adj=1, cex=0.7)
# cm (current meter)
data(cm)
plot(cm, which="uv")
mtext("cm", side=3, line=0, adj=1, cex=0.7)
cmRotated <- rotateAboutZ(cm, 30)
plot(cmRotated, which="uv")
mtext("cm rotated 30 deg", side=3, line=0, adj=1, cex=0.7)

}
\seealso{
Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}

Other things related to adv data: 
\code{\link{[[,adv-method}},
\code{\link{[[<-,adv-method}},
\code{\link{adv-class}},
\code{\link{adv}},
\code{\link{beamName}()},
\code{\link{beamToXyz}()},
\code{\link{enuToOtherAdv}()},
\code{\link{enuToOther}()},
\code{\link{plot,adv-method}},
\code{\link{read.adv.nortek}()},
\code{\link{read.adv.sontek.adr}()},
\code{\link{read.adv.sontek.serial}()},
\code{\link{read.adv.sontek.text}()},
\code{\link{read.adv}()},
\code{\link{subset,adv-method}},
\code{\link{summary,adv-method}},
\code{\link{toEnuAdv}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdv}()},
\code{\link{xyzToEnu}()}

Other things related to cm data: 
\code{\link{[[,cm-method}},
\code{\link{[[<-,cm-method}},
\code{\link{as.cm}()},
\code{\link{cm-class}},
\code{\link{cm}},
\code{\link{plot,cm-method}},
\code{\link{read.cm}()},
\code{\link{subset,cm-method}},
\code{\link{summary,cm-method}}
}
\author{
Dan Kelley
}
\concept{things related to adp data}
\concept{things related to adv data}
\concept{things related to cm data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/units.R
\name{as.unit}
\alias{as.unit}
\title{Convert a String to a Unit}
\usage{
as.unit(u, default = list(unit = expression(), scale = ""))
}
\arguments{
\item{u}{A character string indicating a variable name. The following
names are recognized: \code{"DBAR"},
\code{"IPTS-68"}, \code{"ITS-90"}, \code{"PSS-78"}, and \code{"UMOL/KG"}.
All other names yield a return value equal to the value of the
\code{default} argument.}

\item{default}{A default to be used for the return value, if \code{u}
is not a recognized string.}
}
\value{
A list with elements \code{unit}, an \code{\link[=expression]{expression()}},
and \code{scale}, a string.
}
\description{
This function is not presently used by any \code{oce} functions, and is
provided as a convenience function for users.
}
\examples{
as.unit("DBAR")
as.unit("IPTS-68")
as.unit("ITS-90")
as.unit("PSS-78")
as.unit("UMOL/KG")
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{despike}
\alias{despike}
\title{Remove spikes from a time series}
\usage{
despike(
  x,
  reference = c("median", "smooth", "trim"),
  n = 4,
  k = 7,
  min = NA,
  max = NA,
  replace = c("reference", "NA"),
  skip
)
}
\arguments{
\item{x}{a vector of (time-series) values, a list of vectors, a data frame,
or an \linkS4class{oce} object.}

\item{reference}{indication of the type of reference time series to be used
in the detection of spikes; see \sQuote{Details}.}

\item{n}{an indication of the limit to differences between \code{x} and the
reference time series, used for \code{reference="median"} or
\code{reference="smooth"}; see \sQuote{Details.}}

\item{k}{length of running median used with \code{reference="median"}, and
ignored for other values of \code{reference}.}

\item{min}{minimum non-spike value of \code{x}, used with
\code{reference="trim"}.}

\item{max}{maximum non-spike value of \code{x}, used with
\code{reference="trim"}.}

\item{replace}{an indication of what to do with spike values, with
\code{"reference"} indicating to replace them with the reference time
series, and \code{"NA"} indicating to replace them with \code{NA}.}

\item{skip}{optional vector naming columns to be skipped. This is ignored if
\code{x} is a simple vector. Any items named in \code{skip} will be passed
through to the return value without modification.  In some cases,
\code{despike} will set up reasonable defaults for \code{skip}, e.g. for a
\code{ctd} object, \code{skip} will be set to \code{c("time", "scan",
"pressure")} if it is not supplied as an argument.}
}
\value{
A new vector in which spikes are replaced as described above.
}
\description{
The method identifies spikes with respect to a "reference" time-series, and
replaces these spikes with the reference value, or with \code{NA} according
to the value of \code{action}; see \dQuote{Details}.
}
\details{
Three modes of operation are permitted, depending on the value of
\code{reference}.
\enumerate{
\item For \code{reference="median"}, the first step is to linearly interpolate
across any gaps (spots where \code{x==NA}), using \code{\link[=approx]{approx()}} with
\code{rule=2}. The second step is to pass this through
\code{\link[=runmed]{runmed()}} to get a running median spanning \code{k}
elements. The result of these two steps is the "reference" time-series.
Then, the standard deviation of the difference between \code{x}
and the reference is calculated.  Any \code{x} values that differ from
the reference by more than \code{n} times this standard deviation are considered
to be spikes.  If \code{replace="reference"}, the spike values are
replaced with the reference, and the resultant time series is
returned.  If \code{replace="NA"}, the spikes are replaced with \code{NA},
and that result is returned.
\item For \code{reference="smooth"}, the processing is the same as for
\code{"median"}, except that \code{\link[=smooth]{smooth()}} is used to calculate the
reference time series.
\item For \code{reference="trim"}, the reference time series is constructed by
linear interpolation across any regions in which \code{x<min} or
\code{x>max}.  (Again, this is done with \code{\link[=approx]{approx()}} with
\code{rule=2}.) In this case, the value of \code{n} is ignored, and the
return value is the same as \code{x}, except that spikes are replaced
with the reference series (if \code{replace="reference"} or with
\code{NA}, if \code{replace="NA"}.
}
}
\examples{
n <- 50
x <- 1:n
y <- rnorm(n=n)
y[n/2] <- 10                    # 10 standard deviations
plot(x, y, type='l')
lines(x, despike(y), col='red')
lines(x, despike(y, reference="smooth"), col='darkgreen')
lines(x, despike(y, reference="trim", min=-3, max=3), col='blue')
legend("topright", lwd=1, col=c("black", "red", "darkgreen", "blue"),
       legend=c("raw", "median", "smooth", "trim"))

# add a spike to a CTD object
data(ctd)
plot(ctd)
T <- ctd[["temperature"]]
T[10] <- T[10] + 10
ctd[["temperature"]] <- T
CTD <- despike(ctd)
plot(CTD)
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/argo.R
\name{handleFlags,argo-method}
\alias{handleFlags,argo-method}
\alias{handleFlags.argo}
\title{Handle Flags in ARGO Objects}
\usage{
\S4method{handleFlags}{argo}(
  object,
  flags = NULL,
  actions = NULL,
  where = NULL,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{object}{an \linkS4class{argo} object.}

\item{flags}{A \link{list} specifying flag values upon which
actions will be taken. This can take two forms.
\itemize{
\item In the first form, the
list has named elements each containing a vector of integers. For example,
salinities flagged with values of 1 or 3:9 would be specified
by \code{flags=list(salinity=c(1,3:9))}. Several data items can be specified,
e.g. \code{flags=list(salinity=c(1,3:9), temperature=c(1,3:9))} indicates
that the actions are to take place for both salinity and temperature.
\item In the second form, \code{flags} is a list holding a single \emph{unnamed} vector, and
this means to apply the actions to \emph{all} the data entries.  For example,
\code{flags=list(c(1,3:9))} means to apply not just to salinity and temperature,
but to everything within the \code{data} slot.
}

If \code{flags} is not provided, then \code{\link[=defaultFlags]{defaultFlags()}} is called, to try to
determine a reasonable default.}

\item{actions}{an optional \link{list} that contains items with
names that match those in the \code{flags} argument.  If \code{actions}
is not supplied, the default will be to set all values identified by
\code{flags} to \code{NA}; this can also be specified by
specifying \code{actions=list("NA")}. It is also possible to specify
functions that calculate replacement values. These are provided
with \code{object} as the single argument, and must return a
replacement for the data item in question.
See \dQuote{Details} for the default that is used if \code{actions} is not supplied.}

\item{where}{an optional character value that permits the function to work with
objects that store flags in e.g. \code{object@metadata$flags$where}
instead of in \code{object@metadata$flags}, and data within
\code{object@data$where} instead of within \code{object@data}. The
default value of \code{NULL} means to look withing \code{object@metadata}
itself, and this is the default within \code{oce}.  (The purpose of \code{where}
is to make \code{oce} extensible by other packages, which may choose to store
data two levels deep in the \code{data} slot.)}

\item{debug}{An optional integer specifying the degree of debugging, with
value 0 meaning to skip debugging and 1 or higher meaning to print some
information about the arguments and the data. It is usually a good idea to set
this to 1 for initial work with a dataset, to see which flags are being
handled for each data item. If not supplied, this defaults to the value of
\code{\link{getOption}}\code{("oceDebug")}.}
}
\description{
Data-quality flags are stored in the \code{metadata}
slot of \linkS4class{oce} objects in a
\link{list} named \code{flags}.
The present function (a generic that has specialized versions
for various data classes) provides a way to
manipulate the contents of the \code{data} slot, based on
such data-quality flags. For example, a common operation is to replace
erroneous data with \code{NA}.

If the \code{flags} within \code{object}'s \code{metadata} slot is empty,
then \code{object} is returned, unaltered.
Otherwise, \code{handleFlags} examines \code{object@metadata$flags}
in the context of the \code{flags} argument, and then
carries out actions that are specified by the \code{actions} argument.
By default, this sets the returned \code{data} entries to \code{NA},
wherever the corresponding \code{metadata$flag} values
signal unreliable data. To maintain a hint as to why
\code{data} were changed, \code{metadata$flags} in the
returned value is a direct copy of the corresponding
entry in \code{object}.
}
\examples{
library(oce)
data(argo)
argoNew <- handleFlags(argo)
# Demonstrate replacement, looking at the second profile
f <- argo[["salinityFlag"]][,2]
df <- data.frame(flag=f, orig=argo[["salinity"]][,2], new=argoNew[["salinity"]][,2])
df[11:15,] # notice line 13

}
\references{
\enumerate{
\item Wong, Annie, Robert Keeley, Thierry Carval, and Argo Data Management Team.
"Argo Quality Control Manual for CTD and Trajectory Data," January 1, 2020.
\url{https://archimer.ifremer.fr/doc/00228/33951/}.
}
}
\seealso{
Other functions relating to data-quality flags: 
\code{\link{defaultFlags}()},
\code{\link{handleFlags,adp-method}},
\code{\link{handleFlags,ctd-method}},
\code{\link{handleFlags,oce-method}},
\code{\link{handleFlags,section-method}},
\code{\link{handleFlags}()},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{initializeFlagScheme,oce-method}},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{initializeFlagSchemeInternal}()},
\code{\link{initializeFlagScheme}()},
\code{\link{initializeFlags,adp-method}},
\code{\link{initializeFlags,oce-method}},
\code{\link{initializeFlagsInternal}()},
\code{\link{initializeFlags}()},
\code{\link{setFlags,adp-method}},
\code{\link{setFlags,ctd-method}},
\code{\link{setFlags,oce-method}},
\code{\link{setFlags}()}

Other things related to argo data: 
\code{\link{[[,argo-method}},
\code{\link{[[<-,argo-method}},
\code{\link{argo-class}},
\code{\link{argoGrid}()},
\code{\link{argoNames2oceNames}()},
\code{\link{argo}},
\code{\link{as.argo}()},
\code{\link{plot,argo-method}},
\code{\link{read.argo}()},
\code{\link{subset,argo-method}},
\code{\link{summary,argo-method}}
}
\author{
Dan Kelley
}
\concept{functions relating to data-quality flags}
\concept{things related to argo data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/argo.R
\docType{class}
\name{argo-class}
\alias{argo-class}
\title{Class to Store Argo Data}
\description{
This class stores data from Argo floats.
}
\details{
An \code{argo} object may be read with \code{\link[=read.argo]{read.argo()}} or
created with \code{\link[=as.argo]{as.argo()}}.  Argo data can be gridded to constant
pressures with \code{\link[=argoGrid]{argoGrid()}} or subsetted with
\code{\link[=subset,argo-method]{subset,argo-method()}}.  Plots can be made with
\code{\link[=plot,argo-method]{plot,argo-method()}}, while \code{\link[=summary,argo-method]{summary,argo-method()}}
produces statistical summaries and \code{show} produces overviews.
}
\section{Slots}{

\describe{
\item{\code{data}}{As with all \code{oce} objects, the \code{data} slot
for \code{argo} objects is a
\link{list} containing the main data for the object.
The key items stored in this slot include  equal-length vectors \code{time}, \code{longitude}, \code{latitude} and equal-dimension matrices \code{pressure}, \code{salinity}, and \code{temperature}.}

\item{\code{metadata}}{As with all \code{oce} objects, the \code{metadata} slot
for \code{argo} objects is a \link{list} containing
information about the \code{data} or about the object itself.
Examples that are of common interest include \code{id}, a vector of ID codes for the profiles, and \code{dataMode}, a vector of strings indicating whether the profile is in archived mode (\code{"A"}), realtime mode (\code{"R"}), or delayed mode (\code{"D"}).}

\item{\code{processingLog}}{As with all \code{oce} objects, the \code{processingLog} slot
for \code{argo} objects is a
\link{list} with entries describing the creation and evolution
of the object. The contents are updated by various \code{oce} functions to
keep a record of processing steps.  Object summaries and
\code{\link[=processingLogShow]{processingLogShow()}} both display the log.}
}}

\section{Modifying slot contents}{


Although the \code{[[<-} operator may permit modification of the contents
of \linkS4class{argo} objects (see \code{\link{[[<-,argo-method}}),
it is better to use \code{\link[=oceSetData]{oceSetData()}} and \code{\link[=oceSetMetadata]{oceSetMetadata()}},
because those functions save an entry in the \code{processingLog}
that describes the change.
}

\section{Retrieving slot contents}{


The full contents of the \code{data} and \code{metadata} slots of a \linkS4class{argo}
object may be retrieved in the standard R way using \code{\link[=slot]{slot()}}. For
example \code{slot(o,"data")} returns the \code{data} slot of an object named \code{o},
and similarly \code{slot(o,"metadata")} returns
the \code{metadata} slot.

The slots may also be obtained with the \code{\link{[[,argo-method}}
operator, as e.g. \code{o[["data"]]} and \code{o[["metadata"]]}, respectively.

The \code{\link{[[,argo-method}} operator can also
be used to retrieve items from within the \code{data} and \code{metadata} slots.
For example, \code{o[["temperature"]]} can be used to retrieve temperature
from an object containing that quantity. The rule is that a named
quantity is sought first within the object's \code{metadata} slot,
with the \code{data} slot being checked only if \code{metadata} does not
contain the item. This \code{[[} method can also be used to get
certain derived quantities, if the object contains sufficient
information to calculate them. For example, an object that holds
(practical) salinity, temperature and pressure, along with
longitude and latitude, has sufficient information to compute
Absolute Salinity, and so \code{o[["SA"]]} will yield the
calculated Absolute Salinity.

It is also possible to find items more directly, using \code{\link[=oceGetData]{oceGetData()}} and
\code{\link[=oceGetMetadata]{oceGetMetadata()}}, but neither of these functions can
retrieve derived items.
}

\seealso{
Other classes provided by oce: 
\code{\link{adp-class}},
\code{\link{adv-class}},
\code{\link{bremen-class}},
\code{\link{cm-class}},
\code{\link{coastline-class}},
\code{\link{ctd-class}},
\code{\link{lisst-class}},
\code{\link{lobo-class}},
\code{\link{met-class}},
\code{\link{oce-class}},
\code{\link{odf-class}},
\code{\link{rsk-class}},
\code{\link{sealevel-class}},
\code{\link{section-class}},
\code{\link{topo-class}},
\code{\link{windrose-class}},
\code{\link{xbt-class}}

Other things related to argo data: 
\code{\link{[[,argo-method}},
\code{\link{[[<-,argo-method}},
\code{\link{argoGrid}()},
\code{\link{argoNames2oceNames}()},
\code{\link{argo}},
\code{\link{as.argo}()},
\code{\link{handleFlags,argo-method}},
\code{\link{plot,argo-method}},
\code{\link{read.argo}()},
\code{\link{subset,argo-method}},
\code{\link{summary,argo-method}}
}
\author{
Dan Kelley and Clark Richards
}
\concept{classes provided by oce}
\concept{things related to argo data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swSpecificHeat}
\alias{swSpecificHeat}
\title{Seawater specific heat}
\usage{
swSpecificHeat(
  salinity,
  temperature = NULL,
  pressure = 0,
  longitude = NULL,
  latitude = NULL,
  eos = getOption("oceEOS", default = "gsw")
)
}
\arguments{
\item{salinity}{either practical salinity (in which case \code{temperature}
and \code{pressure} must be provided) \emph{or} an \code{oce} object (in
which case \code{salinity}, etc. are inferred from the object).}

\item{temperature}{\emph{in-situ} temperature (\eqn{^\circ}{deg}C), defined
on the ITS-90 scale.}

\item{pressure}{seawater pressure (dbar)}

\item{longitude}{longitude of observation (only used if \code{eos="gsw"};
see \sQuote{Details}).}

\item{latitude}{latitude of observation (only used if \code{eos="gsw"}; see
\sQuote{Details}).}

\item{eos}{equation of state, either \code{"unesco"} or \code{"gsw"}.}
}
\value{
Specific heat \eqn{J kg^{-1}\,^\circ C^{-1}}{J/(kg degC)}
}
\description{
Compute specific heat of seawater.
}
\details{
If the first argument is a \code{ctd} object, then salinity, etc, are
extracted from it, and used for the calculation.
}
\examples{
swSpecificHeat(40, T90fromT68(40), 10000, eos="unesco") # 3949.499

}
\references{
Millero et. al., J. Geophys. Res. 78 (1973), 4499-4507

Millero et. al., UNESCO report 38 (1981), 99-188.
}
\seealso{
Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{oce.write.table}
\alias{oce.write.table}
\title{Write the Data Portion of Object to a File}
\usage{
oce.write.table(x, file = "", ...)
}
\arguments{
\item{x}{an \linkS4class{oce} object.}

\item{file}{file name, as passed to \code{\link[utils:write.table]{utils::write.table()}}.  Use
\code{""} to get a listing in the terminal window.}

\item{...}{optional arguments passed to \code{\link[utils:write.table]{utils::write.table()}}.}
}
\value{
The value returned by \code{\link[utils:write.table]{utils::write.table()}}.
}
\description{
The output has a line containing the names of the columns in \code{x$data},
each enclosed in double quotes.  After that line are lines for the data
themselves.  The default is to separate data items by a single space
character, but this can be altered by using a \code{sep} argument in the
\code{...} list; see \code{\link[utils:write.table]{utils::write.table()}}.
}
\details{
This function is little more than a thin wrapper around
\code{\link[utils:write.table]{utils::write.table()}}, the only difference being that row names
are omitted here, making for a file format that is more conventional in
Oceanography.
}
\seealso{
`\code{\link[utils:write.table]{utils::write.table()}}, which does the actual work.
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{oceDebug}
\alias{oceDebug}
\alias{oce.debug}
\title{Print a debugging message}
\usage{
oceDebug(debug = 0, ..., style = "plain", unindent = 0, sep = "")
}
\arguments{
\item{debug}{an integer, less than or equal to zero for no message, and
greater than zero for increasing levels of debugging.  Values greater than 4
are treated like 4.}

\item{\dots}{items to be supplied to \code{\link[=cat]{cat()}}, which does the
printing.  Note that no newline will be printed unless \dots
contains a string with a newline character (as in the example).}

\item{style}{either a string or a function. If a string,
it must be \code{"plain"} (the default) for plain text,
\code{"bold"}, \code{"italic"}, \code{"red"}, \code{"green"} or \code{"blue"} (with
obvious meanings).
If \code{style} is a function, it must prepend and postpend the text
with control codes, as in the cyan-coloured example; note that
\CRANpkg{crayon} provides many functions that work well for \code{style}.}

\item{unindent}{integer giving the number of levels to un-indent,
e.g. for start and end lines from a called function.}

\item{sep}{character to insert between elements of \code{...}, by passing
it to \code{\link[=cat]{cat()}}.}
}
\description{
Print an indented debugging message.
Many oce functions decrease the \code{debug} level by 1 when they call other
functions, so the effect is a nesting, with more space for deeper function
level.
}
\examples{
oceDebug(debug=1, "Example", 1, "Plain text")
oceDebug(debug=1, "Example", 2, "Bold", style="bold")
oceDebug(debug=1, "Example", 3, "Italic", style="italic")
oceDebug(debug=1, "Example", 4, "Red", style="red")
oceDebug(debug=1, "Example", 5, "Green", style="green")
oceDebug(debug=1, "Example", 6, "Blue", style="blue")
mycyan <- function(...) paste("\033[36m", paste(..., sep=" "), "\033[0m", sep="")
oceDebug(debug=1, "Example", 7, "User-set cyan", style=mycyan)
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/bremen.R
\name{[[<-,bremen-method}
\alias{[[<-,bremen-method}
\title{Replace Parts of a Bremen Object}
\usage{
\S4method{[[}{bremen}(x, i, j, ...) <- value
}
\arguments{
\item{x}{a \linkS4class{bremen} object.}

\item{i}{character value naming the item to replace.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{optional additional information (ignored).}

\item{value}{The value to be placed into \code{x}, somewhere.}
}
\description{
The \code{[[<-} method works for all \linkS4class{oce} objects.
The purpose, as with the related extraction method, \code{[[},
is to insulate users from the internal details of \linkS4class{oce}
objects, by looking for items within the various storage
slots of the object. Items not actually stored can also be
replaced, including units and data-quality
flags.
}
\details{
As with \code{[[} method, the procedure works in steps.

First, the \code{metadata} slot of \code{x} is checked to
see whether it contains something named with \code{i}.
If so, then the named item is replaced with \code{value}.

Otherwise, if the string value of \code{i} ends in \code{Unit}, then the
characters preceding that are taken as the name of a variable, and
the \code{metadata} slot of \code{x} is updated to store that unit, e.g.\preformatted{x[["temperatureUnits"]] <- list(unit=expression(degree*F),scale="")
}

Similarly, if \code{i} ends in \code{Flag}, then quality-control
flags are set up as defined by \code{result}, e.g.\preformatted{o[["temperatureFlags"]] <- c(2,4,2,2)
}

Otherwise, \code{\link[=pmatch]{pmatch()}} is used for a partial-string match with
the names of the items that are in the \code{data} slot of \code{x}.
The first item found (if any) is then updated to hold the value \code{result}.

If none of these conditions is met, a warning is issued.
}
\seealso{
Other functions that replace parts of oce objects: 
\code{\link{[[<-,adp-method}},
\code{\link{[[<-,amsr-method}},
\code{\link{[[<-,argo-method}},
\code{\link{[[<-,cm-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{[[<-,echosounder-method}},
\code{\link{[[<-,g1sst-method}},
\code{\link{[[<-,gps-method}},
\code{\link{[[<-,ladp-method}},
\code{\link{[[<-,landsat-method}},
\code{\link{[[<-,lisst-method}},
\code{\link{[[<-,lobo-method}},
\code{\link{[[<-,met-method}},
\code{\link{[[<-,oce-method}},
\code{\link{[[<-,odf-method}},
\code{\link{[[<-,rsk-method}},
\code{\link{[[<-,sealevel-method}},
\code{\link{[[<-,section-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{[[<-,topo-method}},
\code{\link{[[<-,windrose-method}},
\code{\link{[[<-,xbt-method}}

Other things related to bremen data: 
\code{\link{[[,bremen-method}},
\code{\link{bremen-class}},
\code{\link{plot,bremen-method}},
\code{\link{read.bremen}()},
\code{\link{summary,bremen-method}}
}
\concept{functions that replace parts of oce objects}
\concept{things related to bremen data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/AllClass.R
\name{setFlags}
\alias{setFlags}
\title{Set data-quality flags within a oce object}
\usage{
setFlags(object, name = NULL, i = NULL, value = NULL, debug = 0)
}
\arguments{
\item{object}{An oce object.}

\item{name}{Character string indicating the name of the variable to be flagged. If
this variable is not contained in the object's \code{data} slot, an error is reported.}

\item{i}{Indication of where to insert the flags; see \dQuote{Description} for
general rules and \dQuote{Details} for rules for \linkS4class{oce}
objects.}

\item{value}{The value to be inserted in the flag.}

\item{debug}{Integer set to 0 for quiet action or to 1 for some debugging.}
}
\value{
An object with flags set as indicated.
}
\description{
This function changes specified entries in the data-quality
flags of a oce object, which are stored within
a list named \code{flags} that resides in the \code{metadata}
slot. If the object already has a flag set up for \code{name},
then only the specified entries are altered. If not, the flag
entry is first created and its entries set to \code{default},
after which the entries specified by \code{i}
are changed to \code{value}.

The specification is made with \code{i}, the form of which
is determined by the data item in question. Generally,
the rules are as follows:
\enumerate{
\item If the data item is a vector, then \code{i} must be (a)
an integer vector specifying indices to be set to \code{value},
(b) a logical vector of length matching the data item, with
\code{TRUE} meaning to set the flag to \code{value}, or (c)
a function that takes an \code{oce} object as its single
argument, and returns a vector in either of the forms
just described.
\item If the data item is an array, then \code{i} must be
(a) a data frame of integers whose rows specify spots to change
(where the number of columns matches the number of dimensions
of the data item), (b) a logical array that has dimension equal to
that of the data item, or (c) a function that takes an \code{oce}
object as its single input and returns such a data frame or array.
}

See \dQuote{Details} for the particular case of \linkS4class{oce} objects.
}
\details{
This generic function is overridden by specialized functions for some object classes.
}
\seealso{
Other functions relating to data-quality flags: 
\code{\link{defaultFlags}()},
\code{\link{handleFlags,adp-method}},
\code{\link{handleFlags,argo-method}},
\code{\link{handleFlags,ctd-method}},
\code{\link{handleFlags,oce-method}},
\code{\link{handleFlags,section-method}},
\code{\link{handleFlags}()},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{initializeFlagScheme,oce-method}},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{initializeFlagSchemeInternal}()},
\code{\link{initializeFlagScheme}()},
\code{\link{initializeFlags,adp-method}},
\code{\link{initializeFlags,oce-method}},
\code{\link{initializeFlagsInternal}()},
\code{\link{initializeFlags}()},
\code{\link{setFlags,adp-method}},
\code{\link{setFlags,ctd-method}},
\code{\link{setFlags,oce-method}}
}
\concept{functions relating to data-quality flags}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/xbt.R
\name{subset,xbt-method}
\alias{subset,xbt-method}
\title{Subset an xbt Object}
\usage{
\S4method{subset}{xbt}(x, subset, ...)
}
\arguments{
\item{x}{an \linkS4class{xbt} object.}

\item{subset}{a condition to be applied to the \code{data} portion of \code{x}.
See \sQuote{Details}.}

\item{...}{ignored.}
}
\value{
A new \code{xbt} object.
}
\description{
This function is somewhat analogous to \code{\link[=subset.data.frame]{subset.data.frame()}}.
}
\examples{
library(oce)
data(xbt)
plot(xbt)
plot(subset(xbt, depth < mean(range(xbt[["depth"]]))))

}
\seealso{
Other things related to xbt data: 
\code{\link{[[,xbt-method}},
\code{\link{[[<-,xbt-method}},
\code{\link{as.xbt}()},
\code{\link{plot,xbt-method}},
\code{\link{read.xbt.noaa1}()},
\code{\link{read.xbt}()},
\code{\link{summary,xbt-method}},
\code{\link{xbt-class}},
\code{\link{xbt.edf}},
\code{\link{xbt}}

Other functions that subset oce objects: 
\code{\link{subset,adp-method}},
\code{\link{subset,adv-method}},
\code{\link{subset,amsr-method}},
\code{\link{subset,argo-method}},
\code{\link{subset,cm-method}},
\code{\link{subset,coastline-method}},
\code{\link{subset,ctd-method}},
\code{\link{subset,echosounder-method}},
\code{\link{subset,lobo-method}},
\code{\link{subset,met-method}},
\code{\link{subset,oce-method}},
\code{\link{subset,odf-method}},
\code{\link{subset,rsk-method}},
\code{\link{subset,sealevel-method}},
\code{\link{subset,section-method}},
\code{\link{subset,topo-method}}
}
\author{
Dan Kelley
}
\concept{functions that subset oce objects}
\concept{things related to xbt data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swSstar}
\alias{swSstar}
\title{Seawater Preformed Salinity, in GSW formulation}
\usage{
swSstar(salinity, pressure = NULL, longitude = NULL, latitude = NULL)
}
\arguments{
\item{salinity}{either practical salinity (in which case
\code{pressure} must be provided) \emph{or} an \code{oce} object with
\code{salinity} and \code{pressure} in its data slot, and with
\code{longitude} and \code{latitude} either there, or in the metadata
slot.}

\item{pressure}{pressure in dbar.}

\item{longitude}{longitude of observation.}

\item{latitude}{latitude of observation.}
}
\value{
Preformed Salinity, S*, in \eqn{g/kg}{g/kg}.
}
\description{
Compute seawater Preformed Salinity (S*), according to the GSW/TEOS-10
formulation with \code{\link[gsw:gsw_Sstar_from_SA]{gsw::gsw_Sstar_from_SA()}} in the \CRANpkg{gsw} package.
}
\examples{
\dontrun{
sa <- swAbsoluteSalinity(35.5, 300, 260, 16)
stopifnot(abs(35.671358392019094 - sa) < 00.000000000000010)
}

}
\references{
McDougall, T.J. and P.M. Barker, 2011: Getting started with
TEOS-10 and the Gibbs Seawater (GSW) Oceanographic Toolbox, 28pp.,
SCOR/IAPSO WG127, ISBN 978-0-646-55621-5.
}
\seealso{
For some objects, S-star may also be recovered by
indexing as e.g. \code{ctd[["Sstar"]]}.

Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/coastline.R
\name{[[<-,coastline-method}
\alias{[[<-,coastline-method}
\title{Replace Parts of a Coastline Object}
\usage{
\S4method{[[}{coastline}(x, i, j, ...) <- value
}
\arguments{
\item{x}{a \linkS4class{coastline} object.}

\item{i}{character value naming the item to replace.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{optional additional information (ignored).}

\item{value}{The value to be placed into \code{x}, somewhere.}
}
\description{
The \code{[[<-} method works for all \linkS4class{oce} objects.
The purpose, as with the related extraction method, \code{[[},
is to insulate users from the internal details of \linkS4class{oce}
objects, by looking for items within the various storage
slots of the object. Items not actually stored can also be
replaced, including units and data-quality
flags.
}
\details{
As with \code{[[} method, the procedure works in steps.

First, the \code{metadata} slot of \code{x} is checked to
see whether it contains something named with \code{i}.
If so, then the named item is replaced with \code{value}.

Otherwise, if the string value of \code{i} ends in \code{Unit}, then the
characters preceding that are taken as the name of a variable, and
the \code{metadata} slot of \code{x} is updated to store that unit, e.g.\preformatted{x[["temperatureUnits"]] <- list(unit=expression(degree*F),scale="")
}

Similarly, if \code{i} ends in \code{Flag}, then quality-control
flags are set up as defined by \code{result}, e.g.\preformatted{o[["temperatureFlags"]] <- c(2,4,2,2)
}

Otherwise, \code{\link[=pmatch]{pmatch()}} is used for a partial-string match with
the names of the items that are in the \code{data} slot of \code{x}.
The first item found (if any) is then updated to hold the value \code{result}.

If none of these conditions is met, a warning is issued.
}
\seealso{
Other things related to coastline data: 
\code{\link{[[,coastline-method}},
\code{\link{as.coastline}()},
\code{\link{coastline-class}},
\code{\link{coastlineBest}()},
\code{\link{coastlineCut}()},
\code{\link{coastlineWorld}},
\code{\link{download.coastline}()},
\code{\link{plot,coastline-method}},
\code{\link{read.coastline.openstreetmap}()},
\code{\link{read.coastline.shapefile}()},
\code{\link{subset,coastline-method}},
\code{\link{summary,coastline-method}}

Other functions that replace parts of oce objects: 
\code{\link{[[<-,adp-method}},
\code{\link{[[<-,amsr-method}},
\code{\link{[[<-,argo-method}},
\code{\link{[[<-,bremen-method}},
\code{\link{[[<-,cm-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{[[<-,echosounder-method}},
\code{\link{[[<-,g1sst-method}},
\code{\link{[[<-,gps-method}},
\code{\link{[[<-,ladp-method}},
\code{\link{[[<-,landsat-method}},
\code{\link{[[<-,lisst-method}},
\code{\link{[[<-,lobo-method}},
\code{\link{[[<-,met-method}},
\code{\link{[[<-,oce-method}},
\code{\link{[[<-,odf-method}},
\code{\link{[[<-,rsk-method}},
\code{\link{[[<-,sealevel-method}},
\code{\link{[[<-,section-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{[[<-,topo-method}},
\code{\link{[[<-,windrose-method}},
\code{\link{[[<-,xbt-method}}
}
\author{
Dan Kelley
}
\concept{functions that replace parts of oce objects}
\concept{things related to coastline data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/astronomy.R
\name{equatorialToLocalHorizontal}
\alias{equatorialToLocalHorizontal}
\title{Convert equatorial to local horizontal coordinate}
\usage{
equatorialToLocalHorizontal(
  rightAscension,
  declination,
  t,
  longitude,
  latitude
)
}
\arguments{
\item{rightAscension}{right ascension, e.g. calculated with
\code{\link[=eclipticalToEquatorial]{eclipticalToEquatorial()}}.}

\item{declination}{declination, e.g. calculated with
\code{\link[=eclipticalToEquatorial]{eclipticalToEquatorial()}}.}

\item{t}{time of observation.}

\item{longitude}{longitude of observation, positive in eastern hemisphere.}

\item{latitude}{latitude of observation, positive in northern hemisphere.}
}
\value{
A data frame containing columns \code{altitude} (angle above
horizon, in degrees) and \code{azimuth} (angle anticlockwise from south, in
degrees).
}
\description{
Convert from equatorial coordinates to local horizontal coordinates, i.e.
azimuth and altitude.
The method is taken from equations 8.5 and 8.6 of reference 1, or, equivalently,
from equations 12.5 and 12.6 of reference 2.
}
\references{
\itemize{
\item Meeus, Jean. Astronomical Formulas for Calculators. Second Edition.
Richmond, Virginia, USA: Willmann-Bell, 1982.
\item Meeus, Jean. Astronomical Algorithms. Second Edition.
Richmond, Virginia, USA: Willmann-Bell, 1991.
}
}
\seealso{
Other things related to astronomy: 
\code{\link{angle2hms}()},
\code{\link{eclipticalToEquatorial}()},
\code{\link{julianCenturyAnomaly}()},
\code{\link{julianDay}()},
\code{\link{moonAngle}()},
\code{\link{siderealTime}()},
\code{\link{sunAngle}()},
\code{\link{sunDeclinationRightAscension}()}
}
\author{
Dan Kelley, based on formulae in references 1 and 2.
}
\concept{things related to astronomy}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/landsat.R
\name{landsatTrim}
\alias{landsatTrim}
\title{Trim a landsat Image to a Geographical Region}
\usage{
landsatTrim(x, ll, ur, box, debug = getOption("oceDebug"))
}
\arguments{
\item{x}{a \linkS4class{landsat} object.}

\item{ll}{A list containing \code{longitude} and \code{latitude}, for the
lower-left corner of the portion of the image to retain, or a vector
with first element longitude and second element latitude. If provided,
then \code{ur} must also be provided, but \code{box} cannot.}

\item{ur}{A list containing \code{longitude} and \code{latitude}, for the
upper-right corner of the portion of the image to retain, or a vector
with first element longitude and second element latitude. If provided,
then \code{ll} must also be provided, but \code{box} cannot.}

\item{box}{A list containing \code{x} and \code{y} (each of length 2),
corresponding to the values for \code{ll} and \code{ur}, such as would
be produced by a call to \code{locator(2)}. If provided, neither
\code{ll} nor \code{ur} may be provided.}

\item{debug}{A flag that turns on debugging.  Set to 1 to get a moderate
amount of debugging information, or a higher value for more debugging.}
}
\value{
A \linkS4class{landsat} object, with data having
been trimmed as specified.
}
\description{
Trim a landsat image to a latitude-longitude box. This is only an approximate
operation, because landsat images are provided in x-y coordinates, not
longitude-latitude coordinates.
}
\details{
As of June 25, 2015, the matrices storing the image data are trimmed to indices
determined by linear interpolation based on the location of the \code{ll} and
\code{ur} corners within the lon-lat corners specified in the image data. (A
previous version trimmed in UTM space, and in fact this may be done in future
also, if a problem in lonlat/utm conversion is resolved.) An error results if
there is no intersection between the trimming box and the image box.
}
\seealso{
The documentation for the \linkS4class{landsat} class explains the
structure of landsat objects, and also outlines the other functions dealing
with them.

Other things related to landsat data: 
\code{\link{[[,landsat-method}},
\code{\link{[[<-,landsat-method}},
\code{\link{landsat-class}},
\code{\link{landsatAdd}()},
\code{\link{landsat}},
\code{\link{plot,landsat-method}},
\code{\link{read.landsat}()},
\code{\link{summary,landsat-method}}
}
\author{
Dan Kelley and Clark Richards
}
\concept{things related to landsat data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/AllClass.R
\name{plot,oce-method}
\alias{plot,oce-method}
\alias{plot.oce}
\title{Plot an oce Object}
\usage{
\S4method{plot}{oce}(x, y, ...)
}
\arguments{
\item{x}{a basic \linkS4class{oce} object,
but not from any subclass that derive from this base, because
subclasses have their own plot methods, e.g. calling \code{plot()} on a
\linkS4class{ctd} object dispatches to \code{\link[=plot,ctd-method]{plot,ctd-method()}}.}

\item{y}{Ignored; only present here because S4 object for generic \code{plot}
need to have a second parameter before the \code{...} parameter.}

\item{...}{Passed to \code{\link[=hist]{hist()}}, \code{\link[=plot]{plot()}}, or to
\code{\link[=pairs]{pairs()}}, according to whichever does the plotting.}
}
\description{
This creates a \code{\link[=pairs]{pairs()}} plot of the elements in the \code{data}
slot, if there are more than 2 elements there, or a simple xy plot if 2
elements, or a histogram if 1 element.
}
\examples{
library(oce)
o <- new("oce")
o <- oceSetData(o, 'x', rnorm(10))
o <- oceSetData(o, 'y', rnorm(10))
o <- oceSetData(o, 'z', rnorm(10))
plot(o)
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.R
\name{plot,ctd-method}
\alias{plot,ctd-method}
\alias{plot.ctd}
\title{Plot a ctd Object}
\usage{
\S4method{plot}{ctd}(
  x,
  which,
  col = par("fg"),
  fill,
  borderCoastline = NA,
  colCoastline = "lightgray",
  eos = getOption("oceEOS", default = "gsw"),
  ref.lat = NaN,
  ref.lon = NaN,
  grid = TRUE,
  coastline = "best",
  Slim,
  Clim,
  Tlim,
  plim,
  densitylim,
  N2lim,
  Rrholim,
  dpdtlim,
  timelim,
  drawIsobaths = FALSE,
  clongitude,
  clatitude,
  span,
  showHemi = TRUE,
  lonlabels = TRUE,
  latlabels = TRUE,
  latlon.pch = 20,
  latlon.cex = 1.5,
  latlon.col = "red",
  projection = NULL,
  cex = 1,
  cex.axis = par("cex.axis"),
  pch = 1,
  useSmoothScatter = FALSE,
  df,
  keepNA = FALSE,
  type,
  mgp = getOption("oceMgp"),
  mar = c(mgp[1] + 1.5, mgp[1] + 1.5, mgp[1] + 1.5, mgp[1] + 1),
  inset = FALSE,
  add = FALSE,
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{x}{a \linkS4class{ctd} object.}

\item{which}{a numeric or character vector specifying desired plot types. If
\code{which} is not supplied, a default will be used. This default depends on
\code{deploymentType} in the \code{metadata} slot of \code{x}.  If \code{deploymentType} is
\code{"profile"} or missing, then \code{which} defaults to \code{c(1,2,3,5)}.  If
\code{deploymentType} is \code{"moored"} or \code{"thermosalinograph"} then \code{which} defaults
to \code{c(30, 3, 31, 5)}. Finally, if \code{deploymentType} is \code{towyo} then \code{which}
defaults to \code{c(30, 31, 32, 3)}.

The details of individual \code{which} values are as follows. Some of the entries
refer to the EOS (equation of state for seawater), which may either \code{"gsw"}
for the modern Gibbs Seawater system, or \code{"unesco"} for the older UNESCO
system. The EOS may be set with the \code{eos} argument to \verb{plot,ctd-method()} or
by using \code{\link[=options]{options()}}, with \code{options(oceEOS="unesco")} or
\code{options(oceEOS="unesco")}. The default EOS is \code{"gsw"}.
\itemize{
\item \code{which=1} or \code{which="salinity+temperature"} gives a combined profile of
temperature and salinity.  If the EOS is \code{"gsw"} then Conservative
Temperature and Absolute Salinity are shown; otherwise in-situ temperature
and practical salinity are shown.
\item \code{which=2} or \code{which="density+N2"} gives a combined profile of density
anomaly, computed with \code{\link[=swSigma0]{swSigma0()}}, along with the square of the buoyancy
frequency, computed with \code{\link[=swN2]{swN2()}}. The \code{eos} parameter is passed
to each of these functions, so the desired EOS is used.
\item \code{which=3} or \code{which="TS"} gives a TS plot. If the EOS is \code{"gsw"}, T is
Conservative Temperature and S is Absolute Salinity; otherwise, they are
in-situ temperature and practical salinity, respectively.
\item \code{which=4} or \code{which="text"} gives a textual  summary of
some aspects of the data.
\item \code{which=5} or \code{which="map"} gives a map plotted with
\code{\link[=plot,coastline-method]{plot,coastline-method()}}, with a dot for the station location.  Notes near
the top boundary of the map give the station number, the sampling date, and
the name of the chief scientist, if these are known. Note that the longitude
will be converted to a value between -180 and 180 before plotting.  (See also
notes about \code{span}.)
\item \code{which=5.1} as for \code{which=5}, except that the file name
is drawn above the map.
\item \code{which=6} or \code{which="density+dpdt"} gives a profile of density and
\eqn{dP/dt}{dP/dt}, which is useful for evaluating whether the instrument is
dropping properly through the water column.  If the EOS is \code{"gsw"} then
\eqn{\sigma_0}{sigma[0]} is shown; otherwise,
\eqn{\sigma_\theta}{sigma[theta]} is shown.
\item \code{which=7} or \code{which="density+time"} gives a profile of density and time.
\item \code{which=8} or \code{which="index"} gives a profile of index number, which can
provide useful information for trimming with \code{\link[=ctdTrim]{ctdTrim()}}.
\item \code{which=9} or \code{which="salinity"} gives a profile of Absolute
Salinity if the EOS is \code{"gsw"}, or practical salinity otherwise.
\item \code{which=10} or \code{which="temperature"} gives a profile
of Conservative Temperature if the EOS is \code{"gsw"}, or
in-situ temperature otherwise.
\item \code{which=11} or \code{which="density"} gives a profile
of density as computed with \code{\link[=swRho]{swRho()}}, to which the \code{eos}
parameter is passed.
\item \code{which=12} or \code{which="N2"} gives an \eqn{N^2}{N^2} profile.
\item \code{which=13} or \code{which="spice"} gives a spiciness profile.
\item \code{which=14} or \code{which="tritium"} gives a tritium profile.
\item \code{which=15} or \code{which="Rrho"} gives a diffusive-case density
ratio profile.
\item \code{which=16} or \code{which="RrhoSF"} gives a salt-finger case
density ratio profile.
\item \code{which=17} or \code{which="conductivity"} gives a conductivity profile.
\item \code{which=20} or \code{which="CT"} gives a profile of Conservative Temperature.
\item \code{which=21} or \code{which="SA"} gives a profile of Absolute Salinity.
\item \code{which=30} or \code{which="Sts"} gives a time series of Salinity Absolute
Salinity if the EOS is \code{"gsw"} or practical salinity otherwise.
\item \code{which=31} or \code{which="Tts"} gives a time series of Conservative Temperature
if the EOS is \code{"gsw"} or in-situ temperature otherwise.
\item \code{which=32} or \code{which="pts"} gives a time series of pressure
\item \code{which=33} or \code{which="rhots"} gives a time series of density anomaly,
\eqn{\sigma_0}{sigma[0]} if the EOS is \code{"gsw"}
or \eqn{\sigma_\theta}{sigma[theta]} otherwise.
\item otherwise, \code{which} is interpreted as a character value to be checked
against the \code{data} and \code{dataDerived} fields returned by \verb{x[["?"]}. If a match
is found then a profile of the corresponding quantity is plotted.  If there
is no match, an error is reported.
}}

\item{col}{color of lines or symbols.}

\item{fill}{a legacy parameter that will be permitted only temporarily; see
\dQuote{History}.}

\item{borderCoastline}{color of coastlines and international borders, passed
to \code{\link[=plot,coastline-method]{plot,coastline-method()}} if a map is included in \code{which}.}

\item{colCoastline}{fill color of coastlines and international borders, passed
to \code{\link[=plot,coastline-method]{plot,coastline-method()}} if a map is included in \code{which}. Set to
\code{NULL} to avoid filling.}

\item{eos}{character value indicating the equation of state to be used, either
\code{"unesco"} or \code{"gsw"}.  The default is to use a value stored with
\code{\link[=options]{options()}} as e.g. \code{options(oceEOS="unesco")}.}

\item{ref.lat}{latitude of reference point for distance calculation.
The permitted range is -90 to 90.}

\item{ref.lon}{longitude of reference point for distance calculation.
The permitted range is -180 to 180.}

\item{grid}{logical value indicating whether to draw a grid on the plot.}

\item{coastline}{a specification of the coastline to be used for
\code{which="map"}.  This may be a coastline object, whether built-in or
supplied by the user, or a character string.  If the later, it may be the
name of a built-in coastline (\code{"coastlineWorld"},
\code{"coastlineWorldFine"}, or
\code{"coastlineWorldCoarse"}), or \code{"best"}, to choose
a suitable coastline for the locale, or \code{"none"} to prevent
the drawing of a coastline.  There is a speed penalty for providing
\code{coastline} as a character string, because it forces
\code{\link[=plot,coastline-method]{plot,coastline-method()}} to load it on every call.  So, if
\code{\link[=plot,coastline-method]{plot,coastline-method()}} is to be called several times for a given
coastline, it makes sense to load it in before the first call, and to
supply the object as an argument, as opposed to the name of the object.}

\item{Slim, Clim, Tlim, plim, densitylim, N2lim, Rrholim, dpdtlim, timelim}{optional
numeric vectors of length 2, that give axis limits for salinity,
conductivity, temperature, pressure, the square of buoyancy frequency,
density ratio, dp/dt, and time, respectively.}

\item{drawIsobaths}{logical value indicating whether to draw depth contours on
maps, in addition to the coastline. The argument has no effect except
for panels in which the value of \code{which} equals \code{"map"} or
the equivalent numerical code, \code{5}. If \code{drawIsobaths} is
\code{FALSE}, then no contours are drawn. If \code{drawIsobaths}
is \code{TRUE}, then contours are selected automatically,
using \link{pretty}\code{(c(0,300))} if the station depth is
under 100m or \link{pretty}\code{(c(0,5500))} otherwise.
If \code{drawIsobaths} is a numerical vector,
then the indicated depths are drawn. For plots drawn with \code{projection}
set to \code{NULL}, the contours are added with \code{\link[=contour]{contour()}}
and otherwise \code{\link[=mapContour]{mapContour()}} is used. To customize
the resultant contours, e.g. setting particular line types or colors,
users should call these functions directly (see e.g. Example 2).}

\item{clongitude, clatitude, span}{controls for the map area view,
used only if \code{which="map"}.
\code{clongitude} and \code{clatitude} specify the centre of the view, and
\code{span} specifies the approximate extend of the view, in
kilometres.  (If \code{span} is not given, it is be determined as a small
multiple of the distance to the nearest point of land, in an attempt to show
the station in familiar geographical context.)}

\item{showHemi, lonlabels, latlabels}{controls for axis labelling, used only if \code{which="map"}.
\code{showHemi} is logical value indicating whether to show hemisphere in axis tick
labels. \code{lonlabels} and \code{latlabels} are numeric and character values that control the
axis labelling.}

\item{latlon.pch, latlon.cex, latlon.col}{controls for station location,
used only if \code{which="map"}.  \code{latlon.pch} sets the symbol code,
\code{latlon.cex} sets the character expansion
factor, and \code{latlon.col} sets the colour.}

\item{projection}{controls the map projection (if any), and ignored unless
\code{which="map"}. The possibilities are as follows.  (1) If \code{projection=NULL}
(the default) then no projection will be used; the map will simply show
longitude and latitude in a Cartesian frame, scaled to retain shapes at the
centre.  (2) If \verb{projection=}"automatic"\verb{then either a Mercator or Stereographic projection will be used, depending on whether the CTD station is within 70 degrees of the equator or at higher latitudes. (3) If}projection` is a string in the format used by \code{\link[=mapPlot]{mapPlot()}}, then it is is
passed to that function.}

\item{cex}{size to be used for plot symbols (see \code{\link[=par]{par()}}).}

\item{cex.axis}{size factor for axis labels (see \code{\link[=par]{par()}}).}

\item{pch}{code for plotting symbol (see \code{\link[=par]{par()}}).}

\item{useSmoothScatter}{logical value indicating whether to
use \code{\link[=smoothScatter]{smoothScatter()}} instead of \code{\link[=plot]{plot()}} to draw the plot.}

\item{df}{optional numeric argument that is ignored except for plotting buoyancy
frequency; in that case, it is passed to \code{\link[=swN2]{swN2()}}.}

\item{keepNA}{logical value indicating whether \code{NA} values
will yield breaks in lines drawn if \code{type} is \code{b}, \code{l}, or \code{o}.
The default value is \code{FALSE}.  Setting \code{keepNA} to \code{TRUE}
strung together into one \linkS4class{ctd} object, which otherwise
would have extraneous lines joining the deepest point in one
profile to the shallowest in the next profile.}

\item{type}{the type of plot to draw, using the same scheme as
\code{\link[=plot]{plot()}}. If supplied, this is increased to be the
same length as \code{which}, if necessary, and then supplied to
each of the individual plot calls. If it is not supplied,
then those plot calls use defaults (e.g. using a line for
\code{\link[=plotProfile]{plotProfile()}}, using dots for \code{\link[=plotTS]{plotTS()}},
etc).}

\item{mgp}{three-element numerical vector specifying axis-label geometry,
passed to \code{\link[=par]{par()}}.
The default establishes tighter margins than in the usual R setup.}

\item{mar}{four-element numerical vector specifying margin geometry,
passed to \code{\link[=par]{par()}}.
The default establishes tighter margins than in the usual R setup.
Note that the value of \code{mar} is ignored for the map panel
of multi-panel maps; instead, the present value of
\link{par}\code{("mar")} is used, which in the default call will
make the map plot region equal that of the previously-drawn
profiles and TS plot.}

\item{inset}{logical value indicating whether this
function is being used as an inset.  The
effect is to prevent the present function from adjusting margins, which is
necessary because margin adjustment is the basis for the method used by
\code{\link[=plotInset]{plotInset()}}.}

\item{add}{logical value indicating whether to add to an existing plot.  This
only works if \code{length(which)=1}, and it will yield odd results if the
value of \code{which} does not match that in the previous plots.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}

\item{...}{optional arguments passed to plotting functions.}
}
\description{
Plot CTD data in any of many different ways. In many cases,
the best choice is to use default values for all parameters
other than the first.  This yields a 4-panel plot that
displays a basic overview of the data, with a combined
profile of salinity and temperature at the top
left, a combined plot of density and the square of buoyancy
frequency at top right, a TS diagram at bottom left, and
a map at bottom right.
}
\details{
The
default values of \code{which} and other arguments are chosen to be useful
for quick overviews of data. However, for detailed work it is common
to call the present function with just a single value of \code{which}, e.g.
with four calls to get four panels. The advantage of this is that it provides
much more control over the display, and also it permits the addition of extra
display elements (lines, points, margin notes, etc.) to the individual panels.

Note that panels that draw more than one curve (e.g. \code{which="salinity+temperature"}
draws temperature and salinity profiles in one graph), the value of \link{par}\code{("usr")}
is established by the second profile to have been drawn. Some experimentation will
reveal what this profile is, for each permitted \code{which} case, although
it seems unlikely that this will help much ... the simple fact is that drawing two
profiles in one graph is useful for a quick overview, but not useful for e.g. interactive
analysis with \code{\link[=locator]{locator()}} to flag bad data, etc.
}
\section{History of Changes}{

\itemize{
\item January 2022:
\itemize{
\item Add ability to profile anything stored in the \code{data} slot, and anything
that can be computed from information in that slot. The list of
possibilities is found by examining the \code{data} and \code{dataDerived} elements
of \code{x[["?"]]}.
\item Drop the \code{lonlim} and \code{latlim} parameters, marked for removal in 2014;
use \code{clongitude}, \code{clatitude} and \code{span} instead (see
\code{\link[=plot,coastline-method]{plot,coastline-method()}}).
}
\item February 2016:
\itemize{
\item Drop the \code{fill} parameter for land colour; use \code{colCoastline} instead.
\item Add the \code{borderCoastline} argument, to control the colour of coastlines
and international boundaries.
}
}
}

\examples{
## 1. simple plot
library(oce)
data(ctd)
plot(ctd)

## 2. how to customize depth contours
par(mfrow=c(1,2))
data(section)
stn <- section[["station", 105]]
plot(stn, which="map", drawIsobaths=TRUE)
plot(stn, which="map")
data(topoWorld)
tlon <- topoWorld[["longitude"]]
tlat <- topoWorld[["latitude"]]
tdep <- -topoWorld[["z"]]
contour(tlon, tlat, tdep, drawlabels=FALSE,
        levels=seq(1000,6000,1000), col='lightblue', add=TRUE)
contour(tlon, tlat, tdep, vfont=c("sans serif", "bold"),
        levels=stn[['waterDepth']], col='red', lwd=2, add=TRUE)

}
\seealso{
The documentation for \linkS4class{ctd} explains the structure of CTD
objects, and also outlines the other functions dealing with them.

Other functions that plot oce data: 
\code{\link{download.amsr}()},
\code{\link{plot,adp-method}},
\code{\link{plot,adv-method}},
\code{\link{plot,amsr-method}},
\code{\link{plot,argo-method}},
\code{\link{plot,bremen-method}},
\code{\link{plot,cm-method}},
\code{\link{plot,coastline-method}},
\code{\link{plot,gps-method}},
\code{\link{plot,ladp-method}},
\code{\link{plot,landsat-method}},
\code{\link{plot,lisst-method}},
\code{\link{plot,lobo-method}},
\code{\link{plot,met-method}},
\code{\link{plot,odf-method}},
\code{\link{plot,rsk-method}},
\code{\link{plot,satellite-method}},
\code{\link{plot,sealevel-method}},
\code{\link{plot,section-method}},
\code{\link{plot,tidem-method}},
\code{\link{plot,topo-method}},
\code{\link{plot,windrose-method}},
\code{\link{plot,xbt-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{tidem-class}}

Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}
}
\author{
Dan Kelley
}
\concept{functions that plot oce data}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/tides.R
\name{webtide}
\alias{webtide}
\title{Get a Tidal Prediction from a WebTide Database}
\source{
The WebTide software may be downloaded for free at the
Department of Fisheries and Oceans (Canada) website at
\verb{http://www.bio.gc.ca/science/research-recherche/ocean/webtide/index-en.php}
(checked February 2016 and May 2017).
}
\usage{
webtide(
  action = c("map", "predict"),
  longitude,
  latitude,
  node,
  time,
  basedir = getOption("webtide"),
  region = "nwatl",
  plot = TRUE,
  tformat,
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{action}{An indication of the action, either \code{action="map"} to
draw a map or \code{action="predict"} to get a prediction; see
\sQuote{Details}.}

\item{longitude, latitude}{optional location at which prediction is required (ignored if
\code{node} is given).}

\item{node}{optional integer relating to a node in the database. If \code{node}
is given, then neither \code{latitude} nor \code{longitude} may be given.
If \code{node} is positive, then specifies indicates the node. If it is negative,
\code{\link[=locator]{locator()}} is called so that the user can click (once) on the map, after
which the node is displayed on the map.}

\item{time}{a vector of times (in the UTC timezone)
at which prediction is to be made.
If not supplied, this will be the week starting at the present time,
computed with \code{\link[=presentTime]{presentTime()}}, with a 15 minute increment.}

\item{basedir}{directory containing the \code{WebTide} application.}

\item{region}{database region, given as a directory name in the WebTide
directory.  For example, \code{h3o} is for Halifax Harbour, \code{nwatl} is
for the northwest Atlantic, and \code{sshelf} is for the Scotian Shelf and
Gulf of Maine.}

\item{plot}{boolean indicating whether to plot.}

\item{tformat}{optional argument passed to \code{\link[=oce.plot.ts]{oce.plot.ts()}}, for
plot types that call that function.  (See \code{\link[=strptime]{strptime()}} for the
format used.)}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}

\item{\dots}{optional arguments passed to plotting functions. A common
example is to set \code{xlim} and \code{ylim}, to focus a map region.}
}
\value{
The value depends on \code{action}:
\itemize{
\item If \code{action="map"} the return value is a
list containing the index of the nearest node, along with the
\code{latitude} and \code{longitude} of that node.  If
\code{plot} is \code{FALSE}, this value is returned invisibly.
\item If \code{action="predict"}, the return value is a list containing a vector
of times (\code{time}), as well as vectors of the predicted \code{elevation}
in metres and the predicted horizontal components of velocity, \code{u} and
\code{v}, along with the \code{node} number, and the \code{basedir} and
\code{region} as supplied to this function. If \code{plot} is \code{FALSE},
this value is returned invisibly.
}
}
\description{
Get a tidal prediction from a WebTide database. This only
works if the standalone WebTide application is installed,
and if it is installed in a standard location. The details
of installation are not within the oce purview.
}
\details{
There are two methods of using this function.
\emph{Case 1:} \code{action="map"}. In this case, if
\code{plot} is \code{FALSE}, a list is returned, containing
all the \code{node}s in the selected database, along with all
the \code{latitude}s and \code{longitude}s. This value is
also returned (silently) if \code{plot} is true, but in that case,
a plot is drawn to indicate the node locations. If \code{latitude} and
\code{longitude} are given, then the node nearest that spot is indicated on
the map; otherwise, if \code{node} is given, then the location of that
node is indicated. There is also a special case: if \code{node} is negative
and \code{interactive()} is \code{TRUE},
then \code{\link[=locator]{locator()}} is called, and the node nearest the spot
where the user clicks the mouse is indicated in the plot and in the
return value.

\emph{Case 2:} \code{action="predict"}. If \code{plot} is \code{FALSE},
then a list is returned, indicating \code{time}, predicted
\code{elevation}, velocity components \code{u} and \code{v},
\code{node} number, the name of the \code{basedir}, and
the \code{region}. If \code{plot} is \code{TRUE}, this list is returned
silently, and time-series plots are drawn for elevation, u, and v.

Naturally, \code{webtide} will not work unless WebTide has been installed on
the computer.
}
\section{Caution}{

WebTide is not an open-source application, so the present function was
designed based on little more than guesses about the WebTide file structure.
Users should be on the lookout for odd results.
}

\examples{
\dontrun{
## needs WebTide at the system level
library(oce)
## 1. prediction at Halifax NS
longitude <- -63.57
latitude <- 44.65
prediction <- webtide("predict", longitude=longitude, latitude=latitude)
mtext(sprintf("prediction at \%fN \%fE", latitude, longitude), line=0.75, side=3)
## 2. map
webtide(lon=-63.57,lat=44.65,xlim=c(-64,-63),ylim=c(43.0,46))
}

}
\seealso{
Other things related to tides: 
\code{\link{[[,tidem-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{as.tidem}()},
\code{\link{plot,tidem-method}},
\code{\link{predict.tidem}()},
\code{\link{summary,tidem-method}},
\code{\link{tidalCurrent}},
\code{\link{tidedata}},
\code{\link{tidem-class}},
\code{\link{tidemAstron}()},
\code{\link{tidemVuf}()},
\code{\link{tidem}}
}
\author{
Dan Kelley
}
\concept{things related to tides}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/extdata.R
\docType{data}
\encoding{UTF-8}
\name{CTD_BCD2014666_008_1_DN.ODF.gz}
\alias{CTD_BCD2014666_008_1_DN.ODF.gz}
\title{Sample ctd dataset in odf format}
\description{
The location is approximately 30km southeast of Halifax Harbour,
at "Station 2" of the Halifax Line on the Scotian Shelf.
}
\examples{
ctd <- read.ctd(system.file("extdata", "CTD_BCD2014666_008_1_DN.ODF.gz", package="oce"))
summary(ctd)
plot(ctd)

}
\seealso{
Other raw datasets: 
\code{\link{adp_rdi.000}},
\code{\link{ctd.cnv}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{test_met_csv1.csv}},
\code{\link{test_met_csv2.csv}},
\code{\link{test_met_xml2.xml}},
\code{\link{xbt.edf}}

Other things related to ctd data: 
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}

Other things related to odf data: 
\code{\link{ODF2oce}()},
\code{\link{ODFListFromHeader}()},
\code{\link{ODFNames2oceNames}()},
\code{\link{[[,odf-method}},
\code{\link{[[<-,odf-method}},
\code{\link{odf-class}},
\code{\link{plot,odf-method}},
\code{\link{read.ctd.odf}()},
\code{\link{read.odf}()},
\code{\link{subset,odf-method}},
\code{\link{summary,odf-method}}
}
\concept{raw datasets}
\concept{things related to ctd data}
\concept{things related to odf data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{dataLabel}
\alias{dataLabel}
\title{Try to associate data names with units, for use by summary()}
\usage{
dataLabel(names, units)
}
\arguments{
\item{names}{the names of data within an object}

\item{units}{the units from metadata}
}
\value{
a vector of strings, with blank entries for data with unknown units
}
\description{
Note that the whole object is not being given as an argument;
possibly this will reduce copying and thus storage impact.
}
\examples{
library(oce)
data(ctd)
dataLabel(names(ctd@data), ctd@metadata$units)
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.R
\name{beamUnspreadAdp}
\alias{beamUnspreadAdp}
\title{Adjust ADP Signal for Spherical Spreading}
\usage{
beamUnspreadAdp(
  x,
  count2db = c(0.45, 0.45, 0.45, 0.45),
  asMatrix = FALSE,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{x}{an \linkS4class{adp} object.}

\item{count2db}{a set of coefficients, one per beam, to convert from beam
echo intensity to decibels.}

\item{asMatrix}{a boolean that indicates whether to return a numeric matrix,
as opposed to returning an updated object (in which the matrix is cast to a
raw value).}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\value{
An \linkS4class{adp} object.
}
\description{
Compensate ADP signal strength for spherical spreading.
}
\details{
First, beam echo intensity is converted from counts to decibels, by
multiplying by \code{count2db}.  Then, the signal decrease owing to
spherical spreading is compensated for by adding the term
\eqn{20\log10(r)}{20*log10(r)}, where \eqn{r}{r} is the distance from the
sensor head to the water from which scattering is occurring.  \eqn{r}{r} is
given by \code{x[["distance"]]}.
}
\examples{
library(oce)
data(adp)
plot(adp, which=5) # beam 1 echo intensity
adp.att <- beamUnspreadAdp(adp)
plot(adp.att, which=5) # beam 1 echo intensity
## Profiles
par(mar=c(4, 4, 1, 1))
a <- adp[["a", "numeric"]]             # second arg yields matrix return value
distance <- adp[["distance"]]
plot(apply(a,2,mean), distance, type='l', xlim=c(0,256))
lines(apply(a,2,median), distance, type='l',col='red')
legend("topright",lwd=1,col=c("black","red"),legend=c("original","attenuated"))
## Image
plot(adp.att, which="amplitude",col=oce.colorsViridis(100))

}
\references{
The coefficient to convert to decibels is a personal
communication.  The logarithmic term is explained in textbooks on acoustics,
optics, etc.
}
\seealso{
Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley
}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/argo.R
\name{argoNames2oceNames}
\alias{argoNames2oceNames}
\title{Convert Argo Data Name to Oce Name}
\usage{
argoNames2oceNames(names, ignore.case = TRUE)
}
\arguments{
\item{names}{vector of character strings containing names in the Argo convention.}

\item{ignore.case}{a logical value passed to \code{\link[=gsub]{gsub()}}, indicating whether to
ignore the case of input strings. The default is set to \code{TRUE} because some data
files use lower-case names, despite the fact that the Argo documentation specifies
upper-case.}
}
\value{
A character vector of the same length as \code{names}, but with
replacements having been made for all known quantities.
}
\description{
This function is used internally by \code{\link[=read.argo]{read.argo()}} to convert Argo-convention
data names to oce-convention names. Users should not call this directly, since
its return value may be changed at any moment (e.g. to include units as well
as names).
}
\details{
The inference of names was done
by inspection of some data files, based on reference 1. It should be noted,
however, that the data files examined contain some names that are not
documented in reference 1, and others that are listed only in its changelog,
with no actual definitions being given. For example, the files had six distinct
variable names that seem to relate to phase in the oxygen sensor, but
these are not translated by the present function because these
variable names are not defined in reference 1, or not defined uniquely
in reference 2.

The names are converted with
\code{\link[=gsub]{gsub()}}, using the \code{ignore.case} argument of the present
function.
The procedure
is to first handle the items listed in the following table, with string
searches anchored to the start of the string. After that,
the qualifiers
\verb{_ADJUSTED}, \verb{_ERROR} and \verb{_QC},
are translated to \code{Adjusted}, \code{Error}, and \code{QC}, respectively.

\tabular{ll}{
\strong{Argo name} \tab \strong{oce name}\cr
\code{BBP} \tab \code{bbp}\cr
\code{BETA_BACKSCATTERING} \tab \code{betaBackscattering}\cr
\code{BPHASE_OXY} \tab \code{bphaseOxygen}\cr
\code{CDOM} \tab \code{CDOM}\cr
\code{CNDC} \tab \code{conductivity}\cr
\code{CHLA} \tab \code{chlorophyllA}\cr
\code{CP} \tab \code{beamAttenuation}\cr
\code{CYCLE_NUMBER} \tab \code{cycleNumber} (both this and \code{cycle} are handled by the [[ operator)\cr
\code{DATA_CENTRE} \tab \code{dataCentre}\cr
\code{DATA_MODE} \tab \code{dataMode}\cr
\code{DATA_STATE_INDICATOR} \tab \code{dataStateIndicator}\cr
\code{DC_REFERENCE} \tab \code{DCReference}\cr
\code{DIRECTION} \tab \code{direction}\cr
\code{DOWN_IRRADIANCE} \tab \code{downwellingIrradiance}\cr
\code{DOWNWELLING_PAR} \tab \code{downwellingPAR}\cr
\code{FIRMWARE_VERSION} \tab \code{firmwareVersion}\cr
\code{FIT_ERROR_NITRATE} \tab \code{fitErrorNitrate}\cr
\code{FLUORESCENCE_CDOM} \tab \code{fluorescenceCDOM}\cr
\code{FLUORESCENCE_CHLA} \tab \code{fluorescenceChlorophyllA}\cr
\code{INST_REFERENCE} \tab \code{instReference}\cr
\code{JULD} \tab \code{juld} (and used to compute \code{time})\cr
\code{JULD_QC_LOCATION} \tab \code{juldQCLocation}\cr
\code{LATITUDE} \tab \code{latitude}\cr
\code{LONGITUDE} \tab \code{longitude}\cr
\code{MOLAR_DOXY} \tab \code{oxygenUncompensated}\cr
\code{PH_IN_SITU_FREE} \tab \code{pHFree}\cr
\code{PH_IN_SITU_TOTAL} \tab \code{pH}\cr
\code{PI_NAME} \tab \code{PIName}\cr
\code{PLATFORM_NUMBER} \tab \code{id}\cr
\code{POSITION_ACCURACY} \tab \code{positionAccuracy}\cr
\code{POSITIONING_SYSTEM} \tab \code{positioningSystem}\cr
\code{PROFILE} \tab \code{profile}\cr
\code{PROJECT_NAME} \tab \code{projectName}\cr
\code{RAW_DOWNWELLING_IRRADIANCE} \tab \code{rawDownwellingIrradiance}\cr
\code{RAW_DOWNWELLING_PAR} \tab \code{rawDownwellingPAR}\cr
\code{RAW_UPWELLING_RADIANCE} \tab \code{rawUpwellingRadiance}\cr
\code{STATION_PARAMETERS} \tab \code{stationParameters}\cr
\code{TEMP} \tab \code{temperature}\cr
\code{TEMP_CPU_CHLA} \tab \code{temperatureCPUChlorophyllA}\cr
\code{TEMP_DOXY} \tab \code{temperatureOxygen}\cr
\code{TEMP_NITRATE} \tab \code{temperatureNitrate}\cr
\code{TEMP_PH} \tab \code{temperaturePH}\cr
\code{TEMP_SPECTROPHOTOMETER_NITRATE} \tab \code{temperatureSpectrophotometerNitrate}\cr
\code{TILT} \tab \code{tilt}\cr
\code{TURBIDITY} \tab \code{turbidity}\cr
\code{UP_RADIANCE} \tab \code{upwellingRadiance}\cr
\code{UV_INTENSITY} \tab \code{UVIntensity}\cr
\code{UV_INTENSITY_DARK_NITRATE} \tab \code{UVIntensityDarkNitrate}\cr
\code{UV_INTENSITY_NITRATE} \tab \code{UVIntensityNitrate}\cr
\code{VRS_PH} \tab \code{VRSpH}\cr
\code{WMO_INST_TYPE} \tab \code{WMOInstType}\cr
}
}
\references{
\enumerate{
\item Argo User's Manual Version 3.3, Nov 89th, 2019, available at
\url{https://archimer.ifremer.fr/doc/00187/29825/} online.
\item Argo list of parameters in an excel spreadsheet, available at
\url{http://www.argodatamgt.org/content/download/27444/187206/file/argo-parameters-list-core-and-b.xlsx}
}
}
\seealso{
Other things related to argo data: 
\code{\link{[[,argo-method}},
\code{\link{[[<-,argo-method}},
\code{\link{argo-class}},
\code{\link{argoGrid}()},
\code{\link{argo}},
\code{\link{as.argo}()},
\code{\link{handleFlags,argo-method}},
\code{\link{plot,argo-method}},
\code{\link{read.argo}()},
\code{\link{subset,argo-method}},
\code{\link{summary,argo-method}}
}
\concept{things related to argo data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{titleCase}
\alias{titleCase}
\title{Capitalize first letter of each of a vector of words}
\usage{
titleCase(w)
}
\arguments{
\item{w}{vector of character strings}
}
\value{
vector of strings patterned on \code{w} but with first letter
in upper case and others in lower case
}
\description{
This is used in making labels for data names in some ctd functions
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/met.R
\name{summary,met-method}
\alias{summary,met-method}
\title{Summarize a met Object}
\usage{
\S4method{summary}{met}(object, ...)
}
\arguments{
\item{object}{a \linkS4class{met} object.}

\item{\dots}{further arguments passed to or from other methods.}
}
\description{
Pertinent summary information is presented, including the sampling location,
data ranges, etc.
}
\seealso{
Other things related to met data: 
\code{\link{[[,met-method}},
\code{\link{[[<-,met-method}},
\code{\link{as.met}()},
\code{\link{download.met}()},
\code{\link{met-class}},
\code{\link{met}},
\code{\link{plot,met-method}},
\code{\link{read.met}()},
\code{\link{subset,met-method}},
\code{\link{test_met_csv1.csv}},
\code{\link{test_met_csv2.csv}},
\code{\link{test_met_xml2.xml}}
}
\author{
Dan Kelley
}
\concept{things related to met data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/map.R
\name{mapPlot}
\alias{mapPlot}
\title{Draw a Map}
\usage{
mapPlot(
  longitude,
  latitude,
  longitudelim,
  latitudelim,
  grid = TRUE,
  geographical = 0,
  bg,
  fill,
  border = NULL,
  col = NULL,
  clip = TRUE,
  type = "polygon",
  axes = TRUE,
  axisStyle = 1,
  cex,
  cex.axis = 1,
  mgp = c(0, 0.5, 0),
  drawBox = TRUE,
  showHemi = TRUE,
  polarCircle = 0,
  lonlabels = TRUE,
  latlabels = TRUE,
  projection = "+proj=moll",
  tissot = FALSE,
  trim = TRUE,
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{longitude}{either a numeric vector of longitudes of points to be plotted, or
something (an \code{oce} object, a list, or a data frame) from which both
longitude and latitude may be inferred (in which case the \code{latitude}
argument is ignored).  If \code{longitude} is missing, both it and
\code{latitude} are taken from the built-in \link{coastlineWorld} dataset.}

\item{latitude}{numeric vector of latitudes of points to be plotted (ignored
if the first argument contains both latitude and longitude).}

\item{longitudelim}{optional numeric vector of length two, indicating the
longitude limits of the plot. This value is used in the selection of
longitude lines that are shown (and possibly
labelled on the axes). In some cases, e.g. for polar views,
this can lead to odd results, with some expected longitude lines
being left out of the plot.  Altering \code{longitudelim} can
often help in such cases, e.g. \code{longitudelim=c(-180, 180)} will
force the drawing of lines all around the globe.}

\item{latitudelim}{optional vector of length two, indicating
the latitude limits of the plot. This, together with \code{longitudelim}
(and, importantly, the geometry of the plot device) is used in the
selection of map scale.}

\item{grid}{either a number (or pair of numbers) indicating the spacing of
longitude and latitude lines, in degrees, or a logical value (or pair of
values) indicating whether to draw an auto-scaled grid, or whether to skip
the grid drawing.  In the case of numerical values, \code{NA} can be used to
turn off the grid in longitude or latitude.  Grids are set up based on
examination of the scale used in middle 10 percent of the plot area, and for
most projections this works quite well.  If not, one may set
\code{grid=FALSE} and add a grid later with \code{\link[=mapGrid]{mapGrid()}}.}

\item{geographical}{flag indicating the style of axes.  With
\code{geographical=0}, the axes are conventional, with decimal degrees as
the unit, and negative signs indicating the southern and western
hemispheres.  With \code{geographical=1}, the signs are dropped, with axis
values being in decreasing order within the southern and western
hemispheres.  With \code{geographical=2}, the signs are dropped and the axes
are labelled with degrees, minutes and seconds, as appropriate, and
hemispheres are indicated with letters. With \code{geographical=3}, things
are the same as for \code{geographical=2}, but the hemisphere indication
is omitted. Finally, with \code{geographical=4}, unsigned numbers are used,
followed by letters \code{N} in the northern hemisphere, \code{S} in the southern,
\code{E} in the eastern, and \code{W} in the western.}

\item{bg}{color of the background (ignored).}

\item{fill}{is a deprecated argument; see \link{oce-deprecated}.}

\item{border}{color of coastlines and international borders (ignored unless
\code{type="polygon"}.}

\item{col}{either the color for filling polygons (if \code{type="polygon"})
or the color of the points and line segments (if \code{type="p"},
\code{type="l"}, or \code{type="o"}). If \code{col=NULL} then a default
will be set: no coastline filling for the \code{type="polygon"} case,
or black coastlines, for \code{type="p"}, \code{type="l"}, or
\code{type="o"}.}

\item{clip}{logical value indicating whether to trim any coastline elements that lie wholly
outside the plot region. This can prevent e.g. a problem of filling the whole plot area of
an Arctic stereopolar view, because the projected trace for Antarctica lies outside all
other regions so the whole of the world ends up being "land".  Setting \code{clip=FALSE}
disables this action, which may be of benefit in rare instances in the line connecting
two points on a coastline may cross the plot domain, even if those points are outside
that domain.}

\item{type}{indication of type; may be \code{"polygon"}, for a filled polygon,
\code{"p"} for points, \code{"l"} for line segments, or \code{"o"} for points
overlain with line segments.}

\item{axes}{a logical value indicating whether to draw longitude and latitude
values in the lower and left margin, respectively.  This may not work well
for some projections or scales.  See also \code{axisStyle}, \code{lonlabels}
and \code{latlabels}, which offer more granular control of labelling.}

\item{axisStyle}{an integer specifying the style of labels for the numbers
on axes.  The choices are:
1 for signed numbers without additional labels;
2 (the default) for unsigned numbers followed by letters indicating the hemisphere;
3 for signed numbers followed by a degree sign;
4 for unsigned numbers followed by a degree sign; and
5 for signed numbers followed by a degree sign and letters indicating the hemisphere.}

\item{cex}{character expansion factor for plot symbols,
used if \code{type='p'} or any other value that yields symbols.}

\item{cex.axis}{axis-label expansion factor (see \code{\link[=par]{par()}}).}

\item{mgp}{three-element numerical vector describing axis-label
placement, passed to \code{\link[=mapAxis]{mapAxis()}}.}

\item{drawBox}{logical value indicating whether to draw a box around the plot.
This is helpful for many projections at sub-global scale.}

\item{showHemi}{logical value indicating whether to show the hemisphere in
axis tick labels.}

\item{polarCircle}{a number indicating the number of degrees of latitude
extending from the poles, within which zones are not drawn.}

\item{lonlabels}{An optional logical value or numeric vector that controls
the labelling along the horizontal axis. There are four possibilities:
(1) If \code{lonlabels} is \code{TRUE} (the default), then reasonable values are inferred
and axes are drawn with ticks and labels alongside those ticks;
(2) if \code{lonlabels} is \code{FALSE}, then ticks are drawn, but no labels;
(3) if \code{lonlabels} is \code{NULL}, then no axis ticks or labels are drawn; and
(4) if \code{lonlabels} is  a vector of finite numerical values, then tick marks
are placed  at those longitudes, and labels are put alongside them.
Note that R tries to avoid overwriting labels on axes, so the instructions
See also \code{latlabels}, and note that setting \code{axes=FALSE}
ensures that no longitude or latitude axes will be drawn regardless
of the values of \code{lonlabels} and \code{latlabels}.}

\item{latlabels}{As \code{lonlabels}, but for latitude, on the left
plot axis.}

\item{projection}{optional indication of projection, in one of two
forms. First, it may be a character string in the "CRS" format that is
used by the \CRANpkg{sf} package (and in much of modern computer-based
cartography). For example, \code{projection="+proj=merc"} specifies a
Mercator projection. The second format is the output from
\code{\link[sp:CRS-class]{sp::CRS()}} in the \CRANpkg{sp} package, which is an object
with a slot named \code{projarg} that gets used as a projection string.
See \dQuote{Details}.}

\item{tissot}{logical value indicating whether to use \code{\link[=mapTissot]{mapTissot()}}
to plot Tissot indicatrices, i.e. ellipses at grid intersection points, which
indicate map distortion.}

\item{trim}{logical value indicating whether to trim islands or lakes
containing only points that are off-scale of the current plot box.  This
solves the problem of Antarctica overfilling the entire domain, for an
Arctic-centred stereographic projection.  It is not a perfect solution,
though, because the line segment joining two off-scale points might
intersect the plotting box.}

\item{debug}{a flag that turns on debugging.  Set to 1 to get a moderate
amount of debugging information, or to 2 to get more.}

\item{...}{optional arguments passed to some plotting functions.  This can
be useful in many ways, e.g.  Example 5 shows how to use \code{xlim} etc to
reproduce a scale exactly between two plots.}
}
\description{
Plot coordinates as a map, using one of the subset of projections
provided by the \CRANpkg{sf} package.  The projection information specified
with the \code{mapPlot()} call is stored in a global variable that can be retrieved
by related functions, making it easy to add points, lines, text, images
or contours to an existing map.
See the \dQuote{Details} for a list of
available projections.
}
\details{
The calculations for map projections are done with
the \CRANpkg{sf} package.  Importantly, though, not all
the \CRANpkg{sf} projections are
available in \code{oce}, for reasons relating to limitations
of \CRANpkg{sf}, for example relating to inverse-projection
calculations. The \code{oce} choices are tabulated
below, e.g. \code{projection="+proj=aea"} selects the Albers equal area
projection.  (See also the warning, below, about a problem
with \CRANpkg{sf} version 0.9-8.)

Further details of the vast array of map projections provided by PROJ
are given in reference 4.  This system has been in rapid development
since about 2018, and reference 5 provides a helpful overview of the changes
and the reasons why they were necessary.  Practical examples of map projections in
\CRANpkg{oce} are provided in reference 6, along with some notes on problems.
A fascinating treatment of the history of map projections is provided
in reference 7.  To get an idea of how projections are being created
nowadays, see reference 8, about the \code{eqearth} projection that was added
to \CRANpkg{oce} in August 2020.
}
\section{A warning about 'sf' version 0.9-8}{

This version of \CRANpkg{sf}, released in March of 2021, has errors
with respect to some projections.  This was noticed for the \code{"ortho"}
projection, but the problem may occur for other projections as well.
Therefore, the user ought to use \CRANpkg{sf} versions prior to 0.9-8,
or subsequent to it.  Most likely, this message will become moot
in the summer of 2021, when a new version of \CRANpkg{sf} will
become available on CRAN.
}

\section{Available Projections}{


The following table lists projections available in \CRANpkg{oce}, and was generated
by reformatting a subset of the output of the unix
command \code{proj -lP}. Most of the arguments have default values, and many
projections also have optional arguments.  Although e.g. \code{proj -l=aea}
provides a little more information about particular projections,
users ought to consult reference 4 for fuller details and illustrations.\tabular{lll}{
   \strong{Projection} \tab \strong{Code} \tab \strong{Arguments} \cr
   Albers equal area \tab \code{aea} \tab \code{lat_1}, \code{lat_2} \cr
   Azimuthal equidistant \tab \code{aeqd} \tab \code{lat_0}, \code{guam} \cr
   Aitoff \tab \code{aitoff} \tab - \cr
   Mod. stererographics of Alaska \tab \code{alsk} \tab - \cr
   Bipolar conic of western hemisphere \tab \code{bipc} \tab - \cr
   Bonne Werner \tab \code{bonne} \tab \code{lat_1} \cr
   Cassini \tab \code{cass} \tab - \cr
   Central cylindrical \tab \code{cc} \tab - \cr
   Equal area cylindrical \tab \code{cea} \tab \code{lat_ts} \cr
   Collignon \tab \code{collg} \tab - \cr
   Craster parabolic Putnins P4 \tab \code{crast} \tab - \cr
   Eckert I \tab \code{eck1} \tab - \cr
   Eckert II \tab \code{eck2} \tab - \cr
   Eckert III \tab \code{eck3} \tab - \cr
   Eckert IV \tab \code{eck4} \tab - \cr
   Eckert V \tab \code{eck5} \tab - \cr
   Eckert VI \tab \code{eck6} \tab - \cr
   Equidistant cylindrical plate (Caree) \tab \code{eqc} \tab \code{lat_ts}, \code{lat_0} \cr
   Equidistant conic \tab \code{eqdc} \tab \code{lat_1}, \code{lat_2} \cr
   Equal earth \tab \code{eqearth} \tab - \cr
   Euler \tab \code{euler} \tab \code{lat_1}, \code{lat_2} \cr
   Extended transverse Mercator \tab \code{etmerc} \tab - \cr
   Fahey \tab \code{fahey} \tab - \cr
   Foucault \tab \code{fouc} \tab - \cr
   Foucault sinusoidal \tab \code{fouc_s} \tab - \cr
   Gall stereographic \tab \code{gall} \tab - \cr
   Geostationary satellite view \tab \code{geos} \tab \code{h} \cr
   General sinusoidal series \tab \code{gn_sinu} \tab \code{m}, \code{n} \cr
   Gnomonic \tab \code{gnom} \tab - \cr
   Goode homolosine \tab \code{goode} \tab - \cr
   Hatano asymmetrical equal area \tab \code{hatano} \tab - \cr
   Interrupted Goode homolosine \tab \code{igh} \tab - \cr
   Kavraisky V \tab \code{kav5} \tab - \cr
   Kavraisky VII \tab \code{kav7} \tab - \cr
   Lambert azimuthal equal area \tab \code{laea} \tab - \cr
   Longitude and latitude \tab \code{longlat} \tab - \cr
   Longitude and latitude \tab \code{latlong} \tab - \cr
   Lambert conformal conic \tab \code{lcc} \tab \code{lat_1}, \code{lat_2} or \code{lat_0}, \code{k_0} \cr
   Lambert equal area conic \tab \code{leac} \tab \code{lat_1}, \code{south} \cr
   Loximuthal \tab \code{loxim} \tab - \cr
   Space oblique for Landsat \tab \code{lsat} \tab \code{lsat}, \code{path} \cr
   McBryde-Thomas flat-polar sine, no. 1 \tab \code{mbt_s} \tab - \cr
   McBryde-Thomas flat-polar sine, no. 2 \tab \code{mbt_fps} \tab - \cr
   McBryde-Thomas flat-polar parabolic \tab \code{mbtfpp} \tab - \cr
   McBryde-Thomas flat-polar quartic \tab \code{mbtfpq} \tab - \cr
   McBryde-Thomas flat-polar sinusoidal \tab \code{mbtfps} \tab - \cr
   Mercator \tab \code{merc} \tab \code{lat_ts} \cr
   Miller oblated stereographic \tab \code{mil_os} \tab - \cr
   Miller cylindrical \tab \code{mill} \tab - \cr
   Mollweide \tab \code{moll} \tab - \cr
   Murdoch I \tab \code{murd1} \tab \code{lat_1}, \code{lat_2} \cr
   Murdoch II \tab \code{murd2} \tab \code{lat_1}, \code{lat_2} \cr
   murdoch III \tab \code{murd3} \tab \code{lat_1}, \code{lat_2} \cr
   Natural earth \tab \code{natearth} \tab - \cr
   Nell \tab \code{nell} \tab - \cr
   Nell-Hammer \tab \code{nell_h} \tab - \cr
   Near-sided perspective \tab \code{nsper} \tab \code{h} \cr
   New Zealand map grid \tab \code{nzmg} \tab - \cr
   General oblique transformation \tab \code{ob_tran} \tab \code{o_proj}, \code{o_lat_p}, \code{o_lon_p}, \cr
    \tab  \tab \code{o_alpha}, \code{o_lon_c}, \code{o_lat_c}, \cr
    \tab  \tab \code{o_lon_1}, \code{o_lat_1}, \cr
    \tab  \tab \code{o_lon_2}, \code{o_lat_2} \cr
   Oblique cylindrical equal area \tab \code{ocea} \tab \code{lat_1}, \code{lat_2}, \code{lon_1}, \code{lon_2} \cr
   Oblated equal area \tab \code{oea} \tab \code{n}, \code{m}, \code{theta} \cr
   Oblique Mercator \tab \code{omerc} \tab \code{alpha}, \code{gamma}, \code{no_off}, \cr
    \tab  \tab \code{lonc}, \code{lon_1}, \code{lat_1}, \cr
    \tab  \tab \code{lon_2}, \code{lat_2} \cr
   Orthographic \tab \code{ortho} \tab - \cr
   Polyconic American \tab \code{poly} \tab - \cr
   Putnins P1 \tab \code{putp1} \tab - \cr
   Putnins P2 \tab \code{putp2} \tab - \cr
   Putnins P3 \tab \code{putp3} \tab - \cr
   Putnins P3' \tab \code{putp3p} \tab - \cr
   Putnins P4' \tab \code{putp4p} \tab - \cr
   Putnins P5 \tab \code{putp5} \tab - \cr
   Putnins P5' \tab \code{putp5p} \tab - \cr
   Putnins P6 \tab \code{putp6} \tab - \cr
   Putnins P6' \tab \code{putp6p} \tab - \cr
   Quartic authalic \tab \code{qua_aut} \tab - \cr
   Quadrilateralized spherical cube \tab \code{qsc} \tab - \cr
   Robinson \tab \code{robin} \tab - \cr
   Roussilhe stereographic \tab \code{rouss} \tab - \cr
   Sinusoidal aka Sanson-Flamsteed \tab \code{sinu} \tab - \cr
   Swiss. oblique Mercator \tab \code{somerc} \tab - \cr
   Stereographic \tab \code{stere} \tab \code{lat_ts} \cr
   Oblique stereographic alternative \tab \code{sterea} \tab - \cr
   Transverse cylindrical equal area \tab \code{tcea} \tab - \cr
   Tissot \tab \code{tissot} \tab \code{lat_1}, \code{lat_2} \cr
   Transverse Mercator \tab \code{tmerc} \tab \code{approx} \cr
   Two point equidistant \tab \code{tpeqd} \tab \code{lat_1}, \code{lon_1}, \code{lat_2}, \code{lon_2} \cr
   Tilted perspective \tab \code{tpers} \tab \code{tilt}, \code{azi}, \code{h} \cr
   Universal polar stereographic \tab \code{ups} \tab \code{south} \cr
   Urmaev flat-polar sinusoidal \tab \code{urmfps} \tab \code{n} \cr
   Universal transverse Mercator \tab \code{utm} \tab \code{zone}, \code{south}, \code{approx} \cr
   van der Grinten I \tab \code{vandg} \tab - \cr
   Vitkovsky I \tab \code{vitk1} \tab \code{lat_1}, \code{lat_2} \cr
   Wagner I Kavraisky VI \tab \code{wag1} \tab - \cr
   Wagner II \tab \code{wag2} \tab - \cr
   Wagner III \tab \code{wag3} \tab \code{lat_ts} \cr
   Wagner IV \tab \code{wag4} \tab - \cr
   Wagner V \tab \code{wag5} \tab - \cr
   Wagner VI \tab \code{wag6} \tab - \cr
   Werenskiold I \tab \code{weren} \tab - \cr
   Winkel I \tab \code{wink1} \tab \code{lat_ts} \cr
   Winkel Tripel \tab \code{wintri} \tab \code{lat_ts} \cr
}
}

\section{Choosing a projection}{

The best choice of projection depends on the application.
Users may find \code{projection="+proj=moll"} useful for world-wide
plots, \code{ortho} for hemispheres viewed from the equator, \code{stere}
for polar views, \code{lcc} for wide meridional ranges in mid latitudes,
\code{merc} in limited-area cases where angle preservation is
important, or either \code{aea} or \code{eqearth} (on local and global
scales, respectively) where area preservation is important.
The choice becomes more important, the larger the size of the region
represented.  When it comes to publication, it can be sensible to use the
same projection as used in previous reports.
}

\section{Problems}{

Map projection is a complicated matter that is addressed here
in a limited and pragmatic way.  For example, \code{mapPlot} tries to draw
axes along a box containing the map, instead of trying to find spots along
the ``edge'' of the map at which to put longitude and latitude labels.
This design choice greatly simplifies the coding effort, freeing up time to
work on issues regarded as more pressing.  Chief among those issues are (a)
the occurrence of horizontal lines in maps that have prime meridians
(b) inaccurate filling of land regions that (again) occur with shifted
meridians and (c) inaccurate filling of Antarctica in some projections.
Generally, issues are tackled first for commonly used projections, such as
those used in the examples.
}

\section{Changes}{

\itemize{
\item 2020-12-24: complete switch from \CRANpkg{rgdal} to \CRANpkg{sf},
removing the testing scheme created on 2020-08-03.
\item 2020-08-03: added support for the \code{eqearth} projection (like \code{robin} but
an equal-area method).
\item 2020-08-03: dropped support for the \code{healpix}, \code{pconic}
and \code{rhealpix} projections, which caused errors with the
\CRANpkg{sf} package.  (This is not a practical loss, since these
interrupted projections were handled badly by \code{mapPlot()}
in any case.)
\item 2020-08-03: switch from \CRANpkg{rgdal} to \CRANpkg{sf} for
calculations related to map projection, owing to some
changes in the former package that broke \CRANpkg{oce}
code.  (To catch problems, \CRANpkg{oce} was set up to use
both packages temporarily, issuing warnings if the results differed
by more than 1 metre in easting or northing values.)
\item 2019-03-20: the test code provided the \dQuote{Examples} section
is disabled on i386/windows machines, on which the requisite
\CRANpkg{rgdal} package continues to fail on common projections.
\item 2017-11-19: \code{imw_p} removed, because it has problems doing
inverse calculations.
This is a also problem in the standalone PROJ.4 application version
4.9.3, downloaded and built on OSX.
See \url{https://github.com/dankelley/oce/issues/1319} for details.
\item 2017-11-17: \code{lsat} removed, because it does not work in
\CRANpkg{rgdal} or in the latest standalone PROJ.4 application.
This is a also problem in the standalone PROJ.4 application version
4.9.3, downloaded and built on OSX.
See \url{https://github.com/dankelley/oce/issues/1337} for details.
\item 2017-09-30: \code{lcca} removed, because its inverse was
wildly inaccurate in a Pacific Antarctic-Alaska application
(see \url{https://github.com/dankelley/oce/issues/1303}).
}
}

\examples{
\donttest{
library(oce)
data(coastlineWorld)

# Example 1.
# Mollweide (referenc 1 page 54) is an equal-area projection that works well
# for whole-globe views.
mapPlot(coastlineWorld, projection="+proj=moll", col='gray')
mtext("Mollweide", adj=1)

# Example 2.
# Note that filling is not employed (`col` is not
# given) when the prime meridian is shifted, because
# this causes a problem with Antarctica
cl180 <- coastlineCut(coastlineWorld, lon_0=-180)
mapPlot(cl180, projection="+proj=moll +lon_0=-180")
mtext("Mollweide with coastlineCut", adj=1)

# Example 3.
# Orthographic projections resemble a globe, making them attractive for
# non-technical use, but they are neither conformal nor equal-area, so they
# are somewhat limited for serious use on large scales.  See Section 20 of
# reference 1. Note that filling is not employed because it causes a problem with
# Antarctica.
if (utils::packageVersion("sf") != "0.9.8") {
    # sf version 0.9-8 has a problem with this projection
    par(mar=c(3, 3, 1, 1))
    mapPlot(coastlineWorld, projection="+proj=ortho +lon_0=-180")
    mtext("Orthographic", adj=1)
}

# Example 4.
# The Lambert conformal conic projection is an equal-area projection
# recommended by reference 1, page 95, for regions of large east-west extent
# away from the equator, here illustrated for the USA and Canada.
par(mar=c(3, 3, 1, 1))
mapPlot(coastlineCut(coastlineWorld, -100),
        longitudelim=c(-130,-55), latitudelim=c(35, 60),
        projection="+proj=lcc +lat_0=30 +lat_1=60 +lon_0=-100", col='gray')
mtext("Lambert conformal", adj=1)

# Example 5.
# The stereographic projection (reference 1, page 120) is conformal, used
# below for an Arctic view with a Canadian focus.  Note the trick of going
# past the pole: the second latitudelim value is 180 minus the first, and the
# second longitudelim is 180 plus the first; this uses image points "over"
# the pole.
par(mar=c(3, 3, 1, 1))
mapPlot(coastlineCut(coastlineWorld, -135),
        longitudelim=c(-130, 50), latitudelim=c(70, 110),
        projection="+proj=stere +lat_0=90 +lon_0=-135", col='gray')
mtext("Stereographic", adj=1)

# Example 6.
# Spinning globe: create PNG files that can be assembled into a movie
}

\dontrun{
if (utils::packageVersion("sf") != "0.9.8") {
    # sf version 0.9-8 has a problem with this projection
    png("globe-\%03d.png")
    lons <- seq(360, 0, -15)
    par(mar=rep(0, 4))
    for (i in seq_along(lons)) {
        p <- paste("+proj=ortho +lat_0=30 +lon_0=", lons[i], sep="")
        if (i == 1) {
            mapPlot(coastlineCut(coastlineWorld, lons[i]), projection=p, col="gray")
            xlim <- par("usr")[1:2]
            ylim <- par("usr")[3:4]
        } else {
            mapPlot(coastlineCut(coastlineWorld, lons[i]), projection=p, col="gray",
                    xlim=xlim, ylim=ylim, xaxs="i", yaxs="i")
        }
    }
    dev.off()
}
}

}
\references{
\enumerate{
\item Snyder, John P., 1987.  Map Projections: A Working Manual.  USGS
Professional Paper: 1395
\url{https://pubs.er.usgs.gov/publication/pp1395}
\doi{10.3133/pp1395}
\item Natural Resources Canada
\url{https://www.nrcan.gc.ca/earth-sciences/geography/topographic-information/maps/9805}
\item "List of Map Projections." In Wikipedia, January 26, 2021.
\url{https://en.wikipedia.org/w/index.php?title=List_of_map_projections}.
\item PROJ contributors (2020).
"PROJ Coordinate Transformation Software Library."
Open Source Geospatial Foundation, n.d.
\url{https://proj.org}.
\item Bivand, Roger (2020) Why have CRS, projections and transformations changed?
Vignette for rgdal 1.5-13
\url{https://rgdal.r-forge.r-project.org/articles/CRS_projections_transformations.html}
\item A gallery of map plots is provided at
\url{https://dankelley.github.io/r/2020/08/02/oce-proj.html}
\item Snyder, John Parr.
Flattening the Earth: Two Thousand Years of Map Projections.
Chicago, IL: University of Chicago Press, 1993.
\url{https://press.uchicago.edu/ucp/books/book/chicago/F/bo3632853.html}
\item ≈†avriƒç, Bojan, Tom Patterson, and Bernhard Jenny.
"The Equal Earth Map Projection."
International Journal of Geographical Information Science 33, no. 3 (March 4, 2019): 454-65.
\doi{10.1080/13658816.2018.1504949}
}
}
\seealso{
Points may be added to a map with \code{\link[=mapPoints]{mapPoints()}}, lines with
\code{\link[=mapLines]{mapLines()}}, text with \code{\link[=mapText]{mapText()}}, polygons with
\code{\link[=mapPolygon]{mapPolygon()}}, images with \code{\link[=mapImage]{mapImage()}}, and scale bars
with \code{\link[=mapScalebar]{mapScalebar()}}.  Points on a map may be determined with mouse
clicks using \code{\link[=mapLocator]{mapLocator()}}.  Great circle paths can be calculated
with \code{\link[=geodGc]{geodGc()}}.  See reference 8 for a demonstration of the available map
projections (with graphs).

Other functions related to maps: 
\code{\link{formatPosition}()},
\code{\link{lonlat2map}()},
\code{\link{lonlat2utm}()},
\code{\link{map2lonlat}()},
\code{\link{mapArrows}()},
\code{\link{mapAxis}()},
\code{\link{mapContour}()},
\code{\link{mapCoordinateSystem}()},
\code{\link{mapDirectionField}()},
\code{\link{mapGrid}()},
\code{\link{mapImage}()},
\code{\link{mapLines}()},
\code{\link{mapLocator}()},
\code{\link{mapLongitudeLatitudeXY}()},
\code{\link{mapPoints}()},
\code{\link{mapPolygon}()},
\code{\link{mapScalebar}()},
\code{\link{mapText}()},
\code{\link{mapTissot}()},
\code{\link{oceCRS}()},
\code{\link{shiftLongitude}()},
\code{\link{usrLonLat}()},
\code{\link{utm2lonlat}()}
}
\author{
Dan Kelley and Clark Richards
}
\concept{functions related to maps}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/tides.R
\docType{data}
\name{tidedata}
\alias{tidedata}
\title{Tidal Constituent Information}
\source{
The data come from the \code{tide3.dat} file of the \code{T_TIDE}
package (Pawlowicz et al., 2002), and derive from Appendices provided by
Foreman (1978).  The data are scanned using \file{tests/tide.R} in this
package, which also performs some tests using \code{T_TIDE} values as a
reference.
}
\description{
The \code{tidedata} dataset contains Tide-constituent information that is
use by \code{\link[=tidem]{tidem()}} to fit tidal models.  \code{tidedata} is a list
containing
\describe{
\item{\code{const}}{
a list containing vectors
\code{name} (a string with constituent name),
\code{freq} (the frequency, in cycles per hour),
\code{kmpr} (a string naming the comparison constituent, blank if there is none),
\code{ikmpr} (index of comparison constituent, or \code{0} if there is none),
\code{df} (frequency difference between constituent and its
comparison, used in the Rayleigh criterion),
\code{d1} through \code{d6} (the first through sixth Doodson numbers),
\code{semi},
\code{nsat} (number of satellite constituents),
\code{ishallow},
\code{nshallow},
\code{doodsonamp},
and
\code{doodsonspecies}.
}
\item{\code{sat}}{
a list containing vectors
\code{deldood},
\code{phcorr},
\code{amprat},
\code{ilatfac},
and
\code{iconst}.
}
\item{\code{shallow}}{
a list containing vectors
\code{iconst},
\code{coef},
and
\code{iname}.
}
}
Apart from the use of \code{d1} through \code{d6}, the naming and content
follows \code{T_TIDE} (see Pawlowicz et al. 2002), which in turn builds upon
the analysis of Foreman (1978).
}
\references{
Foreman, M. G. G., 1978. Manual for Tidal Currents Analysis and Prediction.
Pacific Marine Science Report. British Columbia, Canada: Institute of Ocean
Sciences, Patricia Bay.

Pawlowicz, Rich, Bob Beardsley, and Steve Lentz, 2002.  Classical tidal
harmonic analysis including error estimates in MATLAB using \code{T_TIDE}.
Computers and Geosciences, 28, 929-937.
}
\seealso{
Other things related to tides: 
\code{\link{[[,tidem-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{as.tidem}()},
\code{\link{plot,tidem-method}},
\code{\link{predict.tidem}()},
\code{\link{summary,tidem-method}},
\code{\link{tidalCurrent}},
\code{\link{tidem-class}},
\code{\link{tidemAstron}()},
\code{\link{tidemVuf}()},
\code{\link{tidem}},
\code{\link{webtide}()}
}
\author{
Dan Kelley
}
\concept{things related to tides}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/section.R
\name{[[<-,section-method}
\alias{[[<-,section-method}
\title{Replace Parts of a Section Object}
\usage{
\S4method{[[}{section}(x, i, j, ...) <- value
}
\arguments{
\item{x}{a \linkS4class{section} object.}

\item{i}{character value naming the item to replace.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{optional additional information (ignored).}

\item{value}{The value to be placed into \code{x}, somewhere.}
}
\description{
The \code{[[<-} method works for all \linkS4class{oce} objects.
The purpose, as with the related extraction method, \code{[[},
is to insulate users from the internal details of \linkS4class{oce}
objects, by looking for items within the various storage
slots of the object. Items not actually stored can also be
replaced, including units and data-quality
flags.
}
\details{
As with \code{[[} method, the procedure works in steps.

First, the \code{metadata} slot of \code{x} is checked to
see whether it contains something named with \code{i}.
If so, then the named item is replaced with \code{value}.

Otherwise, if the string value of \code{i} ends in \code{Unit}, then the
characters preceding that are taken as the name of a variable, and
the \code{metadata} slot of \code{x} is updated to store that unit, e.g.\preformatted{x[["temperatureUnits"]] <- list(unit=expression(degree*F),scale="")
}

Similarly, if \code{i} ends in \code{Flag}, then quality-control
flags are set up as defined by \code{result}, e.g.\preformatted{o[["temperatureFlags"]] <- c(2,4,2,2)
}

Otherwise, \code{\link[=pmatch]{pmatch()}} is used for a partial-string match with
the names of the items that are in the \code{data} slot of \code{x}.
The first item found (if any) is then updated to hold the value \code{result}.

If none of these conditions is met, a warning is issued.
}
\examples{
# 1. Change section ID from a03 to A03
data(section)
section[["sectionId"]]
section[["sectionId"]] <- toupper(section[["sectionId"]])
section[["sectionId"]]
# 2. Add a millidegree to temperatures at station 10
section[["station", 10]][["temperature"]] <-
    1e-3 + section[["station", 10]][["temperature"]]

}
\seealso{
Other functions that replace parts of oce objects: 
\code{\link{[[<-,adp-method}},
\code{\link{[[<-,amsr-method}},
\code{\link{[[<-,argo-method}},
\code{\link{[[<-,bremen-method}},
\code{\link{[[<-,cm-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{[[<-,echosounder-method}},
\code{\link{[[<-,g1sst-method}},
\code{\link{[[<-,gps-method}},
\code{\link{[[<-,ladp-method}},
\code{\link{[[<-,landsat-method}},
\code{\link{[[<-,lisst-method}},
\code{\link{[[<-,lobo-method}},
\code{\link{[[<-,met-method}},
\code{\link{[[<-,oce-method}},
\code{\link{[[<-,odf-method}},
\code{\link{[[<-,rsk-method}},
\code{\link{[[<-,sealevel-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{[[<-,topo-method}},
\code{\link{[[<-,windrose-method}},
\code{\link{[[<-,xbt-method}}

Other things related to section data: 
\code{\link{[[,section-method}},
\code{\link{as.section}()},
\code{\link{handleFlags,section-method}},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{plot,section-method}},
\code{\link{read.section}()},
\code{\link{section-class}},
\code{\link{sectionAddStation}()},
\code{\link{sectionGrid}()},
\code{\link{sectionSmooth}()},
\code{\link{sectionSort}()},
\code{\link{section}},
\code{\link{subset,section-method}},
\code{\link{summary,section-method}}
}
\author{
Dan Kelley
}
\concept{functions that replace parts of oce objects}
\concept{things related to section data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.R
\name{xyzToEnuAdp}
\alias{xyzToEnuAdp}
\title{Convert ADP From XYZ to ENU Coordinates}
\usage{
xyzToEnuAdp(x, declination = 0, debug = getOption("oceDebug"))
}
\arguments{
\item{x}{an \linkS4class{adp} object.}

\item{declination}{magnetic declination to be added to the heading after
"righting" (see below), to get ENU with N as "true" north.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\value{
An object with \code{data$v[,,1:3]} altered appropriately, and
\code{x[["oceCoordinate"]]} changed from \code{xyz} to \code{enu}.
}
\description{
Convert ADP velocity components from a xyz-based coordinate system to
an enu-based coordinate system, by using the instrument's recording of
information relating to heading, pitch, and roll. The action is based
on what is stored in the data, and so it depends greatly on instrument type
and the style of original data format. This function handles data from
RDI Teledyne, Sontek, and some Nortek instruments directly. However, Nortek
data stored in in the AD2CP format are handled by the specialized
function \code{\link[=xyzToEnuAdpAD2CP]{xyzToEnuAdpAD2CP()}}, the documentation for which
should be consulted, rather than the material given blow.
}
\details{
The first step is to convert the (x,y,z) velocity components (stored in the
three columns of \code{x[["v"]][,,1:3]}) into what RDI (reference 1, pages 11 and 12)
calls "ship" (or "righted") components.  For example, the z coordinate,
which may point upwards or downwards depending on instrument orientation, is
mapped onto a "mast" coordinate that points more nearly upwards than
downward.  The other ship coordinates are called "starboard" and "forward",
the meanings of which will be clear to mariners.  Once the (x,y,z)
velocities are converted to ship velocities, the orientation of the
instrument is extracted from heading, pitch, and roll vectors stored in the
object.  These angles are defined differently for RDI and Sontek profilers.

The code handles every case individually, based on the table given below.
The table comes from Clark Richards, a former PhD student at Dalhousie
University (reference 2), who developed it based on instrument documentation,
discussion on user groups, and analysis of measurements acquired with RDI
and Sontek acoustic current profilers in the SLEIWEX experiment.  In the
table, (X, Y, Z) denote instrument-coordinate velocities, (S, F, M) denote
ship-coordinate velocities, and (H, P, R) denote heading, pitch, and roll.

\tabular{rrrrrrrrrrrr}{ \strong{Case} \tab \strong{Mfr.} \tab
\strong{Instr.} \strong{Orient.} \tab \strong{H} \tab \strong{P} \tab
\strong{R} \tab \strong{S} \tab \strong{F} \tab \strong{M}\cr 1 \tab RDI
\tab ADCP \tab up \tab H \tab arctan(tan(P)*cos(R)) \tab R \tab -X \tab Y
\tab -Z\cr 2 \tab RDI \tab ADCP \tab down \tab H \tab arctan(tan(P)*cos(R))
\tab -R \tab X \tab Y \tab Z\cr 3 \tab Nortek \tab ADP \tab up \tab H-90
\tab R \tab -P \tab X \tab Y \tab Z\cr 4 \tab Nortek \tab ADP \tab down \tab
H-90 \tab R \tab -P \tab X \tab -Y \tab -Z\cr 5 \tab Sontek \tab ADP \tab up
\tab H-90 \tab -P \tab -R \tab X \tab Y \tab Z\cr 6 \tab Sontek \tab ADP
\tab down \tab H-90 \tab -P \tab -R \tab X \tab Y \tab Z\cr 7 \tab Sontek
\tab PCADP \tab up \tab H-90 \tab R \tab -P \tab X \tab Y \tab Z\cr 8 \tab
Sontek \tab PCADP \tab down \tab H-90 \tab R \tab -P \tab X \tab Y \tab Z\cr
}

Finally, a standardized rotation matrix is used to convert from ship
coordinates to earth coordinates.  As described in the RDI coordinate
transformation manual (reference 1, pages 13 and 14), this matrix is based on sines
and cosines of heading, pitch, and roll If \code{CH} and \code{SH} denote
cosine and sine of heading (after adjusting for declination), with similar
terms for pitch and roll using second letters \code{P} and \code{R}, the
rotation matrix is

\preformatted{ rbind(c( CH*CR + SH*SP*SR, SH*CP, CH*SR - SH*SP*CR), c(-SH*CR
+ CH*SP*SR, CH*CP, -SH*SR - CH*SP*CR), c( -CP*SR, SP, CP*CR)) }

This matrix is left-multiplied by a matrix with three rows, the top a vector
of "starboard" values, the middle a vector of "forward" values, and the
bottom a vector of "mast" values.  Finally, the columns of
\code{data$v[,,1:3]} are filled in with the result of the matrix
multiplication.
}
\references{
\enumerate{
\item Teledyne RD Instruments. \dQuote{ADCP Coordinate Transformation: Formulas and Calculations,}
January 2010. P/N 951-6079-00.
\item Clark Richards, 2012, PhD Dalhousie University Department of
Oceanography.
}
}
\seealso{
Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley and Clark Richards
}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/cm.R
\docType{class}
\name{cm-class}
\alias{cm-class}
\title{Class to Store Current Meter Data}
\description{
This class stores current meter data, e.g. from an Interocean/S4 device
or an Aanderaa/RCM device.
}
\section{Slots}{

\describe{
\item{\code{data}}{As with all \code{oce} objects, the \code{data} slot
for \code{cm} objects is a
\link{list} containing the main data for the object.
The key items stored in this slot are \code{time}, \code{u} and \code{v}.}

\item{\code{metadata}}{As with all \code{oce} objects, the \code{metadata} slot
for \code{cm} objects is a \link{list} containing
information about the \code{data} or about the object itself.
{}}

\item{\code{processingLog}}{As with all \code{oce} objects, the \code{processingLog} slot
for \code{cm} objects is a
\link{list} with entries describing the creation and evolution
of the object. The contents are updated by various \code{oce} functions to
keep a record of processing steps.  Object summaries and
\code{\link[=processingLogShow]{processingLogShow()}} both display the log.}
}}

\section{Modifying slot contents}{


Although the \code{[[<-} operator may permit modification of the contents
of \linkS4class{cm} objects (see \code{\link{[[<-,cm-method}}),
it is better to use \code{\link[=oceSetData]{oceSetData()}} and \code{\link[=oceSetMetadata]{oceSetMetadata()}},
because those functions save an entry in the \code{processingLog}
that describes the change.
}

\section{Retrieving slot contents}{


The full contents of the \code{data} and \code{metadata} slots of a \linkS4class{cm}
object may be retrieved in the standard R way using \code{\link[=slot]{slot()}}. For
example \code{slot(o,"data")} returns the \code{data} slot of an object named \code{o},
and similarly \code{slot(o,"metadata")} returns
the \code{metadata} slot.

The slots may also be obtained with the \code{\link{[[,cm-method}}
operator, as e.g. \code{o[["data"]]} and \code{o[["metadata"]]}, respectively.

The \code{\link{[[,cm-method}} operator can also
be used to retrieve items from within the \code{data} and \code{metadata} slots.
For example, \code{o[["temperature"]]} can be used to retrieve temperature
from an object containing that quantity. The rule is that a named
quantity is sought first within the object's \code{metadata} slot,
with the \code{data} slot being checked only if \code{metadata} does not
contain the item. This \code{[[} method can also be used to get
certain derived quantities, if the object contains sufficient
information to calculate them. For example, an object that holds
(practical) salinity, temperature and pressure, along with
longitude and latitude, has sufficient information to compute
Absolute Salinity, and so \code{o[["SA"]]} will yield the
calculated Absolute Salinity.

It is also possible to find items more directly, using \code{\link[=oceGetData]{oceGetData()}} and
\code{\link[=oceGetMetadata]{oceGetMetadata()}}, but neither of these functions can
retrieve derived items.
}

\seealso{
Other things related to cm data: 
\code{\link{[[,cm-method}},
\code{\link{[[<-,cm-method}},
\code{\link{as.cm}()},
\code{\link{cm}},
\code{\link{plot,cm-method}},
\code{\link{read.cm}()},
\code{\link{rotateAboutZ}()},
\code{\link{subset,cm-method}},
\code{\link{summary,cm-method}}

Other classes provided by oce: 
\code{\link{adp-class}},
\code{\link{adv-class}},
\code{\link{argo-class}},
\code{\link{bremen-class}},
\code{\link{coastline-class}},
\code{\link{ctd-class}},
\code{\link{lisst-class}},
\code{\link{lobo-class}},
\code{\link{met-class}},
\code{\link{oce-class}},
\code{\link{odf-class}},
\code{\link{rsk-class}},
\code{\link{sealevel-class}},
\code{\link{section-class}},
\code{\link{topo-class}},
\code{\link{windrose-class}},
\code{\link{xbt-class}}
}
\author{
Dan Kelley
}
\concept{classes provided by oce}
\concept{things related to cm data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.R
\name{handleFlags,adp-method}
\alias{handleFlags,adp-method}
\title{Handle Flags in adp Objects}
\usage{
\S4method{handleFlags}{adp}(
  object,
  flags = NULL,
  actions = NULL,
  where = NULL,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{object}{an \linkS4class{adp} object.}

\item{flags}{A \link{list} specifying flag values upon which
actions will be taken. This can take two forms.
\itemize{
\item In the first form, the
list has named elements each containing a vector of integers. For example,
salinities flagged with values of 1 or 3:9 would be specified
by \code{flags=list(salinity=c(1,3:9))}. Several data items can be specified,
e.g. \code{flags=list(salinity=c(1,3:9), temperature=c(1,3:9))} indicates
that the actions are to take place for both salinity and temperature.
\item In the second form, \code{flags} is a list holding a single \emph{unnamed} vector, and
this means to apply the actions to \emph{all} the data entries.  For example,
\code{flags=list(c(1,3:9))} means to apply not just to salinity and temperature,
but to everything within the \code{data} slot.
}

If \code{flags} is not provided, then \code{\link[=defaultFlags]{defaultFlags()}} is called, to try to
determine a reasonable default.}

\item{actions}{an optional \link{list} that contains items with
names that match those in the \code{flags} argument.  If \code{actions}
is not supplied, the default will be to set all values identified by
\code{flags} to \code{NA}; this can also be specified by
specifying \code{actions=list("NA")}. It is also possible to specify
functions that calculate replacement values. These are provided
with \code{object} as the single argument, and must return a
replacement for the data item in question.
See \dQuote{Details} for the default that is used if \code{actions} is not supplied.}

\item{where}{an optional character value that permits the function to work with
objects that store flags in e.g. \code{object@metadata$flags$where}
instead of in \code{object@metadata$flags}, and data within
\code{object@data$where} instead of within \code{object@data}. The
default value of \code{NULL} means to look withing \code{object@metadata}
itself, and this is the default within \code{oce}.  (The purpose of \code{where}
is to make \code{oce} extensible by other packages, which may choose to store
data two levels deep in the \code{data} slot.)}

\item{debug}{An optional integer specifying the degree of debugging, with
value 0 meaning to skip debugging and 1 or higher meaning to print some
information about the arguments and the data. It is usually a good idea to set
this to 1 for initial work with a dataset, to see which flags are being
handled for each data item. If not supplied, this defaults to the value of
\code{\link{getOption}}\code{("oceDebug")}.}
}
\description{
Data-quality flags are stored in the \code{metadata}
slot of \linkS4class{oce} objects in a
\link{list} named \code{flags}.
The present function (a generic that has specialized versions
for various data classes) provides a way to
manipulate the contents of the \code{data} slot, based on
such data-quality flags. For example, a common operation is to replace
erroneous data with \code{NA}.

If the \code{flags} within \code{object}'s \code{metadata} slot is empty,
then \code{object} is returned, unaltered.
Otherwise, \code{handleFlags} examines \code{object@metadata$flags}
in the context of the \code{flags} argument, and then
carries out actions that are specified by the \code{actions} argument.
By default, this sets the returned \code{data} entries to \code{NA},
wherever the corresponding \code{metadata$flag} values
signal unreliable data. To maintain a hint as to why
\code{data} were changed, \code{metadata$flags} in the
returned value is a direct copy of the corresponding
entry in \code{object}.
}
\details{
If \code{flags} and \code{actions} are not provided, the
default is to consider a flag value of 1 to indicate bad data,
and 0 to indicate good data. Note that it only makes sense to use
velocity (\code{v}) flags, because other flags are, at least
for some instruments, stored as \code{raw} quantities, and such
quantities may not be set to \code{NA}.
}
\examples{
# Flag low "goodness" or high "error beam" values.
library(oce)
data(adp)
# Same as Example 2 of ?'setFlags,adp-method'
v <- adp[["v"]]
i2 <- array(FALSE, dim=dim(v))
g <- adp[["g", "numeric"]]
# Set thresholds on percent "goodness" and error "velocity".
G <- 25
V4 <- 0.45
for (k in 1:3)
    i2[,,k] <- ((g[,,k]+g[,,4]) < G) | (v[,,4] > V4)
adpQC <- initializeFlags(adp, "v", 2)
adpQC <- setFlags(adpQC, "v", i2, 3)
adpClean <- handleFlags(adpQC, flags=list(3), actions=list("NA"))
# Demonstrate (subtle) change graphically.
par(mfcol=c(2, 1))
plot(adp, which="u1", drawTimeRange=FALSE)
plot(adpClean, which="u1", drawTimeRange=FALSE)
t0 <- 1214510000 # from locator()
arrows(t0, 20, t0, 35, length=0.1, lwd=3, col="magenta")
mtext("Slight change above arrow", col="magenta", font=2)

}
\seealso{
Other functions relating to data-quality flags: 
\code{\link{defaultFlags}()},
\code{\link{handleFlags,argo-method}},
\code{\link{handleFlags,ctd-method}},
\code{\link{handleFlags,oce-method}},
\code{\link{handleFlags,section-method}},
\code{\link{handleFlags}()},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{initializeFlagScheme,oce-method}},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{initializeFlagSchemeInternal}()},
\code{\link{initializeFlagScheme}()},
\code{\link{initializeFlags,adp-method}},
\code{\link{initializeFlags,oce-method}},
\code{\link{initializeFlagsInternal}()},
\code{\link{initializeFlags}()},
\code{\link{setFlags,adp-method}},
\code{\link{setFlags,ctd-method}},
\code{\link{setFlags,oce-method}},
\code{\link{setFlags}()}

Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\concept{functions relating to data-quality flags}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.woce.R
\name{oceNames2whpNames}
\alias{oceNames2whpNames}
\title{Translate Oce Data Names to WHP Data Names}
\usage{
oceNames2whpNames(names)
}
\arguments{
\item{names}{vector of strings holding oce-style names.}
}
\value{
vector of strings holding WHP-style names.
}
\description{
Translate oce-style names to WOCE names, using \code{\link[=gsub]{gsub()}}
to match patterns. For example, the pattern \code{"oxygen"}
is taken to mean \code{"CTDOXY"}.
}
\references{
Several online sources list WHP names. An example is
\url{https://cchdo.github.io/hdo-assets/documentation/manuals/pdf/90_1/chap4.pdf}
}
\seealso{
Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}

Other functions that interpret variable names and units from headers: 
\code{\link{ODFNames2oceNames}()},
\code{\link{cnvName2oceName}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{unitFromStringRsk}()},
\code{\link{unitFromString}()},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()}
}
\author{
Dan Kelley
}
\concept{functions that interpret variable names and units from headers}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ladp.R
\name{[[,ladp-method}
\alias{[[,ladp-method}
\title{Extract Something From an ladp Object}
\usage{
\S4method{[[}{ladp}(x, i, j, ...)
}
\arguments{
\item{x}{an \linkS4class{ladp} object.}

\item{i}{character value indicating the name of an item to extract.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{ignored.}
}
\description{
Generally, the \code{[[} method lets users extract information from \code{oce}
objects, without having to know the details of the internal storage.  For
many \code{oce} sub-classes, \code{[[} can also return quantities that are computed
from the object's contents.
}
\details{
A two-step process is used to try to find the requested information. First, a
class-specific function is used (see \dQuote{Details of the Specialized
Method}).  If this yields nothing, then a general method is used (see
\dQuote{Details of the General Method}).  If both methods fail, then \code{[[}
returns NULL.

Some understanding of the subclass is required to know what can be retrieved
with \code{[[}. When dealing with an unfamiliar subclass, it can be useful to
first use \code{x[["?"]]} to get a listing of the retrievable items. See
\dQuote{Details of the Specialized Method} for more information.
}
\section{Details of the Specialized Method}{


If \code{i} is \code{"?"}, then the return value is a list containing four items, each
of which is a character vector holding the names of things that can be
accessed with \code{[[}. The \code{data} and \code{metadata} items hold the names of entries
in the object's data and metadata slots, respectively. The \code{metadataDerived}
item is NULL, and the \code{dataDerived} item holds the following synonyms: \code{"p"}
for \code{"pressure"}, \code{"t"} for \code{"temperature"} and \code{"S"} for \code{"salinity"}.
}

\section{Details of the General Method}{


Note: the text of this section is identical for all \code{oce} subclasses, and so
some of what you read here may not be relevant to the class being described
in this help page.

If the specialized method produces no matches, the following generalized
method is applied. As with the specialized method, the procedure hinges first
on the values of \code{i} and, optionally, \code{j}. The work proceeds in steps, by
testing a sequence of possible conditions in sequence.
\enumerate{
\item A check is made as to whether \code{i} names one of the standard \code{oce} slots.
If so, \code{[[} returns the slot contents of that slot.  Thus, \code{x[["metadata"]]}
will retrieve the \code{metadata} slot, while \code{x[["data"]]} and
\code{x[["processingLog"]]} return those slots.
\item If \code{i} is a string ending in the \code{"Unit"}, then the characters preceding
that string are taken to be the name of an item in the data object, and a
list containing the unit is returned (or \code{NULL} if there is no such unit).
This list consists of an item named \code{unit}, which is an \code{\link[=expression]{expression()}}, and
an item named \code{scale}, which is a string describing the measurement scale.
If the string ends in \code{" unit"}, e.g.  \code{x[["temperature unit"]]} (note the
space), then just the expression is returned, and if it ends in \code{" scale"},
then just the scale is returned.
\item If \code{i} is a string ending in \code{"Flag"}, then the corresponding data-quality
flag is returned (or \code{NULL} if there is no such flag).
\item If the object holds hydrographic information (pressure, salinity,
temperature, longitude and latitude) then another set of possibilities
arises.  If \code{i} is \code{"sigmaTheta"}, then the value of \code{\link[=swSigmaTheta]{swSigmaTheta()}} is
called with \code{x} as the sole argument, and the results are returned.
Similarly, \code{\link[=swSigma0]{swSigma0()}} is used if \code{i="sigma0"}, and \code{\link[=swSpice]{swSpice()}} is used if
\code{i="spice"}. Of course, these actions only make sense for objects that
contain the relevant items within their \code{data} slot.
\item After these possibilities are eliminated, the action depends on whether
\code{j} has been provided.  If \code{j} is not provided, or is the string \code{""}, then
\code{i} is sought in the \code{metadata} slot, and then in the \code{data} slot, returning
whichever is found first.  In other words, if \code{j} is not provided, the
\code{metadata} slot takes preference over the \code{data} slot. However, if \code{j} is
provided, then it must be either the string \code{"metadata"} or \code{"data"}, and it
directs where to look.
\item If none of the above-listed conditions holds, then \code{NULL} is returned.
}
}

\seealso{
Other functions that extract parts of oce objects: 
\code{\link{[[,adp-method}},
\code{\link{[[,adv-method}},
\code{\link{[[,amsr-method}},
\code{\link{[[,argo-method}},
\code{\link{[[,bremen-method}},
\code{\link{[[,cm-method}},
\code{\link{[[,coastline-method}},
\code{\link{[[,ctd-method}},
\code{\link{[[,echosounder-method}},
\code{\link{[[,g1sst-method}},
\code{\link{[[,gps-method}},
\code{\link{[[,landsat-method}},
\code{\link{[[,lisst-method}},
\code{\link{[[,lobo-method}},
\code{\link{[[,met-method}},
\code{\link{[[,oce-method}},
\code{\link{[[,odf-method}},
\code{\link{[[,rsk-method}},
\code{\link{[[,sealevel-method}},
\code{\link{[[,section-method}},
\code{\link{[[,tidem-method}},
\code{\link{[[,topo-method}},
\code{\link{[[,windrose-method}},
\code{\link{[[,xbt-method}},
\code{\link{[[<-,adv-method}}

Other things related to ladp data: 
\code{\link{[[<-,ladp-method}},
\code{\link{as.ladp}()},
\code{\link{ladp-class}},
\code{\link{plot,ladp-method}},
\code{\link{summary,ladp-method}}
}
\author{
Dan Kelley
}
\concept{functions that extract parts of oce objects}
\concept{things related to ladp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/geod.R
\name{geodDist}
\alias{geodDist}
\title{Compute Geodesic Distance on Surface of Earth}
\usage{
geodDist(
  longitude1,
  latitude1 = NULL,
  longitude2 = NULL,
  latitude2 = NULL,
  alongPath = FALSE
)
}
\arguments{
\item{longitude1}{longitude or a vector of longitudes, \emph{or} a
\code{section} object, from which longitude and latitude are extracted and
used instead of the next three arguments}

\item{latitude1}{latitude or vector of latitudes (ignored if
\code{longitude1} is a \code{section} object)}

\item{longitude2}{optional longitude or vector of longitudes (ignored if
\code{alongPath=TRUE})}

\item{latitude2}{optional latitude or vector of latitudes (ignored if
\code{alongPath=TRUE})}

\item{alongPath}{boolean indicating whether to compute distance along the
path, as opposed to distance from the reference point.  If
\code{alongPath=TRUE}, any values provided for \code{latitude2} and
\code{longitude2} will be ignored.}
}
\value{
Vector of distances in kilometres.
}
\description{
This calculates geodesic distance between points on the earth, i.e.
distance measured along the (presumed ellipsoidal) surface. The method
involves the solution of the geodetic inverse problem, using T. Vincenty's
modification of Rainsford's method with Helmert's elliptical terms.
}
\details{
The function may be used in several different ways.

Case 1: \code{longitude1} is a \code{section} object. The values of
\code{latitude1}, \code{longitude2}, and \code{latitude2} arguments are
ignored, and the behaviour depends on the value of the \code{alongPath}
argument.  If \code{alongPath=FALSE}, the return value contains the geodetic
distances of each station from the first one.  If \code{alongPath=TRUE}, the
return value is the geodetic distance along the path connecting the
stations, in the order in which they are stored in the section.

Case 2: \code{longitude1} is a vector.  If \code{longitude2} and
\code{latitude2} are not given, then the return value is a vector containing
the distances of each point from the first one, \emph{or} the distance
along the path connecting the points, according to the value of
\code{alongPath}.  On the other hand, if both \code{longitude2} and
\code{latitude2} are specified, then the return result depends on the length
of these arguments.  If they are each of length 1, then they are taken as a
reference point, from which the distances to \code{longitude1} and
\code{latitude1} are calculated (ignoring the value of \code{alongPath}).
However, if they are of the same length as \code{longitude1} and
\code{latitude1}, then the return value is the distance between
corresponding (\code{longitude1},\code{latitude1}) and
(\code{longitude2},\code{latitude2}) values.
}
\examples{
library(oce)
km <- geodDist(100, 45, 100, 46)
data(section)
geodDist(section)
geodDist(section, alongPath=TRUE)

}
\references{
T. Vincenty, "Direct and Inverse Solutions of Ellipsoid on the
Ellipsoid with Application of Nested Equations", \emph{Survey Review}, April
1975.
}
\seealso{
\code{\link[=geodXy]{geodXy()}}

Other functions relating to geodesy: 
\code{\link{geodGc}()},
\code{\link{geodXyInverse}()},
\code{\link{geodXy}()}
}
\author{
Dan Kelley based this on R code sent to him by Darren Gillis, who in
2003 had modified Fortran code that, according to comments in the source,
had been written in 1974 by L. Pfeifer and J. G. Gergen.
}
\concept{functions relating to geodesy}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/satellite.R
\name{plot,satellite-method}
\alias{plot,satellite-method}
\alias{plot.satellite}
\title{Plot a satellite Object}
\usage{
\S4method{plot}{satellite}(x, y, asp, debug = getOption("oceDebug"), ...)
}
\arguments{
\item{x}{a \linkS4class{satellite} object.}

\item{y}{String indicating the quantity to be plotted.}

\item{asp}{Optional aspect ratio for plot.}

\item{debug}{A debugging flag, integer.}

\item{...}{extra arguments passed to \code{\link[=imagep]{imagep()}}, e.g. set
\code{col} to control colors.}
}
\description{
For an example using \code{g1sst} data, see \code{\link[=read.g1sst]{read.g1sst()}}.
}
\seealso{
Other functions that plot oce data: 
\code{\link{download.amsr}()},
\code{\link{plot,adp-method}},
\code{\link{plot,adv-method}},
\code{\link{plot,amsr-method}},
\code{\link{plot,argo-method}},
\code{\link{plot,bremen-method}},
\code{\link{plot,cm-method}},
\code{\link{plot,coastline-method}},
\code{\link{plot,ctd-method}},
\code{\link{plot,gps-method}},
\code{\link{plot,ladp-method}},
\code{\link{plot,landsat-method}},
\code{\link{plot,lisst-method}},
\code{\link{plot,lobo-method}},
\code{\link{plot,met-method}},
\code{\link{plot,odf-method}},
\code{\link{plot,rsk-method}},
\code{\link{plot,sealevel-method}},
\code{\link{plot,section-method}},
\code{\link{plot,tidem-method}},
\code{\link{plot,topo-method}},
\code{\link{plot,windrose-method}},
\code{\link{plot,xbt-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{tidem-class}}
}
\author{
Dan Kelley
}
\concept{functions that plot oce data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ad.R
\name{toEnu}
\alias{toEnu}
\title{Rotate acoustic-Doppler data to the enu coordinate system}
\usage{
toEnu(x, ...)
}
\arguments{
\item{x}{an \linkS4class{adp} or \linkS4class{adv} object.}

\item{\dots}{extra arguments that are passed on to \code{\link[=toEnuAdp]{toEnuAdp()}}
or \code{\link[=toEnuAdv]{toEnuAdv()}}.}
}
\value{
An object of the same class as \code{x}, but with velocities
in the enu coordinate system
}
\description{
Rotate acoustic-Doppler data to the enu coordinate system
}
\seealso{
Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}

Other things related to adv data: 
\code{\link{[[,adv-method}},
\code{\link{[[<-,adv-method}},
\code{\link{adv-class}},
\code{\link{adv}},
\code{\link{beamName}()},
\code{\link{beamToXyz}()},
\code{\link{enuToOtherAdv}()},
\code{\link{enuToOther}()},
\code{\link{plot,adv-method}},
\code{\link{read.adv.nortek}()},
\code{\link{read.adv.sontek.adr}()},
\code{\link{read.adv.sontek.serial}()},
\code{\link{read.adv.sontek.text}()},
\code{\link{read.adv}()},
\code{\link{rotateAboutZ}()},
\code{\link{subset,adv-method}},
\code{\link{summary,adv-method}},
\code{\link{toEnuAdv}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdv}()},
\code{\link{xyzToEnu}()}
}
\concept{things related to adp data}
\concept{things related to adv data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/section.R
\name{summary,section-method}
\alias{summary,section-method}
\alias{summary.section}
\title{Summarize a Section Object}
\usage{
\S4method{summary}{section}(object, ...)
}
\arguments{
\item{object}{An object of class \code{"section"}, usually, a result of a call
to \code{\link[=read.section]{read.section()}}, \code{\link[=read.oce]{read.oce()}}, or
\code{\link[=as.section]{as.section()}}.}

\item{...}{Further arguments passed to or from other methods.}
}
\value{
\code{NULL}
}
\description{
Pertinent summary information is presented, including station locations,
distance along track, etc.
}
\examples{
library(oce)
data(section)
summary(section)

}
\seealso{
Other things related to section data: 
\code{\link{[[,section-method}},
\code{\link{[[<-,section-method}},
\code{\link{as.section}()},
\code{\link{handleFlags,section-method}},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{plot,section-method}},
\code{\link{read.section}()},
\code{\link{section-class}},
\code{\link{sectionAddStation}()},
\code{\link{sectionGrid}()},
\code{\link{sectionSmooth}()},
\code{\link{sectionSort}()},
\code{\link{section}},
\code{\link{subset,section-method}}
}
\author{
Dan Kelley
}
\concept{things related to section data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/colors.R
\name{colormap}
\alias{colormap}
\title{Calculate color map}
\usage{
colormap(
  z = NULL,
  zlim,
  zclip = FALSE,
  breaks,
  col = oceColorsViridis,
  name,
  x0,
  x1,
  col0,
  col1,
  blend = 0,
  missingColor,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{z}{an optional vector or other set of numerical values to be examined.
If \code{z} is given, the return value will contain an item named
\code{zcol} that will be a vector of the same length as \code{z}, containing
a color for each point.  If \code{z} is not given, \code{zcol} will contain
just one item, the color \code{"black"}.}

\item{zlim}{optional vector containing two numbers that specify the \code{z}
limits for the color scale. This can only be provided in cases A
and B, as defined in \dQuote{Details}.  For case A, if \code{zlim} is not
provided, then it is inferred by using \code{\link[=rangeExtended]{rangeExtended()}}
on \code{breaks}, if that is provided, or from \code{z} otherwise.  Also,
in case A, it is an error to provide both \code{zlim} and \code{breaks},
unless the latter is of length 1, meaning the number of
subdivisions to use within the range set by \code{zlim}.
In case B, \code{zlim} is inferred from using \code{\link[=rangeExtended]{rangeExtended()}} on \code{c(x0,x1)}.
In case C, providing \code{zlim} yields an error message, because it makes no
sense in the context of a named, predefined color scheme.}

\item{zclip}{logical, with \code{TRUE} indicating that z values outside the
range of \code{zlim} or \code{breaks} should be painted with
\code{missingColor} and \code{FALSE} indicating that these values should be
painted with the nearest in-range color.}

\item{breaks}{an optional indication of break points between color levels
(see \code{\link[=image]{image()}}).  If this is provided, the arguments \code{name}
through \code{blend} are all ignored (see \dQuote{Details}).  If it is
provided, then it may either be a vector of break points, or a single number
indicating the desired number of break points to be computed with
\code{\link{pretty}}\verb{(z,breaks)}.  In either case of non-missing
\code{breaks}, the resultant break points must number 1 plus the number of
colors (see \code{col}).}

\item{col}{either a vector of colors or a function taking a numerical value
as its single argument and returning a vector of colors.  Prior to 2021-02-08,
the default for \code{col} was \code{oceColorsJet}, but it was switched to
\code{oceColorsViridis} on that date.  The value of
\code{col} is ignored if \code{name} is provided, or if \code{x0} through
\code{col1} are provided.}

\item{name}{an optional string naming a built-in colormap (one of
\code{"gmt_relief"}, \code{"gmt_ocean"}, \code{"gmt_globe"} or
\code{"gmt_gebco"}) or the name of a file or URL that contains a color map
specification in GMT format.  If \code{name} is given, then it is
passed to \code{\link[=colormapGMT]{colormapGMT()}}, which creates the colormap.
Note that the colormap thus created has a fixed
relationship between value and color, and \code{zlim},
only other argument that is examined is \code{z} (which may be used
so that \code{zcol} will be defined in the return value), and warnings
are issued if some irrelevant arguments are provided.}

\item{x0, x1, col0, col1}{Vectors that specify a color map.  They must all be
the same length, with \code{x0} and \code{x1} being numerical values, and
\code{col0} and \code{col1} being colors.  The colors may be strings (e.g.
\code{"red"}) or colors as defined by \code{\link[=rgb]{rgb()}} or \code{\link[=hsv]{hsv()}}.}

\item{blend}{a number indicating how to blend colors within each band.
This is ignored except when \code{x0} through \code{col1} are supplied.  A
value of 0 means to use \code{col0[i]} through the interval \code{x0[i]} to
\code{x1[i]}.  A value of 1 means to use \code{col1[i]} in that interval.  A
value between 0 and 1 means to blend between the two colors according to
the stated fraction.  Values exceeding 1 are an error at present, but there
is a plan to use this to indicate sub-intervals, so a smooth palette can be
created from a few colors.}

\item{missingColor}{color to use for missing values. This cannot be provided
if \code{name} is also provided (case C), because named schemes have pre-defined
colors.  For other cases, \code{missingColor} defaults to \code{"gray"}, if it
is not provided as an argument.}

\item{debug}{a flag that turns on debugging.  Set to 1 to get a moderate
amount of debugging information, or to 2 to get more.}
}
\value{
a list containing the following (not necessarily in this order)
\itemize{
\item \code{zcol}, a vector of colors for \code{z}, if \code{z} was
provided, otherwise \code{"black"}
\item \code{zlim}, a two-element vector suitable as the argument of the same
name supplied to \code{\link[=image]{image()}} or \code{\link[=imagep]{imagep()}}
\item \code{breaks} and \code{col}, vectors of breakpoints and colors,
suitable as the same-named arguments to \code{\link[=image]{image()}} or
\code{\link[=imagep]{imagep()}}
\item \code{zclip} the provided value of \code{zclip}.
\item \code{x0} and \code{x1}, numerical vectors of the sides of color
intervals, and \code{col0} and \code{col1}, vectors of corresponding
colors.  The meaning is the same as on input.  The purpose of returning
these four vectors is to permit users to alter color mapping, as in example
3 in \dQuote{Examples}.
\item \code{missingColor}, a color that could be used to specify missing
values, e.g. as the same-named argument to \code{\link[=imagep]{imagep()}}.
\item \code{colfunction}, a univariate function that returns a vector
of colors, given a vector of \code{z} values; see Example 6.
}
}
\description{
Create a mapping between numeric values and colors, for use in palettes and plots.
The return value can be used in various ways, including colorizing points
on scattergraphs, controlling images created by \code{\link[=image]{image()}} or \code{\link[=imagep]{imagep()}},
drawing palettes with \code{\link[=drawPalette]{drawPalette()}}, etc.
}
\details{
\code{colormap} can be used in a variety of ways, including the following.
\itemize{
\item \strong{Case A.} Supply some combination of arguments that
is sufficient to define a mapping of value to color, without
providing \code{x0}, \code{col0}, \code{x1} or \code{col1} (see case B for these),
or providing \code{name} (see Case C). There are several ways to
do this.  One approach is to supply \code{z} but no
other argument, in which case \code{zlim}, and \code{breaks} will be determined
from \code{z}, and the default \code{col} will be used.  Another approach is
to specify \code{breaks} and \code{col} together, in the same way as they
might be specified for the base R function \code{\link[=image]{image()}}.  It is
also possible to supply only \code{zlim}, in which case \code{breaks} is
inferred from that value.
\if{html}{The figure below explains the
(`breaks`, `col`) method of specifying a color mapping.  Note
that there must be one more break than color.  This is the method used by
e.g. [image()].}
\if{html}{\figure{colormap_fig_1.png}}
\item \strong{Case B.} Supply \code{x0}, \code{col0}, \code{x1}, and \code{col1}, but \emph{not}
\code{zlim}, \code{breaks}, \code{col} or \code{name}.
The \code{x0}, \code{col0}, \code{x1} and \code{col1} values specify a
value-color mapping that is similar to that used
for GMT color maps.  The method works by using \code{\link[=seq]{seq()}} to
interpolate between the elements of the \code{x0} vector.  The same is done
for \code{x1}.  Similarly, \code{\link[=colorRampPalette]{colorRampPalette()}} is used to
interpolate between the colors in the \code{col0} vector, and the same is
done for \code{col1}.  \if{html}{The figure above explains the (`x0`,
`x1`, `col0`, `col1`) method of specifying a color mapping.
Note that the each of the items has the same length. The case of
`blend=0`, which has color `col0[i]` between `x0[i]` and
`x1[i]`, is illustrated below.}
\if{html}{\figure{colormap_fig_2.png}}
\item \strong{Case C.} Supply \code{name} and possibly also \code{z}, but \emph{not}
\code{zlim}, \code{breaks}, \code{col}, \code{x0}, \code{col0}, \code{x1} or \code{col1}.
The \code{name} may be the name of a pre-defined color palette
(\code{"gmt_relief"}, \code{"gmt_ocean"}, \code{"gmt_globe"} or
\code{"gmt_gebco"}), or it may be the name of a file (or URL pointing to a file)
that contains a color map in the GMT format (see \dQuote{References}). If
\code{z} is supplied along with \code{name}, then \code{zcol} will be set up in the
return value, e.g. for use in colorizing points.  Another method
for finding colors for data points is to use the \code{colfunction()}
function in the return value.
}
}
\examples{
library(oce)
## Example 1. color scheme for points on xy plot
x <- seq(0, 1, length.out=40)
y <- sin(2 * pi * x)
par(mar=c(3, 3, 1, 1))
mar <- par('mar') # prevent margin creep by drawPalette()
## First, default breaks
c <- colormap(y)
drawPalette(c$zlim, col=c$col, breaks=c$breaks)
plot(x, y, bg=c$zcol, pch=21, cex=1)
grid()
par(mar=mar)
## Second, 100 breaks, yielding a smoother palette
c <- colormap(y, breaks=100)
drawPalette(c$zlim, col=c$col, breaks=c$breaks)
plot(x, y, bg=c$zcol, pch=21, cex=1)
grid()
par(mar=mar)

\dontrun{
## Example 2. topographic image with a standard color scheme
par(mfrow=c(1,1))
data(topoWorld)
cm <- colormap(name="gmt_globe")
imagep(topoWorld, breaks=cm$breaks, col=cm$col)

## Example 3. topographic image with modified colors,
## black for depths below 4km.
cm <- colormap(name="gmt_globe")
deep <- cm$x0 < -4000
cm$col0[deep] <- 'black'
cm$col1[deep] <- 'black'
cm <- colormap(x0=cm$x0, x1=cm$x1, col0=cm$col0, col1=cm$col1)
imagep(topoWorld, breaks=cm$breaks, col=cm$col)

## Example 4. image of world topography with water colorized
## smoothly from violet at 8km depth to blue
## at 4km depth, then blending in 0.5km increments
## to white at the coast, with tan for land.
cm <- colormap(x0=c(-8000, -4000,   0,  100),
               x1=c(-4000,     0, 100, 5000),
               col0=c("violet","blue","white","tan"),
               col1=c("blue","white","tan","yellow"))
lon <- topoWorld[['longitude']]
lat <- topoWorld[['latitude']]
z <- topoWorld[['z']]
imagep(lon, lat, z, breaks=cm$breaks, col=cm$col)
contour(lon, lat, z, levels=0, add=TRUE)

## Example 5. visualize GMT style color map
cm <- colormap(name="gmt_globe", debug=4)
plot(seq_along(cm$x0), cm$x0, pch=21, bg=cm$col0)
grid()
points(seq_along(cm$x1), cm$x1, pch=21, bg=cm$col1)

## Example 6. colfunction
cm <- colormap(c(0, 1))
x <- 1:10
y <- (x - 5.5)^2
z <- seq(0, 1, length.out=length(x))
drawPalette(colormap=cm)
plot(x, y, pch=21, bg=cm$colfunction(z), cex=3)
}

}
\references{
The following references provide information on choosing
colour schemes, that are suitable for viewers who have
colour deficiencies.

Light, Adam, and Patrick J. Bartlein. "The End of the Rainbow? Color
Schemes for Improved Data Graphics." \emph{Eos, Transactions American Geophysical
Union} 85, no. 40 (2004): 385.
DOI: 10.1029/2004EO400002

Stephenson, David B. "Comment on 'Color Schemes for Improved Data
Graphics', by A Light and P.J. Bartlein." \emph{Eos, Transactions American
Geophysical Union} 86, no.  20 (2005): 196.
DOI: 10.1029/2005EO200005

Light, Adam, and Patrick J. Bartlein. "Reply to 'Comment on Color
Schemes for Improved Data Graphics,' by A. Light and P.J.  Bartlein'."
\emph{Eos, Transactions American Geophysical Union} 86, no. 20 (2005): 196‚Äì196.
DOI: 10.1029/2005EO200006
}
\seealso{
Other things related to colors: 
\code{\link{colormapGMT}()},
\code{\link{oceColors9B}()},
\code{\link{oceColorsCDOM}()},
\code{\link{oceColorsChlorophyll}()},
\code{\link{oceColorsClosure}()},
\code{\link{oceColorsDensity}()},
\code{\link{oceColorsFreesurface}()},
\code{\link{oceColorsGebco}()},
\code{\link{oceColorsJet}()},
\code{\link{oceColorsOxygen}()},
\code{\link{oceColorsPAR}()},
\code{\link{oceColorsPalette}()},
\code{\link{oceColorsPhase}()},
\code{\link{oceColorsSalinity}()},
\code{\link{oceColorsTemperature}()},
\code{\link{oceColorsTurbidity}()},
\code{\link{oceColorsTurbo}()},
\code{\link{oceColorsTwo}()},
\code{\link{oceColorsVelocity}()},
\code{\link{oceColorsViridis}()},
\code{\link{oceColorsVorticity}()},
\code{\link{ocecolors}}
}
\author{
Dan Kelley
}
\concept{things related to colors}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/AllClass.R
\name{concatenate,oce-method}
\alias{concatenate,oce-method}
\title{Concatenate oce objects}
\usage{
\S4method{concatenate}{oce}(object, ...)
}
\arguments{
\item{object}{An object of \linkS4class{oce}, or a list containing such
objects (in which case the remaining arguments are ignored).}

\item{...}{Optional additional objects of \linkS4class{oce}.}
}
\value{
An object of \linkS4class{oce}.
}
\description{
This function concatenates oce objects. It is intended
for objects holding data sampled through time, and it works by pasting
together data linearly if they are vectors, by row if they are matrices,
and by second index if they are arrays. It has been tested for the following
classes:
\linkS4class{adp},
\linkS4class{adv},
\linkS4class{ctd},
and
\linkS4class{met}. It may do useful things for other classes, and
so users are encouraged to try, and to report problems to the developers.
It is unlikely that the function will do anything even remotely useful
for image and topographic data, to name just two cases that do not fit
the sampled-over-time category.
}
\examples{
## 1. Split, then recombine, a ctd object.
data(ctd)
ctd1 <- subset(ctd, scan <= median(ctd[["scan"]]))
ctd2 <- subset(ctd, scan > median(ctd[["scan"]]))
CTD <- concatenate(ctd1, ctd2)

## 2. Split, then recombine, an adp object.
data(adp)
midtime <- median(adp[["time"]])
adp1 <- subset(adp, time <= midtime)
adp2 <- subset(adp, time > midtime)
ADP <- concatenate(adp1, adp2)

\dontrun{
## 3. Download two met files and combine them.
met1 <- read.met(download.met(id=6358, year=2003, month=8))
met2 <- read.met(download.met(id=6358, year=2003, month=9))
MET <- concatenate(met1, met2)
}

}
\seealso{
Other functions that concatenate oce objects: 
\code{\link{concatenate,adp-method}},
\code{\link{concatenate,list-method}},
\code{\link{concatenate}()}
}
\author{
Dan Kelley
}
\concept{functions that concatenate oce objects}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/AllClass.R
\name{defaultFlags}
\alias{defaultFlags}
\title{Suggest a default flag vector for bad or suspicious data}
\usage{
defaultFlags(object)
}
\arguments{
\item{object}{An oce object}
}
\value{
A vector of one or more flag values, or \code{NULL} if \code{object}
\code{metadata} slot lacks a \code{flagScheme} as set by \code{\link[=initializeFlagScheme]{initializeFlagScheme()}},
or if it has a scheme that is not in the list provide in \dQuote{Description}.
}
\description{
\code{defaultFlags} tries to suggest a reasonable default \code{flag} scheme
for use by \code{\link[=handleFlags]{handleFlags()}}. It does this by looking for an item
named \code{flagScheme} in the \code{metadata} slot of \code{object}.
If \code{flagScheme} is found, and if the scheme is recognized, then a numeric
vector is returned that indicates bad or questionable data. If
\code{flagScheme$default} exists, then that scheme is returned. However,
if that does not exist, and if \code{flagScheme$name} is recognized,
then a pre-defined (very conservative) scheme is used,
as listed below.
}
\details{
\itemize{
\item for \code{argo}, the default is
\code{c(0,3,4,6,7,9)}, meaning to act upon \code{not_assessed} (0), \code{probably_bad} (3),
\code{bad} (4), \code{not_used_6} (6), \code{not_used_7} (7) and \code{missing} (9).  See Section
3.2.2 of Carval et al. (2019).
\item for \code{BODC}, the default is
\code{c(0,2,3,4,5,6,7,8,9)}, i.e. all flags except \code{good}.
\item for \code{DFO}, the default is
\code{c(0,2,3,4,5,8,9)}, i.e. all flags except \code{appears_correct}.
\item for \verb{WHP bottle}, the default is
\code{c(1,3,4,5,6,7,8,9)}, i.e. all flags except \code{no_problems_noted}.
\item for \verb{WHP ctd}, the default is
\code{c(1,3,4,5,6,7,9)}, i.e. all flags except \code{acceptable}.
}
}
\references{
\itemize{
\item Carval, Thierry, Bob Keeley, Yasushi Takatsuki, Takashi Yoshida, Stephen Loch Loch,
Claudia Schmid, and Roger Goldsmith. Argo User's Manual V3.3. Ifremer, 2019.
\doi{10.13155/29825}
}
}
\seealso{
Other functions relating to data-quality flags: 
\code{\link{handleFlags,adp-method}},
\code{\link{handleFlags,argo-method}},
\code{\link{handleFlags,ctd-method}},
\code{\link{handleFlags,oce-method}},
\code{\link{handleFlags,section-method}},
\code{\link{handleFlags}()},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{initializeFlagScheme,oce-method}},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{initializeFlagSchemeInternal}()},
\code{\link{initializeFlagScheme}()},
\code{\link{initializeFlags,adp-method}},
\code{\link{initializeFlags,oce-method}},
\code{\link{initializeFlagsInternal}()},
\code{\link{initializeFlags}()},
\code{\link{setFlags,adp-method}},
\code{\link{setFlags,ctd-method}},
\code{\link{setFlags,oce-method}},
\code{\link{setFlags}()}
}
\concept{functions relating to data-quality flags}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/map.R
\name{mapPoints}
\alias{mapPoints}
\title{Add Points to a Map}
\usage{
mapPoints(longitude, latitude, debug = getOption("oceDebug"), ...)
}
\arguments{
\item{longitude}{Longitudes of points to be plotted, or an object from which
longitude and latitude can be inferred in which case the following two
arguments are ignored.  This objects that are possible include those of type
\code{coastline}.}

\item{latitude}{numeric vector of latitudes of points to be plotted.}

\item{debug}{A flag that turns on debugging.  Set to 1 to get a moderate amount
of debugging information, or to 2 to get more.}

\item{...}{Optional arguments passed to \code{\link[=points]{points()}}.}
}
\description{
Plot points on an existing map, by analogy to \code{\link[=points]{points()}}.
}
\examples{
\donttest{
library(oce)
data(coastlineWorld)
mapPlot(coastlineWorld, longitudelim=c(-80, 0), latitudelim=c(20, 50),
        col="lightgray", projection="+proj=laea +lon_0=-35")
data(section)
mapPoints(section)
}

}
\seealso{
A map must first have been created with \code{\link[=mapPlot]{mapPlot()}}.

A map must first have been created with \code{\link[=mapPlot]{mapPlot()}}.

Other functions related to maps: 
\code{\link{formatPosition}()},
\code{\link{lonlat2map}()},
\code{\link{lonlat2utm}()},
\code{\link{map2lonlat}()},
\code{\link{mapArrows}()},
\code{\link{mapAxis}()},
\code{\link{mapContour}()},
\code{\link{mapCoordinateSystem}()},
\code{\link{mapDirectionField}()},
\code{\link{mapGrid}()},
\code{\link{mapImage}()},
\code{\link{mapLines}()},
\code{\link{mapLocator}()},
\code{\link{mapLongitudeLatitudeXY}()},
\code{\link{mapPlot}()},
\code{\link{mapPolygon}()},
\code{\link{mapScalebar}()},
\code{\link{mapText}()},
\code{\link{mapTissot}()},
\code{\link{oceCRS}()},
\code{\link{shiftLongitude}()},
\code{\link{usrLonLat}()},
\code{\link{utm2lonlat}()}
}
\author{
Dan Kelley
}
\concept{functions related to maps}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{oceColorsPalette}
\alias{oceColorsPalette}
\alias{oce.colorsPalette}
\title{Create a vector of colors}
\usage{
oceColorsPalette(n, which = 1)
}
\arguments{
\item{n}{number of colors to create}

\item{which}{integer or character string indicating the palette
to use; see \dQuote{Details}.}
}
\description{
The available schemes are:
\itemize{
\item \code{which=1} for a red-white-blue scheme.
\item \code{which=2} for a red-yellow-blue scheme.
\item \code{which=9.01}, \code{which="9A"} or \code{which="jet"} for \code{\link{oceColorsJet}}\code{(n)}.
\item \code{which=9.02} or \code{which="9B"} for \code{\link{oceColors9B}}\code{(n)}.
}
}
\references{
The following references provide information on choosing
colour schemes, that are suitable for viewers who have
colour deficiencies.

Light, Adam, and Patrick J. Bartlein. "The End of the Rainbow? Color
Schemes for Improved Data Graphics." \emph{Eos, Transactions American Geophysical
Union} 85, no. 40 (2004): 385.
DOI: 10.1029/2004EO400002

Stephenson, David B. "Comment on 'Color Schemes for Improved Data
Graphics', by A Light and P.J. Bartlein." \emph{Eos, Transactions American
Geophysical Union} 86, no.  20 (2005): 196.
DOI: 10.1029/2005EO200005

Light, Adam, and Patrick J. Bartlein. "Reply to 'Comment on Color
Schemes for Improved Data Graphics,' by A. Light and P.J.  Bartlein'."
\emph{Eos, Transactions American Geophysical Union} 86, no. 20 (2005): 196‚Äì196.
DOI: 10.1029/2005EO200006
}
\seealso{
Other things related to colors: 
\code{\link{colormapGMT}()},
\code{\link{colormap}()},
\code{\link{oceColors9B}()},
\code{\link{oceColorsCDOM}()},
\code{\link{oceColorsChlorophyll}()},
\code{\link{oceColorsClosure}()},
\code{\link{oceColorsDensity}()},
\code{\link{oceColorsFreesurface}()},
\code{\link{oceColorsGebco}()},
\code{\link{oceColorsJet}()},
\code{\link{oceColorsOxygen}()},
\code{\link{oceColorsPAR}()},
\code{\link{oceColorsPhase}()},
\code{\link{oceColorsSalinity}()},
\code{\link{oceColorsTemperature}()},
\code{\link{oceColorsTurbidity}()},
\code{\link{oceColorsTurbo}()},
\code{\link{oceColorsTwo}()},
\code{\link{oceColorsVelocity}()},
\code{\link{oceColorsViridis}()},
\code{\link{oceColorsVorticity}()},
\code{\link{ocecolors}}
}
\concept{things related to colors}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sealevel.R
\name{[[<-,sealevel-method}
\alias{[[<-,sealevel-method}
\title{Replace Parts of a Sealevel Object}
\usage{
\S4method{[[}{sealevel}(x, i, j, ...) <- value
}
\arguments{
\item{x}{a \linkS4class{sealevel} object.}

\item{i}{character value naming the item to replace.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{optional additional information (ignored).}

\item{value}{The value to be placed into \code{x}, somewhere.}
}
\description{
The \code{[[<-} method works for all \linkS4class{oce} objects.
The purpose, as with the related extraction method, \code{[[},
is to insulate users from the internal details of \linkS4class{oce}
objects, by looking for items within the various storage
slots of the object. Items not actually stored can also be
replaced, including units and data-quality
flags.
}
\details{
As with \code{[[} method, the procedure works in steps.

First, the \code{metadata} slot of \code{x} is checked to
see whether it contains something named with \code{i}.
If so, then the named item is replaced with \code{value}.

Otherwise, if the string value of \code{i} ends in \code{Unit}, then the
characters preceding that are taken as the name of a variable, and
the \code{metadata} slot of \code{x} is updated to store that unit, e.g.\preformatted{x[["temperatureUnits"]] <- list(unit=expression(degree*F),scale="")
}

Similarly, if \code{i} ends in \code{Flag}, then quality-control
flags are set up as defined by \code{result}, e.g.\preformatted{o[["temperatureFlags"]] <- c(2,4,2,2)
}

Otherwise, \code{\link[=pmatch]{pmatch()}} is used for a partial-string match with
the names of the items that are in the \code{data} slot of \code{x}.
The first item found (if any) is then updated to hold the value \code{result}.

If none of these conditions is met, a warning is issued.
}
\seealso{
Other functions that replace parts of oce objects: 
\code{\link{[[<-,adp-method}},
\code{\link{[[<-,amsr-method}},
\code{\link{[[<-,argo-method}},
\code{\link{[[<-,bremen-method}},
\code{\link{[[<-,cm-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{[[<-,echosounder-method}},
\code{\link{[[<-,g1sst-method}},
\code{\link{[[<-,gps-method}},
\code{\link{[[<-,ladp-method}},
\code{\link{[[<-,landsat-method}},
\code{\link{[[<-,lisst-method}},
\code{\link{[[<-,lobo-method}},
\code{\link{[[<-,met-method}},
\code{\link{[[<-,oce-method}},
\code{\link{[[<-,odf-method}},
\code{\link{[[<-,rsk-method}},
\code{\link{[[<-,section-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{[[<-,topo-method}},
\code{\link{[[<-,windrose-method}},
\code{\link{[[<-,xbt-method}}

Other things related to sealevel data: 
\code{\link{[[,sealevel-method}},
\code{\link{as.sealevel}()},
\code{\link{plot,sealevel-method}},
\code{\link{read.sealevel}()},
\code{\link{sealevel-class}},
\code{\link{sealevelTuktoyaktuk}},
\code{\link{sealevel}},
\code{\link{subset,sealevel-method}},
\code{\link{summary,sealevel-method}}
}
\concept{functions that replace parts of oce objects}
\concept{things related to sealevel data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/map.R
\name{mapGrid}
\alias{mapGrid}
\title{Add a Longitude and Latitude Grid to an Existing Map}
\usage{
mapGrid(
  dlongitude = 15,
  dlatitude = 15,
  longitude,
  latitude,
  col = "darkgray",
  lty = "solid",
  lwd = 0.5 * par("lwd"),
  polarCircle = 0,
  longitudelim,
  latitudelim,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{dlongitude}{increment in longitude, ignored if \code{longitude}
is supplied, but otherwise determines the longitude sequence.}

\item{dlatitude}{increment in latitude, ignored if \code{latitude}
is supplied, but otherwise determines the latitude sequence.}

\item{longitude}{numeric vector of longitudes, or \code{NULL} to prevent drawing
longitude lines.}

\item{latitude}{numeric vector of latitudes, or \code{NULL} to prevent drawing
latitude lines.}

\item{col}{color of lines}

\item{lty}{line type}

\item{lwd}{line width}

\item{polarCircle}{a number indicating the number of degrees of latitude
extending from the poles, within which zones are not drawn.}

\item{longitudelim}{optional argument specifying suggested longitude limits
for the grid. If this is not supplied, grid lines are drawn for the
whole globe, which can yield excessively slow drawing speeds for
small-region plots. This, and \code{latitudelim}, are both set by
\code{\link[=mapPlot]{mapPlot()}} if the arguments of the same name are passed to
that function.}

\item{latitudelim}{similar to \code{longitudelim}.}

\item{debug}{a flag that turns on debugging.  Set to 1 to get a moderate
amount of debugging information, 2 to go two function levels deep, or
3 to go all the way to the core functions. Any value above 3 will be
truncated to 3.}
}
\value{
A \link{data.frame}, returned silently, containing
\code{"side"}, \code{"value"}, \code{"type"}, and \code{"at"}.
A default call to \code{\link[=mapPlot]{mapPlot()}} ensures agreement of grid and axes by using
this return value in a call to \code{\link[=mapAxis]{mapAxis()}}.
}
\description{
Plot longitude and latitude grid on an existing map.
This is an advanced function, requiring
coordination with \code{\link[=mapPlot]{mapPlot()}} and (possibly) also with \code{\link[=mapAxis]{mapAxis()}},
and so it is best avoided by novices, who may be satisfied
with the defaults used by \code{\link[=mapPlot]{mapPlot()}}.
}
\details{
This is somewhat analogous to \code{\link[=grid]{grid()}}, except that the
first two arguments of the latter supply the number of lines in the grid,
whereas the present function has increments for the first two arguments.
}
\examples{
\donttest{
if (utils::packageVersion("sf") != "0.9.8") {
    # sf version 0.9-8 has a problem with this projection
    library(oce)
    data(coastlineWorld)
    par(mar=c(2, 2, 1, 1))
    # In mapPlot() call, note axes and grid args, to
    # prevent over-plotting of defaults.
    mapPlot(coastlineWorld, type="l", projection="+proj=ortho",
            axes=FALSE, grid=FALSE)
    mapGrid(15, 15)
}
}

}
\seealso{
A map must first have been created with \code{\link[=mapPlot]{mapPlot()}}.

Other functions related to maps: 
\code{\link{formatPosition}()},
\code{\link{lonlat2map}()},
\code{\link{lonlat2utm}()},
\code{\link{map2lonlat}()},
\code{\link{mapArrows}()},
\code{\link{mapAxis}()},
\code{\link{mapContour}()},
\code{\link{mapCoordinateSystem}()},
\code{\link{mapDirectionField}()},
\code{\link{mapImage}()},
\code{\link{mapLines}()},
\code{\link{mapLocator}()},
\code{\link{mapLongitudeLatitudeXY}()},
\code{\link{mapPlot}()},
\code{\link{mapPoints}()},
\code{\link{mapPolygon}()},
\code{\link{mapScalebar}()},
\code{\link{mapText}()},
\code{\link{mapTissot}()},
\code{\link{oceCRS}()},
\code{\link{shiftLongitude}()},
\code{\link{usrLonLat}()},
\code{\link{utm2lonlat}()}
}
\author{
Dan Kelley
}
\concept{functions related to maps}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/met.R
\name{download.met}
\alias{download.met}
\title{Download and Cache a met File}
\usage{
download.met(
  id,
  year,
  month,
  deltat,
  type = "xml",
  destdir = ".",
  destfile,
  force = FALSE,
  quiet = FALSE,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{id}{A number giving the "Station ID" of the station of interest. If not
provided, \code{id} defaults to 6358, for Halifax International Airport. See
\dQuote{Details}.}

\item{year}{A number giving the year of interest. Ignored unless \code{deltat}
is \code{"hour"}. If \code{year} is not given, it defaults to the present year.}

\item{month}{A number giving the month of interest. Ignored unless \code{deltat}
is \code{"hour"}. If \code{month} is not given, it defaults to the present
month.}

\item{deltat}{Optional character string indicating the time step of the
desired dataset. This may be \code{"hour"} or \code{"month"}.
If \code{deltat} is not given, it defaults to \code{"hour"}.}

\item{type}{String indicating which type of file to download, either
\code{"xml"} (the default) for an XML file or \code{"csv"} for a CSV file.}

\item{destdir}{Optional string indicating the directory in which to store
downloaded files. If not supplied, \code{"."} is used, i.e. the data file
is stored in the present working directory.}

\item{destfile}{Optional string indicating the name of the file.
If not supplied, the file name is constructed from the other
parameters of the function call, so subsequent calls with the same
parameters will yield the same result, thus providing the key
to the caching scheme.}

\item{force}{Logical value indicating whether to force a download, even
if the file already exists locally.}

\item{quiet}{Logical value passed to \code{\link[=download.file]{download.file()}}; a \code{TRUE} value
silences output.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\value{
String indicating the full pathname to the downloaded file.
}
\description{
Data are downloaded from the Environment Canada's historical data
website and cached locally.
}
\details{
The data are downloaded
using \code{\link[utils:download.file]{utils::download.file()}} based on a query devised by reverse-engineering
queries create by the Environment Canada interface
\verb{https://climate.weather.gc.ca/historical_data/search_historic_data_e.html}.
The constructed query contains Station ID, as provided in the \code{id} argument.
Note that this seems to be a creation of Environment Canada, alone;
it is distinct from the more standard "Climate ID" and "WMO ID".
To make things more difficult, Environment Canada states that the
Station ID is subject to change over time. (Whether this applies to existing
data is unclear.)

Given these difficulties with Station ID, users are advised to consult
the Environment Canada website (reference 1) before downloading any data,
and to check it from time to time
during the course of a research project, to see if the Station ID has changed.
Another approach would be to use Gavin Simpson's
\code{canadaHCD} package (reference 2) to look up Station IDs. This package maintains
a copy of the Environment Canada listing of stations, and its
\code{find_station} function provides an easy way to determine Station IDs.
After that, its \code{hcd_hourly} function (and related functions) make
it easy to read data. These data can then be converted to the
\code{met} class with \code{\link[=as.met]{as.met()}}, although doing so leaves
many important metadata blank.
}
\examples{
\dontrun{
library(oce)
## Download data for Halifax International Airport, in September
## of 2003. (This dataset is used for data(met) provided with oce.)
metFile <- download.met(6358, 2003, 9, destdir=".")
met <- read.met(metFile)
}

}
\references{
\enumerate{
\item Environment Canada website for Historical Climate Data
\verb{https://climate.weather.gc.ca/index_e.html}
\item Gavin Simpson's \code{canadaHCD} package on GitHub
\url{https://github.com/gavinsimpson/canadaHCD}
}
}
\seealso{
The work is done with \code{\link[utils:download.file]{utils::download.file()}}.

Other functions that download files: 
\code{\link{download.amsr}()},
\code{\link{download.coastline}()},
\code{\link{download.topo}()}

Other things related to met data: 
\code{\link{[[,met-method}},
\code{\link{[[<-,met-method}},
\code{\link{as.met}()},
\code{\link{met-class}},
\code{\link{met}},
\code{\link{plot,met-method}},
\code{\link{read.met}()},
\code{\link{subset,met-method}},
\code{\link{summary,met-method}},
\code{\link{test_met_csv1.csv}},
\code{\link{test_met_csv2.csv}},
\code{\link{test_met_xml2.xml}}
}
\author{
Dan Kelley
}
\concept{functions that download files}
\concept{things related to met data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adv.R
\name{enuToOtherAdv}
\alias{enuToOtherAdv}
\title{Convert ENU to Other Coordinate}
\usage{
enuToOtherAdv(
  x,
  heading = 0,
  pitch = 0,
  roll = 0,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{x}{an \linkS4class{adv} object.}

\item{heading}{number or vector of numbers, giving the angle, in degrees, to
be added to the heading. If this has length less than the number of velocity
sampling times, then it will be extended using \code{\link[=rep]{rep()}}.}

\item{pitch}{as \code{heading} but for pitch.}

\item{roll}{as \code{heading} but for roll.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\description{
Convert ADV velocity components from an enu-based coordinate system to
another system, perhaps to align axes with the coastline.
}
\details{
The supplied angles specify rotations to be made around the axes for which
heading, pitch, and roll are defined.  For example, an eastward current will
point southeast if \code{heading=45} is used.

The returned value has heading, pitch, and roll matching those of \code{x},
so these angles retain their meaning as the instrument orientation.

NOTE: this function works similarly to \code{\link[=xyzToEnuAdv]{xyzToEnuAdv()}}, except
that in the present function, it makes no difference whether the instrument
points up or down, etc.
}
\seealso{
Other things related to adv data: 
\code{\link{[[,adv-method}},
\code{\link{[[<-,adv-method}},
\code{\link{adv-class}},
\code{\link{adv}},
\code{\link{beamName}()},
\code{\link{beamToXyz}()},
\code{\link{enuToOther}()},
\code{\link{plot,adv-method}},
\code{\link{read.adv.nortek}()},
\code{\link{read.adv.sontek.adr}()},
\code{\link{read.adv.sontek.serial}()},
\code{\link{read.adv.sontek.text}()},
\code{\link{read.adv}()},
\code{\link{rotateAboutZ}()},
\code{\link{subset,adv-method}},
\code{\link{summary,adv-method}},
\code{\link{toEnuAdv}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdv}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley
}
\concept{things related to adv data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/argo.R
\name{as.argo}
\alias{as.argo}
\title{Coerce Data Into an Argo Dataset}
\usage{
as.argo(
  time,
  longitude,
  latitude,
  salinity,
  temperature,
  pressure,
  units = NULL,
  id,
  filename = "",
  missingValue
)
}
\arguments{
\item{time}{a vector of POSIXct times.}

\item{longitude}{a vector of longitudes.}

\item{latitude}{a vector of latitudes.}

\item{salinity}{a vector of salinities.}

\item{temperature}{a vector of temperatures.}

\item{pressure}{a vector of pressures.}

\item{units}{an optional list containing units. If \code{NULL}, the default,
then \code{"degree east"} is used for \code{longitude},
\code{"degree north"} for \code{latitude},
\code{""} for \code{salinity},
\code{"ITS-90"} for \code{temperature}, and
\code{"dbar"} for \code{pressure}.}

\item{id}{an identifier for the argo float, typically a number, but stored within
the object in a character form. (For example, the dataset retrieved with \code{data(argo)}
has an \code{id} of \code{"6900388"}.)}

\item{filename}{a source filename, which defaults to an empty string.}

\item{missingValue}{an optional missing value, indicating data values that should be
taken as \code{NA}.}
}
\value{
An \linkS4class{argo} object.
}
\description{
Coerce a dataset into an argo dataset. This is not the right way to
read official argo datasets, which are provided in NetCDF format and may
be read with \code{\link[=read.argo]{read.argo()}}.
}
\seealso{
The documentation for the \linkS4class{argo} class explains the structure of argo
objects, and also outlines the other functions dealing with them.

Other things related to argo data: 
\code{\link{[[,argo-method}},
\code{\link{[[<-,argo-method}},
\code{\link{argo-class}},
\code{\link{argoGrid}()},
\code{\link{argoNames2oceNames}()},
\code{\link{argo}},
\code{\link{handleFlags,argo-method}},
\code{\link{plot,argo-method}},
\code{\link{read.argo}()},
\code{\link{subset,argo-method}},
\code{\link{summary,argo-method}}
}
\author{
Dan Kelley
}
\concept{things related to argo data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.R
\name{ctdFindProfiles}
\alias{ctdFindProfiles}
\title{Find Profiles within a Tow-Yow CTD Record}
\usage{
ctdFindProfiles(
  x,
  cutoff = 0.5,
  minLength = 10,
  minHeight = 0.1 * diff(range(x[["pressure"]])),
  smoother = smooth.spline,
  direction = c("descending", "ascending"),
  breaks,
  arr.ind = FALSE,
  distinct,
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{x}{a \linkS4class{ctd} object.}

\item{cutoff}{criterion on pressure difference; see \dQuote{Details}.}

\item{minLength}{lower limit on number of points in candidate profiles.}

\item{minHeight}{lower limit on height of candidate profiles.}

\item{smoother}{The smoothing function to use for identifying down/up
casts. The default is \code{smooth.spline}, which performs well for
a small number of cycles; see \dQuote{Examples} for a method that is
better for a long tow-yo. The return value from \code{smoother} must
be either a list containing an element named \code{y} or something
that can be coerced to a vector with \code{\link[=as.vector]{as.vector()}}. To
turn smoothing off, so that cycles in pressure are determined by
simple first difference, set \code{smoother} to \code{NULL}.}

\item{direction}{String indicating the travel direction to be selected.}

\item{breaks}{optional integer vector indicating the indices of last
datum in each profile stored within \code{x}. Thus, the first profile
in the return value will contain the \code{x} data from indices 1
to \code{breaks[1]}.  If \code{breaks} is given, then all
other arguments except \code{x} are ignored. Using \code{breaks}
is handy in cases where other schemes fail, or when the author
has independent knowledge of how the profiles are strung together
in \code{x}.}

\item{arr.ind}{Logical indicating whether the array indices should be returned;
the alternative is to return a vector of ctd objects.}

\item{distinct}{An optional string indicating how to identify profiles
by unique values. Use \code{"location"}
to find profiles by a change in longitude and latitude, or use the name of any
of item in the \code{data} slot in \code{x}. In these cases, all the
other arguments except \code{x} are ignored. However, if \code{distinct}
is not supplied, the other arguments are handled as described above.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}

\item{...}{Optional extra arguments that are passed to the smoothing function, \code{smoother}.}
}
\value{
If \code{arr.ind=TRUE}, a data frame with columns \code{start} and \code{end}, the indices
of the downcasts.  Otherwise, a vector of \code{ctd} objects. In this second case,
the station names are set to a form like \code{"10/3"}, for the third profile within an
original ctd object with station name \code{"10"}, or to \code{"3"}, if the original
ctd object had no station name defined.
}
\description{
Examine the pressure record looking for extended periods of either ascent or descent, and return
either indices to these events or a vector of CTD records containing the events.
}
\details{
The method works by examining the pressure record.  First, this is smoothed using
\code{smoother()} (see \dQuote{Arguments}), and then the result is first-differenced
using \code{\link[=diff]{diff()}}.  Median values of the positive and
negative first-difference values are then multiplied by \code{cutoff}.  This establishes criteria
for any given point to be in an ascending profile, a descending profile, or a non-profile.
Contiguous regions are then found, and those that have fewer than \code{minLength} points are
discarded.  Then, those that have pressure ranges less than \code{minHeight} are discarded.

Caution: this method is not well-suited to all datasets. For example, the default
value of \code{smoother} is \code{\link[=smooth.spline]{smooth.spline()}}, and this works well for just a few
profiles, but poorly for a tow-yo with a long sequence of profiles; in the latter case,
it can be preferable to use simpler smoothers (see \dQuote{Examples}). Also, depending
on the sampling protocol, it is often necessary to pass the resultant profiles through
\code{\link[=ctdTrim]{ctdTrim()}}, to remove artifacts such as an equilibration phase, etc.
Generally, one is well-advised to use the present function for a quick look at the data,
relying on e.g. \code{\link[=plotScan]{plotScan()}} to identify profiles visually, for a final product.
}
\examples{
library(oce)

\dontrun{
# Example 1. Find profiles within a towyo file, as can result
# if the CTD is cycled within the water column as the ship
# moves.
profiles <- ctdFindProfiles(towyo)
}

\dontrun{
# Example 2. Use a moving average to smooth pressure, instead of the
# default smooth.spline() method. This might avoid a tendency of
# the default scheme to miss some profiles in a long towyo.
movingAverage <- function(x, n = 11, ...)
{
    f <- rep(1/n, n)
    stats::filter(x, f, ...)
}
casts <- ctdFindProfiles(towyo, smoother=movingAverage)
}

\dontrun{
# Example 3: glider data read into a ctd object. Chop
# into profiles by looking for pressure jumps exceeding
# 10 dbar.
breaks <- which(diff(gliderAsCtd[["pressure"]]) > 10)
profiles <- ctdFindProfiles(gliderAsCtd, breaks=breaks)
}

}
\seealso{
The documentation for \linkS4class{ctd} explains the structure
of CTD objects, and also outlines the other functions dealing with them.

Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}
}
\author{
Dan Kelley and Clark Richards
}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/air.R
\name{airRho}
\alias{airRho}
\title{Air density}
\usage{
airRho(temperature, pressure, humidity)
}
\arguments{
\item{temperature}{\emph{in-situ} temperature, in \eqn{^\circ}{deg}C.}

\item{pressure}{numeric value for pressure in Pa (\emph{not} the kPa used in public
weather forecasts).}

\item{humidity}{ignored at present}
}
\value{
\emph{In-situ} dry-air density, in kg/m\eqn{^3}{^3}.
}
\description{
Compute \eqn{\rho}{rho}, the \emph{in-situ} density of dry air.
}
\details{
This will eventually be a proper equation of state, but for now it just
uses a dry-air formula posted on wikipedia (i.e. not trustworthy).
}
\examples{
degC <- seq(0,30,length.out=100)
p <- seq(98,102,length.out=100) * 1e3
contour(x=degC, y=p, z=outer(degC,p,airRho), labcex=1)
}
\references{
\enumerate{
\item \url{https://en.wikipedia.org/wiki/Density_of_air}
\item National Oceanographic and Atmospheric Agency, 1976.  U.S.
Standard Atmosphere, 1976.  NOAA-S/T 76-1562.  (A PDF of this document may
be available at
\verb{http://ntrs.nasa.gov/archive/nasa/casi.ntrs.nasa.gov/19770009539_1977009539.pdf}
or
\verb{http://www.dtic.mil/cgi-bin/GetTRDoc?Location=U2&doc=GetTRDoc.pdf&AD=ADA035728}
although neither link has proven to be reliable.)
}
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/windrose.R
\name{as.windrose}
\alias{as.windrose}
\title{Create a Windrose Object}
\usage{
as.windrose(x, y, dtheta = 15, debug = getOption("oceDebug"))
}
\arguments{
\item{x}{The x component of wind speed (or stress) \emph{or} an object of class
\code{met} (see \linkS4class{met}), in which case the \code{u} and
\code{v} components of that object are used for the components of wind speed,
and \code{y} here is ignored.}

\item{y}{The y component of wind speed (or stress).}

\item{dtheta}{The angle increment (in degrees) within which to classify the data.}

\item{debug}{A flag that turns on debugging.  Set to 1 to get a moderate amount
of debugging information, or to 2 to get more.}
}
\value{
A \linkS4class{windrose} object, with \code{data} slot containing

\tabular{ll}{
\strong{Item}   \tab \strong{Meaning}\cr
\code{n}       \tab the number of \code{x} values\cr
\code{x.mean}  \tab the mean of the \code{x} values\cr
\code{y.mean}  \tab the mean of the \code{y} values\cr
\code{theta}   \tab the central angle (in degrees) for the class\cr
\code{count}   \tab the number of observations in this class\cr
\code{mean}    \tab the mean of the observations in this class\cr
\code{fivenum} \tab the \code{\link[=fivenum]{fivenum()}} vector for
observations in this class (the min, the lower hinge, the
median, the upper hinge, and the max)\cr
}
}
\description{
Create a wind-rose object, typically for plotting with
\code{\link[=plot,windrose-method]{plot,windrose-method()}}.
}
\examples{
library(oce)
xcomp <- rnorm(360) + 1
ycomp <- rnorm(360)
wr <- as.windrose(xcomp, ycomp)
summary(wr)
plot(wr)

}
\seealso{
Other things related to windrose data: 
\code{\link{[[,windrose-method}},
\code{\link{[[<-,windrose-method}},
\code{\link{plot,windrose-method}},
\code{\link{summary,windrose-method}},
\code{\link{windrose-class}}
}
\author{
Dan Kelley, with considerable help from Alex Deckmyn.
}
\concept{things related to windrose data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.R
\name{subset,adp-method}
\alias{subset,adp-method}
\title{Subset an ADP Object}
\usage{
\S4method{subset}{adp}(x, subset, ...)
}
\arguments{
\item{x}{an \linkS4class{adp} object.}

\item{subset}{A condition to be applied to the \code{data} portion of
\code{x}.  See \sQuote{Details}.}

\item{...}{Ignored.}
}
\value{
An \linkS4class{adp} object.
}
\description{
Subset an adp (acoustic Doppler profile) object, in a manner that is function
is somewhat analogous to \code{\link[=subset.data.frame]{subset.data.frame()}}.
}
\details{
For any data type,
subsetting can be by \code{time}, \code{ensembleNumber}, or \code{distance}.
These may not be combined, but it is easy to use a string of calls to
carry out combined operations, e.g.
\code{subset(subset(adp,distance<d0), time<t0)}

For the special
case of AD2CP data (see \code{\link[=read.adp.ad2cp]{read.adp.ad2cp()}}), it is possible to subset
to the "average" data records with \code{subset="average"}, to the
"burst" records with \code{subset="burst"}, or to the "interleavedBurst"
with \code{subset="interleavedBurst"}; note that no warning is issued,
if this leaves an object with no useful data.
}
\examples{
library(oce)
data(adp)
# 1. Look at first part of time series, organized by time
earlyTime <- subset(adp, time < mean(range(adp[['time']])))
plot(earlyTime)

# 2. Look at first ten ensembles (AKA profiles)
en <- adp[["ensembleNumber"]]
firstTen <- subset(adp, ensembleNumber < en[11])
plot(firstTen)

}
\seealso{
Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}

Other functions that subset oce objects: 
\code{\link{subset,adv-method}},
\code{\link{subset,amsr-method}},
\code{\link{subset,argo-method}},
\code{\link{subset,cm-method}},
\code{\link{subset,coastline-method}},
\code{\link{subset,ctd-method}},
\code{\link{subset,echosounder-method}},
\code{\link{subset,lobo-method}},
\code{\link{subset,met-method}},
\code{\link{subset,oce-method}},
\code{\link{subset,odf-method}},
\code{\link{subset,rsk-method}},
\code{\link{subset,sealevel-method}},
\code{\link{subset,section-method}},
\code{\link{subset,topo-method}},
\code{\link{subset,xbt-method}}
}
\author{
Dan Kelley
}
\concept{functions that subset oce objects}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{as.oce}
\alias{as.oce}
\title{Coerce Something Into an Oce Object}
\usage{
as.oce(x, ...)
}
\arguments{
\item{x}{an item containing data. This may be data frame, list, or an oce object.}

\item{\dots}{optional extra arguments, passed to conversion functions
\code{\link[=as.coastline]{as.coastline()}} or \code{\link[=ODF2oce]{ODF2oce()}}, if these are used.}
}
\value{
An \linkS4class{oce} object.
}
\description{
Coerce Something Into an Oce Object
}
\details{
This function is limited and not intended for common use.
In most circumstances, users should employ a function such
as \code{\link[=as.ctd]{as.ctd()}} to construct specialized oce sub-classes.

\code{as.oce} creates an oce object from data contained within its
first argument, which may be a list, a data frame, or an object
of \linkS4class{oce}.  (In the last case, \code{x} is
simply returned, without modification.)

If \code{x} is a list containing items named \code{longitude} and
\code{latitude}, then \code{\link[=as.coastline]{as.coastline()}} is called (with
the specified \dots value) to create a coastline object.

If \code{x} is a list created by \code{read_odf} from the (as
yet unreleased) ODF package developed by the Bedford Institute of
Oceanography, then \code{\link[=ODF2oce]{ODF2oce()}} is called (with
no arguments other than the first) to calculate a return value.
If the sub-class inference made by \code{\link[=ODF2oce]{ODF2oce()}} is
incorrect, users should call that function directly, specifying
a value for its \code{coerce} argument.

If \code{x} has not been created by \code{read_odf}, then the names
of the items it contains are examined, and used to try to infer
the proper return value.  There
are only a few cases (although more may be added if there is
sufficient user demand). The cases are as follows.
\itemize{
\item If \code{x} contains items named \code{temperature},
\code{pressure} and either \code{salinity} or \code{conductivity},
then an object of type \linkS4class{ctd} will be returned.
\item If \code{x} contains columns named \code{longitude} and \code{latitude},
but no other columns, then an object of class \linkS4class{coastline}
is returned.
}
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adv.R
\name{toEnuAdv}
\alias{toEnuAdv}
\title{Convert an ADV Object to ENU Coordinates}
\usage{
toEnuAdv(x, declination = 0, debug = getOption("oceDebug"))
}
\arguments{
\item{x}{an \linkS4class{adv} object.}

\item{declination}{magnetic declination to be added to the heading, to get
ENU with N as "true" north.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\description{
Convert an ADV Object to ENU Coordinates
}
\references{
\enumerate{
\item @template nortekCoordTemplate
}
}
\seealso{
See \code{\link[=read.adv]{read.adv()}} for notes on functions relating to
\code{"adv"} objects.  Also, see \code{\link[=beamToXyzAdv]{beamToXyzAdv()}} and
\code{\link[=xyzToEnuAdv]{xyzToEnuAdv()}}.

Other things related to adv data: 
\code{\link{[[,adv-method}},
\code{\link{[[<-,adv-method}},
\code{\link{adv-class}},
\code{\link{adv}},
\code{\link{beamName}()},
\code{\link{beamToXyz}()},
\code{\link{enuToOtherAdv}()},
\code{\link{enuToOther}()},
\code{\link{plot,adv-method}},
\code{\link{read.adv.nortek}()},
\code{\link{read.adv.sontek.adr}()},
\code{\link{read.adv.sontek.serial}()},
\code{\link{read.adv.sontek.text}()},
\code{\link{read.adv}()},
\code{\link{rotateAboutZ}()},
\code{\link{subset,adv-method}},
\code{\link{summary,adv-method}},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdv}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley
}
\concept{things related to adv data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/landsat.R
\name{read.landsat}
\alias{read.landsat}
\title{Read a landsat File Directory}
\usage{
read.landsat(
  file,
  band = "all",
  emissivity = 0.984,
  decimate,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{file}{A connection or a character string giving the name of the file to
load.  This is a directory name containing the data.}

\item{band}{The bands to be read, by default all of the bands.  Use
\code{band=NULL} to skip the reading of bands, instead reading only the
image metadata, which is often enough to check if the image is of
interest in a given study. See \sQuote{Details} for the names of the
bands, some of which are pseudo-bands, computed from the actual data.}

\item{emissivity}{Value of the emissivity of the surface, stored as
\code{emissivity} in the \code{metadata} slot of the
resultant object. This is used in the
calculation of surface temperature, as explained in the discussion of
accessor functions for \linkS4class{landsat}. The default value is
from Konda et al.  (1994). These authors suggest an uncertainty of 0.04,
but a wider range of values can be found in the literature.  The value of
\code{metadata$emissivity} is easy to alter, either as a single value or
as a matrix, yielding flexibility of calculation.}

\item{decimate}{optional positive integer indicating the degree to which
the data should be subsampled after reading and before storage. Setting
this to 10 can speed up reading by a factor of 3 or more, but higher values
have diminishing effect.  In exploratory work, it is useful to set
\code{decimate=10}, to plot the image to determine a subregion
of interest, and then to use \code{\link[=landsatTrim]{landsatTrim()}} to trim the image.}

\item{debug}{a flag that turns on debugging.  Set to 1 to get a moderate
amount of debugging information, or to 2 to get more.}
}
\value{
A \linkS4class{landsat} object, with the conventional Oce
slots \code{metadata}, \code{data} and \code{processingLog}.  The
\code{metadata} is mainly intended for use by Oce functions, but for generality
it also contains an entry named \code{header} that represents the full image
header in a list (with names made lower-case).  The \code{data} slot holds
matrices of the data in the requested bands, and users may add extra matrices
if desired, e.g. to store calculated quantities.
}
\description{
Read a landsat data file, producing an object of \linkS4class{landsat}.
The actual reading is done with \code{\link[tiff:readTIFF]{tiff::readTIFF()}} in the
\CRANpkg{tiff} package, so that package must be installed for
\code{read.landsat} to work.
}
\details{
Landsat data are provided in directories that contain TIFF files and header
information, and \code{read.landsat} relies on a strict convention for the
names of the files in those directories.  Those file names were found by
inspection of some data, on the assumption that similar patterns will hold for
other datasets for any given satellite. This is a brittle approach and it
should be born in mind if \code{read.landsat} fails for a given dataset.

For Landsat 8, there are 11 bands, with names \code{"aerosol"} (band 1),
\code{"blue"} (band 2), \code{"green"} (band 3), \code{"red"} (band 4),
\code{"nir"} (band 5), \code{"swir1"} (band 6), \code{"swir2"} (band 7),
\code{"panchromatic"} (band 8), \code{"cirrus"} (band 9), \code{"tirs1"} (band
10), and \code{"tirs2"} (band 11).
In addition to the above, setting \code{band="terralook"} may be used as
an abbreviation for \code{band=c("red", "green", "nir")}.

For Landsat 7, there 8 bands, with names \code{"blue"} (band 1), \code{"green"}
(band 2), \code{"red"} (band 3), \code{"nir"} (band 4), \code{"swir1"} (band
5), \code{"tir1"} (band 6A), \code{"tir2"} (band 6B), \code{"swir2"} (band 7)
and \code{"panchromatic"} (band 8).

For Landsat 4 and 5, the bands similar to Landsat 7 but without
\code{"panchromatic"} (band 8).
}
\section{Storage requirements}{


Landsat data files (directories, really) are large, accounting for
approximately 1 gigabyte each.  The storage of the Oce object is
similar (see \linkS4class{landsat}).  In R, many operations involving
copying data, so that dealing with full-scale landsat images can overwhelm
computers with storage under 8GB.  For this reason, it is typical to read just
the bands that are of interest.  It is also helpful to use
\code{\link[=landsatTrim]{landsatTrim()}} to trim the data to a geographical range, or
to use \code{\link[=decimate]{decimate()}} to get a coarse view of the domain, especially
early in an analysis.
}

\references{
\enumerate{
\item Konda, M. Imasato N., Nishi, K., and T. Toda, 1994.  Measurement of the Sea
Surface Emissivity.  \emph{Journal of Oceanography}, 50, 17:30.
\doi{10.1007/BF02233853}
}
}
\seealso{
See the documentation for the \linkS4class{landsat} class
for more information on \code{landsat} objects,
especially band information.  Use \code{\link[=landsatTrim]{landsatTrim()}} to trim Landsat
objects geographically and \code{\link[=landsatAdd]{landsatAdd()}} to add new ``bands.''  The
accessor operator (\code{[[}) is used to access band information, full or
decimated, and to access certain derived quantities.  A sample dataset named
\code{\link[=landsat]{landsat()}} is provided by the \CRANpkg{oce} package.

Other things related to landsat data: 
\code{\link{[[,landsat-method}},
\code{\link{[[<-,landsat-method}},
\code{\link{landsat-class}},
\code{\link{landsatAdd}()},
\code{\link{landsatTrim}()},
\code{\link{landsat}},
\code{\link{plot,landsat-method}},
\code{\link{summary,landsat-method}}
}
\author{
Dan Kelley
}
\concept{things related to landsat data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sealevel.R
\docType{class}
\name{sealevel-class}
\alias{sealevel-class}
\title{Class to Store Sealevel Data}
\description{
This class stores sealevel data, e.g. from a tide gauge.
}
\section{Slots}{

\describe{
\item{\code{data}}{As with all \code{oce} objects, the \code{data} slot
for \code{sealevel} objects is a
\link{list} containing the main data for the object.
The key items stored in this slot are \code{time} and \code{elevation}.}

\item{\code{metadata}}{As with all \code{oce} objects, the \code{metadata} slot
for \code{sealevel} objects is a \link{list} containing
information about the \code{data} or about the object itself.
An example of the former might be the location at which a \code{sealevel} measurement was made, stored in \code{longitude} and \code{latitude}, and of the latter might be \code{filename}, the name of the data source.}

\item{\code{processingLog}}{As with all \code{oce} objects, the \code{processingLog} slot
for \code{sealevel} objects is a
\link{list} with entries describing the creation and evolution
of the object. The contents are updated by various \code{oce} functions to
keep a record of processing steps.  Object summaries and
\code{\link[=processingLogShow]{processingLogShow()}} both display the log.}
}}

\section{Modifying slot contents}{


Although the \code{[[<-} operator may permit modification of the contents
of \linkS4class{sealevel} objects (see \code{\link{[[<-,sealevel-method}}),
it is better to use \code{\link[=oceSetData]{oceSetData()}} and \code{\link[=oceSetMetadata]{oceSetMetadata()}},
because those functions save an entry in the \code{processingLog}
that describes the change.
}

\section{Retrieving slot contents}{


The full contents of the \code{data} and \code{metadata} slots of a \linkS4class{sealevel}
object may be retrieved in the standard R way using \code{\link[=slot]{slot()}}. For
example \code{slot(o,"data")} returns the \code{data} slot of an object named \code{o},
and similarly \code{slot(o,"metadata")} returns
the \code{metadata} slot.

The slots may also be obtained with the \code{\link{[[,sealevel-method}}
operator, as e.g. \code{o[["data"]]} and \code{o[["metadata"]]}, respectively.

The \code{\link{[[,sealevel-method}} operator can also
be used to retrieve items from within the \code{data} and \code{metadata} slots.
For example, \code{o[["temperature"]]} can be used to retrieve temperature
from an object containing that quantity. The rule is that a named
quantity is sought first within the object's \code{metadata} slot,
with the \code{data} slot being checked only if \code{metadata} does not
contain the item. This \code{[[} method can also be used to get
certain derived quantities, if the object contains sufficient
information to calculate them. For example, an object that holds
(practical) salinity, temperature and pressure, along with
longitude and latitude, has sufficient information to compute
Absolute Salinity, and so \code{o[["SA"]]} will yield the
calculated Absolute Salinity.

It is also possible to find items more directly, using \code{\link[=oceGetData]{oceGetData()}} and
\code{\link[=oceGetMetadata]{oceGetMetadata()}}, but neither of these functions can
retrieve derived items.
}

\seealso{
Other classes provided by oce: 
\code{\link{adp-class}},
\code{\link{adv-class}},
\code{\link{argo-class}},
\code{\link{bremen-class}},
\code{\link{cm-class}},
\code{\link{coastline-class}},
\code{\link{ctd-class}},
\code{\link{lisst-class}},
\code{\link{lobo-class}},
\code{\link{met-class}},
\code{\link{oce-class}},
\code{\link{odf-class}},
\code{\link{rsk-class}},
\code{\link{section-class}},
\code{\link{topo-class}},
\code{\link{windrose-class}},
\code{\link{xbt-class}}

Other things related to sealevel data: 
\code{\link{[[,sealevel-method}},
\code{\link{[[<-,sealevel-method}},
\code{\link{as.sealevel}()},
\code{\link{plot,sealevel-method}},
\code{\link{read.sealevel}()},
\code{\link{sealevelTuktoyaktuk}},
\code{\link{sealevel}},
\code{\link{subset,sealevel-method}},
\code{\link{summary,sealevel-method}}
}
\author{
Dan Kelley
}
\concept{classes provided by oce}
\concept{things related to sealevel data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ad.R
\name{xyzToEnu}
\alias{xyzToEnu}
\title{Convert Acoustic-Doppler Data From xyz to enu Coordinates}
\usage{
xyzToEnu(x, ...)
}
\arguments{
\item{x}{an \linkS4class{adp} or \linkS4class{adv} object.}

\item{\dots}{extra arguments that are passed on to \code{\link[=xyzToEnuAdp]{xyzToEnuAdp()}}
or \code{\link[=xyzToEnuAdv]{xyzToEnuAdv()}}; see the documentation for those functions,
for the details.}
}
\value{
An object of the same class as \code{x}, but with velocities
in east-north-up coordinates instead of xyz coordinates.
}
\description{
Convert Acoustic-Doppler Data From xyz to enu Coordinates
}
\seealso{
Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()}

Other things related to adv data: 
\code{\link{[[,adv-method}},
\code{\link{[[<-,adv-method}},
\code{\link{adv-class}},
\code{\link{adv}},
\code{\link{beamName}()},
\code{\link{beamToXyz}()},
\code{\link{enuToOtherAdv}()},
\code{\link{enuToOther}()},
\code{\link{plot,adv-method}},
\code{\link{read.adv.nortek}()},
\code{\link{read.adv.sontek.adr}()},
\code{\link{read.adv.sontek.serial}()},
\code{\link{read.adv.sontek.text}()},
\code{\link{read.adv}()},
\code{\link{rotateAboutZ}()},
\code{\link{subset,adv-method}},
\code{\link{summary,adv-method}},
\code{\link{toEnuAdv}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdv}()}
}
\concept{things related to adp data}
\concept{things related to adv data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.R
\name{adpConvertRawToNumeric}
\alias{adpConvertRawToNumeric}
\title{Convert Raw to Numeric Values For adp Objects}
\usage{
adpConvertRawToNumeric(
  object = NULL,
  variables = NULL,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{object}{an \linkS4class{adp} object.}

\item{variables}{variables stored in an \linkS4class{adp} object that has
the same dimensional as \code{v} and is stored in a raw format.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\value{
\code{adpConvertRawToNumeric} returns an \linkS4class{adp} object whose specified
variables have been converted from raw to numerical format.
}
\description{
Convert variables in an \linkS4class{adp} object from raw to numeric format.
}
\examples{
library(oce)
data(adp)
adp[["a"]][,,1][,1]
ADP <- adpConvertRawToNumeric(adp)
ADP[["a"]][,,1][,1]
}
\seealso{
Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\author{
Jaimie Harbin and Dan Kelley
}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/processingLog.R
\name{processingLogItem}
\alias{processingLogItem}
\title{Create an item that can be inserted into a processing log}
\usage{
processingLogItem(value = "")
}
\arguments{
\item{value}{A string that will be used for the item.}
}
\value{
A \code{\link[=list]{list()}} containing \code{time}, which is
the time in UTC (calculated with \code{\link[=presentTime]{presentTime()}})
at the moment the function is called and
\code{value}, a string that is set to the argument of the same name.
}
\description{
A function is used internally to initialize processing logs.
Users will probably prefer to use \code{\link[=processingLogAppend]{processingLogAppend()}}
instead.
}
\seealso{
Other things related to processing logs: 
\code{\link{processingLog<-}()},
\code{\link{processingLogAppend}()},
\code{\link{processingLogShow}()}
}
\concept{things related to processing logs}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.R
\name{summary,adp-method}
\alias{summary,adp-method}
\alias{summary.adp}
\alias{summary,adp,missing-method}
\title{Summarize an ADP Object}
\usage{
\S4method{summary}{adp}(object, ...)
}
\arguments{
\item{object}{an object of class \code{"adp"}, usually, a result of a call
to \code{\link[=read.oce]{read.oce()}}, \code{\link[=read.adp.rdi]{read.adp.rdi()}}, or
\code{\link[=read.adp.nortek]{read.adp.nortek()}}.}

\item{\dots}{further arguments passed to or from other methods.}
}
\value{
A matrix containing statistics of the elements of the \code{data}
slot.
}
\description{
Summarize data in an \code{adp} object.
}
\details{
Pertinent summary information is presented.
}
\seealso{
Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley
}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{makeFilter}
\alias{makeFilter}
\title{Make a digital filter}
\usage{
makeFilter(
  type = c("blackman-harris", "rectangular", "hamming", "hann"),
  m,
  asKernel = TRUE
)
}
\arguments{
\item{type}{a string indicating the type of filter to use.  (See Harris
(1978) for a comparison of these and similar filters.)
\itemize{
\item \code{"blackman-harris"} yields a modified raised-cosine filter designated
as "4-Term (-92 dB) Blackman-Harris" by Harris (1978; coefficients given in
the table on page 65).  This is also called "minimum 4-sample Blackman
Harris" by that author, in his Table 1, which lists figures of merit as
follows: highest side lobe level -92dB; side lobe fall off -6 db/octave;
coherent gain 0.36; equivalent noise bandwidth 2.00 bins; 3.0-dB bandwidth
1.90 bins; scallop loss 0.83 dB; worst case process loss 3.85 dB; 6.0-db
bandwidth 2.72 bins; overlap correlation 46 percent for 75\\% overlap and 3.8
for 50\\% overlap.  Note that the equivalent noise bandwidth is the width of
a spectral peak, so that a value of 2 indicates a cutoff frequency of
\code{1/m}, where \code{m} is as given below.
\item \code{"rectangular"} for a flat filter.  (This is just for convenience.  Note that
\code{\link{kernel}}\verb{("daniell",....)} gives the same result, in kernel form.)
\code{"hamming"} for a Hamming filter (a raised-cosine that does not taper
to zero at the ends)
\item \code{"hann"} (a raised cosine that tapers to zero at the ends).
}}

\item{m}{length of filter.  This should be an odd number, for any
non-rectangular filter.}

\item{asKernel}{boolean, set to \code{TRUE} to get a smoothing kernel for
the return value.}
}
\value{
If \code{asKernel} is \code{FALSE}, this returns a list of filter
coefficients, symmetric about the midpoint and summing to 1.  These may be
used with \code{\link[=filter]{filter()}}, which should be provided with argument
\code{circular=TRUE} to avoid phase offsets.  If \code{asKernel} is
\code{TRUE}, the return value is a smoothing kernel, which can be applied to
a timeseries with \code{\link[=kernapply]{kernapply()}}, whose bandwidth can be determined
with \code{\link[=bandwidth.kernel]{bandwidth.kernel()}}, and which has both print and plot
methods.
}
\description{
The filter is suitable for use by \code{\link[=filter]{filter()}},
\code{\link[=convolve]{convolve()}} or (for the \code{asKernal=TRUE} case) with
\code{\link[=kernapply]{kernapply()}}.  Note that \code{\link[=convolve]{convolve()}} should be faster
than \code{\link[=filter]{filter()}}, but it cannot be used if the time series has
missing values.  For the Blackman-Harris filter, the half-power frequency is
at \code{1/m} cycles per time unit, as shown in the \dQuote{Examples}
section.  When using \code{\link[=filter]{filter()}} or \code{\link[=kernapply]{kernapply()}} with
these filters, use \code{circular=TRUE}.
}
\examples{
library(oce)

# 1. Demonstrate step-function response
y <- c(rep(1, 10), rep(-1, 10))
x <- seq_along(y)
plot(x, y, type='o', ylim=c(-1.05, 1.05))
BH <- makeFilter("blackman-harris", 11, asKernel=FALSE)
H <- makeFilter("hamming", 11, asKernel=FALSE)
yBH <- stats::filter(y, BH)
points(x, yBH, col=2, type='o')
yH <- stats::filter(y, H)
points(yH, col=3, type='o')
legend("topright", col=1:3, cex=2/3, pch=1,
       legend=c("input", "Blackman Harris", "Hamming"))

# 2. Show theoretical and practical filter gain, where
#    the latter is based on random white noise, and
#    includes a particular value for the spans
#    argument of spectrum(), etc.
\dontrun{
# need signal package for this example
r <- rnorm(2048)
rh <- stats::filter(r, H)
rh <- rh[is.finite(rh)] # kludge to remove NA at start/end
sR <- spectrum(r, plot=FALSE, spans=c(11, 5, 3))
sRH <- spectrum(rh, plot=FALSE, spans=c(11, 5, 3))
par(mfrow=c(2, 1), mar=c(3, 3, 1, 1), mgp=c(2, 0.7, 0))
plot(sR$freq, sRH$spec/sR$spec, xlab="Frequency", ylab="Power Transfer",
     type='l', lwd=5, col='gray')
theory <- freqz(H, n=seq(0,pi,length.out=100))
# Note we must square the modulus for the power spectrum
lines(theory$f/pi/2, Mod(theory$h)^2, lwd=1, col='red')
grid()
legend("topright", col=c("gray", "red"), lwd=c(5, 1), cex=2/3,
       legend=c("Practical", "Theory"), bg="white")
plot(log10(sR$freq), log10(sRH$spec/sR$spec),
     xlab="log10 Frequency", ylab="log10 Power Transfer",
     type='l', lwd=5, col='gray')
theory <- freqz(H, n=seq(0,pi,length.out=100))
# Note we must square the modulus for the power spectrum
lines(log10(theory$f/pi/2), log10(Mod(theory$h)^2), lwd=1, col='red')
grid()
legend("topright", col=c("gray", "red"), lwd=c(5, 1), cex=2/3,
       legend=c("Practical", "Theory"), bg="white")
}
}
\references{
F. J. Harris, 1978.  On the use of windows for harmonic analysis
with the discrete Fourier Transform.  \emph{Proceedings of the IEEE}, 66(1),
51-83 (\url{http://web.mit.edu/xiphmont/Public/windows.pdf}.)
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.R
\name{plotTS}
\alias{plotTS}
\title{Plot Temperature-Salinity Diagram}
\usage{
plotTS(
  x,
  inSitu = FALSE,
  type = "p",
  referencePressure = 0,
  nlevels = 6,
  levels,
  grid = TRUE,
  col.grid = "lightgray",
  lty.grid = "dotted",
  rho1000 = FALSE,
  eos = getOption("oceEOS", default = "gsw"),
  cex = par("cex"),
  col = par("col"),
  pch = par("pch"),
  bg = "white",
  pt.bg = "transparent",
  col.rho = gray(0.5),
  cex.rho = 3/4 * par("cex"),
  rotate = TRUE,
  useSmoothScatter = FALSE,
  xlab,
  ylab,
  Slim,
  Tlim,
  drawFreezing = TRUE,
  trimIsopycnals = TRUE,
  mgp = getOption("oceMgp"),
  mar = c(mgp[1] + 1.5, mgp[1] + 1.5, mgp[1], mgp[1]),
  lwd = par("lwd"),
  lty = par("lty"),
  lwd.rho = par("lwd"),
  lty.rho = par("lty"),
  add = FALSE,
  inset = FALSE,
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{x}{a \linkS4class{ctd}, \linkS4class{argo} or \linkS4class{section} object, or a list
containing solely \linkS4class{ctd} objects or \linkS4class{argo} objects.}

\item{inSitu}{A boolean indicating whether to use in-situ temperature or
(the default) potential temperature, calculated with reference pressure
given by \code{referencePressure}.  This is ignored if \code{eos="gsw"},
because those cases the y axis is necessarily the conservative formulation
of temperature.}

\item{type}{representation of data, \code{"p"} for points, \code{"l"} for
connecting lines, or \code{"n"} for no indication.}

\item{referencePressure}{reference pressure, to be used in calculating
potential temperature, if \code{inSitu} is \code{FALSE}.}

\item{nlevels}{Number of automatically-selected isopycnal levels (ignored if
\code{levels} is supplied).}

\item{levels}{Optional vector of desired isopycnal levels.}

\item{grid}{a flag that can be set to \code{TRUE} to get a grid.}

\item{col.grid}{color for grid.}

\item{lty.grid}{line type for grid.}

\item{rho1000}{if TRUE, label isopycnals as e.g. 1024; if FALSE, label as
e.g. 24}

\item{eos}{equation of state to be used, either \code{"unesco"} or
\code{"gsw"}.}

\item{cex}{character-expansion factor for symbols, as in \link{par}\code{("cex")}.}

\item{col}{color for symbols.}

\item{pch}{symbol type, as in \link{par}\code{("pch")}.}

\item{bg}{optional color to be painted under plotting area, before
plotting.  (This is useful for cases in which \code{inset=TRUE}.)}

\item{pt.bg}{inside color for symbols with \code{pch} in 21:25}

\item{col.rho}{color for isopycnal lines and their labels.}

\item{cex.rho}{size of the isopycnal labels.}

\item{rotate}{if TRUE, labels in right-hand margin are written vertically}

\item{useSmoothScatter}{if TRUE, use \code{\link[=smoothScatter]{smoothScatter()}} to plot the
points.}

\item{xlab}{optional label for the x axis, with default "Salinity [PSU]".}

\item{ylab}{optional label for the y axis, with default "Temperature [C]".}

\item{Slim}{optional limits for salinity axis, otherwise inferred from visible data
(i.e. the data that have finite values for both salinity and temperature).}

\item{Tlim}{as \code{Slim}, but for temperature.}

\item{drawFreezing}{logical indication of whether to draw a freezing-point
line. This is based on zero pressure. If \code{eos="unesco"} then
\code{\link[=swTFreeze]{swTFreeze()}} is used to compute the curve, whereas if
\code{eos="gsw"} then \code{\link[gsw:gsw_CT_freezing]{gsw::gsw_CT_freezing()}} is used;
in each case, zero pressure is used.}

\item{trimIsopycnals}{logical value (\code{TRUE} by default) that
indicates whether to trim isopycnal curves
to the region of temperature-salinity space for which density
computations are considered to be valid in the context of the
chosen \code{eos}; see \dQuote{Details}.}

\item{mgp}{3-element numerical vector to use for \verb{[par](mgp)}, and also
for \link{par}\code{(mar)}, computed from this.  The default is tighter than the R
default, in order to use more space for the data and less for the axes.}

\item{mar}{value to be used with \link{par}\code{("mar")}.  If set to
\code{NULL}, then \link{par}\code{("mar")} is used.  A good choice for a TS diagram
with a palette to the right is \verb{mar=par("mar")+c(0, 0, 0, 1))}.}

\item{lwd}{line width of lines or symbols.}

\item{lty}{line type of lines or symbols.}

\item{lwd.rho}{line width for density curves.}

\item{lty.rho}{line type for density curves.}

\item{add}{a flag that controls whether to add to an existing plot.  (It
makes sense to use \code{add=TRUE} in the \code{panel} argument of a
\code{\link[=coplot]{coplot()}}, for example.)}

\item{inset}{set to \code{TRUE} for use within \code{\link[=plotInset]{plotInset()}}.  The
effect is to prevent the present function from adjusting margins, which is
necessary because margin adjustment is the basis for the method used by
\code{\link[=plotInset]{plotInset()}}.}

\item{debug}{a flag that turns on debugging.  Set to 1 to get a moderate
amount of debugging information, or to 2 to get more.}

\item{\dots}{optional arguments passed to plotting functions.}
}
\value{
A list is silently returned, containing \code{xat} and \code{yat},
values that can be used by \code{\link[=oce.grid]{oce.grid()}} to add a grid to the plot.
}
\description{
Creates a temperature-salinity plot for a CTD cast, with labeled isopycnals.
}
\details{
The isopycnal curves (along which density is constant) are
drawn with \code{\link[=drawIsopycnals]{drawIsopycnals()}}, which also places
labels in the margins showing density minus 1000 \eqn{kg/m^3}{kg/m^3}.
If \code{trimIsopycnals} is \code{TRUE} (which is the default), these curves
are trimmed to the region within which the results of density calculation
in the chosen equation of state (\code{eos}) are considered to be reliable.

With \code{eos="unesco"} this region includes
Practical Salinity from 0 to 42 and Potential Temperature
from -2C to 40C, in accordance with Fofonoff and Millard
(1983, page 23).

With \code{eos="gsw"} the lower
limit of Absolute Salinity validity is taken as 0 g/kg,
in accordance with both McDougall et al. (2003 section 3)
and the TEOS-10/gsw Matlab code for the so-called "funnel" of validity.
However, an appropriate upper limit on Absolute Salinity is not as clear.
Here, the value 42 g/kg is chosen to match the "funnel" Matlab code
as of July 2020, but two other choices might have been
made. One is 50 g/kg, since \code{\link[gsw:gsw_SA_from_rho]{gsw::gsw_SA_from_rho()}} returns \code{NA} values
for Absolute Salinities exceeding that value, and another is
40 g/kg, as in McDougall et al. (2003 section 3).
The Conservative Temperature range is set to run from -2C
to 33C, as in McDougall et al. (2003 section 3), even though the
"funnel" imposes no upper limit on this variable.
}
\examples{
## For a simple ctd object
library(oce)
data(ctd)
plotTS(ctd)

## For a section object (note the outlier!)
data(section)
plotTS(section)

## Adding a colormap based on a different variable, e.g. oxygen
marOrig <- par("mar") # so later plots with palettes have same margins
cm <- colormap(section[['oxygen']])
drawPalette(colormap=cm, zlab='Oxygen')
plotTS(section, pch=19, col=cm$zcol, mar=par('mar')) # the mar adjusts for the palette

## Coloring based on station:
Tlim <- range(section[['temperature']], na.rm=TRUE)
Slim <- range(section[['salinity']], na.rm=TRUE)
cm <- colormap(seq_along(section[['latitude', 'byStation']]))
par(mar=marOrig) # same as previous plot
drawPalette(colormap=cm, zlab='Latitude')
plotTS(section, Tlim=Tlim, Slim=Slim, pch=NA, mar=par('mar'))
jnk <- mapply(
    function(s, col) {
        plotTS(s, col=col, add=TRUE, type='l')
    },
    section[['station']], col=cm$zcol)

## Show TS for an argo object
data(argo)
plotTS(handleFlags(argo))

}
\references{
\itemize{
\item Fofonoff, N. P., and R. C. Millard.
"Algorithms for Computation of Fundamental Properties of Seawater."
UNESCO Technical Papers in Marine Research. SCOR working group on Evaluation of CTD data;
UNESCO/ICES/SCOR/IAPSO Joint Panel on Oceanographic Tables and Standards, 1983.
\url{https://unesdoc.unesco.org/ark:/48223/pf0000059832}.
\item McDougall, Trevor J., David R. Jackett, Daniel G. Wright, and Rainer Feistel.
"Accurate and Computationally Efficient Algorithms for Potential Temperature and Density of Seawater."
Journal of Atmospheric and Oceanic Technology 20, no. 5 (May 1, 2003): 730-41.
\code{https://journals.ametsoc.org/jtech/article/20/5/730/2543/Accurate-and-Computationally-Efficient-Algorithms}.
}
}
\seealso{
\code{\link[=summary,ctd-method]{summary,ctd-method()}} summarizes the information, while
\code{\link[=read.ctd]{read.ctd()}} scans it from a file.

Other functions that plot oce data: 
\code{\link{download.amsr}()},
\code{\link{plot,adp-method}},
\code{\link{plot,adv-method}},
\code{\link{plot,amsr-method}},
\code{\link{plot,argo-method}},
\code{\link{plot,bremen-method}},
\code{\link{plot,cm-method}},
\code{\link{plot,coastline-method}},
\code{\link{plot,ctd-method}},
\code{\link{plot,gps-method}},
\code{\link{plot,ladp-method}},
\code{\link{plot,landsat-method}},
\code{\link{plot,lisst-method}},
\code{\link{plot,lobo-method}},
\code{\link{plot,met-method}},
\code{\link{plot,odf-method}},
\code{\link{plot,rsk-method}},
\code{\link{plot,satellite-method}},
\code{\link{plot,sealevel-method}},
\code{\link{plot,section-method}},
\code{\link{plot,tidem-method}},
\code{\link{plot,topo-method}},
\code{\link{plot,windrose-method}},
\code{\link{plot,xbt-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{tidem-class}}

Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}
}
\author{
Dan Kelley
}
\concept{functions that plot oce data}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/g1sst.R
\name{[[<-,g1sst-method}
\alias{[[<-,g1sst-method}
\title{Replace Parts of a G1SST Object}
\usage{
\S4method{[[}{g1sst}(x, i, j, ...) <- value
}
\arguments{
\item{x}{a \linkS4class{g1sst} object.}

\item{i}{character value naming the item to replace.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{optional additional information (ignored).}

\item{value}{The value to be placed into \code{x}, somewhere.}
}
\description{
The \code{[[<-} method works for all \linkS4class{oce} objects.
The purpose, as with the related extraction method, \code{[[},
is to insulate users from the internal details of \linkS4class{oce}
objects, by looking for items within the various storage
slots of the object. Items not actually stored can also be
replaced, including units and data-quality
flags.
}
\details{
As with \code{[[} method, the procedure works in steps.

First, the \code{metadata} slot of \code{x} is checked to
see whether it contains something named with \code{i}.
If so, then the named item is replaced with \code{value}.

Otherwise, if the string value of \code{i} ends in \code{Unit}, then the
characters preceding that are taken as the name of a variable, and
the \code{metadata} slot of \code{x} is updated to store that unit, e.g.\preformatted{x[["temperatureUnits"]] <- list(unit=expression(degree*F),scale="")
}

Similarly, if \code{i} ends in \code{Flag}, then quality-control
flags are set up as defined by \code{result}, e.g.\preformatted{o[["temperatureFlags"]] <- c(2,4,2,2)
}

Otherwise, \code{\link[=pmatch]{pmatch()}} is used for a partial-string match with
the names of the items that are in the \code{data} slot of \code{x}.
The first item found (if any) is then updated to hold the value \code{result}.

If none of these conditions is met, a warning is issued.
}
\seealso{
Other functions that replace parts of oce objects: 
\code{\link{[[<-,adp-method}},
\code{\link{[[<-,amsr-method}},
\code{\link{[[<-,argo-method}},
\code{\link{[[<-,bremen-method}},
\code{\link{[[<-,cm-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{[[<-,echosounder-method}},
\code{\link{[[<-,gps-method}},
\code{\link{[[<-,ladp-method}},
\code{\link{[[<-,landsat-method}},
\code{\link{[[<-,lisst-method}},
\code{\link{[[<-,lobo-method}},
\code{\link{[[<-,met-method}},
\code{\link{[[<-,oce-method}},
\code{\link{[[<-,odf-method}},
\code{\link{[[<-,rsk-method}},
\code{\link{[[<-,sealevel-method}},
\code{\link{[[<-,section-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{[[<-,topo-method}},
\code{\link{[[<-,windrose-method}},
\code{\link{[[<-,xbt-method}}

Other things related to g1sst data: 
\code{\link{[[,g1sst-method}},
\code{\link{g1sst-class}},
\code{\link{read.g1sst}()}
}
\concept{functions that replace parts of oce objects}
\concept{things related to g1sst data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/amsr.R
\name{plot,amsr-method}
\alias{plot,amsr-method}
\alias{plot.amsr}
\alias{both}
\alias{xlim}
\alias{and}
\alias{ylim}
\alias{to}
\alias{compute}
\alias{zrange}
\title{Plot an amsr Object}
\usage{
\S4method{plot}{amsr}(
  x,
  y,
  asp = NULL,
  breaks,
  col,
  colormap,
  zlim,
  missingColor = list(land = "papayaWhip", none = "lightGray", bad = "gray", rain =
    "plum", ice = "mediumVioletRed"),
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{x}{an \linkS4class{amsr} object.}

\item{y}{character value indicating the name of the band to plot; if not provided,
\code{SST} is used; see the documentation for the \linkS4class{amsr} class for a list of bands.}

\item{asp}{optional numerical value giving the aspect ratio for plot.  The
default value, \code{NULL}, means to use an aspect ratio of 1 for world views,
and a value computed from \code{ylim}, if the latter is specified in the
\code{...} argument.}

\item{breaks}{optional numeric vector of the z values for breaks in the color scheme.
If \code{colormap} is provided, it takes precedence over \code{breaks} and \code{col}.}

\item{col}{optional argument, either a vector of colors corresponding to the breaks, of length
1 less than the number of breaks, or a function specifying colors.
If neither \code{col} or \code{colormap} is provided, then \code{col} defaults to
\code{\link[=oceColorsTemperature]{oceColorsTemperature()}}.
If \code{colormap} is provided, it takes precedence over \code{breaks} and \code{col}.}

\item{colormap}{a specification of the colormap to use, as created
with \code{\link[=colormap]{colormap()}}.  If \code{colormap} is NULL, which is the default, then
a colormap is created to cover the range of data values, using
\link{oceColorsTemperature} colour scheme.
If \code{colormap} is provided, it takes precedence over \code{breaks} and \code{col}.
See \dQuote{Examples} for an example of using the "turbo" colour scheme.}

\item{zlim}{optional numeric vector of length 2, giving the limits
of the plotted quantity.  A reasonable default is computed, if this
is not given.}

\item{missingColor}{List of colors for problem cases. The names of the
elements in this list must be as in the default, but the colors may
be changed to any desired values. These default values work reasonably
well for SST images, which are the default image, and which employ a
blue-white-red blend of colors, no mixture of which matches the
default values in \code{missingColor}.}

\item{debug}{A debugging flag, integer.}

\item{...}{extra arguments passed to \code{\link[=imagep]{imagep()}}, e.g. to control
the view with \code{xlim} (for longitude) and \code{ylim} (for latitude).}
}
\description{
Plot an image of a component of an \linkS4class{amsr} object.
}
\details{
In addition to fields named directly in the object, such as \code{SSTDay} and
\code{SSTNight}, it is also possible to plot computed fields, such as \code{SST},
which combines the day and night fields.
}
\examples{
library(oce)
data(coastlineWorld)
data(amsr) # see ?amsr for how to read and composite such objects

# Example 1: plot with default colour scheme, oceColorsTemperature()
plot(amsr, "SST")
lines(coastlineWorld[['longitude']], coastlineWorld[['latitude']])

# Example 2: 'turbo' colour scheme
plot(amsr, "SST", col=oceColorsTurbo)
lines(coastlineWorld[['longitude']], coastlineWorld[['latitude']])

}
\seealso{
Other things related to amsr data: 
\code{\link{[[,amsr-method}},
\code{\link{[[<-,amsr-method}},
\code{\link{amsr-class}},
\code{\link{amsr}},
\code{\link{composite,amsr-method}},
\code{\link{download.amsr}()},
\code{\link{read.amsr}()},
\code{\link{subset,amsr-method}},
\code{\link{summary,amsr-method}}

Other functions that plot oce data: 
\code{\link{download.amsr}()},
\code{\link{plot,adp-method}},
\code{\link{plot,adv-method}},
\code{\link{plot,argo-method}},
\code{\link{plot,bremen-method}},
\code{\link{plot,cm-method}},
\code{\link{plot,coastline-method}},
\code{\link{plot,ctd-method}},
\code{\link{plot,gps-method}},
\code{\link{plot,ladp-method}},
\code{\link{plot,landsat-method}},
\code{\link{plot,lisst-method}},
\code{\link{plot,lobo-method}},
\code{\link{plot,met-method}},
\code{\link{plot,odf-method}},
\code{\link{plot,rsk-method}},
\code{\link{plot,satellite-method}},
\code{\link{plot,sealevel-method}},
\code{\link{plot,section-method}},
\code{\link{plot,tidem-method}},
\code{\link{plot,topo-method}},
\code{\link{plot,windrose-method}},
\code{\link{plot,xbt-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{tidem-class}}
}
\author{
Dan Kelley
}
\concept{functions that plot oce data}
\concept{things related to amsr data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/odf.R
\name{ODFListFromHeader}
\alias{ODFListFromHeader}
\title{Create a list of ODF header metadata}
\usage{
ODFListFromHeader(header)
}
\arguments{
\item{header}{Vector of character strings, holding the header}
}
\value{
A list holding the metadata, with item names matching
those in the ODF header, except that duplicates are transformed
through the use of \code{\link[=unduplicateNames]{unduplicateNames()}}.
}
\description{
Create a list of ODF header metadata
}
\seealso{
Other things related to odf data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{ODF2oce}()},
\code{\link{ODFNames2oceNames}()},
\code{\link{[[,odf-method}},
\code{\link{[[<-,odf-method}},
\code{\link{odf-class}},
\code{\link{plot,odf-method}},
\code{\link{read.ctd.odf}()},
\code{\link{read.odf}()},
\code{\link{subset,odf-method}},
\code{\link{summary,odf-method}}
}
\concept{things related to odf data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ladp.R
\name{as.ladp}
\alias{as.ladp}
\title{Coerce data into an ladp object}
\usage{
as.ladp(
  longitude,
  latitude,
  station,
  time,
  pressure,
  u,
  v,
  uz,
  vz,
  salinity,
  temperature,
  ...
)
}
\arguments{
\item{longitude}{longitude in degrees east, or an \code{oce} object that
contains the data otherwise given by \code{longitude} and the
other arguments.}

\item{latitude}{latitude in degrees east (use negative in southern hemisphere).}

\item{station}{number or string indicating station ID.}

\item{time}{time at the start of the profile, constructed by e.g. \code{\link[=as.POSIXct]{as.POSIXct()}}.}

\item{pressure}{pressure in decibars, through the water column.}

\item{u}{eastward velocity (m/s).}

\item{v}{northward velocity (m/s).}

\item{uz}{vertical derivative of eastward velocity (1/s).}

\item{vz}{vertical derivative of northward velocity (1/s).}

\item{salinity}{salinity through the water column, in practical salinity units.}

\item{temperature}{temperature through the water column.}

\item{\dots}{optional additional data columns.}
}
\value{
An \linkS4class{ladp} object.
}
\description{
This function assembles vectors of pressure and velocity, possibly also with
shears, salinity, temperature, etc.
}
\seealso{
Other things related to ladp data: 
\code{\link{[[,ladp-method}},
\code{\link{[[<-,ladp-method}},
\code{\link{ladp-class}},
\code{\link{plot,ladp-method}},
\code{\link{summary,ladp-method}}
}
\author{
Dan Kelley
}
\concept{things related to ladp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/map.R
\name{mapText}
\alias{mapText}
\title{Add Text to a Map}
\usage{
mapText(longitude, latitude, labels, ...)
}
\arguments{
\item{longitude}{numeric vector of longitudes of text to be plotted.}

\item{latitude}{numeric vector of latitudes of text to be plotted.}

\item{labels}{vector of labels of text to be plotted.}

\item{...}{optional arguments passed to \code{\link[=text]{text()}}, e.g. \code{adj},
\code{pos}, etc.}
}
\description{
Plot text on an existing map, by analogy to \code{\link[=text]{text()}}.
}
\examples{
\donttest{
library(oce)
data(coastlineWorld)
longitude <- coastlineWorld[['longitude']]
latitude <- coastlineWorld[['latitude']]
mapPlot(longitude, latitude, type='l', grid=5,
        longitudelim=c(-70,-50), latitudelim=c(45, 50),
        projection="+proj=merc")
lon <- -63.5744 # Halifax
lat <- 44.6479
mapPoints(lon, lat, pch=20, col="red")
mapText(lon, lat, "Halifax", col="red", pos=1, offset=1)
}

}
\seealso{
A map must first have been created with \code{\link[=mapPlot]{mapPlot()}}.

Other functions related to maps: 
\code{\link{formatPosition}()},
\code{\link{lonlat2map}()},
\code{\link{lonlat2utm}()},
\code{\link{map2lonlat}()},
\code{\link{mapArrows}()},
\code{\link{mapAxis}()},
\code{\link{mapContour}()},
\code{\link{mapCoordinateSystem}()},
\code{\link{mapDirectionField}()},
\code{\link{mapGrid}()},
\code{\link{mapImage}()},
\code{\link{mapLines}()},
\code{\link{mapLocator}()},
\code{\link{mapLongitudeLatitudeXY}()},
\code{\link{mapPlot}()},
\code{\link{mapPoints}()},
\code{\link{mapPolygon}()},
\code{\link{mapScalebar}()},
\code{\link{mapTissot}()},
\code{\link{oceCRS}()},
\code{\link{shiftLongitude}()},
\code{\link{usrLonLat}()},
\code{\link{utm2lonlat}()}
}
\author{
Dan Kelley
}
\concept{functions related to maps}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/astronomy.R
\name{siderealTime}
\alias{siderealTime}
\title{Convert a POSIXt time to a sidereal time}
\usage{
siderealTime(t)
}
\arguments{
\item{t}{a time, in POSIXt format, e.g. as created by
\code{\link[=as.POSIXct]{as.POSIXct()}}, \code{\link[=as.POSIXlt]{as.POSIXlt()}}, or
\code{\link[=numberAsPOSIXct]{numberAsPOSIXct()}}.  If this is provided, the other arguments are
ignored.}
}
\value{
A sidereal time, in hours in the range from 0 to 24.
}
\description{
Convert a POSIXt time to a sidereal time, using
the method in Chapter 7 of reference 1.  The small correction
that he discusses after his equation 7.1 is not applied here.
}
\examples{

t <- ISOdatetime(1978, 11, 13, 0, 0, 0, tz="UTC")
print(siderealTime(t))

}
\references{
\itemize{
\item Meeus, Jean. Astronomical Formulas for Calculators. Second Edition.
Richmond, Virginia, USA: Willmann-Bell, 1982.
}
}
\seealso{
Other things related to astronomy: 
\code{\link{angle2hms}()},
\code{\link{eclipticalToEquatorial}()},
\code{\link{equatorialToLocalHorizontal}()},
\code{\link{julianCenturyAnomaly}()},
\code{\link{julianDay}()},
\code{\link{moonAngle}()},
\code{\link{sunAngle}()},
\code{\link{sunDeclinationRightAscension}()}
}
\author{
Dan Kelley
}
\concept{things related to astronomy}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.R
\docType{class}
\name{adp-class}
\alias{adp-class}
\title{Class to Store adp (ADCP) Data}
\description{
This class stores data from acoustic Doppler profilers. Some manufacturers
call these ADCPs, while others call them ADPs; here the shorter form is
used by analogy to ADVs.
}
\section{Slots}{

\describe{
\item{\code{data}}{As with all \code{oce} objects, the \code{data} slot
for \code{adp} objects is a
\link{list} containing the main data for the object.
The key items stored in this slot include \code{time}, \code{distance}, and \code{v}, along with angles \code{heading}, \code{pitch} and \code{roll}.}

\item{\code{metadata}}{As with all \code{oce} objects, the \code{metadata} slot
for \code{adp} objects is a \link{list} containing
information about the \code{data} or about the object itself.
Examples that are of common interest include \code{oceCoordinate}, \code{orientation}, \code{frequency}, and \code{beamAngle}.}

\item{\code{processingLog}}{As with all \code{oce} objects, the \code{processingLog} slot
for \code{adp} objects is a
\link{list} with entries describing the creation and evolution
of the object. The contents are updated by various \code{oce} functions to
keep a record of processing steps.  Object summaries and
\code{\link[=processingLogShow]{processingLogShow()}} both display the log.}
}}

\section{Modifying slot contents}{


Although the \code{[[<-} operator may permit modification of the contents
of \linkS4class{adp} objects (see \code{\link{[[<-,adp-method}}),
it is better to use \code{\link[=oceSetData]{oceSetData()}} and \code{\link[=oceSetMetadata]{oceSetMetadata()}},
because those functions save an entry in the \code{processingLog}
that describes the change.
}

\section{Retrieving slot contents}{


The full contents of the \code{data} and \code{metadata} slots of a \linkS4class{adp}
object may be retrieved in the standard R way using \code{\link[=slot]{slot()}}. For
example \code{slot(o,"data")} returns the \code{data} slot of an object named \code{o},
and similarly \code{slot(o,"metadata")} returns
the \code{metadata} slot.

The slots may also be obtained with the \code{\link{[[,adp-method}}
operator, as e.g. \code{o[["data"]]} and \code{o[["metadata"]]}, respectively.

The \code{\link{[[,adp-method}} operator can also
be used to retrieve items from within the \code{data} and \code{metadata} slots.
For example, \code{o[["temperature"]]} can be used to retrieve temperature
from an object containing that quantity. The rule is that a named
quantity is sought first within the object's \code{metadata} slot,
with the \code{data} slot being checked only if \code{metadata} does not
contain the item. This \code{[[} method can also be used to get
certain derived quantities, if the object contains sufficient
information to calculate them. For example, an object that holds
(practical) salinity, temperature and pressure, along with
longitude and latitude, has sufficient information to compute
Absolute Salinity, and so \code{o[["SA"]]} will yield the
calculated Absolute Salinity.

It is also possible to find items more directly, using \code{\link[=oceGetData]{oceGetData()}} and
\code{\link[=oceGetMetadata]{oceGetMetadata()}}, but neither of these functions can
retrieve derived items.
}

\section{Reading/creating \code{adp} objects}{


The \code{metadata} slot contains various
items relating to the dataset, including source file name, sampling rate,
velocity resolution, velocity maximum value, and so on.  Some of these are
particular to particular instrument types, and prudent researchers will take
a moment to examine the whole contents of the metadata, either in summary
form (with \code{str(adp[["metadata"]])}) or in detail (with
\code{adp[["metadata"]]}).  Perhaps the most useful general properties are
\code{adp[["bin1Distance"]]} (the distance, in metres, from the sensor to
the bottom of the first bin), \code{adp[["cellSize"]]} (the cell height, in
metres, in the vertical direction, \emph{not} along the beam), and
\code{adp[["beamAngle"]]} (the angle, in degrees, between beams and an
imaginary centre line that bisects all beam pairs).

The diagram provided below indicates the coordinate-axis and beam-numbering
conventions for three- and four-beam ADP devices, viewed as though the
reader were looking towards the beams being emitted from the transducers.

\if{html}{\figure{adp_beams.png}{options: width="400" alt="Figure: adp_beams.png"}}

The bin geometry of a four-beam profiler is illustrated below, for
\code{adp[["beamAngle"]]} equal to 20 degrees, \code{adp[["bin1Distance"]]}
equal to 2m, and \code{adp[["cellSize"]]} equal to 1m.   In the diagram, the
viewer is in the plane containing two beams that are not shown, so the two
visible beams are separated by 40 degrees.  Circles indicate the centres of
the range-gated bins within the beams.  The lines enclosing those circles
indicate the coverage of beams that spread plus and minus 2.5 degrees from
their centreline.

\if{html}{\figure{adpgeometry2.png}{options: width="400" alt="Figure: adpgeometry2.png"}}

Note that \code{adp[["oceCoordinate"]]} stores the present coordinate system
of the object, and it has possible values \code{"beam"}, \code{"xyz"}, \code{"sfm"} or
\code{"enu"}.  (This should not be confused with
\code{adp[["originalCoordinate"]]}, which stores the coordinate system used
in the original data file.)

The \code{data} slot holds some standardized items, and
many that vary from instrument to instrument.  One standard item is
\code{adp[["v"]]}, a three-dimensional numeric array of velocities in
m/s.  In this matrix, the first index indicates time, the second bin
number, and the third beam number. The meaning of beams number depends on
whether the object is in beam coordinates, frame coordinates, or earth
coordinates.  For example, if in earth coordinates, then beam 1 is
the eastward component of velocity.
Thus, for example,
\preformatted{
library(oce)
data(adp)
t <- adp[['time']]
d <- adp[['distance']]
eastward <- adp[['v']][,,1]
imagep(t, d, eastward, missingColor="gray")
}
plots an image of the eastward component of velocity as a function of time (the x axis)
and distance from sensor (y axis), since the \code{adp} dataset is
in earth coordinates. Note the semidurnal tidal signal, and the pattern of missing
data at the ocean surface (gray blotches at the top).

Corresponding to the velocity array are two arrays of type raw, and
identical dimension, accessed by \code{adp[["a"]]} and \code{adp[["q"]]},
holding measures of signal strength and data quality quality,
respectively.  (The exact meanings of these depend on the particular type
of instrument, and it is assumed that users will be familiar enough with
instruments to know both the meanings and their practical consequences in
terms of data-quality assessment, etc.)

In addition to the arrays, there are time-based vectors.  The vector
\code{adp[["time"]]} (of length equal to the first index of
\code{adp[["v"]]}, etc.) holds times of observation.  Depending on type of
instrument and its configuration, there may also be corresponding vectors
for sound speed (\code{adp[["soundSpeed"]]}), pressure
(\code{adp[["pressure"]]}), temperature (\code{adp[["temperature"]]}),
heading (\code{adp[["heading"]]}) pitch (\code{adp[["pitch"]]}), and roll
(\code{adp[["roll"]]}), depending on the setup of the instrument.

The precise meanings of the data items depend on the instrument type.  All
instruments have \code{v} (for velocity), \code{q} (for a measure of data
quality) and \code{a} (for a measure of backscatter amplitude, also called
echo intensity).
Teledyne-RDI profilers have an additional item \code{g} (for
percent-good).

VmDas-equipped Teledyne-RDI profilers additional navigation data, with
details listed in the table below; note that the RDI documentation (reference 2) and
the RDI gui use inconsistent names for most items.

\tabular{lll}{
\strong{Oce name}\tab \strong{RDI doc name}\tab \strong{RDI GUI name}\cr
\code{avgSpeed}\tab Avg Speed\tab Speed/Avg/Mag\cr
\code{avgMagnitudeVelocityEast}\tab Avg Mag Vel East\tab ?\cr
\code{avgMagnitudeVelocityNorth}\tab Avg Mag Vel North\tab ?\cr
\code{avgTrackMagnetic}\tab Avg Track Magnetic\tab Speed/Avg/Dir (?)\cr
\code{avgTrackTrue}\tab Avg Track True\tab Speed/Avg/Dir (?)\cr
\code{avgTrueVelocityEast}\tab Avg True Vel East\tab ?\cr
\code{avgTrueVelocityNorth}\tab Avg True Vel North\tab ?\cr
\code{directionMadeGood}\tab Direction Made Good\tab Speed/Made Good/Dir\cr
\code{firstLatitude}\tab First latitude\tab Start Lat\cr
\code{firstLongitude}\tab First longitude\tab Start Lon\cr
\code{firstTime}\tab UTC Time of last fix\tab End Time\cr
\code{lastLatitude}\tab Last latitude\tab End Lat\cr
\code{lastLongitude}\tab Last longitude\tab End Lon\cr
\code{lastTime}\tab UTC Time of last fix\tab End Time\cr
\code{numberOfHeadingSamplesAveraged}\tab Number heading samples averaged\tab ?\cr
\code{numberOfMagneticTrackSamplesAveraged}\tab Number of magnetic track samples averaged\tab ? \cr
\code{numberOfPitchRollSamplesAvg}\tab Number of magnetic track samples averaged\tab ? \cr
\code{numberOfSpeedSamplesAveraged}\tab Number of speed samples averaged\tab ? \cr
\code{numberOfTrueTrackSamplesAvg}\tab Number of true track samples averaged\tab ? \cr
\code{primaryFlags}\tab Primary Flags\tab ?\cr
\code{shipHeading}\tab Heading\tab ?\cr
\code{shipPitch}\tab Pitch\tab ?\cr
\code{shipRoll}\tab Roll\tab ?\cr
\code{speedMadeGood}\tab Speed Made Good\tab Speed/Made Good/Mag\cr
\code{speedMadeGoodEast}\tab Speed MG East\tab ?\cr
\code{speedMadeGoodNorth}\tab Speed MG North\tab ?\cr
}

For Teledyne-RDI profilers, there are four three-dimensional arrays
holding beamwise data.  In these, the first index indicates time, the
second bin number, and the third beam number (or coordinate number, for
data in \code{xyz}, \code{sfm}, \code{enu} or \code{other} coordinate systems).  In
the list below, the quoted phrases are quantities as defined in Figure 9
of reference 1.
\itemize{
\item \code{v} is ``velocity'' in m/s, inferred from two-byte signed
integer values (multiplied by the scale factor that is stored in
\code{velocityScale} in the metadata).
\item \code{q} is ``correlation magnitude'' a one-byte quantity stored
as type \code{raw} in the object. The values may range from 0 to 255.
\item \code{a} is \verb{backscatter amplitude}, also known as ``echo
intensity'' a one-byte quantity stored as type \code{raw} in the object.
The values may range from 0 to 255.
\item \code{g} is ``percent good'' a one-byte quantity stored as \code{raw}
in the object.  The values may range from 0 to 100.
}

Finally, there is a vector \code{adp[["distance"]]} that indicates the bin
distances from the sensor, measured in metres along an imaginary centre
line bisecting beam pairs.  The length of this vector equals
\code{dim(adp[["v"]])[2]}.
}

\section{Teledyne-RDI Sentinel V ADCPs}{
 As of 2016-09-27 there is
provisional support for the TRDI "SentinelV" ADCPs, which are 5
beam ADCPs with a vertical centre beam. Relevant vertical beam
fields are called \code{adp[["vv"]]}, \code{adp[["va"]]},
\code{adp[["vq"]]}, and \code{adp[["vg"]]} in analogy with the
standard 4-beam fields.
}

\section{Accessing and altering information within \linkS4class{adp} objects}{

\emph{Extracting values} Matrix data may be accessed as illustrated
above, e.g.  or an adp object named \code{adv}, the data are provided by
\code{adp[["v"]]}, \code{adp[["a"]]}, and \code{adp[["q"]]}.  As a
convenience, the last two of these can be accessed as numeric (as opposed to
raw) values by e.g.  \code{adp[["a", "numeric"]]}.  The vectors are accessed
in a similar way, e.g. \code{adp[["heading"]]}, etc.  Quantities in the
\code{metadata} slot are also available by name, e.g.
\code{adp[["velocityResolution"]]}, etc.

\emph{Assigning values.} This follows the standard form, e.g. to increase
all velocity data by 1 cm/s, use \code{adp[["v"]] <- 0.01 + adp[["v"]]}.

\emph{Overview of contents} The \code{show} method (e.g.
\code{show(d)}) displays information about an ADP object named \code{d}.
}

\section{Dealing with suspect data}{

There are many possibilities for confusion
with \code{adp} devices, owing partly to the flexibility that manufacturers
provide in the setup.  Prudent users will undertake many tests before trusting
the details of the data.  Are mean currents in the expected direction, and of
the expected magnitude, based on other observations or physical constraints?
Is the phasing of currents as expected?  If the signals are suspect, could an
incorrect scale account for it?  Could the transformation matrix be incorrect?
Might the data have exceeded the maximum value, and then ``wrapped around'' to
smaller values?  Time spent on building confidence in data quality is seldom
time wasted.
}

\section{References}{

\enumerate{
\item Teledyne-RDI, 2007.
\emph{WorkHorse commands and output data format.}
P/N 957-6156-00 (November 2007).
\item Teledyne-RDI, 2012. \emph{VmDas User's Guide, Ver. 1.46.5}.
}
}

\seealso{
A file containing ADP data is usually recognized by Oce, and so
\code{\link[=read.oce]{read.oce()}} will usually read the data.  If not, one may use the
general ADP function \code{\link[=read.adp]{read.adp()}} or specialized variants
\code{\link[=read.adp.rdi]{read.adp.rdi()}}, \code{\link[=read.adp.nortek]{read.adp.nortek()}},
\code{\link[=read.adp.ad2cp]{read.adp.ad2cp()}},
\code{\link[=read.adp.sontek]{read.adp.sontek()}} or \code{\link[=read.adp.sontek.serial]{read.adp.sontek.serial()}}.

ADP data may be plotted with \code{\link[=plot,adp-method]{plot,adp-method()}}, which is a
generic function so it may be called simply as \code{plot}.

Statistical summaries of ADP data are provided by the generic function
\code{summary}, while briefer overviews are provided with \code{show}.

Conversion from beam to xyz coordinates may be done with
\code{\link[=beamToXyzAdp]{beamToXyzAdp()}}, and from xyz to enu (east north up) may be done
with \code{\link[=xyzToEnuAdp]{xyzToEnuAdp()}}.  \code{\link[=toEnuAdp]{toEnuAdp()}} may be used to
transfer either beam or xyz to enu.  Enu may be converted to other coordinates
(e.g. aligned with a coastline) with \code{\link[=enuToOtherAdp]{enuToOtherAdp()}}.

Other classes provided by oce: 
\code{\link{adv-class}},
\code{\link{argo-class}},
\code{\link{bremen-class}},
\code{\link{cm-class}},
\code{\link{coastline-class}},
\code{\link{ctd-class}},
\code{\link{lisst-class}},
\code{\link{lobo-class}},
\code{\link{met-class}},
\code{\link{oce-class}},
\code{\link{odf-class}},
\code{\link{rsk-class}},
\code{\link{sealevel-class}},
\code{\link{section-class}},
\code{\link{topo-class}},
\code{\link{windrose-class}},
\code{\link{xbt-class}}

Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\concept{classes provided by oce}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swAlphaOverBeta}
\alias{swAlphaOverBeta}
\title{Ratio of seawater thermal expansion coefficient to haline contraction
coefficient}
\usage{
swAlphaOverBeta(
  salinity,
  temperature = NULL,
  pressure = NULL,
  longitude = NULL,
  latitude = NULL,
  eos = getOption("oceEOS", default = "gsw")
)
}
\arguments{
\item{salinity}{either practical salinity (in which case \code{temperature}
and \code{pressure} must be provided) \emph{or} an \code{oce} object (in
which case \code{salinity}, etc. are inferred from the object).}

\item{temperature}{\emph{in-situ} temperature (\eqn{^\circ}{deg}C)}

\item{pressure}{pressure (dbar)}

\item{longitude}{longitude of observation (only used if \code{eos="gsw"};
see \sQuote{Details}).}

\item{latitude}{latitude of observation (only used if \code{eos="gsw"}; see
\sQuote{Details}).}

\item{eos}{equation of state, either \code{"unesco"} or \code{"gsw"}.}
}
\value{
Value in psu/\eqn{^\circ}{deg}C.
}
\description{
Compute \eqn{\alpha/\beta}{alpha/beta} using McDougall's (1987) algorithm.
}
\examples{
swAlphaOverBeta(40, 10, 4000, eos="unesco") # 0.3476

}
\references{
The \code{eos="unesco"} formulae are based on the UNESCO
equation of state, but are formulated empirically by Trevor J. McDougall,
1987, Neutral Surfaces, Journal of Physical Oceanography, volume 17, pages
1950-1964. The \code{eos="gsw"} formulae come from GSW; see references in
the \code{\link[=swRho]{swRho()}} documentation.
}
\seealso{
Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.woce.R
\name{read.ctd.woce.other}
\alias{read.ctd.woce.other}
\title{Read a WOCE-exchange EXPOCODE File}
\usage{
read.ctd.woce.other(
  file,
  columns = NULL,
  station = NULL,
  missingValue,
  deploymentType = "unknown",
  monitor = FALSE,
  debug = getOption("oceDebug"),
  processingLog,
  ...
)
}
\arguments{
\item{file}{a connection or a character string giving the name of the file to
load.  For \code{\link[=read.ctd.sbe]{read.ctd.sbe()}} and \code{\link[=read.ctd.woce]{read.ctd.woce()}}, this may be a
wildcard (e.g. \code{"*.cnv"} or \code{"*.csv"}) in which case the return
value is a vector containing CTD objects created by reading the files from
\code{\link[=list.files]{list.files()}} with \code{pattern} set to the specified wildcard
pattern.}

\item{columns}{an optional \link{list} that can be used to convert unrecognized
data names to resultant variable names.  This is used only by
\code{\link[=read.ctd.sbe]{read.ctd.sbe()}} and \code{\link[=read.ctd.odf]{read.ctd.odf()}}. For example,
if a data file named salinity as \code{"SAL"}, then using\preformatted{d <- read.ctd(f, columns=list(
    salinity=list(name="SAL",
                  unit=list(unit=expression(),
                  scale="PSS-78"))))
}

would assign the \code{"SAL"} column to the \code{salinity} entry in the data
slot of the CTD object returned by the \verb{read.*} function.}

\item{station}{optional character string containing an identifying name or
number for the station. This can be useful if the routine cannot determine the
name automatically, or if another name is preferred.}

\item{missingValue}{optional missing-value flag; data matching this value will
be set to \code{NA} upon reading. If this is provided, then it overrules any
missing-value flag found in the data. For Seabird (\code{.cnv}) files, there is
usually no need to set \code{missingValue}, because it can be inferred from the
header (typically as -9.990e-29). Set \code{missingValue=NULL} to turn off
missing-value detection, even in \code{.cnv} files that contain missing-value
codes in their headers. If \code{missingValue} is not specified,
then an attempt is made to infer such a value from the data, by testing
whether salinity and/or temperature has a minimum that is under -8 in value;
this should catch common values in files, without false positives. A warning
will be issued in this case, and a note inserted in the processing log of
the return value.}

\item{deploymentType}{character string indicating the type of deployment. Use
\code{"unknown"} if this is not known, \code{"profile"} for a profile (in
which the data were acquired during a downcast, while the device was lowered
into the water column, perhaps also including an upcast; \code{"moored"} if
the device is installed on a fixed mooring, \code{"thermosalinograph"} (or
\code{"tsg"}) if the device is mounted on a moving vessel, to record
near-surface properties, or \code{"towyo"} if the device is repeatedly
lowered and raised.}

\item{monitor}{boolean, set to \code{TRUE} to provide an indication of
progress.  This is useful if \code{filename} is a wildcard.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed.}

\item{processingLog}{if provided, the action item to be stored in the log.
This is typically only provided for internal calls; the default that it provides is
better for normal calls by a user.}

\item{...}{additional arguments, passed to called routines.}
}
\value{
a \linkS4class{ctd} object. The details of the contents
depend on the source file. The \code{metadata} slot is particularly
variable across data formats, because the meta-information provided
in those formats varies so widely.
}
\description{
This reads WOCE exchange files that start with the string \code{"EXPOCODE"}.
}
\seealso{
Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}

Other functions that read ctd data: 
\code{\link{read.ctd.aml}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()}
}
\author{
Dan Kelley
}
\concept{functions that read ctd data}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swAbsoluteSalinity}
\alias{swAbsoluteSalinity}
\title{Seawater absolute salinity, in GSW formulation}
\usage{
swAbsoluteSalinity(
  salinity,
  pressure = NULL,
  longitude = NULL,
  latitude = NULL
)
}
\arguments{
\item{salinity}{either practical salinity (in which case \code{temperature}
and \code{pressure} must be provided) \emph{or} an \code{oce} object (in
which case \code{salinity}, etc. are inferred from the object).}

\item{pressure}{pressure in dbar.}

\item{longitude}{longitude of observation.}

\item{latitude}{latitude of observation.}
}
\value{
Absolute Salinity in \eqn{g/kg}{g/kg}.
}
\description{
Compute the seawater Absolute Salinity, according to the GSW/TEOS-10
formulation with \code{\link[gsw:gsw_SA_from_SP]{gsw::gsw_SA_from_SP()}} in the \CRANpkg{gsw} package.
Typically, this is a fraction of a unit
higher than practical salinity as defined in the UNESCO formulae.
}
\examples{
\dontrun{
sa <- swAbsoluteSalinity(35.5, 300, 260, 16)
stopifnot(abs(35.671358392019094 - sa) < 00.000000000000010)
}

}
\references{
McDougall, T.J. and P.M. Barker, 2011: Getting started with
TEOS-10 and the Gibbs Seawater (GSW) Oceanographic Toolbox, 28pp.,
SCOR/IAPSO WG127, ISBN 978-0-646-55621-5.
}
\seealso{
The related TEOS-10 quantity ``conservative temperature'' may be
computed with \code{\link[=swConservativeTemperature]{swConservativeTemperature()}}.  For a ctd object,
absolute salinity may also be recovered by indexing as e.g.
\code{ctd[["absoluteSalinity"]]} or \code{ctd[["SA"]]}.

Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/section.R
\name{sectionSmooth}
\alias{sectionSmooth}
\title{Smooth a Section}
\usage{
sectionSmooth(
  section,
  method = "spline",
  x,
  xg,
  yg,
  xgl,
  ygl,
  xr,
  yr,
  df,
  gamma = 0.5,
  iterations = 2,
  trim = 0,
  pregrid = FALSE,
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{section}{A \code{section} object containing the section to be smoothed.
For \code{method="spline"}, the pressure levels must match for each station in
the section.}

\item{method}{A string or a function that specifies the method to use; see \sQuote{Details}.}

\item{x}{Optional numerical vector, of the same length as the number of stations in \code{section},
which will be used in gridding in the lateral direction. If not provided, this
defaults to \code{\link{geodDist}}\code{(section)}.}

\item{xg, xgl}{ignored in the \code{method="spline"} case, but passed to
\code{\link[=interpBarnes]{interpBarnes()}} if \code{method="barnes"}, to kriging
functions if \code{method="kriging"}, or to \code{method} itself, if it
is a function.
If \code{xg} is supplied, it defines the x component of the grid, which by
default is the terms of station distances, x, along the track of the section. (Note
that the grid \code{xg} is trimmed to the range of the data \code{x}, because otherwise
it would be impossible to interpolate between stations to infer water depth,
longitude, and latitude, which are all stored within the stations in the
returned \code{section} object.)
Alternatively, if \code{xgl} is supplied, the x grid is established using \code{\link[=seq]{seq()}},
to span the data with \code{xgl} elements. If neither of these is supplied, the output
x grid will equal the input x grid.}

\item{yg, ygl}{similar to \code{xg} and \code{xgl}, but for pressure. (Note that
trimming to the input \code{y} is not done, as it is for \code{xg} and \code{x}.)
If \code{ygl} was not given, then a grid is constructed to span the pressures
of that deepest station with \code{ygl} elements. On the other hand,
if \code{ygl} was not given, then the y grid will constructed from the
pressure levels in the deepest station.}

\item{xr, yr}{influence ranges in x (along-section distance) and y (pressure),
passed to \code{\link[=interpBarnes]{interpBarnes()}} if \code{method="barnes"} or to
\code{method}, if the latter is a function. If missing, \code{xr} defaults to
1.5X the median inter-station distance and \code{yr} defaults to 0.2X
the pressure range. Since these defaults have changed over the evolution
of \code{sectionSmooth}, analysts ought to supply \code{xr} and \code{yr}
in the function call, tailoring them to particular applications, and
making the code more resistant to changes in \code{sectionSmooth}.}

\item{df}{Degree-of-freedom parameter, passed to \code{\link[=smooth.spline]{smooth.spline()}}
if \code{method="spline"}, and ignored otherwise. If \code{df} is not provided,
it defaults to 1/5-th of the number of stations containing non-\code{NA}
data at the particular pressure level being processed, as \code{sectionSmooth}
works its way through the water column.}

\item{gamma, iterations, trim, pregrid}{Values passed to
\code{\link[=interpBarnes]{interpBarnes()}}, if \code{method="barnes"}, and
ignored otherwise. \code{gamma} is the factor by which
\code{xr} and \code{yr} are reduced on each of succeeding iterations.
\code{iterations} is the number of iterations to do.
\code{trim} controls whether the gridded data are set to
\code{NA} in regions with sparse data
coverage. \code{pregrid} controls whether data are to be
pre-gridded with \code{\link[=binMean2D]{binMean2D()}} before being passed to
\code{\link[=interpBarnes]{interpBarnes()}}.}

\item{debug}{A flag that turns on debugging.  Set to 1 to get a moderate amount
of debugging information, or to 2 to get more.}

\item{...}{Optional extra arguments, passed to either
\code{\link[=smooth.spline]{smooth.spline()}}, if \code{method="spline"}, and ignored otherwise.}
}
\value{
A \linkS4class{section} object of that has been smoothed in some way.
Every data field that is in even a single station of the input object
is inserted into every station in the returned value, and therefore
the units represent all the units in any of the stations, as do the
flag names. However, the flags are all set to \code{NA} values.
}
\description{
Smooth a section, in any of several ways, working either in the vertical
direction or in both the vertical and lateral directions.
}
\details{
This function produces smoothed fields that might be useful in
simplifying graphical elements created with \code{\link[=plot,section-method]{plot,section-method()}}.
As with any smoothing method, a careful analyst will compare the results
against the raw data, e.g. using \code{\link[=plot,section-method]{plot,section-method()}}.
In addition the problem of falsely widening narrow features such as
fronts, there is also the potential to get unphysical results with
spars sampling near topographic features such as bottom slopes and ridges.

The \code{method} argument selects between three possible methods.
\itemize{
\item For \code{method="spline"}, i.e. the default, the section is smoothed
laterally, using \code{\link[=smooth.spline]{smooth.spline()}} on individual pressure levels.
(If the pressure levels do not match up, \code{\link[=sectionGrid]{sectionGrid()}} should
be used first to create a \code{section} object that can be used here.)
The \code{df} argument sets the degree of freedom of the spline, with
larger values indicating less smoothing.
\item For the (much slower) \code{method="barnes"} method, smoothing is done across
both horizontal and vertical coordinates, using \code{\link[=interpBarnes]{interpBarnes()}}.
The output station locations are computed by linear interpolation of
input locations, using \code{\link[=approx]{approx()}} on the original
longitudes and longitudes of stations, with the independent variable
being the distance along the track, computed with \code{\link[=geodDist]{geodDist()}}.
The values of \code{xg}, \code{yg}, \code{xgl} and \code{ygl} control
the smoothing.
\item If \code{method} is a function, then that function is applied to
the (distance, pressure) data for each variable at a grid defined by
\code{xg}, \code{xgl}, \code{yg} and \code{ygl}. The function must
be of the form \verb{function(x,y,F,xg,xr,yg,yr)}, and must
return a matrix of the gridded result, with first index indicating
the "grid" station number and second index indicating "grid" pressure.
The \code{x} value that is supplied to this function is set as
the distance along the section, as computed with \code{\link[=geodDist]{geodDist()}},
and repeated for each of the points at each station.  The corresponding
pressures are provided in \code{y}, and the value to be gridded is
in \code{v}, which will be \code{temperture} on one call to the function,
\code{salinity} on another call, etc. The other quantities
have the meanings as described below.  See the \dQuote{Examples}
for a description of how to set up and use a function for the gridding
method known as Kriging.
}
}
\examples{
# Unsmoothed (Gulf Stream)
library(oce)
data(section)
gs <- subset(section, 115<=stationId&stationId<=125)
par(mfrow=c(2, 2))

plot(gs, which="temperature")
mtext("unsmoothed")

# Spline
gsg <- sectionGrid(gs, p=seq(0, 5000, 100))
gsSpline <- sectionSmooth(gsg, "spline")
plot(gsSpline, which="temperature")
mtext("spline-smoothed")

# Barnes
gsBarnes <- sectionSmooth(gs, "barnes", xr=50, yr=200)
plot(gsBarnes, which="temperature")
mtext("Barnes-smoothed")

# Kriging.
# As of early March, 2022, two CRAN test machines (running R-devel
# on linux-fedora) started to fail on the code shown below, so it
# was deactivated, as a cautionary measure.  It may reactivated
# later, if the CRAN machines start handling it properly again.
\dontrun{
if (requireNamespace("automap",quietly=TRUE)&&requireNamespace("sp",quietly=TRUE)) {
  krig <- function(x, y, F, xg, xr, yg, yr) {
    xy <- data.frame(x=x/xr, y=y/yr)
    K <- automap::autoKrige(F~1, remove_duplicates=TRUE,
                            input_data=sp::SpatialPointsDataFrame(xy, data.frame(F)),
                            new_data=sp::SpatialPoints(expand.grid(xg/xr, yg/yr)))
    matrix(K$krige_output@data$var1.pred, nrow=length(xg), ncol=length(yg))
  }
  gsKrig <- sectionSmooth(gs, krig)
  plot(gsKrig, which="temperature")
  mtext("Kriging-smoothed")
}
}

}
\seealso{
Other things related to section data: 
\code{\link{[[,section-method}},
\code{\link{[[<-,section-method}},
\code{\link{as.section}()},
\code{\link{handleFlags,section-method}},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{plot,section-method}},
\code{\link{read.section}()},
\code{\link{section-class}},
\code{\link{sectionAddStation}()},
\code{\link{sectionGrid}()},
\code{\link{sectionSort}()},
\code{\link{section}},
\code{\link{subset,section-method}},
\code{\link{summary,section-method}}
}
\author{
Dan Kelley
}
\concept{things related to section data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/odf.R
\name{read.ctd.odf}
\alias{read.ctd.odf}
\title{Read a CTD file in ODF format}
\usage{
read.ctd.odf(
  file,
  columns = NULL,
  station = NULL,
  missingValue,
  deploymentType = "unknown",
  monitor = FALSE,
  exclude = NULL,
  debug = getOption("oceDebug"),
  processingLog,
  ...
)
}
\arguments{
\item{file}{a connection or a character string giving the name of the file to
load.  For \code{\link[=read.ctd.sbe]{read.ctd.sbe()}} and \code{\link[=read.ctd.woce]{read.ctd.woce()}}, this may be a
wildcard (e.g. \code{"*.cnv"} or \code{"*.csv"}) in which case the return
value is a vector containing CTD objects created by reading the files from
\code{\link[=list.files]{list.files()}} with \code{pattern} set to the specified wildcard
pattern.}

\item{columns}{an optional \link{list} that can be used to convert unrecognized
data names to resultant variable names.  This is used only by
\code{\link[=read.ctd.sbe]{read.ctd.sbe()}} and \code{\link[=read.ctd.odf]{read.ctd.odf()}}. For example,
if a data file named salinity as \code{"SAL"}, then using\preformatted{d <- read.ctd(f, columns=list(
    salinity=list(name="SAL",
                  unit=list(unit=expression(),
                  scale="PSS-78"))))
}

would assign the \code{"SAL"} column to the \code{salinity} entry in the data
slot of the CTD object returned by the \verb{read.*} function.}

\item{station}{optional character string containing an identifying name or
number for the station. This can be useful if the routine cannot determine the
name automatically, or if another name is preferred.}

\item{missingValue}{optional missing-value flag; data matching this value will
be set to \code{NA} upon reading. If this is provided, then it overrules any
missing-value flag found in the data. For Seabird (\code{.cnv}) files, there is
usually no need to set \code{missingValue}, because it can be inferred from the
header (typically as -9.990e-29). Set \code{missingValue=NULL} to turn off
missing-value detection, even in \code{.cnv} files that contain missing-value
codes in their headers. If \code{missingValue} is not specified,
then an attempt is made to infer such a value from the data, by testing
whether salinity and/or temperature has a minimum that is under -8 in value;
this should catch common values in files, without false positives. A warning
will be issued in this case, and a note inserted in the processing log of
the return value.}

\item{deploymentType}{character string indicating the type of deployment. Use
\code{"unknown"} if this is not known, \code{"profile"} for a profile (in
which the data were acquired during a downcast, while the device was lowered
into the water column, perhaps also including an upcast; \code{"moored"} if
the device is installed on a fixed mooring, \code{"thermosalinograph"} (or
\code{"tsg"}) if the device is mounted on a moving vessel, to record
near-surface properties, or \code{"towyo"} if the device is repeatedly
lowered and raised.}

\item{monitor}{boolean, set to \code{TRUE} to provide an indication of
progress.  This is useful if \code{filename} is a wildcard.}

\item{exclude}{either a character value holding a regular
expression that is used with \code{\link[=grep]{grep()}} to remove lines from the
header before processing, or \code{NULL} (the default), meaning
not to exclude any such lines.  The purpose of this argument
is to solve problems with some files, which can have
thousands of lines that indicate details that are may be of
little value in processing.  For example, some files have thousands
of lines that would be excluded by using
\code{exclude="PROCESS='Nulled the .* value"} in the function call.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed.}

\item{processingLog}{if provided, the action item to be stored in the log.
This is typically only provided for internal calls; the default that it provides is
better for normal calls by a user.}

\item{...}{additional arguments, passed to called routines.}
}
\value{
a \linkS4class{ctd} object. The details of the contents
depend on the source file. The \code{metadata} slot is particularly
variable across data formats, because the meta-information provided
in those formats varies so widely.
}
\description{
Read a CTD file in ODF format
}
\details{
\code{read.ctd.odf} reads files stored in Ocean Data Format, used in
some Canadian hydrographic databases.
}
\section{Caution}{


Lacking detailed documentation of the ODF file format, the \code{\link[=read.odf]{read.odf()}} and
\code{\link[=read.ctd.odf]{read.ctd.odf()}} functions were crafted based on inspection of data files, and
so some guesses had to be made.

The \code{PARAMETER_HEADER} chunks describing quality-control flags are
a case in point.  These contain \code{NAME} components that
refer to other \code{PARAMETER_HEADER} chunks that hold measured data.  However,
those references are not always matched well with the data names, and
even if they do match, the cross-reference syntax used by
the Bedford Institute of Oceanography differs from that used by
l‚ÄôInstitut Maurice-Lamontagne. To simplify coding, it was assumed that
each quality-control sequence applies to the data sequence
immediately preceding it.  (This assumption is made in other
analysis systems.)

It is also prudent to pay attention to the units decoding,
which \code{\link[=read.odf]{read.odf()}} handles by calling \code{\link[=unitFromString]{unitFromString()}}.
Be on the lookout for incorrect temperature scales, which
are sometimes reported with nonstandard strings in ODF files.
Also, note that you may see warnings about conductivity ratios,
which some ODF files incorrectly suggest have dimensions.
}

\references{
For sources that describe the ODF format, see the documentation
for the \linkS4class{odf} class.
}
\seealso{
Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}

Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}

Other things related to odf data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{ODF2oce}()},
\code{\link{ODFListFromHeader}()},
\code{\link{ODFNames2oceNames}()},
\code{\link{[[,odf-method}},
\code{\link{[[<-,odf-method}},
\code{\link{odf-class}},
\code{\link{plot,odf-method}},
\code{\link{read.odf}()},
\code{\link{subset,odf-method}},
\code{\link{summary,odf-method}}

Other functions that read ctd data: 
\code{\link{read.ctd.aml}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()}
}
\author{
Dan Kelley
}
\concept{functions that read ctd data}
\concept{things related to ctd data}
\concept{things related to odf data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{rangeExtended}
\alias{rangeExtended}
\title{Calculate Range, Extended a Little, as is Done for Axes}
\usage{
rangeExtended(x, extend = 0.04)
}
\arguments{
\item{x}{a numeric vector.}

\item{extend}{fraction to extend on either end}
}
\value{
A two-element vector with the extended range of \code{x}.
}
\description{
This is analogous to what is done as part of the R axis range calculation,
in the case where \code{xaxs="r"}.
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{prettyPosition}
\alias{prettyPosition}
\title{Pretty lat/lon in deg, min, sec}
\usage{
prettyPosition(x, debug = getOption("oceDebug"))
}
\arguments{
\item{x}{a series of one or more values of a latitude or longitude, in
decimal degrees}

\item{debug}{set to a positive value to get debugging information during
processing.}
}
\value{
A vector of numbers that will yield good axis labels if
\code{\link[=formatPosition]{formatPosition()}} is used.
}
\description{
Round a geographical positions in degrees, minutes, and seconds
Depending on the range of values in \code{x}, rounding is done to degrees,
half-degrees, minutes, etc.
}
\examples{
library(oce)
formatPosition(prettyPosition(10+1:10/60+2.8/3600))
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{oceColorsVorticity}
\alias{oceColorsVorticity}
\alias{oce.colorsVorticity}
\title{Create colors suitable for vorticity fields}
\usage{
oceColorsVorticity(n)
}
\arguments{
\item{n}{number of colors to create.}
}
\value{
A vector of color specifications.
}
\description{
Create a set of colors for displaying vorticity values,
based on the scheme devised by Thyng et al. (2016) and
presented in a python package by Thyng (2019).
The color specifications were transliterated from
python to R on 2015-09-29, but have not been adjusted
since, even though the python source has changed.  This
is to prevent breaking old \code{oce} code.  To get the
latest versions of these colours or other colours,
use the \CRANpkg{cmocean} R package
(Thyng, Richards, and Krylov, 2019) directly, as is
illustrated (with the "matter" scheme) in Example 2.
Note that the \CRANpkg{cmocean} core functions provide a way
to select between various versions of the colour schemes.
It is also worth considering the palettes provided by the
\CRANpkg{viridis} package, as illustrated (with the
"inferno" scheme) in Example 3.
}
\examples{
library(oce)

# Example 1
imagep(volcano, col=oceColorsVorticity(128),
       zlab="oceColorsVorticity")
\dontrun{
# Example 2 (requires the cmocean package)
imagep(volcano, col=cmocean::cmocean("matter"),
       zlab='cmocean::cmocean("matter")')}

\dontrun{
# Example 3 (requires the viridis package)
imagep(volcano, col=viridis::inferno,
       zlab='viridis::inferno')}

}
\references{
\itemize{
\item Thyng, Kristen, Chad Greene, Robert Hetland, Heather Zimmerle, and Steven DiMarco.
\dQuote{True Colors of Oceanography: Guidelines for Effective and Accurate Colormap Selection.}
Oceanography 29, no. 3 (September 1, 2016): 9‚Äì13.
\doi{10.5670/oceanog.2016.66}
\item Thyng, Kristen. Kthyng/Cmocean. Python, 2019.
\url{https://github.com/kthyng/cmocean}.
\item Thyng, Kristen, Clark Richards, and Ivan Krylov.
Cmocean: Beautiful Colour Maps for Oceanography (version 0.2), 2019.
\url{https://CRAN.R-project.org/package=cmocean}.
}

The following references provide information on choosing
colour schemes, that are suitable for viewers who have
colour deficiencies.

Light, Adam, and Patrick J. Bartlein. "The End of the Rainbow? Color
Schemes for Improved Data Graphics." \emph{Eos, Transactions American Geophysical
Union} 85, no. 40 (2004): 385.
DOI: 10.1029/2004EO400002

Stephenson, David B. "Comment on 'Color Schemes for Improved Data
Graphics', by A Light and P.J. Bartlein." \emph{Eos, Transactions American
Geophysical Union} 86, no.  20 (2005): 196.
DOI: 10.1029/2005EO200005

Light, Adam, and Patrick J. Bartlein. "Reply to 'Comment on Color
Schemes for Improved Data Graphics,' by A. Light and P.J.  Bartlein'."
\emph{Eos, Transactions American Geophysical Union} 86, no. 20 (2005): 196‚Äì196.
DOI: 10.1029/2005EO200006
}
\seealso{
Other things related to colors: 
\code{\link{colormapGMT}()},
\code{\link{colormap}()},
\code{\link{oceColors9B}()},
\code{\link{oceColorsCDOM}()},
\code{\link{oceColorsChlorophyll}()},
\code{\link{oceColorsClosure}()},
\code{\link{oceColorsDensity}()},
\code{\link{oceColorsFreesurface}()},
\code{\link{oceColorsGebco}()},
\code{\link{oceColorsJet}()},
\code{\link{oceColorsOxygen}()},
\code{\link{oceColorsPAR}()},
\code{\link{oceColorsPalette}()},
\code{\link{oceColorsPhase}()},
\code{\link{oceColorsSalinity}()},
\code{\link{oceColorsTemperature}()},
\code{\link{oceColorsTurbidity}()},
\code{\link{oceColorsTurbo}()},
\code{\link{oceColorsTwo}()},
\code{\link{oceColorsVelocity}()},
\code{\link{oceColorsViridis}()},
\code{\link{ocecolors}}
}
\author{
Krysten M. Thyng (Python version), Dan Kelley (R transliteration)
}
\concept{things related to colors}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/map.R
\name{shiftLongitude}
\alias{shiftLongitude}
\title{Shift Longitude to Range -180 to 180}
\usage{
shiftLongitude(longitudes)
}
\arguments{
\item{longitudes}{numerical vector of longitudes.}
}
\value{
vector of longitudes, shifted to the desired range.
}
\description{
This is a utility function used by \code{\link[=mapGrid]{mapGrid()}}. It works
simply by subtracting 180 from each longitude, if any longitude
in the vector exceeds 180.
}
\seealso{
\code{\link[=matrixShiftLongitude]{matrixShiftLongitude()}} and \code{\link[=standardizeLongitude]{standardizeLongitude()}}.

Other functions related to maps: 
\code{\link{formatPosition}()},
\code{\link{lonlat2map}()},
\code{\link{lonlat2utm}()},
\code{\link{map2lonlat}()},
\code{\link{mapArrows}()},
\code{\link{mapAxis}()},
\code{\link{mapContour}()},
\code{\link{mapCoordinateSystem}()},
\code{\link{mapDirectionField}()},
\code{\link{mapGrid}()},
\code{\link{mapImage}()},
\code{\link{mapLines}()},
\code{\link{mapLocator}()},
\code{\link{mapLongitudeLatitudeXY}()},
\code{\link{mapPlot}()},
\code{\link{mapPoints}()},
\code{\link{mapPolygon}()},
\code{\link{mapScalebar}()},
\code{\link{mapText}()},
\code{\link{mapTissot}()},
\code{\link{oceCRS}()},
\code{\link{usrLonLat}()},
\code{\link{utm2lonlat}()}
}
\concept{functions related to maps}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{interpBarnes}
\alias{interpBarnes}
\title{Grid data using Barnes algorithm}
\usage{
interpBarnes(
  x,
  y,
  z,
  w,
  xg,
  yg,
  xgl,
  ygl,
  xr,
  yr,
  gamma = 0.5,
  iterations = 2,
  trim = 0,
  pregrid = FALSE,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{x, y}{a vector of x and y locations.}

\item{z}{a vector of z values, one at each (x,y) location.}

\item{w}{a optional vector of weights at the (x,y) location.  If not
supplied, then a weight of 1 is used for each point, which means equal
weighting.  Higher weights give data points more influence. If \code{pregrid}
is \code{TRUE}, then any supplied value of \code{w} is ignored, and instead
each of the pregriddd points is given equal weight.}

\item{xg, yg}{optional vectors defining the x and y grids.  If not supplied,
these values are inferred from the data, using e.g. \code{pretty(x, n=50)}.}

\item{xgl, ygl}{optional lengths of the x and y grids, to be constructed with
\code{\link[=seq]{seq()}} spanning the data range.  These values \code{xgl} are only
examined if \code{xg} and \code{yg} are not supplied.}

\item{xr, yr}{optional values defining the x and y radii of the weighting ellipse.
If not supplied, these are calculated as the span of x
and y over the square root of the number of data.}

\item{gamma}{grid-focussing parameter.  At each successive iteration, \code{xr} and
\code{yr} are reduced by a factor of \code{sqrt(gamma)}.}

\item{iterations}{number of iterations.  Set this to 1 to perform just
one iteration, using the radii as described at \verb{xr,yr} above.}

\item{trim}{a number between 0 and 1, indicating the quantile of data weight
to be used as a criterion for blanking out the gridded value (using
\code{NA}).  If 0, the whole \code{zg} grid is returned.  If >0, any spots
on the grid where the data weight is less than the \code{trim}-th
\code{\link[=quantile]{quantile()}} are set to \code{NA}.  See examples.}

\item{pregrid}{an indication of whether to pre-grid the data. If
\code{FALSE}, this is not done, i.e. conventional Barnes interpolation is
performed.  Otherwise, then the data are first averaged within grid cells
using \code{\link[=binMean2D]{binMean2D()}}.  If \code{pregrid} is \code{TRUE} or
\code{4}, then this averaging is done within a grid that is 4 times finer
than the grid that will be used for the Barnes interpolation. Otherwise,
\code{pregrid} may be a single integer indicating the grid refinement (4
being the result if \code{TRUE} had been supplied), or a vector of two
integers, for the grid refinement in x and y. The purpose of using
\code{pregrid} is to speed processing on large datasets, and to remove
spatial bias (e.g. with a single station that is repeated frequently in an
otherwise seldom-sampled region).  A form of pregridding is done in the
World Ocean Atlas, for example.}

\item{debug}{a flag that turns on debugging.  Set to 0 for no debugging
information, to 1 for more, etc; the value is reduced by 1 for each
descendent function call.}
}
\value{
A list containing: \code{xg}, a vector holding the x-grid);
\code{yg}, a vector holding the y-grid; \code{zg}, a matrix holding the
gridded values; \code{wg}, a matrix holding the weights used in the
interpolation at its final iteration; and \code{zd}, a vector of the same
length as \code{x}, which holds the interpolated values at the data points.
}
\description{
The algorithm follows that described by Koch et al. (1983), except
that \code{interpBarnes} adds (1) the ability to
blank out the grid where data are
sparse, using the \code{trim} argument, and (2) the ability to
pre-grid, with the \code{pregrid} argument.
}
\examples{
library(oce)

# 1. contouring example, with wind-speed data from Koch et al. (1983)
data(wind)
u <- interpBarnes(wind$x, wind$y, wind$z)
contour(u$xg, u$yg, u$zg, labcex=1)
text(wind$x, wind$y, wind$z, cex=0.7, col="blue")
title("Numbers are the data")

# 2. As 1, but blank out spots where data are sparse
u <- interpBarnes(wind$x, wind$y, wind$z, trim=0.1)
contour(u$xg, u$yg, u$zg, level=seq(0, 30, 1))
points(wind$x, wind$y, cex=1.5, pch=20, col="blue")

# 3. As 1, but interpolate back to points, and display the percent mismatch
u <- interpBarnes(wind$x, wind$y, wind$z)
contour(u$xg, u$yg, u$zg, labcex=1)
mismatch <- 100 * (wind$z - u$zd) / wind$z
text(wind$x, wind$y, round(mismatch), col="blue")
title("Numbers are percent mismatch between grid and data")

# 4. As 3, but contour the mismatch
mismatchGrid <- interpBarnes(wind$x, wind$y, mismatch)
contour(mismatchGrid$xg, mismatchGrid$yg, mismatchGrid$zg, labcex=1)

# 5. One-dimensional example, smoothing a salinity profile
data(ctd)
p <- ctd[["pressure"]]
y <- rep(1, length(p)) # fake y data, with arbitrary value
S <- ctd[["salinity"]]
pg <- pretty(p, n=100)
g <- interpBarnes(p, y, S, xg=pg, xr=1)
plot(S, p, cex=0.5, col="blue", ylim=rev(range(p)))
lines(g$zg, g$xg, col="red")
}
\references{
S. E.  Koch and M.  DesJardins and P. J. Kocin, 1983.  ``An
interactive Barnes objective map analysis scheme for use with satellite and
conventional data,'' \emph{J.  Climate Appl.  Met.}, vol 22, p. 1487-1503.
}
\seealso{
See \code{\link[=wind]{wind()}}.
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{oceColorsTwo}
\alias{oceColorsTwo}
\alias{oce.colorsTwo}
\title{Create two-color palette}
\usage{
oceColorsTwo(n, low = 2/3, high = 0, smax = 1, alpha = 1)
}
\arguments{
\item{n}{number of colors to generate.}

\item{low, high}{numerical values (in range 0 to 1) specifying the hue
for the low and high ends of the color scale.}

\item{smax}{numerical value (in range 0 to 1) for the color saturation.}

\item{alpha}{numerical value (in ragne 0 to 1) for the alpha (transparency)
of the colors.}
}
\description{
Create colors ranging between two specified limits, with white
in the middle.
}
\examples{
library(oce)
imagep(volcano-mean(range(volcano)), col=oceColorsTwo(128),
       zlim="symmetric", zlab="oceColorsTwo")
}
\seealso{
Other things related to colors: 
\code{\link{colormapGMT}()},
\code{\link{colormap}()},
\code{\link{oceColors9B}()},
\code{\link{oceColorsCDOM}()},
\code{\link{oceColorsChlorophyll}()},
\code{\link{oceColorsClosure}()},
\code{\link{oceColorsDensity}()},
\code{\link{oceColorsFreesurface}()},
\code{\link{oceColorsGebco}()},
\code{\link{oceColorsJet}()},
\code{\link{oceColorsOxygen}()},
\code{\link{oceColorsPAR}()},
\code{\link{oceColorsPalette}()},
\code{\link{oceColorsPhase}()},
\code{\link{oceColorsSalinity}()},
\code{\link{oceColorsTemperature}()},
\code{\link{oceColorsTurbidity}()},
\code{\link{oceColorsTurbo}()},
\code{\link{oceColorsVelocity}()},
\code{\link{oceColorsViridis}()},
\code{\link{oceColorsVorticity}()},
\code{\link{ocecolors}}
}
\concept{things related to colors}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/extdata.R
\docType{data}
\encoding{UTF-8}
\name{test_met_csv2.csv}
\alias{test_met_csv2.csv}
\title{Sample met dataset (CSV format as of October 2019)}
\description{
This is a subset of a file downloaded with \code{\link[=download.met]{download.met()}}
from the Environment Canada "Climate Data"
website \verb{https://climate.weather.gc.ca/index_e.html}
on October 13, 2019.
}
\examples{
\dontrun{
read.met(system.file("extdata", "test_met_csv2.csv", package="oce"))
}

}
\seealso{
Other raw datasets: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{adp_rdi.000}},
\code{\link{ctd.cnv}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{test_met_csv1.csv}},
\code{\link{test_met_xml2.xml}},
\code{\link{xbt.edf}}

Other things related to met data: 
\code{\link{[[,met-method}},
\code{\link{[[<-,met-method}},
\code{\link{as.met}()},
\code{\link{download.met}()},
\code{\link{met-class}},
\code{\link{met}},
\code{\link{plot,met-method}},
\code{\link{read.met}()},
\code{\link{subset,met-method}},
\code{\link{summary,met-method}},
\code{\link{test_met_csv1.csv}},
\code{\link{test_met_xml2.xml}}
}
\concept{raw datasets}
\concept{things related to met data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/xbt.R
\name{read.xbt.edf}
\alias{read.xbt.edf}
\title{Read a Sippican '.edf' format xbt file}
\usage{
read.xbt.edf(
  file,
  longitude = NA,
  latitude = NA,
  debug = getOption("oceDebug"),
  processingLog
)
}
\arguments{
\item{file}{a connection or a character string giving the name of the file to
load.}

\item{longitude}{optional signed number indicating the longitude in degrees
East.}

\item{latitude}{optional signed number indicating the latitude in degrees North.}

\item{debug}{a flag that turns on debugging.  The value indicates the depth
within the call stack to which debugging applies.}

\item{processingLog}{if provided, the action item to be stored in the log.  This
parameter is typically only provided for internal calls; the default that it
provides is better for normal calls by a user.}
}
\value{
An \linkS4class{xbt} object.
}
\description{
The function was written by inspection of a particular file, and might
be wrong for other files; see \dQuote{Details} for a note on character
translation.
}
\details{
The header is converted to ASCII format prior to storage in
the \code{metadata} slot, so that e.g. a degree sign in the original file will
become a \verb{?} character in the \code{header}.  This is to prevent problems
with submission of \code{oce} to the CRAN system, which produces NOTEs
about UTF-8 strings in data (on some build machines, evidently depending
on the locale on those machines).  This character substitution
is at odds with the \code{oce} philosophy of leaving data intact, so
it will be reverted, if CRAN policy changes or if the developers
can find a way to otherwise silence the NOTE.
}
\examples{
library(oce)
xbt <- read.oce(system.file("extdata", "xbt.edf", package="oce"))
summary(xbt)
plot(xbt)

}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/imagep.R
\name{drawPalette}
\alias{drawPalette}
\title{Draw a palette, leaving margins suitable for accompanying plot}
\usage{
drawPalette(
  zlim,
  zlab = "",
  breaks,
  col,
  colormap,
  mai,
  cex = par("cex"),
  pos = 4,
  las = 0,
  labels = NULL,
  at = NULL,
  levels,
  drawContours = FALSE,
  plot = TRUE,
  fullpage = FALSE,
  drawTriangles = FALSE,
  axisPalette,
  tformat,
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{zlim}{two-element vector containing the lower and upper limits of z.
This may also be a vector of any length exceeding 1, in which case its range
is used.}

\item{zlab}{label for the palette scale.}

\item{breaks}{optional numeric vector of the z values for breaks in the color scheme.
If \code{colormap} is provided, it takes precedence over \code{breaks} and \code{col}.}

\item{col}{optional argument, either a vector of colors corresponding to the breaks, of length
1 less than the number of breaks, or a function specifying colors.
If \code{col} is not provided, and if \code{colormap} is also not provided, then \code{col}
defaults to \code{\link[=oceColorsViridis]{oceColorsViridis()}}.
If \code{colormap} is provided, it takes precedence over \code{breaks} and \code{col}.}

\item{colormap}{an optional color map as created by \code{\link[=colormap]{colormap()}}.
If \code{colormap} is provided, it takes precedence over \code{breaks} and \code{col}.}

\item{mai}{margins for palette, as defined in the usual way; see
\code{\link[=par]{par()}}.  If not given, reasonable values are inferred from the
existence of a non-blank \code{zlab}.}

\item{cex}{numeric character expansion value for text labels}

\item{pos}{an integer indicating the location of the palette within the
plotting area, 1 for near the bottom, 2 for near the left-hand side, 3 for
near the top side, and 4 (the default) for near the right-hand side.}

\item{las}{optional argument, passed to \code{\link[=axis]{axis()}}, to control the orientation
of numbers along the axis. As explained in the help for \code{\link[=par]{par()}}, the
meaning of \code{las} is as follows: \code{las=0} (the default) means to put labels
parallel to the axis, \code{las=1} means horizontal (regardless of
axis orientation), \code{las=2} means perpendicular to the axis,
and \code{las=3} means to vertical (regardless of axis orientation).  Note that
the automatic computation of margin spacing parameter \code{mai}
assumes that \code{las=0}, and so for other cases, the user may need to
specify the \code{mai} argument directly.}

\item{labels}{optional vector of labels for ticks on palette axis (must
correspond with \code{at})}

\item{at}{optional vector of positions for the \code{label}s}

\item{levels}{optional contour levels, in preference to \code{breaks}
values, to be added to the image if \code{drawContours} is \code{TRUE}.}

\item{drawContours}{logical value indicating whether to draw contours on the
palette, at the color breaks.}

\item{plot}{logical value indicating whether to plot the palette, the
default, or whether to just alter the margins to make space for where the
palette would have gone.  The latter case may be useful in lining up plots,
as in example 1 of \dQuote{Examples}.}

\item{fullpage}{logical value indicating whether to draw the palette filling
the whole plot width (apart from \code{mai}, of course).  This can be
helpful if the palette panel is to be created with \code{\link[=layout]{layout()}}, as
illustrated in the \dQuote{Examples}.}

\item{drawTriangles}{logical value indicating whether to draw triangles on
the top and bottom of the palette.  If a single value is provided, it applies
to both ends of the palette.  If a pair is provided, the first refers to the
lower range of the palette, and the second to the upper range.}

\item{axisPalette}{optional replacement function for \code{axis()}, e.g.
for exponential notation on large or small values.}

\item{tformat}{optional format for axis labels, if the variable is a time
type (ignored otherwise).}

\item{debug}{a flag that turns on debugging.  Set to 1 to get a moderate
amount of debugging information, or to 2 to get more.}

\item{\dots}{optional arguments passed to plotting functions.}
}
\value{
None.
}
\description{
In the normal use, \code{\link[=drawPalette]{drawPalette()}} draws an image palette near the
right-hand side of the plotting device, and then adjusts the global margin
settings in such a way as to cause the next plot to appear (with much larger
width) to the left of the palette. The function can also be used, if
\code{zlim} is not provided, to adjust the margin without drawing anything;
this is useful in lining up the x axes of a stack of plots, some some of
which will have palettes and others not.
}
\details{
The plot positioning is done entirely with margins, not with
\code{par(mfrow)} or other R schemes for multi-panel plots.  This means that
the user is free to use those schemes without worrying about nesting or
conflicts.
}
\section{Use with multi-panel plots}{
 An important consequence of the margin
adjustment is that multi-panel plots require that the initial margin be
stored prior to the first call to \code{\link[=drawPalette]{drawPalette()}}, and reset after each
palette-plot pair.  This method is illustrated in \dQuote{Examples}.
}

\examples{

library(oce)
par(mgp=getOption("oceMgp"))

## 1. A three-panel plot
par(mfrow=c(3, 1), mar=c(3, 3, 1, 1))
omar <- par('mar')                 # save initial margin

## 1a. top panel: simple case with Viridis scheme
drawPalette(zlim=c(0, 1), col=oce.colorsViridis(10))
plot(1:10, 1:10, col=oce.colorsViridis(10)[1:10],pch=20,cex=3,xlab='x', ylab='y')
par(mar=omar)                      # reset margin

## 1b. middle panel: colormap
cm <- colormap(name="gmt_globe")
drawPalette(colormap=cm)
icol <- seq_along(cm$col)
plot(icol, cm$breaks[icol], pch=20, cex=2, col=cm$col,
     xlab="Palette index", ylab="Palette breaks")
par(mar=omar)                      # reset margin

## 1c. bottom panel: space for palette (to line up graphs)
drawPalette(plot=FALSE)
plot(1:10, 1:10, col=oce.colorsViridis(10)[1:10],pch=20,cex=3,xlab='x', ylab='y')
par(mar=omar)                      # reset margin

# 2. Use layout to mimic the action of imagep(), with the width
# of the palette region being 14 percent of figure width.
d <- 0.14
layout(matrix(1:2,nrow=1), widths=c(1-d,d))
image(volcano, col=oce.colorsViridis(100), zlim=c(90, 200))
contour(volcano, add=TRUE)
drawPalette(c(90, 200), fullpage=TRUE, col=oce.colorsViridis)
}
\seealso{
This is used by \code{\link[=imagep]{imagep()}}.
}
\author{
Dan Kelley, with help from Clark Richards
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{plotInset}
\alias{plotInset}
\title{Plot an Inset Diagram}
\usage{
plotInset(
  xleft,
  ybottom,
  xright,
  ytop,
  expr,
  mar = c(2, 2, 1, 1),
  debug = getOption("oceDebug")
)
}
\arguments{
\item{xleft}{location of left-hand of the inset diagram, in the existing
plot units.  (PROVISIONAL FEATURE: this may also be \code{"bottomleft"}, to
put the inset there.  Eventually, other positions may be added.)}

\item{ybottom}{location of bottom side of the inset diagram, in the existing
plot units.}

\item{xright}{location of right-hand side of the inset diagram, in the
existing plot units.}

\item{ytop}{location of top side of the inset diagram, in the existing plot
units.}

\item{expr}{An expression that draws the inset plot.  This may be a single
plot command, or a sequence of commands enclosed in curly braces.}

\item{mar}{margins, in line heights, to be used at the four sides of the
inset diagram.  (This is often helpful to save space.)}

\item{debug}{a flag that turns on debugging.  Set to 1 to get a moderate
amount of debugging information, or to 2 to get more.}
}
\description{
Adds an inset diagram to an existing plot.  Note that if the inset is a map
or coastline, it will be necessary to supply \code{inset=TRUE} to prevent
the inset diagram from occupying the whole device width.  After
\code{plotInset()} has been called, any further plotting will take place
within the inset, so it is essential to finish a plot before drawing an
inset.
}
\examples{
library(oce)
# power law in linear and log form
x <- 1:10
y <- x^2
plot(x, y, log='xy',type='l')
plotInset(3, 1, 10, 8,
          expr=plot(x,y,type='l',cex.axis=3/4,mgp=c(3/2, 1/2, 0)),
          mar=c(2.5, 2.5, 1, 1))

# CTD data with location
data(ctd)
plot(ctd, which="TS")
plotInset(29.9, 2.7, 31, 10,
          expr=plot(ctd, which='map',
          coastline="coastlineWorld",
          span=5000, mar=NULL, cex.axis=3/4))
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/section.R
\name{subset,section-method}
\alias{subset,section-method}
\alias{subset.section}
\title{Subset a Section Object}
\usage{
\S4method{subset}{section}(x, subset, ...)
}
\arguments{
\item{x}{a \linkS4class{section} object.}

\item{subset}{an optional indication of either the stations to be kept,
or the data to be kept within the stations.  See \dQuote{Details}.}

\item{...}{optional arguments, of which only the first is examined. The
possibilities for this argument are \code{indices}, which must be a
vector of station indices (see Example 6), or \code{within}, which must be
a list or data frame, containing items named either \code{x} and \code{y}
or \code{longitude} and \code{latitude} (see Example 7). If \code{within}
is given, then \code{subset} is ignored.}
}
\value{
A \linkS4class{section} object.
}
\description{
Return a subset of a section object.
}
\details{
This function is used to subset data within the
stations of a section, or to choose a subset of the stations
themselves. The first case is handled with the \code{subset} argument,
while the second is handled if \code{...} contains a vector named
\code{indices}. Either \code{subset} or \code{indices} must be
provided, but not both.

\strong{In the "subset" method}, \code{subset} indicates
either stations to be kept, or data to be kept within the stations.

The first step in processing is to check for the presence of certain
key words in the \code{subset} expression. If \code{distance}
is present, then stations are selected according to a condition on the
distance (in km) from the first station to the given station (Example 1).
If either \code{longitude} or \code{latitude} is given, then
stations are selected according to the stated condition (Example 2).
If \code{stationId} is present, then selection is in terms of the
station ID (not the sequence number) is used (Example 3). In all of these
cases, stations are either selected in their entirety or dropped
in their entirety.

If none of these keywords is present, then the \code{subset} expression is
evaluated in the context of the \code{data} slot of each of the CTD stations
stored within \code{x}. (Note that this slot does not normally
contain any of the keywords that are listed in the previous
paragraph; it does, then odd results may occur.) Each station is examined
in turn, with \code{subset} being evaluated individually in each. The evaluation
produces a logical vector. If that vector has length 1 (Examples 4 and 5)
then the station is retained or discarded, accordingly. If the vector is longer,
then the logical vector is used as a sieve to subsample that individual CTD
profile.

\strong{In the "indices" method}, stations are selected using
\code{indices}, which may be a vector of integers that indicate sequence
number, or a logical vector, again indicating which stations to keep.
}
\examples{
library(oce)
data(section)

# Example 1. Stations within 500 km of the first station
starting <- subset(section, distance < 500)

# Example 2. Stations east of 50W
east <- subset(section, longitude > (-50))

# Example 3. Gulf Stream
GS <- subset(section, 109 <= stationId & stationId <= 129)

# Example 4. Only stations with more than 5 pressure levels
long <- subset(section, length(pressure) > 5)

# Example 5. Only stations that have some data in top 50 dbar
surfacing <- subset(section, min(pressure) < 50)

# Example 6. Similar to #4, but done in more detailed way
long <- subset(section,
   indices=unlist(lapply(section[["station"]],
                  function(s)
                    5 < length(s[["pressure"]]))))

# Example 7. Subset by a polygon determined with locator()
\dontrun{
par(mfrow=c(2, 1))
plot(section, which="map")
bdy <- locator(4) # choose a polygon near N. America
GS <- subset(section, within=bdy)
plot(GS, which="map")
}

}
\seealso{
Other functions that subset oce objects: 
\code{\link{subset,adp-method}},
\code{\link{subset,adv-method}},
\code{\link{subset,amsr-method}},
\code{\link{subset,argo-method}},
\code{\link{subset,cm-method}},
\code{\link{subset,coastline-method}},
\code{\link{subset,ctd-method}},
\code{\link{subset,echosounder-method}},
\code{\link{subset,lobo-method}},
\code{\link{subset,met-method}},
\code{\link{subset,oce-method}},
\code{\link{subset,odf-method}},
\code{\link{subset,rsk-method}},
\code{\link{subset,sealevel-method}},
\code{\link{subset,topo-method}},
\code{\link{subset,xbt-method}}

Other things related to section data: 
\code{\link{[[,section-method}},
\code{\link{[[<-,section-method}},
\code{\link{as.section}()},
\code{\link{handleFlags,section-method}},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{plot,section-method}},
\code{\link{read.section}()},
\code{\link{section-class}},
\code{\link{sectionAddStation}()},
\code{\link{sectionGrid}()},
\code{\link{sectionSmooth}()},
\code{\link{sectionSort}()},
\code{\link{section}},
\code{\link{summary,section-method}}
}
\author{
Dan Kelley
}
\concept{functions that subset oce objects}
\concept{things related to section data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swTheta}
\alias{swTheta}
\title{Seawater potential temperature}
\usage{
swTheta(
  salinity,
  temperature = NULL,
  pressure = NULL,
  referencePressure = 0,
  longitude = NULL,
  latitude = NULL,
  eos = getOption("oceEOS", default = "gsw")
)
}
\arguments{
\item{salinity}{either salinity (PSU) (in which case \code{temperature} and
\code{pressure} must be provided) \emph{or} an \code{oce} object (in which
case \code{salinity}, etc. are inferred from the object).}

\item{temperature}{\emph{in-situ} temperature (\eqn{^\circ}{deg}C), defined
on the ITS-90 scale; see \dQuote{Temperature units} in the documentation for
\code{\link[=swRho]{swRho()}}, and the examples below.}

\item{pressure}{pressure (dbar)}

\item{referencePressure}{reference pressure (dbar)}

\item{longitude}{longitude of observation (only used if \code{eos="gsw"};
see \sQuote{Details}).}

\item{latitude}{latitude of observation (only used if \code{eos="gsw"}; see
\sQuote{Details}).}

\item{eos}{equation of state, either \code{"unesco"} (references 1 and 2) or \code{"gsw"}
(references 3 and 4).}
}
\value{
Potential temperature (\eqn{^\circ}{deg}C) of seawater, referenced
to pressure \code{referencePressure}.
}
\description{
Compute the potential temperature of seawater, denoted \eqn{\theta}{theta}
in the UNESCO system, and \code{pt} in the GSW system.
}
\details{
Different formulae are used depending on the equation of state. If \code{eos}
is \code{"unesco"}, the method of Fofonoff \emph{et al.} (1983) is used
(see references 1 and 2).
Otherwise, \code{swTheta} uses \code{\link[gsw:gsw_pt_from_t]{gsw::gsw_pt_from_t()}} from
the \CRANpkg{gsw} package.

If the first argument is a \code{ctd} or \code{section} object, then values
for salinity, etc., are extracted from it, and used for the calculation, and
the corresponding arguments to the present function are ignored.
}
\examples{
library(oce)
## test value from Fofonoff et al., 1983
stopifnot(abs(36.8818748026 - swTheta(40, T90fromT68(40), 10000, 0, eos="unesco")) < 0.0000000001)

# Example from a cross-Atlantic section
data(section)
stn <- section[['station', 70]]
plotProfile(stn, 'theta', ylim=c(6000, 1000))
lines(stn[['temperature']], stn[['pressure']], lty=2)
legend("bottomright", lty=1:2,
       legend=c("potential", "in-situ"),
       bg='white', title="Station 70")

}
\references{
\enumerate{
\item Fofonoff, P. and R. C. Millard Jr, 1983. Algorithms for computation of
fundamental properties of seawater.
\emph{Unesco Technical Papers in Marine Science}, \emph{44}, 53 pp
\item Gill, A.E., 1982. \emph{Atmosphere-ocean Dynamics}, Academic Press, New
York, 662 pp.
\item IOC, SCOR, and IAPSO (2010). The international thermodynamic equation of
seawater-2010: Calculation and use of thermodynamic properties.  Technical
Report 56, Intergovernmental Oceanographic Commission, Manuals and Guide.
\item McDougall, T.J. and P.M. Barker, 2011: Getting started with TEOS-10 and
the Gibbs Seawater (GSW) Oceanographic Toolbox, 28pp., SCOR/IAPSO WG127,
ISBN 978-0-646-55621-5.
}
}
\seealso{
Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/lisst.R
\name{as.lisst}
\alias{as.lisst}
\title{Coerce Data Into a LISST Object}
\usage{
as.lisst(
  data,
  filename = "",
  year = 0,
  tz = "UTC",
  longitude = NA,
  latitude = NA
)
}
\arguments{
\item{data}{A table (or matrix) containing 42 columns, as in a LISST data
file.}

\item{filename}{Name of file containing the data.}

\item{year}{Year in which the first observation was made.  This is necessary
because LISST timestamps do not indicate the year of observation.  The
default value is odd enough to remind users to include this argument.}

\item{tz}{Timezone of observations.  This is necessary because LISST
timestamps do not indicate the timezone.}

\item{longitude}{Longitude of observation.}

\item{latitude}{Latitude of observation.}
}
\value{
A \linkS4class{lisst} object.
}
\description{
Coerce data into a lisst object
If \code{data} contains fewer than 42 columns, an error is reported.  If it
contains more than 42 columns, only the first 42 are used.  This is used by
\code{\link[=read.lisst]{read.lisst()}}, the documentation on which explains the meanings
of the columns.
}
\seealso{
Other things related to lisst data: 
\code{\link{[[,lisst-method}},
\code{\link{[[<-,lisst-method}},
\code{\link{lisst-class}},
\code{\link{plot,lisst-method}},
\code{\link{read.lisst}()},
\code{\link{summary,lisst-method}}
}
\author{
Dan Kelley
}
\concept{things related to lisst data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/coastline.R
\name{summary,coastline-method}
\alias{summary,coastline-method}
\alias{summary.coastline}
\title{Summarize a Coastline Object}
\usage{
\S4method{summary}{coastline}(object, ...)
}
\arguments{
\item{object}{a \linkS4class{coastline} object.}

\item{\dots}{further arguments passed to or from other methods.}
}
\description{
Summarizes coastline length, bounding box, etc.
}
\seealso{
Other things related to coastline data: 
\code{\link{[[,coastline-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{as.coastline}()},
\code{\link{coastline-class}},
\code{\link{coastlineBest}()},
\code{\link{coastlineCut}()},
\code{\link{coastlineWorld}},
\code{\link{download.coastline}()},
\code{\link{plot,coastline-method}},
\code{\link{read.coastline.openstreetmap}()},
\code{\link{read.coastline.shapefile}()},
\code{\link{subset,coastline-method}}
}
\author{
Dan Kelley
}
\concept{things related to coastline data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{oceColorsChlorophyll}
\alias{oceColorsChlorophyll}
\alias{oce.colorsChlorophyll}
\title{Create colors suitable for chlorophyll fields}
\usage{
oceColorsChlorophyll(n)
}
\arguments{
\item{n}{number of colors to create.}
}
\value{
A vector of color specifications.
}
\description{
Create a set of colors for displaying chlorophyll values,
based on the scheme devised by Thyng et al. (2016) and
presented in a python package by Thyng (2019).
The color specifications were transliterated from
python to R on 2015-09-29, but have not been adjusted
since, even though the python source has changed.  This
is to prevent breaking old \code{oce} code.  To get the
latest versions of these colours or other colours,
use the \CRANpkg{cmocean} R package
(Thyng, Richards, and Krylov, 2019) directly, as is
illustrated (with the "matter" scheme) in Example 2.
Note that the \CRANpkg{cmocean} core functions provide a way
to select between various versions of the colour schemes.
It is also worth considering the palettes provided by the
\CRANpkg{viridis} package, as illustrated (with the
"inferno" scheme) in Example 3.
}
\examples{
library(oce)

# Example 1
imagep(volcano, col=oceColorsChlorophyll(128),
       zlab="oceColorsChlorophyll")
\dontrun{
# Example 2 (requires the cmocean package)
imagep(volcano, col=cmocean::cmocean("matter"),
       zlab='cmocean::cmocean("matter")')}

\dontrun{
# Example 3 (requires the viridis package)
imagep(volcano, col=viridis::inferno,
       zlab='viridis::inferno')}

}
\references{
\itemize{
\item Thyng, Kristen, Chad Greene, Robert Hetland, Heather Zimmerle, and Steven DiMarco.
\dQuote{True Colors of Oceanography: Guidelines for Effective and Accurate Colormap Selection.}
Oceanography 29, no. 3 (September 1, 2016): 9‚Äì13.
\doi{10.5670/oceanog.2016.66}
\item Thyng, Kristen. Kthyng/Cmocean. Python, 2019.
\url{https://github.com/kthyng/cmocean}.
\item Thyng, Kristen, Clark Richards, and Ivan Krylov.
Cmocean: Beautiful Colour Maps for Oceanography (version 0.2), 2019.
\url{https://CRAN.R-project.org/package=cmocean}.
}

The following references provide information on choosing
colour schemes, that are suitable for viewers who have
colour deficiencies.

Light, Adam, and Patrick J. Bartlein. "The End of the Rainbow? Color
Schemes for Improved Data Graphics." \emph{Eos, Transactions American Geophysical
Union} 85, no. 40 (2004): 385.
DOI: 10.1029/2004EO400002

Stephenson, David B. "Comment on 'Color Schemes for Improved Data
Graphics', by A Light and P.J. Bartlein." \emph{Eos, Transactions American
Geophysical Union} 86, no.  20 (2005): 196.
DOI: 10.1029/2005EO200005

Light, Adam, and Patrick J. Bartlein. "Reply to 'Comment on Color
Schemes for Improved Data Graphics,' by A. Light and P.J.  Bartlein'."
\emph{Eos, Transactions American Geophysical Union} 86, no. 20 (2005): 196‚Äì196.
DOI: 10.1029/2005EO200006
}
\seealso{
Other things related to colors: 
\code{\link{colormapGMT}()},
\code{\link{colormap}()},
\code{\link{oceColors9B}()},
\code{\link{oceColorsCDOM}()},
\code{\link{oceColorsClosure}()},
\code{\link{oceColorsDensity}()},
\code{\link{oceColorsFreesurface}()},
\code{\link{oceColorsGebco}()},
\code{\link{oceColorsJet}()},
\code{\link{oceColorsOxygen}()},
\code{\link{oceColorsPAR}()},
\code{\link{oceColorsPalette}()},
\code{\link{oceColorsPhase}()},
\code{\link{oceColorsSalinity}()},
\code{\link{oceColorsTemperature}()},
\code{\link{oceColorsTurbidity}()},
\code{\link{oceColorsTurbo}()},
\code{\link{oceColorsTwo}()},
\code{\link{oceColorsVelocity}()},
\code{\link{oceColorsViridis}()},
\code{\link{oceColorsVorticity}()},
\code{\link{ocecolors}}
}
\author{
Krysten M. Thyng (Python version), Dan Kelley (R transliteration)
}
\concept{things related to colors}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/echosounder.R
\name{plot,echosounder-method}
\alias{plot,echosounder-method}
\alias{plot.echosounder}
\title{Plot an echosounder Object}
\usage{
\S4method{plot}{echosounder}(
  x,
  which = 1,
  beam = "a",
  newx,
  xlab,
  ylab,
  xlim,
  ylim,
  zlim,
  type = "l",
  col = oceColorsJet,
  lwd = 2,
  despike = FALSE,
  drawBottom,
  ignore = 5,
  drawTimeRange = FALSE,
  drawPalette = TRUE,
  radius,
  coastline,
  mgp = getOption("oceMgp"),
  mar = c(mgp[1], mgp[1] + 1.5, mgp[2] + 1/2, 1/2),
  atTop,
  labelsTop,
  tformat,
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{x}{an \linkS4class{echosounder} object.}

\item{which}{list of desired plot types: \code{which=1} or \code{which="zt
image"} gives a z-time image, \code{which=2} or \code{which="zx image"}
gives a z-distance image, and \code{which=3} or \code{which="map"} gives a
map showing the cruise track.  In the image plots, the display is of
\code{\link[=log10]{log10()}} of amplitude, trimmed to zero for any amplitude values
less than 1 (including missing values, which equal 0).  Add 10 to the
numeric codes to get the secondary data (non-existent for single-beam files,}

\item{beam}{a more detailed specification of the data to be plotted.  For
single-beam data, this may only be \code{"a"}.  For dual-beam data, this may
be \code{"a"} for the narrow-beam signal, or \code{"b"} for the wide-beam
signal.  For split-beam data, this may be \code{"a"} for amplitude,
\code{"b"} for x-angle data, or \code{"c"} for y-angle data.}

\item{newx}{optional vector of values to appear on the horizontal axis if
\code{which=1}, instead of time.  This must be of the same length as the
time vector, because the image is remapped from time to \code{newx} using
\code{\link[=approx]{approx()}}.}

\item{xlab, ylab}{optional labels for the horizontal and vertical axes; if
not provided, the labels depend on the value of \code{which}.}

\item{xlim}{optional range for x axis.}

\item{ylim}{optional range for y axis.}

\item{zlim}{optional range for color scale.}

\item{type}{type of graph, \code{"l"} for line, \code{"p"} for points, or
\code{"b"} for both.}

\item{col}{color scale for image, a function}

\item{lwd}{line width (ignored if \code{type="p"})}

\item{despike}{remove vertical banding by using \code{\link[=smooth]{smooth()}} to
smooth across image columns, row by row.}

\item{drawBottom}{optional flag used for section images.  If \code{TRUE},
then the bottom is inferred as a smoothed version of the ridge of highest
image value, and data below that are grayed out after the image is drawn.
If \code{drawBottom} is a color, then that color is used, instead of
white.  The bottom is detected with \code{\link[=findBottom]{findBottom()}}, using the
\code{ignore} value described next.}

\item{ignore}{optional flag specifying the thickness in metres of a surface
region to be ignored during the bottom-detection process.  This is ignored
unless \code{drawBottom=TRUE}.}

\item{drawTimeRange}{if \code{TRUE}, the time range will be drawn at the
top.  Ignored except for \code{which=2}, i.e. distance-depth plots.}

\item{drawPalette}{if \code{TRUE}, the palette will be drawn.}

\item{radius}{radius to use for maps; ignored unless \code{which=3} or
\code{which="map"}.}

\item{coastline}{coastline to use for maps; ignored unless \code{which=3} or
\code{which="map"}.}

\item{mgp}{3-element numerical vector to use for \code{\link{par}}\code{("mgp")}, and also
for \code{\link{par}}\code{("mar")}, computed from this.  The default is tighter than the R
default, in order to use more space for the data and less for the axes.}

\item{mar}{value to be used with \code{\link{par}}\code{("mar")}.}

\item{atTop}{optional vector of time values, for labels at the top of the
plot produced with \code{which=2}.  If \code{labelsTop} is provided, then it
will hold the labels.  If \code{labelsTop} is not provided, the labels will
be constructed with the \code{\link[=format]{format()}} function, and these may be
customized by supplying a \code{format} in the \code{...} arguments.}

\item{labelsTop}{optional vector of character strings to be plotted above
the \code{atTop} times.  Ignored unless \code{atTop} was provided.}

\item{tformat}{optional argument passed to \code{\link[=imagep]{imagep()}}, for plot
types that call that function.  (See \code{\link[=strptime]{strptime()}} for the format
used.)}

\item{debug}{set to an integer exceeding zero, to get debugging information
during processing.}

\item{\dots}{optional arguments passed to plotting functions.  For example,
for maps, it is possible to specify the radius of the view in kilometres,
with \code{radius}.}
}
\value{
A list is silently returned, containing \code{xat} and \code{yat},
values that can be used by \code{\link[=oce.grid]{oce.grid()}} to add a grid to the plot.
}
\description{
Plot echosounder data.
Simple linear approximation is used when a \code{newx} value is specified
with the \code{which=2} method, but arguably a gridding method should be
used, and this may be added in the future.
}
\examples{
library(oce)
data(echosounder)
plot(echosounder, drawBottom=TRUE)

}
\seealso{
Other things related to echosounder data: 
\code{\link{[[,echosounder-method}},
\code{\link{[[<-,echosounder-method}},
\code{\link{as.echosounder}()},
\code{\link{echosounder-class}},
\code{\link{echosounder}},
\code{\link{findBottom}()},
\code{\link{read.echosounder}()},
\code{\link{subset,echosounder-method}},
\code{\link{summary,echosounder-method}}
}
\author{
Dan Kelley, with extensive help from Clark Richards
}
\concept{things related to echosounder data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{lowpass}
\alias{lowpass}
\title{Perform lowpass digital filtering}
\usage{
lowpass(x, filter = "hamming", n, replace = TRUE, coefficients = FALSE)
}
\arguments{
\item{x}{a vector to be smoothed}

\item{filter}{name of filter; at present, \code{"hamming"}, \code{"hanning"}, and \code{"boxcar"} are permitted.}

\item{n}{length of filter (must be an odd integer exceeding 1)}

\item{replace}{a logical value indicating whether points near the
ends of \code{x} should be copied into the end regions, replacing
the \code{NA} values that would otherwise be placed there by
\code{\link[stats:filter]{stats::filter()}}.}

\item{coefficients}{logical value indicating whether to return
the filter coefficients, instead of the filtered values. In accordance
with conventions in the literature, the returned values are not
normalized to sum to 1, although of course that normalization
is done in the actual filtering.}
}
\value{
By default, \code{lowpass} returns a filtered version
of \code{x}, but if \code{coefficients} is \code{TRUE} then it
returns the filter coefficients.
}
\description{
The filter coefficients are constructed using standard definitions,
and then \code{\link[stats:filter]{stats::filter()}} is
used to filter the data. This leaves \code{NA}
values within half the filter length of the ends of the time series, but
these may be replaced with the original \code{x} values, if the argument
\code{replace} is set to \code{TRUE}.
}
\section{Caution}{
 This function was added in June of 2017,
and it may be extended during the rest of 2017. New arguments
may appear between \code{n} and \code{replace}, so users are
advised to call this function with named arguments, not positional
arguments.
}

\examples{
library(oce)
par(mfrow=c(1, 2), mar=c(4, 4, 1, 1))
coef <- lowpass(n=5, coefficients=TRUE)
plot(-2:2, coef, ylim=c(0, 1), xlab="Lag", ylab="Coefficient")
x <- seq(-5, 5) + rnorm(11)
plot(1:11, x, type='o', xlab="time", ylab="x and X")
X <- lowpass(x, n=5)
lines(1:11, X, col=2)
points(1:11, X, col=2)
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.R
\name{enuToOtherAdp}
\alias{enuToOtherAdp}
\title{Convert ADP ENU to Rotated Coordinate}
\usage{
enuToOtherAdp(x, heading = 0, pitch = 0, roll = 0)
}
\arguments{
\item{x}{an \linkS4class{adp} object.}

\item{heading}{number or vector of numbers, giving the angle, in degrees, to
be added to the heading.  See \dQuote{Details}.}

\item{pitch}{as \code{heading} but for pitch.}

\item{roll}{as \code{heading} but for roll.}
}
\value{
An object with \code{data$v[,1:3,]} altered appropriately, and
\code{metadata$oce.coordinate} changed from \code{enu} to \code{other}.
}
\description{
Convert ADP velocity components from an enu-based coordinate system to
another system, perhaps to align axes with the coastline.
}
\details{
The supplied angles specify rotations to be made around the axes for which
heading, pitch, and roll are defined.  For example, an eastward current will
point southeast if \code{heading=45} is used.

The returned value has heading, pitch, and roll matching those of \code{x},
so these angles retain their meaning as the instrument orientation.

NOTE: this function works similarly to \code{\link[=xyzToEnuAdp]{xyzToEnuAdp()}}, except
that in the present function, it makes no difference whether the instrument
points up or down, etc.
}
\examples{

library(oce)
data(adp)
o <- enuToOtherAdp(adp, heading=-31.5)
plot(o, which=1:3)

}
\references{
\enumerate{
\item Teledyne RD Instruments. \dQuote{ADCP Coordinate Transformation: Formulas and
Calculations,} January 2010. P/N 951-6079-00.
}
}
\seealso{
See \code{\link[=read.adp]{read.adp()}} for other functions that relate to
objects of class \code{"adp"}.

Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley
}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/satellite.R
\name{summary,satellite-method}
\alias{summary,satellite-method}
\title{Summarize a satellite object}
\usage{
\S4method{summary}{satellite}(object, ...)
}
\arguments{
\item{object}{a \linkS4class{satellite} object.}

\item{...}{Ignored.}
}
\description{
Summarize a satellite object
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/map.R
\name{mapAxis}
\alias{mapAxis}
\title{Add Axis Labels to an Existing Map}
\usage{
mapAxis(
  side = 1:2,
  longitude = TRUE,
  latitude = TRUE,
  axisStyle = 1,
  tick = TRUE,
  line = NA,
  pos = NA,
  outer = FALSE,
  font = NA,
  lty = "solid",
  lwd = 1,
  lwd.ticks = lwd,
  col = NULL,
  col.ticks = NULL,
  hadj = NA,
  padj = NA,
  tcl = -0.3,
  cex.axis = 1,
  mgp = c(0, 0.5, 0),
  debug = getOption("oceDebug")
)
}
\arguments{
\item{side}{the side at which labels are to be drawn.  If not provided,
sides 1 and 2 will be used (i.e. bottom and left-hand sides).}

\item{longitude}{either a logical value or a numeric vector of longitudes. There
are three possible cases:
(1) If \code{longitude=TRUE} (the default) then ticks and nearby numbers will occur at the
longitude grid established by the previous call to \code{\link[=mapPlot]{mapPlot()}};
(2) if \code{longitude=FALSE} then no longitude ticks or numbers are
drawn;
(3) if \code{longitude} is a vector of numerical values, then those ticks
are placed at those values, and numbers are written beside them.
Note that in cases 1 and 3, efforts are made to avoid overdrawing text,
so some longitude values might get ticks but not numbers. To get ticks
but not numbers, set \code{cex.axis=0}.}

\item{latitude}{similar to \code{longitude} but for latitude.}

\item{axisStyle}{an integer specifying the style of labels for the numbers
on axes.  The choices are:
1 for signed numbers without additional labels;
2 (the default) for unsigned numbers followed by letters indicating the hemisphere;
3 for signed numbers followed by a degree sign;
4 for unsigned numbers followed by a degree sign; and
5 for signed numbers followed by a degree sign and letters indicating the hemisphere.}

\item{tick}{parameter passed to \code{\link[=axis]{axis()}}.}

\item{line}{parameter passed to \code{\link[=axis]{axis()}}.}

\item{pos}{parameter passed to \code{\link[=axis]{axis()}}.}

\item{outer}{parameter passed to \code{\link[=axis]{axis()}}.}

\item{font}{axis font, passed to \code{\link[=axis]{axis()}}.}

\item{lty}{axis line type, passed to \code{\link[=axis]{axis()}}.}

\item{lwd}{axis line width, passed to \code{\link[=axis]{axis()}}).}

\item{lwd.ticks}{tick line width, passed to \code{\link[=axis]{axis()}}.}

\item{col}{axis color, passed to \code{\link[=axis]{axis()}}.}

\item{col.ticks}{axis tick color, passed to \code{\link[=axis]{axis()}}.}

\item{hadj}{an argument that is transmitted to \code{\link[=axis]{axis()}}.}

\item{padj}{an argument that is transmitted to \code{\link[=axis]{axis()}}.}

\item{tcl}{axis-tick size (see \code{\link[=par]{par()}}).}

\item{cex.axis}{axis-label expansion factor (see \code{\link[=par]{par()}}); set to 0
to prevent numbers from being placed in axes.}

\item{mgp}{three-element numerical vector describing axis-label
placement (see \code{\link[=par]{par()}}). It usually makes sense to set
the first and third elements to zero.}

\item{debug}{a flag that turns on debugging.  Set to 1 to get a moderate
amount of debugging information, or to 2 to get more.}
}
\description{
Plot axis labels on an existing map.
This is an advanced function, requiring
coordination with \code{\link[=mapPlot]{mapPlot()}} and (possibly) also with \code{\link[=mapGrid]{mapGrid()}},
and so it is best avoided by novices, who may be satisfied
with the defaults used by \code{\link[=mapPlot]{mapPlot()}}.
}
\examples{
\donttest{
library(oce)
data(coastlineWorld)
par(mar=c(2, 2, 1, 1))
lonlim <- c(-180, 180)
latlim <- c(70, 110)
# In mapPlot() call, note axes and grid args, to
# prevent over-plotting of defaults.  Some adjustments
# might be required to the mapGrid() arguments, to
# get agreement with the axis. This is why both
# mapGrid() and mapAxis() are best avoided; it is
# simpler to let mapPlot() handle these things.
mapPlot(coastlineWorld, projection="+proj=stere +lat_0=90",
        longitudelim=lonlim, latitudelim=latlim,
        col="tan", axes=FALSE, grid=FALSE)
mapGrid(15, 15)
mapAxis(axisStyle=5)
}

}
\seealso{
A map must first have been created with \code{\link[=mapPlot]{mapPlot()}}.

Other functions related to maps: 
\code{\link{formatPosition}()},
\code{\link{lonlat2map}()},
\code{\link{lonlat2utm}()},
\code{\link{map2lonlat}()},
\code{\link{mapArrows}()},
\code{\link{mapContour}()},
\code{\link{mapCoordinateSystem}()},
\code{\link{mapDirectionField}()},
\code{\link{mapGrid}()},
\code{\link{mapImage}()},
\code{\link{mapLines}()},
\code{\link{mapLocator}()},
\code{\link{mapLongitudeLatitudeXY}()},
\code{\link{mapPlot}()},
\code{\link{mapPoints}()},
\code{\link{mapPolygon}()},
\code{\link{mapScalebar}()},
\code{\link{mapText}()},
\code{\link{mapTissot}()},
\code{\link{oceCRS}()},
\code{\link{shiftLongitude}()},
\code{\link{usrLonLat}()},
\code{\link{utm2lonlat}()}
}
\author{
Dan Kelley
}
\concept{functions related to maps}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/cm.R
\name{[[<-,cm-method}
\alias{[[<-,cm-method}
\title{Replace Parts of a CM Object}
\usage{
\S4method{[[}{cm}(x, i, j, ...) <- value
}
\arguments{
\item{x}{a \linkS4class{cm} object.}

\item{i}{character value naming the item to replace.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{optional additional information (ignored).}

\item{value}{The value to be placed into \code{x}, somewhere.}
}
\description{
The \code{[[<-} method works for all \linkS4class{oce} objects.
The purpose, as with the related extraction method, \code{[[},
is to insulate users from the internal details of \linkS4class{oce}
objects, by looking for items within the various storage
slots of the object. Items not actually stored can also be
replaced, including units and data-quality
flags.
}
\details{
As with \code{[[} method, the procedure works in steps.

First, the \code{metadata} slot of \code{x} is checked to
see whether it contains something named with \code{i}.
If so, then the named item is replaced with \code{value}.

Otherwise, if the string value of \code{i} ends in \code{Unit}, then the
characters preceding that are taken as the name of a variable, and
the \code{metadata} slot of \code{x} is updated to store that unit, e.g.\preformatted{x[["temperatureUnits"]] <- list(unit=expression(degree*F),scale="")
}

Similarly, if \code{i} ends in \code{Flag}, then quality-control
flags are set up as defined by \code{result}, e.g.\preformatted{o[["temperatureFlags"]] <- c(2,4,2,2)
}

Otherwise, \code{\link[=pmatch]{pmatch()}} is used for a partial-string match with
the names of the items that are in the \code{data} slot of \code{x}.
The first item found (if any) is then updated to hold the value \code{result}.

If none of these conditions is met, a warning is issued.
}
\seealso{
Other functions that replace parts of oce objects: 
\code{\link{[[<-,adp-method}},
\code{\link{[[<-,amsr-method}},
\code{\link{[[<-,argo-method}},
\code{\link{[[<-,bremen-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{[[<-,echosounder-method}},
\code{\link{[[<-,g1sst-method}},
\code{\link{[[<-,gps-method}},
\code{\link{[[<-,ladp-method}},
\code{\link{[[<-,landsat-method}},
\code{\link{[[<-,lisst-method}},
\code{\link{[[<-,lobo-method}},
\code{\link{[[<-,met-method}},
\code{\link{[[<-,oce-method}},
\code{\link{[[<-,odf-method}},
\code{\link{[[<-,rsk-method}},
\code{\link{[[<-,sealevel-method}},
\code{\link{[[<-,section-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{[[<-,topo-method}},
\code{\link{[[<-,windrose-method}},
\code{\link{[[<-,xbt-method}}

Other things related to cm data: 
\code{\link{[[,cm-method}},
\code{\link{as.cm}()},
\code{\link{cm-class}},
\code{\link{cm}},
\code{\link{plot,cm-method}},
\code{\link{read.cm}()},
\code{\link{rotateAboutZ}()},
\code{\link{subset,cm-method}},
\code{\link{summary,cm-method}}
}
\concept{functions that replace parts of oce objects}
\concept{things related to cm data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/extdata.R
\docType{data}
\encoding{UTF-8}
\name{test_met_csv1.csv}
\alias{test_met_csv1.csv}
\title{Sample met dataset (CSV format prior to October 2019)}
\description{
This is a subset of a file downloaded with \code{\link[=download.met]{download.met()}}
from the Environment Canada "Climate Data"
website \verb{https://climate.weather.gc.ca/index_e.html}
sometime before October 13, 2019.
}
\examples{
\dontrun{
read.met(system.file("extdata", "test_met_csv1.csv", package="oce"))
}

}
\seealso{
Other raw datasets: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{adp_rdi.000}},
\code{\link{ctd.cnv}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{test_met_csv2.csv}},
\code{\link{test_met_xml2.xml}},
\code{\link{xbt.edf}}

Other things related to met data: 
\code{\link{[[,met-method}},
\code{\link{[[<-,met-method}},
\code{\link{as.met}()},
\code{\link{download.met}()},
\code{\link{met-class}},
\code{\link{met}},
\code{\link{plot,met-method}},
\code{\link{read.met}()},
\code{\link{subset,met-method}},
\code{\link{summary,met-method}},
\code{\link{test_met_csv2.csv}},
\code{\link{test_met_xml2.xml}}
}
\concept{raw datasets}
\concept{things related to met data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{standardDepths}
\alias{standardDepths}
\title{Standard Oceanographic Depths}
\usage{
standardDepths(n = 0)
}
\arguments{
\item{n}{Integer specifying the number of subdivisions to insert between
each of the stated levels. For exmple, setting \code{n=1} puts a 5m level
between the 0 and 10m levels, and \code{n=2} puts 3.33 and 6.66 between
0 and 10m.}
}
\value{
A vector of depths that are more closely spaced for small values,
i.e. a finer grid near the ocean surface.
}
\description{
This returns a vector of numbers that build upon the shorter lists
provided in Chapter 10 of reference 1 and the more modern World
Ocean Atlases (e.g. reference 2).
With the default call,
i.e. with \code{n=0}, the result is
\code{c(0, 10, 20, 30, 40, 50, 75, 100, 125, 150, 200, 250,
seq(300, 1500, by=100), 1750, seq(2000, 10000, by=500))}.
For higher values of \code{n}, progressively more and more values
are added between each pair in this sequence.
See the documentation for
\code{\link[=sectionGrid]{sectionGrid()}} for how \code{standardDepths} can be used
in gridding data for section plots.
}
\examples{
depth  <- standardDepths()
depth1 <- standardDepths(1)
plot(depth, depth)
points(depth1, depth1, col=2, pch=20, cex=1/2)

}
\references{
\enumerate{
\item Sverdrup, H U, Martin W Johnson, and Richard H Fleming. The Oceans,
Their Physics, Chemistry, and General Biology. New York: Prentice-Hall, 1942.
\code{https://publishing.cdlib.org/ucpressebooks/view?docId=kt167nb66r}
}

2.Locarnini, R. A., A. V. Mishonov, J. I. Antonov, T. P. Boyer,
H. E. Garcia, O. K. Baranova, M. M. Zweng, D. R. Johnson, and
S. Levitus. \dQuote{World Ocean Atlas 2009 Temperature.}
US Government printing Office, 2010.
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/tides.R
\name{tidemConstituentNameFix}
\alias{tidemConstituentNameFix}
\title{Change tidal constituent name from T-TIDE to Foreman convention}
\usage{
tidemConstituentNameFix(names, debug = 1)
}
\arguments{
\item{names}{a vector of character values, holding constituent names}

\item{debug}{an integer controlling warnings. If this is zero, then no warnings
are issued during processing; otherwise, as is the default, warnings are
issued for each conversion that is required.}
}
\value{
A vector of character values of tidal constituent names, in the Foreman naming convention.
}
\description{
This is used by \code{\link[=tidem]{tidem()}} to permit users to specify constituent names in either
the T-TIDE convention (see Pawlowicz et al. 2002) or Foreman convention
(see Foreman (1978). There are only two such instances:
\code{"MS"}, which gets translated to \code{"M8"}, and \code{"UPSI"},
which gets translated to \code{"UPS1"}.
}
\references{
Foreman, M. G. G., 1978. Manual for Tidal Currents Analysis and Prediction.
Pacific Marine Science Report. British Columbia, Canada: Institute of Ocean
Sciences, Patricia Bay.

Pawlowicz, Rich, Bob Beardsley, and Steve Lentz, 2002.  Classical tidal
harmonic analysis including error estimates in MATLAB using \code{T_TIDE}.
Computers and Geosciences, 28, 929-937.
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/AllClass.R
\name{concatenate,list-method}
\alias{concatenate,list-method}
\title{Concatenate a list of oce objects}
\usage{
\S4method{concatenate}{list}(object)
}
\arguments{
\item{object}{a \link{list} of \linkS4class{oce} objects.}
}
\value{
An object of class corresponding to that in \code{object}.
}
\description{
Concatenate a list of oce objects
}
\seealso{
Other functions that concatenate oce objects: 
\code{\link{concatenate,adp-method}},
\code{\link{concatenate,oce-method}},
\code{\link{concatenate}()}
}
\concept{functions that concatenate oce objects}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{drawDirectionField}
\alias{drawDirectionField}
\title{Draw a Direction Field}
\usage{
drawDirectionField(
  x,
  y,
  u,
  v,
  scalex,
  scaley,
  skip,
  length = 0.05,
  add = FALSE,
  type = 1,
  col = par("fg"),
  pch = 1,
  cex = par("cex"),
  lwd = par("lwd"),
  lty = par("lty"),
  xlab = "",
  ylab = "",
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{x, y}{coordinates at which velocities are specified. The
length of \code{x} and \code{y} depends on the form of \code{u}
and \code{v} (vectors or matrices).}

\item{u, v}{velocity components in the x and y directions. Can be
either vectors with the same length as \verb{x, y}, or
matrices, of dimension \code{length(x)} by \code{length(y)}.}

\item{scalex, scaley}{scale to be used for the velocity arrows.
Exactly one of these must be specified.  Arrows that have
\code{u^2+v^2=1} will have length \code{scalex} along the x
axis, or \code{scaley} along the y axis, according to which
argument is given.}

\item{skip}{either an integer, or a two-element vector indicating
the number of points to skip when plotting arrows (for the
matrix \verb{u, v} case). If a single value, the same
\code{skip} is applied to both the \code{x} and \code{y}
directions. If a two-element vector, specifies different values
for the \code{x} and \code{y} directions.}

\item{length}{indication of \emph{width} of arrowheads. The
somewhat confusing name of this argument is a consequence of
the fact that it is passed to \code{\link[=arrows]{arrows()}} for drawing
arrows.  Note that the present default is smaller than the
default used by \code{\link[=arrows]{arrows()}}.}

\item{add}{if \code{TRUE}, the arrows are added to an existing
plot; otherwise, a new plot is started by calling
\code{\link[=plot]{plot()}} with \code{x}, \code{y} and \code{type="n"}.
In other words, the plot will be very basic. In most cases, the
user will probably want to draw a diagram first, and \code{add}
the direction field later.}

\item{type}{indication of the style of arrow-like indication of the
direction.}

\item{col}{color of line segments or arrows; see \code{\link[=par]{par()}} for meaning}

\item{pch, cex}{plot character and expansion factor, used for
\code{type=1}; see \code{\link[=par]{par()}} for meanings}

\item{lwd, lty}{line width and type, used for \code{type=2}; see \code{\link[=par]{par()}} for meaning}

\item{xlab, ylab}{\code{x} and \code{y} axis labels}

\item{debug}{debugging value; set to a positive integer to get
debugging information.}

\item{...}{other arguments to be passed to plotting functions
(e.g. axis labels, etc).}
}
\value{
None.
}
\description{
The direction field is indicated variously, depending on the value of
\code{type}:
}
\details{
\itemize{
\item For \code{type=1}, each indicator is drawn with a symbol, according to the
value of \code{pch} (either supplied globally, or as an element of the
\code{...} list) and of size \code{cex}, and color \code{col}.   Then, a
line segment is drawn for each, and for this \code{lwd} and \code{col} may
be set globally or in the \code{...} list.
\item For \code{type=2}, the points are not drawn, but arrows are drawn instead
of the line segments.  Again, \code{lwd} and \code{col} control the type of
the line.
}
}
\examples{
library(oce)
plot(c(-1.5, 1.5), c(-1.5, 1.5), xlab="", ylab="", type='n')
drawDirectionField(x=rep(0, 2), y=rep(0, 2), u=c(1, 1), v=c(1, -1), scalex=0.5, add=TRUE)
plot(c(-1.5, 1.5), c(-1.5, 1.5), xlab="", ylab="", type='n')
drawDirectionField(x=rep(0, 2), y=rep(0, 2), u=c(1, 1), v=c(1, -1), scalex=0.5, add=TRUE,
                   type=2)

# 2D example
x <- seq(-2, 2, 0.1)
y <- x
xx <- expand.grid(x, y)[,1]
yy <- expand.grid(x, y)[,2]
z <- matrix(xx*exp(-xx^2 -yy^2), nrow=length(x))
gz <- grad(z, x, y)
drawDirectionField(x, y, gz$gx, gz$gy, scalex=0.5, type=2, len=0.02)
oceContour(x, y, z, add=TRUE)

}
\author{
Dan Kelley and Clark Richards
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{tail.oce}
\alias{tail.oce}
\title{Extract the End of an Oce Object}
\usage{
\method{tail}{oce}(x, n = 6L, ...)
}
\arguments{
\item{x}{an \linkS4class{oce} object.}

\item{n}{Number of elements to extract, as for \code{\link[=tail]{tail()}}.}

\item{...}{ignored}
}
\description{
Extract the End of an Oce Object

This function handles the following object classes directly:
\linkS4class{adp},
\linkS4class{adv},
\linkS4class{argo} (selection by profile),
\linkS4class{coastline},
\linkS4class{ctd},
\linkS4class{echosounder} (selection by ping),
\linkS4class{section} (selection by station)
and
\linkS4class{topo} (selection by longitude and latitude).
It does not handle
\linkS4class{amsr} or
\linkS4class{landsat}
yet, instead issuing a warning and returning \code{x} in those cases.
For all other classes, it calls \code{\link[=tail]{tail()}}
with \code{n} as provided, for each item in the \code{data}
slot, issuing a warning if that item is not a vector; the
author is quite aware that this may not work well for all classes.
The plan is to handle all appropriate classes by July 2018.
Please contact the author if there is a class you need handled
before that date.
}
\seealso{
\code{\link[=head.oce]{head.oce()}}, which yields the start of an \code{oce} object.
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swSpice}
\alias{swSpice}
\title{Seawater spiciness}
\usage{
swSpice(
  salinity,
  temperature = NULL,
  pressure = NULL,
  longitude = NULL,
  latitude = NULL,
  eos = getOption("oceEOS", default = "gsw")
)
}
\arguments{
\item{salinity}{either salinity (PSU) (in which case \code{temperature} and
\code{pressure} must be provided) \emph{or} a \code{ctd} object (in which
case \code{salinity}, \code{temperature} and \code{pressure} are determined
from the object, and must not be provided in the argument list).}

\item{temperature}{\emph{in-situ} temperature (\eqn{^\circ}{deg}C) on the
ITS-90 scale; see \dQuote{Temperature units} in the documentation for
\code{\link[=swRho]{swRho()}}.}

\item{pressure}{Seawater pressure (dbar) (only used if \code{eos} is
\code{"gsw"}); see \sQuote{Details}..}

\item{longitude}{longitude of observation (only used if \code{eos} is
\code{"gsw"}; see \sQuote{Details}).}

\item{latitude}{latitude of observation (only used if \code{eos} is
\code{"gsw"}; see \sQuote{Details}).}

\item{eos}{Character value specifying the equation of state,
either \code{"unesco"} (for the Flament formulation, although this
is not actually part of UNESCO)
or \code{"gsw"} for the Gibbs SeaWater formulation.}
}
\value{
Flament-formulated spice \eqn{kg/m^3} if \code{eos} is \code{"unesco"}
or surface-referenced GSW spiciness0 \eqn{kg/m^3} if \code{eos} is \code{"gsw"},
the latter provided by \code{\link[gsw:gsw_spiciness0]{gsw::gsw_spiciness0()}}, and hence aimed
at application within the top half-kilometre of the ocean.
}
\description{
Compute seawater "spice", also called "spiciness" (a variable orthogonal
to density in TS space), in either of two formulations, depending on
the value of the \code{eos} argument. If \code{eos="unesco"} then
Flament's (reference 1) formulation is used. If \code{eos="gsw"}
then the Gibbs SeaWater formulation for "spiciness0" is used
(see reference 2).
}
\details{
If the first argument is a \code{ctd} object, then salinity, temperature and
pressure values are extracted from it, and used for the calculation. (For
the \code{eos="gsw"} case, longitude and latitude are also extracted, because
these are required for the formulation of spiciness0.

Roughly speaking, seawater with a high spiciness is relatively warm and
salty compared with less spicy water. Another interpretation is that spice
is a variable measuring distance orthogonal to isopycnal lines on TS
diagrams (if the diagrams are scaled to make the isopycnals run at 45
degrees). Note that pressure, longitude and latitude are all
ignored in the Flament definition.
}
\examples{
## Contrast the two formulations.
library(oce)
data(ctd)
p <- ctd[["pressure"]]
plot(swSpice(ctd, eos="unesco"), p,
     xlim=c(-2.7, -1.5), ylim=rev(range(p)),
     xlab="Spice", ylab="Pressure (dbar)")
points(swSpice(ctd, eos="gsw"), p,col=2)
mtext("black=unesco, red=gsw")

}
\references{
\enumerate{
\item Flament, P. \dQuote{A State Variable for Characterizing Water Masses and Their
Diffusive Stability: Spiciness.} Progress in Oceanography, Observations of the
1997-98 El Nino along the West Coast of North America, 54, no. 1
(July 1, 2002):493-501.
\doi{10.1016/S0079-6611(02)00065-4}
}

2.McDougall, Trevor J., and Oliver A. Krzysik. \dQuote{Spiciness.}
Journal of Marine Research 73, no. 5 (September 1, 2015): 141-52.
\doi{10.1357/002224015816665589}
}
\seealso{
Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley coded this, merely an interface to the code described
by references 1 and 2.
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.nortek.R
\name{read.adp.nortek}
\alias{read.adp.nortek}
\title{Read a Nortek ADP File}
\usage{
read.adp.nortek(
  file,
  from = 1,
  to,
  by = 1,
  tz = getOption("oceTz"),
  longitude = NA,
  latitude = NA,
  type = c("aquadoppHR", "aquadoppProfiler", "aquadopp", "aquadoppPlusMagnetometer"),
  orientation,
  distance,
  monitor = FALSE,
  despike = FALSE,
  processingLog,
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{file}{a connection or a character string giving the name of the file
to load.  (For \code{read.adp.sontek.serial}, this is generally a list of
files, which will be concatenated.)}

\item{from}{indication of the first profile to read.  This can be an
integer, the sequence number of the first profile to read, or a POSIXt time
before which profiles should be skipped, or a character string that converts
to a POSIXt time (assuming UTC timezone).  See \dQuote{Examples}, and make
careful note of the use of the \code{tz} argument. If \code{from} is not
supplied, it defaults to 1.}

\item{to}{an optional indication of the last profile to read, in a
format as described for \code{from}.  As a special case, \code{to=0} means
to read the file to the end.  If \code{to} is not supplied, then it defaults
to 0.}

\item{by}{an optional indication of the stride length to use while walking through
the file.  If this is an integer, then \code{by-1} profiles are skipped
between each pair of profiles that is read, e.g. the default \code{by=1}
means to read all the data.  (For RDI files \emph{only}, there are some
extra features to avoid running out of memory; see \dQuote{Memory considerations}.)}

\item{tz}{character string indicating time zone to be assumed in the data.}

\item{longitude}{optional signed number indicating the longitude in degrees
East.}

\item{latitude}{optional signed number indicating the latitude in degrees
North.}

\item{type}{a character string indicating the type of instrument.}

\item{orientation}{an optional character string specifying the orientation of
the sensor, provided for those cases in which it cannot be inferred from the
data file.  The valid choices are \code{"upward"}, \code{"downward"}, and
\code{"sideward"}.}

\item{distance}{an optional vector holding the distances of bin centres from
the sensor.  This argument is ignored except for Nortek profilers, and need
not be given if the function determines the distances correctly from the
data.  The problem is that the distance is poorly documented in the Nortek
System Integrator Guide (2008 edition, page 31), so the function must rely
on word-of-mouth formulae that do not work in all cases.}

\item{monitor}{boolean value indicating whether to indicate the progress
of reading the file, by using \code{\link[=txtProgressBar]{txtProgressBar()}} or otherwise.  The value
of \code{monitor} is changed to \code{FALSE} automatically, for non-interactive
sessions.}

\item{despike}{a logical value indicating whether to use \code{\link[=despike]{despike()}} to remove
anomalous spikes in heading, etc.}

\item{processingLog}{if provided, the action item to be stored in the log.
(Typically only provided for internal calls; the default that it provides is
better for normal calls by a user.)}

\item{debug}{a flag that turns on debugging.  Set to 1 to get a moderate
amount of debugging information, or to 2 to get more.}

\item{\dots}{optional additional arguments that some (but not all)
\verb{read.adp.*()} functions pass to lower-level functions.}
}
\value{
An \linkS4class{adp} object.
The contents of that object make sense for the particular instrument
type under study, e.g. if the data file contains
NMEA strings, then navigational data will be stored in an item
called \code{nmea} in the \code{data} slot).
}
\description{
Read a Nortek ADP File
}
\references{
\enumerate{
\item Information on Nortek profilers (including the System Integrator Guide,
which explains the data format byte-by-byte) is available at
\url{https://www.nortekusa.com/}.  (One must join the site to see the
manuals.)
\item The Nortek Knowledge Center
\url{https://www.nortekusa.com/en/knowledge-center} may be of help if
problems arise in dealing with data from Nortek instruments.
}
}
\seealso{
Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley
}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{oceColorsTurbo}
\alias{oceColorsTurbo}
\alias{oce.colorsTurbo}
\title{Create colors similar to the google turbo scheme}
\usage{
oceColorsTurbo(n)
}
\arguments{
\item{n}{number of colors to create.}
}
\description{
This uses the coefficients published (with Apache license) by google,
as described by Mikhailo (2019).
}
\examples{
library(oce)
imagep(volcano, col=oceColorsTurbo(128),
       zlab="oceColorsTurbo")

}
\references{
Mikhailo, Anton.
\dQuote{Turbo, An Improved Rainbow Colormap for Visualization.}
Google AI (blog), August 20, 2019.
\url{https://ai.googleblog.com/2019/08/turbo-improved-rainbow-colormap-for.html}

The following references provide information on choosing
colour schemes, that are suitable for viewers who have
colour deficiencies.

Light, Adam, and Patrick J. Bartlein. "The End of the Rainbow? Color
Schemes for Improved Data Graphics." \emph{Eos, Transactions American Geophysical
Union} 85, no. 40 (2004): 385.
DOI: 10.1029/2004EO400002

Stephenson, David B. "Comment on 'Color Schemes for Improved Data
Graphics', by A Light and P.J. Bartlein." \emph{Eos, Transactions American
Geophysical Union} 86, no.  20 (2005): 196.
DOI: 10.1029/2005EO200005

Light, Adam, and Patrick J. Bartlein. "Reply to 'Comment on Color
Schemes for Improved Data Graphics,' by A. Light and P.J.  Bartlein'."
\emph{Eos, Transactions American Geophysical Union} 86, no. 20 (2005): 196‚Äì196.
DOI: 10.1029/2005EO200006
}
\seealso{
Other things related to colors: 
\code{\link{colormapGMT}()},
\code{\link{colormap}()},
\code{\link{oceColors9B}()},
\code{\link{oceColorsCDOM}()},
\code{\link{oceColorsChlorophyll}()},
\code{\link{oceColorsClosure}()},
\code{\link{oceColorsDensity}()},
\code{\link{oceColorsFreesurface}()},
\code{\link{oceColorsGebco}()},
\code{\link{oceColorsJet}()},
\code{\link{oceColorsOxygen}()},
\code{\link{oceColorsPAR}()},
\code{\link{oceColorsPalette}()},
\code{\link{oceColorsPhase}()},
\code{\link{oceColorsSalinity}()},
\code{\link{oceColorsTemperature}()},
\code{\link{oceColorsTurbidity}()},
\code{\link{oceColorsTwo}()},
\code{\link{oceColorsVelocity}()},
\code{\link{oceColorsViridis}()},
\code{\link{oceColorsVorticity}()},
\code{\link{ocecolors}}
}
\author{
Dan Kelley
}
\concept{things related to colors}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.nortek.R
\name{read.adp.ad2cp}
\alias{read.adp.ad2cp}
\title{Read an AD2CP File}
\usage{
read.adp.ad2cp(
  file,
  from = 1,
  to = 0,
  by = 1,
  tz = getOption("oceTz"),
  longitude = NA,
  latitude = NA,
  orientation,
  distance,
  plan,
  type,
  monitor = FALSE,
  despike = FALSE,
  processingLog,
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{file}{A connection or a character string giving the name of the file to load.}

\item{from}{An integer indicating the index number of the first record to
read. This must equal 1, for this version of \code{read.adp.ad2cp}.
(If not provided, \code{from} defaults to 1.)}

\item{to}{An integer indicating the final record to read.
(If not provided, \code{by} defaults to 1e9.)}

\item{by}{An integer indicating the step from record to record. This
must equal 1, for this version of \code{read.adp.ad2cp}.
(If not provided, \code{by} defaults to 1.)}

\item{tz}{Character value indicating time zone. This is used in interpreting
times stored in the file.}

\item{longitude}{Numerical value indicating the longitude of observation.}

\item{latitude}{Numerical value indicating the latitude of observation.}

\item{orientation}{Ignored by \code{read.adp.ad2cp}, and provided only for similarity
to other adp-reading functions.}

\item{distance}{Ignored by \code{read.adp.ad2cp}, and provided only for similarity
to other adp-reading functions.}

\item{plan}{Optional integer specifying which 'plan' to focus on (see reference 1
for the meaning of 'plan').  If this is not given, it defaults to the most
common plan in the requested subset of the data.}

\item{type}{Optional character value indicating the type of Nortek instrument.
If this is not provided, an attempt is made to infer it
from the file header (if there is one), and \code{"Signature1000"}
is used, otherwise. The importance
of knowing the type is for inferring the slantwise beam angle, which is usd in the
conversion from beam coordinates to xyz coordinates. If \code{type} is
provided, it must be one of \code{"Signature250"}, \code{"Signature500"},
or \code{"Signature1000"}; the first of these has a 20 degree
slant-beam angle, while the others each have 20 degrees (see reference 2,
section 2 on page 6). Note that \code{\link[=oceSetMetadata]{oceSetMetadata()}}
can be used to alter the slantwise beam angle of an existing object,
and this will alter any later conversion from beam to xyz coordinates.}

\item{monitor}{boolean value indicating whether to indicate the progress
of reading the file, by using \code{\link[=txtProgressBar]{txtProgressBar()}} or otherwise.  The value
of \code{monitor} is changed to \code{FALSE} automatically, for non-interactive
sessions.}

\item{despike}{Ignored by this function, and provided only for similarity
to other adp-reading functions.}

\item{processingLog}{Character value that, if provided, is saved
within the \code{processingLog} slot of th returned value.}

\item{debug}{Integer value indicating the level of debugging.
Set to 1 to get a moderate amount of debugging information, from
the R code only, to 2 to get some debugging information from the C++
code that is used to parse the data chunks, or to 3 for intensive
debugging at both levels.}

\item{\dots}{Ignored by this function.}
}
\description{
This function may be incomplete in some important ways,
because AD2CP data formats are not described clearly enough
in references 1, 2 and 3 to be certain of how to handle
the range of file configurations that may be possible. The code
has been tested with a small number of files that are available
to the author, but these do not cover some cases that users might
require, e.g. involving echosounder and altimeter data streams.
Please be on the lookout for problems and contact
the author if you need help. Also, note that some of the standard
\verb{read.adp.*} arguments are handled differently with
this function, e.g. \code{by} must equal 1, because skipping
records makes little sense with blended multiple streams;
see the \dQuote{Arguments} section for other limitations
that stem from the specifics of this file format.
}
\examples{
\dontrun{
d <- read.adp.ad2cp("~/test.ad2cp", to=100) # or read.oce()
}

}
\references{
\enumerate{
\item Nortek AS. \dQuote{Signature Integration 55|250|500|1000kHz.} Nortek AS, 2017.
\item Nortek AS. \dQuote{Operations Manual - Signature250, 500 and 1000.} Nortek AS, September 21, 2018.
\item Nortek AS. \dQuote{Signature Integration 55|250|500|1000kHz.} Nortek AS, 2018. (This revision of
reference 1 is useful in including new information about instrument orientation. Note that
most of the comments within the \code{read.adp.ad2cp} code refer to reference 1, which has different
page numbers than reference 3.)
}
}
\seealso{
Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley
}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/accessors.R
\name{oceDeleteMetadata}
\alias{oceDeleteMetadata}
\title{Delete Something in an oce metadata Slot}
\usage{
oceDeleteMetadata(object, name)
}
\arguments{
\item{object}{an \linkS4class{oce} object.}

\item{name}{String indicating the name of the item to be deleted.}
}
\description{
Return a copy of the supplied object that lacks the named
element in its \code{metadata} slot, and that has a note
about the deletion in its processing log.
}
\seealso{
Other things related to the metadata slot: 
\code{\link{oceGetMetadata}()},
\code{\link{oceRenameMetadata}()},
\code{\link{oceSetMetadata}()}
}
\author{
Dan Kelley
}
\concept{things related to the metadata slot}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/lobo.R
\name{summary,lobo-method}
\alias{summary,lobo-method}
\title{Summarize a LOBO Object}
\usage{
\S4method{summary}{lobo}(object, ...)
}
\arguments{
\item{object}{a \linkS4class{lobo} object.}

\item{\dots}{further arguments passed to or from other methods.}
}
\value{
A matrix containing statistics of the elements of the \code{data}
slot.
}
\description{
Pertinent summary information is presented, including the sampling interval,
data ranges, etc.
}
\examples{

library(oce)
data(lobo)
summary(lobo)
}
\seealso{
The documentation for \linkS4class{lobo} explains the
structure of LOBO objects, and also outlines the other functions dealing
with them.

Other things related to lobo data: 
\code{\link{[[,lobo-method}},
\code{\link{[[<-,lobo-method}},
\code{\link{as.lobo}()},
\code{\link{lobo-class}},
\code{\link{lobo}},
\code{\link{plot,lobo-method}},
\code{\link{read.lobo}()},
\code{\link{subset,lobo-method}}
}
\author{
Dan Kelley
}
\concept{things related to lobo data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{magneticField}
\alias{magneticField}
\title{Earth magnetic declination, inclination, and intensity}
\usage{
magneticField(longitude, latitude, time, version = 13)
}
\arguments{
\item{longitude}{longitude in degrees east (negative for degrees west).  The
dimensions must conform to lat.}

\item{latitude}{latitude in degrees north, a number, vector, or matrix.}

\item{time}{The time at which the field is desired. This may be a
single value or a vector or matrix that is structured to match
\code{longitude} and \code{latitude}. The value may a decimal year,
a POSIXt time, or a Date time.}

\item{version}{an integer that must be either 12 or 13, to specify
the version number of the formulae. Note that 13 became the default
on 2020 March 3, so to old code will need to specify \code{version=12}
to work as it did before that date.}
}
\value{
A list containing \code{declination}, \code{inclination}, and
\code{intensity}.
}
\description{
Implements the 12th and 13th generations of the
International Geomagnetic Reference Field
(IGRF), based on a reworked version of a Fortran program downloaded from a
NOAA website (see reference 1).
}
\details{
The code (subroutines \code{igrf12syn} and \code{igrf13syn}) seem to have
been written by Susan Macmillan of the British Geological Survey.  Comments
in the source code \code{igrf13syn} (the current default used here)
indicate that its coefficients were agreed to in
December 2019 by the IAGA Working Group V-MOD.  Other comments in that code
suggest that the proposed application time interval is from years 1900 to 2025, inclusive,
but that only dates from 1945 to 2015 are to be considered definitive.
}
\section{Historical Notes}{

For about a decade, \code{magneticField} used the version 12 formulae provided
by IAGA, but the code was updated on March 3, 2020, to version 13.  Example
3 shows that the differences in declination are typically under 2 degrees
(with 95 percent of the data lying between -1.7 and 0.7 degrees).
}

\examples{
library(oce)
# 1. Today's value at Halifax NS
magneticField(-(63+36/60), 44+39/60, Sys.Date())

# 2. World map of declination in year 2000.
\donttest{
data(coastlineWorld)
par(mar=rep(0.5, 4)) # no axes on whole-world projection
mapPlot(coastlineWorld, projection="+proj=robin", col="lightgray")
# Construct matrix holding declination
lon <- seq(-180, 180)
lat <- seq(-90, 90)
dec2000 <- function(lon, lat)
    magneticField(lon, lat, 2000)$declination
dec <- outer(lon, lat, dec2000) # hint: outer() is very handy!
# Contour, unlabelled for small increments, labeled for
# larger increments.
mapContour(lon, lat, dec, col='blue', levels=seq(-180, -5, 5),
           lty=3, drawlabels=FALSE)
mapContour(lon, lat, dec, col='blue', levels=seq(-180, -20, 20))
mapContour(lon, lat, dec, col='red', levels=seq(5, 180, 5),
           lty=3, drawlabels=FALSE)
mapContour(lon, lat, dec, col='red', levels=seq(20, 180, 20))
mapContour(lon, lat, dec, levels=180, col='black', lwd=2, drawlabels=FALSE)
mapContour(lon, lat, dec, levels=0, col='black', lwd=2)
}

# 3. Declination differences between versions 12 and 13
\donttest{
lon <- seq(-180, 180)
lat <- seq(-90, 90)
decDiff <- function(lon, lat) {
    old <- magneticField(lon, lat, 2020, version=13)$declination
    new <- magneticField(lon, lat, 2020, version=12)$declination
    new - old
}
decDiff <- outer(lon, lat, decDiff)
decDiff <- ifelse(decDiff > 180, decDiff - 360, decDiff)
# Overall (mean) shift -0.1deg
t.test(decDiff)
# View histogram, narrowed to small differences
par(mar=c(3.5, 3.5, 2, 2), mgp=c(2, 0.7, 0))
hist(decDiff, breaks=seq(-180, 180, 0.05), xlim=c(-2, 2),
     xlab="Declination difference [deg] from version=12 to version=13",
     main="Predictions for year 2020")
print(quantile(decDiff, c(0.025, 0.975)))
# Note that the large differences are at high latitudes
imagep(lon,lat,decDiff, zlim=c(-1,1)*max(abs(decDiff)))
lines(coastlineWorld[["longitude"]], coastlineWorld[["latitude"]])
}
}
\references{
\enumerate{
\item The underlying Fortran code for version 12 is from \code{igrf12.f}, downloaded the NOAA
website (\url{https://www.ngdc.noaa.gov/IAGA/vmod/igrf.html}) on June 7,
\item That for version 13 is \code{igrf13.f}, downloaded from the NOAA website
(\url{https://www.ngdc.noaa.gov/IAGA/vmod/igrf.html} on March 3, 2020.
\item Witze, Alexandra. \dQuote{Earth's Magnetic Field Is Acting up and Geologists Don't Know Why.}
Nature 565 (January 9, 2019): 143.
\doi{10.1038/d41586-019-00007-1}
}
}
\seealso{
Other things related to magnetism: 
\code{\link{applyMagneticDeclination}()}
}
\author{
Dan Kelley wrote the R code and a fortran wrapper to the
\code{igrf12.f} subroutine, which was written by Susan Macmillan of the
British Geological Survey and distributed ``without limitation'' (email from
SM to DK dated June 5, 2015).
}
\concept{things related to magnetism}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/xbt.R
\name{summary,xbt-method}
\alias{summary,xbt-method}
\title{Summarize an xbt Object}
\usage{
\S4method{summary}{xbt}(object, ...)
}
\arguments{
\item{object}{A \linkS4class{xbt} object.}

\item{...}{Further arguments passed to or from other methods.}
}
\description{
Summarizes some of the data in a \code{xbt} object.
}
\seealso{
The documentation for the \linkS4class{xbt} class explains the structure
of \code{xbt} objects, and also outlines the other functions dealing with them.

Other things related to xbt data: 
\code{\link{[[,xbt-method}},
\code{\link{[[<-,xbt-method}},
\code{\link{as.xbt}()},
\code{\link{plot,xbt-method}},
\code{\link{read.xbt.noaa1}()},
\code{\link{read.xbt}()},
\code{\link{subset,xbt-method}},
\code{\link{xbt-class}},
\code{\link{xbt.edf}},
\code{\link{xbt}}
}
\author{
Dan Kelley
}
\concept{things related to xbt data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/section.R
\name{plot,section-method}
\alias{plot,section-method}
\alias{plot.section}
\title{Plot a section Object}
\usage{
\S4method{plot}{section}(
  x,
  which = c(1, 2, 3, 99),
  eos,
  at = NULL,
  labels = TRUE,
  grid = FALSE,
  contourLevels = NULL,
  contourLabels = NULL,
  stationIndices,
  coastline = "best",
  xlim = NULL,
  ylim = NULL,
  zlim = NULL,
  zbreaks = NULL,
  zcol = NULL,
  map.xlim = NULL,
  map.ylim = NULL,
  clongitude,
  clatitude,
  span,
  projection = NULL,
  xtype = "distance",
  ytype = "depth",
  ztype = "contour",
  longitude0,
  latitude0,
  legend.loc = "bottomright",
  legend.text = NULL,
  showStations = FALSE,
  showStart = TRUE,
  stationTicks = TRUE,
  showBottom = TRUE,
  showSpine = TRUE,
  drawPalette = TRUE,
  axes = TRUE,
  mgp,
  mar,
  col,
  cex,
  pch,
  labcex = 1,
  debug,
  ...
)
}
\arguments{
\item{x}{a \linkS4class{section} object.}

\item{which}{a list of desired plot types, as explained in \dQuote{Details}.
There may be up to four panels in total, and the desired plots are placed in
these panels, in reading order.  If only one panel is plotted, \code{par} is
not adjusted, which makes it easy to add to the plot with subsequent plotting
commands.}

\item{eos}{Character indication of the seawater equation of state
to use. The permitted choices are \code{"gsw"} and \code{"unesco"}. If \code{eos} is
not supplied, it defaults to \code{\link{getOption}}\verb{("oceEOS",default="gsw")}.}

\item{at}{If \code{NULL} (the default), the x axis will indicate the distance
of the stations from the first in the section.  (This may give errors in the
contouring routine, if the stations are not present in a geographical order.)
If a list, then it indicates the values at which stations will be plotted.}

\item{labels}{Either a logical, indicating whether to put labels on the x axis,
or a vector that is a list of labels to be placed at the x positions indicated
by \code{at}.}

\item{grid}{If \code{TRUE}, points are drawn at data locations.}

\item{contourLevels}{Optional contour levels.}

\item{contourLabels}{Optional contour labels.}

\item{stationIndices}{Optional list of the indices of stations to use.  Note
that an index is \emph{not} a station number, e.g. to show the first 4
stations, use \code{station.indices=1:4}.}

\item{coastline}{Either a \linkS4class{coastline} object to be used,
or a string.  In the second case, the permitted
choices are \code{"best"} (the default) to pick
a variant that suits the scale, \code{"coastlineWorld"} for the coarse
version that is provided by \CRANpkg{oce},
\code{"coastlineWorldMedium"} or \code{"coastlineWorldFine"} for two
coastlines provided by the \CRANpkg{ocedata} package, or \code{"none"}, to avoid
drawing a coastline.}

\item{xlim}{Optional limit for x axis (only in sections, not map).}

\item{ylim}{Optional limit for y axis (only in sections, not map)}

\item{zlim, zbreaks, zcol}{Elements that control colours for \code{image} and \code{points}
plot types, i.e. if \code{ztype} is either \code{"points"} or \code{"image"}.
\code{zlim} is a two-element numerical vector specifying the
limit on the plotted field.  If not provided, it defaults to the data
range.
\code{zbreaks} controls the colour breaks, in a manner that is similar to
the \code{\link[=image]{image()}} parameter named \code{breaks}.  If not provided, \code{zbreaks} is
inferred from \code{zlim}.
\code{zcol}, which controls the colour scheme, may be a vector of colours
(of length 1 less than \code{zbreaks}), or a function that takes an
integer as its sole argument and returns that number of colours.
If not provided, \code{zcol} defaults to \code{\link[=oceColorsViridis]{oceColorsViridis()}}.
These three parameters are used in Example 6, an illustration of
Atlantic salinity along 36N.}

\item{map.xlim, map.ylim}{Optional limits for station map; \code{map.ylim} is
ignored if \code{map.xlim} is provided.}

\item{clongitude, clatitude, span}{Optional map centre position and span (km).}

\item{projection}{Parameter specifying map
projection; see \code{\link[=mapPlot]{mapPlot()}}.  If \code{projection="automatic"},
however, a projection is devised from the data, with \code{stereographic} if
the mean latitude exceeds 70N and \code{mollweide} otherwise.}

\item{xtype}{Type of x axis, for contour plots, either \code{"distance"} for
distance (in km) to the first point in the section, \code{"track"} for distance
along the cruise track, \code{"longitude"}, \code{"latitude"},
\code{"time"} or \code{"spine"} (distance along a spine that was added
with \code{\link[=addSpine]{addSpine()}}).  Note that if the x values are not in order, they will be put in
order, and since that might not make physical sense, a warning will be issued.}

\item{ytype}{Type of y axis for contour plots, either \code{"pressure"} for
pressure (in dbar, with zero at the surface) or \code{"depth"} for depth (in m
below the surface, calculated from pressure with \code{\link[=swDepth]{swDepth()}}).}

\item{ztype}{String indicating whether to how to indicate the "z"
data (in the R sense, i.e. this could be salinity, temperature, etc; it does
not mean the vertical coordinate) The choices are: \code{"contour"} for
contours, \code{"image"} for an image (drawn with \code{\link[=imagep]{imagep()}} with
\code{filledContours=TRUE}), or \code{"points"} to draw points.
In the first two cases, the data must be gridded, with identical pressures at
each station.}

\item{longitude0, latitude0}{Location of the point from which distance is measured.
These values are ignored unless \code{xtype} is \code{"distance"}.}

\item{legend.loc}{Location of legend, as supplied to \code{\link[=legend]{legend()}}, or
set to the empty string to avoid plotting a legend.}

\item{legend.text}{character value indicating the text for the legend.
If this is NULL (the default) then the legend is automatically
constructed by \code{\link[=labelWithUnit]{labelWithUnit()}}, based on the value of \code{which}.}

\item{showStations}{Logical indicating whether to draw station numbers on maps.}

\item{showStart}{Logical indicating whether to indicate the first station with}

\item{stationTicks}{A logical value indicating whether to indicate station
locations with ticks at the top margin of cross-section plots. Setting this
parameter to \code{FALSE} frees the user up to do their own labelling
at this spot.}

\item{showBottom}{a value indicating whether (and how) to indicate the
ocean bottom on cross-section views.  There are three possibilities.
(a) If \code{showBottom} is \code{FALSE}, then the bottom is not rendered.  If it
is \code{TRUE}, then the  bottom is rendered with a gray polygon.
(b) If \code{showBottom} is the character value \code{"polygon"}, then a polygon is drawn,
and similarly lines are drawn for \code{"lines"}, and points for \code{"points"}.
(c) If \code{showBottom} is a \linkS4class{topo} object, then the station locations are
interpolated to that topography and the results are shown with a polygon.
See \dQuote{Examples}.}

\item{showSpine}{logical value used if \code{which="map"}.  If \code{showSpine} is
\code{TRUE} and \code{section} has had a spine added with \code{\link[=addSpine]{addSpine()}}, then
the spine is drawn in blue.}

\item{drawPalette}{Logical value indicating whether to draw a palette when \code{ztype="image"}
ignored otherwise.}

\item{axes}{Logical value indicating whether to draw axes.}

\item{mgp}{A 3-element numerical vector to use for \code{par(mgp)}, and also for
\code{par(mar)}, computed from this. If not provided, this defaults to
\code{\link{getOption}}\code{("oceMgp")}.}

\item{mar}{Value to be used with \code{\link{par}}\code{("mar")}. If not provided,
a default is set up.}

\item{col}{Color for line types.  If not provided, this defaults to
\code{\link{par}}\code{("col")}.  See \code{zcol}, for \code{ztype="image"} and \code{ztype="points"}.}

\item{cex}{Numerical character-expansion factor, which defaults to \code{\link{par}}\code{("cex")}.}

\item{pch}{Indication of symbol type; defaults to \code{\link{par}}\code{("pch")}.}

\item{labcex}{Size of characters in contour labels (passed to
\code{\link[=contour]{contour()}}).}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If \code{debug} is not supplied, it defaults to
\code{\link{getOption}}\code{("oceDebug")}.}

\item{...}{Optional arguments passed to the contouring function.}
}
\value{
If the original section was gridded, the return value is that section.
Otherwise, the gridded section that was constructed for the plot is returned.
In both cases, the value is returned silently. The
purpose of returning the section is to enable subsequent processing
of the grid, including adding elements to the plot (see example 5).
}
\description{
Creates a summary plot for a CTD section, with one panel for each value of
\code{which}.
}
\details{
The type of plot is governed by \code{which}, as follows.
\itemize{
\item \code{which=0} or \code{"potential temperature"} for potential temperature contours
\item \code{which=1} or \code{"temperature"} for in-situ temperature contours (the default)
\item \code{which=2} or \code{"salinity"} for salinity contours
\item \code{which=3} or \code{"sigmaTheta"} for sigma-theta contours
\item \code{which=4} or \code{"nitrate"} for nitrate concentration contours
\item \code{which=5} or \code{"nitrite"} for nitrite concentration contours
\item \code{which=6} or \code{"oxygen"} for oxygen concentration  contours
\item \code{which=7} or \code{"phosphate"} for phosphate concentration contours
\item \code{which=8} or \code{"silicate"} for silicate concentration contours
\item \code{which=9} or \code{"u"} for eastward velocity
\item \code{which=10} or \code{"uz"} for vertical derivative of eastward velocity
\item \code{which=11} or \code{"v"} for northward velocity
\item \code{which=12} or \code{"vz"} for vertical derivative of northward velocity
\item \code{which=20} or \code{"data"} for a dot for each data location
\item \code{which=99} or \code{"map"} for a location map
}

The y-axis for the contours is pressure, plotted in the conventional reversed
form, so that the water surface appears at the top of the plot.  The x-axis is
more complicated. If \code{at} is not supplied, then the routine calculates x
as the distance between the first station in the section and each of the other
stations. (This will produce an error if the stations are not ordered
geographically, because the \code{\link[=contour]{contour()}} routine cannot handle
non-increasing axis coordinates.) If \code{at} is specified, then it is taken
to be the location, in arbitrary units, along the x-axis of labels specified by
\code{labels}; the way this works is designed to be the same as for
\code{\link[=axis]{axis()}}.
}
\examples{
library(oce)
data(section)
sg <- sectionGrid(section)

# 1. start of section, default fields.
plot(head(section))

# 2. Gulf Stream
GS <- subset(section, 109<=stationId&stationId<=129)
GSg <- sectionGrid(GS, p=seq(0, 2000, 100))
plot(GSg, which=c(1, 99), map.ylim=c(34, 42))
par(mfrow=c(2, 1))
plot(GS, which=1, ylim=c(2000, 0), ztype='points',
     zbreaks=seq(0,30,2), pch=20, cex=3)
plot(GSg, which=1, ztype='image', zbreaks=seq(0,30,2))

par(mfrow=c(1, 1))

# 3. Image, with colored dots to indicate grid-data mismatch.
\dontrun{
plot(GSg, which=1, ztype='image')
T <- GS[['temperature']]
col <- oceColorsViridis(100)[rescale(T, rlow=1, rhigh=100)]
points(GS[['distance']],GS[['depth']],pch=20,cex=3,col='white')
points(GS[['distance']],GS[['depth']],pch=20,cex=2.5,col=col)
}


\dontrun{
# 4. Image of Absolute Salinity, with 4-minute bathymetry
# It's easy to calculate the desired area for the bathymetry,
# but for brevity we'll hard-code it. Note that download.topo()
# caches the file locally.
f <- download.topo(west=-80, east=0, south=35, north=40, resolution=4)
t <- read.topo(f)
plot(section, which="SA", xtype="longitude", ztype="image", showBottom=t)
}

\dontrun{
# 5. Temperature with salinity added in red
s <- plot(section, which="temperature")
distance <- s[["distance", "byStation"]]
depth <- s[["station", 1]][["depth"]]
salinity <- matrix(s[["salinity"]], byrow=TRUE, nrow=length(s[["station"]]))
contour(distance, depth, salinity, col=2, add=TRUE)
}

\dontrun{
# 6. Image with controlled colours
plot(section, which="salinity", ztype="image",
    zlim=c(35, 37.5),
    zbreaks=seq(35, 37.5, 0.25),
    zcol=oceColorsTurbo)
}

}
\seealso{
The documentation for \linkS4class{section} explains the
structure of section objects, and also outlines the other functions dealing
with them.

Other functions that plot oce data: 
\code{\link{download.amsr}()},
\code{\link{plot,adp-method}},
\code{\link{plot,adv-method}},
\code{\link{plot,amsr-method}},
\code{\link{plot,argo-method}},
\code{\link{plot,bremen-method}},
\code{\link{plot,cm-method}},
\code{\link{plot,coastline-method}},
\code{\link{plot,ctd-method}},
\code{\link{plot,gps-method}},
\code{\link{plot,ladp-method}},
\code{\link{plot,landsat-method}},
\code{\link{plot,lisst-method}},
\code{\link{plot,lobo-method}},
\code{\link{plot,met-method}},
\code{\link{plot,odf-method}},
\code{\link{plot,rsk-method}},
\code{\link{plot,satellite-method}},
\code{\link{plot,sealevel-method}},
\code{\link{plot,tidem-method}},
\code{\link{plot,topo-method}},
\code{\link{plot,windrose-method}},
\code{\link{plot,xbt-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{tidem-class}}

Other things related to section data: 
\code{\link{[[,section-method}},
\code{\link{[[<-,section-method}},
\code{\link{as.section}()},
\code{\link{handleFlags,section-method}},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{read.section}()},
\code{\link{section-class}},
\code{\link{sectionAddStation}()},
\code{\link{sectionGrid}()},
\code{\link{sectionSmooth}()},
\code{\link{sectionSort}()},
\code{\link{section}},
\code{\link{subset,section-method}},
\code{\link{summary,section-method}}
}
\author{
Dan Kelley
}
\concept{functions that plot oce data}
\concept{things related to section data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/landsat.R
\docType{class}
\name{landsat-class}
\alias{landsat-class}
\title{Class to Store landsat Data}
\description{
This class holds landsat data. Such are available at several
websites (e.g. reference 1).
Although the various functions may work for other satellites, the
discussion here focusses on Landsat 8 and Landsat 7.
}
\section{Slots}{

\describe{
\item{\code{data}}{As with all \code{oce} objects, the \code{data} slot
for \code{landsat} objects is a
\link{list} containing the main data for the object.
{}}

\item{\code{metadata}}{As with all \code{oce} objects, the \code{metadata} slot
for \code{landsat} objects is a \link{list} containing
information about the \code{data} or about the object itself.
{}}

\item{\code{processingLog}}{As with all \code{oce} objects, the \code{processingLog} slot
for \code{landsat} objects is a
\link{list} with entries describing the creation and evolution
of the object. The contents are updated by various \code{oce} functions to
keep a record of processing steps.  Object summaries and
\code{\link[=processingLogShow]{processingLogShow()}} both display the log.}
}}

\section{Modifying slot contents}{


Although the \code{[[<-} operator may permit modification of the contents
of \linkS4class{landsat} objects (see \code{\link{[[<-,landsat-method}}),
it is better to use \code{\link[=oceSetData]{oceSetData()}} and \code{\link[=oceSetMetadata]{oceSetMetadata()}},
because those functions save an entry in the \code{processingLog}
that describes the change.
}

\section{Retrieving slot contents}{


The full contents of the \code{data} and \code{metadata} slots of a \linkS4class{landsat}
object may be retrieved in the standard R way using \code{\link[=slot]{slot()}}. For
example \code{slot(o,"data")} returns the \code{data} slot of an object named \code{o},
and similarly \code{slot(o,"metadata")} returns
the \code{metadata} slot.

The slots may also be obtained with the \code{\link{[[,landsat-method}}
operator, as e.g. \code{o[["data"]]} and \code{o[["metadata"]]}, respectively.

The \code{\link{[[,landsat-method}} operator can also
be used to retrieve items from within the \code{data} and \code{metadata} slots.
For example, \code{o[["temperature"]]} can be used to retrieve temperature
from an object containing that quantity. The rule is that a named
quantity is sought first within the object's \code{metadata} slot,
with the \code{data} slot being checked only if \code{metadata} does not
contain the item. This \code{[[} method can also be used to get
certain derived quantities, if the object contains sufficient
information to calculate them. For example, an object that holds
(practical) salinity, temperature and pressure, along with
longitude and latitude, has sufficient information to compute
Absolute Salinity, and so \code{o[["SA"]]} will yield the
calculated Absolute Salinity.

It is also possible to find items more directly, using \code{\link[=oceGetData]{oceGetData()}} and
\code{\link[=oceGetMetadata]{oceGetMetadata()}}, but neither of these functions can
retrieve derived items.
}

\section{Data storage}{


The data are stored with 16-bit resolution.  Oce
breaks these 16 bits up into most-significant and least-significant bytes.
For example, the aerosol band of a Landsat object named \code{x} are
contained within \code{x@data$aerosol$msb} and \code{x@data$aerosol$lsb},
each of which is a matrix of raw values.  The results may be combined as e.g.
\preformatted{
256L*as.integer(x@data[[i]]$msb) + as.integer(x@data[[i]]$lsb)
}
and this is what is returned by executing \code{x[["aerosol"]]}.

Landsat data files typically occupy approximately a
gigabyte of storage.  That means that corresponding Oce objects are about
the same size, and this can pose significant problems on computers with
less than 8GB of memory.  It is sensible to specify bands of interest when
reading data with \code{\link[=read.landsat]{read.landsat()}}, and also to use
\code{\link[=landsatTrim]{landsatTrim()}} to isolate geographical regions that need
processing.

Experts may need to get direct access to the data, and this is easy because
all Landsat objects (regardless of satellite) use a similar storage form.
Band information is stored in byte form, to conserve space.  Two bytes are
used for each pixel in Landsat-8 objects, with just one for other objects.
For example, if a Landsat-8 object named \code{L} contains the \code{tirs1}
band, the most- and least-significant bytes will be stored in matrices
\code{L@data$tirs1$msb} and \code{L@data$tirs1$lsb}.  A similar Landsat-7
object would have the same items, but \code{msb} would be just the value
\code{0x00}.

Derived bands, which may be added to a landsat object with
\code{\link[=landsatAdd]{landsatAdd()}}, are not stored in byte matrices.  Instead they
are stored in numerical matrices, which means that they use 4X more storage
space for Landsat-8 images, and 8X more storage space for other satellites.
A computer needs at least 8GB of RAM to work with such data.
}

\section{Landsat 8}{


The Landsat 8 satellite has 11 frequency bands, listed below (see reference 2]).
\preformatted{
.------------------------------------------------------------------------------.
| Band | Band                      | Band         | Wavelength    | Resolution |
| No.  | Contents                  | Name         | (micrometers) |   (meters) |
|------+---------------------------+--------------+---------------+------------|
|    1 | Coastal aerosol           | aerosol      |  0.43 -  0.45 |         30 |
|    2 | Blue                      | blue         |  0.45 -  0.51 |         30 |
|    3 | Green                     | green        |  0.53 -  0.59 |         30 |
|    4 | Red                       | red          |  0.64 -  0.67 |         30 |
|    5 | Near Infrared (NIR)       | nir          |  0.85 -  0.88 |         30 |
|    6 | SWIR 1                    | swir1        |  1.57 -  1.65 |         30 |
|    7 | SWIR 2                    | swir2        |  2.11 -  2.29 |         30 |
|    8 | Panchromatic              | panchromatic |  0.50 -  0.68 |         15 |
|    9 | Cirrus                    | cirrus       |  1.36 -  1.38 |         30 |
|   10 | Thermal Infrared (TIRS) 1 | tirs1        | 10.60 - 11.19 |        100 |
|   11 | Thermal Infrared (TIRS) 2 | tirs2        | 11.50 - 12.51 |        100 |
.------------------------------------------------------------------------------.
}
In addition to the above, setting \code{band="terralook"} may be used as
an abbreviation for \code{band=c("red", "green", "nir")}.

Band 8 is panchromatic, and has the highest resolution.  For convenience of
programming, \code{\link[=read.landsat]{read.landsat()}} subsamples the \code{tirs1} and
\code{tirs2} bands to the 30m resolution of the other bands.  See Reference
3 for information about the evolution of Landsat 8 calibration
coefficients, which as of summer 2014 are still subject to change.
}

\section{Landsat 7}{


Band information is as follows (from reference 8).  The names are not official, but
are set up to roughly correspond with Landsat-8 names, according to wavelength.
An exception is the Landsat-7 bands named \code{tirs1} and \code{tirs2}, which
are at two different gain settings, with identical wavelength span for
each, which roughly matches the range of the Landsat-8 bands \code{tirs1}
and \code{tirs2} combined.  This may seem confusing, but it lets code like
\code{plot(im, band="tirs1")} to work with both Landsat-8 and Landsat-7.

\preformatted{
.------------------------------------------------------------------------------.
| Band | Band                      | Band         | Wavelength    | Resolution |
| No.  | Contents                  | Name         | (micrometers) |   (meters) |
|------+---------------------------+--------------+---------------+------------|
|    1 | Blue                      | blue         |  0.45 -  0.52 |         30 |
|    2 | Green                     | green        |  0.52 -  0.60 |         30 |
|    3 | Red                       | red          |  0.63 -  0.69 |         30 |
|    4 | Near IR                   | nir          |  0.77 -  0.90 |         30 |
|    5 | SWIR                      | swir1        |  1.55 -  1.75 |         30 |
|    6 | Thermal IR                | tirs1        | 10.4  - 12.50 |         30 |
|    7 | Thermal IR                | tirs2        | 10.4  - 12.50 |         30 |
|    8 | SWIR                      | swir2        |  2.09 -  2.35 |         30 |
|    9 | Panchromatic              | panchromatic |  0.52 -  0.90 |         15 |
.------------------------------------------------------------------------------.
}
}

\references{
\enumerate{
\item See the USGS "glovis" web site.
\item see landsat.gsfc.nasa.gov/?page_id=5377
\item see landsat.usgs.gov/calibration_notices.php
\item \url{https://dankelley.github.io/r/2014/07/01/landsat.html}
\item \url{https://scienceofdoom.com/2010/12/27/emissivity-of-the-ocean/}
\item see landsat.usgs.gov/Landsat8_Using_Product.php
\item see landsathandbook.gsfc.nasa.gov/pdfs/Landsat7_Handbook.pdf
\item see landsat.usgs.gov/band_designations_landsat_satellites.php
\item Yu, X. X. Guo and Z. Wu., 2014. Land Surface Temperature Retrieval from
Landsat 8 TIRS-Comparison between Radiative Transfer Equation-Based Method,
Split Window Algorithm and Single Channel Method, \emph{Remote Sensing}, 6,
9829-9652.  \verb{https://www.mdpi.com/2072-4292/6/10/9829}
\item Rajeshwari, A., and N. D. Mani, 2014.  Estimation of land surface
temperature of Dindigul district using Landsat 8 data. \emph{International
Journal of Research in Engineering and Technology}, 3(5), 122-126.
\verb{http://www.academia.edu/7655089/ESTIMATION_OF_LAND_SURFACE_TEMPERATURE_OF_DINDIGUL_DISTRICT_USING_LANDSAT_8_DATA}
\item Konda, M. Imasato N., Nishi, K., and T. Toda, 1994.  Measurement of the Sea
Surface Emissivity.  \emph{Journal of Oceanography}, 50, 17:30.
\doi{10.1007/BF02233853}
}
}
\seealso{
Data from AMSR satellites are handled with \linkS4class{amsr}.

A file containing Landsat data may be read with \code{\link[=read.landsat]{read.landsat()}} or
\code{\link[=read.oce]{read.oce()}}, and one such file is provided by the \CRANpkg{ocedata}
package as a dataset named \code{landsat}.

Plots may be made with \code{\link[=plot,landsat-method]{plot,landsat-method()}}.  Since plotting can be quite
slow, decimation is available both in the plotting function and as the separate
function \code{\link[=decimate]{decimate()}}.  Images may be subsetted with
\code{\link[=landsatTrim]{landsatTrim()}}.

Other classes holding satellite data: 
\code{\link{amsr-class}},
\code{\link{g1sst-class}},
\code{\link{satellite-class}}

Other things related to landsat data: 
\code{\link{[[,landsat-method}},
\code{\link{[[<-,landsat-method}},
\code{\link{landsatAdd}()},
\code{\link{landsatTrim}()},
\code{\link{landsat}},
\code{\link{plot,landsat-method}},
\code{\link{read.landsat}()},
\code{\link{summary,landsat-method}}
}
\author{
Dan Kelley and Clark Richards
}
\concept{classes holding satellite data}
\concept{things related to landsat data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{binMean1D}
\alias{binMean1D}
\title{Bin-average f=f(x)}
\usage{
binMean1D(x, f, xbreaks)
}
\arguments{
\item{x}{vector of numerical values.}

\item{f}{vector of numerical values.}

\item{xbreaks}{Vector of values of x at the boundaries between bins, calculated using
\code{\link[=pretty]{pretty()}} if not supplied.}
}
\value{
A list with the following elements: the breaks (\code{xbreaks},
midpoints (\code{xmids}) between those breaks,
the count (\code{number}) of \code{x} values between successive breaks,
and the resultant average (\code{result}) of \code{f}, classified by the
\code{x} breaks.
}
\description{
Average the values of a vector \code{f} in bins defined on another
vector \code{x}. A common example might be averaging CTD profile
data into pressure bins (see \dQuote{Examples}).
}
\examples{
library(oce)
data(ctd)
z <- ctd[["z"]]
T <- ctd[["temperature"]]
plot(T, z)
TT <- binMean1D(z, T, seq(-100, 0, 1))
lines(TT$result, TT$xmids, col='red')

}
\seealso{
Other bin-related functions: 
\code{\link{binApply1D}()},
\code{\link{binApply2D}()},
\code{\link{binAverage}()},
\code{\link{binCount1D}()},
\code{\link{binCount2D}()},
\code{\link{binMean2D}()}
}
\author{
Dan Kelley
}
\concept{bin-related functions}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/xbt.R
\name{plot,xbt-method}
\alias{plot,xbt-method}
\alias{plot.xbt}
\title{Plot an xbt Object}
\usage{
\S4method{plot}{xbt}(
  x,
  which = 1,
  type = "l",
  mgp = getOption("oceMgp"),
  mar,
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{x}{an \linkS4class{xbt} object.}

\item{which}{list of desired plot types; see \dQuote{Details} for the meanings of various
values of \code{which}.}

\item{type}{type of plot, as for \code{\link[=plot]{plot()}}.}

\item{mgp}{3-element numerical vector to use for \code{par(mgp)}, and also for
\code{par(mar)}, computed from this.  The default is tighter than the R default,
in order to use more space for the data and less for the axes.}

\item{mar}{value to be used with \code{\link{par}}\code{("mar")}.}

\item{debug}{a flag that turns on debugging.  Set to 1 to get a moderate amount
of debugging information, or to 2 to get more.}

\item{...}{optional arguments passed to plotting functions.}
}
\description{
Plots data contained in an \linkS4class{xbt} object.
}
\details{
The panels are controlled by the \code{which} argument, with choices as follows.
\itemize{
\item \code{which=1} for a temperature profile as a function of depth.
\item \code{which=2} for a soundSpeed profile as a function of depth.
}
}
\examples{
library(oce)
data(xbt)
summary(xbt)
plot(xbt)

}
\seealso{
Other functions that plot oce data: 
\code{\link{download.amsr}()},
\code{\link{plot,adp-method}},
\code{\link{plot,adv-method}},
\code{\link{plot,amsr-method}},
\code{\link{plot,argo-method}},
\code{\link{plot,bremen-method}},
\code{\link{plot,cm-method}},
\code{\link{plot,coastline-method}},
\code{\link{plot,ctd-method}},
\code{\link{plot,gps-method}},
\code{\link{plot,ladp-method}},
\code{\link{plot,landsat-method}},
\code{\link{plot,lisst-method}},
\code{\link{plot,lobo-method}},
\code{\link{plot,met-method}},
\code{\link{plot,odf-method}},
\code{\link{plot,rsk-method}},
\code{\link{plot,satellite-method}},
\code{\link{plot,sealevel-method}},
\code{\link{plot,section-method}},
\code{\link{plot,tidem-method}},
\code{\link{plot,topo-method}},
\code{\link{plot,windrose-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{tidem-class}}

Other things related to xbt data: 
\code{\link{[[,xbt-method}},
\code{\link{[[<-,xbt-method}},
\code{\link{as.xbt}()},
\code{\link{read.xbt.noaa1}()},
\code{\link{read.xbt}()},
\code{\link{subset,xbt-method}},
\code{\link{summary,xbt-method}},
\code{\link{xbt-class}},
\code{\link{xbt.edf}},
\code{\link{xbt}}
}
\author{
Dan Kelley
}
\concept{functions that plot oce data}
\concept{things related to xbt data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/landsat.R
\name{summary,landsat-method}
\alias{summary,landsat-method}
\title{Summarize a landsat Object}
\usage{
\S4method{summary}{landsat}(object, ...)
}
\arguments{
\item{object}{A \linkS4class{landsat} object.}

\item{...}{Ignored.}
}
\description{
Provides a summary of a some information about a \linkS4class{landsat} object.
}
\seealso{
Other things related to landsat data: 
\code{\link{[[,landsat-method}},
\code{\link{[[<-,landsat-method}},
\code{\link{landsat-class}},
\code{\link{landsatAdd}()},
\code{\link{landsatTrim}()},
\code{\link{landsat}},
\code{\link{plot,landsat-method}},
\code{\link{read.landsat}()}
}
\author{
Dan Kelley
}
\concept{things related to landsat data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{coriolis}
\alias{coriolis}
\title{Coriolis parameter on rotating earth}
\usage{
coriolis(latitude, degrees = TRUE)
}
\arguments{
\item{latitude}{Vector of latitudes in \eqn{^\circ}{deg}N or radians north of the equator.}

\item{degrees}{Flag indicating whether degrees are used for latitude; if set
to \code{FALSE}, radians are used.}
}
\value{
Coriolis parameter, in radian/s.
}
\description{
Compute \eqn{f}{f}, the Coriolis parameter as a function of latitude
(see reference 1),
assuming earth siderial angular rotation rate
\eqn{omega}{omega}=7292115e-11 rad/s. See reference 1 for general notes, and
see reference 2 for comments on temporal variations
of \eqn{omega}{omega}.
}
\examples{
C <- coriolis(45) # 1e-4
}
\references{
\enumerate{
\item Gill, A.E., 1982. \emph{Atmosphere-ocean Dynamics}, Academic
Press, New York, 662 pp.
\item Groten, E., 2004: Fundamental Parameters and Current, 2004. Best
Estimates of the Parameters of Common Relevance to Astronomy, Geodesy,
and Geodynamics. Journal of Geodesy, 77:724-797.
(downloaded from
\verb{http://www.iag-aig.org/attach/e354a3264d1e420ea0a9920fe762f2a0/51-groten.pdf}
March 11, 2017).
}
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/AllClass.R
\name{handleFlags}
\alias{handleFlags}
\title{Handle flags in oce objects}
\usage{
handleFlags(
  object,
  flags = NULL,
  actions = NULL,
  where = NULL,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{object}{an \linkS4class{oce} object.}

\item{flags}{A \link{list} specifying flag values upon which
actions will be taken. This can take two forms.
\itemize{
\item In the first form, the
list has named elements each containing a vector of integers. For example,
salinities flagged with values of 1 or 3:9 would be specified
by \code{flags=list(salinity=c(1,3:9))}. Several data items can be specified,
e.g. \code{flags=list(salinity=c(1,3:9), temperature=c(1,3:9))} indicates
that the actions are to take place for both salinity and temperature.
\item In the second form, \code{flags} is a list holding a single \emph{unnamed} vector, and
this means to apply the actions to \emph{all} the data entries.  For example,
\code{flags=list(c(1,3:9))} means to apply not just to salinity and temperature,
but to everything within the \code{data} slot.
}

If \code{flags} is not provided, then \code{\link[=defaultFlags]{defaultFlags()}} is called, to try to
determine a reasonable default.}

\item{actions}{an optional \link{list} that contains items with
names that match those in the \code{flags} argument.  If \code{actions}
is not supplied, the default will be to set all values identified by
\code{flags} to \code{NA}; this can also be specified by
specifying \code{actions=list("NA")}. It is also possible to specify
functions that calculate replacement values. These are provided
with \code{object} as the single argument, and must return a
replacement for the data item in question.
See \dQuote{Details} for the default that is used if \code{actions} is not supplied.}

\item{where}{an optional character value that permits the function to work with
objects that store flags in e.g. \code{object@metadata$flags$where}
instead of in \code{object@metadata$flags}, and data within
\code{object@data$where} instead of within \code{object@data}. The
default value of \code{NULL} means to look withing \code{object@metadata}
itself, and this is the default within \code{oce}.  (The purpose of \code{where}
is to make \code{oce} extensible by other packages, which may choose to store
data two levels deep in the \code{data} slot.)}

\item{debug}{An optional integer specifying the degree of debugging, with
value 0 meaning to skip debugging and 1 or higher meaning to print some
information about the arguments and the data. It is usually a good idea to set
this to 1 for initial work with a dataset, to see which flags are being
handled for each data item. If not supplied, this defaults to the value of
\code{\link{getOption}}\code{("oceDebug")}.}
}
\description{
Data-quality flags are stored in the \code{metadata}
slot of \linkS4class{oce} objects in a
\link{list} named \code{flags}.
The present function (a generic that has specialized versions
for various data classes) provides a way to
manipulate the contents of the \code{data} slot, based on
such data-quality flags. For example, a common operation is to replace
erroneous data with \code{NA}.

If the \code{flags} within \code{object}'s \code{metadata} slot is empty,
then \code{object} is returned, unaltered.
Otherwise, \code{handleFlags} examines \code{object@metadata$flags}
in the context of the \code{flags} argument, and then
carries out actions that are specified by the \code{actions} argument.
By default, this sets the returned \code{data} entries to \code{NA},
wherever the corresponding \code{metadata$flag} values
signal unreliable data. To maintain a hint as to why
\code{data} were changed, \code{metadata$flags} in the
returned value is a direct copy of the corresponding
entry in \code{object}.
}
\details{
Each specialized variant of this function has its own defaults
for \code{flags} and \code{actions}.
}
\seealso{
Other functions relating to data-quality flags: 
\code{\link{defaultFlags}()},
\code{\link{handleFlags,adp-method}},
\code{\link{handleFlags,argo-method}},
\code{\link{handleFlags,ctd-method}},
\code{\link{handleFlags,oce-method}},
\code{\link{handleFlags,section-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{initializeFlagScheme,oce-method}},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{initializeFlagSchemeInternal}()},
\code{\link{initializeFlagScheme}()},
\code{\link{initializeFlags,adp-method}},
\code{\link{initializeFlags,oce-method}},
\code{\link{initializeFlagsInternal}()},
\code{\link{initializeFlags}()},
\code{\link{setFlags,adp-method}},
\code{\link{setFlags,ctd-method}},
\code{\link{setFlags,oce-method}},
\code{\link{setFlags}()}
}
\concept{functions relating to data-quality flags}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/gps.R
\name{[[<-,gps-method}
\alias{[[<-,gps-method}
\title{Replace Parts of a GPS Object}
\usage{
\S4method{[[}{gps}(x, i, j, ...) <- value
}
\arguments{
\item{x}{a \linkS4class{gps} object.}

\item{i}{character value naming the item to replace.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{optional additional information (ignored).}

\item{value}{The value to be placed into \code{x}, somewhere.}
}
\description{
The \code{[[<-} method works for all \linkS4class{oce} objects.
The purpose, as with the related extraction method, \code{[[},
is to insulate users from the internal details of \linkS4class{oce}
objects, by looking for items within the various storage
slots of the object. Items not actually stored can also be
replaced, including units and data-quality
flags.
}
\details{
As with \code{[[} method, the procedure works in steps.

First, the \code{metadata} slot of \code{x} is checked to
see whether it contains something named with \code{i}.
If so, then the named item is replaced with \code{value}.

Otherwise, if the string value of \code{i} ends in \code{Unit}, then the
characters preceding that are taken as the name of a variable, and
the \code{metadata} slot of \code{x} is updated to store that unit, e.g.\preformatted{x[["temperatureUnits"]] <- list(unit=expression(degree*F),scale="")
}

Similarly, if \code{i} ends in \code{Flag}, then quality-control
flags are set up as defined by \code{result}, e.g.\preformatted{o[["temperatureFlags"]] <- c(2,4,2,2)
}

Otherwise, \code{\link[=pmatch]{pmatch()}} is used for a partial-string match with
the names of the items that are in the \code{data} slot of \code{x}.
The first item found (if any) is then updated to hold the value \code{result}.

If none of these conditions is met, a warning is issued.
}
\seealso{
Other functions that replace parts of oce objects: 
\code{\link{[[<-,adp-method}},
\code{\link{[[<-,amsr-method}},
\code{\link{[[<-,argo-method}},
\code{\link{[[<-,bremen-method}},
\code{\link{[[<-,cm-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{[[<-,echosounder-method}},
\code{\link{[[<-,g1sst-method}},
\code{\link{[[<-,ladp-method}},
\code{\link{[[<-,landsat-method}},
\code{\link{[[<-,lisst-method}},
\code{\link{[[<-,lobo-method}},
\code{\link{[[<-,met-method}},
\code{\link{[[<-,oce-method}},
\code{\link{[[<-,odf-method}},
\code{\link{[[<-,rsk-method}},
\code{\link{[[<-,sealevel-method}},
\code{\link{[[<-,section-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{[[<-,topo-method}},
\code{\link{[[<-,windrose-method}},
\code{\link{[[<-,xbt-method}}

Other things related to gps data: 
\code{\link{[[,gps-method}},
\code{\link{as.gps}()},
\code{\link{gps-class}},
\code{\link{plot,gps-method}},
\code{\link{read.gps}()},
\code{\link{summary,gps-method}}
}
\concept{functions that replace parts of oce objects}
\concept{things related to gps data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swSigma1}
\alias{swSigma1}
\title{Seawater potential density anomaly referenced to 1000db pressure}
\usage{
swSigma1(
  salinity,
  temperature = NULL,
  pressure = NULL,
  longitude = NULL,
  latitude = NULL,
  eos = getOption("oceEOS", default = "gsw")
)
}
\arguments{
\item{salinity}{either practical salinity (in which case \code{temperature}
and \code{pressure} must be provided) \emph{or} an \code{oce} object, in
which case \code{salinity}, \code{temperature} (in the ITS-90 scale; see
next item), etc. are inferred from the object, ignoring the
other parameters, if they are supplied.}

\item{temperature}{\emph{in-situ} temperature (\eqn{^\circ}{deg}C), defined
on the ITS-90 scale.  This scale is used by GSW-style calculation (as
requested by setting \code{eos="gsw"}), and is the value contained within
\code{ctd} objects (and probably most other objects created with data
acquired in the past decade or two). Since the UNESCO-style calculation is
based on IPTS-68, the temperature is converted within the present function,
using \code{\link[=T68fromT90]{T68fromT90()}}.}

\item{pressure}{pressure (dbar)}

\item{longitude}{longitude of observation (only used if \code{eos="gsw"};
see \sQuote{Details}).}

\item{latitude}{latitude of observation (only used if \code{eos="gsw"}; see
\sQuote{Details}).}

\item{eos}{equation of state, either \code{"unesco"} (references 1 and 2)
or \code{"gsw"} (references 3 and 4).}
}
\value{
Potential density anomaly (kg/m\eqn{^3}{^3}).
}
\description{
Compute \eqn{\sigma_\theta}{sigma}, the potential density of seawater (minus
1000 kg/m\eqn{^3}{^3}), referenced to 1000db pressure.
}
\details{
Definition:
\eqn{\sigma_1=\sigma_\theta=\rho(S,\theta(S,t,p),1000}{sigma_1=sigma_theta=rho(S,theta(S,t,p),1000)}
\itemize{
\item 1000 kg/m\eqn{^3}{^3}.
}
}
\references{
See citations provided in the \code{\link[=swRho]{swRho()}} documentation.
}
\seealso{
Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{numberAsHMS}
\alias{numberAsHMS}
\title{Convert a Numeric Time to Hour, Minute, and Second}
\usage{
numberAsHMS(t, default = 0)
}
\arguments{
\item{t}{a vector of factors or character strings, in the format 1200 for
12:00, 0900 for 09:00, etc.}

\item{default}{value to be used for the returned hour, minute and second if
there is something wrong with the input value (e.g. its length exceeds 4
characters, or it contains non-numeric characters)}
}
\value{
A list containing \code{hour}, \code{minute}, and \code{second}, the
last of which is always zero.
}
\description{
Convert a Numeric Time to Hour, Minute, and Second
}
\examples{
t <- c("0900", "1234")
numberAsHMS(t)

}
\seealso{
Other things related to time: 
\code{\link{ctimeToSeconds}()},
\code{\link{julianCenturyAnomaly}()},
\code{\link{julianDay}()},
\code{\link{numberAsPOSIXct}()},
\code{\link{secondsToCtime}()},
\code{\link{unabbreviateYear}()}
}
\author{
Dan Kelley
}
\concept{things related to time}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{retime}
\alias{retime}
\title{Adjust the time within Oce object}
\usage{
retime(x, a, b, t0, debug = getOption("oceDebug"))
}
\arguments{
\item{x}{an \linkS4class{oce} object.}

\item{a}{intercept (in seconds) in linear model of time drift (see
\dQuote{Details}).}

\item{b}{slope (unitless) in linear model of time drift (unitless) (see
\dQuote{Details}).}

\item{t0}{reference time (in \code{\link[=POSIXct]{POSIXct()}} format) used in linear model of time
drift (see \dQuote{Details}).}

\item{debug}{a flag that, if nonzero, turns on debugging.}
}
\value{
A new object, with time and other data adjusted.
}
\description{
This function compensates for drifting instrument clocks, according to
\eqn{t'=t + a + b (t-t0)}{t'=t + a + b*(t-t0)}, where \eqn{t'}{t'} is the
true time and \eqn{t}{t} is the time stored in the object.  A single check
on time mismatch can be described by a simple time offset, with a non-zero
value of \code{a}, a zero value of \code{b}, and an arbitrary value of
\code{t0}.  Checking the mismatch before and after an experiment yields
sufficient information to specify a linear drift, via \code{a}, \code{b},
and \code{t0}.  Note that \code{t0} is just a convenience parameter, which
avoids the user having to know the "zero time" used in R and clarifies the
values of the other two parameters.  It makes sense for \code{t0} to have
the same timezone as the time within \code{x}.
}
\details{
The returned object is computed by linear interpolation, using
\code{\link[=approx]{approx()}} with \code{rule=2}, to avoid \code{NA} values at the
start or end.  The new time will be as given by the formula above. Note that
if the drift is large enough, the sampling rate will be changed.  It is a
good idea to start with an object that has an extended time range, so that,
after this is called, \code{\link[=subset]{subset()}} can be used to trim to a desired
time range.
}
\examples{
library(oce)
data(adv)
adv2 <- retime(adv,0,1e-4,as.POSIXct("2008-07-01 00:00:00",tz="UTC"))
plot(adv[["time"]], adv2[["time"]]-adv[["time"]], type='l')
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{oceAxis}
\alias{oceAxis}
\title{Draw an axis, possibly with decade-style logarithmic scaling}
\usage{
oceAxis(side, labels = TRUE, logStyle = "r", ...)
}
\arguments{
\item{side}{an integer specifying which axis to draw, with 1 for bottom axis, 2 for left axis,
3 for top axis, and 4 for right axis (as with \code{\link[=axis]{axis()}}).}

\item{labels}{either a vector of character values used for labels or a logical value indicating
whether to draw such labels.  The first form only works if the coordinate is not logarithmic,
and if \code{logStyle} is \code{"r"}.}

\item{logStyle}{a character value that indicates how to draw the y axis, if
\code{log="y"}.  If it is \code{"r"} (the default) then the conventional R style is used,
in which a logarithmic transform connects y values to position on the "page"
of the plot device, so that tics will be nonlinearly spaced, but not
organized by integral powers of 10.  However, if it is \code{"decade"}, then
the style will be that used in the scientific literature, in which large
tick marks are used for integral powers of 10, with smaller tick marks
at integral multiples of those powers, and with labels that use exponential
format for values above 100 or below 0.01.}

\item{\dots}{other graphical parameters, passed to \code{\link[=axis]{axis()}}.}
}
\value{
Numerical values at which tick marks were drawn (or would have been drawn, if \code{labels}
specified to draw them).
}
\description{
Draw an axis, possibly with decade-style logarithmic scaling
}
\examples{
library(oce)
Ra <- 10^seq(4, 10, 0.1)
Nu <- 0.085 * Ra^(1/3)
plot(Ra, Nu, log="xy", axes=FALSE)
box()
oceAxis(1, logStyle="decade")
oceAxis(2, logStyle="decade")

}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.R
\name{plotProfile}
\alias{plotProfile}
\title{Plot a CTD Profile}
\usage{
plotProfile(
  x,
  xtype = "salinity+temperature",
  ytype = "pressure",
  eos = getOption("oceEOS", default = "gsw"),
  lty = 1,
  xlab = NULL,
  ylab = NULL,
  col = "black",
  col.salinity = "darkgreen",
  col.temperature = "red",
  col.rho = "blue",
  col.N2 = "brown",
  col.dpdt = "darkgreen",
  col.time = "darkgreen",
  pt.bg = "transparent",
  grid = TRUE,
  col.grid = "lightgray",
  lty.grid = "dotted",
  Slim,
  Clim,
  Tlim,
  densitylim,
  N2lim,
  Rrholim,
  dpdtlim,
  timelim,
  plim,
  xlim,
  ylim,
  lwd = par("lwd"),
  xaxs = "r",
  yaxs = "r",
  cex = 1,
  pch = 1,
  useSmoothScatter = FALSE,
  df,
  keepNA = FALSE,
  type = "l",
  mgp = getOption("oceMgp"),
  mar,
  add = FALSE,
  inset = FALSE,
  debug = getOption("oceDebug", 0),
  ...
)
}
\arguments{
\item{x}{a \linkS4class{ctd} object.}

\item{xtype}{Item(s) plotted on the x axis, either a vector of length equal
to that of \code{pressure} in the \code{data} slot,
or a text code from the list below.
\itemize{
\item \code{"salinity"} Profile of salinity.
\item \code{"conductivity"} Profile of conductivity.
\item \code{"temperature"} Profile of \emph{in-situ} temperature.
\item \code{"theta"} Profile of potential temperature.
\item \code{"density"} Profile of density (expressed as \eqn{\sigma_\theta}{sigma_theta}).
\item \code{"index"} Index of sample (useful for working with \code{\link[=ctdTrim]{ctdTrim()}}).
\item \code{"salinity+temperature"} Profile of salinity and temperature within a single axis frame.
\item \code{"N2"} Profile of square of buoyancy frequency \eqn{N^2}{N^2},
calculated with \code{\link[=swN2]{swN2()}} with
an optional argument setting of \code{df=length(x[["pressure"]])/4} to do
some smoothing.
\item \code{"density+N2"} Profile of sigma0 and
the square of buoyancy frequency within a single axis frame.
\item \code{"density+dpdt"} Profile of sigma0 and dP/dt for the
sensor.  The latter is useful in indicating problems with the deployment.
It is calculated by first differencing pressure and then using a smoothing
spline on the result (to avoid grid-point wiggles that result because the
SBE software only writes 3 decimal places in pressure).  Note that dP/dt may
be off by a scale factor; this should not be a problem if there is a
\code{time} column in the \code{data} slot, or a \code{sample.rate} in the
\code{metadata} slot.
\item \code{"sigma0"}, \code{"sigma1"}, \code{"sigma2"}, \code{"sigma3"}
and \code{"sigma4"} Profile of potential density referenced
to 0dbar (i.e. the surface), 1000dbar, 2000dbar, 3000dbar, and 4000dbar.
\item \code{"spice"} Profile of spice.
\item \code{"Rrho"} Profile of Rrho, defined in the diffusive sense.
\item \code{"RrhoSF"} Profile of Rrho, defined in the salt-finger sense.
}}

\item{ytype}{variable to use on y axis. The valid choices are:
\code{"pressure"} (the default), \code{"z"},
\code{"depth"} and \code{"sigmaTheta"}.}

\item{eos}{equation of state to be used, either \code{"unesco"} or
\code{"gsw"}.}

\item{lty}{line type for the profile.}

\item{xlab}{optional label for x axis (at top of plot). If not
provided, the value of \code{xtype} is used (along with the units
of the variable) to construct a label. For user-supplied \code{xlab},
the unit (if any) must be specified as in e.g.
\code{xlab=expression("Speed ["*m/s*"]")}.}

\item{ylab}{optional label for y axis.  See \code{xlab} for a note on
units.  Setting \code{ylab=""} prevents labelling the axis.}

\item{col}{color for a general profile.}

\item{col.salinity}{color for salinity profile (see \dQuote{Details}).}

\item{col.temperature}{color for temperature (see \dQuote{Details}).}

\item{col.rho}{color for density (see \dQuote{Details}).}

\item{col.N2}{color for square of buoyancy frequency (see
\dQuote{Details}).}

\item{col.dpdt}{color for dP/dt.}

\item{col.time}{color for delta-time.}

\item{pt.bg}{inside color for symbols with \code{pch} in 21:25}

\item{grid}{logical, set to \code{TRUE} to get a grid.}

\item{col.grid}{color for grid.}

\item{lty.grid}{line type for grid.}

\item{Slim}{Optional limit for S axis}

\item{Clim}{Optional limit for conductivity axis}

\item{Tlim}{Optional limit for T axis}

\item{densitylim}{Optional limit for density axis}

\item{N2lim}{Optional limit for N2 axis}

\item{Rrholim}{Optional limit for Rrho axis}

\item{dpdtlim}{Optional limit for dp/dt axis}

\item{timelim}{Optional limit for delta-time axis}

\item{plim}{Optional limit for pressure axis, ignored unless
\code{ytype=="pressure"}, in which case it takes precedence over
\code{ylim}.}

\item{xlim}{Optional limit for x axis, which can apply to any plot type.
This is ignored if the plotted x variable is something for which a limit
may be specified with an argument, e.g. \code{xlim} is ignored for a
salinity profile, because \code{Slim} ought to be given in such a case.}

\item{ylim}{Optional limit for y axis, which can apply to any plot type,
although is overridden by \code{plim} if \code{ytype} is \code{"pressure"}
or by \code{densitylim} if \code{ytype} is \code{"sigmaTheta"}.}

\item{lwd}{lwd value for data line}

\item{xaxs}{value of \code{\link[=par]{par()}} \code{xaxs} to use}

\item{yaxs}{value of \code{\link[=par]{par()}} \code{yaxs} to use}

\item{cex}{size to be used for plot symbols (see \code{\link[=par]{par()}})}

\item{pch}{code for plotting symbol (see \code{\link[=par]{par()}}).}

\item{useSmoothScatter}{boolean, set to \code{TRUE} to use
\code{\link[=smoothScatter]{smoothScatter()}} instead of \code{\link[=plot]{plot()}} to draw the plot.}

\item{df}{optional argument, passed to \code{\link[=swN2]{swN2()}} if provided, and
if a plot using \eqn{N^2}{N^2} is requested.}

\item{keepNA}{FALSE}

\item{type}{type of plot to draw, using the same scheme as
\code{\link[=plot]{plot()}}.}

\item{mgp}{3-element numerical vector to use for \link{par}\code{(mgp)}, and also
for \link{par}\code{(mar)}, computed from this.  The default is tighter than the R
default, in order to use more space for the data and less for the axes.}

\item{mar}{Four-element numerical value to be used to set the plot
margins, with a call to \link{par}\code{(mar)} prior to the plot.
If this is not supplied, a reasonable default will be set up.}

\item{add}{A logical value that controls whether to add to an existing plot.  (It
makes sense to use \code{add=TRUE} in the \code{panel} argument of a
\code{\link[=coplot]{coplot()}}, for example.)}

\item{inset}{A logical value indicating whether to draw an inset plot.
Setting this to \code{TRUE} will prevent the present function from adjusting
the margins, which is
necessary because margin adjustment is the basis for the method used by
\code{\link[=plotInset]{plotInset()}}.}

\item{debug}{a flag that turns on debugging.  Set to 1 to get a moderate
amount of debugging information, or to 2 to get more.}

\item{\dots}{optional arguments passed to other functions.  A common example
is to set \code{df}, for use in \code{\link[=swN2]{swN2()}} calculations.}
}
\value{
None.
}
\description{
Plot a profile, showing variation of some quantity (or quantities) with
pressure, using the oceanographic convention of putting lower pressures
nearer the top of the plot. This works for any \code{oce} object that has a
pressure column in its \code{data} slot.
The colors (\code{col.salinity}, etc.) are only used if two profiles appear
on a plot.
}
\examples{
library(oce)
data(ctd)
plotProfile(ctd, xtype="temperature")

}
\seealso{
\code{\link[=read.ctd]{read.ctd()}} scans ctd information from a file,
\code{\link[=plot,ctd-method]{plot,ctd-method()}} is a general plotting function for \code{ctd}
objects, and \code{\link[=plotTS]{plotTS()}} plots a temperature-salinity diagrams.

Other functions that plot oce data: 
\code{\link{download.amsr}()},
\code{\link{plot,adp-method}},
\code{\link{plot,adv-method}},
\code{\link{plot,amsr-method}},
\code{\link{plot,argo-method}},
\code{\link{plot,bremen-method}},
\code{\link{plot,cm-method}},
\code{\link{plot,coastline-method}},
\code{\link{plot,ctd-method}},
\code{\link{plot,gps-method}},
\code{\link{plot,ladp-method}},
\code{\link{plot,landsat-method}},
\code{\link{plot,lisst-method}},
\code{\link{plot,lobo-method}},
\code{\link{plot,met-method}},
\code{\link{plot,odf-method}},
\code{\link{plot,rsk-method}},
\code{\link{plot,satellite-method}},
\code{\link{plot,sealevel-method}},
\code{\link{plot,section-method}},
\code{\link{plot,tidem-method}},
\code{\link{plot,topo-method}},
\code{\link{plot,windrose-method}},
\code{\link{plot,xbt-method}},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{tidem-class}}

Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}
}
\author{
Dan Kelley
}
\concept{functions that plot oce data}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/cm.R
\name{summary,cm-method}
\alias{summary,cm-method}
\title{Summarize a CM Object}
\usage{
\S4method{summary}{cm}(object, ...)
}
\arguments{
\item{object}{A \linkS4class{cm} object.}

\item{...}{Further arguments passed to or from other methods.}
}
\description{
Summarizes some of the data in a \code{cm} object, presenting such information
as the station name, sampling location, data ranges, etc.
}
\examples{
library(oce)
data(cm)
summary(cm)

}
\seealso{
The documentation for the \linkS4class{cm} class explains the structure
of \code{cm} objects, and also outlines the other functions dealing with them.

Other things related to cm data: 
\code{\link{[[,cm-method}},
\code{\link{[[<-,cm-method}},
\code{\link{as.cm}()},
\code{\link{cm-class}},
\code{\link{cm}},
\code{\link{plot,cm-method}},
\code{\link{read.cm}()},
\code{\link{rotateAboutZ}()},
\code{\link{subset,cm-method}}
}
\author{
Dan Kelley
}
\concept{things related to cm data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{detrend}
\alias{detrend}
\title{Detrend a set of observations}
\usage{
detrend(x, y)
}
\arguments{
\item{x}{a vector of numerical values.  If \code{y} is not given, then
\code{x} is taken for \code{y}.}

\item{y}{an optional vector}
}
\value{
A list containing \code{Y}, the detrended version of \code{y}, and
the intercept \code{a} and slope \code{b} of the linear function of \code{x}
that is subtracted from \code{y} to yield \code{Y}.
}
\description{
Detrends \code{y} by subtracting a linear trend in \code{x}, to create
a vector that is zero for its first and last finite value.
If the second parameter (\code{y}) is missing, then \code{x} is
taken to be \code{y}, and a new \code{x} is constructed with
\code{\link[=seq_along]{seq_along()}}.  Any \code{NA} values are left as-is.
}
\details{
A common application is to bring the end points of a time series
down to zero, prior to applying a digital filter. (See examples.)
}
\examples{
x <- seq(0, 0.9 * pi, length.out=50)
y <- sin(x)
y[1] <- NA
y[10] <- NA
plot(x, y, ylim=c(0, 1))
d <- detrend(x, y)
points(x, d$Y, pch=20)
abline(d$a, d$b, col='blue')
abline(h=0)
points(x, d$Y + d$a + d$b * x, col='blue', pch='+')
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/gps.R
\name{plot,gps-method}
\alias{plot,gps-method}
\alias{plot.gps}
\title{Plot a gps Object}
\usage{
\S4method{plot}{gps}(
  x,
  xlab = "",
  ylab = "",
  asp,
  clongitude,
  clatitude,
  span,
  projection,
  expand = 1,
  mgp = getOption("oceMgp"),
  mar = c(mgp[1] + 1, mgp[1] + 1, 1, 1),
  bg,
  axes = TRUE,
  cex.axis = par("cex.axis"),
  add = FALSE,
  inset = FALSE,
  geographical = 0,
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{x}{a \linkS4class{gps} object.}

\item{xlab}{label for x axis}

\item{ylab}{label for y axis}

\item{asp}{Aspect ratio for plot.  The default is for \code{plot.gps} to set
the aspect ratio to give natural latitude-longitude scaling somewhere near
the centre latitude on the plot. Often, it makes sense to set \code{asp}
yourself, e.g. to get correct shapes at 45N, use
\code{asp=1/cos(45*pi/180)}.  Note that the land mass is not symmetric about
the equator, so to get good world views you should set \code{asp=1} or set
\code{ylim} to be symmetric about zero. Any given value of \code{asp} is
ignored, if \code{clongitude} and \code{clatitude} are given.}

\item{clongitude, clatitude}{optional center latitude of map, in decimal
degrees.  If both \code{clongitude} and \code{clatitude} are provided, then
any provided value of \code{asp} is ignored, and instead the plot aspect
ratio is computed based on the center latitude.  If \code{clongitude} and
\code{clatitude} are provided, then \code{span} must also be provided.}

\item{span}{optional suggested span of plot, in kilometers.  The suggestion
is an upper limit on the scale; depending on the aspect ratio of the
plotting device, the radius may be smaller than \code{span}.  A value for
\code{span} must be supplied, if \code{clongitude} and \code{clatitude} are
supplied.}

\item{projection}{optional map projection to use (see
\code{\link[=mapPlot]{mapPlot()}}); if not given, a cartesian frame is used, scaled so
that gps shapes near the centre of the plot are preserved.  If a projection
is provided, the coordinate system will bear an indirect relationship to
longitude and longitude, and further adornment of the plot must be done with
e.g.  \code{\link[=mapPoints]{mapPoints()}} instead of \code{\link[=points]{points()}}.}

\item{expand}{numerical factor for the expansion of plot limits, showing
area outside the plot, e.g. if showing a ship track as a gps, and then an
actual gps to show the ocean boundary.  The value of \code{expand} is
ignored if either \code{xlim} or \code{ylim} is given.}

\item{mgp}{3-element numerical vector to use for \code{par(mgp)}, and also
for \code{par(mar)}, computed from this.  The default is tighter than the R
default, in order to use more space for the data and less for the axes.}

\item{mar}{value to be used with \code{\link{par}}\code{("mar")}.}

\item{bg}{optional color to be used for the background of the map.  This
comes in handy for drawing insets (see \dQuote{details}).}

\item{axes}{boolean, set to \code{TRUE} to plot axes.}

\item{cex.axis}{value for axis font size factor.}

\item{add}{boolean, set to \code{TRUE} to draw the gps on an existing plot.
Note that this retains the aspect ratio of that existing plot, so it is
important to set that correctly, e.g. with \code{asp=1/cos(lat * pi / 180)},
where \code{clat} is the central latitude of the plot.}

\item{inset}{set to \code{TRUE} for use within \code{\link[=plotInset]{plotInset()}}.  The
effect is to prevent the present function from adjusting margins, which is
necessary because margin adjustment is the basis for the method used by
\code{\link[=plotInset]{plotInset()}}.}

\item{geographical}{flag indicating the style of axes.  If
\code{geographical=0}, the axes are conventional, with decimal degrees as
the unit, and negative signs indicating the southern and western
hemispheres.  If \code{geographical=1}, the signs are dropped, with axis
values being in decreasing order within the southern and western
hemispheres.  If \code{geographical=2}, the signs are dropped and the axes
are labelled with degrees, minutes and seconds, as appropriate.}

\item{debug}{set to \code{TRUE} to get debugging information during
processing.}

\item{\dots}{optional arguments passed to plotting functions.  For example,
set \code{yaxp=c(-90,90,4)} for a plot extending from pole to pole.}
}
\description{
This function plots a gps object.  An attempt is made to use the whole space
of the plot, and this is done by limiting either the longitude range or the
latitude range, as appropriate, by modifying the eastern or northern limit,
as appropriate.
To get an inset map inside another map, draw the first map, do
\code{par(new=TRUE)}, and then call \code{plot.gps} with a value of
\code{mar} that moves the inset plot to a desired location on the existing
plot, and with \code{bg="white"}.
}
\seealso{
Other functions that plot oce data: 
\code{\link{download.amsr}()},
\code{\link{plot,adp-method}},
\code{\link{plot,adv-method}},
\code{\link{plot,amsr-method}},
\code{\link{plot,argo-method}},
\code{\link{plot,bremen-method}},
\code{\link{plot,cm-method}},
\code{\link{plot,coastline-method}},
\code{\link{plot,ctd-method}},
\code{\link{plot,ladp-method}},
\code{\link{plot,landsat-method}},
\code{\link{plot,lisst-method}},
\code{\link{plot,lobo-method}},
\code{\link{plot,met-method}},
\code{\link{plot,odf-method}},
\code{\link{plot,rsk-method}},
\code{\link{plot,satellite-method}},
\code{\link{plot,sealevel-method}},
\code{\link{plot,section-method}},
\code{\link{plot,tidem-method}},
\code{\link{plot,topo-method}},
\code{\link{plot,windrose-method}},
\code{\link{plot,xbt-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{tidem-class}}

Other things related to gps data: 
\code{\link{[[,gps-method}},
\code{\link{[[<-,gps-method}},
\code{\link{as.gps}()},
\code{\link{gps-class}},
\code{\link{read.gps}()},
\code{\link{summary,gps-method}}
}
\author{
Dan Kelley
}
\concept{functions that plot oce data}
\concept{things related to gps data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adv.R
\name{xyzToEnuAdv}
\alias{xyzToEnuAdv}
\title{Convert an ADP from XYZ to ENU Coordinates}
\usage{
xyzToEnuAdv(
  x,
  declination = 0,
  cabled = FALSE,
  horizontalCase,
  sensorOrientation,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{x}{an \linkS4class{adv} object.}

\item{declination}{magnetic declination to be added to the heading, to get
ENU with N as "true" north.}

\item{cabled}{boolean value indicating whether the sensor head is connected
to the pressure case with a cable.  If \code{cabled=FALSE}, then
\code{horizontalCase} is ignored.  See \dQuote{Details}.}

\item{horizontalCase}{optional boolean value indicating whether the sensor
case is oriented horizontally.  Ignored unless \code{cabled} is \code{TRUE}.
See \dQuote{Details}.}

\item{sensorOrientation}{optional string indicating the direction in which
the sensor points.  The value, which must be \code{"upward"} or
\code{"downward"}, over-rides the value of \code{orientation},
in the \code{metadata} slot,
which will have been set by \code{\link[=read.adv]{read.adv()}}, \emph{provided} that the
data file contained the full header.  See \dQuote{Details}.}

\item{debug}{a flag that, if non-zero, turns on debugging.  Higher values
yield more extensive debugging.}
}
\description{
Convert ADV velocity components from a xyz-based coordinate system to
an enu-based coordinate system.
}
\details{
The coordinate transformation is done using the heading, pitch, and roll
information contained within \code{x}.  The algorithm is similar to that
used for Teledyne/RDI ADCP units, taking into account the different
definitions of heading, pitch, and roll as they are defined for the
velocimeters.

Generally, the transformation must be done on a time-by-time basis, which is
a slow operation.  However, this function checks whether the vectors for
heading, pitch and roll, are all of unit length, and in that case, the
calculation is altered, resulting in shorter execution times.  Note that
the angles are held in (\code{data$timeSlow}, \code{data$headingSlow}, ...)
for Nortek instruments and (\code{data$time}, \code{data$heading}, ...) for
Sontek instruments.

Since the documentation provided by instrument manufacturers can be vague on
the coordinate transformations, the method used here had to be developed
indirectly.  (This is in contrast to the RDI ADCP instruments, for which
there are clear instructions.)  documents that manufacturers provide.  If
results seem incorrect (e.g. if currents go east instead of west), users
should examine the code in detail for the case at hand.  The first step is
to set \code{debug} to 1, so that the processing will print a trail of
processing steps.  The next step should be to consult the table below, to
see if it matches the understanding (or empirical tests) of the user.  It
should not be difficult to tailor the code, if needed.

The code handles every case individually, based on the table given below.
The table comes from Clark Richards, a former PhD student at Dalhousie
University (reference 2), who developed it based on instrument documentation,
discussion on user groups, and analysis of measurements acquired with Nortek
and Sontek velocimeters in the SLEIWEX experiment.

The column labelled \verb{Cabled'' indicates whether the sensor and the pressure case are connected with a flexible cable, and the column labelled }H.case''
indicates whether the pressure case is oriented horizontally.  These two
properties are not discoverable in the headers of the data files, and so
they must be supplied with the arguments \code{cabled} and
\code{horizontalCase}.  The source code refers to the information in this
table by case numbers.  (Cases 5 and 6 are not handled.)  Angles are
abbreviated as follows:: heading \verb{H,'' pitch }P,'' and roll ``R''.
Entries X, Y and Z refer to instrument coordinates of the same names.
Entries S, F and M refer to so-called ship coordinates starboard, forward,
and mast; it is these that are used together with a rotation matrix to get
velocity components in the east, north, and upward directions.

\tabular{rrrrrrrrrrrr}{
\strong{\code{Case}} \tab \strong{\code{Mfr.}} \tab
\strong{\code{Instr.}} \tab \strong{\code{Cabled}} \tab \strong{\verb{H. case}} \tab
\strong{\code{Orient.}} \tab \strong{\code{H}} \tab \strong{\code{P}} \tab \strong{\code{R}} \tab
\strong{\code{S}} \tab \strong{\code{F}} \tab \strong{\code{M}}\cr

1 \tab Nortek \tab vector \tab
no \tab - \tab up \tab H-90 \tab R \tab -P \tab X \tab -Y \tab -Z\cr

2 \tab Nortek \tab vector \tab no \tab - \tab down \tab H-90 \tab R \tab -P \tab X
\tab Y \tab Z\cr

3 \tab Nortek \tab vector \tab yes \tab yes \tab up \tab
H-90 \tab R \tab -P \tab X \tab Y \tab Z\cr

4 \tab Nortek \tab vector \tab
yes \tab yes \tab down \tab H-90 \tab R \tab P \tab X \tab -Y \tab -Z\cr

5 \tab Nortek \tab vector \tab yes \tab no \tab up \tab - \tab - \tab - \tab -
\tab - \tab -\cr

6 \tab Nortek \tab vector \tab yes \tab no \tab down \tab -
\tab - \tab - \tab - \tab - \tab -\cr

7 \tab Sontek \tab adv \tab - \tab -
\tab up \tab H-90 \tab R \tab -P \tab X \tab -Y \tab -Z\cr

8 \tab Sontek \tab adv \tab - \tab - \tab down \tab H-90 \tab R \tab -P \tab X \tab Y \tab Z\cr
}
}
\references{
\enumerate{
\item @template nortekCoordTemplate
\item Clark Richards, 2012, PhD Dalhousie University Department of
Oceanography.
}
}
\seealso{
See \code{\link[=read.adv]{read.adv()}} for notes on functions relating to
\code{adv} objects.

Other things related to adv data: 
\code{\link{[[,adv-method}},
\code{\link{[[<-,adv-method}},
\code{\link{adv-class}},
\code{\link{adv}},
\code{\link{beamName}()},
\code{\link{beamToXyz}()},
\code{\link{enuToOtherAdv}()},
\code{\link{enuToOther}()},
\code{\link{plot,adv-method}},
\code{\link{read.adv.nortek}()},
\code{\link{read.adv.sontek.adr}()},
\code{\link{read.adv.sontek.serial}()},
\code{\link{read.adv.sontek.text}()},
\code{\link{read.adv}()},
\code{\link{rotateAboutZ}()},
\code{\link{subset,adv-method}},
\code{\link{summary,adv-method}},
\code{\link{toEnuAdv}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley, in collaboration with Clark Richards
}
\concept{things related to adv data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swThermalConductivity}
\alias{swThermalConductivity}
\title{Seawater thermal conductivity}
\usage{
swThermalConductivity(salinity, temperature = NULL, pressure = NULL)
}
\arguments{
\item{salinity}{salinity (PSU), or a \code{ctd} object, in which case
\code{temperature} and \code{pressure} will be ignored.}

\item{temperature}{\emph{in-situ} temperature (\eqn{^\circ}{deg}C), defined
on the ITS-90 scale; see \dQuote{Temperature units} in the documentation for
\code{\link[=swRho]{swRho()}}.}

\item{pressure}{pressure (dbar)}
}
\value{
Conductivity of seawater in \eqn{W m^{-1\,\circ}C^{-1}}{W/(m*degC)}.
To calculate thermal diffusivity in \eqn{m^2/s}{m^2/s}, divide by the
product of density and specific heat, as in the example.
}
\description{
Compute seawater thermal conductivity, in \eqn{W
m^{-1\circ}C^{-1}}{W/(m*degC)}
}
\details{
Caldwell's (1974) detailed formulation is used.  To be specific, his
equation 6 to calculate K, and his two sentences above that equation are
used to infer this to be K(0,T,S) in his notation of equation 7. Then,
application of his equations 7 and 8 is straightforward. He states an
accuracy for this method of 0.3 percent.  (See the check against his Table 1
in the \dQuote{Examples}.)
}
\examples{
library(oce)
# Values in m^2/s, a unit that is often used instead of W/(m*degC).
swThermalConductivity(35, 10, 100) / (swRho(35,10,100) * swSpecificHeat(35,10,100)) # ocean
swThermalConductivity(0, 20, 0) / (swRho(0, 20, 0) * swSpecificHeat(0, 20, 0)) # lab
# Caldwell Table 1 gives 1478e-6 cal/(cm*sec*degC) at 31.5 o/oo, 10degC, 1kbar
joulePerCalorie <- 4.18400
cmPerM <- 100
swThermalConductivity(31.5,10,1000) / joulePerCalorie / cmPerM

}
\references{
Caldwell, Douglas R., 1974. Thermal conductivity of seawater,
\emph{Deep-sea Research}, \emph{21}, 131-137.
}
\seealso{
Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.R
\name{initializeFlagScheme,ctd-method}
\alias{initializeFlagScheme,ctd-method}
\title{Establish a data-quality scheme for a ctd object}
\usage{
\S4method{initializeFlagScheme}{ctd}(
  object,
  name = NULL,
  mapping = NULL,
  default = NULL,
  update = NULL,
  debug = 0
)
}
\arguments{
\item{object}{An oce object.}

\item{name}{a character value naming the scheme. If this refers
to a pre-defined scheme, then \code{mapping} must not be provided,
because doing so would contradict the pre-defined scheme, defeating
its purpose of providing concreteness and clarity.}

\item{mapping}{a list of named items describing the mapping from
flag meaning to flag numerical value, e.g \code{list(good=1, bad=2)}
might be used for a hypothetical class.}

\item{default}{an integer vector of flag values that are not considered
to be good. If this is not provided, but if \code{name} is \code{"argo"},
\code{"BODC"},
\code{"DFO"},
\code{"WHP bottle"}, or
\code{"WHP CTD"}, then a conservative value will be set automatically,
equal to the list of flag values that designate bad or questionable data.
For example, for \code{name="WHP CTD"}, the setting will be
\code{c(1,3,4,5,6,7,9)}, leaving only value \code{2}, which corresponds
with "acceptable" in the notation used for that flag scheme.}

\item{update}{a logical value indicating whether the scheme provided is
to update an existing scheme.  The default value, \code{FALSE}, prevents such
an attempt to alter an existing flag scheme, if one is already embedded
in \code{object}.}

\item{debug}{an integer set to 0 for quiet action or to 1 for some debugging.}
}
\value{
An object with the \code{metadata} slot containing \code{flagScheme}.
}
\description{
This function stores add an item named \code{flagScheme}
to the \code{metadata} slot of an object inheriting from
\linkS4class{ctd}. This is a list containing two
items: \code{name} and \code{mapping}, as provided in the
function arguments.
The purpose is both to document a flag scheme
and to make it so that \code{\link[=initializeFlags]{initializeFlags()}},
\code{\link[=setFlags]{setFlags()}} and \code{\link[=handleFlags]{handleFlags()}}
can specify flags by
name, as opposed to number. This is a generic function,
that may be specialized to the class of \code{object}
(see \dQuote{Details}).
}
\details{
The following pre-defined schemes are available (note that the
names are simplified from the phrases used in defining
documentation):
\itemize{
\item \code{name="argo"} defaults \code{mapping} to
OLD (prior to June 10, 2020)
}\preformatted{list(not_assessed=0, passed_all_tests=1, probably_good=2,
     probably_bad=3, bad=4, averaged=7,
     interpolated=8, missing=9)
}

NEW (after June 10, 2020)\preformatted{list(not_assessed=0, passed_all_tests=1, probably_good=2,
     probably_bad=3, bad=4, changed=5, not_used_6=6, not_used_7=7,
     estimated=8, missing=9)
}

See reference 1 for a deeper explanation of the meanings of these codes.
\itemize{
\item \code{name="BODC"} defaults \code{mapping} to
}\preformatted{list(no_quality_control=0, good=1, probably_good=2,
     probably_bad=3, bad=4, changed=5,
     below_detection=6, in_excess=7, interpolated=8,
     missing=9)
}

See reference 2 for a deeper explanation of the meanings of these codes,
and note that codes \code{A} and \code{Q} are not provided in
oce.
\itemize{
\item \code{name="DFO"} defaults \code{mapping} to
}\preformatted{list(no_quality_control=0, appears_correct=1, appears_inconsistent=2,
     doubtful=3, erroneous=4, changed=5,
     qc_by_originator=8, missing=9)
}

See reference 3 for a deeper explanation of the meanings of these codes.
\itemize{
\item \code{name="WHP bottle"} defaults \code{mapping} to
}\preformatted{list(no_information=1, no_problems_noted=2, leaking=3,
     did_not_trip=4, not_reported=5, discrepency=6,
     unknown_problem=7, did_not_trip=8, no_sample=9)
}

See reference 4 for a deeper explanation of the meanings of these codes.
\itemize{
\item \code{name="WHP CTD"} defaults \code{mapping} to
}\preformatted{list(not_calibrated=1, acceptable=2, questionable=3,
    bad=4, not_reported=5, interpolated=6,
    despiked=7, missing=9)
}

See reference 4 for a deeper explanation of the meanings of these codes.
}
\references{
\enumerate{
\item The codes for \code{"argo"} are derived from information in Table 4.1
of Wong, Annie, Robert Keeley, Thierry Carval, and Argo Data Management Team
(8 January 2020), "Argo Quality Control Manual for CTD and Trajectory Data, Version 3.3,"
available at \url{https://archimer.ifremer.fr/doc/00228/33951/}
as of June 2020.
\item The codes for \code{"BODC"} are defined at
http://seadatanet.maris2.nl/v_bodc_vocab_v2/browse.asp?order=conceptid&formname=search&screen=0&lib=l20
\item The codes for \code{"DFO"} are defined at
http://www.dfo-mpo.gc.ca/science/data-donnees/code/list/014-eng.html
\item The codes for \code{"WHP CTD"} and \code{"WHP bottle"} are defined at
https://www.nodc.noaa.gov/woce/woce_v3/wocedata_1/whp/exchange/exchange_format_desc.htm
}
}
\seealso{
Other functions relating to data-quality flags: 
\code{\link{defaultFlags}()},
\code{\link{handleFlags,adp-method}},
\code{\link{handleFlags,argo-method}},
\code{\link{handleFlags,ctd-method}},
\code{\link{handleFlags,oce-method}},
\code{\link{handleFlags,section-method}},
\code{\link{handleFlags}()},
\code{\link{initializeFlagScheme,oce-method}},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{initializeFlagSchemeInternal}()},
\code{\link{initializeFlagScheme}()},
\code{\link{initializeFlags,adp-method}},
\code{\link{initializeFlags,oce-method}},
\code{\link{initializeFlagsInternal}()},
\code{\link{initializeFlags}()},
\code{\link{setFlags,adp-method}},
\code{\link{setFlags,ctd-method}},
\code{\link{setFlags,oce-method}},
\code{\link{setFlags}()}

Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}
}
\concept{functions relating to data-quality flags}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.R
\name{xyzToEnuAdpAD2CP}
\alias{xyzToEnuAdpAD2CP}
\title{Convert ADP2CP adp object From XYZ to ENU Coordinates}
\usage{
xyzToEnuAdpAD2CP(x, declination = 0, debug = getOption("oceDebug"))
}
\arguments{
\item{x}{an \linkS4class{adp} object created by \code{\link[=read.adp.ad2cp]{read.adp.ad2cp()}}.}

\item{declination}{IGNORED at present, but will be used at some later time.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\value{
An object with \code{data$v[,,1:3]} altered appropriately, and
\code{x[["oceCoordinate"]]} changed from \code{xyz} to \code{enu}.
}
\description{
\strong{This function will b in active development through the early
months of 2019, and both the methodology and user interface may change
without notice. Only developers (or invitees) should be trying to
use this function.}
}
\section{Limitations}{

This only works if the instrument orientation is \code{"AHRS"}, and even
that is not tested yet. Plus, as noted, the declination is ignored.
}

\references{
\enumerate{
\item Nortek AS. \dQuote{Signature Integration 55|250|500|1000kHz.} Nortek AS, 2017.
\item Nortek AS. \dQuote{Signature Integration 55|250|500|1000kHz.} Nortek AS, 2018.
https://www.nortekgroup.com/assets/software/N3015-007-Integrators-Guide-AD2CP_1018.pdf.
}
}
\seealso{
Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley
}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adv.R
\docType{class}
\name{adv-class}
\alias{adv-class}
\title{Class to Store adv Data}
\description{
This class holds data from acoustic-Doppler velocimeters.
}
\details{
A file containing ADV data is usually recognized by Oce, and so
\code{\link[=read.oce]{read.oce()}} will usually read the data.  If not, one may use the
general ADV function \code{\link[=read.adv]{read.adv()}} or specialized variants
\code{\link[=read.adv.nortek]{read.adv.nortek()}}, \code{\link[=read.adv.sontek.adr]{read.adv.sontek.adr()}} or
\code{\link[=read.adv.sontek.text]{read.adv.sontek.text()}}.

ADV data may be plotted with \code{\link[=plot,adv-method]{plot,adv-method()}} function, which is a
generic function so it may be called simply as \code{plot(x)}, where
\code{x} is an \linkS4class{adv} object.

Statistical summaries of ADV data are provided by the generic function
\code{\link[=summary,adv-method]{summary,adv-method()}}.

Conversion from beam to xyz coordinates may be done with
\code{\link[=beamToXyzAdv]{beamToXyzAdv()}}, and from xyz to enu (east north up) may be done
with \code{\link[=xyzToEnuAdv]{xyzToEnuAdv()}}.  \code{\link[=toEnuAdv]{toEnuAdv()}} may be used to
transfer either beam or xyz to enu.  Enu may be converted to other
coordinates (e.g. aligned with a coastline) with
\code{\link[=enuToOtherAdv]{enuToOtherAdv()}}.
}
\section{Slots}{

\describe{
\item{\code{data}}{As with all \code{oce} objects, the \code{data} slot
for \code{adv} objects is a
\link{list} containing the main data for the object.
The key items stored in this slot include \code{time} and \code{v}.}

\item{\code{metadata}}{As with all \code{oce} objects, the \code{metadata} slot
for \code{adv} objects is a \link{list} containing
information about the \code{data} or about the object itself.
Examples that are of common interest include \code{frequency}, \code{oceCordinate}, and \code{frequency}.}

\item{\code{processingLog}}{As with all \code{oce} objects, the \code{processingLog} slot
for \code{adv} objects is a
\link{list} with entries describing the creation and evolution
of the object. The contents are updated by various \code{oce} functions to
keep a record of processing steps.  Object summaries and
\code{\link[=processingLogShow]{processingLogShow()}} both display the log.}
}}

\section{Modifying slot contents}{


Although the \code{[[<-} operator may permit modification of the contents
of \linkS4class{adv} objects (see \code{\link{[[<-,adv-method}}),
it is better to use \code{\link[=oceSetData]{oceSetData()}} and \code{\link[=oceSetMetadata]{oceSetMetadata()}},
because those functions save an entry in the \code{processingLog}
that describes the change.
}

\section{Retrieving slot contents}{


The full contents of the \code{data} and \code{metadata} slots of a \linkS4class{adv}
object may be retrieved in the standard R way using \code{\link[=slot]{slot()}}. For
example \code{slot(o,"data")} returns the \code{data} slot of an object named \code{o},
and similarly \code{slot(o,"metadata")} returns
the \code{metadata} slot.

The slots may also be obtained with the \code{\link{[[,adv-method}}
operator, as e.g. \code{o[["data"]]} and \code{o[["metadata"]]}, respectively.

The \code{\link{[[,adv-method}} operator can also
be used to retrieve items from within the \code{data} and \code{metadata} slots.
For example, \code{o[["temperature"]]} can be used to retrieve temperature
from an object containing that quantity. The rule is that a named
quantity is sought first within the object's \code{metadata} slot,
with the \code{data} slot being checked only if \code{metadata} does not
contain the item. This \code{[[} method can also be used to get
certain derived quantities, if the object contains sufficient
information to calculate them. For example, an object that holds
(practical) salinity, temperature and pressure, along with
longitude and latitude, has sufficient information to compute
Absolute Salinity, and so \code{o[["SA"]]} will yield the
calculated Absolute Salinity.

It is also possible to find items more directly, using \code{\link[=oceGetData]{oceGetData()}} and
\code{\link[=oceGetMetadata]{oceGetMetadata()}}, but neither of these functions can
retrieve derived items.
}

\examples{
data(adv)
adv[["v"]] <- 0.001 + adv[["v"]] # add 1mm/s to all velocity components

}
\seealso{
Other classes provided by oce: 
\code{\link{adp-class}},
\code{\link{argo-class}},
\code{\link{bremen-class}},
\code{\link{cm-class}},
\code{\link{coastline-class}},
\code{\link{ctd-class}},
\code{\link{lisst-class}},
\code{\link{lobo-class}},
\code{\link{met-class}},
\code{\link{oce-class}},
\code{\link{odf-class}},
\code{\link{rsk-class}},
\code{\link{sealevel-class}},
\code{\link{section-class}},
\code{\link{topo-class}},
\code{\link{windrose-class}},
\code{\link{xbt-class}}

Other things related to adv data: 
\code{\link{[[,adv-method}},
\code{\link{[[<-,adv-method}},
\code{\link{adv}},
\code{\link{beamName}()},
\code{\link{beamToXyz}()},
\code{\link{enuToOtherAdv}()},
\code{\link{enuToOther}()},
\code{\link{plot,adv-method}},
\code{\link{read.adv.nortek}()},
\code{\link{read.adv.sontek.adr}()},
\code{\link{read.adv.sontek.serial}()},
\code{\link{read.adv.sontek.text}()},
\code{\link{read.adv}()},
\code{\link{rotateAboutZ}()},
\code{\link{subset,adv-method}},
\code{\link{summary,adv-method}},
\code{\link{toEnuAdv}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdv}()},
\code{\link{xyzToEnu}()}
}
\concept{classes provided by oce}
\concept{things related to adv data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/map.R
\name{mapArrows}
\alias{mapArrows}
\title{Add Arrows to a Map}
\usage{
mapArrows(
  longitude0,
  latitude0,
  longitude1 = longitude0,
  latitude1 = latitude0,
  length = 0.25,
  angle = 30,
  code = 2,
  col = par("fg"),
  lty = par("lty"),
  lwd = par("lwd"),
  ...
)
}
\arguments{
\item{longitude0, latitude0}{starting points for arrows.}

\item{longitude1, latitude1}{ending points for arrows.}

\item{length}{length of the arrow heads, passed to \code{\link[=arrows]{arrows()}}.}

\item{angle}{angle of the arrow heads, passed to \code{\link[=arrows]{arrows()}}.}

\item{code}{numerical code indicating the type of arrows, passed to \code{\link[=arrows]{arrows()}}.}

\item{col}{arrow color, passed to \code{\link[=arrows]{arrows()}}.}

\item{lty}{arrow line type, passed to \code{\link[=arrows]{arrows()}}.}

\item{lwd}{arrow line width, passed to \code{\link[=arrows]{arrows()}}.}

\item{...}{optional arguments passed to \code{\link[=arrows]{arrows()}}.}
}
\description{
Plot arrows on an existing map, e.g. to indicate a place location.
This is not well-suited for drawing direction fields, e.g. of
velocities; for that, see \code{\link[=mapDirectionField]{mapDirectionField()}}.
Adds arrows to an existing map, by analogy to \code{\link[=arrows]{arrows()}}.
}
\examples{
\donttest{
library(oce)
data(coastlineWorld)
mapPlot(coastlineWorld, longitudelim=c(-120, -60), latitudelim=c(30, 60),
        col="lightgray", projection="+proj=lcc +lat_1=45 +lon_0=-100")
lon <- seq(-120, -75, 15)
n <- length(lon)
lat <- 45 + rep(0, n)
# Draw meridional arrows in N America, from 45N to 60N.
mapArrows(lon, lat, lon, lat+15, length=0.05, col="blue")
}

}
\seealso{
A map must first have been created with \code{\link[=mapPlot]{mapPlot()}}.

Other functions related to maps: 
\code{\link{formatPosition}()},
\code{\link{lonlat2map}()},
\code{\link{lonlat2utm}()},
\code{\link{map2lonlat}()},
\code{\link{mapAxis}()},
\code{\link{mapContour}()},
\code{\link{mapCoordinateSystem}()},
\code{\link{mapDirectionField}()},
\code{\link{mapGrid}()},
\code{\link{mapImage}()},
\code{\link{mapLines}()},
\code{\link{mapLocator}()},
\code{\link{mapLongitudeLatitudeXY}()},
\code{\link{mapPlot}()},
\code{\link{mapPoints}()},
\code{\link{mapPolygon}()},
\code{\link{mapScalebar}()},
\code{\link{mapText}()},
\code{\link{mapTissot}()},
\code{\link{oceCRS}()},
\code{\link{shiftLongitude}()},
\code{\link{usrLonLat}()},
\code{\link{utm2lonlat}()}
}
\author{
Dan Kelley
}
\concept{functions related to maps}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{fullFilename}
\alias{fullFilename}
\title{Full name of file, including path}
\usage{
fullFilename(filename)
}
\arguments{
\item{filename}{name of file}
}
\value{
Full file name
}
\description{
Determines the full name of a file, including the path.  Used by many
\code{read.X} routines, where \code{X} is the name of a class of object.
This is a wrapper around \code{\link[=normalizePath]{normalizePath()}}, with warnings turned
off so that messages are not printed for unfound files (e.g. URLs).
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/map.R
\name{mapTissot}
\alias{mapTissot}
\title{Add Tissot Indicatrices to a Map}
\usage{
mapTissot(grid = rep(15, 2), scale = 0.2, crosshairs = FALSE, ...)
}
\arguments{
\item{grid}{numeric vector of length 2, specifying the increment in
longitude and latitude for the grid. Indicatrices are drawn at e.g.
longitudes \code{seq(-180, 180, grid[1])}.}

\item{scale}{numerical scale factor for ellipses. This is multiplied by
\code{min(grid)} and the result is the radius of the circle on the
earth, in latitude degrees.}

\item{crosshairs}{logical value indicating whether to draw constant-latitude
and constant-longitude crosshairs within the ellipses.  (These are drawn
with 10 line segments each.) This can be helpful in cases where it is
not desired to use \code{\link[=mapGrid]{mapGrid()}} to draw the longitude/latitude
grid.}

\item{\dots}{extra arguments passed to plotting functions, e.g.
\code{col="red"} yields red indicatrices.}
}
\description{
Plot ellipses at grid intersection points, as a method for
indicating the distortion inherent in the projection, somewhat
analogous to the scheme used in reference 1.
(Each ellipse is drawn with 64 segments.)
}
\examples{
\donttest{
library(oce)
data(coastlineWorld)
par(mfrow=c(1, 1), mar=c(2, 2, 1, 1))
p  <- "+proj=aea +lat_1=10 +lat_2=60 +lon_0=-45"
mapPlot(coastlineWorld, projection=p, col="gray",
longitudelim=c(-90,0), latitudelim=c(0, 50))
mapTissot(c(15, 15), col='red')
}

}
\references{
\enumerate{
\item Snyder, John P., 1987.  Map Projections: A Working Manual.  USGS
Professional Paper: 1395
\url{https://pubs.er.usgs.gov/publication/pp1395}
\doi{10.3133/pp1395}
}
}
\seealso{
A map must first have been created with \code{\link[=mapPlot]{mapPlot()}}.

Other functions related to maps: 
\code{\link{formatPosition}()},
\code{\link{lonlat2map}()},
\code{\link{lonlat2utm}()},
\code{\link{map2lonlat}()},
\code{\link{mapArrows}()},
\code{\link{mapAxis}()},
\code{\link{mapContour}()},
\code{\link{mapCoordinateSystem}()},
\code{\link{mapDirectionField}()},
\code{\link{mapGrid}()},
\code{\link{mapImage}()},
\code{\link{mapLines}()},
\code{\link{mapLocator}()},
\code{\link{mapLongitudeLatitudeXY}()},
\code{\link{mapPlot}()},
\code{\link{mapPoints}()},
\code{\link{mapPolygon}()},
\code{\link{mapScalebar}()},
\code{\link{mapText}()},
\code{\link{oceCRS}()},
\code{\link{shiftLongitude}()},
\code{\link{usrLonLat}()},
\code{\link{utm2lonlat}()}
}
\author{
Dan Kelley
}
\concept{functions related to maps}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/topo.R
\name{summary,topo-method}
\alias{summary,topo-method}
\title{Summarize A Topo Object}
\usage{
\S4method{summary}{topo}(object, ...)
}
\arguments{
\item{object}{A \linkS4class{topo} object.}

\item{\dots}{Further arguments passed to or from other methods.}
}
\value{
A matrix containing statistics of the elements of the \code{data} slot.
}
\description{
Pertinent summary information is presented, including the longitude and
latitude range, and the range of elevation.
}
\examples{
library(oce)
data(topoWorld)
summary(topoWorld)

}
\seealso{
Other things related to topo data: 
\code{\link{[[,topo-method}},
\code{\link{[[<-,topo-method}},
\code{\link{as.topo}()},
\code{\link{download.topo}()},
\code{\link{plot,topo-method}},
\code{\link{read.topo}()},
\code{\link{subset,topo-method}},
\code{\link{topo-class}},
\code{\link{topoInterpolate}()},
\code{\link{topoWorld}}
}
\author{
Dan Kelley
}
\concept{things related to topo data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sealevel.R
\name{subset,sealevel-method}
\alias{subset,sealevel-method}
\title{Subset a Sealevel Object}
\usage{
\S4method{subset}{sealevel}(x, subset, ...)
}
\arguments{
\item{x}{a \linkS4class{sealevel} object.}

\item{subset}{a condition to be applied to the \code{data} portion of
\code{x}.}

\item{\dots}{ignored.}
}
\value{
A new \code{sealevel} object.
}
\description{
This function is somewhat analogous to \code{\link[=subset.data.frame]{subset.data.frame()}}, but
subsetting is only permitted by time.
}
\examples{
library(oce)
data(sealevel)
plot(sealevel)
plot(subset(sealevel, time < mean(range(sealevel[['time']]))))

}
\seealso{
Other things related to sealevel data: 
\code{\link{[[,sealevel-method}},
\code{\link{[[<-,sealevel-method}},
\code{\link{as.sealevel}()},
\code{\link{plot,sealevel-method}},
\code{\link{read.sealevel}()},
\code{\link{sealevel-class}},
\code{\link{sealevelTuktoyaktuk}},
\code{\link{sealevel}},
\code{\link{summary,sealevel-method}}

Other functions that subset oce objects: 
\code{\link{subset,adp-method}},
\code{\link{subset,adv-method}},
\code{\link{subset,amsr-method}},
\code{\link{subset,argo-method}},
\code{\link{subset,cm-method}},
\code{\link{subset,coastline-method}},
\code{\link{subset,ctd-method}},
\code{\link{subset,echosounder-method}},
\code{\link{subset,lobo-method}},
\code{\link{subset,met-method}},
\code{\link{subset,oce-method}},
\code{\link{subset,odf-method}},
\code{\link{subset,rsk-method}},
\code{\link{subset,section-method}},
\code{\link{subset,topo-method}},
\code{\link{subset,xbt-method}}
}
\author{
Dan Kelley
}
\concept{functions that subset oce objects}
\concept{things related to sealevel data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{oceColorsPhase}
\alias{oceColorsPhase}
\alias{oce.colorsPhase}
\title{Create colors suitable for phase fields}
\usage{
oceColorsPhase(n)
}
\arguments{
\item{n}{number of colors to create.}
}
\value{
A vector of color specifications.
}
\description{
Create a set of colors for displaying phase values,
based on the scheme devised by Thyng et al. (2016) and
presented in a python package by Thyng (2019).
The color specifications were transliterated from
python to R on 2015-09-29, but have not been adjusted
since, even though the python source has changed.  This
is to prevent breaking old \code{oce} code.  To get the
latest versions of these colours or other colours,
use the \CRANpkg{cmocean} R package
(Thyng, Richards, and Krylov, 2019) directly, as is
illustrated (with the "matter" scheme) in Example 2.
Note that the \CRANpkg{cmocean} core functions provide a way
to select between various versions of the colour schemes.
It is also worth considering the palettes provided by the
\CRANpkg{viridis} package, as illustrated (with the
"inferno" scheme) in Example 3.
}
\examples{
library(oce)

# Example 1
imagep(volcano, col=oceColorsPhase(128),
       zlab="oceColorsPhase")
\dontrun{
# Example 2 (requires the cmocean package)
imagep(volcano, col=cmocean::cmocean("matter"),
       zlab='cmocean::cmocean("matter")')}

\dontrun{
# Example 3 (requires the viridis package)
imagep(volcano, col=viridis::inferno,
       zlab='viridis::inferno')}

}
\references{
\itemize{
\item Thyng, Kristen, Chad Greene, Robert Hetland, Heather Zimmerle, and Steven DiMarco.
\dQuote{True Colors of Oceanography: Guidelines for Effective and Accurate Colormap Selection.}
Oceanography 29, no. 3 (September 1, 2016): 9‚Äì13.
\doi{10.5670/oceanog.2016.66}
\item Thyng, Kristen. Kthyng/Cmocean. Python, 2019.
\url{https://github.com/kthyng/cmocean}.
\item Thyng, Kristen, Clark Richards, and Ivan Krylov.
Cmocean: Beautiful Colour Maps for Oceanography (version 0.2), 2019.
\url{https://CRAN.R-project.org/package=cmocean}.
}

The following references provide information on choosing
colour schemes, that are suitable for viewers who have
colour deficiencies.

Light, Adam, and Patrick J. Bartlein. "The End of the Rainbow? Color
Schemes for Improved Data Graphics." \emph{Eos, Transactions American Geophysical
Union} 85, no. 40 (2004): 385.
DOI: 10.1029/2004EO400002

Stephenson, David B. "Comment on 'Color Schemes for Improved Data
Graphics', by A Light and P.J. Bartlein." \emph{Eos, Transactions American
Geophysical Union} 86, no.  20 (2005): 196.
DOI: 10.1029/2005EO200005

Light, Adam, and Patrick J. Bartlein. "Reply to 'Comment on Color
Schemes for Improved Data Graphics,' by A. Light and P.J.  Bartlein'."
\emph{Eos, Transactions American Geophysical Union} 86, no. 20 (2005): 196‚Äì196.
DOI: 10.1029/2005EO200006
}
\seealso{
Other things related to colors: 
\code{\link{colormapGMT}()},
\code{\link{colormap}()},
\code{\link{oceColors9B}()},
\code{\link{oceColorsCDOM}()},
\code{\link{oceColorsChlorophyll}()},
\code{\link{oceColorsClosure}()},
\code{\link{oceColorsDensity}()},
\code{\link{oceColorsFreesurface}()},
\code{\link{oceColorsGebco}()},
\code{\link{oceColorsJet}()},
\code{\link{oceColorsOxygen}()},
\code{\link{oceColorsPAR}()},
\code{\link{oceColorsPalette}()},
\code{\link{oceColorsSalinity}()},
\code{\link{oceColorsTemperature}()},
\code{\link{oceColorsTurbidity}()},
\code{\link{oceColorsTurbo}()},
\code{\link{oceColorsTwo}()},
\code{\link{oceColorsVelocity}()},
\code{\link{oceColorsViridis}()},
\code{\link{oceColorsVorticity}()},
\code{\link{ocecolors}}
}
\author{
Krysten M. Thyng (Python version), Dan Kelley (R transliteration)
}
\concept{things related to colors}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/lobo.R
\name{read.lobo}
\alias{read.lobo}
\title{Read a LOBO File}
\usage{
read.lobo(file, cols = 7, processingLog)
}
\arguments{
\item{file}{a connection or a character string giving the name of the file
to load.}

\item{cols}{number of columns in dataset.}

\item{processingLog}{if provided, the action item to be stored in the log.
(Typically only provided for internal calls; the default that it provides is
better for normal calls by a user.)}
}
\value{
A \linkS4class{lobo} object.
}
\description{
Read a data file created by a LOBO instrument.
}
\details{
This version of \code{read.lobo} is really quite crude, having been
developed mainly for a ``predict the Spring bloom'' contest at Dalhousie
University.  In particular, the function assumes that the data columns are
exactly as specified in the Examples section; if you reorder the columns or
add new ones, this function is unlikely to work correctly. Furthermore, it
should be noted that the file format was inferred simply by downloading
files; the supplier makes no claims that the format will be fixed in time.
It is also worth noting that there is no \code{\link[=read.oce]{read.oce()}} equivalent
to \code{read.lobo}, because the file format has no recognizable header.
}
\examples{
\dontrun{
library(oce)
uri <- paste("http://lobo.satlantic.com/cgi-bin/nph-data.cgi?",
  "min_date=20070220&max_date=20070305",
  "&x=date&",
  "y=current_across1,current_along1,nitrate,fluorescence,salinity,temperature&",
  "data_format=text",sep="")
lobo <- read.lobo(uri)
}
}
\seealso{
Other things related to lobo data: 
\code{\link{[[,lobo-method}},
\code{\link{[[<-,lobo-method}},
\code{\link{as.lobo}()},
\code{\link{lobo-class}},
\code{\link{lobo}},
\code{\link{plot,lobo-method}},
\code{\link{subset,lobo-method}},
\code{\link{summary,lobo-method}}
}
\author{
Dan Kelley
}
\concept{things related to lobo data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{vectorShow}
\alias{vectorShow}
\title{Show some values from a list, vector or matrix}
\usage{
vectorShow(
  v,
  msg = "",
  postscript = "",
  digits = 5,
  n = 2L,
  showNA = FALSE,
  showNewline = TRUE
)
}
\arguments{
\item{v}{the item to be summarized.  If this is a list of a vector of named
items, then information is provided for each element. Otherwise, information
is provided for the first and last \code{n} values.}

\item{msg}{optional character value indicating a message to show,
introducing the vector.  If not provided, then
a message is created from \code{v}. If \code{msg} is a non-empty string,
then that string is pasted together with a colon (unless \code{msg} already
contains a colon), before pasting a summary of data values.}

\item{postscript}{optional character value indicating an optional message
to append at the end of the return value.}

\item{digits}{for numerical values of \code{v}, this is the number of digits
to use, in formatting the numbers with \code{\link[=format]{format()}}; otherwise,
\code{digits} is ignored.}

\item{n}{number of elements to show at start and end. If \code{n}
is negative, then all the elements are shown.}

\item{showNA}{logical value indicating whether to show the number
of \code{NA} values. This is done only if the output contains ellipses,
meaning that some values are skipped, because if all values are shown,
it will be perfectly obvious whether there are any \code{NA} values.}

\item{showNewline}{logical value indicating whether to put a newline
character at the end of the output string.  The default, TRUE, is
convenient for printing, but using FALSE makes more sense if
the result is to be used with, e.g. \code{\link[=mtext]{mtext()}}.}
}
\value{
A string ending in a newline character, suitable for
display with \code{\link[=cat]{cat()}} or \code{\link[=oceDebug]{oceDebug()}}.
}
\description{
This is similar to \code{\link[=str]{str()}}, but it shows data at the first and last of the
vector, which can be quite helpful in debugging.
}
\examples{
# List
limits <- list(low=0, high=1)
vectorShow(limits)

# Vector of named items
planktonCount <- c(phytoplankton=100, zooplankton=20)
vectorShow(planktonCount)

# Vector
vectorShow(pi)

# Matrix
vectorShow(volcano)

# Other arguments
knot2mps <- 0.5144444
vectorShow(knot2mps, postscript="knots per m/s")
vectorShow("January", msg="The first month is")

}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swSigma0}
\alias{swSigma0}
\title{Seawater potential density anomaly referenced to surface pressure}
\usage{
swSigma0(
  salinity,
  temperature = NULL,
  pressure = NULL,
  longitude = NULL,
  latitude = NULL,
  eos = getOption("oceEOS", default = "gsw")
)
}
\arguments{
\item{salinity}{either practical salinity (in which case \code{temperature}
and \code{pressure} must be provided) \emph{or} an \code{oce} object, in
which case \code{salinity}, \code{temperature} (in the ITS-90 scale; see
next item), etc. are inferred from the object, ignoring the
other parameters, if they are supplied.}

\item{temperature}{\emph{in-situ} temperature (\eqn{^\circ}{deg}C), defined
on the ITS-90 scale.  This scale is used by GSW-style calculation (as
requested by setting \code{eos="gsw"}), and is the value contained within
\code{ctd} objects (and probably most other objects created with data
acquired in the past decade or two). Since the UNESCO-style calculation is
based on IPTS-68, the temperature is converted within the present function,
using \code{\link[=T68fromT90]{T68fromT90()}}.}

\item{pressure}{pressure (dbar)}

\item{longitude}{longitude of observation (only used if \code{eos="gsw"};
see \sQuote{Details}).}

\item{latitude}{latitude of observation (only used if \code{eos="gsw"}; see
\sQuote{Details}).}

\item{eos}{equation of state, either \code{"unesco"} (references 1 and 2)
or \code{"gsw"} (references 3 and 4).}
}
\value{
Potential density anomaly (kg/m\eqn{^3}{^3}).
}
\description{
Compute \eqn{\sigma_\theta}{sigma}, the potential density of seawater (minus
1000 kg/m\eqn{^3}{^3}), referenced to surface pressure.
}
\details{
Definition:
\eqn{\sigma_0=\sigma_\theta=\rho(S,\theta(S,t,p),0}{sigma_0=sigma_theta=rho(S,theta(S,t,p),0)}
\itemize{
\item 1000 kg/m\eqn{^3}{^3}.
}
}
\references{
See citations provided in the \code{\link[=swRho]{swRho()}} documentation.
}
\seealso{
Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{T68fromT90}
\alias{T68fromT90}
\title{Convert from ITS-90 to IPTS-68 temperature}
\usage{
T68fromT90(temperature)
}
\arguments{
\item{temperature}{numeric vector of temperatures]in \eqn{^\circ}{deg}C on the ITS-90 scale.}
}
\value{
Corresponding temperatures in \eqn{^\circ}{deg}C on the IPTS-68 scale.
}
\description{
Today's instruments typically record in the ITS-90 scale, but some old
datasets will be in the IPTS-68 scale. \code{T90fromT68()} converts from the
IPTS-68 to the ITS-90 scale, using Saunders' (1990) formula, while
\code{T68fromT90()} does the reverse. The difference between IPTS-68 and
ITS-90 values is typically a few millidegrees (see \sQuote{Examples}), which
is seldom visible on a typical temperature profile, but may be of interest
in some precise work.  Mostly for historical interest, \code{T90fromT48()}
is provided to convert from the ITS-48 system to ITS-90.
}
\examples{
library(oce)
T68 <- seq(3, 20, 1)
T90 <- T90fromT68(T68)
sqrt(mean((T68-T90)^2))

}
\references{
P. M. Saunders, 1990. The international temperature scale of
1990, ITS-90.  WOCE Newsletter, volume 10, September 1990, page 10.
http://www.nodc.noaa.gov/woce/wdiu/wocedocs/newsltr/news10/contents.htm
}
\seealso{
Other functions that calculate seawater properties: 
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/odf.R
\name{[[,odf-method}
\alias{[[,odf-method}
\title{Extract Something From an ODF Object}
\usage{
\S4method{[[}{odf}(x, i, j, ...)
}
\arguments{
\item{x}{an \linkS4class{odf} object.}

\item{i}{character value indicating the name of an item to extract.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{ignored.}
}
\description{
Generally, the \code{[[} method lets users extract information from \code{oce}
objects, without having to know the details of the internal storage.  For
many \code{oce} sub-classes, \code{[[} can also return quantities that are computed
from the object's contents.
}
\details{
A two-step process is used to try to find the requested information. First, a
class-specific function is used (see \dQuote{Details of the Specialized
Method}).  If this yields nothing, then a general method is used (see
\dQuote{Details of the General Method}).  If both methods fail, then \code{[[}
returns NULL.

Some understanding of the subclass is required to know what can be retrieved
with \code{[[}. When dealing with an unfamiliar subclass, it can be useful to
first use \code{x[["?"]]} to get a listing of the retrievable items. See
\dQuote{Details of the Specialized Method} for more information.
}
\section{Details of the Specialized Method}{


If \code{i} is \code{"?"}, then the return value is a list
containing four items, each of which is a character vector
holding the names of things that can be accessed with \code{[[}.
The \code{data} and \code{metadata} items hold the names of
entries in the object's data and metadata
slots, respectively. The \code{dataDerived}
and \code{metadataDerived} items are each NULL, because
no derived values are defined by \linkS4class{odf} objects.
}

\section{Details of the General Method}{


Note: the text of this section is identical for all \code{oce} subclasses, and so
some of what you read here may not be relevant to the class being described
in this help page.

If the specialized method produces no matches, the following generalized
method is applied. As with the specialized method, the procedure hinges first
on the values of \code{i} and, optionally, \code{j}. The work proceeds in steps, by
testing a sequence of possible conditions in sequence.
\enumerate{
\item A check is made as to whether \code{i} names one of the standard \code{oce} slots.
If so, \code{[[} returns the slot contents of that slot.  Thus, \code{x[["metadata"]]}
will retrieve the \code{metadata} slot, while \code{x[["data"]]} and
\code{x[["processingLog"]]} return those slots.
\item If \code{i} is a string ending in the \code{"Unit"}, then the characters preceding
that string are taken to be the name of an item in the data object, and a
list containing the unit is returned (or \code{NULL} if there is no such unit).
This list consists of an item named \code{unit}, which is an \code{\link[=expression]{expression()}}, and
an item named \code{scale}, which is a string describing the measurement scale.
If the string ends in \code{" unit"}, e.g.  \code{x[["temperature unit"]]} (note the
space), then just the expression is returned, and if it ends in \code{" scale"},
then just the scale is returned.
\item If \code{i} is a string ending in \code{"Flag"}, then the corresponding data-quality
flag is returned (or \code{NULL} if there is no such flag).
\item If the object holds hydrographic information (pressure, salinity,
temperature, longitude and latitude) then another set of possibilities
arises.  If \code{i} is \code{"sigmaTheta"}, then the value of \code{\link[=swSigmaTheta]{swSigmaTheta()}} is
called with \code{x} as the sole argument, and the results are returned.
Similarly, \code{\link[=swSigma0]{swSigma0()}} is used if \code{i="sigma0"}, and \code{\link[=swSpice]{swSpice()}} is used if
\code{i="spice"}. Of course, these actions only make sense for objects that
contain the relevant items within their \code{data} slot.
\item After these possibilities are eliminated, the action depends on whether
\code{j} has been provided.  If \code{j} is not provided, or is the string \code{""}, then
\code{i} is sought in the \code{metadata} slot, and then in the \code{data} slot, returning
whichever is found first.  In other words, if \code{j} is not provided, the
\code{metadata} slot takes preference over the \code{data} slot. However, if \code{j} is
provided, then it must be either the string \code{"metadata"} or \code{"data"}, and it
directs where to look.
\item If none of the above-listed conditions holds, then \code{NULL} is returned.
}
}

\seealso{
Other functions that extract parts of oce objects: 
\code{\link{[[,adp-method}},
\code{\link{[[,adv-method}},
\code{\link{[[,amsr-method}},
\code{\link{[[,argo-method}},
\code{\link{[[,bremen-method}},
\code{\link{[[,cm-method}},
\code{\link{[[,coastline-method}},
\code{\link{[[,ctd-method}},
\code{\link{[[,echosounder-method}},
\code{\link{[[,g1sst-method}},
\code{\link{[[,gps-method}},
\code{\link{[[,ladp-method}},
\code{\link{[[,landsat-method}},
\code{\link{[[,lisst-method}},
\code{\link{[[,lobo-method}},
\code{\link{[[,met-method}},
\code{\link{[[,oce-method}},
\code{\link{[[,rsk-method}},
\code{\link{[[,sealevel-method}},
\code{\link{[[,section-method}},
\code{\link{[[,tidem-method}},
\code{\link{[[,topo-method}},
\code{\link{[[,windrose-method}},
\code{\link{[[,xbt-method}},
\code{\link{[[<-,adv-method}}

Other things related to odf data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{ODF2oce}()},
\code{\link{ODFListFromHeader}()},
\code{\link{ODFNames2oceNames}()},
\code{\link{[[<-,odf-method}},
\code{\link{odf-class}},
\code{\link{plot,odf-method}},
\code{\link{read.ctd.odf}()},
\code{\link{read.odf}()},
\code{\link{subset,odf-method}},
\code{\link{summary,odf-method}}
}
\concept{functions that extract parts of oce objects}
\concept{things related to odf data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.R
\name{write.ctd}
\alias{write.ctd}
\title{Write a CTD Data Object as a CSV File}
\usage{
write.ctd(object, file, metadata = TRUE, flags = TRUE, format = "csv")
}
\arguments{
\item{object}{a \linkS4class{ctd} object.}

\item{file}{Either a character string (the file name) or a connection. If not
provided, \code{file} defaults to \code{\link[=stdout]{stdout()}}.}

\item{metadata}{a logical value indicating whether to put some selected
metadata elements at the start of the output file.}

\item{flags}{a logical value indicating whether to show data-quality flags
as well as data.}

\item{format}{string indicating the format to use. This may be \code{"csv"}
for a simple CSV format, or \code{"whp"} for the World Hydrographic
Program format, described in reference 1 and exemplified in reference 2.}
}
\description{
Writes a comma-separated file containing the data frame stored in
the \code{data} slot of the first argument.  The file is suitable
for reading with a spreadsheet, or
with \code{\link[=read.csv]{read.csv()}}.  This output file will contain
some of the metadata in \code{x}, if \code{metadata} is \code{TRUE}.
}
\examples{
\dontrun{
library(oce)
data(ctd)
write.ctd(ctd, "ctd.csv")
d <- read.csv("ctd.csv")
plot(as.ctd(d$salinity, d$temperature, d$pressure))
}

}
\references{
The following links used to work, but failed as of December 2020.
\enumerate{
\item \verb{https://www.nodc.noaa.gov/woce/woce_v3/wocedata_1/whp/exchange/exchange_format_desc.htm}
\item \verb{https://www.nodc.noaa.gov/woce/woce_v3/wocedata_1/whp/exchange/example_ct1.csv}
}
}
\seealso{
The documentation for \linkS4class{ctd} explains the structure
of CTD objects.

Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()}
}
\author{
Dan Kelley
}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{binApply1D}
\alias{binApply1D}
\title{Apply a function to vector data}
\usage{
binApply1D(x, f, xbreaks, FUN, ...)
}
\arguments{
\item{x}{a vector of numerical values.}

\item{f}{a vector of data to which the elements of \code{FUN} may be
supplied}

\item{xbreaks}{values of x at the boundaries between bins; calculated using
\code{\link[=pretty]{pretty()}} if not supplied.}

\item{FUN}{function to apply to the data}

\item{\dots}{arguments to pass to the function \code{FUN}}
}
\value{
A list with the following elements: the breaks in x and y
(\code{xbreaks} and \code{ybreaks}), the break mid-points (\code{xmids} and
\code{ymids}), and a matrix containing the result of applying function
\code{FUN} to \code{f} subsetted by these breaks.
}
\description{
The function \code{FUN} is applied to \code{f} in bins specified by
\code{xbreaks}.  (If \code{FUN} is \code{\link[=mean]{mean()}},
consider using \code{\link[=binMean2D]{binMean2D()}} instead, since it should be faster.)
}
\examples{
library(oce)
## salinity profile with median and quartile 1 and 3
data(ctd)
p <- ctd[["pressure"]]
S <- ctd[["salinity"]]
q1 <- binApply1D(p, S, pretty(p, 30), function(x) quantile(x, 1/4))
q3 <- binApply1D(p, S, pretty(p, 30), function(x) quantile(x, 3/4))
plotProfile(ctd, "salinity", col='gray', type='n')
polygon(c(q1$result, rev(q3$result)),
c(q1$xmids, rev(q1$xmids)), col='gray')
points(S, p, pch=20)

}
\seealso{
Other bin-related functions: 
\code{\link{binApply2D}()},
\code{\link{binAverage}()},
\code{\link{binCount1D}()},
\code{\link{binCount2D}()},
\code{\link{binMean1D}()},
\code{\link{binMean2D}()}
}
\author{
Dan Kelley
}
\concept{bin-related functions}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.woce.R
\name{woceNames2oceNames}
\alias{woceNames2oceNames}
\title{Translate WOCE Data Names to Oce Data Names}
\usage{
woceNames2oceNames(names)
}
\arguments{
\item{names}{vector of strings holding WOCE-style names.}
}
\value{
vector of strings holding \code{oce}-style names.
}
\description{
Translate WOCE-style names to \code{oce} names, using \code{\link[=gsub]{gsub()}}
to match patterns. For example, the pattern \code{"CTDOXY.*"} is taken
to mean \code{oxygen}.
}
\references{
Several online sources list WOCE names. An example is
\url{https://cchdo.github.io/hdo-assets/documentation/manuals/pdf/90_1/chap4.pdf}
}
\seealso{
Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}

Other functions that interpret variable names and units from headers: 
\code{\link{ODFNames2oceNames}()},
\code{\link{cnvName2oceName}()},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{unitFromStringRsk}()},
\code{\link{unitFromString}()},
\code{\link{woceUnit2oceUnit}()}
}
\author{
Dan Kelley
}
\concept{functions that interpret variable names and units from headers}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{oceColorsJet}
\alias{oceColorsJet}
\alias{oce.colorsJet}
\alias{oceColors9A}
\alias{oce.colors9A}
\title{Create colors similar to the Matlab Jet scheme}
\usage{
oceColorsJet(n)
}
\arguments{
\item{n}{number of colors}
}
\description{
Create colors similar to the Matlab Jet scheme
}
\examples{
library(oce)
imagep(volcano, col=oceColorsJet, zlab="oceColorsJet")

}
\references{
The following references provide information on choosing
colour schemes, that are suitable for viewers who have
colour deficiencies.

Light, Adam, and Patrick J. Bartlein. "The End of the Rainbow? Color
Schemes for Improved Data Graphics." \emph{Eos, Transactions American Geophysical
Union} 85, no. 40 (2004): 385.
DOI: 10.1029/2004EO400002

Stephenson, David B. "Comment on 'Color Schemes for Improved Data
Graphics', by A Light and P.J. Bartlein." \emph{Eos, Transactions American
Geophysical Union} 86, no.  20 (2005): 196.
DOI: 10.1029/2005EO200005

Light, Adam, and Patrick J. Bartlein. "Reply to 'Comment on Color
Schemes for Improved Data Graphics,' by A. Light and P.J.  Bartlein'."
\emph{Eos, Transactions American Geophysical Union} 86, no. 20 (2005): 196‚Äì196.
DOI: 10.1029/2005EO200006
}
\seealso{
Other things related to colors: 
\code{\link{colormapGMT}()},
\code{\link{colormap}()},
\code{\link{oceColors9B}()},
\code{\link{oceColorsCDOM}()},
\code{\link{oceColorsChlorophyll}()},
\code{\link{oceColorsClosure}()},
\code{\link{oceColorsDensity}()},
\code{\link{oceColorsFreesurface}()},
\code{\link{oceColorsGebco}()},
\code{\link{oceColorsOxygen}()},
\code{\link{oceColorsPAR}()},
\code{\link{oceColorsPalette}()},
\code{\link{oceColorsPhase}()},
\code{\link{oceColorsSalinity}()},
\code{\link{oceColorsTemperature}()},
\code{\link{oceColorsTurbidity}()},
\code{\link{oceColorsTurbo}()},
\code{\link{oceColorsTwo}()},
\code{\link{oceColorsVelocity}()},
\code{\link{oceColorsViridis}()},
\code{\link{oceColorsVorticity}()},
\code{\link{ocecolors}}
}
\concept{things related to colors}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/amsr.R
\docType{data}
\name{amsr}
\alias{amsr}
\title{An amsr dataset for waters near Nova Scotia}
\usage{
data(amsr)
}
\description{
This is a composite satellite image combining views for
2020 August 9, 10 and 11, trimmed from a world view to a view
spanning 30N to 60N and 80W to 40W; see \dQuote{Details}.
}
\details{
The following code was used to create this dataset.
\preformatted{
library(oce)
data(coastlineWorldFine, package="ocedata")
d1 <- read.amsr(download.amsr(2020, 8,  9, "~/data/amsr"))
d2 <- read.amsr(download.amsr(2020, 8, 10, "~/data/amsr"))
d3 <- read.amsr(download.amsr(2020, 8, 11, "~/data/amsr"))
d <- composite(d1, d2, d3)
amsr <- subset(d,    -80 < longitude & longitude < -40)
amsr <- subset(amsr,  30 < latitude  &  latitude <  60)
}
}
\examples{
library(oce)
data(coastlineWorld)
data(amsr)
plot(amsr, "SST")
lines(coastlineWorld[["longitude"]], coastlineWorld[["latitude"]])

}
\seealso{
Other satellite datasets provided with oce: 
\code{\link{landsat}}

Other datasets provided with oce: 
\code{\link{adp}},
\code{\link{adv}},
\code{\link{argo}},
\code{\link{cm}},
\code{\link{coastlineWorld}},
\code{\link{ctdRaw}},
\code{\link{ctd}},
\code{\link{echosounder}},
\code{\link{landsat}},
\code{\link{lisst}},
\code{\link{lobo}},
\code{\link{met}},
\code{\link{ocecolors}},
\code{\link{rsk}},
\code{\link{sealevelTuktoyaktuk}},
\code{\link{sealevel}},
\code{\link{section}},
\code{\link{topoWorld}},
\code{\link{wind}},
\code{\link{xbt}}

Other things related to amsr data: 
\code{\link{[[,amsr-method}},
\code{\link{[[<-,amsr-method}},
\code{\link{amsr-class}},
\code{\link{composite,amsr-method}},
\code{\link{download.amsr}()},
\code{\link{plot,amsr-method}},
\code{\link{read.amsr}()},
\code{\link{subset,amsr-method}},
\code{\link{summary,amsr-method}}
}
\concept{datasets provided with oce}
\concept{satellite datasets provided with oce}
\concept{things related to amsr data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{lookWithin}
\alias{lookWithin}
\title{Look Within the First Element of a List for Replacement Values}
\usage{
lookWithin(list)
}
\arguments{
\item{list}{A list of elements, typically arguments that will be used in sw functions.}
}
\value{
A list with elements of the same names but possibly filled in from the first element.
}
\description{
This is a helper function used by some seawater functions
(with names starting with \code{sw}) to
facilitate the specification of water properties either with
distinct arguments, or with data stored within an \code{oce}
object that is the first argument.
}
\details{
If \code{list[1]} is not an \code{oce} object, then the
return value of \code{lookWithin} is the same as the input
value, except that (a) \code{eos} is completed to either
\code{"gsw"} or \code{"unesco"} and (b) if \code{longitude}
and \code{latitude} are within \code{list[1]}, then they
are possibly lengthened, to have the same length as the first
item in the \code{data} slot of \code{list[1]}.

The examples may clarify this somewhat.
}
\examples{
## 1. If first item is not a CTD object, just return the input
lookWithin(list(a=1, b=2)) # returns a list
## 2. Extract salinity from a CTD object
data(ctd)
str(lookWithin(list(salinity=ctd)))
## 3. Extract salinity and temperature. Note that the
## value specified for temperature is ignored; all that matters
## is that temperature is named.
str(lookWithin(list(salinity=ctd, temperature=NULL)))
## 4. How it is used by swRho()
rho1 <- swRho(ctd, eos="unesco")
rho2 <- swRho(ctd[["salinity"]], ctd[["temperature"]], ctd[["pressure"]], eos="unesco")
stopifnot(all.equal(rho1, rho2))
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/section.R
\name{handleFlags,section-method}
\alias{handleFlags,section-method}
\alias{handleFlags.section}
\title{Handle flags in Section Objects}
\usage{
\S4method{handleFlags}{section}(
  object,
  flags = NULL,
  actions = NULL,
  where = where,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{object}{A \linkS4class{section} object.}

\item{flags}{A \link{list} specifying flag values upon which
actions will be taken. This can take two forms.
\itemize{
\item In the first form, the
list has named elements each containing a vector of integers. For example,
salinities flagged with values of 1 or 3:9 would be specified
by \code{flags=list(salinity=c(1,3:9))}. Several data items can be specified,
e.g. \code{flags=list(salinity=c(1,3:9), temperature=c(1,3:9))} indicates
that the actions are to take place for both salinity and temperature.
\item In the second form, \code{flags} is a list holding a single \emph{unnamed} vector, and
this means to apply the actions to \emph{all} the data entries.  For example,
\code{flags=list(c(1,3:9))} means to apply not just to salinity and temperature,
but to everything within the \code{data} slot.
}

If \code{flags} is not provided, then \code{\link[=defaultFlags]{defaultFlags()}} is called, to try to
determine a reasonable default.}

\item{actions}{an optional \link{list} that contains items with
names that match those in the \code{flags} argument.  If \code{actions}
is not supplied, the default will be to set all values identified by
\code{flags} to \code{NA}; this can also be specified by
specifying \code{actions=list("NA")}. It is also possible to specify
functions that calculate replacement values. These are provided
with \code{object} as the single argument, and must return a
replacement for the data item in question.
See \dQuote{Details} for the default that is used if \code{actions} is not supplied.}

\item{where}{an optional character value that permits the function to work with
objects that store flags in e.g. \code{object@metadata$flags$where}
instead of in \code{object@metadata$flags}, and data within
\code{object@data$where} instead of within \code{object@data}. The
default value of \code{NULL} means to look withing \code{object@metadata}
itself, and this is the default within \code{oce}.  (The purpose of \code{where}
is to make \code{oce} extensible by other packages, which may choose to store
data two levels deep in the \code{data} slot.)}

\item{debug}{An optional integer specifying the degree of debugging, with
value 0 meaning to skip debugging and 1 or higher meaning to print some
information about the arguments and the data. It is usually a good idea to set
this to 1 for initial work with a dataset, to see which flags are being
handled for each data item. If not supplied, this defaults to the value of
\code{\link{getOption}}\code{("oceDebug")}.}
}
\description{
Data-quality flags are stored in the \code{metadata}
slot of \linkS4class{oce} objects in a
\link{list} named \code{flags}.
The present function (a generic that has specialized versions
for various data classes) provides a way to
manipulate the contents of the \code{data} slot, based on
such data-quality flags. For example, a common operation is to replace
erroneous data with \code{NA}.

If the \code{flags} within \code{object}'s \code{metadata} slot is empty,
then \code{object} is returned, unaltered.
Otherwise, \code{handleFlags} examines \code{object@metadata$flags}
in the context of the \code{flags} argument, and then
carries out actions that are specified by the \code{actions} argument.
By default, this sets the returned \code{data} entries to \code{NA},
wherever the corresponding \code{metadata$flag} values
signal unreliable data. To maintain a hint as to why
\code{data} were changed, \code{metadata$flags} in the
returned value is a direct copy of the corresponding
entry in \code{object}.
}
\details{
The default for \code{flags} is based on
calling \code{\link[=defaultFlags]{defaultFlags()}} based on the
\code{metadata} in the first station in the section. If the
other stations have different flag schemes (which seems highly
unlikely for archived data), this will not work well, and indeed
the only way to proceed would be to use \code{\link[=handleFlags,ctd-method]{handleFlags,ctd-method()}}
on the stations, individually.
}
\examples{
library(oce)
data(section)
section2 <- handleFlags(section, flags=c(1,3:9))
par(mfrow=c(2, 1))
plotTS(section)
plotTS(section2)

}
\references{
The following link used to work, but was found to fail in December 2020.
\enumerate{
\item \verb{https://www.nodc.noaa.gov/woce/woce_v3/wocedata_1/whp/exchange/exchange_format_desc.htm}
}
}
\seealso{
Other functions relating to data-quality flags: 
\code{\link{defaultFlags}()},
\code{\link{handleFlags,adp-method}},
\code{\link{handleFlags,argo-method}},
\code{\link{handleFlags,ctd-method}},
\code{\link{handleFlags,oce-method}},
\code{\link{handleFlags}()},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{initializeFlagScheme,oce-method}},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{initializeFlagSchemeInternal}()},
\code{\link{initializeFlagScheme}()},
\code{\link{initializeFlags,adp-method}},
\code{\link{initializeFlags,oce-method}},
\code{\link{initializeFlagsInternal}()},
\code{\link{initializeFlags}()},
\code{\link{setFlags,adp-method}},
\code{\link{setFlags,ctd-method}},
\code{\link{setFlags,oce-method}},
\code{\link{setFlags}()}

Other things related to section data: 
\code{\link{[[,section-method}},
\code{\link{[[<-,section-method}},
\code{\link{as.section}()},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{plot,section-method}},
\code{\link{read.section}()},
\code{\link{section-class}},
\code{\link{sectionAddStation}()},
\code{\link{sectionGrid}()},
\code{\link{sectionSmooth}()},
\code{\link{sectionSort}()},
\code{\link{section}},
\code{\link{subset,section-method}},
\code{\link{summary,section-method}}
}
\concept{functions relating to data-quality flags}
\concept{things related to section data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/g1sst.R
\name{[[,g1sst-method}
\alias{[[,g1sst-method}
\title{Extract Something From a G1SST Object}
\usage{
\S4method{[[}{g1sst}(x, i, j, ...)
}
\arguments{
\item{x}{a \linkS4class{g1sst} object.}

\item{i}{character value indicating the name of an item to extract.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{ignored.}
}
\description{
Generally, the \code{[[} method lets users extract information from \code{oce}
objects, without having to know the details of the internal storage.  For
many \code{oce} sub-classes, \code{[[} can also return quantities that are computed
from the object's contents.
}
\details{
A two-step process is used to try to find the requested information. First, a
class-specific function is used (see \dQuote{Details of the Specialized
Method}).  If this yields nothing, then a general method is used (see
\dQuote{Details of the General Method}).  If both methods fail, then \code{[[}
returns NULL.

Some understanding of the subclass is required to know what can be retrieved
with \code{[[}. When dealing with an unfamiliar subclass, it can be useful to
first use \code{x[["?"]]} to get a listing of the retrievable items. See
\dQuote{Details of the Specialized Method} for more information.
}
\section{Details of the Specialized Method}{

\itemize{
\item If \code{i} is \code{"?"}, then the return value is a list
containing four items, each of which is a character vector
holding the names of things that can be accessed with \code{[[}.
The \code{data} and \code{metadata} items hold the names of
entries in the object's data and metadata
slots, respectively. The \code{dataDerived}
and \code{metadataDerived} items are each NULL, because
no derived values are defined by \linkS4class{g1sst} objects.
}
}

\section{Details of the General Method}{


Note: the text of this section is identical for all \code{oce} subclasses, and so
some of what you read here may not be relevant to the class being described
in this help page.

If the specialized method produces no matches, the following generalized
method is applied. As with the specialized method, the procedure hinges first
on the values of \code{i} and, optionally, \code{j}. The work proceeds in steps, by
testing a sequence of possible conditions in sequence.
\enumerate{
\item A check is made as to whether \code{i} names one of the standard \code{oce} slots.
If so, \code{[[} returns the slot contents of that slot.  Thus, \code{x[["metadata"]]}
will retrieve the \code{metadata} slot, while \code{x[["data"]]} and
\code{x[["processingLog"]]} return those slots.
\item If \code{i} is a string ending in the \code{"Unit"}, then the characters preceding
that string are taken to be the name of an item in the data object, and a
list containing the unit is returned (or \code{NULL} if there is no such unit).
This list consists of an item named \code{unit}, which is an \code{\link[=expression]{expression()}}, and
an item named \code{scale}, which is a string describing the measurement scale.
If the string ends in \code{" unit"}, e.g.  \code{x[["temperature unit"]]} (note the
space), then just the expression is returned, and if it ends in \code{" scale"},
then just the scale is returned.
\item If \code{i} is a string ending in \code{"Flag"}, then the corresponding data-quality
flag is returned (or \code{NULL} if there is no such flag).
\item If the object holds hydrographic information (pressure, salinity,
temperature, longitude and latitude) then another set of possibilities
arises.  If \code{i} is \code{"sigmaTheta"}, then the value of \code{\link[=swSigmaTheta]{swSigmaTheta()}} is
called with \code{x} as the sole argument, and the results are returned.
Similarly, \code{\link[=swSigma0]{swSigma0()}} is used if \code{i="sigma0"}, and \code{\link[=swSpice]{swSpice()}} is used if
\code{i="spice"}. Of course, these actions only make sense for objects that
contain the relevant items within their \code{data} slot.
\item After these possibilities are eliminated, the action depends on whether
\code{j} has been provided.  If \code{j} is not provided, or is the string \code{""}, then
\code{i} is sought in the \code{metadata} slot, and then in the \code{data} slot, returning
whichever is found first.  In other words, if \code{j} is not provided, the
\code{metadata} slot takes preference over the \code{data} slot. However, if \code{j} is
provided, then it must be either the string \code{"metadata"} or \code{"data"}, and it
directs where to look.
\item If none of the above-listed conditions holds, then \code{NULL} is returned.
}
}

\seealso{
Other functions that extract parts of oce objects: 
\code{\link{[[,adp-method}},
\code{\link{[[,adv-method}},
\code{\link{[[,amsr-method}},
\code{\link{[[,argo-method}},
\code{\link{[[,bremen-method}},
\code{\link{[[,cm-method}},
\code{\link{[[,coastline-method}},
\code{\link{[[,ctd-method}},
\code{\link{[[,echosounder-method}},
\code{\link{[[,gps-method}},
\code{\link{[[,ladp-method}},
\code{\link{[[,landsat-method}},
\code{\link{[[,lisst-method}},
\code{\link{[[,lobo-method}},
\code{\link{[[,met-method}},
\code{\link{[[,oce-method}},
\code{\link{[[,odf-method}},
\code{\link{[[,rsk-method}},
\code{\link{[[,sealevel-method}},
\code{\link{[[,section-method}},
\code{\link{[[,tidem-method}},
\code{\link{[[,topo-method}},
\code{\link{[[,windrose-method}},
\code{\link{[[,xbt-method}},
\code{\link{[[<-,adv-method}}

Other things related to g1sst data: 
\code{\link{[[<-,g1sst-method}},
\code{\link{g1sst-class}},
\code{\link{read.g1sst}()}
}
\author{
Dan Kelley
}
\concept{functions that extract parts of oce objects}
\concept{things related to g1sst data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/coastline.R
\name{read.coastline.openstreetmap}
\alias{read.coastline.openstreetmap}
\title{Read a Coastline File in Openstreetmap Format}
\usage{
read.coastline.openstreetmap(
  file,
  lonlim = c(-180, 180),
  latlim = c(-90, 90),
  debug = getOption("oceDebug"),
  monitor = FALSE,
  processingLog
)
}
\arguments{
\item{file}{name of file containing coastline data (a file ending in \code{.shp})
or a zipfile that contains such a file, with a corresponding name.
The second scheme is useful for files downloaded from the NaturalEarth
website (see reference 2).}

\item{lonlim}{numerical vectors specifying the
west and east edges (and south and north edges) of a focus window.
Coastline polygons that do not intersect the defined box are
skipped, which can be useful in narrowing high-resolution world-scale
data to a local application.}

\item{latlim}{numerical vectors specifying the
west and east edges (and south and north edges) of a focus window.
Coastline polygons that do not intersect the defined box are
skipped, which can be useful in narrowing high-resolution world-scale
data to a local application.}

\item{debug}{set to TRUE to print information about the header, etc.}

\item{monitor}{Logical indicating whether to print an indication of progress through
the file.}

\item{processingLog}{if provided, the action item to be stored in the log.
(Typically only provided for internal calls; the default that it provides is
better for normal calls by a user.)}
}
\value{
a \linkS4class{coastline} object.
}
\description{
Read coastline data stored in the openstreetmap format.
}
\seealso{
Other things related to coastline data: 
\code{\link{[[,coastline-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{as.coastline}()},
\code{\link{coastline-class}},
\code{\link{coastlineBest}()},
\code{\link{coastlineCut}()},
\code{\link{coastlineWorld}},
\code{\link{download.coastline}()},
\code{\link{plot,coastline-method}},
\code{\link{read.coastline.shapefile}()},
\code{\link{subset,coastline-method}},
\code{\link{summary,coastline-method}}
}
\author{
Dan Kelley
}
\concept{things related to coastline data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/echosounder.R
\name{findBottom}
\alias{findBottom}
\title{Find the Ocean Bottom in an Echosounder Object}
\usage{
findBottom(x, ignore = 5, clean = despike)
}
\arguments{
\item{x}{an \linkS4class{echosounder} object.}

\item{ignore}{number of metres of data to ignore, near the surface.}

\item{clean}{a function to clean the inferred depth of spikes.}
}
\value{
A list with elements: the \code{time} of a ping, the \code{depth} of
the inferred depth in metres, and the \code{index} of the inferred bottom
location, referenced to the object's \code{depth} vector.
}
\description{
Finds the depth in a Biosonics echosounder file, by finding the strongest
reflector and smoothing its trace.
}
\seealso{
The documentation for \linkS4class{echosounder} explains the
structure of \code{echosounder} objects, and also outlines the other
functions dealing with them.

Other things related to echosounder data: 
\code{\link{[[,echosounder-method}},
\code{\link{[[<-,echosounder-method}},
\code{\link{as.echosounder}()},
\code{\link{echosounder-class}},
\code{\link{echosounder}},
\code{\link{plot,echosounder-method}},
\code{\link{read.echosounder}()},
\code{\link{subset,echosounder-method}},
\code{\link{summary,echosounder-method}}
}
\author{
Dan Kelley
}
\concept{things related to echosounder data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/cm.R
\name{as.cm}
\alias{as.cm}
\title{Coerce data into a CM object}
\usage{
as.cm(
  time,
  u = NULL,
  v = NULL,
  pressure = NULL,
  conductivity = NULL,
  temperature = NULL,
  salinity = NULL,
  longitude = NA,
  latitude = NA,
  filename = "",
  debug = getOption("oceDebug")
)
}
\arguments{
\item{time}{A vector of times of observation, or an \code{oce} object from which time
and two velocity components can be inferred, e.g. an \linkS4class{adv} object, or
an \linkS4class{adp} object that has only one distance bin.  If \code{time} is an \code{oce} object,
then all of the following arguments are ignored.}

\item{u, v}{optional numerical vectors containing the x and y components of velocity (m/s).}

\item{pressure, conductivity, salinity, temperature}{optional numerical vectors
containing pressure (dbar), electrical conductivity, practical salinity,
and in-situ temperature (degree C).}

\item{longitude, latitude}{optional position specified in degrees East and North.}

\item{filename}{optional source file name.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\description{
Coerce data into a CM object
}
\examples{
library(oce)
# Example 1: creation from scratch
t <- Sys.time() + 0:50
u <- sin(2*pi*0:50/5) + rnorm(51)
v <- cos(2*pi*0:50/5) + rnorm(51)
p <- 100 + rnorm(51)
summary(as.cm(t, u, v, p))

# Example 2: creation from an adv object
data(adv)
summary(as.cm(adv))

}
\seealso{
Other things related to cm data: 
\code{\link{[[,cm-method}},
\code{\link{[[<-,cm-method}},
\code{\link{cm-class}},
\code{\link{cm}},
\code{\link{plot,cm-method}},
\code{\link{read.cm}()},
\code{\link{rotateAboutZ}()},
\code{\link{subset,cm-method}},
\code{\link{summary,cm-method}}
}
\concept{things related to cm data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/coastline.R
\name{plot,coastline-method}
\alias{plot,coastline-method}
\alias{plot.coastline}
\title{Plot a coastline Object}
\usage{
\S4method{plot}{coastline}(
  x,
  xlab = "",
  ylab = "",
  showHemi = TRUE,
  asp,
  clongitude,
  clatitude,
  span,
  lonlabels = TRUE,
  latlabels = TRUE,
  projection = NULL,
  expand = 1,
  mgp = getOption("oceMgp"),
  mar = c(mgp[1] + 1, mgp[1] + 1, 1, 1),
  bg,
  fill,
  type = "polygon",
  border = NULL,
  col = NULL,
  axes = TRUE,
  cex.axis = par("cex.axis"),
  add = FALSE,
  inset = FALSE,
  geographical = 0,
  longitudelim,
  latitudelim,
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{x}{a \linkS4class{coastline} object.}

\item{xlab}{label for x axis}

\item{ylab}{label for y axis}

\item{showHemi}{logical indicating whether to show the hemisphere in axis
tick labels.}

\item{asp}{Aspect ratio for plot.  The default is for \verb{plot,coastline-method}
to set the aspect ratio to give natural latitude-longitude scaling somewhere
near the centre latitude on the plot. Often, it makes sense to set
\code{asp} yourself, e.g. to get correct shapes at 45N, use
\code{asp=1/cos(45*pi/180)}.  Note that the land mass is not symmetric about
the equator, so to get good world views you should set \code{asp=1} or set
\code{ylim} to be symmetric about zero. Any given value of \code{asp} is
ignored, if \code{clongitude} and \code{clatitude} are given (or
if the latter two are inferred from \code{projection}.}

\item{clongitude, clatitude}{optional center latitude of map, in decimal
degrees.  If both \code{clongitude} and \code{clatitude} are provided,
or alternatively if they can be inferred from substrings \code{+lon_0}
and \code{+lat_0} in \code{projection}, then
any provided value of \code{asp} is ignored, and instead the plot aspect
ratio is computed based on the center latitude.  If \code{clongitude} and
\code{clatitude} are known, then \code{span} must also be provided, and
in this case, it is not permitted to also specify \code{longitudelim} and
\code{latitudelim}.}

\item{span}{optional suggested diagonal span of the plot, in kilometers.
The plotted span is usually close to the suggestion, although the details
depend on the plot aspect ratio and other factors, so some adjustment may be
required to fine-tune a plot.  A value for \code{span} must be supplied, if
\code{clongitude} and \code{clatitude} are supplied
(or inferred from \code{projection}).}

\item{lonlabels, latlabels}{optional vectors of longitude and latitude to
label on the sides of plot, passed to \code{\link[=mapPlot]{mapPlot()}} to control
axis labelling, for plots done with map projections (i.e. for
cases in which \code{projection} is not \code{NULL}).}

\item{projection}{optional map projection to use (see
the \code{\link[=mapPlot]{mapPlot()}} argument of the same name).
If set to \code{FALSE} then no projection is used,
and the data are plotted in a cartesion frame, with aspect ratio set to
reduce distortion near the middle of the plot.  This option is useful if the
coastline produces spurious horizontal lines owing to islands crossing the
plot edges (a problem that plagues map projections).  If \code{projection}
is not set, a Mercator projection is used for latitudes below about 70
degrees, as if \code{projection="+proj=merc"} had been supplied, or a
Stereopolar one is used as if \code{projection="+proj=stere"}.  Otherwise,
\code{projection} must be a character string identifying a projection
accepted by \code{\link[=mapPlot]{mapPlot()}}.}

\item{expand}{numerical factor for the expansion of plot limits, showing
area outside the plot, e.g. if showing a ship track as a coastline, and then
an actual coastline to show the ocean boundary.  The value of \code{expand}
is ignored if either \code{xlim} or \code{ylim} is given.}

\item{mgp}{3-element numerical vector to use for \code{\link{par}}\code{("mgp")}, and also
for \code{par(mar)}, computed from this.  The default is tighter than the R
default, in order to use more space for the data and less for the axes.}

\item{mar}{value to be used with \code{\link{par}}\code{("mar")}.}

\item{bg}{optional color to be used for the background of the map.  This
comes in handy for drawing insets (see \dQuote{details}).}

\item{fill}{a legacy parameter that will be permitted only temporarily; see
\dQuote{History}.}

\item{type}{indication of type; may be \code{"polygon"}, for a filled polygon,
\code{"p"} for points, \code{"l"} for line segments, or \code{"o"} for points
overlain with line segments. See \code{color} for a note on how
the the value of \code{type} alters the meaning of the \code{color}
argument.}

\item{border}{color used to indicate land (if \code{type="polygon"}) or
the coastline and international borders (if \code{type="l"}).}

\item{col}{either the color for filling polygons (if \code{type="polygon"})
or the color of the points and line segments (if \code{type="p"},
\code{type="l"}, or \code{type="o"}).}

\item{axes}{boolean, set to \code{TRUE} to plot axes.}

\item{cex.axis}{value for axis font size factor.}

\item{add}{boolean, set to \code{TRUE} to draw the coastline on an existing
plot.  Note that this retains the aspect ratio of that existing plot, so it
is important to set that correctly, e.g. with \code{asp=1/cos(lat * pi /
180)}, where \code{clat} is the central latitude of the plot.}

\item{inset}{set to \code{TRUE} for use within \code{\link[=plotInset]{plotInset()}}.  The
effect is to prevent the present function from adjusting margins, which is
necessary because margin adjustment is the basis for the method used by
\code{\link[=plotInset]{plotInset()}}.}

\item{geographical}{flag indicating the style of axes.  With
\code{geographical=0}, the axes are conventional, with decimal degrees as
the unit, and negative signs indicating the southern and western
hemispheres.  With \code{geographical=1}, the signs are dropped, with axis
values being in decreasing order within the southern and western
hemispheres.  With \code{geographical=2}, the signs are dropped and the axes
are labelled with degrees, minutes and seconds, as appropriate, and
hemispheres are indicated with letters. With \code{geographical=3}, things
are the same as for \code{geographical=2}, but the hemisphere indication
is omitted. Finally, with \code{geographical=4}, unsigned numbers are used,
followed by letters \code{N} in the northern hemisphere, \code{S} in the southern,
\code{E} in the eastern, and \code{W} in the western.}

\item{longitudelim}{this and \code{latitudelim} provide a second way to
suggest plot ranges. Note that these may not be supplied if
\code{clongitude}, \code{clatitude} and \code{span} are given.}

\item{latitudelim}{see \code{longitudelim}.}

\item{debug}{set to \code{TRUE} to get debugging information during
processing.}

\item{\dots}{optional arguments passed to plotting functions.  For example,
set \code{yaxp=c(-90,90,4)} for a plot extending from pole to pole.}
}
\value{
None.
}
\description{
This function plots a coastline.  An attempt is made to fill the space of
the plot, and this is done by limiting either the longitude range or the
latitude range, as appropriate, by modifying the eastern or northern limit,
as appropriate.
}
\details{
If \code{longitudelim}, \code{latitudelim} and \code{projection} are all given,
then these arguments are passed to \code{\link[=mapPlot]{mapPlot()}} to produce the plot.
(The call uses \code{bg} for \code{col}, and uses \code{col}, \code{fill}
and \code{border} directly.) If the results need further customization,
users should use \code{\link[=mapPlot]{mapPlot()}} directly.

If \code{projection} is provided without \code{longitudelim} or \code{latitudelim},
then \code{\link[=mapPlot]{mapPlot()}} is still called, but \code{longitudelim} and
\code{latitudelim} are computed from \code{clongitude}, \code{clatitude} and \code{span}.

If \code{projection} is not provided, much simpler plots are produced. These are
Cartesian, with aspect ratio set to minimize shape distortion at the central latitude.
Although these are crude, they have the benefit of always working, which cannot
be said of true map projections, which can be problematic in various ways owing
to difficulties in inverting projection calculations.

To get an inset map inside another map, draw the first map, do
\code{\link{par}}\code{(new=TRUE)}, and then call \code{\link[=plot,coastline-method]{plot,coastline-method()}} with a value of
\code{mar} that moves the inset plot to a desired location on the existing
plot, and with \code{bg="white"}.
}
\section{History}{
 Until February, 2016, \verb{plot,coastline-method} relied on a
now-defunct argument \code{fill} to control colors; \code{col} is to be
used now, instead.
}

\examples{
\donttest{
library(oce)
par(mar=c(2, 2, 1, 1))
data(coastlineWorld)
plot(coastlineWorld)
plot(coastlineWorld, clongitude=-63.6, clatitude=44.6, span=1000)

## Canada in Lambert projection
plot(coastlineWorld, clongitude=-95, clatitude=65, span=5500,
     grid=10, projection='+proj=laea +lon_0=-100 +lat_0=55')
}

}
\seealso{
The documentation for the \linkS4class{coastline} class explains the
structure of coastline objects, and also outlines the other functions
dealing with them.

Other functions that plot oce data: 
\code{\link{download.amsr}()},
\code{\link{plot,adp-method}},
\code{\link{plot,adv-method}},
\code{\link{plot,amsr-method}},
\code{\link{plot,argo-method}},
\code{\link{plot,bremen-method}},
\code{\link{plot,cm-method}},
\code{\link{plot,ctd-method}},
\code{\link{plot,gps-method}},
\code{\link{plot,ladp-method}},
\code{\link{plot,landsat-method}},
\code{\link{plot,lisst-method}},
\code{\link{plot,lobo-method}},
\code{\link{plot,met-method}},
\code{\link{plot,odf-method}},
\code{\link{plot,rsk-method}},
\code{\link{plot,satellite-method}},
\code{\link{plot,sealevel-method}},
\code{\link{plot,section-method}},
\code{\link{plot,tidem-method}},
\code{\link{plot,topo-method}},
\code{\link{plot,windrose-method}},
\code{\link{plot,xbt-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{tidem-class}}

Other things related to coastline data: 
\code{\link{[[,coastline-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{as.coastline}()},
\code{\link{coastline-class}},
\code{\link{coastlineBest}()},
\code{\link{coastlineCut}()},
\code{\link{coastlineWorld}},
\code{\link{download.coastline}()},
\code{\link{read.coastline.openstreetmap}()},
\code{\link{read.coastline.shapefile}()},
\code{\link{subset,coastline-method}},
\code{\link{summary,coastline-method}}
}
\author{
Dan Kelley
}
\concept{functions that plot oce data}
\concept{things related to coastline data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/AllClass.R
\name{initializeFlagScheme,oce-method}
\alias{initializeFlagScheme,oce-method}
\title{Establish a data-quality scheme for a oce object}
\usage{
\S4method{initializeFlagScheme}{oce}(
  object,
  name = NULL,
  mapping = NULL,
  default = NULL,
  update = NULL,
  debug = 0
)
}
\arguments{
\item{object}{An oce object.}

\item{name}{a character value naming the scheme. If this refers
to a pre-defined scheme, then \code{mapping} must not be provided,
because doing so would contradict the pre-defined scheme, defeating
its purpose of providing concreteness and clarity.}

\item{mapping}{a list of named items describing the mapping from
flag meaning to flag numerical value, e.g \code{list(good=1, bad=2)}
might be used for a hypothetical class.}

\item{default}{an integer vector of flag values that are not considered
to be good. If this is not provided, but if \code{name} is \code{"argo"},
\code{"BODC"},
\code{"DFO"},
\code{"WHP bottle"}, or
\code{"WHP CTD"}, then a conservative value will be set automatically,
equal to the list of flag values that designate bad or questionable data.
For example, for \code{name="WHP CTD"}, the setting will be
\code{c(1,3,4,5,6,7,9)}, leaving only value \code{2}, which corresponds
with "acceptable" in the notation used for that flag scheme.}

\item{update}{a logical value indicating whether the scheme provided is
to update an existing scheme.  The default value, \code{FALSE}, prevents such
an attempt to alter an existing flag scheme, if one is already embedded
in \code{object}.}

\item{debug}{an integer set to 0 for quiet action or to 1 for some debugging.}
}
\value{
An object with the \code{metadata} slot containing \code{flagScheme}.
}
\description{
This function stores add an item named \code{flagScheme}
to the \code{metadata} slot of an object inheriting from
\linkS4class{oce}. This is a list containing two
items: \code{name} and \code{mapping}, as provided in the
function arguments.
The purpose is both to document a flag scheme
and to make it so that \code{\link[=initializeFlags]{initializeFlags()}},
\code{\link[=setFlags]{setFlags()}} and \code{\link[=handleFlags]{handleFlags()}}
can specify flags by
name, as opposed to number. This is a generic function,
that may be specialized to the class of \code{object}
(see \dQuote{Details}).
}
\details{
The following pre-defined schemes are available (note that the
names are simplified from the phrases used in defining
documentation):
\itemize{
\item \code{name="argo"} defaults \code{mapping} to
OLD (prior to June 10, 2020)
}\preformatted{list(not_assessed=0, passed_all_tests=1, probably_good=2,
     probably_bad=3, bad=4, averaged=7,
     interpolated=8, missing=9)
}

NEW (after June 10, 2020)\preformatted{list(not_assessed=0, passed_all_tests=1, probably_good=2,
     probably_bad=3, bad=4, changed=5, not_used_6=6, not_used_7=7,
     estimated=8, missing=9)
}

See reference 1 for a deeper explanation of the meanings of these codes.
\itemize{
\item \code{name="BODC"} defaults \code{mapping} to
}\preformatted{list(no_quality_control=0, good=1, probably_good=2,
     probably_bad=3, bad=4, changed=5,
     below_detection=6, in_excess=7, interpolated=8,
     missing=9)
}

See reference 2 for a deeper explanation of the meanings of these codes,
and note that codes \code{A} and \code{Q} are not provided in
oce.
\itemize{
\item \code{name="DFO"} defaults \code{mapping} to
}\preformatted{list(no_quality_control=0, appears_correct=1, appears_inconsistent=2,
     doubtful=3, erroneous=4, changed=5,
     qc_by_originator=8, missing=9)
}

See reference 3 for a deeper explanation of the meanings of these codes.
\itemize{
\item \code{name="WHP bottle"} defaults \code{mapping} to
}\preformatted{list(no_information=1, no_problems_noted=2, leaking=3,
     did_not_trip=4, not_reported=5, discrepency=6,
     unknown_problem=7, did_not_trip=8, no_sample=9)
}

See reference 4 for a deeper explanation of the meanings of these codes.
\itemize{
\item \code{name="WHP CTD"} defaults \code{mapping} to
}\preformatted{list(not_calibrated=1, acceptable=2, questionable=3,
    bad=4, not_reported=5, interpolated=6,
    despiked=7, missing=9)
}

See reference 4 for a deeper explanation of the meanings of these codes.
}
\references{
\enumerate{
\item The codes for \code{"argo"} are derived from information in Table 4.1
of Wong, Annie, Robert Keeley, Thierry Carval, and Argo Data Management Team
(8 January 2020), "Argo Quality Control Manual for CTD and Trajectory Data, Version 3.3,"
available at \url{https://archimer.ifremer.fr/doc/00228/33951/}
as of June 2020.
\item The codes for \code{"BODC"} are defined at
http://seadatanet.maris2.nl/v_bodc_vocab_v2/browse.asp?order=conceptid&formname=search&screen=0&lib=l20
\item The codes for \code{"DFO"} are defined at
http://www.dfo-mpo.gc.ca/science/data-donnees/code/list/014-eng.html
\item The codes for \code{"WHP CTD"} and \code{"WHP bottle"} are defined at
https://www.nodc.noaa.gov/woce/woce_v3/wocedata_1/whp/exchange/exchange_format_desc.htm
}
}
\seealso{
Other functions relating to data-quality flags: 
\code{\link{defaultFlags}()},
\code{\link{handleFlags,adp-method}},
\code{\link{handleFlags,argo-method}},
\code{\link{handleFlags,ctd-method}},
\code{\link{handleFlags,oce-method}},
\code{\link{handleFlags,section-method}},
\code{\link{handleFlags}()},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{initializeFlagSchemeInternal}()},
\code{\link{initializeFlagScheme}()},
\code{\link{initializeFlags,adp-method}},
\code{\link{initializeFlags,oce-method}},
\code{\link{initializeFlagsInternal}()},
\code{\link{initializeFlags}()},
\code{\link{setFlags,adp-method}},
\code{\link{setFlags,ctd-method}},
\code{\link{setFlags,oce-method}},
\code{\link{setFlags}()}

Other things related to oce data: 
\code{\link{initializeFlagSchemeInternal}()},
\code{\link{initializeFlagScheme}()}
}
\concept{functions relating to data-quality flags}
\concept{things related to oce data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.R
\name{adpEnsembleAverage}
\alias{adpEnsembleAverage}
\title{Ensemble Average an ADP Object in Time}
\usage{
adpEnsembleAverage(x, n = 5, leftover = FALSE, na.rm = TRUE, ...)
}
\arguments{
\item{x}{an \linkS4class{adp} object.}

\item{n}{number of pings to average together.}

\item{leftover}{a logical value indicating how to proceed in cases
where \code{n} does not divide evenly into the number of ensembles
in \code{x}. If \code{leftover} is \code{FALSE} (the default) then any extra
ensembles at the end of \code{x} are ignored. Otherwise, they are used
to create a final ensemble in the returned value.}

\item{na.rm}{a logical value indicating whether NA values should be stripped
before the computation proceeds}

\item{...}{extra arguments to be passed to the \code{mean()} function.}
}
\value{
A new \linkS4class{adp} object with ensembles averaged as specified. E.g. for an \code{adp} object with 100 pings and \code{n=5} the number of rows of the data arrays will be reduced by a factor of 5.
}
\description{
Ensemble averaging of \code{adp} objects is often necessary to
reduce the uncertainty in velocity estimates from single
pings. Many types of ADPs can be configured to perform the
ensemble averaging during the data collection, due to memory
limitations for long deployments. In cases where the instrument is
not memory limited, it may be desirable to perform the ensemble
averaging during post-processing, thereby reducing the overall
size of the data set and decreasing the uncertainty of the
velocity estimates (by averaging out Doppler noise).
}
\examples{
library(oce)
data(adp)
adpAvg <- adpEnsembleAverage(adp, n=2)
plot(adpAvg)

}
\seealso{
Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\author{
Clark Richards and Dan Kelley
}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{argShow}
\alias{argShow}
\title{Show an argument to a function, e.g. for debugging}
\usage{
argShow(x, nshow = 4, last = FALSE, sep = "=")
}
\arguments{
\item{x}{the argument}

\item{nshow}{number of values to show at first (if length(x)> 1)}

\item{last}{indicates whether this is the final argument to the function}

\item{sep}{the separator between name and value}
}
\description{
Show an argument to a function, e.g. for debugging
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/extdata.R
\docType{data}
\encoding{UTF-8}
\name{test_met_xml2.xml}
\alias{test_met_xml2.xml}
\title{Sample met dataset (XML format as of October 2019)}
\description{
This is a subset of a file downloaded with \code{\link[=download.met]{download.met()}}
from the Environment Canada "Climate Data"
website \verb{https://climate.weather.gc.ca/index_e.html}
on October 13, 2019.
}
\examples{
\dontrun{
read.met(system.file("extdata", "test_met_xml2.xml", package="oce"))
}

}
\seealso{
Other raw datasets: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{adp_rdi.000}},
\code{\link{ctd.cnv}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{test_met_csv1.csv}},
\code{\link{test_met_csv2.csv}},
\code{\link{xbt.edf}}

Other things related to met data: 
\code{\link{[[,met-method}},
\code{\link{[[<-,met-method}},
\code{\link{as.met}()},
\code{\link{download.met}()},
\code{\link{met-class}},
\code{\link{met}},
\code{\link{plot,met-method}},
\code{\link{read.met}()},
\code{\link{subset,met-method}},
\code{\link{summary,met-method}},
\code{\link{test_met_csv1.csv}},
\code{\link{test_met_csv2.csv}}
}
\concept{raw datasets}
\concept{things related to met data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sealevel.R
\name{as.sealevel}
\alias{as.sealevel}
\title{Coerce Data Into a Sealevel Object}
\usage{
as.sealevel(
  elevation,
  time,
  header = NULL,
  stationNumber = NA,
  stationVersion = NA,
  stationName = NULL,
  region = NULL,
  year = NA,
  longitude = NA,
  latitude = NA,
  GMTOffset = NA,
  decimationMethod = NA,
  referenceOffset = NA,
  referenceCode = NA,
  deltat
)
}
\arguments{
\item{elevation}{a list of sea-level heights in metres, in an hourly
sequence.}

\item{time}{optional list of times, in POSIXct format.  If missing, the list
will be constructed assuming hourly samples, starting at 0000-01-01
00:00:00.}

\item{header}{a character string as read from first line of a standard data
file.}

\item{stationNumber}{three-character string giving station number.}

\item{stationVersion}{single character for version of station.}

\item{stationName}{the name of station (at most 18 characters).}

\item{region}{the name of the region or country of station (at most 19
characters).}

\item{year}{the year of observation.}

\item{longitude}{the longitude in decimal degrees, positive east of
Greenwich.}

\item{latitude}{the latitude in decimal degrees, positive north of the
equator.}

\item{GMTOffset}{offset from GMT, in hours.}

\item{decimationMethod}{a coded value, with 1 meaning filtered, 2 meaning a
simple average of all samples, 3 meaning spot readings, and 4 meaning some
other method.}

\item{referenceOffset}{?}

\item{referenceCode}{?}

\item{deltat}{optional interval between samples, in hours (as for the
\code{\link[=ts]{ts()}} timeseries function). If this is not provided, and \code{t}
can be understood as a time, then the difference between the first two times
is used.  If this is not provided, and \code{t} cannot be understood as a
time, then 1 hour is assumed.}
}
\value{
A \linkS4class{sealevel} object (for details, see \code{\link[=read.sealevel]{read.sealevel()}}).
}
\description{
Coerces a dataset (minimally, a sequence of times and heights) into a
sealevel dataset.
The arguments are based on the standard data format, as were described in a
file formerly available at reference 1.
}
\examples{
library(oce)

# Construct a year of M2 tide, starting at the default time
# 0000-01-01T00:00:00.
h <- seq(0, 24*365)
elevation <- 2.0 * sin(2*pi*h/12.4172)
sl <- as.sealevel(elevation)
summary(sl)

# As above, but start at the Y2K time.
time <- as.POSIXct("2000-01-01") + h * 3600
sl <- as.sealevel(elevation, time)
summary(sl)
}
\references{
\verb{http://ilikai.soest.hawaii.edu/rqds/hourly.fmt} (this link
worked for years but failed at least temporarily on December 4, 2016).
}
\seealso{
The documentation for the \linkS4class{sealevel} class explains the
structure of sealevel objects, and also outlines the other functions dealing
with them.

Other things related to sealevel data: 
\code{\link{[[,sealevel-method}},
\code{\link{[[<-,sealevel-method}},
\code{\link{plot,sealevel-method}},
\code{\link{read.sealevel}()},
\code{\link{sealevel-class}},
\code{\link{sealevelTuktoyaktuk}},
\code{\link{sealevel}},
\code{\link{subset,sealevel-method}},
\code{\link{summary,sealevel-method}}
}
\author{
Dan Kelley
}
\concept{things related to sealevel data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/index.R
\name{read.index}
\alias{read.index}
\title{Read a NOAA ocean index file}
\usage{
read.index(
  file,
  format,
  missingValue,
  tz = getOption("oceTz"),
  debug = getOption("oceDebug")
)
}
\arguments{
\item{file}{a connection or a character string giving the name of the file
to load.  May be a URL.}

\item{format}{optional character string indicating the format type. If not
supplied, a guess will be made, based on examination of the first few lines
of the file. If supplied, the possibilities are \code{"noaa"} and
\code{"ucar"}. See \dQuote{Details}.}

\item{missingValue}{If supplied, this is a numerical value that indicates
invalid data. In some datasets, this is -99.9, but other values may be used.
If \code{missingValue} is not supplied, any data that have value equal
to or less than -99 are considered invalid. Set \code{missingValue}
to \code{TRUE} to turn of the processing of missing values.}

\item{tz}{character string indicating time zone to be assumed in the data.}

\item{debug}{a flag that turns on debugging, ignored in the present version
of the function.}
}
\value{
A data frame containing \code{t}, a POSIX time, and \code{index},
the numerical index.  The times are set to the 15th day of each month, which
is a guess that may need to be changed if so indicated by documentation (yet
to be located).
}
\description{
Read an ocean index file, in the format used by NOAA.
}
\details{
Reads a text-format index file, in either of two formats. If
\code{format} is missing, then the first line of the file is examined. If
that line contains 2 (whitespace-separated) tokens, then \code{"noaa"}
format is assumed. If it contains 13 tokens, then \code{"ucar"} format is assumed.
Otherwise, an error is reported.

In the \code{"noaa"} format, the two tokens on the first line are taken to
be the start year and the end year, respectively. The second line
must contain a single token, the missing value. All further lines must contain
12 tokens, for the values in January, February, etc.

In the \code{"ucar"} format, all data lines must contain the 13 tokens,
the first of which is the year, with the following 12 being the values
for January, February, etc.
}
\examples{

\dontrun{
library(oce)
par(mfrow=c(2, 1))
# 1. AO, Arctic oscillation
#  Note that data used to be at https://www.esrl.noaa.gov/psd/data/correlation/ao.data
ao <- read.index("https://psl.noaa.gov/data/correlation/ao.data")
aorecent <- subset(ao, t > as.POSIXct("2000-01-01"))
oce.plot.ts(aorecent$t, aorecent$index)
# 2. SOI, probably more up-to-date then data(soi, package="ocedata")
soi <- read.index("https://www.cgd.ucar.edu/cas/catalog/climind/SOI.signal.ascii")
soirecent <- subset(soi, t > as.POSIXct("2000-01-01"))
oce.plot.ts(soirecent$t, soirecent$index)
}
}
\references{
See \url{https://psl.noaa.gov/data/climateindices/list/}
for a list of indices.
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/amsr.R
\docType{class}
\name{amsr-class}
\alias{amsr-class}
\title{Class to Store AMSR-2 Satellite Data}
\description{
This class stores data from the AMSR-2 satellite.
}
\details{
The Advanced Microwave Scanning Radiometer (AMSR-2) is in current operation on
the Japan Aerospace Exploration Agency (JAXA) GCOM-W1 space craft, launched in
May 2012. Data are processed by Remote Sensing Systems. The satellite
completes an ascending and descending pass during local daytime and nighttime
hours respectively. Each daily file contains 7 daytime and 7 nighttime
maps of variables named as follows within the \code{data}
slot of amsr objects: \code{timeDay},
\code{SSTDay}, \code{LFwindDay} (wind at 10m sensed in
the 10.7GHz band), \code{MFwindDay} (wind at 10m sensed at 18.7GHz),
\code{vaporDay}, \code{cloudDay}, and \code{rainDay}, along with
similarly-named items that end in \code{Night}.
See reference 1 for additional information on the instrument, how
to cite the data source in a paper, etc.

The bands are stored in \code{\link[=raw]{raw()}} form, to save storage. The accessor
function \code{\link{[[,amsr-method}} can provide these values in \code{raw}
form or in physical units; \code{\link[=plot,amsr-method]{plot,amsr-method()}}, and
\code{\link[=summary,amsr-method]{summary,amsr-method()}} work with physical units.
}
\section{Slots}{

\describe{
\item{\code{data}}{As with all \code{oce} objects, the \code{data} slot
for \code{amsr} objects is a
\link{list} containing the main data for the object.
{}}

\item{\code{metadata}}{As with all \code{oce} objects, the \code{metadata} slot
for \code{amsr} objects is a \link{list} containing
information about the \code{data} or about the object itself.
Examples that are of common interest include  \code{longitude} and \code{latitude}, which define the grid.}

\item{\code{processingLog}}{As with all \code{oce} objects, the \code{processingLog} slot
for \code{amsr} objects is a
\link{list} with entries describing the creation and evolution
of the object. The contents are updated by various \code{oce} functions to
keep a record of processing steps.  Object summaries and
\code{\link[=processingLogShow]{processingLogShow()}} both display the log.}
}}

\section{Modifying slot contents}{


Although the \code{[[<-} operator may permit modification of the contents
of \linkS4class{amsr} objects (see \code{\link{[[<-,amsr-method}}),
it is better to use \code{\link[=oceSetData]{oceSetData()}} and \code{\link[=oceSetMetadata]{oceSetMetadata()}},
because those functions save an entry in the \code{processingLog}
that describes the change.
}

\section{Retrieving slot contents}{


The full contents of the \code{data} and \code{metadata} slots of a \linkS4class{amsr}
object may be retrieved in the standard R way using \code{\link[=slot]{slot()}}. For
example \code{slot(o,"data")} returns the \code{data} slot of an object named \code{o},
and similarly \code{slot(o,"metadata")} returns
the \code{metadata} slot.

The slots may also be obtained with the \code{\link{[[,amsr-method}}
operator, as e.g. \code{o[["data"]]} and \code{o[["metadata"]]}, respectively.

The \code{\link{[[,amsr-method}} operator can also
be used to retrieve items from within the \code{data} and \code{metadata} slots.
For example, \code{o[["temperature"]]} can be used to retrieve temperature
from an object containing that quantity. The rule is that a named
quantity is sought first within the object's \code{metadata} slot,
with the \code{data} slot being checked only if \code{metadata} does not
contain the item. This \code{[[} method can also be used to get
certain derived quantities, if the object contains sufficient
information to calculate them. For example, an object that holds
(practical) salinity, temperature and pressure, along with
longitude and latitude, has sufficient information to compute
Absolute Salinity, and so \code{o[["SA"]]} will yield the
calculated Absolute Salinity.

It is also possible to find items more directly, using \code{\link[=oceGetData]{oceGetData()}} and
\code{\link[=oceGetMetadata]{oceGetMetadata()}}, but neither of these functions can
retrieve derived items.
}

\references{
\enumerate{
\item Information on the satellite, how to cite the data, etc. is
provided at \verb{http://www.remss.com/missions/amsr/}.
\item A simple interface for viewing and downloading data is at
\verb{http://images.remss.com/amsr/amsr2_data_daily.html}.
}
}
\seealso{
Other classes holding satellite data: 
\code{\link{g1sst-class}},
\code{\link{landsat-class}},
\code{\link{satellite-class}}

Other things related to amsr data: 
\code{\link{[[,amsr-method}},
\code{\link{[[<-,amsr-method}},
\code{\link{amsr}},
\code{\link{composite,amsr-method}},
\code{\link{download.amsr}()},
\code{\link{plot,amsr-method}},
\code{\link{read.amsr}()},
\code{\link{subset,amsr-method}},
\code{\link{summary,amsr-method}}
}
\author{
Dan Kelley and Chantelle Layton
}
\concept{classes holding satellite data}
\concept{things related to amsr data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/met.R
\docType{data}
\name{met}
\alias{met}
\title{Sample met Object}
\source{
Environment Canada website on October 19, 2019.
}
\description{
This is sample \linkS4class{met} object containing data for Halifax, Nova Scotia,
during September of 2003 (the period during which Hurricane Juan struck the
city).
}
\details{
The data file was downloaded\if{html}{\out{<div class="sourceCode R">}}\preformatted{metFile <- download.met(id=6358, year=2003, month=9, destdir=".", type="xml")
}\if{html}{\out{</div>}}

Note that using \code{\link[=download.met]{download.met()}} avoids having to navigate the
the awkward Environment Canada website, but it imposes the burden
of having to know the station ID number.  With the data in-hand,
the object was then created (and its timezone adjusted) with\if{html}{\out{<div class="sourceCode R">}}\preformatted{met <- read.met(metFile)
met <- oceSetData(met, "time", met[["time"]]+4*3600,
                 note="add 4h to local time to get UTC time")
}\if{html}{\out{</div>}}

\emph{Historical note.} The \code{data(met)} object was changed on October 19,
2019, based on the data provided by Environment
Canada at that time. The previous version of \code{data(met)},
created in 2017, had been based on a data format that
Environment Canada no longer provided in 2019.  See the
notes on the \code{type} argument of \code{\link[=read.met]{read.met()}} for more on this
shift in the Environment Canada data format.
}
\seealso{
Other datasets provided with oce: 
\code{\link{adp}},
\code{\link{adv}},
\code{\link{amsr}},
\code{\link{argo}},
\code{\link{cm}},
\code{\link{coastlineWorld}},
\code{\link{ctdRaw}},
\code{\link{ctd}},
\code{\link{echosounder}},
\code{\link{landsat}},
\code{\link{lisst}},
\code{\link{lobo}},
\code{\link{ocecolors}},
\code{\link{rsk}},
\code{\link{sealevelTuktoyaktuk}},
\code{\link{sealevel}},
\code{\link{section}},
\code{\link{topoWorld}},
\code{\link{wind}},
\code{\link{xbt}}

Other things related to met data: 
\code{\link{[[,met-method}},
\code{\link{[[<-,met-method}},
\code{\link{as.met}()},
\code{\link{download.met}()},
\code{\link{met-class}},
\code{\link{plot,met-method}},
\code{\link{read.met}()},
\code{\link{subset,met-method}},
\code{\link{summary,met-method}},
\code{\link{test_met_csv1.csv}},
\code{\link{test_met_csv2.csv}},
\code{\link{test_met_xml2.xml}}
}
\concept{datasets provided with oce}
\concept{things related to met data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/echosounder.R
\name{as.echosounder}
\alias{as.echosounder}
\title{Coerce Data into an Echosounder Object}
\usage{
as.echosounder(
  time,
  depth,
  a,
  src = "",
  sourceLevel = 220,
  receiverSensitivity = -55.4,
  transmitPower = 0,
  pulseDuration = 400,
  beamwidthX = 6.5,
  beamwidthY = 6.5,
  frequency = 41800,
  correction = 0
)
}
\arguments{
\item{time}{times of pings}

\item{depth}{depths of samples within pings}

\item{a}{matrix of amplitudes}

\item{src}{optional string indicating data source}

\item{sourceLevel}{source level, in dB (uPa at 1m), denoted \code{sl} in
reference 1 p15, where it is in units 0.1dB (uPa at 1m)}

\item{receiverSensitivity}{receiver sensitivity of the main element, in
dB(counts/uPa), denoted \code{rs} in reference 1 p15, where it is in units of
0.1dB(counts/uPa)}

\item{transmitPower}{transmit power reduction factor, in dB, denoted
\code{tpow} in reference 1 p10, where it is in units 0.1 dB.}

\item{pulseDuration}{duration of transmitted pulse in us}

\item{beamwidthX}{x-axis -3dB one-way beamwidth in deg, denoted \code{bwx}
in reference 1 p16, where the unit is 0.2 deg}

\item{beamwidthY}{y-axis -3dB one-way beamwidth in deg, denoted \code{bwx}
in reference 1 p16, where the unit is 0.2 deg}

\item{frequency}{transducer frequency in Hz, denoted \code{fq} in reference 1 p16}

\item{correction}{user-defined calibration correction in dB, denoted
\code{corr} in reference 1 p14, where the unit is 0.01dB.}
}
\value{
An \linkS4class{echosounder} object.
}
\description{
Coerces a dataset into a echosounder dataset.
}
\details{
Creates an echosounder file.  The defaults for e.g.  \code{transmitPower}
are taken from the \code{echosounder} dataset, and they are unlikely to make
sense generally.
}
\seealso{
Other things related to echosounder data: 
\code{\link{[[,echosounder-method}},
\code{\link{[[<-,echosounder-method}},
\code{\link{echosounder-class}},
\code{\link{echosounder}},
\code{\link{findBottom}()},
\code{\link{plot,echosounder-method}},
\code{\link{read.echosounder}()},
\code{\link{subset,echosounder-method}},
\code{\link{summary,echosounder-method}}
}
\author{
Dan Kelley
}
\concept{things related to echosounder data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{decimate}
\alias{decimate}
\title{Smooth and Decimate, or Subsample, an Oce Object}
\usage{
decimate(x, by = 10, to, filter, debug = getOption("oceDebug"))
}
\arguments{
\item{x}{an \linkS4class{oce} object.}

\item{by}{an indication of the subsampling.  If this is a single number,
then it indicates the spacing between elements of \code{x} that are
selected.  If it is two numbers (a condition only applicable if \code{x} is
an \code{echosounder} object, at present), then the first number indicates
the time spacing and the second indicates the depth spacing.}

\item{to}{Indices at which to subsample.  If given, this over-rides
\code{by}.}

\item{filter}{optional list of numbers representing a digital filter to be
applied to each variable in the \code{data} slot of \code{x}, before
decimation is done. If not supplied, then the decimation is done strictly by
sub-sampling.}

\item{debug}{a flag that turns on debugging.  Set to 1 to get a moderate
amount of debugging information, or to 2 to get more.}
}
\value{
An \linkS4class{oce} object that has been subsampled appropriately.
}
\description{
Later on, other methods will be added, and \code{\link[=ctdDecimate]{ctdDecimate()}} will be
retired in favour of this, a more general, function.  The filtering is done
with the \code{\link[=filter]{filter()}} function of the stats package.
}
\section{Bugs}{
 Only a preliminary version of this function is provided in
the present package.  It only works for objects of class \code{echosounder},
for which the decimation is done after applying a running median filter and
then a boxcar filter, each of length equal to the corresponding component of
\code{by}.
}

\examples{
library(oce)
data(adp)
plot(adp)
adpDec <- decimate(adp,by=2,filter=c(1/4, 1/2, 1/4))
plot(adpDec)
}
\seealso{
Filter coefficients may be calculated using
\code{\link[=makeFilter]{makeFilter()}}.  (Note that \code{\link[=ctdDecimate]{ctdDecimate()}} will be
retired when the present function gains equivalent functionality.)
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/accessors.R
\name{oceGetData}
\alias{oceGetData}
\title{Get Something from an oce data Slot}
\usage{
oceGetData(object, name)
}
\arguments{
\item{object}{an \linkS4class{oce} object.}

\item{name}{String indicating the name of the item to be found.}
}
\description{
In contrast to the various \code{[[} functions, this is
guaranteed to look only within the \code{data} slot. If
the named item is not found, \code{NULL} is returned.
}
\seealso{
Other things related to the data slot: 
\code{\link{oceDeleteData}()},
\code{\link{oceRenameData}()},
\code{\link{oceSetData}()}
}
\author{
Dan Kelley
}
\concept{things related to the data slot}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/section.R
\name{initializeFlagScheme,section-method}
\alias{initializeFlagScheme,section-method}
\title{Establish a data-quality scheme for a section object}
\usage{
\S4method{initializeFlagScheme}{section}(
  object,
  name = NULL,
  mapping = NULL,
  default = NULL,
  update = NULL,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{object}{An oce object.}

\item{name}{a character value naming the scheme. If this refers
to a pre-defined scheme, then \code{mapping} must not be provided,
because doing so would contradict the pre-defined scheme, defeating
its purpose of providing concreteness and clarity.}

\item{mapping}{a list of named items describing the mapping from
flag meaning to flag numerical value, e.g \code{list(good=1, bad=2)}
might be used for a hypothetical class.}

\item{default}{an integer vector of flag values that are not considered
to be good. If this is not provided, but if \code{name} is \code{"argo"},
\code{"BODC"},
\code{"DFO"},
\code{"WHP bottle"}, or
\code{"WHP CTD"}, then a conservative value will be set automatically,
equal to the list of flag values that designate bad or questionable data.
For example, for \code{name="WHP CTD"}, the setting will be
\code{c(1,3,4,5,6,7,9)}, leaving only value \code{2}, which corresponds
with "acceptable" in the notation used for that flag scheme.}

\item{update}{a logical value indicating whether the scheme provided is
to update an existing scheme.  The default value, \code{FALSE}, prevents such
an attempt to alter an existing flag scheme, if one is already embedded
in \code{object}.}

\item{debug}{an integer set to 0 for quiet action or to 1 for some debugging.}
}
\value{
An object with the \code{metadata} slot containing \code{flagScheme}.
}
\description{
This function stores add an item named \code{flagScheme}
to the \code{metadata} slot of an object inheriting from
\linkS4class{section}. This is a list containing two
items: \code{name} and \code{mapping}, as provided in the
function arguments.
The purpose is both to document a flag scheme
and to make it so that \code{\link[=initializeFlags]{initializeFlags()}},
\code{\link[=setFlags]{setFlags()}} and \code{\link[=handleFlags]{handleFlags()}}
can specify flags by
name, as opposed to number. This is a generic function,
that may be specialized to the class of \code{object}
(see \dQuote{Details}).
}
\details{
The following pre-defined schemes are available (note that the
names are simplified from the phrases used in defining
documentation):
\itemize{
\item \code{name="argo"} defaults \code{mapping} to
OLD (prior to June 10, 2020)
}\preformatted{list(not_assessed=0, passed_all_tests=1, probably_good=2,
     probably_bad=3, bad=4, averaged=7,
     interpolated=8, missing=9)
}

NEW (after June 10, 2020)\preformatted{list(not_assessed=0, passed_all_tests=1, probably_good=2,
     probably_bad=3, bad=4, changed=5, not_used_6=6, not_used_7=7,
     estimated=8, missing=9)
}

See reference 1 for a deeper explanation of the meanings of these codes.
\itemize{
\item \code{name="BODC"} defaults \code{mapping} to
}\preformatted{list(no_quality_control=0, good=1, probably_good=2,
     probably_bad=3, bad=4, changed=5,
     below_detection=6, in_excess=7, interpolated=8,
     missing=9)
}

See reference 2 for a deeper explanation of the meanings of these codes,
and note that codes \code{A} and \code{Q} are not provided in
oce.
\itemize{
\item \code{name="DFO"} defaults \code{mapping} to
}\preformatted{list(no_quality_control=0, appears_correct=1, appears_inconsistent=2,
     doubtful=3, erroneous=4, changed=5,
     qc_by_originator=8, missing=9)
}

See reference 3 for a deeper explanation of the meanings of these codes.
\itemize{
\item \code{name="WHP bottle"} defaults \code{mapping} to
}\preformatted{list(no_information=1, no_problems_noted=2, leaking=3,
     did_not_trip=4, not_reported=5, discrepency=6,
     unknown_problem=7, did_not_trip=8, no_sample=9)
}

See reference 4 for a deeper explanation of the meanings of these codes.
\itemize{
\item \code{name="WHP CTD"} defaults \code{mapping} to
}\preformatted{list(not_calibrated=1, acceptable=2, questionable=3,
    bad=4, not_reported=5, interpolated=6,
    despiked=7, missing=9)
}

See reference 4 for a deeper explanation of the meanings of these codes.
}
\examples{
\dontrun{
data(section)
section <- read.section("a03_hy1.csv", sectionId="a03", institute="SIO",
                        ship="R/V Professor Multanovskiy", scientist="Vladimir Tereschenkov")
sectionWithFlags <- initializeFlagScheme(section, "WHP bottle")
station1 <- sectionWithFlags[["station", 1]]
str(station1[["flagScheme"]])
}
}
\references{
\enumerate{
\item The codes for \code{"argo"} are derived from information in Table 4.1
of Wong, Annie, Robert Keeley, Thierry Carval, and Argo Data Management Team
(8 January 2020), "Argo Quality Control Manual for CTD and Trajectory Data, Version 3.3,"
available at \url{https://archimer.ifremer.fr/doc/00228/33951/}
as of June 2020.
\item The codes for \code{"BODC"} are defined at
http://seadatanet.maris2.nl/v_bodc_vocab_v2/browse.asp?order=conceptid&formname=search&screen=0&lib=l20
\item The codes for \code{"DFO"} are defined at
http://www.dfo-mpo.gc.ca/science/data-donnees/code/list/014-eng.html
\item The codes for \code{"WHP CTD"} and \code{"WHP bottle"} are defined at
https://www.nodc.noaa.gov/woce/woce_v3/wocedata_1/whp/exchange/exchange_format_desc.htm
}
}
\seealso{
Other functions relating to data-quality flags: 
\code{\link{defaultFlags}()},
\code{\link{handleFlags,adp-method}},
\code{\link{handleFlags,argo-method}},
\code{\link{handleFlags,ctd-method}},
\code{\link{handleFlags,oce-method}},
\code{\link{handleFlags,section-method}},
\code{\link{handleFlags}()},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{initializeFlagScheme,oce-method}},
\code{\link{initializeFlagSchemeInternal}()},
\code{\link{initializeFlagScheme}()},
\code{\link{initializeFlags,adp-method}},
\code{\link{initializeFlags,oce-method}},
\code{\link{initializeFlagsInternal}()},
\code{\link{initializeFlags}()},
\code{\link{setFlags,adp-method}},
\code{\link{setFlags,ctd-method}},
\code{\link{setFlags,oce-method}},
\code{\link{setFlags}()}

Other things related to section data: 
\code{\link{[[,section-method}},
\code{\link{[[<-,section-method}},
\code{\link{as.section}()},
\code{\link{handleFlags,section-method}},
\code{\link{plot,section-method}},
\code{\link{read.section}()},
\code{\link{section-class}},
\code{\link{sectionAddStation}()},
\code{\link{sectionGrid}()},
\code{\link{sectionSmooth}()},
\code{\link{sectionSort}()},
\code{\link{section}},
\code{\link{subset,section-method}},
\code{\link{summary,section-method}}
}
\concept{functions relating to data-quality flags}
\concept{things related to section data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/AllClass.R
\name{setFlags,oce-method}
\alias{setFlags,oce-method}
\title{Set data-quality flags within a oce object}
\usage{
\S4method{setFlags}{oce}(
  object,
  name = NULL,
  i = NULL,
  value = NULL,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{object}{An oce object.}

\item{name}{Character string indicating the name of the variable to be flagged. If
this variable is not contained in the object's \code{data} slot, an error is reported.}

\item{i}{Indication of where to insert the flags; see \dQuote{Description} for
general rules and \dQuote{Details} for rules for \linkS4class{oce}
objects.}

\item{value}{The value to be inserted in the flag.}

\item{debug}{Integer set to 0 for quiet action or to 1 for some debugging.}
}
\value{
An object with flags set as indicated.
}
\description{
This function changes specified entries in the data-quality
flags of a oce object, which are stored within
a list named \code{flags} that resides in the \code{metadata}
slot. If the object already has a flag set up for \code{name},
then only the specified entries are altered. If not, the flag
entry is first created and its entries set to \code{default},
after which the entries specified by \code{i}
are changed to \code{value}.

The specification is made with \code{i}, the form of which
is determined by the data item in question. Generally,
the rules are as follows:
\enumerate{
\item If the data item is a vector, then \code{i} must be (a)
an integer vector specifying indices to be set to \code{value},
(b) a logical vector of length matching the data item, with
\code{TRUE} meaning to set the flag to \code{value}, or (c)
a function that takes an \code{oce} object as its single
argument, and returns a vector in either of the forms
just described.
\item If the data item is an array, then \code{i} must be
(a) a data frame of integers whose rows specify spots to change
(where the number of columns matches the number of dimensions
of the data item), (b) a logical array that has dimension equal to
that of the data item, or (c) a function that takes an \code{oce}
object as its single input and returns such a data frame or array.
}

See \dQuote{Details} for the particular case of \linkS4class{oce} objects.
}
\details{
This generic function is overridden by specialized functions for some object classes.
}
\seealso{
Other functions relating to data-quality flags: 
\code{\link{defaultFlags}()},
\code{\link{handleFlags,adp-method}},
\code{\link{handleFlags,argo-method}},
\code{\link{handleFlags,ctd-method}},
\code{\link{handleFlags,oce-method}},
\code{\link{handleFlags,section-method}},
\code{\link{handleFlags}()},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{initializeFlagScheme,oce-method}},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{initializeFlagSchemeInternal}()},
\code{\link{initializeFlagScheme}()},
\code{\link{initializeFlags,adp-method}},
\code{\link{initializeFlags,oce-method}},
\code{\link{initializeFlagsInternal}()},
\code{\link{initializeFlags}()},
\code{\link{setFlags,adp-method}},
\code{\link{setFlags,ctd-method}},
\code{\link{setFlags}()}
}
\concept{functions relating to data-quality flags}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/landsat.R
\name{plot,landsat-method}
\alias{plot,landsat-method}
\alias{plot.landsat}
\title{Plot a landsat Object}
\usage{
\S4method{plot}{landsat}(
  x,
  band,
  which = 1,
  decimate = TRUE,
  zlim,
  utm = FALSE,
  col = oce.colorsPalette,
  drawPalette = TRUE,
  showBandName = TRUE,
  alpha.f = 1,
  red.f = 1.7,
  green.f = 1.5,
  blue.f = 6,
  offset = c(0, -0.05, -0.2, 0),
  transform = diag(c(red.f, green.f, blue.f, alpha.f)),
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{x}{a \linkS4class{landsat} object.}

\item{band}{If given, the name of the band.  For Landsat-8 data, this may be
one of: \code{"aerosol"}, \code{"blue"}, \code{"green"}, \code{"red"},
\code{"nir"}, \code{"swir1"}, \code{"swir2"}, \code{"panchromatic"},
\code{"cirrus"}, \code{"tirs1"}, or \code{"tirs2"}.  For Landsat-7 data,
this may be one of \code{"blue"}, \code{"green"}, \code{"red"},
\code{"nir"}, \code{"swir1"}, \code{"tirs1"}, \code{"tirs2"},
\code{"swir2"}, or \code{"panchromatic"}.  For Landsat data prior to
Landsat-7, this may be one of \code{"blue"}, \code{"green"},
\code{"red"}, \code{"nir"}, \code{"swir1"}, \code{"tirs1"},
\code{"tirs2"}, or \code{"swir2"}.  If \code{band} is not given, the
(\code{"tirs1"}) will be used if it exists in the object data, or
otherwise the first band will be used.  In addition to the above, using
\code{band="temperature"} will plot an estimate of at-satellite
brightness temperature, computed from the \code{tirs1} band, and
\code{band="terralook"} will plot a sort of natural color by combining
the \code{red}, \code{green}, \code{blue} and \code{nir} bands
according to the formula provided at
\verb{https://lta.cr.usgs.gov/terralook/what_is_terralook} (a
website that worked once, but failed as of Feb 2, 2017).}

\item{which}{Desired plot type; 1=image, 2=histogram.}

\item{decimate}{An indication of the desired decimation,
passed to \code{\link[=imagep]{imagep()}} for image plots.
The default yields faster plotting.  Some decimation is sensible for
full-size images, since no graphical displays can show 16 thousand pixels
on a side.}

\item{zlim}{Either a pair of numbers giving the limits for the colorscale,
or \code{"histogram"} to have a flattened histogram (i.e. to maximally
increase contrast throughout the domain.)  If not given, the 1 and 99
percent quantiles are calculated and used as limits.}

\item{utm}{A logical value indicating whether to use UTS (easting and northing) instead
of longitude and latitude on plot.}

\item{col}{Either a function yielding colors, taking a single integer
argument with the desired number of colors, or the string
\code{"natural"}, which combines the information in the \code{red},
\code{green} and \code{blue} bands and produces a natural-hue image.  In
the latter case, the band designation is ignored, and the object must
contain the three color bands.}

\item{drawPalette}{Indication of the type of palette to draw, if
any. See \code{\link[=imagep]{imagep()}} for details.}

\item{showBandName}{A logical indicating whether the band name is to
plotted in the top margin, near the right-hand side.}

\item{alpha.f}{Argument used if \code{col="natural"}, to adjust colors
with \code{\link[=adjustcolor]{adjustcolor()}}.}

\item{red.f}{Argument used if \code{col="natural"}, to adjust colors with
\code{\link[=adjustcolor]{adjustcolor()}}.  Higher values of \code{red.f} cause red hues
to be emphasized (e.g. dry land).}

\item{green.f}{Argument used if \code{col="natural"}, to adjust colors with
\code{\link[=adjustcolor]{adjustcolor()}}.  Higher values of \code{green.f} emphasize
green hues (e.g. forests).}

\item{blue.f}{Argument used if \code{band="terralook"}, to adjust colors with
\code{\link[=adjustcolor]{adjustcolor()}}.  Higher values of \code{blue.f} emphasize blue
hues (e.g. ocean).}

\item{offset}{Argument used if \code{band="terralook"}, to adjust colors with
\code{\link[=adjustcolor]{adjustcolor()}}.}

\item{transform}{Argument used if \code{band="terralook"}, to adjust colors
with \code{\link[=adjustcolor]{adjustcolor()}}.}

\item{debug}{Set to a positive value to get debugging information during
processing.}

\item{...}{optional arguments passed to plotting functions.}
}
\description{
Plot the data within a landsat image, or information computed from the
data. The second category includes possibilities such as an estimate of
surface temperature and the \code{"terralook"} estimate of a natural-color
view.
}
\details{
For Landsat-8 data, the \code{band} may be
one of: \code{"aerosol"}, \code{"blue"}, \code{"green"}, \code{"red"},
\code{"nir"}, \code{"swir1"}, \code{"swir2"}, \code{"panchromatic"},
\code{"cirrus"}, \code{"tirs1"}, or \code{"tirs2"}.

For Landsat-7 data, \code{band} may be one of \code{"blue"}, \code{"green"}, \code{"red"},
\code{"nir"}, \code{"swir1"}, \code{"tirs1"}, \code{"tirs2"},
\code{"swir2"}, or \code{"panchromatic"}.

For Landsat data prior to
Landsat-7, \code{band} may be one of \code{"blue"}, \code{"green"},
\code{"red"}, \code{"nir"}, \code{"swir1"}, \code{"tirs1"},
\code{"tirs2"}, or \code{"swir2"}.

If \code{band} is not given, the
(\code{"tirs1"}) will be used if it exists in the object data, or
otherwise the first band will be used.

In addition to the above there are also some pseudo-bands that
can be plotted, as follows.
\itemize{
\item Setting \code{band="temperature"} will plot an estimate
of at-satellite brightness temperature, computed from the
\code{tirs1} band.
\item Setting \code{band="terralook"} will plot a sort of natural
color by combining the \code{red}, \code{green}, \code{blue} and
\code{nir} bands according to the formula provided at
\verb{https://lta.cr.usgs.gov/terralook/what_is_terralook} (a
website that worked once, but failed as of Feb 2, 2017), namely
that the \code{red}-band data are provided as the \code{red}
argument of the \code{\link[=rgb]{rgb()}} function, while
the \code{green} argument is computed as
2/3 of the \code{green}-band data plus 1/3 of the \code{nir}-band data, and
the \code{blue} argument is computed as 2/3 of the \code{green}-band
data minus 1/3 of the \code{nir}-band data. (This is not a typo: the
\code{blue} band is not used.)
}
}
\seealso{
Other things related to landsat data: 
\code{\link{[[,landsat-method}},
\code{\link{[[<-,landsat-method}},
\code{\link{landsat-class}},
\code{\link{landsatAdd}()},
\code{\link{landsatTrim}()},
\code{\link{landsat}},
\code{\link{read.landsat}()},
\code{\link{summary,landsat-method}}

Other functions that plot oce data: 
\code{\link{download.amsr}()},
\code{\link{plot,adp-method}},
\code{\link{plot,adv-method}},
\code{\link{plot,amsr-method}},
\code{\link{plot,argo-method}},
\code{\link{plot,bremen-method}},
\code{\link{plot,cm-method}},
\code{\link{plot,coastline-method}},
\code{\link{plot,ctd-method}},
\code{\link{plot,gps-method}},
\code{\link{plot,ladp-method}},
\code{\link{plot,lisst-method}},
\code{\link{plot,lobo-method}},
\code{\link{plot,met-method}},
\code{\link{plot,odf-method}},
\code{\link{plot,rsk-method}},
\code{\link{plot,satellite-method}},
\code{\link{plot,sealevel-method}},
\code{\link{plot,section-method}},
\code{\link{plot,tidem-method}},
\code{\link{plot,topo-method}},
\code{\link{plot,windrose-method}},
\code{\link{plot,xbt-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{tidem-class}}
}
\author{
Dan Kelley
}
\concept{functions that plot oce data}
\concept{things related to landsat data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/lobo.R
\name{plot,lobo-method}
\alias{plot,lobo-method}
\alias{plot.lobo}
\title{Plot a lobo object}
\usage{
\S4method{plot}{lobo}(
  x,
  which = c(1, 2, 3),
  mgp = getOption("oceMgp"),
  mar = c(mgp[2] + 1, mgp[1] + 1, 1, mgp[1] + 1.25),
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{x}{a \linkS4class{lobo} object.}

\item{which}{A vector of numbers or character strings, indicating the
quantities to plot.  These are stacked in a single column.  The possible
values for \code{which} are as follows: \code{1} or \code{"temperature"} for
a time series of temperature; \code{2} or \code{"salinity"} for salinity;
\code{3} or \code{"TS"} for a TS diagram (which uses \code{eos="unesco"}),
\code{4} or \code{"u"} for a
timeseries of the u component of velocity; \code{5} or \code{"v"} for a
timeseries of the v component of velocity; \code{6} or \code{"nitrate"} for
a timeseries of nitrate concentration; \code{7} or \code{"fluorescence"} for
a timeseries of fluorescence value.}

\item{mgp}{3-element numerical vector to use for \code{par(mgp)}, and also
for \code{par(mar)}, computed from this.  The default is tighter than the R
default, in order to use more space for the data and less for the axes.}

\item{mar}{value to be used with \code{\link{par}}\code{("mar")}.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}

\item{\dots}{optional arguments passed to plotting functions.}
}
\description{
Plot a summary diagram for lobo data.
}
\seealso{
Other functions that plot oce data: 
\code{\link{download.amsr}()},
\code{\link{plot,adp-method}},
\code{\link{plot,adv-method}},
\code{\link{plot,amsr-method}},
\code{\link{plot,argo-method}},
\code{\link{plot,bremen-method}},
\code{\link{plot,cm-method}},
\code{\link{plot,coastline-method}},
\code{\link{plot,ctd-method}},
\code{\link{plot,gps-method}},
\code{\link{plot,ladp-method}},
\code{\link{plot,landsat-method}},
\code{\link{plot,lisst-method}},
\code{\link{plot,met-method}},
\code{\link{plot,odf-method}},
\code{\link{plot,rsk-method}},
\code{\link{plot,satellite-method}},
\code{\link{plot,sealevel-method}},
\code{\link{plot,section-method}},
\code{\link{plot,tidem-method}},
\code{\link{plot,topo-method}},
\code{\link{plot,windrose-method}},
\code{\link{plot,xbt-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{tidem-class}}

Other things related to lobo data: 
\code{\link{[[,lobo-method}},
\code{\link{[[<-,lobo-method}},
\code{\link{as.lobo}()},
\code{\link{lobo-class}},
\code{\link{lobo}},
\code{\link{read.lobo}()},
\code{\link{subset,lobo-method}},
\code{\link{summary,lobo-method}}
}
\author{
Dan Kelley
}
\concept{functions that plot oce data}
\concept{things related to lobo data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.R
\name{initializeFlags,adp-method}
\alias{initializeFlags,adp-method}
\title{Create storage for a flag, and initialize values, for a adp object}
\usage{
\S4method{initializeFlags}{adp}(
  object,
  name = NULL,
  value = NULL,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{object}{An \linkS4class{oce} object.}

\item{name}{Character value indicating the name of a variable within the
\code{data} slot of \code{object}.}

\item{value}{Numerical or character value to be stored in the newly-created
entry within \code{flags}. (A character value will only work if
\code{\link[=initializeFlags]{initializeFlags()}} has been used first on \code{object}.)}

\item{debug}{Integer set to 0 for quiet action or to 1 for some debugging.}
}
\value{
An object with the \code{flags} item within the
\code{metadata} slot set up as indicated.
}
\description{
This function creates an item for a named variable within
the \code{flags} entry in the object's \code{metadata} slot.
The purpose is both to document a flag scheme
and to make it so that \code{\link[=initializeFlags]{initializeFlags()}}
and \code{\link[=setFlags]{setFlags()}} can specify flags by
name, in addition to number. A generic function, it is
specialized for some classes via interpretation of the
\code{scheme} argument (see \dQuote{Details}, for those
object classes that have such specializations).
}
\details{
If \code{object} already contains a \code{flags} entry with
the indicated name, then it is returned unaltered, and a warning
is issued.
}
\seealso{
Other functions relating to data-quality flags: 
\code{\link{defaultFlags}()},
\code{\link{handleFlags,adp-method}},
\code{\link{handleFlags,argo-method}},
\code{\link{handleFlags,ctd-method}},
\code{\link{handleFlags,oce-method}},
\code{\link{handleFlags,section-method}},
\code{\link{handleFlags}()},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{initializeFlagScheme,oce-method}},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{initializeFlagSchemeInternal}()},
\code{\link{initializeFlagScheme}()},
\code{\link{initializeFlags,oce-method}},
\code{\link{initializeFlagsInternal}()},
\code{\link{initializeFlags}()},
\code{\link{setFlags,adp-method}},
\code{\link{setFlags,ctd-method}},
\code{\link{setFlags,oce-method}},
\code{\link{setFlags}()}
}
\concept{functions relating to data-quality flags}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/lobo.R
\name{subset,lobo-method}
\alias{subset,lobo-method}
\title{Subset a LOBO Object}
\usage{
\S4method{subset}{lobo}(x, subset, ...)
}
\arguments{
\item{x}{a \linkS4class{lobo} object.}

\item{subset}{a condition to be applied to the \code{data} portion of
\code{x}.  See \sQuote{Details}.}

\item{\dots}{ignored.}
}
\value{
A \linkS4class{lobo} object.
}
\description{
Subset an lobo object, in a way that is somewhat
analogous to \code{\link[=subset.data.frame]{subset.data.frame()}}.
}
\seealso{
Other things related to lobo data: 
\code{\link{[[,lobo-method}},
\code{\link{[[<-,lobo-method}},
\code{\link{as.lobo}()},
\code{\link{lobo-class}},
\code{\link{lobo}},
\code{\link{plot,lobo-method}},
\code{\link{read.lobo}()},
\code{\link{summary,lobo-method}}

Other functions that subset oce objects: 
\code{\link{subset,adp-method}},
\code{\link{subset,adv-method}},
\code{\link{subset,amsr-method}},
\code{\link{subset,argo-method}},
\code{\link{subset,cm-method}},
\code{\link{subset,coastline-method}},
\code{\link{subset,ctd-method}},
\code{\link{subset,echosounder-method}},
\code{\link{subset,met-method}},
\code{\link{subset,oce-method}},
\code{\link{subset,odf-method}},
\code{\link{subset,rsk-method}},
\code{\link{subset,sealevel-method}},
\code{\link{subset,section-method}},
\code{\link{subset,topo-method}},
\code{\link{subset,xbt-method}}
}
\author{
Dan Kelley
}
\concept{functions that subset oce objects}
\concept{things related to lobo data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ladp.R
\name{summary,ladp-method}
\alias{summary,ladp-method}
\title{Summarize an ladp object}
\usage{
\S4method{summary}{ladp}(object, ...)
}
\arguments{
\item{object}{an \linkS4class{ladp} object.}

\item{...}{Further arguments passed to or from other methods.}
}
\value{
A matrix containing statistics of the elements of the \code{data} slot.
}
\description{
Pertinent summary information is presented, including the station name,
sampling location, data ranges, etc.
}
\seealso{
Other things related to ladp data: 
\code{\link{[[,ladp-method}},
\code{\link{[[<-,ladp-method}},
\code{\link{as.ladp}()},
\code{\link{ladp-class}},
\code{\link{plot,ladp-method}}
}
\author{
Dan Kelley
}
\concept{things related to ladp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/AllClass.R
\name{summary,oce-method}
\alias{summary,oce-method}
\title{Summarize an oce Object}
\usage{
\S4method{summary}{oce}(object, ...)
}
\arguments{
\item{object}{The object to be summarized.}

\item{...}{Extra arguments (ignored)}
}
\description{
Provide a textual summary of some pertinent aspects of the object, including
selected components of its \code{metadata} slot, statistical and
dimensional information on the entries in the \code{data} slot,
and a listing of the contents of its \code{processingLog} slot.
The details depend on the class of the object, especially for
the \code{metadata} slot, so it can help to consult the specialized
documentation, e.g. \link{summary,ctd-method} for CTD objects
(i.e. objects inheriting from the \linkS4class{ctd} class.)
It is important to note that this is not
a good way to learn the details of the object contents. Instead,
for an object named \code{object}, say, one might use \code{str(object)}
to learn about all the contents, or \code{str(object[["metadata"]])}
to learn about the \code{metadata}, etc.
}
\examples{
o <- new("oce")
summary(o)
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.R
\name{beamToXyzAdp}
\alias{beamToXyzAdp}
\title{Convert ADP From Beam to XYZ Coordinates}
\usage{
beamToXyzAdp(x, debug = getOption("oceDebug"))
}
\arguments{
\item{x}{an \linkS4class{adp} object.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\value{
An object with the first 3 velocity indices having been altered to
represent velocity components in xyz (or instrument) coordinates.  (For
\code{rdi} data, the values at the 4th velocity index are changed to
represent the "error" velocity.)
To indicate the change, the value of \code{x[["oceCoordinate"]]} is
changed from \code{beam} to \code{xyz}.
}
\description{
Convert ADP velocity components from a beam-based coordinate system to a
xyz-based coordinate system. The action depends on the type of object.
Objects creating by reading RDI Teledyne, Sontek, and some Nortek
instruments are handled directly. However, Nortek
data stored in in the AD2CP format are handled by the specialized
function \code{\link[=beamToXyzAdpAD2CP]{beamToXyzAdpAD2CP()}}, the documentation for which
should be consulted, rather than the material given blow.
}
\details{
For a 3-beam Nortek \code{aquadopp} object, the beams are transformed into
velocities using the matrix stored in the header.

For 4-beam objects (and for the slanted 4 beams of 5-beam
objects), the along-beam velocity components \eqn{B_1}{B1}
\eqn{B_2}{B1}, \eqn{B_3}{B3}, and \eqn{B_4}{B4}
are converted to Cartesian velocity components \eqn{u}{u}
\eqn{v}{v} and \eqn{w}{w}
using formulae from section 5.5 of \emph{RD Instruments} (1998), viz. the
along-beam velocity components \eqn{B_1}{B1}, \eqn{B_2}{B2}, \eqn{B_3}{B3},
and \eqn{B_4}{B4} are used to calculate velocity components in a cartesian
system referenced to the instrument using the following formulae:
\eqn{u=ca(B_1-B_2)}{u=c*a*(B1-B2)}, \eqn{v=ca(B_4-B_3)}{v=c*a*(B4-B3)},
\eqn{w=-b(B_1+B_2+B_3+B_4)}{w=-b*(B1+B2+B3+B4)}. In addition to these,
an estimate of the
error in velocity is computed as
\eqn{e=d(B_1+B_2-B_3-B_4)}{e=d*(B1+B2-B3-B4)}.
The geometrical factors in these formulae are:
\code{c} is +1 for convex beam geometry or -1 for concave beam geometry,
\eqn{a=1/(2\sin\theta)}{a=1/(2*sin(theta))}
where \eqn{\theta}{theta} is the angle the beams make to the axial direction
(which is available as \code{x[["beamAngle"]]}),
\eqn{b=1/(4\cos\theta)}{b=1/(4*cos(theta))}, and
\eqn{d=a/\sqrt{2}}{d=a/sqrt(2)}.
}
\references{
\enumerate{
\item Teledyne RD Instruments. \dQuote{ADCP Coordinate Transformation: Formulas and
Calculations,} January 2010. P/N 951-6079-00.
\item WHOI/USGS-provided Matlab code for beam-enu transformation
\samp{http://woodshole.er.usgs.gov/pubs/of2005-1429/MFILES/AQDPTOOLS/beam2enu.m}
}
}
\seealso{
See \code{\link[=read.adp]{read.adp()}} for other functions that relate to
objects of class \code{"adp"}.

Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley
}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{renameData}
\alias{renameData}
\title{Rename items in the data slot of an oce object (defunct)}
\usage{
renameData(x, old = NULL, new = NULL)
}
\arguments{
\item{x}{Ignored, since this function is defunct.}

\item{old}{Ignored, since this function is defunct.}

\item{new}{Ignored, since this function is defunct.}
}
\description{
\strong{WARNING:} This function will be removed in oce 1.6.0; see \link{oce-defunct}.
}
\author{
Dan Kelley
}
\concept{deprecated functions}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{oceSpectrum}
\alias{oceSpectrum}
\alias{oce.spectrum}
\title{Wrapper to give normalized spectrum}
\usage{
oceSpectrum(x, ...)
}
\arguments{
\item{x}{a univariate or multivariate time series, as for \code{\link[=spectrum]{spectrum()}}.}

\item{\dots}{extra arguments passed on to \code{\link[=spectrum]{spectrum()}}.}
}
\value{
A spectrum that has values that integrate to the variance.
}
\description{
This is a wrapper around the R \code{\link[=spectrum]{spectrum()}} function, which
returns spectral values that are adjusted so that the integral of those
values equals the variance of the input \code{x}.
}
\examples{
  x <- rnorm(1e3)
  s <- spectrum(x, plot=FALSE)
  ss <- oce.spectrum(x, plot=FALSE)
  cat("variance of x=", var(x), "\n")
  cat("integral of     spectrum=", sum(s$spec)*diff(s$freq[1:2]), "\n")
  cat("integral of oce.spectrum=", sum(ss$spec)*diff(ss$freq[1:2]), "\n")
}
\seealso{
\code{\link[=spectrum]{spectrum()}}.
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.R
\name{plot,adp-method}
\alias{plot,adp-method}
\alias{plot.adp}
\title{Plot an adp Object}
\usage{
\S4method{plot}{adp}(
  x,
  which,
  j,
  col,
  breaks,
  zlim,
  titles,
  lwd = par("lwd"),
  type = "l",
  ytype = c("profile", "distance"),
  drawTimeRange = getOption("oceDrawTimeRange"),
  useSmoothScatter,
  missingColor = "gray",
  mgp = getOption("oceMgp"),
  mar = c(mgp[1] + 1.5, mgp[1] + 1.5, 1.5, 1.5),
  mai.palette = rep(0, 4),
  tformat,
  marginsAsImage = FALSE,
  cex = par("cex"),
  cex.axis = par("cex.axis"),
  cex.lab = par("cex.lab"),
  xlim,
  ylim,
  control,
  useLayout = FALSE,
  coastline = "coastlineWorld",
  span = 300,
  main = "",
  grid = FALSE,
  grid.col = "darkgray",
  grid.lty = "dotted",
  grid.lwd = 1,
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{x}{an \linkS4class{adp} object.}

\item{which}{list of desired plot types.  These are graphed in panels
running down from the top of the page.  If \code{which} is not given,
the plot will show images of the distance-time dependence of velocity
for each beam. See \dQuote{Details} for the meanings of various values of \code{which}.}

\item{j}{optional string specifying a sub-class of \code{which}. For
Nortek Aquadopp profilers, this may either be \code{"default"} (or missing)
to get the main signal, or \code{"diagnostic"} to get a diagnostic
signal. For Nortek AD2CP profiles, this may be any one of
\code{"average"} (or missing) for averaged data, \code{"burst"}
for burst data, or \code{"interleaved burst"} for interleaved burst data;
more data types are provided by that instrument, and may be added here
at some future time.}

\item{col}{optional indication of color(s) to use.  If not provided, the
default for images is \code{oce.colorsPalette(128,1)}, and for lines and
points is black.}

\item{breaks}{optional breaks for color scheme}

\item{zlim}{a range to be used as the \code{zlim} parameter to the
\code{\link[=imagep]{imagep()}} call that is used to create the image.  If omitted,
\code{zlim} is set for each panel individually, to encompass the data of the
panel and to be centred around zero.  If provided as a two-element vector,
then that is used for each panel.  If provided as a two-column matrix, then
each panel of the graph uses the corresponding row of the matrix; for
example, setting \code{zlim=rbind(c(-1,1),c(-1,1),c(-.1,.1))} might make
sense for \code{which=1:3}, so that the two horizontal velocities have one
scale, and the smaller vertical velocity has another.}

\item{titles}{optional vector of character strings to be used as labels for
the plot panels.  For images, these strings will be placed in the right hand
side of the top margin.  For timeseries, these strings are ignored.  If this
is provided, its length must equal that of \code{which}.}

\item{lwd}{if the plot is of a time-series or scattergraph format with
lines, this is used in the usual way; otherwise, e.g. for image formats,
this is ignored.}

\item{type}{if the plot is of a time-series or scattergraph format, this is
used in the usual way, e.g. \code{"l"} for lines, etc.; otherwise, as for
image formats, this is ignored.}

\item{ytype}{character string controlling the type of the y axis for images
(ignored for time series).  If \code{"distance"}, then the y axis will be
distance from the sensor head, with smaller distances nearer the bottom of
the graph.  If \code{"profile"}, then this will still be true for
upward-looking instruments, but the y axis will be flipped for
downward-looking instruments, so that in either case, the top of the graph
will represent the sample nearest the sea surface.}

\item{drawTimeRange}{boolean that applies to panels with time as the
horizontal axis, indicating whether to draw the time range in the top-left
margin of the plot.}

\item{useSmoothScatter}{boolean that indicates whether to use
\code{\link[=smoothScatter]{smoothScatter()}} in various plots, such as \code{which="uv"}.  If
not provided a default is used, with \code{\link[=smoothScatter]{smoothScatter()}} being used
if there are more than 2000 points to plot.}

\item{missingColor}{color used to indicate \code{NA} values in images (see
\code{\link[=imagep]{imagep()}}); set to \code{NULL} to avoid this indication.}

\item{mgp}{A 3-element numerical vector used with \code{\link{par}}\code{("mgp")}
to control the spacing of axis elements.  The default is tighter than the R default.}

\item{mar}{A 4-element numerical vector used with \code{\link{par}}\code{("mar")}
to control the plot margins. The default is tighter than the R default.}

\item{mai.palette}{margins, in inches, to be added to those calculated for
the palette; alter from the default only with caution}

\item{tformat}{optional argument passed to \code{\link[=oce.plot.ts]{oce.plot.ts()}}, for
plot types that call that function.  (See \code{\link[=strptime]{strptime()}} for the
format used.)}

\item{marginsAsImage}{boolean, \code{TRUE} to put a wide margin to the right
of time-series plots, even if there are no images in the \code{which} list.
(The margin is made wide if there are some images in the sequence.)}

\item{cex}{numeric character expansion factor for plot symbols; see \code{\link[=par]{par()}}.}

\item{cex.axis, cex.lab}{character expansion factors for axis numbers and axis names; see \code{\link[=par]{par()}}.}

\item{xlim}{optional 2-element list for \code{xlim}, or 2-column matrix, in
which case the rows are used, in order, for the panels of the graph.}

\item{ylim}{optional 2-element list for \code{ylim}, or 2-column matrix, in
which case the rows are used, in order, for the panels of the graph.}

\item{control}{optional list of parameters that may be used for different
plot types.  Possibilities are \code{drawBottom} (a boolean that indicates
whether to draw the bottom) and \code{bin} (a numeric giving the index of
the bin on which to act, as explained in \dQuote{Details}).}

\item{useLayout}{set to \code{FALSE} to prevent using \code{\link[=layout]{layout()}}
to set up the plot.  This is needed if the call is to be part of a sequence
set up by e.g. \code{par(mfrow)}.}

\item{coastline}{a \code{coastline} object, or a character string naming
one.  This is used only for \code{which="map"}.  See notes at
\code{\link[=plot,ctd-method]{plot,ctd-method()}} for more information on built-in coastlines.}

\item{span}{approximate span of map in km}

\item{main}{main title for plot, used just on the top panel, if there are
several panels.}

\item{grid}{if \code{TRUE}, a grid will be drawn for each panel.  (This
argument is needed, because calling \code{\link[=grid]{grid()}} after doing a
sequence of plots will not result in useful results for the individual
panels.}

\item{grid.col}{color of grid}

\item{grid.lty}{line type of grid}

\item{grid.lwd}{line width of grid}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}

\item{\dots}{optional arguments passed to plotting functions.  For example,
supplying \code{despike=TRUE} will cause time-series panels to be de-spiked
with \code{\link[=despike]{despike()}}.  Another common action is to set the color for
missing values on image plots, with the argument \code{missingColor} (see
\code{\link[=imagep]{imagep()}}).  Note that it is an error to give \code{breaks} in
\dots{}, if the formal argument \code{zlim} was also given, because they
could contradict each other.}
}
\value{
A list is silently returned, containing \code{xat} and \code{yat},
values that can be used by \code{\link[=oce.grid]{oce.grid()}} to add a grid to the plot.
}
\description{
Create a summary plot of data measured by an acoustic doppler profiler.
}
\details{
The plot may have one or more panels, with the content being controlled by
the \code{which} argument.
\itemize{
\item \code{which=1:4} (or \code{which="u1"} to \code{"u4"}) yield a
distance-time image plot of a velocity component.  If \code{x} is in
\code{beam} coordinates (signalled by
\code{metadata$oce.coordinate=="beam"}), this will be the beam velocity,
labelled \code{b[1]} etc.  If \code{x} is in xyz coordinates (sometimes
called frame coordinates, or ship coordinates), it will be the velocity
component to the right of the frame or ship (labelled \code{u} etc).
Finally, if \code{x} is in \code{"enu"} coordinates, the image will show the
the eastward component (labelled \code{east}).  If \code{x} is in
\code{"other"} coordinates, it will be component corresponding to east,
after rotation (labelled \verb{u\\'}).  Note that the coordinate is set by
\code{\link[=read.adp]{read.adp()}}, or by \code{\link[=beamToXyzAdp]{beamToXyzAdp()}},
\code{\link[=xyzToEnuAdp]{xyzToEnuAdp()}}, or \code{\link[=enuToOtherAdp]{enuToOtherAdp()}}.
\item \code{which=5:8} (or \code{which="a1"} to \code{"a4"}) yield
distance-time images of backscatter intensity of the respective beams.  (For
data derived from Teledyne-RDI instruments, this is the item called ``echo
intensity.'')
\item \code{which=9:12} (or \code{which="q1"} to \code{"q4"}) yield
distance-time images of signal quality for the respective beams.  (For RDI
data derived from instruments, this is the item called ``correlation
magnitude.'')
\item \code{which=60} or \code{which="map"} draw a map of location(s).
\item \code{which=70:73} (or \code{which="g1"} to \code{"g4"}) yield
distance-time images of percent-good for the respective beams.  (For data
derived from Teledyne-RDI instruments, which are the only instruments that
yield this item, it is called ``percent good.'')
\item \code{which=80:83} (or \code{which="vv"}, \code{which="va"},
\code{which="vq"}, and \code{which="vg"}) yield distance-time
images of the vertical beam fields for a 5 beam "SentinelV" ADCP
from Teledyne RDI.
\item \code{which="vertical"} yields a two panel distance-time
image of vertical beam velocity and amplitude.
\item \code{which=13} (or \code{which="salinity"}) yields a time-series plot
of salinity.
\item \code{which=14} (or \code{which="temperature"}) yields a time-series
plot of temperature.
\item \code{which=15} (or \code{which="pressure"}) yields a time-series plot
of pressure.
\item \code{which=16} (or \code{which="heading"}) yields a time-series plot
of instrument heading.
\item \code{which=17} (or \code{which="pitch"}) yields a time-series plot of
instrument pitch.
\item \code{which=18} (or \code{which="roll"}) yields a time-series plot of
instrument roll.
\item \code{which=19} yields a time-series plot of distance-averaged
velocity for beam 1, rightward velocity, eastward velocity, or
rotated-eastward velocity, depending on the coordinate system.
\item \code{which=20} yields a time-series of distance-averaged velocity for
beam 2, foreward velocity, northward velocity, or rotated-northward
velocity, depending on the coordinate system.
\item \code{which=21} yields a time-series of distance-averaged velocity for
beam 3, up-frame velocity, upward velocity, or rotated-upward velocity,
depending on the coordinate system.
\item \code{which=22} yields a time-series of distance-averaged velocity for
beam 4, for \code{beam} coordinates, or velocity estimate, for other
coordinates.  (This is ignored for 3-beam data.)
\item \code{which="progressiveVector"} (or \code{which=23}) yields a progressive-vector diagram in the horizontal
plane, plotted with \code{asp=1}.  Normally, the depth-averaged velocity
components are used, but if the \code{control} list contains an item named
\code{bin}, then the depth bin will be used (with an error resulting if the
bin is out of range).
\item \code{which=24} yields a time-averaged profile of the first component
of velocity (see \code{which=19} for the meaning of the component, in
various coordinate systems).
\item \code{which=25} as for 24, but the second component.
\item \code{which=26} as for 24, but the third component.
\item \code{which=27} as for 24, but the fourth component (if that makes
sense, for the given instrument).
\item \code{which=28} or \code{"uv"} yields velocity plot in the horizontal
plane, i.e. \code{u[2]} versus \code{u[1]}.  If the number of data points is small, a
scattergraph is used, but if it is large, \code{\link[=smoothScatter]{smoothScatter()}} is
used.
\item \code{which=29} or \code{"uv+ellipse"} as the \code{"uv"} case, but
with an added indication of the tidal ellipse, calculated from the eigen
vectors of the covariance matrix.
\item \code{which=30} or \code{"uv+ellipse+arrow"} as the
\code{"uv+ellipse"} case, but with an added arrow indicating the mean
current.
\item \code{which=40} or \code{"bottomRange"} for average bottom range from
all beams of the instrument.
\item \code{which=41} to \code{44} (or \code{"bottomRange1"} to
\code{"bottomRange4"}) for bottom range from beams 1 to 4.
\item \code{which=50} or \code{"bottomVelocity"} for average bottom velocity
from all beams of the instrument.
\item \code{which=51} to \code{54} (or \code{"bottomVelocity1"} to
\code{"bottomVelocity4"}) for bottom velocity from beams 1 to 4.
\item \code{which=55} (or \code{"heaving"}) for time-integrated,
depth-averaged, vertical velocity, i.e. a time series of heaving.
\item \code{which=100} (or \code{"soundSpeed"}) for a time series of sound speed.
}

In addition to the above, the following shortcuts are defined:
\itemize{
\item \code{which="velocity"} equivalent to \code{which=1:3} or \code{1:4}
(depending on the device) for velocity components.
\item \code{which="amplitude"} equivalent to \code{which=5:7}
or \code{5:8} (depending on the device) for backscatter intensity
components.
\item \code{which="quality"} equivalent to \code{which=9:11} or \code{9:12}
(depending on the device) for quality components.
\item \code{which="hydrography"} equivalent to \code{which=14:15}
for temperature and pressure.
\item \code{which="angles"} equivalent to \code{which=16:18} for
heading, pitch and roll.
}

The color scheme for image plots (\code{which} in 1:12) is provided by the
\code{col} argument, which is passed to \code{\link[=image]{image()}} to do the actual
plotting.  See \dQuote{Examples} for some comparisons.

A common quick-look plot to assess mooring movement is to use
\code{which=15:18} (pressure being included to signal the tide, and tidal
currents may dislodge a mooring or cause it to settle).

By default, \verb{plot,adp-method} uses a \code{zlim} value for the
\code{\link[=image]{image()}} that is constructed to contain all the data, but to be
symmetric about zero.  This is done on a per-panel basis, and the scale is
plotted at the top-right corner, along with the name of the variable being
plotted. You may also supply \code{zlim} as one of the \dots{} arguments,
but be aware that a reasonable limit on horizontal velocity components is
unlikely to be of much use for the vertical component.

A good first step in the analysis of measurements made from a moored device
(stored in \code{d}, say) is to do \code{plot(d, which=14:18)}.  This shows
time series of water properties and sensor orientation, which is helpful in
deciding which data to trim at the start and end of the deployment, because
they were measured on the dock or on the ship as it travelled to the mooring
site.
}
\examples{
library(oce)
data(adp)
plot(adp, which=1:3)
plot(adp, which='temperature', tformat='\%H:\%M')

}
\seealso{
Other functions that plot oce data: 
\code{\link{download.amsr}()},
\code{\link{plot,adv-method}},
\code{\link{plot,amsr-method}},
\code{\link{plot,argo-method}},
\code{\link{plot,bremen-method}},
\code{\link{plot,cm-method}},
\code{\link{plot,coastline-method}},
\code{\link{plot,ctd-method}},
\code{\link{plot,gps-method}},
\code{\link{plot,ladp-method}},
\code{\link{plot,landsat-method}},
\code{\link{plot,lisst-method}},
\code{\link{plot,lobo-method}},
\code{\link{plot,met-method}},
\code{\link{plot,odf-method}},
\code{\link{plot,rsk-method}},
\code{\link{plot,satellite-method}},
\code{\link{plot,sealevel-method}},
\code{\link{plot,section-method}},
\code{\link{plot,tidem-method}},
\code{\link{plot,topo-method}},
\code{\link{plot,windrose-method}},
\code{\link{plot,xbt-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{tidem-class}}

Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley
}
\concept{functions that plot oce data}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.rdi.R
\name{read.adp.rdi}
\alias{read.adp.rdi}
\title{Read a Teledyne/RDI ADP File}
\usage{
read.adp.rdi(
  file,
  from,
  to,
  by,
  tz = getOption("oceTz"),
  longitude = NA,
  latitude = NA,
  type = c("workhorse"),
  monitor = FALSE,
  despike = FALSE,
  processingLog,
  testing = FALSE,
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{file}{a connection or a character string giving the name of the file
to load.  (For \code{read.adp.sontek.serial}, this is generally a list of
files, which will be concatenated.)}

\item{from}{indication of the first profile to read.  This can be an
integer, the sequence number of the first profile to read, or a POSIXt time
before which profiles should be skipped, or a character string that converts
to a POSIXt time (assuming UTC timezone).  See \dQuote{Examples}, and make
careful note of the use of the \code{tz} argument. If \code{from} is not
supplied, it defaults to 1.}

\item{to}{an optional indication of the last profile to read, in a
format as described for \code{from}.  As a special case, \code{to=0} means
to read the file to the end.  If \code{to} is not supplied, then it defaults
to 0.}

\item{by}{an optional indication of the stride length to use while walking through
the file.  If this is an integer, then \code{by-1} profiles are skipped
between each pair of profiles that is read, e.g. the default \code{by=1}
means to read all the data.  (For RDI files \emph{only}, there are some
extra features to avoid running out of memory; see \dQuote{Memory considerations}.)}

\item{tz}{character string indicating time zone to be assumed in the data.}

\item{longitude}{optional signed number indicating the longitude in degrees
East.}

\item{latitude}{optional signed number indicating the latitude in degrees
North.}

\item{type}{character string indicating the type of instrument.}

\item{monitor}{boolean value indicating whether to indicate the progress
of reading the file, by using \code{\link[=txtProgressBar]{txtProgressBar()}} or otherwise.  The value
of \code{monitor} is changed to \code{FALSE} automatically, for non-interactive
sessions.}

\item{despike}{if \code{TRUE}, \code{\link[=despike]{despike()}} will be used to clean
anomalous spikes in heading, etc.}

\item{processingLog}{if provided, the action item to be stored in the log.
(Typically only provided for internal calls; the default that it provides is
better for normal calls by a user.)}

\item{testing}{logical value (IGNORED).}

\item{debug}{a flag that turns on debugging.  Set to 1 to get a moderate
amount of debugging information, or to 2 to get more.}

\item{\dots}{optional additional arguments that some (but not all)
\verb{read.adp.*()} functions pass to lower-level functions.}
}
\value{
An \linkS4class{adp} object.
The contents of that object make sense for the particular instrument
type under study, e.g. if the data file contains
NMEA strings, then navigational data will be stored in an item
called \code{nmea} in the \code{data} slot).
}
\description{
Read a Teledyne/RDI ADCP file (called 'adp' in oce).
}
\details{
As of 2016-09-25, this function has provisional functionality to
read data from the new "SentinelV" series ADCP -- essentially a
combination of a 4 beam workhorse with an additional vertical
centre beam.

If a heading bias had been set with the \code{EB} command during the setup
for the deployment, then a heading bias will have been stored in the file's
header.  This value is stored in the object's metadata as
\code{metadata$heading.bias}.  \strong{Importantly}, this value is
subtracted from the headings stored in the file, and the result of this
subtraction is stored in the objects heading value (in \code{data$heading}).
It should be noted that \code{read.adp.rdi()} was tested for firmware
version 16.30.  For other versions, there may be problems.  For example, the
serial number is not recognized properly for version 16.28.

In Teledyne/RDI ADP data files, velocities are coded to signed 2-byte integers, with a
scale factor being used to convert to velocity in metres per second.  These
two facts control the maximum recordable velocity and the velocity
resolution, values that may be retrieved for an ADP object name \code{d}
with \code{d[["velocityMaximum"]]} and \code{d[["velocityResolution"]]}.
}
\section{Handling of old file formats}{

\enumerate{
\item Early PD0 file formats stored the year of sampling with a different
base year than that used in modern files.  To accommodate this,
\code{read.adp.rdi} examines the inferred year, and if it is greater than
2050, then 100 years are subtracted from the time. This offset was
inferred by tests with sample files, but \emph{not} from RDI documentation,
so it is somewhat risky.  If the authors can find RDI documentation that
indicates the condition in which this century offset is required, then
a change will be made to the code.  Even if not, the method should
not cause problems for a long time.
}
}

\section{Names of items in data slot}{


The names of items in the \code{data} slot are below. Not all items are present
for ll file varieties; use e.g. \code{names(d[["data"]])} to determine the
names used in an object named \code{d}. In this list, items are either
a vector (with one sample per time of measurement), a
\link{matrix} with first index for time and second for bin number,
or an \link{array} with first index for time, second for bin number,
and third for beam number. Items are of vector type, unless
otherwise indicated.

\tabular{rr}{
\strong{Item} \tab \strong{Meaning} \cr
\code{a}                                    \tab signal amplitude array (units?)\cr
\code{ambientTemp}                          \tab ambient temperature (degC)\cr
\code{attitude}                             \tab attitude (deg)\cr
\code{attitudeTemp}                         \tab (FIXME add a description here)\cr
\code{avgMagnitudeVelocityEast}             \tab (FIXME add a description here)\cr
\code{avgMagnitudeVelocityNorth}            \tab (FIXME add a description here)\cr
\code{avgSpeed}                             \tab (FIXME add a description here)\cr
\code{avgTrackMagnetic}                     \tab (FIXME add a description here)\cr
\code{avgTrackTrue}                         \tab (FIXME add a description here)\cr
\code{avgTrueVelocityEast}                  \tab (FIXME add a description here)\cr
\code{avgTrueVelocityNorth}                 \tab (FIXME add a description here)\cr
\code{br}                                   \tab bottom range matrix (m)\cr
\code{bv}                                   \tab bottom velocity matrix (m/s)\cr
\code{contaminationSensor}                  \tab (FIXME add a description here)\cr
\code{depth}                                \tab depth (m)\cr
\code{directionMadeGood}                    \tab (FIXME add a description here)\cr
\code{distance}                             \tab (FIXME add a description here)\cr
\code{firstLatitude}                        \tab latitude at start of profile (deg)\cr
\code{firstLongitude}                       \tab longitude at start of profile (deg)\cr
\code{firstTime}                            \tab (FIXME add a description here)\cr
\code{g}                                    \tab data goodness matrix (units?)\cr
\code{heading}                              \tab instrument heading (degrees)\cr
\code{headingStd}                           \tab instrument heading std-dev (deg)\cr
\code{lastLatitude}                         \tab latitude at end of profile (deg)\cr
\code{lastLongitude}                        \tab longitude at end of profile (deg)\cr
\code{lastTime}                             \tab (FIXME add a description here)\cr
\code{numberOfHeadingSamplesAveraged}       \tab (FIXME add a description here)\cr
\code{numberOfMagneticTrackSamplesAveraged} \tab (FIXME add a description here)\cr
\code{numberOfPitchRollSamplesAveraged}     \tab (FIXME add a description here)\cr
\code{numberOfSpeedSamplesAveraged}         \tab (FIXME add a description here)\cr
\code{numberOfTrueTrackSamplesAveraged}     \tab (FIXME add a description here)\cr
\code{pitch}                                \tab instrument pitch (deg)\cr
\code{pitchStd}                             \tab instrument pitch std-dev (deg)\cr
\code{pressure}                             \tab pressure (dbar)\cr
\code{pressureMinus}                        \tab (FIXME add a description here)\cr
\code{pressurePlus}                         \tab (FIXME add a description here)\cr
\code{pressureStd}                          \tab pressure std-dev (dbar)\cr
\code{primaryFlags}                         \tab (FIXME add a description here)\cr
\code{q}                                    \tab data quality array\cr
\code{roll}                                 \tab instrument roll (deg)\cr
\code{rollStd}                              \tab instrument roll std-dev (deg)\cr
\code{salinity}                             \tab salinity\cr
\code{shipHeading}                          \tab ship heading (deg)\cr
\code{shipPitch}                            \tab ship pitch (deg)\cr
\code{shipRoll}                             \tab ship roll (deg)\cr
\code{soundSpeed}                           \tab sound speed (m/s)\cr
\code{speedMadeGood}                        \tab speed over ground (?) (m/s)\cr
\code{speedMadeGoodEast}                    \tab (FIXME add a description here)\cr
\code{speedMadeGoodNorth}                   \tab (FIXME add a description here)\cr
\code{temperature}                          \tab temperature (degC)\cr
\code{time}                                 \tab profile time (POSIXct)\cr
\code{v}                                    \tab velocity array (m/s)\cr
\code{xmitCurrent}                          \tab transmit current (unit?)\cr
\code{xmitVoltage}                          \tab transmit voltage\cr
}
}

\section{Memory considerations}{


For \code{RDI} files only, and only in the case where \code{by} is not specified,
an attempt is made to avoid running out of memory by skipping some profiles
in large input files. This only applies if \code{from} and \code{to} are both
integers; if they are times, none of the rest of this section applies.

A key issue is that RDI files store velocities in 2-byte values, which is
not a format that R supports. These velocities become 8-byte (numeric) values
in R. Thus, the R object created by \code{read.adp.rdi} will require more memory
than that of the data file. A scale factor can be estimated by ignoring
vector quantities (e.g. time, which has just one value per profile) and concentrating on matrix properties
such as velocity, backscatter, and correlation. These three elements have equal dimensions.
Thus, each 4-byte slide in the data file (2 bytes + 1 byte + 1 byte)
corresponds to 10 bytes in the object (8 bytes + 1 byte + 1 byte).
Rounding up the resultant 10/4 to 3 for safety, we conclude that any limit on the
size of the R object corresponds to a 3X smaller limit on file size.

Various things can limit the size of objects in R, but a strong upper limit
is set by the space the operating system provides to R. The least-performant machines
in typical use appear to be Microsoft-Windows systems, which limit R objects to
about 2e6 bytes (see \code{?Memory-limits}).  Since R routinely duplicates objects for certain tasks
(e.g. for call-by-value in function evaluation), \code{read.adp.rdi} uses a safety
factor in its calculation of when to auto-decimate a file. This factor is set to 3,
based partly on the developers' experience with datasets in their possession.
Multiplied by the previously stated safety factor of 3,
this suggests that the 2 GB limit on R objects corresponds to approximately a
222 MB limit on file size. In the present version of \code{read.adp.rdi}, this
value is lowered to 200 MB for simplicity. Larger files are considered to be "big",
and are decimated unless the user supplies a value for the \code{by} argument.

The decimation procedure has two cases.
\enumerate{
\item If \code{from=1} and \code{to=0} (or if neither \code{from} or \code{to} is given), then the
intention is to process the full span of the data.  If the input file is
under 200 MB, then \code{by} defaults to 1, so that all profiles are read.
For larger files, \code{by} is set to the \code{\link[=ceiling]{ceiling()}} of the
ratio of input file size to 200 MB.
\item If \code{from} exceeds 1, and/or \code{to} is nonzero, then
the intention is to process only an interior subset of the file. In this
case, \code{by} is calculated as the \code{\link[=ceiling]{ceiling()}} of
the ratio of \code{bbp*(1+to-from)} to 200 MB, where \code{bbp} is the number
of file bytes per profile. Of course, \code{by} is set to 1, if this
ratio is less than 1.
}

If the result of these calculations is that \code{by} exceeds 1, then
messages are printed to alert the user that the file will be decimated,
and also \code{monitor} is set to \code{TRUE}, so that a textual progress bar
is shown (if the session is interactive).
}

\section{Development Notes}{

An important part of the work of this function is to recognize what
will be called "data chunks" by two-byte ID sequences. This function is
developed in a practical way, with emphasis being focussed on
data files in the possession of the developers. Since Teledyne-RDI tends
to introduce new ID codes with new instruments, that means that
\code{read.adp.rdi} may not work on recently-developed instruments.

The following two-byte ID codes are recognized by \code{read.adp.rdi}
at this time (with bytes listed in natural order, LSB byte before
MSB). Items preceded by an asterisk are recognized, but not handled,
and so produce a warning.
\tabular{rrrr}{
\tab \strong{Byte 1} \tab \strong{Byte 2} \tab \strong{Meaning}\cr
\tab 0x00 \tab 0x01 \tab velocity\cr
\tab 0x00 \tab 0x01 \tab velocity\cr
\tab 0x00 \tab 0x02 \tab correlation\cr
\tab 0x00 \tab 0x03 \tab echo intensity\cr
\tab 0x00 \tab 0x04 \tab percent good\cr
\tab 0x00 \tab 0x06 \tab bottom track\cr
\tab 0x00 \tab 0x0a \tab Sentinel vertical beam velocity\cr
\tab 0x00 \tab 0x0b \tab Sentinel vertical beam correlation\cr
\tab 0x00 \tab 0x0c \tab Sentinel vertical beam amplitude\cr
\tab 0x00 \tab 0x0d \tab Sentinel vertical beam percent good\cr
\tab 0x00 \tab 0x20 \tab VMDASS\cr
\tab 0x00 \tab 0x30 \tab Binary Fixed Attitude header\cr
\tab 0x00 \tab 0x32 \tab Sentinel transformation matrix\cr
\tab 0x00 \tab 0x0a \tab Sentinel data\cr
\tab 0x00 \tab 0x0b \tab Sentinel correlation\cr
\tab 0x00 \tab 0x0c \tab Sentinel amplitude\cr
\tab 0x00 \tab 0x0d \tab Sentinel percent good\cr
\tab 0x01 \tab 0x0f \tab ?? something to do with V series and 4-beam\cr
}

Lacking a comprehensive Teledyne-RDI listing of ID codes,
the authors have cobbled together a listing from documents to which they
have access, as follows.
\itemize{
\item Table 33 of reference 3 lists codes as follows:
\tabular{rrr}{
\strong{Standard ID}  \tab  \strong{Standard plus 1D} \tab \strong{DESCRIPTION}\cr
MSB    LSB  \tab        MSB    LSB\tab\cr
---    ---  \tab        ---    ---\tab\cr
7F     7F  \tab         7F     7F \tab   Header\cr
00     00  \tab         00     01 \tab   Fixed Leader\cr
00     80  \tab         00     81 \tab   Variable Leader\cr
01     00  \tab         01     01 \tab   Velocity Profile Data\cr
02     00  \tab         02     01 \tab   Correlation Profile Data\cr
03     00  \tab         03     01 \tab   Echo Intensity Profile Data\cr
04     00  \tab         04     01 \tab   Percent Good Profile Data\cr
05     00  \tab         05     01 \tab   Status Profile Data\cr
06     00  \tab         06     01 \tab   Bottom Track Data\cr
20     00  \tab         20     00 \tab   Navigation\cr
30     00  \tab         30     00 \tab   Binary Fixed Attitude\cr
30  40-F0  \tab         30  40-F0 \tab   Binary Variable Attitude\cr
}
\item Table 6 on p90 of reference 4 lists "Fixed Leader Navigation" ID
codes (none of which are handled by \code{read.adp.rdi} yet)
as follows (the format is reproduced literally; note that
e.g. 0x2100 is 0x00,0x21 in the oce notation):
\tabular{rr}{
\strong{ID} \tab \strong{Description}\cr
0x2100 \tab  $xxDBT\cr
0x2101 \tab  $xxGGA\cr
0x2102 \tab  $xxVTG\cr
0x2103 \tab  $xxGSA\cr
0x2104 \tab  $xxHDT, $xxHGD or $PRDID\cr
}
and following pages in that manual reveal the following meanings
\tabular{rr}{
\strong{Symbol} \tab \strong{Meaning}\cr
\code{DBT}   \tab depth below transducer\cr
\code{GGA}   \tab global positioning system\cr
\code{VTA}   \tab track made good and ground speed\cr
\code{GSA}   \tab GPS DOP and active satellites\cr
\code{HDT}   \tab heading, true\cr
\code{HDG}   \tab heading, deviation, and variation\cr
\code{PRDID} \tab heading, pitch and roll\cr
}
}
}

\section{Error recovery}{


Files can sometimes be corrupted, and \code{read.adp.rdi} has ways to handle two types
of error that have been noticed in files supplied by users.
\enumerate{
\item There are two bytes within each ensemble that indicate the number of bytes to check within
that ensemble, to get the checksum. Sometimes, those two bytes can be erroneous, so that
the wrong number of bytes are checked, leading to a failed checksum. As a preventative
measure, \code{read.adp.rdi} checks the stated ensemble length, whenever it detects a
failed checksum. If that length agrees with the length of the most recent ensemble that
had a good checksum, then the ensemble is declared as faulty and is ignored. However,
if the length differs from that of the most recent accepted ensemble, then \code{read.adp.rdi}
goes back to just after the start of the ensemble, and searches forward for the next two-byte
pair, namely \verb{0x7f 0x7f}, that designates the start of an ensemble.  Distinct notifications
are given about these two cases, and they give the byte numbers in the original file, as a way
to help analysts who want to look at the data stream with other tools.
\item At the end of an ensemble, the next two characters ought to be \verb{0x7f 0x7f}, and if they
are not, then the next ensemble is faulty. If this error occurs, \code{read.adp.rdi} attempts
to recover by searching forward to the next instance of this two-byte pair, discarding any
information that is present in the mangled ensemble.
}

In each of these cases, warnings are printed about ensembles that seem problematic.
Advanced users who want to diagnose the problem further might find it helpful to
examine the original data file using other tools. To this end, \code{read.adp.rdi}
inserts an element named \code{ensembleInFile} into the \code{metadata} slot.
This gives the starting byte number of each inferred ensemble within the original data
file.  For example, if \code{d} is an object read with \code{read.adp.rdi}, then using\preformatted{plot(d[["time"]][-1], diff(d[["ensembleInFile"]]))
}

can be a good way to narrow in on problems.
}

\examples{
adp <- read.adp.rdi(system.file("extdata", "adp_rdi.000", package="oce"))
summary(adp)

}
\references{
\enumerate{
\item Teledyne-RDI, 2007. \emph{WorkHorse commands and output data
format.} P/N 957-6156-00 (November 2007).  (Section 5.3 h details the binary
format, e.g. the file should start with the byte \code{0x7f} repeated twice,
and each profile starts with the bytes \code{0x80}, followed by \code{0x00},
followed by the sequence number of the profile, represented as a
little-endian two-byte short integer.  \code{read.adp.rdi} uses these
sequences to interpret data files.)
\item Teledyne RD Instruments, 2015. \emph{V Series monitor, sentinel Output Data Format.}
P/N 95D-6022-00 (May 2015). \code{SV_ODF_May15.pdf}
\item Teledyne RD Instruments, 2014. \emph{Ocean Surveyor / Ocean Observer Technical Manual.}
P/N 95A-6012-00 (April 2014). \code{OS_TM_Apr14.pdf}
\item Teledyne RD Instruments, 2001. \emph{WinRiver User's Guide International Version.}
P/N 957-6171-00 (June 2001) \verb{WinRiver User Guide International Version.pdf.pdf}
}
}
\seealso{
Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley and Clark Richards
}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/accessors.R
\name{oceSetMetadata}
\alias{oceSetMetadata}
\title{Set Something in an oce metadata Slot}
\usage{
oceSetMetadata(object, name, value, note = "")
}
\arguments{
\item{object}{an \linkS4class{oce} object.}

\item{name}{String indicating the name of the \code{metadata} item to be set.}

\item{value}{Value for the item.}

\item{note}{Either empty (the default), a character string, or \code{NULL},
to control additions made to the processing log of the return value. If
\code{note=""} then an entry is created based on deparsing the function call.
If \code{note} is a non-empty string, then that string gets added added
to the processing log. Finally, if \code{note=NULL}, then nothing is
added to the processing log.  This last form is useful in cases where
\code{oceSetData} is to be called many times in succession, resulting
in an overly verbose processing log; in which case, it might helpful
to use \link{processingLog<-} to add a summary entry to the object's
processing log.}
}
\value{
An \linkS4class{oce} object, the \code{metadata} slot of which has
been altered either by adding a new item or modifying an existing
item.
}
\description{
Create a copy of an object in which some element of its
\code{metadata} slot has been altered, or added.
}
\examples{
# Add an estimate of MLD (mixed layer depth) to a ctd object
library(oce)
data(ctd)
ctdWithMLD <- oceSetMetadata(ctd, "MLD", 3)
ctdWithMLD[["MLD"]] # 3

}
\seealso{
Other things related to the metadata slot: 
\code{\link{oceDeleteMetadata}()},
\code{\link{oceGetMetadata}()},
\code{\link{oceRenameMetadata}()}
}
\author{
Dan Kelley
}
\concept{things related to the metadata slot}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/xbt.R
\name{[[<-,xbt-method}
\alias{[[<-,xbt-method}
\title{Replace Parts of an xbt Object}
\usage{
\S4method{[[}{xbt}(x, i, j, ...) <- value
}
\arguments{
\item{x}{an \linkS4class{xbt} object.}

\item{i}{character value naming the item to replace.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{optional additional information (ignored).}

\item{value}{The value to be placed into \code{x}, somewhere.}
}
\description{
The \code{[[<-} method works for all \linkS4class{oce} objects.
The purpose, as with the related extraction method, \code{[[},
is to insulate users from the internal details of \linkS4class{oce}
objects, by looking for items within the various storage
slots of the object. Items not actually stored can also be
replaced, including units and data-quality
flags.
}
\details{
As with \code{[[} method, the procedure works in steps.

First, the \code{metadata} slot of \code{x} is checked to
see whether it contains something named with \code{i}.
If so, then the named item is replaced with \code{value}.

Otherwise, if the string value of \code{i} ends in \code{Unit}, then the
characters preceding that are taken as the name of a variable, and
the \code{metadata} slot of \code{x} is updated to store that unit, e.g.\preformatted{x[["temperatureUnits"]] <- list(unit=expression(degree*F),scale="")
}

Similarly, if \code{i} ends in \code{Flag}, then quality-control
flags are set up as defined by \code{result}, e.g.\preformatted{o[["temperatureFlags"]] <- c(2,4,2,2)
}

Otherwise, \code{\link[=pmatch]{pmatch()}} is used for a partial-string match with
the names of the items that are in the \code{data} slot of \code{x}.
The first item found (if any) is then updated to hold the value \code{result}.

If none of these conditions is met, a warning is issued.
}
\seealso{
Other functions that replace parts of oce objects: 
\code{\link{[[<-,adp-method}},
\code{\link{[[<-,amsr-method}},
\code{\link{[[<-,argo-method}},
\code{\link{[[<-,bremen-method}},
\code{\link{[[<-,cm-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{[[<-,echosounder-method}},
\code{\link{[[<-,g1sst-method}},
\code{\link{[[<-,gps-method}},
\code{\link{[[<-,ladp-method}},
\code{\link{[[<-,landsat-method}},
\code{\link{[[<-,lisst-method}},
\code{\link{[[<-,lobo-method}},
\code{\link{[[<-,met-method}},
\code{\link{[[<-,oce-method}},
\code{\link{[[<-,odf-method}},
\code{\link{[[<-,rsk-method}},
\code{\link{[[<-,sealevel-method}},
\code{\link{[[<-,section-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{[[<-,topo-method}},
\code{\link{[[<-,windrose-method}}

Other things related to xbt data: 
\code{\link{[[,xbt-method}},
\code{\link{as.xbt}()},
\code{\link{plot,xbt-method}},
\code{\link{read.xbt.noaa1}()},
\code{\link{read.xbt}()},
\code{\link{subset,xbt-method}},
\code{\link{summary,xbt-method}},
\code{\link{xbt-class}},
\code{\link{xbt.edf}},
\code{\link{xbt}}
}
\concept{functions that replace parts of oce objects}
\concept{things related to xbt data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/lisst.R
\name{read.lisst}
\alias{read.lisst}
\title{Read a LISST File}
\usage{
read.lisst(file, year = 0, tz = "UTC", longitude = NA, latitude = NA)
}
\arguments{
\item{file}{a connection or a character string giving the name of the file
to load.}

\item{year}{year in which the measurement of the series was made.}

\item{tz}{time zone.}

\item{longitude}{longitude of observation (stored in metadata)}

\item{latitude}{latitude of observation (stored in metadata)}
}
\value{
x A \linkS4class{lisst} object.
}
\description{
Read a LISST data file.
The file should contain 42 columns, with no header.  If there are fewer than
42 columns, an error results.  If there are more, only the first 42 are
used.  Note that \code{\link[=read.oce]{read.oce()}} can recognize LISST files by their
having a name ending in \code{".asc"} and by having 42 elements on the first
line.  Even so, it is preferred to use the present function, because it
gives the opportunity to specify the year and timezone, so that times can be
calculated properly.
}
\seealso{
Other things related to lisst data: 
\code{\link{[[,lisst-method}},
\code{\link{[[<-,lisst-method}},
\code{\link{as.lisst}()},
\code{\link{lisst-class}},
\code{\link{plot,lisst-method}},
\code{\link{summary,lisst-method}}
}
\author{
Dan Kelley
}
\concept{things related to lisst data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/echosounder.R
\name{read.echosounder}
\alias{read.echosounder}
\title{Read an Echosounder File}
\usage{
read.echosounder(
  file,
  channel = 1,
  soundSpeed,
  tz = getOption("oceTz"),
  debug = getOption("oceDebug"),
  processingLog
)
}
\arguments{
\item{file}{a connection or a character string giving the name of the file
to load.}

\item{channel}{sequence number of channel to extract, for multi-channel
files.}

\item{soundSpeed}{sound speed, in m/s. If not provided, this is calculated
using \code{\link{swSoundSpeed}}\verb{(35,15,30,eos="unesco")}.  (In theory,
it could be calculated using the temperature and salinity that are stored
in the data file, but these will just be nominal values, anyway.}

\item{tz}{character string indicating time zone to be assumed in the data.}

\item{debug}{a flag that turns on debugging.  Set to 1 to get a moderate
amount of debugging information, or to 2 to get more.}

\item{processingLog}{if provided, the action item to be stored in the log,
typically only provided for internal calls.}
}
\value{
An \linkS4class{echosounder} object.
}
\description{
Reads a biosonics echosounder file.  This function was written for and
tested with single-beam, dual-beam, and split-beam Biosonics files of type
V3, and may not work properly with other file formats.
}
\section{Bugs}{
 Only the amplitude information (in counts) is determined.  A
future version of this function may provide conversion to dB, etc.  The
handling of dual-beam and split-beam files is limited.  In the dual-beam
cse, only the wide beam signal is processed (I think ... it could be the
narrow beam, actually, given the confusing endian tricks being played).  In
the split-beam case, only amplitude is read, with the x-axis and y-axis
angle data being ignored.
}

\references{
Various echosounder instruments provided by BioSonics are
described at the company website, \url{https://www.biosonicsinc.com/}.  The
document listed as reference 1 below was provided to the author of this function in
November 2011, which suggests that the data format was not changed since
July 2010.
\enumerate{
\item Biosonics, 2010.  DT4 Data File Format Specification.  BioSonics
Advanced Digital Hydroacoustics. July, 2010.  SOFTWARE AND ENGINEERING
LIBRARY REPORT BS&E-2004-07-0009-2.0.
}
}
\seealso{
The documentation for \linkS4class{echosounder} explains the
structure of \code{ctd} objects, and also outlines the other functions
dealing with them.

Other things related to echosounder data: 
\code{\link{[[,echosounder-method}},
\code{\link{[[<-,echosounder-method}},
\code{\link{as.echosounder}()},
\code{\link{echosounder-class}},
\code{\link{echosounder}},
\code{\link{findBottom}()},
\code{\link{plot,echosounder-method}},
\code{\link{subset,echosounder-method}},
\code{\link{summary,echosounder-method}}
}
\author{
Dan Kelley, with help from Clark Richards
}
\concept{things related to echosounder data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{fillGap}
\alias{fillGap}
\title{Fill a gap in an oce object}
\usage{
fillGap(x, method = c("linear"), rule = 1)
}
\arguments{
\item{x}{an \linkS4class{oce} object.}

\item{method}{to use; see \dQuote{Details}.}

\item{rule}{integer controlling behaviour at start and end of \code{x}.  If
\code{rule=1}, \code{NA} values at the ends are left in the return value.
If \code{rule=2}, they are replaced with the nearest non-NA point.}
}
\value{
A new \code{oce} object, with gaps removed.
}
\description{
Sequences of \code{NA} values, are filled by linear interpolation between
the non-\code{NA} values that bound the gap.
}
\section{Bugs}{

\enumerate{
\item Eventually, this will be expanded to work
with any \code{oce} object.  But, for now, it only works for vectors that
can be coerced to numeric.
\item If the first or last point is \code{NA}, then \code{x} is returned unaltered.
\item Only method \code{linear} is permitted now.
}
}

\examples{
library(oce)
# Integers
x <- c(1:2, NA, NA, 5:6)
y <- fillGap(x)
print(data.frame(x,y))
# Floats
x <- x + 0.1
y <- fillGap(x)
print(data.frame(x,y))
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/tides.R
\name{[[<-,tidem-method}
\alias{[[<-,tidem-method}
\title{Replace Parts of a Tidem Object}
\usage{
\S4method{[[}{tidem}(x, i, j, ...) <- value
}
\arguments{
\item{x}{a \linkS4class{tidem} object.}

\item{i}{character value naming the item to replace.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{optional additional information (ignored).}

\item{value}{The value to be placed into \code{x}, somewhere.}
}
\description{
The \code{[[<-} method works for all \linkS4class{oce} objects.
The purpose, as with the related extraction method, \code{[[},
is to insulate users from the internal details of \linkS4class{oce}
objects, by looking for items within the various storage
slots of the object. Items not actually stored can also be
replaced, including units and data-quality
flags.
}
\details{
As with \code{[[} method, the procedure works in steps.

First, the \code{metadata} slot of \code{x} is checked to
see whether it contains something named with \code{i}.
If so, then the named item is replaced with \code{value}.

Otherwise, if the string value of \code{i} ends in \code{Unit}, then the
characters preceding that are taken as the name of a variable, and
the \code{metadata} slot of \code{x} is updated to store that unit, e.g.\preformatted{x[["temperatureUnits"]] <- list(unit=expression(degree*F),scale="")
}

Similarly, if \code{i} ends in \code{Flag}, then quality-control
flags are set up as defined by \code{result}, e.g.\preformatted{o[["temperatureFlags"]] <- c(2,4,2,2)
}

Otherwise, \code{\link[=pmatch]{pmatch()}} is used for a partial-string match with
the names of the items that are in the \code{data} slot of \code{x}.
The first item found (if any) is then updated to hold the value \code{result}.

If none of these conditions is met, a warning is issued.
}
\seealso{
Other functions that replace parts of oce objects: 
\code{\link{[[<-,adp-method}},
\code{\link{[[<-,amsr-method}},
\code{\link{[[<-,argo-method}},
\code{\link{[[<-,bremen-method}},
\code{\link{[[<-,cm-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{[[<-,echosounder-method}},
\code{\link{[[<-,g1sst-method}},
\code{\link{[[<-,gps-method}},
\code{\link{[[<-,ladp-method}},
\code{\link{[[<-,landsat-method}},
\code{\link{[[<-,lisst-method}},
\code{\link{[[<-,lobo-method}},
\code{\link{[[<-,met-method}},
\code{\link{[[<-,oce-method}},
\code{\link{[[<-,odf-method}},
\code{\link{[[<-,rsk-method}},
\code{\link{[[<-,sealevel-method}},
\code{\link{[[<-,section-method}},
\code{\link{[[<-,topo-method}},
\code{\link{[[<-,windrose-method}},
\code{\link{[[<-,xbt-method}}

Other things related to tides: 
\code{\link{[[,tidem-method}},
\code{\link{as.tidem}()},
\code{\link{plot,tidem-method}},
\code{\link{predict.tidem}()},
\code{\link{summary,tidem-method}},
\code{\link{tidalCurrent}},
\code{\link{tidedata}},
\code{\link{tidem-class}},
\code{\link{tidemAstron}()},
\code{\link{tidemVuf}()},
\code{\link{tidem}},
\code{\link{webtide}()}
}
\concept{functions that replace parts of oce objects}
\concept{things related to tides}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/map.R
\name{utm2lonlat}
\alias{utm2lonlat}
\title{Convert UTM to Longitude and Latitude}
\usage{
utm2lonlat(easting, northing, zone = 1, hemisphere = "N", km = FALSE)
}
\arguments{
\item{easting}{easting coordinate (in km or m, depending on value of
\code{km}).  Alternatively, a list containing items named \code{easting},
\code{northing}, and \code{zone}, in which case these are taken from the
list and the arguments named \code{northing}, \code{zone} and are ignored.}

\item{northing}{northing coordinate (in km or m, depending on value of
\code{km}).}

\item{zone}{UTM zone}

\item{hemisphere}{indication of hemisphere; \code{"N"} for North, anything
else for South.}

\item{km}{logical value indicating whether \code{easting} and
\code{northing} are in kilometers or meters.}
}
\value{
A list containing \code{longitude} and \code{latitude}.
}
\description{
Convert UTM to Longitude and Latitude
}
\examples{
library(oce)
## Cape Split, in the Minas Basin of the Bay of Fundy
utm2lonlat(852863, 5029997, 19)

}
\references{
\url{https://en.wikipedia.org/wiki/Universal_Transverse_Mercator_coordinate_system},
downloaded May 31, 2014.
}
\seealso{
\code{\link[=lonlat2utm]{lonlat2utm()}} does the inverse operation.  For general
projections and their inverses, use \code{\link[=lonlat2map]{lonlat2map()}} and
\code{\link[=map2lonlat]{map2lonlat()}}.

Other functions related to maps: 
\code{\link{formatPosition}()},
\code{\link{lonlat2map}()},
\code{\link{lonlat2utm}()},
\code{\link{map2lonlat}()},
\code{\link{mapArrows}()},
\code{\link{mapAxis}()},
\code{\link{mapContour}()},
\code{\link{mapCoordinateSystem}()},
\code{\link{mapDirectionField}()},
\code{\link{mapGrid}()},
\code{\link{mapImage}()},
\code{\link{mapLines}()},
\code{\link{mapLocator}()},
\code{\link{mapLongitudeLatitudeXY}()},
\code{\link{mapPlot}()},
\code{\link{mapPoints}()},
\code{\link{mapPolygon}()},
\code{\link{mapScalebar}()},
\code{\link{mapText}()},
\code{\link{mapTissot}()},
\code{\link{oceCRS}()},
\code{\link{shiftLongitude}()},
\code{\link{usrLonLat}()}
}
\author{
Dan Kelley
}
\concept{functions related to maps}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adv.nortek.R
\name{read.adv.nortek}
\alias{read.adv.nortek}
\title{Read an ADV data file}
\usage{
read.adv.nortek(
  file,
  from = 1,
  to,
  by = 1,
  tz = getOption("oceTz"),
  header = TRUE,
  longitude = NA,
  latitude = NA,
  type = c("vector", "aquadopp"),
  haveAnalog1 = FALSE,
  haveAnalog2 = FALSE,
  debug = getOption("oceDebug"),
  monitor = FALSE,
  processingLog = NULL
)
}
\arguments{
\item{file}{a connection or a character string giving the name of the file
to load.  It is also possible to give \code{file} as a vector of filenames,
to handle the case of data split into files by a data logger.  In the
multi-file case, \code{header} must be \code{FALSE}, \code{start} must be a
vector of times, and \code{deltat} must be provided.}

\item{from}{index number of the first profile to be read, or the time of
that profile, as created with \code{\link[=as.POSIXct]{as.POSIXct()}} (hint: use
\code{tz="UTC"}).  This argument is ignored if \code{header==FALSE}. See
\dQuote{Examples}.}

\item{to}{indication of the last profile to read, in a format matching that
of \code{from}.  This is ignored if \code{header==FALSE}.}

\item{by}{an indication of the stride length to use while walking through
the file.  This is ignored if \code{header==FALSE}.  Otherwise, if this is
an integer, then \code{by-1} profiles are skipped between each pair of
profiles that is read. This may not make much sense, if the data are not
equi-spaced in time.  If \code{by} is a string representing a time interval,
in colon-separated format, then this interval is divided by the sampling
interval, to get the stride length. \emph{BUG:} \code{by} only partially
works; see \dQuote{Bugs}.}

\item{tz}{character string indicating time zone to be assumed in the data.}

\item{header}{A logical value indicating whether the file starts with a header.
(This will not be the case for files that are created by data loggers that
chop the raw data up into a series of sub-files, e.g. once per hour.)}

\item{longitude}{optional signed number indicating the longitude in degrees
East.}

\item{latitude}{optional signed number indicating the latitude in degrees
North.}

\item{type}{A string indicating which type of Nortek device produced the data file, \code{vector}
or \code{aquadopp}.}

\item{haveAnalog1}{A logical value indicating whether the data file has 'analog1' data.}

\item{haveAnalog2}{A logical value indicating whether the data file has 'analog2' data.}

\item{debug}{a flag that turns on debugging.  The value indicates the depth
within the call stack to which debugging applies.  For example,
\code{read.adv.nortek()} calls \code{read.header.nortek()}, so that
\code{read.adv.nortek(...,debug=2)} provides information about not just the
main body of the data file, but also the details of the header.}

\item{monitor}{boolean value indicating whether to indicate the progress
of reading the file, by using \code{\link[=txtProgressBar]{txtProgressBar()}} or otherwise.  The value
of \code{monitor} is changed to \code{FALSE} automatically, for non-interactive
sessions.}

\item{processingLog}{if provided, the action item to be stored in the log.
This parameter is typically only provided for internal calls; the default
that it provides is better for normal calls by a user.}
}
\value{
An \linkS4class{adv} object that contains measurements made with an ADV device.

The \code{metadata} contains information as given in the following table.
The \verb{Nortek name'' is the name used in the Nortek System Integrator Guide (reference 1) and the }Sontek name'' is the name used in the relevant
Sontek documentation.  References are given in square brackets.

\tabular{llll}{

\strong{\code{metadata} name}\tab \strong{Nortek name} \tab \strong{Sontek name} \tab \strong{Meaning}\cr

\code{manufacturer}\tab - \tab - \tab Either \code{"nortek"} or \code{"sontek"}\cr

\code{instrumentType}\tab  - \tab - \tab Either \code{"vector"} or \code{"adv"}\cr

\code{filename}\tab  - \tab - \tab Name of data file(s)\cr

\code{latitude}\tab - \tab - \tab  Latitude of mooring (if applicable)\cr

\code{longitude}\tab - \tab - \tab Longitude of mooring (if applicable)\cr

\code{numberOfSamples}\tab - \tab - \tab Number of data samples in file\cr

\code{numberOfBeams}\tab NBeams (reference 1, p18) \tab - \tab Number of beams (always 3)\cr

\code{numberOfBeamSequencesPerBurst}\tab NPings \tab - \tab number of beam sequences per burst\cr

\code{measurementInterval}\tab MeasInterval (reference 1 p31) \tab - \tab \cr

\code{samplingRate}\tab 512/(AvgInterval) (reference 1 p30; reference 4) \tab - #' \tab \cr

}

The \code{data} list contains items with names corresponding to \code{adp}
objects, with an exception for Nortek data.  Nortek instruments report some
things at a time interval that is longer than the velocity sampling, and
these are stored in \code{data} as \code{timeSlow}, \code{headingSlow},
\code{pitchSlow}, \code{rollSlow}, and \code{temperatureSlow}; if burst
sampling was used, there will also be items \code{recordsBurst} and
\code{timeBurst}.

The \code{processingLog} is in the standard format.
}
\description{
Read an ADV data file, producing an object of type \code{adv}. This
function works by transferring control to a more specialized function,
e.g. \code{\link[=read.adp.nortek]{read.adp.nortek()}} and \code{\link[=read.adp.sontek]{read.adp.sontek()}},
and in many cases users will find it preferable to either use these or
the several even more specialized functions, if the file type is
known.
}
\details{
Files \emph{without} headers may be created in experiments in which a data
logger was set up to monitor the serial data stream from an instrument.  The
lack of header information places a burden on the user, who must supply such
basic information as the times of observations, the instrument orientation,
the instrument coordinate system, etc.  Example 3 below shows how to deal
with such files.  Three things should be noted.
\enumerate{
\item The user must choose the appropriate \code{read.adv} variant
corresponding to the instrument in question.  (This is necessary because
\code{\link[=oceMagic]{oceMagic()}}, which is used by the generic \code{\link[=read.oce]{read.oce()}}
routine, cannot determine the type of instrument by examining a file that
lacks a header.)
\item The call to the \code{read} function must include a start time
(\code{start}) and the number of seconds between data (\code{deltat}),
again, because the instrument data stream may lack those things when the
device is set to a serial mode.  Also, of course, it is necessary to set
\code{header=FALSE} in the function call.
\item Once the file has been read in, the user will be obliged to specify
other information, for the object to be well-formed.  For example, the
\code{read} function will have no way of knowing the instrument orientation,
the coordinate system being used, the transformation matrix to go from
\code{"beam"} to \code{"xyz"} coordinates, or the instrument heading, pitch,
and roll, to go from \code{"xyz"} coordinates to \code{"enu"} coordinates.
Such things are illustrated in example 3 below.
}

In ADV data files, velocities are coded to signed 2-byte integers, with a
scale factor being used to convert to velocity in metres per second.  These
two facts control the maximum recordable velocity and the velocity
resolution, values that may be retrieved for an ADV object name \code{d}
with \code{d[["velocityMaximum"]]} and \code{d[["velocityResolution"]]}.
}
\section{Nortek files}{


\strong{Sampling-rate and similar issues}

The data format is inferred from the System Integrator Guide (reference 1A) and System
Integrator Manual (reference 1B).  These document lacks clarity in spots, and so
\code{read.adv.nortek} contains some assumptions that are noted here, so
that users will be aware of possible problems.

A prominent example is the specification of the sampling rate, stored in
\code{metadata$sampingRate} in the return value.  Repeated examination of
the System Integrator Guide (reference 1) failed to indicate where this value is
stored in the various headers contained in Vector datasets.  After some
experimentation with a few data files, \code{read.adv.nortek} was set up to
calculate \code{metadata$samplingRate} as \code{512/AvgInterval} where
\code{AvgInterval} is a part of the \verb{User Configuration'' header (reference 1 p30), where the explanation is }average interval in seconds'').  This formula was
developed through trial and error, but it was confirmed later on the Nortek
discussion group, and it should appear in upcoming versions of (reference 1).

Another basic issue is the determination of whether an instrument had
recorded in continuous mode or burst mode.  One might infer that
\code{TimCtrlReg} in the \verb{User Configuration'' header (reference 1 p30) determines this, in bits 1 and 2.  However, this was the case in test files available to the author.  For this reason, `read.adv.nortek` infers the mode by reverse engineering of data files of known configuration.  The present version of `read.adv.nortek` determines the sampling mode from the ```NRecords`'' item of the }Vector Velocity Data'' header, which seems
to be 0 for data collected continuously, and non-zero for data collected in
bursts.

Taking these things together, we come upon the issue of how to infer
sampling times for Nortek instruments.  There do not seem to be definitive
documents on this, and so \code{read.adv.nortek} is based partly on
information (of unknown quality) found on Nortek discussion boards.  The
present version of \code{read.adv.nortek} infers the times of velocity
observations differently, depending on whether the instrument was set to
record in burst mode or continuous mode.  For burst mode, times stated in
the burst headers are used, but for continuous mode, times stated in the
``vector system data'' are used.  On the advice found on a Nortek discussion
board, the burst-mode times are offset by 2 seconds to allow for the
instrument warm-up period.

\strong{Handling IMU (inertial measurement unit) data}

Starting in March
2016, \code{read.adv.nortek} has offered some support for handling IMU
(inertial measurement unit) data incorporated into Nortek binary files. This
is not described in the Nortek document named \verb{System Integrator Guide'' (reference 1A) but it appeared in }System Integrator Manual'' (reference 1B;
reference 1C). Confusingly, 1B described 3 varieties of data, whereas 1C does not
describe any of these, but describes instead a fourth variety. As of March
2016, \code{read.adv.nortek} handles all 4 varieties, because files in the
various schemes appear to exist. In \code{oce}, the varieties are named
after the byte code that flags them.  (Variety \code{c3} is the one
described in (reference 1C); the others were described in (reference 1B).)
The variety is stored in the \code{metadata} slot of the returned object as
a string named \code{IMUtype}.

For each variety, the reader is cautioned that strong tests have not been
performed on the code.  One way to test the code is to compare with textual
data files produced by the Nortek software.  In March 2016, an \code{oce}
user shared a dataset of the \code{c3} variety, and this permitted detailed
comparison between the text file and the values inferred by
\code{read.adv.nortek}. The test suggested agreement (to within the
resolution printed in the text file) for velocity (\code{v} in the
\code{data} slot), signal amplitude (\code{a}), correlation (\code{q}),
pressure (\code{p}), the three components of IMU delta angle
(\code{IMUdeltaAngleX} etc), and all components of the rotation matrix
(\code{IMUrotation}).  However, the delta velocity signals did not match,
with \code{IMUdeltaVelocityX} disagreeing in the second decimal place,
\code{IMUdeltaVelocityY} component disagreeing in the first, and
\code{IMUdeltaVelocityZ} being out by a factor of about 10. This is github
issue 893 (\url{https://github.com/dankelley/oce/issues/893}).
\itemize{
\item Variety \code{c3} (signalled by byte 5 of a sequence being
\code{0xc3}) provides information on what Nortek calls DeltaAngle,
DeltaVelocity and Orientation Matrix. (Apart from the orientation matrix,
Nortek provides no documentation on what these quantities mean.) In the
object returned by \code{read.adv.nortek}, these are stored in the
\code{data} slot as \code{IMUdeltaAngleX}, \code{IMUdeltaAngleY},
\code{IMUdeltaAngleZ}, \code{IMUdeltaVelocityX}, \code{IMUdeltaVelocityY},
\code{IMUdeltaVelocityZ}, and \code{IMUrotation}, all vectors except the
last, which is a 3D array. In addition to these, \code{IMUtimestamp} is a
time-stamp, which is not defined in the Nortek documents but seems, from IMU
documents (reference 5), to be defined based on a clock that ticks once per 16
microseconds. Caution may be required in dealing with this timestamp, since
it seemed sensible in one test case (variety \code{d3}) but kept reseting to
zero in another (variety \code{c3}). The lack of Nortek documentation on
most of these quantities is a roadblock to implementing \code{oce} functions
dealing with IMU-enabled datasets
\item Variety \code{cc} (signalled by byte 5 of a sequence being
\code{0xcc}) provides information on acceleration, angular rotation rate,
magnetic vector and orientation matrix.  Each is a timeseries. Acceleration
is stored in the \code{data} slot as \code{IMUaccelX}, \code{IMUaccelY},
\code{IMUaccelz}. The angular rotation components are \code{IMUngrtX},
\code{IMUngrtY} and \code{IMUngrtz}. The magnetic data are in
\code{IMUmagrtx}, \code{IMUmagrty} and \code{IMUmagrtz}. Finally,
\code{IMUmatrix} is a rotation matrix made up from elements named
\code{M11}, \code{M12}, etc in the Nortek documentation.  In addition to all
of these, \code{IMUtime} stores time in seconds, with an origin whose
definition is not stated in reference 1B.
\item Variety \code{d2} (signalled by byte 5 being \code{0xd2}) provides
information on gyro-stabilized acceleration, angular rate and magnetometer
vectors.  The data stored \code{MUaccelX}, \code{IMUangrtX},
\code{IMUmagrtX}, with similar for \code{Y} and \code{Z}.  Again, time is in
\code{IMUtime}. This data type has not been tested as of mid-March 2016,
because the developers do not have a test file with which to test.
\item Variety \code{d3} (signalled by byte 5 being \code{0xd3}) provides
information on DeltaAngle, DeltaVelocity and magnetometer vectors, stored in
\code{IMUdeltaAngleX}, \code{IMUdeltaVelocityX}, and
\code{IMUdeltaMagVectorX}, with similar for \code{Y} and \code{Z}. Again,
time is in \code{IMUtime}. This data type has not been tested as of
mid-March 2016, because the developers do not have a test file with which to
test.
}
}

\examples{
\dontrun{
library(oce)
# A nortek Vector file
d <- read.oce("/data/archive/sleiwex/2008/moorings/m05/adv/nortek_1943/raw/adv_nortek_1943.vec",
              from=as.POSIXct("2008-06-26 00:00:00", tz="UTC"),
              to=as.POSIXct("2008-06-26 00:00:10", tz="UTC"))
plot(d, which=c(1:3,15))
}
}
\references{
1A. Nortek AS.  System Integrator Guide (paradopp family of products). June
2008.  (Doc No: PSI00-0101-0608).  (Users may find it helpful to also
examine newer versions of the guide.)

1B. Nortek AS.  System Integrator Manual. Dec 2014.
(\code{system-integrator-manual_Dec2014_jan.pdf})

1C. Nortek AS.  System Integrator Manual. March 2016.
(\code{system-integrator-manual_Mar2016.pdf})
\enumerate{
\item SonTek/YSI ADVField/Hydra Acoustic Doppler Velocimeter (Field) Technical
Documentation (Sept 1, 2001).
\item Appendix 2.2.3 of the Sontek ADV operation Manual, Firmware Version 4.0
(Oct 1997).
\item Nortek Knowledge Center (http://www.nortekusa.com/en/knowledge-center)
\item A document describing an IMU unit that seems to be close to the one named
in (references 1B and C) as being an adjunct to Nortek Vector systems is at
\verb{http://files.microstrain.com/3DM-GX3-35-Data-Communications-Protocol.pdf}
}
}
\seealso{
Other things related to adv data: 
\code{\link{[[,adv-method}},
\code{\link{[[<-,adv-method}},
\code{\link{adv-class}},
\code{\link{adv}},
\code{\link{beamName}()},
\code{\link{beamToXyz}()},
\code{\link{enuToOtherAdv}()},
\code{\link{enuToOther}()},
\code{\link{plot,adv-method}},
\code{\link{read.adv.sontek.adr}()},
\code{\link{read.adv.sontek.serial}()},
\code{\link{read.adv.sontek.text}()},
\code{\link{read.adv}()},
\code{\link{rotateAboutZ}()},
\code{\link{subset,adv-method}},
\code{\link{summary,adv-method}},
\code{\link{toEnuAdv}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdv}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley
}
\concept{things related to adv data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/air.R
\docType{data}
\name{wind}
\alias{wind}
\title{Wind dataset}
\description{
Wind data inferred from Figure 5 of Koch et al. (1983), provided to
illustrate the \code{\link[=interpBarnes]{interpBarnes()}} function.
Columns \code{wind$x} and \code{wind$y} are location, while \code{wind$z} is
the wind speed, in m/s.
}
\references{
S. E.  Koch and M.  DesJardins and P. J. Kocin, 1983.  ``An
interactive Barnes objective map analysis scheme for use with satellite and
conventional data,'' \emph{J.  Climate Appl.  Met.}, vol 22, p. 1487-1503.
}
\seealso{
Other datasets provided with oce: 
\code{\link{adp}},
\code{\link{adv}},
\code{\link{amsr}},
\code{\link{argo}},
\code{\link{cm}},
\code{\link{coastlineWorld}},
\code{\link{ctdRaw}},
\code{\link{ctd}},
\code{\link{echosounder}},
\code{\link{landsat}},
\code{\link{lisst}},
\code{\link{lobo}},
\code{\link{met}},
\code{\link{ocecolors}},
\code{\link{rsk}},
\code{\link{sealevelTuktoyaktuk}},
\code{\link{sealevel}},
\code{\link{section}},
\code{\link{topoWorld}},
\code{\link{xbt}}
}
\concept{datasets provided with oce}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/met.R
\name{subset,met-method}
\alias{subset,met-method}
\title{Subset a met Object}
\usage{
\S4method{subset}{met}(x, subset, ...)
}
\arguments{
\item{x}{a \linkS4class{met} object.}

\item{subset}{An expression indicating how to subset \code{x}.}

\item{\dots}{ignored.}
}
\value{
A \linkS4class{met} object.
}
\description{
This function is somewhat analogous to \code{\link[=subset.data.frame]{subset.data.frame()}}.
}
\examples{
library(oce)
data(met)
# Few days surrounding Hurricane Juan
plot(subset(met, time > as.POSIXct("2003-09-27", tz="UTC")))

}
\seealso{
Other things related to met data: 
\code{\link{[[,met-method}},
\code{\link{[[<-,met-method}},
\code{\link{as.met}()},
\code{\link{download.met}()},
\code{\link{met-class}},
\code{\link{met}},
\code{\link{plot,met-method}},
\code{\link{read.met}()},
\code{\link{summary,met-method}},
\code{\link{test_met_csv1.csv}},
\code{\link{test_met_csv2.csv}},
\code{\link{test_met_xml2.xml}}

Other functions that subset oce objects: 
\code{\link{subset,adp-method}},
\code{\link{subset,adv-method}},
\code{\link{subset,amsr-method}},
\code{\link{subset,argo-method}},
\code{\link{subset,cm-method}},
\code{\link{subset,coastline-method}},
\code{\link{subset,ctd-method}},
\code{\link{subset,echosounder-method}},
\code{\link{subset,lobo-method}},
\code{\link{subset,oce-method}},
\code{\link{subset,odf-method}},
\code{\link{subset,rsk-method}},
\code{\link{subset,sealevel-method}},
\code{\link{subset,section-method}},
\code{\link{subset,topo-method}},
\code{\link{subset,xbt-method}}
}
\author{
Dan Kelley
}
\concept{functions that subset oce objects}
\concept{things related to met data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.R
\name{setFlags,adp-method}
\alias{setFlags,adp-method}
\title{Set data-quality flags within a adp object}
\usage{
\S4method{setFlags}{adp}(
  object,
  name = NULL,
  i = NULL,
  value = NULL,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{object}{An oce object.}

\item{name}{Character string indicating the name of the variable to be flagged. If
this variable is not contained in the object's \code{data} slot, an error is reported.}

\item{i}{Indication of where to insert the flags; see \dQuote{Description} for
general rules and \dQuote{Details} for rules for \linkS4class{adp}
objects.}

\item{value}{The value to be inserted in the flag.}

\item{debug}{Integer set to 0 for quiet action or to 1 for some debugging.}
}
\value{
An object with flags set as indicated.
}
\description{
This function changes specified entries in the data-quality
flags of a adp object, which are stored within
a list named \code{flags} that resides in the \code{metadata}
slot. If the object already has a flag set up for \code{name},
then only the specified entries are altered. If not, the flag
entry is first created and its entries set to \code{default},
after which the entries specified by \code{i}
are changed to \code{value}.

The specification is made with \code{i}, the form of which
is determined by the data item in question. Generally,
the rules are as follows:
\enumerate{
\item If the data item is a vector, then \code{i} must be (a)
an integer vector specifying indices to be set to \code{value},
(b) a logical vector of length matching the data item, with
\code{TRUE} meaning to set the flag to \code{value}, or (c)
a function that takes an \code{oce} object as its single
argument, and returns a vector in either of the forms
just described.
\item If the data item is an array, then \code{i} must be
(a) a data frame of integers whose rows specify spots to change
(where the number of columns matches the number of dimensions
of the data item), (b) a logical array that has dimension equal to
that of the data item, or (c) a function that takes an \code{oce}
object as its single input and returns such a data frame or array.
}

See \dQuote{Details} for the particular case of \linkS4class{adp} objects.
}
\details{
The only flag that may be set is \code{v}, for the array holding velocity. See \dQuote{Indexing rules}, noting that adp data are stored in 3D arrays; Example 1 shows using a data frame for \code{i}, while Example 2 shows using an array.
}
\examples{
library(oce)
data(adp)

## Example 1: flag first 10 samples in a mid-depth bin of beam 1
i1 <- data.frame(1:20, 40, 1)
adpQC <- initializeFlags(adp, "v", 2)
adpQC <- setFlags(adpQC, "v", i1, 3)
adpClean1 <- handleFlags(adpQC, flags=list(3), actions=list("NA"))
par(mfrow=c(2, 1))
## Top: original, bottom: altered
plot(adp, which="u1")
plot(adpClean1, which="u1")

## Example 2: percent-good and error-beam scheme
v <- adp[["v"]]
i2 <- array(FALSE, dim=dim(v))
g <- adp[["g", "numeric"]]
# Thresholds on percent "goodness" and error "velocity"
G <- 25
V4 <- 0.45
for (k in 1:3)
    i2[,,k] <- ((g[,,k]+g[,,4]) < G) | (v[,,4] > V4)
adpQC2 <- initializeFlags(adp, "v", 2)
adpQC2 <- setFlags(adpQC2, "v", i2, 3)
adpClean2 <- handleFlags(adpQC2, flags=list(3), actions=list("NA"))
## Top: original, bottom: altered
plot(adp, which="u1")
plot(adpClean2, which="u1") # differs at 8h and 20h

}
\seealso{
Other functions relating to data-quality flags: 
\code{\link{defaultFlags}()},
\code{\link{handleFlags,adp-method}},
\code{\link{handleFlags,argo-method}},
\code{\link{handleFlags,ctd-method}},
\code{\link{handleFlags,oce-method}},
\code{\link{handleFlags,section-method}},
\code{\link{handleFlags}()},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{initializeFlagScheme,oce-method}},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{initializeFlagSchemeInternal}()},
\code{\link{initializeFlagScheme}()},
\code{\link{initializeFlags,adp-method}},
\code{\link{initializeFlags,oce-method}},
\code{\link{initializeFlagsInternal}()},
\code{\link{initializeFlags}()},
\code{\link{setFlags,ctd-method}},
\code{\link{setFlags,oce-method}},
\code{\link{setFlags}()}

Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\concept{functions relating to data-quality flags}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/rsk.R
\name{as.rsk}
\alias{as.rsk}
\title{Coerce Data Into a Rsk Object}
\usage{
as.rsk(
  time,
  columns,
  filename = "",
  instrumentType = "rbr",
  serialNumber = "",
  model = "",
  sampleInterval = NA,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{time}{a vector of times for the data.}

\item{columns}{a list or data frame containing the measurements at the indicated
times; see \dQuote{Details}.}

\item{filename}{optional name of file containing the data.}

\item{instrumentType}{type of instrument.}

\item{serialNumber}{serial number for instrument.}

\item{model}{instrument model type, e.g. \code{"RBRduo"}.}

\item{sampleInterval}{sampling interval. If given as \code{NA}, then this is
estimated as the median difference in times.}

\item{debug}{a flag that can be set to \code{TRUE} to turn on debugging.}
}
\value{
An \linkS4class{rsk} object.
}
\description{
Create a rsk object.
}
\details{
The contents of \code{columns} are be copied into the \code{data} slot
of the returned object directly, so it is critical that the names and units
correspond to those expected by other code dealing with
\linkS4class{rsk} objects. If there is a conductivity, it must be called
\code{conductivity}, and it must be in units of mS/cm. If there is a
temperature, it must be called \code{temperature}, and it must be an in-situ
value recorded in ITS-90 units.  And if there is a pressure, it must be
\emph{absolute} pressure (sea pressure plus atmospheric pressure) and it must
be named \code{pressure}. No checks are made within \code{as.rsk} on any of
these rules, but if they are broken, you may expect problems with any further
processing.
}
\seealso{
Other things related to rsk data: 
\code{\link{[[,rsk-method}},
\code{\link{[[<-,rsk-method}},
\code{\link{plot,rsk-method}},
\code{\link{read.rsk}()},
\code{\link{rsk-class}},
\code{\link{rskPatm}()},
\code{\link{rskToc}()},
\code{\link{rsk}},
\code{\link{subset,rsk-method}},
\code{\link{summary,rsk-method}}
}
\author{
Dan Kelley
}
\concept{things related to rsk data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.nortek.R
\name{read.aquadopp}
\alias{read.aquadopp}
\title{Read a Nortek Aquadopp File}
\usage{
read.aquadopp(
  file,
  from = 1,
  to,
  by = 1,
  tz = getOption("oceTz"),
  longitude = NA,
  latitude = NA,
  type = "aquadopp",
  orientation,
  distance,
  monitor = FALSE,
  despike = FALSE,
  processingLog,
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{file}{a connection or a character string giving the name of the file
to load.  (For \code{read.adp.sontek.serial}, this is generally a list of
files, which will be concatenated.)}

\item{from}{indication of the first profile to read.  This can be an
integer, the sequence number of the first profile to read, or a POSIXt time
before which profiles should be skipped, or a character string that converts
to a POSIXt time (assuming UTC timezone).  See \dQuote{Examples}, and make
careful note of the use of the \code{tz} argument. If \code{from} is not
supplied, it defaults to 1.}

\item{to}{an optional indication of the last profile to read, in a
format as described for \code{from}.  As a special case, \code{to=0} means
to read the file to the end.  If \code{to} is not supplied, then it defaults
to 0.}

\item{by}{an optional indication of the stride length to use while walking through
the file.  If this is an integer, then \code{by-1} profiles are skipped
between each pair of profiles that is read, e.g. the default \code{by=1}
means to read all the data.  (For RDI files \emph{only}, there are some
extra features to avoid running out of memory; see \dQuote{Memory considerations}.)}

\item{tz}{character string indicating time zone to be assumed in the data.}

\item{longitude}{optional signed number indicating the longitude in degrees
East.}

\item{latitude}{optional signed number indicating the latitude in degrees
North.}

\item{type}{Either "aquadopp" for a standard aquadopp file (the default), or
"aquadoppPlusMagnetometer" for a file which includes the raw magnetometer
data.}

\item{orientation}{Optional character string specifying the orientation of the
sensor, provided for those cases in which it cannot be inferred from the
data file.  The valid choices are \code{"upward"}, \code{"downward"}, and
\code{"sideward"}.}

\item{distance}{Optional vector holding the distances of bin centres from the
sensor.  This argument is ignored except for Nortek profilers, and need not
be given if the function determines the distances correctly from the data.
The problem is that the distance is poorly documented in the Nortek System
Integrator Guide (2008 edition, page 31), so the function must rely on
word-of-mouth formulae that do not work in all cases.}

\item{monitor}{boolean value indicating whether to indicate the progress
of reading the file, by using \code{\link[=txtProgressBar]{txtProgressBar()}} or otherwise.  The value
of \code{monitor} is changed to \code{FALSE} automatically, for non-interactive
sessions.}

\item{despike}{if \code{TRUE}, \code{\link[=despike]{despike()}} will be used to clean
anomalous spikes in heading, etc.}

\item{processingLog}{if provided, the action item to be stored in the log.
(Typically only provided for internal calls; the default that it provides is
better for normal calls by a user.)}

\item{debug}{a flag that turns on debugging.  Set to 1 to get a moderate
amount of debugging information, or to 2 to get more.}

\item{\dots}{optional additional arguments that some (but not all)
\verb{read.adp.*()} functions pass to lower-level functions.}
}
\value{
An \linkS4class{adp} object.
The contents of that object make sense for the particular instrument
type under study, e.g. if the data file contains
NMEA strings, then navigational data will be stored in an item
called \code{nmea} in the \code{data} slot).
}
\description{
The R code is based on information in the Nortek System Integrator Guide
(2017), postings on the Nortek ``knowledge center'' discussion board, and
discussions with Nortek engineers (Dec. 2020).
}
\references{
\enumerate{
\item Information on Nortek profilers (including the System Integrator Guide,
which explains the data format byte-by-byte) is available at
\url{https://www.nortekusa.com/}.  (One must join the site to see the
manuals.)
\item The Nortek Knowledge Center
\url{https://www.nortekusa.com/en/knowledge-center} may be of help if
problems arise in dealing with data from Nortek instruments.
}
}
\seealso{
Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley and Clark Richards
}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.R
\name{summary,ctd-method}
\alias{summary,ctd-method}
\title{Summarize a ctd Object}
\usage{
\S4method{summary}{ctd}(object, ...)
}
\arguments{
\item{object}{a \linkS4class{ctd} object.}

\item{...}{Further arguments passed to or from other methods.}
}
\description{
Summarizes some of the data in a \code{ctd} object, presenting such information
as the station name, sampling location, data ranges, etc. If the object was read
from a \code{.cnv} file or a \code{.rsk} file, then the \code{OriginalName}
column for the data summary will contain the original names of data within
the source file.
}
\examples{
library(oce)
data(ctd)
summary(ctd)

}
\seealso{
Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}
}
\author{
Dan Kelley
}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{rangeLimit}
\alias{rangeLimit}
\title{Substitute NA for data outside a range}
\usage{
rangeLimit(x, min, max)
}
\arguments{
\item{x}{vector of values}

\item{min}{minimum acceptable value.  If not supplied, and if \code{max} is
also not supplied, a \code{min} of the 0.5 percentile will be used.}

\item{max}{maximum acceptable value.  If not supplied, and if \code{min} is
also not supplied, a \code{min} of the 0.995 percentile will be used.}
}
\description{
Substitute NA for data outside a range, e.g. to remove wild spikes in data.
}
\examples{

ten.to.twenty <- rangeLimit(1:100, 10, 20)
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/cm.R
\name{subset,cm-method}
\alias{subset,cm-method}
\title{Subset a CM Object}
\usage{
\S4method{subset}{cm}(x, subset, ...)
}
\arguments{
\item{x}{a \linkS4class{cm} object.}

\item{subset}{a condition to be applied to the \code{data} portion of \code{x}.
See \sQuote{Details}.}

\item{...}{ignored.}
}
\value{
A new \code{cm} object.
}
\description{
This function is somewhat analogous to \code{\link[=subset.data.frame]{subset.data.frame()}}.
}
\examples{
library(oce)
data(cm)
plot(cm)
plot(subset(cm, time < mean(range(cm[['time']]))))

}
\seealso{
Other things related to cm data: 
\code{\link{[[,cm-method}},
\code{\link{[[<-,cm-method}},
\code{\link{as.cm}()},
\code{\link{cm-class}},
\code{\link{cm}},
\code{\link{plot,cm-method}},
\code{\link{read.cm}()},
\code{\link{rotateAboutZ}()},
\code{\link{summary,cm-method}}

Other functions that subset oce objects: 
\code{\link{subset,adp-method}},
\code{\link{subset,adv-method}},
\code{\link{subset,amsr-method}},
\code{\link{subset,argo-method}},
\code{\link{subset,coastline-method}},
\code{\link{subset,ctd-method}},
\code{\link{subset,echosounder-method}},
\code{\link{subset,lobo-method}},
\code{\link{subset,met-method}},
\code{\link{subset,oce-method}},
\code{\link{subset,odf-method}},
\code{\link{subset,rsk-method}},
\code{\link{subset,sealevel-method}},
\code{\link{subset,section-method}},
\code{\link{subset,topo-method}},
\code{\link{subset,xbt-method}}
}
\author{
Dan Kelley
}
\concept{functions that subset oce objects}
\concept{things related to cm data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/imagep.R
\name{abbreviateTimeLabels}
\alias{abbreviateTimeLabels}
\title{Abbreviate a vector of times by removing commonalities}
\usage{
abbreviateTimeLabels(t, ...)
}
\arguments{
\item{t}{vector of times.}

\item{\dots}{optional arguments passed to the \code{\link[=format]{format()}}, e.g.
\code{format}.}
}
\value{
None.
}
\description{
Abbreviate a vector of times by removing commonalities (e.g. year)
}
\seealso{
This is used by various functions that draw time labels on axes,
e.g.  \code{\link[=plot,adp-method]{plot,adp-method()}}.
}
\author{
Dan Kelley, with help from Clark Richards
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/g1sst.R
\name{read.g1sst}
\alias{read.g1sst}
\title{Read a G1SST file}
\usage{
read.g1sst(file)
}
\arguments{
\item{file}{character value containing the name of a netcdf file containing G1SST data.}
}
\value{
A \linkS4class{g1sst} object.
}
\description{
Read a G1SST file in the netcdf format provided by the ERDDAP server
(see reference 1).
}
\details{
As noted in the documentation for the \linkS4class{g1sst} class, one
must be aware of the incorporation of model simulations in the
\code{g1sst} product. For example, the code presented below
might lead one to believe that the mapped field represents
observations, whereas in fact it can be verified by
consulting reference 2 (clicking and unclicking the radio button to
show just the data) that the field mostly derives from simulation.
}
\examples{
\dontrun{
# Construct query, making it easier to understand and modify.
day <- "2016-01-02"
lon0 <- -66.5
lon1 <- -64.0
lat0 <- 44
lat1 <- 46
source <- paste("https://coastwatch.pfeg.noaa.gov/erddap/griddap/",
                "jplG1SST.nc?",
                "SST\%5B(", day, "T12:00:00Z)",
                "\%5D\%5B(", lat0, "):(", lat1, ")",
                "\%5D\%5B(", lon0, "):(", lon1, ")",
                "\%5D", sep="")
if (!length(list.files(pattern="^a.nc$")))
    download.file(source, "a.nc")
d <- read.g1sst("a.nc")
plot(d, "SST", col=oceColorsTemperature)
if (requireNamespace("ocedata", quietly=TRUE)) {
    data(coastlineWorldFine, package="ocedata")
    lines(coastlineWorldFine[['longitude']],coastlineWorldFine[['latitude']])
}
}

}
\references{
\enumerate{
\item ERDDAP Portal \verb{https://coastwatch.pfeg.noaa.gov/erddap/}
\item JPO OurOcean Portal \verb{https://ourocean.jpl.nasa.gov/SST/}
(link worked in 2016 but was seen to fail 2017 Feb 2).
}
}
\seealso{
Other things related to g1sst data: 
\code{\link{[[,g1sst-method}},
\code{\link{[[<-,g1sst-method}},
\code{\link{g1sst-class}}
}
\author{
Dan Kelley
}
\concept{things related to g1sst data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/bremen.R
\name{[[,bremen-method}
\alias{[[,bremen-method}
\title{Extract Something From a Bremen Object}
\usage{
\S4method{[[}{bremen}(x, i, j, ...)
}
\arguments{
\item{x}{a \linkS4class{bremen} object.}

\item{i}{character value indicating the name of an item to extract.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{ignored.}
}
\description{
Generally, the \code{[[} method lets users extract information from \code{oce}
objects, without having to know the details of the internal storage.  For
many \code{oce} sub-classes, \code{[[} can also return quantities that are computed
from the object's contents.
}
\details{
A two-step process is used to try to find the requested information. First, a
class-specific function is used (see \dQuote{Details of the Specialized
Method}).  If this yields nothing, then a general method is used (see
\dQuote{Details of the General Method}).  If both methods fail, then \code{[[}
returns NULL.

Some understanding of the subclass is required to know what can be retrieved
with \code{[[}. When dealing with an unfamiliar subclass, it can be useful to
first use \code{x[["?"]]} to get a listing of the retrievable items. See
\dQuote{Details of the Specialized Method} for more information.
}
\section{Details of the Specialized Method}{

\itemize{
\item If \code{i} is \code{"?"}, then the return value is a list
containing four items, each of which is a character vector
holding the names of things that can be accessed with \code{[[}.
The \code{data} and \code{metadata} items hold the names of
entries in the object's data and metadata
slots, respectively. The \code{dataDerived}
and \code{metadataDerived} items are each NULL, because
no derived values are defined by \linkS4class{bremen} objects.
}
}

\section{Details of the General Method}{


Note: the text of this section is identical for all \code{oce} subclasses, and so
some of what you read here may not be relevant to the class being described
in this help page.

If the specialized method produces no matches, the following generalized
method is applied. As with the specialized method, the procedure hinges first
on the values of \code{i} and, optionally, \code{j}. The work proceeds in steps, by
testing a sequence of possible conditions in sequence.
\enumerate{
\item A check is made as to whether \code{i} names one of the standard \code{oce} slots.
If so, \code{[[} returns the slot contents of that slot.  Thus, \code{x[["metadata"]]}
will retrieve the \code{metadata} slot, while \code{x[["data"]]} and
\code{x[["processingLog"]]} return those slots.
\item If \code{i} is a string ending in the \code{"Unit"}, then the characters preceding
that string are taken to be the name of an item in the data object, and a
list containing the unit is returned (or \code{NULL} if there is no such unit).
This list consists of an item named \code{unit}, which is an \code{\link[=expression]{expression()}}, and
an item named \code{scale}, which is a string describing the measurement scale.
If the string ends in \code{" unit"}, e.g.  \code{x[["temperature unit"]]} (note the
space), then just the expression is returned, and if it ends in \code{" scale"},
then just the scale is returned.
\item If \code{i} is a string ending in \code{"Flag"}, then the corresponding data-quality
flag is returned (or \code{NULL} if there is no such flag).
\item If the object holds hydrographic information (pressure, salinity,
temperature, longitude and latitude) then another set of possibilities
arises.  If \code{i} is \code{"sigmaTheta"}, then the value of \code{\link[=swSigmaTheta]{swSigmaTheta()}} is
called with \code{x} as the sole argument, and the results are returned.
Similarly, \code{\link[=swSigma0]{swSigma0()}} is used if \code{i="sigma0"}, and \code{\link[=swSpice]{swSpice()}} is used if
\code{i="spice"}. Of course, these actions only make sense for objects that
contain the relevant items within their \code{data} slot.
\item After these possibilities are eliminated, the action depends on whether
\code{j} has been provided.  If \code{j} is not provided, or is the string \code{""}, then
\code{i} is sought in the \code{metadata} slot, and then in the \code{data} slot, returning
whichever is found first.  In other words, if \code{j} is not provided, the
\code{metadata} slot takes preference over the \code{data} slot. However, if \code{j} is
provided, then it must be either the string \code{"metadata"} or \code{"data"}, and it
directs where to look.
\item If none of the above-listed conditions holds, then \code{NULL} is returned.
}
}

\seealso{
Other functions that extract parts of oce objects: 
\code{\link{[[,adp-method}},
\code{\link{[[,adv-method}},
\code{\link{[[,amsr-method}},
\code{\link{[[,argo-method}},
\code{\link{[[,cm-method}},
\code{\link{[[,coastline-method}},
\code{\link{[[,ctd-method}},
\code{\link{[[,echosounder-method}},
\code{\link{[[,g1sst-method}},
\code{\link{[[,gps-method}},
\code{\link{[[,ladp-method}},
\code{\link{[[,landsat-method}},
\code{\link{[[,lisst-method}},
\code{\link{[[,lobo-method}},
\code{\link{[[,met-method}},
\code{\link{[[,oce-method}},
\code{\link{[[,odf-method}},
\code{\link{[[,rsk-method}},
\code{\link{[[,sealevel-method}},
\code{\link{[[,section-method}},
\code{\link{[[,tidem-method}},
\code{\link{[[,topo-method}},
\code{\link{[[,windrose-method}},
\code{\link{[[,xbt-method}},
\code{\link{[[<-,adv-method}}

Other things related to bremen data: 
\code{\link{[[<-,bremen-method}},
\code{\link{bremen-class}},
\code{\link{plot,bremen-method}},
\code{\link{read.bremen}()},
\code{\link{summary,bremen-method}}
}
\author{
Dan Kelley
}
\concept{functions that extract parts of oce objects}
\concept{things related to bremen data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/rsk.R
\name{unitFromStringRsk}
\alias{unitFromStringRsk}
\title{Infer Rsk units from a vector of strings}
\usage{
unitFromStringRsk(s)
}
\arguments{
\item{s}{Vector of character strings, holding the \code{units} entry in the
\code{channels} table of the \code{.rsk} database.}
}
\value{
List of unit lists.
}
\description{
This is used by \code{\link[=read.rsk]{read.rsk()}} to infer the units of data, based
on strings stored in \code{.rsk} files. Lacking a definitive guide
to the format of these file, this function was based on visual inspection
of the data contained within a few sample files; unusual sensors may
not be handled properly.
}
\seealso{
Other functions that interpret variable names and units from headers: 
\code{\link{ODFNames2oceNames}()},
\code{\link{cnvName2oceName}()},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{unitFromString}()},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()}
}
\concept{functions that interpret variable names and units from headers}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{GMTOffsetFromTz}
\alias{GMTOffsetFromTz}
\title{Determine time offset from timezone}
\usage{
GMTOffsetFromTz(tz)
}
\arguments{
\item{tz}{a timezone, e.g. \code{UTC}.}
}
\value{
Number of hours in offset, e.g. \code{AST} yields 4.
}
\description{
The data are from
\url{https://www.timeanddate.com/time/zones/} and were
hand-edited to develop this code, so there may be errors.  Also, note that
some of these contradict; if you examine the code, you'll see some
commented-out portions that represent solving conflicting definitions by
choosing the more common timezone abbreviation over a the less common one.
}
\examples{
library(oce)
cat("Atlantic Standard Time is ", GMTOffsetFromTz("AST"), "hours after UTC")
}
\author{
Dan Kelley
}
\concept{things relating to time}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sealevel.R
\name{[[,sealevel-method}
\alias{[[,sealevel-method}
\title{Extract Something From a Sealevel Object}
\usage{
\S4method{[[}{sealevel}(x, i, j, ...)
}
\arguments{
\item{x}{a \linkS4class{sealevel} object.}

\item{i}{character value indicating the name of an item to extract.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{ignored.}
}
\description{
Generally, the \code{[[} method lets users extract information from \code{oce}
objects, without having to know the details of the internal storage.  For
many \code{oce} sub-classes, \code{[[} can also return quantities that are computed
from the object's contents.
}
\details{
A two-step process is used to try to find the requested information. First, a
class-specific function is used (see \dQuote{Details of the Specialized
Method}).  If this yields nothing, then a general method is used (see
\dQuote{Details of the General Method}).  If both methods fail, then \code{[[}
returns NULL.

Some understanding of the subclass is required to know what can be retrieved
with \code{[[}. When dealing with an unfamiliar subclass, it can be useful to
first use \code{x[["?"]]} to get a listing of the retrievable items. See
\dQuote{Details of the Specialized Method} for more information.
}
\section{Details of the Specialized Method}{

\itemize{
\item If \code{i} is \code{"?"}, then the return value is a list
containing four items, each of which is a character vector
holding the names of things that can be accessed with \code{[[}.
The \code{data} and \code{metadata} items hold the names of
entries in the object's data and metadata
slots, respectively. The \code{dataDerived}
and \code{metadataDerived} items are each NULL, because
no derived values are defined by \linkS4class{sealevel} objects.
\item In many cases, the focus will be on variations of
sealevel elevation over time, so it is common to use e.g.
\code{x[["time"]]} and \code{x[["elevation"]]} to retrieve vectors
of these quantities. Another common task is to retrieve
the location of the observations, using e.g. \code{x[["longitude"]]}
and \code{x[["latitude"]]}.
}
}

\section{Details of the General Method}{


Note: the text of this section is identical for all \code{oce} subclasses, and so
some of what you read here may not be relevant to the class being described
in this help page.

If the specialized method produces no matches, the following generalized
method is applied. As with the specialized method, the procedure hinges first
on the values of \code{i} and, optionally, \code{j}. The work proceeds in steps, by
testing a sequence of possible conditions in sequence.
\enumerate{
\item A check is made as to whether \code{i} names one of the standard \code{oce} slots.
If so, \code{[[} returns the slot contents of that slot.  Thus, \code{x[["metadata"]]}
will retrieve the \code{metadata} slot, while \code{x[["data"]]} and
\code{x[["processingLog"]]} return those slots.
\item If \code{i} is a string ending in the \code{"Unit"}, then the characters preceding
that string are taken to be the name of an item in the data object, and a
list containing the unit is returned (or \code{NULL} if there is no such unit).
This list consists of an item named \code{unit}, which is an \code{\link[=expression]{expression()}}, and
an item named \code{scale}, which is a string describing the measurement scale.
If the string ends in \code{" unit"}, e.g.  \code{x[["temperature unit"]]} (note the
space), then just the expression is returned, and if it ends in \code{" scale"},
then just the scale is returned.
\item If \code{i} is a string ending in \code{"Flag"}, then the corresponding data-quality
flag is returned (or \code{NULL} if there is no such flag).
\item If the object holds hydrographic information (pressure, salinity,
temperature, longitude and latitude) then another set of possibilities
arises.  If \code{i} is \code{"sigmaTheta"}, then the value of \code{\link[=swSigmaTheta]{swSigmaTheta()}} is
called with \code{x} as the sole argument, and the results are returned.
Similarly, \code{\link[=swSigma0]{swSigma0()}} is used if \code{i="sigma0"}, and \code{\link[=swSpice]{swSpice()}} is used if
\code{i="spice"}. Of course, these actions only make sense for objects that
contain the relevant items within their \code{data} slot.
\item After these possibilities are eliminated, the action depends on whether
\code{j} has been provided.  If \code{j} is not provided, or is the string \code{""}, then
\code{i} is sought in the \code{metadata} slot, and then in the \code{data} slot, returning
whichever is found first.  In other words, if \code{j} is not provided, the
\code{metadata} slot takes preference over the \code{data} slot. However, if \code{j} is
provided, then it must be either the string \code{"metadata"} or \code{"data"}, and it
directs where to look.
\item If none of the above-listed conditions holds, then \code{NULL} is returned.
}
}

\seealso{
Other functions that extract parts of oce objects: 
\code{\link{[[,adp-method}},
\code{\link{[[,adv-method}},
\code{\link{[[,amsr-method}},
\code{\link{[[,argo-method}},
\code{\link{[[,bremen-method}},
\code{\link{[[,cm-method}},
\code{\link{[[,coastline-method}},
\code{\link{[[,ctd-method}},
\code{\link{[[,echosounder-method}},
\code{\link{[[,g1sst-method}},
\code{\link{[[,gps-method}},
\code{\link{[[,ladp-method}},
\code{\link{[[,landsat-method}},
\code{\link{[[,lisst-method}},
\code{\link{[[,lobo-method}},
\code{\link{[[,met-method}},
\code{\link{[[,oce-method}},
\code{\link{[[,odf-method}},
\code{\link{[[,rsk-method}},
\code{\link{[[,section-method}},
\code{\link{[[,tidem-method}},
\code{\link{[[,topo-method}},
\code{\link{[[,windrose-method}},
\code{\link{[[,xbt-method}},
\code{\link{[[<-,adv-method}}

Other things related to sealevel data: 
\code{\link{[[<-,sealevel-method}},
\code{\link{as.sealevel}()},
\code{\link{plot,sealevel-method}},
\code{\link{read.sealevel}()},
\code{\link{sealevel-class}},
\code{\link{sealevelTuktoyaktuk}},
\code{\link{sealevel}},
\code{\link{subset,sealevel-method}},
\code{\link{summary,sealevel-method}}
}
\author{
Dan Kelley
}
\concept{functions that extract parts of oce objects}
\concept{things related to sealevel data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/astronomy.R
\name{eclipticalToEquatorial}
\alias{eclipticalToEquatorial}
\title{Convert ecliptical to equatorial coordinate}
\usage{
eclipticalToEquatorial(lambda, beta, epsilon)
}
\arguments{
\item{lambda}{longitude, in degrees, or a data frame containing
\code{lambda}, \code{beta}, and \code{epsilon}, in which case the next to
arguments are ignored}

\item{beta}{geocentric latitude, in degrees}

\item{epsilon}{obliquity of the ecliptic, in degrees}
}
\value{
A data frame containing columns \code{rightAscension} and
\code{declination} both in degrees.
}
\description{
Convert from ecliptical to equatorial coordinates, using
equations 8.3 and 8.4 of reference 1, or, equivalently,
equations 12.3 and 12.4 of reference 2.
}
\details{
The code is based on reference 1; see \code{\link[=moonAngle]{moonAngle()}} for comments on
the differences in formulae found in reference 2.  Indeed,
reference 2 is only cited
here in case readers want to check the ideas of the formulae; DK
has found that reference 2 is available to him via his university library
inter-library loan system, whereas he owns a copy of reference 1.
}
\references{
\itemize{
\item Meeus, Jean. Astronomical Formulas for Calculators. Second Edition.
Richmond, Virginia, USA: Willmann-Bell, 1982.
\item Meeus, Jean. Astronomical Algorithms. Second Edition.
Richmond, Virginia, USA: Willmann-Bell, 1991.
}
}
\seealso{
Other things related to astronomy: 
\code{\link{angle2hms}()},
\code{\link{equatorialToLocalHorizontal}()},
\code{\link{julianCenturyAnomaly}()},
\code{\link{julianDay}()},
\code{\link{moonAngle}()},
\code{\link{siderealTime}()},
\code{\link{sunAngle}()},
\code{\link{sunDeclinationRightAscension}()}
}
\author{
Dan Kelley, based on formulae in references 1 and 2.
}
\concept{things related to astronomy}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/topo.R
\name{[[<-,topo-method}
\alias{[[<-,topo-method}
\title{Replace Parts of a Topo Object}
\usage{
\S4method{[[}{topo}(x, i, j, ...) <- value
}
\arguments{
\item{x}{a \linkS4class{topo} object.}

\item{i}{character value naming the item to replace.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{optional additional information (ignored).}

\item{value}{The value to be placed into \code{x}, somewhere.}
}
\description{
The \code{[[<-} method works for all \linkS4class{oce} objects.
The purpose, as with the related extraction method, \code{[[},
is to insulate users from the internal details of \linkS4class{oce}
objects, by looking for items within the various storage
slots of the object. Items not actually stored can also be
replaced, including units and data-quality
flags.
}
\details{
As with \code{[[} method, the procedure works in steps.

First, the \code{metadata} slot of \code{x} is checked to
see whether it contains something named with \code{i}.
If so, then the named item is replaced with \code{value}.

Otherwise, if the string value of \code{i} ends in \code{Unit}, then the
characters preceding that are taken as the name of a variable, and
the \code{metadata} slot of \code{x} is updated to store that unit, e.g.\preformatted{x[["temperatureUnits"]] <- list(unit=expression(degree*F),scale="")
}

Similarly, if \code{i} ends in \code{Flag}, then quality-control
flags are set up as defined by \code{result}, e.g.\preformatted{o[["temperatureFlags"]] <- c(2,4,2,2)
}

Otherwise, \code{\link[=pmatch]{pmatch()}} is used for a partial-string match with
the names of the items that are in the \code{data} slot of \code{x}.
The first item found (if any) is then updated to hold the value \code{result}.

If none of these conditions is met, a warning is issued.
}
\seealso{
Other things related to topo data: 
\code{\link{[[,topo-method}},
\code{\link{as.topo}()},
\code{\link{download.topo}()},
\code{\link{plot,topo-method}},
\code{\link{read.topo}()},
\code{\link{subset,topo-method}},
\code{\link{summary,topo-method}},
\code{\link{topo-class}},
\code{\link{topoInterpolate}()},
\code{\link{topoWorld}}

Other functions that replace parts of oce objects: 
\code{\link{[[<-,adp-method}},
\code{\link{[[<-,amsr-method}},
\code{\link{[[<-,argo-method}},
\code{\link{[[<-,bremen-method}},
\code{\link{[[<-,cm-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{[[<-,echosounder-method}},
\code{\link{[[<-,g1sst-method}},
\code{\link{[[<-,gps-method}},
\code{\link{[[<-,ladp-method}},
\code{\link{[[<-,landsat-method}},
\code{\link{[[<-,lisst-method}},
\code{\link{[[<-,lobo-method}},
\code{\link{[[<-,met-method}},
\code{\link{[[<-,oce-method}},
\code{\link{[[<-,odf-method}},
\code{\link{[[<-,rsk-method}},
\code{\link{[[<-,sealevel-method}},
\code{\link{[[<-,section-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{[[<-,windrose-method}},
\code{\link{[[<-,xbt-method}}
}
\concept{functions that replace parts of oce objects}
\concept{things related to topo data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/cm.R
\docType{data}
\name{cm}
\alias{cm}
\title{A Current Meter (cm) Object}
\usage{
data(cm)
}
\description{
The result of using \code{\link[=read.cm]{read.cm()}} on a current meter file holding measurements made with an
Interocean S4 device.  See \code{\link[=read.cm]{read.cm()}} for some general cautionary notes on reading such
files. Note that the salinities in this sample dataset are known to be incorrect, perhaps
owing to a lack of calibration of an old instrument that had not been used in a long time.
}
\examples{
library(oce)
data(cm)
summary(cm)
plot(cm)

}
\seealso{
Other datasets provided with oce: 
\code{\link{adp}},
\code{\link{adv}},
\code{\link{amsr}},
\code{\link{argo}},
\code{\link{coastlineWorld}},
\code{\link{ctdRaw}},
\code{\link{ctd}},
\code{\link{echosounder}},
\code{\link{landsat}},
\code{\link{lisst}},
\code{\link{lobo}},
\code{\link{met}},
\code{\link{ocecolors}},
\code{\link{rsk}},
\code{\link{sealevelTuktoyaktuk}},
\code{\link{sealevel}},
\code{\link{section}},
\code{\link{topoWorld}},
\code{\link{wind}},
\code{\link{xbt}}

Other things related to cm data: 
\code{\link{[[,cm-method}},
\code{\link{[[<-,cm-method}},
\code{\link{as.cm}()},
\code{\link{cm-class}},
\code{\link{plot,cm-method}},
\code{\link{read.cm}()},
\code{\link{rotateAboutZ}()},
\code{\link{subset,cm-method}},
\code{\link{summary,cm-method}}
}
\concept{datasets provided with oce}
\concept{things related to cm data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/rsk.R
\name{summary,rsk-method}
\alias{summary,rsk-method}
\title{Summarize a Rsk Object}
\usage{
\S4method{summary}{rsk}(object, ...)
}
\arguments{
\item{object}{An \linkS4class{rsk} object.}

\item{...}{Further arguments passed to or from other methods.}
}
\description{
Summarizes some of the data in a \linkS4class{rsk} object, presenting such information
as the station name, sampling location, data ranges, etc.
}
\examples{
library(oce)
data(rsk)
summary(rsk)

}
\seealso{
The documentation for \linkS4class{rsk} explains the structure
of CTD objects, and also outlines the other functions dealing with them.

Other things related to rsk data: 
\code{\link{[[,rsk-method}},
\code{\link{[[<-,rsk-method}},
\code{\link{as.rsk}()},
\code{\link{plot,rsk-method}},
\code{\link{read.rsk}()},
\code{\link{rsk-class}},
\code{\link{rskPatm}()},
\code{\link{rskToc}()},
\code{\link{rsk}},
\code{\link{subset,rsk-method}}
}
\author{
Dan Kelley
}
\concept{things related to rsk data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sealevel.R
\docType{data}
\name{sealevel}
\alias{sealevel}
\title{Sealevel data for Halifax Harbour}
\source{
The data were created as \preformatted{ sealevel <-
read.oce("490-01-JAN-2003_slev.csv") sealevel <- oce.edit(sealevel,
"longitude", -sealevel[["longitude"]], reason="Fix longitude hemisphere") }
where the csv file was downloaded from reference 1. Note the correction of longitude
sign, which is required because the data file has no indication that this is
the western hemisphere.
}
\description{
This sample sea-level dataset is the 2003 record from Halifax Harbour in
Nova Scotia, Canada.  For reasons that are not mentioned on the data archive
website, the record ends on the 8th of October.
}
\references{
\enumerate{
\item Fisheries and Oceans Canada \code{http://www.meds-sdmm.dfo-mpo.gc.ca/isdm-gdsi/index-eng.html}
}
}
\seealso{
Other datasets provided with oce: 
\code{\link{adp}},
\code{\link{adv}},
\code{\link{amsr}},
\code{\link{argo}},
\code{\link{cm}},
\code{\link{coastlineWorld}},
\code{\link{ctdRaw}},
\code{\link{ctd}},
\code{\link{echosounder}},
\code{\link{landsat}},
\code{\link{lisst}},
\code{\link{lobo}},
\code{\link{met}},
\code{\link{ocecolors}},
\code{\link{rsk}},
\code{\link{sealevelTuktoyaktuk}},
\code{\link{section}},
\code{\link{topoWorld}},
\code{\link{wind}},
\code{\link{xbt}}

Other things related to sealevel data: 
\code{\link{[[,sealevel-method}},
\code{\link{[[<-,sealevel-method}},
\code{\link{as.sealevel}()},
\code{\link{plot,sealevel-method}},
\code{\link{read.sealevel}()},
\code{\link{sealevel-class}},
\code{\link{sealevelTuktoyaktuk}},
\code{\link{subset,sealevel-method}},
\code{\link{summary,sealevel-method}}
}
\author{
Dan Kelley
}
\concept{datasets provided with oce}
\concept{things related to sealevel data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/tides.R
\name{tidemVuf}
\alias{tidemVuf}
\title{Nodal Modulation Calculations for Tidem}
\usage{
tidemVuf(t, j, latitude = NULL)
}
\arguments{
\item{t}{a single time in \code{\link[=POSIXct]{POSIXct()}} format, with timezone \code{"UTC"}.}

\item{j}{integer vector, giving indices of tidal constituents to use.}

\item{latitude}{optional numerical value containing the latitude in degrees North.
If not provided, \code{u} in the return value will be a vector consisting of
repeated 0 value, and \code{f} will be a vector of repeated 1 value.}
}
\value{
A \code{list} containing
items named \code{v}, \code{u} and \code{f} as described in the \code{T_TIDE} documentation,
as well in Pawlowicz et al. (2002) and Foreman (1978).
}
\description{
Carry out nodal modulation calculations for \code{\link[=tidem]{tidem()}}. This function is based directly
on \code{t_vuf} in the \code{T_TIDE} Matlab package (Pawlowicz et al., 2002),
which inherits from the Fortran code described by Foreman (1978).
}
\examples{
## Look up values for the M2 constituent in Halifax Harbour, Canada.
library(oce)
data("tidedata")
j <- with(tidedata$const, which(name=="M2"))
tidemVuf(t=as.POSIXct("2008-01-22 18:50:24"), j=j, lat=44.63)

}
\references{
\itemize{
\item Foreman, M. G. G., 1978. Manual for Tidal Currents Analysis and Prediction.
Pacific Marine Science Report. British Columbia, Canada: Institute of Ocean
Sciences, Patricia Bay.
\item Pawlowicz, Rich, Bob Beardsley, and Steve Lentz, 2002.  Classical tidal
harmonic analysis including error estimates in MATLAB using \code{T_TIDE}.
Computers and Geosciences, 28, 929-937.
}
}
\seealso{
Other things related to tides: 
\code{\link{[[,tidem-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{as.tidem}()},
\code{\link{plot,tidem-method}},
\code{\link{predict.tidem}()},
\code{\link{summary,tidem-method}},
\code{\link{tidalCurrent}},
\code{\link{tidedata}},
\code{\link{tidem-class}},
\code{\link{tidemAstron}()},
\code{\link{tidem}},
\code{\link{webtide}()}
}
\author{
Dan Kelley translated this from the \code{t_vuf} function
of the \code{T_TIDE} Matlab package (see Pawlowicz et al. 2002).
}
\concept{things related to tides}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{oceColorsTemperature}
\alias{oceColorsTemperature}
\alias{oce.colorsTemperature}
\title{Create colors suitable for temperature fields}
\usage{
oceColorsTemperature(n)
}
\arguments{
\item{n}{number of colors to create.}
}
\value{
A vector of color specifications.
}
\description{
Create a set of colors for displaying temperature values,
based on the scheme devised by Thyng et al. (2016) and
presented in a python package by Thyng (2019).
The color specifications were transliterated from
python to R on 2015-09-29, but have not been adjusted
since, even though the python source has changed.  This
is to prevent breaking old \code{oce} code.  To get the
latest versions of these colours or other colours,
use the \CRANpkg{cmocean} R package
(Thyng, Richards, and Krylov, 2019) directly, as is
illustrated (with the "matter" scheme) in Example 2.
Note that the \CRANpkg{cmocean} core functions provide a way
to select between various versions of the colour schemes.
It is also worth considering the palettes provided by the
\CRANpkg{viridis} package, as illustrated (with the
"inferno" scheme) in Example 3.
}
\examples{
library(oce)

# Example 1
imagep(volcano, col=oceColorsTemperature(128),
       zlab="oceColorsTemperature")
\dontrun{
# Example 2 (requires the cmocean package)
imagep(volcano, col=cmocean::cmocean("matter"),
       zlab='cmocean::cmocean("matter")')}

\dontrun{
# Example 3 (requires the viridis package)
imagep(volcano, col=viridis::inferno,
       zlab='viridis::inferno')}

}
\references{
\itemize{
\item Thyng, Kristen, Chad Greene, Robert Hetland, Heather Zimmerle, and Steven DiMarco.
\dQuote{True Colors of Oceanography: Guidelines for Effective and Accurate Colormap Selection.}
Oceanography 29, no. 3 (September 1, 2016): 9‚Äì13.
\doi{10.5670/oceanog.2016.66}
\item Thyng, Kristen. Kthyng/Cmocean. Python, 2019.
\url{https://github.com/kthyng/cmocean}.
\item Thyng, Kristen, Clark Richards, and Ivan Krylov.
Cmocean: Beautiful Colour Maps for Oceanography (version 0.2), 2019.
\url{https://CRAN.R-project.org/package=cmocean}.
}

The following references provide information on choosing
colour schemes, that are suitable for viewers who have
colour deficiencies.

Light, Adam, and Patrick J. Bartlein. "The End of the Rainbow? Color
Schemes for Improved Data Graphics." \emph{Eos, Transactions American Geophysical
Union} 85, no. 40 (2004): 385.
DOI: 10.1029/2004EO400002

Stephenson, David B. "Comment on 'Color Schemes for Improved Data
Graphics', by A Light and P.J. Bartlein." \emph{Eos, Transactions American
Geophysical Union} 86, no.  20 (2005): 196.
DOI: 10.1029/2005EO200005

Light, Adam, and Patrick J. Bartlein. "Reply to 'Comment on Color
Schemes for Improved Data Graphics,' by A. Light and P.J.  Bartlein'."
\emph{Eos, Transactions American Geophysical Union} 86, no. 20 (2005): 196‚Äì196.
DOI: 10.1029/2005EO200006
}
\seealso{
Other things related to colors: 
\code{\link{colormapGMT}()},
\code{\link{colormap}()},
\code{\link{oceColors9B}()},
\code{\link{oceColorsCDOM}()},
\code{\link{oceColorsChlorophyll}()},
\code{\link{oceColorsClosure}()},
\code{\link{oceColorsDensity}()},
\code{\link{oceColorsFreesurface}()},
\code{\link{oceColorsGebco}()},
\code{\link{oceColorsJet}()},
\code{\link{oceColorsOxygen}()},
\code{\link{oceColorsPAR}()},
\code{\link{oceColorsPalette}()},
\code{\link{oceColorsPhase}()},
\code{\link{oceColorsSalinity}()},
\code{\link{oceColorsTurbidity}()},
\code{\link{oceColorsTurbo}()},
\code{\link{oceColorsTwo}()},
\code{\link{oceColorsVelocity}()},
\code{\link{oceColorsViridis}()},
\code{\link{oceColorsVorticity}()},
\code{\link{ocecolors}}
}
\author{
Krysten M. Thyng (Python version), Dan Kelley (R transliteration)
}
\concept{things related to colors}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/lobo.R
\docType{data}
\name{lobo}
\alias{lobo}
\title{Sample LOBO Dataset}
\source{
The data were downloaded from a web interface at Satlantic LOBO web
server and then read with \code{\link[=read.lobo]{read.lobo()}}.
}
\description{
This is sample lobo dataset obtained in the Northwest Arm of Halifax by
Satlantic.
}
\examples{
library(oce)
data(lobo)
summary(lobo)
plot(lobo)

}
\seealso{
Other datasets provided with oce: 
\code{\link{adp}},
\code{\link{adv}},
\code{\link{amsr}},
\code{\link{argo}},
\code{\link{cm}},
\code{\link{coastlineWorld}},
\code{\link{ctdRaw}},
\code{\link{ctd}},
\code{\link{echosounder}},
\code{\link{landsat}},
\code{\link{lisst}},
\code{\link{met}},
\code{\link{ocecolors}},
\code{\link{rsk}},
\code{\link{sealevelTuktoyaktuk}},
\code{\link{sealevel}},
\code{\link{section}},
\code{\link{topoWorld}},
\code{\link{wind}},
\code{\link{xbt}}

Other things related to lobo data: 
\code{\link{[[,lobo-method}},
\code{\link{[[<-,lobo-method}},
\code{\link{as.lobo}()},
\code{\link{lobo-class}},
\code{\link{plot,lobo-method}},
\code{\link{read.lobo}()},
\code{\link{subset,lobo-method}},
\code{\link{summary,lobo-method}}
}
\author{
Dan Kelley
}
\concept{datasets provided with oce}
\concept{things related to lobo data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/satellite.R
\docType{class}
\name{satellite-class}
\alias{satellite-class}
\title{Class to Store Satellite Data}
\description{
This class holds satellite data of various types, including
\linkS4class{amsr} and \linkS4class{g1sst}.
}
\seealso{
Other classes holding satellite data: 
\code{\link{amsr-class}},
\code{\link{g1sst-class}},
\code{\link{landsat-class}}
}
\author{
Dan Kelley and Chantelle Layton
}
\concept{classes holding satellite data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{latFormat}
\alias{latFormat}
\title{Format a latitude}
\usage{
latFormat(lat, digits = max(6, getOption("digits") - 1))
}
\arguments{
\item{lat}{latitude in \eqn{^\circ}{deg}N north of the equator.}

\item{digits}{the number of significant digits to use when printing.}
}
\value{
A character string.
}
\description{
Format a latitude, using "S" for negative latitude.
}
\seealso{
\code{\link[=lonFormat]{lonFormat()}} and \code{\link[=latlonFormat]{latlonFormat()}}.
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/windrose.R
\name{summary,windrose-method}
\alias{summary,windrose-method}
\title{Summarize a \code{windrose} object}
\usage{
\S4method{summary}{windrose}(object, ...)
}
\arguments{
\item{object}{A \linkS4class{windrose} object.}

\item{...}{Further arguments passed to or from other methods.}
}
\description{
Summarizes some of the data in a \code{windrose} object.
}
\seealso{
Other things related to windrose data: 
\code{\link{[[,windrose-method}},
\code{\link{[[<-,windrose-method}},
\code{\link{as.windrose}()},
\code{\link{plot,windrose-method}},
\code{\link{windrose-class}}
}
\author{
Dan Kelley
}
\concept{things related to windrose data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/map.R
\name{lonlat2utm}
\alias{lonlat2utm}
\title{Convert Longitude and Latitude to UTM}
\usage{
lonlat2utm(longitude, latitude, zone, km = FALSE)
}
\arguments{
\item{longitude}{numeric vector of decimal longitude.  May also be a list containing items
named \code{longitude} and \code{latitude}, in which case the indicated
values are used, and next argument is ignored.}

\item{latitude}{numeric vector of decimal latitude (ignored if \code{longitude} is a list
containing both coordinates)}

\item{zone}{optional indication of UTM zone.  Normally this is inferred from
the longitude, but specifying it can be helpful in dealing with Landsat
images, which may cross zones and which therefore are described by a single
zone.}

\item{km}{logical value indicating whether \code{easting} and
\code{northing} are in kilometers or meters.}
}
\value{
A list containing \code{easting}, \code{northing}, \code{zone} and
\code{hemisphere}.
}
\description{
Convert Longitude and Latitude to UTM
}
\examples{
library(oce)
## Cape Split, in the Minas Basin of the Bay of Fundy
lonlat2utm(-64.496567, 45.334626)

}
\references{
\url{https://en.wikipedia.org/wiki/Universal_Transverse_Mercator_coordinate_system},
downloaded May 31, 2014.
}
\seealso{
\code{\link[=utm2lonlat]{utm2lonlat()}} does the inverse operation.  For general
projections and their inverses, use \code{\link[=lonlat2map]{lonlat2map()}} and
\code{\link[=map2lonlat]{map2lonlat()}}.

Other functions related to maps: 
\code{\link{formatPosition}()},
\code{\link{lonlat2map}()},
\code{\link{map2lonlat}()},
\code{\link{mapArrows}()},
\code{\link{mapAxis}()},
\code{\link{mapContour}()},
\code{\link{mapCoordinateSystem}()},
\code{\link{mapDirectionField}()},
\code{\link{mapGrid}()},
\code{\link{mapImage}()},
\code{\link{mapLines}()},
\code{\link{mapLocator}()},
\code{\link{mapLongitudeLatitudeXY}()},
\code{\link{mapPlot}()},
\code{\link{mapPoints}()},
\code{\link{mapPolygon}()},
\code{\link{mapScalebar}()},
\code{\link{mapText}()},
\code{\link{mapTissot}()},
\code{\link{oceCRS}()},
\code{\link{shiftLongitude}()},
\code{\link{usrLonLat}()},
\code{\link{utm2lonlat}()}
}
\author{
Dan Kelley
}
\concept{functions related to maps}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adv.R
\docType{data}
\name{adv}
\alias{adv}
\title{Sample adv (acoustic-doppler velocimeter) dataset}
\source{
This file came from the SLEIWEX-2008 experiment.
}
\usage{
data(adv)
}
\description{
This \linkS4class{adv} object is a sampling of measurements made with a
Nortek Vector acoustic Doppler velocimeter deployed as part of the St Lawrence
Internal Wave Experiment (SLEIWEX).  Various identifying features have been
redacted.
}
\examples{
\donttest{
library(oce)
data(adv)

# Velocity time-series
plot(adv)

# Spectrum of upward component of velocity, with ``turbulent'' reference line
s <- spectrum(adv[["v"]][,3],plot=FALSE)
plot(log10(s$freq), log10(s$spec), type="l")
for (a in seq(-20, 20, by=1))
    abline(a=a, b=-5/3, col="gray", lty="dotted")
}

}
\seealso{
Other datasets provided with oce: 
\code{\link{adp}},
\code{\link{amsr}},
\code{\link{argo}},
\code{\link{cm}},
\code{\link{coastlineWorld}},
\code{\link{ctdRaw}},
\code{\link{ctd}},
\code{\link{echosounder}},
\code{\link{landsat}},
\code{\link{lisst}},
\code{\link{lobo}},
\code{\link{met}},
\code{\link{ocecolors}},
\code{\link{rsk}},
\code{\link{sealevelTuktoyaktuk}},
\code{\link{sealevel}},
\code{\link{section}},
\code{\link{topoWorld}},
\code{\link{wind}},
\code{\link{xbt}}

Other things related to adv data: 
\code{\link{[[,adv-method}},
\code{\link{[[<-,adv-method}},
\code{\link{adv-class}},
\code{\link{beamName}()},
\code{\link{beamToXyz}()},
\code{\link{enuToOtherAdv}()},
\code{\link{enuToOther}()},
\code{\link{plot,adv-method}},
\code{\link{read.adv.nortek}()},
\code{\link{read.adv.sontek.adr}()},
\code{\link{read.adv.sontek.serial}()},
\code{\link{read.adv.sontek.text}()},
\code{\link{read.adv}()},
\code{\link{rotateAboutZ}()},
\code{\link{subset,adv-method}},
\code{\link{summary,adv-method}},
\code{\link{toEnuAdv}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdv}()},
\code{\link{xyzToEnu}()}
}
\concept{datasets provided with oce}
\concept{things related to adv data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.R
\name{read.ctd}
\alias{read.ctd}
\title{Read a General CTD File}
\usage{
read.ctd(
  file,
  type = NULL,
  columns = NULL,
  station = NULL,
  missingValue,
  deploymentType = "unknown",
  monitor = FALSE,
  debug = getOption("oceDebug"),
  processingLog,
  ...
)
}
\arguments{
\item{file}{a connection or a character string giving the name of the file to
load.  For \code{\link[=read.ctd.sbe]{read.ctd.sbe()}} and \code{\link[=read.ctd.woce]{read.ctd.woce()}}, this may be a
wildcard (e.g. \code{"*.cnv"} or \code{"*.csv"}) in which case the return
value is a vector containing CTD objects created by reading the files from
\code{\link[=list.files]{list.files()}} with \code{pattern} set to the specified wildcard
pattern.}

\item{type}{If \code{NULL}, then the first line is studied, in order to
determine the file type, and control is dispatched to a specialized
function to handle that type.  Otherwise, \code{type} must be a string.
If \code{type="SBE19"} then a Seabird file format is assumed,
and control is dispatched to \code{\link[=read.ctd.sbe]{read.ctd.sbe()}}.
If \code{type="WOCE"} then a WOCE-exchange file is assumed,
and control is dispatched to \code{\link[=read.ctd.woce]{read.ctd.woce()}}.
If \code{type="ITP"} then an ice-tethered profiler file is assumed,
and control is dispatched to \code{\link[=read.ctd.itp]{read.ctd.itp()}}.
If \code{type="ODF"} then an ODF file (used by the Canadian Department of
Fisheries and Ocean) is assumed,
and control is dispatched to \code{\link[=read.ctd.odf]{read.ctd.odf()}}.
Finally, if \code{type="ODV"} then an ODV file (used by Ocean Data View software) is assumed,
and control is dispatched to \code{\link[=read.ctd.odv]{read.ctd.odv()}}.}

\item{columns}{an optional \link{list} that can be used to convert unrecognized
data names to resultant variable names.  This is used only by
\code{\link[=read.ctd.sbe]{read.ctd.sbe()}} and \code{\link[=read.ctd.odf]{read.ctd.odf()}}. For example,
if a data file named salinity as \code{"SAL"}, then using\preformatted{d <- read.ctd(f, columns=list(
    salinity=list(name="SAL",
                  unit=list(unit=expression(),
                  scale="PSS-78"))))
}

would assign the \code{"SAL"} column to the \code{salinity} entry in the data
slot of the CTD object returned by the \verb{read.*} function.}

\item{station}{optional character string containing an identifying name or
number for the station. This can be useful if the routine cannot determine the
name automatically, or if another name is preferred.}

\item{missingValue}{optional missing-value flag; data matching this value will
be set to \code{NA} upon reading. If this is provided, then it overrules any
missing-value flag found in the data. For Seabird (\code{.cnv}) files, there is
usually no need to set \code{missingValue}, because it can be inferred from the
header (typically as -9.990e-29). Set \code{missingValue=NULL} to turn off
missing-value detection, even in \code{.cnv} files that contain missing-value
codes in their headers. If \code{missingValue} is not specified,
then an attempt is made to infer such a value from the data, by testing
whether salinity and/or temperature has a minimum that is under -8 in value;
this should catch common values in files, without false positives. A warning
will be issued in this case, and a note inserted in the processing log of
the return value.}

\item{deploymentType}{character string indicating the type of deployment. Use
\code{"unknown"} if this is not known, \code{"profile"} for a profile (in
which the data were acquired during a downcast, while the device was lowered
into the water column, perhaps also including an upcast; \code{"moored"} if
the device is installed on a fixed mooring, \code{"thermosalinograph"} (or
\code{"tsg"}) if the device is mounted on a moving vessel, to record
near-surface properties, or \code{"towyo"} if the device is repeatedly
lowered and raised.}

\item{monitor}{boolean, set to \code{TRUE} to provide an indication of
progress.  This is useful if \code{filename} is a wildcard.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed.}

\item{processingLog}{if provided, the action item to be stored in the log.
This is typically only provided for internal calls; the default that it provides is
better for normal calls by a user.}

\item{...}{additional arguments, passed to called routines.}
}
\value{
a \linkS4class{ctd} object. The details of the contents
depend on the source file. The \code{metadata} slot is particularly
variable across data formats, because the meta-information provided
in those formats varies so widely.
}
\description{
Read a General CTD File
}
\seealso{
Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}

Other functions that read ctd data: 
\code{\link{read.ctd.aml}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()}
}
\author{
Dan Kelley
}
\concept{functions that read ctd data}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/AllClass.R
\docType{class}
\name{oce-class}
\alias{oce-class}
\title{Base Class for oce Objects}
\description{
This is mainly used within oce to create sub-classes, although
users can use \code{new("oce")} to create a blank \code{oce}
object, if desired.
}
\section{Slots}{

\describe{
\item{\code{metadata}}{A list containing information about the data. The
contents vary across sub-classes, e.g. an \linkS4class{adp}
object has information about beam patterns, which obviously would
not make sense for a \linkS4class{ctd} object In addition,
all classes have items named \code{units} and \code{flags}, used
to store information on the units of the data, and the data quality.}

\item{\code{data}}{A list containing the data.}

\item{\code{processingLog}}{A list containing time-stamped processing steps,
typically stored in the object by oce functions.}
}}

\examples{
str(new("oce"))

}
\seealso{
Other classes provided by oce: 
\code{\link{adp-class}},
\code{\link{adv-class}},
\code{\link{argo-class}},
\code{\link{bremen-class}},
\code{\link{cm-class}},
\code{\link{coastline-class}},
\code{\link{ctd-class}},
\code{\link{lisst-class}},
\code{\link{lobo-class}},
\code{\link{met-class}},
\code{\link{odf-class}},
\code{\link{rsk-class}},
\code{\link{sealevel-class}},
\code{\link{section-class}},
\code{\link{topo-class}},
\code{\link{windrose-class}},
\code{\link{xbt-class}}
}
\concept{classes provided by oce}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/cm.R
\name{plot,cm-method}
\alias{plot,cm-method}
\alias{plot.cm}
\title{Plot a cm Object}
\usage{
\S4method{plot}{cm}(
  x,
  which = c(1:2),
  type = "l",
  drawTimeRange = getOption("oceDrawTimeRange"),
  drawZeroLine = FALSE,
  mgp = getOption("oceMgp"),
  mar = c(mgp[1] + 1.5, mgp[1] + 1.5, 1.5, 1.5),
  small = 2000,
  main = "",
  tformat,
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{x}{a \linkS4class{cm} object.}

\item{which}{list of desired plot types.  These are graphed in panels running
down from the top of the page.  See \dQuote{Details} for the meanings of various
values of \code{which}.}

\item{type}{type of plot, as for \code{\link[=plot]{plot()}}.}

\item{drawTimeRange}{boolean that applies to panels with time as the horizontal
axis, indicating whether to draw the time range in the top-left margin of the
plot.}

\item{drawZeroLine}{boolean that indicates whether to draw zero lines on
velocities.}

\item{mgp}{3-element numerical vector to use for \code{par(mgp)}, and also for
\code{par(mar)}, computed from this.  The default is tighter than the R default,
in order to use more space for the data and less for the axes.}

\item{mar}{value to be used with \code{\link{par}}\code{("mar")}.}

\item{small}{an integer indicating the size of data set to be considered
"small", to be plotted with points or lines using the standard
\code{\link[=plot]{plot()}} function.  Data sets with more than \code{small} points will
be plotted with \code{\link[=smoothScatter]{smoothScatter()}} instead.}

\item{main}{main title for plot, used just on the top panel, if there are
several panels.}

\item{tformat}{optional argument passed to \code{\link[=oce.plot.ts]{oce.plot.ts()}}, for plot
types that call that function.  (See \code{\link[=strptime]{strptime()}} for the format
used.)}

\item{debug}{a flag that turns on debugging.  Set to 1 to get a moderate amount
of debugging information, or to 2 to get more.}

\item{...}{Optional arguments passed to plotting functions.}
}
\description{
Creates a multi-panel summary plot of data measured by a current meter.
}
\details{
The panels are controlled by the \code{which} argument, as follows.
\itemize{
\item \code{which=1} or \code{which="u"} for a time-series graph of eastward
velocity, \code{u}, as a function of time.
\item \code{which=2} or \code{which="v"} for a time-series graph of
northward velocity, \code{u}, as a function of time.
\item \code{which=3} or \code{"progressive vector"} for progressive-vector
plot
\item \code{which=4} or \code{"uv"} for a plot of \code{v} versus \code{u}.
(Dots are used for small datasets, and smoothScatter for large ones.)
\item \code{which=5} or \code{"uv+ellipse"} as the \code{"uv"} case, but
with an added indication of the tidal ellipse, calculated from the eigen
vectors of the covariance matrix.
\item \code{which=6} or \code{"uv+ellipse+arrow"} as the \code{"uv+ellipse"}
case, but with an added arrow indicating the mean current.
\item \code{which=7} or \code{"pressure"} for pressure
\item \code{which=8} or \code{"salinity"} for salinity
\item \code{which=9} or \code{"temperature"} for temperature
\item \code{which=10} or \code{"TS"} for a TS diagram
\item \code{which=11} or \code{"conductivity"} for conductivity
\item \code{which=20} or \code{"direction"} for the direction of flow
}
}
\examples{
  library(oce)
  data(cm)
  summary(cm)
  plot(cm)

}
\seealso{
Other functions that plot oce data: 
\code{\link{download.amsr}()},
\code{\link{plot,adp-method}},
\code{\link{plot,adv-method}},
\code{\link{plot,amsr-method}},
\code{\link{plot,argo-method}},
\code{\link{plot,bremen-method}},
\code{\link{plot,coastline-method}},
\code{\link{plot,ctd-method}},
\code{\link{plot,gps-method}},
\code{\link{plot,ladp-method}},
\code{\link{plot,landsat-method}},
\code{\link{plot,lisst-method}},
\code{\link{plot,lobo-method}},
\code{\link{plot,met-method}},
\code{\link{plot,odf-method}},
\code{\link{plot,rsk-method}},
\code{\link{plot,satellite-method}},
\code{\link{plot,sealevel-method}},
\code{\link{plot,section-method}},
\code{\link{plot,tidem-method}},
\code{\link{plot,topo-method}},
\code{\link{plot,windrose-method}},
\code{\link{plot,xbt-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{tidem-class}}

Other things related to cm data: 
\code{\link{[[,cm-method}},
\code{\link{[[<-,cm-method}},
\code{\link{as.cm}()},
\code{\link{cm-class}},
\code{\link{cm}},
\code{\link{read.cm}()},
\code{\link{rotateAboutZ}()},
\code{\link{subset,cm-method}},
\code{\link{summary,cm-method}}
}
\author{
Dan Kelley
}
\concept{functions that plot oce data}
\concept{things related to cm data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/moon.R
\name{moonAngle}
\alias{moonAngle}
\title{Lunar Angle as Function of Space and Time}
\usage{
moonAngle(t, longitude = 0, latitude = 0, useRefraction = TRUE)
}
\arguments{
\item{t}{time, a POSIXt object (converted to timezone \code{"UTC"},
if it is not already in that timezone), a character or numeric value that
corresponds to such a time.}

\item{longitude}{observer longitude in degrees east}

\item{latitude}{observer latitude in degrees north}

\item{useRefraction}{boolean, set to \code{TRUE} to apply a correction for
atmospheric refraction.  (Ignored at present.)}
}
\value{
A list containing the following.
\itemize{
\item \code{time}
\item \code{azimuth} moon azimuth, in degrees eastward of north, from 0 to 360.
Note: this is not the convention used by Meeus, who uses degrees westward of
South. Here, the convention is chosen to more closely match the expectation
of oceanographers.
\item \code{altitude} moon altitude, in degrees from -90 to 90.
\item \code{rightAscension} in' degrees.
\item \code{declination} in degrees.
\item \code{lambda} geocentric longitude, in degrees.
\item \code{beta} geocentric latitude, in degrees.
\item \code{diameter} lunar diameter, in degrees.
\item \code{distance} earth-moon distance, in kilometers.
\item \code{illuminatedFraction} fraction of moon's visible disk that is illuminated.
\item \code{phase} phase of the moon, defined in equation 32.3 of Meeus (1982).
The fractional part of which is 0 for new moon, 1/4 for first
quarter, 1/2 for full moon, and 3/4 for last quarter.
}

\if{html}{\figure{starCoords.png}{options: width="400"}}
}
\description{
The calculations are based on formulae provided by Meeus (1982), primarily
in chapters 6, 18, and 30.  The first step is to compute sidereal time as
formulated in Meeus (1982) chapter 7, which in turn uses Julian day computed
according to as formulae in Meeus (1982) chapter 3.  Using these quantities,
formulae in Meeus (1982) chapter 30 are then used to compute geocentric
longitude (\eqn{lambda}{lambda}, in the Meeus notation), geocentric latitude
(\eqn{beta}{beta}), and parallax.  Then the \code{obliquity} of the ecliptic
is computed with Meeus (1982) equation 18.4.  Equatorial coordinates (right
ascension and declination) are computed with equations 8.3 and 8.4 from
Meeus (1982), using \code{\link[=eclipticalToEquatorial]{eclipticalToEquatorial()}}.  The hour angle
(\eqn{H}{H}) is computed using the unnumbered equation preceding Meeus's
(1982) equation 8.1.  Finally, Meeus (1982) equations 8.5 and 8.6 are used
to calculate the local azimuth and altitude of the moon, using
\code{\link[=equatorialToLocalHorizontal]{equatorialToLocalHorizontal()}}.
}
\section{Alternate formulations}{

Formulae provide by Meeus (1982) are used
for all calculations here.  Meeus (1991) provides formulae that are similar,
but that differ in the 5th or 6th digits.  For example, the formula for
ephemeris time in Meeus (1991) differs from that in Meeus (1992) at the 5th
digit, and almost all of the approximately 200 coefficients in the relevant
formulae also differ in the 5th and 6th digits.  Discussion of the changing
formulations is best left to members of the astronomical community.  For the
present purpose, it may be sufficient to note that \code{moonAngle}, based
on Meeus (1982), reproduces the values provided in example 45.a of Meeus
(1991) to 4 significant digits, e.g. with all angles matching to under 2
minutes of arc.
}

\examples{

library(oce)
par(mfrow=c(3,2))
y <- 2012
m <- 4
days <- 1:3
## Halifax sunrise/sunset (see e.g. https://www.timeanddate.com/worldclock)
rises <- ISOdatetime(y, m, days,c(13,15,16), c(55, 04, 16),0,tz="UTC") + 3 * 3600 # ADT
sets <- ISOdatetime(y, m,days,c(3,4,4), c(42, 15, 45),0,tz="UTC") + 3 * 3600
azrises <- c(69, 75, 82)
azsets <- c(293, 288, 281)
latitude <- 44.65
longitude <- -63.6
for (i in 1:3) {
    t <- ISOdatetime(y, m, days[i],0,0,0,tz="UTC") + seq(0, 24*3600, 3600/4)
    ma <- moonAngle(t, longitude, latitude)
    oce.plot.ts(t, ma$altitude, type='l', mar=c(2, 3, 1, 1), cex=1/2, ylab="Altitude")
    abline(h=0)
    points(rises[i], 0, col='red', pch=3, lwd=2, cex=1.5)
    points(sets[i], 0, col='blue', pch=3, lwd=2, cex=1.5)
    oce.plot.ts(t, ma$azimuth, type='l', mar=c(2, 3, 1, 1), cex=1/2, ylab="Azimuth")
    points(rises[i], -180+azrises[i], col='red', pch=3, lwd=2, cex=1.5)
    points(sets[i], -180+azsets[i], col='blue', pch=3, lwd=2, cex=1.5)
}

}
\references{
\itemize{
\item Meeus, Jean. Astronomical Formulas for Calculators. Second Edition.
Richmond, Virginia, USA: Willmann-Bell, 1982.
\item Meeus, Jean. Astronomical Algorithms. Second Edition.
Richmond, Virginia, USA: Willmann-Bell, 1991.
}
}
\seealso{
The equivalent function for the sun is \code{\link[=sunAngle]{sunAngle()}}.

Other things related to astronomy: 
\code{\link{angle2hms}()},
\code{\link{eclipticalToEquatorial}()},
\code{\link{equatorialToLocalHorizontal}()},
\code{\link{julianCenturyAnomaly}()},
\code{\link{julianDay}()},
\code{\link{siderealTime}()},
\code{\link{sunAngle}()},
\code{\link{sunDeclinationRightAscension}()}
}
\author{
Dan Kelley, based on formulae in Meeus (1982).
}
\concept{things related to astronomy}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{numberAsPOSIXct}
\alias{numberAsPOSIXct}
\title{Convert a Numeric Time to a POSIXct Time}
\usage{
numberAsPOSIXct(t, type, tz = "UTC")
}
\arguments{
\item{t}{an integer corresponding to a time, in a way that depends on
\code{type}.}

\item{type}{the type of time (see \dQuote{Details}).}

\item{tz}{a string indicating the time zone, used only for unix and matlab
times, since GPS times are always referenced to the UTC timezone.}
}
\value{
A \code{\link[=POSIXct]{POSIXct()}} time vector.
}
\description{
There are many varieties, according to the value of \code{type} as defined
in \sQuote{Details}.
}
\details{
\itemize{
\item \code{"unix"} handles Unix times, measured in seconds since the start
of the year 1970.
\item \code{"matlab"} handles Matlab times, measured in days since what
MathWorks (reference 1) calls ``January 0, 0000'' (i.e.  \code{ISOdatetime(0, 1, 1, 0,
0, 0)} in R notation).
\item \code{"gps"} handles the GPS convention. For this, \code{t} is a
two-column matrix, with the first column being the the GPS "week"
(referenced to 1999-08-22) and the second being the GPS "second" (i.e. the
second within the week). Since the GPS satellites do not handle leap
seconds, the R-defined \code{.leap.seconds} is used for corrections.
\item \code{"argo"} handles Argo times, measured in days since the start of
the year 1900.
\item \code{"excel"} handles Excel times, measured in days since the start of
the year 1900. (Note that excel incorrectly regards 1900 as a leap year,
so 1 day is subtracted from \code{t} unless the time is less than or equal
to 1900 Feb 28.  Note that NA is returned for the day 60, which is
what excel codes for "Feb 29, 1900", the non-existing day that excel
accepts.
\item \code{"ncep1"} handles NCEP times, measured in hours since the start
of the year 1800.
\item \code{"ncep2"} handles NCEP times, measured in days since the start of
the year 1. (Note that, for reasons that are unknown at this time, a simple
R expression of this definition is out by two days compared with the UDUNITS
library, which is used by NCEP. Therefore, a two-day offset is applied. See
references 2 and 3.)
\item \code{"sas"} handles SAS times, indicated by \code{type="sas"}, have
origin at the start of 1960.
\item \code{"spss"} handles SPSS times, in seconds after 1582-10-14.
\item \code{"yearday"} handles a convention in which \code{t} is a
two-column matrix, with the first column being the year, and the second the
yearday (starting at 1 for the first second of January 1, to match the
convention used by Sea-Bird CTD software).
\item \code{"epic"} handles a convention used in the EPIC software library,
from the Pacific Marine Environmental Laboratory, in which \code{t} is a
two-column matrix, with the first column being the julian Day (as defined in
\code{\link[=julianDay]{julianDay()}}, for example), and with the second column being the
millisecond within that day. See reference 4.
}

\code{"vms"} handles a convention used in the VMS operating system and
for Modified Julian Day, in which \code{t} is the number of seconds
past 1859-11-17T00:00:00 UTC. See reference 5.
}
\examples{
numberAsPOSIXct(0)                     # unix time 0
numberAsPOSIXct(1, type="matlab")      # matlab time 1
numberAsPOSIXct(cbind(566, 345615), type="gps") # Canada Day, zero hour UTC
numberAsPOSIXct(cbind(2013, 1), type="yearday") # start of 2013

# Epic time, one hour into Canada Day of year 2018. In computing the
# Julian day, note that this starts at noon.
jd <- julianDay(as.POSIXct("2018-07-01 12:00:00", tz="UTC"))
numberAsPOSIXct(cbind(jd, 1e3 * 1 * 3600), type="epic", tz="UTC")

}
\references{
\enumerate{
\item Matlab times:
\url{https://www.mathworks.com/help/matlab/ref/datenum.html}
\item NCEP times: \url{https://psl.noaa.gov/data/gridded/faq.html}
\item problem with NCEP times:
\url{https://github.com/dankelley/oce/issues/738}
\item EPIC times: software and manuals at \url{https://www.pmel.noaa.gov/epic/download/index.html#epslib};
see also Denbo, Donald W., and Nancy N. Soreide. \dQuote{EPIC.} Oceanography 9 (1996).
\doi{10.5670/oceanog.1996.10}
\item VMS times: https://en.wikipedia.org/wiki/Epoch_(computing)
}
}
\seealso{
\code{\link[=numberAsHMS]{numberAsHMS()}}

Other things related to time: 
\code{\link{ctimeToSeconds}()},
\code{\link{julianCenturyAnomaly}()},
\code{\link{julianDay}()},
\code{\link{numberAsHMS}()},
\code{\link{secondsToCtime}()},
\code{\link{unabbreviateYear}()}
}
\author{
Dan Kelley
}
\concept{things related to time}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{useHeading}
\alias{useHeading}
\title{Replace the Heading for One Instrument With That of Another}
\usage{
useHeading(b, g, add = 0)
}
\arguments{
\item{b}{object holding data from an instrument whose heading is bad, but
whose other data are good.}

\item{g}{object holding data from an instrument whose heading is good, and
should be interpolated to the time base of \code{b}.}

\item{add}{an angle, in degrees, to be added to the heading.}
}
\value{
A copy of \code{b}, but with \code{b$data$heading} replaced with
heading angles that result from linear interpolation of the headings in
\code{g}, and then adding the angle \code{add}.
}
\description{
Replace the heading angles in one oce object with that from another,
possibly with a constant adjustment.
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.nortek.R
\name{ad2cpHeaderValue}
\alias{ad2cpHeaderValue}
\title{Decode an item from a Nortek AD2CP file header}
\usage{
ad2cpHeaderValue(x, key, item, numeric = TRUE, default)
}
\arguments{
\item{x}{an \linkS4class{adp} object that holds AD2CP data.}

\item{key}{Character value that identifies a particular line in \code{x[["text"]]}.}

\item{item}{Character value indicating the name of the item sought.}

\item{numeric}{Logical value indicating whether to convert the return value
from a string to a numerical value.}

\item{default}{Optional value to be used if the item is not found in the
header, or if the header is \code{NULL} (as in the case of a split-up file
that lacks the initial header information)}
}
\value{
String or number interpreted from the \code{x[["text"]]}, or \code{NULL},
if the desired item is not found there, or if \code{x} is not of the required
class and variety.
}
\description{
Decode an item from a Nortek AD2CP file header
}
\examples{
\dontrun{
d <- read.oce("a.ad2cp")
# The examples start with the line in x[["text"]][[1]]; note that in the second
# example, it would be insuficient to use a key of "BEAMCFGLIST", because that will
# yield 4 lines, and the function is not designed to handle that.

# ID,STR=\"Signature1000\",SN=123456
type <- ad2cpHeaderValue(d, "ID", "STR", numeric=FALSE)
serialNumber <- ad2cpHeaderValue(d, "ID", "SN")

# BEAMCFGLIST,BEAM=1,THETA=25.00,PHI=0.00,FREQ=1000,BW=25,BRD=1,HWBEAM=1,ZNOM=60.00
beam1Angle <- ad2cpHeaderValue(d, "BEAMCFGLIST,BEAM=1", "THETA")
frequency <- ad2cpHeaderValue(d, "BEAMCFGLIST,BEAM=1", "FREQ", default=NA)
}

}
\seealso{
Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/echosounder.R
\name{subset,echosounder-method}
\alias{subset,echosounder-method}
\title{Subset an Echosounder Object}
\usage{
\S4method{subset}{echosounder}(x, subset, ...)
}
\arguments{
\item{x}{an \linkS4class{echosounder} object.}

\item{subset}{a condition to be applied to the \code{data} portion of
\code{x}.  See \sQuote{Details}.}

\item{\dots}{ignored.}
}
\value{
An \linkS4class{echosounder} object.
}
\description{
This function is somewhat analogous to \code{\link[=subset.data.frame]{subset.data.frame()}}.
Subsetting can be by \code{time} or \code{depth}, but these may not be
combined; use a sequence of calls to subset by both.
}
\examples{
library(oce)
data(echosounder)
plot(echosounder)
plot(subset(echosounder, depth < 10))
plot(subset(echosounder, time < mean(range(echosounder[['time']]))))

}
\seealso{
Other things related to echosounder data: 
\code{\link{[[,echosounder-method}},
\code{\link{[[<-,echosounder-method}},
\code{\link{as.echosounder}()},
\code{\link{echosounder-class}},
\code{\link{echosounder}},
\code{\link{findBottom}()},
\code{\link{plot,echosounder-method}},
\code{\link{read.echosounder}()},
\code{\link{summary,echosounder-method}}

Other functions that subset oce objects: 
\code{\link{subset,adp-method}},
\code{\link{subset,adv-method}},
\code{\link{subset,amsr-method}},
\code{\link{subset,argo-method}},
\code{\link{subset,cm-method}},
\code{\link{subset,coastline-method}},
\code{\link{subset,ctd-method}},
\code{\link{subset,lobo-method}},
\code{\link{subset,met-method}},
\code{\link{subset,oce-method}},
\code{\link{subset,odf-method}},
\code{\link{subset,rsk-method}},
\code{\link{subset,sealevel-method}},
\code{\link{subset,section-method}},
\code{\link{subset,topo-method}},
\code{\link{subset,xbt-method}}
}
\author{
Dan Kelley
}
\concept{functions that subset oce objects}
\concept{things related to echosounder data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{latlonFormat}
\alias{latlonFormat}
\title{Format a latitude-longitude pair}
\usage{
latlonFormat(lat, lon, digits = max(6, getOption("digits") - 1))
}
\arguments{
\item{lat}{latitude in \eqn{^\circ}{deg}N north of the equator.}

\item{lon}{longitude in \eqn{^\circ}{deg}N east of Greenwich.}

\item{digits}{the number of significant digits to use when printing.}
}
\value{
A character string.
}
\description{
Format a latitude-longitude pair, using "S" for negative latitudes, etc.
}
\seealso{
\code{\link[=latFormat]{latFormat()}} and \code{\link[=lonFormat]{lonFormat()}}.
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/extdata.R
\docType{data}
\encoding{UTF-8}
\name{ctd.cnv}
\alias{ctd.cnv}
\title{Sample ctd dataset in .cnv format}
\description{
Sample ctd dataset in .cnv format
}
\examples{
\dontrun{
read.oce(system.file("extdata", "ctd.cnv", package="oce"))
}

}
\seealso{
Other raw datasets: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{adp_rdi.000}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{test_met_csv1.csv}},
\code{\link{test_met_csv2.csv}},
\code{\link{test_met_xml2.xml}},
\code{\link{xbt.edf}}

Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}
}
\concept{raw datasets}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/echosounder.R
\name{summary,echosounder-method}
\alias{summary,echosounder-method}
\title{Summarize an Echosounder Object}
\usage{
\S4method{summary}{echosounder}(object, ...)
}
\arguments{
\item{object}{an object of class \code{"echosounder"}, usually, a result of
a call to \code{\link[=read.echosounder]{read.echosounder()}}, \code{\link[=read.oce]{read.oce()}}, or
\code{\link[=as.echosounder]{as.echosounder()}}.}

\item{\dots}{further arguments passed to or from other methods.}
}
\description{
Summarizes some of the data in an \linkS4class{echosounder} object.
}
\seealso{
Other things related to echosounder data: 
\code{\link{[[,echosounder-method}},
\code{\link{[[<-,echosounder-method}},
\code{\link{as.echosounder}()},
\code{\link{echosounder-class}},
\code{\link{echosounder}},
\code{\link{findBottom}()},
\code{\link{plot,echosounder-method}},
\code{\link{read.echosounder}()},
\code{\link{subset,echosounder-method}}
}
\author{
Dan Kelley
}
\concept{things related to echosounder data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adv.R
\name{[[<-,adv-method}
\alias{[[<-,adv-method}
\title{Replace Parts of an ADV Object}
\usage{
\S4method{[[}{adv}(x, i, j, ...) <- value
}
\arguments{
\item{x}{an \linkS4class{adv} object.}

\item{i}{character value indicating the name of an item to extract.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{ignored.}

\item{value}{The value to be inserted into \code{x}.}
}
\description{
Generally, the \code{[[} method lets users extract information from \code{oce}
objects, without having to know the details of the internal storage.  For
many \code{oce} sub-classes, \code{[[} can also return quantities that are computed
from the object's contents.
}
\details{
If the \code{adv} object holds slow variables (i.e. if \code{timeSlow} is
in the \code{data} slot), then assigning to .e.g. \code{heading} will not
actually assign to a variable of that name, but instead assigns to
\code{headingSlow}. To catch misapplication of this rule, an error
message will be issued if the assigned value is not of the same length
as \code{timeSlow}.

A two-step process is used to try to find the requested information. First, a
class-specific function is used (see \dQuote{Details of the Specialized
Method}).  If this yields nothing, then a general method is used (see
\dQuote{Details of the General Method}).  If both methods fail, then \code{[[}
returns NULL.

Some understanding of the subclass is required to know what can be retrieved
with \code{[[}. When dealing with an unfamiliar subclass, it can be useful to
first use \code{x[["?"]]} to get a listing of the retrievable items. See
\dQuote{Details of the Specialized Method} for more information.
}
\section{Details of the General Method}{


Note: the text of this section is identical for all \code{oce} subclasses, and so
some of what you read here may not be relevant to the class being described
in this help page.

If the specialized method produces no matches, the following generalized
method is applied. As with the specialized method, the procedure hinges first
on the values of \code{i} and, optionally, \code{j}. The work proceeds in steps, by
testing a sequence of possible conditions in sequence.
\enumerate{
\item A check is made as to whether \code{i} names one of the standard \code{oce} slots.
If so, \code{[[} returns the slot contents of that slot.  Thus, \code{x[["metadata"]]}
will retrieve the \code{metadata} slot, while \code{x[["data"]]} and
\code{x[["processingLog"]]} return those slots.
\item If \code{i} is a string ending in the \code{"Unit"}, then the characters preceding
that string are taken to be the name of an item in the data object, and a
list containing the unit is returned (or \code{NULL} if there is no such unit).
This list consists of an item named \code{unit}, which is an \code{\link[=expression]{expression()}}, and
an item named \code{scale}, which is a string describing the measurement scale.
If the string ends in \code{" unit"}, e.g.  \code{x[["temperature unit"]]} (note the
space), then just the expression is returned, and if it ends in \code{" scale"},
then just the scale is returned.
\item If \code{i} is a string ending in \code{"Flag"}, then the corresponding data-quality
flag is returned (or \code{NULL} if there is no such flag).
\item If the object holds hydrographic information (pressure, salinity,
temperature, longitude and latitude) then another set of possibilities
arises.  If \code{i} is \code{"sigmaTheta"}, then the value of \code{\link[=swSigmaTheta]{swSigmaTheta()}} is
called with \code{x} as the sole argument, and the results are returned.
Similarly, \code{\link[=swSigma0]{swSigma0()}} is used if \code{i="sigma0"}, and \code{\link[=swSpice]{swSpice()}} is used if
\code{i="spice"}. Of course, these actions only make sense for objects that
contain the relevant items within their \code{data} slot.
\item After these possibilities are eliminated, the action depends on whether
\code{j} has been provided.  If \code{j} is not provided, or is the string \code{""}, then
\code{i} is sought in the \code{metadata} slot, and then in the \code{data} slot, returning
whichever is found first.  In other words, if \code{j} is not provided, the
\code{metadata} slot takes preference over the \code{data} slot. However, if \code{j} is
provided, then it must be either the string \code{"metadata"} or \code{"data"}, and it
directs where to look.
\item If none of the above-listed conditions holds, then \code{NULL} is returned.
}
}

\seealso{
Other functions that extract parts of oce objects: 
\code{\link{[[,adp-method}},
\code{\link{[[,adv-method}},
\code{\link{[[,amsr-method}},
\code{\link{[[,argo-method}},
\code{\link{[[,bremen-method}},
\code{\link{[[,cm-method}},
\code{\link{[[,coastline-method}},
\code{\link{[[,ctd-method}},
\code{\link{[[,echosounder-method}},
\code{\link{[[,g1sst-method}},
\code{\link{[[,gps-method}},
\code{\link{[[,ladp-method}},
\code{\link{[[,landsat-method}},
\code{\link{[[,lisst-method}},
\code{\link{[[,lobo-method}},
\code{\link{[[,met-method}},
\code{\link{[[,oce-method}},
\code{\link{[[,odf-method}},
\code{\link{[[,rsk-method}},
\code{\link{[[,sealevel-method}},
\code{\link{[[,section-method}},
\code{\link{[[,tidem-method}},
\code{\link{[[,topo-method}},
\code{\link{[[,windrose-method}},
\code{\link{[[,xbt-method}}

Other things related to adv data: 
\code{\link{[[,adv-method}},
\code{\link{adv-class}},
\code{\link{adv}},
\code{\link{beamName}()},
\code{\link{beamToXyz}()},
\code{\link{enuToOtherAdv}()},
\code{\link{enuToOther}()},
\code{\link{plot,adv-method}},
\code{\link{read.adv.nortek}()},
\code{\link{read.adv.sontek.adr}()},
\code{\link{read.adv.sontek.serial}()},
\code{\link{read.adv.sontek.text}()},
\code{\link{read.adv}()},
\code{\link{rotateAboutZ}()},
\code{\link{subset,adv-method}},
\code{\link{summary,adv-method}},
\code{\link{toEnuAdv}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdv}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley
}
\concept{functions that extract parts of oce objects}
\concept{things related to adv data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{oceMagic}
\alias{oceMagic}
\title{Find the Type of an Oceanographic Data File}
\usage{
oceMagic(file, debug = getOption("oceDebug"))
}
\arguments{
\item{file}{a connection or a character string giving the name of the file
to be checked.}

\item{debug}{an integer, set non-zero to turn on debugging.  Higher values
indicate more debugging.}
}
\value{
A character string indicating the file type, or \code{"unknown"}, if
the type cannot be determined. If the result contains \code{"/"} characters,
these separate a list describing the file type, with the first element being
the general type, the second element being the manufacturer, and the third
element being the manufacturer's name for the instrument. For example,
\code{"adp/nortek/aquadopp"} indicates a acoustic-doppler profiler made by
NorTek, of the model type called Aquadopp.
}
\description{
\code{oceMagic} tries to infer the file type, based on the data
within the file, the file name, or a combination of the two.
}
\details{
\code{oceMagic} was previously called \code{oce.magic}, but that
alias was removed in version 0.9.24; see \link{oce-defunct}.
}
\seealso{
This is used mainly by \code{\link[=read.oce]{read.oce()}}.
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/echosounder.R
\name{[[<-,echosounder-method}
\alias{[[<-,echosounder-method}
\title{Replace Parts of an Echosounder Object}
\usage{
\S4method{[[}{echosounder}(x, i, j, ...) <- value
}
\arguments{
\item{x}{an \linkS4class{echosounder} object.}

\item{i}{character value naming the item to replace.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{optional additional information (ignored).}

\item{value}{The value to be placed into \code{x}, somewhere.}
}
\description{
The \code{[[<-} method works for all \linkS4class{oce} objects.
The purpose, as with the related extraction method, \code{[[},
is to insulate users from the internal details of \linkS4class{oce}
objects, by looking for items within the various storage
slots of the object. Items not actually stored can also be
replaced, including units and data-quality
flags.
}
\details{
As with \code{[[} method, the procedure works in steps.

First, the \code{metadata} slot of \code{x} is checked to
see whether it contains something named with \code{i}.
If so, then the named item is replaced with \code{value}.

Otherwise, if the string value of \code{i} ends in \code{Unit}, then the
characters preceding that are taken as the name of a variable, and
the \code{metadata} slot of \code{x} is updated to store that unit, e.g.\preformatted{x[["temperatureUnits"]] <- list(unit=expression(degree*F),scale="")
}

Similarly, if \code{i} ends in \code{Flag}, then quality-control
flags are set up as defined by \code{result}, e.g.\preformatted{o[["temperatureFlags"]] <- c(2,4,2,2)
}

Otherwise, \code{\link[=pmatch]{pmatch()}} is used for a partial-string match with
the names of the items that are in the \code{data} slot of \code{x}.
The first item found (if any) is then updated to hold the value \code{result}.

If none of these conditions is met, a warning is issued.
}
\seealso{
Other functions that replace parts of oce objects: 
\code{\link{[[<-,adp-method}},
\code{\link{[[<-,amsr-method}},
\code{\link{[[<-,argo-method}},
\code{\link{[[<-,bremen-method}},
\code{\link{[[<-,cm-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{[[<-,g1sst-method}},
\code{\link{[[<-,gps-method}},
\code{\link{[[<-,ladp-method}},
\code{\link{[[<-,landsat-method}},
\code{\link{[[<-,lisst-method}},
\code{\link{[[<-,lobo-method}},
\code{\link{[[<-,met-method}},
\code{\link{[[<-,oce-method}},
\code{\link{[[<-,odf-method}},
\code{\link{[[<-,rsk-method}},
\code{\link{[[<-,sealevel-method}},
\code{\link{[[<-,section-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{[[<-,topo-method}},
\code{\link{[[<-,windrose-method}},
\code{\link{[[<-,xbt-method}}

Other things related to echosounder data: 
\code{\link{[[,echosounder-method}},
\code{\link{as.echosounder}()},
\code{\link{echosounder-class}},
\code{\link{echosounder}},
\code{\link{findBottom}()},
\code{\link{plot,echosounder-method}},
\code{\link{read.echosounder}()},
\code{\link{subset,echosounder-method}},
\code{\link{summary,echosounder-method}}
}
\concept{functions that replace parts of oce objects}
\concept{things related to echosounder data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/tides.R
\name{tidemAstron}
\alias{tidemAstron}
\title{Astronomical Calculations for Tidem}
\usage{
tidemAstron(t)
}
\arguments{
\item{t}{Either a time in \code{POSIXct} format (with \code{"UTC"} timezone,
or else odd behaviours may result),
or an integer. In the second case, it is converted to a time with
\code{\link[=numberAsPOSIXct]{numberAsPOSIXct()}}, using \code{tz="UTC"}.}
}
\value{
A \code{list} containing items named
\code{astro} and \code{ader} (see the \code{T_TIDE} documentation).
}
\description{
Do some astronomical calculations for \code{\link[=tidem]{tidem()}}.  This function is based directly
on \code{t_astron} in the \code{T_TIDE} Matlab package (see Pawlowicz et al. 2002),
which inherits from the Fortran code described by Foreman (1978).
}
\examples{
tidemAstron(as.POSIXct("2008-01-22 18:50:24"))

}
\references{
\itemize{
\item Foreman, M. G. G., 1978. Manual for Tidal Currents Analysis and Prediction.
Pacific Marine Science Report. British Columbia, Canada: Institute of Ocean
Sciences, Patricia Bay.
\item Pawlowicz, Rich, Bob Beardsley, and Steve Lentz, 2002.  Classical tidal
harmonic analysis including error estimates in MATLAB using \code{T_TIDE}.
Computers and Geosciences, 28, 929-937.
}
}
\seealso{
Other things related to tides: 
\code{\link{[[,tidem-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{as.tidem}()},
\code{\link{plot,tidem-method}},
\code{\link{predict.tidem}()},
\code{\link{summary,tidem-method}},
\code{\link{tidalCurrent}},
\code{\link{tidedata}},
\code{\link{tidem-class}},
\code{\link{tidemVuf}()},
\code{\link{tidem}},
\code{\link{webtide}()}
}
\author{
Dan Kelley translated this from the \code{t_astron} function
of the \code{T_TIDE} Matlab package (see Pawlowicz et al. 2002).
}
\concept{things related to tides}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/coastline.R
\name{[[,coastline-method}
\alias{[[,coastline-method}
\title{Extract Something From a Coastline Object}
\usage{
\S4method{[[}{coastline}(x, i, j, ...)
}
\arguments{
\item{x}{a \linkS4class{coastline} object.}

\item{i}{character value indicating the name of an item to extract.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{ignored.}
}
\description{
Generally, the \code{[[} method lets users extract information from \code{oce}
objects, without having to know the details of the internal storage.  For
many \code{oce} sub-classes, \code{[[} can also return quantities that are computed
from the object's contents.
}
\details{
A two-step process is used to try to find the requested information. First, a
class-specific function is used (see \dQuote{Details of the Specialized
Method}).  If this yields nothing, then a general method is used (see
\dQuote{Details of the General Method}).  If both methods fail, then \code{[[}
returns NULL.

Some understanding of the subclass is required to know what can be retrieved
with \code{[[}. When dealing with an unfamiliar subclass, it can be useful to
first use \code{x[["?"]]} to get a listing of the retrievable items. See
\dQuote{Details of the Specialized Method} for more information.
}
\section{Details of the Specialized Method}{

\itemize{
\item If \code{i} is \code{"?"}, then the return value is a list
containing four items, each of which is a character vector
holding the names of things that can be accessed with \code{[[}.
The \code{data} and \code{metadata} items hold the names of
entries in the object's data and metadata
slots, respectively. The \code{dataDerived}
and \code{metadataDerived} items are each NULL, because
no derived values are defined for \linkS4class{coastline} objects.
\item In many cases, the focus will be on the coastline trace
in longitude-latitude space, so \code{x[["longitude"]]}
and \code{x[["latitude"]]} are commonly used.
}
}

\section{Details of the General Method}{


Note: the text of this section is identical for all \code{oce} subclasses, and so
some of what you read here may not be relevant to the class being described
in this help page.

If the specialized method produces no matches, the following generalized
method is applied. As with the specialized method, the procedure hinges first
on the values of \code{i} and, optionally, \code{j}. The work proceeds in steps, by
testing a sequence of possible conditions in sequence.
\enumerate{
\item A check is made as to whether \code{i} names one of the standard \code{oce} slots.
If so, \code{[[} returns the slot contents of that slot.  Thus, \code{x[["metadata"]]}
will retrieve the \code{metadata} slot, while \code{x[["data"]]} and
\code{x[["processingLog"]]} return those slots.
\item If \code{i} is a string ending in the \code{"Unit"}, then the characters preceding
that string are taken to be the name of an item in the data object, and a
list containing the unit is returned (or \code{NULL} if there is no such unit).
This list consists of an item named \code{unit}, which is an \code{\link[=expression]{expression()}}, and
an item named \code{scale}, which is a string describing the measurement scale.
If the string ends in \code{" unit"}, e.g.  \code{x[["temperature unit"]]} (note the
space), then just the expression is returned, and if it ends in \code{" scale"},
then just the scale is returned.
\item If \code{i} is a string ending in \code{"Flag"}, then the corresponding data-quality
flag is returned (or \code{NULL} if there is no such flag).
\item If the object holds hydrographic information (pressure, salinity,
temperature, longitude and latitude) then another set of possibilities
arises.  If \code{i} is \code{"sigmaTheta"}, then the value of \code{\link[=swSigmaTheta]{swSigmaTheta()}} is
called with \code{x} as the sole argument, and the results are returned.
Similarly, \code{\link[=swSigma0]{swSigma0()}} is used if \code{i="sigma0"}, and \code{\link[=swSpice]{swSpice()}} is used if
\code{i="spice"}. Of course, these actions only make sense for objects that
contain the relevant items within their \code{data} slot.
\item After these possibilities are eliminated, the action depends on whether
\code{j} has been provided.  If \code{j} is not provided, or is the string \code{""}, then
\code{i} is sought in the \code{metadata} slot, and then in the \code{data} slot, returning
whichever is found first.  In other words, if \code{j} is not provided, the
\code{metadata} slot takes preference over the \code{data} slot. However, if \code{j} is
provided, then it must be either the string \code{"metadata"} or \code{"data"}, and it
directs where to look.
\item If none of the above-listed conditions holds, then \code{NULL} is returned.
}
}

\seealso{
Other functions that extract parts of oce objects: 
\code{\link{[[,adp-method}},
\code{\link{[[,adv-method}},
\code{\link{[[,amsr-method}},
\code{\link{[[,argo-method}},
\code{\link{[[,bremen-method}},
\code{\link{[[,cm-method}},
\code{\link{[[,ctd-method}},
\code{\link{[[,echosounder-method}},
\code{\link{[[,g1sst-method}},
\code{\link{[[,gps-method}},
\code{\link{[[,ladp-method}},
\code{\link{[[,landsat-method}},
\code{\link{[[,lisst-method}},
\code{\link{[[,lobo-method}},
\code{\link{[[,met-method}},
\code{\link{[[,oce-method}},
\code{\link{[[,odf-method}},
\code{\link{[[,rsk-method}},
\code{\link{[[,sealevel-method}},
\code{\link{[[,section-method}},
\code{\link{[[,tidem-method}},
\code{\link{[[,topo-method}},
\code{\link{[[,windrose-method}},
\code{\link{[[,xbt-method}},
\code{\link{[[<-,adv-method}}

Other things related to coastline data: 
\code{\link{[[<-,coastline-method}},
\code{\link{as.coastline}()},
\code{\link{coastline-class}},
\code{\link{coastlineBest}()},
\code{\link{coastlineCut}()},
\code{\link{coastlineWorld}},
\code{\link{download.coastline}()},
\code{\link{plot,coastline-method}},
\code{\link{read.coastline.openstreetmap}()},
\code{\link{read.coastline.shapefile}()},
\code{\link{subset,coastline-method}},
\code{\link{summary,coastline-method}}
}
\author{
Dan Kelley
}
\concept{functions that extract parts of oce objects}
\concept{things related to coastline data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swTSrho}
\alias{swTSrho}
\title{Seawater temperature from salinity and density}
\usage{
swTSrho(
  salinity,
  density,
  pressure = NULL,
  eos = getOption("oceEOS", default = "gsw")
)
}
\arguments{
\item{salinity}{\emph{in-situ} salinity (PSU)}

\item{density}{\emph{in-situ} density or sigma value (kg/m\eqn{^3}{^3})}

\item{pressure}{\emph{in-situ} pressure (dbar)}

\item{eos}{equation of state to be used, either \code{"unesco"} or
\code{"gsw"} (ignored at present).}
}
\value{
\emph{In-situ} temperature in \eqn{^\circ}{deg}C on the ITS-90
scale.
}
\description{
Compute \emph{in-situ} temperature, given salinity, density, and pressure.
}
\details{
Finds the temperature that yields the given density, with the given salinity
and pressure.  The method is a bisection search with temperature tolerance
0.001 \eqn{^\circ}{deg}C.
}
\examples{
swTSrho(35, 23, 0, eos="unesco") # 26.11301

}
\references{
Fofonoff, P. and R. C. Millard Jr, 1983. Algorithms for
computation of fundamental properties of seawater. \emph{Unesco Technical
Papers in Marine Science}, \emph{44}, 53 pp

Gill, A.E., 1982. \emph{Atmosphere-ocean Dynamics}, Academic Press, New
York, 662 pp.
}
\seealso{
\code{\link[=swSTrho]{swSTrho()}}

Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swPressure}
\alias{swPressure}
\title{Water pressure}
\usage{
swPressure(depth, latitude = 45, eos = getOption("oceEOS", default = "gsw"))
}
\arguments{
\item{depth}{distance below the surface in metres.}

\item{latitude}{Latitude in \eqn{^\circ}{deg}N or radians north of the
equator.}

\item{eos}{indication of formulation to be used, either \code{"unesco"} or
\code{"gsw"}.}
}
\value{
Pressure in dbar.
}
\description{
Compute seawater pressure from depth by inverting \code{\link[=swDepth]{swDepth()}}
using \code{\link[=uniroot]{uniroot()}}.
}
\details{
If \code{eos="unesco"} this is done by numerical inversion of
\code{\link[=swDepth]{swDepth()}} is done using \code{\link[=uniroot]{uniroot()}}. If
\code{eos="gsw"}, it is done using \code{\link[gsw:gsw_p_from_z]{gsw::gsw_p_from_z()}} in the
\CRANpkg{gsw} package.
}
\examples{
swPressure(9712.653, 30, eos="unesco") # 10000
swPressure(9712.653, 30, eos="gsw")    #  9998.863

}
\references{
Unesco 1983. Algorithms for computation of fundamental
properties of seawater, 1983. \emph{Unesco Tech. Pap. in Mar. Sci.}, No. 44,
53 pp.
}
\seealso{
Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/accessors.R
\name{oceGetMetadata}
\alias{oceGetMetadata}
\title{Get Something From an oce metadata Slot}
\usage{
oceGetMetadata(object, name)
}
\arguments{
\item{object}{an \linkS4class{oce} object.}

\item{name}{String indicating the name of the item to be found.}
}
\description{
In contrast to the various \code{[[} functions, this is
guaranteed to look only within the \code{metadata} slot. If
the named item is not found, \code{NULL} is returned.
}
\seealso{
Other things related to the metadata slot: 
\code{\link{oceDeleteMetadata}()},
\code{\link{oceRenameMetadata}()},
\code{\link{oceSetMetadata}()}
}
\author{
Dan Kelley
}
\concept{things related to the metadata slot}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/met.R
\name{[[<-,met-method}
\alias{[[<-,met-method}
\title{Replace Parts of a met Object}
\usage{
\S4method{[[}{met}(x, i, j, ...) <- value
}
\arguments{
\item{x}{a \linkS4class{met} object.}

\item{i}{character value naming the item to replace.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{optional additional information (ignored).}

\item{value}{The value to be placed into \code{x}, somewhere.}
}
\description{
The \code{[[<-} method works for all \linkS4class{oce} objects.
The purpose, as with the related extraction method, \code{[[},
is to insulate users from the internal details of \linkS4class{oce}
objects, by looking for items within the various storage
slots of the object. Items not actually stored can also be
replaced, including units and data-quality
flags.
}
\details{
As with \code{[[} method, the procedure works in steps.

First, the \code{metadata} slot of \code{x} is checked to
see whether it contains something named with \code{i}.
If so, then the named item is replaced with \code{value}.

Otherwise, if the string value of \code{i} ends in \code{Unit}, then the
characters preceding that are taken as the name of a variable, and
the \code{metadata} slot of \code{x} is updated to store that unit, e.g.\preformatted{x[["temperatureUnits"]] <- list(unit=expression(degree*F),scale="")
}

Similarly, if \code{i} ends in \code{Flag}, then quality-control
flags are set up as defined by \code{result}, e.g.\preformatted{o[["temperatureFlags"]] <- c(2,4,2,2)
}

Otherwise, \code{\link[=pmatch]{pmatch()}} is used for a partial-string match with
the names of the items that are in the \code{data} slot of \code{x}.
The first item found (if any) is then updated to hold the value \code{result}.

If none of these conditions is met, a warning is issued.
}
\seealso{
Other functions that replace parts of oce objects: 
\code{\link{[[<-,adp-method}},
\code{\link{[[<-,amsr-method}},
\code{\link{[[<-,argo-method}},
\code{\link{[[<-,bremen-method}},
\code{\link{[[<-,cm-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{[[<-,echosounder-method}},
\code{\link{[[<-,g1sst-method}},
\code{\link{[[<-,gps-method}},
\code{\link{[[<-,ladp-method}},
\code{\link{[[<-,landsat-method}},
\code{\link{[[<-,lisst-method}},
\code{\link{[[<-,lobo-method}},
\code{\link{[[<-,oce-method}},
\code{\link{[[<-,odf-method}},
\code{\link{[[<-,rsk-method}},
\code{\link{[[<-,sealevel-method}},
\code{\link{[[<-,section-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{[[<-,topo-method}},
\code{\link{[[<-,windrose-method}},
\code{\link{[[<-,xbt-method}}

Other things related to met data: 
\code{\link{[[,met-method}},
\code{\link{as.met}()},
\code{\link{download.met}()},
\code{\link{met-class}},
\code{\link{met}},
\code{\link{plot,met-method}},
\code{\link{read.met}()},
\code{\link{subset,met-method}},
\code{\link{summary,met-method}},
\code{\link{test_met_csv1.csv}},
\code{\link{test_met_csv2.csv}},
\code{\link{test_met_xml2.xml}}
}
\concept{functions that replace parts of oce objects}
\concept{things related to met data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/AllClass.R
\name{concatenate}
\alias{concatenate}
\title{Concatenate oce objects}
\usage{
concatenate(object, ...)
}
\arguments{
\item{object}{an \linkS4class{oce} object.}

\item{...}{optional additional \linkS4class{oce} objects.}
}
\value{
An object of class corresponding to that of \code{object}.
}
\description{
Concatenate oce objects
}
\seealso{
Other functions that concatenate oce objects: 
\code{\link{concatenate,adp-method}},
\code{\link{concatenate,list-method}},
\code{\link{concatenate,oce-method}}
}
\concept{functions that concatenate oce objects}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{firstFinite}
\alias{firstFinite}
\title{Get first finite value in a vector or array, or NULL if none}
\usage{
firstFinite(v)
}
\arguments{
\item{v}{A numerical vector or array.}
}
\description{
Get first finite value in a vector or array, or NULL if none
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/coastline.R
\name{coastlineBest}
\alias{coastlineBest}
\title{Find the Name of the Best Coastline Object}
\usage{
coastlineBest(lonRange, latRange, span, debug = getOption("oceDebug"))
}
\arguments{
\item{lonRange}{range of longitude for locale}

\item{latRange}{range of latitude for locale}

\item{span}{span of domain in km (if provided, previous two arguments are
ignored).}

\item{debug}{set to a positive value to get debugging information during
processing.}
}
\value{
The name of a coastline that can be loaded with \code{data()}.
}
\description{
Find the name of the most appropriate coastline for a given locale
Checks \code{coastlineWorld}, \code{coastlineWorldFine} and
\code{coastlineWorldCoarse}, in that order, to find the one most appropriate
for the locale.
}
\seealso{
Other things related to coastline data: 
\code{\link{[[,coastline-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{as.coastline}()},
\code{\link{coastline-class}},
\code{\link{coastlineCut}()},
\code{\link{coastlineWorld}},
\code{\link{download.coastline}()},
\code{\link{plot,coastline-method}},
\code{\link{read.coastline.openstreetmap}()},
\code{\link{read.coastline.shapefile}()},
\code{\link{subset,coastline-method}},
\code{\link{summary,coastline-method}}
}
\author{
Dan Kelley
}
\concept{things related to coastline data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/tides.R
\docType{data}
\name{tidalCurrent}
\alias{tidalCurrent}
\title{Tidal Current Dataset}
\source{
The data come from the \code{tide8.dat} and \code{tide9.dat} files provided
at reference 2.
}
\description{
The \code{tidalCurrent} dataset contains tidal velocities reported in
Foreman's (1978) report (reference 1) on his Fortran code for the analysis of
tidal currents and provided in an associated webpage (reference 2).
Here, \code{tidalCurrent} is data frame containing
\itemize{
\item \code{time} a POSIXct time.
\item \code{u} the eastward component of velocity in m/s.
\item \code{v} the northward component of velocity in m/s.
}
}
\examples{
library(oce)
data(tidalCurrent)
par(mfrow=c(2, 1))
oce.plot.ts(tidalCurrent$time, tidalCurrent$u, ylab="u [m/s]")
abline(h=0, col=2)
oce.plot.ts(tidalCurrent$time, tidalCurrent$v, ylab="v [m/s]")
abline(h=0, col=2)

}
\references{
\enumerate{
\item Foreman, M. G. G. "Manual for Tidal Currents Analysis and Prediction."
Pacific Marine Science Report.
British Columbia, Canada: Institute of Ocean Sciences, Patricia Bay, 1978.
\item \code{https://www.dfo-mpo.gc.ca/science/documents/data-donnees/tidal-marees/tidpack.zip}
}
}
\seealso{
Other things related to tides: 
\code{\link{[[,tidem-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{as.tidem}()},
\code{\link{plot,tidem-method}},
\code{\link{predict.tidem}()},
\code{\link{summary,tidem-method}},
\code{\link{tidedata}},
\code{\link{tidem-class}},
\code{\link{tidemAstron}()},
\code{\link{tidemVuf}()},
\code{\link{tidem}},
\code{\link{webtide}()}
}
\author{
Dan Kelley (reformatting data provided by Michael Foreman)
}
\concept{things related to tides}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.R
\docType{data}
\name{ctd}
\alias{ctd}
\title{A CTD profile in Halifax Harbour}
\usage{
data(ctd)
}
\description{
This is a CTD profile measured in Halifax Harbour in 2003, based
on \code{\link[=ctdRaw]{ctdRaw()}}, but trimmed to just the downcast with
\code{\link[=ctdTrim]{ctdTrim()}}, using indices inferred by inspection of the
results from \code{\link[=plotScan]{plotScan()}}.
}
\details{
This station was sampled by students enrolled in the Dan Kelley's
Physical Oceanography class at Dalhousie University.
The data were acquired near the centre of the Bedford Basin of the
Halifax Harbour, during an October 2003 field trip of Dalhousie University's
Oceanography 4120/5120 class. The original \code{.cnv} data file had
temperature in the IPTS-68 scale, but this was converted to the more modern
scale using \code{\link[=T90fromT68]{T90fromT68()}}.
}
\examples{
\dontrun{
library(oce)
data(ctd)
plot(ctd)
}

}
\seealso{
The full profile (not trimmed to the downcast) is available as
\code{data(ctdRaw)}.

Other datasets provided with oce: 
\code{\link{adp}},
\code{\link{adv}},
\code{\link{amsr}},
\code{\link{argo}},
\code{\link{cm}},
\code{\link{coastlineWorld}},
\code{\link{ctdRaw}},
\code{\link{echosounder}},
\code{\link{landsat}},
\code{\link{lisst}},
\code{\link{lobo}},
\code{\link{met}},
\code{\link{ocecolors}},
\code{\link{rsk}},
\code{\link{sealevelTuktoyaktuk}},
\code{\link{sealevel}},
\code{\link{section}},
\code{\link{topoWorld}},
\code{\link{wind}},
\code{\link{xbt}}

Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}
}
\concept{datasets provided with oce}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/amsr.R
\name{[[<-,amsr-method}
\alias{[[<-,amsr-method}
\title{Replace Parts of an amsr Object}
\usage{
\S4method{[[}{amsr}(x, i, j, ...) <- value
}
\arguments{
\item{x}{an \linkS4class{amsr} object.}

\item{i}{character value naming the item to replace.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{optional additional information (ignored).}

\item{value}{The value to be placed into \code{x}, somewhere.}
}
\description{
The \code{[[<-} method works for all \linkS4class{oce} objects.
The purpose, as with the related extraction method, \code{[[},
is to insulate users from the internal details of \linkS4class{oce}
objects, by looking for items within the various storage
slots of the object. Items not actually stored can also be
replaced, including units and data-quality
flags.
}
\details{
As with \code{[[} method, the procedure works in steps.

First, the \code{metadata} slot of \code{x} is checked to
see whether it contains something named with \code{i}.
If so, then the named item is replaced with \code{value}.

Otherwise, if the string value of \code{i} ends in \code{Unit}, then the
characters preceding that are taken as the name of a variable, and
the \code{metadata} slot of \code{x} is updated to store that unit, e.g.\preformatted{x[["temperatureUnits"]] <- list(unit=expression(degree*F),scale="")
}

Similarly, if \code{i} ends in \code{Flag}, then quality-control
flags are set up as defined by \code{result}, e.g.\preformatted{o[["temperatureFlags"]] <- c(2,4,2,2)
}

Otherwise, \code{\link[=pmatch]{pmatch()}} is used for a partial-string match with
the names of the items that are in the \code{data} slot of \code{x}.
The first item found (if any) is then updated to hold the value \code{result}.

If none of these conditions is met, a warning is issued.
}
\seealso{
Other functions that replace parts of oce objects: 
\code{\link{[[<-,adp-method}},
\code{\link{[[<-,argo-method}},
\code{\link{[[<-,bremen-method}},
\code{\link{[[<-,cm-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{[[<-,echosounder-method}},
\code{\link{[[<-,g1sst-method}},
\code{\link{[[<-,gps-method}},
\code{\link{[[<-,ladp-method}},
\code{\link{[[<-,landsat-method}},
\code{\link{[[<-,lisst-method}},
\code{\link{[[<-,lobo-method}},
\code{\link{[[<-,met-method}},
\code{\link{[[<-,oce-method}},
\code{\link{[[<-,odf-method}},
\code{\link{[[<-,rsk-method}},
\code{\link{[[<-,sealevel-method}},
\code{\link{[[<-,section-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{[[<-,topo-method}},
\code{\link{[[<-,windrose-method}},
\code{\link{[[<-,xbt-method}}

Other things related to amsr data: 
\code{\link{[[,amsr-method}},
\code{\link{amsr-class}},
\code{\link{amsr}},
\code{\link{composite,amsr-method}},
\code{\link{download.amsr}()},
\code{\link{plot,amsr-method}},
\code{\link{read.amsr}()},
\code{\link{subset,amsr-method}},
\code{\link{summary,amsr-method}}
}
\concept{functions that replace parts of oce objects}
\concept{things related to amsr data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/bremen.R
\name{read.bremen}
\alias{read.bremen}
\title{Read a Bremen File}
\usage{
read.bremen(file)
}
\arguments{
\item{file}{a connection or a character string giving the name of the file
to load.}
}
\value{
A \linkS4class{bremen} object.
}
\description{
Read a file in Bremen format.
}
\details{
Velocities are assumed to be in
cm/s, and are converted to m/s to follow the oce convention. Shears
(which is what the variables named \code{uz} and \code{vz} are assumed
to represent) are assumed to be in (cm/s)/m, although they could be in 1/s
or something else; the lack of documentation is a problem here. Also,
note that the assumed shears are not just first-difference estimates
of velocity, given the results of a sample dataset:\preformatted{> head(data.frame(b[["data"]]))
  pressure      u      v       uz       vz
1        0  0.092 -0.191  0.00000  0.00000
2       10  0.092 -0.191  0.02183 -0.35412
3       20  0.092 -0.191  0.03046 -0.09458
4       30  0.026 -0.246 -0.03948  0.02169
5       40 -0.003 -0.212 -0.02614  0.03111
6       50 -0.023 -0.169 -0.03791  0.01706
}
}
\section{Issues}{
 This function may be renamed (or removed) without notice.
It was created to read some data being used in a particular research
project, and will be rendered useless if Bremen changes this data format.
}

\seealso{
Other things related to bremen data: 
\code{\link{[[,bremen-method}},
\code{\link{[[<-,bremen-method}},
\code{\link{bremen-class}},
\code{\link{plot,bremen-method}},
\code{\link{summary,bremen-method}}
}
\author{
Dan Kelley
}
\concept{things related to bremen data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{oceFilter}
\alias{oceFilter}
\alias{oce.filter}
\title{Filter a Time Series}
\usage{
oceFilter(x, a = 1, b, zero.phase = FALSE)
}
\arguments{
\item{x}{a vector of numeric values, to be filtered as a time series.}

\item{a}{a vector of numeric values, giving the \eqn{a}{a} coefficients (see
\dQuote{Details}).}

\item{b}{a vector of numeric values, giving the \eqn{b}{b} coefficients (see
\dQuote{Details}).}

\item{zero.phase}{boolean, set to \code{TRUE} to run the filter forwards,
and then backwards, thus removing any phase shifts associated with the
filter.}
}
\value{
A numeric vector of the filtered results, \eqn{y}{y}, as denoted in
\dQuote{Details}.
}
\description{
Filter a time-series, possibly recursively
}
\details{
The filter is defined as e.g.  \eqn{y[i]=b[1]*x[i] + }{y[i]=b[1]*x[i] +
b[2]*x[i-1] + b[3]*x[i-2] + ... - a[2]*y[i-1] - a[3]*y[i-2] - a[4]*y[i-3] -
...}\eqn{ b[2]*x[i-1] + b[3]*x[i-2] + ... - a[2]*y[i-1] - a[3]*y[i-2] -
}{y[i]=b[1]*x[i] + b[2]*x[i-1] + b[3]*x[i-2] + ... - a[2]*y[i-1] -
a[3]*y[i-2] - a[4]*y[i-3] - ...}\eqn{ a[4]*y[i-3] - ...}{y[i]=b[1]*x[i] +
b[2]*x[i-1] + b[3]*x[i-2] + ... - a[2]*y[i-1] - a[3]*y[i-2] - a[4]*y[i-3] -
...}, where some of the illustrated terms will be omitted if the lengths of
\code{a} and \code{b} are too small, and terms are dropped at the start of
the time series where the index on \code{x} would be less than 1.

By contrast with the \code{\link[=filter]{filter()}} function of R, \code{oce.filter}
lacks the option to do a circular filter.  As a consequence,
\code{oceFilter} introduces a phase lag.  One way to remove this lag is to
run the filter forwards and then backwards, as in the \dQuote{Examples}.
However, the result is still problematic, in the sense that applying it in
the reverse order would yield a different result.  (Matlab's \code{filtfilt}
shares this problem.)
}
\note{
The first value in the \code{a} vector is ignored, and if
\code{length(a)} equals 1, a non-recursive filter results.
}
\examples{
library(oce)
par(mar=c(4, 4, 1, 1))
b <- rep(1, 5)/5
a <- 1
x <- seq(0, 10)
y <- ifelse(x == 5, 1, 0)
f1 <- oceFilter(y, a, b)
plot(x, y, ylim=c(-0, 1.5), pch="o", type='b')
points(x, f1, pch="x", col="red")

# remove the phase lag
f2 <- oceFilter(y, a, b, TRUE)
points(x, f2, pch="+", col="blue")

legend("topleft", col=c("black","red","blue"), pch=c("o","x","+"),
       legend=c("data","normal filter", "zero-phase filter"))
mtext("note that normal filter rolls off at end")
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\docType{package}
\name{oce}
\alias{oce}
\title{oce: A Package for Oceanographic Analysis.}
\description{
The oce package provides functions for working with
Oceanographic data, for calculations that are specific
to Oceanography, and for producing graphics that
match the conventions of the field.
}
\details{
Over a dozen specialized data types are handled by oce,
with generic plots and summaries for each, along with
the specialized functions needed for typical Oceanographic
analysis.

See \linkS4class{oce} for a summary of the class structure
and links to documentation for the many subclasses of
oce objects, each aligned with a class of instrument or
or type of dataset.  For a more task-oriented approach,
see the several vignettes that are provided with oce,
and a book
(Kelley, Dan E. Oceanographic Analysis with R. New York: Springer-Verlag, 2018.
https://link.springer.com/us/book/9781493988426) written
by one of the oce co-authors.
}
\section{Specialized Functions}{

A key function is \code{\link[=read.oce]{read.oce()}}, which will attempt
to read Oceanographic data in raw format. This uses
\code{\link[=oceMagic]{oceMagic()}} to try to detect the file type,
based on the file name and contents. If this detection
is not possible, users will need to go beyond \code{\link[=read.oce]{read.oce()}},
\code{\link[=read.ctd.sbe]{read.ctd.sbe()}} for Teledyne-Seabird files, etc.
}

\section{Generic Methods}{

A list of the generic methods in oce is provided by
\link{methods}\code{(class="oce")}; a few that are used frequently
are as follows.
\itemize{
\item \code{[[} Finds the value of an item in the object's
\code{metadata} or \code{data} slot. If the item does
not exist, but can be calculated from the other items,
then the calculated value is returned. As an example of the
latter, consider the built-in \code{ctd} dataset, which does
not contain potential temperature, "\code{theta}". Using
\code{ctd[["theta"]]} therefore causes \code{\link[=swTheta]{swTheta()}}
to be called, to calculate \code{theta}.
See \link{[[,oce-method} or type \code{?"[[,oce-method"}
to learn more about general functioning, or a specialized
method like \link{[[,ctd-method} for CTD data, etc.
\item \verb{[[<-} Alters the named item in the object's \code{metadata} or
\code{data} slot.  If the item does not exist, it is created.
See \link{[[<-,oce-method} or type \code{?"[[<-,oce-method"}
to learn more about the general methodology, or a specialized
method like \link{[[<-,ctd-method} for CTD data, etc.
\item \code{summary()} Displays some information about the object named as an
argument, including a few elements from its \code{metadata} slot
and some statistics of the contents of its \code{data} slot.
See \link{summary,oce-method} or type \code{?"summary,oce-method"}
to learn more about general functioning, or a specialized
method like \link{summary,ctd-method} for CTD data, etc.
\item \code{subset()} Takes a subset of an oce object.
See \link{subset,oce-method} or type \code{?"subset,oce-method"}
to learn more about general functioning, or a specialized
method like \link{subset,ctd-method} for CTD data, etc.
}
}

% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/topo.R
\name{[[,topo-method}
\alias{[[,topo-method}
\title{Extract Something From a Topo Object}
\usage{
\S4method{[[}{topo}(x, i, j, ...)
}
\arguments{
\item{x}{a \linkS4class{topo} object.}

\item{i}{character value indicating the name of an item to extract.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{ignored.}
}
\description{
Generally, the \code{[[} method lets users extract information from \code{oce}
objects, without having to know the details of the internal storage.  For
many \code{oce} sub-classes, \code{[[} can also return quantities that are computed
from the object's contents.
}
\details{
A two-step process is used to try to find the requested information. First, a
class-specific function is used (see \dQuote{Details of the Specialized
Method}).  If this yields nothing, then a general method is used (see
\dQuote{Details of the General Method}).  If both methods fail, then \code{[[}
returns NULL.

Some understanding of the subclass is required to know what can be retrieved
with \code{[[}. When dealing with an unfamiliar subclass, it can be useful to
first use \code{x[["?"]]} to get a listing of the retrievable items. See
\dQuote{Details of the Specialized Method} for more information.
}
\section{Details of the Specialized Method}{

\itemize{
\item If \code{i} is \code{"?"}, then the return value is a list
containing four items, each of which is a character vector
holding the names of things that can be accessed with \code{[[}.
The \code{data} and \code{metadata} items hold the names of
entries in the object's data and metadata
slots, respectively. The \code{dataDerived}
and \code{metadataDerived} items are each NULL, because
no derived values are available for \code{topo} objects.
}
}

\section{Details of the General Method}{


Note: the text of this section is identical for all \code{oce} subclasses, and so
some of what you read here may not be relevant to the class being described
in this help page.

If the specialized method produces no matches, the following generalized
method is applied. As with the specialized method, the procedure hinges first
on the values of \code{i} and, optionally, \code{j}. The work proceeds in steps, by
testing a sequence of possible conditions in sequence.
\enumerate{
\item A check is made as to whether \code{i} names one of the standard \code{oce} slots.
If so, \code{[[} returns the slot contents of that slot.  Thus, \code{x[["metadata"]]}
will retrieve the \code{metadata} slot, while \code{x[["data"]]} and
\code{x[["processingLog"]]} return those slots.
\item If \code{i} is a string ending in the \code{"Unit"}, then the characters preceding
that string are taken to be the name of an item in the data object, and a
list containing the unit is returned (or \code{NULL} if there is no such unit).
This list consists of an item named \code{unit}, which is an \code{\link[=expression]{expression()}}, and
an item named \code{scale}, which is a string describing the measurement scale.
If the string ends in \code{" unit"}, e.g.  \code{x[["temperature unit"]]} (note the
space), then just the expression is returned, and if it ends in \code{" scale"},
then just the scale is returned.
\item If \code{i} is a string ending in \code{"Flag"}, then the corresponding data-quality
flag is returned (or \code{NULL} if there is no such flag).
\item If the object holds hydrographic information (pressure, salinity,
temperature, longitude and latitude) then another set of possibilities
arises.  If \code{i} is \code{"sigmaTheta"}, then the value of \code{\link[=swSigmaTheta]{swSigmaTheta()}} is
called with \code{x} as the sole argument, and the results are returned.
Similarly, \code{\link[=swSigma0]{swSigma0()}} is used if \code{i="sigma0"}, and \code{\link[=swSpice]{swSpice()}} is used if
\code{i="spice"}. Of course, these actions only make sense for objects that
contain the relevant items within their \code{data} slot.
\item After these possibilities are eliminated, the action depends on whether
\code{j} has been provided.  If \code{j} is not provided, or is the string \code{""}, then
\code{i} is sought in the \code{metadata} slot, and then in the \code{data} slot, returning
whichever is found first.  In other words, if \code{j} is not provided, the
\code{metadata} slot takes preference over the \code{data} slot. However, if \code{j} is
provided, then it must be either the string \code{"metadata"} or \code{"data"}, and it
directs where to look.
\item If none of the above-listed conditions holds, then \code{NULL} is returned.
}
}

\examples{
data(topoWorld)
dim(topoWorld[['z']])

}
\seealso{
Other functions that extract parts of oce objects: 
\code{\link{[[,adp-method}},
\code{\link{[[,adv-method}},
\code{\link{[[,amsr-method}},
\code{\link{[[,argo-method}},
\code{\link{[[,bremen-method}},
\code{\link{[[,cm-method}},
\code{\link{[[,coastline-method}},
\code{\link{[[,ctd-method}},
\code{\link{[[,echosounder-method}},
\code{\link{[[,g1sst-method}},
\code{\link{[[,gps-method}},
\code{\link{[[,ladp-method}},
\code{\link{[[,landsat-method}},
\code{\link{[[,lisst-method}},
\code{\link{[[,lobo-method}},
\code{\link{[[,met-method}},
\code{\link{[[,oce-method}},
\code{\link{[[,odf-method}},
\code{\link{[[,rsk-method}},
\code{\link{[[,sealevel-method}},
\code{\link{[[,section-method}},
\code{\link{[[,tidem-method}},
\code{\link{[[,windrose-method}},
\code{\link{[[,xbt-method}},
\code{\link{[[<-,adv-method}}

Other things related to topo data: 
\code{\link{[[<-,topo-method}},
\code{\link{as.topo}()},
\code{\link{download.topo}()},
\code{\link{plot,topo-method}},
\code{\link{read.topo}()},
\code{\link{subset,topo-method}},
\code{\link{summary,topo-method}},
\code{\link{topo-class}},
\code{\link{topoInterpolate}()},
\code{\link{topoWorld}}
}
\author{
Dan Kelley
}
\concept{functions that extract parts of oce objects}
\concept{things related to topo data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/section.R
\name{sectionSort}
\alias{sectionSort}
\title{Sort a Section}
\usage{
sectionSort(section, by)
}
\arguments{
\item{section}{A \code{section} object containing the section whose stations
are to be sorted.}

\item{by}{An optional string indicating how to reorder.  If not provided,
\code{"stationID"} will be assumed.  Other choices are \code{"distance"}, for
distance from the first station, \code{"longitude"}, for longitude,
\code{"latitude"} for latitude, and \code{"time"}, for time.}
}
\value{
object A \linkS4class{section} object that has been smoothed,
so its data fields will station-to-station variation than
is the case for the input section, \code{x}.
}
\description{
Sections created with \code{\link[=as.section]{as.section()}} have "stations" that are in the
order of the CTD objects (or filenames for such objects) provided.  Sometimes,
this is not the desired order, e.g. if file names discovered with
\code{\link[=dir]{dir()}} are in an order that makes no sense.  (For example, a
practioner might name stations \code{"stn1"}, \code{"stn2"}, etc., not
realizing that this will yield an unhelpful ordering, by file name, if there
are more than 9 stations.) The purpose of \code{sectionSort} is to permit
reordering the constituent stations in sensible ways.
}
\examples{
\dontrun{
# Eastern North Atlantic, showing Mediterranean water;
# sorting by longitude makes it easier to compare
# the map and the section.
library(oce)
data(section)
s <- sectionGrid(subset(section, -30 <= longitude))
ss <- sectionSort(ss, by="longitude")
plot(ss)
}

}
\seealso{
Other things related to section data: 
\code{\link{[[,section-method}},
\code{\link{[[<-,section-method}},
\code{\link{as.section}()},
\code{\link{handleFlags,section-method}},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{plot,section-method}},
\code{\link{read.section}()},
\code{\link{section-class}},
\code{\link{sectionAddStation}()},
\code{\link{sectionGrid}()},
\code{\link{sectionSmooth}()},
\code{\link{section}},
\code{\link{subset,section-method}},
\code{\link{summary,section-method}}
}
\author{
Dan Kelley
}
\concept{things related to section data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{oceColors9B}
\alias{oceColors9B}
\alias{oce.colors9B}
\title{Create colors in a red-yellow-blue color scheme}
\usage{
oceColors9B(n)
}
\arguments{
\item{n}{number of colors}
}
\description{
The results are similar to those of \code{\link[=oceColorsJet]{oceColorsJet()}}, but
with white hues in the centre, rather than green ones. The scheme
may be useful in displaying signed quantities, and thus is somewhat
analogous to \code{\link[=oceColorsTwo]{oceColorsTwo()}}, except that they (average)
eye may be more able to distinguish colors with \code{oceColors9B}.
}
\examples{
library(oce)
imagep(volcano, col=oceColors9B(128),
       zlab="oceColors9B")

}
\references{
The following references provide information on choosing
colour schemes, that are suitable for viewers who have
colour deficiencies.

Light, Adam, and Patrick J. Bartlein. "The End of the Rainbow? Color
Schemes for Improved Data Graphics." \emph{Eos, Transactions American Geophysical
Union} 85, no. 40 (2004): 385.
DOI: 10.1029/2004EO400002

Stephenson, David B. "Comment on 'Color Schemes for Improved Data
Graphics', by A Light and P.J. Bartlein." \emph{Eos, Transactions American
Geophysical Union} 86, no.  20 (2005): 196.
DOI: 10.1029/2005EO200005

Light, Adam, and Patrick J. Bartlein. "Reply to 'Comment on Color
Schemes for Improved Data Graphics,' by A. Light and P.J.  Bartlein'."
\emph{Eos, Transactions American Geophysical Union} 86, no. 20 (2005): 196‚Äì196.
DOI: 10.1029/2005EO200006
}
\seealso{
Other things related to colors: 
\code{\link{colormapGMT}()},
\code{\link{colormap}()},
\code{\link{oceColorsCDOM}()},
\code{\link{oceColorsChlorophyll}()},
\code{\link{oceColorsClosure}()},
\code{\link{oceColorsDensity}()},
\code{\link{oceColorsFreesurface}()},
\code{\link{oceColorsGebco}()},
\code{\link{oceColorsJet}()},
\code{\link{oceColorsOxygen}()},
\code{\link{oceColorsPAR}()},
\code{\link{oceColorsPalette}()},
\code{\link{oceColorsPhase}()},
\code{\link{oceColorsSalinity}()},
\code{\link{oceColorsTemperature}()},
\code{\link{oceColorsTurbidity}()},
\code{\link{oceColorsTurbo}()},
\code{\link{oceColorsTwo}()},
\code{\link{oceColorsVelocity}()},
\code{\link{oceColorsViridis}()},
\code{\link{oceColorsVorticity}()},
\code{\link{ocecolors}}
}
\concept{things related to colors}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/map.R
\name{mapLongitudeLatitudeXY}
\alias{mapLongitudeLatitudeXY}
\title{Convert From Longitude and Latitude to X and Y}
\usage{
mapLongitudeLatitudeXY(longitude, latitude)
}
\arguments{
\item{longitude}{numeric vector of the longitudes of points, or an object from which
both latitude and longitude can be inferred (e.g. a coastline file, or the
return value from \code{\link[=mapLocator]{mapLocator()}}), in which case the following
two arguments are ignored.}

\item{latitude}{numeric vector of latitudes of points, needed only if they cannot
be inferred from the first argument.}
}
\value{
A list containing \code{x} and \code{y}.
}
\description{
Find (x, y) values corresponding to (longitude, latitude) values, using the
present projection.
}
\details{
This is mainly a wrapper around \code{\link[=lonlat2map]{lonlat2map()}}.
}
\examples{
\donttest{
library(oce)
data(coastlineWorld)
par(mfrow=c(2, 1), mar=rep(2, 4))
mapPlot(coastlineWorld, projection="+proj=moll") # sets a projection
xy <- mapLongitudeLatitudeXY(coastlineWorld)
plot(xy, type='l', asp=1)
}

}
\seealso{
A map must first have been created with \code{\link[=mapPlot]{mapPlot()}}.

Other functions related to maps: 
\code{\link{formatPosition}()},
\code{\link{lonlat2map}()},
\code{\link{lonlat2utm}()},
\code{\link{map2lonlat}()},
\code{\link{mapArrows}()},
\code{\link{mapAxis}()},
\code{\link{mapContour}()},
\code{\link{mapCoordinateSystem}()},
\code{\link{mapDirectionField}()},
\code{\link{mapGrid}()},
\code{\link{mapImage}()},
\code{\link{mapLines}()},
\code{\link{mapLocator}()},
\code{\link{mapPlot}()},
\code{\link{mapPoints}()},
\code{\link{mapPolygon}()},
\code{\link{mapScalebar}()},
\code{\link{mapText}()},
\code{\link{mapTissot}()},
\code{\link{oceCRS}()},
\code{\link{shiftLongitude}()},
\code{\link{usrLonLat}()},
\code{\link{utm2lonlat}()}
}
\author{
Dan Kelley
}
\concept{functions related to maps}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/argo.R
\name{[[,argo-method}
\alias{[[,argo-method}
\title{Extract Something From an Argo Object}
\usage{
\S4method{[[}{argo}(x, i, j, ...)
}
\arguments{
\item{x}{an \linkS4class{argo} object.}

\item{i}{character value indicating the name of an item to extract.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{ignored.}
}
\description{
Generally, the \code{[[} method lets users extract information from \code{oce}
objects, without having to know the details of the internal storage.  For
many \code{oce} sub-classes, \code{[[} can also return quantities that are computed
from the object's contents.
}
\details{
A two-step process is used to try to find the requested information. First, a
class-specific function is used (see \dQuote{Details of the Specialized
Method}).  If this yields nothing, then a general method is used (see
\dQuote{Details of the General Method}).  If both methods fail, then \code{[[}
returns NULL.

Some understanding of the subclass is required to know what can be retrieved
with \code{[[}. When dealing with an unfamiliar subclass, it can be useful to
first use \code{x[["?"]]} to get a listing of the retrievable items. See
\dQuote{Details of the Specialized Method} for more information.
}
\section{Details of the Specialized Method}{


Note that \linkS4class{argo} data may contain both unadjusted data and adjusted
data.  By default, this extraction function refers to the former, but a
preference for the latter may be set with \code{\link[=preferAdjusted]{preferAdjusted()}}, the
documentation of which explains (fairly complex) details.

The results from \code{argo[[i]]} or \code{argo[[i,j]]} depend on the
nature of \code{i} and (if provided) \code{j}. The details are as follows.
\itemize{
\item If \code{i} is \code{"?"}, then the return value is a list
containing four items, each of which is a character vector
holding the names of things that can be accessed with \code{[[}.
The \code{data} and \code{metadata} items hold the names of
entries in the object's data and metadata
slots, respectively. The \code{dataDerived}
and \code{metadataDerived} items hold the names of things
that can be inferred from the object's contents, e.g.
\code{"SA"} is named in \code{dataDerived}, indicating that
\code{argo[["SA"]]} is permitted (to compute Absolute Salinity).
\item If \code{i} is \code{"profile"} and \code{j} is an integer vector,
then an argo object is returned, as specified by \code{j}. For example,
\code{argo[["profile", 2:5]]} is equivalent to
\code{subset(argo, profile \%in\% 2:5)}.
\item If \code{i} is \code{"CT"}, then
Conservative Temperature is returned, as computed with
\code{\link[gsw:gsw_CT_from_t]{gsw::gsw_CT_from_t}}\verb{(SA,t,p)}, where
first \code{SA} is computed as explained
in the next item, \code{t} is in-situ temperature,
and \code{p} is pressure.
\item If \code{i} is \code{"N2"}, then the square of buoyancy is returned,
as computed with \code{\link[=swN2]{swN2()}}.
\item If \code{i} is \code{"SA"}, then
Absolute Salinity is returned, as computed with
\code{\link[gsw:gsw_SA_from_SP]{gsw::gsw_SA_from_SP()}}.
\item If \code{i} is \code{"sigmaTheta"}, then
potential density anomaly (referenced to zero
pressure) is computed, with \code{\link[=swSigmaTheta]{swSigmaTheta()}}, where the
equation of state is taken to be
\code{\link{getOption}}\verb{("oceEOS",default="gsw")}.
\item If \code{i} is \code{"sigma0"}, \code{"sigma1"}, \code{"sigma2"}, \code{"sigma3"} or \code{"sigma4"},
then the associated function in the \CRANpkg{gsw} package.
For example, \code{"sigma0"} uses \code{\link[gsw:gsw_sigma0]{gsw::gsw_sigma0()}}, which returns
potential density anomaly referenced to 0 dbar,
according to the gsw equation of state.
\item If \code{i} is \code{"theta"}, then
potential temperature (referenced to zero
pressure) is computed, with \code{\link[=swTheta]{swTheta()}}, where the
equation of state is taken to be
\code{\link{getOption}}\verb{("oceEOS",default="gsw")}.
\item If \code{i} is \code{"depth"}, then
a matrix of depths is returned.
\item If \code{i} is \code{"id"} or \code{"ID"}, then the \code{id} element within
the \code{metadata} slot is returned.
\item If \code{i} is in the \code{data} slot of \code{x},
then it is returned, otherwise if it is in the \code{metadata} slot,
then that is returned, otherwise \code{NULL} is returned.
}
}

\section{Details of the General Method}{


Note: the text of this section is identical for all \code{oce} subclasses, and so
some of what you read here may not be relevant to the class being described
in this help page.

If the specialized method produces no matches, the following generalized
method is applied. As with the specialized method, the procedure hinges first
on the values of \code{i} and, optionally, \code{j}. The work proceeds in steps, by
testing a sequence of possible conditions in sequence.
\enumerate{
\item A check is made as to whether \code{i} names one of the standard \code{oce} slots.
If so, \code{[[} returns the slot contents of that slot.  Thus, \code{x[["metadata"]]}
will retrieve the \code{metadata} slot, while \code{x[["data"]]} and
\code{x[["processingLog"]]} return those slots.
\item If \code{i} is a string ending in the \code{"Unit"}, then the characters preceding
that string are taken to be the name of an item in the data object, and a
list containing the unit is returned (or \code{NULL} if there is no such unit).
This list consists of an item named \code{unit}, which is an \code{\link[=expression]{expression()}}, and
an item named \code{scale}, which is a string describing the measurement scale.
If the string ends in \code{" unit"}, e.g.  \code{x[["temperature unit"]]} (note the
space), then just the expression is returned, and if it ends in \code{" scale"},
then just the scale is returned.
\item If \code{i} is a string ending in \code{"Flag"}, then the corresponding data-quality
flag is returned (or \code{NULL} if there is no such flag).
\item If the object holds hydrographic information (pressure, salinity,
temperature, longitude and latitude) then another set of possibilities
arises.  If \code{i} is \code{"sigmaTheta"}, then the value of \code{\link[=swSigmaTheta]{swSigmaTheta()}} is
called with \code{x} as the sole argument, and the results are returned.
Similarly, \code{\link[=swSigma0]{swSigma0()}} is used if \code{i="sigma0"}, and \code{\link[=swSpice]{swSpice()}} is used if
\code{i="spice"}. Of course, these actions only make sense for objects that
contain the relevant items within their \code{data} slot.
\item After these possibilities are eliminated, the action depends on whether
\code{j} has been provided.  If \code{j} is not provided, or is the string \code{""}, then
\code{i} is sought in the \code{metadata} slot, and then in the \code{data} slot, returning
whichever is found first.  In other words, if \code{j} is not provided, the
\code{metadata} slot takes preference over the \code{data} slot. However, if \code{j} is
provided, then it must be either the string \code{"metadata"} or \code{"data"}, and it
directs where to look.
\item If none of the above-listed conditions holds, then \code{NULL} is returned.
}
}

\examples{
data(argo)
# 1. show that dataset has 223 profiles, each with 56 levels
dim(argo[['temperature']])

# 2. show importance of focussing on data flagged 'good'
fivenum(argo[["salinity"]],na.rm=TRUE)
fivenum(argo[["salinity"]][argo[["salinityFlag"]]==1],na.rm=TRUE)

}
\seealso{
Other functions that extract parts of oce objects: 
\code{\link{[[,adp-method}},
\code{\link{[[,adv-method}},
\code{\link{[[,amsr-method}},
\code{\link{[[,bremen-method}},
\code{\link{[[,cm-method}},
\code{\link{[[,coastline-method}},
\code{\link{[[,ctd-method}},
\code{\link{[[,echosounder-method}},
\code{\link{[[,g1sst-method}},
\code{\link{[[,gps-method}},
\code{\link{[[,ladp-method}},
\code{\link{[[,landsat-method}},
\code{\link{[[,lisst-method}},
\code{\link{[[,lobo-method}},
\code{\link{[[,met-method}},
\code{\link{[[,oce-method}},
\code{\link{[[,odf-method}},
\code{\link{[[,rsk-method}},
\code{\link{[[,sealevel-method}},
\code{\link{[[,section-method}},
\code{\link{[[,tidem-method}},
\code{\link{[[,topo-method}},
\code{\link{[[,windrose-method}},
\code{\link{[[,xbt-method}},
\code{\link{[[<-,adv-method}}

Other things related to argo data: 
\code{\link{[[<-,argo-method}},
\code{\link{argo-class}},
\code{\link{argoGrid}()},
\code{\link{argoNames2oceNames}()},
\code{\link{argo}},
\code{\link{as.argo}()},
\code{\link{handleFlags,argo-method}},
\code{\link{plot,argo-method}},
\code{\link{read.argo}()},
\code{\link{subset,argo-method}},
\code{\link{summary,argo-method}}
}
\author{
Dan Kelley
}
\concept{functions that extract parts of oce objects}
\concept{things related to argo data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{matrixShiftLongitude}
\alias{matrixShiftLongitude}
\title{Rearrange areal matrix so Greenwich is near the centre}
\usage{
matrixShiftLongitude(m, longitude)
}
\arguments{
\item{m}{The matrix to be modified.}

\item{longitude}{A vector containing the longitude in the 0-360 convention. If missing, this is constructed to range from 0 to 360, with as many elements as the first index of \code{m}.}
}
\value{
A list containing \code{m} and \code{longitude}, both rearranged as appropriate.
}
\description{
Sometimes datasets are provided in matrix form, with first
index corresponding to longitudes ranging from 0 to 360.
\code{matrixShiftLongitude} cuts such matrices at
longitude=180, and swaps the pieces so that the dateline
is at the left of the matrix, not in the middle.
}
\seealso{
\code{\link[=shiftLongitude]{shiftLongitude()}} and \code{\link[=standardizeLongitude]{standardizeLongitude()}}.
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/map.R
\name{mapScalebar}
\alias{mapScalebar}
\title{Add a Scalebar to a Map}
\usage{
mapScalebar(
  x,
  y = NULL,
  length,
  lwd = 1.5 * par("lwd"),
  cex = par("cex"),
  col = "black"
)
}
\arguments{
\item{x, y}{position of the scalebar.  Eventually this may be similar to
the corresponding arguments in \code{\link[=legend]{legend()}}, but at the moment
\code{y} must be \code{NULL} and \code{x} must be \code{"topleft"} or \code{"topright"}.}

\item{length}{the distance to indicate, in kilometres.  If not provided, a
reasonable choice is made, based on the existing plot.}

\item{lwd}{line width of the scalebar.}

\item{cex}{character expansion factor for the scalebar text.}

\item{col}{color of the scalebar.}
}
\description{
Draw a scalebar on a map created by \code{\link[=mapPlot]{mapPlot()}} or otherwise.
}
\details{
The scale is appropriate to the centre of the plot, and will become
increasingly inaccurate away from that spot, with the error depending on
the projection and the fraction of the earth that is shown.

Until December 2020, it was required that the map had been drawn by \code{\link[=mapPlot]{mapPlot()}},
but now it can be any diagram showing longitude and latitude in degrees.
}
\examples{
\donttest{
library(oce)
data(coastlineWorld)
## Arctic Ocean
par(mar=c(2.5, 2.5, 1, 1))
mapPlot(coastlineWorld, latitudelim=c(60, 120), longitudelim=c(-130,-50),
        col="lightgray", projection="+proj=stere +lat_0=90")
mapScalebar()
}

}
\seealso{
A map must first have been created with \code{\link[=mapPlot]{mapPlot()}}.

Other functions related to maps: 
\code{\link{formatPosition}()},
\code{\link{lonlat2map}()},
\code{\link{lonlat2utm}()},
\code{\link{map2lonlat}()},
\code{\link{mapArrows}()},
\code{\link{mapAxis}()},
\code{\link{mapContour}()},
\code{\link{mapCoordinateSystem}()},
\code{\link{mapDirectionField}()},
\code{\link{mapGrid}()},
\code{\link{mapImage}()},
\code{\link{mapLines}()},
\code{\link{mapLocator}()},
\code{\link{mapLongitudeLatitudeXY}()},
\code{\link{mapPlot}()},
\code{\link{mapPoints}()},
\code{\link{mapPolygon}()},
\code{\link{mapText}()},
\code{\link{mapTissot}()},
\code{\link{oceCRS}()},
\code{\link{shiftLongitude}()},
\code{\link{usrLonLat}()},
\code{\link{utm2lonlat}()}
}
\author{
Dan Kelley
}
\concept{functions related to maps}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/extdata.R
\docType{data}
\encoding{UTF-8}
\name{xbt.edf}
\alias{xbt.edf}
\title{Sample xbt dataset}
\description{
Sample xbt dataset
}
\examples{
\dontrun{
xbt <- read.oce(system.file("extdata", "xbt.edf", package="oce"))
}

}
\seealso{
Other raw datasets: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{adp_rdi.000}},
\code{\link{ctd.cnv}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{test_met_csv1.csv}},
\code{\link{test_met_csv2.csv}},
\code{\link{test_met_xml2.xml}}

Other things related to xbt data: 
\code{\link{[[,xbt-method}},
\code{\link{[[<-,xbt-method}},
\code{\link{as.xbt}()},
\code{\link{plot,xbt-method}},
\code{\link{read.xbt.noaa1}()},
\code{\link{read.xbt}()},
\code{\link{subset,xbt-method}},
\code{\link{summary,xbt-method}},
\code{\link{xbt-class}},
\code{\link{xbt}}
}
\concept{raw datasets}
\concept{things related to xbt data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.woce.R
\name{woceUnit2oceUnit}
\alias{woceUnit2oceUnit}
\title{Translate WOCE units to oce units}
\usage{
woceUnit2oceUnit(woceUnit)
}
\arguments{
\item{woceUnit}{string holding a WOCE unit}
}
\value{
expression in oce unit form
}
\description{
Translate WOCE-style units to \code{oce} units.
}
\seealso{
Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{write.ctd}()}

Other functions that interpret variable names and units from headers: 
\code{\link{ODFNames2oceNames}()},
\code{\link{cnvName2oceName}()},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{unitFromStringRsk}()},
\code{\link{unitFromString}()},
\code{\link{woceNames2oceNames}()}
}
\author{
Dan Kelley
}
\concept{functions that interpret variable names and units from headers}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/cm.R
\name{[[,cm-method}
\alias{[[,cm-method}
\title{Extract Something From a cm Object}
\usage{
\S4method{[[}{cm}(x, i, j, ...)
}
\arguments{
\item{x}{a \linkS4class{cm} object.}

\item{i}{character value indicating the name of an item to extract.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{ignored.}
}
\description{
Generally, the \code{[[} method lets users extract information from \code{oce}
objects, without having to know the details of the internal storage.  For
many \code{oce} sub-classes, \code{[[} can also return quantities that are computed
from the object's contents.
}
\details{
A two-step process is used to try to find the requested information. First, a
class-specific function is used (see \dQuote{Details of the Specialized
Method}).  If this yields nothing, then a general method is used (see
\dQuote{Details of the General Method}).  If both methods fail, then \code{[[}
returns NULL.

Some understanding of the subclass is required to know what can be retrieved
with \code{[[}. When dealing with an unfamiliar subclass, it can be useful to
first use \code{x[["?"]]} to get a listing of the retrievable items. See
\dQuote{Details of the Specialized Method} for more information.
}
\section{Details of the Specialized Method}{

\itemize{
\item If \code{i} is \code{"?"}, then the return value is a list
containing four items, each of which is a character vector
holding the names of things that can be accessed with \code{[[}.
The \code{data} and \code{metadata} items hold the names of
entries in the object's data and metadata
slots, respectively. The \code{dataDerived}
and \code{metadataDerived} items are each NULL, because
no derived values are defined by \linkS4class{cm} objects.
}
}

\section{Details of the General Method}{


Note: the text of this section is identical for all \code{oce} subclasses, and so
some of what you read here may not be relevant to the class being described
in this help page.

If the specialized method produces no matches, the following generalized
method is applied. As with the specialized method, the procedure hinges first
on the values of \code{i} and, optionally, \code{j}. The work proceeds in steps, by
testing a sequence of possible conditions in sequence.
\enumerate{
\item A check is made as to whether \code{i} names one of the standard \code{oce} slots.
If so, \code{[[} returns the slot contents of that slot.  Thus, \code{x[["metadata"]]}
will retrieve the \code{metadata} slot, while \code{x[["data"]]} and
\code{x[["processingLog"]]} return those slots.
\item If \code{i} is a string ending in the \code{"Unit"}, then the characters preceding
that string are taken to be the name of an item in the data object, and a
list containing the unit is returned (or \code{NULL} if there is no such unit).
This list consists of an item named \code{unit}, which is an \code{\link[=expression]{expression()}}, and
an item named \code{scale}, which is a string describing the measurement scale.
If the string ends in \code{" unit"}, e.g.  \code{x[["temperature unit"]]} (note the
space), then just the expression is returned, and if it ends in \code{" scale"},
then just the scale is returned.
\item If \code{i} is a string ending in \code{"Flag"}, then the corresponding data-quality
flag is returned (or \code{NULL} if there is no such flag).
\item If the object holds hydrographic information (pressure, salinity,
temperature, longitude and latitude) then another set of possibilities
arises.  If \code{i} is \code{"sigmaTheta"}, then the value of \code{\link[=swSigmaTheta]{swSigmaTheta()}} is
called with \code{x} as the sole argument, and the results are returned.
Similarly, \code{\link[=swSigma0]{swSigma0()}} is used if \code{i="sigma0"}, and \code{\link[=swSpice]{swSpice()}} is used if
\code{i="spice"}. Of course, these actions only make sense for objects that
contain the relevant items within their \code{data} slot.
\item After these possibilities are eliminated, the action depends on whether
\code{j} has been provided.  If \code{j} is not provided, or is the string \code{""}, then
\code{i} is sought in the \code{metadata} slot, and then in the \code{data} slot, returning
whichever is found first.  In other words, if \code{j} is not provided, the
\code{metadata} slot takes preference over the \code{data} slot. However, if \code{j} is
provided, then it must be either the string \code{"metadata"} or \code{"data"}, and it
directs where to look.
\item If none of the above-listed conditions holds, then \code{NULL} is returned.
}
}

\seealso{
Other functions that extract parts of oce objects: 
\code{\link{[[,adp-method}},
\code{\link{[[,adv-method}},
\code{\link{[[,amsr-method}},
\code{\link{[[,argo-method}},
\code{\link{[[,bremen-method}},
\code{\link{[[,coastline-method}},
\code{\link{[[,ctd-method}},
\code{\link{[[,echosounder-method}},
\code{\link{[[,g1sst-method}},
\code{\link{[[,gps-method}},
\code{\link{[[,ladp-method}},
\code{\link{[[,landsat-method}},
\code{\link{[[,lisst-method}},
\code{\link{[[,lobo-method}},
\code{\link{[[,met-method}},
\code{\link{[[,oce-method}},
\code{\link{[[,odf-method}},
\code{\link{[[,rsk-method}},
\code{\link{[[,sealevel-method}},
\code{\link{[[,section-method}},
\code{\link{[[,tidem-method}},
\code{\link{[[,topo-method}},
\code{\link{[[,windrose-method}},
\code{\link{[[,xbt-method}},
\code{\link{[[<-,adv-method}}

Other things related to cm data: 
\code{\link{[[<-,cm-method}},
\code{\link{as.cm}()},
\code{\link{cm-class}},
\code{\link{cm}},
\code{\link{plot,cm-method}},
\code{\link{read.cm}()},
\code{\link{rotateAboutZ}()},
\code{\link{subset,cm-method}},
\code{\link{summary,cm-method}}
}
\author{
Dan Kelley
}
\concept{functions that extract parts of oce objects}
\concept{things related to cm data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{gravity}
\alias{gravity}
\title{Acceleration due to earth gravity}
\usage{
gravity(latitude = 45, degrees = TRUE)
}
\arguments{
\item{latitude}{Latitude in \eqn{^\circ}{deg}N or radians north of the
equator.}

\item{degrees}{Flag indicating whether degrees are used for latitude; if set
to \code{FALSE}, radians are used.}
}
\value{
Acceleration due to gravity, in \eqn{m^2/s}{m^2/s}.
}
\description{
Compute \eqn{g}{g}, the acceleration due to gravity, as a function of
latitude.
}
\details{
Value not verified yet, except roughly.
}
\examples{
g <- gravity(45) # 9.8
}
\references{
Gill, A.E., 1982. \emph{Atmosphere-ocean Dynamics}, Academic
Press, New York, 662 pp.

\strong{Caution:} Fofonoff and Millard (1983 UNESCO) use a different formula.
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swSoundSpeed}
\alias{swSoundSpeed}
\title{Seawater sound speed}
\usage{
swSoundSpeed(
  salinity,
  temperature = NULL,
  pressure = NULL,
  longitude = NULL,
  latitude = NULL,
  eos = getOption("oceEOS", default = "gsw")
)
}
\arguments{
\item{salinity}{either practical salinity (in which case \code{temperature}
and \code{pressure} must be provided) \emph{or} an \code{oce} object, in
which case \code{salinity}, \code{temperature} (in the ITS-90 scale; see
next item), etc. are inferred from the object, ignoring the
other parameters, if they are supplied.}

\item{temperature}{\emph{in-situ} temperature (\eqn{^\circ}{deg}C), defined
on the ITS-90 scale.  This scale is used by GSW-style calculation (as
requested by setting \code{eos="gsw"}), and is the value contained within
\code{ctd} objects (and probably most other objects created with data
acquired in the past decade or two). Since the UNESCO-style calculation is
based on IPTS-68, the temperature is converted within the present function,
using \code{\link[=T68fromT90]{T68fromT90()}}.}

\item{pressure}{pressure (dbar)}

\item{longitude}{longitude of observation (only used if \code{eos="gsw"};
see \sQuote{Details}).}

\item{latitude}{latitude of observation (only used if \code{eos="gsw"}; see
\sQuote{Details}).}

\item{eos}{equation of state, either \code{"unesco"} (references 1 and 2)
or \code{"gsw"} (references 3 and 4).}
}
\value{
Sound speed (m/s).
}
\description{
Compute the seawater speed of sound.
}
\details{
If \code{eos="unesco"}, the sound speed is calculated using the formulation
in section 9 of Fofonoff and Millard (1983). If \code{eos="gsw"}, then the
\code{\link[gsw:gsw_sound_speed]{gsw::gsw_sound_speed()}} function from the
\CRANpkg{gsw} package is used.
}
\examples{
swSoundSpeed(40, T90fromT68(40), 10000) # 1731.995 (p48 of Fofonoff + Millard 1983)

}
\references{
Fofonoff, P. and R. C. Millard Jr, 1983. Algorithms for
computation of fundamental properties of seawater.
\emph{Unesco Technical Papers in Marine Science},
\emph{44}, 53 pp.  (See section 9.)
}
\seealso{
Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adv.R
\name{summary,adv-method}
\alias{summary,adv-method}
\title{Summarize an ADV object}
\usage{
\S4method{summary}{adv}(object, ...)
}
\arguments{
\item{object}{an object of class \code{"adv"}, usually, a result of a call to
\code{\link[=read.adv]{read.adv()}}.}

\item{...}{further arguments passed to or from other methods.}
}
\description{
Summarize data in an \code{adv} object.
}
\examples{
library(oce)
data(adv)
summary(adv)

}
\seealso{
Other things related to adv data: 
\code{\link{[[,adv-method}},
\code{\link{[[<-,adv-method}},
\code{\link{adv-class}},
\code{\link{adv}},
\code{\link{beamName}()},
\code{\link{beamToXyz}()},
\code{\link{enuToOtherAdv}()},
\code{\link{enuToOther}()},
\code{\link{plot,adv-method}},
\code{\link{read.adv.nortek}()},
\code{\link{read.adv.sontek.adr}()},
\code{\link{read.adv.sontek.serial}()},
\code{\link{read.adv.sontek.text}()},
\code{\link{read.adv}()},
\code{\link{rotateAboutZ}()},
\code{\link{subset,adv-method}},
\code{\link{toEnuAdv}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdv}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley
}
\concept{things related to adv data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.R
\name{beamName}
\alias{beamName}
\title{Get names of Acoustic-Doppler Beams}
\usage{
beamName(x, which)
}
\arguments{
\item{x}{an \linkS4class{adp} object.}

\item{which}{an integer indicating beam number.}
}
\value{
A character string containing a reasonable name for the beam, of the
form \code{"beam 1"}, etc., for beam coordinates, \code{"east"}, etc. for
enu coordinates, \code{"u"}, etc. for \code{"xyz"}, or \code{"u'"}, etc.,
for \code{"other"} coordinates.  The coordinate system is determined
with \code{x[["coordinate"]]}.
}
\description{
Get names of Acoustic-Doppler Beams
}
\seealso{
This is used by \code{\link[=read.oce]{read.oce()}}.

Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}

Other things related to adv data: 
\code{\link{[[,adv-method}},
\code{\link{[[<-,adv-method}},
\code{\link{adv-class}},
\code{\link{adv}},
\code{\link{beamToXyz}()},
\code{\link{enuToOtherAdv}()},
\code{\link{enuToOther}()},
\code{\link{plot,adv-method}},
\code{\link{read.adv.nortek}()},
\code{\link{read.adv.sontek.adr}()},
\code{\link{read.adv.sontek.serial}()},
\code{\link{read.adv.sontek.text}()},
\code{\link{read.adv}()},
\code{\link{rotateAboutZ}()},
\code{\link{subset,adv-method}},
\code{\link{summary,adv-method}},
\code{\link{toEnuAdv}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdv}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley
}
\concept{things related to adp data}
\concept{things related to adv data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/lisst.R
\name{[[,lisst-method}
\alias{[[,lisst-method}
\title{Extract Something From a LISST Object}
\usage{
\S4method{[[}{lisst}(x, i, j, ...)
}
\arguments{
\item{x}{a \linkS4class{lisst} object.}

\item{i}{character value indicating the name of an item to extract.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{ignored.}
}
\description{
Generally, the \code{[[} method lets users extract information from \code{oce}
objects, without having to know the details of the internal storage.  For
many \code{oce} sub-classes, \code{[[} can also return quantities that are computed
from the object's contents.
}
\details{
A two-step process is used to try to find the requested information. First, a
class-specific function is used (see \dQuote{Details of the Specialized
Method}).  If this yields nothing, then a general method is used (see
\dQuote{Details of the General Method}).  If both methods fail, then \code{[[}
returns NULL.

Some understanding of the subclass is required to know what can be retrieved
with \code{[[}. When dealing with an unfamiliar subclass, it can be useful to
first use \code{x[["?"]]} to get a listing of the retrievable items. See
\dQuote{Details of the Specialized Method} for more information.
}
\section{Details of the Specialized Method}{

\itemize{
\item If \code{i} is \code{"?"}, then the return value is a list
containing four items, each of which is a character vector
holding the names of things that can be accessed with \code{[[}.
The \code{data} and \code{metadata} items hold the names of
entries in the object's data and metadata
slots, respectively. The \code{dataDerived}
and \code{metadataDerived} items are each NULL, because
no derived values are defined by \linkS4class{lisst} objects.
}
}

\section{Details of the General Method}{


Note: the text of this section is identical for all \code{oce} subclasses, and so
some of what you read here may not be relevant to the class being described
in this help page.

If the specialized method produces no matches, the following generalized
method is applied. As with the specialized method, the procedure hinges first
on the values of \code{i} and, optionally, \code{j}. The work proceeds in steps, by
testing a sequence of possible conditions in sequence.
\enumerate{
\item A check is made as to whether \code{i} names one of the standard \code{oce} slots.
If so, \code{[[} returns the slot contents of that slot.  Thus, \code{x[["metadata"]]}
will retrieve the \code{metadata} slot, while \code{x[["data"]]} and
\code{x[["processingLog"]]} return those slots.
\item If \code{i} is a string ending in the \code{"Unit"}, then the characters preceding
that string are taken to be the name of an item in the data object, and a
list containing the unit is returned (or \code{NULL} if there is no such unit).
This list consists of an item named \code{unit}, which is an \code{\link[=expression]{expression()}}, and
an item named \code{scale}, which is a string describing the measurement scale.
If the string ends in \code{" unit"}, e.g.  \code{x[["temperature unit"]]} (note the
space), then just the expression is returned, and if it ends in \code{" scale"},
then just the scale is returned.
\item If \code{i} is a string ending in \code{"Flag"}, then the corresponding data-quality
flag is returned (or \code{NULL} if there is no such flag).
\item If the object holds hydrographic information (pressure, salinity,
temperature, longitude and latitude) then another set of possibilities
arises.  If \code{i} is \code{"sigmaTheta"}, then the value of \code{\link[=swSigmaTheta]{swSigmaTheta()}} is
called with \code{x} as the sole argument, and the results are returned.
Similarly, \code{\link[=swSigma0]{swSigma0()}} is used if \code{i="sigma0"}, and \code{\link[=swSpice]{swSpice()}} is used if
\code{i="spice"}. Of course, these actions only make sense for objects that
contain the relevant items within their \code{data} slot.
\item After these possibilities are eliminated, the action depends on whether
\code{j} has been provided.  If \code{j} is not provided, or is the string \code{""}, then
\code{i} is sought in the \code{metadata} slot, and then in the \code{data} slot, returning
whichever is found first.  In other words, if \code{j} is not provided, the
\code{metadata} slot takes preference over the \code{data} slot. However, if \code{j} is
provided, then it must be either the string \code{"metadata"} or \code{"data"}, and it
directs where to look.
\item If none of the above-listed conditions holds, then \code{NULL} is returned.
}
}

\seealso{
Other functions that extract parts of oce objects: 
\code{\link{[[,adp-method}},
\code{\link{[[,adv-method}},
\code{\link{[[,amsr-method}},
\code{\link{[[,argo-method}},
\code{\link{[[,bremen-method}},
\code{\link{[[,cm-method}},
\code{\link{[[,coastline-method}},
\code{\link{[[,ctd-method}},
\code{\link{[[,echosounder-method}},
\code{\link{[[,g1sst-method}},
\code{\link{[[,gps-method}},
\code{\link{[[,ladp-method}},
\code{\link{[[,landsat-method}},
\code{\link{[[,lobo-method}},
\code{\link{[[,met-method}},
\code{\link{[[,oce-method}},
\code{\link{[[,odf-method}},
\code{\link{[[,rsk-method}},
\code{\link{[[,sealevel-method}},
\code{\link{[[,section-method}},
\code{\link{[[,tidem-method}},
\code{\link{[[,topo-method}},
\code{\link{[[,windrose-method}},
\code{\link{[[,xbt-method}},
\code{\link{[[<-,adv-method}}

Other things related to lisst data: 
\code{\link{[[<-,lisst-method}},
\code{\link{as.lisst}()},
\code{\link{lisst-class}},
\code{\link{plot,lisst-method}},
\code{\link{read.lisst}()},
\code{\link{summary,lisst-method}}
}
\author{
Dan Kelley
}
\concept{functions that extract parts of oce objects}
\concept{things related to lisst data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{integerToAscii}
\alias{integerToAscii}
\title{Decode integer to corresponding ASCII code}
\usage{
integerToAscii(i)
}
\arguments{
\item{i}{an integer, or integer vector.}
}
\value{
A character, or character vector.
}
\description{
Decode integer to corresponding ASCII code
}
\examples{
library(oce)
A <- integerToAscii(65)
cat("A=", A, "\n")
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{oce.contour}
\alias{oce.contour}
\alias{oceContour}
\title{Oce Variant of contour}
\usage{
oce.contour(
  x,
  y,
  z,
  revx = FALSE,
  revy = FALSE,
  add = FALSE,
  tformat,
  drawTimeRange = getOption("oceDrawTimeRange"),
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{x}{values for x grid.}

\item{y}{values for y grid.}

\item{z}{matrix for values to be contoured.  The first dimension of \code{z}
must equal the number of items in \code{x}, etc.}

\item{revx}{set to \code{TRUE} to reverse the order in which the labels on
the x axis are drawn}

\item{revy}{set to \code{TRUE} to reverse the order in which the labels on
the y axis are drawn}

\item{add}{logical value indicating whether the contours should be added to
a pre-existing plot.}

\item{tformat}{time format; if not supplied, a reasonable choice will be
made by \code{\link[=oce.axis.POSIXct]{oce.axis.POSIXct()}}, which draws time axes.}

\item{drawTimeRange}{logical, only used if the \code{x} axis is a time.  If
\code{TRUE}, then an indication of the time range of the data (not the axis)
is indicated at the top-left margin of the graph.  This is useful because
the labels on time axes only indicate hours if the range is less than a day,
etc.}

\item{debug}{a flag that turns on debugging; set to 1 to information about
the processing.}

\item{\dots}{optional arguments passed to plotting functions.}
}
\description{
This provides something analogous to \code{\link[=contour]{contour()}}, but with the
ability to flip x and y.
Setting \code{revy=TRUE} can be helpful if the \code{y} data represent
pressure or depth below the surface.
}
\examples{
library(oce)
data(topoWorld)
# coastline now, and in last glacial maximum
lon <- topoWorld[["longitude"]]
lat <- topoWorld[["latitude"]]
z <- topoWorld[["z"]]
oce.contour(lon, lat, z, levels=0, drawlabels=FALSE)
oce.contour(lon, lat, z, levels=-130, drawlabels=FALSE, col='blue', add=TRUE)
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/coastline.R
\name{coastlineCut}
\alias{coastlineCut}
\title{Cut a Coastline Object at Specified Longitude}
\usage{
coastlineCut(coastline, lon_0 = 0)
}
\arguments{
\item{coastline}{a \linkS4class{coastline} object.}

\item{lon_0}{longitude as would be given in a \verb{+lon_0=} item in a
call to \code{\link[sf:sf_project]{sf::sf_project()}}.}
}
\value{
a new coastline object
}
\description{
This can be helpful in preventing \code{\link[=mapPlot]{mapPlot()}} from producing ugly
horizontal lines in world maps. These lines occur when a coastline segment
is intersected by longitude lon_0+180.  Since the coastline files in the oce
and ocedata packages are already "cut" at longitudes of -180 and 180, the present
function is not needed for default maps, which have \code{+lon_0=0}. However,
may help with other values of \code{lon_0}.
}
\section{Caution}{

This function is provisional. Its behaviour, name and very existence
may change.  Part of the development plan is to see if there is common
ground between this and the \code{clipPolys} function in the
\CRANpkg{PBSmapping} package.
}

\examples{
\donttest{
library(oce)
data(coastlineWorld)
mapPlot(coastlineCut(coastlineWorld, lon_0=100),
        projection="+proj=moll +lon_0=100", col="gray")
}

}
\seealso{
Other things related to coastline data: 
\code{\link{[[,coastline-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{as.coastline}()},
\code{\link{coastline-class}},
\code{\link{coastlineBest}()},
\code{\link{coastlineWorld}},
\code{\link{download.coastline}()},
\code{\link{plot,coastline-method}},
\code{\link{read.coastline.openstreetmap}()},
\code{\link{read.coastline.shapefile}()},
\code{\link{subset,coastline-method}},
\code{\link{summary,coastline-method}}
}
\author{
Dan Kelley
}
\concept{things related to coastline data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.R
\name{[[<-,adp-method}
\alias{[[<-,adp-method}
\title{Replace Parts of an ADP Object}
\usage{
\S4method{[[}{adp}(x, i, j, ...) <- value
}
\arguments{
\item{x}{an \linkS4class{adp} object.}

\item{i}{character value naming the item to replace.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{optional additional information (ignored).}

\item{value}{The value to be placed into \code{x}, somewhere.}
}
\description{
In addition to the usual insertion of elements by name, note
that e.g. \code{pitch} gets stored into \code{pitchSlow}.

The \code{[[<-} method works for all \linkS4class{oce} objects.
The purpose, as with the related extraction method, \code{[[},
is to insulate users from the internal details of \linkS4class{oce}
objects, by looking for items within the various storage
slots of the object. Items not actually stored can also be
replaced, including units and data-quality
flags.
}
\details{
As with \code{[[} method, the procedure works in steps.

First, the \code{metadata} slot of \code{x} is checked to
see whether it contains something named with \code{i}.
If so, then the named item is replaced with \code{value}.

Otherwise, if the string value of \code{i} ends in \code{Unit}, then the
characters preceding that are taken as the name of a variable, and
the \code{metadata} slot of \code{x} is updated to store that unit, e.g.\preformatted{x[["temperatureUnits"]] <- list(unit=expression(degree*F),scale="")
}

Similarly, if \code{i} ends in \code{Flag}, then quality-control
flags are set up as defined by \code{result}, e.g.\preformatted{o[["temperatureFlags"]] <- c(2,4,2,2)
}

Otherwise, \code{\link[=pmatch]{pmatch()}} is used for a partial-string match with
the names of the items that are in the \code{data} slot of \code{x}.
The first item found (if any) is then updated to hold the value \code{result}.

If none of these conditions is met, a warning is issued.
}
\seealso{
Other functions that replace parts of oce objects: 
\code{\link{[[<-,amsr-method}},
\code{\link{[[<-,argo-method}},
\code{\link{[[<-,bremen-method}},
\code{\link{[[<-,cm-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{[[<-,echosounder-method}},
\code{\link{[[<-,g1sst-method}},
\code{\link{[[<-,gps-method}},
\code{\link{[[<-,ladp-method}},
\code{\link{[[<-,landsat-method}},
\code{\link{[[<-,lisst-method}},
\code{\link{[[<-,lobo-method}},
\code{\link{[[<-,met-method}},
\code{\link{[[<-,oce-method}},
\code{\link{[[<-,odf-method}},
\code{\link{[[<-,rsk-method}},
\code{\link{[[<-,sealevel-method}},
\code{\link{[[<-,section-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{[[<-,topo-method}},
\code{\link{[[<-,windrose-method}},
\code{\link{[[<-,xbt-method}}

Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley
}
\concept{functions that replace parts of oce objects}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.R
\name{concatenate,adp-method}
\alias{concatenate,adp-method}
\title{Concatenate adp objects}
\usage{
\S4method{concatenate}{adp}(object, ...)
}
\arguments{
\item{object}{An object of \linkS4class{adp}, or a list containing such
objects (in which case the remaining arguments are ignored).}

\item{...}{Optional additional objects of \linkS4class{adp}.}
}
\value{
An object of \linkS4class{adp}.
}
\description{
This function concatenates adp objects. It is intended
for objects holding data sampled through time, and it works by pasting
together data linearly if they are vectors, by row if they are matrices,
and by second index if they are arrays. It has been tested for the following
classes:
\linkS4class{adp},
\linkS4class{adv},
\linkS4class{ctd},
and
\linkS4class{met}. It may do useful things for other classes, and
so users are encouraged to try, and to report problems to the developers.
It is unlikely that the function will do anything even remotely useful
for image and topographic data, to name just two cases that do not fit
the sampled-over-time category.
}
\examples{
## 1. Split, then recombine, a ctd object.
data(ctd)
ctd1 <- subset(ctd, scan <= median(ctd[["scan"]]))
ctd2 <- subset(ctd, scan > median(ctd[["scan"]]))
CTD <- concatenate(ctd1, ctd2)

## 2. Split, then recombine, an adp object.
data(adp)
midtime <- median(adp[["time"]])
adp1 <- subset(adp, time <= midtime)
adp2 <- subset(adp, time > midtime)
ADP <- concatenate(adp1, adp2)

\dontrun{
## 3. Download two met files and combine them.
met1 <- read.met(download.met(id=6358, year=2003, month=8))
met2 <- read.met(download.met(id=6358, year=2003, month=9))
MET <- concatenate(met1, met2)
}

}
\seealso{
Other functions that concatenate oce objects: 
\code{\link{concatenate,list-method}},
\code{\link{concatenate,oce-method}},
\code{\link{concatenate}()}
}
\author{
Dan Kelley
}
\concept{functions that concatenate oce objects}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/map.R
\name{mapDirectionField}
\alias{mapDirectionField}
\title{Add a Direction Field to an Existing Map}
\usage{
mapDirectionField(
  longitude,
  latitude,
  u,
  v,
  scale = 1,
  length = 0.05,
  code = 2,
  col = par("fg"),
  ...
)
}
\arguments{
\item{longitude, latitude}{numeric vectors of the starting points for arrows.}

\item{u, v}{numeric vectors of the components of a vector to be shown as a direction
field.}

\item{scale}{latitude degrees per unit of \code{u} or \code{v}.}

\item{length}{length of arrow heads, passed to \code{\link[=arrows]{arrows()}}.}

\item{code}{code of arrows, passed to \code{\link[=arrows]{arrows()}}.}

\item{col}{color of arrows.  This may be a single color, or a matrix
of colors of the same dimension as \code{u}.}

\item{\dots}{optional arguments passed to \code{\link[=arrows]{arrows()}}, e.g.
\code{angle} and \code{lwd} can be useful in differentiating different
fields.}
}
\description{
Plot a direction field on a existing map.
}
\details{
Adds arrows for a direction field on an existing map.  There are different
possibilities for how \code{longitude}, \code{latitude} and \code{u} and
\code{v} match up.  In one common case, all four of these are matrices, e.g.
output from a numerical model.  In another, \code{longitude} and
\code{latitude} are the coordinates along the matrices, and are thus stored in
vectors with lengths that match appropriately.
}
\examples{
\dontrun{
library(oce)
data(coastlineWorld)
par(mar=rep(2, 4))
mapPlot(coastlineWorld, longitudelim=c(-120,-55), latitudelim=c(35, 50),
        projection="+proj=laea +lat0=40 +lat1=60 +lon_0=-110")
lon <- seq(-120, -60, 15)
lat <- 45 + seq(-15, 15, 5)
lonm <- matrix(expand.grid(lon, lat)[, 1], nrow=length(lon))
latm <- matrix(expand.grid(lon, lat)[, 2], nrow=length(lon))
## vectors pointed 45 degrees clockwise from north
u <- matrix(1/sqrt(2), nrow=length(lon), ncol=length(lat))
v <- matrix(1/sqrt(2), nrow=length(lon), ncol=length(lat))
mapDirectionField(lon, lat, u, v, scale=3)
mapDirectionField(lonm, latm, 0, 1, scale=3, col='red')
# Color code by longitude, using thick lines
col <- colormap(lonm)$zcol
mapDirectionField(lonm, latm, 1, 0, scale=3, col=col, lwd=2)
}

}
\seealso{
A map must first have been created with \code{\link[=mapPlot]{mapPlot()}}.

Other functions related to maps: 
\code{\link{formatPosition}()},
\code{\link{lonlat2map}()},
\code{\link{lonlat2utm}()},
\code{\link{map2lonlat}()},
\code{\link{mapArrows}()},
\code{\link{mapAxis}()},
\code{\link{mapContour}()},
\code{\link{mapCoordinateSystem}()},
\code{\link{mapGrid}()},
\code{\link{mapImage}()},
\code{\link{mapLines}()},
\code{\link{mapLocator}()},
\code{\link{mapLongitudeLatitudeXY}()},
\code{\link{mapPlot}()},
\code{\link{mapPoints}()},
\code{\link{mapPolygon}()},
\code{\link{mapScalebar}()},
\code{\link{mapText}()},
\code{\link{mapTissot}()},
\code{\link{oceCRS}()},
\code{\link{shiftLongitude}()},
\code{\link{usrLonLat}()},
\code{\link{utm2lonlat}()}
}
\author{
Dan Kelley
}
\concept{functions related to maps}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.R
\name{initialize,ctd-method}
\alias{initialize,ctd-method}
\title{Initialize storage for a ctd object}
\usage{
\S4method{initialize}{ctd}(
  .Object,
  pressure,
  salinity,
  temperature,
  conductivity,
  units,
  pressureType,
  deploymentType,
  ...
)
}
\arguments{
\item{.Object}{the string \code{"ctd"}}

\item{pressure}{optional numerical vector of pressures.}

\item{salinity}{optional numerical vector of salinities.}

\item{temperature}{optional numerical vector of temperatures.}

\item{conductivity}{optional numerical vector of conductivities.}

\item{units}{optional list indicating units for the quantities specified
in the previous arguments. If this
is not supplied, a default is set up, based on which of the
\code{pressure} to \code{conductivity} arguments were specified.
If all of those 4 arguments were specified, then \code{units} is set
up as if the call included the following:
\code{units=list(temperature=list(unit=expression(degree*C), scale="ITS-90"),
     salinity=list(unit=expression(), scale="PSS-78"),
     conductivity=list(unit=expression(), scale=""),
     pressure=list(unit=expression(dbar), scale=""),
     depth=list(unit=expression(m), scale=""))}. This list is trimmed
of any of the 4 items that were not specified in the previous
arguments. Note that if \code{units} is specified, then it is just
copied into the \code{metadata} slot of the returned object, so the user
must be careful to set up values that will make sense to other \code{oce}
functions.}

\item{pressureType}{optional character string indicating the type of pressure;
if not supplied, this defaults to \code{"sea"}, which indicates the excess of
pressure over the atmospheric value, in dbar.}

\item{deploymentType}{optional character string indicating the type of deployment, which may
be \code{"unknown"}, \code{"profile"}, \code{"towyo"}, or \code{"thermosalinograph"}.
If this is not set, the value defaults to \code{"unknown"}.}

\item{...}{Ignored.}
}
\description{
This function creates \linkS4class{ctd} objects. It is mainly
used by \code{oce} functions such as \code{\link[=read.ctd]{read.ctd()}} and \code{\link[=as.ctd]{as.ctd()}},
and it is not intended for novice users, so it may change at any time, without
following the usual rules for transitioning to deprecated and defunct status
(see \link{oce-deprecated}).
}
\details{
To save storage, this function has arguments only for quantities that are often present in data
files all cases. For example, not
all data files will have oxygen, so that's not present here.
Extra data may be added after the object is created, using
\code{\link[=oceSetData]{oceSetData()}}.
Similarly, \code{\link[=oceSetMetadata]{oceSetMetadata()}} may be used to add metadata (station ID, etc),
while bearing in mind that other functions look for such information
in very particular places (e.g. the station ID is a string named \code{station}
within the \code{metadata} slot). See \linkS4class{ctd} for more information
on elements stored in \code{ctd} objects.
}
\examples{

# 1. empty
new("ctd")

# 2. fake data with no location information, so can only
#    plot with the UNESCO equation of state.
#    NOTE: always name arguments, in case the default order gets changed
ctd <- new("ctd", salinity=35+1:3/10, temperature=10-1:3/10, pressure=1:3)
summary(ctd)
plot(ctd, eos="unesco")

# 3. as 2, but insert location and plot with GSW equation of state.
ctd <- oceSetMetadata(ctd, "latitude", 44)
ctd <- oceSetMetadata(ctd, "longitude", -63)
plot(ctd, eos="gsw")

}
\seealso{
Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}
}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{oce.axis.POSIXct}
\alias{oce.axis.POSIXct}
\title{Oce Version of axis.POSIXct}
\usage{
oce.axis.POSIXct(
  side,
  x,
  at,
  tformat,
  labels = TRUE,
  drawTimeRange,
  abbreviateTimeRange = FALSE,
  drawFrequency = FALSE,
  cex.axis = par("cex.axis"),
  cex.lab = par("cex.lab"),
  cex.main = par("cex.main"),
  mar = par("mar"),
  mgp = par("mgp"),
  main = "",
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{side}{as for \code{\link[=axis.POSIXct]{axis.POSIXct()}}.}

\item{x}{as for \code{\link[=axis.POSIXct]{axis.POSIXct()}}.}

\item{at}{as for \code{\link[=axis.POSIXct]{axis.POSIXct()}}.}

\item{tformat}{as \code{format} for \code{\link[=axis.POSIXct]{axis.POSIXct()}} for now, but
may eventually have new features for multiline labels, e.g. day on one line
and month on another.}

\item{labels}{as for \code{\link[=axis.POSIXct]{axis.POSIXct()}}.}

\item{drawTimeRange}{Optional indication of whether/how to draw the time range
in the margin on the side of the the plot opposite the time axis. If this is
not supplied, it defaults to the value returned by
\code{\link{getOption}}\code{("oceDrawTimeRange")}, and if that option is not set,
it defaults to \code{TRUE}. No time range is drawn if \code{drawTimeRange} is \code{FALSE}.
If it is \code{TRUE}, the range will be shown. This range refers to
range of the x axis (not the data). The format of the elements of that range is set by
\code{\link{getOption}}\code{("oceTimeFormat")} (or with the default value
of an empty string, if this option has not been set). The timezone will
be indicated if the time range is under a week.  For preliminary work, it makes
sense to use \code{drawTimeRange=TRUE}, but for published work it can be better
to drop this label and indicate something about the time in the figure caption.}

\item{abbreviateTimeRange}{boolean, \code{TRUE} to abbreviate the second
number in the time range, e.g. dropping the year if it is the same in the
first number.}

\item{drawFrequency}{boolean, \code{TRUE} to show the frequency of sampling
in the data}

\item{cex.axis, cex.lab, cex.main}{character expansion factors for axis numbers, axis names and plot titles; see \code{\link[=par]{par()}}.}

\item{mar}{value for \code{par(mar)} for axis}

\item{mgp}{value for \code{par(mgp)} for axis}

\item{main}{title of plot}

\item{debug}{a flag that turns on debugging.  Set to 1 to get a moderate
amount of debugging information, or to 2 to get more.}

\item{\dots}{as for \code{\link[=axis.POSIXct]{axis.POSIXct()}}.}
}
\value{
A vector of times corresponding to axis ticks is returned silently.
}
\description{
A specialized variant of \code{\link[=axis.POSIXct]{axis.POSIXct()}} that produces
results with less ambiguity in axis labels.
}
\details{
The tick marks are set automatically based on examination of the time range on
the axis. The scheme was devised by constructing test cases with a typical plot
size and font size, and over a wide range of time scales. In some categories,
both small tick marks are interspersed between large ones.

The user may set the format of axis numbers with the \code{tformat} argument.
If this is not supplied, the format is set based on the time span of the axis:
\itemize{
\item If this time span is less than a minute, the time axis labels are in
seconds (fractional seconds, if the interval is less than 2 seconds), with
leading zeros on small integers. (Fractional seconds are enabled with a trick:
the usual R format \code{"\\\%S"} is supplemented with a new format e.g.
\code{"\\\%.2S"}, meaning to use two digits after the decimal.)
\item If the time span exceeds a minute but is less than 1.5 days, the label
format is \code{"\\\%H:\\\%M:\\\%S"}.
\item If the time span exceeds 1.5 days but is less than 1 year, the format is
\code{"\\\%b \\\%d"} (e.g. Jul 15) and, again, the tick marks are set up for several
subcategories.
\item If the time span exceeds a year, the format is \code{"\\\%Y"}, i.e. the year
is displayed with 4 digits.
}

It should be noted that this scheme differs from the R approach in several
ways. First, R writes day names for some time ranges, in a convention that is
seldom seen in the literature. Second, R will write nn:mm for both HH:MM and
MM:SS, an ambiguity that might confuse readers. Third, the use of both large
and small tick marks is not something that R does.

Bear in mind that \code{tformat} may be set to alter the number format, but
that the tick mark scheme cannot (presently) be controlled.
}
\seealso{
This is used mainly by \code{\link[=oce.plot.ts]{oce.plot.ts()}}.
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/tides.R
\name{predict.tidem}
\alias{predict.tidem}
\title{Predict a Tidal Signal}
\usage{
\method{predict}{tidem}(object, newdata, ...)
}
\arguments{
\item{object}{a \linkS4class{tidem} object.}

\item{newdata}{vector of POSIXt times at which to make the
prediction.  For models created with \code{\link[=tidem]{tidem()}},
the \code{newdata} argument is optional, and if it is not provided, then
the predictions are at the observation times given to
\code{\link[=tidem]{tidem()}}. However, \code{newdata} is required  if \code{\link[=as.tidem]{as.tidem()}}
had been used to create \code{object}.}

\item{\dots}{optional arguments passed on to children.}
}
\value{
A vector of predictions.
}
\description{
This creates a time-series of predicted tides, based on a
tidal model object that was created by \code{\link[=as.tidem]{as.tidem()}} or \code{\link[=tidem]{tidem()}}.
}
\examples{

\dontrun{
library(oce)
# 1. tidal anomaly
data(sealevelTuktoyaktuk)
time <- sealevelTuktoyaktuk[["time"]]
elevation <- sealevelTuktoyaktuk[["elevation"]]
oce.plot.ts(time, elevation, type='l', ylab="Height [m]", ylim=c(-2, 6))
tide <- tidem(sealevelTuktoyaktuk)
lines(time, elevation - predict(tide), col="red")
abline(h=0, col="red")

# 2. prediction at specified times
data(sealevel)
m <- tidem(sealevel)
## Check fit over 2 days (interpolating to finer timescale)
look <- 1:48
time <- sealevel[["time"]]
elevation <- sealevel[["elevation"]]
oce.plot.ts(time[look], elevation[look])
# 360s = 10 minute timescale
t <- seq(from=time[1], to=time[max(look)], by=360)
lines(t, predict(m, newdata=t), col='red')
legend("topright", col=c("black","red"),
legend=c("data","model"),lwd=1)
}

}
\seealso{
Other things related to tides: 
\code{\link{[[,tidem-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{as.tidem}()},
\code{\link{plot,tidem-method}},
\code{\link{summary,tidem-method}},
\code{\link{tidalCurrent}},
\code{\link{tidedata}},
\code{\link{tidem-class}},
\code{\link{tidemAstron}()},
\code{\link{tidemVuf}()},
\code{\link{tidem}},
\code{\link{webtide}()}
}
\author{
Dan Kelley
}
\concept{things related to tides}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/lobo.R
\name{as.lobo}
\alias{as.lobo}
\title{Coerce Data into a Lobo Object}
\usage{
as.lobo(
  time,
  u,
  v,
  salinity,
  temperature,
  pressure,
  nitrate,
  fluorescence,
  filename = ""
)
}
\arguments{
\item{time}{vector of times of observation}

\item{u}{vector of x velocity component observations}

\item{v}{vector of y velocity component observations}

\item{salinity}{vector of salinity observations}

\item{temperature}{vector of temperature observations}

\item{pressure}{vector of pressure observations}

\item{nitrate}{vector of nitrate observations}

\item{fluorescence}{vector of fluorescence observations}

\item{filename}{source filename}
}
\value{
A \linkS4class{lobo} object.
}
\description{
Coerce a dataset into a lobo dataset.
}
\seealso{
Other things related to lobo data: 
\code{\link{[[,lobo-method}},
\code{\link{[[<-,lobo-method}},
\code{\link{lobo-class}},
\code{\link{lobo}},
\code{\link{plot,lobo-method}},
\code{\link{read.lobo}()},
\code{\link{subset,lobo-method}},
\code{\link{summary,lobo-method}}
}
\author{
Dan Kelley
}
\concept{things related to lobo data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swRho}
\alias{swRho}
\title{Seawater density}
\usage{
swRho(
  salinity,
  temperature = NULL,
  pressure = NULL,
  longitude = NULL,
  latitude = NULL,
  eos = getOption("oceEOS", default = "gsw")
)
}
\arguments{
\item{salinity}{either practical salinity (in which case \code{temperature}
and \code{pressure} must be provided) \emph{or} an \code{oce} object, in
which case \code{salinity}, \code{temperature} (in the ITS-90 scale; see
next item), etc. are inferred from the object, ignoring the
other parameters, if they are supplied.}

\item{temperature}{\emph{in-situ} temperature (\eqn{^\circ}{deg}C), defined
on the ITS-90 scale.  This scale is used by GSW-style calculation (as
requested by setting \code{eos="gsw"}), and is the value contained within
\code{ctd} objects (and probably most other objects created with data
acquired in the past decade or two). Since the UNESCO-style calculation is
based on IPTS-68, the temperature is converted within the present function,
using \code{\link[=T68fromT90]{T68fromT90()}}.}

\item{pressure}{pressure (dbar)}

\item{longitude}{longitude of observation (only used if \code{eos="gsw"};
see \sQuote{Details}).}

\item{latitude}{latitude of observation (only used if \code{eos="gsw"}; see
\sQuote{Details}).}

\item{eos}{equation of state, either \code{"unesco"} (references 1 and 2)
or \code{"gsw"} (references 3 and 4).}
}
\value{
\emph{In-situ} density (kg/m\eqn{^3}{^3}).
}
\description{
Compute \eqn{\rho}{rho}, the \emph{in-situ} density of seawater.
}
\details{
If \code{eos="unesco"}, the density is calculated using the UNESCO equation
of state for seawater (references 1 and 2), and if \code{eos="gsw"}, the GSW formulation
(references 3 and 4) is used.
}
\section{Temperature units}{
 The UNESCO formulae are defined in terms of
temperature measured on the IPTS-68 scale, whereas the replacement GSW
formulae are based on the ITS-90 scale. Prior to the addition of GSW
capabilities, the various \verb{sw*} functions took temperature to be in
IPTS-68 units. As GSW capabilities were added in early 2015, the assumed
unit of \code{temperature} was taken to be ITS-90.  This change means that
old code has to be modified, by replacing e.g. \code{swRho(S, T, p)} with
\code{swRho(S, T90fromT68(T), p)}. At typical oceanic values, the difference
between the two scales is a few millidegrees.
}

\examples{
library(oce)
# The numbers in the comments are the check values listed in reference 1;
# note that temperature in that reference was on the T68 scale, but that
# the present function works with the ITS-90 scale, so a conversion
# is required.
swRho(35, T90fromT68(5),      0, eos="unesco") # 1027.67547
swRho(35, T90fromT68(5),  10000, eos="unesco") # 1069.48914
swRho(35, T90fromT68(25),     0, eos="unesco") # 1023.34306
swRho(35, T90fromT68(25), 10000, eos="unesco") # 1062.53817

}
\references{
\enumerate{
\item Fofonoff, P. and R. C. Millard Jr, 1983. Algorithms for computation of
fundamental properties of seawater.
\emph{Unesco Technical Papers in Marine Science},
\emph{44}, 53 pp
\item Gill, A.E., 1982. \emph{Atmosphere-ocean Dynamics}, Academic Press, New
York, 662 pp.
\item IOC, SCOR, and IAPSO (2010). The international thermodynamic equation of
seawater-2010: Calculation and use of thermodynamic properties.  Technical
Report 56, Intergovernmental Oceanographic Commission, Manuals and Guide.
\item McDougall, T.J. and P.M. Barker, 2011: Getting started with TEOS-10 and
the Gibbs Seawater (GSW) Oceanographic Toolbox, 28pp., SCOR/IAPSO WG127,
ISBN 978-0-646-55621-5.
}
}
\seealso{
Related density routines include \code{\link[=swSigma0]{swSigma0()}} (and
equivalents at other pressure horizons), \code{\link[=swSigmaT]{swSigmaT()}}, and
\code{\link[=swSigmaTheta]{swSigmaTheta()}}.

Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/argo.R
\name{timeToArgoJuld}
\alias{timeToArgoJuld}
\title{Convert time to Argo Julian Day (juld)}
\usage{
timeToArgoJuld(t)
}
\arguments{
\item{t}{A POSIXct time or a string that can be converted to a POSIXct time}
}
\description{
This converts a POSIXct time into an Argo julian day, with the convention
that juld=0 at 1950-01-01.
}
\examples{
timeToArgoJuld("2020-07-01")

}
\author{
Jaimie Harbin and Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ladp.R
\docType{class}
\name{ladp-class}
\alias{ladp-class}
\title{Class to Store Lowered-adp Data}
\description{
This class stores data measured with a lowered ADP (also
known as ADCP) device.
}
\section{Slots}{

\describe{
\item{\code{data}}{As with all \code{oce} objects, the \code{data} slot
for \code{ladp} objects is a
\link{list} containing the main data for the object.
{}}

\item{\code{metadata}}{As with all \code{oce} objects, the \code{metadata} slot
for \code{ladp} objects is a \link{list} containing
information about the \code{data} or about the object itself.
{}}

\item{\code{processingLog}}{As with all \code{oce} objects, the \code{processingLog} slot
for \code{ladp} objects is a
\link{list} with entries describing the creation and evolution
of the object. The contents are updated by various \code{oce} functions to
keep a record of processing steps.  Object summaries and
\code{\link[=processingLogShow]{processingLogShow()}} both display the log.}
}}

\section{Modifying slot contents}{


Although the \code{[[<-} operator may permit modification of the contents
of \linkS4class{ladp} objects (see \code{\link{[[<-,ladp-method}}),
it is better to use \code{\link[=oceSetData]{oceSetData()}} and \code{\link[=oceSetMetadata]{oceSetMetadata()}},
because those functions save an entry in the \code{processingLog}
that describes the change.
}

\section{Retrieving slot contents}{


The full contents of the \code{data} and \code{metadata} slots of a \linkS4class{ladp}
object may be retrieved in the standard R way using \code{\link[=slot]{slot()}}. For
example \code{slot(o,"data")} returns the \code{data} slot of an object named \code{o},
and similarly \code{slot(o,"metadata")} returns
the \code{metadata} slot.

The slots may also be obtained with the \code{\link{[[,ladp-method}}
operator, as e.g. \code{o[["data"]]} and \code{o[["metadata"]]}, respectively.

The \code{\link{[[,ladp-method}} operator can also
be used to retrieve items from within the \code{data} and \code{metadata} slots.
For example, \code{o[["temperature"]]} can be used to retrieve temperature
from an object containing that quantity. The rule is that a named
quantity is sought first within the object's \code{metadata} slot,
with the \code{data} slot being checked only if \code{metadata} does not
contain the item. This \code{[[} method can also be used to get
certain derived quantities, if the object contains sufficient
information to calculate them. For example, an object that holds
(practical) salinity, temperature and pressure, along with
longitude and latitude, has sufficient information to compute
Absolute Salinity, and so \code{o[["SA"]]} will yield the
calculated Absolute Salinity.

It is also possible to find items more directly, using \code{\link[=oceGetData]{oceGetData()}} and
\code{\link[=oceGetMetadata]{oceGetMetadata()}}, but neither of these functions can
retrieve derived items.
}

\seealso{
Other things related to ladp data: 
\code{\link{[[,ladp-method}},
\code{\link{[[<-,ladp-method}},
\code{\link{as.ladp}()},
\code{\link{plot,ladp-method}},
\code{\link{summary,ladp-method}}
}
\author{
Dan Kelley
}
\concept{things related to ladp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/windrose.R
\name{[[<-,windrose-method}
\alias{[[<-,windrose-method}
\title{Replace Parts of a Windrose Object}
\usage{
\S4method{[[}{windrose}(x, i, j, ...) <- value
}
\arguments{
\item{x}{a \linkS4class{windrose} object.}

\item{i}{character value naming the item to replace.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{optional additional information (ignored).}

\item{value}{The value to be placed into \code{x}, somewhere.}
}
\description{
The \code{[[<-} method works for all \linkS4class{oce} objects.
The purpose, as with the related extraction method, \code{[[},
is to insulate users from the internal details of \linkS4class{oce}
objects, by looking for items within the various storage
slots of the object. Items not actually stored can also be
replaced, including units and data-quality
flags.
}
\details{
As with \code{[[} method, the procedure works in steps.

First, the \code{metadata} slot of \code{x} is checked to
see whether it contains something named with \code{i}.
If so, then the named item is replaced with \code{value}.

Otherwise, if the string value of \code{i} ends in \code{Unit}, then the
characters preceding that are taken as the name of a variable, and
the \code{metadata} slot of \code{x} is updated to store that unit, e.g.\preformatted{x[["temperatureUnits"]] <- list(unit=expression(degree*F),scale="")
}

Similarly, if \code{i} ends in \code{Flag}, then quality-control
flags are set up as defined by \code{result}, e.g.\preformatted{o[["temperatureFlags"]] <- c(2,4,2,2)
}

Otherwise, \code{\link[=pmatch]{pmatch()}} is used for a partial-string match with
the names of the items that are in the \code{data} slot of \code{x}.
The first item found (if any) is then updated to hold the value \code{result}.

If none of these conditions is met, a warning is issued.
}
\seealso{
Other functions that replace parts of oce objects: 
\code{\link{[[<-,adp-method}},
\code{\link{[[<-,amsr-method}},
\code{\link{[[<-,argo-method}},
\code{\link{[[<-,bremen-method}},
\code{\link{[[<-,cm-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{[[<-,echosounder-method}},
\code{\link{[[<-,g1sst-method}},
\code{\link{[[<-,gps-method}},
\code{\link{[[<-,ladp-method}},
\code{\link{[[<-,landsat-method}},
\code{\link{[[<-,lisst-method}},
\code{\link{[[<-,lobo-method}},
\code{\link{[[<-,met-method}},
\code{\link{[[<-,oce-method}},
\code{\link{[[<-,odf-method}},
\code{\link{[[<-,rsk-method}},
\code{\link{[[<-,sealevel-method}},
\code{\link{[[<-,section-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{[[<-,topo-method}},
\code{\link{[[<-,xbt-method}}

Other things related to windrose data: 
\code{\link{[[,windrose-method}},
\code{\link{as.windrose}()},
\code{\link{plot,windrose-method}},
\code{\link{summary,windrose-method}},
\code{\link{windrose-class}}
}
\concept{functions that replace parts of oce objects}
\concept{things related to windrose data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/section.R
\docType{data}
\name{section}
\alias{section}
\title{Hydrographic section}
\source{
This is based on the WOCE file named \code{a03_hy1.csv}, downloaded
from \url{https://cchdo.ucsd.edu/cruise/90CT40_1}, 13 April 2015.
}
\usage{
data(section)
}
\description{
This is line A03 (ExpoCode 90CT40_1, with nominal sampling date 1993-09-11).
The chief scientist was Tereschenkov of SOI, working aboard the Russian ship
Multanovsky, undertaking a westward transect from the Mediterranean outflow
region across to North America, with a change of heading in the last few dozen
stations to run across the nominal Gulf Stream axis.
The data flags follow the "WHP Bottle"convention, set by
\code{\link[=initializeFlagScheme,section-method]{initializeFlagScheme,section-method()}} to \code{"WHP bottle"}.  This convention
used to be described at the link
\verb{https://www.nodc.noaa.gov/woce/woce_v3/wocedata_1/whp/exchange/exchange_format_desc.htm}
but that was found to fail in December 2020.
}
\examples{
\dontrun{
library(oce)
# Gulf Stream
data(section)
GS <- subset(section, 109<=stationId&stationId<=129)
GSg <- sectionGrid(GS, p=seq(0, 5000, 100))
plot(GSg, map.xlim=c(-80,-60))
}

}
\seealso{
Other datasets provided with oce: 
\code{\link{adp}},
\code{\link{adv}},
\code{\link{amsr}},
\code{\link{argo}},
\code{\link{cm}},
\code{\link{coastlineWorld}},
\code{\link{ctdRaw}},
\code{\link{ctd}},
\code{\link{echosounder}},
\code{\link{landsat}},
\code{\link{lisst}},
\code{\link{lobo}},
\code{\link{met}},
\code{\link{ocecolors}},
\code{\link{rsk}},
\code{\link{sealevelTuktoyaktuk}},
\code{\link{sealevel}},
\code{\link{topoWorld}},
\code{\link{wind}},
\code{\link{xbt}}

Other things related to section data: 
\code{\link{[[,section-method}},
\code{\link{[[<-,section-method}},
\code{\link{as.section}()},
\code{\link{handleFlags,section-method}},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{plot,section-method}},
\code{\link{read.section}()},
\code{\link{section-class}},
\code{\link{sectionAddStation}()},
\code{\link{sectionGrid}()},
\code{\link{sectionSmooth}()},
\code{\link{sectionSort}()},
\code{\link{subset,section-method}},
\code{\link{summary,section-method}}
}
\concept{datasets provided with oce}
\concept{things related to section data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/accessors.R
\name{oceSetData}
\alias{oceSetData}
\title{Set Something in an oce data Slot}
\usage{
oceSetData(object, name, value, unit, originalName, note = "")
}
\arguments{
\item{object}{an \linkS4class{oce} object.}

\item{name}{String indicating the name of the \code{data} item to be set.}

\item{value}{Value for the item.}

\item{unit}{An optional indication of the units for the item. This has
three possible forms (see \dQuote{Details}).}

\item{originalName}{Optional character string giving an 'original' name (e.g.
as stored in the header of a data file).}

\item{note}{Either empty (the default), a character string, or \code{NULL},
to control additions made to the processing log of the return value. If
\code{note=""} then the an entry is created based on deparsing the function call.
If \code{note} is a non-empty string, then that string gets added added
to the processing log. Finally, if \code{note=NULL}, then nothing is
added to the processing log. This last form is useful in cases where
\code{oceSetData} is to be called many times in succession, resulting
in an overly verbose processing log; in such cases, it might help
to add a note by e.g. \code{processingLog(a) <- "QC (memo dek-2018-01/31)"}}
}
\value{
An \linkS4class{oce} object, the \code{data} slot of which has
been altered either by adding a new item or modifying an existing
item.
}
\description{
Create a copy of an object in which some element of its
\code{data} slot has been altered, or added.
}
\details{
The trickiest argument to set is the \code{unit}.  There are three
possibilities for this:
\enumerate{
\item \code{unit} is a named or unnamed \code{\link[=list]{list()}} that contains two items.
If the list is named, the names must be
\code{unit} and \code{scale}. If the list is unnamed, the stated names are assigned
to the items, in the stated order. Either way, the \code{unit}
item must be an \code{\link[=expression]{expression()}} that specifies the unit,
and the \code{scale} item must be a string that describes the scale. For
example, modern temperatures have
\code{unit=list(unit=expression(degree*C), scale="ITS-90")}.
\item \code{unit} is an \code{\link[=expression]{expression()}} giving the unit as above. In this
case, the scale will be set to \code{""}.
\item \code{unit} is a character string that is converted
into an expression with \link{parse}\code{(text=unit)},
and the scale set to \code{""}.
}
}
\examples{
data(ctd)
Tf <- swTFreeze(ctd)
ctd <- oceSetData(ctd, "freezing", Tf,
    unit=list(unit=expression(degree*C), scale="ITS-90"))
plotProfile(ctd, "freezing")

}
\seealso{
Other things related to the data slot: 
\code{\link{oceDeleteData}()},
\code{\link{oceGetData}()},
\code{\link{oceRenameData}()}
}
\author{
Dan Kelley
}
\concept{things related to the data slot}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{window.oce}
\alias{window.oce}
\title{Window an Oce Object by Time or Distance}
\usage{
\method{window}{oce}(
  x,
  start = NULL,
  end = NULL,
  frequency = NULL,
  deltat = NULL,
  extend = FALSE,
  which = c("time", "distance"),
  indexReturn = FALSE,
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{x}{an \linkS4class{oce} object.}

\item{start}{the start time (or distance) of the time (or space) region of
interest.  This may be a single value or a vector.}

\item{end}{the end time (or distance) of the time (or space) region of
interest.  This may be a single value or a vector.}

\item{frequency}{not permitted yet.}

\item{deltat}{not permitted yet}

\item{extend}{not permitted yet}

\item{which}{string containing the name of the quantity on which sampling is
done.  Possibilities are \code{"time"}, which applies the windowing on the
\code{time} entry of the \code{data} slot, and \code{"distance"}, for the
\code{distance} entry (for those objects, such as \code{adp}, that have this
entry).}

\item{indexReturn}{boolean flag indicating whether to return a list of the
"kept" indices for the \code{time} entry of the \code{data} slot, as well as
the \code{timeSlow} entry, if there is one..  Either of these lists will be
\code{NULL}, if the object lacks the relevant items.}

\item{debug}{a flag that turns on debugging.}

\item{\dots}{ignored}
}
\value{
Normally, this is new \code{oce} object.  However, if
\code{indexReturn=TRUE}, the return value is two-element list containing
items named \code{index} and \code{indexSlow}, which are the indices for the
\code{time} entry of the \code{data} slot (and the \code{timeSlow}, if it
exists).
}
\description{
Windows \code{x} on either time or distance, depending on the value of
\code{which}.  In each case, values of \code{start} and \code{end} may be
integers, to indicate a portion of the time or distance range.  If
\code{which} is \code{"time"}, then the \code{start} and \code{end} values
may also be provided as POSIX times, or character strings indicating times
(in time zone given by the value of \code{getOption("oceTz")}).
Note that \code{\link[=subset]{subset()}} may be more useful than this function.
}
\examples{
library(oce)
data(adp)
plot(adp)
early <- window(adp, start="2008-06-26 00:00:00", end="2008-06-26 12:00:00")
plot(early)
bottom <- window(adp, start=0, end=20, which="distance")
plot(bottom)
}
\seealso{
\code{\link[=subset]{subset()}} provides more flexible selection of subsets.
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/odf.R
\name{read.odf}
\alias{read.odf}
\title{Read an ODF file}
\usage{
read.odf(
  file,
  columns = NULL,
  header = "list",
  exclude = NULL,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{file}{the file containing the data.}

\item{columns}{An optional \link{list} that can be used to convert unrecognized
data names to resultant variable names.  For example,
\verb{columns=list(salinity=list(name="salt", unit=list(unit=expression(), scale="PSS-78"))}
states that a short-name of \code{"salt"} represents salinity, and that the unit is
as indicated. This is passed to \code{\link[=cnvName2oceName]{cnvName2oceName()}} or \code{\link[=ODFNames2oceNames]{ODFNames2oceNames()}},
as appropriate, and takes precedence over the lookup table in that function.}

\item{header}{An indication of whether, or how, to store the entire
ODF file header in the \code{metadata} slot of the returned object.
There are three choices for the \code{header} argument.
(1) If it is \code{NULL}, then the ODF header is not stored in
the \code{metadata} slot (although some of its contents are).
(2) If it is \code{"character"}, the header is stored within
the \code{metadata} as a vector named \code{header}, comprising
a character string for each line of the header within the ODF file.
(3) If it is \code{"list"}, then the \code{metadata} slot of the
returned object will contain a \code{list} named \code{header} that
has lists as its entries. (The sub-lists are in the form of
key-value pairs.) The naming of list entries is patterned on
that in the ODF header, except that \code{\link[=unduplicateNames]{unduplicateNames()}}
is used to transform repeated names by adding numerical suffices.
Note: on June 6, 2019, the default value of \code{header} was
changed from \code{NULL} to \code{"list"}; in addition, the resultant
list was made to contain every single item in the ODF header, with
\code{\link[=unduplicateNames]{unduplicateNames()}} being used to append integers to
distinguish between repeated names in the ODF format.}

\item{exclude}{either a character value holding a regular
expression that is used with \code{\link[=grep]{grep()}} to remove lines from the
header before processing, or \code{NULL} (the default), meaning
not to exclude any such lines.  The purpose of this argument
is to solve problems with some files, which can have
thousands of lines that indicate details that are may be of
little value in processing.  For example, some files have thousands
of lines that would be excluded by using
\code{exclude="PROCESS='Nulled the .* value"} in the function call.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\value{
An \linkS4class{oce} object.
}
\description{
ODF (Ocean Data Format) is a
format developed at the Bedford Institute of Oceanography and also used
at other Canadian Department of Fisheries and Oceans (DFO) facilities
(see references 1 and 2).
It can hold various types of time-series data, which includes a variety
of instrument types. Thus, \code{\link[=read.odf]{read.odf()}}
is used by \code{read.ctd.odf} for CTD data, etc.
}
\details{
Note that some elements of the metadata are particular to ODF objects,
e.g. \code{depthMin}, \code{depthMax} and \code{sounding}, which
are inferred from ODF items named \code{MIN_DEPTH}, \code{MAX_DEPTH}
and \code{SOUNDING}, respectively. In addition, the more common metadata
item \code{waterDepth}, which is used in \code{ctd} objects to refer to
the total water depth, is set to \code{sounding} if that is finite,
or to \code{maxDepth} otherwise.

The function \code{\link[=ODFNames2oceNames]{ODFNames2oceNames()}} is used to translate
data names from the ODF file to standard \code{oce} names.
}
\section{Metadata conventions}{


Some metadata items may be specific to certain instruments, and
certain research groups. It can be important for analysts to be aware of
the conventions used in datasets that are under study.
For example, as of June 2018, \code{adp}
objects created at the Bedford Institute of Oceanography may
have a metadata item named \code{depthOffBottom} (called
\code{DEPTH_OFF_BOTTOM} in ODF files), which is not typically
present in \code{ctd} files. This item illustrates the renaming
convention, from the CAMEL_CASE used in ODF files to the snakeCase
used in oce. Bearing this conversion in mind, users should not
find it difficult to understand the meaning of items that \code{\link[=read.odf]{read.odf()}}
stores within the \code{metadata} slot. Users should bear in mind
that the whole ODF header is saved as a list by
calling the function with \code{header="list"}, after which
e.g. \link{str}\code{(rval[["header"]])} or \link{View}\code{(rval[["header"]])}
can be used to isolate any information of interest (but bear in mind
that suffices are used to disambiguate sibling items of identical
name in the ODF header).
}

\section{Handling of temperature scales}{

\code{read.odf()} stores temperature data directly as read from the file, which
might mean the IPTS-68 scale.  These values should not be used to calculate
other seawater quantities, because formulae are generally based in ITS90
temperatures. To avoid problems, the accessor function converts to the modern
scale, e.g. \code{x[["temperature"]]} yields temperature in the ITS90
scale, whether temperatures in the original file were reported on that scale
or the older IPTS-68 scale.
}

\section{Caution}{


Lacking detailed documentation of the ODF file format, the \code{\link[=read.odf]{read.odf()}} and
\code{\link[=read.ctd.odf]{read.ctd.odf()}} functions were crafted based on inspection of data files, and
so some guesses had to be made.

The \code{PARAMETER_HEADER} chunks describing quality-control flags are
a case in point.  These contain \code{NAME} components that
refer to other \code{PARAMETER_HEADER} chunks that hold measured data.  However,
those references are not always matched well with the data names, and
even if they do match, the cross-reference syntax used by
the Bedford Institute of Oceanography differs from that used by
l‚ÄôInstitut Maurice-Lamontagne. To simplify coding, it was assumed that
each quality-control sequence applies to the data sequence
immediately preceding it.  (This assumption is made in other
analysis systems.)

It is also prudent to pay attention to the units decoding,
which \code{\link[=read.odf]{read.odf()}} handles by calling \code{\link[=unitFromString]{unitFromString()}}.
Be on the lookout for incorrect temperature scales, which
are sometimes reported with nonstandard strings in ODF files.
Also, note that you may see warnings about conductivity ratios,
which some ODF files incorrectly suggest have dimensions.
}

\examples{
library(oce)
#
# 1. Read a CTD cast made on the Scotian Shelf. Note that the file's metadata
# states that conductivity is in S/m, but it is really conductivity ratio,
# so we must alter the unit before converting to a CTD object. Note that
# read.odf() on this data file produces a warning suggesting that the user
# repair the unit, using the method outlined here.
odf <- read.odf(system.file("extdata", "CTD_BCD2014666_008_1_DN.ODF.gz", package="oce"))
ctd <- as.ctd(odf) ## so we can e.g. extract potential temperature
ctd[["conductivityUnit"]] <- list(unit=expression(), scale="")
#
# 2. Make a CTD, and plot (with span to show NS)
plot(ctd, span=500)
#
# 3. Highlight bad data on TS diagram. (Note that the eos
# is specified, because we will extract practical-salinity and
# UNESCO-defined potential temperatures for the added points.)
plotTS(ctd, type="o", eos="unesco") # use a line to show loops
bad <- ctd[["QCFlag"]]!=0
points(ctd[['salinity']][bad],ctd[['theta']][bad],col='red',pch=20)

}
\references{
For sources that describe the ODF format, see the documentation
for the \linkS4class{odf} class.
}
\seealso{
\code{\link[=ODF2oce]{ODF2oce()}} will be an alternative to this, once (or perhaps if) a \code{ODF}
package is released by the Canadian Department of Fisheries and Oceans.

Other things related to odf data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{ODF2oce}()},
\code{\link{ODFListFromHeader}()},
\code{\link{ODFNames2oceNames}()},
\code{\link{[[,odf-method}},
\code{\link{[[<-,odf-method}},
\code{\link{odf-class}},
\code{\link{plot,odf-method}},
\code{\link{read.ctd.odf}()},
\code{\link{subset,odf-method}},
\code{\link{summary,odf-method}}
}
\author{
Dan Kelley, with help from Chantelle Layton
}
\concept{things related to odf data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/map.R
\name{mapImage}
\alias{mapImage}
\title{Add an Image to a Map}
\usage{
mapImage(
  longitude,
  latitude,
  z,
  zlim,
  zclip = FALSE,
  breaks,
  col,
  colormap,
  border = NA,
  lwd = par("lwd"),
  lty = par("lty"),
  missingColor = NA,
  filledContour = FALSE,
  gridder = "binMean2D",
  debug = getOption("oceDebug")
)
}
\arguments{
\item{longitude}{numeric vector of longitudes corresponding to \code{z} matrix.}

\item{latitude}{numeric vector of latitudes corresponding to \code{z} matrix.}

\item{z}{numeric matrix to be represented as an image.}

\item{zlim}{limit for z (color).}

\item{zclip}{A logical value, \code{TRUE} indicating that out-of-range
\code{z} values should be painted with \code{missingColor} and \code{FALSE}
indicating that these values should be painted with the nearest
in-range color.  If \code{zlim} is given then its min and max set the
range.  If \code{zlim} is not given but \code{breaks} is given, then
the min and max of \code{breaks} sets the range used for z.  If neither
\code{zlim} nor \code{breaks} is given, clipping is not done, i.e. the
action is as if \code{zclip} were \code{FALSE}.}

\item{breaks}{The z values for breaks in the color scheme.  If this is of
length 1, the value indicates the desired number of breaks, which is
supplied to \code{\link[=pretty]{pretty()}}, in determining clean break points.}

\item{col}{Either a vector of colors corresponding to the breaks, of length
1 plus the number of breaks, or a function specifying colors,
e.g. \code{\link[=oce.colorsViridis]{oce.colorsViridis()}} for the Viridis scheme.}

\item{colormap}{optional colormap, as created by \code{\link[=colormap]{colormap()}}.
If a \code{colormap} is provided, then its properties takes precedence
over \code{breaks}, \code{col}, \code{missingColor}, and \code{zclip}
specified to \code{mapImage}.}

\item{border}{Color used for borders of patches (passed to
\code{\link[=polygon]{polygon()}}); the default \code{NA} means no border.}

\item{lwd}{line width, used if borders are drawn.}

\item{lty}{line type, used if borders are drawn.}

\item{missingColor}{a color to be used to indicate missing data, or
\code{NA} to skip the drawing of such regions (which will retain
whatever material has already been drawn at the regions).}

\item{filledContour}{either a logical value indicating whether to use
filled contours to plot the image, or a numerical value indicating the
resampling rate to be used in interpolating from lon-lat coordinates to
x-y coordinates.  See \dQuote{Details} for how this interacts with
\code{gridder}.}

\item{gridder}{Name of gridding function used if \code{filledContour} is
\code{TRUE}.  This can be either \code{"binMean2D"} to select
\code{\link[=binMean2D]{binMean2D()}} or \code{"interp"} for
\code{\link[akima:interp]{akima::interp()}}. If not provided, then a selection is made
automatically, with \code{\link[=binMean2D]{binMean2D()}} being used if there are
more than 10,000 data points in the present graphical view. This
\code{"binMean2D"} method is much faster than \code{"interp"}.}

\item{debug}{A flag that turns on debugging.  Set to 1 to get a
moderate amount of debugging information, or to 2 to get more.}
}
\description{
Plot an image on an existing map that was created with \code{\link[=mapPlot]{mapPlot()}}.
(See example 4 for a way to start  with a blank map.)
}
\details{
The data are on a regular grid in lon-lat space, but not in the projected
x-y space.  This means that \code{\link[=image]{image()}} cannot be used.  Instead,
there are two approaches, depending on the value of \code{filledContour}.

If \code{filledContour} is \code{FALSE}, the image \verb{pixels'' are with [polygon()], which can be prohibitively slow for fine grids. However, if `filledContour` is `TRUE` or a numerical value, then the }pixels'' are remapped into a regular grid and then displayed with
\code{\link[=.filled.contour]{.filled.contour()}}.  The remapping starts by converting the
regular lon-lat grid to an irregular x-y grid using
\code{\link[=lonlat2map]{lonlat2map()}}.  This irregular grid is then interpolated onto a
regular x-y grid  with \code{\link[=binMean2D]{binMean2D()}} or with
\code{\link[akima:interp]{akima::interp()}} from the \code{akima} package, as determined by
the \code{gridder} argument.   If \code{filledContour} is \code{TRUE}, the
dimensions of the regular x-y grid is the same as that of the original
lon-lat grid; otherwise, the number of rows and columns are multiplied by
the numerical value of \code{filledContour}, e.g. the value 2 means to make
the grid twice as fine.

Filling contours can produce aesthetically-pleasing results, but the method
involves interpolation, so the data are not represented exactly and
analysts are advised to compare the results from the two methods (and
perhaps various grid refinement values) to guard against misinterpretation.

If a \code{\link[=png]{png()}} device is to be used, it is advised to supply
arguments \code{type="cairo"} and \code{antialias="none"}; see reference 1.
}
\examples{
\dontrun{
library(oce)
data(coastlineWorld)
data(topoWorld)

## Northern polar region, with color-coded bathymetry
par(mfrow=c(1,1), mar=c(2,2,1,1))
cm <- colormap(zlim=c(-5000, 0), col=oceColorsGebco)
drawPalette(colormap=cm)
mapPlot(coastlineWorld, projection="+proj=stere +lat_0=90",
        longitudelim=c(-180,180), latitudelim=c(70,110))
mapImage(topoWorld, colormap=cm)
mapGrid(15, 15, polarCircle=1, col=gray(0.2))
mapPolygon(coastlineWorld[['longitude']], coastlineWorld[['latitude']], col="tan")
}

}
\references{
\enumerate{
\item \url{https://codedocean.wordpress.com/2014/02/03/anti-aliasing-and-image-plots/}
}
}
\seealso{
A map must first have been created with \code{\link[=mapPlot]{mapPlot()}}.

Other functions related to maps: 
\code{\link{formatPosition}()},
\code{\link{lonlat2map}()},
\code{\link{lonlat2utm}()},
\code{\link{map2lonlat}()},
\code{\link{mapArrows}()},
\code{\link{mapAxis}()},
\code{\link{mapContour}()},
\code{\link{mapCoordinateSystem}()},
\code{\link{mapDirectionField}()},
\code{\link{mapGrid}()},
\code{\link{mapLines}()},
\code{\link{mapLocator}()},
\code{\link{mapLongitudeLatitudeXY}()},
\code{\link{mapPlot}()},
\code{\link{mapPoints}()},
\code{\link{mapPolygon}()},
\code{\link{mapScalebar}()},
\code{\link{mapText}()},
\code{\link{mapTissot}()},
\code{\link{oceCRS}()},
\code{\link{shiftLongitude}()},
\code{\link{usrLonLat}()},
\code{\link{utm2lonlat}()}
}
\author{
Dan Kelley
}
\concept{functions related to maps}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/rsk.R
\name{subset,rsk-method}
\alias{subset,rsk-method}
\title{Subset a Rsk Object}
\usage{
\S4method{subset}{rsk}(x, subset, ...)
}
\arguments{
\item{x}{an \linkS4class{rsk} object.}

\item{subset}{a condition to be applied to the \code{data} portion of \code{x}.
See \sQuote{Details}.}

\item{\dots}{ignored.}
}
\value{
An \linkS4class{rsk} object.
}
\description{
Subset a rsk object.  This function is somewhat analogous to
\code{\link[=subset.data.frame]{subset.data.frame()}}, but subsetting is only permitted by time.
}
\examples{
library(oce)
data(rsk)
plot(rsk)
plot(subset(rsk, time < mean(range(rsk[['time']]))))

}
\seealso{
Other things related to rsk data: 
\code{\link{[[,rsk-method}},
\code{\link{[[<-,rsk-method}},
\code{\link{as.rsk}()},
\code{\link{plot,rsk-method}},
\code{\link{read.rsk}()},
\code{\link{rsk-class}},
\code{\link{rskPatm}()},
\code{\link{rskToc}()},
\code{\link{rsk}},
\code{\link{summary,rsk-method}}

Other functions that subset oce objects: 
\code{\link{subset,adp-method}},
\code{\link{subset,adv-method}},
\code{\link{subset,amsr-method}},
\code{\link{subset,argo-method}},
\code{\link{subset,cm-method}},
\code{\link{subset,coastline-method}},
\code{\link{subset,ctd-method}},
\code{\link{subset,echosounder-method}},
\code{\link{subset,lobo-method}},
\code{\link{subset,met-method}},
\code{\link{subset,oce-method}},
\code{\link{subset,odf-method}},
\code{\link{subset,sealevel-method}},
\code{\link{subset,section-method}},
\code{\link{subset,topo-method}},
\code{\link{subset,xbt-method}}
}
\author{
Dan Kelley
}
\concept{functions that subset oce objects}
\concept{things related to rsk data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/windrose.R
\name{[[,windrose-method}
\alias{[[,windrose-method}
\title{Extract Something From a Windrose Object}
\usage{
\S4method{[[}{windrose}(x, i, j, ...)
}
\arguments{
\item{x}{a \linkS4class{windrose} object.}

\item{i}{character value indicating the name of an item to extract.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{ignored.}
}
\description{
Generally, the \code{[[} method lets users extract information from \code{oce}
objects, without having to know the details of the internal storage.  For
many \code{oce} sub-classes, \code{[[} can also return quantities that are computed
from the object's contents.
}
\details{
A two-step process is used to try to find the requested information. First, a
class-specific function is used (see \dQuote{Details of the Specialized
Method}).  If this yields nothing, then a general method is used (see
\dQuote{Details of the General Method}).  If both methods fail, then \code{[[}
returns NULL.

Some understanding of the subclass is required to know what can be retrieved
with \code{[[}. When dealing with an unfamiliar subclass, it can be useful to
first use \code{x[["?"]]} to get a listing of the retrievable items. See
\dQuote{Details of the Specialized Method} for more information.
}
\section{Details of the Specialized Method}{

\itemize{
\item If \code{i} is \code{"?"}, then the return value is a list
containing four items, each of which is a character vector
holding the names of things that can be accessed with \code{[[}.
The \code{data} and \code{metadata} items hold the names of
entries in the object's data and metadata
slots, respectively. The \code{metadataDerived} and
\code{dataDerived} items are both NULL.
}
}

\section{Details of the General Method}{


Note: the text of this section is identical for all \code{oce} subclasses, and so
some of what you read here may not be relevant to the class being described
in this help page.

If the specialized method produces no matches, the following generalized
method is applied. As with the specialized method, the procedure hinges first
on the values of \code{i} and, optionally, \code{j}. The work proceeds in steps, by
testing a sequence of possible conditions in sequence.
\enumerate{
\item A check is made as to whether \code{i} names one of the standard \code{oce} slots.
If so, \code{[[} returns the slot contents of that slot.  Thus, \code{x[["metadata"]]}
will retrieve the \code{metadata} slot, while \code{x[["data"]]} and
\code{x[["processingLog"]]} return those slots.
\item If \code{i} is a string ending in the \code{"Unit"}, then the characters preceding
that string are taken to be the name of an item in the data object, and a
list containing the unit is returned (or \code{NULL} if there is no such unit).
This list consists of an item named \code{unit}, which is an \code{\link[=expression]{expression()}}, and
an item named \code{scale}, which is a string describing the measurement scale.
If the string ends in \code{" unit"}, e.g.  \code{x[["temperature unit"]]} (note the
space), then just the expression is returned, and if it ends in \code{" scale"},
then just the scale is returned.
\item If \code{i} is a string ending in \code{"Flag"}, then the corresponding data-quality
flag is returned (or \code{NULL} if there is no such flag).
\item If the object holds hydrographic information (pressure, salinity,
temperature, longitude and latitude) then another set of possibilities
arises.  If \code{i} is \code{"sigmaTheta"}, then the value of \code{\link[=swSigmaTheta]{swSigmaTheta()}} is
called with \code{x} as the sole argument, and the results are returned.
Similarly, \code{\link[=swSigma0]{swSigma0()}} is used if \code{i="sigma0"}, and \code{\link[=swSpice]{swSpice()}} is used if
\code{i="spice"}. Of course, these actions only make sense for objects that
contain the relevant items within their \code{data} slot.
\item After these possibilities are eliminated, the action depends on whether
\code{j} has been provided.  If \code{j} is not provided, or is the string \code{""}, then
\code{i} is sought in the \code{metadata} slot, and then in the \code{data} slot, returning
whichever is found first.  In other words, if \code{j} is not provided, the
\code{metadata} slot takes preference over the \code{data} slot. However, if \code{j} is
provided, then it must be either the string \code{"metadata"} or \code{"data"}, and it
directs where to look.
\item If none of the above-listed conditions holds, then \code{NULL} is returned.
}
}

\seealso{
Other functions that extract parts of oce objects: 
\code{\link{[[,adp-method}},
\code{\link{[[,adv-method}},
\code{\link{[[,amsr-method}},
\code{\link{[[,argo-method}},
\code{\link{[[,bremen-method}},
\code{\link{[[,cm-method}},
\code{\link{[[,coastline-method}},
\code{\link{[[,ctd-method}},
\code{\link{[[,echosounder-method}},
\code{\link{[[,g1sst-method}},
\code{\link{[[,gps-method}},
\code{\link{[[,ladp-method}},
\code{\link{[[,landsat-method}},
\code{\link{[[,lisst-method}},
\code{\link{[[,lobo-method}},
\code{\link{[[,met-method}},
\code{\link{[[,oce-method}},
\code{\link{[[,odf-method}},
\code{\link{[[,rsk-method}},
\code{\link{[[,sealevel-method}},
\code{\link{[[,section-method}},
\code{\link{[[,tidem-method}},
\code{\link{[[,topo-method}},
\code{\link{[[,xbt-method}},
\code{\link{[[<-,adv-method}}

Other things related to windrose data: 
\code{\link{[[<-,windrose-method}},
\code{\link{as.windrose}()},
\code{\link{plot,windrose-method}},
\code{\link{summary,windrose-method}},
\code{\link{windrose-class}}
}
\concept{functions that extract parts of oce objects}
\concept{things related to windrose data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sun.R
\name{sunDeclinationRightAscension}
\alias{sunDeclinationRightAscension}
\title{Sun Declination and Right Ascension}
\usage{
sunDeclinationRightAscension(time, apparent = FALSE)
}
\arguments{
\item{time}{a POSIXct time. This ought to be in UTC timezone; if not,
the behaviour of this function is unlikely to be correct.}

\item{apparent}{logical value indicating whether to return
the 'apparent' angles.}
}
\value{
A list containing \code{declination} and \code{rightAscension}, in degrees.
}
\description{
The formulae are from Meeus (1991), chapter 24 (which uses chapter 21).
}
\examples{
## Example 24.a in Meeus (1991) (page 158 PDF, 153 print)
time <- as.POSIXct("1992-10-13 00:00:00", tz="UTC")
a <- sunDeclinationRightAscension(time, apparent=TRUE)
stopifnot(abs(a$declination - (-7.78507)) < 0.00004)
stopifnot(abs(a$rightAscension - (-161.61919)) < 0.00003)
b <- sunDeclinationRightAscension(time)
## check against previous results, to protect aginst code-drift errors
stopifnot(abs(b$declination - (-7.785464443)) < 0.000000001)
stopifnot(abs(b$rightAscension - (-161.6183305)) < 0.0000001)

}
\references{
\itemize{
\item Meeus, Jean. Astronomical Algorithms. Second Edition.
Richmond, Virginia, USA: Willmann-Bell, 1991.
}
}
\seealso{
Other things related to astronomy: 
\code{\link{angle2hms}()},
\code{\link{eclipticalToEquatorial}()},
\code{\link{equatorialToLocalHorizontal}()},
\code{\link{julianCenturyAnomaly}()},
\code{\link{julianDay}()},
\code{\link{moonAngle}()},
\code{\link{siderealTime}()},
\code{\link{sunAngle}()}
}
\author{
Dan Kelley, based on formulae in Meeus (1991).
}
\concept{things related to astronomy}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{grad}
\alias{grad}
\title{Calculate Matrix Gradient}
\usage{
grad(
  h,
  x = seq(0, 1, length.out = nrow(h)),
  y = seq(0, 1, length.out = ncol(h))
)
}
\arguments{
\item{h}{a matrix of values}

\item{x}{vector of coordinates along matrix columns (defaults to integers)}

\item{y}{vector of coordinates along matrix rows (defaults to integers)}
}
\value{
A list containing \eqn{|\nabla h|}{abs(grad(h))} as \code{g},
\eqn{\partial h/\partial x}{dh/dx} as \code{gx},
and \eqn{\partial h/\partial y}{dh/dy} as \code{gy},
each of which is a matrix of the same dimension as \code{h}.
}
\description{
In the interior of the matrix, centred second-order differences are used to
infer the components of the grad.  Along the edges, first-order differences
are used.
}
\examples{
## 1. Built-in volcano dataset
g <- grad(volcano)
par(mfrow=c(2, 2), mar=c(3, 3, 1, 1), mgp=c(2, 0.7, 0))
imagep(volcano, zlab="h")
imagep(g$g, zlab="|grad(h)|")
zlim <- c(-1, 1) * max(g$g)
imagep(g$gx, zlab="dh/dx", zlim=zlim)
imagep(g$gy, zlab="dh/dy", zlim=zlim)

## 2. Geostrophic flow around an eddy
library(oce)
dx <- 5e3
dy <- 10e3
x <- seq(-200e3, 200e3, dx)
y <- seq(-200e3, 200e3, dy)
R <- 100e3
h <- outer(x, y, function(x, y) 500*exp(-(x^2+y^2)/R^2))
grad <- grad(h, x, y)
par(mfrow=c(2, 2), mar=c(3, 3, 1, 1), mgp=c(2, 0.7, 0))
contour(x,y,h,asp=1, main=expression(h))
f <- 1e-4
gprime <- 9.8 * 1 / 1024
u <- -(gprime / f) * grad$gy
v <-  (gprime / f) * grad$gx
contour(x, y, u, asp=1, main=expression(u))
contour(x, y, v, asp=1, main=expression(v))
contour(x, y, sqrt(u^2+v^2), asp=1, main=expression(speed))

}
\seealso{
Other things relating to vector calculus: 
\code{\link{curl}()}
}
\author{
Dan Kelley, based on advice of Clark Richards, and mimicking a matlab function.
}
\concept{things relating to vector calculus}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{oceColorsViridis}
\alias{oceColorsViridis}
\alias{oce.colorsViridis}
\title{Create colors similar to the matlab Viridis scheme}
\usage{
oceColorsViridis(n)
}
\arguments{
\item{n}{number of colors to create.}
}
\description{
This is patterned on a \proglang{matlab}/\proglang{python} scheme that blends
from yellow to blue in a way that is designed to reproduce well
in black-and-white, and to be interpretable by those with
certain forms of color blindness.  See the references for
notes about issues of colour blindness in computer graphics.
An alternative
to \link{oceColorsViridis} is provided in the \CRANpkg{viridis} package, as illustrated
in Example 2.
}
\examples{
library(oce)
# Example 1: oceColorsViridis
imagep(volcano, col=oceColorsViridis(128),
       zlab="oceColorsViridis")
# Example 2: viridis::viridis
\dontrun{
imagep(volcano, col=viridis::viridis,
       zlab="viridis::viridis")}

}
\references{
The following references provide information on choosing
colour schemes, that are suitable for viewers who have
colour deficiencies.

Light, Adam, and Patrick J. Bartlein. "The End of the Rainbow? Color
Schemes for Improved Data Graphics." \emph{Eos, Transactions American Geophysical
Union} 85, no. 40 (2004): 385.
DOI: 10.1029/2004EO400002

Stephenson, David B. "Comment on 'Color Schemes for Improved Data
Graphics', by A Light and P.J. Bartlein." \emph{Eos, Transactions American
Geophysical Union} 86, no.  20 (2005): 196.
DOI: 10.1029/2005EO200005

Light, Adam, and Patrick J. Bartlein. "Reply to 'Comment on Color
Schemes for Improved Data Graphics,' by A. Light and P.J.  Bartlein'."
\emph{Eos, Transactions American Geophysical Union} 86, no. 20 (2005): 196‚Äì196.
DOI: 10.1029/2005EO200006

The following references provide information on choosing
colour schemes, that are suitable for viewers who have
colour deficiencies.

Light, Adam, and Patrick J. Bartlein. "The End of the Rainbow? Color
Schemes for Improved Data Graphics." \emph{Eos, Transactions American Geophysical
Union} 85, no. 40 (2004): 385.
DOI: 10.1029/2004EO400002

Stephenson, David B. "Comment on 'Color Schemes for Improved Data
Graphics', by A Light and P.J. Bartlein." \emph{Eos, Transactions American
Geophysical Union} 86, no.  20 (2005): 196.
DOI: 10.1029/2005EO200005

Light, Adam, and Patrick J. Bartlein. "Reply to 'Comment on Color
Schemes for Improved Data Graphics,' by A. Light and P.J.  Bartlein'."
\emph{Eos, Transactions American Geophysical Union} 86, no. 20 (2005): 196‚Äì196.
DOI: 10.1029/2005EO200006
}
\seealso{
Other things related to colors: 
\code{\link{colormapGMT}()},
\code{\link{colormap}()},
\code{\link{oceColors9B}()},
\code{\link{oceColorsCDOM}()},
\code{\link{oceColorsChlorophyll}()},
\code{\link{oceColorsClosure}()},
\code{\link{oceColorsDensity}()},
\code{\link{oceColorsFreesurface}()},
\code{\link{oceColorsGebco}()},
\code{\link{oceColorsJet}()},
\code{\link{oceColorsOxygen}()},
\code{\link{oceColorsPAR}()},
\code{\link{oceColorsPalette}()},
\code{\link{oceColorsPhase}()},
\code{\link{oceColorsSalinity}()},
\code{\link{oceColorsTemperature}()},
\code{\link{oceColorsTurbidity}()},
\code{\link{oceColorsTurbo}()},
\code{\link{oceColorsTwo}()},
\code{\link{oceColorsVelocity}()},
\code{\link{oceColorsVorticity}()},
\code{\link{ocecolors}}
}
\author{
Dan Kelley
}
\concept{things related to colors}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/amsr.R
\name{read.amsr}
\alias{read.amsr}
\title{Read an amsr File}
\usage{
read.amsr(file, debug = getOption("oceDebug"))
}
\arguments{
\item{file}{String indicating the name of a compressed file. See
\dQuote{File sources}.}

\item{debug}{A debugging flag, integer.}
}
\description{
Read a compressed amsr file, generating an \linkS4class{amsr} object.
Note that only compressed files are read in this version.
}
\details{
AMSR files are provided at the FTP site, at
\code{ftp.ssmi.com/amsr2/bmaps_v07.2} as of April 2021.
To acquire such files,
log in as "guest",
then enter a year-based directory (e.g. \code{y2016} for the year 2016),
then enter a month-based directory (e.g. \code{m08} for August, the 8th
month), and then download a file for the present date, e.g.
\code{f34_20160803v7.2.gz} for August 3rd, 2016. Do not uncompress
this file, since \code{read.amsr} can only read the raw files from the server.
If \code{read.amsr} reports an error on the number of chunks, try
downloading a similarly-named file (e.g. in the present example,
\code{read.amsr("f34_20160803v7.2_d3d.gz")} will report an error
about inability to read a 6-chunk file, but
\code{read.amsr("f34_20160803v7.2.gz")} will work properly.
}
\seealso{
\code{\link[=plot,amsr-method]{plot,amsr-method()}} for an example.

Other things related to amsr data: 
\code{\link{[[,amsr-method}},
\code{\link{[[<-,amsr-method}},
\code{\link{amsr-class}},
\code{\link{amsr}},
\code{\link{composite,amsr-method}},
\code{\link{download.amsr}()},
\code{\link{plot,amsr-method}},
\code{\link{subset,amsr-method}},
\code{\link{summary,amsr-method}}
}
\author{
Dan Kelley and Chantelle Layton
}
\concept{things related to amsr data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swTFreeze}
\alias{swTFreeze}
\title{Seawater freezing temperature}
\usage{
swTFreeze(
  salinity,
  pressure = NULL,
  longitude = NULL,
  latitude = NULL,
  saturation_fraction = 1,
  eos = getOption("oceEOS", default = "gsw")
)
}
\arguments{
\item{salinity}{Either practical salinity (PSU) or a \code{ctd} object from which
practical salinity and pressure (plus in the \code{eos="gsw"} case,
longitude and latitude) are inferred, using \code{\link[=lookWithin]{lookWithin()}}.}

\item{pressure}{Seawater pressure (dbar).}

\item{longitude}{Longitude of observation (only used if \code{eos="gsw"};
see \sQuote{Details}).}

\item{latitude}{Latitude of observation (only used if \code{eos="gsw"}; see
\sQuote{Details}).}

\item{saturation_fraction}{The saturation fraction of dissolved air in seawater,
ignored if \code{eos="unesco"}).}

\item{eos}{The equation of state, either \code{"unesco"} (references 1 and 2) or \code{"gsw"}.
(references 3 and 4).}
}
\value{
Temperature (\eqn{^\circ}{deg}C), defined on the ITS-90 scale.
}
\description{
Compute in-situ freezing temperature of seawater, using either the UNESCO formulation
(computed as in Section 5 of reference 1) or the GSW formulation (computed
by using \code{\link[gsw:gsw_SA_from_SP]{gsw::gsw_SA_from_SP()}} to get Absolute Salinity, and
then \code{\link[gsw:gsw_t_freezing]{gsw::gsw_t_freezing()}} to get the freezing temperature).
}
\details{
If the first argument is an \code{oce} object, and if the \code{pressure}
argument is \code{NULL}, then the pressure is sought within the first
argument. In the case of \code{eos="gsw"}, then a similar procedure
also applies to the \code{longitude} and \code{latitude} arguments.
}
\examples{
# 1. Test for a check-value given in reference 1. This value, -2.588567 degC,
# is in the 1968 temperature scale (IPTS-68), but swTFreeze reports
# in the newer ITS-90 scale, so we must convert before checking.
Tcheck <- -2.588567 # IPTS-68
T <- swTFreeze(salinity=40, pressure=500, eos="unesco")
stopifnot(abs(Tcheck - T68fromT90(T)) < 1e-6)

# 2. Compare unesco and gsw formulations.
data(ctd)
p <- ctd[["pressure"]]
par(mfrow=c(1, 2), mar=c(3, 3, 1, 2), mgp=c(2, 0.7, 0))
plot(swTFreeze(ctd, eos="unesco"),
     p, xlab="unesco", ylim=rev(range(p)))
plot(swTFreeze(ctd, eos="unesco") - swTFreeze(ctd, eos="gsw"),
     p, xlab="unesco-gsw", ylim=rev(range(p)))

}
\references{
\enumerate{
\item Fofonoff, N. P., and R. C. Millard. \dQuote{Algorithms for Computation of
Fundamental Properties of Seawater.} UNESCO Technical Papers in Marine
Research. SCOR working group on Evaluation of CTD data; UNESCO/ICES/SCOR/IAPSO
Joint Panel on Oceanographic Tables and Standards, 1983.
https://unesdoc.unesco.org/ark:/48223/pf0000059832.
\item Gill, A E. Atmosphere-Ocean Dynamics. New York, NY, USA: Academic Press,
\item 
\item IOC, SCOR, and IAPSO (2010). The international thermodynamic equation of
seawater-2010: Calculation and use of thermodynamic properties.  Technical
Report 56, Intergovernmental Oceanographic Commission, Manuals and Guide, 2010.
\item McDougall, Trevor J., and Paul M. Barker. Getting Started with TEOS-10 and
the Gibbs Seawater (GSW) Oceanographic Toolbox. SCOR/IAPSO WG127, 2011.
}
}
\seealso{
Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/map.R
\name{usrLonLat}
\alias{usrLonLat}
\title{Calculate lon-lat coordinates of plot-box trace}
\usage{
usrLonLat(n = 25, debug = getOption("oceDebug"))
}
\arguments{
\item{n}{number of points to check along each side of the plot box}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\value{
A list containing numerical
values \code{lonmin}, \code{lonmax}, \code{latmin}, and \code{latmax}, along
with logical value \code{ok}. The last of these indicates
whether at least one point on the plot box is invertible.
Note that longitudes are in the range from -180 to 180 degrees.
}
\description{
Trace along the plot box, converting from xy coordinates to lonlat
coordinates. The results are used by \code{\link[=mapGrid]{mapGrid()}}
and \code{\link[=mapAxis]{mapAxis()}} to ignore out-of-frame grid
lines and axis labels.
}
\details{
Some projections, such as \code{"wintri"}, have trouble inverting
points that are "off the globe".  In such cases,
the returned value has \code{lonmin}, \code{lonmax}, \code{latmin} and
\code{latmax} set to \code{NA}, and \code{ok} set to \code{FALSE}.
}
\seealso{
Other functions related to maps: 
\code{\link{formatPosition}()},
\code{\link{lonlat2map}()},
\code{\link{lonlat2utm}()},
\code{\link{map2lonlat}()},
\code{\link{mapArrows}()},
\code{\link{mapAxis}()},
\code{\link{mapContour}()},
\code{\link{mapCoordinateSystem}()},
\code{\link{mapDirectionField}()},
\code{\link{mapGrid}()},
\code{\link{mapImage}()},
\code{\link{mapLines}()},
\code{\link{mapLocator}()},
\code{\link{mapLongitudeLatitudeXY}()},
\code{\link{mapPlot}()},
\code{\link{mapPoints}()},
\code{\link{mapPolygon}()},
\code{\link{mapScalebar}()},
\code{\link{mapText}()},
\code{\link{mapTissot}()},
\code{\link{oceCRS}()},
\code{\link{shiftLongitude}()},
\code{\link{utm2lonlat}()}
}
\author{
Dan Kelley
}
\concept{functions related to maps}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{threenum}
\alias{threenum}
\title{Calculate min, mean, and max values}
\usage{
threenum(x)
}
\arguments{
\item{x}{a vector or matrix of numerical values.}
}
\value{
A character vector of three values: the minimum, the mean, the
maximum.
}
\description{
This is a simpler cousin of the standard \code{\link[=fivenum]{fivenum()}} function,
used in \code{\link[=summary]{summary()}} functions for \code{oce} objects.
}
\section{Historical note}{

On Aug 5, 2019, the dimension was dropped as the fourth column, and
this function returned to the original intention (revealed by its
name).  Another change is that the function now returns numerical
results, leaving the task of setting the number of digits to
\code{\link[=summary]{summary()}}.
}

\examples{
library(oce)
threenum(1:10)
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swSigma4}
\alias{swSigma4}
\title{Seawater potential density anomaly referenced to 4000db pressure}
\usage{
swSigma4(
  salinity,
  temperature = NULL,
  pressure = NULL,
  longitude = NULL,
  latitude = NULL,
  eos = getOption("oceEOS", default = "gsw")
)
}
\arguments{
\item{salinity}{either practical salinity (in which case \code{temperature}
and \code{pressure} must be provided) \emph{or} an \code{oce} object, in
which case \code{salinity}, \code{temperature} (in the ITS-90 scale; see
next item), etc. are inferred from the object, ignoring the
other parameters, if they are supplied.}

\item{temperature}{\emph{in-situ} temperature (\eqn{^\circ}{deg}C), defined
on the ITS-90 scale.  This scale is used by GSW-style calculation (as
requested by setting \code{eos="gsw"}), and is the value contained within
\code{ctd} objects (and probably most other objects created with data
acquired in the past decade or two). Since the UNESCO-style calculation is
based on IPTS-68, the temperature is converted within the present function,
using \code{\link[=T68fromT90]{T68fromT90()}}.}

\item{pressure}{pressure (dbar)}

\item{longitude}{longitude of observation (only used if \code{eos="gsw"};
see \sQuote{Details}).}

\item{latitude}{latitude of observation (only used if \code{eos="gsw"}; see
\sQuote{Details}).}

\item{eos}{equation of state, either \code{"unesco"} (references 1 and 2)
or \code{"gsw"} (references 3 and 4).}
}
\value{
Potential density anomaly (kg/m\eqn{^3}{^3}).
}
\description{
Compute \eqn{\sigma_\theta}{sigma}, the potential density of seawater (minus
1000 kg/m\eqn{^3}{^3}), referenced to 4000db pressures.
This is defined as
\eqn{\sigma_1=\sigma_\theta=\rho(S,\theta(S,t,p),4000}{sigma_1=sigma_theta=rho(S,theta(S,t,p),4000)}
\itemize{
\item 1000 kg/m\eqn{^3}{^3}.
}
}
\references{
See citations provided in the \code{\link[=swRho]{swRho()}} documentation.
}
\seealso{
Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{snakeToCamel}
\alias{snakeToCamel}
\title{Convert each of a vector of strings from SNAKE_CASE to camelCase}
\usage{
snakeToCamel(s, specialCases = NULL)
}
\arguments{
\item{s}{A vector of character values.}

\item{specialCases}{A vector of character values that tell which
special-cases to apply, or \code{NULL} (the default) to turn off special
cases.  The only permitted special case at the moment is \code{"QC"} (see
\dQuote{Details}) but the idea of this argument is that other cases
can be added later, if needed.}
}
\value{
A vector of character values
}
\description{
\code{snakeToCamel} converts "snake-case" characters such as \code{"NOVA_SCOTIA"}
to "camel-case" values, such as \code{"NovaScotia"}.  It was written for
use by \code{\link[=read.argo]{read.argo()}}, but it also may prove helpful in other contexts.
}
\details{
The basic procedure is to chop the string up into substrings separated by
the underline character, then to upper-case the first letter of
all substrings except the first, and then to paste the substrings
together.

However, there are exceptions.  First, any upper-case string that contains no
underlines is converted to lower case, but any mixed-case string with no
underlines is returned as-is (see the second example). Second, if
the \code{specialCases} argument contains \code{"QC"}, then the
\code{QC} is passed through directly (since it is an acronym) and
if the first letter of remaining text is upper-cased (contrast
see the four examples).
}
\examples{
library(oce)
snakeToCamel("PARAMETER_DATA_MODE")   # "parameterDataMode"
snakeToCamel("PARAMETER")             # "parameter"
snakeToCamel("HISTORY_QCTEST")        # "historyQctest"
snakeToCamel("HISTORY_QCTEST", "QC")  # "historyQCTest"
snakeToCamel("PROFILE_DOXY_QC")       # "profileDoxyQc"
snakeToCamel("PROFILE_DOXY_QC", "QC") # "profileDoxyQC"
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/topo.R
\name{download.topo}
\alias{download.topo}
\title{Download and Cache a topo File}
\usage{
download.topo(
  west,
  east,
  south,
  north,
  resolution = 4,
  destdir = ".",
  destfile,
  format,
  server = "https://gis.ngdc.noaa.gov",
  debug = getOption("oceDebug")
)
}
\arguments{
\item{west, east}{numeric values for the limits of the data-selection box, in degrees.
These are converted to the -180 to 180 degree notation, if needed.
Then, \code{west} is rounded down to the nearest 1/100th degree, and \code{east}
is rounded up to the the nearest 1/100th degree. The results of these
operations are used in constructing the query for the NOAA data server.}

\item{south, north}{latitude limits, treated in a way that
corresponds to the longitude limits.}

\item{resolution}{numeric value of grid spacing, in geographical minutes.
The default value is 4 minutes, corresponding to 4 nautical miles (approx. 7.4km)
in the north-south direction, and less in the east-west direction.}

\item{destdir}{Optional string indicating the directory in which to store
downloaded files. If not supplied, \code{"."} is used, i.e. the data file
is stored in the present working directory.}

\item{destfile}{Optional string indicating the name of the file.
If not supplied, the file name is constructed from the other
parameters of the function call, so subsequent calls with the same
parameters will yield the same result, thus providing the key
to the caching scheme.}

\item{format}{Deprecated, and ignored, as of June 2020.}

\item{server}{character value specifying the base from which a
download URL will be constructed.  It is unlikely that any value
other than the default will work, unless it is a similarly-constructed
mirrored site.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\value{
String indicating the full pathname to the downloaded file.
}
\description{
Topographic data are downloaded from a data server that holds the ETOPO1
dataset (Amante, C. and B.W. Eakins, 2009), and saved as a netCDF file whose
name specifies the data request, if a file of that name is not already
present on the local file system.  The return value is the name of the data
file, and its typical use is as the filename for a call to \code{\link[=read.topo]{read.topo()}}.
Given the rules on file naming, subsequent calls to \code{download.topo}
with identical parameters will simply return the name of the cached file,
assuming the user has not deleted it in the meantime.
}
\details{
The specified longitude and latitude limits are rounded to 2 digits
(corresponding to a footprint of approximately 1km), and these are used
in the server request. If the resultant request would generate under
1 row or column in the result, \code{download.topo} generates an
error message and stops.
}
\section{Historical note relating to NOAA server changes}{

In May of 2020, \code{download.topo} stopped working, evidently owing
to changes in the NOAA server API, which had been inferred by reverse
engineering a NOAA data-request website. However, the
\code{marmap} function \code{getNOAA.bathy}
was found to be working at that time, and so \code{download.topo} was revised based on
that function.  The problem of keeping up with changing data-server APIs should
be easier in the future, since NOAA has made the API public.
}

\examples{
\dontrun{
library(oce)
topoFile <- download.topo(west=-66, east=-60, south=43, north=47,
                          resolution=1, destdir="~/data/topo")
topo <- read.topo(topoFile)
imagep(topo, zlim=c(-400, 400), col=oceColorsTwo, drawTriangles=TRUE)
if (requireNamespace("ocedata", quietly=TRUE)) {
    data(coastlineWorldFine, package="ocedata")
    lines(coastlineWorldFine[["longitude"]], coastlineWorldFine[["latitude"]])
}
}

}
\references{
\itemize{
\item Amante, C. and B.W. Eakins, 2009. ETOPO1 1 Arc-Minute Global Relief
Model: Procedures, Data Sources and Analysis. NOAA Technical Memorandum
NESDIS NGDC-24. National Geophysical Data Center, NOAA. doi:10.7289/V5C8276M
}
}
\seealso{
Other functions that download files: 
\code{\link{download.amsr}()},
\code{\link{download.coastline}()},
\code{\link{download.met}()}

Other things related to topo data: 
\code{\link{[[,topo-method}},
\code{\link{[[<-,topo-method}},
\code{\link{as.topo}()},
\code{\link{plot,topo-method}},
\code{\link{read.topo}()},
\code{\link{subset,topo-method}},
\code{\link{summary,topo-method}},
\code{\link{topo-class}},
\code{\link{topoInterpolate}()},
\code{\link{topoWorld}}
}
\author{
Dan Kelley
}
\concept{functions that download files}
\concept{things related to topo data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/spectral.R
\name{pwelch}
\alias{pwelch}
\title{Welch periodogram}
\usage{
pwelch(
  x,
  window,
  noverlap,
  nfft,
  fs,
  spec,
  demean = FALSE,
  detrend = TRUE,
  plot = TRUE,
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{x}{a vector or timeseries to be analyzed.  If \code{x} is a timeseries, then
it there is no need to \code{fs}, and doing so will result in an error if it does
not match the value inferred from \code{x}.}

\item{window}{optional numeric vector specifying a window to be applied
to the timeseries subsamples.  This is ignored if \code{spec} is provided.
Otherwise, if \code{window} is provided, then it must either
be of the same length as \code{nfft} or be of length 1.  In the first case,
the vector is multiplied into the timeseries subsample, and the length
of \code{window} must equal \code{nfft} is that is supplied.
In the second then \code{window} is taken to be the number
of sub-intervals into which the time series is to be broken up, with a
hamming window being used for each sub-interval.  If \code{window} is not
specified and \code{nfft} is given, then the window is constructed as
a hamming window with length \code{nfft}.  And, if neither \code{window}
nor \code{nfft} are specified, then \code{x} will be broken up
into 8 portions.}

\item{noverlap}{number of points to overlap between windows.  If not
specified, this will be set to half the window length.}

\item{nfft}{length of FFT. See \code{window} for how \code{nfft} interacts with
that argument.}

\item{fs}{frequency of time-series.  If \code{x} is a time-series, and if
\code{fs} is supplied, then time-series is altered to have frequency
\code{fs}.}

\item{spec}{optional function to be used for the computation of the spectrum,
to allow finer-grained control of the processing.
If provided, \code{spec} must accept a time-series as its first argument, and
must return a list containing the spectrum in \code{spec} and the
frequency in \code{freq}.
Note that no window will be applied to the data after subsampling,
and an error will be reported if \code{window} and \code{spec} are both given.
An error will be reported if \code{spec} is given but \code{nfft} is not given.
Note that the values of \code{demean}, \code{detrend} and \code{plot} are ignored if \code{spec}
is given. However, the \dots argument \emph{is} passed to \code{spec}.}

\item{demean, detrend}{logical values that can control the spectrum calculation,
in the default case of \code{spec}.  These are passed to \code{\link[=spectrum]{spectrum()}} and thence
\code{\link[=spec.pgram]{spec.pgram()}}; see the help pages for the latter for an explanation.}

\item{plot}{logical, set to \code{TRUE} to plot the spectrum.}

\item{debug}{a flag that turns on debugging.  Set to 1 to get a moderate
amount of debugging information, or to 2 to get more.}

\item{\dots}{optional extra arguments to be passed to
\code{\link[=spectrum]{spectrum()}}, or to \code{spec}, if the latter is given.}
}
\value{
\code{pwelch} returns a list mimicking the return value from \code{\link[=spectrum]{spectrum()}},
containing frequency \code{freq}, spectral power \code{spec}, degrees of
freedom \code{df}, bandwidth \code{bandwidth}, etc.
}
\description{
Compute periodogram using the Welch (1967) method. This is
somewhat analogous to the Matlab function of the same name,
but it is \emph{not} intended as a drop-in replacement.
}
\details{
First, \code{x} is broken up into chunks,
overlapping as specified by \code{noverlap}.  These chunks are then
multiplied by the window, and then
passed to \code{\link[=spectrum]{spectrum()}}.  The resulting spectra are then averaged,
with the results being stored in \code{spec} of the return value.  Other
entries of the return value mimic those returned by \code{\link[=spectrum]{spectrum()}}.

It should be noted that the actions of several parameters are interlocked,
so this can be a complex function to use.  For example, if \code{window} is
given and has length exceeding 1, then its length must equal \code{nfft}, if the
latter is also provided.
}
\section{Bugs}{

Both bandwidth and degrees of freedom are just copied from
the values for one of the chunk spectra, and are thus incorrect.  That means
the cross indicated on the graph is also incorrect.
}

\section{Historical notes}{

\itemize{
\item \strong{2021-06-26:} Until this date, \code{\link[=pwelch]{pwelch()}} passed the
subsampled timeseries portions through \code{\link[=detrend]{detrend()}}
before applying the window. This practice was dropped
because it could lead to over-estimates of low frequency
energy (as noticed by Holger Foysi of the University of Siegen),
perhaps because \code{\link[=detrend]{detrend()}} considers only endpoints and
therefore can yield inaccurate trend estimates.
In a related change, \code{demean} and \code{detrend} were added
as formal arguments, to avoid users having to trace the documentation
for \code{\link[=spectrum]{spectrum()}} and then \code{\link[=spec.pgram]{spec.pgram()}}, to learn how to
remove means and trends from data.
For more control, the \code{spec} argument was
added to let users sidestep \code{\link[=spectrum]{spectrum()}} entirely, by providing
their own spectral computation functions.
}
}

\examples{
library(oce)
Fs <- 1000
t <- seq(0, 0.296, 1/Fs)
x <- cos(2 * pi * t * 200) + rnorm(n=length(t))
X <- ts(x, frequency=Fs)
s <- spectrum(X, spans=c(3,2), main="random + 200 Hz", log='no')
w <- pwelch(X, plot=FALSE)
lines(w$freq, w$spec, col="red")
w2 <- pwelch(X, nfft=75, plot=FALSE)
lines(w2$freq, w2$spec, col='green')
abline(v=200, col="blue", lty="dotted")
cat("Checking spectral levels with Parseval's theorem:\n")
cat("var(x)                              = ", var(x), "\n")
cat("2 * sum(s$spec) * diff(s$freq[1:2]) = ", 2 * sum(s$spec) * diff(s$freq[1:2]), "\n")
cat("sum(w$spec) * diff(s$freq[1:2])     = ", sum(w$spec) * diff(w$freq[1:2]), "\n")
cat("sum(w2$spec) * diff(s$freq[1:2])    = ", sum(w2$spec) * diff(w2$freq[1:2]), "\n")
## co2
par(mar=c(3,3,2,1), mgp=c(2,0.7,0))
s <- spectrum(co2, plot=FALSE)
plot(log10(s$freq), s$spec * s$freq,
     xlab=expression(log[10]*Frequency), ylab="Power*Frequency", type='l')
title("Variance-preserving spectrum")
pw <- pwelch(co2, nfft=256, plot=FALSE)
lines(log10(pw$freq), pw$spec * pw$freq, col='red')
}
\references{
Welch, P. D., 1967. The Use of Fast Fourier Transform for the
Estimation of Power Spectra: A Method Based on Time Averaging Over Short,
Modified Periodograms. \emph{IEEE Transactions on Audio Electroacoustics},
AU-15, 70--73.
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/section.R
\name{longitudeTighten}
\alias{longitudeTighten}
\title{Try to Reduce Section Longitude Range}
\usage{
longitudeTighten(section)
}
\arguments{
\item{section}{a \linkS4class{section} object.}
}
\value{
A \linkS4class{section} object based on its first argument, but with
longitudes shifted in its \code{metadata} slot, and also in the \code{metadata} slots
of each of the \linkS4class{ctd} objects that are stored in the \code{station} item
in its \code{data} slot.
}
\description{
\link{longitudeTighten} shifts some longitudes in its first argument by 360 degrees, if doing
so will reduce the overall longitude span.
}
\details{
This function can be helpful in cases where the CTD stations within a section
cross the cut point of the longitude convention, which otherwise might
yield ugly plots if \code{\link[=plot,section-method]{plot,section-method()}} is used with \code{xtype="longitude"}.
This problem does occur with CTD objects ordered by time of sampling,
but was observed in December 2020 for a GO-SHIPS dataset downloaded from
\verb{https://cchdo.ucsd.edu/data/15757/a10_1992_ct1}.
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{approx3d}
\alias{approx3d}
\title{Trilinear interpolation in a 3D array}
\usage{
approx3d(x, y, z, f, xout, yout, zout)
}
\arguments{
\item{x}{vector of x values for grid (must be equi-spaced)}

\item{y}{vector of y values for grid (must be equi-spaced)}

\item{z}{vector of z values for grid (must be equi-spaced)}

\item{f}{matrix of rank 3, with the gridded values mapping to the \code{x}
values (first index of \code{f}), etc.}

\item{xout}{vector of x values for output.}

\item{yout}{vector of y values for output (length must match that of
\code{xout}).}

\item{zout}{vector of z values for output (length must match that of
\code{xout}).}
}
\value{
A vector of interpolated values (or \code{NA} values), with length
matching that of \code{xout}.
}
\description{
Interpolate within a 3D array, using the trilinear approximation.
}
\details{
Trilinear interpolation is used to interpolate within the \code{f} array,
for those (\code{xout}, \code{yout} and \code{zout}) triplets that are
inside the region specified by \code{x}, \code{y} and \code{z}.  Triplets
that lie outside the range of \code{x}, \code{y} or \code{z} result in
\code{NA} values.
}
\examples{
## set up a grid
library(oce)
n <- 5
x <- seq(0, 1, length.out=n)
y <- seq(0, 1, length.out=n)
z <- seq(0, 1, length.out=n)
f <- array(1:n^3, dim=c(length(x), length(y), length(z)))
## interpolate along a diagonal line
m <- 100
xout <- seq(0, 1, length.out=m)
yout <- seq(0, 1, length.out=m)
zout <- seq(0, 1, length.out=m)
approx <- approx3d(x, y, z, f, xout, yout, zout)
## graph the results
plot(xout, approx, type='l')
points(xout[1], f[1, 1, 1])
points(xout[m], f[n,n,n])
}
\author{
Dan Kelley and Clark Richards
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/map.R
\name{oceProject}
\alias{oceProject}
\title{Wrapper to sf::sf_project()}
\usage{
oceProject(xy, proj, inv = FALSE, debug = getOption("oceDebug"))
}
\arguments{
\item{xy}{two-column numeric matrix specifying locations.  If \code{inv} is False, then \code{xy[,1]} will hold
longitude and \code{xy[,2]} will hold latitude, but if \code{inv} is True, then the columns will be easting
and northing values (in metres).}

\item{proj}{character string indicating the desired map projection, or an object of class \code{crs};
see the documentation for \code{\link[sf:sf_project]{sf::sf_project()}}.}

\item{inv}{logical value, False by default, indicating whether an inverse projection is requested.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\value{
A two-column matrix, with first column holding either
\code{longitude} or \code{x}, and second column holding either
\code{latitude} or \code{y}.
}
\description{
This function is used to isolate other oce functions from
changes to the map-projection functions that are done in the \CRANpkg{sf}
package.  (Until 2020 December, the \CRANpkg{rgdal} package was used,
after a year of tests ensuring that the results of the two packages were
the same.)
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/met.R
\name{metNames2oceNames}
\alias{metNames2oceNames}
\title{Convert met Data Name to Oce Name}
\usage{
metNames2oceNames(names, scheme)
}
\arguments{
\item{names}{a vector of character strings with original names}

\item{scheme}{an optional indication of the scheme that is employed. This may
be \code{"ODF"}, in which case \code{\link[=ODFNames2oceNames]{ODFNames2oceNames()}} is used,
or \code{"met"}, in which case some tentative code for met files is used.}
}
\value{
Vector of strings for the decoded names. If an unknown scheme is provided,
this will just be \code{names}.
}
\description{
Interoperability between oce functions requires that standardized data names
be used, e.g. \code{"temperature"} for in-situ temperature. Very few
data-file headers name the temperature column in exactly that way, however,
and this function is provided to try to guess the names. The task is complicated
by the fact that Environment Canada seems to change the names of the columns,
e.g. sometimes a symbol is used for the degree sign, other times not.
}
\details{
Several quantities in the returned object differ from their values in the source
file. For example, speed is converted from km/h to m/s, and angles are converted
from tens of degrees to degrees. Also, some items are created from scratch, e.g.
\code{u} and \code{v}, the eastward and northward velocity, are computed from speed
and direction. (Note that e.g. u is positive if the wind blows to the east; the
data are thus in the normal Physics convention.)
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/gps.R
\docType{class}
\name{gps-class}
\alias{gps-class}
\title{Class to Store GPS Data}
\description{
This class stores GPS data. These objects may be read with
\code{\link[=read.gps]{read.gps()}} or assembled with \code{\link[=as.gps]{as.gps()}}.
}
\section{Slots}{

\describe{
\item{\code{data}}{As with all \code{oce} objects, the \code{data} slot
for \code{gps} objects is a
\link{list} containing the main data for the object.
{}}

\item{\code{metadata}}{As with all \code{oce} objects, the \code{metadata} slot
for \code{gps} objects is a \link{list} containing
information about the \code{data} or about the object itself.
{}}

\item{\code{processingLog}}{As with all \code{oce} objects, the \code{processingLog} slot
for \code{gps} objects is a
\link{list} with entries describing the creation and evolution
of the object. The contents are updated by various \code{oce} functions to
keep a record of processing steps.  Object summaries and
\code{\link[=processingLogShow]{processingLogShow()}} both display the log.}
}}

\section{Modifying slot contents}{


Although the \code{[[<-} operator may permit modification of the contents
of \linkS4class{gps} objects (see \code{\link{[[<-,gps-method}}),
it is better to use \code{\link[=oceSetData]{oceSetData()}} and \code{\link[=oceSetMetadata]{oceSetMetadata()}},
because those functions save an entry in the \code{processingLog}
that describes the change.
}

\section{Retrieving slot contents}{


The full contents of the \code{data} and \code{metadata} slots of a \linkS4class{gps}
object may be retrieved in the standard R way using \code{\link[=slot]{slot()}}. For
example \code{slot(o,"data")} returns the \code{data} slot of an object named \code{o},
and similarly \code{slot(o,"metadata")} returns
the \code{metadata} slot.

The slots may also be obtained with the \code{\link{[[,gps-method}}
operator, as e.g. \code{o[["data"]]} and \code{o[["metadata"]]}, respectively.

The \code{\link{[[,gps-method}} operator can also
be used to retrieve items from within the \code{data} and \code{metadata} slots.
For example, \code{o[["temperature"]]} can be used to retrieve temperature
from an object containing that quantity. The rule is that a named
quantity is sought first within the object's \code{metadata} slot,
with the \code{data} slot being checked only if \code{metadata} does not
contain the item. This \code{[[} method can also be used to get
certain derived quantities, if the object contains sufficient
information to calculate them. For example, an object that holds
(practical) salinity, temperature and pressure, along with
longitude and latitude, has sufficient information to compute
Absolute Salinity, and so \code{o[["SA"]]} will yield the
calculated Absolute Salinity.

It is also possible to find items more directly, using \code{\link[=oceGetData]{oceGetData()}} and
\code{\link[=oceGetMetadata]{oceGetMetadata()}}, but neither of these functions can
retrieve derived items.
}

\seealso{
Other things related to gps data: 
\code{\link{[[,gps-method}},
\code{\link{[[<-,gps-method}},
\code{\link{as.gps}()},
\code{\link{plot,gps-method}},
\code{\link{read.gps}()},
\code{\link{summary,gps-method}}
}
\author{
Dan Kelley
}
\concept{things related to gps data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/extdata.R
\docType{data}
\encoding{UTF-8}
\name{d201211_0011.cnv}
\alias{d201211_0011.cnv}
\title{Sample ctd dataset in .cnv format}
\description{
Sample ctd dataset in .cnv format
}
\examples{
\dontrun{
read.oce(system.file("extdata", "d201211_0011.cnv", package="oce"))
}

}
\seealso{
Other raw datasets: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{adp_rdi.000}},
\code{\link{ctd.cnv}},
\code{\link{d200321-001.ctd}},
\code{\link{test_met_csv1.csv}},
\code{\link{test_met_csv2.csv}},
\code{\link{test_met_xml2.xml}},
\code{\link{xbt.edf}}

Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}
}
\concept{raw datasets}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.nortek.R
\name{read.aquadoppProfiler}
\alias{read.aquadoppProfiler}
\title{Read a Nortek Aquadopp-Profiler File}
\usage{
read.aquadoppProfiler(
  file,
  from = 1,
  to,
  by = 1,
  tz = getOption("oceTz"),
  longitude = NA,
  latitude = NA,
  orientation,
  distance,
  monitor = FALSE,
  despike = FALSE,
  processingLog,
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{file}{a connection or a character string giving the name of the file
to load.  (For \code{read.adp.sontek.serial}, this is generally a list of
files, which will be concatenated.)}

\item{from}{indication of the first profile to read.  This can be an
integer, the sequence number of the first profile to read, or a POSIXt time
before which profiles should be skipped, or a character string that converts
to a POSIXt time (assuming UTC timezone).  See \dQuote{Examples}, and make
careful note of the use of the \code{tz} argument. If \code{from} is not
supplied, it defaults to 1.}

\item{to}{an optional indication of the last profile to read, in a
format as described for \code{from}.  As a special case, \code{to=0} means
to read the file to the end.  If \code{to} is not supplied, then it defaults
to 0.}

\item{by}{an optional indication of the stride length to use while walking through
the file.  If this is an integer, then \code{by-1} profiles are skipped
between each pair of profiles that is read, e.g. the default \code{by=1}
means to read all the data.  (For RDI files \emph{only}, there are some
extra features to avoid running out of memory; see \dQuote{Memory considerations}.)}

\item{tz}{character string indicating time zone to be assumed in the data.}

\item{longitude}{optional signed number indicating the longitude in degrees
East.}

\item{latitude}{optional signed number indicating the latitude in degrees
North.}

\item{orientation}{Optional character string specifying the orientation of the
sensor, provided for those cases in which it cannot be inferred from the
data file.  The valid choices are \code{"upward"}, \code{"downward"}, and
\code{"sideward"}.}

\item{distance}{Optional vector holding the distances of bin centres from the
sensor.  This argument is ignored except for Nortek profilers, and need not
be given if the function determines the distances correctly from the data.
The problem is that the distance is poorly documented in the Nortek System
Integrator Guide (2008 edition, page 31), so the function must rely on
word-of-mouth formulae that do not work in all cases.}

\item{monitor}{boolean value indicating whether to indicate the progress
of reading the file, by using \code{\link[=txtProgressBar]{txtProgressBar()}} or otherwise.  The value
of \code{monitor} is changed to \code{FALSE} automatically, for non-interactive
sessions.}

\item{despike}{if \code{TRUE}, \code{\link[=despike]{despike()}} will be used to clean
anomalous spikes in heading, etc.}

\item{processingLog}{if provided, the action item to be stored in the log.
(Typically only provided for internal calls; the default that it provides is
better for normal calls by a user.)}

\item{debug}{a flag that turns on debugging.  Set to 1 to get a moderate
amount of debugging information, or to 2 to get more.}

\item{\dots}{optional additional arguments that some (but not all)
\verb{read.adp.*()} functions pass to lower-level functions.}
}
\value{
An \linkS4class{adp} object.
The contents of that object make sense for the particular instrument
type under study, e.g. if the data file contains
NMEA strings, then navigational data will be stored in an item
called \code{nmea} in the \code{data} slot).
}
\description{
The R code is based on information in
the Nortek System Integrator Guide (2008) and on postings on the Nortek
``knowledge center'' discussion board.  One might assume that the latter is
less authoritative than the former.  For example, the inference of cell size
follows advice found at
\url{https://www.nortekusa.com/en/knowledge-center/forum/hr-profilers/736804717},
which contains a typo in an early posting that is
corrected later on.
}
\references{
\enumerate{
\item Information on Nortek profilers (including the System Integrator Guide,
which explains the data format byte-by-byte) is available at
\url{https://www.nortekusa.com/}.  (One must join the site to see the
manuals.)
\item The Nortek Knowledge Center
\url{https://www.nortekusa.com/en/knowledge-center} may be of help if
problems arise in dealing with data from Nortek instruments.
}
}
\seealso{
Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley
}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.R
\name{adpFlagPastBoundary}
\alias{adpFlagPastBoundary}
\title{Flag adp Data Past Water Column Boundary}
\usage{
adpFlagPastBoundary(
  x = NULL,
  fields = NULL,
  df = 20,
  trim = 0.15,
  good = 1,
  bad = 4,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{x}{an \linkS4class{adp} object containing bottom ranges.}

\item{fields}{a variable contained within \code{x} indicating which field to flag.}

\item{df}{the degrees of freedom to use during the smoothing spline operation.}

\item{trim}{a scale factor for boundary trimming (see \dQuote{Details}).}

\item{good}{number stored in flags to indicate good data.}

\item{bad}{number stored in flags to indicate bad data.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\value{
\code{adpFlagPastBoundary} returns an \linkS4class{adp} object with flags
adjusted in the specified fields if data are beyond the water column boundary.
}
\description{
Flag variables with the same dimension of \code{v} in
an \linkS4class{adp} object that are beyond the water column boundary.
Currently, this operation can only be performed on \linkS4class{adp}
objects that contain bottom ranges. Commonly, \code{\link[=handleFlags]{handleFlags()}} would
then be used to remove such data.
}
\details{
This works by fitting a smoothing spline to a bottom range with a defined
number of degrees of freedom. For each time, it then searches to determine
which associated distances are greater than the predicted smooth spline
multiplied by \eqn{1-trim}.
}
\seealso{
Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\author{
Jaimie Harbin, Clark Richards, and Dan Kelley
}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/odf.R
\docType{class}
\name{odf-class}
\alias{odf-class}
\title{Class to Store ODF Data}
\description{
This class is for data stored in a format used at Canadian
Department of Fisheries and Oceans laboratories. It is somewhat
similar to the \linkS4class{bremen} class, in the sense
that it does not apply just to a particular instrument.
}
\section{Slots}{

\describe{
\item{\code{data}}{As with all \code{oce} objects, the \code{data} slot
for \code{odf} objects is a
\link{list} containing the main data for the object.
{}}

\item{\code{metadata}}{As with all \code{oce} objects, the \code{metadata} slot
for \code{odf} objects is a \link{list} containing
information about the \code{data} or about the object itself.
{}}

\item{\code{processingLog}}{As with all \code{oce} objects, the \code{processingLog} slot
for \code{odf} objects is a
\link{list} with entries describing the creation and evolution
of the object. The contents are updated by various \code{oce} functions to
keep a record of processing steps.  Object summaries and
\code{\link[=processingLogShow]{processingLogShow()}} both display the log.}
}}

\section{Modifying slot contents}{


Although the \code{[[<-} operator may permit modification of the contents
of \linkS4class{odf} objects (see \code{\link{[[<-,odf-method}}),
it is better to use \code{\link[=oceSetData]{oceSetData()}} and \code{\link[=oceSetMetadata]{oceSetMetadata()}},
because those functions save an entry in the \code{processingLog}
that describes the change.
}

\section{Retrieving slot contents}{


The full contents of the \code{data} and \code{metadata} slots of a \linkS4class{odf}
object may be retrieved in the standard R way using \code{\link[=slot]{slot()}}. For
example \code{slot(o,"data")} returns the \code{data} slot of an object named \code{o},
and similarly \code{slot(o,"metadata")} returns
the \code{metadata} slot.

The slots may also be obtained with the \code{\link{[[,odf-method}}
operator, as e.g. \code{o[["data"]]} and \code{o[["metadata"]]}, respectively.

The \code{\link{[[,odf-method}} operator can also
be used to retrieve items from within the \code{data} and \code{metadata} slots.
For example, \code{o[["temperature"]]} can be used to retrieve temperature
from an object containing that quantity. The rule is that a named
quantity is sought first within the object's \code{metadata} slot,
with the \code{data} slot being checked only if \code{metadata} does not
contain the item. This \code{[[} method can also be used to get
certain derived quantities, if the object contains sufficient
information to calculate them. For example, an object that holds
(practical) salinity, temperature and pressure, along with
longitude and latitude, has sufficient information to compute
Absolute Salinity, and so \code{o[["SA"]]} will yield the
calculated Absolute Salinity.

It is also possible to find items more directly, using \code{\link[=oceGetData]{oceGetData()}} and
\code{\link[=oceGetMetadata]{oceGetMetadata()}}, but neither of these functions can
retrieve derived items.
}

\references{
\enumerate{
\item Anthony W. Isenor and David Kellow, 2011. \emph{ODF Format Specification
Version 2.0}. (This is a .doc file obtained in June 2011 by Dan Kelley,
which no longer seems to be made available at any DFO website.)
\item (Unknown authors), October 2014. \emph{ODF Format Description (MLI)},
\url{https://ogsl.ca/wp-content/uploads/ODF_format_desc_en_0.pdf},
(Link verified on May 1, 2021.)
\item A sample ODF file in the DFO format is available at
\code{system.file("extdata","CTD_BCD2014666_008_1_DN.ODF.gz",package="oce")}
\item A sample ODF file in the MLI format is available at
\url{https://ogsl.ca/wp-content/uploads/ODF_file_example_en_0.pdf}.
(Link verified on May 1, 2021.)
}
}
\seealso{
Other things related to odf data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{ODF2oce}()},
\code{\link{ODFListFromHeader}()},
\code{\link{ODFNames2oceNames}()},
\code{\link{[[,odf-method}},
\code{\link{[[<-,odf-method}},
\code{\link{plot,odf-method}},
\code{\link{read.ctd.odf}()},
\code{\link{read.odf}()},
\code{\link{subset,odf-method}},
\code{\link{summary,odf-method}}

Other classes provided by oce: 
\code{\link{adp-class}},
\code{\link{adv-class}},
\code{\link{argo-class}},
\code{\link{bremen-class}},
\code{\link{cm-class}},
\code{\link{coastline-class}},
\code{\link{ctd-class}},
\code{\link{lisst-class}},
\code{\link{lobo-class}},
\code{\link{met-class}},
\code{\link{oce-class}},
\code{\link{rsk-class}},
\code{\link{sealevel-class}},
\code{\link{section-class}},
\code{\link{topo-class}},
\code{\link{windrose-class}},
\code{\link{xbt-class}}
}
\author{
Dan Kelley
}
\concept{classes provided by oce}
\concept{things related to odf data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/AllClass.R
\name{composite,list-method}
\alias{composite,list-method}
\title{Composite by Averaging Across Data}
\usage{
\S4method{composite}{list}(object)
}
\arguments{
\item{object}{a \link{list} of \linkS4class{oce} objects.}
}
\description{
This is done
by calling a specialized version of the function defined in the given
class. In the present
version, the objects must inherit from \linkS4class{amsr}, so the
action is to call
\code{\link[=composite,amsr-method]{composite,amsr-method()}}.

objects that are supplied as arguments are averaged in a way that
makes sense for the object class, i.e. taking into account
the particular bad-data codes of that particular class.
}
\seealso{
Other functions that create composite objects: 
\code{\link{composite,amsr-method}},
\code{\link{composite}()}
}
\concept{functions that create composite objects}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.R
\name{drawIsopycnals}
\alias{drawIsopycnals}
\title{Add Isopycnal Curves to TS Plot}
\usage{
drawIsopycnals(
  nlevels = 6,
  levels,
  rotate = TRUE,
  rho1000 = FALSE,
  digits = 2,
  eos = getOption("oceEOS", default = "gsw"),
  trimIsopycnals = TRUE,
  cex = 0.75 * par("cex"),
  col = "darkgray",
  lwd = par("lwd"),
  lty = par("lty")
)
}
\arguments{
\item{nlevels}{suggested number of density levels (i.e. isopycnal curves);
ignored if \code{levels} is supplied.}

\item{levels}{optional density levels to draw.}

\item{rotate}{boolean, set to \code{TRUE} to write all density labels
horizontally.}

\item{rho1000}{boolean, set to \code{TRUE} to write isopycnal labels as e.g.
1024 instead of 24.}

\item{digits}{number of decimal digits to use in label (supplied to
\code{\link[=round]{round()}}).}

\item{eos}{equation of state to be used, either \code{"unesco"} or
\code{"gsw"}.}

\item{trimIsopycnals}{logical value (\code{TRUE} by default) that
indicates whether to trim isopycnal curves (if drawn)
to the region of temperature-salinity space for which density
computations are considered to be valid in the context of the
chosen \code{eos}; see the \dQuote{Details} of the documentation
for \code{\link[=plotTS]{plotTS()}}.}

\item{cex}{size for labels.}

\item{col}{color for lines and labels.}

\item{lwd}{line width for isopycnal curves}

\item{lty}{line type for isopycnal curves}
}
\value{
None.
}
\description{
Adds isopycnal lines to an existing temperature-salinity plot.  This is
called by \code{\link[=plotTS]{plotTS()}}, and may be called by the user also, e.g. if
an image plot is used to show TS data density.
}
\references{
\itemize{
\item Fofonoff, N. P., and R. C. Millard.
"Algorithms for Computation of Fundamental Properties of Seawater."
UNESCO Technical Papers in Marine Research. SCOR working group on Evaluation of CTD data;
UNESCO/ICES/SCOR/IAPSO Joint Panel on Oceanographic Tables and Standards, 1983.
\url{https://unesdoc.unesco.org/ark:/48223/pf0000059832}.
\item McDougall, Trevor J., David R. Jackett, Daniel G. Wright, and Rainer Feistel.
"Accurate and Computationally Efficient Algorithms for Potential Temperature and Density of Seawater."
Journal of Atmospheric and Oceanic Technology 20, no. 5 (May 1, 2003): 730-41.
\code{https://journals.ametsoc.org/jtech/article/20/5/730/2543/Accurate-and-Computationally-Efficient-Algorithms}.
}
}
\seealso{
\code{\link[=plotTS]{plotTS()}}, which calls this.
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{bound125}
\alias{bound125}
\title{Calculate a rounded bound, rounded up to mantissa 1, 2, or 5}
\usage{
bound125(x)
}
\arguments{
\item{x}{a single positive number}
}
\value{
for positive x, a value exceeding x that has mantissa 1, 2, or 5; otherwise, x
}
\description{
Calculate a rounded bound, rounded up to mantissa 1, 2, or 5
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swSCTp}
\alias{swSCTp}
\title{Practical salinity from electrical conductivity, temperature and pressure}
\usage{
swSCTp(
  conductivity,
  temperature = NULL,
  pressure = NULL,
  conductivityUnit,
  eos = getOption("oceEOS", default = "gsw")
)
}
\arguments{
\item{conductivity}{a measure of conductivity (see also \code{conductivityUnit})
or an \code{oce} object holding hydrographic information. In the second case,
all the other arguments to \code{swSCTp} are ignored.}

\item{temperature}{\emph{in-situ} temperature (\eqn{^\circ}{deg}C), defined
on the ITS-90 scale; see \dQuote{Temperature units} in the documentation for
\code{\link[=swRho]{swRho()}}.}

\item{pressure}{pressure (dbar).}

\item{conductivityUnit}{string indicating the unit used for conductivity.
This may be \code{"ratio"} or \code{""} (meaning conductivity ratio),
\code{"mS/cm"} or \code{"S/m"}.  Note that the ratio mode assumes that
measured conductivity has been divided by the standard conductivity
of 4.2914 S/m.}

\item{eos}{equation of state, either \code{"unesco"} or \code{"gsw"}.#'}
}
\value{
Practical Salinity.
}
\description{
Calculate salinity from what is actually measured by a CTD, \emph{i.e.}
conductivity, \emph{in-situ} temperature and pressure.  Often this is done
by the CTD processing software, but sometimes it is helpful to do this
directly, \emph{e.g.} when there is a concern about mismatches in sensor
response times.  Two variants are provided. First, if \code{eos} is
\code{"unesco"}, then salinity is calculated using
the UNESCO algorithm described by Fofonoff and Millard (1983) as in
reference 1. Second, if \code{eos} is \code{"gsw"}, then the
Gibbs-SeaWater formulation is used, via \code{\link[gsw:gsw_SP_from_C]{gsw::gsw_SP_from_C()}}
in the \CRANpkg{gsw} package. The latter starts with the same formula
as the former, but if this yields a Practical Salinity less than 2,
then the result is instead calculated using
formulae provided by Hill et al. (1986; reference 2), modified to match the
\code{"unesco"} value at Practical salinity equal to 2 (reference 3).
}
\examples{
# 1. Demonstrate agreement with test value in UNESCO documents
swSCTp(1, T90fromT68(15), 0, eos="unesco") # expect 35
# 2. Demonstrate agreement of gsw and unesco, S>2 case
swSCTp(1, T90fromT68(15), 0, eos="gsw") # again, expect 35
# 3. Demonstrate close values even in very brackish water
swSCTp(0.02, 10, 100, eos="gsw")    # 0.6013981
swSCTp(0.02, 10, 100, eos="unesco") # 0.6011721

}
\references{
\enumerate{
\item Fofonoff, P. and R. C. Millard Jr, 1983. Algorithms for
computation of fundamental properties of seawater. \emph{Unesco Technical
Papers in Marine Science}, \emph{44}, 53 pp.
\item K. Hill, T. Dauphinee, and D. Woods. \dQuote{The Extension of the Practical
Salinity Scale 1978 to Low Salinities.} IEEE Journal of Oceanic Engineering 11,
no. 1 (January 1986): 109-12.
\doi{10.1109/JOE.1986.1145154}
\item \code{gsw_from_SP} online documentation, available at
\verb{http://www.teos-10.org/pubs/gsw/html/gsw_C_from_SP.html}
}
}
\seealso{
For thermal (as opposed to electrical) conductivity, see
\code{\link[=swThermalConductivity]{swThermalConductivity()}}.  For computation of electrical
conductivity from salinity, see \code{\link[=swCSTp]{swCSTp()}}.

Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/cm.R
\name{read.cm}
\alias{read.cm}
\title{Read a CM file}
\usage{
read.cm(
  file,
  from = 1,
  to,
  by = 1,
  tz = getOption("oceTz"),
  type = c("s4"),
  longitude = NA,
  latitude = NA,
  debug = getOption("oceDebug"),
  monitor = FALSE,
  processingLog
)
}
\arguments{
\item{file}{a connection or a character string giving the name of the file to
load.}

\item{from}{index number of the first measurement to be read, or the time of
that measurement, as created with \code{\link[=as.POSIXct]{as.POSIXct()}} (hint: use
\code{tz="UTC"}).}

\item{to}{indication of the last measurement to read, in a format matching that
of \code{from}.}

\item{by}{an indication of the stride length to use while walking through the
file. If this is an integer, then \code{by-1} measurements are skipped between
each pair of profiles that is read. This may not make much sense, if the data
are not equi-spaced in time.  If \code{by} is a string representing a time
interval, in colon-separated format, then this interval is divided by the
sampling interval, to get the stride length. \emph{BUG:} if the data are not
equi-spaced, then odd results will occur.}

\item{tz}{character string indicating time zone to be assumed in the data.}

\item{type}{character string indicating type of file (ignored at present).}

\item{longitude}{optional signed number indicating the longitude in degrees
East.}

\item{latitude}{optional signed number indicating the latitude in degrees North.}

\item{debug}{a flag that turns on debugging.  The value indicates the depth
within the call stack to which debugging applies.}

\item{monitor}{ignored.}

\item{processingLog}{if provided, the action item to be stored in the log.  This
parameter is typically only provided for internal calls; the default that it
provides is better for normal calls by a user.}
}
\value{
An \linkS4class{cm} object.

The \code{data} slot will contain all the data in the file, with names
determined from the tokens in line 3 in that file, passed through
\code{\link[=make.names]{make.names()}}, except that
\code{Vnorth} is renamed \code{v} (after conversion from cm/s to m/s),
\code{Veast} is renamed \code{u} (after conversion from cm/s to m/s),
\code{Cond} is renamed \code{conductivity},
\code{T.Temp} is renamed \code{temperature}
and
\code{Sal} is renamed \code{salinity}, and a new
column named \code{time} (a POSIX time) is constructed
from the information in the file header, and another named
\code{pressure} is constructed from the column named \code{Depth}.
At least in the single file studied in the creation of this function,
there are some columns that are unnamed in line 3 of the header;
these yield data items with names \code{X}, \code{X.1}, etc.
}
\description{
Read a current-meter data file, producing a \linkS4class{cm} object.
}
\details{
There function has been tested on only a single file, and the data-scanning
algorithm was based on visual inspection of that file.  Whether it will work
generally is an open question. It should be noted that the sample file had
several odd characteristics, some of which are listed below.
\itemize{
\item file contained two columns named \code{"Cond"}, which was guessed
to stand for conductivity. Since only the first contained data, the second was
ignored, but this may not be the case for all files.
\item The unit for \code{"Cond"} was stated in the file to be \code{"mS"},
which makes no sense, so the unit was assumed to be mS/cm.
\item The file contained a column named \code{"T-Temp"}, which is not
something the author has seen in his career. It was assumed to stand for
in-situ temperature.
\item The file contained a column named \code{"Depth"}, which is not something
an instrument can measure. Presumably it was calculated from pressure (with
what atmospheric offset, though?) and so pressure was inferred from it using
\code{\link[=swPressure]{swPressure()}}.
\item The file contained several columns that lacked names. These were ignored.
\item The file contained several columns that seem to be derived from the
actual measured data, such as \code{"Speed"}, \code{"Dir"}, \code{"N-S Dist"},
etc. These are ignored.
\item The file contained several columns that were basically a mystery to the
author, e.g. \code{"Hx"}, \code{"Hy"}, \code{"Vref"}, etc. These were ignored.
}

Based on such considerations, \code{\link[=read.cm]{read.cm()}} reads only the columns that
were reasonably well-understood based on the sample file. Users who need more
columns should contact the author. And a user who could produce a document
explaining the data format would be especially appreciated!
}
\section{Historical note}{

Prior to late July, 2016, the direction of current flow was stored in the
return value, but it is no longer stored, since it can be derived from the
\code{u} and \code{v} values.
}

\examples{
\dontrun{
  library(oce)
  cm <- read.oce("cm_interocean_0811786.s4a.tab")
  summary(cm)
  plot(cm)
}

}
\seealso{
Other things related to cm data: 
\code{\link{[[,cm-method}},
\code{\link{[[<-,cm-method}},
\code{\link{as.cm}()},
\code{\link{cm-class}},
\code{\link{cm}},
\code{\link{plot,cm-method}},
\code{\link{rotateAboutZ}()},
\code{\link{subset,cm-method}},
\code{\link{summary,cm-method}}
}
\author{
Dan Kelley
}
\concept{things related to cm data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.aml.R
\name{read.ctd.aml}
\alias{read.ctd.aml}
\title{Read an AML CTD file}
\usage{
read.ctd.aml(file, debug = getOption("oceDebug"), processingLog, ...)
}
\arguments{
\item{file}{a connection or a character string giving the name of the file to
load.}

\item{debug}{an integer controlling whether to print debugging
information during processing. Use 0 for quiet processing, or
any positive number to see some output.}

\item{processingLog}{ignored.}

\item{...}{ignored.}
}
\value{
\code{\link[=read.ctd.aml]{read.ctd.aml()}} returns a \linkS4class{ctd} object.
}
\description{
\code{\link[=read.ctd.aml]{read.ctd.aml()}} reads files that hold data acquired with an
AML Oceanographic BaseX2 CTD instrument. In its present form,
the fuction only works for .txt files that were created with
certain settings in the AML software, as explained in
\sQuote{Details} section.
See \sQuote{Alternatives to this function} for hints on dealing
with files that do not match the \code{\link[=read.ctd.aml]{read.ctd.aml()}}
requirements.
}
\details{
The first step in processing is to examine the first line of
the file, and to interpret it as a comma-separated list of column
names.  If this list does not contain each of \code{"Date"},
\code{"Time"}, \code{"Conductivity (mS/cm)"} and \code{"Temperature (C)"},
then an error message is printed.

The next step is to break up each line in the file into
comma-separated elements, and to take note of the lines
that have the same number of elements as the first line.
Those lines are considered to be data, and are read with
\code{\link[=read.csv]{read.csv()}}.  Lines after the first, and before the first
data line are stored as \code{header} in the \code{metadata} slot
of the returned value, while the data (with columns renamed
to match oce conventions) are stored in the \code{data} slot.
If pressure is not found in the data, but depth is, then
pressure is computed with \code{\link[=swPressure]{swPressure()}} and stored in the
\code{data} slot.  Salinity is then inferred with \code{\link[=swSCTp]{swSCTp()}}
and inserted into the \code{data} slot.

Some extra processing is done to take account of units,
and to extract certain elements from the header, particularly
position (stored as \code{longitude} and \code{latitude} in the \code{metadata}
slot), and these are inserted into the return value, which
is constructed with \code{\link[=as.ctd]{as.ctd()}}.
}
\section{Alternatives to this function}{

If \code{\link[=read.ctd.aml]{read.ctd.aml()}} fails to read a user's data file, all is
not lost, because a bit of work in a plain-text editor is likely
to reveal the data structure, so a user can extract elements
to be used with \code{\link[=as.ctd]{as.ctd()}} using procedures akin to those listed
in the \sQuote{Details} section.  Of course, the file must contain
conductivity, temperature and pressure (or depth, from which pressure
can be computed) because these three things are needed for
\linkS4class{ctd} objects.  Adding longitude and latitude to the \code{\link[=as.ctd]{as.ctd()}}
call is recommended, if the desire is to use the modern TEOS-10
equation of state.
}

\seealso{
Other functions that read ctd data: 
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()}
}
\author{
Dan Kelley
}
\concept{functions that read ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{oceColorsClosure}
\alias{oceColorsClosure}
\title{Create color functions}
\usage{
oceColorsClosure(spec)
}
\arguments{
\item{spec}{Specification of the color scheme. This may be a
character string, in which case it must be the name of an item stored
in \code{data(ocecolors)}, or either a 3-column data frame or
matrix, in which case the columns specify red, green and blue values
(in range from 0 to 1).}
}
\description{
This function generates other functions that are used to specify colors.
It is used within oce to create \code{\link[=oceColorsTemperature]{oceColorsTemperature()}}
and its many cousins. Users may also find it helpful, for creating
custom color schemes (see \dQuote{Examples}).
}
\examples{
\dontrun{
# Update oxygen color scheme to latest matplotlib value.
library(oce)
oxy <- "https://raw.githubusercontent.com/matplotlib/cmocean/master/cmocean/rgb/oxy-rgb.txt"
oxyrgb <- read.table(oxy, header=FALSE)
oceColorsOxygenUpdated <- oceColorsClosure(oxyrgb)
par(mfrow=c(1, 2))
m <- matrix(1:256)
imagep(m, col=oceColorsOxygen, zlab="oxygen")
imagep(m, col=oceColorsOxygenUpdated, zlab="oxygenUpdated")
}
}
\seealso{
Other things related to colors: 
\code{\link{colormapGMT}()},
\code{\link{colormap}()},
\code{\link{oceColors9B}()},
\code{\link{oceColorsCDOM}()},
\code{\link{oceColorsChlorophyll}()},
\code{\link{oceColorsDensity}()},
\code{\link{oceColorsFreesurface}()},
\code{\link{oceColorsGebco}()},
\code{\link{oceColorsJet}()},
\code{\link{oceColorsOxygen}()},
\code{\link{oceColorsPAR}()},
\code{\link{oceColorsPalette}()},
\code{\link{oceColorsPhase}()},
\code{\link{oceColorsSalinity}()},
\code{\link{oceColorsTemperature}()},
\code{\link{oceColorsTurbidity}()},
\code{\link{oceColorsTurbo}()},
\code{\link{oceColorsTwo}()},
\code{\link{oceColorsVelocity}()},
\code{\link{oceColorsViridis}()},
\code{\link{oceColorsVorticity}()},
\code{\link{ocecolors}}
}
\concept{things related to colors}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.R
\name{subset,ctd-method}
\alias{subset,ctd-method}
\title{Subset a ctd Object}
\usage{
\S4method{subset}{ctd}(x, subset, ...)
}
\arguments{
\item{x}{a \linkS4class{ctd} object.}

\item{subset}{An expression indicating how to subset \code{x}.}

\item{...}{optional arguments, of which only the first is examined. The only
possibility is that this argument be named \code{indices}. See \dQuote{Details}.}
}
\value{
A \linkS4class{ctd} object.
}
\description{
Return a subset of a \linkS4class{ctd} object.
}
\details{
This function is used to subset data within
a \linkS4class{ctd} object. There are two ways of working. If
\code{subset} is supplied, then it is a logical expression
that is evaluated within the environment of the \code{data}
slot of the object (see Example 1). Alternatively, if the
\code{...} list contains an expression defining \code{indices},
then that expression is used to subset each item within the
\code{data} slot (see Example 2).
}
\examples{
library(oce)
data(ctd)
plot(ctd)
## Example 1
plot(subset(ctd, pressure<10))
## Example 2
plot(subset(ctd, indices=1:10))

}
\seealso{
Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdRepair}()},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}

Other functions that subset oce objects: 
\code{\link{subset,adp-method}},
\code{\link{subset,adv-method}},
\code{\link{subset,amsr-method}},
\code{\link{subset,argo-method}},
\code{\link{subset,cm-method}},
\code{\link{subset,coastline-method}},
\code{\link{subset,echosounder-method}},
\code{\link{subset,lobo-method}},
\code{\link{subset,met-method}},
\code{\link{subset,oce-method}},
\code{\link{subset,odf-method}},
\code{\link{subset,rsk-method}},
\code{\link{subset,sealevel-method}},
\code{\link{subset,section-method}},
\code{\link{subset,topo-method}},
\code{\link{subset,xbt-method}}
}
\author{
Dan Kelley
}
\concept{functions that subset oce objects}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{computableWaterProperties}
\alias{computableWaterProperties}
\title{Available derived water properties}
\usage{
computableWaterProperties(x)
}
\arguments{
\item{x}{An \linkS4class{oce} object.}
}
\value{
A character vector listing the names of computable
water properties, or NULL, if there are none.
}
\description{
This checks to see whether \code{x} is an \code{oce} object containing
\code{salinity}, \code{temperature}, \code{pressure}, \code{latitude} and \code{longitude}.
If this holds, then it returns a list of items that can be accessed
with \code{[[}.
}
\seealso{
Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/run.R
\name{runlm}
\alias{runlm}
\title{Calculate running linear models}
\usage{
runlm(x, y, xout, window = c("hanning", "boxcar"), L, deriv)
}
\arguments{
\item{x}{a vector holding x values.}

\item{y}{a vector holding y values.}

\item{xout}{optional vector of x values at which the derivative is to be
found.  If not provided, \code{x} is used.}

\item{window}{type of weighting function used to weight data within the
window; see \sQuote{Details}.}

\item{L}{width of running window, in x units.  If not provided, a reasonable
default will be used.}

\item{deriv}{an optional indicator of the desired return value; see
\sQuote{Examples}.}
}
\value{
If \code{deriv} is not specified, a list containing vectors of
output values \code{y} and \code{y}, derivative (\code{dydx}), along with
the scalar length scale \code{L}.  If \code{deriv=0}, a vector of values is
returned, and if \code{deriv=1}, a vector of derivatives is returned.
}
\description{
The linear model is calculated from the slope of a localized least-squares
regression model y=y(x).  The localization is defined by the x difference
from the point in question, with data at distance exceeding L/2 being
ignored.  With a \code{boxcar} window, all data within the local domain are
treated equally, while with a \code{hanning} window, a raised-cosine
weighting function is used; the latter produces smoother derivatives, which
can be useful for noisy data.  The function is based on internal
calculation, not on \code{\link[=lm]{lm()}}.
}
\examples{

library(oce)

# Case 1: smooth a noisy signal
x <- 1:100
y <- 1 + x/100 + sin(x/5)
yn <- y + rnorm(100, sd=0.1)
L <- 4
calc <- runlm(x, y, L=L)
plot(x, y, type='l', lwd=7, col='gray')
points(x, yn, pch=20, col='blue')
lines(x, calc$y, lwd=2, col='red')

# Case 2: square of buoyancy frequency
data(ctd)
par(mfrow=c(1,1))
plot(ctd, which="N2")
rho <- swRho(ctd)
z <- swZ(ctd)
zz <- seq(min(z), max(z), 0.1)
N2 <- -9.8 / mean(rho) * runlm(z, rho, zz, deriv=1)
lines(N2, -zz, col='red')
legend("bottomright", lwd=2, bg="white",
       col=c("black", "red"),
       legend=c("swN2()", "using runlm()"))
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/topo.R
\name{plot,topo-method}
\alias{plot,topo-method}
\alias{plot.topo}
\title{Plot a topo Object}
\usage{
\S4method{plot}{topo}(
  x,
  xlab = "",
  ylab = "",
  asp,
  clongitude,
  clatitude,
  span,
  expand = 1.5,
  water.z,
  col.water,
  lty.water,
  lwd.water,
  land.z,
  col.land,
  lty.land,
  lwd.land,
  geographical = FALSE,
  location = "topright",
  mgp = getOption("oceMgp"),
  mar = c(mgp[1] + 1, mgp[1] + 1, 1, 1),
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{x}{a \linkS4class{topo} object.}

\item{xlab, ylab}{Character strings giving a label for the x and y axes.}

\item{asp}{Aspect ratio for plot.  The default is for \code{plot.coastline} to
set the aspect ratio to give natural latitude-longitude scaling somewhere near
the centre latitude on the plot. Often, it makes sense to set \code{asp}
yourself, e.g. to get correct shapes at 45N, use \code{asp=1/cos(45*pi/180)}.
Note that the land mass is not symmetric about the equator, so to get good
world views you should set \code{asp=1} or set \code{ylim} to be symmetric
about zero.  Any given value of \code{asp} is ignored, if \code{clongitude} and
\code{clatitude} are given.}

\item{clongitude}{Optional center longitude of map, in degrees east; see
\code{clatitude}.}

\item{clatitude}{Optional center latitude of map, in degrees north.  If this
and \code{clongitude} are provided, then any provided value of \code{asp} is
ignored, and instead the plot aspect ratio is computed based on the center
latitude.  Also, if \code{clongitude} and \code{clatitude} are provided, then
\code{span} must be, also.}

\item{span}{Optional suggested span of plot, in kilometers (must be supplied,
if \code{clongitude} and \code{clatitude} are supplied).}

\item{expand}{Numerical factor for the expansion of plot limits, showing area
outside the plot, e.g. if showing a ship track as a coastline, and then an
actual coastline to show the ocean boundary.  The value of \code{expand} is
ignored if either \code{xlim} or \code{ylim} is given.}

\item{water.z}{Depths at which to plot water contours.  If not provided, these
are inferred from the data.}

\item{col.water}{Colors corresponding to \code{water.z} values.  If not
provided, these will be \code{"fill"} colors from
\code{\link[=oce.colorsGebco]{oce.colorsGebco()}}.}

\item{lty.water}{Line type(s) for water contours.}

\item{lwd.water}{Line width(s) for water contours.}

\item{land.z}{Depths at which to plot land contours.  If not provided, these
are inferred from the data.  If set to \code{NULL}, no land contours will be
plotted.}

\item{col.land}{Colors corresponding to \code{land.z} values.  If not
provided, these will be \code{"fill"} colors from
\code{\link[=oce.colorsGebco]{oce.colorsGebco()}}.}

\item{lty.land}{Line type(s) for land contours.}

\item{lwd.land}{Line width(s) for land contours.}

\item{geographical}{Logical, indicating whether to plot latitudes and
longitudes without minus signs.}

\item{location}{Location for a legend (or \code{"none"}, for no legend).}

\item{mgp}{3-element numerical vector to use for \code{par(mgp)}, and also for
\code{par(mar)}, computed from this.  The default is tighter than the R
default, in order to use more space for the data and less for the axes.}

\item{mar}{Four-element numerical vector to be used with
\code{\link{par}}\code{("mar")}.}

\item{debug}{Numerical value, with positive values indicating higher levels of
debugging.}

\item{...}{Additional arguments passed on to plotting functions.}
}
\description{
This plots contours of topographic elevation.  The plot aspect ratio is set
based on the middle latitude in the plot.  The line properties, such as
\code{land.lwd}, may either be a single item, or a vector; in the latter case,
the length must match the length of the corresponding properties, e.g.
\code{land.z}.
}
\examples{
library(oce)
data(topoWorld)
plot(topoWorld, clongitude=-60, clatitude=45, span=10000)

}
\seealso{
Other functions that plot oce data: 
\code{\link{download.amsr}()},
\code{\link{plot,adp-method}},
\code{\link{plot,adv-method}},
\code{\link{plot,amsr-method}},
\code{\link{plot,argo-method}},
\code{\link{plot,bremen-method}},
\code{\link{plot,cm-method}},
\code{\link{plot,coastline-method}},
\code{\link{plot,ctd-method}},
\code{\link{plot,gps-method}},
\code{\link{plot,ladp-method}},
\code{\link{plot,landsat-method}},
\code{\link{plot,lisst-method}},
\code{\link{plot,lobo-method}},
\code{\link{plot,met-method}},
\code{\link{plot,odf-method}},
\code{\link{plot,rsk-method}},
\code{\link{plot,satellite-method}},
\code{\link{plot,sealevel-method}},
\code{\link{plot,section-method}},
\code{\link{plot,tidem-method}},
\code{\link{plot,windrose-method}},
\code{\link{plot,xbt-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{tidem-class}}

Other things related to topo data: 
\code{\link{[[,topo-method}},
\code{\link{[[<-,topo-method}},
\code{\link{as.topo}()},
\code{\link{download.topo}()},
\code{\link{read.topo}()},
\code{\link{subset,topo-method}},
\code{\link{summary,topo-method}},
\code{\link{topo-class}},
\code{\link{topoInterpolate}()},
\code{\link{topoWorld}}
}
\author{
Dan Kelley
}
\concept{functions that plot oce data}
\concept{things related to topo data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/argo.R
\name{subset,argo-method}
\alias{subset,argo-method}
\alias{subset.argo}
\title{Subset an Argo Object}
\usage{
\S4method{subset}{argo}(x, subset, ...)
}
\arguments{
\item{x}{an \linkS4class{argo} object.}

\item{subset}{An expression indicating how to subset \code{x}.}

\item{...}{optional arguments, of which only the first is examined. The
only possibility is \code{within}, a polygon enclosing data to be
retained. This must be either a list or data frame, containing items
named either \code{x} and \code{y} or \code{longitude} and
\code{latitude}; see Example 4.  If \code{within} is given,
then \code{subset} is ignored.}
}
\value{
An \linkS4class{argo} object.
}
\description{
Subset an argo object, either by selecting just the "adjusted" data
or by subsetting by pressure or other variables.
}
\details{
If \code{subset} is the string \code{"adjusted"}, then \code{subset}
replaces the station variables with their adjusted counterparts. In
the argo notation, e.g. \code{PSAL} is replaced with \code{PSAL_ADJUSTED};
in the present notation, this means that \code{salinity} in the \code{data}
slot is replaced with \code{salinityAdjusted}, and the latter is deleted.
Similar replacements are also done with the flags stored in the \code{metadata}
slot.

If \code{subset} is an expression, then the action is somewhat similar
to other \code{subset} functions, but with the restriction that
only one independent variable may be
used in in any call to the function, so that
repeated calls will be necessary to subset based on more than one
independent variable.  Subsetting may be done by anything
stored in the data, e.g. \code{time},
\code{latitude}, \code{longitude}, \code{profile}, \code{dataMode},
or \code{pressure} or by \code{profile} (a made-up variable),
\code{id} (from the \code{metadata} slot) or \code{ID} (a synonym for \code{id}).
Note that subsetting by \code{pressure}
preserves matrix shape, by setting discarded values to \code{NA}, as opposed
to dropping data (as is the case with \code{time}, for example).
}
\examples{
library(oce)
data(argo)

# Example 1: subset by time, longitude, and pressure
par(mfrow=c(2,2))
plot(argo)
plot(subset(argo, time > mean(time)))
plot(subset(argo, longitude > mean(longitude)))
plot(subset(argoGrid(argo), pressure > 500 & pressure < 1000), which=5)

# Example 2: restrict attention to delayed-mode profiles.
\dontrun{
par(mfrow=c(1, 1))
plot(subset(argo, dataMode == "D"))
}

# Example 3: contrast adjusted and unadjusted data
\dontrun{
par(mfrow=c(1, 2))
plotTS(argo)
plotTS(subset(argo, "adjusted"))
}

# Example 4. Subset by a polygon determined with locator()
\dontrun{
par(mfrow=c(1, 2))
plot(argo, which="map")
## Can get a boundary with e.g. locator(4)
boundary <- list(x=c(-65, -40, -40, -65), y=c(65, 65, 45, 45))
argoSubset <- subset(argo, within=boundary)
plot(argoSubset, which="map")
}
}
\seealso{
Other things related to argo data: 
\code{\link{[[,argo-method}},
\code{\link{[[<-,argo-method}},
\code{\link{argo-class}},
\code{\link{argoGrid}()},
\code{\link{argoNames2oceNames}()},
\code{\link{argo}},
\code{\link{as.argo}()},
\code{\link{handleFlags,argo-method}},
\code{\link{plot,argo-method}},
\code{\link{read.argo}()},
\code{\link{summary,argo-method}}

Other functions that subset oce objects: 
\code{\link{subset,adp-method}},
\code{\link{subset,adv-method}},
\code{\link{subset,amsr-method}},
\code{\link{subset,cm-method}},
\code{\link{subset,coastline-method}},
\code{\link{subset,ctd-method}},
\code{\link{subset,echosounder-method}},
\code{\link{subset,lobo-method}},
\code{\link{subset,met-method}},
\code{\link{subset,oce-method}},
\code{\link{subset,odf-method}},
\code{\link{subset,rsk-method}},
\code{\link{subset,sealevel-method}},
\code{\link{subset,section-method}},
\code{\link{subset,topo-method}},
\code{\link{subset,xbt-method}}
}
\author{
Dan Kelley
}
\concept{functions that subset oce objects}
\concept{things related to argo data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/gps.R
\name{as.gps}
\alias{as.gps}
\title{Coerce data into a GPS dataset}
\usage{
as.gps(longitude, latitude, filename = "")
}
\arguments{
\item{longitude}{the longitude in decimal degrees, positive east of
Greenwich, or a data frame with columns named \code{latitude} and
\code{longitude}, in which case these values are extracted from the data
frame and the second argument is ignored.}

\item{latitude}{the latitude in decimal degrees, positive north of the
Equator.}

\item{filename}{name of file containing data (if applicable).}
}
\value{
A \linkS4class{gps} object.
}
\description{
Coerces a sequence of longitudes and latitudes into a GPS dataset.
This may be used when \code{\link[=read.gps]{read.gps()}} cannot read a file, or when the
data have been manipulated.
}
\examples{
# Location of the Tower Tank at Dalhousie University
towerTank <- as.gps(-63.59428, 44.63572)

}
\seealso{
Other things related to gps data: 
\code{\link{[[,gps-method}},
\code{\link{[[<-,gps-method}},
\code{\link{gps-class}},
\code{\link{plot,gps-method}},
\code{\link{read.gps}()},
\code{\link{summary,gps-method}}
}
\author{
Dan Kelley
}
\concept{things related to gps data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{oceColorsSalinity}
\alias{oceColorsSalinity}
\alias{oce.colorsSalinity}
\title{Create colors suitable for salinity fields}
\usage{
oceColorsSalinity(n)
}
\arguments{
\item{n}{number of colors to create.}
}
\value{
A vector of color specifications.
}
\description{
Create a set of colors for displaying salinity values,
based on the scheme devised by Thyng et al. (2016) and
presented in a python package by Thyng (2019).
The color specifications were transliterated from
python to R on 2015-09-29, but have not been adjusted
since, even though the python source has changed.  This
is to prevent breaking old \code{oce} code.  To get the
latest versions of these colours or other colours,
use the \CRANpkg{cmocean} R package
(Thyng, Richards, and Krylov, 2019) directly, as is
illustrated (with the "matter" scheme) in Example 2.
Note that the \CRANpkg{cmocean} core functions provide a way
to select between various versions of the colour schemes.
It is also worth considering the palettes provided by the
\CRANpkg{viridis} package, as illustrated (with the
"inferno" scheme) in Example 3.
}
\examples{
library(oce)

# Example 1
imagep(volcano, col=oceColorsSalinity(128),
       zlab="oceColorsSalinity")
\dontrun{
# Example 2 (requires the cmocean package)
imagep(volcano, col=cmocean::cmocean("matter"),
       zlab='cmocean::cmocean("matter")')}

\dontrun{
# Example 3 (requires the viridis package)
imagep(volcano, col=viridis::inferno,
       zlab='viridis::inferno')}

}
\references{
\itemize{
\item Thyng, Kristen, Chad Greene, Robert Hetland, Heather Zimmerle, and Steven DiMarco.
\dQuote{True Colors of Oceanography: Guidelines for Effective and Accurate Colormap Selection.}
Oceanography 29, no. 3 (September 1, 2016): 9‚Äì13.
\doi{10.5670/oceanog.2016.66}
\item Thyng, Kristen. Kthyng/Cmocean. Python, 2019.
\url{https://github.com/kthyng/cmocean}.
\item Thyng, Kristen, Clark Richards, and Ivan Krylov.
Cmocean: Beautiful Colour Maps for Oceanography (version 0.2), 2019.
\url{https://CRAN.R-project.org/package=cmocean}.
}

The following references provide information on choosing
colour schemes, that are suitable for viewers who have
colour deficiencies.

Light, Adam, and Patrick J. Bartlein. "The End of the Rainbow? Color
Schemes for Improved Data Graphics." \emph{Eos, Transactions American Geophysical
Union} 85, no. 40 (2004): 385.
DOI: 10.1029/2004EO400002

Stephenson, David B. "Comment on 'Color Schemes for Improved Data
Graphics', by A Light and P.J. Bartlein." \emph{Eos, Transactions American
Geophysical Union} 86, no.  20 (2005): 196.
DOI: 10.1029/2005EO200005

Light, Adam, and Patrick J. Bartlein. "Reply to 'Comment on Color
Schemes for Improved Data Graphics,' by A. Light and P.J.  Bartlein'."
\emph{Eos, Transactions American Geophysical Union} 86, no. 20 (2005): 196‚Äì196.
DOI: 10.1029/2005EO200006
}
\seealso{
Other things related to colors: 
\code{\link{colormapGMT}()},
\code{\link{colormap}()},
\code{\link{oceColors9B}()},
\code{\link{oceColorsCDOM}()},
\code{\link{oceColorsChlorophyll}()},
\code{\link{oceColorsClosure}()},
\code{\link{oceColorsDensity}()},
\code{\link{oceColorsFreesurface}()},
\code{\link{oceColorsGebco}()},
\code{\link{oceColorsJet}()},
\code{\link{oceColorsOxygen}()},
\code{\link{oceColorsPAR}()},
\code{\link{oceColorsPalette}()},
\code{\link{oceColorsPhase}()},
\code{\link{oceColorsTemperature}()},
\code{\link{oceColorsTurbidity}()},
\code{\link{oceColorsTurbo}()},
\code{\link{oceColorsTwo}()},
\code{\link{oceColorsVelocity}()},
\code{\link{oceColorsViridis}()},
\code{\link{oceColorsVorticity}()},
\code{\link{ocecolors}}
}
\author{
Krysten M. Thyng (Python version), Dan Kelley (R transliteration)
}
\concept{things related to colors}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/windrose.R
\name{plot,windrose-method}
\alias{plot,windrose-method}
\alias{plot.windrose}
\title{Plot a windrose Object}
\usage{
\S4method{plot}{windrose}(
  x,
  type = c("count", "mean", "median", "fivenum"),
  convention = c("meteorological", "oceanographic"),
  mgp = getOption("oceMgp"),
  mar = c(mgp[1], mgp[1], 1 + mgp[1], mgp[1]),
  col,
  ...
)
}
\arguments{
\item{x}{a \linkS4class{windrose} object.}

\item{type}{The thing to be plotted, either the number of counts in the angle
interval, the mean of the values in the interval, the median of the values, or
a \code{\link[=fivenum]{fivenum()}} representation of the values.}

\item{convention}{String indicating whether to use meteorological convention or
oceanographic convention for the arrows that emanate from the centre of the
rose.  In meteorological convection, an arrow emanates towards the right on
the diagram if the wind is from the east; in oceanographic convention, such an
arrow indicates flow \emph{to} the east.}

\item{mgp}{Three-element numerical vector to use for \code{\link{par}}\code{(mgp)}, and also
for \code{\link{par}}\code{(mar)}, computed from this.  The default is tighter than the R
default, in order to use more space for the data and less for the axes.}

\item{mar}{Four-element numerical vector to be used with \code{\link{par}}\code{("mar")}.}

\item{col}{Optional list of colors to use.  If not set, the colors will be
\code{c("red", "pink", "blue", "lightgray")}.  For the first three types of
plot, the first color in this list is used to fill in the rose, the third is
used for the petals of the rose, and the fourth is used for grid lines. For the
\code{"fivenum"} type, the first color is used for the inter-quartile range,
the second is used outside this range, the third is used for the median, and
the fourth is, again, used for the grid lines.}

\item{...}{Optional arguments passed to plotting functions.}
}
\description{
Plot a \linkS4class{windrose} object.
}
\examples{
library(oce)
opar <- par(no.readonly = TRUE)
xcomp <- rnorm(360) + 1
ycomp <- rnorm(360)
wr <- as.windrose(xcomp, ycomp)
par(mfrow=c(1, 2))
plot(wr)
plot(wr, "fivenum")
par(opar)

}
\seealso{
Other functions that plot oce data: 
\code{\link{download.amsr}()},
\code{\link{plot,adp-method}},
\code{\link{plot,adv-method}},
\code{\link{plot,amsr-method}},
\code{\link{plot,argo-method}},
\code{\link{plot,bremen-method}},
\code{\link{plot,cm-method}},
\code{\link{plot,coastline-method}},
\code{\link{plot,ctd-method}},
\code{\link{plot,gps-method}},
\code{\link{plot,ladp-method}},
\code{\link{plot,landsat-method}},
\code{\link{plot,lisst-method}},
\code{\link{plot,lobo-method}},
\code{\link{plot,met-method}},
\code{\link{plot,odf-method}},
\code{\link{plot,rsk-method}},
\code{\link{plot,satellite-method}},
\code{\link{plot,sealevel-method}},
\code{\link{plot,section-method}},
\code{\link{plot,tidem-method}},
\code{\link{plot,topo-method}},
\code{\link{plot,xbt-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{tidem-class}}

Other things related to windrose data: 
\code{\link{[[,windrose-method}},
\code{\link{[[<-,windrose-method}},
\code{\link{as.windrose}()},
\code{\link{summary,windrose-method}},
\code{\link{windrose-class}}
}
\author{
Dan Kelley
}
\concept{functions that plot oce data}
\concept{things related to windrose data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/argo.R
\docType{data}
\name{argo}
\alias{argo}
\title{ARGO float dataset}
\source{
The netcdf file used by \code{\link[=read.argo]{read.argo()}} to create this \linkS4class{argo}
object was downloaded using FTP to
\code{ftp.ifremer.fr/ifremer/argo/dac/bodc/6900388/6900388_prof.nc}
on 2020 June 24.
}
\description{
This holds data from ARGO 6900388 in the North Atlantic.
}
\details{
Below is the official citation (note that this DOI has web links for
downloads):
Argo (2017). Argo float data and metadata from Global Data Assembly Centre
(Argo GDAC) - Snapshot of Argo GDAC of July, 8st 2017. SEANOE.
\doi{10.17882/42182#50865}
}
\examples{
library(oce)
data(argo)
summary(argo)
data(coastlineWorld)
plot(argo, which="trajectory")

}
\seealso{
Other datasets provided with oce: 
\code{\link{adp}},
\code{\link{adv}},
\code{\link{amsr}},
\code{\link{cm}},
\code{\link{coastlineWorld}},
\code{\link{ctdRaw}},
\code{\link{ctd}},
\code{\link{echosounder}},
\code{\link{landsat}},
\code{\link{lisst}},
\code{\link{lobo}},
\code{\link{met}},
\code{\link{ocecolors}},
\code{\link{rsk}},
\code{\link{sealevelTuktoyaktuk}},
\code{\link{sealevel}},
\code{\link{section}},
\code{\link{topoWorld}},
\code{\link{wind}},
\code{\link{xbt}}

Other things related to argo data: 
\code{\link{[[,argo-method}},
\code{\link{[[<-,argo-method}},
\code{\link{argo-class}},
\code{\link{argoGrid}()},
\code{\link{argoNames2oceNames}()},
\code{\link{as.argo}()},
\code{\link{handleFlags,argo-method}},
\code{\link{plot,argo-method}},
\code{\link{read.argo}()},
\code{\link{subset,argo-method}},
\code{\link{summary,argo-method}}
}
\concept{datasets provided with oce}
\concept{things related to argo data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/processingLog.R
\name{processingLog<-}
\alias{processingLog<-}
\title{Add an item to a processing log (in place)}
\usage{
processingLog(x) <- value
}
\arguments{
\item{x}{an \linkS4class{oce} object.}

\item{value}{A character string with the description of the logged activity.}
}
\description{
Add an item to a processing log (in place)
}
\examples{
data(ctd)
processingLogShow(ctd)
processingLog(ctd) <- "test"
processingLogShow(ctd)

}
\seealso{
Other things related to processing logs: 
\code{\link{processingLogAppend}()},
\code{\link{processingLogItem}()},
\code{\link{processingLogShow}()}
}
\concept{things related to processing logs}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/lisst.R
\docType{data}
\name{lisst}
\alias{lisst}
\title{LISST Dataset}
\source{
This was constructed artificially using \code{\link[=as.lisst]{as.lisst()}},
to approximately match values that might be measured in the field.
}
\usage{
data(lisst)
}
\description{
LISST (Laser in-situ scattering and transmissometry) dataset, constructed
artificially.
}
\seealso{
Other datasets provided with oce: 
\code{\link{adp}},
\code{\link{adv}},
\code{\link{amsr}},
\code{\link{argo}},
\code{\link{cm}},
\code{\link{coastlineWorld}},
\code{\link{ctdRaw}},
\code{\link{ctd}},
\code{\link{echosounder}},
\code{\link{landsat}},
\code{\link{lobo}},
\code{\link{met}},
\code{\link{ocecolors}},
\code{\link{rsk}},
\code{\link{sealevelTuktoyaktuk}},
\code{\link{sealevel}},
\code{\link{section}},
\code{\link{topoWorld}},
\code{\link{wind}},
\code{\link{xbt}}
}
\author{
Dan Kelley
}
\concept{datasets provided with oce}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adv.R
\name{plot,adv-method}
\alias{plot,adv-method}
\alias{plot.adv}
\title{Plot an adv Object}
\usage{
\S4method{plot}{adv}(
  x,
  which = c(1:3, 14, 15),
  col,
  titles,
  type = "l",
  lwd = par("lwd"),
  drawTimeRange = getOption("oceDrawTimeRange"),
  drawZeroLine = FALSE,
  useSmoothScatter,
  mgp = getOption("oceMgp"),
  mar = c(mgp[1] + 1.5, mgp[1] + 1.5, 1.5, 1.5),
  tformat,
  marginsAsImage = FALSE,
  cex = par("cex"),
  cex.axis = par("cex.axis"),
  cex.lab = par("cex.lab"),
  cex.main = par("cex.main"),
  xlim,
  ylim,
  brushCorrelation,
  colBrush = "red",
  main = "",
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{x}{an \linkS4class{adv} object.}

\item{which}{List of desired plot types.  These are graphed in panels running
down from the top of the page.  See \dQuote{Details} for the meanings of
various values of \code{which}.}

\item{col}{Optional indication of color(s) to use.  If not provided, the
default for images is \code{oce.colorsPalette(128,1)}, and for lines and points
is black.}

\item{titles}{Optional vector of character strings to be used as labels for the
plot panels.  For images, these strings will be placed in the right hand side
of the top margin.  For timeseries, these strings are ignored.  If this is
provided, its length must equal that of \code{which}.}

\item{type}{Type of plot, as for \code{\link[=plot]{plot()}}.}

\item{lwd}{If the plot is of a time-series or scattergraph format with lines,
this is used in the usual way; otherwise, e.g. for image formats, this is
ignored.}

\item{drawTimeRange}{Logical value that applies to panels with time as the horizontal
axis, indicating whether to draw the time range in the top-left margin of the
plot.}

\item{drawZeroLine}{Logical value indicating whether to draw zero lines on
velocities.}

\item{useSmoothScatter}{Logical value indicating whether to use
\code{\link[=smoothScatter]{smoothScatter()}} in various plots, such as \code{which="uv"}.  If
not provided a default is used, with \code{\link[=smoothScatter]{smoothScatter()}} being used if
there are more than 2000 points to plot.}

\item{mgp}{3-element numerical
vector to use for \code{par(mgp)}, and also for \code{par(mar)}, computed from
this.  The default is tighter than the R default, in order to use more space
for the data and less for the axes.}

\item{mar}{Value to be used with \code{\link{par}}\code{("mar")}.}

\item{tformat}{Optional argument passed to \code{\link[=oce.plot.ts]{oce.plot.ts()}}, for plot
types that call that function.  (See \code{\link[=strptime]{strptime()}} for the format
used.)}

\item{marginsAsImage}{Logical value indicating whether to put a wide margin to
the right of time-series plots, matching the space used up by a palette in an
\code{\link[=imagep]{imagep()}} plot.}

\item{cex}{numeric character expansion factor for plot symbols; see \code{\link[=par]{par()}}.}

\item{cex.axis, cex.lab, cex.main}{character expansion factors for axis numbers, axis names and plot titles; see \code{\link[=par]{par()}}.}

\item{xlim}{Optional 2-element list for \code{xlim}, or 2-column matrix, in
which case the rows are used, in order, for the panels of the graph.}

\item{ylim}{Optional 2-element list for \code{ylim}, or 2-column matrix, in
which case the rows are used, in order, for the panels of the graph.}

\item{brushCorrelation}{Optional number between 0 and 100, indicating a
per-beam correlation threshold below which data are to be considered suspect.
If the plot type is \code{p}, the suspect points (velocity, backscatter
amplitude, or correlation) will be colored red; otherwise, this argument is
ignored.}

\item{colBrush}{Color to use for brushed (bad) data, if
\code{brushCorrelation} is active.}

\item{main}{Main title for plot, used just on the top panel, if there are
several panels.}

\item{debug}{A flag that turns on debugging.  Set to 1 to get a moderate amount
of debugging information, or to 2 to get more.}

\item{...}{Optional arguments passed to plotting functions.}
}
\description{
Plot \linkS4class{adv} data.
}
\details{
Creates a multi-panel summary plot of data measured by an ADV.
The panels are controlled by the \code{which} argument.  (Note the
gaps in the sequence, e.g. 4 and 8 are not used.)
\itemize{
\item \code{which=1} to \code{3} (or \code{"u1"} to \code{"u3"})
yield timeseries of the first, second, and third components of
velocity (in beam, xyz or enu coordinates).
\item \code{which=4} is not permitted (since ADV are 3-beam devices)
\item \code{which=5} to \code{7} (or \code{"a1"} to \code{"a3"})
yield timeseries of the amplitudes of beams 1 to 3.  (Note that
the data are called \code{data$a[,1]}, \code{data$a[,2]} and
\code{data$a[,3]}, for these three timeseries.)
\item \code{which=8} is not permitted (since ADV are 3-beam devices)
\item \code{which=9} to \code{11} (or \code{"q1"} to \code{"q3"})
yield timeseries of correlation for beams 1 to 3.  (Note that the
data are called \code{data$c[,1]}, \code{data$c[,2]} and
\code{data$c[,3]}, for these three timeseries.)
\item \code{which=12} is not permitted (since ADVs are 3-beam devices)
\item \code{which=13} is not permitted (since ADVs do not measure salinity)
\item \code{which=14} or \code{which="temperature"} yields a timeseries of temperature.
\item \code{which=15} or \code{which="pressure"} yields a timeseries of pressure.
\item \code{which=16} or \code{which="heading"} yields a timeseries of heading.
\item \code{which=17} or \code{which="pitch"}yields a timeseries of pitch.
\item \code{which=18} or \code{which="roll"}yields a timeseries of roll.
\item \code{which=19} to \code{21} yields plots of correlation versus
amplitude, for beams 1 through 3, using \code{\link[=smoothScatter]{smoothScatter()}}.
\item \code{which=22} is not permitted (since ADVs are 3-beam devices)
\item \code{which=23} or \code{"progressive vector"} yields a
progressive-vector diagram in the horizontal plane, plotted with
\code{asp=1}, and taking beam1 and beam2 as the eastward and
northward components of velocity, respectively.
\item \code{which=28} or \code{"uv"} yields velocity plot in the
horizontal plane, i.e. \code{u[2]} versus \code{u[1]}.  If the number of data
points is small, a scattergraph is used, but if it is large,
\code{\link[=smoothScatter]{smoothScatter()}} is used.
\item \code{which=29} or \code{"uv+ellipse"} as the \code{"uv"}
case, but with an added indication of the tidal ellipse,
calculated from the eigen vectors of the covariance matrix.
\item \code{which=30} or \code{"uv+ellipse+arrow"} as the
\code{"uv+ellipse"} case, but with an added arrow indicating the
mean current.
\item \code{which=50} or \code{"analog1"} plots a time series of the
analog1 signal, if there is one.
\item \code{which=51} or \code{"analog2"} plots a time series of the
analog2 signal, if there is one.
\item \code{which=100} or \code{"voltage"} plots the voltage as a
timeseries, if voltage exists in the dataset.
}

In addition to the above, there are some groupings defined:
\itemize{
\item \code{which="velocity"} equivalent to \code{which=1:3} (three velocity components)
\item \code{which="amplitude"} equivalent to \code{which=5:7} (three amplitude components)
\item \code{which="backscatter"} equivalent to \code{which=9:11} (three backscatter components)
\item \code{which="hydrography"} equivalent to \code{which=14:15} (temperature and pressure)
\item \code{which="angles"} equivalent to \code{which=16:18} (heading, pitch and roll)
}
}
\examples{
library(oce)
data(adv)
plot(adv)

}
\seealso{
The documentation for \linkS4class{adv} explains the structure
of ADV objects, and also outlines the other functions dealing with them.

Other functions that plot oce data: 
\code{\link{download.amsr}()},
\code{\link{plot,adp-method}},
\code{\link{plot,amsr-method}},
\code{\link{plot,argo-method}},
\code{\link{plot,bremen-method}},
\code{\link{plot,cm-method}},
\code{\link{plot,coastline-method}},
\code{\link{plot,ctd-method}},
\code{\link{plot,gps-method}},
\code{\link{plot,ladp-method}},
\code{\link{plot,landsat-method}},
\code{\link{plot,lisst-method}},
\code{\link{plot,lobo-method}},
\code{\link{plot,met-method}},
\code{\link{plot,odf-method}},
\code{\link{plot,rsk-method}},
\code{\link{plot,satellite-method}},
\code{\link{plot,sealevel-method}},
\code{\link{plot,section-method}},
\code{\link{plot,tidem-method}},
\code{\link{plot,topo-method}},
\code{\link{plot,windrose-method}},
\code{\link{plot,xbt-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{tidem-class}}

Other things related to adv data: 
\code{\link{[[,adv-method}},
\code{\link{[[<-,adv-method}},
\code{\link{adv-class}},
\code{\link{adv}},
\code{\link{beamName}()},
\code{\link{beamToXyz}()},
\code{\link{enuToOtherAdv}()},
\code{\link{enuToOther}()},
\code{\link{read.adv.nortek}()},
\code{\link{read.adv.sontek.adr}()},
\code{\link{read.adv.sontek.serial}()},
\code{\link{read.adv.sontek.text}()},
\code{\link{read.adv}()},
\code{\link{rotateAboutZ}()},
\code{\link{subset,adv-method}},
\code{\link{summary,adv-method}},
\code{\link{toEnuAdv}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdv}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley
}
\concept{functions that plot oce data}
\concept{things related to adv data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/windrose.R
\docType{class}
\name{windrose-class}
\alias{windrose-class}
\title{Class to Store Windrose Data}
\description{
This class stores \code{windrose} objects, which
store statistical information about winds, mainly for
plotting as "wind rose" plots with \code{\link[=plot,windrose-method]{plot,windrose-method()}}.
Unlike most other \linkS4class{oce} objects, there is no reading
method for \code{windrose} objects, because there is no standard way to store
wind data in files; instead, \code{\link[=as.windrose]{as.windrose()}} is provided
to construct \code{windrose} objects.
}
\section{Slots}{

\describe{
\item{\code{data}}{As with all \code{oce} objects, the \code{data} slot
for \code{windrose} objects is a
\link{list} containing the main data for the object.
{}}

\item{\code{metadata}}{As with all \code{oce} objects, the \code{metadata} slot
for \code{windrose} objects is a \link{list} containing
information about the \code{data} or about the object itself.
{}}

\item{\code{processingLog}}{As with all \code{oce} objects, the \code{processingLog} slot
for \code{windrose} objects is a
\link{list} with entries describing the creation and evolution
of the object. The contents are updated by various \code{oce} functions to
keep a record of processing steps.  Object summaries and
\code{\link[=processingLogShow]{processingLogShow()}} both display the log.}
}}

\section{Modifying slot contents}{


Although the \code{[[<-} operator may permit modification of the contents
of \linkS4class{windrose} objects (see \code{\link{[[<-,windrose-method}}),
it is better to use \code{\link[=oceSetData]{oceSetData()}} and \code{\link[=oceSetMetadata]{oceSetMetadata()}},
because those functions save an entry in the \code{processingLog}
that describes the change.
}

\section{Retrieving slot contents}{


The full contents of the \code{data} and \code{metadata} slots of a \linkS4class{windrose}
object may be retrieved in the standard R way using \code{\link[=slot]{slot()}}. For
example \code{slot(o,"data")} returns the \code{data} slot of an object named \code{o},
and similarly \code{slot(o,"metadata")} returns
the \code{metadata} slot.

The slots may also be obtained with the \code{\link{[[,windrose-method}}
operator, as e.g. \code{o[["data"]]} and \code{o[["metadata"]]}, respectively.

The \code{\link{[[,windrose-method}} operator can also
be used to retrieve items from within the \code{data} and \code{metadata} slots.
For example, \code{o[["temperature"]]} can be used to retrieve temperature
from an object containing that quantity. The rule is that a named
quantity is sought first within the object's \code{metadata} slot,
with the \code{data} slot being checked only if \code{metadata} does not
contain the item. This \code{[[} method can also be used to get
certain derived quantities, if the object contains sufficient
information to calculate them. For example, an object that holds
(practical) salinity, temperature and pressure, along with
longitude and latitude, has sufficient information to compute
Absolute Salinity, and so \code{o[["SA"]]} will yield the
calculated Absolute Salinity.

It is also possible to find items more directly, using \code{\link[=oceGetData]{oceGetData()}} and
\code{\link[=oceGetMetadata]{oceGetMetadata()}}, but neither of these functions can
retrieve derived items.
}

\seealso{
Other classes provided by oce: 
\code{\link{adp-class}},
\code{\link{adv-class}},
\code{\link{argo-class}},
\code{\link{bremen-class}},
\code{\link{cm-class}},
\code{\link{coastline-class}},
\code{\link{ctd-class}},
\code{\link{lisst-class}},
\code{\link{lobo-class}},
\code{\link{met-class}},
\code{\link{oce-class}},
\code{\link{odf-class}},
\code{\link{rsk-class}},
\code{\link{sealevel-class}},
\code{\link{section-class}},
\code{\link{topo-class}},
\code{\link{xbt-class}}

Other things related to windrose data: 
\code{\link{[[,windrose-method}},
\code{\link{[[<-,windrose-method}},
\code{\link{as.windrose}()},
\code{\link{plot,windrose-method}},
\code{\link{summary,windrose-method}}
}
\concept{classes provided by oce}
\concept{things related to windrose data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{binCount1D}
\alias{binCount1D}
\title{Bin-count vector data}
\usage{
binCount1D(x, xbreaks)
}
\arguments{
\item{x}{vector of numerical values.}

\item{xbreaks}{Vector of values of x at the boundaries between bins, calculated using
\code{\link[=pretty]{pretty()}} if not supplied.}
}
\value{
A list with the following elements: the breaks (\code{xbreaks},
midpoints (\code{xmids}) between those breaks, and
the count (\code{number}) of \code{x} values between successive breaks.
}
\description{
Count the number of elements of a given vector that fall within
successive pairs of values within a second vector.
}
\seealso{
Other bin-related functions: 
\code{\link{binApply1D}()},
\code{\link{binApply2D}()},
\code{\link{binAverage}()},
\code{\link{binCount2D}()},
\code{\link{binMean1D}()},
\code{\link{binMean2D}()}
}
\author{
Dan Kelley
}
\concept{bin-related functions}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/astronomy.R
\name{angle2hms}
\alias{angle2hms}
\title{Convert astronomical angle in degrees to hours, minutes and seconds}
\usage{
angle2hms(angle)
}
\arguments{
\item{angle}{numerical value giving an angle in degrees}
}
\value{
\link{angle2hms} returns a list containing
values \code{time} (a numerical value for decimal hour, between 0 and 24),
\code{hour}, \code{minute}, and \code{second} (the last of which may have a fractional
part), and \code{string}, a character value indicates the time in hour-minute-second
notation, with the second part to two decimal places and intervening \code{h},
\code{m} and \code{s} characters between the units.
}
\description{
The purpose of \link{angle2hms} is to facilitate comparison of
\code{rightAscension} angles computed by \code{\link[=sunAngle]{sunAngle()}} and \code{\link[=moonAngle]{moonAngle()}}
with angles reported in astronomical sources and software, which
often employ an hour-minute-second notation. In that notation,
decimal hour is computed as 24/360 times the angle in degrees,
and from that decimal hour are compute integer hour and minute
values, plus a decimal second value. It is common in the
astronomical literature to use strings to represent the results,
e.g. with \eqn{11^h40^m48^s.10}{11h40m48s.10} for the
value used in the \dQuote{Examples}; see Chapter 1 of
Meeus (1991) for more on angle calculation and representation.
}
\examples{
# A randomly-chosen example on page 99 of Meeus (1991).
angle2hms(177.74208) # string component 11h50m58s.10

}
\references{
\itemize{
\item Meeus, Jean. Astronomical Algorithms. Second Edition.
Richmond, Virginia, USA: Willmann-Bell, 1991.
}
}
\seealso{
Other things related to astronomy: 
\code{\link{eclipticalToEquatorial}()},
\code{\link{equatorialToLocalHorizontal}()},
\code{\link{julianCenturyAnomaly}()},
\code{\link{julianDay}()},
\code{\link{moonAngle}()},
\code{\link{siderealTime}()},
\code{\link{sunAngle}()},
\code{\link{sunDeclinationRightAscension}()}
}
\author{
Dan Kelley
}
\concept{things related to astronomy}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/lobo.R
\name{[[<-,lobo-method}
\alias{[[<-,lobo-method}
\title{Replace Parts of a LOBO Object}
\usage{
\S4method{[[}{lobo}(x, i, j, ...) <- value
}
\arguments{
\item{x}{a \linkS4class{lobo} object.}

\item{i}{character value naming the item to replace.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{optional additional information (ignored).}

\item{value}{The value to be placed into \code{x}, somewhere.}
}
\description{
The \code{[[<-} method works for all \linkS4class{oce} objects.
The purpose, as with the related extraction method, \code{[[},
is to insulate users from the internal details of \linkS4class{oce}
objects, by looking for items within the various storage
slots of the object. Items not actually stored can also be
replaced, including units and data-quality
flags.
}
\details{
As with \code{[[} method, the procedure works in steps.

First, the \code{metadata} slot of \code{x} is checked to
see whether it contains something named with \code{i}.
If so, then the named item is replaced with \code{value}.

Otherwise, if the string value of \code{i} ends in \code{Unit}, then the
characters preceding that are taken as the name of a variable, and
the \code{metadata} slot of \code{x} is updated to store that unit, e.g.\preformatted{x[["temperatureUnits"]] <- list(unit=expression(degree*F),scale="")
}

Similarly, if \code{i} ends in \code{Flag}, then quality-control
flags are set up as defined by \code{result}, e.g.\preformatted{o[["temperatureFlags"]] <- c(2,4,2,2)
}

Otherwise, \code{\link[=pmatch]{pmatch()}} is used for a partial-string match with
the names of the items that are in the \code{data} slot of \code{x}.
The first item found (if any) is then updated to hold the value \code{result}.

If none of these conditions is met, a warning is issued.
}
\seealso{
Other functions that replace parts of oce objects: 
\code{\link{[[<-,adp-method}},
\code{\link{[[<-,amsr-method}},
\code{\link{[[<-,argo-method}},
\code{\link{[[<-,bremen-method}},
\code{\link{[[<-,cm-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{[[<-,echosounder-method}},
\code{\link{[[<-,g1sst-method}},
\code{\link{[[<-,gps-method}},
\code{\link{[[<-,ladp-method}},
\code{\link{[[<-,landsat-method}},
\code{\link{[[<-,lisst-method}},
\code{\link{[[<-,met-method}},
\code{\link{[[<-,oce-method}},
\code{\link{[[<-,odf-method}},
\code{\link{[[<-,rsk-method}},
\code{\link{[[<-,sealevel-method}},
\code{\link{[[<-,section-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{[[<-,topo-method}},
\code{\link{[[<-,windrose-method}},
\code{\link{[[<-,xbt-method}}

Other things related to lobo data: 
\code{\link{[[,lobo-method}},
\code{\link{as.lobo}()},
\code{\link{lobo-class}},
\code{\link{lobo}},
\code{\link{plot,lobo-method}},
\code{\link{read.lobo}()},
\code{\link{subset,lobo-method}},
\code{\link{summary,lobo-method}}
}
\concept{functions that replace parts of oce objects}
\concept{things related to lobo data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/rsk.R
\name{[[,rsk-method}
\alias{[[,rsk-method}
\title{Extract Something From a Rsk Object}
\usage{
\S4method{[[}{rsk}(x, i, j, ...)
}
\arguments{
\item{x}{an \linkS4class{rsk} object.}

\item{i}{character value indicating the name of an item to extract.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{ignored.}
}
\description{
Generally, the \code{[[} method lets users extract information from \code{oce}
objects, without having to know the details of the internal storage.  For
many \code{oce} sub-classes, \code{[[} can also return quantities that are computed
from the object's contents.
}
\details{
A two-step process is used to try to find the requested information. First, a
class-specific function is used (see \dQuote{Details of the Specialized
Method}).  If this yields nothing, then a general method is used (see
\dQuote{Details of the General Method}).  If both methods fail, then \code{[[}
returns NULL.

Some understanding of the subclass is required to know what can be retrieved
with \code{[[}. When dealing with an unfamiliar subclass, it can be useful to
first use \code{x[["?"]]} to get a listing of the retrievable items. See
\dQuote{Details of the Specialized Method} for more information.
}
\section{Details of the Specialized Method}{

\itemize{
\item If \code{i} is \code{"?"}, then the return value is a list
containing four items, each of which is a character vector
holding the names of things that can be accessed with \code{[[}.
The \code{data} and \code{metadata} items hold the names of
entries in the object's data and metadata
slots, respectively. The \code{dataDerived}
and \code{metadataDerived} items are each NULL, because
no derived values are defined by \linkS4class{rsk} objects.
}
}

\section{Details of the General Method}{


Note: the text of this section is identical for all \code{oce} subclasses, and so
some of what you read here may not be relevant to the class being described
in this help page.

If the specialized method produces no matches, the following generalized
method is applied. As with the specialized method, the procedure hinges first
on the values of \code{i} and, optionally, \code{j}. The work proceeds in steps, by
testing a sequence of possible conditions in sequence.
\enumerate{
\item A check is made as to whether \code{i} names one of the standard \code{oce} slots.
If so, \code{[[} returns the slot contents of that slot.  Thus, \code{x[["metadata"]]}
will retrieve the \code{metadata} slot, while \code{x[["data"]]} and
\code{x[["processingLog"]]} return those slots.
\item If \code{i} is a string ending in the \code{"Unit"}, then the characters preceding
that string are taken to be the name of an item in the data object, and a
list containing the unit is returned (or \code{NULL} if there is no such unit).
This list consists of an item named \code{unit}, which is an \code{\link[=expression]{expression()}}, and
an item named \code{scale}, which is a string describing the measurement scale.
If the string ends in \code{" unit"}, e.g.  \code{x[["temperature unit"]]} (note the
space), then just the expression is returned, and if it ends in \code{" scale"},
then just the scale is returned.
\item If \code{i} is a string ending in \code{"Flag"}, then the corresponding data-quality
flag is returned (or \code{NULL} if there is no such flag).
\item If the object holds hydrographic information (pressure, salinity,
temperature, longitude and latitude) then another set of possibilities
arises.  If \code{i} is \code{"sigmaTheta"}, then the value of \code{\link[=swSigmaTheta]{swSigmaTheta()}} is
called with \code{x} as the sole argument, and the results are returned.
Similarly, \code{\link[=swSigma0]{swSigma0()}} is used if \code{i="sigma0"}, and \code{\link[=swSpice]{swSpice()}} is used if
\code{i="spice"}. Of course, these actions only make sense for objects that
contain the relevant items within their \code{data} slot.
\item After these possibilities are eliminated, the action depends on whether
\code{j} has been provided.  If \code{j} is not provided, or is the string \code{""}, then
\code{i} is sought in the \code{metadata} slot, and then in the \code{data} slot, returning
whichever is found first.  In other words, if \code{j} is not provided, the
\code{metadata} slot takes preference over the \code{data} slot. However, if \code{j} is
provided, then it must be either the string \code{"metadata"} or \code{"data"}, and it
directs where to look.
\item If none of the above-listed conditions holds, then \code{NULL} is returned.
}
}

\seealso{
Other functions that extract parts of oce objects: 
\code{\link{[[,adp-method}},
\code{\link{[[,adv-method}},
\code{\link{[[,amsr-method}},
\code{\link{[[,argo-method}},
\code{\link{[[,bremen-method}},
\code{\link{[[,cm-method}},
\code{\link{[[,coastline-method}},
\code{\link{[[,ctd-method}},
\code{\link{[[,echosounder-method}},
\code{\link{[[,g1sst-method}},
\code{\link{[[,gps-method}},
\code{\link{[[,ladp-method}},
\code{\link{[[,landsat-method}},
\code{\link{[[,lisst-method}},
\code{\link{[[,lobo-method}},
\code{\link{[[,met-method}},
\code{\link{[[,oce-method}},
\code{\link{[[,odf-method}},
\code{\link{[[,sealevel-method}},
\code{\link{[[,section-method}},
\code{\link{[[,tidem-method}},
\code{\link{[[,topo-method}},
\code{\link{[[,windrose-method}},
\code{\link{[[,xbt-method}},
\code{\link{[[<-,adv-method}}

Other things related to rsk data: 
\code{\link{[[<-,rsk-method}},
\code{\link{as.rsk}()},
\code{\link{plot,rsk-method}},
\code{\link{read.rsk}()},
\code{\link{rsk-class}},
\code{\link{rskPatm}()},
\code{\link{rskToc}()},
\code{\link{rsk}},
\code{\link{subset,rsk-method}},
\code{\link{summary,rsk-method}}
}
\author{
Dan Kelley
}
\concept{functions that extract parts of oce objects}
\concept{things related to rsk data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{lonFormat}
\alias{lonFormat}
\title{Format a longitude}
\usage{
lonFormat(lon, digits = max(6, getOption("digits") - 1))
}
\arguments{
\item{lon}{longitude in \eqn{^\circ}{deg}N east of Greenwich.}

\item{digits}{the number of significant digits to use when printing.}
}
\value{
A character string.
}
\description{
Format a longitude, using "W" for west longitude.
}
\seealso{
\code{\link[=latFormat]{latFormat()}} and \code{\link[=latlonFormat]{latlonFormat()}}.
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/amsr.R
\name{subset,amsr-method}
\alias{subset,amsr-method}
\title{Subset an amsr Object}
\usage{
\S4method{subset}{amsr}(x, subset, ...)
}
\arguments{
\item{x}{an \linkS4class{amsr} object.}

\item{subset}{an expression indicating how to subset \code{x}.}

\item{...}{ignored.}
}
\value{
An \linkS4class{amsr} object.
}
\description{
Return a subset of a \linkS4class{amsr} object.
}
\details{
This function is used to subset data within an \linkS4class{amsr}
object by \code{longitude} or by \code{latitude}.  These two methods cannot
be combined in a single call, so two calls are required, as shown
in the Example.
}
\examples{
library(oce)
data(amsr) # see ?amsr for how to read and composite such objects
sub <- subset(amsr, -75 < longitude & longitude < -45)
sub <- subset(sub,   40 < latitude  &  latitude <  50)
plot(sub)
data(coastlineWorld)
lines(coastlineWorld[['longitude']], coastlineWorld[['latitude']])

}
\seealso{
Other things related to amsr data: 
\code{\link{[[,amsr-method}},
\code{\link{[[<-,amsr-method}},
\code{\link{amsr-class}},
\code{\link{amsr}},
\code{\link{composite,amsr-method}},
\code{\link{download.amsr}()},
\code{\link{plot,amsr-method}},
\code{\link{read.amsr}()},
\code{\link{summary,amsr-method}}

Other functions that subset oce objects: 
\code{\link{subset,adp-method}},
\code{\link{subset,adv-method}},
\code{\link{subset,argo-method}},
\code{\link{subset,cm-method}},
\code{\link{subset,coastline-method}},
\code{\link{subset,ctd-method}},
\code{\link{subset,echosounder-method}},
\code{\link{subset,lobo-method}},
\code{\link{subset,met-method}},
\code{\link{subset,oce-method}},
\code{\link{subset,odf-method}},
\code{\link{subset,rsk-method}},
\code{\link{subset,sealevel-method}},
\code{\link{subset,section-method}},
\code{\link{subset,topo-method}},
\code{\link{subset,xbt-method}}
}
\author{
Dan Kelley
}
\concept{functions that subset oce objects}
\concept{things related to amsr data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/xbt.R
\docType{class}
\name{xbt-class}
\alias{xbt-class}
\title{Class to Store XBT Data}
\description{
This class stores expendable bathythermograph (XBT) data, e.g. from a Sippican
device.  Reference 1 gives some information on Sippican
devices, and reference 2 is a useful introduction to the
modern literature on XBTs in general.
}
\section{Slots}{

\describe{
\item{\code{data}}{As with all \code{oce} objects, the \code{data} slot
for \code{xbt} objects is a
\link{list} containing the main data for the object.
The key items stored in this slot are \code{depth} (or \code{z}) and \code{temperature}, although some datasets also have \code{soundSpeed}.  Note that \code{depth} and \code{z} are inferred from time in water, using an empirical formula for instrument descent rate, and that \code{soundSpeed} is #' calculated using a fixed  practical salinity of 35. Note that the \code{[[} accessor will compute any of \code{depth}, \code{z} or \code{pressure}, based on whatever is in the data object.  Similarly, \code{soundspeed} will compute sound speed (assuming a practical salinity of 35), if that that item is present in the \code{data} slot.}

\item{\code{metadata}}{As with all \code{oce} objects, the \code{metadata} slot
for \code{xbt} objects is a \link{list} containing
information about the \code{data} or about the object itself.
{}}

\item{\code{processingLog}}{As with all \code{oce} objects, the \code{processingLog} slot
for \code{xbt} objects is a
\link{list} with entries describing the creation and evolution
of the object. The contents are updated by various \code{oce} functions to
keep a record of processing steps.  Object summaries and
\code{\link[=processingLogShow]{processingLogShow()}} both display the log.}
}}

\section{Modifying slot contents}{


Although the \code{[[<-} operator may permit modification of the contents
of \linkS4class{xbt} objects (see \code{\link{[[<-,xbt-method}}),
it is better to use \code{\link[=oceSetData]{oceSetData()}} and \code{\link[=oceSetMetadata]{oceSetMetadata()}},
because those functions save an entry in the \code{processingLog}
that describes the change.
}

\section{Retrieving slot contents}{


The full contents of the \code{data} and \code{metadata} slots of a \linkS4class{xbt}
object may be retrieved in the standard R way using \code{\link[=slot]{slot()}}. For
example \code{slot(o,"data")} returns the \code{data} slot of an object named \code{o},
and similarly \code{slot(o,"metadata")} returns
the \code{metadata} slot.

The slots may also be obtained with the \code{\link{[[,xbt-method}}
operator, as e.g. \code{o[["data"]]} and \code{o[["metadata"]]}, respectively.

The \code{\link{[[,xbt-method}} operator can also
be used to retrieve items from within the \code{data} and \code{metadata} slots.
For example, \code{o[["temperature"]]} can be used to retrieve temperature
from an object containing that quantity. The rule is that a named
quantity is sought first within the object's \code{metadata} slot,
with the \code{data} slot being checked only if \code{metadata} does not
contain the item. This \code{[[} method can also be used to get
certain derived quantities, if the object contains sufficient
information to calculate them. For example, an object that holds
(practical) salinity, temperature and pressure, along with
longitude and latitude, has sufficient information to compute
Absolute Salinity, and so \code{o[["SA"]]} will yield the
calculated Absolute Salinity.

It is also possible to find items more directly, using \code{\link[=oceGetData]{oceGetData()}} and
\code{\link[=oceGetMetadata]{oceGetMetadata()}}, but neither of these functions can
retrieve derived items.
}

\references{
\enumerate{
\item Sippican, Inc. "Bathythermograph Data Acquisition System: Installation, Operation and Maintenance
Manual (P/N 308195, Rev. A)," 2003.
https://pages.uoregon.edu/drt/MGL0910_Science_Report/attachments/MK21_ISA_Manual_Rev_A.pdf.
\item Cheng, Lijing, John Abraham, Gustavo Goni, Timothy Boyer, Susan Wijffels, Rebecca
Cowley, Viktor Gouretski, et al. "XBT Science: Assessment of Instrumental Biases and Errors."
Bulletin of the American Meteorological Society 97, no. 6 (June 2016): 924-33.
\code{10.1175/BAMS-D-15-00031.1}
}
}
\seealso{
Other things related to xbt data: 
\code{\link{[[,xbt-method}},
\code{\link{[[<-,xbt-method}},
\code{\link{as.xbt}()},
\code{\link{plot,xbt-method}},
\code{\link{read.xbt.noaa1}()},
\code{\link{read.xbt}()},
\code{\link{subset,xbt-method}},
\code{\link{summary,xbt-method}},
\code{\link{xbt.edf}},
\code{\link{xbt}}

Other classes provided by oce: 
\code{\link{adp-class}},
\code{\link{adv-class}},
\code{\link{argo-class}},
\code{\link{bremen-class}},
\code{\link{cm-class}},
\code{\link{coastline-class}},
\code{\link{ctd-class}},
\code{\link{lisst-class}},
\code{\link{lobo-class}},
\code{\link{met-class}},
\code{\link{oce-class}},
\code{\link{odf-class}},
\code{\link{rsk-class}},
\code{\link{sealevel-class}},
\code{\link{section-class}},
\code{\link{topo-class}},
\code{\link{windrose-class}}
}
\author{
Dan Kelley
}
\concept{classes provided by oce}
\concept{things related to xbt data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/echosounder.R
\docType{data}
\name{echosounder}
\alias{echosounder}
\title{Echosounder Dataset}
\source{
This file came from the SLEIWEX-2008 experiment, and was decimated
using \code{\link[=decimate]{decimate()}} with \code{by=c()}.
}
\description{
This is degraded subsample of measurements that were made with a Biosonics
scientific echosounder, as part of the St Lawrence Internal Wave Experiment
(SLEIWEX).
}
\seealso{
Other datasets provided with oce: 
\code{\link{adp}},
\code{\link{adv}},
\code{\link{amsr}},
\code{\link{argo}},
\code{\link{cm}},
\code{\link{coastlineWorld}},
\code{\link{ctdRaw}},
\code{\link{ctd}},
\code{\link{landsat}},
\code{\link{lisst}},
\code{\link{lobo}},
\code{\link{met}},
\code{\link{ocecolors}},
\code{\link{rsk}},
\code{\link{sealevelTuktoyaktuk}},
\code{\link{sealevel}},
\code{\link{section}},
\code{\link{topoWorld}},
\code{\link{wind}},
\code{\link{xbt}}

Other things related to echosounder data: 
\code{\link{[[,echosounder-method}},
\code{\link{[[<-,echosounder-method}},
\code{\link{as.echosounder}()},
\code{\link{echosounder-class}},
\code{\link{findBottom}()},
\code{\link{plot,echosounder-method}},
\code{\link{read.echosounder}()},
\code{\link{subset,echosounder-method}},
\code{\link{summary,echosounder-method}}
}
\author{
Dan Kelley
}
\concept{datasets provided with oce}
\concept{things related to echosounder data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/amsr.R
\name{download.amsr}
\alias{download.amsr}
\title{Download and Cache an amsr File}
\usage{
download.amsr(
  year,
  month,
  day,
  destdir = ".",
  server = "http://data.remss.com/amsr2/bmaps_v08"
)
}
\arguments{
\item{year, month, day}{Numerical values of the year, month, and day
of the desired dataset. Note that one file is archived per day,
so these three values uniquely identify a dataset.
If \code{day} and \code{month} are not provided but \code{day} is,
then the time is provided in a relative sense, based on the present
date, with \code{day} indicating the number of days in the past.
Owing to issues with timezones and the time when the data
are uploaded to the server, \code{day=3} may yield the
most recent available data. For this reason, there is a
third option, which is to leave \code{day} unspecified, which
works as though \code{day=3} had been given.}

\item{destdir}{A string naming the directory in which to cache the downloaded file.
The default is to store in the present directory, but many users find it more
helpful to use something like \code{"~/data/amsr"} for this, to collect all
downloaded amsr files in one place.}

\item{server}{A string naming the server from which data
are to be acquired. See \dQuote{History}.}
}
\value{
A character value indicating the filename of the result; if
there is a problem of any kind, the result will be the empty
string.
}
\description{
If the file is already present in \code{destdir}, then it is not
downloaded again. The default \code{destdir} is the present directory,
but it probably makes more sense to use something like \code{"~/data/amsr"}
to make it easy for scripts in other directories to use the cached data.
The file is downloaded with \code{\link[=download.file]{download.file()}}.
}
\section{History}{

Until 25 March 2017, the default server was
\code{"ftp.ssmi.com/amsr2/bmaps_v07.2"}, but this was changed when the author
discovered that this FTP site had been changed to require users to create
accounts to register for downloads.  The default was changed to
\code{"http://data.remss.com/amsr2/bmaps_v07.2"} on the named date.
This site was found by a web search, but it seems to provide proper data.
It is assumed that users will do some checking on the best source.

On 23 January 2018, it was noticed that the server-url naming convention
had changed, e.g.
\verb{http://data.remss.com/amsr2/bmaps_v07.2/y2017/m01/f34_20170114v7.2.gz}
becoming
\verb{http://data.remss.com/amsr2/bmaps_v08/y2017/m01/f34_20170114v8.gz}
}

\examples{
\dontrun{
## The download takes several seconds.
f <- download.amsr(2017, 1, 14) # Jan 14, 2017
d <- read.amsr(f)
plot(d)
mtext(d[["filename"]], side=3, line=0, adj=0)
}

}
\references{
\verb{http://images.remss.com/amsr/amsr2_data_daily.html}
provides daily images going back to 2012. Three-day,
monthly, and monthly composites are also provided on that site.
}
\seealso{
Other functions that download files: 
\code{\link{download.coastline}()},
\code{\link{download.met}()},
\code{\link{download.topo}()}

Other functions that plot oce data: 
\code{\link{plot,adp-method}},
\code{\link{plot,adv-method}},
\code{\link{plot,amsr-method}},
\code{\link{plot,argo-method}},
\code{\link{plot,bremen-method}},
\code{\link{plot,cm-method}},
\code{\link{plot,coastline-method}},
\code{\link{plot,ctd-method}},
\code{\link{plot,gps-method}},
\code{\link{plot,ladp-method}},
\code{\link{plot,landsat-method}},
\code{\link{plot,lisst-method}},
\code{\link{plot,lobo-method}},
\code{\link{plot,met-method}},
\code{\link{plot,odf-method}},
\code{\link{plot,rsk-method}},
\code{\link{plot,satellite-method}},
\code{\link{plot,sealevel-method}},
\code{\link{plot,section-method}},
\code{\link{plot,tidem-method}},
\code{\link{plot,topo-method}},
\code{\link{plot,windrose-method}},
\code{\link{plot,xbt-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{tidem-class}}

Other things related to amsr data: 
\code{\link{[[,amsr-method}},
\code{\link{[[<-,amsr-method}},
\code{\link{amsr-class}},
\code{\link{amsr}},
\code{\link{composite,amsr-method}},
\code{\link{plot,amsr-method}},
\code{\link{read.amsr}()},
\code{\link{subset,amsr-method}},
\code{\link{summary,amsr-method}}
}
\concept{functions that download files}
\concept{functions that plot oce data}
\concept{things related to amsr data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.R
\name{read.adp}
\alias{read.adp}
\title{Read an ADP File}
\usage{
read.adp(
  file,
  from,
  to,
  by,
  tz = getOption("oceTz"),
  longitude = NA,
  latitude = NA,
  manufacturer,
  monitor = FALSE,
  despike = FALSE,
  processingLog,
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{file}{a connection or a character string giving the name of the file
to load.  (For \code{read.adp.sontek.serial}, this is generally a list of
files, which will be concatenated.)}

\item{from}{indication of the first profile to read.  This can be an
integer, the sequence number of the first profile to read, or a POSIXt time
before which profiles should be skipped, or a character string that converts
to a POSIXt time (assuming UTC timezone).  See \dQuote{Examples}, and make
careful note of the use of the \code{tz} argument. If \code{from} is not
supplied, it defaults to 1.}

\item{to}{an optional indication of the last profile to read, in a
format as described for \code{from}.  As a special case, \code{to=0} means
to read the file to the end.  If \code{to} is not supplied, then it defaults
to 0.}

\item{by}{an optional indication of the stride length to use while walking through
the file.  If this is an integer, then \code{by-1} profiles are skipped
between each pair of profiles that is read, e.g. the default \code{by=1}
means to read all the data.  (For RDI files \emph{only}, there are some
extra features to avoid running out of memory; see \dQuote{Memory considerations}.)}

\item{tz}{character string indicating time zone to be assumed in the data.}

\item{longitude}{optional signed number indicating the longitude in degrees
East.}

\item{latitude}{optional signed number indicating the latitude in degrees
North.}

\item{manufacturer}{an optional character string indicating the manufacturer, used by
the general function \code{read.adp} to select a subsidiary function to use. If this
is not given, then \code{\link[=oceMagic]{oceMagic()}} is used to try to infer the type. If this
is provided, then the value \code{"rdi"} will cause \code{\link[=read.adp.rdi]{read.adp.rdi()}}
to be used, \code{"nortek"} will cause \code{\link[=read.adp.nortek]{read.adp.nortek()}} to be used,
and \code{"sontek"} will cause \code{\link[=read.adp.sontek]{read.adp.sontek()}} to be used.}

\item{monitor}{boolean value indicating whether to indicate the progress
of reading the file, by using \code{\link[=txtProgressBar]{txtProgressBar()}} or otherwise.  The value
of \code{monitor} is changed to \code{FALSE} automatically, for non-interactive
sessions.}

\item{despike}{if \code{TRUE}, \code{\link[=despike]{despike()}} will be used to clean
anomalous spikes in heading, etc.}

\item{processingLog}{if provided, the action item to be stored in the log.
(Typically only provided for internal calls; the default that it provides is
better for normal calls by a user.)}

\item{debug}{a flag that turns on debugging.  Set to 1 to get a moderate
amount of debugging information, or to 2 to get more.}

\item{\dots}{optional additional arguments that some (but not all)
\verb{read.adp.*()} functions pass to lower-level functions.}
}
\value{
An \linkS4class{adp} object.
The contents of that object make sense for the particular instrument
type under study, e.g. if the data file contains
NMEA strings, then navigational data will be stored in an item
called \code{nmea} in the \code{data} slot).
}
\description{
Read an ADP data file, producing an \linkS4class{adp} object.
}
\details{
Several file types can be handled.  Some of
these functions are wrappers that map to device names, e.g.
\code{read.aquadoppProfiler} does its work by calling
\code{read.adp.nortek}; in this context, it is worth noting that the
``aquadopp'' instrument is a one-cell profiler that might just as well have
been documented under the heading \code{\link[=read.adv]{read.adv()}}.
}
\seealso{
Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\author{
Dan Kelley and Clark Richards
}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{oce.grid}
\alias{oce.grid}
\title{Add a Grid to an Existing Oce Plot}
\usage{
oce.grid(xat, yat, col = "lightgray", lty = "dotted", lwd = par("lwd"))
}
\arguments{
\item{xat}{either a list of x values at which to draw the grid, or the return value from an oce plotting function}

\item{yat}{a list of y values at which to plot the grid (ignored if \code{gx} was a return value from an oce plotting function)}

\item{col}{color of grid lines (see \code{\link[=par]{par()}})}

\item{lty}{type for grid lines (see \code{\link[=par]{par()}})}

\item{lwd}{width for grid lines (see \code{\link[=par]{par()}})}
}
\description{
Add a Grid to an Existing Oce Plot
}
\details{
For plots not created by oce functions, or for missing \code{xat} and \code{yat},
this is the same as a call to \code{\link[=grid]{grid()}} with missing \code{nx} and
\code{ny}. However, if \code{xat} is the return value from certain oce functions,
a more sophisticated grid is constructed. The problem with \code{\link[=grid]{grid()}} is
that it cannot handle axes with non-uniform grids, e.g. those with time axes
that span months of differing lengths.

As of early February 2015, \code{oce.grid} handles \code{xat} produced as the
return value from the following functions: \code{\link[=imagep]{imagep()}} and
\code{\link[=oce.plot.ts]{oce.plot.ts()}}, \code{\link[=plot,adp-method]{plot,adp-method()}},
\code{\link[=plot,echosounder-method]{plot,echosounder-method()}}, and \code{\link[=plotTS]{plotTS()}}.
It makes no sense to try to use \code{oce.grid} for multipanel oce plots,
e.g. the default plot from \code{\link[=plot,adp-method]{plot,adp-method()}}.
}
\examples{
library(oce)
i <- imagep(volcano)
oce.grid(i, lwd=2)

data(sealevel)
i <- oce.plot.ts(sealevel[["time"]], sealevel[["elevation"]])
oce.grid(i, col='red')

data(ctd)
i <- plotTS(ctd)
oce.grid(i, col='red')

data(adp)
i <- plot(adp, which=1)
oce.grid(i, col='gray', lty=1)

data(echosounder)
i <- plot(echosounder)
oce.grid(i, col='pink', lty=1)

}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/xbt.R
\docType{data}
\name{xbt}
\alias{xbt}
\title{An XBT Object}
\usage{
data(xbt)
}
\description{
An \linkS4class{xbt} object created by using \code{\link[=read.xbt]{read.xbt()}} on a Sippican file created by extracting the near-surface
fraction of the sample provided in Section 5.5.6 of reference 1.
}
\examples{
library(oce)
data(xbt)
summary(xbt)
plot(xbt)

}
\references{
\enumerate{
\item Sippican, Inc. "Bathythermograph Data Acquisition System: Installation, Operation and Maintenance
Manual (P/N 308195, Rev. A)," 2003.
https://pages.uoregon.edu/drt/MGL0910_Science_Report/attachments/MK21_ISA_Manual_Rev_A.pdf.
}
}
\seealso{
Other datasets provided with oce: 
\code{\link{adp}},
\code{\link{adv}},
\code{\link{amsr}},
\code{\link{argo}},
\code{\link{cm}},
\code{\link{coastlineWorld}},
\code{\link{ctdRaw}},
\code{\link{ctd}},
\code{\link{echosounder}},
\code{\link{landsat}},
\code{\link{lisst}},
\code{\link{lobo}},
\code{\link{met}},
\code{\link{ocecolors}},
\code{\link{rsk}},
\code{\link{sealevelTuktoyaktuk}},
\code{\link{sealevel}},
\code{\link{section}},
\code{\link{topoWorld}},
\code{\link{wind}}

Other things related to xbt data: 
\code{\link{[[,xbt-method}},
\code{\link{[[<-,xbt-method}},
\code{\link{as.xbt}()},
\code{\link{plot,xbt-method}},
\code{\link{read.xbt.noaa1}()},
\code{\link{read.xbt}()},
\code{\link{subset,xbt-method}},
\code{\link{summary,xbt-method}},
\code{\link{xbt-class}},
\code{\link{xbt.edf}}
}
\concept{datasets provided with oce}
\concept{things related to xbt data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{oceColorsPAR}
\alias{oceColorsPAR}
\alias{oce.colorsPAR}
\title{Create colors suitable for PAR fields}
\usage{
oceColorsPAR(n)
}
\arguments{
\item{n}{number of colors to create.}
}
\value{
A vector of color specifications.
}
\description{
Create a set of colors for displaying PAR values,
based on the scheme devised by Thyng et al. (2016) and
presented in a python package by Thyng (2019).
The color specifications were transliterated from
python to R on 2015-09-29, but have not been adjusted
since, even though the python source has changed.  This
is to prevent breaking old \code{oce} code.  To get the
latest versions of these colours or other colours,
use the \CRANpkg{cmocean} R package
(Thyng, Richards, and Krylov, 2019) directly, as is
illustrated (with the "matter" scheme) in Example 2.
Note that the \CRANpkg{cmocean} core functions provide a way
to select between various versions of the colour schemes.
It is also worth considering the palettes provided by the
\CRANpkg{viridis} package, as illustrated (with the
"inferno" scheme) in Example 3.
}
\examples{
library(oce)

# Example 1
imagep(volcano, col=oceColorsPAR(128),
       zlab="oceColorsPAR")
\dontrun{
# Example 2 (requires the cmocean package)
imagep(volcano, col=cmocean::cmocean("matter"),
       zlab='cmocean::cmocean("matter")')}

\dontrun{
# Example 3 (requires the viridis package)
imagep(volcano, col=viridis::inferno,
       zlab='viridis::inferno')}

}
\references{
\itemize{
\item Thyng, Kristen, Chad Greene, Robert Hetland, Heather Zimmerle, and Steven DiMarco.
\dQuote{True Colors of Oceanography: Guidelines for Effective and Accurate Colormap Selection.}
Oceanography 29, no. 3 (September 1, 2016): 9‚Äì13.
\doi{10.5670/oceanog.2016.66}
\item Thyng, Kristen. Kthyng/Cmocean. Python, 2019.
\url{https://github.com/kthyng/cmocean}.
\item Thyng, Kristen, Clark Richards, and Ivan Krylov.
Cmocean: Beautiful Colour Maps for Oceanography (version 0.2), 2019.
\url{https://CRAN.R-project.org/package=cmocean}.
}

The following references provide information on choosing
colour schemes, that are suitable for viewers who have
colour deficiencies.

Light, Adam, and Patrick J. Bartlein. "The End of the Rainbow? Color
Schemes for Improved Data Graphics." \emph{Eos, Transactions American Geophysical
Union} 85, no. 40 (2004): 385.
DOI: 10.1029/2004EO400002

Stephenson, David B. "Comment on 'Color Schemes for Improved Data
Graphics', by A Light and P.J. Bartlein." \emph{Eos, Transactions American
Geophysical Union} 86, no.  20 (2005): 196.
DOI: 10.1029/2005EO200005

Light, Adam, and Patrick J. Bartlein. "Reply to 'Comment on Color
Schemes for Improved Data Graphics,' by A. Light and P.J.  Bartlein'."
\emph{Eos, Transactions American Geophysical Union} 86, no. 20 (2005): 196‚Äì196.
DOI: 10.1029/2005EO200006
}
\seealso{
Other things related to colors: 
\code{\link{colormapGMT}()},
\code{\link{colormap}()},
\code{\link{oceColors9B}()},
\code{\link{oceColorsCDOM}()},
\code{\link{oceColorsChlorophyll}()},
\code{\link{oceColorsClosure}()},
\code{\link{oceColorsDensity}()},
\code{\link{oceColorsFreesurface}()},
\code{\link{oceColorsGebco}()},
\code{\link{oceColorsJet}()},
\code{\link{oceColorsOxygen}()},
\code{\link{oceColorsPalette}()},
\code{\link{oceColorsPhase}()},
\code{\link{oceColorsSalinity}()},
\code{\link{oceColorsTemperature}()},
\code{\link{oceColorsTurbidity}()},
\code{\link{oceColorsTurbo}()},
\code{\link{oceColorsTwo}()},
\code{\link{oceColorsVelocity}()},
\code{\link{oceColorsViridis}()},
\code{\link{oceColorsVorticity}()},
\code{\link{ocecolors}}
}
\author{
Krysten M. Thyng (Python version), Dan Kelley (R transliteration)
}
\concept{things related to colors}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{oceColorsFreesurface}
\alias{oceColorsFreesurface}
\alias{oce.colorsFreesurface}
\title{Create colors suitable for freesurface fields}
\usage{
oceColorsFreesurface(n)
}
\arguments{
\item{n}{number of colors to create.}
}
\value{
A vector of color specifications.
}
\description{
Create a set of colors for displaying freesurface values,
based on the scheme devised by Thyng et al. (2016) and
presented in a python package by Thyng (2019).
The color specifications were transliterated from
python to R on 2015-09-29, but have not been adjusted
since, even though the python source has changed.  This
is to prevent breaking old \code{oce} code.  To get the
latest versions of these colours or other colours,
use the \CRANpkg{cmocean} R package
(Thyng, Richards, and Krylov, 2019) directly, as is
illustrated (with the "matter" scheme) in Example 2.
Note that the \CRANpkg{cmocean} core functions provide a way
to select between various versions of the colour schemes.
It is also worth considering the palettes provided by the
\CRANpkg{viridis} package, as illustrated (with the
"inferno" scheme) in Example 3.
}
\examples{
library(oce)

# Example 1
imagep(volcano, col=oceColorsFreesurface(128),
       zlab="oceColorsFreesurface")
\dontrun{
# Example 2 (requires the cmocean package)
imagep(volcano, col=cmocean::cmocean("matter"),
       zlab='cmocean::cmocean("matter")')}

\dontrun{
# Example 3 (requires the viridis package)
imagep(volcano, col=viridis::inferno,
       zlab='viridis::inferno')}

}
\references{
\itemize{
\item Thyng, Kristen, Chad Greene, Robert Hetland, Heather Zimmerle, and Steven DiMarco.
\dQuote{True Colors of Oceanography: Guidelines for Effective and Accurate Colormap Selection.}
Oceanography 29, no. 3 (September 1, 2016): 9‚Äì13.
\doi{10.5670/oceanog.2016.66}
\item Thyng, Kristen. Kthyng/Cmocean. Python, 2019.
\url{https://github.com/kthyng/cmocean}.
\item Thyng, Kristen, Clark Richards, and Ivan Krylov.
Cmocean: Beautiful Colour Maps for Oceanography (version 0.2), 2019.
\url{https://CRAN.R-project.org/package=cmocean}.
}

The following references provide information on choosing
colour schemes, that are suitable for viewers who have
colour deficiencies.

Light, Adam, and Patrick J. Bartlein. "The End of the Rainbow? Color
Schemes for Improved Data Graphics." \emph{Eos, Transactions American Geophysical
Union} 85, no. 40 (2004): 385.
DOI: 10.1029/2004EO400002

Stephenson, David B. "Comment on 'Color Schemes for Improved Data
Graphics', by A Light and P.J. Bartlein." \emph{Eos, Transactions American
Geophysical Union} 86, no.  20 (2005): 196.
DOI: 10.1029/2005EO200005

Light, Adam, and Patrick J. Bartlein. "Reply to 'Comment on Color
Schemes for Improved Data Graphics,' by A. Light and P.J.  Bartlein'."
\emph{Eos, Transactions American Geophysical Union} 86, no. 20 (2005): 196‚Äì196.
DOI: 10.1029/2005EO200006
}
\seealso{
Other things related to colors: 
\code{\link{colormapGMT}()},
\code{\link{colormap}()},
\code{\link{oceColors9B}()},
\code{\link{oceColorsCDOM}()},
\code{\link{oceColorsChlorophyll}()},
\code{\link{oceColorsClosure}()},
\code{\link{oceColorsDensity}()},
\code{\link{oceColorsGebco}()},
\code{\link{oceColorsJet}()},
\code{\link{oceColorsOxygen}()},
\code{\link{oceColorsPAR}()},
\code{\link{oceColorsPalette}()},
\code{\link{oceColorsPhase}()},
\code{\link{oceColorsSalinity}()},
\code{\link{oceColorsTemperature}()},
\code{\link{oceColorsTurbidity}()},
\code{\link{oceColorsTurbo}()},
\code{\link{oceColorsTwo}()},
\code{\link{oceColorsVelocity}()},
\code{\link{oceColorsViridis}()},
\code{\link{oceColorsVorticity}()},
\code{\link{ocecolors}}
}
\author{
Krysten M. Thyng (Python version), Dan Kelley (R transliteration)
}
\concept{things related to colors}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/coastline.R
\name{read.coastline}
\alias{read.coastline}
\title{Read a Coastline File}
\usage{
read.coastline(
  file,
  type = c("R", "S", "mapgen", "shapefile", "openstreetmap"),
  debug = getOption("oceDebug"),
  monitor = FALSE,
  processingLog
)
}
\arguments{
\item{file}{name of file containing coastline data.}

\item{type}{type of file, one of \code{"R"}, \code{"S"}, \code{"mapgen"},
\code{"shapefile"} or \code{"openstreetmap"}.}

\item{debug}{set to TRUE to print information about the header, etc.}

\item{monitor}{print a dot for every coastline segment read (ignored except
for reading "shapefile" type)}

\item{processingLog}{if provided, the action item to be stored in the log.
(Typically only provided for internal calls; the default that it provides is
better for normal calls by a user.)}
}
\value{
a \linkS4class{coastline} object.
}
\description{
Read a coastline file in R, Splus, mapgen, shapefile, or openstreetmap
format.
The S and R formats are identical, and consist of two columns, lon and lat,
with land-jump segments separated by lines with two NAs.
The MapGen format is of the form \preformatted{ # -b -16.179081 28.553943
-16.244793 28.563330 } BUG: the 'arc/info ungenerate' format is not yet
understood.
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{secondsToCtime}
\alias{secondsToCtime}
\title{Time interval as colon-separated string}
\usage{
secondsToCtime(sec)
}
\arguments{
\item{sec}{length of time interval in seconds.}
}
\value{
A string with a colon-separated time interval.
}
\description{
Convert a time interval to a colon-separated string
}
\examples{
library(oce)
cat("   10 s = ", secondsToCtime(10), "\n", sep="")
cat("   61 s = ", secondsToCtime(61), "\n", sep="")
cat("86400 s = ", secondsToCtime(86400), "\n", sep="")
}
\seealso{
See \code{\link[=ctimeToSeconds]{ctimeToSeconds()}}, the inverse of this.

Other things related to time: 
\code{\link{ctimeToSeconds}()},
\code{\link{julianCenturyAnomaly}()},
\code{\link{julianDay}()},
\code{\link{numberAsHMS}()},
\code{\link{numberAsPOSIXct}()},
\code{\link{unabbreviateYear}()}
}
\author{
Dan Kelley
}
\concept{things related to time}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{unwrapAngle}
\alias{unwrapAngle}
\title{Unwrap an angle that suffers modulo-360 problems}
\usage{
unwrapAngle(angle)
}
\arguments{
\item{angle}{an angle (in degrees) that is thought be near 360 degrees, with
added noise}
}
\value{
A list with two estimates: \code{mean} is based on an arithmetic
mean, and \code{median} is based on the median. Both are mapped to the range
0 to 360.
}
\description{
This is mostly used for instrument heading angles, in cases where the
instrument is aligned nearly northward, so that small variations in heading
(e.g. due to mooring motion) can yield values that swing from small angles
to large angles, because of the modulo-360 cut point.
The method is to use the cosine and sine of the angle, to construct "x" and
"y" values on a unit circle, then to find means and medians of x and y
respectively, and finally to use \code{\link[=atan2]{atan2()}} to infer the angles.
}
\examples{
library(oce)
true <- 355
a <- true + rnorm(100, sd=10)
a <- ifelse(a > 360, a - 360, a)
a2 <- unwrapAngle(a)
par(mar=c(3, 3, 5, 3))
hist(a, breaks=360)
abline(v=a2$mean, col="blue", lty="dashed")
abline(v=true, col="blue")
mtext("true (solid)\n estimate (dashed)", at=true, side=3, col="blue")
abline(v=mean(a), col="red")
mtext("mean", at=mean(a), side=3, col="red")
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/AllClass.R
\name{[[,oce-method}
\alias{[[,oce-method}
\title{Extract Something From an oce Object}
\usage{
\S4method{[[}{oce}(x, i, j, ...)
}
\arguments{
\item{x}{an \linkS4class{oce} object.}

\item{i}{character value indicating the name of an item to extract.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{ignored.}
}
\description{
Generally, the \code{[[} method lets users extract information from \code{oce}
objects, without having to know the details of the internal storage.  For
many \code{oce} sub-classes, \code{[[} can also return quantities that are computed
from the object's contents.
}
\details{
A two-step process is used to try to find the requested information. First, a
class-specific function is used (see \dQuote{Details of the Specialized
Method}).  If this yields nothing, then a general method is used (see
\dQuote{Details of the General Method}).  If both methods fail, then \code{[[}
returns NULL.

Some understanding of the subclass is required to know what can be retrieved
with \code{[[}. When dealing with an unfamiliar subclass, it can be useful to
first use \code{x[["?"]]} to get a listing of the retrievable items. See
\dQuote{Details of the Specialized Method} for more information.
}
\section{Details of the General Method}{


Note: the text of this section is identical for all \code{oce} subclasses, and so
some of what you read here may not be relevant to the class being described
in this help page.

If the specialized method produces no matches, the following generalized
method is applied. As with the specialized method, the procedure hinges first
on the values of \code{i} and, optionally, \code{j}. The work proceeds in steps, by
testing a sequence of possible conditions in sequence.
\enumerate{
\item A check is made as to whether \code{i} names one of the standard \code{oce} slots.
If so, \code{[[} returns the slot contents of that slot.  Thus, \code{x[["metadata"]]}
will retrieve the \code{metadata} slot, while \code{x[["data"]]} and
\code{x[["processingLog"]]} return those slots.
\item If \code{i} is a string ending in the \code{"Unit"}, then the characters preceding
that string are taken to be the name of an item in the data object, and a
list containing the unit is returned (or \code{NULL} if there is no such unit).
This list consists of an item named \code{unit}, which is an \code{\link[=expression]{expression()}}, and
an item named \code{scale}, which is a string describing the measurement scale.
If the string ends in \code{" unit"}, e.g.  \code{x[["temperature unit"]]} (note the
space), then just the expression is returned, and if it ends in \code{" scale"},
then just the scale is returned.
\item If \code{i} is a string ending in \code{"Flag"}, then the corresponding data-quality
flag is returned (or \code{NULL} if there is no such flag).
\item If the object holds hydrographic information (pressure, salinity,
temperature, longitude and latitude) then another set of possibilities
arises.  If \code{i} is \code{"sigmaTheta"}, then the value of \code{\link[=swSigmaTheta]{swSigmaTheta()}} is
called with \code{x} as the sole argument, and the results are returned.
Similarly, \code{\link[=swSigma0]{swSigma0()}} is used if \code{i="sigma0"}, and \code{\link[=swSpice]{swSpice()}} is used if
\code{i="spice"}. Of course, these actions only make sense for objects that
contain the relevant items within their \code{data} slot.
\item After these possibilities are eliminated, the action depends on whether
\code{j} has been provided.  If \code{j} is not provided, or is the string \code{""}, then
\code{i} is sought in the \code{metadata} slot, and then in the \code{data} slot, returning
whichever is found first.  In other words, if \code{j} is not provided, the
\code{metadata} slot takes preference over the \code{data} slot. However, if \code{j} is
provided, then it must be either the string \code{"metadata"} or \code{"data"}, and it
directs where to look.
\item If none of the above-listed conditions holds, then \code{NULL} is returned.
}
}

\seealso{
Many \code{oce} object classes have specialized versions
of \code{[[} that handle the details in specialized way.

Other functions that extract parts of oce objects: 
\code{\link{[[,adp-method}},
\code{\link{[[,adv-method}},
\code{\link{[[,amsr-method}},
\code{\link{[[,argo-method}},
\code{\link{[[,bremen-method}},
\code{\link{[[,cm-method}},
\code{\link{[[,coastline-method}},
\code{\link{[[,ctd-method}},
\code{\link{[[,echosounder-method}},
\code{\link{[[,g1sst-method}},
\code{\link{[[,gps-method}},
\code{\link{[[,ladp-method}},
\code{\link{[[,landsat-method}},
\code{\link{[[,lisst-method}},
\code{\link{[[,lobo-method}},
\code{\link{[[,met-method}},
\code{\link{[[,odf-method}},
\code{\link{[[,rsk-method}},
\code{\link{[[,sealevel-method}},
\code{\link{[[,section-method}},
\code{\link{[[,tidem-method}},
\code{\link{[[,topo-method}},
\code{\link{[[,windrose-method}},
\code{\link{[[,xbt-method}},
\code{\link{[[<-,adv-method}}
}
\author{
Dan Kelley
}
\concept{functions that extract parts of oce objects}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swSoundAbsorption}
\alias{swSoundAbsorption}
\title{Seawater sound absorption in dB/m}
\usage{
swSoundAbsorption(
  frequency,
  salinity,
  temperature,
  pressure,
  pH = 8,
  formulation = c("fisher-simmons", "francois-garrison")
)
}
\arguments{
\item{frequency}{The frequency of sound, in Hz.}

\item{salinity}{either practical salinity (in which case \code{temperature}
and \code{pressure} must be provided) \emph{or} an \code{oce} object, in
which case \code{salinity}, \code{temperature} (in the ITS-90 scale; see
next item), etc. are inferred from the object, ignoring the
other parameters, if they are supplied.}

\item{temperature}{\emph{in-situ} temperature (\eqn{^\circ}{deg}C), defined
on the ITS-90 scale.  This scale is used by GSW-style calculation (as
requested by setting \code{eos="gsw"}), and is the value contained within
\code{ctd} objects (and probably most other objects created with data
acquired in the past decade or two). Since the UNESCO-style calculation is
based on IPTS-68, the temperature is converted within the present function,
using \code{\link[=T68fromT90]{T68fromT90()}}.}

\item{pressure}{pressure (dbar)}

\item{pH}{seawater pH}

\item{formulation}{character string indicating the formulation to use, either
of \code{"fischer-simmons"} or \code{"francois-garrison"}; see \dQuote{References}.}
}
\value{
Sound absorption in dB/m.
}
\description{
Compute the sound absorption of seawater, in dB/m
}
\details{
Salinity and pH are ignored in this formulation.  Several formulae may be
found in the literature, and they give results differing by 10 percent, as
shown in reference 3 for example.  For this reason, it is likely that more
formulations will be added to this function, and entirely possible that the
default may change.
}
\examples{
## Fisher & Simmons (1977 table IV) gives 0.52 dB/km for 35 PSU, 5 degC, 500 atm
## (4990 dbar of water)a and 10 kHz
alpha <- swSoundAbsorption(35, 4, 4990, 10e3)

## reproduce part of Fig 8 of Francois and Garrison (1982 Fig 8)
f <- 1e3 * 10^(seq(-1,3,0.1)) # in KHz
plot(f/1000, 1e3*swSoundAbsorption(f, 35, 10, 0, formulation='fr'),
     xlab=" Freq [kHz]", ylab=" dB/km", type='l', log='xy')
lines(f/1000, 1e3*swSoundAbsorption(f, 0, 10, 0, formulation='fr'), lty='dashed')
legend("topleft", lty=c("solid", "dashed"), legend=c("S=35", "S=0"))

}
\references{
\enumerate{
\item F. H. Fisher and V. P. Simmons, 1977.  Sound absorption in
sea water.  Journal of the Acoustical Society of America, 62(3), 558-564.
\item R. E. Francois and G. R. Garrison, 1982.  Sound absorption based on
ocean measurements.  Part II: Boric acid contribution and equation for total
absorption.  Journal of the Acoustical Society of America, 72(6):1879-1890.
\item \url{http://resource.npl.co.uk/acoustics/techguides/seaabsorption/}
}
}
\seealso{
Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.R
\docType{data}
\name{adp}
\alias{adp}
\title{Sample adp (acoustic-doppler profiler) dataset}
\source{
This file came from the SLEIWEX-2008 experiment.
}
\usage{
data(adp)
}
\description{
This is degraded subsample of measurements that were made with an
upward-pointing, moored, ADP manufactured by Teledyne-RDI, as part of the St Lawrence
Internal Wave Experiment (SLEIWEX).
}
\examples{
\donttest{
library(oce)
data(adp)

# Velocity components.  (Note: we should probably trim some bins at top.)
plot(adp)

# Note that tides have moved the mooring.
plot(adp, which=15:18)
}


}
\seealso{
Other datasets provided with oce: 
\code{\link{adv}},
\code{\link{amsr}},
\code{\link{argo}},
\code{\link{cm}},
\code{\link{coastlineWorld}},
\code{\link{ctdRaw}},
\code{\link{ctd}},
\code{\link{echosounder}},
\code{\link{landsat}},
\code{\link{lisst}},
\code{\link{lobo}},
\code{\link{met}},
\code{\link{ocecolors}},
\code{\link{rsk}},
\code{\link{sealevelTuktoyaktuk}},
\code{\link{sealevel}},
\code{\link{section}},
\code{\link{topoWorld}},
\code{\link{wind}},
\code{\link{xbt}}

Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp_rdi.000}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\concept{datasets provided with oce}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/amsr.R
\name{composite,amsr-method}
\alias{composite,amsr-method}
\title{Create a composite of amsr satellite data}
\usage{
\S4method{composite}{amsr}(object, ...)
}
\arguments{
\item{object}{An \linkS4class{amsr} object.}

\item{...}{Other amsr objects.}
}
\description{
Form averages for each item in the \code{data} slot of the supplied objects,
taking into account the bad-data codes.

objects that are supplied as arguments are averaged in a way that
makes sense for the object class, i.e. taking into account
the particular bad-data codes of that particular class.
}
\details{
If none of the objects has good
data at any particular pixel (i.e. particular latitude and longitude),
the resultant will have the bad-data code of the last item in the argument
list.
The metadata in the result are taken directly from the metadata of the
final argument, except that the filename is set to a comma-separated list
of the component filenames.
}
\seealso{
Other things related to amsr data: 
\code{\link{[[,amsr-method}},
\code{\link{[[<-,amsr-method}},
\code{\link{amsr-class}},
\code{\link{amsr}},
\code{\link{download.amsr}()},
\code{\link{plot,amsr-method}},
\code{\link{read.amsr}()},
\code{\link{subset,amsr-method}},
\code{\link{summary,amsr-method}}

Other functions that create composite objects: 
\code{\link{composite,list-method}},
\code{\link{composite}()}
}
\concept{functions that create composite objects}
\concept{things related to amsr data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/bremen.R
\docType{class}
\name{bremen-class}
\alias{bremen-class}
\title{Class to Store Bremen-formatted Data}
\description{
This class is for data stored in a format used at Bremen.
It is somewhat similar to the \linkS4class{odf}, in the sense
that it does not apply just to a particular instrument.
Although some functions are provided for dealing with these
data (see \dQuote{Details}), the most common action is to read the
data with \code{\link[=read.bremen]{read.bremen()}}, and then to coerce the object to
another storage class (e.g. using \code{\link[=as.ctd]{as.ctd()}} for CTD-style
data) so that specialized functions can be used thereafter.
}
\section{Slots}{

\describe{
\item{\code{data}}{As with all \code{oce} objects, the \code{data} slot
for \code{bremen} objects is a
\link{list} containing the main data for the object.
{}}

\item{\code{metadata}}{As with all \code{oce} objects, the \code{metadata} slot
for \code{bremen} objects is a \link{list} containing
information about the \code{data} or about the object itself.
{}}

\item{\code{processingLog}}{As with all \code{oce} objects, the \code{processingLog} slot
for \code{bremen} objects is a
\link{list} with entries describing the creation and evolution
of the object. The contents are updated by various \code{oce} functions to
keep a record of processing steps.  Object summaries and
\code{\link[=processingLogShow]{processingLogShow()}} both display the log.}
}}

\section{Modifying slot contents}{


Although the \code{[[<-} operator may permit modification of the contents
of \linkS4class{bremen} objects (see \code{\link{[[<-,bremen-method}}),
it is better to use \code{\link[=oceSetData]{oceSetData()}} and \code{\link[=oceSetMetadata]{oceSetMetadata()}},
because those functions save an entry in the \code{processingLog}
that describes the change.
}

\section{Retrieving slot contents}{


The full contents of the \code{data} and \code{metadata} slots of a \linkS4class{bremen}
object may be retrieved in the standard R way using \code{\link[=slot]{slot()}}. For
example \code{slot(o,"data")} returns the \code{data} slot of an object named \code{o},
and similarly \code{slot(o,"metadata")} returns
the \code{metadata} slot.

The slots may also be obtained with the \code{\link{[[,bremen-method}}
operator, as e.g. \code{o[["data"]]} and \code{o[["metadata"]]}, respectively.

The \code{\link{[[,bremen-method}} operator can also
be used to retrieve items from within the \code{data} and \code{metadata} slots.
For example, \code{o[["temperature"]]} can be used to retrieve temperature
from an object containing that quantity. The rule is that a named
quantity is sought first within the object's \code{metadata} slot,
with the \code{data} slot being checked only if \code{metadata} does not
contain the item. This \code{[[} method can also be used to get
certain derived quantities, if the object contains sufficient
information to calculate them. For example, an object that holds
(practical) salinity, temperature and pressure, along with
longitude and latitude, has sufficient information to compute
Absolute Salinity, and so \code{o[["SA"]]} will yield the
calculated Absolute Salinity.

It is also possible to find items more directly, using \code{\link[=oceGetData]{oceGetData()}} and
\code{\link[=oceGetMetadata]{oceGetMetadata()}}, but neither of these functions can
retrieve derived items.
}

\seealso{
Other classes provided by oce: 
\code{\link{adp-class}},
\code{\link{adv-class}},
\code{\link{argo-class}},
\code{\link{cm-class}},
\code{\link{coastline-class}},
\code{\link{ctd-class}},
\code{\link{lisst-class}},
\code{\link{lobo-class}},
\code{\link{met-class}},
\code{\link{oce-class}},
\code{\link{odf-class}},
\code{\link{rsk-class}},
\code{\link{sealevel-class}},
\code{\link{section-class}},
\code{\link{topo-class}},
\code{\link{windrose-class}},
\code{\link{xbt-class}}

Other things related to bremen data: 
\code{\link{[[,bremen-method}},
\code{\link{[[<-,bremen-method}},
\code{\link{plot,bremen-method}},
\code{\link{read.bremen}()},
\code{\link{summary,bremen-method}}
}
\author{
Dan Kelley
}
\concept{classes provided by oce}
\concept{things related to bremen data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/extdata.R
\docType{data}
\encoding{UTF-8}
\name{adp_rdi.000}
\alias{adp_rdi.000}
\title{Sample adp (acoustic-doppler profiler) file in RDI format}
\description{
Sample adp (acoustic-doppler profiler) file in RDI format
}
\examples{
\dontrun{
read.oce(system.file("extdata", "adp_rdi.000", package="oce"))
}

}
\seealso{
Other raw datasets: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{ctd.cnv}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{test_met_csv1.csv}},
\code{\link{test_met_csv2.csv}},
\code{\link{test_met_xml2.xml}},
\code{\link{xbt.edf}}

Other things related to adp data: 
\code{\link{[[,adp-method}},
\code{\link{[[<-,adp-method}},
\code{\link{ad2cpHeaderValue}()},
\code{\link{adp-class}},
\code{\link{adpConvertRawToNumeric}()},
\code{\link{adpEnsembleAverage}()},
\code{\link{adpFlagPastBoundary}()},
\code{\link{adp}},
\code{\link{as.adp}()},
\code{\link{beamName}()},
\code{\link{beamToXyzAdpAD2CP}()},
\code{\link{beamToXyzAdp}()},
\code{\link{beamToXyzAdv}()},
\code{\link{beamToXyz}()},
\code{\link{beamUnspreadAdp}()},
\code{\link{binmapAdp}()},
\code{\link{enuToOtherAdp}()},
\code{\link{enuToOther}()},
\code{\link{handleFlags,adp-method}},
\code{\link{is.ad2cp}()},
\code{\link{plot,adp-method}},
\code{\link{read.adp.ad2cp}()},
\code{\link{read.adp.nortek}()},
\code{\link{read.adp.rdi}()},
\code{\link{read.adp.sontek.serial}()},
\code{\link{read.adp.sontek}()},
\code{\link{read.adp}()},
\code{\link{read.aquadoppHR}()},
\code{\link{read.aquadoppProfiler}()},
\code{\link{read.aquadopp}()},
\code{\link{rotateAboutZ}()},
\code{\link{setFlags,adp-method}},
\code{\link{subset,adp-method}},
\code{\link{subtractBottomVelocity}()},
\code{\link{summary,adp-method}},
\code{\link{toEnuAdp}()},
\code{\link{toEnu}()},
\code{\link{velocityStatistics}()},
\code{\link{xyzToEnuAdpAD2CP}()},
\code{\link{xyzToEnuAdp}()},
\code{\link{xyzToEnu}()}
}
\concept{raw datasets}
\concept{things related to adp data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/lisst.R
\name{plot,lisst-method}
\alias{plot,lisst-method}
\alias{plot.lisst}
\title{Plot a lisst Object}
\usage{
\S4method{plot}{lisst}(x, which = c(16, 37, 38), tformat, debug = getOption("oceDebug"), ...)
}
\arguments{
\item{x}{a \linkS4class{lisst} object.}

\item{which}{list of desired plot types.  These are graphed in panels
running down from the top of the page.  See \dQuote{Details} for the
meanings of various values of \code{which}.}

\item{tformat}{optional argument passed to \code{\link[=oce.plot.ts]{oce.plot.ts()}}, for
plot types that call that function.  (See \code{\link[=strptime]{strptime()}} for the
format used.)}

\item{debug}{a flag that turns on debugging.  The value indicates the depth
within the call stack to which debugging applies.}

\item{\dots}{optional arguments passed to plotting functions.}
}
\description{
Creates a multi-panel summary plot of data measured by LISST instrument.
}
\details{
The panels are controlled by the \code{which} argument, as follows.
\itemize{
\item \code{which=1} to \code{32}, or \code{which="C1"} to \code{"C32"} for
a time-series graph of the named column (a size class).
\item \code{which=33} or \code{which="lts"} for a time-series plot of laser
transmission sensor.
\item \code{which=34} or \code{which="voltage"} for a time-series plot of
instrument voltage.
\item \code{which=35} or \code{which="aux"} for a time-series plot of the
external auxiliary input.
\item \code{which=36} or \code{which="lrs"} for a time-series plot of the
laser reference sensor.
\item \code{which=37} or \code{which="pressure"} for a time-series plot of
pressure.
\item \code{which=38} or \code{which="temperature"} for a time-series plot
of temperature.
\item \code{which=41} or \code{which="transmission"} for a time-series plot
of transmission, in percent.
\item \code{which=42} or \code{which="beam"} for a time-series plot of
beam-C, in 1/metre.
}
}
\examples{
library(oce)
data(lisst)
plot(lisst)

}
\seealso{
The documentation for \linkS4class{lisst} explains the
structure of lisst objects, and also outlines the other functions dealing
with them.

Other functions that plot oce data: 
\code{\link{download.amsr}()},
\code{\link{plot,adp-method}},
\code{\link{plot,adv-method}},
\code{\link{plot,amsr-method}},
\code{\link{plot,argo-method}},
\code{\link{plot,bremen-method}},
\code{\link{plot,cm-method}},
\code{\link{plot,coastline-method}},
\code{\link{plot,ctd-method}},
\code{\link{plot,gps-method}},
\code{\link{plot,ladp-method}},
\code{\link{plot,landsat-method}},
\code{\link{plot,lobo-method}},
\code{\link{plot,met-method}},
\code{\link{plot,odf-method}},
\code{\link{plot,rsk-method}},
\code{\link{plot,satellite-method}},
\code{\link{plot,sealevel-method}},
\code{\link{plot,section-method}},
\code{\link{plot,tidem-method}},
\code{\link{plot,topo-method}},
\code{\link{plot,windrose-method}},
\code{\link{plot,xbt-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{tidem-class}}

Other things related to lisst data: 
\code{\link{[[,lisst-method}},
\code{\link{[[<-,lisst-method}},
\code{\link{as.lisst}()},
\code{\link{lisst-class}},
\code{\link{read.lisst}()},
\code{\link{summary,lisst-method}}
}
\author{
Dan Kelley
}
\concept{functions that plot oce data}
\concept{things related to lisst data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/xbt.R
\name{read.xbt}
\alias{read.xbt}
\title{Read an xbt file}
\usage{
read.xbt(
  file,
  type = "sippican",
  longitude = NA,
  latitude = NA,
  debug = getOption("oceDebug"),
  processingLog
)
}
\arguments{
\item{file}{a connection or a character string giving the name of the file to
load.}

\item{type}{character string indicating type of file, with valid choices being
\code{"sippican"} and \code{"noaa1"}.}

\item{longitude, latitude}{optional signed numbers indicating the longitude in degrees
East and latitude in degrees North. These values are used if \code{type="sippican"},
but ignored if \code{type="noaa1"}, because those files contain location information.}

\item{debug}{a flag that turns on debugging.  The value indicates the depth
within the call stack to which debugging applies.}

\item{processingLog}{if provided, the action item to be stored in the log.  This
parameter is typically only provided for internal calls; the default that it
provides is better for normal calls by a user.}
}
\value{
An \linkS4class{xbt} object.
}
\description{
Two file types are handled: (1) the \code{"sippican"} format, used
for Sippican XBT files, handled with \code{\link[=read.xbt.edf]{read.xbt.edf()}}, and (2)
the \code{"noaa1"} format, handled with \code{\link[=read.xbt.noaa1]{read.xbt.noaa1()}}. The first of
these is recognized by \code{\link[=read.oce]{read.oce()}}, but the second must be called
directly with \code{\link[=read.xbt.noaa1]{read.xbt.noaa1()}}.
}
\examples{
library(oce)
xbt <- read.oce(system.file("extdata", "xbt.edf", package="oce"))
summary(xbt)
plot(xbt)

}
\references{
\enumerate{
\item Sippican, Inc. "Bathythermograph Data Acquisition System: Installation, Operation and Maintenance
Manual (P/N 308195, Rev. A)," 2003.
https://pages.uoregon.edu/drt/MGL0910_Science_Report/attachments/MK21_ISA_Manual_Rev_A.pdf.
}
}
\seealso{
Other things related to xbt data: 
\code{\link{[[,xbt-method}},
\code{\link{[[<-,xbt-method}},
\code{\link{as.xbt}()},
\code{\link{plot,xbt-method}},
\code{\link{read.xbt.noaa1}()},
\code{\link{subset,xbt-method}},
\code{\link{summary,xbt-method}},
\code{\link{xbt-class}},
\code{\link{xbt.edf}},
\code{\link{xbt}}
}
\author{
Dan Kelley
}
\concept{things related to xbt data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swViscosity}
\alias{swViscosity}
\title{Seawater viscosity}
\usage{
swViscosity(salinity, temperature)
}
\arguments{
\item{salinity}{either salinity (PSU) (in which case \code{temperature} and
\code{pressure} must be provided) \emph{or} a \code{ctd} object (in which
case \code{salinity}, \code{temperature} and \code{pressure} are determined
from the object, and must not be provided in the argument list).}

\item{temperature}{\emph{in-situ} temperature (\eqn{^\circ}{deg}C), defined
on the ITS-90 scale; see \dQuote{Temperature units} in the documentation for
\code{\link[=swRho]{swRho()}}, and the examples below.}
}
\value{
Viscosity of seawater in \eqn{Pa\cdot s}{Pa*s}.  Divide by density
to get kinematic viscosity in \eqn{m^2/s}{m^2/s}.
}
\description{
Compute viscosity of seawater, in \eqn{Pa\cdot s}{Pa*s}
}
\details{
If the first argument is a \code{ctd} object, then salinity, temperature and
pressure values are extracted from it, and used for the calculation.

The result is determined from a regression of the data provided in Table 87
of Dorsey (1940). The fit matches the table to within 0.2 percent at worst,
and with average absolute error of 0.07 percent. The maximum deviation from
the table is one unit in the last decimal place.

No pressure dependence was reported by Dorsey (1940).
}
\examples{
swViscosity(30, 10) # 0.001383779

}
\references{
N. Ernest Dorsey (1940),
\emph{Properties of ordinary Water-substance},
American Chemical Society Monograph Series.  Reinhold
Publishing.
}
\seealso{
Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{read.netcdf}
\alias{read.netcdf}
\title{Read a NetCDF File}
\usage{
read.netcdf(file, ...)
}
\arguments{
\item{file}{the name of a file}

\item{...}{unused}
}
\value{
An \linkS4class{oce} object.
}
\description{
Read a netcdf file, trying to interpret its contents sensibly.
}
\details{
It is important to note that this is a preliminary version of
this function, and much about it may change without notice.
Indeed, it may be removed entirely.

Below are some features that may be changed.
\enumerate{
\item The names of data items are not changed from those in the netcdf
file on the assumption that this will offer the least surprise to
the user.
\item An attempt is made to find some common metadata from global
attributes in the netcdf file. These attributes include
\code{Longitude}, \code{Latitude}, \code{Ship} and \code{Cruise}.
Before they are stored in the metadata, they are converted to
lower case, since that is the oce convention.
}
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{applyMagneticDeclination}
\alias{applyMagneticDeclination}
\title{Earth magnetic declination}
\usage{
applyMagneticDeclination(x, declination = 0, debug = getOption("oceDebug"))
}
\arguments{
\item{x}{an \linkS4class{oce} object.}

\item{declination}{magnetic declination (to be added to the heading)}

\item{debug}{a debugging flag, set to a positive value to get debugging.}
}
\value{
Object, with velocity components adjusted to be aligned with
geographic north and east.
}
\description{
Instruments that use magnetic compasses to determine current direction need
to have corrections applied for magnetic declination, to get currents with
the y component oriented to geographic, not magnetic, north.  Sometimes, and
for some instruments, the declination is specified when the instrument is
set up, so that the velocities as recorded are already.  Other times, the
data need to be adjusted.  This function is for the latter case.
}
\references{
\enumerate{
\item \samp{https://www.ngdc.noaa.gov/IAGA/vmod/igrf.html}
}
}
\seealso{
Use \code{\link[=magneticField]{magneticField()}} to determine the declination,
inclination and intensity at a given spot on the world, at a given time.

Other things related to magnetism: 
\code{\link{magneticField}()}
}
\author{
Dan Kelley
}
\concept{things related to magnetism}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{head.oce}
\alias{head.oce}
\title{Extract The Start of an Oce Object}
\usage{
\method{head}{oce}(x, n = 6L, ...)
}
\arguments{
\item{x}{an \linkS4class{oce} object.}

\item{n}{Number of elements to extract, as for \code{\link[=head]{head()}}.}

\item{...}{ignored}
}
\description{
Extract The Start of an Oce Object

This function handles the following object classes directly:
\linkS4class{adp},
\linkS4class{adv},
\linkS4class{argo} (selection by profile),
\linkS4class{coastline},
\linkS4class{ctd},
\linkS4class{echosounder} (selection by ping),
\linkS4class{section} (selection by station)
and
\linkS4class{topo} (selection by longitude and latitude).
It does not handle
\linkS4class{amsr} or
\linkS4class{landsat}
yet, instead issuing a warning and returning \code{x} in those cases.
For all other classes, it calls \code{\link[=head]{head()}}
with \code{n} as provided, for each item in the \code{data}
slot, issuing a warning if that item is not a vector; the
author is quite aware that this may not work well for all classes.
The plan is to handle all appropriate classes by July 2018.
Please contact the author if there is a class you need handled
before that date.
}
\seealso{
\code{\link[=tail.oce]{tail.oce()}}, which yields the end of an \code{oce} object.
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{unitFromString}
\alias{unitFromString}
\title{Decode units, from strings}
\usage{
unitFromString(unit, scale = NULL)
}
\arguments{
\item{unit}{a character value indicating the unit. These
are matched according to rules developed to work with actual
data files, and so the list is not by any means exhaustive.}

\item{scale}{a character value indicating the scale.  The default value
of \code{NULL} dictates that the scale is to be inferred from the unit. If
a non-\code{NULL} value is supplied, it will be used, even if it makes no sense
in relation to value of \code{unit}.}
}
\value{
A \code{\link[=list]{list()}} of two items: \code{unit} which is an
\code{\link[=expression]{expression()}}, and \code{scale}, which is a string.
}
\description{
This is mainly intended for internal use within the package, e.g. by
\code{\link[=read.odf]{read.odf()}}, and so the list of string-to-unit mappings is not
documented, since developers can learn it from simple examination
of the code.  The focus of \code{unitFromString()} is on strings that are
found in oceanographic files available to the author, \emph{not} on all
possible units.
}
\examples{
unitFromString("dbar")   # dbar (no scale)
unitFromString("deg c")  # modern temperature (ITS-90 scale)
}
\seealso{
Other functions that interpret variable names and units from headers: 
\code{\link{ODFNames2oceNames}()},
\code{\link{cnvName2oceName}()},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{unitFromStringRsk}()},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()}
}
\concept{functions that interpret variable names and units from headers}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/oce.R
\name{oceColorsOxygen}
\alias{oceColorsOxygen}
\alias{oce.colorsOxygen}
\title{Create colors suitable for oxygen fields}
\usage{
oceColorsOxygen(n)
}
\arguments{
\item{n}{number of colors to create.}
}
\value{
A vector of color specifications.
}
\description{
Create a set of colors for displaying oxygen values,
based on the scheme devised by Thyng et al. (2016) and
presented in a python package by Thyng (2019).
The color specifications were transliterated from
python to R on 2015-09-29, but have not been adjusted
since, even though the python source has changed.  This
is to prevent breaking old \code{oce} code.  To get the
latest versions of these colours or other colours,
use the \CRANpkg{cmocean} R package
(Thyng, Richards, and Krylov, 2019) directly, as is
illustrated (with the "matter" scheme) in Example 2.
Note that the \CRANpkg{cmocean} core functions provide a way
to select between various versions of the colour schemes.
It is also worth considering the palettes provided by the
\CRANpkg{viridis} package, as illustrated (with the
"inferno" scheme) in Example 3.
}
\examples{
library(oce)

# Example 1
imagep(volcano, col=oceColorsOxygen(128),
       zlab="oceColorsOxygen")
\dontrun{
# Example 2 (requires the cmocean package)
imagep(volcano, col=cmocean::cmocean("matter"),
       zlab='cmocean::cmocean("matter")')}

\dontrun{
# Example 3 (requires the viridis package)
imagep(volcano, col=viridis::inferno,
       zlab='viridis::inferno')}

}
\references{
\itemize{
\item Thyng, Kristen, Chad Greene, Robert Hetland, Heather Zimmerle, and Steven DiMarco.
\dQuote{True Colors of Oceanography: Guidelines for Effective and Accurate Colormap Selection.}
Oceanography 29, no. 3 (September 1, 2016): 9‚Äì13.
\doi{10.5670/oceanog.2016.66}
\item Thyng, Kristen. Kthyng/Cmocean. Python, 2019.
\url{https://github.com/kthyng/cmocean}.
\item Thyng, Kristen, Clark Richards, and Ivan Krylov.
Cmocean: Beautiful Colour Maps for Oceanography (version 0.2), 2019.
\url{https://CRAN.R-project.org/package=cmocean}.
}

The following references provide information on choosing
colour schemes, that are suitable for viewers who have
colour deficiencies.

Light, Adam, and Patrick J. Bartlein. "The End of the Rainbow? Color
Schemes for Improved Data Graphics." \emph{Eos, Transactions American Geophysical
Union} 85, no. 40 (2004): 385.
DOI: 10.1029/2004EO400002

Stephenson, David B. "Comment on 'Color Schemes for Improved Data
Graphics', by A Light and P.J. Bartlein." \emph{Eos, Transactions American
Geophysical Union} 86, no.  20 (2005): 196.
DOI: 10.1029/2005EO200005

Light, Adam, and Patrick J. Bartlein. "Reply to 'Comment on Color
Schemes for Improved Data Graphics,' by A. Light and P.J.  Bartlein'."
\emph{Eos, Transactions American Geophysical Union} 86, no. 20 (2005): 196‚Äì196.
DOI: 10.1029/2005EO200006
}
\seealso{
Other things related to colors: 
\code{\link{colormapGMT}()},
\code{\link{colormap}()},
\code{\link{oceColors9B}()},
\code{\link{oceColorsCDOM}()},
\code{\link{oceColorsChlorophyll}()},
\code{\link{oceColorsClosure}()},
\code{\link{oceColorsDensity}()},
\code{\link{oceColorsFreesurface}()},
\code{\link{oceColorsGebco}()},
\code{\link{oceColorsJet}()},
\code{\link{oceColorsPAR}()},
\code{\link{oceColorsPalette}()},
\code{\link{oceColorsPhase}()},
\code{\link{oceColorsSalinity}()},
\code{\link{oceColorsTemperature}()},
\code{\link{oceColorsTurbidity}()},
\code{\link{oceColorsTurbo}()},
\code{\link{oceColorsTwo}()},
\code{\link{oceColorsVelocity}()},
\code{\link{oceColorsViridis}()},
\code{\link{oceColorsVorticity}()},
\code{\link{ocecolors}}
}
\author{
Krysten M. Thyng (Python version), Dan Kelley (R transliteration)
}
\concept{things related to colors}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/section.R
\name{as.section}
\alias{as.section}
\title{Create a Section}
\usage{
as.section(
  salinity,
  temperature,
  pressure,
  longitude,
  latitude,
  station,
  sectionId = "",
  debug = getOption("oceDebug")
)
}
\arguments{
\item{salinity}{This may be a numerical vector, in which case it is interpreted
as the salinity, and the other arguments are used for the other components of
\linkS4class{ctd} objects. Alternatively, it may be one of a variety of
other objects from which the CTD objects can be inferred, in which case the
other arguments are ignored; see \sQuote{Details}.}

\item{temperature}{Temperature, in a vector holding values for all stations.}

\item{pressure}{Pressure, in a vector holding values for all stations.}

\item{longitude}{Longitude, in a vector holding values for all stations.}

\item{latitude}{Latitude, in a vector holding values for all stations.}

\item{station}{Station identifiers, in a vector holding values for all stations.}

\item{sectionId}{Section identifier.}

\item{debug}{an integer value that controls whether \code{as.section()} prints information
during its work.  The function works quietly if this is 0 and prints out some
information if it is positive.}
}
\value{
An object of \linkS4class{section}.
}
\description{
Create a section based on columnar data, or a set of \linkS4class{oce}
objects that can be coerced to a section. There are three cases.
}
\details{
Case 1. If the first argument is a numerical vector, then it is taken to be the
salinity, and \code{\link[=factor]{factor()}} is applied to \code{station} to break the
data up into chunks that are assembled into \linkS4class{ctd} objects with
\code{\link[=as.ctd]{as.ctd()}} and combined to make a \linkS4class{section} object
to be returned. This mode of operation is provided as a convenience for datasets
that are already partly processed; if original CTD data are available, the next
mode is preferred, because it permits the storage of much more data and metadata
in the CTD object.

Case 2. If the first argument is a list containing oce objects, then those
objects are taken as profiles of something.  A requirement for this
to work is that every element of the list contains both \code{longitude}
and \code{latitude} in either the \code{metadata} or \code{data} slot (in
the latter case, the mean value is recorded in the section object)
and that every element also contains \code{pressure} in its \code{data} slot.

Case 3. If the first argument is a \linkS4class{argo} object, then
the profiles it contains are turned into \linkS4class{ctd} objects,
and these are assembled into a section to be returned.
}
\examples{
library(oce)
data(ctd)
## vector of names of CTD objects
fake <- ctd
fake[["temperature"]] <- ctd[["temperature"]] + 0.5
fake[["salinity"]] <- ctd[["salinity"]] + 0.1
fake[["longitude"]] <- ctd[["longitude"]] + 0.01
fake[["station"]] <- "fake"
sec1 <- as.section(c("ctd", "fake"))
summary(sec1)
## vector of CTD objects
ctds <- vector("list", 2)
ctds[[1]] <- ctd
ctds[[2]] <- fake
sec2 <- as.section(ctds)
summary(sec2)
## argo data (a subset)
data(argo)
sec3 <- as.section(subset(argo, profile<5))
summary(sec3)

}
\seealso{
Other things related to section data: 
\code{\link{[[,section-method}},
\code{\link{[[<-,section-method}},
\code{\link{handleFlags,section-method}},
\code{\link{initializeFlagScheme,section-method}},
\code{\link{plot,section-method}},
\code{\link{read.section}()},
\code{\link{section-class}},
\code{\link{sectionAddStation}()},
\code{\link{sectionGrid}()},
\code{\link{sectionSmooth}()},
\code{\link{sectionSort}()},
\code{\link{section}},
\code{\link{subset,section-method}},
\code{\link{summary,section-method}}
}
\author{
Dan Kelley
}
\concept{things related to section data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/argo.R
\name{read.argo}
\alias{read.argo}
\title{Read an Argo Data File}
\usage{
read.argo(file, debug = getOption("oceDebug"), processingLog, ...)
}
\arguments{
\item{file}{A character string giving the name of the file to load.}

\item{debug}{A flag that turns on debugging.  Set to 1 to get a moderate amount
of debugging information, or to 2 to get more.}

\item{processingLog}{If provided, the action item to be stored in the log.
(Typically only provided for internal calls; the default that it provides is
better for normal calls by a user.)}

\item{...}{additional arguments, passed to called routines.}
}
\value{
An \linkS4class{argo} object.
}
\description{
\code{read.argo} is used to read an Argo file, producing an \linkS4class{argo} object.
The file must be in the ARGO-style NetCDF format described
in the Argo documentation (see references 2 and 3).
}
\details{
See the Argo documentation (see references 2 and 3) for some details on what files contain.
Many items listed in section 2.2.3 of reference 3 are read from the
file and stored in the \code{metadata} slot, with the exception of
\code{longitude} and \code{latitude}, which are stored in the
\code{data} slot, alongside hydrographic information.

The names of several data parameters stored within the netCDF file
are altered to fit the oce context. For example, \code{PRES} becomes \code{pressure},
matching the name of this variable in other oce data types.
The original names are reported by \verb{summary,argo-method}, and
data may be extracted with \verb{[[,argo-method} using those names, so
the renaming should not be too inconvenient to Argo experts who
are new to oce.

Several of the netCDF global attributes are also renamed before
placement in the \code{metadata} slot of the return value.  These include
\code{conventions}, \code{featureType}, \code{history}, \code{institution},
\code{nParameters}, \code{nProfiles},  \code{references}, \code{source}, \code{title},
and \code{userManualVersion}.
These names are derived from those in the netcdf
file, and mainly follow the pattern explained in the
\dQuote{Variable renaming convention} section.

For profile data (as indicated by the NetCDF global attribute
named \code{"featureType"} being equal to \code{"trajectoryProfile"}),
the NetCDF item named \code{"STATION_PARAMETERS"} controls
whether variables in the source file will be stored in the
\code{metadata} or \code{data} slot of the returned object.
If \code{STATION_PARAMETERS} is not present, as is the case for
trajectory files (which are detected by \code{featureType} being
\code{"trajectory"}), some guesses are made as to what goes in
\code{data} and \code{metadata} slots.

Each data item can have variants, as
described in Sections 2.3.4 of reference 3.
For example, if \code{"PRES"} is found in \code{STATION_PARAMETERS},
then \code{PRES} (pressure) data are sought in the file, along with
\code{PRES_QC}, \code{PRES_ADJUSTED}, \code{PRES_ADJUSTED_QC}, and
\code{PRES_ERROR}. The same pattern works for other profile data. The variables
are stored with names created as explained in the
\dQuote{Variable renaming convention} section below. Note that
flags, which are stored variables ending in \code{"_QC"} in the netcdf
file, are stored in the \code{flags} item within the \code{metadata} slot
of the returned object; thus, for example,
\code{PRES_QC} is stored as \code{pressure} in \code{flags}.
}
\section{Variable renaming convention}{

Argo netcdf files employ a \code{"SNAKE_CASE"} naming scheme (sometimes
using lower case) that is inconsistent with the \code{"camelCase"} scheme
used in oce. Since argo objects are just a small part of oce, a decision
was made to rename argo items. For example, \code{"CYCLE_NUMBER"} in the netcdf file
becomes \code{"cycleNumber"} in the oce object returned by \code{read.argo}.
(Note that \verb{[[,argo-method} also accepts \code{"cycle"} for this item.)
The conversion for objects in the \code{data} slot often also involves
expanding on argo abbreviations, e.g. \code{"PSAL"} becomes \code{"salinity"}.
The renaming work is carried out with \code{\link[=argoNames2oceNames]{argoNames2oceNames()}} for
handles both name expansion for several dozen special cases,
and with \code{\link[=snakeToCamel]{snakeToCamel()}} with the \code{specialCase} argument
set to \code{"QC"}. While this results in variable names that should make
sense in the general oce context (where, for example, salinity is expected
to be stored in a variable named \code{"salinity"}), it may be confusing
to argo experts who are just starting to use oce.  Such people might
find it helpful to use e.g. \code{sort(names(x[["metadata"]]))} to get a list
of all items in the \code{metadata} slot (or similar with \code{"data"}), since working
in reverse may be easier than simply guessing at what names oce has chosen.
(Note that prior to 2020 June 24, some metadata items were stored in
\code{"SNAKE_CASE"}.)
}

\section{Data sources}{

Argo data are made available at several websites. A bit of detective
work can be required to track down the data.

Some servers provide data for floats that surfaced in a given ocean
on a given day, the anonymous FTP server
\code{usgodae.org/pub/outgoing/argo/geo/} being an example.

Other servers provide data on a per-float basis. A complicating
factor is that these data tend to be categorized by "dac" (data
archiving centre), which makes it difficult to find a particular
float. For example,
\code{https://www.usgodae.org/ftp/outgoing/argo/} is the top level of
a such a repository. If the ID of a float is known but not the
"dac", then a first step is to download the text file
\code{https://www.usgodae.org/ftp/outgoing/argo/ar_index_global_meta.txt}
and search for the ID. The first few lines of that file are header,
and after that the format is simple, with columns separated by slash
(\code{/}). The dac is in the first such column and the float ID in the
second. A simple search will reveal the dac.
For example \code{data(argo)} is based on float 6900388, and the line
containing that token is
\verb{bodc/6900388/6900388_meta.nc,846,BO,20120225005617}, from
which the dac is seen to be the British Oceanographic Data Centre
(\code{bodc}). Armed with that information, visit
\code{https://www.usgodae.org/ftp/outgoing/argo/dac/bodc/6900388}
and see a directory called \code{profiles} that contains a NetCDF
file for each profile the float made. These can be read with
\code{read.argo}. It is also possible, and probably more common,
to read a NetCDF file containing all the profiles together and for
that purpose the file
\code{https://www.usgodae.org/ftp/outgoing/argo/dac/bodc/6900388/6900388_prof.nc}
should be downloaded and provided as the \code{file} argument to
\code{read.argo}.  This can be automated as in Example 2,
although readers are cautioned that URL structures tend to change
over time.

Similar steps can be followed on other servers.
}

\examples{
\dontrun{
## Example 1: read from a local file
library(oce)
d <- read.argo("/data/OAR/6900388_prof.nc")
summary(d)
plot(d)

## Example 2: construct URL for download (brittle)
id <- "6900388"
url <- "https://www.usgodae.org/ftp/outgoing/argo"
if (!length(list.files(pattern="argo_index.txt")))
    download.file(paste(url, "ar_index_global_meta.txt", sep="/"), "argo_index.txt")
index <- readLines("argo_index.txt")
line <- grep(id, index)
if (0 == length(line)) stop("id ", id, " not found")
if (1 < length(line)) stop("id ", id, " found multiple times")
dac <- strsplit(index[line], "/")[[1]][1]
profile <- paste(id, "_prof.nc", sep="")
float <- paste(url, "dac", dac, id, profile, sep="/")
download.file(float, profile)
argo <- read.argo(profile)
summary(argo)
}


}
\references{
\enumerate{
\item \url{https://argo.ucsd.edu}
\item Argo User's Manual Version 3.2, Dec 29th, 2015, available at
\url{https://archimer.ifremer.fr/doc/00187/29825/} online.
\item User's Manual (ar-um-02-01) 13 July 2010, available at
\url{http://www.argodatamgt.org/content/download/4729/34634/file/argo-dm-user-manual-version-2.3.pdf},
which is the main document describing argo data.
}
}
\seealso{
The documentation for the \linkS4class{argo} class explains the structure of argo
objects, and also outlines the other functions dealing with them.

Other things related to argo data: 
\code{\link{[[,argo-method}},
\code{\link{[[<-,argo-method}},
\code{\link{argo-class}},
\code{\link{argoGrid}()},
\code{\link{argoNames2oceNames}()},
\code{\link{argo}},
\code{\link{as.argo}()},
\code{\link{handleFlags,argo-method}},
\code{\link{plot,argo-method}},
\code{\link{subset,argo-method}},
\code{\link{summary,argo-method}}
}
\author{
Dan Kelley
}
\concept{things related to argo data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctd.R
\name{ctdRepair}
\alias{ctdRepair}
\title{Repair a malformed ctd object}
\usage{
ctdRepair(x, debug = getOption("oceDebug"))
}
\arguments{
\item{x}{a \linkS4class{ctd} object.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\value{
A \linkS4class{ctd} object that is based on \code{x}, but possibly
with some elements changed as described in the \dQuote{Details}
section.
}
\description{
Make a \linkS4class{ctd} object adhere more closely with the expected form, e.g. by
moving certain things from the \code{data} slot to the \code{metadata} slot, where
other oce functions may assume they will be located.
This can be handy for objects that were set up
incorrectly, perhaps by inappropriate user insertions.
}
\details{
The possible changes fall into the following categories.
\enumerate{
\item If unit-length values for \code{latitude}, \code{longitude}, \code{time}, or \code{station}
exist in the \code{data} slot, move them to the \code{metadata} slot.  However,
leave them in \code{data}  if their length exceeds 1, because this can
arise with towyo data.
\item If the \code{metadata} or \code{data} slot contains items named
\code{time}, \code{recoveryTime}, \code{startTime}, or \code{systemUploadTime},
and if these are not in POSIXt format, then use \code{\link[=as.POSIXct]{as.POSIXct()}} with
\code{tz="UTC"} to convert them to POSIXt format. If that conversion fails,
owing to an unrecognizable format, then the original value
is retained, unaltered.
}
}
\examples{
library(oce)
data(ctd)
# Insert location information into 'data', although it belongs in 'metadata'.
ctd@data$latitude <- ctd@metadata$latitude   # do NOT do this!
ctd@data$longitude <- ctd@metadata$longitude # do NOT do this!
repaired <- ctdRepair(ctd)

}
\seealso{
Other things related to ctd data: 
\code{\link{CTD_BCD2014666_008_1_DN.ODF.gz}},
\code{\link{[[,ctd-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{as.ctd}()},
\code{\link{cnvName2oceName}()},
\code{\link{ctd-class}},
\code{\link{ctd.cnv}},
\code{\link{ctdDecimate}()},
\code{\link{ctdFindProfiles}()},
\code{\link{ctdRaw}},
\code{\link{ctdTrim}()},
\code{\link{ctd}},
\code{\link{d200321-001.ctd}},
\code{\link{d201211_0011.cnv}},
\code{\link{handleFlags,ctd-method}},
\code{\link{initialize,ctd-method}},
\code{\link{initializeFlagScheme,ctd-method}},
\code{\link{oceNames2whpNames}()},
\code{\link{oceUnits2whpUnits}()},
\code{\link{plot,ctd-method}},
\code{\link{plotProfile}()},
\code{\link{plotScan}()},
\code{\link{plotTS}()},
\code{\link{read.ctd.itp}()},
\code{\link{read.ctd.odf}()},
\code{\link{read.ctd.odv}()},
\code{\link{read.ctd.sbe}()},
\code{\link{read.ctd.woce.other}()},
\code{\link{read.ctd.woce}()},
\code{\link{read.ctd}()},
\code{\link{setFlags,ctd-method}},
\code{\link{subset,ctd-method}},
\code{\link{summary,ctd-method}},
\code{\link{woceNames2oceNames}()},
\code{\link{woceUnit2oceUnit}()},
\code{\link{write.ctd}()}
}
\author{
Dan Kelley
}
\concept{things related to ctd data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/gps.R
\name{summary,gps-method}
\alias{summary,gps-method}
\title{Summarize a GPS Object}
\usage{
\S4method{summary}{gps}(object, ...)
}
\arguments{
\item{object}{an object of class \code{"gps"}}

\item{\dots}{further arguments passed to or from other methods.}
}
\description{
Summarize a \linkS4class{gps} object.
}
\seealso{
Other things related to gps data: 
\code{\link{[[,gps-method}},
\code{\link{[[<-,gps-method}},
\code{\link{as.gps}()},
\code{\link{gps-class}},
\code{\link{plot,gps-method}},
\code{\link{read.gps}()}
}
\author{
Dan Kelley
}
\concept{things related to gps data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/gps.R
\name{read.gps}
\alias{read.gps}
\title{Read a GPS File}
\usage{
read.gps(file, type = NULL, debug = getOption("oceDebug"), processingLog)
}
\arguments{
\item{file}{name of file containing gps data.}

\item{type}{type of file, which will be inferred from examination of the
data if not supplied.  In the present version, the only choice for
\code{type} is \code{"gpx"}.}

\item{debug}{set to TRUE to print information about the header, etc.}

\item{processingLog}{if provided, the action item to be stored in the log.
(Typically only provided for internal calls; the default that it provides is
better for normal calls by a user.)}
}
\value{
A \linkS4class{gps} object.
}
\description{
Reads GPX format files by simply finding all longitudes and latitudes; in
other words, information on separate tracks, or waypoints, etc., is lost.
}
\seealso{
Other things related to gps data: 
\code{\link{[[,gps-method}},
\code{\link{[[<-,gps-method}},
\code{\link{as.gps}()},
\code{\link{gps-class}},
\code{\link{plot,gps-method}},
\code{\link{summary,gps-method}}
}
\author{
Dan Kelley
}
\concept{things related to gps data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/tides.R
\name{summary,tidem-method}
\alias{summary,tidem-method}
\title{Summarize a Tidem Object}
\usage{
\S4method{summary}{tidem}(object, p, constituent, ...)
}
\arguments{
\item{object}{an object of class \link{tidem}, as created by
\code{\link[=as.tidem]{as.tidem()}} or \code{\link[=tidem]{tidem()}}.}

\item{p}{optional value of the maximum p value for the display of an
individual coefficient.  If not given, all coefficients are shown.}

\item{constituent}{optional character vector holding the names
of constituents on which to focus.}

\item{\dots}{further arguments passed to or from other methods.}
}
\value{
\code{NULL}
}
\description{
By default, all fitted constituents are plotted, but it is quite useful to
set e.g. p=0.05 To see just those constituents that are significant at the 5
percent level.
Note that the p values are estimated as the average of the p values for the
sine and cosine components at a given frequency.
}
\examples{
\dontrun{
library(oce)
data(sealevel)
tide <- tidem(sealevel)
summary(tide)
}

}
\seealso{
Other things related to tides: 
\code{\link{[[,tidem-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{as.tidem}()},
\code{\link{plot,tidem-method}},
\code{\link{predict.tidem}()},
\code{\link{tidalCurrent}},
\code{\link{tidedata}},
\code{\link{tidem-class}},
\code{\link{tidemAstron}()},
\code{\link{tidemVuf}()},
\code{\link{tidem}},
\code{\link{webtide}()}
}
\author{
Dan Kelley
}
\concept{things related to tides}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{plotTaylor}
\alias{plotTaylor}
\title{Plot a Model-data Comparison Diagram}
\usage{
plotTaylor(x, y, scale, pch, col, labels, pos, ...)
}
\arguments{
\item{x}{a vector of reference values of some quantity, e.g. measured over
time or space.}

\item{y}{a matrix whose columns hold values of values to be compared with
those in x.  (If \code{y} is a vector, it is converted first to a one-column
matrix).}

\item{scale}{optional scale, interpreted as the maximum value of standard
deviation.}

\item{pch}{list of characters to plot, one for each column of \code{y}.}

\item{col}{list of colors for points on the plot, one for each column of
\code{y}.}

\item{labels}{optional vector of strings to use for labelling the points.}

\item{pos}{optional vector of positions for labelling strings.  If not
provided, labels will be to the left of the symbols.}

\item{\dots}{optional arguments passed by \code{plotTaylor} to more child
functions.}
}
\description{
Creates a diagram as described by Taylor (2001).  The graph is in the form
of a semicircle, with radial lines and spokes connecting at a focus point on
the flat (lower) edge.  The radius of a point on the graph indicates the
standard deviation of the corresponding quantity, i.e. x and the columns in
y.  The angle connecting a point on the graph to the focus provides an
indication of correlation coefficient with respect to x.  The ``east'' side
of the graph indicates \eqn{R=1}{R=1}, while \eqn{R=0}{R=0} is at the
"north" edge and \eqn{R=-1}{R=-1} is at the "west" side.  The \code{x}
data are indicated with a bullet on the graph, appearing on the lower edge
to the right of the focus at a distance indicating the standard deviation of
`x`.  The other points on the graph represent the columns of `y`,
coded automatically or with the supplied values of `pch` and
`col`.
The example shows two tidal models of the Halifax sealevel data, computed
with \code{\link[=tidem]{tidem()}} with just the M2 component and the S2 component;
the graph indicates that the M2 model is much better than the S2 model.
}
\examples{
library(oce)
data(sealevel)
x <- sealevel[["elevation"]]
M2 <- predict(tidem(sealevel, constituents="M2"))
S2 <- predict(tidem(sealevel, constituents=c("S2")))
plotTaylor(x, cbind(M2, S2))
}
\references{
Taylor, Karl E., 2001.  Summarizing multiple aspects of model
performance in a single diagram, \emph{J. Geophys. Res.}, 106:D7, 7183--7192.
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/imagep.R
\name{imagep}
\alias{imagep}
\title{Plot an Image with a Color Palette}
\usage{
imagep(
  x,
  y,
  z,
  xlim,
  ylim,
  zlim,
  zclip = FALSE,
  flipy = FALSE,
  xlab = "",
  ylab = "",
  zlab = "",
  zlabPosition = c("top", "side"),
  las.palette = 0,
  decimate = TRUE,
  breaks,
  col,
  colormap,
  labels = NULL,
  at = NULL,
  drawContours = FALSE,
  drawPalette = TRUE,
  drawTriangles = FALSE,
  tformat,
  drawTimeRange = getOption("oceDrawTimeRange"),
  filledContour = FALSE,
  missingColor = NULL,
  useRaster,
  mgp = getOption("oceMgp"),
  mar,
  mai.palette,
  xaxs = "i",
  yaxs = "i",
  asp = NA,
  cex = par("cex"),
  cex.axis = cex,
  cex.lab = cex,
  cex.main = cex,
  axes = TRUE,
  main = "",
  axisPalette,
  add = FALSE,
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{x, y}{These have different meanings in different modes of operation.
\emph{Mode 1.} One
mode has them meaning the locations of coordinates along which values
matrix \code{z} are defined. In this case, both \code{x} and \code{y}
must be supplied and, within each, the values must be finite and
distinct; if values are out of order, they (and \code{z}) will be
transformed to put them in order.
ordered in a matching way).  \emph{Mode 2.}
If \code{z} is provided but not \code{x} and \code{y}, then the latter
are constructed to
indicate the indices of the matrix, in contrast
to the range of 0 to 1, as is the case for \code{\link[=image]{image()}}.
\emph{Mode 3.} If
\code{x} is a list, its components \code{x$x} and \code{x$y} are used
for \code{x} and \code{y}, respectively. If the list has component
\code{z} this is used for \code{z}. (NOTE: these arguments are meant to
mimic those of \code{\link[=image]{image()}}, which explains the same description
here.)  \emph{Mode 4.} There are also some special cases, e.g. if \code{x} is a
topographic object such as can be created with \code{\link[=read.topo]{read.topo()}}
or \code{\link[=as.topo]{as.topo()}}, then longitude and latitude are used for
axes, and topographic height is drawn.}

\item{z}{A matrix containing the values to be plotted (NAs are allowed). Note
that x can be used instead of z for convenience. (NOTE: these arguments
are meant to mimic those of \code{\link[=image]{image()}}, which explains the same
description here.)}

\item{xlim, ylim}{Limits on x and y axes.}

\item{zlim}{If missing, the z scale is determined by the range of the data.
If provided, \code{zlim} may take several forms. First, it may be a pair
of numbers that specify the limits for the color scale.  Second,
it could be the string \code{"histogram"}, to yield a flattened
histogram (i.e. to increase contrast). Third, it could be the
string \code{"symmetric"}, to yield limits that are symmetric
about zero, which can be helpful in drawing velocity fields,
for which a zero value has a particular meaning (in which case,
a good color scheme might be \code{col=oceColorsTwo}).}

\item{zclip}{Logical, indicating whether to clip the colors to those
corresponding to \code{zlim}. This only works if \code{zlim} is
provided. Clipped regions will be colored with \code{missingColor}.
Thus, clipping an image is somewhat analogous to clipping in an xy
plot, with clipped data being ignored, which in an image means to be be
colored with \code{missingColor}.}

\item{flipy}{Logical, with \code{TRUE} indicating that the graph
should have the y axis reversed, i.e. with smaller values at
the bottom of the page. (\emph{Historical note:} until 2019 March 26,
the meaning of \code{flipy} was different; it meant to reverse the
range of the y axis, so that if \code{ylim} were given as a reversed
range, then setting \code{flipy=TRUE} would reverse the flip, yielding
a conventional axis with smaller values at the bottom.)}

\item{xlab, ylab, zlab}{Names for x axis, y axis, and the image values.}

\item{zlabPosition}{String indicating where to put the label for the z axis,
either at the top-right of the main image, or on the side, in the axis
for the palette.}

\item{las.palette}{Parameter controlling the orientation of labels on the
image palette, passed as the \code{las} argument to \code{\link[=drawPalette]{drawPalette()}}.  See the
documentation for \code{\link[=drawPalette]{drawPalette()}} for details.}

\item{decimate}{Controls whether the image will be decimated before plotting,
in three possible cases.
\enumerate{
\item If \code{decimate=FALSE} then every grid cell in the matrix will
be represented by a pixel in the image.
\item If \code{decimate=TRUE} (the default), then decimation will be done
in the horizontal or vertical direction (or both) if the length of the
corresponding edge of the \code{z} matrix exceeds 800. (This also creates
a warning message.) The decimation
factor is computed as the integer just below the ratio of \code{z} dimension
to 400. Thus, no decimation is done if the dimension is less than 800,
but if the dimension s between 800 and 1199, only every second grid
point is mapped to a pixel in the image.
\item If \code{decimate} is an integer, then that \code{z} is subsampled
at \code{seq.int(1L, dim(z)[1], by=decimate)} (as is \code{x}), and
the same is done for the \code{y} direction.
\item If \code{decimate} is a vector of two integers, the first is used for
the first index of \code{z}, and the second is used for the second
index.
}}

\item{breaks}{The z values for breaks in the color scheme.  If this is of
length 1, the value indicates the desired number of breaks, which is
supplied to \code{\link[=pretty]{pretty()}}, in determining clean break points.
If \code{colormap} is provided, it takes precedence over \code{breaks} and \code{col}.}

\item{col}{Either a vector of colors corresponding to the breaks, of length
1 plus the number of breaks, or a function specifying colors.
If \code{col} is not provided, and if \code{colormap} is also not provided, then \code{col}
defaults to \code{\link[=oceColorsViridis]{oceColorsViridis()}}.
If \code{colormap} is provided, it takes precedence over \code{breaks} and \code{col}.}

\item{colormap}{A color map as created by \code{\link[=colormap]{colormap()}}.  If
provided, then \code{colormap$breaks} and \code{colormap$col} take
precedence over the present arguments \code{breaks} and \code{col}.
(All of the other contents of \code{colormap} are ignored, though.)
If \code{colormap} is provided, it takes precedence over \code{breaks} and \code{col}.}

\item{labels}{Optional vector of labels for ticks on palette axis (must
correspond with \code{at}).}

\item{at}{Optional vector of positions for the \code{label}s.}

\item{drawContours}{Logical value indicating whether to draw contours on the
image, and palette, at the color breaks.  Images with a great deal of
high-wavenumber variation look poor with contours.}

\item{drawPalette}{Indication of the type of palette to draw, if any.  If
\code{drawPalette=TRUE}, a palette is drawn at the right-hand side of the
main image.  If \code{drawPalette=FALSE}, no palette is drawn, and the
right-hand side of the plot has a thin margin.  If
\code{drawPalette="space"}, then no palette is drawn, but space is put in
the right-hand margin to occupy the region in which the palette would
have been drawn.  This last form is useful for producing stacked plots
with uniform left and right margins, but with palettes on only some of
the images.}

\item{drawTriangles}{Logical value indicating whether to draw
triangles on the top and bottom of the palette.  This is passed to
\code{\link[=drawPalette]{drawPalette()}}.}

\item{tformat}{Optional argument passed to \code{\link[=oce.plot.ts]{oce.plot.ts()}}, for
plot types that call that function.  (See \code{\link[=strptime]{strptime()}} for the
format used.)}

\item{drawTimeRange}{Logical, only used if the \code{x} axis is a
time.  If \code{TRUE}, then an indication of the time range of the
data (not the axis) is indicated at the top-left margin of the
graph.  This is useful because the labels on time axes only indicate
hours if the range is less than a day, etc.}

\item{filledContour}{Boolean value indicating whether to use filled
contours to plot the image.}

\item{missingColor}{A color to be used to indicate missing data, or
\code{NULL} for transparent (to see this, try setting
\code{par("bg")<-"red"}).}

\item{useRaster}{A logical value passed to \code{\link[=image]{image()}}, in cases
where \code{filledContour} is \code{FALSE}. Setting \code{useRaster=TRUE}
can alleviate some anti-aliasing effects on some plot devices;
see the documentation for \code{\link[=image]{image()}}.}

\item{mgp}{A 3-element numerical vector to use for \code{par(mgp)}, and
also for \code{par(mar)}, computed from this.  The default is
tighter than the R default, in order to use more space for the
data and less for the axes.}

\item{mar}{A 4-element Value to be used with \code{\link{par}}\code{("mar")}.  If not
given, a reasonable value is calculated based on whether \code{xlab} and
\code{ylab} are empty strings.}

\item{mai.palette}{Palette margin corrections (in inches), added to the
\code{mai} value used for the palette.  Use with care.}

\item{xaxs}{Character indicating whether image should extend to edge
of x axis (with value \code{"i"}) or not; see
\code{\link{par}}\code{("xaxs")}.}

\item{yaxs}{As \code{xaxs} but for y axis.}

\item{asp}{Aspect ratio of the plot, as for \code{\link[=plot.default]{plot.default()}}. If
\code{x} inherits from \linkS4class{topo} and \code{asp=NA} (the
default) then \code{asp} is redefined to be the reciprocal of the
mean latitude in \code{x}, as a way to reduce geographical distortion.
Otherwise, if \code{asp} is not \code{NA}, then it is used directly.}

\item{cex}{numeric character expansion factor, used for \code{cex.axis}, \code{cex.lab} and \code{cex.main},
if those values are not supplied.}

\item{cex.axis, cex.lab, cex.main}{numeric character expansion factors for axis numbers,
axis names and plot titles; see \code{\link[=par]{par()}}.}

\item{axes}{Logical, set \code{TRUE} to get axes on the main image.}

\item{main}{Title for plot.}

\item{axisPalette}{Optional replacement function for \code{\link[=axis]{axis()}}, passed to
\code{\link[=drawPalette]{drawPalette()}}.}

\item{add}{Logical value indicating whether to add to an existing plot.
The default value, \code{FALSE} indicates that a new plot is to be created.
However, if \code{add} is \code{TRUE}, the idea is to add an image (but not
its palette or its axes) to an existing plot. Clearly, then, arguments
such \code{xlim} are to be ignored. Indeed, if \code{add=TRUE}, the only
arguments examined are \code{x} (which must be a vector; the mode of providing
a matrix or \code{oce} object does not work), \code{y}, \code{z},
\code{decimate}, plus either \code{colormap} or
both \code{breaks} and \code{col}.}

\item{debug}{A flag that turns on debugging.  Set to 1 to get a
moderate amount of debugging information, or to 2 to get more.}

\item{\dots}{Optional arguments passed to plotting functions.}
}
\value{
A list is silently returned, containing \code{xat} and \code{yat},
values that can be used by \code{\link[=oce.grid]{oce.grid()}} to add a grid to the
plot.
}
\description{
Plot an image with a color palette, in a way that does not conflict with
\code{\link{par}}\code{("mfrow")} or \code{\link[=layout]{layout()}}.  To plot just a palette,
e.g. to get an x-y plot with points colored according to a palette, use
\code{\link[=drawPalette]{drawPalette()}} and then draw the main diagram.
}
\details{
By default, creates an image with a color palette to the right.  The effect is similar to
\code{\link[=filled.contour]{filled.contour()}} except that with \code{imagep} it is possible to
set the \code{\link[=layout]{layout()}} outside the function, which enables the creation
of plots with many image-palette panels.  Note that the contour lines may not
coincide with the color transitions, in the case of coarse images.

Note that this does not use \code{\link[=layout]{layout()}} or any of the other screen
splitting methods.  It simply manipulates margins, and draws two plots
together.  This lets users employ their favourite layout schemes.

NOTE: \code{imagep} is an analogue of \code{\link[=image]{image()}}, and from that
it borrows a the convention that the number of rows in the matrix corresponds to
to \code{x} axis, not the \code{y} axis.  (Actually, \code{\link[=image]{image()}} permits
the length of \code{x} to match either \code{nrow(z)} or \code{1+nrow(z)}, but
here only the first is permitted.)
}
\examples{
library(oce)

# 1. simplest use
imagep(volcano)

# 2. something oceanographic (internal-wave speed)
h <- seq(0, 50, length.out=100)
drho <- seq(1, 3, length.out=200)
speed <- outer(h, drho, function(drho, h) sqrt(9.8 * drho * h / 1024))
imagep(h, drho, speed, xlab="Equivalent depth [m]",
ylab=expression(paste(Delta*rho, " [kg/m^3]")),
zlab="Internal-wave speed [m/s]")

# 3. fancy labelling on atan() function
x <- seq(0, 1, 0.01)
y <- seq(0, 1, 0.01)
angle <- outer(x,y,function(x,y) atan2(y,x))
imagep(x, y, angle, filledContour=TRUE, breaks=c(0, pi/4, pi/2),
       col=c("lightgray", "darkgray"),
       at=c(0, pi/4, pi/2),
       labels=c(0, expression(pi/4), expression(pi/2)))

# 5. y-axis flipping
par(mfrow=c(2,2))
data(adp)
d <- adp[["distance"]]
t <- adp[["time"]]
u <- adp[["v"]][ , ,1]
imagep(t, d, u, drawTimeRange=FALSE)
mtext("normal")
imagep(t, d, u, flipy=TRUE, drawTimeRange=FALSE)
mtext("flipy")
imagep(t, d, u, ylim=rev(range(d)), drawTimeRange=FALSE)
mtext("ylim")
imagep(t, d, u, ylim=rev(range(d)), flipy=TRUE, drawTimeRange=FALSE)
mtext("flipy and ylim")
par(mfrow=c(1,1))

# 6. a colormap case
data(topoWorld)
cm <- colormap(name="gmt_globe")
imagep(topoWorld, colormap=cm)

}
\seealso{
This uses \code{\link[=drawPalette]{drawPalette()}}, and is used by \code{\link[=plot,adp-method]{plot,adp-method()}},
\code{\link[=plot,landsat-method]{plot,landsat-method()}}, and other image-generating functions.
}
\author{
Dan Kelley and Clark Richards
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/colors.R
\name{colormapGMT}
\alias{colormapGMT}
\title{Create a GMT-type (CPT) colormap}
\usage{
colormapGMT(name, debug = getOption("oceDebug"))
}
\arguments{
\item{name}{character value specifying the GMT scheme, or a source for such
a scheme. Four pre-defined schemes are available, accessed by setting \code{name} to
\code{"gmt_gebco"}, \code{"gmt_globe"}, \code{"gmt_ocean"}, or \code{"gmt_relief"}. If \code{name} is
not one of these values, then it is taken to be the name of a local file
in GMT format or, if no such file is found, a URL holding such a file.}

\item{debug}{integer that, if positive, indicates to print some debugging output}
}
\value{
\code{colormap} returns a list, in the same format as the return value for \code{\link[=colormap]{colormap()}}.
}
\description{
\code{colormapGMT} creates colormaps in the Generic Mapping Tools (GMT)
scheme (see References 1 to 4).  A few such schemes are built-in, and may be referred to
by name (\code{"gmt_gebco"}, \code{"gmt_globe"}, \code{"gmt_ocean"}, or \code{"gmt_relief"})
while others are handled by reading local files that are in GMT
format, or URLs providing such files (see Reference 3).
}
\details{
The GMT files understood by \link{colormapGMT} are what GMT calls
"Regular CPT files" (see reference 4).  This is a text format that
can be read and (with care) edited in a text editor.  There
are three categories of lines within this file.  (1) Any
line starting with the \code{"#"} character is a comment, and is ignored
by \link{colormapGMT}. (2) Lines with 8 numbers specify colour bands.
The first number is a z value, and the three numbers after that
are red, green and blue values in the range from 0 to 255. This
set of 4 numbers is followed on the same line with similar values.
Think of this sequence as describing a band of colours between
two z values. (3) Lines starting with a character, followed
by three numbers, specify particular codings.  The character \code{"B"}
specifies background colour, while \code{"F"} specifies foreground
colour, and \code{"N"} specifies the colour to be used for missing
data (the letter stands for not-a-number).  Only \code{"N"} is used
by \link{colormapGMT}, and it takes on the role that the \code{missingColor}
argument would otherwise have.  (This is why \code{missingColor} is not
permitted if \code{name} is given.)
}
\references{
\enumerate{
\item General overview of GMT system
\url{https://www.generic-mapping-tools.org}.
\item Information on GMT color schemes
\url{https://docs.generic-mapping-tools.org/dev/cookbook/cpts.html}
\item Source of GMT specification files
https://beamreach.org/maps/gmt/share/cpt/ (this link worked
for years, but failed to respond in October 2021)
\item CPT (color palette table) format
\url{https://www.soest.hawaii.edu/gmt/gmt/html/GMT_Docs.html#x1-820004.15}
}
}
\seealso{
Other things related to colors: 
\code{\link{colormap}()},
\code{\link{oceColors9B}()},
\code{\link{oceColorsCDOM}()},
\code{\link{oceColorsChlorophyll}()},
\code{\link{oceColorsClosure}()},
\code{\link{oceColorsDensity}()},
\code{\link{oceColorsFreesurface}()},
\code{\link{oceColorsGebco}()},
\code{\link{oceColorsJet}()},
\code{\link{oceColorsOxygen}()},
\code{\link{oceColorsPAR}()},
\code{\link{oceColorsPalette}()},
\code{\link{oceColorsPhase}()},
\code{\link{oceColorsSalinity}()},
\code{\link{oceColorsTemperature}()},
\code{\link{oceColorsTurbidity}()},
\code{\link{oceColorsTurbo}()},
\code{\link{oceColorsTwo}()},
\code{\link{oceColorsVelocity}()},
\code{\link{oceColorsViridis}()},
\code{\link{oceColorsVorticity}()},
\code{\link{ocecolors}}
}
\author{
Dan Kelley
}
\concept{things related to colors}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swN2}
\alias{swN2}
\title{Squared buoyancy frequency for seawater}
\usage{
swN2(
  pressure,
  sigmaTheta = NULL,
  derivs,
  df,
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{pressure}{either pressure (dbar) (in which case \code{sigmaTheta} must
be provided) \emph{or} an object of class \code{ctd} object (in which case
\code{sigmaTheta} is inferred from the object.}

\item{sigmaTheta}{Surface-referenced potential density minus 1000
(kg/m\eqn{^3}{^3}).}

\item{derivs}{optional argument to control how the derivative
\eqn{d\sigma_\theta/dp}{d(sigmaTheta)/d(pressure)} is calculated.  This may
be a character string or a function of two arguments.  See \dQuote{Details}.}

\item{df}{argument passed to \code{\link[=smooth.spline]{smooth.spline()}} if this function is
used for smoothing; set to \code{NA} to prevent smoothing.}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}

\item{\dots}{additional argument, passed to \code{\link[=smooth.spline]{smooth.spline()}}, in
the case that \code{derivs="smoothing"}.  See \dQuote{Details}.}
}
\value{
Square of buoyancy frequency (\eqn{radian^2/s^2}{radian^2/s^2}).
}
\description{
Compute \eqn{N^2}{N^2}, the square of the buoyancy frequency for a seawater
profile.
}
\details{
Smoothing is often useful prior to computing buoyancy frequency, and so this
may optionally be done with \code{\link[=smooth.spline]{smooth.spline()}}, unless
\code{df=NA}, in which case raw data are used.  If \code{df} is not
provided, a possibly reasonable value computed from an analysis of the
profile, based on the number of pressure levels.

The core of the method involves computing potential density referenced to median
pressure, using the UNESCO-style \link{swSigmaTheta} function, and then differentiating
this with respect to pressure. The \code{derivs} argument is used
to control how this is done, as follows.
\itemize{
\item If \code{derivs} is not supplied, the action is as though it were
given as the string \code{"smoothing"}
\item If \code{derivs} equals \code{"simple"}, then the derivative of
density with respect to pressure is calculated as the ratio of first-order
derivatives of density and pressure, each calculated using
\code{\link[=diff]{diff()}}.  (A zero is appended at the top level.)
\item If \code{derivs} equals \code{"smoothing"}, then the processing
depends on the number of data in the profile, and on whether \code{df} is
given as an optional argument.  When the number of points exceeds 4, and
when \code{df} exceeds 1, \code{\link[=smooth.spline]{smooth.spline()}} is used to calculate
smoothing spline representation the variation of density as a function of
pressure, and derivatives are extracted from the spline using
\code{predict}.  Otherwise, density is smoothed using \code{\link[=smooth]{smooth()}},
and derivatives are calculated as with the \code{"simple"} method.
\item If \code{derivs} is a function taking two arguments (first pressure,
then density) then that function is called directly to calculate the
derivative, and no smoothing is done before or after that call.
}

For precise work, it makes sense to skip \code{swN2} entirely, choosing
whether, what, and how to smooth based on an understanding of fundamental
principles as well as data practicalities.
}
\section{Deprecation Notice}{

Until 2019 April 11, \code{swN2} had an argument named \code{eos}. However,
this did not work as stated, unless the first argument was a \code{ctd}
object. Besides, the argument name was inherently deceptive, because the UNESCO
scheme does not specify how N2 is to be calculated.
Nothing is really lost by making this change, because the new default is the
same as was previously available with the \code{eos="unesco"}
setup, and the gsw-formulated estimate of N2 is provided
by \code{\link[gsw:gsw_Nsquared]{gsw::gsw_Nsquared()}} in the \CRANpkg{gsw} package.
}

\examples{

library(oce)
data(ctd)
# Left panel: density
p <- ctd[["pressure"]]
ylim <- rev(range(p))
par(mfrow=c(1, 2), mar=c(3, 3, 1, 1), mgp=c(2, 0.7, 0))
plot(ctd[["sigmaTheta"]], p, ylim=ylim, type='l', xlab=expression(sigma[theta]))
# Right panel: N2, with default settings (black) and with df=2 (red)
N2 <- swN2(ctd)
plot(N2, p, ylim=ylim, xlab="N2 [1/s^2]", ylab="p", type="l")
lines(swN2(ctd, df=3), p, col=2)

}
\seealso{
The \code{\link[gsw:gsw_Nsquared]{gsw::gsw_Nsquared()}} function of the \CRANpkg{gsw}
provides an alternative to this, as formulated in the GSW system. It
has a more sophisticated treatment of potential density, but it is based
on simple first-difference derivatives, so its results may require
smoothing, depending on the dataset and purpose of the analysis.

Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swLapseRate}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/lisst.R
\name{[[<-,lisst-method}
\alias{[[<-,lisst-method}
\title{Replace Parts of a LISST Object}
\usage{
\S4method{[[}{lisst}(x, i, j, ...) <- value
}
\arguments{
\item{x}{a \linkS4class{lisst} object.}

\item{i}{character value naming the item to replace.}

\item{j}{optional additional information on the \code{i} item.}

\item{...}{optional additional information (ignored).}

\item{value}{The value to be placed into \code{x}, somewhere.}
}
\description{
The \code{[[<-} method works for all \linkS4class{oce} objects.
The purpose, as with the related extraction method, \code{[[},
is to insulate users from the internal details of \linkS4class{oce}
objects, by looking for items within the various storage
slots of the object. Items not actually stored can also be
replaced, including units and data-quality
flags.
}
\details{
As with \code{[[} method, the procedure works in steps.

First, the \code{metadata} slot of \code{x} is checked to
see whether it contains something named with \code{i}.
If so, then the named item is replaced with \code{value}.

Otherwise, if the string value of \code{i} ends in \code{Unit}, then the
characters preceding that are taken as the name of a variable, and
the \code{metadata} slot of \code{x} is updated to store that unit, e.g.\preformatted{x[["temperatureUnits"]] <- list(unit=expression(degree*F),scale="")
}

Similarly, if \code{i} ends in \code{Flag}, then quality-control
flags are set up as defined by \code{result}, e.g.\preformatted{o[["temperatureFlags"]] <- c(2,4,2,2)
}

Otherwise, \code{\link[=pmatch]{pmatch()}} is used for a partial-string match with
the names of the items that are in the \code{data} slot of \code{x}.
The first item found (if any) is then updated to hold the value \code{result}.

If none of these conditions is met, a warning is issued.
}
\seealso{
Other functions that replace parts of oce objects: 
\code{\link{[[<-,adp-method}},
\code{\link{[[<-,amsr-method}},
\code{\link{[[<-,argo-method}},
\code{\link{[[<-,bremen-method}},
\code{\link{[[<-,cm-method}},
\code{\link{[[<-,coastline-method}},
\code{\link{[[<-,ctd-method}},
\code{\link{[[<-,echosounder-method}},
\code{\link{[[<-,g1sst-method}},
\code{\link{[[<-,gps-method}},
\code{\link{[[<-,ladp-method}},
\code{\link{[[<-,landsat-method}},
\code{\link{[[<-,lobo-method}},
\code{\link{[[<-,met-method}},
\code{\link{[[<-,oce-method}},
\code{\link{[[<-,odf-method}},
\code{\link{[[<-,rsk-method}},
\code{\link{[[<-,sealevel-method}},
\code{\link{[[<-,section-method}},
\code{\link{[[<-,tidem-method}},
\code{\link{[[<-,topo-method}},
\code{\link{[[<-,windrose-method}},
\code{\link{[[<-,xbt-method}}

Other things related to lisst data: 
\code{\link{[[,lisst-method}},
\code{\link{as.lisst}()},
\code{\link{lisst-class}},
\code{\link{plot,lisst-method}},
\code{\link{read.lisst}()},
\code{\link{summary,lisst-method}}
}
\concept{functions that replace parts of oce objects}
\concept{things related to lisst data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sw.R
\name{swLapseRate}
\alias{swLapseRate}
\title{Seawater lapse rate}
\usage{
swLapseRate(
  salinity,
  temperature = NULL,
  pressure = NULL,
  longitude = NULL,
  latitude = NULL,
  eos = getOption("oceEOS", default = "gsw")
)
}
\arguments{
\item{salinity}{either salinity (PSU) (in which case \code{temperature} and
\code{pressure} must be provided) \emph{or} a \code{ctd} object (in which
case \code{salinity}, \code{temperature} and \code{pressure} are determined
from the object, and must not be provided in the argument list).}

\item{temperature}{\emph{in-situ} temperature (\eqn{^\circ}{deg}C), defined
on the ITS-90 scale; see \dQuote{Temperature units} in the documentation for
\code{\link[=swRho]{swRho()}}.}

\item{pressure}{pressure (dbar)}

\item{longitude}{longitude of observation (only used if \code{eos="gsw"};
see \sQuote{Details}).}

\item{latitude}{latitude of observation (only used if \code{eos="gsw"}; see
\sQuote{Details}).}

\item{eos}{equation of state, either \code{"unesco"} (references 1 and 2)
or \code{"gsw"} (references 3 and 4).}
}
\value{
Lapse rate (\eqn{deg}{deg}C/m).
}
\description{
Compute adiabatic lapse rate
}
\details{
If \code{eos="unesco"}, the density is calculated using the UNESCO equation
of state for seawater (references 1 and 2), and if \code{eos="gsw"}, the GSW formulation
(references 3 and 4) is used.
}
\examples{
lr <- swLapseRate(40, 40, 10000) # 3.255976e-4

}
\references{
Fofonoff, P. and R. C. Millard Jr, 1983. Algorithms for
computation of fundamental properties of seawater.
\emph{Unesco Technical Papers in Marine Science},
\emph{44}, 53 pp.  (Section 7, pages 38-40)
}
\seealso{
Other functions that calculate seawater properties: 
\code{\link{T68fromT90}()},
\code{\link{T90fromT48}()},
\code{\link{T90fromT68}()},
\code{\link{computableWaterProperties}()},
\code{\link{locationForGsw}()},
\code{\link{swAbsoluteSalinity}()},
\code{\link{swAlphaOverBeta}()},
\code{\link{swAlpha}()},
\code{\link{swBeta}()},
\code{\link{swCSTp}()},
\code{\link{swConservativeTemperature}()},
\code{\link{swDepth}()},
\code{\link{swDynamicHeight}()},
\code{\link{swN2}()},
\code{\link{swPressure}()},
\code{\link{swRho}()},
\code{\link{swRrho}()},
\code{\link{swSCTp}()},
\code{\link{swSR}()},
\code{\link{swSTrho}()},
\code{\link{swSigma0}()},
\code{\link{swSigma1}()},
\code{\link{swSigma2}()},
\code{\link{swSigma3}()},
\code{\link{swSigma4}()},
\code{\link{swSigmaTheta}()},
\code{\link{swSigmaT}()},
\code{\link{swSigma}()},
\code{\link{swSoundAbsorption}()},
\code{\link{swSoundSpeed}()},
\code{\link{swSpecificHeat}()},
\code{\link{swSpice}()},
\code{\link{swSstar}()},
\code{\link{swTFreeze}()},
\code{\link{swTSrho}()},
\code{\link{swThermalConductivity}()},
\code{\link{swTheta}()},
\code{\link{swViscosity}()},
\code{\link{swZ}()}
}
\author{
Dan Kelley
}
\concept{functions that calculate seawater properties}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sealevel.R
\name{read.sealevel}
\alias{read.sealevel}
\title{Read a Sealevel File}
\usage{
read.sealevel(
  file,
  tz = getOption("oceTz"),
  processingLog,
  debug = getOption("oceDebug")
)
}
\arguments{
\item{file}{a connection or a character string giving the name of the file
to load.  See Details for the types of files that are recognized.}

\item{tz}{time zone.  The default value, \code{oceTz}, is set to \code{UTC}
at setup.  (If a time zone is present in the file header, this will
supercede the value given here.)}

\item{processingLog}{if provided, the action item to be stored in the log.
(Typically only provided for internal calls; the default that it provides is
better for normal calls by a user.)}

\item{debug}{an integer specifying whether debugging information is
to be printed during the processing. This is a general parameter that
is used by many \code{oce} functions. Generally, setting \code{debug=0}
turns off the printing, while higher values suggest that more information
be printed. If one function calls another, it usually reduces the value of
\code{debug} first, so that a user can often obtain deeper debugging
by specifying higher \code{debug} values.}
}
\value{
A \linkS4class{sealevel} object.
}
\description{
Read a data file holding sea level data.  BUG: the time vector assumes GMT,
regardless of the GMT.offset value.
}
\details{
This function starts by scanning the first line of the file, from which it
determines whether the file is in one of two known formats: type 1, the
format used at the Hawaii archive centre, and type 2, the
comma-separated-value format used by the Marine Environmental Data Service.
The file type is inferred by examination of its first line. If that contains
the string \code{Station_Name} the file is of type 2. If
the file is in neither of these formats, the user might wish to scan it
directly, and then to use \code{\link[=as.sealevel]{as.sealevel()}} to create a
\code{sealevel} object.
The Hawaii archive site at
\verb{http://ilikai.soest.hawaii.edu/uhslc/datai.html} at one time provided a graphical
interface for downloading sealevel data in Type 1, with format that was once
described at \verb{http://ilikai.soest.hawaii.edu/rqds/hourly.fmt} (although that link
was observed to no longer work, on December 4, 2016).
Examination of data retrieved from what seems to be a replacement Hawaii server
(https://uhslc.soest.hawaii.edu/data/?rq) in September 2019 indicated that the
format had been changed to what is called Type 3 by \code{read.sealevel}.
Web searches did not uncover documentation on this format, so the
decoding scheme was developed solely through examination of
data files, which means that it might be not be correct.
The MEDS repository (\code{http://www.isdm-gdsi.gc.ca/isdm-gdsi/index-eng.html})
provides Type 2 data.
}
\seealso{
Other things related to sealevel data: 
\code{\link{[[,sealevel-method}},
\code{\link{[[<-,sealevel-method}},
\code{\link{as.sealevel}()},
\code{\link{plot,sealevel-method}},
\code{\link{sealevel-class}},
\code{\link{sealevelTuktoyaktuk}},
\code{\link{sealevel}},
\code{\link{subset,sealevel-method}},
\code{\link{summary,sealevel-method}}
}
\author{
Dan Kelley
}
\concept{things related to sealevel data}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/adp.nortek.R
\name{decodeHeaderNortek}
\alias{decodeHeaderNortek}
\title{Decode a Nortek Header}
\usage{
decodeHeaderNortek(
  buf,
  type = c("aquadoppHR", "aquadoppProfiler", "aquadopp", "aquadoppPlusMagnetometer",
    "vector"),
  debug = getOption("oceDebug"),
  ...
)
}
\arguments{
\item{buf}{a ``raw'' buffer containing the header}

\item{type}{type of device}

\item{debug}{a flag that turns on debugging.  Set to 1 to get a moderate
amount of debugging information, or to 2 to get more.}

\item{\dots}{additional arguments, passed to called routines.}
}
\value{
A list containing elements \code{hardware}, \code{head}, \code{user}
and \code{offset}.  The easiest way to find the contents of these is to run
this function with \code{debug=3}.
}
\description{
Decode data in a Nortek ADV or ADP header.
}
\details{
Decodes the header in a binary-format Nortek ADV/ADP file.  This function is
designed to be used by \code{\link[=read.adp]{read.adp()}} and \code{\link[=read.adv]{read.adv()}},
but can be used directly as well.  The code is based on information in the
Nortek System Integrator Guide (2008) and on postings on the Nortek
``knowledge center'' discussion board.  One might assume that the latter is
less authoritative than the former.  For example, the inference of cell size
follows advice found at
https://www.nortekusa.com/en/knowledge-center/forum/hr-profilers/736804717,
which contains a typo in an early posting that is
corrected later on.
}
\references{
\enumerate{
\item Information on Nortek profilers (including the System
Integrator Guide, which explains the data format byte-by-byte) is available
at \url{https://www.nortekusa.com/usa?set_language=usa} after login.
\item The Nortek Knowledge Center
https://www.nortekusa.com/en/knowledge-center
may be of help if problems arise in dealing with data from Nortek instruments.
}
}
\seealso{
Most users should employ the functions \code{\link[=read.adp]{read.adp()}} and
\code{\link[=read.adv]{read.adv()}} instead of this one.
}
\author{
Dan Kelley and Clark Richards
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/misc.R
\name{oce.as.raw}
\alias{oce.as.raw}
\title{Version of as.raw() that clips data}
\usage{
oce.as.raw(x)
}
\arguments{
\item{x}{values to be converted to raw}
}
\value{
Raw values corresponding to \code{x}.
}
\description{
A version of as.raw() that clips data to prevent warnings
}
\details{
Negative values are clipped to 0, while values above 255 are clipped to 255;
the result is passed to \code{\link[=as.raw]{as.raw()}} and returned.
}
\examples{
x <- c(-0.1, 0, 1, 255, 255.1)
data.frame(x, oce.as.raw(x))
}
\author{
Dan Kelley
}
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/AllClass.R
\name{composite}
\alias{composite}
\title{Create a composite object by averaging across good data}
\usage{
composite(object, ...)
}
\arguments{
\item{object}{either a \link{list} of \linkS4class{oce} objects, in
which case this is the only argument, or a single \linkS4class{oce} object,
in which case at least one other argument (an object of the same size)
must be supplied.}

\item{...}{Ignored, if \code{object} is a list. Otherwise, one or more
\linkS4class{oce} objects of the same sub-class as the first argument.}
}
\description{
objects that are supplied as arguments are averaged in a way that
makes sense for the object class, i.e. taking into account
the particular bad-data codes of that particular class.
}
\seealso{
Other functions that create composite objects: 
\code{\link{composite,amsr-method}},
\code{\link{composite,list-method}}
}
\concept{functions that create composite objects}
