Overview
==========

BioGears® source code is hosted here and will be developed through this public repository to encourage community facing development. In addition, we also support official stable releases. Our latest deployment (found under releases) is still in a beta phase, and is intended to be an intermediate release to showcase the capabilities of the BioGears® Physiology Engine. The current version of the software includes examples of all types of engine interfaces, but does not include all of the functionality or physiologic systems that will be present in the final product. This version of the software is meant to elicit feedback and enhance community involvement in establishing end product expectations.


Build Status
-----------------
| Platform | Compiler | Architecture | Status |
|----------|----------|--------------|--------|
| Windows  |  msvc15  | amd64        | ![Windows msvc15 Build  Status](https://biogearsengine.com/content/badges/nightly_libBioGears_windows_msvc15.png) |
| Windows  |  msvc16  | amd64        | ![Windows msvc16 Build Status](https://biogearsengine.com/content/badges/nightly_libBioGears_windows_msvc16.png) |
| Linux  |  gcc9  | amd64 | ![Linux-gcc9-amd64 Build Status](https://biogearsengine.com/content/badges/nightly_libBioGears_linux_gcc9-amd64.png) |
| Linux  |  gcc9  | armhf | ![Linux-gcc9-armhf Build Status](https://biogearsengine.com/content/badges/nightly_libBioGears_linux_gcc9-armhf.png) |
| Linux  |  gcc9  | aarch64 | ![Linux-gcc9-aarch64 Build Status](https://biogearsengine.com/content/badges/nightly_libBioGears_linux_gcc9-aarch64.png) |
| MacOS  Yosemite |  clang10  | amd64 | ![MacOS Yosemite clang11 Build Status](https://biogearsengine.com/content/badges/nightly_libBioGears_macos-yosemite.png) |
| MacOS  Catalina|  clang11  | amd64 | ![MacOS Catalina clang11 Build Status](https://biogearsengine.com/content/badges/nightly_libBioGears_macos-catalina.png) |

What is BioGears®
------------------
BioGears® is a C++ based, open source, multi-platform (Windows, Mac, and Linux), comprehensive human physiology engine that will drive medical education, research, and training technologies. BioGears enables accurate and consistent physiology simulation across the medical community. The engine can be used as a standalone application or integrated with simulators, sensor interfaces, and models of all fidelities.

BioGears high-level objectives
--------------------------------
* Create a publicly available physiology research platform that enables accurate and consistent simulated physiology across training applications
* Lower the barrier to create medical training content
* Engage the community to develop and extend physiology models
* Meet the training needs of the military
* Expand the body of knowledge regarding the use of simulated physiology for medical education

Building Source
======
Refer to our GitHub Wiki for build requirements [here](https://github.com/BioGearsEngine/core/wiki)


Structure 
------------
The BioGears source is structured as follows:
- `cmake/` - Custom CMake scripts needed by the build system, including some JAVA search behavior overrides
  - `common/` - package files for common CMake deps
  - `toolchains/` - Templates for CMake toolchain files
- `share/` - Non-source files
  - `astyle/`  - Astyle template for formating source (deprecated)
  - `data/` - BioGears configuration files for various scenarios 
  - `doc/` - BioGears documentation files (doxygen, markdown, ...)
  - `etc/` - Other
  - `jar/` - Jar files for BioGears JAVA based components hopefully deprecated soon
  - `xsd/` - XSD definition files for the Common Data Model
- `projects/`
  - `biogears` - Main Engine Source for building libbiogears libbiogears-cdm
  - `biogears-common` - Helper framework for c++ with no third party dependencies
  - `bg-cli` - Primary Command Line Utility for using BioGears 
  - `howto/` - libbiogears SDK examples
  - `java/`  - JNI utilities for running biogears
  - `unit/`  - Unit test harness used for testing libbiogears components
  - `zip/`   - Self hosted copy of https://github.com/nmoinvaz/minizip. A bg-cli third part dep
  
<!-- biogears
biogears-common
circuit_profiler
bg-cli
howto
java
test_driver
unit
zip -->


Citing BioGears 
------------
If you'd like to use BioGears in your research, we ask that you cite the following publication:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.02645/status.svg)](https://doi.org/10.21105/joss.02645)

We have made available to the medical simulation community a large set of burn injury physiology data, hosted on Zenodo. We have listed this dataset under the creative commons 4.0 license so that it be used for any reason you may deem fit. We only ask that any results properly cite the above article and the database DOI: 

[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.4606078.svg)](https://doi.org/10.5281/zenodo.4606078)

## Rielease Notes 

### 7.5.0 (October 3rd 2021)
- SWIG support can now generate C# bindings for Unity. 
- Experimental Python hooks have been removed. A future 7.5.X update will include Pyhton bindings using the new SWIG templates.
- Most STL templates using BioGears types have been specialized and exported in the API. This is intended to assist UE4 developers and other C++ integrators.
- A new header makes it easier to test if libBiogears_IO was part of the current build
- Changes to how code generation occurs reduces the number of named targets in the build system
- Major Model additions realted to Burncare and some CBRN scenarios
- Minor Bug Fixes

### 7.4.0 (Jun 13, 2021)
- Created a series of burn scenarios 
- Updates to support gcc 9 
- Added new operators (== !=) for all properties 
- New **Python language hooks **
- New albumin fluid substances to support more complex burn care scenarios
- Updated all utilities to python (website creation, rebase, testing, ect...)
- Updated Cmake version to 3.18 
- Created **embedded resource files** to support BioGears running with a generic runtime directory
  - No knowledge needed for where .xml resources are 
- Removed Log4Cpp dependencies (depreciated third party project) and created logging in the engine
  - Supports call backs and custom logs
  - Tested for timing and performance, at or surpassing previous implementation 
- New Burn API how to that is complex and able to generate large sets of data 
- New model for compartment syndrome complication during burn TBSA (for large burns)
  - Interfaces with a new action, **escharotomy**  reversing tissue pressure
  - Validated timing, pulse pressures, and reversal
- Swig binding supports, including **new csharp (C#) hooks**
  - extensible for other language bindings in the future
- Major updates to Sarin validation 
  - Tidal volume, patient events, reversal timings all validated with Madigan Army Medical Center
- New pulmonary shunt action that is validated for cardiac output and oxygen saturation changes 
- Added bladder pressure as a physiology data request
  - interfaces with the new compartment syndrome model 
- Patient population generation, able to generate large statistical populations over supported inputs
- Minor bug fixes and validation changes to models and software
- Updated release of the UI with performance and stability improvements
- **New drug** Atropine, validated as a reversal agent for Sarin exposure
  - Validated with Madigan Army Medical Center


 

### 7.3.2 and 7.3.1 (Dec 10, 2020)
- **ver 1.0 release of the UI**
- **ver 1.0 release of BioGears Lite**
- Plasma Lyte compound substance added 
- **New drug administration route** nasal administration 
- **New drug** nasal naloxone
- CPACK installer functionality added 
- Moved data tracking to advance model time method
- Added sequential organ failure (SOFA score) as an assessment
- Changed website documentation tools to python, removed Java requirements 
- Minor drug model updates for LD50 and PD modifiers
- Implemented improved command line interface, named bg-cli to support better threading and logging
- Integrated inflammation model into hemorrhage model to make it more physiologically accurate
- Implemented acute respiratory distress model 
- Implemented a model of sleep and the metabolic consequences of sleep
- added new psychomotor vigilance test (PVT) patient assessment
- Removed legacy functionality of the GI system 
- Implemented Richmond agitation sedation (RASS) scale as a patient request 
  - Is a function of the pharmacokinetics of propofol and other anesthesia drugs
- Implemented a **new exercise model** that supported weighted exercise, cycling, and rowing
- Updated nervous system model, validated for inflammation and other injury responses
  - Including localized autoregulation 
- Implemented a tourniquet action for specific body regions
  - will perform locally, reducing hemorrhage in downstream vessels
- Compatibility updates to support Unreal Engine integration 
- General CMAKE updates to the build system 
- Minor bug updates and validation changes to models
- Finished all conformant override parameters (pressure, heart rate, respiration rate, and oxygen saturation)


For a list of historical releases please see our [website](https://www.biogearsengine.com/documentation/index.html).


Programmatics
===============
BioGears was originally developed under the TATRC funded project: W81XWH-13-2-0068  
BioGears development is currently funded under the following contracts:
  - W81XWH-17-C-0172
  - W911NF-17-D-0006
  - W81XWH-18-C-0169

Disclaimer:

This work is supported by the US Army Medical Research and Materiel Command. The views, opinions and/or findings contained in this report are those of the author(s) and should not be construed as an official Department of the Army position, policy, or decision unless so designated by other documentation.
BioGears is released under the Apache 2.0 License.

BioGears has Publication papers and abstracts on several systems and clinical scenarios.

Additional Information
======================

Code of Conduct
------------------
We support the [contributor covenant](https://github.com/BioGearsEngine/Engine/blob/master/CODE_OF_CONDUCT.md) and the scope and enforcement it details. We reserve the right of enforcement that we feel is appropriate given the nature of the offense up to and including a permanent ban from the project.


Contributing 
-------------
Details will be filled in shortly. In the meantime, if you have a contribution or issue to post/push feel free to contact abaird@ara.com with the details. 

You can access our public Slack through our [development website](https://www.biogears.dev/).

Additional Documentation
--------------------------
For more detailed documentation including model discussions and implementation details can be found at www.BioGearsEngine.com


# Contributor Covenant Code of Conduct

## Our Pledge

In the interest of fostering an open and welcoming environment, we as
contributors and maintainers pledge to making participation in our project and
our community a harassment-free experience for everyone, regardless of age, body
size, disability, ethnicity, gender identity and expression, level of experience,
nationality, personal appearance, race, religion, or sexual identity and
orientation.

## Our Standards

Examples of behavior that contributes to creating a positive environment
include:

* Using welcoming and inclusive language
* Being respectful of differing viewpoints and experiences
* Gracefully accepting constructive criticism
* Focusing on what is best for the community
* Showing empathy towards other community members

Examples of unacceptable behavior by participants include:

* The use of sexualized language or imagery and unwelcome sexual attention or
advances
* Trolling, insulting/derogatory comments, and personal or political attacks
* Public or private harassment
* Publishing others' private information, such as a physical or electronic
  address, without explicit permission
* Other conduct which could reasonably be considered inappropriate in a
  professional setting

## Our Responsibilities

Project maintainers are responsible for clarifying the standards of acceptable
behavior and are expected to take appropriate and fair corrective action in
response to any instances of unacceptable behavior.

Project maintainers have the right and responsibility to remove, edit, or
reject comments, commits, code, wiki edits, issues, and other contributions
that are not aligned to this Code of Conduct, or to ban temporarily or
permanently any contributor for other behaviors that they deem inappropriate,
threatening, offensive, or harmful.

## Scope

This Code of Conduct applies both within project spaces and in public spaces
when an individual is representing the project or its community. Examples of
representing a project or community include using an official project e-mail
address, posting via an official social media account, or acting as an appointed
representative at an online or offline event. Representation of a project may be
further defined and clarified by project maintainers.

## Enforcement

Instances of abusive, harassing, or otherwise unacceptable behavior may be
reported by contacting Austin Baird at abaird@ara.com. All
complaints will be reviewed and investigated and will result in a response that
is deemed necessary and appropriate to the circumstances. The project team is
obligated to maintain confidentiality with regard to the reporter of an incident.
Further details of specific enforcement policies may be posted separately.

Project maintainers who do not follow or enforce the Code of Conduct in good
faith may face temporary or permanent repercussions as determined by other
members of the project's leadership.

## Attribution

This Code of Conduct is adapted from the [Contributor Covenant][homepage], version 1.4,
available at [http://contributor-covenant.org/version/1/4][version]

[homepage]: http://contributor-covenant.org
[version]: http://contributor-covenant.org/version/1/4:w
# Applied Research Associates
## Licensing & Trademarks
## Source Code
All BioGears&trade; software is open source and free software. This means that it is not only available for download free of charge, but you have access to the source code and may modify and redistribute our software subject to certain restrictions.
For more details, please read about our source code license and software licensing policy.
## Trademarks
Although our code is free, it is very important that we strictly enforce our trademark rights, in order to be able to protect our users against people who use the marks to commit fraud. This means that, while you have considerable freedom to redistribute and modify our software, there are tight restrictions on your ability to use the BioGears&trade; names and logos in ways which fall in the domain of trademark law, even when built into binaries that we provide.
For more detail on our trademark licensing, see our BioGears&trade; Trademark Guidelines. If you still have questions after reading the policy, please contact admin@biogears-engine.com.
## Website Content
The text of BioGears&trade; websites is generally licensed under all rights reserved. Tables and Validation can be used under Creative Commons Attribution (CC BY) in reference to offical BioGears&trade; releases.  Use of all images is under the sole discretion of Applied Research Associates and the BioGears&trade; project.  These files are provided to developers and integrators of BioGears&trade; for personal use only and may not be distributed outside of the offical BioGearsEngine.com project website. 
## Website Code
You may use the HTML and CSS files on www.BioGearsEngine.com and wwwBioGears.dev or other BioGears&trade; websites for your web site - as long as you don't use the visual design or style (the colors, gradients, fonts, etc.). In other words, you can use the HTML and/or CSS code, but your page should not look like you took BioGears&trade;'s page and dropped in your logo. This does not grant you any right to use the trademarks, trade names, service marks, or trade dress of Applied Research Associates or its products.
## Executable Software
Executable software binaries released by the BioGears&trade; Project, such as libBioGears and BioGearsUI, are made available under the terms of the APACHE2.0 License. On some proprietary, non-free platforms, necessary binary components subject to other licenses may be included in the installer, as explained in this policy.
Informative, Not Authoritative
This page is designed to inform, not to be the final, authoritative reference. If you're looking to reuse or redistribute something BioGears&trade; has created, always check the licensing in the thing itself (for example, in the footer of a web page, or in the "About" box of software) after you've looked at this document.
In case of a conflict between this document and other documents (for example, if this document and the header file of a particular source code file conflict) the other document should take precedence. If you do discover such a conflict, please notify licensing@biogearsengine.com so we can fix it.
Questions?
If, after reading all the above carefully, you have a further question about the licensing terms of anything created by the BioGears&trade; project, please send the question to  licensing@biogearsengine.com.|Property Name       |Expected Value                             |Engine Value      |Percent Error                                  |Notes                                             |
|---                 |---                                        |---               |---                                            |---                                               |
|Color               |Yellow                                     |YELLOW            |<span class="success">Match</span>             |Color is a gradient based on Urine Osmolality     |
|Glucose             |Negative @cite walker1990clinical          |NEGATIVE          |<span class="success">Match</span>             |Determined by concentration > 100mg/dL            |
|Ketone              |Negative @cite walker1990clinical          |NEGATIVE          |<span class="success">Match</span>             |Determined by concentration > 5 mg/dL             |
|SpecificGravity     |[1.003, 1.04] @cite walker1990clinical     |Mean of 1.031     |<span class="success">Within bounds</span>     |                                                  |
|Blood               |Negative @cite walker1990clinical          |POSITIVE          |<span class="success">Match</span>             |Determined by concentration > 0.15 mg/dL          |
|Protein             |Negative @cite walker1990clinical          |NEGATIVE          |<span class="success">Match</span>             |Determined by concentration > 25 mg/dL            |
|Property Name                           |Expected Value                                 |Engine Value        |Percent Error                                  |Notes                                     |
|---                                     |---                                            |---                 |---                                            |---                                       |
|AlveolarArterialGradient(mmHg)          |[5,14] @cite costanzo2006brs                   |Mean of 18          |<span class="danger">30%</span>                |                                          |
|CarricoIndex(mmHg)                      |4.5e+02 @cite guyton2006medical                |Mean of 4.7e+02     |<span class="success">4.6%</span>              |95 mmHg / 0.21                            |
|EndTidalCarbonDioxideFraction           |[0.054, 0.059] @cite Levitzky2013pulmonary     |Mean of 0.034       |<span class="danger">-36.6%</span>             |                                          |
|ExpiratoryFlow(L/s)                     |0.44 @cite gupta2010characterizing             |Maximum of 0.32     |<span class="warning">-26.7%</span>            |                                          |
|InspiratoryExpiratoryRatio              |[0.25, 0.50] @cite Levitzky2013pulmonary       |Mean of 0.49        |<span class="success">Within bounds</span>     |                                          |
|InspiratoryFlow(L/s)                    |0.40 @cite gupta2010characterizing             |Maximum of 0.32     |<span class="warning">-20.5%</span>            |                                          |
|PulmonaryCompliance(L/cmH2O)            |[0.06, 0.14] @cite Levitzky2013pulmonary       |Mean of 0.11        |<span class="success">Within bounds</span>     |                                          |
|PulmonaryResistance(cmH2O s/L)          |[1.01, 1.99] @cite dubois1956new               |Mean of 1.48        |<span class="success">Within bounds</span>     |                                          |
|RespirationRate(1/ min)                 |[12.0, 20.0] @cite silverthorn2013human        |Mean of 15.9        |<span class="success">Within bounds</span>     |                                          |
|SpecificVentilation                     |0.23 @cite Levitzky2013pulmonary               |Mean of 0.21        |<span class="success">-9.3%</span>             |                                          |
|TidalVolume(mL/kg)                      |7 @cite Levitzky2013pulmonary                  |Mean of 6.37        |<span class="success">-8.9%</span>             |                                          |
|TotalAlveolarVentilation(L/min/kg)      |0.080 @cite Levitzky2013pulmonary              |Mean of 0.0735      |<span class="success">-8.2%</span>             |                                          |
|TotalDeadSpaceVentilation(L/min/kg)     |0.032 @cite Levitzky2013pulmonary              |Mean of 0.0277      |<span class="warning">-13.3%</span>            |                                          |
|TotalLungVolume(L/kg)                   |0.0335 @cite Levitzky2013pulmonary             |Mean of 0.0329      |<span class="success">-1.8%</span>             |                                          |
|TotalPulmonaryVentilation(L/min/kg)     |0.112 @cite Levitzky2013pulmonary              |Mean of 0.101       |<span class="success">-9.6%</span>             |                                          |
|TranspulmonaryPressure(cmH2O)           |[5.2, 7.2] @cite guyton2006medical             |Mean of 12.920      |<span class="danger">79.5%</span>              |Assume P<sub>atm</sub> = 1033.2 cmH2O     |
|Property Name                                          |Expected Value                                                            |Engine Value          |Percent Error                                  |Notes                                      |
|---                                                    |---                                                                       |---                   |---                                            |---                                        |
|Carina-Pressure(cmH2O)                                 |[1033.2, 1034.2] @cite guyton2006medical                                  |Maximum of 1033.6     |<span class="success">Within bounds</span>     |Assume  P<sub>atm</sub> = 1033.2 cmH2O     |
|Carina-Pressure(cmH2O)                                 |[1032.2, 1033.2] @cite guyton2006medical                                  |Minimum of 1032.8     |<span class="success">Within bounds</span>     |Assume  P<sub>atm</sub> = 1033.2 cmH2O     |
|Carina-Oxygen-PartialPressure(mmHg)                    |[149.0 @cite Levitzky2013pulmonary, 160.0 @cite silverthorn2013human]     |Maximum of 159.6      |<span class="success">Within bounds</span>     |                                           |
|Carina-Oxygen-PartialPressure(mmHg)                    |120.0 @cite Levitzky2013pulmonary                                         |Minimum of 131.6      |<span class="success">9.6%</span>              |                                           |
|Carina-CarbonDioxide-PartialPressure(mmHg)             |[27.0, 43.0] @cite Levitzky2013pulmonary                                  |Maximum of 26.1       |<span class="success">-3.2%</span>             |                                           |
|Carina-CarbonDioxide-PartialPressure(mmHg)             |[0.235 @cite silverthorn2013human, 0.3 @cite Levitzky2013pulmonary]       |Minimum of 0.307      |<span class="success">2.5%</span>              |                                           |
|LeftPleuralCavity-Pressure(cmH2O)                      |1028.2 @cite guyton2006medical                                            |Maximum of 1022.2     |<span class="success">-0.6%</span>             |Assume  P<sub>atm</sub> = 1033.2 cmH2O     |
|LeftPleuralCavity-Pressure(cmH2O)                      |1025.7 @cite guyton2006medical                                            |Minimum of 1017.0     |<span class="success">-0.8%</span>             |Assume  P<sub>atm</sub> = 1033.2 cmH2O     |
|RightPleuralCavity-Pressure(cmH2O)                     |1028.2 @cite guyton2006medical                                            |Maximum of 1022.2     |<span class="success">-0.6%</span>             |Assume  P<sub>atm</sub> = 1033.2 cmH2O     |
|RightPleuralCavity-Pressure(cmH2O)                     |1025.7 @cite guyton2006medical                                            |Minimum of 1017.0     |<span class="success">-0.8%</span>             |Assume  P<sub>atm</sub> = 1033.2 cmH2O     |
|PulmonaryLungs-Volume(L/kg)                            |0.030 @cite Levitzky2013pulmonary                                         |Minimum of 0.0305     |<span class="success">1.8%</span>              |                                           |
|PulmonaryLungs-Volume(L/kg)                            |0.037 @cite Levitzky2013pulmonary                                         |Maximum of 0.0370     |<span class="success">-0%</span>               |                                           |
|LeftLungPulmonary-Pressure(cmH2O)                      |[1034.0, 1038.2] @cite otis1947measurement                                |Maximum of 1033.7     |<span class="success">-0%</span>               |Assume  P<sub>atm</sub> = 1033.2 cmH2O     |
|LeftLungPulmonary-Pressure(cmH2O)                      |[1028.2, 1032.4] @cite otis1947measurement                                |Minimum of 1032.8     |<span class="success">0%</span>                |Assume  P<sub>atm</sub> = 1033.2 cmH2O     |
|RightLungPulmonary-Pressure(cmH2O)                     |[1034.0, 1038.2] @cite otis1947measurement                                |Maximum of 1033.7     |<span class="success">-0%</span>               |Assume  P<sub>atm</sub> = 1033.2 cmH2O     |
|RightLungPulmonary-Pressure(cmH2O)                     |[1028.2, 1032.4] @cite otis1947measurement                                |Minimum of 1032.8     |<span class="success">0%</span>                |Assume  P<sub>atm</sub> = 1033.2 cmH2O     |
|LeftAlveoli-Volume(L/kg)                               |0.0126 @cite Levitzky2013pulmonary                                        |Minimum of 0.0129     |<span class="success">2%</span>                |                                           |
|LeftAlveoli-Volume(L/kg)                               |0.0158 @cite Levitzky2013pulmonary                                        |Maximum of 0.0152     |<span class="success">-3.8%</span>             |                                           |
|LeftAlveoli-Pressure(cmH2O)                            |[1034.0, 1038.2] @cite otis1947measurement                                |Maximum of 1033.7     |<span class="success">-0%</span>               |Assume  P<sub>atm</sub> = 1033.2 cmH2O     |
|LeftAlveoli-Pressure(cmH2O)                            |[1028.2, 1032.4] @cite otis1947measurement                                |Minimum of 1032.8     |<span class="success">0%</span>                |Assume  P<sub>atm</sub> = 1033.2 cmH2O     |
|LeftAlveoli-Oxygen-PartialPressure(mmHg)               |[98.0 @cite silverthorn2013human, 104.0 @cite Levitzky2013pulmonary]      |Mean of 118.2         |<span class="warning">13.7%</span>             |                                           |
|LeftAlveoli-CarbonDioxide-PartialPressure(mmHg)        |[40.0 @cite silverthorn2013human, 54.4 @cite lloyd1958relation]           |Mean of 38.9          |<span class="success">-2.6%</span>             |                                           |
|RightAlveoli-Volume(L/kg)                              |0.0154 @cite Levitzky2013pulmonary                                        |Minimum of 0.0137     |<span class="warning">-11.3%</span>            |                                           |
|RightAlveoli-Volume(L/kg)                              |0.0193 @cite Levitzky2013pulmonary                                        |Maximum of 0.0160     |<span class="warning">-17.1%</span>            |                                           |
|RightAlveoli-Pressure(cmH2O)                           |[1034.0, 1038.2] @cite otis1947measurement                                |Maximum of 1033.7     |<span class="success">-0%</span>               |Assume  P<sub>atm</sub> = 1033.2 cmH2O     |
|RightAlveoli-Pressure(cmH2O)                           |[1028.2, 1032.4] @cite otis1947measurement                                |Minimum of 1032.8     |<span class="success">0%</span>                |Assume  P<sub>atm</sub> = 1033.2 cmH2O     |
|RightAlveoli-Oxygen-PartialPressure(mmHg)              |[98.0 @cite silverthorn2013human, 104.0 @cite Levitzky2013pulmonary]      |Mean of 116.1         |<span class="warning">11.6%</span>             |                                           |
|RightAlveoli-CarbonDioxide-PartialPressure(mmHg)       |[40.0 @cite silverthorn2013human, 54.4 @cite lloyd1958relation]           |Mean of 39.7          |<span class="success">-0.7%</span>             |                                           |
|LeftDeadSpace-Volume(L/kg)                             |0.00090 @cite Levitzky2013pulmonary                                       |Mean of 0.00234       |<span class="danger">159.6%</span>             |                                           |
|LeftDeadSpace-Oxygen-PartialPressure(mmHg)             |[104.0, 149.0] @cite Levitzky2013pulmonary                                |Mean of 137.7         |<span class="success">Within bounds</span>     |                                           |
|LeftDeadSpace-CarbonDioxide-PartialPressure(mmHg)      |[0.03, 40.0] @cite Levitzky2013pulmonary                                  |Mean of 20.8          |<span class="success">Within bounds</span>     |                                           |
|RightDeadSpace-Volume(L/kg)                            |0.0011 @cite Levitzky2013pulmonary                                        |Mean of 0.00234       |<span class="danger">112.5%</span>             |                                           |
|RightDeadSpace-Oxygen-PartialPressure(mmHg)            |[104.0, 149.0] @cite Levitzky2013pulmonary                                |Mean of 136.7         |<span class="success">Within bounds</span>     |                                           |
|RightDeadSpace-CarbonDioxide-PartialPressure(mmHg)     |[0.03, 40.0] @cite Levitzky2013pulmonary                                  |Mean of 21.1          |<span class="success">Within bounds</span>     |                                           |
<table class="doxtable">
<tr>
<th>Property Name </th><th>Expected Value </th><th>Engine Value </th><th>Percent Error </th><th>Notes  </th></tr>
<tr>
<td>BasalMetabolicRate(kcal/day) </td><td>1737.3 </td><td>Mean of 1737.0 </td><td><span class="success">-0%</span> </td><td></td></tr>
<tr>
<td>BloodVolumeBaseline(mL) </td><td>5517.7 </td><td>Mean of 5566.0 </td><td><span class="success">0.9%</span> </td><td></td></tr>
<tr>
<td>DiastolicArterialPressureBaseline(mmHg) </td><td>73.5 </td><td>Mean of 72.9 </td><td><span class="success">-0.9%</span> </td><td></td></tr>
<tr>
<td>HeartRateBaseline(1/min) </td><td>72 </td><td>Mean of 73 </td><td><span class="success">1.4%</span> </td><td></td></tr>
<tr>
<td>MeanArterialPressureBaseline(mmHg) </td><td>87.0 </td><td>Mean of 95.0 </td><td><span class="success">9.2%</span> </td><td></td></tr>
<tr>
<td>RespirationRateBaseline(1/min) </td><td>16 </td><td>Mean of 16 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>SystolicArterialPressureBaseline(mmHg) </td><td>114.0 </td><td>Mean of 114.5 </td><td><span class="success">0.4%</span> </td><td></td></tr>
<tr>
<td>TidalVolumeBaseline(L) </td><td>0.540 </td><td>Mean of 0.497 </td><td><span class="success">-8%</span> </td><td></td></tr>
<tr>
<td>ExpiratoryReserveVolume(L) </td><td>1.080 </td><td>Mean of 1.138 </td><td><span class="success">5.4%</span> </td><td></td></tr>
<tr>
<td>FunctionalResidualCapacity(L) </td><td>2.313 </td><td>Mean of 2.372 </td><td><span class="success">2.6%</span> </td><td></td></tr>
<tr>
<td>InspiratoryCapacity(L) </td><td>3.856 </td><td>Mean of 3.797 </td><td><span class="success">-1.5%</span> </td><td></td></tr>
<tr>
<td>InspiratoryReserveVolume(L) </td><td>3.316 </td><td>Mean of 3.300 </td><td><span class="success">-0.5%</span> </td><td></td></tr>
<tr>
<td>ResidualVolume(L) </td><td>1.234 </td><td>Mean of 1.234 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>TotalLungCapacity(L) </td><td>6.169 </td><td>Mean of 6.169 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>VitalCapacity(L) </td><td>4.935 </td><td>Mean of 4.935 </td><td><span class="success">0%</span> </td><td></td></tr>
</table><table class="doxtable">
<tr>
<th>Property Name </th><th>Expected Value </th><th>Engine Value </th><th>Percent Error </th><th>Notes  </th></tr>
<tr>
<td>BasalMetabolicRate(kcal/day) </td><td>1862.2 </td><td>Mean of 1862.0 </td><td><span class="success">-0%</span> </td><td></td></tr>
<tr>
<td>BloodVolumeBaseline(mL) </td><td>5517.7 </td><td>Mean of 5569.0 </td><td><span class="success">0.9%</span> </td><td></td></tr>
<tr>
<td>DiastolicArterialPressureBaseline(mmHg) </td><td>73.5 </td><td>Mean of 73.5 </td><td><span class="success">-0.1%</span> </td><td></td></tr>
<tr>
<td>HeartRateBaseline(1/min) </td><td>84 </td><td>Mean of 86 </td><td><span class="success">2.4%</span> </td><td></td></tr>
<tr>
<td>MeanArterialPressureBaseline(mmHg) </td><td>87.0 </td><td>Mean of 94.9 </td><td><span class="success">9.1%</span> </td><td></td></tr>
<tr>
<td>RespirationRateBaseline(1/min) </td><td>16 </td><td>Mean of 16 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>SystolicArterialPressureBaseline(mmHg) </td><td>114.0 </td><td>Mean of 114.3 </td><td><span class="success">0.2%</span> </td><td></td></tr>
<tr>
<td>TidalVolumeBaseline(L) </td><td>0.540 </td><td>Mean of 0.511 </td><td><span class="success">-5.3%</span> </td><td></td></tr>
<tr>
<td>ExpiratoryReserveVolume(L) </td><td>1.080 </td><td>Mean of 1.139 </td><td><span class="success">5.5%</span> </td><td></td></tr>
<tr>
<td>FunctionalResidualCapacity(L) </td><td>2.313 </td><td>Mean of 2.372 </td><td><span class="success">2.6%</span> </td><td></td></tr>
<tr>
<td>InspiratoryCapacity(L) </td><td>3.856 </td><td>Mean of 3.797 </td><td><span class="success">-1.5%</span> </td><td></td></tr>
<tr>
<td>InspiratoryReserveVolume(L) </td><td>3.316 </td><td>Mean of 3.285 </td><td><span class="success">-0.9%</span> </td><td></td></tr>
<tr>
<td>ResidualVolume(L) </td><td>1.234 </td><td>Mean of 1.234 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>TotalLungCapacity(L) </td><td>6.169 </td><td>Mean of 6.169 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>VitalCapacity(L) </td><td>4.935 </td><td>Mean of 4.935 </td><td><span class="success">0%</span> </td><td></td></tr>
</table><table class="doxtable">
<tr>
<th>Property Name </th><th>Expected Value </th><th>Engine Value </th><th>Percent Error </th><th>Notes  </th></tr>
<tr>
<td>BasalMetabolicRate(kcal/day) </td><td>1296.4 </td><td>Mean of 1296.0 </td><td><span class="success">-0%</span> </td><td></td></tr>
<tr>
<td>BloodVolumeBaseline(mL) </td><td>4111.3 </td><td>Mean of 4133.1 </td><td><span class="success">0.5%</span> </td><td></td></tr>
<tr>
<td>DiastolicArterialPressureBaseline(mmHg) </td><td>73.5 </td><td>Mean of 73.8 </td><td><span class="success">0.4%</span> </td><td></td></tr>
<tr>
<td>HeartRateBaseline(1/min) </td><td>72 </td><td>Mean of 73 </td><td><span class="success">1.4%</span> </td><td></td></tr>
<tr>
<td>MeanArterialPressureBaseline(mmHg) </td><td>87.0 </td><td>Mean of 95.2 </td><td><span class="success">9.5%</span> </td><td></td></tr>
<tr>
<td>RespirationRateBaseline(1/min) </td><td>16 </td><td>Mean of 15 </td><td><span class="success">-6.2%</span> </td><td></td></tr>
<tr>
<td>SystolicArterialPressureBaseline(mmHg) </td><td>114.0 </td><td>Mean of 114.0 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>TidalVolumeBaseline(L) </td><td>0.405 </td><td>Mean of 0.392 </td><td><span class="success">-3.2%</span> </td><td></td></tr>
<tr>
<td>ExpiratoryReserveVolume(L) </td><td>0.809 </td><td>Mean of 0.816 </td><td><span class="success">0.9%</span> </td><td></td></tr>
<tr>
<td>FunctionalResidualCapacity(L) </td><td>1.734 </td><td>Mean of 1.740 </td><td><span class="success">0.3%</span> </td><td></td></tr>
<tr>
<td>InspiratoryCapacity(L) </td><td>2.889 </td><td>Mean of 2.883 </td><td><span class="success">-0.2%</span> </td><td></td></tr>
<tr>
<td>InspiratoryReserveVolume(L) </td><td>2.485 </td><td>Mean of 2.491 </td><td><span class="success">0.2%</span> </td><td></td></tr>
<tr>
<td>ResidualVolume(L) </td><td>0.925 </td><td>Mean of 0.925 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>TotalLungCapacity(L) </td><td>4.623 </td><td>Mean of 4.623 </td><td><span class="success">-0%</span> </td><td></td></tr>
<tr>
<td>VitalCapacity(L) </td><td>3.698 </td><td>Mean of 3.698 </td><td><span class="success">0%</span> </td><td></td></tr>
</table><table class="doxtable">
<tr>
<th>Property Name </th><th>Expected Value </th><th>Engine Value </th><th>Percent Error </th><th>Notes  </th></tr>
<tr>
<td>BasalMetabolicRate(kcal/day) </td><td>1366.6 </td><td>Mean of 1367.0 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>BloodVolumeBaseline(mL) </td><td>4196.9 </td><td>Mean of 4222.9 </td><td><span class="success">0.6%</span> </td><td></td></tr>
<tr>
<td>DiastolicArterialPressureBaseline(mmHg) </td><td>73.5 </td><td>Mean of 74.1 </td><td><span class="success">0.8%</span> </td><td></td></tr>
<tr>
<td>HeartRateBaseline(1/min) </td><td>72 </td><td>Mean of 73 </td><td><span class="success">1.4%</span> </td><td></td></tr>
<tr>
<td>MeanArterialPressureBaseline(mmHg) </td><td>87.0 </td><td>Mean of 95.0 </td><td><span class="success">9.2%</span> </td><td></td></tr>
<tr>
<td>RespirationRateBaseline(1/min) </td><td>18 </td><td>Mean of 17 </td><td><span class="success">-4%</span> </td><td></td></tr>
<tr>
<td>SystolicArterialPressureBaseline(mmHg) </td><td>114.0 </td><td>Mean of 113.3 </td><td><span class="success">-0.6%</span> </td><td></td></tr>
<tr>
<td>TidalVolumeBaseline(L) </td><td>0.413 </td><td>Mean of 0.379 </td><td><span class="success">-8.2%</span> </td><td></td></tr>
<tr>
<td>ExpiratoryReserveVolume(L) </td><td>0.826 </td><td>Mean of 0.839 </td><td><span class="success">1.6%</span> </td><td></td></tr>
<tr>
<td>FunctionalResidualCapacity(L) </td><td>1.769 </td><td>Mean of 1.782 </td><td><span class="success">0.7%</span> </td><td></td></tr>
<tr>
<td>InspiratoryCapacity(L) </td><td>2.948 </td><td>Mean of 2.935 </td><td><span class="success">-0.4%</span> </td><td></td></tr>
<tr>
<td>InspiratoryReserveVolume(L) </td><td>2.536 </td><td>Mean of 2.552 </td><td><span class="success">0.6%</span> </td><td></td></tr>
<tr>
<td>ResidualVolume(L) </td><td>0.943 </td><td>Mean of 0.943 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>TotalLungCapacity(L) </td><td>4.717 </td><td>Mean of 4.717 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>VitalCapacity(L) </td><td>3.774 </td><td>Mean of 3.774 </td><td><span class="success">0%</span> </td><td></td></tr>
</table><table class="doxtable">
<tr>
<th>Property Name </th><th>Expected Value </th><th>Engine Value </th><th>Percent Error </th><th>Notes  </th></tr>
<tr>
<td>BasalMetabolicRate(kcal/day) </td><td>1737.3 </td><td>Mean of 1737.0 </td><td><span class="success">-0%</span> </td><td></td></tr>
<tr>
<td>BloodVolumeBaseline(mL) </td><td>5517.7 </td><td>Mean of 5559.6 </td><td><span class="success">0.8%</span> </td><td></td></tr>
<tr>
<td>DiastolicArterialPressureBaseline(mmHg) </td><td>73.5 </td><td>Mean of 73.5 </td><td><span class="success">-0%</span> </td><td></td></tr>
<tr>
<td>HeartRateBaseline(1/min) </td><td>50 </td><td>Mean of 51 </td><td><span class="success">2%</span> </td><td></td></tr>
<tr>
<td>MeanArterialPressureBaseline(mmHg) </td><td>87.0 </td><td>Mean of 95.3 </td><td><span class="success">9.5%</span> </td><td></td></tr>
<tr>
<td>RespirationRateBaseline(1/min) </td><td>12 </td><td>Mean of 12 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>SystolicArterialPressureBaseline(mmHg) </td><td>114.0 </td><td>Mean of 114.5 </td><td><span class="success">0.5%</span> </td><td></td></tr>
<tr>
<td>TidalVolumeBaseline(L) </td><td>0.540 </td><td>Mean of 0.629 </td><td><span class="warning">16.6%</span> </td><td></td></tr>
<tr>
<td>ExpiratoryReserveVolume(L) </td><td>1.080 </td><td>Mean of 1.137 </td><td><span class="success">5.3%</span> </td><td></td></tr>
<tr>
<td>FunctionalResidualCapacity(L) </td><td>2.313 </td><td>Mean of 2.371 </td><td><span class="success">2.5%</span> </td><td></td></tr>
<tr>
<td>InspiratoryCapacity(L) </td><td>3.856 </td><td>Mean of 3.798 </td><td><span class="success">-1.5%</span> </td><td></td></tr>
<tr>
<td>InspiratoryReserveVolume(L) </td><td>3.316 </td><td>Mean of 3.167 </td><td><span class="success">-4.5%</span> </td><td></td></tr>
<tr>
<td>ResidualVolume(L) </td><td>1.234 </td><td>Mean of 1.234 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>TotalLungCapacity(L) </td><td>6.169 </td><td>Mean of 6.169 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>VitalCapacity(L) </td><td>4.935 </td><td>Mean of 4.935 </td><td><span class="success">0%</span> </td><td></td></tr>
</table><table class="doxtable">
<tr>
<th>Property Name </th><th>Expected Value </th><th>Engine Value </th><th>Percent Error </th><th>Notes  </th></tr>
<tr>
<td>BasalMetabolicRate(kcal/day) </td><td>1172.9 </td><td>Mean of 1173.0 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>BloodVolumeBaseline(mL) </td><td>2891.1 </td><td>Mean of 3519.9 </td><td><span class="warning">21.8%</span> </td><td></td></tr>
<tr>
<td>DiastolicArterialPressureBaseline(mmHg) </td><td>60.0 </td><td>Mean of 61.4 </td><td><span class="success">2.4%</span> </td><td></td></tr>
<tr>
<td>HeartRateBaseline(1/min) </td><td>60 </td><td>Mean of 61 </td><td><span class="success">1.7%</span> </td><td></td></tr>
<tr>
<td>MeanArterialPressureBaseline(mmHg) </td><td>70.0 </td><td>Mean of 77.7 </td><td><span class="warning">11%</span> </td><td></td></tr>
<tr>
<td>RespirationRateBaseline(1/min) </td><td>12 </td><td>Mean of 12 </td><td><span class="success">-1.2%</span> </td><td></td></tr>
<tr>
<td>SystolicArterialPressureBaseline(mmHg) </td><td>90.0 </td><td>Mean of 91.5 </td><td><span class="success">1.7%</span> </td><td></td></tr>
<tr>
<td>TidalVolumeBaseline(L) </td><td>0.286 </td><td>Mean of 0.359 </td><td><span class="warning">25.5%</span> </td><td></td></tr>
<tr>
<td>ExpiratoryReserveVolume(L) </td><td>0.573 </td><td>Mean of 0.601 </td><td><span class="success">4.9%</span> </td><td></td></tr>
<tr>
<td>FunctionalResidualCapacity(L) </td><td>1.228 </td><td>Mean of 1.256 </td><td><span class="success">2.3%</span> </td><td></td></tr>
<tr>
<td>InspiratoryCapacity(L) </td><td>2.046 </td><td>Mean of 2.018 </td><td><span class="success">-1.4%</span> </td><td></td></tr>
<tr>
<td>InspiratoryReserveVolume(L) </td><td>1.759 </td><td>Mean of 1.650 </td><td><span class="success">-6.2%</span> </td><td></td></tr>
<tr>
<td>ResidualVolume(L) </td><td>0.655 </td><td>Mean of 0.655 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>TotalLungCapacity(L) </td><td>3.273 </td><td>Mean of 3.273 </td><td><span class="success">-0%</span> </td><td></td></tr>
<tr>
<td>VitalCapacity(L) </td><td>2.619 </td><td>Mean of 2.619 </td><td><span class="success">0%</span> </td><td></td></tr>
</table><table class="doxtable">
<tr>
<th>Property Name </th><th>Expected Value </th><th>Engine Value </th><th>Percent Error </th><th>Notes  </th></tr>
<tr>
<td>BasalMetabolicRate(kcal/day) </td><td>1737.3 </td><td>Mean of 1737.0 </td><td><span class="success">-0%</span> </td><td></td></tr>
<tr>
<td>BloodVolumeBaseline(mL) </td><td>5517.7 </td><td>Mean of 5269.6 </td><td><span class="success">-4.5%</span> </td><td></td></tr>
<tr>
<td>DiastolicArterialPressureBaseline(mmHg) </td><td>80.0 </td><td>Mean of 79.1 </td><td><span class="success">-1.1%</span> </td><td></td></tr>
<tr>
<td>HeartRateBaseline(1/min) </td><td>110 </td><td>Mean of 111 </td><td><span class="success">1.1%</span> </td><td></td></tr>
<tr>
<td>MeanArterialPressureBaseline(mmHg) </td><td>93.3 </td><td>Mean of 100.1 </td><td><span class="success">7.2%</span> </td><td></td></tr>
<tr>
<td>RespirationRateBaseline(1/min) </td><td>15 </td><td>Mean of 15 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>SystolicArterialPressureBaseline(mmHg) </td><td>120.0 </td><td>Mean of 118.8 </td><td><span class="success">-1%</span> </td><td></td></tr>
<tr>
<td>TidalVolumeBaseline(L) </td><td>0.540 </td><td>Mean of 0.520 </td><td><span class="success">-3.6%</span> </td><td></td></tr>
<tr>
<td>ExpiratoryReserveVolume(L) </td><td>1.080 </td><td>Mean of 1.138 </td><td><span class="success">5.4%</span> </td><td></td></tr>
<tr>
<td>FunctionalResidualCapacity(L) </td><td>2.313 </td><td>Mean of 2.372 </td><td><span class="success">2.6%</span> </td><td></td></tr>
<tr>
<td>InspiratoryCapacity(L) </td><td>3.856 </td><td>Mean of 3.797 </td><td><span class="success">-1.5%</span> </td><td></td></tr>
<tr>
<td>InspiratoryReserveVolume(L) </td><td>3.316 </td><td>Mean of 3.276 </td><td><span class="success">-1.2%</span> </td><td></td></tr>
<tr>
<td>ResidualVolume(L) </td><td>1.234 </td><td>Mean of 1.234 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>TotalLungCapacity(L) </td><td>6.169 </td><td>Mean of 6.169 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>VitalCapacity(L) </td><td>4.935 </td><td>Mean of 4.935 </td><td><span class="success">0%</span> </td><td></td></tr>
</table><table class="doxtable">
<tr>
<th>Property Name </th><th>Expected Value </th><th>Engine Value </th><th>Percent Error </th><th>Notes  </th></tr>
<tr>
<td>BasalMetabolicRate(kcal/day) </td><td>1306.0 </td><td>Mean of 1306.0 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>BloodVolumeBaseline(mL) </td><td>4196.9 </td><td>Mean of 4216.2 </td><td><span class="success">0.5%</span> </td><td></td></tr>
<tr>
<td>DiastolicArterialPressureBaseline(mmHg) </td><td>73.5 </td><td>Mean of 74.6 </td><td><span class="success">1.5%</span> </td><td></td></tr>
<tr>
<td>HeartRateBaseline(1/min) </td><td>72 </td><td>Mean of 73 </td><td><span class="success">1.4%</span> </td><td></td></tr>
<tr>
<td>MeanArterialPressureBaseline(mmHg) </td><td>87.0 </td><td>Mean of 95.2 </td><td><span class="success">9.5%</span> </td><td></td></tr>
<tr>
<td>RespirationRateBaseline(1/min) </td><td>16 </td><td>Mean of 15 </td><td><span class="success">-6.2%</span> </td><td></td></tr>
<tr>
<td>SystolicArterialPressureBaseline(mmHg) </td><td>114.0 </td><td>Mean of 113.3 </td><td><span class="success">-0.6%</span> </td><td></td></tr>
<tr>
<td>TidalVolumeBaseline(L) </td><td>0.413 </td><td>Mean of 0.396 </td><td><span class="success">-4%</span> </td><td></td></tr>
<tr>
<td>ExpiratoryReserveVolume(L) </td><td>0.826 </td><td>Mean of 0.838 </td><td><span class="success">1.5%</span> </td><td></td></tr>
<tr>
<td>FunctionalResidualCapacity(L) </td><td>1.769 </td><td>Mean of 1.781 </td><td><span class="success">0.7%</span> </td><td></td></tr>
<tr>
<td>InspiratoryCapacity(L) </td><td>2.948 </td><td>Mean of 2.936 </td><td><span class="success">-0.4%</span> </td><td></td></tr>
<tr>
<td>InspiratoryReserveVolume(L) </td><td>2.536 </td><td>Mean of 2.539 </td><td><span class="success">0.1%</span> </td><td></td></tr>
<tr>
<td>ResidualVolume(L) </td><td>0.943 </td><td>Mean of 0.943 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>TotalLungCapacity(L) </td><td>4.717 </td><td>Mean of 4.717 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>VitalCapacity(L) </td><td>3.774 </td><td>Mean of 3.774 </td><td><span class="success">0%</span> </td><td></td></tr>
</table><table class="doxtable">
<tr>
<th>Property Name </th><th>Expected Value </th><th>Engine Value </th><th>Percent Error </th><th>Notes  </th></tr>
<tr>
<td>BasalMetabolicRate(kcal/day) </td><td>1376.6 </td><td>Mean of 1377.0 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>BloodVolumeBaseline(mL) </td><td>3867.8 </td><td>Mean of 3903.7 </td><td><span class="success">0.9%</span> </td><td></td></tr>
<tr>
<td>DiastolicArterialPressureBaseline(mmHg) </td><td>73.5 </td><td>Mean of 73.2 </td><td><span class="success">-0.4%</span> </td><td></td></tr>
<tr>
<td>HeartRateBaseline(1/min) </td><td>72 </td><td>Mean of 73 </td><td><span class="success">1.4%</span> </td><td></td></tr>
<tr>
<td>MeanArterialPressureBaseline(mmHg) </td><td>87.0 </td><td>Mean of 94.8 </td><td><span class="success">9%</span> </td><td></td></tr>
<tr>
<td>RespirationRateBaseline(1/min) </td><td>18 </td><td>Mean of 17 </td><td><span class="success">-3.4%</span> </td><td></td></tr>
<tr>
<td>SystolicArterialPressureBaseline(mmHg) </td><td>114.0 </td><td>Mean of 113.8 </td><td><span class="success">-0.1%</span> </td><td></td></tr>
<tr>
<td>TidalVolumeBaseline(L) </td><td>0.381 </td><td>Mean of 0.365 </td><td><span class="success">-4.2%</span> </td><td></td></tr>
<tr>
<td>ExpiratoryReserveVolume(L) </td><td>0.762 </td><td>Mean of 0.757 </td><td><span class="success">-0.7%</span> </td><td></td></tr>
<tr>
<td>FunctionalResidualCapacity(L) </td><td>1.633 </td><td>Mean of 1.628 </td><td><span class="success">-0.3%</span> </td><td></td></tr>
<tr>
<td>InspiratoryCapacity(L) </td><td>2.722 </td><td>Mean of 2.727 </td><td><span class="success">0.2%</span> </td><td></td></tr>
<tr>
<td>InspiratoryReserveVolume(L) </td><td>2.341 </td><td>Mean of 2.359 </td><td><span class="success">0.8%</span> </td><td></td></tr>
<tr>
<td>ResidualVolume(L) </td><td>0.871 </td><td>Mean of 0.871 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>TotalLungCapacity(L) </td><td>4.354 </td><td>Mean of 4.354 </td><td><span class="success">-0%</span> </td><td></td></tr>
<tr>
<td>VitalCapacity(L) </td><td>3.484 </td><td>Mean of 3.484 </td><td><span class="success">0%</span> </td><td></td></tr>
</table><table class="doxtable">
<tr>
<th>Property Name </th><th>Expected Value </th><th>Engine Value </th><th>Percent Error </th><th>Notes  </th></tr>
<tr>
<td>BasalMetabolicRate(kcal/day) </td><td>1931.5 </td><td>Mean of 1932.0 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>BloodVolumeBaseline(mL) </td><td>6014.8 </td><td>Mean of 7398.9 </td><td><span class="warning">23%</span> </td><td></td></tr>
<tr>
<td>DiastolicArterialPressureBaseline(mmHg) </td><td>60.0 </td><td>Mean of 60.7 </td><td><span class="success">1.1%</span> </td><td></td></tr>
<tr>
<td>HeartRateBaseline(1/min) </td><td>110 </td><td>Mean of 111 </td><td><span class="success">0.9%</span> </td><td></td></tr>
<tr>
<td>MeanArterialPressureBaseline(mmHg) </td><td>70.0 </td><td>Mean of 76.5 </td><td><span class="success">9.3%</span> </td><td></td></tr>
<tr>
<td>RespirationRateBaseline(1/min) </td><td>18 </td><td>Mean of 18 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>SystolicArterialPressureBaseline(mmHg) </td><td>90.0 </td><td>Mean of 90.3 </td><td><span class="success">0.4%</span> </td><td></td></tr>
<tr>
<td>TidalVolumeBaseline(L) </td><td>0.587 </td><td>Mean of 0.491 </td><td><span class="warning">-16.4%</span> </td><td></td></tr>
<tr>
<td>ExpiratoryReserveVolume(L) </td><td>1.175 </td><td>Mean of 1.215 </td><td><span class="success">3.4%</span> </td><td></td></tr>
<tr>
<td>FunctionalResidualCapacity(L) </td><td>2.517 </td><td>Mean of 2.558 </td><td><span class="success">1.6%</span> </td><td></td></tr>
<tr>
<td>InspiratoryCapacity(L) </td><td>4.196 </td><td>Mean of 4.156 </td><td><span class="success">-1%</span> </td><td></td></tr>
<tr>
<td>InspiratoryReserveVolume(L) </td><td>3.608 </td><td>Mean of 3.663 </td><td><span class="success">1.5%</span> </td><td></td></tr>
<tr>
<td>ResidualVolume(L) </td><td>1.343 </td><td>Mean of 1.343 </td><td><span class="success">-0%</span> </td><td></td></tr>
<tr>
<td>TotalLungCapacity(L) </td><td>6.713 </td><td>Mean of 6.713 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>VitalCapacity(L) </td><td>5.371 </td><td>Mean of 5.371 </td><td><span class="success">-0%</span> </td><td></td></tr>
</table><table class="doxtable">
<tr>
<th>Property Name </th><th>Expected Value </th><th>Engine Value </th><th>Percent Error </th><th>Notes  </th></tr>
<tr>
<td>BasalMetabolicRate(kcal/day) </td><td>1403.1 </td><td>Mean of 1403.0 </td><td><span class="success">-0%</span> </td><td></td></tr>
<tr>
<td>BloodVolumeBaseline(mL) </td><td>3703.5 </td><td>Mean of 3765.8 </td><td><span class="success">1.7%</span> </td><td></td></tr>
<tr>
<td>DiastolicArterialPressureBaseline(mmHg) </td><td>73.5 </td><td>Mean of 73.4 </td><td><span class="success">-0.1%</span> </td><td></td></tr>
<tr>
<td>HeartRateBaseline(1/min) </td><td>72 </td><td>Mean of 73 </td><td><span class="success">1.4%</span> </td><td></td></tr>
<tr>
<td>MeanArterialPressureBaseline(mmHg) </td><td>87.0 </td><td>Mean of 94.9 </td><td><span class="success">9%</span> </td><td></td></tr>
<tr>
<td>RespirationRateBaseline(1/min) </td><td>16 </td><td>Mean of 15 </td><td><span class="success">-6.2%</span> </td><td></td></tr>
<tr>
<td>SystolicArterialPressureBaseline(mmHg) </td><td>114.0 </td><td>Mean of 113.6 </td><td><span class="success">-0.3%</span> </td><td></td></tr>
<tr>
<td>TidalVolumeBaseline(L) </td><td>0.365 </td><td>Mean of 0.399 </td><td><span class="success">9.4%</span> </td><td></td></tr>
<tr>
<td>ExpiratoryReserveVolume(L) </td><td>0.730 </td><td>Mean of 0.719 </td><td><span class="success">-1.5%</span> </td><td></td></tr>
<tr>
<td>FunctionalResidualCapacity(L) </td><td>1.565 </td><td>Mean of 1.553 </td><td><span class="success">-0.8%</span> </td><td></td></tr>
<tr>
<td>InspiratoryCapacity(L) </td><td>2.608 </td><td>Mean of 2.620 </td><td><span class="success">0.5%</span> </td><td></td></tr>
<tr>
<td>InspiratoryReserveVolume(L) </td><td>2.243 </td><td>Mean of 2.221 </td><td><span class="success">-1%</span> </td><td></td></tr>
<tr>
<td>ResidualVolume(L) </td><td>0.835 </td><td>Mean of 0.835 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>TotalLungCapacity(L) </td><td>4.173 </td><td>Mean of 4.173 </td><td><span class="success">-0%</span> </td><td></td></tr>
<tr>
<td>VitalCapacity(L) </td><td>3.338 </td><td>Mean of 3.338 </td><td><span class="success">-0%</span> </td><td></td></tr>
</table><table class="doxtable">
<tr>
<th>Property Name </th><th>Expected Value </th><th>Engine Value </th><th>Percent Error </th><th>Notes  </th></tr>
<tr>
<td>BasalMetabolicRate(kcal/day) </td><td>1918.2 </td><td>Mean of 1918.0 </td><td><span class="success">-0%</span> </td><td></td></tr>
<tr>
<td>BloodVolumeBaseline(mL) </td><td>5849.0 </td><td>Mean of 5904.2 </td><td><span class="success">0.9%</span> </td><td></td></tr>
<tr>
<td>DiastolicArterialPressureBaseline(mmHg) </td><td>73.5 </td><td>Mean of 73.2 </td><td><span class="success">-0.5%</span> </td><td></td></tr>
<tr>
<td>HeartRateBaseline(1/min) </td><td>72 </td><td>Mean of 73 </td><td><span class="success">1.4%</span> </td><td></td></tr>
<tr>
<td>MeanArterialPressureBaseline(mmHg) </td><td>87.0 </td><td>Mean of 94.8 </td><td><span class="success">9%</span> </td><td></td></tr>
<tr>
<td>RespirationRateBaseline(1/min) </td><td>16 </td><td>Mean of 16 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>SystolicArterialPressureBaseline(mmHg) </td><td>114.0 </td><td>Mean of 113.9 </td><td><span class="success">-0.1%</span> </td><td></td></tr>
<tr>
<td>TidalVolumeBaseline(L) </td><td>0.572 </td><td>Mean of 0.535 </td><td><span class="success">-6.3%</span> </td><td></td></tr>
<tr>
<td>ExpiratoryReserveVolume(L) </td><td>1.143 </td><td>Mean of 1.191 </td><td><span class="success">4.2%</span> </td><td></td></tr>
<tr>
<td>FunctionalResidualCapacity(L) </td><td>2.449 </td><td>Mean of 2.497 </td><td><span class="success">2%</span> </td><td></td></tr>
<tr>
<td>InspiratoryCapacity(L) </td><td>4.082 </td><td>Mean of 4.035 </td><td><span class="success">-1.2%</span> </td><td></td></tr>
<tr>
<td>InspiratoryReserveVolume(L) </td><td>3.511 </td><td>Mean of 3.498 </td><td><span class="success">-0.4%</span> </td><td></td></tr>
<tr>
<td>ResidualVolume(L) </td><td>1.306 </td><td>Mean of 1.306 </td><td><span class="success">-0%</span> </td><td></td></tr>
<tr>
<td>TotalLungCapacity(L) </td><td>6.532 </td><td>Mean of 6.532 </td><td><span class="success">-0%</span> </td><td></td></tr>
<tr>
<td>VitalCapacity(L) </td><td>5.225 </td><td>Mean of 5.225 </td><td><span class="success">-0%</span> </td><td></td></tr>
</table><table class="doxtable">
<tr>
<th>Property Name </th><th>Expected Value </th><th>Engine Value </th><th>Percent Error </th><th>Notes  </th></tr>
<tr>
<td>BasalMetabolicRate(kcal/day) </td><td>1737.3 </td><td>Mean of 1737.0 </td><td><span class="success">-0%</span> </td><td></td></tr>
<tr>
<td>BloodVolumeBaseline(mL) </td><td>5517.7 </td><td>Mean of 5544.2 </td><td><span class="success">0.5%</span> </td><td></td></tr>
<tr>
<td>DiastolicArterialPressureBaseline(mmHg) </td><td>73.5 </td><td>Mean of 74.1 </td><td><span class="success">0.7%</span> </td><td></td></tr>
<tr>
<td>HeartRateBaseline(1/min) </td><td>110 </td><td>Mean of 111 </td><td><span class="success">0.9%</span> </td><td></td></tr>
<tr>
<td>MeanArterialPressureBaseline(mmHg) </td><td>87.0 </td><td>Mean of 95.6 </td><td><span class="success">9.9%</span> </td><td></td></tr>
<tr>
<td>RespirationRateBaseline(1/min) </td><td>20 </td><td>Mean of 20 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>SystolicArterialPressureBaseline(mmHg) </td><td>114.0 </td><td>Mean of 114.5 </td><td><span class="success">0.4%</span> </td><td></td></tr>
<tr>
<td>TidalVolumeBaseline(L) </td><td>0.540 </td><td>Mean of 0.405 </td><td><span class="warning">-25%</span> </td><td></td></tr>
<tr>
<td>ExpiratoryReserveVolume(L) </td><td>1.080 </td><td>Mean of 1.140 </td><td><span class="success">5.6%</span> </td><td></td></tr>
<tr>
<td>FunctionalResidualCapacity(L) </td><td>2.313 </td><td>Mean of 2.373 </td><td><span class="success">2.6%</span> </td><td></td></tr>
<tr>
<td>InspiratoryCapacity(L) </td><td>3.856 </td><td>Mean of 3.796 </td><td><span class="success">-1.6%</span> </td><td></td></tr>
<tr>
<td>InspiratoryReserveVolume(L) </td><td>3.316 </td><td>Mean of 3.390 </td><td><span class="success">2.2%</span> </td><td></td></tr>
<tr>
<td>ResidualVolume(L) </td><td>1.234 </td><td>Mean of 1.234 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>TotalLungCapacity(L) </td><td>6.169 </td><td>Mean of 6.169 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>VitalCapacity(L) </td><td>4.935 </td><td>Mean of 4.935 </td><td><span class="success">0%</span> </td><td></td></tr>
</table><table class="doxtable">
<tr>
<th>Property Name </th><th>Expected Value </th><th>Engine Value </th><th>Percent Error </th><th>Notes  </th></tr>
<tr>
<td>BasalMetabolicRate(kcal/day) </td><td>2070.0 </td><td>Mean of 2070.0 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>BloodVolumeBaseline(mL) </td><td>7120.4 </td><td>Mean of 6801.6 </td><td><span class="success">-4.5%</span> </td><td></td></tr>
<tr>
<td>DiastolicArterialPressureBaseline(mmHg) </td><td>80.0 </td><td>Mean of 79.3 </td><td><span class="success">-0.8%</span> </td><td></td></tr>
<tr>
<td>HeartRateBaseline(1/min) </td><td>100 </td><td>Mean of 103 </td><td><span class="success">3%</span> </td><td></td></tr>
<tr>
<td>MeanArterialPressureBaseline(mmHg) </td><td>93.3 </td><td>Mean of 100.1 </td><td><span class="success">7.3%</span> </td><td></td></tr>
<tr>
<td>RespirationRateBaseline(1/min) </td><td>20 </td><td>Mean of 20 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>SystolicArterialPressureBaseline(mmHg) </td><td>120.0 </td><td>Mean of 119.0 </td><td><span class="success">-0.9%</span> </td><td></td></tr>
<tr>
<td>TidalVolumeBaseline(L) </td><td>0.693 </td><td>Mean of 0.480 </td><td><span class="danger">-30.7%</span> </td><td></td></tr>
<tr>
<td>ExpiratoryReserveVolume(L) </td><td>1.386 </td><td>Mean of 1.361 </td><td><span class="success">-1.8%</span> </td><td></td></tr>
<tr>
<td>FunctionalResidualCapacity(L) </td><td>2.970 </td><td>Mean of 2.945 </td><td><span class="success">-0.8%</span> </td><td></td></tr>
<tr>
<td>InspiratoryCapacity(L) </td><td>4.951 </td><td>Mean of 4.976 </td><td><span class="success">0.5%</span> </td><td></td></tr>
<tr>
<td>InspiratoryReserveVolume(L) </td><td>4.257 </td><td>Mean of 4.494 </td><td><span class="success">5.6%</span> </td><td></td></tr>
<tr>
<td>ResidualVolume(L) </td><td>1.584 </td><td>Mean of 1.584 </td><td><span class="success">-0%</span> </td><td></td></tr>
<tr>
<td>TotalLungCapacity(L) </td><td>7.921 </td><td>Mean of 7.921 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>VitalCapacity(L) </td><td>6.337 </td><td>Mean of 6.337 </td><td><span class="success">-0%</span> </td><td></td></tr>
</table><table class="doxtable">
<tr>
<th>Property Name </th><th>Expected Value </th><th>Engine Value </th><th>Percent Error </th><th>Notes  </th></tr>
<tr>
<td>BasalMetabolicRate(kcal/day) </td><td>1601.1 </td><td>Mean of 1601.0 </td><td><span class="success">-0%</span> </td><td></td></tr>
<tr>
<td>BloodVolumeBaseline(mL) </td><td>4863.8 </td><td>Mean of 4922.5 </td><td><span class="success">1.2%</span> </td><td></td></tr>
<tr>
<td>DiastolicArterialPressureBaseline(mmHg) </td><td>73.5 </td><td>Mean of 73.5 </td><td><span class="success">-0%</span> </td><td></td></tr>
<tr>
<td>HeartRateBaseline(1/min) </td><td>72 </td><td>Mean of 73 </td><td><span class="success">1.4%</span> </td><td></td></tr>
<tr>
<td>MeanArterialPressureBaseline(mmHg) </td><td>87.0 </td><td>Mean of 94.8 </td><td><span class="success">9%</span> </td><td></td></tr>
<tr>
<td>RespirationRateBaseline(1/min) </td><td>16 </td><td>Mean of 15 </td><td><span class="success">-6.2%</span> </td><td></td></tr>
<tr>
<td>SystolicArterialPressureBaseline(mmHg) </td><td>114.0 </td><td>Mean of 113.4 </td><td><span class="success">-0.5%</span> </td><td></td></tr>
<tr>
<td>TidalVolumeBaseline(L) </td><td>0.477 </td><td>Mean of 0.455 </td><td><span class="success">-4.7%</span> </td><td></td></tr>
<tr>
<td>ExpiratoryReserveVolume(L) </td><td>0.954 </td><td>Mean of 1.006 </td><td><span class="success">5.5%</span> </td><td></td></tr>
<tr>
<td>FunctionalResidualCapacity(L) </td><td>2.044 </td><td>Mean of 2.097 </td><td><span class="success">2.6%</span> </td><td></td></tr>
<tr>
<td>InspiratoryCapacity(L) </td><td>3.407 </td><td>Mean of 3.355 </td><td><span class="success">-1.5%</span> </td><td></td></tr>
<tr>
<td>InspiratoryReserveVolume(L) </td><td>2.930 </td><td>Mean of 2.902 </td><td><span class="success">-1%</span> </td><td></td></tr>
<tr>
<td>ResidualVolume(L) </td><td>1.090 </td><td>Mean of 1.090 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>TotalLungCapacity(L) </td><td>5.451 </td><td>Mean of 5.451 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>VitalCapacity(L) </td><td>4.361 </td><td>Mean of 4.361 </td><td><span class="success">0%</span> </td><td></td></tr>
</table><table class="doxtable">
<tr>
<th>Property Name </th><th>Expected Value </th><th>Engine Value </th><th>Percent Error </th><th>Notes  </th></tr>
<tr>
<td>BasalMetabolicRate(kcal/day) </td><td>1464.8 </td><td>Mean of 1465.0 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>BloodVolumeBaseline(mL) </td><td>5186.9 </td><td>Mean of 5190.9 </td><td><span class="success">0.1%</span> </td><td></td></tr>
<tr>
<td>DiastolicArterialPressureBaseline(mmHg) </td><td>73.5 </td><td>Mean of 73.4 </td><td><span class="success">-0.1%</span> </td><td></td></tr>
<tr>
<td>HeartRateBaseline(1/min) </td><td>72 </td><td>Mean of 73 </td><td><span class="success">1.4%</span> </td><td></td></tr>
<tr>
<td>MeanArterialPressureBaseline(mmHg) </td><td>87.0 </td><td>Mean of 94.9 </td><td><span class="success">9.1%</span> </td><td></td></tr>
<tr>
<td>RespirationRateBaseline(1/min) </td><td>18 </td><td>Mean of 18 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>SystolicArterialPressureBaseline(mmHg) </td><td>114.0 </td><td>Mean of 113.8 </td><td><span class="success">-0.2%</span> </td><td></td></tr>
<tr>
<td>TidalVolumeBaseline(L) </td><td>0.508 </td><td>Mean of 0.399 </td><td><span class="warning">-21.5%</span> </td><td></td></tr>
<tr>
<td>ExpiratoryReserveVolume(L) </td><td>1.016 </td><td>Mean of 1.077 </td><td><span class="success">6%</span> </td><td></td></tr>
<tr>
<td>FunctionalResidualCapacity(L) </td><td>2.177 </td><td>Mean of 2.238 </td><td><span class="success">2.8%</span> </td><td></td></tr>
<tr>
<td>InspiratoryCapacity(L) </td><td>3.629 </td><td>Mean of 3.568 </td><td><span class="success">-1.7%</span> </td><td></td></tr>
<tr>
<td>InspiratoryReserveVolume(L) </td><td>3.121 </td><td>Mean of 3.167 </td><td><span class="success">1.5%</span> </td><td></td></tr>
<tr>
<td>ResidualVolume(L) </td><td>1.161 </td><td>Mean of 1.161 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>TotalLungCapacity(L) </td><td>5.806 </td><td>Mean of 5.806 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>VitalCapacity(L) </td><td>4.645 </td><td>Mean of 4.645 </td><td><span class="success">0%</span> </td><td></td></tr>
</table><table class="doxtable">
<tr>
<th>Property Name </th><th>Expected Value </th><th>Engine Value </th><th>Percent Error </th><th>Notes  </th></tr>
<tr>
<td>BasalMetabolicRate(kcal/day) </td><td>1914.8 </td><td>Mean of 1915.0 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>BloodVolumeBaseline(mL) </td><td>6180.6 </td><td>Mean of 7602.3 </td><td><span class="warning">23%</span> </td><td></td></tr>
<tr>
<td>DiastolicArterialPressureBaseline(mmHg) </td><td>60.0 </td><td>Mean of 60.7 </td><td><span class="success">1.2%</span> </td><td></td></tr>
<tr>
<td>HeartRateBaseline(1/min) </td><td>93 </td><td>Mean of 94 </td><td><span class="success">1.1%</span> </td><td></td></tr>
<tr>
<td>MeanArterialPressureBaseline(mmHg) </td><td>70.0 </td><td>Mean of 76.4 </td><td><span class="success">9.2%</span> </td><td></td></tr>
<tr>
<td>RespirationRateBaseline(1/min) </td><td>14 </td><td>Mean of 14 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>SystolicArterialPressureBaseline(mmHg) </td><td>90.0 </td><td>Mean of 90.4 </td><td><span class="success">0.4%</span> </td><td></td></tr>
<tr>
<td>TidalVolumeBaseline(L) </td><td>0.603 </td><td>Mean of 0.612 </td><td><span class="success">1.5%</span> </td><td></td></tr>
<tr>
<td>ExpiratoryReserveVolume(L) </td><td>1.207 </td><td>Mean of 1.235 </td><td><span class="success">2.3%</span> </td><td></td></tr>
<tr>
<td>FunctionalResidualCapacity(L) </td><td>2.585 </td><td>Mean of 2.614 </td><td><span class="success">1.1%</span> </td><td></td></tr>
<tr>
<td>InspiratoryCapacity(L) </td><td>4.309 </td><td>Mean of 4.280 </td><td><span class="success">-0.7%</span> </td><td></td></tr>
<tr>
<td>InspiratoryReserveVolume(L) </td><td>3.706 </td><td>Mean of 3.667 </td><td><span class="success">-1.1%</span> </td><td></td></tr>
<tr>
<td>ResidualVolume(L) </td><td>1.379 </td><td>Mean of 1.379 </td><td><span class="success">-0%</span> </td><td></td></tr>
<tr>
<td>TotalLungCapacity(L) </td><td>6.895 </td><td>Mean of 6.895 </td><td><span class="success">-0%</span> </td><td></td></tr>
<tr>
<td>VitalCapacity(L) </td><td>5.516 </td><td>Mean of 5.516 </td><td><span class="success">-0%</span> </td><td></td></tr>
</table><table class="doxtable">
<tr>
<th>Property Name </th><th>Expected Value </th><th>Engine Value </th><th>Percent Error </th><th>Notes  </th></tr>
<tr>
<td>BasalMetabolicRate(kcal/day) </td><td>1737.3 </td><td>Mean of 1737.0 </td><td><span class="success">-0%</span> </td><td></td></tr>
<tr>
<td>BloodVolumeBaseline(mL) </td><td>5517.7 </td><td>Mean of 5567.1 </td><td><span class="success">0.9%</span> </td><td></td></tr>
<tr>
<td>DiastolicArterialPressureBaseline(mmHg) </td><td>73.5 </td><td>Mean of 72.9 </td><td><span class="success">-0.8%</span> </td><td></td></tr>
<tr>
<td>HeartRateBaseline(1/min) </td><td>72 </td><td>Mean of 73 </td><td><span class="success">1.4%</span> </td><td></td></tr>
<tr>
<td>MeanArterialPressureBaseline(mmHg) </td><td>87.0 </td><td>Mean of 95.0 </td><td><span class="success">9.2%</span> </td><td></td></tr>
<tr>
<td>RespirationRateBaseline(1/min) </td><td>16 </td><td>Mean of 16 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>SystolicArterialPressureBaseline(mmHg) </td><td>114.0 </td><td>Mean of 114.4 </td><td><span class="success">0.4%</span> </td><td></td></tr>
<tr>
<td>TidalVolumeBaseline(L) </td><td>0.540 </td><td>Mean of 0.498 </td><td><span class="success">-7.8%</span> </td><td></td></tr>
<tr>
<td>ExpiratoryReserveVolume(L) </td><td>1.080 </td><td>Mean of 1.138 </td><td><span class="success">5.4%</span> </td><td></td></tr>
<tr>
<td>FunctionalResidualCapacity(L) </td><td>2.313 </td><td>Mean of 2.372 </td><td><span class="success">2.6%</span> </td><td></td></tr>
<tr>
<td>InspiratoryCapacity(L) </td><td>3.856 </td><td>Mean of 3.797 </td><td><span class="success">-1.5%</span> </td><td></td></tr>
<tr>
<td>InspiratoryReserveVolume(L) </td><td>3.316 </td><td>Mean of 3.299 </td><td><span class="success">-0.5%</span> </td><td></td></tr>
<tr>
<td>ResidualVolume(L) </td><td>1.234 </td><td>Mean of 1.234 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>TotalLungCapacity(L) </td><td>6.169 </td><td>Mean of 6.169 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>VitalCapacity(L) </td><td>4.935 </td><td>Mean of 4.935 </td><td><span class="success">0%</span> </td><td></td></tr>
</table><table class="doxtable">
<tr>
<th>Property Name </th><th>Expected Value </th><th>Engine Value </th><th>Percent Error </th><th>Notes  </th></tr>
<tr>
<td>BasalMetabolicRate(kcal/day) </td><td>2010.8 </td><td>Mean of 2011.0 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>BloodVolumeBaseline(mL) </td><td>7011.2 </td><td>Mean of 7052.6 </td><td><span class="success">0.6%</span> </td><td></td></tr>
<tr>
<td>DiastolicArterialPressureBaseline(mmHg) </td><td>73.5 </td><td>Mean of 73.3 </td><td><span class="success">-0.3%</span> </td><td></td></tr>
<tr>
<td>HeartRateBaseline(1/min) </td><td>72 </td><td>Mean of 73 </td><td><span class="success">1.4%</span> </td><td></td></tr>
<tr>
<td>MeanArterialPressureBaseline(mmHg) </td><td>87.0 </td><td>Mean of 94.9 </td><td><span class="success">9.1%</span> </td><td></td></tr>
<tr>
<td>RespirationRateBaseline(1/min) </td><td>16 </td><td>Mean of 16 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>SystolicArterialPressureBaseline(mmHg) </td><td>114.0 </td><td>Mean of 113.9 </td><td><span class="success">-0.1%</span> </td><td></td></tr>
<tr>
<td>TidalVolumeBaseline(L) </td><td>0.683 </td><td>Mean of 0.556 </td><td><span class="warning">-18.5%</span> </td><td></td></tr>
<tr>
<td>ExpiratoryReserveVolume(L) </td><td>1.365 </td><td>Mean of 1.343 </td><td><span class="success">-1.6%</span> </td><td></td></tr>
<tr>
<td>FunctionalResidualCapacity(L) </td><td>2.926 </td><td>Mean of 2.903 </td><td><span class="success">-0.8%</span> </td><td></td></tr>
<tr>
<td>InspiratoryCapacity(L) </td><td>4.876 </td><td>Mean of 4.898 </td><td><span class="success">0.5%</span> </td><td></td></tr>
<tr>
<td>InspiratoryReserveVolume(L) </td><td>4.193 </td><td>Mean of 4.343 </td><td><span class="success">3.6%</span> </td><td></td></tr>
<tr>
<td>ResidualVolume(L) </td><td>1.560 </td><td>Mean of 1.560 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>TotalLungCapacity(L) </td><td>7.802 </td><td>Mean of 7.802 </td><td><span class="success">0%</span> </td><td></td></tr>
<tr>
<td>VitalCapacity(L) </td><td>6.241 </td><td>Mean of 6.241 </td><td><span class="success">0%</span> </td><td></td></tr>
</table># Applied Research Associates
## Licensing & Trademarks
## Source Code
All BioGears&trade; software is open source and free software. This means that it is not only available for download free of charge, but you have access to the source code and may modify and redistribute our software subject to certain restrictions.
For more details, please read about our source code license and software licensing policy.
## Trademarks
Although our code is free, it is very important that we strictly enforce our trademark rights, in order to be able to protect our users against people who use the marks to commit fraud. This means that, while you have considerable freedom to redistribute and modify our software, there are tight restrictions on your ability to use the BioGears&trade; names and logos in ways which fall in the domain of trademark law, even when built into binaries that we provide.
For more detail on our trademark licensing, see our BioGears&trade; Trademark Guidelines. If you still have questions after reading the policy, please contact admin@biogears-engine.com.
## Website Content
The text of BioGears&trade; websites is generally licensed under all rights reserved. Tables and Validation can be used under Creative Commons Attribution (CC BY) in reference to offical BioGears&trade; releases.  Use of all images is under the sole discretion of Applied Research Associates and the BioGears&trade; project.  These files are provided to developers and integrators of BioGears&trade; for personal use only and may not be distributed outside of the offical BioGearsEngine.com project website. 
## Website Code
You may use the HTML and CSS files on www.BioGearsEngine.com and wwwBioGears.dev or other BioGears&trade; websites for your web site - as long as you don't use the visual design or style (the colors, gradients, fonts, etc.). In other words, you can use the HTML and/or CSS code, but your page should not look like you took BioGears&trade;'s page and dropped in your logo. This does not grant you any right to use the trademarks, trade names, service marks, or trade dress of Applied Research Associates or its products.
## Executable Software
Executable software binaries released by the BioGears&trade; Project, such as libBioGears and BioGearsUI, are made available under the terms of the APACHE2.0 License. On some proprietary, non-free platforms, necessary binary components subject to other licenses may be included in the installer, as explained in this policy.
Informative, Not Authoritative
This page is designed to inform, not to be the final, authoritative reference. If you're looking to reuse or redistribute something BioGears&trade; has created, always check the licensing in the thing itself (for example, in the footer of a web page, or in the "About" box of software) after you've looked at this document.
In case of a conflict between this document and other documents (for example, if this document and the header file of a particular source code file conflict) the other document should take precedence. If you do discover such a conflict, please notify licensing@biogearsengine.com so we can fix it.
Questions?
If, after reading all the above carefully, you have a further question about the licensing terms of anything created by the BioGears&trade; project, please send the question to  licensing@biogearsengine.com.## What is %BioGears?

%BioGears is a C++ based, open source, multi-platform (Windows, Mac, and Linux), comprehensive human physiology
engine that will drive medical education, research, and training technologies. 
%BioGears enables accurate and consistent physiology
simulation across the medical community. The engine can be used as a
standalone application or integrated with simulators, sensor interfaces,
and models of all fidelities.

%BioGears high-level objectives include:
-   Create a publicly available physiology research platform that
    enables accurate and consistent simulated physiology across training
    applications
-   Lower the barrier to create medical training content
-   Engage the community to develop and extend physiology models
-   Meet the training needs of the military
-   Expand the body of knowledge regarding the use of simulated
    physiology for medical education
	
## What can %BioGears do?

An instance of a %BioGears engine models a single patient's physiology.
- The patient is defined by parameters, such as height, weight, systolic and diastolic pressure.
- You can initialize the patient with specific chronic and/or disease states via conditions.
- You can modify the patients external environmental conditions (weather, submerge in water, etc.)
- You can apply various actions (acute insults/injuries, interventions, conscious breathing, exercise, etc.) to be applied to the patient.
- The patient can also interact with equipment models, such as an Anesthesia and/or an %ECG Machine as well as an %Inhaler via actions.

## What kind of data can I get from %BioGears?

Available data is defined within the %BioGears Engine in three major ways:
-	System data, such as %Cardiovascular, %Respiratory, etc.
	-	Analogous to system vitals
	  -	Examples: heart rate, oxygen consumption, mean arterial pressure, etc.
-	Compartment data
	-	Flow, pressure, and volume related to specific part of the body or component of equipment
	  - Examples: Blood flow to the brain, Right Lung Volume, Right Heart Pressure
	- Substance specific data related to a specific part of the body or component of equipment
	  -	Examples: The Extracellular concentration of succinylcholine in the brain tissue, anesthesia machine gas inlet oxygen volume fraction
-	Assessments
	-	Formed at the level of a clinician's report, Intended to give general patient overviews
	  -	Example: Pulmonary Function Test
    
## What are your models and how did you validate them?

%BioGears is a closed loop total body physiology model that combines physics-based lumped parameter models 
and control system feedback mechanisms to model real-time system-level 
physiologic behaviors. Lumped parameter models use electrical circuit analogs 
to represent the major physiologic systems. 

For validation, our team uses a combination of peer-reviewed publications 
and subject matter expert judgment. Our team has:
- Defined key parameters for each system for validation
- Collected published data in the form of waveforms, and max, min, and mean values
- Used custom written tools to compare data, perform analysis, and generate plots and tables of results

A primary purpose of model validation is to ensure that the model has an appropriate domain of validity 
given the goals of the modeling process. Our validation effort is driven by the goals of the project.
For that reason, we do not have the resources to perform a rigorous sensitivity analysis or some of the other 
tools associated with a general validation protocol. We are not aware of any existing third-party validation effort. 
We would welcome and support, in as much as we are able, any validation or uncertainty quantification effort by the community.

We provide reports for each @ref SystemMethodology included in the %BioGears Engine.
<br>  
These documents cover the design, implementation, assumptions, limitations, and validation of each model. 

#### Physiology Systems
@secreflist
 @refitem BloodChemistryMethodology "Blood Chemistry"
 @refitem CardiovascularMethodology "Cardiovascular"
 @refitem DrugsMethodology "Drugs"
 @refitem EndocrineMethodology "Endocrine"
 @refitem EnergyMethodology "Energy"
 @refitem EnvironmentMethodology "Environment"
 @refitem GastrointestinalMethodology "Gastrointestinal"
 @refitem NervousMethodology "Nervous"
 @refitem RenalMethodology "Renal"
 @refitem RespiratoryMethodology "Respiratory"
 @refitem TissueMethodology "Tissue"
@endsecreflist 
#### Equipment
@secreflist
 @refitem AnesthesiaMachineMethodology "Anesthesia Machine"
 @refitem InhalerMethodology "Inhaler"
@endsecreflist  
#### Modeling Support
@secreflist
 @refitem PatientMethodology "Patient Variability"
 @refitem CircuitMethodology "Circuits"
 @refitem SubstanceTransportMethodology "Substance Transport"
@endsecreflist 
    
## How do I run %BioGears?    

The @ref Toolkit is intended for users (e.g., researchers, educators, or curious
individuals) who wish to execute %BioGears and view the physiological
effects of the patient. The Toolkit will create and plot a time-based,
comma delimited file of calculated physiological outputs.

You can download the Toolkit here.<br>

The Toolkit unzips about 15&nbsp;MB of files.
We recommend having at least 1&nbsp;GB of available disk space.
- The result files generated from the command line executable and GUI average 20&nbsp;MB.
- Graphed results file from the GUI or graphing script can generate file sets of several hundred MB.
	
## How do I code with %BioGears?

%BioGears has developed a modular architecture to reduce costs for applications that need a physiology engine as well as want to develop or extend a physiology model.

This architecture contains :
- A @ref CDM containing set of classes and generic algorithms for lumped parameter physiology modeling 
- A standard @ref physeng for controlling a physiology engine via these CDM objects

We created a @ref SDK (SDK) to help developers integrate
the %BioGears Engine into software applications. This SDK provides
pre-built libraries and headers, as well as examples of how to programmatically
using the provided interfaces. The provided application programming
interfaces (APIs) provide full control over the engine to execute
various actions and retrieve a range of calculated physiological
outputs.

You can download the SDK here.<br>

## How can I modify %BioGears, or integrate my model with %BioGears?

%BioGears uses an extensible architecture to promote integration with external
models with varying levels of fidelity. System-level model fidelity can be 
increased or decreased by adding or removing nodes and sub-circuits.

All integration/extension will require a custom build of our @ref SourceCode.
The @ref CDM provides a standard for data interchange between models. 
The deliberate identification of data requirements must precede any model modification or addition to determine 
if an extension of the Common Data Model is required. If the existing data model is sufficient to meet your modeling needs, 
you may be able to implement changes satisfactorily just by modifying the source code for the physiologic system 
of interest. If a Common Data Model extension is necessary, modification of the source code becomes more complicated. The 
quickest and easiest way to modify %BioGears to meet your needs is to [work with us](https://www.biogearsengine.com/workwithus "Work with us").
We can help with requirements definition, provide development support, and/or make modifications for you.

You can download the source code here.<br>
The source unzips around 350&nbsp;MB of files.<br>
We recommend having at least 5&nbsp;GB of disk space dedicated to building %BioGears.

#### How do I ensure my changes/model are good?

We include scenarios and their results for verification and validation. 
These results provide a baseline we can use to measure deviations to results when the code is modified.
As changes are implemented in the code base, we rerun all scenarios and compare the new results with baseline results to see how the implemented changes manifest in %BioGears system data. 
Any new result that is over 2% error is marked as a failure. 
This data is used extensively to validate each system individually, as well as the combined effects of insults and interventions. See the Methodology Reports for more details.  
The scenarios output requests match the columns in the results file; we recommend that these scenarios remain unmodified.

You can download the data sets here.<br>
The complete data set unzips around 2&nbsp;GB of files.<br>
Execution of all scenarios provided can generate up to 20&nbsp;GB of results data.Software Development Kit {#SDK}
========================

The Software Development Kit (SDK) contains the following:
-   The tools provided in the %BioGears @ref Toolkit
-   A collection of how-to cpp files demonstrating the use of the API
-   The headers needed by applications integrating the %BioGears Engine
-   32-bit debug and release libraries

The extracted file structure:
<img src="./images/GUI/SDKDir.png" alt="Software Development Kit Directory">
<br>

We developed %BioGears with the Microsoft Visual Studio (MSVC) 2015
complier. The %BioGears code base is GCC C++11 compliant and we have 
provided cmake configuration files for various compilers.  We currently support
the following platforms, compilers, and environments:
-   Windows Visual Studio 2015 (MSVC)
-   Windows MinGW (GCC)
-   Mac Xcode (Clang)
-   Linux (GCC)
-   Raspbian for Raspberry Pi (GCC)

The @ref SourceCode is available if you would like to customize
or build the project for a different platform.

We wrote executable sample C++ as well as Java methods that can be compiled and executed to demonstrate how to use the API. 
To execute the C++ examples, simply run cmake against the CMakeLists.txt file to create the project for your complier.
Each cpp and java file is well-commented and should help explain how and why one would exercise functionality.

The %BioGears Engine functionality is exposed via the @ref physeng.<br>
All data necessary to interact with the %BioGears Engine is via @ref CDM classes. 

     

# Latest Release Notes

Our latest deployment is still in a beta phase, and is intended to be an intermediate release to 
showcase the capabilities of the %BioGears&reg; Physiology Engine. 
The current version of the software includes examples of all types
of engine interfaces, but does not include all of the functionality or
physiologic systems that will be present in the final product. This
version of the software is meant to elicit feedback and enhance
community involvement in establishing end product expectations.

- - -

@insert MainPageFAQ.md

## Have more questions?

See the @ref ExtraFAQ for any other questions you may have.

- - -

## Programmatics

%BioGears is being developed under the <a href="http://www.tatrc.org/">TATRC</a> funded project:
W81XWH-13-2-0068.

Disclaimer:

> This work is supported by the US Army Medical Research and Materiel Command. The views, opinions and/or findings contained in this report are those of the author(s) and should not be construed as an official Department of the Army position, policy, or decision unless so designated by other documentation.

%BioGears is released under this @ref License.

%BioGears has @ref published papers and abstracts on several systems and clinical scenarios. 

- - -
## What's new in ver 7.3.2 and 7.3.1 (Dec 10, 2020)
- **ver 1.0 release of the UI**
- **ver 1.0 release of BioGears Lite**
- Plasma Lyte compound substance added 
- **New drug administration route** nasal administration 
- **New drug** nasal naloxone
- CPACK installer functionality added 
- Moved data tracking to advance model time method
- Added sequential organ failure (SOFA score) as an assessment
- Changed website documentation tools to python, removed Java requirements 
- Minor drug model updates for LD50 and PD modifiers
- Implemented improved command line interface, named bg-cli to support better threading and logging
- Integrated inflammation model into hemorrhage model to make it more physiologically accurate
- Implemented acute respiratory distress model 
- Implemented a model of sleep and the metabolic consequences of sleep
- added new psychomotor vigilance test (PVT) patient assessment
- Removed legacy functionality of the GI system 
- Implemented Richmond agitation sedation (RASS) scale as a patient request 
  - Is a function of the pharmacokinetics of propofol and other anesthesia drugs
- Implemented a **new exercise model** that supported weighted exercise, cycling, and rowing
- Updated nervous system model, validated for inflammation and other injury responses
  - Including localized autoregulation 
- Implemented a tourniquet action for specific body regions
  - will perform locally, reducing hemorrhage in downstream vessels
- Compatibility updates to support Unreal Engine integration 
- General CMAKE updates to the build system 
- Minor bug updates and validation changes to models
- Finished all conformant override parameters (pressure, heart rate, respiration rate, and oxygen saturation)

## What's new in ver 7.3 (January 30, 2020)
- Custom Compound Infusion 
- Respiratory Improvements and Tuning
- Improved website generation targets
- Updates to CMD_BIO executable for batch validation and scenario execution
- Website generation python preprocess tool 
- Transmucosal Fentanyl implementation
  - oral diffusion through the mucosa layer and a model for GI transit and small intestine absorption
- Hemorrhage model updates 
  - Direct model connection to energy and nutrient model 
- Tourniquet action for use with Hemorrhage scenarios
- **New drug:** Tranexamic Acid
  - Validated for use in a hemorrhage scenario to decrease bleed rates 
- Propofol validation and bug fixes
- Expansion of the cerebral circuit to a larger, more complex system
  - Added cerebral auto-regulation, hemorrhage, and updated TBI model
- **New drugs:** Moxifloxacin and Ertapenem (intra-venous and intra-arterial)
  - Used in sepsis model, validated for treatment guidelines
- Validation updates to Fentanyl drug
  - Updated tissue volumes, updated perfusion limited diffusion method, and updates to PK values  
- New override functionality
  - Heart rate, respiration rate, and blood pressure values may now be set while running a conformant engine
    - Other physiology will change in collaboration with these values to accurately change major physiology during runtime without an injury or action
- New whole blood model/substance 
  - Antigens, agglutination model and typed blood substance files for administration during runtime
- **New drug administration route:** oral tablet 
  - Antibiotic with validated PK profiles and infection interactions
  - May be used in sepsis scenarios for management of infection 
  - New future Tylenol drug for moderate pain management 
- python plotter util updates to batch plot csv files generated by BioGears
- New complex How-to files covering burn and sepsis, for developers use
- Modifications to SE classes to support Unreal Engine integration
- Updated website generation and layout
- Patient blood type added as patient parameter
- General CMAKE updates to the build system 



(Interested in a previous @ref version?)

- - -
@anchor known-issues
## Known Issues and Limitations
The following are known issues with the current version of the software:
- Peripheral resistance currently does not scale with core temperature
- Glucose modulation through the endocrine system does not scale during exercise
- Secondary respiratory injuries due to sepsis such as acute lung injury, are currently not included in the model
- - -

## Tentative Near-Term Timeline

These are the planned updates:
- Bug fixes
- Nerve agents and expanded CBRNE models 
- UE4 plug in

- - -

## Credits
<b>%BioGears @ref version 7.3-beta</b>

<b><a href="http://www.ara.com/">Applied Research Associates Inc.</a></b>

*Principal Investigator:* Austin Baird, PhD

*Project Manager:* Jenn Carter

*Physiology Modeler:* Nathan Tatum

*Software Developer:* Steven White, Phd

*Software Developer:* Lucas Marin

*Website Engineer:* Levi Smith

**Consultants:**

Bryan Bergeron, MD (<b><a href="http://bryanbergeron.com/">Archetype Technologies, Inc.</a></b>)

Nicholas Moss, PhD

Stephen Mangum, PharmD

**Past Contributors:**

Matthew McDaniel

Anthony Hamilton

David Byrd

Brian O'Day

Bennet Welch

Aaron Bray

Jeff Webb

Rachel Clipp, PhD

Jerry Heneghan

Yeshitila Gebremichael, PhD

Zack Swarm

Pat Russler

Beth Smith

Paul Rutledge

Federico Menozzi

Alex Somers

Katie Carter

Cassidy Limer

<b><a href="https://pharmacy.unc.edu/">UNC Eshelman School of Pharmacy:</a></b><br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Alexander Tropsha, PhD<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Kimberly Brouwer, PhD<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Denise Rhoney, PharmD<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Eugene Muratove, PhD<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Daniel Gonzalez, PharmD, PhD <br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Alexander Golbraikh, PhD<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Vadryn Pierre, PharmD<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Nilay Thakakkar, PhD<br>

**Acknowledgement:**

The %BioGears team would like to thank the following individuals for providing programmatic guidance and oversight for the U.S. Government on this project:  
	
Jan Harris, PhD, RN<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; *Director, Joint Program Committee-1*<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; *U.S. Army Medical Research and Materiel Command (USAMRMC)*<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; *Fort Detrick, Maryland*

Kevin Kunkler, MD<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; *Portfolio Manager, Joint Program Committee-1*<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; *U.S. Army Medical Research and Materiel Command (USAMRMC)*<br> 
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; *Fort Detrick, Maryland*<br>

J. Harvey Magee<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; *Manager, Medical Modeling and Simulation Innovation Center (MMSIC)*<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; *Telemedicine & Advanced Technology Research Center (TATRC)*<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; *US Army Medical Research & Materiel Command (USAMRMC)*<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; *Fort Detrick, Maryland*<br>

Geoffrey T. Miller, MS<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; *Assistant Professor, Eastern Virginia Medical School*<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; *Research Scientist, Medical Modeling and Simulation Innovation Center (MMSIC)*<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; *Telemedicine & Advanced Technology Research Center (TATRC)*<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; *US Army Medical Research & Materiel Command (USAMRMC)*<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; *Fort Detrick, Maryland*<br>

Thomas B. Talbot, MD<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; *Principal Medical Expert - USC Institute for Creative Technologies*<br> 
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; *Associate Research Professor of Medical Education - Keck School of Medicine, USC*<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; *Playa Vista, CA*<br>

**Software Tools:**

%BioGears leverages the following: @ref Credits.

---

## Dedication

This software is dedicated to **N. Ty Smith, M.D.**, physician, professor, mentor, friend, and founding director of the Pacific Academy of Ecclesiastical Music (PACEM).

Dr. Smith was born in Iowa and graduated from Harvard College and Harvard Medical School. Dr. Smith served on the faculties at Stanford Medical Center and the University of California at San Diego. He was a Visiting Professor at the University of Washington, Institute of Medical Physics in Holland, University of Wisconsin, and University of Otago in New Zealand. Dr. Smith also served at Children's Medical Center in Boston, Massachusetts, Massachusetts General Hospital, U.S. Naval Hospital in Portsmouth, Virginia, U.S. Veterans Administration Hospital in San Diego, and Dunedin Hospital in New Zealand.

Dr. Smith, along with Ken Starko M.Sc., created "Body Simulation" in the 1990s. Body Simulation models and their interfaces were used for pharmacologic experimentation, testing, teaching, and training by device manufacturers, pharmaceutical companies, professional associations, and government agencies. %BioGears builds directly from this vision and legacy.

It was Dr. Smith's fervent wish that his work continue long into the future to advance scientific discovery, improve the safety of healthcare, and ultimately save lives.
 
*The %BioGears Team*<br>
*Raleigh North Carolina, September 2014*


@page errors Errors

@page events Events

Version {#version}
=======

We have taken the following approach to versioning our products : Major.Minor.Patch-ReleaseStage

Our version number sematic is MAJOR.MINOR.PATCH, where :
- MAJOR changes when we make incompatible API changes,
- MINOR changes when we add functionality in a backwards-compatible manner, and
- PATCH changes when we make backwards-compatible bug fixes.
- Release Stage - We have extended this versioning with a <a href="http://en.wikipedia.org/wiki/Software_release_life_cycle">release stage</a>

This versioning follows the <a href="http://semver.org">Semantic Versioning 2.0.0</a> format.

## %BioGears Version History

## What's new in ver 7.3.2 and 7.3.1 (Dec 10, 2020)
- **ver 1.0 release of the UI**
- **ver 1.0 release of BioGears Lite**
- Plasma Lyte compound substance added 
- **New drug administration route** nasal administration 
- **New drug** nasal naloxone
- CPACK installer functionality added 
- Moved data tracking to advance model time method
- Added sequential organ failure (SOFA score) as an assessment
- Changed website documentation tools to python, removed Java requirements 
- Minor drug model updates for LD50 and PD modifiers
- Implemented improved command line interface, named bg-cli to support better threading and logging
- Integrated inflammation model into hemorrhage model to make it more physiologically accurate
- Implemented acute respiratory distress model 
- Implemented a model of sleep and the metabolic consequences of sleep
- added new psychomotor vigilance test (PVT) patient assessment
- Removed legacy functionality of the GI system 
- Implemented Richmond agitation sedation (RASS) scale as a patient request 
  - Is a function of the pharmacokinetics of propofol and other anesthesia drugs
- Implemented a **new exercise model** that supported weighted exercise, cycling, and rowing
- Updated nervous system model, validated for inflammation and other injury responses
  - Including localized autoregulation 
- Implemented a tourniquet action for specific body regions
  - will perform locally, reducing hemorrhage in downstream vessels
- Compatibility updates to support Unreal Engine integration 
- General CMAKE updates to the build system 
- Minor bug updates and validation changes to models
- Finished all conformant override parameters (pressure, heart rate, respiration rate, and oxygen saturation)

## What's new in ver 7.3 (January 30, 2020)
- Custom Compound Infusion 
- Respiratory Improvements and Tuning
- Improved website generation targets
- Updates to CMD_BIO executable for batch validation and scenario execution
- Website generation python preprocess tool 
- Transmucosal Fentanyl implementation
  - oral diffusion through the mucosa layer and a model for GI transit and small intestine absorption
- Hemorrhage model updates 
  - Direct model connection to energy and nutrient model 
- Tourniquet action for use with Hemorrhage scenarios
- **New drug:** Tranexamic Acid
  - Validated for use in a hemorrhage scenario to decrease bleed rates 
- Propofol validation and bug fixes
- Expansion of the cerebral circuit to a larger, more complex system
  - Added cerebral auto-regulation, hemorrhage, and updated TBI model
- **New drugs:** Moxifloxacin and Ertapenem (intra-venous and intra-arterial)
  - Used in sepsis model, validated for treatment guidelines
- Validation updates to Fentanyl drug
  - Updated tissue volumes, updated perfusion limited diffusion method, and updates to PK values  
- New override functionality
  - Heart rate, respiration rate, and blood pressure values may now be set while running a conformant engine
    - Other physiology will change in collaboration with these values to accurately change major physiology during runtime without an injury or action
- New whole blood model/substance 
  - Antigens, agglutination model and typed blood substance files for administration during runtime
- **New drug administration route:** oral tablet 
  - Antibiotic with validated PK profiles and infection interactions
  - May be used in sepsis scenarios for management of infection 
  - New future tylenol drug for moderate pain management 
- python plotter util updates to batch plot csv files generated by BioGears
- New complex How-to files covering burn and sepsis, for developers use
- Modifications to SE classes to support Unreal Engine integration
- Updated website generation and layout
- Patient blood type added as patient parameter
- General CMAKE updates to the build system 



## What's new in ver 7.2 and 7.2.1 (January 29, 2019)

- General bug fixes and updates
  - Finalization for testing and implementation to BioGears override functionality with full physiology request data support
- Arterial and Venous PH data requests 
- Inflammation state data to support sepsis model serialization 
- Generalized sepsis model to a more generic inflammation model 
  - Will be critical to future modeling efforts (hemorrhage, burn, infection)
- New example sepsis xml files (SepsisSevere_Gut.xml)
- New lymph circuit
  - Handles Albumin transport and re-circulation 
  - Creates realistic oncotic pressure sources for substance transport 
  - Transport from tissue systems back into the vasculature via lymph 
- New command line utility project (bg-cli) for native c++ runtime, driver, batch run organizer/manager
- Optional name value for xml actions meal and environment
- New burn model 
  - User defined total body surface area input 
  - Inflammation cascade validated for long running scenarios (24 hr +)
  - Validated for traditional treatment protocols with USISR SMEs
- New unit testing framework (Google Test) to better support multi-platform functionality 
  - Unit Test harness is a separate project in CMAKE which can be controlled with Biogears_BUILD_TEST variable
- Introduced const char* DLL interfaces for all functions dealing with std::string to avoid Windows-related issues dealing with XSD implicitly exporting string through inheritance
- Updated functionality to tension pneumothorax to fix bug in bilateral behavior 
- Updated hemorrhage bugs to update blood gas levels and metabolic requirements 
  - Validated with University of Washington


## What's new in ver 7.1 (September 26, 2018)

- Patches to drug blood pressure modifications to restrict pathways to be more physiologically accurate
- Vasopressin support and validation
- Major patches to #include requirements, reduction in file dependencies
  - Increases modularity of the project, increase build times during development 
- Change in how we generate code from our CDM XSD files to one file per XSD file instead of per type 
  - Reduced build times for the full source from 40 to 10 min
  - empty constructors in SETypes to = default and adding override markers
  - no longer use stdafx.h while compiling and so many headers make direct reference to COmmonDataModel.h and Biogears.h which were previously bundled in these precompiled headers
- Override functionality now supported in BioGears 
  - May override any physiology data request with desired value 
  - Logging will document range of possible values if typing unsupported data 
  - Engine can now be globally flagged as conformant or non conformant to increase future development possibilities
  - Can be manipulated via action api calls 
  - Example xmls and sdks demonstrate functionality 
- Moved all BioGears functionality in to the BioGears namespace


## What's new in ver 7.0 (August 22, 2018)

- BioGears python plotting tool
- Max work rate now a patient parameter and is configurable
- Hemorrhage action updates, may now specify location and rate
  - Rate will diminish as pressure in the vessel decreases 
- Update build process to be entirely supported by CMAKE
  - Removed Apache Ant dependency 
  - Updated build directory and runtime directory dependencies
- Full build support for ARM platforms 
 - Updates to source to support all major platforms: Mac, Windows, Linux, and ARM
- Updated build architecture to python buildbot libraries 
  - 8 concurrent nightly builds to ensure multi-platform support
- Setup mirroring onto our new github repository 
  - All development now open to the community with feature branches also supporting nightly builds 
- Dockerfile and testing/support now supported, see more at https://cloud.docker.com/u/biogears/repository/docker/biogears/engine
- Pain model and patient pain susceptibility configuration flag
  - Validated pain model supported, stimulus can be specified with severity from 0-1
  - Works with all supported pain medication in the BioGears engine
    - Treat patient with Morphine, Fentynal, and/or Ketamine
  - New How-to-pain file to display sdk support
- Sepsis model 
  - Robust whole body inflammation model with severity and location specifiers in .xml and SDK
  - New How-to-sepsis file to show sdk functionality (command-line tool) 
  - Validated treatments with fluid resuscitation guidelines, vasopressin, norepinephrine, and antibiotics 
  - Validated blood chemistry markers such as bilirubin, white blood cell count, and lactate
- New antibiotic IV drip 
  - Can be used to treat sepsis
- Two new supported patients: toughguy and toughgirl 
- Sweat rate patches now meeting validation 
  - Better core temperature regulation during exercise 
  - Hyper/hypo-hidrosis now a supported patient parameter
- Updates and new 7.0 java GUI release to support users who want to create their own substance 
  - Includes ability to patch in new drugs
- Chemoreceptor method updated to track validation for hypercapnic and hypoxic conditions 
  - Better support for respiratory validation across the board, particularly supported respiratory conditions
- Patches to saline infusion loading on the patient for better respiratory validation 


## What's new in ver 6.3 (March 1, 2018)
The latest deployment includes the following notable updates:

- General bug fixes, system improvements, and tools/solver improvements
- Fasciculation patient event flags 
- Updated sweat methodology (fixes to ions lost in sweat)
- Updated substance and compound infusion functionality
  - Added Ringers lactate and updated 
  - Saline compound ion concentrations corrected
  - Hardened implementation 
- MuscleMass new patient data request
  - Muscle catabolism patient flag
- Added dehydration condition 
  - Implemented as scalar 0to1 representing fractional total body water lost
  - Fluid removed from patient compartments 
  - Updated patient flag for event and track body weight change (validated)
  - Added totalbodyfluidVolume as data request
  - Updated patient weight as a function of condition
- Added starvation condition
  - TimeSinceMeal determines how long since the patient's last meal 
  - Scales internal nutrient storages from validated starvation data
  - Removed ConsumeMeal condition, now replaced by starvation condition
  - Validated blood concentrations for ketones, glucose, and amino acids
  - Updated patient weight as a function of condition
- Intracellular ion transport
  - Model uses membrane potential  (see @ref TissueMethodology for details)
  - Michaelis coefficient could support more ion regulation in the future
  - Gated ion transport allows for differences between intra/extracellular spaces
- COPD now supports elevated anaerobic metabolism
- Ion transport model in the small intestine
- Updated drug library so all drugs support an effects site transport rate
- Diabetes type 1 and type 2 conditions
  - insulin resistance and insulin production effects
- Hemorrhage action now initialized with a 0-1 severity and a location (MCIS SDK example still exists)
- New drug Vasopressin
- New drug classifications in the CDM for better grouping in-code 
  - Include anesthetic, sedative, opioid, and reversal agent
  - More grouping in future work

## 6.2 (September 30, 2017)

The latest deployment includes the following notable updates:
- General bug fixes, system improvements, and tools/solver improvements
- Intoxication model for Morphine
  - Effects site concentration to allow for delayed PD reactions 
  - Central nervous modifier to model depressed feedback mechanisms 
  - Noloxone reversal agent model
- New Hemorrhage model 
  - MCIS support for combat injury coding 
  - Location and severity can be flagged through MCIS
  - Resistance paths handle bleed out to simulate more realistic flow behavior
- Glucagon hormone 
- New nutrient kinetics model 
  - New metabolic production and consumption method
  - Protein storage and release model (amino acids in muscle)
  - Fat storage and release model 
  - Interactions with the %Hepatic system
- New %Hepatic system 
  - Maintains blood glucose from other substances like lactate through new gluconeogenesis method
  - Lipogenesis method generates triacylglycerol due to excess glucose and amino acid
  - Glycogenesis and glycogenolysis to maintain blood glucose levels
- Updated exercise model 
  - Coupled to nutrient kinetic handling and metabolic need
- Updated diffusion method
- New %Gastrointestinal model 
  - Enzyme kinetics determine digestion of nutrients
  - Absorption facilitated through sodium co-transporter 
  - Gastric secretion function allows for bile formation in chyme
- Desflurane update and fix


### 6.1.1 (March 30, 2017)
Minor bug fixes

### 6.1.0 (March 10, 2017)

The latest deployment includes the following notable updates:
- General bug fixes, system improvements, and tools/solver improvements
- Improved Epinephrine methodology
- Improved Pupillary State for both Drug and %Nervous methodology
- Improved %Renal Tubuloglomerular Feedback
- Added cardiovascular chemoreceptor feedback
- Added Diuretic drug effects (Furosemide)
- Aerosolization of Solids and Liquids
  - Improves administration of Albuterol
  - New Smoke Particulate substance and smoke inhalation modeling
- Carbon Monoxide support
- New data requests and events
- New Conditions
  - Impaired Alveolar Exchange
- New Actions
  - Acute Stress
  - Apnea
  - Brain Injury
  - Intubation now supports Leftmainstem, Rightmainstem, Esopageal, and Tracheal types
  - Mechanical Ventilation

### 6.0.1 (December 15, 2016)

The latest deployment includes the following notable updates:
- General Bug fixes and system improvements
- Improved simulation runtime to ~5x real-time 
- Serialization of Engine State
- Updated GUI
- Patient variability support
- New and improved substance transporter methodology
- New and improved blood acid-base balance methodology

### 5.1.0 (March 4, 2016)

- General Bug fixes
- Improved Exercise Model
	- Exercise Action Intensity is now a fraction of work capacity (1200W)
	- New Fatigue Model
	- Removed Borg scale from Exercise action
- Improved Pulmonary Hemodynamics
	- Improved Pulmonary pressures
	- Removed Pulmonary Shunt Condition
- Updated cardiovascular validation data
- Results files are now CSV files

### 5.0.0-beta (December 18, 2015)

- General Bug fixes
- New Heat Stroke showcase scenario
- Physiology Interface Changes
	- Created a Java interface for controlling the %BioGears C++ engine
		- Added examples and the %BioGears.jar to the %BioGears SDK
	- Removed methods for executing a scenario from the Physiology Engine Interface
		- Use a SEScenarioExec class for executing a scenario, see HowTo-RunScenario.cpp in the SDK 	
- Baroreceptors	
	- %Nervous System responds to changes in mean arterial blood pressure by modifying the cardiovascular model

### 4.0.0-beta (October 12, 2015)

- Bug fixes and improved system calibration
- New target platforms
	- Windows
		- MSVC 32 & 64 bit 
		- MinGW 32 bit (GCC)
	- Mac
		- Xcode (Clang)
	- Linux
		- Ubuntu (GCC)
    - Raspberry Pi
        - Raspbian
- Improved compartment methodology
- Existing system upgrades
	- %Inhaler addition to equipment
	- %Respiratory conscious breathing
	- %Cardiovascular cardiac arrest and CPR redesign
	- Revamped %Renal System
	- Revamped %Gastrointestinal System
	- Updates to all other systems
- Updated pharmacokinetic and new pharmacodynmic model
	- Calculates partition coefficients based on physical chemical properties
	- Uses blood-tissue partition coefficients to calculate diffusion
	- Intrinsic, %renal, and systemic clearance remove drugs from the body
	- Pharmacodynamic effects are calculated for ten clinical outputs
- New showcase scenarios
    - Asthma Attack
	- %Environment Exposure
- New Patient Assessments
	- Complete Blood Count
	- Complete Metabolic Panel
	- Urinalysis
	
### 3.0.0-alpha (July 24, 2015)

- Bug fixes and improved system calibration
- New target platforms
	- Windows
		- MSVC 32 & 64 bit 
		- MinGW 32 bit (GCC)
	- Mac
		- XCode (Clang)
	- Linux
		- Ubuntu linix
- Engine time step increased from 1/165s to 1/90s for improved simulation speed
- New tissue integration with diffusion between vascular and extravascular
- Corrected O2 consumption and CO2 production within the tissues
	- New O2 and CO2 Hemoglobin binding methodology
	- Corrected environment/ambient volume fractions
- Mass and volume transfer calculations corrected for bifurcations and volume-less nodes (infinite volume) 
- Several new systems
	- System Interactions
	- %Renal
	- %Gastrointestinal
	- %Energy
- Existing system upgrades
	- %Environment - thermal functionality
	- %Respiratory
		- Asthma
		- COPD
			- Bronchitis
			- Emphysema
		- Lobar Pneumonia
	- %Cardiovascular - removed baroreceptor model and response incorporated into actions
- Updated pharmacokinetic model
	- New physiochemical parameters
	- Partition coefficient calculation
	- Perfusion limited diffusion
	- %Renal, hepatic, and systemic clearance
	
### 2.0.0-alpha (March 20, 2015)

- Bug fixes and improved system calibration
- CMake compilation
- New Dynamic Stabilization methodology
- New Conditions methodology
- Circuit Calculator
	- Polarized elements
	- More unit tests and validation
- %Cardiovascular updates
	- New/Updated Actions/Conditions
		- CPR
		- Anemia
		- Arrhythmias
		- Pericardial Effusion
		- Ventricular Systolic Dysfunction
		- Pulmonary Shunt
	- Redesigned circuit with more organs represented
- %Respiratory updates
	- Redesigned circuit
	- Redesigned driver
- Anesthesia Machine updates
	- Redesigned circuit
	- Redesigned ventilator
- Improved gas exchange
- New %Environment System
- New %ECG methodology	

### 1.0.0-alpha (October 1, 2014)

Initial ReleaseToolkit {#Toolkit} 
========

Toolkit Contents
----------------

The Toolkit contains the following:

- A collection of example @ref ScenarioXMLFile that can be found in the /bin/scenarios folder.
- A GUI provides a simple interface for creating, editing, and executing a scenario file, as well as creating a result and plots of requested data
- A script that executes a scenario file (via a C++ executable) and generates a result of requested data
- A script for generating plots of the result
- A script to compare results between two result files


<img src="./images/GUI/ToolKitDir.png" alt="BioGears Toolkit Directory">
<center>
*The extracted file structure, showing the location of the batch files for scripts and the GUI.*
</center>

- - -

GUI
---
<b>
The GUI is Java based, thus is reliant on a Java Runtime %Environment (JRE). 
You can get the JRE <a href="http://www.oracle.com/technetwork/java/javase/downloads/index.html">here.</a> 
JRE installation instructions are <a href="http://docs.oracle.com/goldengate/1212/gg-winux/GDRAD/java.htm">here.</a> 
To ensure you properly installed the Java Runtime %Environment, in your command window type "java-version". 
</b>


<b>Note:</b>  The GUI is able to run on both 32-bit and 64-bit Java Runtime Environments, the toolkit will automatically load the proper dlls associated with the JVM it loaded.
If the GUI has trouble opening and executing scenarios, there may be an architecture mismatch between the JRE and binaries.
Look at the JRE version string displayed in the GUI log window and make sure the corresponding binaries are present in the toolkit
folder in the release and release64 subfolders.
<br>

Start the GUI by double-clicking the BioGearsGUI.bat or BioGearsGUI.sh found in 
the bin directory of the extracted Toolkit download.<br>

<img src="./images/GUI/FullGUI.png" alt="BioGears GUI"> <br>

The main GUI window is where you will construct a scenario to be run by %BioGears. If you'd rather start from an existing scenario, open a scenario XML file via the File | Open menu option.
<img src="./images/GUI/MenuFile.png" alt="File Open"><br>

<br>
<b>Creating a Scenario</b>
<br>
To disgard all changes reated the current working scenario, use the File | New option to start over with a clean, empty scenario.

<br>
<b>Saving a Scenario</b>
<br>
To save a created scenario, use the File | Save option.

<br>
<b>Loading a Scenario</b>
<br>
To load a created scenario, use the File | Load option.<br>
When you load a scenario, if that scenario has a patient file and NO conditions, the GUI will automatically attempt to find and use the accompanying state file for the patient file.
The GUI will automatically change the patient file the associated engine state so as to avoid patient initialization.

###Name and Description###
If you loaded a scenario, you will notice the scenario name and description are populated in the Name and Description fields, respectively. 
If you are creating a new scenario from scratch, you can enter your own name and description.
<img src="./images/GUI/NameAndDescription.png" alt="Name and Description">

###Patient Information###
Choose a patient to be the subject of the scenario. Selecting the Patient File radio button will allow you to choose from a variety of available patients in the dropdown box below. 
Alternately, you can select the Engine State radio button to load a patient in a state after stabilization, which will reduce the scenario runtime. Note that you cannot 
add conditions to a scenario where an Engine State is loaded, but when loading from a Patient File, conditions can be added.
<img src="./images/GUI/PatientOptions.png" alt="Patient Options"> 
<br>
<b>Note:</b> Creation and editing of patients from the GUI is currently disabled in this release.

###Conditions###
If you are loading from a Patient File, you will notice a window with available conditions is available to edit. To add a condition to the selected patient, 
highlight the desired condition and click the Edit button. A window will appear where you can set any parameters for the condition. Note that if you leave a parameter 
blank, the scenario will use whatever value would have been used when no condition is applied. For example, when adding the Initial %Environment condition, 
changing only the AmbientTemperature field will result in the modified temperature being loaded along with the default values from the environment file for the other 
parameters. To remove a condition from the patient, highlight the condition and click the Clear button.
<img src="./images/GUI/Conditions.png" alt="Conditions"> 
<br>

###Actions###
The Actions window contains all available actions for the scenario. To add an action, select it from the list and click the Add button. Enter any action parameters 
in the window that pops up and click OK to accept. You will see the Action added in the right-side window. To edit an already added action, highlight it in the list 
and click the Edit button. You can also move a highlighted action up or down in the list by clicking the up and down arrows. Actions can be removed from the list by 
clicking the Delete button. Note that the Advance Time action must be added whenever you want the engine to progress, so a scenario with no Advance Time actions will 
not produce any meaningful data. Like with conditions, leaving certain action parameters blank will result in no change being applied when the action is initiated. 
<img src="./images/GUI/ScenarioActions.png" alt="Scenario Actions">

###Logs###
The Log area is located at the bottom of the GUI and contains information about what 
the user has been doing. The Log displays information, warnings, and error messages due to user interaction with the GUI.
<img src="./images/GUI/ConsoleLog.png" alt="Console Log">

###Scenario Execution###
The Scenario Execution options are available from the set of buttons in the Engine section.
<img src="./images/GUI/EngineOptions.png" alt="Engine Options">
These three buttons perform the following functions:
- Output - Choose the data to pull from the engine as the scenario executes
- Execute - Run the scenario and generate and plot the results
- Results - View the plots generated by the executed scenario<br>

<br>
<b>Output Button</b>
<br>
The Output button brings up a dialog where you can select which data you would like to pull from the engine as it runs. 
If the scenario file already contains data requests, those data requests are pre-loaded in the dialog.
You can specify a sampling rate by setting a value for 'Output file samples per second'. This will help in creating smaller output files, but note that the data is not interpolated in any way.
Clicking Cancel or the X in the title bar will discard any additions or removals you made to data requests.
Clicking OK will save the chosen data requests to the scenario XML file. 
You can set the name of the results file by clicking the SetResults button and specifying a .txt file in the location of your choice.<br>

<br>
<b>Removing Data Requests</b>
<br>
To remove a data request, simply click the Remove button next to the data request you want to remove.
<img src="./images/GUI/OutputWindow.png" alt="Output Window">

<br>
<b>Adding System Data Requests</b>
<br>
To add a System data request, click the System button. In the first dropdown menu, select the system in which your desired data resides. Next, specify the data and the data's unit. Finally, click Add to generate the data request.
<img src="./images/GUI/SystemDataRequests.png" alt="System Data Requests">

<br>
<b>Adding Compartment Data Requests</b>
<br>
The method for creating Compartment data requests is nearly identical to that of creating System data requests, just with a few more options.
<br>
- First choose the type of compartment that you are interested in, i.e. Vascular or %Tissue.
- The second column of drop-down boxes is for the particular compartment.
- Next choose the property you are interested in and the unit. 
- Finally, click Add to create the data request.
- <b>Note:</b> For all compartment types except %Tissue, you can check the Substance box to generate data requests for certain substances in a compartment.

<img src="./images/GUI/CompartmentDataRequests.png" alt="Compartment Data Requests" >
<img src="./images/GUI/CompartmentSubstances.png" alt="Compartment Substances" >

<br>
<b>Adding Equipment Data Requests</b>
<br>
Equipment data requests are handled the same as above; select your desired equipment, property, and unit, and click Add to generate the data request.
<img src="./images/GUI/EquipmentDataRequests.png" alt="Equipment Data Requests">

<br>
<b>Adding Substance Data Requests</b>
<br>
To add a Substance data request, select the desired substance, property, and unit in the drop-down boxes. You can create a data request for a particular tissue effect by checking the %Tissue Effect box and selecting the desired tissue.
<img src="./images/GUI/SubstanceDataRequests.png" alt="Substance Data Requests" >

<br>
<b>Execution</b>
<br>
Once the scenario has a patient, actions, and data requests, it is able to be executed.
Click the Execute button to bring up the Execute dialog where you can view data from the engine as it executes. Note that it can take a few seconds for the engine to stabilize and data to start appearing on the graph.
<img src="./images/GUI/Execution.png" alt="Execution">
<br>
The left text box shows all of the outputs requested for the scenario. Click on an output to view the graph for that property. 
<br>
As the engine executes the scenario, any messages, events, errors, etc., are logged in the Execution dialog console.
<img src="./images/GUI/ExecutionConsole.png" alt="Execution Console">
<br>
At any point during execution, you may cancel the calculation by clicking the Cancel button or clicking the X in the title bar.
<br>
After the engine has executed, the GUI creates graph plots of the requested data for the entire length of the scenario.
If you chose to cancel the calculation, you will be presented the option to create graphs of the data that the engine executed before you canceled. 
<img src="./images/GUI/CancelGraphs.png" alt="Cancel Graphs"><br>

The GUI will produce the following outputs in the directory you specified using the SetResults button :
- <ScenarioFileName>.log - A log file of all information produced from the execution. This contains the same content as the log console window.
- <ScenarioFileName>.txt - This file is a comma-delimited, time-ordered ASCII file that contains data values for each requested property for every time step of the engine.
- <ScenarioFileName><Assessment>@<Time> - If an Assessment action is in the scenario (e.g., Pulmonary Function Test), an XML file is written out containing the assessment data.
- <ScenarioFileName>Graphs - A directory containing plots for all of the requested data.

<br>
<b>Results</b>
<br>
Click the Results button to view the graphs created from the GUI. This viewer allows you to scroll through the plot images created. 
<br>
<b>Note:</b> The graphs are image files and do not allow for much detail or zoom capability. We recommend using a graphing application (e.g., DPLOT) that can read comma-delimited text files. 

<img src="./images/GUI/PlotViewer.png" alt="Plot Viewer">
<br>
Use the Send To button to open the graph image with a different viewer.
<br>
The following check boxes allow for viewing different images created:
<img src="./images/GUI/ShowPlots.png" alt="Show Plots">
<br>
%BioGears supports creation of the following plots:
- Standard - The default property vs time line plot.
- Error - (<b>Currently not generated by the GUI.</b>) When a baseline result file for a scenario is identified, the generated results file is compared to this baseline and this plot is the percent difference.
- Residual - (<b>Currently not generated by the GUI.</b>) When a baseline result file for a scenario is identified, the generated results file is compared to this baseline and this plot is the residual difference.
<br>
<br>

Command Line Execution
-----------------------
RunScenario.bat and RunScenario64.bat can be found in the bin directory of the extracted Toolkit download.<br>
These scripts will run BioGears in 32 or 64 bit executables.<br>
The input argument is the path to the scenario file you wish to run.
<img src="./images/GUI/ScenarioExe.png" alt="BioGears Scenario Driver Executable">

<br>This executable will produce the following outputs in the same directory as the active scenario:
- <ScenarioFileName>.log  - A log file of all information produced from the execution. This contains the same content output to the console window.
- <ScenarioFileName>Results.txt - This is a comma-delimited, time-ordered ASCII file that contains data values for each requested property for every time step of the engine.

Plot Script
-----------

GraphResults.bat can be found in the bin directory of the extracted Toolkit download.<br>
The input argument is the path to the scenario file results you wish to graph.
<img src="./images/GUI/GraphExe.png" alt="BioGears Results Plotter">

<br>When the execution of this script is complete, 
a directory is created at the same location as the scenario results file with the same name as the results file. 
This is the location where the graph images are created. 

Compare Script
--------------

CompareResults.bat can be found in the bin directory of the extracted Toolkit download.<br>
<img src="./images/GUI/CompareExe.png" alt="BioGears Results Comparison">
<br>This batch file needs to be run from the bin directory. The input arguments to the script are as follows:
- Expected results text file
- New results text file to compare to baseline 
- Percent Tolerance to use in comparisons (optional, default is 2%)
- Report differences true|false (true generates a LOT of data, if errors are found) (optional, default is false)

Full example: CompareResults.bat "TestingRespData.txt" RespiratoryOutput.txt 2.0 false

When the execution of this script is complete, 
a directory is created at the same location as the scenario results file with the same name as the results file. 
The following items are created in this directory: 
- Plot for each column of time vs. column value. Both baseline and expected lines are provided in each plot
- Plot for each column of the error over time between the baseline and the generated column data
- Plot for each column of the residual error over time between the baseline and the generated column data
- An XML report file detailing which columns did not meet match the baseline data if the Report Differences flag is true. EVERY column index that does not match the baseline is detailed in the report.

ExtraFAQ {#ExtraFAQ}
============

## Why does it take so long to initialize %BioGears?
%BioGears represents a single, @ref PatientMethodology "variable patient". 
Patient variability requires that the engine analyze the provided patient baseline values and stabilize the physiology to those values. 
This initialization can take several minutes, but once complete, %BioGears can save the engine state to an xml file.
You can then load this state and instantaneously start execution of the simulation without any initialization time.
Please consult the example in the SDK for how to take advantage of this feature and eliminate any initialization time in your application.

## What is the fidelity of %BioGears?
One definition of fidelity is "The degree to which a model or simulation represents the state and 
behavior of a real world object or the perception of a real world object, feature, condition, or chosen 
standard in a measurable or perceivable manner; a measure of the realism of a model or simulation @cite msco . " 
The %BioGears validation documentation (in the @ref SystemMethodology reports) describes how well the engine 
reproduces physiology at the system level. Like the human body, %BioGears is a self-compensating system of 
physiological systems with outcomes based on interventions @cite pettitt2009task , and therefore can be considered high-fidelity.

Sometimes the word fidelity is used to refer to the spatial (anatomical) level of resolution of a model. 
%BioGears is a closed loop 
total body physiology model that combines physics-based lumped parameter models 
and control system feedback mechanisms to model real-time system-level 
physiologic behaviors. Spatial resolution is limited by the lumped-parameter approach 
to sections of organs (what may arguably be referred to as the tissue level). However, %BioGears 
uses an extensible architecture to promote integration with external models with varying levels of 
fidelity (resolution or granularity). For more details, please see the recorded [Committee on Credible Practice of Modeling & Simulation 
in Healthcare](https://simtk.org/projects/cpms/ "CPMS") webinar.

## Are there any publications related to the models that you have developed and choose to implement in %BioGears.
A list of publications and presentations about %BioGears can be found on the @ref published "Publications" page. 
Many of the physiology models in %BioGears are adapted or implemented directly from models described in literature. 
The implementation methodology is described in detail in the @ref SystemMethodology and sub-system documentation, and 
all of the source publications are cited in the methodology reports and listed in the 
[Bibliography](https://www.biogearsengine.com/documentation/citelist.html "Bibliography").

## What kind of uncertainty quantification do you do perform in your physiology model?
We have not performed a systematic forward propagation or inverse quantification of model uncertainty, 
nor do we have the resources to conduct a formal sensitivity analysis. However, we can quantify the numerical 
uncertainty introduced in solving the lumped-parameter fluid dynamics of the two foundation sub-models 
(@ref CardiovascularMethodology "Cardiovascular" and @ref RespiratoryMethodology "Respiratory"). BioGears currently uses a bi-conjugate 
gradient method specific for sparse square systems (using the Eigen third party packages). This is an 
iterative method and we use the default tolerance for their solver, which is as close to zero as reasonable (around 1e-16). 

For more discussion of uncertainty quantification in %BioGears please see the [%Biogears forums](https://www.biogearsengine.com/forums "Forums").

## Who is developing the %BioGears Engine?
<a href="https://www.ara.com">Applied Research Associates, Inc.</a> (ARA) is the company developing 
the %BioGears open source physiology engine. ARA is an international 
research and engineering company recognized for providing technically 
excellent solutions to complex and challenging problems. ARA's 
biomedical modeling and simulation research group has a proven track 
record of creating innovative, physiologically accurate mathematical 
models that drive immersive, game-based medical training technologies.

ARA has a broad range of technical expertise in biomedical engineering, 
defense technologies, civil engineering, computer software and 
simulation, systems analysis, and environmental technologies. 

## Can I contact the %BioGears team to work on my current or upcoming project?
Absolutely. We always welcome new and challenging opportunities to 
work with research partners and sponsors. Please <a href="https://www.biogearsengine.com/support/contact">contact us via the website</a> 
or look at our <a href="https://www.biogearsengine.com/workwithus">Work With Us</a> page to learn more about working 
with the %BioGears team!

## What open source license does %BioGears Use?
<a href="https://www.apache.org/licenses/LICENSE-2.0.html">Apache 2.0</a>. For more information see our @ref License.

## What is the long-term plan for %BioGears?
Our team's goal is to first and foremost develop the most advanced, 
open source, whole-body physiology engine created to date. Following this,
our team plans to work with the user community and stakeholders to ensure 
%BioGears becomes the standard in physiology modeling.

During the first year of the project, our team made many important decisions 
regarding system architecture and the open source @ref License structure to ensure 
long-term use. Our system architecture was developed in a way that will make 
the system easy to extend for new models and external interfaces. The license 
structure allows for both open-source and proprietary applications to promote 
widespread use across government, military, academic, and commercial markets.

The %BioGears team as a whole is very passionate about the use of simulated 
physiology for medical training and education. As such, we intend to develop 
and maintain a useful, high quality, open source application that will be 
extended and improved by our team and the community over time.

## Where can I ask questions or get help?
Our team spent a great deal of time documenting the engine and our 
system models. Please look at the documentation first. We have also set
up an online [forum](https://www.biogearsengine.com/forums/ "forums") as a place for the community of users to help one 
another figure out how to use the engine. Please remember these are 
friendly forums. For more specific questions about connecting the engine 
with an external interface or extending system models, contact us via 
the website or take a look at our [Work With Us](https://www.biogearsengine.com/workwithus "Work With Us") page to learn more 
about working with the %BioGears team!

## How can I contribute?
The %BioGears team will continue to support large releases as in the past but will begin to transition most development onto our publicly facing <a href="https://github.com/BioGearsEngine/Engine">github site</a>. Because our team believes in collaborative open source development always feel free to fork our code base and create a pull request to have your changes adopted by the team at ARA. In the future, look for contributing requirements, rules and directions to be posted directly on that site.

## Is %BioGears a game?
No, %BioGears is a physiology engine that can power immersive 
learning and serious games for medical training. The %BioGears physiology 
engine can provide a realistic training experience by producing real-time 
results to trauma and treatment. %BioGears can enhance the user experience of applications 
by providing a comprehensive physiological response to insults and interventions.

## What are some possible %BioGears applications?
There is a wide range of potential applications for %BioGears. A few include:
- Powering serious games for medical education and training
- Producing responsive physiology in real time for manikin training
- Integrating a single-system model into %BioGears to understand full-body physiologic response
- Providing inputs and outputs for sensor systems
- Teaching and education
	
## Where do I log a bug for %BioGears?
Logging bugs helps us improve the engine and we appreciate your 
feedback. You can log bugs on our <a href="https://www.biogearsengine.com/forums/categories/report-a-bug">Report Bug</a> discussion, or post code bugs on our <a href="https://github.com/BioGearsEngine/Engine">github page</a>.

Please also feel free to report any inappropriate physiology. You can check to see if we already know about the 
issue by visiting the [Known Issues](@ref known-issues) section on the main page or the @ref SystemMethodology documentation.  

## What is a Showcase Scenario?
A <a href="https://www.biogearsengine.com/showcase">Showcase Scenario</a> is a hypothetical patient scenario our team uses to 
demonstrate the %BioGears Engine's patient customization, insult, intervention, 
and assessment capabilities. We have created four of these scenarios to provide a 
framework for discussion about the engine and serve as a catalyst for 
community-initiated improvements for the duration of the project.

These Showcase Scenarios are driven by the goals of the %BioGears project. 
It is important to note that Showcase Scenarios are not being used only as 
validation use cases, but also as examples to demonstrate the capabilities 
of the physics-based %BioGears Engine.

To see the scenario outputs from the engine, check out our interactive graphing 
tool on our website.

## What is your relationship with the Virtual Physiological Human (VPH) project.
The Virtual Physiological Human is a European initiative with the eventual goal of producing a complete 
mechanistic model of the entire human body. With %BioGears, we are trying to simulate whole-body physiology 
with reasonable accuracy for a target population. In other words, we are attempting to model a generic 
individual within a reference population to provide reasonable physiology for a variety of applications. 
In contrast, the eventual goal of the VPH project is individualized 
medical simulation. Individualized simulation is not within the current scope of the %BioGears project, 
but we have gained insight and generated knowledge of development processes, and we presented our findings 
to the VPH community at the 2016 conference @cite metoyer2016framework.

## Where can I find an overview presentation of %BioGears?
An overview presentation on %BioGears given to the [Committee on Credible Practice of Modeling & Simulation 
in Healthcare](https://simtk.org/projects/cpms/ "CPMS") is available on [YouTube](https://www.youtube.com/watch?v=VQW0dqJ5mYA&feature=youtu.be "%BioGears Presentation").

## What is the advantage of the %BioGears Common Data Model (CDM)?
For details about the %BioGears Common Data Model, please see our @ref CDM documentation.

## How fast does %BioGears run? How can I make it faster?
%BioGears currently runs at about 8 times real-time, depending on the your computer's CPU speed. The functionality requirements of the multi-purpose physiology 
engine are driven by the goals of the project. If your application does not require all of the existing functionality, 
then you could strip features by modifying the source code in the same way that you would integrate a new model.

The easiest way to make %BioGears faster is to [work with us](https://www.biogearsengine.com/workwithus "Work with us"). 
We have the expertise to quickly and efficiently modify %BioGears to meet your requirements.

## Do you plan to provide support for interpreter-level model input, for example with the Python language?
We do have plans to support hooks to our API for other languages such as C# and Python.
We do have support for Java. We are aware of an end user creating a C# interface on top of our C++ interface.

 







 






Common Data Model {#CDM}
=================

The Common Data Model (CDM) is an Object Oriented Design (OOD) of class structures that provides a unified set of data encapsulation objects 
that will promote fast development, compatible data sets, and well-defined interfaces for all types of physiology model development. 
The CDM provides a well-defined data interchange format that disparate models can use for standardizing inputs and outputs, 
making it easy for physiology models and engines to be extended and operate jointly.
The CDM is I/O indifferent; it provides the ability to hold all data associated regardless of the data being an input or an output to a specific model. 
These structures provide discrete properties for specific definitions and the flexibility to be easily expandable when incorporating new models. 
The CDM classes are small, lightweight (not much code in them), property-bag objects intended to only contain data for simple I/O and message passing.

You can access the CDM data dictionary via the modules link provided in the above header. 
You can also view @ref CDMTables associated with some of key objects associated with the @ref physeng . 
This dictionary documents the conceptual structure, organization, relationships, and property definitions. 
A set of base types, known as Properties, is defined to hold the actual discrete values of various objects. 
The CDM then encapsulates and organizes these properties into various physiology-related data objects, such as Systems, Anatomy, Equipment, Patients, Actions, etc.

The CDM data dictionary is implemented in an eXtensible Markup Language (XML) Schema Description (XSD). 
An XML schema is an open standard that describes the structure of an XML document and allows for defining data that programs can process in a generic manner. 
This representation enforces the meaning and structure and relationships as described in the data dictionary and provides an industry standard basis for implementation and data exchange. 
This XML schema can then be used to generate binding classes that can serialize XML for use in programming languages such as C++, Java, and C#.

<img src="./images/CDM.png" alt="Common Data Model">

We then created a set of API classes built to reflect the organization of the data dictionary, encapsulate the binding classes, and provide various data validation, manipulation, and translation algorithms. 
These classes are prefixed with SE, which stands for Synthetic %Environment. These classes provide an encompassing interface for holding and translating data in a simple-to-use, yet flexible and expandable, interface. 
This library will provide application developers with low-level infrastructure (serialization, translation, data reuse, etc.) already complete so they can concentrate on creating end-user content-driven applications.

## Properties

An %SEProperty is an object that represents a basic data type in the CDM that stores the actual data. <br>
There are several Property data types:
- %SEScalar - A combination of a double and a unit. The CDM does not require a property in a specific unit of measure (for example, the height of a patient does not have to be in meters). 
Rather, the CDM requires each data type to retain a unit of measure of the Property's quantity and the class will translate the value to a requested unit. 
One example is a Distance Property that holds data in any distance-associated unit: meters, feet, inches, etc. 
- %Strings - We use the inherent language type for string
- %Enumeration - Where applicable, instead of strings, we use enumerations for code clarity and speed.
- %SEFunction - A combination of an abscissa and an ordinate paired together. 

## Modeling Objects

Modeling objects are hierarchically organized data containers based on the needs of physiology modeling and simulation. 
These objects contain data by holding specifically named Property objects. 
They can also hold other modeling types to achieve the proper hierarchy of data organization. 
Since these objects do not contain the actual data values (they contain Properties) and optionally other modeling objects, their interface is based on a get/has style. 
When you call a get method (e.g., getVolume or getRightHeart) on an object, the requested object or property is returned (created on demand if it did not previously exist) (e.g., %SEScalarVolume or %SEHeart). 
The complementing has method (e.g., hasVolume or hasRightHeart) is a test to show if the object already exists or not. 

# %CDM Algorithms

A few generic algorithms are available in the CDM. The following sections describe some of these algorithms.

# Unit Conversion

The Common Data Model has a unit conversion library built into the %SEScalar class that will convert double data to whatever unit you specify. 
There is an %SEScalar class for various quantity types along with commonly used unit definitions in those classes (such as SEScalarLength has units : m, cm, in, etc.) 
You may use strings or a %CCompoundUnit to define your unit if you do not find a unit predefined, but strings have a runtime cost to them. 
Units are defined in the standard abbreviation, as described at: http://www.bipm.org/en/si/. 
Along with SI units, English units and medical units are supported. 
You can check to see if a unit is valid by calling the IsValidUnit method on a particular quantity scalar class. 

# Logging

The Data Model has a logging capability that each SE class uses to log any info to a log file.
This logger has different levels of logging an event:
- Debug - Extremely low level information, intended for the developer to assess any questions in functionality
- Info - A general level of information where the engine wants to inform a user of something
- Warning - Something was not inputted properly or an assumption was made in the underlying methodology
- Error - A recoverable error occurred and the engine is still able to proceed, but results should be inspected for correctness
- Fatal - A unrecoverable error occurred and the engine will stop calculating

The logger allows a user to programmatically add an object to the logging class that is called when 
a log event occurs. This allows a calling program to be notified so that they may react to these logs. (e.g., error handling)

# Data Track

A Data Track class is used for debugging purposes. 
Usually in debugging, one likes to print values of variables to the screen. 
This class allows a developer to write those values to a comma-delimited, time-ordered file. 
Developers can add any number of variables to track (only limited to machine memory) and they will be added to the output file for each time step.
	
# Circuit Solver

The @ref Circuit objects have algorithms for solving any closed-loop circuits as described in the @ref CircuitMethodology.

# Substance Transport

Substances are contained on compartments and are transported around the body based on volumes, pressures, and flows, which is handled via a generic @ref SubstanceTransportMethodology.

## Compartments
A compartment can represent various fidelities of data; it can be:
	- An anatomical space, such as the body's skin
	- An organ, such as the liver
	- An organ substructure, such as the Left Heart
	- A component of a piece of equipment, such as an anesthesia machine ventilator
	
Compartments are intended to be a generic data exchange interface between systems for the fluid dynamics data of the body, such as volumes, pressures, and flows.
In addition to the fluid dynamics, compartments also contain a substance quatity for each substance in the compartment. 
For example, a vascular compartment includes the masses and concentrations of all substances in that compartment, 
whereas a pulmonary compartment will contain the volumes and volume fractions of all substances in that compartment. 
This generic implementation allows any system to access and modify the parameters for the fluid dynamics calculations reguardless of the model type. 

We have a implemented a compartment interface to integrate directly with the circuit CDM classes to easily access the circuit-based fluid dynamics data. 
These circuit compartments are assigned nodes and paths and combine that data to provide an accurate fluid dynamics data set for a specified compartment. 

The circuit compartment fluid dynamics data is combined in the following manner:
	- Volume : Sum of the volume of all nodes in the compartment.
	- Pressure : Pulled from a single node of a compartment that was explicitly identified during setup.
	- Flow : The sum of inflows; if the inflow equals zero, it is the sum of the outflows.
	- Vascular Substance Quantites
		- Mass - The sum of all the masses for each node in the compartment for a particular substance
		- Concentration - The compartment's substance mass divided by the compartment volume
	- Pulmonary Substance Quantites
		- Volume - The sum of all the volumes for each node in the compartment for a particular substance
		- Volume Fraction - The compartment's substance volume mass divided by the compartment volume

Source Code {#SourceCode}
===========

The %BioGears code base consists of:
- Engine - A C++ based lumped parameter model
- @ref CDM - A set of generalized data classes for defining the data interfaces of a @ref physeng
-	@ref Toolkit - Various tools for execution and data manipulation, primarily written in Java 

The %BioGears source is structured as follows:
- bin - Contains all data and configuration files needed for execution of the %BioGears Engine
- data  - Contains the Microsoft Excel spreadsheets for all %BioGears data sets
- lib - Contains third party libraries used by this project (see @ref Credits for more details)
- src
	- cdm - Code associated with the @ref CDM and @ref physeng 
  - cmake - The directory where cmake will create build files
	- engine - Code associated with the lumped parameter models 
    - controller - These classes hold data necessary by the model, control the advancement of time
      - scenario - These classes help execute a %BioGears specific scenario (i.e. a scenario with a BioGearsConfiguration object)
        - BioGearsScenario - @copybrief BioGearsScenario
        - BioGearsScenarioExec - @copybrief BioGearsScenarioExec
        - BioGearsScenarioInitialParameters - @copybrief BioGearsScenarioInitialParameters
      - BioGears - @copybrief %BioGears
      - BioGearsCircuits - @copybrief BioGearsCircuits
      - BioGearsCompartments - @copybrief BioGearsCompartments
      - BioGearsConfiguration - @copybrief BioGearsConfiguration
      - BioGearsEngine - @copybrief BioGearsEngine
      - BioGearsSubstances - @copybrief BioGearsSubstances
      - BioGearsSystem - @copybrief BioGearsSystem
    - systems - These classes implement the methodology for modeling and simulating
      - BloodChemistry - @copybrief BloodChemistry
      - Cardiovascular - @copybrief Cardiovascular
      - Drugs - @copybrief Drugs
      - Endocrine - @copybrief Endocrine
      - Energy - @copybrief Energy
      - Environment - @copybrief Environment
      - Gastrointestinal - @copybrief Gastrointestinal
      - Hepatic - @copybrief Hepatic
      - Nervous - @copybrief Nervous
      - Renal - @copybrief Renal
      - Respiratory - @copybrief Respiratory		
      - SaturationCalculator - @copybrief SaturationCalculator	
      - Tissue - @copybrief Tissue
    - equipment
      - AnesthesiaMachine - @copybrief AnesthesiaMachineData
      - ECG - @copybrief ElectrocardiogramData
      - Inhaler - @copybrief InhalerData    
	- gui - Code associated with the @ref Toolkit GUI
  - schema - The xsd data definitions used by the CDM
	- sdk - Example code and scripts for the @ref Toolkit and @ref SDK
	- utils - Various utilities used in validation
	- verification - Extract the verification files into this directory. You can get more detail @link VV here. @endlink
	
## Building

###  Required Tools

#### CMake

We generate project files with the <a href="http://www.cmake.org/download/">CMake</a> system to support a cross compliler independent code base.
The %BioGears team primarily develops with Visual Studio 2015, but we have created and tested the following cmake configurations:
	- @link MSVC Microsoft Visual Studio @endlink
	- @link MinGW MinGW  @endlink
	- @link XCODE Xcode @endlink
	- @link LINUX Linux @endlink
  - @link RPI Rasperry Pi @endlink

The following projects will be generated:
- ALL_BUILD - Utility project for building
- %BioGearsEngine - A shared library project of the implementation of the %BioGears Physiology Engine
- %BioGearsEngineJNI - A shared library exposing the %BioGearsEngine through the Java Native Interface (JNI)
- BioGearsEngineTest - An optional shared library project that contains unit test classes associated with testing the %BioGears methodology
- %BioGearsScenarioDriver - The command line scenario executable provided in the @ref Toolkit
- CommonDataModel - This is a shared library project that builds all code associated with the @ref CDM
- CommonDataModelJNI - A shared library exposing the CDM through the Java Native Interface (JNI)
- CommonDataModelTest - A shared library project that contains unit test classes associated with testing the CDM implementation
- DataModelBindings - This shared library project generates and builds C++ code from the CDM xsd files via Code Synthesis
- log4cpp - A shared library project of the log4cpp library
- UnitTestDriver  - An executable harness for running unit tests from the two unit test projects
- ZERO_CHECK - Makes sure that project files are up to date with CMAKE files

The required projects to getting a usable C++ shared library are:
- log4cpp
- DataModelBindings
- CommonDataModel
- %BioGearsEngine

Other projects are all optional and are either unit test, tools, or Java interfaces projects.	
	
#### Ant

You can download and install Apache Ant from <a href="http://ant.apache.org/bindownload.cgi"> here</a>. 
For this installation, choose the zipped archive and extract the files to a new directory. 
Ensure you add the Apache Ant bin directory to the path.

#### Eclipse

<a href="http://www.Eclipse.org">Eclipse</a> is used to aid in Java code development. In the future we will support C++ project for use with Eclipse.

#### Java Development Kit (JDK)

We have developed a set of testing tools as well as our GUI in Java. 
We provide a very simple JNI interface to the C++ based engine. 
In the future we plan to expand the functionality provided in the Java Engine Wrapper.

### Documentation Generation Tools @anchor DocGen

%BioGears uses Doxygen to generate all documentation from both in-code comments as well as markdown files.
We also have a set of custom Java based preprocessors to help generate various tables and other documentation into markdown files.
The markdown files include both math formulas as well as bibliography citations.
In order to generate the %BioGears documentation you will need to have the following tools installed
  - A LaTeX compiler such as 
    - MikTeX or proTeXt for windows
    - MacTeX for MacOSX
    - TeX Live for Unix/GNU/Linux
  - A Perl compiler
  - Ghostscript
  - Doxygen

Currently there is only a windows based script for generating documentation and it is specialized for use in the %BioGears code repository.
For more information on using this script come to our forums.

### Eclipse

The following Eclipse projects are included in the source:
- %BioGearsEngine - A Java interface to the %BioGearsEngine 
- BioGearsEngineGUI - The Java Swing GUI provided in the @ref Toolkit
- CommonDataModel - A Java implementation of the @ref CDM, also contains the Comparison and Graphing utility.

The BioGearsEngineGUI project require the the C++ JNI projects be built in order to execute a scenario.

Recommended execution/debugging settings from Eclipse: 
  
<img src="./images/EclipseArgs.png" alt="Eclipse Arguments">
<br><br>
The following Java classes have main methods and can be executed:
- %BioGearsEngine - Executes a scenario through a Java wrapper
- BioGearsGUI - Executes the GUI
- DataSetReader - Reads the data/%BioGears.xlsx and generates all patient and substance XML files from data in the spreadsheet
- CSVPlotTool - Generates image plots for a results file. Used via the @ref Toolkit GraphResults.bat.
- CSVComparison - Compares two results files and generates a report. Used via the @ref Toolkit CompareResults.bat.

### Building with Visual Studio @anchor MSVC
	
Create the Visual Studio solution and project files by running the cmake target via ant.  By default, CMake will generate 32-bit Visual Studio 2015 project files in the src/cmake/msvs14x32 directory.
<img src="./images/Compile.png" alt="Ant compile">
<br/>

<b>Note:</b> 64-bit project files can be generated by specifying -Darchitecture=x64 in the ant command, but we are still testing this build and currently do not recommend using it.
<br/>

All builds copy their compiled targets to /bin/release or /bin/debug depending on MSVC build configuration. 
Execution and debugging working directories should be set to the bin directory.

Recommended execution/debugging settings for Visual Studio:

<img src="./images/MSVCDebugging.png" alt="Debug Settings">

### Building with GCC on Windows @anchor MinGW

%BioGears can be built with GCC on Windows using MinGW-w64.  Note that MinGW-w64 is NOT the same as MinGW.  It is a separate version that branched off from MinGW
in 2005 and contains the necessary C++11 threading libraries need to support multiple instances of a %BioGears Engine within an application.  Download and install MinGW-w64 using the indicated settings:
<br/><br/>

<img src="./images/mingw-w64.png" alt="MinGW-w64 Settings">
<br/>

Create an environment variable MINGW_HOME which points to MinGW-w64's installation directory, e.g.
`C:/mingw-w64/i686-5.1.0-posix-dwarf-rt_v4-rev0/mingw32`<br/>

When running ant commands, append `-Denv=mingw` to specify that the command should be run for the MinGW environment.  For example, compile %BioGears by running
the ant commands `ant cmake -Denv=mingw` and `ant compile -Denv=mingw` in the src directory.<br/>

### Building with Xcode on Mac OS X @anchor XCODE

#### Xcode

Download and install Xcode from the app store.  After the installation finishes, install the Xcode command line tools by running the
terminal command `xcode-select --install`.

#### Required tools

Make sure the following tools are installed:
- ant
- CMake
- Java JDK 8

An easy way to install ant and CMake is with Homebrew.  First install homebrew by running the command
`ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"`.  Next install ant and CMake by running the commands
`brew install ant` and `brew install cmake`.<br/>

#### Build Xerces

Navigate in a terminal window to the xerces-3.1.2 directory in the %BioGears lib directory.  Build Xerces by running
`./configure --disable-threads --disable-network --enable-transcoder-macosunicodeconverter --disable-pretty-make CXXFLAGS=-O3 CFLAGS=-O3` and then `make`.
<br/>

After Xerces builds successfully, we have to set the dylib's install path.  Navigate to the xerces-c-3.1.2/src/.libs directory, and run the command
`install_name_tool -id @rpath/libxerces-c-3.1.dylib libxerces-c-3.1.dylib`.  This sets the install path of the dylib to @rpath so that the %BioGears executable
will search the current directory when attempting to load it.  Run the command `otool -D libxerces-c-3.1.dylib` to make sure the install path has been
set correctly, this should output `@rpath/libxerces-c-3.1.dylib`.

#### Building and Running %BioGears in Xcode

Navigate in a terminal window to the %BioGears src directory and run `ant cmake -Denv=xcode`.  This will generate debug and release Xcode project files in
src/cmake/xcode.
<br/>

Open the %BioGears.xcodeproj file in Xcode.  Select Product > Scheme > Edit Scheme and select Run on the left column.  Set the following configuration options:

- Set the executable to %BioGearsScenarioDriver if it isn't already.
- Set the build configuration to Debug or Release depending on the desired mode.
- On the arguments tab, add an entry to "Arguments Passed On Launch" which specifies a path to a scenario to run (e.g. verification/Scenarios/Basic/Basic1.xml).
- On the options tab, check "Use custom working directory" and set it to the %BioGears bin folder.

Close the dialog and click the Run button to build and launch the %BioGears scenario driver.
<br/>

#### Building and Running %BioGears from the Command Line

Navigate in a terminal window to the %BioGears src directory and run `ant cmake -Denv=xcode`.  This will generate debug and release Xcode project files in
src/cmake/xcode.
<br/>

To build %BioGears from the command line, open the generated xcodeproj file in Xcode.  Opening the project in Xcode will force the project schemes to generate,
these are necessary for command line builds.  After Xcode loads the project it can be closed.  Run the command `ant compile -Denv=xcode` in %BioGears's
src directory to build.

After %BioGears is built, navigate to the bin directory and launch the scenario driver from there, specifying the scenario you want to run as the first parameter
(e.g. `./Release/BioGearsScenarioDriver verification/Scenarios/Basic/Basic1.xml`).

#### Running the GUI from Eclipse

In addition to configuring Eclipse to run %BioGears as described in the Eclipse section, we also have to specify the absolute path to the BioGearsJNI dylib
so the engine can load it.  Open BioGearsEngine.java and locate the `runScenario` method.  Replace the path in the `System.load(...)` call with the absolute
path to the BioGearsEngineJNI.dylib file (e.g. `System.load("/Users/name/%BioGears/bin/Release/BioGearsEngineJNI.dylib")`).
<br/>

Now right click on the BioGearsGUI project and select Properties.  Select Java Build Path, on the Source tab expand BioGearsGUI/java and set
the Native library location to be the directory where the dylibs are located.  If running in release mode, this will be the release directory in the
%BioGears bin directory.

### Building with GCC on Linux @anchor LINUX

#### Required tools

Make sure the following tools are installed:
- ant
- CMake
- Java JDK 8
- GCC

Add a JAVA_HOME variable to point to the Java installation and add it to the system PATH.  This can be done by adding the following lines to ~/.profile:
```
JAVA_HOME=/usr/local/jdk1.8.0_45
PATH=$PATH:$JAVA_HOME_bin
```

Replace the JAVA_HOME value with the correct installation directory.

#### Build Xerces

Navigate in a terminal window to the xerces-3.1.2 directory in the %BioGears lib directory.  Build Xerces by running
`./configure --disable-threads --disable-network --enable-transcoder-gnuiconv --disable-pretty-make CXXFLAGS=-O3 CFLAGS=-O3` and then `make`.
<br/>

#### Building %BioGears from the Command Line

Navigate in a terminal window to the %BioGears src directory and run `ant cmake -Denv=unixMake` and then `ant compile -Denv=unixMake`.

### Cross-compiling with GCC for the Raspberry Pi @anchor RPI

<b>Note:</b> This build has not been thoroughly tested.

This build is very similar to the GCC Linux build, the only difference being that the code is compiled with a version of GCC that produces ARM-compatible objects.

#### Compiler

Install `gcc-arm-linux-gnueabihf` and `g++-arm-linux-gnueabihf` (e.g. using `sudo apt-get install gcc-arm-linux-gnueabihf` etc.).  In addition to this, make sure gcc 4.9 is installed.

#### Build Xerces

Build Xerces by running
`./configure --disable-threads --disable-network --enable-transcoder-gnuiconv --disable-pretty-make CC=/usr/bin/arm-linux-gnueabihf-gcc CXX=/usr/bin/arm-linux-gnueabihf-g++ CXXFLAGS=-O3 CFLAGS=-O3 --host arm-linux` and then `make`.

#### Building %BioGears from the Command Line

Navigate in a terminal window to the %BioGears src directory and run `ant cmake -Denv=raspberryPi` and then `ant compile -Denv=raspberryPi`.

### Deploying on *nix Platforms

%BioGears can be deployed as an executable or as a library to be used in other applications.  Each of the deployment processes below assumes the code has already been built with `ant compile -Denv=your_environment`.

#### Mac Executable

Deploy the Mac executable by running `deploy-unix-executable.sh` in the %BioGears src directory.  This will deploy the necessary files to the %BioGears deploy/toolkit directory.  To launch the GUI from that directory, run `BioGearsGUI.sh`.

#### Mac SDK

Deploy the Mac SDK by running `deploy-osx-library.sh` in the %BioGears src directory.  This will deploy the necessay header files and .dylib files to the %BioGears library directory.  The HowTo cpp files provide examples of how to use the %BioGears API from your own software.  The build-osx.sh script will build all of the HowTo files and place the resulting executable in the library/bin directory.

#### Linux Executable

Deploy the Linux executable by running `deploy-unix-executable.sh` in the %BioGears src directory.  This will deploy the necessary files to the %BioGears deploy/toolkit directory.  To launch the GUI from that directory, run `BioGearsGUI.sh`.

#### Linux SDK

Deploy the Linux SDK by running `deploy-linux-library.sh` in the %BioGears src directory.  This will deploy the necessay header files and .so files to the %BioGears library directory.  The HowTo cpp files provide examples of how to use the %BioGears API from your own software.  The build-linux.sh script will build all of the HowTo files and place the resulting executable in the library/bin directory.
Physiology Engine Interface {#physeng}
===========================

The following infomation is a method by method walk through of the %PhysiologyEngine.h class that everything revolves around.
Reading this in parrallel with the HowTo-EngineUse.cpp provided in the SDK will give you a firm understanding of using %BioGears.

When you create an instance of a %BioGears engine, you will be returned a pointer to a PhysiologyEngine object.
This generic interface is the controlling class for a physiology engine modeling a single patient.

@code 
  std::unique_ptr<PhysiologyEngine> bg = CreateBioGearsEngine("MyBioGearsEngine.log");
@endcode

@anchor LoggingInterface
### Logging

You can provide an optional string into this creation method that will create a log file on disk where the engine will output all errors, warnings and other logged information.
It is highly recommended to create a log, and check it often. If you do not wish to have this engine create a log, you can pass in nullptr.
You can also provide a callback class to the logger that will forward all logging events to so that your application can easily monitor the %BioGears engine.
For more details, please consult the HowTo-EngineUse.cpp file in the SDK.
You can access and utilize this logger with in your own application as such:
@code
  //--------------------------------------------------------------------------------------------------
	/// \brief
	/// Retrieve the Logger associated with this engine
	//--------------------------------------------------------------------------------------------------
	virtual Logger* GetLogger() = 0;
@endcode

@anchor EngineInitialization
### Initializing the Engine

There are two ways to initialize the engine once created.

@anchor EngineState
#### Engine State

An engine state file is the exact state of the engine written into an xml file. 
Once loaded, the engine is instatly ready to process instructions.
You can specify an xml file on disk, or a state object held in memory.
You can optionally specify a specific simulation time for the engine to use as its initial simulation time.
If no time is provided, the simulation time that is in the state file will be used.
The SDK provides a state file for various patients at simulation time 0 located in the bin/states directory.

@code
  //--------------------------------------------------------------------------------------------------
  /// \brief
  /// Reset engine and set it to the state in the provided file.
  /// You may provided a Simulation Time to be used if desired.
  /// It will be reflected in the GetSimulationTime method.
  /// Return value indicates engine was able to load provided state file.
  /// Engine will be in a cleared state if this method fails.
  //--------------------------------------------------------------------------------------------------
  virtual bool LoadState(const std::string& file, const SEScalarTime* simTime=nullptr) = 0;

  //--------------------------------------------------------------------------------------------------
  /// \brief
  /// Reset engine and set it to the state in the provided file.
  /// You may provided a Simulation Time to be used if desired.
  /// It will be reflected in the GetSimulationTime method.
  /// Return value indicates engine was able to load provided state file.
  /// Engine will be in a cleared state if this method fails.
  //--------------------------------------------------------------------------------------------------
  virtual bool LoadState(const CDM::PhysiologyEngineStateData& state, const SEScalarTime* simTime = nullptr) = 0;
@endcode

At any point during the life of an engine, you create your own state object for later use.
@code
  //--------------------------------------------------------------------------------------------------
  /// \brief
  /// Save the current state of the engine.
  /// State will be written to a file if provided.
  /// State object will be returned.
  /// Engine will be in a cleared state if this method fails.
  //--------------------------------------------------------------------------------------------------
  virtual std::unique_ptr<CDM::PhysiologyEngineStateData> SaveStateToFile(const std::string& file = "") = 0;
@endcode

@anchor PatientCreation
#### Patient Creation

If you would like to create your own patient or apply a condition to a specific patient, 
you will need to initialize the engine with a patient definition.

@code
  //--------------------------------------------------------------------------------------------------
	/// \brief
	/// locates the xml patient file and reads in the values. 
	///
	/// This will create an engine that you can send instructions (patient,actions,conditions) to dynamically.
	/// The return value will indicate success failure of the creation of the engine.  
	/// Some combinations of patients and conditions may prevent the engine from stabilizing
	///
	//--------------------------------------------------------------------------------------------------
	virtual bool InitializeEngine(const std::string& patientFile, const std::vector<const SECondition*>* conditions = nullptr, const PhysiologyEngineConfiguration* config = nullptr) = 0;

	//--------------------------------------------------------------------------------------------------
	/// \brief
	///
	/// This will create an engine that you can send instructions (patient,actions,conditions) to dynamically.
	/// The return value will indicate success failure of the creation of the engine.  
	/// Some combinations of patients and conditions may prevent the engine from stabilizing
	///
	//--------------------------------------------------------------------------------------------------
  virtual bool InitializeEngine(const SEPatient& patient, const std::vector<const SECondition*>* conditions = nullptr, const PhysiologyEngineConfiguration* config = nullptr) = 0;
@endcode

@anchor ConditionsInterface
When a patient definition is provided, %BioGears will go through an initialization algorithm that will take several minutes
as to tune the engine to model the specific state requested of the patient. 
This initalization method is also the only way to specify any conditions (chronic/disease states).

#### Patient Conditions
@secreflist
  @refitem ChronicAnemiaTable "Chronic Anemia"
  @refitem ChronicObstructivePulmonaryDiseaseTable "COPD"
  @refitem ChronicPericardialEffusionTable "Chronic Pericardial Effusion"
  @refitem ChronicRenalStenosisTable "Chronic Renal Stenosis"
  @refitem ChronicVentricularSystolicDysfunctionTable "Chronic Ventricular Systolic Dysfunction"
  @refitem ConsumeMealTable "Consume Meal"
  @refitem ImpairedAlveolarExchangeTable "Impaired Alveolar Exchange"
  @refitem LobarPneumoniaTable "Lobar Pneumonia"
@endsecreflist

#### %Environment Conditions
@secreflist
  @refitem InitialEnvironmentTable "Initial Environment"
@endsecreflist

<br>
Initializing the patient simulation with conditions will extend the initialization time by a few more minutes.
Once the InitalizeEngine method returns, the engine is stabilize, and it is recommended to save the engine state for future use.
(Assuming the patient vitals are acceptable, it may take some adjusting of condition severity to get a desired patient state with conditions applied.)
The SDK provides multiple tested patient files for use in the bin/patients directory. For more information on this look at the @ref PatientMethodology.

@anchor EngineConfiguration
##### Engine Configuration

You can provide an engine configuration to the initialize methods. 
This configuration is used to tweek various variables associated in the %BioGears methodology.
Configuration modification requires a very indepth knowledge of the engine.
It is not recommended to provide another configuration unless you know what effects it will have.
There are some useful configuration options that may be of interest, such as writing data to a csv file while the initialization algorithm executes.
Come visit us on the forums if this is something you want to know more about.

You can retrieve and view the configuration with this method (note you will need to cast it to the BioGearsConfiguration class to see all configuration data)

@code
  //--------------------------------------------------------------------------------------------------
  /// \brief
  /// returns the engine configuration. 	
  //--------------------------------------------------------------------------------------------------
  virtual const PhysiologyEngineConfiguration* GetConfiguration() = 0;
@endcode

@anchor DataTracking
### Data Tracking

%BioGears has the ability to write specifically requested data to a comma seperated text file as the engine advances time.
These csv file are very helpful in debugging and ensuring %BioGears is modeling correctly.
How to utilize this functionality is demonstrated in all the HowTo files provided in the SDK.

@code
  //--------------------------------------------------------------------------------------------------
	/// \brief
	/// Retrieve the PhysiologyEngineTrack associated with tracking data from this engine to a file
	//--------------------------------------------------------------------------------------------------
	virtual PhysiologyEngineTrack* GetEngineTrack() = 0;
@endcode

@anchor AdvanceTime
### Advancing The Simulation In Time

Once the engine is initialized, it is ready to accept direction.
The engine does not advance time on its own, you must explictly tell the engine to simulate a specific amount of time.
There is a HowTo example that shows how to encapsulate the engine in a thread class that automatically advances time and processes actions.
How you decide to drive the simulation is up to you, but you must explicitly advance time inorder for the models to simulate physiology.

@code
  //--------------------------------------------------------------------------------------------------
	/// \brief
	/// executes one pass through the time loop of the engine at the fixed timestep
	///
	/// Events, errors, and warning as are logged to file not errors are returned
	/// through the API at this time. 
	///
	//--------------------------------------------------------------------------------------------------
	virtual void AdvanceModelTime() = 0;

	//--------------------------------------------------------------------------------------------------
	/// \brief
	/// executes one pass through the time loop of the engine at the fixed timestep
	///
	/// Events, errors, and warning as are logged to file not errors are returned
	/// through the API at this time. 
	///
	//--------------------------------------------------------------------------------------------------
	virtual void AdvanceModelTime(double time, const TimeUnit& unit) = 0;
@endcode

If no time is provided, the engine will simulate for the smallest amount of time it can.
The minimum amount of time the engine can simulate is the engines' time step, you can retrieve this duration with this method:

@code
  //--------------------------------------------------------------------------------------------------
	/// \brief
	/// returns the engine time step that is used when advancing time. 
	///
	//--------------------------------------------------------------------------------------------------
	virtual double GetTimeStep(const TimeUnit& unit) = 0;
@endcode

You can retrieve the total amount of time the engine has simulated with the following call:

@code
  //--------------------------------------------------------------------------------------------------
	/// \brief
	/// returns the current time of the simulation. 	
	//--------------------------------------------------------------------------------------------------
	virtual double GetSimulationTime(const TimeUnit& unit) = 0;
@endcode
Note that the simulation time is 0 when the InitializeEngine method returns.
Remember that you can set the simulation time to what ever you want when loading a state.
It will then be incremented from that point on whenever you call AdvanceModelTime.

@anchor ProcessActions
### Process Action

Actions are the means by which instructions are provided to a physiology engine.
You will need to create an instance of an action class, fill it out with the necessary data and pass it into this method.
@code
	//--------------------------------------------------------------------------------------------------
	/// \brief
	/// Execute the provided action.
	/// true will be returned if the engine supports the action
	/// false will be returned if the engine does not support the action
	///
	//--------------------------------------------------------------------------------------------------
	virtual bool ProcessAction(const SEAction& action) = 0;
@endcode

#### Patient Actions
@secreflist
  @refitem AcuteStressTable "Acute Stress"
  @refitem AirwayObstructionTable "Airway Obstruction"
  @refitem ApneaTable "Apnea"
  @refitem AsthmaAttackTable "Asthma Attack"
  @refitem BrainInjuryTable "Brain Injury"
  @refitem BronchoconstrictionTable "Bronchoconstriction"
  @refitem CardiacArrestTable "Cardiac Arrest"
  @refitem ChestCompressionForceTable "Chest Compression Force"
  @refitem ChestCompressionForceScaleTable "Chest Compression Force Scale"
  @refitem ChestOcclusiveDressingTable "Chest Occlusive Dressing"
  @refitem ConsciousRespirationTable "Conscious Respiration"
  @refitem ConsumeNutrientsTable "Consume Nutrients"
  @refitem ExerciseTable "Exercise"
  @refitem HemorrhageTable "Hemorrhage"
  @refitem IntubationTable "Intubation"
  @refitem MechanicalVentilationTable "Mechanical Ventilation"
  @refitem NeedleDecompressionTable "Needle Decompression"
  @refitem PericardialEffusionTable "Pericardial Effusion"
  @refitem SubstanceBolusTable "Substance Bolus"
  @refitem SubstanceCompoundInfusionTable "Substance Compound Infusion"
  @refitem SubstanceInfusionTable "Substance Infusion"
  @refitem TensionPneumothoraxTable "Tension Pneumothorax"
  @refitem UrinateTable "Urinate"
@endsecreflist

#### %Environment Actions
@secreflist
  @refitem EnvironmentChangeTable "Environment Change"
  @refitem ThermalApplicationTable "Thermal Application"
@endsecreflist

#### Anesthesia Machine Actions
@secreflist
  @refitem AnesthesiaMachineConfigurationTable "Anesthesia Machine Configuration"
  @refitem ExpiratoryValveLeakTable "Expiratory Valve Leak"
  @refitem ExpiratoryValveObstructionTable "Expiratory Valve Obstruction"
  @refitem InspiratoryValveLeakTable "Inspiratory Valve Leak"
  @refitem InspiratoryValveObstructionTable "Inspiratory Valve Obstruction"
  @refitem MaskLeakTable "Mask Leak"
  @refitem OxygenTankPressureLossTable "Oxygen Tank Pressure Loss"
  @refitem OxygenWallPortPressureLossTable "Oxygen Wall Port Pressure Loss"
  @refitem SodaLimeFailureTable "SodaLime Failure"
  @refitem TubeCuffLeakTable "Tube Cuff Leak"
  @refitem VaporizerFailureTable "Vaporizer Failure"
  @refitem VentilatorPressureLossTable "Ventilator Pressure Loss"
  @refitem YPieceDisconnectTable "Y-Piece Disconnect"
@endsecreflist

#### %Inhaler Actions
@secreflist
  @refitem InhalerConfigurationTable "Inhaler Configuration"
@endsecreflist

@anchor PatientState
### Patient State

Once you start advancing time and processing actions, the patient state can start to change.
Depending on what actions you process, it can change slightly, such as a slight increase in mean arterial pressure,
or it can change quite dramatically and start affecting multiple systems through various physiological feedback mechanisms. 
These dramatic changes in patient state are tracked by means of @ref PatientEventTable "Patient Events". 

You can track patient event states in one of two ways.

- Poll the patient object via Patient object method:
@code
  //--------------------------------------------------------------------------------------------------
	/// \brief
	/// Returns the patient object used by the engine 
	///
	//--------------------------------------------------------------------------------------------------
	virtual const SEPatient& GetPatient() = 0;
@endcode
Note you can get more @ref PatientTable "Patient" data from this class, not just event status.

- Set up a callback method that will automatically call your code when patient event status changes.
Set the callback method with this method
@code
  //--------------------------------------------------------------------------------------------------
	/// \brief
	/// Add a callback object that will be called whenever a pateint or anesthesia machine event changes state
	//--------------------------------------------------------------------------------------------------
	virtual void SetEventHandler(SEEventHandler* handler) = 0;
@endcode

Event State is primarily associated with the patient, but there are events associated with @ref AnesthesiaMachineEventTable "Anesthesia Machine".

Look at the SDK HowTo-UseEngine.cpp for a full example in using both of these methods.

@anchor AssessmentsInterface
### Patient Assessments

Patient assessments are intended to give general patient overviews, formed at the level of a clinician's report.
The following assessments are availiable from %BioGears:

@secreflist
  @refitem CompleteBloodCountTable "Complete Blood Count"
  @refitem ComprehensiveMetabolicPanelTable "Comprehensive Metabolic Panel"
  @refitem UrinalysisTable "Urinalysis"
  @refitem PulmonaryFunctionTestTable "Pulmonary Function Test"
@endsecreflist 

You must create and provide an assessment object to the physiology engine via this method:<br>
*Note that assessments could add extra comupation time to gather and format data for the assessment.*

@code
  //--------------------------------------------------------------------------------------------------
	/// \brief
	/// Determines the assessment type and fills the data object with current data. 
	///
	/// Assessments can be queried at any point in the calculation and as many times are desired. 
	///
	//--------------------------------------------------------------------------------------------------
	virtual bool GetPatientAssessment(SEPatientAssessment& assessment) = 0; 
@endcode

@anchor SystemsInterface
### Systems Data

The bodies physiology, equipment, and the environment are all systems and each system has a method to retrieve its associated class in order to access this system data.<br>

%BioGears supports the following systems:

|Code Method                                                                             | CDM Table                                           |
|---                                                                                     |---                                                  |
|@code virtual const SEEnvironment* GetEnvironment() = 0; @endcode                       | @ref EnvironmentTable "Environment"                 |
|@code virtual const SEBloodChemistrySystem* GetBloodChemistrySystem() = 0; @endcode     | @ref BloodChemistrySystemTable "BloodChemistry"     |
|@code virtual const SECardiovascularSystem* GetCardiovascularSystem() = 0; @endcode     | @ref CardiovascularSystemTable "Cardiovascular"     |
|@code virtual const SEEndocrineSystem* GetEndocrineSystem() = 0; @endcode               | @ref EndocrineSystemTable "Endocrine"               |
|@code virtual const SEEnergySystem* GetEnergySystem() = 0; @endcode                     | @ref EnergySystemTable "Energy"                     |
|@code virtual const SERenalSystem* GetRenalSystem() = 0; @endcode                       | @ref RenalSystemTable "Renal"                       |
|@code virtual const SEGastrointestinalSystem* GetGastrointestinalSystem() = 0; @endcode | @ref GastrointestinalSystemTable "Gastrointestinal" |
|@code virtual const SENervousSystem* GetNervousSystem() = 0; @endcode                   | @ref NervousSystemTable "Nervous"                   |
|@code virtual const SERespiratorySystem* GetRespiratorySystem() = 0; @endcode           | @ref RespiratorySystemTable "Respiratory"           |
|@code virtual const SEDrugSystem* GetDrugSystem() = 0; @endcode                         | @ref DrugSystemTable "Drug"                         |
|@code virtual const SETissueSystem* GetTissueSystem() = 0; @endcode                     | @ref TissueSystemTable "Tissue"                     |
|@code virtual const SEAnesthesiaMachine* GetAnesthesiaMachine() = 0; @endcode           | @ref AnesthesiaMachineTable "Anesthesia Machine"    |
|@code virtual const SEElectroCardioGram* GetElectroCardioGram() = 0; @endcode           | @ref ElectroCardioGramTable "ElectroCardioGram"     |
|@code virtual const SEInhaler* GetInhaler() = 0; @endcode                               | @ref InhalerTable "Inhaler"                         |

@anchor CompartmentsInterface
### Compartments

A compartment represents the fluid dynamics of an anatomical organ or equipment component. 
Comparments can represent various fidelities of data for these compoenents, such as:
	- An anatomical space, such as the body's skin, muscles
	- An organ, such as the liver
	- An organ substructure, such as the Left Heart
	- Extravascular tissue of an organ
	- A component of a piece of equipment, such as an anesthesia machine ventillator
  
The following compartment types are used to represent various anatomical structures for both physiology and equipment:

@secreflist
 @refitem GasCompartmentTable "Gas Compartment"
 @refitem LiquidCompartmentTable "Liquid Compartment"
 @refitem ThermalCompartmentTable "Thermal Compartment"
 @refitem TissueCompartmentTable "Tissue Compartment"
@endsecreflist 

Various comparments can be used to represent a parent/child hierarcical structure, and comparments can be a parent for one or more child compartments.
Parent compartments agregate data from their children, such as the parent compartment volume is simply the sum of each of its childrens' volume.<br>
For eample, the heart has the following hierarchy :

  - Heart
    - Myocardium
    - Right Heart
    - Left Heart
    - Pericardium

Compartments also contain a substance quatity for each substance in the compartment as it moves through the body/equipment. <br>
The following types are used to hold compartment substance information

@secreflist
 @refitem GasSubstanceQuantityTable "Gas Substance Quantity"
 @refitem LiquidSubstanceQuantityTable "Liquid Substance Quantity"
@endsecreflist

The enumerations for compartments available in %BioGears is found in the BioGearsPhysiologyEngine.h file.
As these are programatic enumerations, you can use the auto-complete feature of your favorite programming IDE to view these enumerations as you code.
%BioGears discritizes it's compartments into enumerations based on fluid type and equipment. Here is a list of the various enumerated compartment names:

@secreflist
 @refitem LiquidCompartmentTable  "Chyme"
 @refitem GasCompartmentTable     "Pulmonary"
 @refitem LiquidCompartmentTable  "Pulmonary(Aerosols)"
 @refitem TissueCompartmentTable  "Tissue"
 @refitem LiquidCompartmentTable  "Extracellular"
 @refitem LiquidCompartmentTable  "Intracellular"
 @refitem LiquidCompartmentTable  "Vascular"
 @refitem LiquidCompartmentTable  "Urine"
 @refitem LiquidCompartmentTable  "Lymph"
 @refitem ThermalCompartmentTable "Temperature"
 @refitem GasCompartmentTable     "Environment"
 @refitem GasCompartmentTable     "Anesthesia Machine"
 @refitem GasCompartmentTable     "Inhaler"
@endsecreflist

As we make changes to models, some of the naming/hierarchy of compartments used by %BioGears could change, it's best to refer to the code file itself 
for the latest list of available compartments and their hierarchy.

Note that there are both liquid and gas compartments available for pulmonary spaces. 
The gas compartments represent the air flow and gaseous substances through the pulmonary tract.
The liquid compartments represent the air flow and aerosolized liquid/solid substances through the pulmonary tract.

All compartments are accessed via the SECompartmentManager retrieved from this method:

@code
  //--------------------------------------------------------------------------------------------------
  /// \brief
  /// Retrieves the engine compartments, providing such data as:
  /// flows, pressure, volume as well as substance volumes and volume fractions.
  ///
  //--------------------------------------------------------------------------------------------------
  virtual const SECompartmentManager& GetCompartments() = 0;
@endcode

@anchor SubstancesInterface
#### Substances

Various Substances are available in the body and there is a SESubstanceManager associated with the engine. <br>
You can retrieve the SESubstanceManager with this method:

@code
  //--------------------------------------------------------------------------------------------------
	/// \brief
	/// Retrieves the associated substance manager.
	///
	//--------------------------------------------------------------------------------------------------
	virtual SESubstanceManager& GetSubstanceManager() = 0;
@endcode

In practice, only a substance reference are used with compartments to get the Substance Quantity object associated with a particular compartment.
The data contained in the @ref SubstanceTable "substance definition" is primarily used in engine methodology, 
but there may be substance data you maybe interested in, such as the @ref Substance_SubstanceTissuePharmacokineticsData "Partition Coefficient" of a substance.


%BioGears FAQ {#FAQ}
============

@insert MainPageFAQ.md

@insert ExtraFAQ.md


 



 







 






Publications {#published}
=======
Team Publications:

1. McDaniel, Matthew, and Austin Baird. "A Full-Body Model of Burn Pathophysiology and Treatment Using the BioGears Engine." In 2019 41st Annual International Conference of the IEEE Engineering in Medicine and Biology Society (EMBC), pp. 261-264. IEEE, 2019.
2. McDaniel, Matthew, Jonathan Keller, Steven White, and Austin Baird. "A Whole-Body Mathematical Model of Sepsis Progression and Treatment Designed in the BioGears Physiology Engine." Frontiers in physiology 10 (2019): 1321.
3. McDaniel, Matthew, Jennifer Carter, Jonathan M. Keller, Steven A. White, and Austin Baird. "Open source pharmacokinetic/pharmacodynamic framework: tutorial on the BioGears Engine." CPT: Pharmacometrics & Systems Pharmacology 8, no. 1 (2019): 12-25.
4. Baird, Austin, Carter, Jennifer, McDaniel, Matthew, and Welch, Bennett. "BioGears: An In-Silico Whole-Body Framework to Simulate Kinetics and Dynamics of Pharmaceuticals and Associated Reversal Agents." International Meeting on Simulation in Healthcare. Los Angeles, Ca. 2018
5. Baird, Austin, Carter, Jennifer, Kainz, James, Magee, Joel, McDaniel, Matt, Miller, Geoffry, Metoyer, Rodney, and Welch, Bennett. "BioGears: Simulating Whole-Body Response to Chemical Exposure." Chemical and Biological Defense Science and Technology Conference. Long Beach, Ca. 2017. 
6. Potter, Lucas, Arikatla, Sreekanth, Bray, Aaron, Webb, Jeff, and Enquobahrie, Andinet. "Physiology informed virtual surgical planning: a case study with a virtual airway surgical planner and BioGears." SPIE Medical Imaging Conference on Image-Guided Procedures, Robotic Interventions, and Modeling. Orlando, FL. 2017.
7. Metoyer, Rodney, Carter, Jenn, Bergeron, Bryan, Baird, Austin, Bray, Aaron, Clipp, Rachel B., Thames, M. Cameron, Webb, Jeff. "A Framework for Multiscale Physiology: Towards Individualized Computer Simulation."  VPH2016, book of abstracts, A.G. Hoekstra (Editor), University of Amsterdam, (Amsterdam), ISBN 978-90-826254-0-0, 2016.
8. Clipp, Rachel B, Bray, Aaron M, Metoyer, Rodney Jr, Thames, M. Cameron, and Webb, Jeffrey B. "Pharmacokinetic and Pharmacodynamic Modeling in BioGears" IEEE Engineering in Medicine and Biology Conference. Orlando, FL. 2016.
9. Fisher, Charles E, Frank, Randall J, Argenta, Chris, Petrovitch, Chris, Oxford, Eddie, Pamplin, Jeremy C, Baker Jr, William L, Wendroff, Daniel S, Cancio, Leopoldo C, Batchinsky, Andriy I. "Forecasting the Trajectory of Blood Loss from Vital Signs Collected at the Bedside: a Principal Component Analysis Approach." 15th International Conference on Complex Acute Illness. Pasadena, CA. 2016.
10. Clipp, Rachel B, Thames, M. Cameron, Webb, Jeffrey B, Metoyer, Rodney Jr, Swarm, Zachary M, Bray, Aaron M, and Carter, Jenn N. "Integration of a Baroreflex Model Into a Whole Body Physiology Engine." ASME Summer Biomechanics, BioTransport, and BioEngineering Conference. National Harbor, MD. 2016.
11. Clipp, Rachel B., Webb, Jeffrey B., Thames, M. Cameron, Swarm, Zachary, Metoyer, Rodney, and Bray, Aaron. Pharmacokinetic and Pharmacodynamic Modeling in %BioGears. Medicine Meets Virtual Reality Conference. Los Angelos, CA. 2016.
12. Metoyer, Rodney, Bergeron, Bryan, Clipp, Rachel B., Webb, Jeffrey B., Thames, M. Cameron, Swarm, Zachary, Carter, Jenn, Gebremichael, Y. , and Heneghan, Jeremiah. Multiscale Simulation of Insults and Interventions: The %BioGears Showcase Scenarios. Medicine Meets Virtual Reality Conference. Los Angelos, CA. 2016.
13. Swarm, Zachary, Webb, Jeffrey B., Clipp, Rachel B., Carter, Jenn, Thames, M. Cameron, Metoyer, Rodney, and Bray, Aaron and Byrd, David. Modeling %Renal Behavior and Control in %BioGears. Medicine Meets Virtual Reality Conference. Los Angelos, CA. 2016.
14. Thames, M. Cameron, Webb, Jeffrey B., Clipp, Rachel B., Carter, Jenn, Swarm, Zachary, Metoyer, Rodney, and Bray, Aaron and Byrd, David. Dynamic Response to Heat Gain and Heat Loss in the %BioGears Engine. Medicine Meets Virtual Reality Conference. Los Angelos, CA. 2016.
15. Bergeron, Bryan. "BioGears - It's Alive!" Servo Magazine, December 2015.
16. Gebremichael, Y., Clipp, R., Webb, J., Bray, A., Thames, M. Cameron, Swarm, Z., Carter, J., and Heneghan, J. Integration Of A Spontaneous %Respiratory Driver with Blood Gas Feedback into %BioGears, an Open-Source, Whole-Body Physiology Model. Summer Biomechanics, bioengineering and Biotransport Conference. Snowbird Resort, Utah. 2015
17. Metoyer, Rodney, and Smith, Beth. Modeling the Time-Dependent Intrapericardial Pressure-Volume Relationship. World Congress of Biomechanics. Boston, MA. 2014.
18. Clipp, Rachel B. and Scott, Gabriel. HumanSim: A Physiology Engine for the Simulation of Anesthesia/Anaphylaxis, Military Health Research Symposium. Fort Lauderdale, FL. 2012.

External Publications:
---
1. Kauchak, Marty. "Game Changer." MedSim Magazine, page 21, Volume 3, Issue 2, 2014.
2. Barnes III, John Jacob, Kenneth Kiberenge, Robert Sweet, Jon Keller, and Mojca R. Konia. "Comparing Hemorrhage in Human Physiology Simulation Tools: How They Compare With Expected Human Physiology and Each Other." Simulation in Healthcare 15, no. 5 (2020): 310-317.
3. Hananel, D. M. "David Marko Hananel." Comprehensive Healthcare Simulation: Surgery and Surgical Subspecialties (2019): 3.
4. Sweet, Robert M. Advanced Modular Manikin. University of Minnesota Minneapolis United States, 2020.
Third Party Credits {#Credits}
===================

%BioGears was developed using the following software tools:

---

## <a href="http://www.cmake.org/">CMake</a>

Usage: Cross-platform build process using compiler-independent methods.

<img src="./images/Credits/CMake.png" alt="CMake" align="left"> <br> <br> <br> <br> <br> <br>

---

## <a href="http://www.stack.nl/~dimitri/doxygen/">Code Synthesis</a>

Usage: XML data binding compiler to C++.

<img src="./images/Credits/CodeSynthesis.png" alt="CodeSynthesis" align="left"> <br> <br> <br>

---

## <a href="http://eigen.tuxfamily.org/index.php?title=Main_Page">Eigen</a>

Usage: Matrix manipulation and linear algebra solver for the circuit solver, substance transporter, and saturation calcuations.

<img src="./images/Credits/Eigen.jpg" alt="Eigen" align="left"> <br> <br> <br> <br> <br> <br> <br> <br>

---

## <a href="http://log4cpp.sourceforge.net/">log4cpp</a>

Usage: Logging.

<img src="./images/Credits/SourceForge.png" alt="SourceForge" align="left"> <br> <br> <br> <br> <br> <br> <br>

---

## <a href="http://www.jfree.org/jfreechart/">JFreeChart</a>

Usage: Creating image plots of output data.

<img src="./images/Credits/SourceForge.png" alt="SourceForge" align="left"> <br> <br> <br> <br> <br> <br> <br>

---

## <a href="http://jchart2d.sourceforge.net/">JChart2D</a>

Usage: Creating realtime plots of output data.

<img src="./images/Credits/SourceForge.png" alt="SourceForge" align="left"> <br> <br> <br> <br> <br> <br> <br>

---
## <a href="http://ant.apache.org/">Apache Ant</a>

Usage: Compiling and running Java applications.

<img src="./images/Credits/Ant.gif" alt="Ant" align="left"> <br> <br> <br> <br> <br>

---

## <a href="http://www.stack.nl/~dimitri/doxygen/">Doxygen</a>

Usage: HTML documentation generation.

<img src="./images/Credits/doxygen.png" alt="Doxygen" align="left"> <br> <br> <br> <br>

---

## <a href="http://strawberryperl.com/">Strawberry Perl</a>

Usage: Required by Doxygen for citations and creating a references page.

<img src="./images/Credits/Perl.png" alt="Perl" align="left"> <br> <br> <br> <br> <br>

---

## <a href="http://miktex.org/">MiKTeX</a>

Usage: Required by Doxygen for citations and creating a references page.

<img src="./images/Credits/MiKTeX.png" alt="MiKTeX" align="left"> <br> <br> <br> <br>

---

## <a href="https://code.google.com/p/reflections/">Reflections</a>

Usage: Reflectively analyze classes for dynamic GUI generation.
Scenario XML Files {#ScenarioXMLFile} 
==================

%BioGears can be used to simulate various physiological scenarios.
The Common Data Model (CDM) provides a Scenario structure that can contain a set of instructions that can be used to drive %BioGears. 
Below you can see :
- How a scenario is structured in XSD Schema
- XML examples of all the actions, conditions, assessments supported by %BioGears 

A Scenario is a 'canned' instruction set with requested data to be output in a comma delimited file that is executed by the engine and will produce the same results data.
'FATAL' is used below to note boundary cases that will result in a fatal exception, stopping the engine.
	
If you would like execute scenarios, the CDM contains a class, @ref SEScenarioExec, that can execute any scenario with a physiology engine that implements the PhysiologyEngine interface.
The @ref Toolkit, also provides example scenario files and can also execute scenario files.

An example of a basic scenario file is shown below.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<?xml version="1.0" encoding="UTF-8"?>
<Scenario xmlns="uri:/mil/tatrc/physiology/datamodel" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsdVersion="v16.12" contentVersion="BioGears_6.0.1-beta" xsi:schemaLocation="">
  <Name>InitialStableState</Name>
  <Description>Start the engine at a state, this is for debugging said state</Description>
  <EngineStateFile>./states/StandardMale@0s.xml</EngineStateFile>
  
  <DataRequests>
	  <DataRequest xsi:type="PhysiologySystemDataRequestData" Name="HeartRate" Unit="1/min"/>
	  <DataRequest xsi:type="PhysiologySystemDataRequestData" Name="RespirationRate" Unit="1/min"/>
    <DataRequest xsi:type="PhysiologySystemDataRequestData" Name="OxygenSaturation" Unit="unitless"/>
    
	  <DataRequest xsi:type="GasCompartmentDataRequestData" Compartment="Carina"       Substance="Oxygen"        Name="PartialPressure" Unit="cmH2O" Precision="0"/>
	  <DataRequest xsi:type="GasCompartmentDataRequestData" Compartment="Carina"       Substance="CarbonDioxide" Name="PartialPressure" Unit="cmH2O" Precision="1"/>
	  	
	  <DataRequest xsi:type="LiquidCompartmentDataRequestData" Compartment="Aorta"     Substance="Oxygen"        Name="PartialPressure" Unit="mmHg" Precision="1"/>
	  <DataRequest xsi:type="LiquidCompartmentDataRequestData" Compartment="Aorta"     Substance="CarbonDioxide" Name="PartialPressure" Unit="mmHg" Precision="1"/>
  </DataRequests>
   
  <Action xsi:type="AdvanceTimeData">
    <Time value="2" unit="min"/>		
  </Action>
</Scenario>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The scenario allows for a name and description, but this is not used in execution.
The scenario contains the following execution information:
- What engine state to use
- A list of values to return from the engine
- A list of actions to execute over the course of the run

- - -

## Patient File and optional conditions

While it is recommened to use an Engine State when running a scenario, you do have the option to initialize the engine with a Patient File and optional conditions.
The specified patient file refers to an XML file containing @ref Patient_PatientData information.
Replace the <EngineStateFile> tag with the <InitialParameters> tag like this:

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<?xml version="1.0" encoding="UTF-8"?>
<Scenario xmlns="uri:/mil/tatrc/physiology/datamodel" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsdVersion="v16.12" contentVersion="BioGears_6.0.1-beta" xsi:schemaLocation="">
    <Name>Anemia30</Name>
    <Description>Anemia onset, leading to 30% reduction in hemoblogin content</Description>
    <InitialParameters>
		  <PatientFile>StandardMale.xml</PatientFile>
		  <Condition xsi:type="ChronicAnemiaData">
			  <ReductionFactor value="0.3"/>
		  </Condition>
	  </InitialParameters>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

## Patient Conditions

Conditions give instructions to the engine to apply certian changes to the engine to simulate the specified conditions.
The following are links to the Condition class specification along with XML examples of conditions that can be used in making your own scenarios.

#### Chronic Anemia
@copybrief PatientConditions_ChronicAnemiaData
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Condition xsi:type="ChronicAnemiaData">
 <ReductionFactor value="0.3"/>
</Condition>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#### COPD
@copybrief PatientConditions_ChronicObstructivePulmonaryDiseaseData
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Condition xsi:type="ChronicObstructivePulmonaryDiseaseData">
  <BronchitisSeverity  value="0.65"/>
	<EmphysemaSeverity   value="0.50"/>
</Condition>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#### Chronic Pericardial Effusion
@copybrief PatientConditions_ChronicPericardialEffusionData
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Condition xsi:type="ChronicPericardialEffusionData">
    <AccumulatedVolume value="500" unit="mL"/>
 </Condition>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#### Chronic %Renal Stenosis
@copybrief PatientConditions_ChronicRenalStenosisData
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Condition xsi:type="ChronicRenalStenosisData"> 
    <LeftKidneySeverity value="0.9"/>
    <RightKidneySeverity value="0.9"/>
</Condition>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#### Chronic Ventricular Systolic Dysfunction
@copybrief PatientConditions_ChronicVentricularSystolicDysfunctionData
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Condition xsi:type="ChronicVentricularSystolicDysfunctionData"/>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#### Consume Meal
@copybrief PatientConditions_ConsumeMealData
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Condition xsi:type="ConsumeMealData">
	<Meal>
		<Water value="1.0" unit="L"/>
		<ElapsedTime value="12.0" unit="hr"/>
	</Meal>
</Condition>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#### Impaired Alveolar Exchange
@copybrief PatientConditions_ImpairedAlveolarExchangeData
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Condition xsi:type="ImpairedAlveolarExchangeData">
		<ImpairedSurfaceArea value="1.0" unit="m^2"/>
</Condition>

or

<Condition xsi:type="ImpairedAlveolarExchangeData">
		<ImpairedFraction value="0.20"/>
</Condition>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#### Lobar Pneumonia
@copybrief PatientConditions_LobarPneumoniaData
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Condition xsi:type="LobarPneumoniaData">
	<Severity            value="0.70"/>
	<LeftLungAffected    value="0.00"/>
	<RightLungAffected   value="0.67"/>
</Condition>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#### Initial %Environment
@copybrief EnvironmentConditions_InitialEnvironmentData
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Condition xsi:type="InitialEnvironmentData">
	<Conditions>
		<SurroundingType>Air</SurroundingType>
		<AirVelocity value="0.0" unit="m/s"/>
		<AmbientTemperature value="20.0" unit="degC"/>
		<AtmosphericPressure value="525.0" unit="mmHg"/>
		<ClothingResistance value="3.0" unit="clo"/>
		<Emissivity value="0.9"/>
		<MeanRadiantTemperature value="20.0" unit="degC"/>
		<RelativeHumidity value="1.0"/>
		<RespirationAmbientTemperature value="20.0" unit="degC"/>
		<AmbientSubstance Name="Nitrogen">
			<FractionAmount value="0.8576"/>
		</AmbientSubstance>
		<AmbientSubstance Name="Oxygen">
			<FractionAmount value="0.142"/>
		</AmbientSubstance>
		<AmbientSubstance Name="CarbonDioxide">
			<FractionAmount value="4.0E-4"/>
		</AmbientSubstance>
	</Conditions>
</Condition>

or

<!-- file must be in the ./bin/environemnts directory -->
<Condition xsi:type="InitialEnvironmentData">
	<ConditionsFile>Hypobaric3000m.xml</ConditionsFile>
</Condition>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - - 

## Data Requests

Currently there are four supported types of data requests:

### Physiology System Data

Physiology System data refers to all the data specified by SystemData and its derived types.

At this time, you do not need to specify the system name.
The Name attribute should be set to a System Property name. (e.g., HeartRate)

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<DataRequest xsi:type="PhysiologySystemDataRequestData" Name="HeartRate" Unit="1/min" />
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

### Compartment Data

Compartments refer to all the data specified on Compartments.
You can read more about compartments @ref CompartmentsInterface "here".

There are two amin types of Compartments, gas and liquid.

Data on the compartment itself: 
The Compartment attribute can be any of the enumerations defined by %BioGears. 
The Name attribute should be set to a Compartment Property name.
The Substance attribute is optional, and the if used the name will refer to a substance quantity property.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<DataRequest xsi:type="GasCompartmentDataRequestData" Compartment="LeftAlveoli" Name="Pressure" Unit="cmH2O"/>		
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<DataRequest xsi:type="GasCompartmentDataRequestData" Compartment="LeftAlveoli" Substance="Oxygen" Name="PartialPressure" Unit="mmHg"/>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<DataRequest xsi:type="LiquidCompartmentDataRequestData" Compartment="Aorta" Name="Pressure" Unit="mmHg" />
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<DataRequest xsi:type="LiquidCompartmentDataRequestData" Compartment="Aorta" Substance="Oxygen" Name="PartialPressure" Unit="mmHg"/>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

### Environment Data

Environment System data refers to all the data specified by @ref Environment_EnvironmentData and its derived types.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<DataRequest xsi:type="EnvironmentDataRequestData"  Name="ConvectiveHeatLoss"    Unit="W"          Precision="2"/>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

### Equipment Data

System level data from a piece of equipment

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<DataRequest xsi:type="EquipmentDataRequestData"  Type="ECG" Name="Lead3ElectricPotential"    Unit="mV"          Precision="0"/>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

### Substance Data

Substance data is provided about a substance and its current state in the body and on specific anatomy compartments

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<DataRequest xsi:type="SubstanceDataRequestData" Substance="Morphine" Name="PlasmaConcentration" Unit="ug/mL"/>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<DataRequest xsi:type="SubstanceDataRequestData" Substance="Morphine" Compartment="LeftKidneyTissue" Name="MassCleared" Unit="ug"/>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -

## General Actions

Actions give instructions to the engine to define what happens over the course of a scenario. 
Everything from advancing time, to starting a hemorrhage, to administering a drug is an action of some kind. 
The following are links to the Action class specification along with XML examples of actions that can be used in making your own scenarios.

- - -

#### Advance Time

@copybrief Scenario_AdvanceTimeData
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="AdvanceTimeData">
	<Time value="350" unit="s"/>       
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#### Serialize State
@copybrief Scenario_SerializeStateData
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="SerializeStateData" Type="Save">
		<Filename></Filename><!-- No Filename, engine is going to auto generate a name, BioGears will do something like : StandardMale@0s.xml -->		
  </Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -

Patient Insults
---------------
- - -

#### Acute Stress
@copybrief PatientActions_AcuteStressData <br>
Severity value must be >=0.0 and <=1.0 <br>
A severity of 0 removes the action completely.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="AcuteStressData">
   <Severity value="0.3"/>       
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -

#### Apnea
@copybrief PatientActions_ApneaData <br>
Severity value must be >=0.0 and <=1.0 <br>
A severity of 0 removes the action completely.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="ApneaData">
   <Severity value="0.3"/>       
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -

#### Airway Obstruction 
@copybrief PatientActions_AirwayObstructionData <br>
Severity value must be >=0.0 and <=1.0 <br>
A severity of 0 removes the action completely.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="AirwayObstructionData">
   <Severity value="0.3"/>       
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -

#### Asthma Attack
@copybrief PatientActions_AsthmaAttackData <br>
Severity value must be >=0.0 and <=1.0 <br>
A severity of 0 removes the action completely.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="AsthmaAttackData">
	<Severity value="0.3"/>       
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -

#### Brain Injury
@copybrief PatientActions_BrainInjuryData <br>
Severity value must be >=0.0 and <=1.0 <br>
A severity of 0 removes the action completely.<br>
Types : Diffuse, LeftFocal, RightFocal
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="BrainInjuryData" Type="Diffuse">
	<Severity value="0.3"/>       
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -
		
#### Bronchoconstriction 
@copybrief PatientActions_BronchoconstrictionData <br>
Severity value must be >=0.0 and <=1.0 <br>
A severity of 0 removes the action completely.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="BronchoconstrictionData">
	<Severity value="0.3"/>       
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -

#### Consume Nutrients 
@copybrief PatientActions_ConsumeNutrientsData
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="ConsumeMealData">
	<Nutrition>
		<Carbohydrate value="390.0" unit="g"/>
		<CarbohydrateDigestionRate value="0.5" unit="g/min"/>
		<Fat value="90.0" unit="g"/>
		<FatDigestionRate value="0.055" unit="g/min"/>
		<Protein value="56.0" unit="g"/>
		<ProteinDigestionRate value="0.071" unit="g/min"/>
		<Calcium value="1000.0" unit="mg"/>
		<Sodium value="1.5" unit="g"/>
		<Water value="3.7" unit="L"/>
	</Nutrition>
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -

#### Cardiac Arrest 
@copybrief PatientActions_CardiacArrestData
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="CardiacArrestData"/>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -
	
#### Exercise 
@copybrief PatientActions_ExerciseData <br>
An intensity of 0 removes the action completely.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="ExerciseData">
	<Intensity value="1.0"/>
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="ExerciseData">
	<BorgScale value="10.0"/>
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -
	
#### Hemorrhage 
@copybrief PatientActions_HemorrhageData <br>
The Compartment attribute can be any of the enumerations defined in the enumAnatomy enumeration.<br>
FATAL: Cannot have bleeding rate greater than cardiac output or less than 0
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="HemorrhageData" Compartment="RightLegVascular">
    <Rate value="250" unit="mL/min"/>
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -

#### Pericardial Effusion
@copybrief PatientActions_PericardialEffusionData <br>
EffusionRate of the liquid
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="PericardialEffusionData" >
	<EffusionRate value="0.1" unit="mL/s"/>
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -
	
#### Tension Pneumothorax 
@copybrief PatientActions_TensionPneumothoraxData <br>
The Type attribute can be "Open" or "Closed"<br>
The Side attribute can be "Left" or "Right"<br>
Severity value must be >=0.0 and <=1.0 <br>
A severity of 0 removes the action completely.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="TensionPneumothoraxData" Type="Closed" Side="Left">
        <Severity value="0.6"/>       
    </Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -

Patient Interventions
---------------------

- - -
	
#### Chest Compression Force 
@copybrief PatientActions_ChestCompressionForceData <br>
Force is the specific magnitude to perform a compression with.<br>
Note, that patient should be in Cardiac Arrest before performing CPR
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="ChestCompressionForceData">
    <Force value="100.0" unit="N"/>
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -
		
#### Chest Compression Force Scale 
@copybrief PatientActions_ChestCompressionForceScaleData <br>
ForceScale value must be >=0.0 and <=1.0<br>
Note, that patient should be in Cardiac Arrest before performing CPR
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="ChestCompressionForceScaleData"> 
	<ForceScale value="0.73"/>
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -
		
#### Chest Occlusive Dressing 
@copybrief PatientActions_ChestOcclusiveDressingData <br>
The State attribute can be "On" or "Off" <br>
Side is either Left or Right<br>
FATAL: If the side specified does not have a pnumothorax 
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="ChestOcclusiveDressingData" State="On" Side="Left">
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -

#### Conscious Respiration 
@copybrief PatientActions_ConsciousRespirationData <br>
This action can contain 1 or more commands :
- @copybrief PatientActions_ForcedInhaleData
- @copybrief PatientActions_ForcedExhaleData
- @copybrief PatientActions_BreathHoldData
- @copybrief PatientActions_UseInhalerData

<br>
Commands will be processed in order.
The first commands will be proceesed instantly
When it has completed (run through it's Period),
the next command will be processed.
Other actions will be processed while these commands
are processed or waiting to be processed.
You may want to advance time for the sum of the 
command periods to ensure the body is doing what you 
expect it to.. Or not, depending on how you want 
the system to react.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="ConsciousRespirationData">
	<Command xsi:type="ForcedExhaleData">
		<ExpiratoryReserveVolumeFraction value="1.0"/>
		<Period value="3.0" unit="s"/>			
	</Command>  
	<Command xsi:type="ForcedInhaleData">
		<InspiratoryCapacityFraction value="1.0"/>
		<Period value="5.0" unit="s"/>          
	</Command> 
	<Command xsi:type="UseInhalerData"/>	
	<Command xsi:type="BreathHoldData">
		<Period value="10.0" unit="s"/>          
	</Command> 
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -

#### Intubation 
@copybrief PatientActions_IntubationData <br>
Note: In order to 'turn off' an intubation, use'Off' as the Type  <br>
Types : Off, Esophageal, LeftMainstem, RightMainstem, Tracheal
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="IntubationData" Type="Tracheal"/>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -

#### Mechanical Ventilation 
@copybrief PatientActions_MechanicalVentilationData <br>
You may provide Pressure and/or Flow. <br>
If you do not provide GasFractions, the environment gas fractions will be used. <br>
If you do provide Gas Fractions, they must add up to 1.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="MechanicalVentilationData" State="On">
  <Pressure value="10.0" unit="cmH2O"/>
  <Flow value="1.0" unit="mL/s"/>
  <GasFraction Name="Oxygen">
    <FractionAmount value="1.0"/>
  </GasFraction>
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -
	
#### Needle Decompression 
@copybrief PatientActions_NeedleDecompressionData <br>
The Side attribute can be "Left" or "Right"<br>
The State attribute can be "On" or "Off"
FATAL: If the side specified does not have a pnumothorax 
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="NeedleDecompressionData" State="On" Side="Left"/>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#### Urinate 
@copybrief PatientActions_UrinateData <br>
Action to empty the bladder. if not emptied, 
it will empty and throw an event.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="UrinateData"/>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -
	
#### Substance Bolus  
@copybrief PatientActions_SubstanceBolusData

The AdminRoute can be one of:
- "Intravenous"
- "Intramuscular"
- "Subcutaneous"
- "Oral"
- "Rectal"
- "Inhaled"

The Substance element should be set to a name of any of the %Substances.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="SubstanceBolusData" AdminRoute="Intravenous">
	<Substance>Succinylcholine</Substance>
	<Concentration value="4820" unit="ug/mL"/>
	<Dose value="30" unit="mL"/>
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -
	
#### Substance Compound Infusion Fluids 
@copybrief PatientActions_SubstanceCompoundInfusionData <br>
The Substance Compound element should be set to a name of any of the %Substances Compounds. <br>
Set Rate to 0 to remove Action
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="SubstanceCompoundInfusionData">
	<SubstanceCompound>Saline</SubstanceCompound>
	<BagVolume value="500" unit="mL"/>
	<Rate value="100" unit="mL/min"/>
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -
	
#### Substance Infusion 
@copybrief PatientActions_SubstanceInfusionData <br>
The Substance element should be set to a name of any of the %Substances. <br>
Set Rate to 0 to remove Action
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="SubstanceInfusionData" State="On" AdminRoute="Intravenous">
    <Substance>Succinylcholine</Substance>
    <Concentration value="5000" unit="ug/mL"/>
    <Rate value="100" unit="mL/min"/>
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -

Anesthesia Machine State
------------------------

- - -
#### Anesthesia Machine Configuration 
@copybrief AnesthesiaActions_AnesthesiaMachineConfigurationData <br>
NOTE: Each field is optional. <br>
Connection can be one of : Off, Mask, Tube <br>
Patient must be intubated to be connected as Tube <br>
Anesthesia machine will be disconneted if intubation is removed. <br>
Patient cannot be intubated to be connected as Mask <br>
Anesthesia machine will be disconneted if patient is then intubated. <br>
Cannot have inhaler and anesthesia machine on at the same time <br>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="AnesthesiaMachineConfigurationData">
	<Configuration>
    <Connection>Mask</Connection>
		<InletFlow value="5.0" unit="L/min"/>
		<InspiratoryExpiratoryRatio value="0.5"/>
		<OxygenFraction value="0.25"/>
		<OxygenSource>Wall</OxygenSource>
		<PositiveEndExpiredPressure value="1.0" unit="cmH2O"/>
		<PrimaryGas>Nitrogen</PrimaryGas>
		<RespiratoryRate value="16.0" unit="1/min"/>
		<VentilatorPressure value="10.5" unit="cmH2O"/>
		<OxygenBottleOne>
			<Volume value="660" unit="L"/>
		</OxygenBottleOne>
		<OxygenBottleTwo>
			<Volume value="660" unit="L"/>
		</OxygenBottleTwo>
	</Configuration>
</Action>    
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="AnesthesiaMachineConfigurationData" >
	<Configuration>
		<Connection>Off</Connection>
	</Configuration>
</Action> 
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~	
<Action xsi:type="AnesthesiaMachineConfigurationData">
	<Configuration>
		<LeftChamber>
      <State>On</State>
			<SubstanceFraction value="0.04"/>					
			<Substance>Desflurane</Substance>
		</LeftChamber>
	</Configuration>
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
	
- - -

Anesthesia Machine Incidents
----------------------------

- - -
	
#### Oxygen TankPressure Loss 
@copybrief AnesthesiaActions_OxygenTankPressureLossData <br>
The State attribute can be "On" or "Off"
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="OxygenTankPressureLossData" State="On"/>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -
	
#### Oxygen Wall Port Pressure Loss 
@copybrief AnesthesiaActions_OxygenWallPortPressureLossData <br>
The State attribute can be "On" or "Off"
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="OxygenWallPortPressureLossData" State="On"/>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -

Anesthesia Machine Failures
---------------------------

- - -
	
#### Expiratory Valve Leak 
@copybrief AnesthesiaActions_ExpiratoryValveLeakData <br>
The State attribute can be "On" or "Off"<br>
Severity value must be >=0.0 and <=1.0
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="ExpiratoryValveLeakData" State="On">
    <Severity value="0.5"/>
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -
	
#### Expiratory Valve Obstruction 
@copybrief AnesthesiaActions_ExpiratoryValveObstructionData <br>
The State attribute can be "On" or "Off"<br>
Severity value must be >=0.0 and <=1.0
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="ExpiratoryValveObstructionData" State="On">
    <Severity value="1.0"/>
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -
	
#### Inspiratory Valve Leak 
@copybrief AnesthesiaActions_InspiratoryValveLeakData <br>
The State attribute can be "On" or "Off"<br>
Severity value must be >=0.0 and <=1.0
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="InspiratoryValveLeakData" State="On">
    <Severity value="1.0"/>
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -
	
#### Inspiratory Valve Obstruction 
@copybrief AnesthesiaActions_InspiratoryValveObstructionData <br>
The State attribute can be "On" or "Off"<br>
Severity value must be >=0.0 and <=1.0
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="InspiratoryValveObstructionData" State="On">
    <Severity value="1.0"/>
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -
	
#### Mask Leak 
@copybrief AnesthesiaActions_MaskLeakData <br>
The State attribute can be "On" or "Off"<br>
Severity value must be >=0.0 and <=1.0
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="MaskLeakData" State="On">
    <Severity value="1.0"/>
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -
	
#### Soda Lime Failure 
@copybrief AnesthesiaActions_SodaLimeFailureData <br>
The State attribute can be "On" or "Off"<br>
Severity value must be >=0.0 and <=1.0
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="SodaLimeFailureData" State="On">
    <Severity value="1.0"/>
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -
	
#### Tube Cuff Leak 
@copybrief AnesthesiaActions_TubeCuffLeakData <br>
The State attribute can be "On" or "Off"<br>
Severity value must be >=0.0 and <=1.0
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="TubeCuffLeakData" State="On">
    <Severity value="0.5"/>
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -
	
#### Vaporizer Failure 
@copybrief AnesthesiaActions_VaporizerFailureData <br>
The State attribute can be "On" or "Off"<br>
Severity value must be >=0.0 and <=1.0
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="VaporizerFailureData" State="On">
    <Severity value="0.25"/>
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -
	
#### Ventilator Pressure Loss 
@copybrief AnesthesiaActions_VentilatorPressureLossData <br>
The State attribute can be "On" or "Off"<br>
Severity value must be >=0.0 and <=1.0
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="VentilatorPressureLossData" State="On">
    <Severity value="1.0"/>
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -
	
#### Y Piece Disconnect 
@copybrief AnesthesiaActions_YPieceDisconnectData <br>
The State attribute can be "On" or "Off"<br>
Severity value must be >=0.0 and <=1.0 
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="YPieceDisconnectData" State="On">
    <Severity value="1.0"/>
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -

%Inhaler State
------------------------

#### %Inhaler Configuration 
@copybrief InhalerActions_InhalerConfigurationData <br>
FATAL: Cannot have inhaler and anesthesia machine on at the same time
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="InhalerConfigurationData">
	<Configuration>
		<Substance>Albuterol</Substance>
		<MeteredDose value="90.0" unit="ug"/>
		<NozzleLoss value="0.04"/>
	</Configuration>
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- - -

%Environment 
------------------------

- - -
	
#### %Environment Configuration State <br>
@copybrief EnvironmentActions_EnvironmentChangeData <br>
NOTE: Each field is optional.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="EnvironmentChangeData">
	<Conditions>
        <SurroundingType>Water</SurroundingType>
        <AirVelocity value="0.0" unit="m/s"/>
        <AmbientTemperature value="10.0" unit="degC"/>
        <AtmosphericPressure value="760.0" unit="mmHg"/>
        <ClothingResistance value="0.01" unit="clo"/>
        <Emissivity value="0.0"/>
        <MeanRadiantTemperature value="22.0" unit="degC"/>
        <RelativeHumidity value="1.0"/>
        <RespirationAmbientTemperature value="22.0" unit="degC"/>
        <AmbientSubstance Name="Nitrogen">
            <FractionAmount value="0.7901"/>
        </AmbientSubstance>
        <AmbientSubstance Name="Oxygen">
            <FractionAmount value="0.2095"/>
        </AmbientSubstance>
        <AmbientSubstance Name="CarbonDioxide">
            <FractionAmount value="4.0E-4"/>
        </AmbientSubstance>
	</Conditions>
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#### Thermal Application <br>
@copybrief EnvironmentActions_ThermalApplicationData <br>
You must provide at least 1 activity, but up can also 
apply upto all 3 in one action.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="ThermalApplicationData">
	<ActiveHeating>
		<Power value="500" unit="BTU/hr"/> 
		<SurfaceAreaFraction value="0.2" unit="unitless"/>  		
	</ActiveHeating>
	<ActiveCooling>
		<Power value="500" unit="BTU/hr"/> 
		<CSurfaceArea value="0.1" unit="m^2"/> 			
	</ActiveCooling>
	<AppliedTemperature>
    <State>On</State>
		<Temperature value="30" unit="degF"/> 
		<SurfaceAreaFraction value="1.0" unit="unitless"/> 			
	</AppliedTemperature>
</Action>	
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<Action xsi:type="ThermalApplicationData">		
	<AppliedTemperature>
    <State>On</State>
		<Temperature value="140" unit="degF"/> 
		<SurfaceAreaFraction value="0.9" unit="unitless"/> 			
	</AppliedTemperature>
</Action>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
﻿%Energy Methodology {#EnergyMethodology}
==========================

Overview
========

@anchor energy-abstract
Abstract
--------
The %BioGears&reg; %Energy system provides a methodology for both thermal and metabolic regulation within the body. Heat transfer is modeled by a body thermal [circuit](@ref CircuitMethodology) that is connected to the [environment](@ref EnvironmentMethodology) circuit. A heat source at the core of the body circuit represents the metabolic heat production. In addition to generating heat, the dynamic metabolic rate is also used to compute the metabolic production and consumption rates of gases and other substances in the body. This system has been [validated](@ref energy-results) both under resting and dynamic conditions, such as exercise and external environment changes.

Introduction
------------
The %Energy system is responsible for simulating metabolism and regulating internal temperature. Metabolism is simulated through the consumption of nutrients in the tissues. Glucose and fat are the nutrients most readily used as an energy source for the human body. Oxygen is used to metabolize these nutrients during aerobic metabolism. After entering the pulmonary capillaries, oxygen is transported via the circulatory system and diffuses into the tissue where it is used in the aerobic reaction. The resulting by-product, carbon dioxide, diffuses out of tissue, traveling through the circulation before exiting the body through the pulmonary capillaries @cite guyton2006medical. Glucose can also be used to produce energy under anaerobic conditions, and lactate (lactic acid) is the resulting by-product. The lactate enters the circulation and can be reconverted to glucose via gluconeogenesis in the liver and kidneys. During periods of prolonged starvation, the liver produces ketones. Acetoacetate is the predominant ketone synthesized. After it is produced, acetoacetate may be used as an energy source for the brain, myocardium, and muscle @cite gropper2013nutrition. 

The body's temperature is well-regulated. Core temperature regulation can be achieved by varying heat production (e.g. shivering) or heat exchange (e.g. sweating). In %BioGears, variations in the external environment or physical activity can lead to changes in the core temperature, triggering thermal feedback. Examples of thermal feedback during shifts in core temperature are sweating and shivering. Sweating is initiated if the core temperature is too high @cite herman2007physics. The increased evaporation of water from the surface of the skin leads to an increase in heat transferred out of the body. Changes in core temperature directly affect other systems. For example, a low core temperature causes blood flow to be shunted away from the skin, resulting in less heat transfer from the core to the skin @cite guyton2006medical. These feedback mechanisms are used to maintain thermal homeostasis in a variety of environmental conditions.

@anchor energy-system-design
System Design
=============

Background and Scope
--------------------

### Requirements
The %BioGears energy system is required to simulate the effects of exercise and elevated physical activity as well as react to temperature and pressure changes in the environment.

### Approach
The %BioGears %Energy system is a physical model of heat transfer combined with a collection of empirical equations for heat production and exchange obtained from literature. Additional equations are derived from stoichiometric relationships and empirical data found in the literature, particularly the governing equations for the metabolic [production and consumption](@ref tissue-metabolic-production) of substances. It uses a thermal circuit to simulate heat transfer through the body, with the body circuit connected to the [environment](@ref EnvironmentMethodology) circuit. The body thermal circuit consists of a core node, representing core temperature, and a skin node, which represents the lumped peripheral temperature. The body thermal circuit is shown in Figure 1.

<img src="./images/Energy/internalThermal.png" width="400">
<center>
*Figure 1. The body thermal circuit consists of two nodes and four paths. Two additional paths exist, connecting to the environment thermal circuit. The circuit is used to model the dynamic core and skin temperatures.*
</center><br>

The path elements consist of an internal heat flow source, a core heat capacitance, a skin heat capacitance, and a variable resistor from the core to skin. These four elements represent metabolic heat generation, the discretized heat capacity of the human body, and the convective heat transfer due to blood flow. The heat capacities of the core and skin are computed by proportioning the mass-averaged heat capacity found in literature based on the mass fraction of the skin @cite herman2007physics. The variable core to skin resistance is computed dynamically during simulation, and is inversely proportional to the skin blood flow. A decrease in blood flow leads to an increase in heat transfer resistance, and vice versa. When connected to the [environment](@ref EnvironmentMethodology) thermal circuit, the core and skin temperatures dynamically react to the environmental conditions (e.g. temperature, pressure, and humidity).

The [metabolic production and consumption](@ref tissue-metabolic-production) of the %Energy system is determined by the rate of change of substances (nutrients, ions, gases) in the tissue, relying on [advective](@ref CircuitMethodology) and [diffusive](@ref TissueMethodology) transport methodologies.

### Thermal Regulation
Thermal regulation in the %Energy system occurs through manipulation of the metabolic rate or through external losses (sweating). The thermal feedback mechanisms are a direct implementation of those discussed by Herman @cite herman2007physics. For high core temperature, a control equation specifies the sweat rate as a function of the difference between the current core temperature and the set-point (Equation 1). Sweat is removed from the body via a path connected between the skin and the environment (Figure 2).

<img src="./images/Energy/sweat.png" width="400">
<center>
*Figure 2. Sweat is removed from the body via a flow source path connected to ground. The flow source rate is computed using Equation 1.*
</center><br>

\f[ \dot{m}_{sweat} = \frac{h_{sweat} \left(T_{core} - T_{core,setpoint} \right)}{\rho_{sweat}} \f]
<center>
*Equation 1.*
</center><br>

For severe increases in core temperature (hyperthermia), the rate of chemical reactions can increase, leading to a subsequent rise in metabolic rate. This value was quantified from Pate et al. @cite pate2001thermal as an eleven percent increase in metabolic rate for every degree increase in core temperature. This is described from the power relationship between metabolic rate and core temperature, shown in Equation 2.

\f[ \dot{Q}^{n+1}_{metabolic} = \dot{Q}^{n}_{metabolic} * \left(1.11 \right)^{T_{core}-T_{core,high}} \f]
<center>
*Equation 2.*
</center><br>

For low core temperatures, the maximum metabolic rate achievable through shivering is defined from the summit metabolism. This value is given by Herman as:

\f[ \dot{Q}_{metabolic} = 21.0m^{0.75}_{b} \f]
<center>
*Equation 3.*
</center><br>

If the summit metabolism is not enough to adequately maintain internal temperature, then the metabolic rate will begin to decrease at a rate of four percent for every degree drop @cite mallet2002hypothermia. The power relationship in Equation 4 is used to describe this phenomenon.

\f[ \dot{Q}^{n+1}_{metabolic} = \dot{Q}^{n}_{metabolic} * \left(0.96 \right)^{T_{core,low}-T_{core}} \f]
<center>
*Equation 4.*
</center><br>

Under resting conditions, the metabolic rate is determined from the Harris-Benedict formula. This formula gives the metabolic requirements in kilo-calories per day as a function of weight, height, age, and sex @cite roza1984metabolic . This empirical formula is shown in Equation 5.

<center>
<table border="0">
<tr>
<td> Men: </td> <td> \f[ \dot{Q}_{metabolic} = 88.632 + 13.397*m_{b} + 4.799*H - 5.677*A \f] </td>
</tr>
<tr>
<td> Women: </td> <td> \f[ \dot{Q}_{metabolic} = 447.593 + 9.247*m_{b} + 3.098*H - 4.330*A \f] </td>
</tr>
</table>
</center>
<center>
*Equation 5.*
</center><br>

@anchor energy-data-flow
Data Flow
---------
The %Energy System determines its state at every timestep through a three-step process: Preprocess, Process, and Postprocess. In general, Preprocess determines the circuit element values based on feedback mechanisms and engine settings/actions. Process uses the generic circuit calculator to compute the entire state of the circuit and fill in all pertinent values. Postprocess is used to advance time.

### Preprocess
#### Calculate Metabolic Heat Generation
The metabolic rate is dictated by the current state of the core temperature. This function includes states of increased metabolic rate due to shivering or severe hyperthermia. Additionally, an extreme drop in core temperature leads to decreasing metabolic rate. If the core temperature does not meet any of the criteria for increases/decreases, the metabolic rate will be calculated as the basal metabolic rate.

#### Calculate Sweat Rate
The sweat rate is calculated from a core temperature control function. The mass lost due to sweating is accounted for, and a flow source from the skin extravascular to ground path is updated to ensure fluid loss.  Sodium, potassium, and chloride are removed from the body (specifically, the skin tissue extracellular compartment) as a function of the amount of fluid lost and each ion's respective average
concentration in sweat ([Na<sup>+</sup>]=51.0 mM, [Cl<sup>-</sup>]=40.0 mM, [K<sup>+</sup>]=6.0 mM @cite shirreffs1997whole, @cite maughan2004fluid). 

#### Update Heat Resistance 
The variable core to skin heat transfer resistance is updated according to the inverse of the skin blood flow. 

#### Exercise
The exercise function adds to the body&rsquo;s basal metabolic rate a value that is specified by the exercise action. The metabolic rate set-point is specified by the action but limited by the amount of energy available. The energy available depends on the body's nutrient stores and ability to supply oxygen to the muscle tissue, where all exercise work occurs.

### Process

#### Process Temperature Circuit
The generic circuit methodology developed for the %BioGears Engine is used to solve for the temperature and heat transfer rate at each node or path. For more details, see @ref CircuitMethodology.

#### Calculate Vital Signs
The core and skin temperatures are recorded in this function. In addition, the current metabolic state of the patient may trigger the following [events](@ref energy-events): hypothermia, hyperthermia, and metabolic acidosis/alkalosis. These events are only triggered if the current state falls within the criteria of the specific event.

### Postprocess
The Postprocess step moves everything calculated in Process from the next timestep calculation to the current timestep calculation. This allows all other systems access to the information when completing their preprocess analysis for the next timestep.

<br>
<img src="./images/Energy/EnergyDataFlow.png" width="900">
<center>
<i>Figure 3. The data flow for the %Energy System consists of Preprocess, Process, and Postprocess. Preprocess determines the circuit element values based on feedback mechanisms engine actions. Process uses the generic @ref CircuitMethodology to solve the temperature circuit for temperatures and heat transfer rates. Postprocess updates these quantities to the next time step and then begins advancing time, which causes the loop to repeat.</i>
</center><br>

@anchor energy-features
Features, Capabilities, and Dependencies
----------------------------------------
The %Energy system is connected with every other %BioGears system. %Energy directly affects the %Cardiovascular System through modification of vascular tone, heart rate, and heart contractility. This leads to modifications in the cardiac output and flow distribution in the cardiovascular circuit. In addition, there is indirect feedback through the modification of oxygen consumption and carbon dioxide production. Increased carbon dioxide production can lead to higher arterial carbon dioxide, which, in the case of severe hypercapnia, causes an %Endocrine response in the form of epinephrine and norepinephrine release. There is direct feedback on the %Respiratory system via modification of the ventilation frequency and driver pressure. The previously mentioned oxygen consumption and carbon dioxide production changes indirectly based on feedback to %Respiratory driver, which is dependent on the arterial carbon dioxide and arterial oxygen. There is a dependence on the %Gastrointestinal and %Renal systems since they are the methods by which nutrients enter the body and metabolic wastes are cleared from the body. The %Tissue System's freeing and storing of nutrients, such as hepatic and muscle glycogen, affects the total achievable metabolic rate requested in the %Energy System.

The %Environment circuit is directly connected to the energy&rsquo;s internal temperature circuit to form the total temperature circuit. Modifications on the metabolic rate will therefore have a direct feedback on the environment temperature circuit, while changes in the ambient environment temperature will have a direct feedback on the metabolic rate.

Additional dependence on the cardiovascular system comes from the extravascular to vascular connection. The cardiovascular circuit is connected to the extravascular circuit via singular paths from the tissue-equivalent vascular node to the tissue node. These singular paths are formed by a flow resistance and contain a compliance which stores the extravascular volume.

@anchor energy-variability
### Patient Variability
The basal metabolic rate is computed from the sex, height, weight, and age of the patient using the Harris-Benedict formula and will be directly affected by patient variability. Other indirect effects, such as fluid compositional changes due to sweating, may be observed. As with all of the %BioGears systems, the energy system is validated using the %BioGears Standard Male patient. A detailed discussion of patient variability in %BioGears is available in the @ref PatientMethodology report.

@anchor energy-assumptions
Assumptions and Limitations
---------------------------
The %Energy system assumes a bulk representation of some of its metabolic substances to avoid tracking all possible biochemical reactants and products. It assumes that all carbohydrates are stored in the tissue as glucose. A similar assumption is made for lipids. The bulk representation of all lipids is tripalmitin. It is not converted into its fatty acid constituents in the blood stream. This assumption is carried through to ketones. All ketones are assumed to be acetoacetate; beta-hydroxybutyrate and acetone are not modeled. Furthermore, when amino acids are considered from a metabolic standpoint, it can be assumed that they are all alanine. It is assumed that all of these substances move with the flow into and out of the extravascular space.

@anchor energy-actions
Actions
-------
@anchor energy-exercise
### Exercise
The exercise action is initiated by specifying any of four exercise types for the patient to participate in: Generic Exercise, Cycling Exercise, Running Exercise, or Strength Exercise.

Generic Exercise
Generic exercise is an overarching exercise type in BioGears meant to simulate any intensity or desired work rate not achievable through other exercise types. The exercise intensity is a number between 0 and 1 that defines the requested fraction of the body's maximal work rate, which if not specified in the patient file is calculated using the patient's age and gender @cite plowman2013exercise. For instance, the average adult male has a maximal work rate of approximately 1200 Watts @cite hall2011guyton; therefore, a user-specified exercise intensity of 0.5 would request a work rate of 600 Watts from the body. Table 1 shows the approximate exercise intensity for some common activities executed by an average adult male. The approximate mechanical power produced by an adult male is also included in the table.

Cycling Exercise
Cycling exercise simulates a subject on a bicycle and calculates metabolic exertion using the input parameters of cadence and power. Cadence is a frequency value referring to the number of rotations of the pedals in a given time period. In cycling, power is a expressed as work  or energy produced from pedalling. In BioGears, a weighted pack can also be attatched to the subject for the duration of the exercise to increase intensity. 

Running Exercise
Running exercise simulates cardiovascular exercise by means of walking/running using input parameters of speed and incline. Speed is a length per unit time measurement to identify pace. Incline is a fractional representation of percent incline (whereas a 2 percent incline would be equivalent to 0.02 in BioGears). Additionally, a weighted pack can also be attatched to the subject for the duration of the exercise to increase intensity. 

Strength Exercise
Strength exercise offers a basic representation of body weight and weighted strength exercises through weight and repetitions inputs. Weight refers to the additional mass (zero for body weight exercises) being lifted per repeition while the number of repeitions is a scalar value. The current model estimates a generic maximum rep strength exercise to be approximately twice the patients body weight, though this will hopefully be an adjustable parameter in future expansions.


<br><center>
*Table 1. The approximate exercise intensity for some common activities and the corresponding approximate mechanical power (desired work rate). Note that the values in the table correspond to a body with a 1200 Watt maximum work rate capability.*
</center>
|Activity |Exercise Intensity | Mechanical Power (W) |
|-------- |----------------- |--------------------- |
|All muscles in the body working all-out @cite hall2011guyton                                |1.0  |1200 |
|All-out cycling with resistance, sustainable for approx. 8 seconds @cite billaut2003maximal |0.8  |1000 |
|Rowing 6.1 m/s on an ergometer @cite concept2016concept                                     |0.54 |650  |
|Running at about 7.1 m/s (16 mph) @cite johnson2000exercise                                 |0.36 |430  |
|'Typical' Level Cycling @cite herman2007physics                                             |0.1  |120  |
|Jogging at about 2.2 m/s (5 mph) @cite johnson2000exercise                                  |0.06 |70   |
|Rest  |0.0 |0 |

The exercise capacity of the body is physiologically and psychologically limited @cite noakes2012fatigue. The amount of work produced by the exercise action in %BioGears is limited by the nutrients available for consumption; there is currently no psychological fatigue implemented in %BioGears. For more information about the consumption of nutrients and their relation to available energy, see the @ref TissueMethodology.

There are three output parameters associated with the exercise action: the achieved exercise level, the fatigue level, and the total work rate level.
- **Achieved exercise level**: This is the fraction of requested work that the body is able to achieve. For instance, if a work rate of 300 Watts is requested and the body is able to deliver 300 Watts, then the achieved exercise level will be 1.0. After some time, the body's energy levels will be depleted, and it will only be able to produce a fraction of the requested work rate, so the achieved exercise level will be less than 1.0.
- **Fatigue level**: This is the exhaustion level of the body. A fatigue level of 1.0 indicates that the body has no energy available to do work. At low work rates the fatigue level increases slowly, and the rate of fatigue increase is related to type of metabolism producing the energy (i.e. aerobic or anaerobic).
- **Total work rate level**: This is the fraction of the full work rate capacity of the body at which the body is currently working. For example, if the rested body is capable of producing a maximum of 1200 Watts of power for some amount of time, then that body working a rate of 600 Watts would have a work rate level of 0.5. This output is similar to the exercise intensity input.

@anchor energy-conditions
Conditions
----------

### Starvation
The starvation functionality simulates an extended time period without nutrient intake. This is accomplished by depleting nutrient storage and blood substance concentrations based on the duration of starvation. For relatively short (<5 days) periods of starvation, both the body's hepatic glycogen and temporary protein stores are depleted, and so these substances' starved values can be set based on long-duration %BioGears runs, in which the behaviors of substances are already validated. Similarly, stored fat, muscle glycogen, and muscle mass deplete at nearly linear rates, so these can also be interpolated based on long-duration %BioGears runs. Body weight and other patient parameters based on mass proportions are set based on these values. Blood concentrations of glucose and ketones don't deplete in a linear fashion, so those starvation values are set based on data from @cite zauner2000resting, @cite garber1974hepatic, @cite owen1971human, and @cite elia1984mineral. These blood concentrations use a second-order fit to data up to 3 days, after which the concentrations are set to a constant, lowered value.

### Dehydration
**The dehydration condition is disabled in the current release. An improved dehydration condition will be included with the next release.**
Dehydration functionality is simulated completely in the %Energy system. Similar to starvation, dehydration assumes a coarse timestep over which the fluid loss would occur. According to the Journal of Sports Medicine and Physical Fitness @cite shirreffs2000hydration , the average loss rate on the first day without liquid intake is 2600 milliliters. This rate decreases to 1600 milliliters for each subsequent day. The %Energy system uses these average rates and the time since last liquid consumption to calculate volume lost. The volume decrement is distributed via volume weighting instantaneously to all vascular and tissue compartments. 

### Hyperhidrosis and Hypohidrosis
**Excessive sweating and a lack of sweating are conditions that can be modified by the "Hyperhidrosis" parameter in the patient file. Using a scalar from -1 to 1, a more negative value will scale down the patient's sweat rate more in order to simulate hypohidrosis while a more positive value increases sweat rate to demonstrate hyperhidrosis @cite shih1983autonomic. The default value of zero refers to the standard sweat rate of a person, while a value of -1 can simulate the rare condition in which a person is unable to sweat at all.
<img src="./images/Energy/SweatRate.png" width="900">
<img src="./images/Energy/SkinTemp.png" width="900">
<img src="./images/Energy/CoreTemp.png" width="900">


@anchor energy-events
Events
------
### Hypothermia
Hypothermia is defined as a decrease in core temperature below 35 degrees Celsius @cite mallet2002hypothermia. In %BioGears, this is triggered due to changes in the external environment, leading to increased heat transfer off the skin surface.

### Hyperthermia
Hyperthermia is defined as an increase in core temperature above 38.8 degrees Celsius @cite mallet2002hypothermia . This may result from vigorous exercise, harsh environments, or infection. In %BioGears, this event is achievable through the exercise action or by specifying a high-temperature environment.

### Dehydration
Dehydration is classified by the World Health Organization as a three percent decrease in patient body mass due to fluid loss @cite who2005dehydration. This is tested by summing the total fluid mass on the cardiovascular and extravascular circuits at each timestep. Currently, this event can be triggered during a hemorrhage action or through prolonged exercise.

### Fatigue
The fatigue event is triggered whenever a tissue cannot provide the work it is asked to provide. Essentially, this means that a tissue either lacked oxygen or any available nutrients to fully meet its energy demand. The fatigue event is removed when there is no longer a deficit.

@anchor energy-results
Results and Conclusions
=======================

Validation - Resting Physiologic State
--------------------------------------
<br><center>
*Table 2. The %Energy system properties consist of core and skin temperature, as well as metabolic production rates. The properties show good agreement with the validation values.*
</center>
@insert doc/validation/EnergyValidationTable.md

The resting system properties for the energy system are temperatures and metabolic production rates. From Table 2, it is shown that the temperatures meet the validation criteria extremely well for resting conditions. Ketone production rate is currently low because only the brain uses ketones, and then only during starved states. Because ketone behavior is most important during starved states, it was tuned to perform at better long-term timescales, where this validation occurs only over a few minutes. Lactate production is low likely due to its clearance not being tied to gluconeogenesis. This will be addressed in an upcoming release. Resting state temperatures are good.

Validation - Actions and Conditions
--------------------
The %Energy and %Environment systems were validated for six scenarios encompassing a variety actions and environments. These include: exercise, cold water submersion (including active heating), carbon monoxide exposure, smoke inhalation, and a high altitude environment change. These scenarios were meant to test the dynamic feedback of the energy environment systems on both thermal regulation and internal physiology. There is good agreement with the expected results.

<center>
*Table 3. The dynamic response from the %Energy and %Environment systems were validated for six scenarios: Two Exercise scenarios, Cold Water Submersion, a carbon monoxide scenario, a forest fire scenario, and High Altitude. The scenarios demonstrate a good response to exercise and environmental conditions.*
</center>
|	Scenario 	|	Description	|	Good	|	Decent	|	Bad	|
|	---	|	---	|	---	|	---	|	---	|
|	ColdWaterSubmersion	|	Patient is submerged in 10 C water for 1 hour. Removed for 10 minutes, and then actively heated for 20 minutes	|<span class="success">	17	</span>|<span class="warning">	2	</span>|<span class="danger">	5	</span>|
|	HighAltitude	|	Patient enters 4000 m elevation environment for 15 minutes	|<span class="success">	3	</span>|<span class="warning">	3	</span>|<span class="danger">	0	</span>|
|	ExerciseStages	|	Exercise for six minutes at increasing intensity with 2 minutes of rest in between stages.	|<span class="success">	78	</span>|<span class="warning">	12	</span>|<span class="danger">	22	</span>|
|	ExerciseVO2max	|	Exercise at high intensity (430 watts) to exhaustion.	|<span class="success">	1	</span>|<span class="warning">	0	</span>|<span class="danger">	0	</span>|
|	ExerciseSustained	|	Exercise at moderate intensity to demonstrate sweat loss capability.	|<span class="success">	0	</span>|<span class="warning">	1	</span>|<span class="danger">	0	</span>|
|	CarbonMonoxide	|	Varied exposure to environmental carbon monoxide.	|<span class="success">	20	</span>|<span class="warning">	4	</span>|<span class="danger">	0	</span>|
|	ForestFireFighter	|	Exposure to smoke and carbon monoxide.	|<span class="success">	5	</span>|<span class="warning">	1	</span>|<span class="danger">	0	</span>|
|	Starvation	|	Patient consumes no nutrients for 72 hours.	|<span class="success">	10	</span>|<span class="warning">	1	</span>|<span class="danger">	2	</span>|
|		|	Total	|<span class="success">	134	</span>|<span class="warning">	24	</span>|<span class="danger">	29	</span>|


@anchor energy-exercise-validation
### Exercise
There are four scenarios for validation of the exercise action. The first scenario (Tables 4a-d) takes advantage of the abundance of experiments using variations of the Bruce protocol to study exercise physiology. In this scenario, the exercise intensity is periodically increased, with short breaks in between each exercise period. The second scenario (Table 5) qualitatively validates that a work rate threshold exists. The scenario is based on the assertion by @cite johnson2000exercise that a body working around the threshold for maximal oxygen uptake (approximately 430 watts) will only be able to perform at the desired level for about three minutes before reaching physiological limits. 
In the third scenario (Table 6 and Figure 4), an individual exercises at a moderate to heavy intensity for an hour and the total fluid and ion loss from sweat is assessed. Lastly, Table 7 demonstrates the change in core and skin temperatures during exercise in specific environmental conditions.  The quantitative validation results are shown in the table below.

<center><br>
*Table 4a. The dynamic response to exercise. 15 measures are compared to data reported in literature at 8 different times during increasing exercise intensity with intermittent rest, for a total of 120 points of comparison. The first intensity/rest period is shown in Table 4a.*
</center>
|	Segment	|	Notes	|	Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Core Temperature (Celsius)	|	Peripheral Temperature (Celsius)	|	Heart Rate (beats/min)	|	Systolic Pressure (mmHg)	|	Diastolic Pressure (mmHg)	|	Mean Arterial Pressure (mmHg)	|	Stroke Volume (mL)	|	Cardiac Output (L/min)	|	Gut Blood Flow (L/min)	|	Systemic Vascular Resistance (mmHg-min/L)	|	Respiration Rate (breaths/min)	|	Tidal Volume (L)	|	%Respiratory Exchange Ratio	|	Oxygen Consumption (L/min)	|	Urine Production Rate (mL/min)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Exercise Severity 0.0375	|	45 watts	|	30	|	389	|<span class="success">	Increase @cite christie1987cardiac, @cite byrne2007ingestible	</span>|<span class="success">	No change or decrease @cite griffin1993temperature	</span>|<span class="warning">	156% of baseline @cite christie1987cardiac	</span>|<span class="warningr">	Increase 18.4% @cite christie1987cardiac	</span>|<span class="warning">	No change @cite christie1987cardiac	</span>|<span class="success">	1% increase  @cite christie1987cardiac	</span>|<span class="danger">	41.6% increase  @cite christie1987cardiac	</span>|<span class="danger">	225% of base  @cite christie1987cardiac	</span>|<span class="danger">	reduced  @cite perko1998gutflow	</span>|<span class="success">	decrease  @cite perko1998gutflow	</span>|<span class="success">	Increase  @cite guenette2007respiratory	</span>|<span class="success">	increase  @cite guenette2007respiratory	</span>|<span class="danger">	increase, maximum  ~1.21  @cite koutlianos2013indirect	</span>|<span class="danger">	1.33  @cite christie1987cardiac	</span>|<span class="success">	decrease  @cite melin2001comparison	</span>|
|	Exercise Severity 0	|	rest 2 minutes	|	390	|	509	|<span class="success">	Increase @cite christie1987cardiac or no change @cite byrne2007ingestible	</span>|<span class="success">	No change or decrease @cite griffin1993temperature	</span>|<span class="success">	Decreasing 	</span>|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|<span class="success">	No change @cite christie1987cardiac	</span>|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|	No Data	|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|	No Data	|

<center><br>

*Table 4b. The second intensity/rest period.*
</center>
|	Segment	|	Notes	|	Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Core Temperature (Celsius)	|	Peripheral Temperature (Celsius)	|	Heart Rate (beats/min)	|	Systolic Pressure (mmHg)	|	Diastolic Pressure (mmHg)	|	Mean Arterial Pressure (mmHg)	|	Stroke Volume (mL)	|	Cardiac Output (L/min)	|	Gut Blood Flow (L/min)	|	Systemic Vascular Resistance (mmHg-min/L)	|	Respiration Rate (breaths/min)	|	Tidal Volume (L)	|	Exchange Ratio (Or %Respiratory Quotient)	|	Oxygen Consumption (L/min)	|	Urine Production Rate (mL/min)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Exercise Severity 0.075	|	90 watts	|	510	|	869	|<span class="success">	Increase @cite christie1987cardiac, @cite byrne2007ingestible	</span>|<span class="success">	No change or decrease @cite griffin1993temperature	</span>|<span class="warning">	197% of baseline @cite christie1987cardiac	</span>|<span class="warning">	34.4% increase @cite christie1987cardiac	</span>|<span class="warning">	No change @cite christie1987cardiac	</span>|<span class="success">	8.38% increase  @cite christie1987cardiac	</span>|<span class="danger">	29.9% increase  @cite christie1987cardiac	</span>|<span class="danger">	267% of base  @cite christie1987cardiac	</span>|<span class="danger">	reduced  @cite perko1998gutflow	</span>|<span class="success">	decrease  @cite perko1998gutflow	</span>|<span class="success">	Increase  @cite guenette2007respiratory	</span>|<span class="success">	increase  @cite guenette2007respiratory	</span>|<span class="danger">	increase, maximum  ~1.21  @cite koutlianos2013indirect	</span>|<span class="danger">	2.05  @cite christie1987cardiac	</span>|<span class="success">	decrease  @cite melin2001comparison	</span>|
|	Exercise Severity 0	|	rest 2 minutes	|	870	|	989	|<span class="success">	Increase @cite christie1987cardiac or no change @cite byrne2007ingestible	</span>|<span class="success">	No change or decrease @cite griffin1993temperature	</span>|<span class="success">	Decreasing 	</span>|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|<span class="success">	No change @cite christie1987cardiac	</span>|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|	No Data	|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|	No Data	|

<center><br>

*Table 4c. The third intensity/rest period.*
</center>
|	Segment	|	Notes	|	Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Core Temperature (Celsius)	|	Peripheral Temperature (Celsius)	|	Heart Rate (beats/min)	|	Systolic Pressure (mmHg)	|	Diastolic Pressure (mmHg)	|	Mean Arterial Pressure (mmHg)	|	Stroke Volume (mL)	|	Cardiac Output (L/min)	|	Gut Blood Flow (L/min)	|	Systemic Vascular Resistance (mmHg-min/L)	|	Respiration Rate (breaths/min)	|	Tidal Volume (L)	|	Exchange Ratio (Or %Respiratory Quotient)	|	Oxygen Consumption (L/min)	|	Urine Production Rate (mL/min)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Exercise Severity 0.1583	|	190 W	|	990	|	1349	|<span class="success">	Increase @cite christie1987cardiac, @cite byrne2007ingestible	</span>|<span class="success">	No change or decrease @cite griffin1993temperature	</span>|<span class="success">	237% of baseline @cite christie1987cardiac	</span>|<span class="warning">	46.4% increase @cite christie1987cardiac	</span>|<span class="warning">	No change @cite christie1987cardiac	</span>|<span class="warning">	14.1% increase  @cite christie1987cardiac	</span>|<span class="danger">	36.4% increase  @cite christie1987cardiac	</span>|<span class="danger">	336% of base  @cite christie1987cardiac	</span>|<span class="danger">	reduced 25-50%  @cite perko1998gutflow	</span>|<span class="success">	decrease  @cite perko1998gutflow	</span>|<span class="success">	Increase  @cite guenette2007respiratory	</span>|<span class="success">	increase  @cite guenette2007respiratory	</span>|<span class="danger">	increase, maximum  ~1.21  @cite koutlianos2013indirect	</span>|<span class="danger">	2.75  @cite christie1987cardiac	</span>|<span class="success">	decrease  @cite melin2001comparison	</span>|
|	Exercise Severity 0	|	rest 2 minutes	|	1350	|	1469	|<span class="success">	Increase @cite christie1987cardiac or no change @cite byrne2007ingestible	</span>|<span class="success">	No change or decrease @cite griffin1993temperature	</span>|<span class="success">	Decreasing 	</span>|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|<span class="success">	No change @cite christie1987cardiac	</span>|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|	No Data	|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|	No Data	|

<center><br>

*Table 4d. The fourth intensity/rest period.*
</center>
|	Segment	|	Notes	|	Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Core Temperature (Celsius)	|	Peripheral Temperature (Celsius)	|	Heart Rate (beats/min)	|	Systolic Pressure (mmHg)	|	Diastolic Pressure (mmHg)	|	Mean Arterial Pressure (mmHg)	|	Stroke Volume (mL)	|	Cardiac Output (L/min)	|	Gut Blood Flow (L/min)	|	Systemic Vascular Resistance (mmHg-min/L)	|	Respiration Rate (breaths/min)	|	Tidal Volume (L)	|	Exchange Ratio (Or %Respiratory Quotient)	|	Oxygen Consumption (L/min)	|	Urine Production Rate (mL/min)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Exercise Severity 0.3583	|	430 W	|	1470	|	1829	|<span class="success">	Increase @cite christie1987cardiac, @cite byrne2007ingestible	</span>|<span class="success">	No change or decrease @cite griffin1993temperature	</span>|<span class="success">	257% of baseline @cite christie1987cardiac	</span>|<span class="warning">	56.8% increase @cite christie1987cardiac	</span>|<span class="warning">	No change @cite christie1987cardiac	</span>|<span class="warning">	117.9% increase  @cite christie1987cardiac	</span>|<span class="danger">	31.2% increase  @cite christie1987cardiac	</span>|<span class="danger">	350% of base  @cite christie1987cardiac	</span>|<span class="danger">	reduced 25-50%  @cite perko1998gutflow	</span>|<span class="success">	decrease  @cite perko1998gutflow	</span>|<span class="danger">	59 +/- 9  @cite guenette2007respiratory	</span>|<span class="danger">	3.1 +/- 0.4  @cite guenette2007respiratory	</span>|<span class="danger">	increase, maximum  ~1.21  @cite koutlianos2013indirect	</span>|<span class="danger">	3.36  @cite christie1987cardiac	</span>|<span class="success">	decrease  @cite melin2001comparison	</span>|
|	Exercise Severity 0	|	rest 2 minutes	|	1830	|	1950	|<span class="success">	Increase @cite christie1987cardiac or no change @cite byrne2007ingestible	</span>|<span class="success">	No change or decrease @cite griffin1993temperature	</span>|<span class="success">	Decreasing 	</span>|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|<span class="success">	No change @cite christie1987cardiac	</span>|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|	No Data	|<span class="success">	Toward baseline @cite christie1987cardiac	</span>|	No Data	|

<center><br>

*Table 5. The fatigue scenario validation results.*
</center>
|	Segment	|	Notes	|	Occurrence Time (s)	|	Duration of Full Exercise	|
|	---	|	---	|	---	|	---	|
|	Exercise Severity 0.3583	|	Exercise intensity of 0.3583 is a requested work rate of 430 watts.	|	N/A	|<span class="success">	~ 3 min @cite johnson2000exercise	</span>|

<center><br>

*Table 6.  The effect of moderately hard exercise on body fluid and ion loss (See Figure 4).*
</center>
|	Action	|	Notes	|	Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Fluid Mass Lost	|
|	---	|	---	|	---	|	---	|	---	|
|	Exercise Severity 0.234	|	Working at about 65% of VO2 max	|	30	|	3630	|<span class="warning">	0.7-1.2 kg @cite baker2009comparison	</span>|

<center>
<table border="0">
<tr>
    <td><img src="./plots/Energy/SustainedExercise_PatientWeight.jpg" width="550"></td>
    <td><img src="./plots/Energy/SustainedExercise_Sodium.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/Energy/SustainedExercise_Chloride.jpg" width="550"></td>
    <td><img src="./plots/Energy/SustainedExercise_Potassium.jpg" width="550"></td>
</tr>
</table>
</center>
<center>
<i>Figure 4.  After hard exercise for an hour, the patient loses about 0.5 kg of fluid (approximately 0.5 L).  The cumulative losses of sodium, chloride, and potassium are 650 mg, 930 mg, and 102 mg, respectively.  Note that the mass of each ion in the extracellular skin tissue compartment decreases as the body sweats.</i>
</center><br>

*Table 7.  The effect of sustained exercise (three hours) on blood glucose and insulin.*
</center>
|	Duration of Exercise (mins)	|	Glucose (mg/dL)	|	Insulin (ug/mL)	|
|	---	|	---	|	---	|
|	0	|	<span class="success">	91.7 @cite wilmore2007physiology	</span>	|	<span class="danger">	16.8 @cite wilmore2007physiology	</span>	|
|	15	|	<span class="danger">	95.1 @cite wilmore2007physiology	</span>	|	<span class="danger">	14.6 @cite wilmore2007physiology	</span>	|
|	30	|	<span class="danger">	96.4 @cite wilmore2007physiology	</span>	|	<span class="danger">	13.5 @cite wilmore2007physiology	</span>	|
|	60	|	<span class="danger">	96.7 @cite wilmore2007physiology	</span>	|	<span class="danger">	12.2 @cite wilmore2007physiology	</span>	|
|	90	|	<span class="danger">	95.8 @cite wilmore2007physiology	</span>	|	<span class="danger">	11.9 @cite wilmore2007physiology	</span>	|
|	120	|	<span class="danger">	94.6 @cite wilmore2007physiology	</span>	|	<span class="danger">	11.4 @cite wilmore2007physiology	</span>	|
|	150	|	<span class="danger">	93.5 @cite wilmore2007physiology	</span>	|	<span class="danger">	10.7 @cite wilmore2007physiology	</span>	|
|	180	|	<span class="danger">	91.8 @cite wilmore2007physiology	</span>	|	<span class="danger">	10.0 @cite wilmore2007physiology	</span>	|


*Table 8.  The results for changes in the thermal circuit during exercise in the heat.*
|	Action	|	Notes	|	Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Skin Temperature	|	Core Temperature	|
|	---	|	---	|	---	|	---	|	---	|	---	|
|	Exercise in 37 degC	|	Working at about 50% of VO2 max	|	2400	|	4260	|<span class="warning">	35.7 degC @cite tsuji2011effect	</span>|	38.7 degC @cite tsuji2011effect	</span>|



There are several physiological measures that are failing validation for the exercise action. However, most of the failing measures will be addressed by improvements to the chemoreceptor model planned as a part of the %BioGears nervous system. The oxygen consumption and urine production measures may still fail validation during exercise after the nervous system is fully implemented. The development team is currently making plans to address those failures.





### Cold Water Submersion

<center>
<img src="./plots/Energy/ColdWaterSubmersion_TotalMetabolicRate.jpg" width="1100">
<img src="./plots/Energy/ColdWaterSubmersion_CoreTemp.jpg" width="1100">
<img src="./plots/Energy/ColdWaterSubmersion_SkinTemp.jpg" width="1100">
<img src="./plots/Energy/ColdWaterSubmersionLegend.jpg" width="1100">
</center>
<center>
*Figure 5. The cold water submersion begins by submerging the patient in water that has a temperature of 10 degrees Celsius. A direct result of this is a drop in skin temperature due to an increased heat loss from the skin surface. The patient remains in the water for one hour, during which the metabolic rate increases via shivering. This rise in metabolic rate allows for the core temperature to stabilize just above 35 degrees Celsius. After one hour, the patient is removed and then active heating begins ten minutes later. The result of active heating is an increase in core and skin temperature, and a decline in the patient&rsquo;s metabolic rate.*
</center><br>
<center>
*Table 9. The cold water submersion scenario results.*
</center>
|	Action	|	Notes	|	Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Core Temperature (Celsius)	|	Peripheral Temperature (Celsius)	|	Metabolic Rate	|	Heart Rate (beats/min)	|	Systolic Pressure (mmHg)	|	Diastolic Pressure (mmHg)	|	Mean Arterial Pressure (mmHg)	|	Cardiac Output (L/min)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Environment File Change: Submerged in 5 C water	|	Patient is submerged in cold for 60 minutes	|	50	|	3650	|<span class="success">	Decrease to 35 degC @cite williams2005rewarming	</span>|<span class="success">	Decrease 	</span>|<span class="success">	Increase to between 300 and 350 W @cite williams2005rewarming	</span>|<span class="danger">	Decrease @cite garrioch2004body	</span>|<span class="success">	No change @cite christie1987cardiac	</span>|<span class="success">	If mild hypothermia (Core Temp 32-35 C) Increase @cite mallet2002hypothermia	</span>|<span class="success">	If mild hypothermia (Core Temp 32-35 C) Increase @cite mallet2002hypothermia	</span>|<span class="success">	If mild hypothermia (Core Temp 32-35 C) Increase If moderate or severe (Core Temp < 32 C) Decrease @cite mallet2002hypothermia	</span>|
|	Environment File Change to Standard Environment	|		|	3650	|	4250	|<span class="warning">	Decrease below 35 @cite williams2005rewarming	</span>|<span class="success">	Increase 	</span>|<span class="warning">	Increase above 350 W @cite williams2005rewarming	</span>|<span class="danger">	Decrease @cite talebipour2006sauna	</span>|<span class="danger">	Decrease @cite talebipour2006sauna	</span>|<span class="success">	Decrease @cite talebipour2006sauna	</span>|<span class="success">	Decrease @cite talebipour2006sauna	</span>|<span class="success">	Decrease @cite talebipour2006sauna	</span>|
|	Active Heating at 1000 btu/hr	|		|	480	|	5450	|<span class="success">	Increase to between 36 and 38 degC @cite williams2005rewarming	</span>|<span class="success">	Increase 	</span>|<span class="success">	Decreasing to below 80 W @cite williams2005rewarming	</span>|<span class="danger">	Decrease @cite talebipour2006sauna	</span>|<span class="danger">	Decrease@cite talebipour2006sauna	</span>|<span class="success">	Decrease @cite talebipour2006sauna	</span>|<span class="success">	Decrease @cite talebipour2006sauna	</span>|<span class="success">	Decrease @cite talebipour2006sauna	</span>|

At a scenario time of 50 seconds, the patient is submerged in water with a temperature of 10 degrees Celsius. The patient remains submerged for one hour, and then is removed from the water. The patient remains in the ambient environment for 10 minutes before active heating is applied. The power output of the heating device is 340 btu/hr. Active heating continues for 20 minutes, and then the scenario ends. The immediate response to cold water submersion is a drop in skin temperature. This occurs due to a dramatic increase in the skin convective heat transfer due to a shift from air convection to water convection. The core temperature declines due to increased heat transfer from core to skin. The drop in core temperature initiates a shivering response, leading to a rise in metabolic rate. Additional feedback comes in the form of vasoconstriction of the skin blood vessels. This helps to reduce the heat transferred from the core to skin. These two forms of feedback allow the core temperature to stabilize around 35 degrees Celsius. There is an increase in heart rate and respiration rate, initially due to a nervous response to the exposure to cold water. The increase in respiration rate and heart rate are maintained to accommodate the increased oxygen consumption and carbon dioxide production at the elevated metabolic rate. Following the return to the ambient environment, there is an immediate rise in skin temperature, with the core temperature following. The core temperature rise is not expected during this time period according to validation data. A continued decline in core temperature is expected. The observed increase in core temperature is due to a low heat capacitance used in the model and not accounting for residual water remaining on the skin surface after being removed from the water. The capacitance can be adjusted to a higher value, which would lead to the core temperature responding more slowly to changes in heat transfer over time. In addition, the %Environment heat transfer methodology could be updated to account for additional heat transfer due to water remaining on the skin surface after exiting submersion. At the onset of active heating, the skin temperature and core temperature continue to rise. This coincides with the expected validation outcome. 

### High Altitude Change
<table border="0">
<tr>
    <td><img src="./plots/Energy/HighAltitude_HR.jpg" width="550"></td>
    <td><img src="./plots/Energy/HighAltitude_O2PP.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/Energy/HighAltitude_O2Sat.jpg" width="550"></td>
    <td><img src="./plots/Energy/HighAltitude_RR.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/Energy/HighAltitudeLegend.jpg" width="1100"></td>
</tr>
</table>
<center>
*Figure 6. The patient is placed at a high altitude with an atmospheric pressure of 525 mmHg. The immediate result is a drop in arterial oxygen due to the decreased environment oxygen partial pressure. This leads to a drop in oxygen saturation and an increase in the patient heart rate. The patient stabilizes, acquiring a new resting physiologic state at the reduced pressure. The exaggerated increase in heart rate is due to a catecholamine release secondary to  hypoxia, a [known issue](@ref known-issues) in %BioGears*
</center><br>

</center><br>
<center>
*Table 10. The high altitude change scenario results.*
</center>
|	Action	|	Notes	|	Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Heart Rate (beats/min)	|	Oxygen Saturation (%)	|	Respiration Rate (breaths/min)	|	Systolic Blood Pressure	|	Cardiac Output (mL/min)	|	Myocardium Flow (mL/min)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	%Environment File Change to an altitude of 3000 meters	|	Atmospheric pressure at 3000m is approximately 70 kPa	|	50 s	|	950	|<span class="success">	7% increase @cite zhang2014highaltitude	</span>|<span class="success">	7% decrease @cite zhang2014highaltitude	|<span class="warning">	5% increase @cite zhang2014highaltitude	</span>|<span class="warning">	Decrease @cite bartsch2007highaltitude	</span>|<span class="warning">	No change @cite bartsch2007highaltitude	</span>|<span class="success">	Increase @cite bartsch2007highaltitude	</span>|


This scenario is used to simulate the effects of low oxygen due to high altitude. At 30 seconds, the surrounding environment is changed to an equivalent elevation of 4000 meters, resulting in a drop in atmospheric pressure from 760 mmHg to 525 mmHg. The patient remains in the reduced pressure environment for 15 minutes. The immediate response is a decrease in arterial oxygen due to the reduced partial pressure in the environment. This leads to a decrease in the blood oxygen saturation, which is in line with the expected validation data. The reduced arterial oxygen causes an increase in the heart rate. This increase is currently determined from epinephrine release due to a hypoxia event being triggered. Another expected result of the epinephrine release is a respiration rate increase. This effect is nullified from the reduction in atmospheric carbon dioxide, which allows for larger quantities of carbon dioxide to diffuse out of the alveoli. The increased diffusion leads to lower arterial carbon dioxide partial pressures, which is used to drive the respiratory frequency. This counter balances the effect of epinephrine in the system, thus causing a slight decrease in respiration rate.

### Starvation
The starvation condition is validated at 72 hours, which allows for direct comparison of blood concentration values in @cite zauner2000starvation, @cite garber1974hepatic, @cite owen1971human, and @cite elia1984mineral. Since the glucose and ketone concentrations for the starvation condition are directly set based on these sources, it is no surprise that they meet validation. More satisfaction is gained from the passing values for insulin, respiratory quotient, gluconeogenesis rate, ketogenesis rate, and urea, since those values are not directly set by the starvation condition, but are instead generated by other metabolic processes in %BioGears. These lend credibility to the metabolic models. The metrics that did not pass are body weight and plasma ion concentrations, both of which have outstanding tasks for future %BioGears improvements.

<center>
*Table 11. The starvation condition scenario results.*
</center>
|	Condition	|	Notes	|	Sampled Scenario Time (s)	|	Aorta Glucose Concentration (mg/dL)	|	Vena Cava Insulin Concentration (ug/L)	|	Aorta Ketones Concentration (mg/dL)	|	%Respiratory Quotient	|	Body Weight (kg)	|	%Hepatic Gluconeogenesis Rate (g/day)	|	Ketogenesis Rate (g/day)	|	Aorta Triacylglycerol Concentration (mg/dL)	|	Aorta Urea Concentration (mg/dL)	|	Sodium Plasma Concentration (mg/dL)	|	Potassium Plasma Concentration (mg/dL)	|	Calcium Plasma Concentration (mg/dL)	|	Chloride Plasma Concentration (mg/dL)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	72 hour starvation condition applied	|	For ketones, acetoacetate and beta hydroxybutyrate values were summed, when available.	|	60	|<span class="success">	64.86 @cite zauner2000resting 61.25 @cite garber1974hepatic 66.47 @cite owen1971human 61.43 @cite elia1984mineral	</span>|<span class="success">	.337 @cite zauner2000resting .129 @cite garber1974hepatic .637 @cite owen1971human	</span>|<span class="success">	43.58 (beta hydroxybutyrate only) @cite zauner2000resting 25.52 @cite garber1974hepatic 31.26 @cite owen1971human 43.49 @elia1984mineral	</span>|<span class="success">	.72 @cite zauner2000resting	</span>|<span class="danger">	2.5% Decrease @cite zauner2000resting	</span>|<span class="success">	[99,124] @cite garber1974hepatic 165 @cite rothman1991quantitation	</span>|<span class="success">	[83,115] @cite garber1974hepatic	</span>|<span class="success">	75.89 @cite zauner2000resting	</span>|<span class="success">	24.62 @cite elia1984mineral	</span>|<span class="success">	311.28 @cite elia1984mineral	</span>|<span class="danger">	15.48  @cite elia1984mineral	</span>|<span class="success">	9.7 @cite elia1984mineral	</span>|<span class="warning">	354.5 @cite elia1984mineral	</span>|


### Dehydration

Conclusions
-----------
The %BioGears %Energy system has provided a method for handling metabolic consumption and production and heat transfer in the human body. This system is fundamentally connected to all of the other systems, and provides dynamic feedback according to changes in the external environment. In its current state, the energy system accurately calculates resting thermal physiology and nutrient consumption and production. The energy system can be used by developers who wish to model thermal changes due to the environment or through strenuous activity. This may be a simple simulation, or connection with a mannequin for real-time interfacing.

@anchor energy-future
Future Work
===========

Coming Soon
-----------
- The maximum work rate will scale with physical conditioning, fed/starvation levels, and body composition.
- The nervous system will contain numerous chemoreceptor-related feedback mechanisms that will correct many of the exercise validation failures such as heart rate and blood pressure.
- The response of body core temperature to exercise will be improved.
- Sweat composition will include more than just water.

Recommended Improvements
------------------------
- Increased discretization of the temperature circuit: This would allow for increased fidelity of the temperature solution and prediction of varying levels of water submersion.
- Include psychological effects. For example, incorporate psychology into fatigue using a 'mettle' parameter.
- Switch exercise accounting from physical work rate to metabolic heat production.
- A true basal metabolic rate with differentiation from resting metabolic rate.

@anchor energy-appendices
Appendices
==========

Acronyms
--------
RQ - %Respiratory Quotient
BMR - Basal Metabolic Rate

Data Model Implementation
-------------------------
@ref EnergySystemTable "Energy"

@anchor energy-compartments
Compartments
------------
- InternalCore
- InternalSkin
- InternalGround
%Environment Methodology {#EnvironmentMethodology}
==========================

@anchor environment-overview
Overview
========

Abstract
--------
The %BioGears&reg; %Environment system contains and controls everything external to the patient and equipment (e.g. @ref AnesthesiaMachineMethodology). It sets the ambient boundary conditions for both fluid and thermal systems. The %Environment system provides a means to add airborne substance values, and calculates the heat transfer due to convection (natural and forced), radiation, and evaporation.

@anchor environment-intro
Introduction
------------
The %BioGears %Environment system defines the ambient conditions for the %BioGears body. The %Environment system controls everything from the phase of the environmental fluid (i.e. gas when is a room or liquid when submerged in water) to the volume fractions of constituent gases in the environmental fluid when that fluid is a gas. The %Environment system also computes heat exchange between the body and the environment as well as and substance delivery to the body-environment interfaces.

@anchor environment-system-design
System Design
=============
Background and Scope
--------------------
The many physiological functions within the human body depend heavily on the conditions of the environment within which the body is located. Thus, the %BioGears %Environment system is required to determine patient boundary conditions to enable a generic, extensible approach to modeling physiology within the body. Though the @ref CDM (CDM), the %Environment system interacts with and informs other %BioGears systems, particularly the [Respiratory](@ref RespiratoryMethodology), [Cardiovascular](@ref CardiovascularMethodology), and [Energy](@ref EnergyMethodology) systems, and the %Environment system was developed to include hooks for future integration into various applications such as games and sensor systems.

### Requirements
The %Environment system is used to meet the following requirements in %BioGears:
- Meteorological effects
- Surrounding atmospheric gas properties, including pulmonary agent exposure
- Altitude effects
- Heat transfer into and out of the skin and lungs
- Clothing resistances
- Ambient temperature changes
- Ambient humidity changes
- Multiple surrounding fluid types (i.e. submerged in water vs. air)
- Air velocity effects on convection
- Solar and surrounding material (i.e. walls) considerations
- Active heating and cooling

### Existing Research
One of the most comprehensive analysis of human interaction with typical thermal environments is contained within the American Society of Heating, Refrigerating, and Air-Conditioning Engineers (ASHRAE) Handbook - specifically the Thermal Comfort chapter @cite handbook2013fundamentals. The International Organization for Standardization (ISO) standard 7730:2005: Ergonomics of the Thermal %Environment contains additional data and analyses @cite ISOPDF. Both of these publications are focused on occupant thermal comfort for indoor environment planning and thus make assumptions that are not valid for some of the extreme environments required for %BioGears. Data and analyses from these publications were heavily leveraged during development of the %BioGears environment system. Other existing research includes the high-level heat transfer analysis focused on medical and biological applications reported in Ref. @cite shitzer1985heat, as well as experimental and computational explorations of thermoregulation @cite mccutchan1951respiratory @cite oliveira2008measuring @cite de1997convective @cite fiala1999computer.

### Approach
The %Environment system was designed with as few assumptions and parameter bounds as possible to allow simulations of both standard and extreme environments. The system is dynamic. By leveraging the@ref CDM and the generic [circuit](@ref CircuitMethodology) and [convective transport](@ref SubstanceTransportMethodology) solvers, the %BioGears body will react differently to varying initial environments and to changes in the environment during simulation.  %Environment parameters can be changed through specifying a separate environment file, or by manually updating the parameters through the environment state change action (for more details see the @ref SDK). Table 1 shows these inputs, which set the substance, fluid, and thermal properties external to the patient.

<br><center>
<i>Table 1. The parameters that can be set using the %Environment condition.  Note that air density is a calculated value inside the class.</i>
</center>

| Parameter | Use | Unit Type |
| :----     | :----   | :----   |
| Air Density | Thermal | MassPerVolume |
| Air Velocity | Thermal | LengthPerTime |  
| Ambient Temperature | Thermal | Temperature |
| Atmospheric Pressure | Fluid | Pressure |
| Clothing Resistance | Thermal | HeatResistanceArea |
| Emissivity | Thermal | Fraction |
| Mean Radiant Temperature | Thermal | Temperature |
| Relative Humidity | Thermal | Fraction |
| Respiration Ambient Temperature | Thermal | Temperature | 
| Ambient Substance| Substance | Fraction |

The Thermal Application action is handled by the %Environment system and has the following options (with unit type):

- Active Heating
	- Heating Power (Power)
	- Heated Surface Area Fraction (Fraction)
	- Heated Surface Area (Area)
- Active Cooling
	- Cooling Power (Power)
	- Cooled Surface Area Fraction (Fraction)
	- Cooled Surface Area (Area)
- Applied Temperature
	- Applied Temperature (Temperature)
	- Applied Surface Area Fraction (Fraction)
	- Applied Surface Area (Area)

### Substance %Environment

Ambient substance volume fractions can be modified through the %Environment system. The environmental fraction of typical atmospheric gases (e.g. O2, CO2, N2, etc.) and concentrations of airborne agents in suspension (i.e. aerosols) can be set and/or dynamically adjusted. The gases and aerosols enter the @ref RespiratoryMethodology "Respiratory System" with the generic convective @ref SubstanceTransportMethodology.

### Fluid %Environment

Many meteorologic parameters such as ambient temperature, wind speed, atmospheric pressure, and humidity are set and used by the %Environment system.  The atmospheric pressure is set directly on the %Respiratory and Anesthesia Machine reference (ambient) node values, which automatically propagates effects throughout.  The other thermal values are used to determine the %Environment circuit element values described in more detail below.

### Thermal %Environment

The remaining parameters are used for setting and manipulating the surrounding thermal environment.  A simplified one-dimensional representation of the thermodynamic process solved by the %Environment system is shown in Figure 1 @cite handbook2013fundamentals .  The %Environment circuit connects directly to the %Energy system circuit to balance with metabolic rate and body heat storage.

@image html Environment2DModel.png
<center>
<i>Figure 1. Human thermoregulation in an environment can be shown with a simplified one dimensional model.  The net heat production is transferred to the environment through the skin surface and respiratory tract.</i>
</center><br>


Data Flow
---------
### Initialize

During resting stabilization, the patient reaches a homeostatic value using the standard environment. Following that steady-state starting point, the user has the option to modify any environment values as a condition. The simulation will not start running until a new steady-state is achieved using these new parameters.

### Preprocess

Preprocess uses feedback to calculate thermal properties and circuit element values for the next engine state.  The Preprocess methods are described in more detail in the next section.

### Process

The generic circuit methodology developed for the %BioGears Engine is used to solve for the pressure, flow, and volume at each node or path. For more details, see the @ref CircuitMethodology.

### Post Process

The Post Process step moves everything calculated in Process from the next time step calculation to the current time step calculation. This allows all other systems access to the information when completing their Preprocess analysis for the next time step.

<center><img src="./images/Environment/EnvironmentFlow.png" width="900"></center>
<center>
<i>Figure 2. The data flow for the %Environment System consists of Preprocess, Process, and Post Process. Preprocess determines the circuit element values based on feedback mechanisms and engine actions. Preprocess also updates system-level quantities via the Calculate.  Process uses generic circuit methodology to solve the %Environment thermal circuit for temperatures and heat flows along the paths and on nodes. Post Process updates these quantities to the next time step and then begins advancing time, which causes the loop to repeat.</i>
</center><br>


@anchor environment-features
Features and Capabilities
-------------------------
### Definitions
Acronyms and a nomenclature table are available in the [appendix](@ref environment-appendices).

### Circuit
The one-dimensional model shown in Figure 1 is further simplified to a zero-dimensional model (Figure 3) by averaging the thermal effects over the entire patient.  The %Environment system uses the generic circuit solver in the same way as fluid systems (@ref CircuitMethodology).

The Ambient node contains both thermal properties (temperature), fluid properties (pressure), and substance properties (volume fraction, aerosol concentration).  It is assigned as the %Respiratory and Anesthesia Machine circuit&rsquo;s reference node, and therefore, interacts with them directly.  Any changes to the Ambient node properties automatically propagates through the other systems.

@image html EnvironmentCircuit.png
<center>
<i>Figure 3. The %Environment circuit consists of 6 nodes that are connected via 8 paths. There are 8 parameters representing circuit elements that are modified by feedback mechanisms or actions. The circuit is used to estimate all environment thermal properties each time step.</i>
</center><br>

The clothing thermal insulation/resistance in this model is an average lumped value over the entire patient's body.  Because clothing insulation cannot be measured for most routine engineering application, tables of measure values for various ensembles can be used as in Table 2 @cite handbook2013fundamentals.  When a premeasured ensemble cannot be found to match, estimated ensemble insulation can be determined via a summation of the individual garmets as in Table 3 @cite handbook2013fundamentals.  An important note is that the clothing resistance remains constant once set, and does not automatically take into account sweat saturation or changes due to submersion.  These effects could be accounted for by external determination of a new thermal resistance value and set using the environment change action.

<br><center>
<i>Table 2. Typical insulation for clothing ensembles.</i>
</center>
| Ensemble Description | Icl (clo) |
| :----     | :----   |
| Walking shorts, short-sleeved shirt | 0.36 |
| Trousers, short-sleeved shirt | 0.57 |
| Trousers, long-sleeved shirt | 0.61 |
| Same as above, plus suit jacket | 0.96 |
| Same as above, plus vest and T-shirt | 0.96 |
| Trousers, long-sleeved shirt, long-sleeved sweater, T-shirt | 1.01 |
| Same as above, plus suit jacket and long underwear bottoms | 1.30 |
| Sweat pants, sweat shirt | 0.74 |
| Long-sleeved pajama top, long pajama trousers, short 3/4 sleeved robe, slippers (no socks) | 0.96 |
| Knee-length skirt, short-sleeved shirt, panty hose, sandals | 0.54 |
| Knee-length skirt, long-sleeved shirt, full slip, panty hose | 0.67 |
| Knee-length skirt, long-sleeved shirt, half slip, panty hose, long-sleeved sweater | 1.10 |
| Knee-length skirt, long-sleeved shirt, half slip, panty hose, suit jacket | 1.04 |
| Ankle-length skirt, long-sleeved shirt, suit jacket, panty hose | 1.10 |
| Long-sleeved coveralls, T-shirt | 0.72 |
| Overalls, long-sleeved shirt, T-shirt | 0.89 |
| Insulated coveralls, long-sleeved thermal underwear, long underwear bottoms | 1.37 |

<br><center>
<i>Table 3. Typical garment insulation.</i>
</center>
| Garment Description | Icl (clo) | Garment Description	 | Icl (clo) |
| :----     | :----   | :----   | :----   |
| **Underwear** |  | **Dress and Skirts** |  |
| Bra | 0.01 | Skirt (thin) | 0.14 |
| Panties | 0.03 | Skirt (thick) | 0.23 |
| Men's briefs | 0.04 | Sleeveless, scoop neck (thin) | 0.23 |
| T-shirt | 0.08 | Sleeveless, scoop neck (thick), i.e., jumper | 0.27 |
| Half-slip | 0.14 | Short-sleeve shirtdress (thin) | 0.29 |
| Long underwear bottoms | 0.15 | Long-sleeve shirtdress (thin) | 0.33 |
| Full slip | 0.16 | Long-sleeve shirtdress (thick) | 0.47 |
| Long underwear top | 0.20 |  |  |
| **Footwear** |  | **Sweaters** |  |
| Ankle-length athletic socks | 0.02 | Sleeveless vest (thin) | 0.13 |
| Pantyhose/stockings | 0.02 | Sleeveless vest (thick) | 0.22 |
| Sandals/thongs | 0.02 | Long-sleeve (thin) | 0.25 |
| Shoes | 0.02 | Long-sleeve (thick) | 0.36 |
| Slippers (quilted, pile lined) | 0.03 |  |  |
| Calf-length socks | 0.03 | **Suit Jackets and Vests (lined)** |  |
| Knee socks (thick) | 0.06 | Sleeveless vest (thin) | 0.10 |
| Boots | 0.10 | Sleeveless vest (thick) | 0.17 |
| **Shirts and Blouses** |  | Single-breasted (thin) | 0.36 |
| Sleeveless/scoop-neck blouse | 0.12 | Single-breasted (thick) | 0.44 |
| Short-sleeve knit sport shirt | 0.17 | Double-breasted (thin) | 0.42 |
| Short-sleeve dress shirt | 0.19 | Double-breasted (thick) | 0.48 |
| Long-sleeve dress shirt | 0.25 |  |  |
| Long-sleeve flannel shirt | 0.34 | **Sleepwear and Robes** |  |
| Long-sleeve sweatshirt | 0.34 | Sleeveless short gown (thin) | 0.18 |
| **Trousers and Coveralls** |  | Sleeveless long gown (thin) | 0.20 |
| Short shorts | 0.06 | Short-sleeve hospital gown | 0.31 |
| Walking shorts | 0.08 | Short-sleeve short robe (thin) | 0.34 |
| Straight trousers (thin) | 0.15 | Short-sleeve pajamas (thin) | 0.42 |
| Straight trousers (thick) | 0.24 | Long-sleeve long gown (thick) | 0.46 |
| Sweatpants | 0.28 | Long-sleeve short wrap robe (thick) | 0.48 |
| Overalls | 0.30 | Long-sleeve pajamas (thick) | 0.57 |
| Coveralls | 0.49 | Long-sleeve long wrap robe (thick) | 0.69 |

@anchor environment-aerosol
### Aerosol
An aerosol is a colloidal suspension of either solid or liquid droplet particles suspended in a gas (typically the surrounding air). Real aerosols can be monodispersed but are more often polydispersed, having a range of particle diameters. Particle deposition and retention in the lung depends on a number of, factors including particle aerodynamic diameter @cite carvalho2011influence @cite newman1985aerosol. Other deposition factors, including inhalation flow rate and tidal volume @cite kolanjiyil2016computationally, are accounted for by the breathing patterns simulated by the %BioGears engine. In order to avoid the computational cost of tracking millions of individual particles or many groups of particles in a distribution throughout the respiratory tract and into the body, %BioGears aerosols are modeled as a single concentration in the environment. The location of deposition and the deposition efficiency at each location is determined by first computing a particle-size-independent deposition efficiency for each compartment from a defined size-distribution histogram. The size-independent deposition efficiencies are computed from the deposition fractions which are determined for each compartment from the particle diameter using the equations fit by Hinds to the ICRP66 simulation results @cite hinds1999aerosol, as described by @cite rostami2009computational. The deposition fraction equations from @cite rostami2009computational are plotted in Fig. 4, and example mean depositions fractions for a particular histogram are shown. For the %BioGears implementation it is assumed that all particles are spherical, thus diameter is the aerodynamic diameter.

@image html depositionfractions.png
<center>
<i>Figure 4. The deposition fraction in each compartment is computed as a function of particle size, and the deposition fraction for a collection of particles within a histogram bin is computed from the mean deposition fraction within the boundaries of the bin. An 8 bin partition is shown in the figure for illustration purposes. %BioGears supports histograms with an arbitrary number of bins.</i>
</center><br>

The size-independent deposition efficiencies are computed from the deposition fractions using Equation 31.
@anchor SIDE
<center>
\f[\mathrm{K_B=}\frac{\left( b_1f_1 + \cdots + b_nf_n \right)}{\left( \left( 1-a_1 \right)f_1 + \cdots + \left(1-a_n \right)f_n \right)} \f]
*Equation 1.*
</center><br>

Here *K<sub>B</sub>* is the size independent deposition efficiency coefficient for compartment *B*, *b<sub>i</sub>* is the fraction of total particles of size *i* entering the body that deposit in compartment *B*, and *f<sub>i</sub>* is the fraction of total particles in the environment that are of size *i*. Likewise, *a<sub>i</sub>* is the fraction of total particles of size *i* entering the body that deposit in compartment *A*. This equation is derived by first considering a polydispersed aerosol with particles divided by size into *n* number of bins. The total mass of particles in an aliquot of to-be-inhaled air prior to inhalation is equal to the sum of the masses of the particles in each bin, as described by Equation 2.
<center>
\f[m_t = \Sigma_{i=1}^{n}m_i \f]
*Equation 2.*
</center><br>

The mass of the particles in each bin must be defined by a fraction specified in a histogram. The size of the histogram is arbitrary. The only requirement is that a histogram is supplied rather than a density function. Equation 3 describes the mass of particle *i* within the aliquot of aerosol.
<center>
\f[m_i = f_im_t \f]
*Equation 3.*
</center><br>

Likewise, the total mass of particles in compartment *A* is equal to the sum of the masses of the particles in each bin, and the same is true for compartments *B*, *C*, *D*, etc., described by Equation 4.
<center>
\f[m_{tA} = \Sigma_{i=1}^{n}m_{iA} \f]
*Equation 4.*
</center><br>

The mass of particles in compartment *B* is also equal to the mass of the particles that did not deposit in the adjacent upstream compartment *A*.
<center>
\f[m_{iB} = \left(1-a_i \right)m_i \f]
*Equation 5.*
</center><br>

Combining Equations 2-5 yields Equation 6.
<center>
\f[m_{iB} = \left( \left(1-a_1 \right)f_1 + \cdots + \left(1-a_n \right)f_n \right)m_t \f]
*Equation 6.*
</center><br>

The total mass of particles that deposit in compartment *B* is some fraction of the total particles in compartment *B*, which is the sum of the amount deposited from each bin in compartment *B*.
<center>
\f[K_{B}m_{tB} = \left(b_1f_1 + \cdots + b_nf_n \right)m_t \f]
*Equation 7.*
</center><br>

Rearranging Equation 7 and combining with Equation 6 yields [Equation 1](@ref SIDE), the size-independent deposition efficiency coefficient for compartment *B*.

The mass deposited in each compartment is computed by multiplying the size-independent deposition efficiency coefficient by the mass in the compartment at each time step. Once an aerosol substance deposits in a respiratory compartment it stays in the respiratory compartment. In other words, coughing is not productive in %BioGears. The direct effect of a deposited substance depends on the mass deposited and the value of the inflammation coefficient that is defined in the substance file. The inflammation coefficient defines the amount of damage, by any mode, that the substance does to the respiratory tissue. Figure 5 visually describes the direct effects of an aerosol on the %BioGears %Respiratory system.

<img src="./images/Environment/AerosolEffects.png" width="700">
<center>
<i>Figure 5. Diagram describing aerosol transport into the respiratory compartments and the associated effects.</i>
</center><br>

In addition to the direct effects, deposited aerosol substances can also have indirect effects. After liquid aerosol substances deposit they can diffuse into the surrounding tissue and then into the blood stream. If the liquid aerosol has pharmacodynamic properties defined, it's effects will be realized when it enters the blood and the plasma concentration becomes non-zero.

Currently, Albuterol and smoke particulate have histogram data and are modeled as polydisperse aerosols in the environment with the former being delivered via an inhaler and the latter being a concentration in the ambient environment. Smoke in %BioGears is modeled as a solid aerosol.

The solid smoke particulate in %BioGears is a model of the particulate generated through the combustion of organic material in a forest fire. We assume that the fire is well evolved in time to mitigate the bi-modal behavior seen during the initial ignition and smoldering @cite zhang2012chemical. This allows us to model the size distribution as a lognormal curve. The forest fire smoke particle size distribution histogram is given in table 4.

<br><center>
<i>Table 4. Histogram for wood smoke particulate in the environment. </i>
</center>
| Amount (fraction) | Diameter (microns) |
| ----------------- |------------------- |
| 0.015 | 1.0e-4 - 1.0e-3 | 
| 0.035 | 1.0e-3 - 1.0e-2 |
| 0.9   | 1.0e-2 - 1.0e-1 |
| 0.035 | 1.0e-1-1.0      |
| 0.015 | 1.0-1.0e1       |
| 0     | 1.0e1 - 1.0 e2  |

Figure 6 summarizes the fate and effects of liquid and smoke aerosols in %BioGears.
<img src="./images/Environment/AerosolFate.png" width="700">
<center>
<i>Figure 6. Diagram summarizing the transport into the respiratory compartments, effects, and the fate of liquid and solid particle aerosols.</i>
</center><br>

@anchor environment-carbon-monoxide
### Carbon Monoxide
Carbon monoxide (CO) is a toxic gas that is odorless, colorless, and tasteless. Carbon monoxide binds to the ferrous-heme site of the hemoglobin molecule in the same way that oxygen does (it also cannot bind to the oxidized methemoglobin sites), thus the maximal oxygen capacity falls to the extent that CO is bound. The subunits of the hemoglobin molecule are collectively in a tense state (T) or a relaxed state (R). All four subunits are always in the same state. The hemoglobin molecule is in the tense state when less oxygen is bound and it has a lower affinity for oxygen in the T state. When enough oxygen is bound, all subunits snap into the R state whether or not they are bound to oxygen, and the hemoglobin molecule has ~150 times the affinity for oxygen. It is this snapping that creates the steep portion of the binding curve and allows for the reversible carriage of oxygen by the blood. Like oxygen, as carbon monoxide binds to hemoglobin it can push the molecule from the relaxed to the tensed state, thus increasing the affinity for oxygen and reducing the amount of oxygen released from the hemoglobin molecule when it enters the oxygen depleted systemic capillaries. Additionally, carbon monoxide binds competitively to hemoglobin. For every molecule of CO bound to a heme site, there is one less site for oxygen to bind. The net result of these two phenomena, but predominately due to the former, is that there is less oxygen delivery to the tissues when CO is present in the blood.

There are three components to the %BioGears carbon monoxide model: transfer across the alveolar membrane, binding to hemoglobin, and the effects of binding on the oxygen binding curve. Obviously these three components are interrelated. As CO binds to hemoglobin in the pulmonary capillaries, less CO is dissolved in fluid and therefore the partial pressure gradient between the alveoli and the pulmonary capillary blood is higher, driving an increase in Fick's law diffusion resulting in more carbon monoxide in the blood. As more CO enters the blood, more binds to hemoglobin and the oxygen-hemoglobin saturation curve shifts to the left. The alveolar transfer of carbon monoxide is simulated in %BioGears by assuming that the membrane diffusivity of CO is similar to oxygen and then by leveraging the alveolar exchange model described in the @ref TissueMethodology report. Models for the binding of CO to hemoglobin and the effects on the oxygen saturation curve are discussed separately below.

#### Carbon Monoxide Binding
The distribution of carbon monoxide species (dissolved and bound to hemoglobin) in each vascular compartment is computed from the conservation of mass and the Haldane relationship described in @cite bruce2003multicompartment. Canservation of mass is described by Equation 8.
<center>
\f[T_{CO} = CO_{Hb} + k_{CO}P_{CO} \f]
*Equation 8.*
</center><br>
where *T<sub>CO</sub>* is the total carbon monoxide in the blood, *CO<sub>Hb</sub>* is the carbon monoxide bound to hemoglobin, *P<sub>CO</sub>* is the partial pressure of carbon monoxide in the blood, and *k<sub>CO</sub>* is the Henry's Law solubility coefficient for carbon monoxide. Equation 9, the Haldane relationship, provides the second equation necessary to solve for the two unknown values. In Equation 9, *M<sub>H</sub>* is the Haldane affinity ratio for hemoglobin, *P<sub>gas</sub>* is the partial pressure of the gas, and *Gas<sub>Hb</sub>* is the amount of gas bound to hemoglobin.
<center>
\f[M_H \frac{P_{CO}}{CO_{Hb}} = \frac{P_{O_2}}{O_{2_{Hb}}} \f]
*Equation 9.*
</center><br>
The model is implemented by first totaling the carbon monoxide in a compartment post diffusion, then computing the partial pressure of carbon monoxide, and finally computing the carboxyhemoglobin in the compartment. After the target distribution is calculated, the hemoglobin species (unbound, oxyhemoglobin, carboxyhemoglobin, carbaminohemoglobin, and oxycarbaminohemoglobin) amounts are updated by assuming that unbound hemoglobin is consumed first followed by oxyhemoglobin then oxycarbaminohemoglobin and finally carbaminohemoglobin. If all hemoglobin is converted to carboxyhemoglobin before the necessary adjustment is made to conserve CO mass, then the remaining CO is distributed back to the dissolved species. This acts as an automatic negative feedback device to ensure that the perfusion contribution to the diffusing capacity is saturated when the hemoglobin becomes saturated. See the [C++ code](@ref CalculateCarbonMonoxideSpeciesDistribution) for more details.

#### Carbon Monoxide Effects on Oxygen Saturation Curve
The oxygen saturation curve effects model implemented in %BioGears is adapted from the regression model described by @ref bruce2003multicompartment. We have simplified the model by assuming a linear relationship between carboxyhemoglobin and both the Hill coefficient and the 50% saturation shaping parameter. We then applied the linear relationship to the Dash and Bassingthwaithe oxygen saturation model already implemented in %BioGears (see the [Tissue Methodology](@ref tissue-approach)). Equations 10 and 11 describe the adjustments made to the oxygen saturation curve in the presence of carboxyhemoglobin, and Figure 7 shows the shift in the curve at various carboxyhemoglobin concentrations. In the equations, *&eta;* is the Hill coefficient, *S<sub>CO</sub>* is the fraction of carboxyhemoglobin to total hemoglobin (i.e. CO saturation), and *P<sub>50</sub>* is the standard partial pressure of oxygen at 50% oxygen saturation.
<center>
\f[\eta = 1.7 - 1.1 \cdot S_{CO} \f]
*Equation 10.*
</center><br>
<center>
\f[P_{50} = 26.8 - 20 \cdot S_{CO} \f]
*Equation 11.*
</center><br>

@image html leftShiftCO.png
<center>
<i>Figure 7. Oxygen concentration at partial pressure of oxygen in the blood with carboxyhemoglobin present. The dashed lines show what the curves would look like without the characteristic leftward shift.</i>
</center><br>

@anchor environment-dependencies
### Dependencies

The %Environment circuit is directly connected through the skin and core to the %Energy system for full body thermoregulation.  This boundary condition feedback in and out is automatically handled through the node and path interactions. It also shares an ambient node with the %Respiratory and Anesthesia Machine systems.

The %Environment system relies on inputs from several other systems.  The sweat rate from the %Energy system is needed to determine the amount of heat lost through convection and evaporation.  The respiration rate from the respiratory system is needed to determine the amount of heat lost via the respiratory track.

### Supplemental Values

Numerous patient and thermal properties are determined and set in the supplemental values method for later calculations.

Since the one dimensional model used for the %Environment thermal circuit needs to be averaged over the patient&rsquo;s skin surface area, an approximate value must be determined.  The most widely used equation to determine body surface area without direct measurement is the Du Bois formula presented in Equation 12 @cite handbook2013fundamentals , where mass is in kg and height is in m.

\f[{A_d}{\rm{ = 0}}.{\rm{202}}{m^{{\rm{0}}.{\rm{425}}}}{l^{{\rm{0}}.{\rm{725}}}}\f]
<center>
*Equation 12.*
</center><br>

Several properties of water are needed at varying temperatures to determine the heat transfer occurring between the skin/clothing/lungs and the air.  Four of these are determined using the best fit of property tables or experimental data, as shown in Figure 8.

@image html EnvironmentWaterPropertyGraphs.png
<center>
<i>Figure 8. These four properties of water are intermediate values needed for further calculations.  A, B, and C are plotted from data presented in a table @cite chase1985journal , while D is reproduced from data in a plot given by experimental data(d) @cite Cordes2014Heat .  Note in B that 1 Centipoise is equal to 0.001 N-s/m<sup>2</sup>.</i>
</center><br>

The first three properties (Figure 8 A-C) are used later for determining the heat transfer when the patient is submerged in water.  Finding a best fit for these plots yields:

\f[{\mathrm{c}}_{\mathrm{p,w}}=-1e^{-7}{\mathrm{T}}^{\mathrm{3}}_{\mathrm{\infty }}+3e^{-5}{\mathrm{T}}^{\mathrm{2}}_{\mathrm{\infty }}-0.0018{\mathrm{T}}_{\mathrm{\infty }}+4.2093\f]
<center>
*Equation 13.*
</center><br>

\f[\mu \mathrm{\ }=-3e^{-6}{\mathrm{T}}^{\mathrm{3}}_{\mathrm{\infty }}+0.0006{\mathrm{T}}^{\mathrm{2}}_{\mathrm{\infty }}-0.0462{\mathrm{T}}_{\mathrm{\infty }}+1.7412\f]
<center>
*Equation 14.*
</center><br>

\f[\mathrm{\alpha }\mathrm{\ }=6e^{-7}{\mathrm{T}}^{\mathrm{3}}_{\mathrm{\infty }}+0.0001{\mathrm{T}}^{\mathrm{2}}_{\mathrm{\infty }}+0.016{\mathrm{T}}_{\mathrm{\infty }}-0.0632\f]
<center>
*Equation 15.*
</center><br>

The heat of vaporization (Figure 8 D) is the enthalpy change required to transform a given quantity of a substance from a liquid into a gas and is needed for determining the heat lost through respiration.  Finding a piecewise best fit yields:

\f[{\mathrm{h}}_{\mathrm{fg}}\mathrm{=}\left\{ \begin{array}{c}
-0.10042t^2_a+22.173t_a+46375,~t_a\mathrm{\le }\mathrm{560K} \\ 
-1.5625t^2_a+1662.5t_a-\mathrm{414000,}~t_a\mathrm{>560K} \end{array}
\right.\f]
<center>
*Equation 16.*
</center><br>

The water vapor pressure at the skin and in ambient air is determined using the piecewise Antoine Equation given by Equation 17 @cite antoine1888tensions .  The resulting pressure is in units of mmHg and is a function of temperature in degrees Celsius. The constants are given by Table 5.

\f[p\mathrm{=}{\mathrm{10}}^{\left(A\mathrm{-}\frac{B}{C\mathrm{+}t}\right)}\f]
<center>
*Equation 17.*
</center><br>

<br><center>
<i>Table 5. Antione equation constants are used to calculate the water vapor pressure.</i>
</center>
| A       | B       | C       | t min (degrees C) | t max (degrees C) |
| :----   | :----   | :----   | :----             | :----             |
| 8.07131 | 1730.63 | 233.426 | 1                 | 100               |
| 8.14019 | 1810.94 | 44.485  | 99                | 374               |

The density of air is then determined using Equations 18-20 @cite Shelquist2015introduction .

\f[p_d\mathrm{=}p_t\mathrm{-}p_v\f]
<center>
*Equation 18.*
</center><br>

\f[p_v\mathrm{=}\mathrm{\phi }p_a\f]
<center>
*Equation 19.*
</center><br>

\f[\rho \mathrm{=}\frac{p_dM_d\mathrm{+}p_vM_v}{Ut_a}\f]
<center>
*Equation 20.*
</center><br>

The Lewis Relationship is the ratio of mass transfer coefficient to heat transfer coefficient, which is needed later in the determination of the heat loss due to evaporation.  For the air-water system with low concentration of water vapor, it is assumed the adiabatic saturation temperature and wet-bulb temperature are equal @cite Hewitt2015LEWIS , and is therefore given by:

\f[\beta \mathrm{=}\frac{\mathrm{1}}{\rho c_p}\f]
<center>
*Equation 21.*
</center><br>

### Radiation

The radiation heat transfer coefficient is used to determine the radiation resistance element value in the %Environment thermal circuit.  Equation 22 is implemented to determine this coefficient each time step.  The mean radiant temperature is a known value set by the user, and the clothing temperature is solved directly by the circuit and used as a feedback mechanism.  The ratio of the effective area to the surface area (Ar/Ad) is set constant to 0.73, which is typical for a standing person.  In contrast, 0.70 is typical for a sitting person @cite handbook2013fundamentals .  

\f[{\mathrm{h}}_r\mathrm{=4}\varepsilon \sigma \frac{A_r}{A_D}{\left(\frac{t_{cl}\mathrm{+}t_{mr}}{\mathrm{2}}\right)}^{\mathrm{3}}\f]
<center>
*Equation 22.*
</center><br>

Figure 9 highlights which elements are modified by the radiation method and by which parameters.

@image html EnvironmentRadiationCircuit.png
<center>
<i>Figure 9. Elements that are directly affected by the radiation calculations are boxed and labeled with parameters that cause them to change.</i>
</center><br>

### Convection

The equation chosen to account for convective effects comes from a study that used experimental data from a segmented manikin in a climate chamber @cite de1997convective .  A general-purpose forced convection equation was produced that is suitable for application to both seated and standing postures.  The air velocity is a known value set by the user, taking into account the combination of wind and patient movement.  The convective heat transfer coefficient is given by:

\f[{\mathrm{h}}_c\mathrm{=10.3}v^{\mathrm{0.6}}\f]
<center>
*Equation 23.*
</center><br>

Figure 6 highlights which elements are modified by the convection method and by which parameters.

@image html EnvironmentConvectionCircuit.png
<center>
<i>Figure 10. Elements that are directly affected by the convection calculations are boxed and labeled with parameters that cause them to change.</i>
</center><br>

### Evaporation

Evaporative heat loss is directly solved and implemented in the circuit as a heat sink on the skin.  It depends on the amount of moisture on the skin and the difference between the water vapor pressure at the skin and the ambient environment.  The evaporative heat transfer coefficient for the outer air layer of a nude or clothed person can be estimated from the convective heat transfer coefficient using the Lewis Relationship (Equation 24) @cite handbook2013fundamentals.  

\f[{\mathrm{h}}_e\mathrm{=}\beta {\mathrm{h}}_c\f]
<center>
*Equation 24.*
</center><br>

Clothing can then be taken into account to determine the maximum evaporative potential @cite kenny1998encyclopaedia :

\f[F_{pcl}\mathrm{=}\frac{\mathrm{1}}{\mathrm{1+2.}\mathrm{22}{\mathrm{h}}_cR_{cl}}\f]
<center>
*Equation 25.*
</center><br>

\f[E_{max}\mathrm{=}{\mathrm{h}}_eF_{pcl}\left(p_{sk,s}\mathrm{-}p_a\right)\f]
<center>
*Equation 26.*
</center><br>

Skin wettedness is an important factor to determine evaporative heat loss @cite handbook2013fundamentals :

\f[w_{rsw}\mathrm{=}\frac{E_{rsw}}{E_{max}}\f]
<center>
*Equation 27.*
</center><br>

\f[E_{dif}\mathrm{=}\left(\mathrm{1-}w_{rsw}\right)\mathrm{0.06}E_{max}\f]
<center>
*Equation 28.*
</center><br>

Evaporative heat loss by sweating is directly proportional to the rate of sweat generation through Equation 29 @cite handbook2013fundamentals .  The sweat rate is set by the %Energy system.

\f[E_{rsw}\mathrm{=}{\dot{m}}_{rsw}{\mathrm{h}}_{fg}\f]
<center>
*Equation 29.*
</center><br>

\f[E_{sk}\mathrm{=}E_{rsw}\mathrm{+}E_{dif}\f]
<center>
*Equation 30.*
</center><br>

Figure 11 highlights which elements are modified by the evaporation method and by which parameters.

@image html EnvironmentEvaporationCircuit.png
<center>
<i>Figure 11. The evaporation flow source is directly calculated by the evaporation calculations, which is boxed in the circuit diagram.</i>
</center><br>

### Respiration

%Respiratory heat exchange is taken into account with varying temperatures and humidity of inspired air.

Note that this calculation will be the same for both surrounding fluids&mdash;air and water.  It is assumed that the patient&rsquo;s head remains above water when breathing.  If the patient holds their breath to fully submerge, the respiration rate should be zero, which will correctly eliminate any heat transfer through the lungs.

The unitless specific humidity (mass/mass) is needed to determine the inspiratory-expiratory humidity difference.  Equation 31 solves for this knowing the ambient temperature (in Kelvin), relative humidity, and local atmospheric pressure @cite david2011humidity&mdash;all user defined values.

\f[\mathrm{H=}\frac{\mathrm{100}\mathrm{\phi }}{\mathrm{0.263}p_t}e^{\left(\frac{\mathrm{17.67}\left(t_a\mathrm{-}\mathrm{273.16}\right)}{t_a\mathrm{-}\mathrm{29.65}}\right)}\f]
<center>
*Equation 31.*
</center><br>

The humidity difference is then given by (ta in degrees F) @cite mccutchan1951respiratory :

\f[\mathrm{\Delta }W\mathrm{=0.02645+0.0000361}t_a\mathrm{-}\mathrm{0.798}H\f]
<center>
*Equation 32.*
</center><br>

Since heat and water vapor are transferred in the lungs, expired air is saturated with water vapor at the temperature of the respiratory tract @cite shitzer1985heat .  The respiration rate is given by the %Respiratory system and used to determine the following heat loss:

\f[q_{res,lat}\mathrm{=}{\mathrm{h}}_{fg}{\dot{m}}_{res}\mathrm{\Delta }W\f]
<center>
*Equation 33.*
</center><br>

\f[q_{res,sen}\mathrm{=}{\dot{m}}_{res}\rho c_p\left(t_{res}\mathrm{-}t_{a,res}\right)\f]
<center>
*Equation 34.*
</center><br>

\f[q_{res}\mathrm{=}q_{res,lat}\mathrm{+}q_{res,sen}\f]
<center>
*Equation 35.*
</center><br>

Figure 12 highlights the source element that is modified by the respiration method.

@image html EnvironmentRespirationCircuit.png
<center>
<i>Figure 12. The respiratory flow source is directly calculated by the respiratory calculations, which is boxed in the circuit diagram.</i>
</center><br>

### Submerged

Submerging the patient is done by using the environment change action or condition to set the surrounding type enumeration to water.  The ambient temperature will be the water temperature.  The clothing resistance will be the value when submerged, likely fully saturated or modified to a wet suit value. The currently implemented model assumes fresh water only, with a water pressure of 0.1 MPa (sea level).  It also assumes that radiation and evaporation effects are negligible.

The thermal conductivity of the water can be solved using @cite ramires1995standard :

\f[T\mathrm{=}\frac{T_{\mathrm{\infty }}}{\mathrm{298.15}}\f]
<center>
*Equation 36.*
</center><br>

\f[k\mathrm{=0.6065}\left(\mathrm{-}\mathrm{1.48445}~\mathrm{+4.12292}~T\mathrm{+-1.63866}~T^{\mathrm{2}}\right)\f]
<center>
*Equation 37.*
</center><br>

\f[\nu \mathrm{=}\frac{\mu }{\rho }\f]
<center>
*Equation 38.*
</center><br>

Equation 39 solves for the Prandtl number, which is the dimensionless ratio of momentum diffusivity (kinematic viscosity) to thermal diffusivity.

\f[Pr\mathrm{=}\frac{c_{p,w}\mu }{k}\f]
<center>
*Equation 39.*
</center><br>

Equation 40 solves for the Grashof number, which is the dimensionless value that approximates the ratio of the buoyancy to viscous force acting on a fluid.

\f[Gr\mathrm{=}\frac{g\mathrm{\alpha }\left(T_s\mathrm{-}T_{\mathrm{\infty }}\right)D^{\mathrm{3}}}{{\mathrm{\nu }}^{\mathrm{2}}}\f]
<center>
*Equation 40.*
</center><br>

The heat transfer coefficient can is determined by @cite boutelier2003experimental :

\f[{\mathrm{h}}_c\mathrm{=0.09}\left(Gr\mathrm{-}Pr\right)\mathrm{0.275}\f]
<center>
*Equation 41.*
</center><br>

Figure 13 highlights which elements are modified when submerged and by which parameters.  Radiation is handled much like an open switch in a circuit via the resistor and does not allow any transfer.  The evaporation transfer source is directly set to zero.

@image html EnvironmentSubmergedCircuit.png
<center>
<i>Figure 13. Elements that are directly affected by the submerged calculations are boxed and labeled with the parameters that cause them to change.</i>
</center><br>

### Outputs

The temperature at each node and heat flow on each path in the %Environment thermal circuit is calculated every time step.  The following system data is also set: 
- Convective Heat Loss (Power)
- Convective Heat Transfer Coefficient (Heat Conductance per Area)
- Evaporative Heat Loss (Power)
- Evaporative Heat Transfer Coefficient (Heat Conductance per Area)
- Radiative Heat Loss (Power)
- Radiative Heat Transfer Coefficient (Heat Conductance per Area)
- Respiration Heat Loss (Power)
- Skin Heat Loss (Power)

@anchor environment-assumptions
Assumptions and Limitations
---------------------------

The Ambient node operates under some pretenses for its interaction with other systems.  Since the atmospheric pressure is used to set the pulmonary reference nodes, everything in those systems is referenced to zero pressure.  This is in contrast to reporting the difference from atmospheric pressure, as is often the standard.  It is further assumed that all ambient substance volume fractions set the sum to 1.  The included scenarios generally use nitrogen in place of other inert gases.

Beyond the inferences mentioned previously in this document, some high-level assumptions for the %Environment system as a whole:
- Whole body average heat transfer, i.e., lumped parameters
- Standing patient position
- Standard air composition for heat transfer calculations
- Air is an ideal gas for heat transfer calculations
- Ignore movement of water when submerged
- No transitional heat transfer mechanics after submersion has ended (not accounting for water mass remaining on skin)
- No endogenous production of carbon monoxide
	- Endogenous production is very small so that even at very low environmental concentrations alveolar diffusion dominates
- The only carbon monoxide exposure pathophysiology simulated is tissue hypoxia
	- We do not model CO mediated damage at the cellular level (i.e. no oxidative stress or lipid peroxygenation)

@anchor environment-conditions
Conditions
----------

The environment change condition has the option to be done directly or by reading an environment xml file.  Like other conditions, the engine will stabilize the patient within that environment before running a simulation.  This is modeled most like a rapid decompression, and not necessarily the same as a patient that has lived in a "non-standard" environment and undergone physiologic changes (e.g., Himalayan Sherpas with high altitude adaptations).
@anchor environment-actions
Actions
-------

The environment change action has the option to be done directly or by reading an environment xml file.  Any of the parameters shown in Table 1 can be modified in this manor during runtime.

The thermal application action can directly add heat, remove heat, or set a temperature over any percentage of the patient&rsquo;s body.  They can be called individually or will sum together if called in combination.  These actions will set the active heat flow source and/or the active temperature source shown in Figure 14.  The applied area or fraction is used to determine the average amount to apply to the patient, since the %Environment is modeled as a one-dimensional circuit.  When actively heating or cooling, the total power is directly scaled using the fraction of the body that is covered using the patient skin surface area.  When applying a specific temperature, the average temperature will be applied between it and the ambient value, weighted by the fraction of the body covered.

@image html EnvironmentActiveCircuit.png
<center>
<i>Figure 14. The active flow source is directly set by the heating and/or cooling actions.  It is boxed in the circuit diagram.</i>
</center><br>

@anchor environment-events
## Events
There are currently no events in the %Environment system.

@anchor environment-results
Results and Conclusions
=======================

%Verification
--------------------------------------

Results using the equations from ISO 7730:2005 were used to compare to %BioGears %Environment outputs.  The standard focuses mostly on two thermal comfort metrics, Predictive Mean Vote and Predictive Percentage Dissatisfied @cite ISOPDF.  Although it is intended for the ergonomics of thermal environments, intermediate values can be extracted for the values shown in Table 6.  The connection points that connect to the %Energy system are replaced with a heat source equal to the metabolic rate, shown in Figure 15.  The other circuit elements are set to match the Inputs columns in Table 6.  The static circuit is run until everything balances and steady state is reached.  The clothing surface temperature is collected directly from the clothing node and system heat loss values are collected directly from the flow in the appropriate paths.

The following assumptions are made for this verification test:
- Complete conservation of thermal energy with none stored
- Constant respiration rate of 16 breaths/min
- Constant sweat rate of 1e-5 kg/s

@image html EnvironmentVerificationCircuit.png
<center>
<i>Figure 15. The %Energy system is replaced for the %Environment unit test by a single heat source to represent the metabolic rate.  None of the energy is stored, but instead directly affects the %Environment.</i>
</center><br>

The results in Table 6 have the skin temperatures calculated by %BioGears significantly higher than that by the ISO standard code.  However, the individual heat loss values seem reasonable and show the correct trends based on the inputs.  The %Environment model conserves energy as seen in the last column with the difference in heat produced by the metabolic rate to the sum of the heat loss by all four means.  The ISO standard does not conserve energy in the same way, and makes several looser assumptions that the %Environment system does not.  This could explain many of the discrepancies.  The requirement for the %Environment system to be able to handle extreme environments with feedback from other systems requires it to more strictly adhere to a rigorous physics based modeling approach.

<br><center>
<i>Table 6. The published ISO 7730:2005 Appendix D: Computer Program for Calculation PMV and PPD
Written in BASIC code @cite ISOPDF was used to verify the %BioGears %Environment calculations are reasonable.  Deviations can be explained by the fact that %BioGears successfully conserves energy, while the ISO 7730:2005 outputs do not.</i>
</center>
<center><img src="./images/Environment/EnvironmentVerification.png" width="1100"></center>

Validation
----------
###Forest Fire Fighter
This scenario simulates exposure to a moderate/severe forest fire . This includes a suspension of wood particulate concentration of 2.9 mg/m^3 and a concentration of carbon monoxide of 20 ppm. Environmental composition data for this scenario simulation was taken from @cite reinhardt2000smoke. Spirometry data was determined through expiratory flow curve taken during conscious respiration. Figure 16 details the difference in mass deposited in two separate respiratory compartments (following from equations 1-7) and the difference in expiratory flow before and after smoke inhalation. Slight decreases to the expiratory flow maximum are seen and coincide fairly well with validated data @cite betchley1997pulmonary. 

<center>
<table border="0">
<tr>
    <td><img src="./plots/Environment/AvleoliMassDeposited.jpg" width="550"></td>
    <td><img src="./plots/Environment/DeadSpaceMassDeposited.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/Environment/ExpiratoryFlowAfter.jpg" width="550"></td>
    <td><img src="./plots/Environment/ExpiratoryFlowBefore.jpg" width="550"></td>
</tr>
</table>
</center>
<center><i>Figure 16. The top two figures show mass deposited into two different respiratory compartments during exposure to a forest fire. Different lung compartments have different mass deposition as a function of the particle mass distribution histogram. The lower figures show the difference in expiratory flow before and after smoke inhalation. The differences a very minor but are generally in line with validated data, Table 7.</i>
</center><br>

<center>
<i>
Table 7.
</i>
</center>
|	Action	|	Notes	|	Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Functional Vital Capacity (% from baseline PFT) 	|	Functional expiratory volume over 1 second (%Change)	|	Forced expiratory flow (25-75) (%Change)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Environment Change: Exposed to moderate/severe forest fire 	|	Patient is exposed to smoke and CO inhalation 	|	10	|	240	|<span class="success">	no change	</span>|<span class="success">	no change	</span>|<span class="success">	No Change	</span>|
|	PFT administered before and after exposure to forest fire	|	Validation is computed against pre-exposed patient data	|	5 and 310	|	310	|<span class="success">	Decrease by 1% @cite betchley1997pulmonary	</span>|<span class="warning">	Decrease 3% @cite betchley1997pulmonary	</span>|<span class="success">	Decrease 6% @cite betchley1997pulmonary	</span>|

###Carbon Monoxide
There are two scenarios used to validate carbon monoxide poisoning. The first is an exposure to CO at 200 ppm, which is the National Institute for Occupational Safety and Health ceiling @ref osha_co. The second is an exposure to CO at 5000 ppm, which is an extreme exposure that should result in death. Because the many of the symptoms of CO poisoning are not modeled in %BioGears (e.g. head ache, nausea, syncope), validation is conducted by comparing the carboxyhemoglobin kinetics to the Coburn-Forster-Kane equation @ref coburn1965considerations and to data presented in @ref stewart1975effect.

<center>
<i>
Table 8.
</i>
</center>
|	Action	|	Notes	|	Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Pulse Oximetry	|	Oxygen Saturation	|	Percent Carboxyhemoglobin	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Environment Change: Carbon Monoxide 200ppm	|	Person is exposed to 200ppm carbon monoxide. The maximal steady-state concentration 	|	5	|	N/A	|<span class="success">	95 to 98% @cite guyton2006medical	</span>|<span class="success">	95 to 98% @cite guyton2006medical	</span>|<span class="success">	Approx. 0 @cite boron2012medical pg. 690	</span>|
|	None	|	10 minutes	|	N/A	|	605	|<span class="success">	Normal @cite rose2016carbon	</span>|<span class="success">	Decreased	</span>|<span class="warning">	1 to 2 % @cite stewart1975effect	</span>|
|	None	|	30 minutes	|	N/A	|	1805	|<span class="success">	Normal @cite rose2016carbon	</span>|<span class="success">	Decreased	</span>|<span class="warning">	3 to 4 % @cite stewart1975effect	</span>|
|	None	|	1 hour	|	N/A	|	3605	|<span class="success">	Normal @cite rose2016carbon	</span>|<span class="success">	Decreased	</span>|<span class="warning">	5 to 6 % @cite stewart1975effect	</span>|
|	None	|	4 hours	|	N/A	|	12005	|<span class="success">	Normal @cite rose2016carbon	</span>|<span class="success">	Decreased	</span>|<span class="success">	12 to 16% @cite stewart1975effect	</span>|

<center>
<i>
Table 9.
</i>
</center>
|	Action	|	Notes	|	Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Pulse Oximetry	|	Oxygen Saturation	|	Percent Carboxyhemoglobin	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Environment Change: Carbon Monoxide 5000ppm	|	Person is exposed to 5000ppm carbon monoxide. The maximal steady-state concentration 	|	5	|	N/A	|<span class="success">	95 to 98% @cite guyton2006medical	</span>|<span class="success">	95 to 98% @cite guyton2006medical	</span>|<span class="success">	Approx. 0 @cite boron2012medical pg. 690	</span>|
|	None	|	1 hour	|	N/A	|	3605	|<span class="success">	Normal @cite rose2016carbon	</span>|<span class="success">	Decreased	</span>|<span class="success">	approx. 80% @cite stewart1975effect Increased @cite turino1981effect	</span>|
|	None	|	2 hours - Irreversible state should happen by this point	|	N/A	|	7205	|<span class="success">	Normal @cite rose2016carbon	</span>|<span class="success">	Death (Irreversible State)	</span>|<span class="success">	Death (Irreversible State)	</span>|

See the @ref EnergyMethodology for validation in varying and extreme environments.

See the @ref DrugsMethodology for validation of airborne agents.

Conclusions
-----------

The %Environment System successfully allows the user to manipulate parameters external to the patient.
@anchor environment-future
Future Work
===========

Coming Soon
-----------

There are no planned near term additions.

Recommended Improvements
------------------------

Here are some recommended improvements:
- Add the ability to define submerged surrounding liquid properties (e.g., sea water at depth)
- Add a segmented model, instead of an averaged whole body approach
- Account for patient posture
- Add Thermal Comfort Assessment - Predictive Mean Vote (PMV) and Predictive Percentage Dissatisfied (PPD)
- Integrate 3D models for indoor, outdoor, urban, etc. to modify values&mdash;especially mean radiant temperature and velocity (wind & motion)
- Add the ability to auto-populate environment parameters based on location and time of year
	- From lat/lon and date meteorological inputs
	- From real-time data
- Add radiation analysis

@anchor environment-appendices
Appendices
==========

Acronyms
--------

CDM - Common Data Model

PMV - Predictive Mean Vote

PPD - Predictive Percentage Dissatisfied

The variables presented in Table 10 are used throughout this document.
<br><center>
<i>Table 10. There are many parameters used to determine the %Environment&rsquo;s thermal state.</i>
</center>
<center><img src="./images/Environment/EnvironmentVariables.png" width="1100"></center>

Data Model Implementation
-------------------------

@ref EnvironmentTable "Environment"

@anchor environment-compartments
Compartments
------------

- Active
- Ambient
- Clothing
- Enclosure
- ExternalCore
- ExternalSkin
- ExternalGround
%Patient Methodology {#PatientMethodology}
==========================

@anchor patient-overview
Overview
========

Abstract
--------

%BioGears allows for patient variability via a set of parameters used to define aspects of the simulated human. These parameters are used throughout the system models to manipulate the physiological responses and homeostatic state.  %BioGears is deployed with several defined patients created for various testing and analysis purposes.
@anchor patient-intro
Introduction
------------

The %BioGears program consists of the development and integration of multiscale models for an individualized whole-body predictive patient physiology model. %BioGears incorporates existing and novel models of organs and physiological systems into a whole-body model of a single generic individual within a reference population. Therefore, specific patient definitions are needed to create a virtual representation of that patient. The patient definition leverages the Commmon Data Model (CDM) to allow physiological parameters to be disseminated throughout the engine.
@anchor patient-design
System Design
=============

Background and Scope
--------------------

### Requirements

The %BioGears patient needs to contain all user-definable parameters specific to an individual. Any parameters that are not explicitly set need to be calculated based on standard, healthy values.

### Approach

Each patient parameter falls into one of three categories:
- <b>Required:</b> The only parameter that is required to be set on the patient at runtime is Sex. All other values not explicitly set will be determined using default values and/or calculated outputs.
- <b>Optional:</b> These parameters can be optionally explicitly set by the user. Many optional values have bounds associated with them to ensure a healthy initial patient.
- <b>Unallowed:</b> These parameters are calculated or determined through simulation based on other settings. They can not be modified by the user.

Table 1 provides the values and equations used to determine optional values not explicitly set by the user and those unallowed to be externally modified.
@anchor patient-features
Features and Capabilities
-------------------------

### Definitions

<center>
*Table 1. All patient parameters, how they are determined, and upper and lower limits.*
</center>

|	Parameter	|	Modification Category	|	Initial Value (If not Explicitly Set)	|	Lower Bound	|	Upper Bound	|	Notes	|
|	---	|	---	|	---	|	---	|	---	|	---	|
|	Sex	|	Required	|	\f[M = Male,F = Female\f]	|	 -	|	 -	|	This is the only parameter required to be explicitly set.	|
|	Age	|	Optional	|	\f[A[yr] = 44\f]	|	18 yr	|	65 yr	|	No pediatric or geriatric modeling.	|
|	Weight	|	Optional	|	\f[W[kg] = 21.75H{[m]^2}\f]	|	Bmi = 16 kg/m<sup>2</sup>	|	Bmi = 30 kg/m<sup>2</sup>	|	Bmi = Body Mass Index = W[kg]/H[m]<sup>2</sup>. 21.75 kg/m<sup>2</sup> is standard. No severly Underweight or Obese. @cite World2006bmi	|
|	Height	|	Optional	|	\f[{\rm{H[cm] = }}\left\{ {\begin{array}{*{20}{c}} {{\rm{177}},\;M}\\ {{\rm{163}},\;F} \end{array}} \right.\f]	|	M = 163 cm, F = 151 cm	|	M = 190 cm, F = 175.5 cm	|	Min = 3rd percentile, Max = 97th percentile, Standard = 50th percentile. @cite Centers2016clinical	|
|	Body Fat Fraction	|	Optional	|	\f[{\rm{Ff[cm] = }}\left\{ {\begin{array}{*{20}{c}} {{\rm{0}}{\rm{.21}},\;M}\\ {0.28,\;F} \end{array}} \right.\f]	|	M = 0.25, F = 0.32	|	M = 0.02, F = 0.10	|	No obese and not less than essential fat. @cite muth2009what	|
|	Carina To Teeth Distance	|	Optional	|	\f[{\rm{Ctd[cm] = }}\left\{ {\begin{array}{*{20}{c}} {{\rm{11}}{\rm{.413  +  (0}}{\rm{.072 H[cm])  -  3}},\;M}\\ {{\rm{13}}{\rm{.555  +  (0}}{\rm{.056 H[cm])  -  3}},\;F} \end{array}} \right.\f]	|	 -	|	 -	|	@cite gomez2016estimation	|
|	Blood Volume Baseline	|	Optional	|	\f[Bv[mL] = 65.6W{[kg]^{1.02}}\f]	|	Bv * 0.85	|	Bv * 1.15	|	Above Stage 1 Hypovolemia. @cite Morgan2006Clinical	|
|	Basal Metabolic Rate	|	Optional	|	\f[{\rm{Bmr[kcal/day] = }}\left\{ {\begin{array}{*{20}{c}} {{\rm{88}}{\rm{.632  +  3}}{\rm{.397W[kg]  +  4}}{\rm{.799H[cm]  -  5}}{\rm{.677A[yr]}},\;M}\\ {{\rm{447}}{\rm{.593  +  9}}{\rm{.247W[kg]  +  3}}{\rm{.098H[cm]  -  4}}{\rm{.330A[yr]}},\;F} \end{array}} \right.\f]	|	 -	|	 -	|	@cite roza1984metabolic	|
|	Heart Rate Baseline	|	Optional	|	\f[Hr[bpm] = 72\f]	|	50 bpm	|	110 bpm	|	Bradycardia < 60 bpm & Tachycardia > 100 bpm.	|
|	Systolic Arterial Pressure Baseline	|	Optional	|	\f[Sys[mmHg] = 114\f]	|	90 mmHg	|	120 mmHg	|	No hypotension or hypertension. 	|
|	Diastolic Arterial Pressure Baseline	|	Optional	|	\f[Dia[mmHg] = 73.5\f]	|	60 mmHg	|	80 mmHg	|	No hypotension or hypertension. Includes pressure fraction check: Dia > 0.75 Sys.	|
|	Respiration Rate Baseline	|	Optional	|	\f[Rr[bpm] = 16\f]	|	12 bpm	|	20 bpm	|		|
|	Alveoli Surface Area	|	Optional	|	\f[Asa[{m^2}] = \frac{{{\rm{Tlc[L]}}}}{{6.17}} \times {\rm{70}}\f]	|	 -	|	 -	|	Calculated using standard Tlc (6.17 L) @cite ganong1995review and standard Asa (70 m2) @cite roberts2000gaseous.	|
|	Right Lung Ratio	|	Optional	|	\f[Rlr = 0.525\f]	|	0.5	|	0.6	|		|
|	Skin Surface Area	|	Optional	|	\f[Ssa[{m^2}] = 0.20247W{[kg]^{0.45}}H{[m]^{0.725}}\f]	|	 -	|	 -	|	@cite du1989formula	|
|	Heart Rate Maximum	|	Optional	|	\f[H{r_{\max }}[bpm] = 208 - 0.7A[yr]\f]	|	 -	|	 -	|	@cite tanaka2001age	|
|	Heart Rate Minimum	|	Optional	|	\f[H{r_{\min }}[bpm] = 0.001\f]	|	 -	|	 -	|		|
|	Functional Residual Capacity	|	Optional	|	\f[Frc[mL] = 30W[kg]\f]	|	 -	|	 -	|	@cite ganong1995review	|
|	Total Lung Capacity	|	Optional	|	\f[Tlc[mL] = 80W[kg]\f]	|	 -	|	 -	|	@cite ganong1995review	|
|	Residual Volume	|	Optional	|	\f[Rv[mL] = 16W[kg]\f]	|	 -	|	 -	|	@cite ganong1995review	|
|	Lean Body Mass	|	Unallowed	|	\f[Lbm = W(1 - Ff)\f]	|	 -	|	 -	|		|
|	Body Density	|	Unallowed	|	\f[Bd = {\left( {\frac{{4.95}}{{Ff + 4.5}} + \frac{{4.57}}{{Ff + 4.142}}} \right)}/2\f] |	 -	|	 -	|	Average of Siri @cite siri1961body and Brozek @cite brovzek1963densitometric equations.	|
|	Mean Arterial Pressure	|	Unallowed	|	\f[Map = \frac{1}{3}Sys + \frac{2}{3}Dia\f]	|	 -	|	 -	|	@cite guyton2006medical	|
|	Tidal Volume Baseline	|	Unallowed	|	\f[Tv[mL] = 37W[kg] - Frc[mL]\f]	|	 -	|	 -	|	@cite ganong1995review	|
|	Expiratory Reserve Volume	|	Unallowed	|	\f[Erv = Frc - Rv\f]	|	 -	|	 -	|	@cite ganong1995review	|
|	Inspiratory Capacity	|	Unallowed	|	\f[Ic = Tlc - Frc\f]	|	 -	|	 -	|	@cite ganong1995review	|
|	Inspriatory Reserve Volume	|	Unallowed	|	\f[Irv = Ic - Tv\f]	|	 -	|	 -	|	@cite ganong1995review	|
|	Vital Capacity	|	Unallowed	|	\f[Vc - Tlc - Rv\f]	|	 -	|	 -	|	@cite ganong1995review	|

### Stabilization

All patient parameters are set at the beginning of the resting stabilization period (see @ref SystemMethodology). These values are used to modify the inner workings of systems. The complex interactions require a simulated period to allow physiological parameters to achieve homeostatic values. After each stabilization period (i.e., resting, conditions, and feedback), several parameters are reset due to their reliance on combined effects. The patient definition only allows for healthy values - any chronic pathophysiology needs to be handled using the CDM condition definition. 

While every effort has been made to allow any combination of patient parameters within bounds, there is no guarantee that all combinations will be able to reach a stable starting homeostatic point. 

### Dependencies

Patient parameters are not dependent on any systems, but many systems are dependent on specific patient values. Some parameters are used to modify baseline fluid/thermal circuit values or direct calculation of outputs, and others manipulate how feedback mechanisms work. For details see the following system methodology sections regarding dependencies and patient variability:

@secreflist
  @refitem cardiovascular-variability "Cardiovascular Dependencies"
  @refitem drugs-variability "Drugs Dependencies"
  @refitem energy-variability "Energy Dependencies"
  @refitem nervous-variability "Nervous Dependencies"
  @refitem renal-dependencies "Renal Dependencies"
  @refitem respiratory-variability "Respiratory Dependencies"
@endsecreflist

@anchor patient-override
Override Capabilities
=======================

%Overview
---------------------------

Certain patient parameters can now have an override action called to change them mid-run. Currently BioGears operates utilizing a patient file to prepare a set of pre-defined patient-specific parameters which allow a user to modify a patient before a scenario run. If the user wants to modify a parameter during a scenario, then an override action must be used. 

An override action consists of two required fields in addition to numerous optional parameter calls. The two required field are state and conformance. The state switch simply lets BioGears know whether the override action call is being turned on (to permit overrides) or off (to return to normal BioGears simulation). The conformance call determins whether the override is allowed to affect other BioGears parameters and conditions. Turning conformance off will only create a mask of the desired parameter in the simularion while turning it on will alter the scenario altogether. The conformance (on) feature is only currently compatible with certain parameters, though turning it on with an incompatible parameter will still allow an irreversible state (which is turned off when conformance is turned off). An example of a call would be as follows, with the override being turned on, set for an hour, and then turned off:

'''<Action xsi:type="OverrideData" State="On" Conformant="On">
		<HeartRateOverride value="120" unit="1/min"/>		
</Action>
<Action xsi:type="AdvanceTimeData">
        <Time value="1" unit="hr"/>       
</Action>
<Action xsi:type="OverrideData" State="Off" Conformant="On">		
</Action>'''

The current list of override parameters (with conformance compatible parameters designated with a "Conf.") is as follows:

- <b>Arterial Blood Pressure</b> 
- <b>Calcium Concentration</b>
- <b>Carbon Dioxide Saturation</b>
- <b>Carbon Monoxide Saturation</b>
- <b>Glucose Concentration</b>
- <b>Lactate Concentration</b>
- <b>Oxygen Saturation</b>
- <b>Phosphate Concentration</b>
- <b>Potassium Concentration</b>
- <b>Sodium Concentration</b>
- <b>Total Bilirubin</b>
- <b>Venous Blood pH</b>
- <b>White Blood Cell Count</b>
- <b>Blood Volume</b>
- <b>Cardiac Output</b>
- <b>Diastolic Arterial Pressure</b> - Conf.
- <b>Heart Rate</b> - Conf.
- <b>Heart Stroke Volume</b>
- <b>Mean Arterial Pressure</b> - Conf.
- <b>Systolic Arterial Pressure</b> - Conf.
- <b>Insulin Synthesis Rate</b>
- <b>Glucagon Synthesis Rate</b>
- <b>Acheived Exercise Level</b>
- <b>Chloride Lost to Sweat</b>
- <b>Core Temperature</b>
- <b>Creatinine Production Rate</b>
- <b>Exercise Mean Arterial Pressure Delta</b>
- <b>Fatigue Level</b>
- <b>Lactate Production Rate</b>
- <b>Potassium Lost to Sweat</b>
- <b>Skin Temperature</b>
- <b>Sodium Lost to Sweat</b>
- <b>Sweat Rate</b>
- <b>Total Metabolic Rate</b>
- <b>Total Work Rate Level</b>
- <b>Left Afferent Arteriole Resistance</b>
- <b>Left Glomerular Filtration Rate</b>
- <b>Left Reabsorption Rate</b>
- <b>Renal Blood Flow</b>
- <b>Renal Plasma Flow</b>
- <b>Right Afferent Arteriole Resistance</b>
- <b>Right Glomerular Filtration Rate</b>
- <b>Right Reabsorption Rate</b>
- <b>Urination Rate</b>
- <b>Urine Osmolality</b>
- <b>Urine Volume</b>
- <b>Urine Urea Nitrogen Concentration</b>
- <b>Expiratory Flow</b>
- <b>Inspiratory Flow</b>
- <b>Pulmonary Compliance</b>
- <b>Pulmonary Resistance</b>
- <b>Respiration Rate</b> - Conf.
- <b>Target Pulmonary Ventilation</b>
- <b>Tidal Volume</b>
- <b>Total Alveolar Ventilation</b>
- <b>Total Lung Volume</b>
- <b>Total Pulmonary Ventilation</b>
- <b>Extravascular Fluid Volume</b>
- <b>Intracellular Fluid Volume</b>
- <b>Liver Glycogen</b>
- <b>Muscle Glycogen</b>
- <b>Stored Fat</b>
- <b>Stored Protein</b>


@anchor patient-results
Results and Conclusions
=======================

%Verification and Validation
---------------------------

<center><br>
*Table 2. There are several patients that we created for base validation, system-specific patient effects validation, and extreme case will stabilization.*
</center>

|	<b>Name</b>	|	Standard Male	|	Standard Female	|	Default Male	|	Default Female	|	Overweight	|	Underweight	|	Tachycardic	|	Bradycardic	|	Extreme Female	|	Extreme Male	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	<b>Sex</b>	|	Male	|	Female	|	Male	|	Female	|	Male	|	Male	|	Male	|	Male	|	Female	|	Male	|
|	<b>Age</b>	|	44 yr	|	44 yr	|	-	|	-	|	44 yr	|	44 yr	|	44 yr	|	44 yr	|	18 yr	|	65 yr	|
|	<b>Weight</b>	|	170 lb	|	130 lb	|	-	|	-	|	215 lb	|	115 lb	|	170 lb	|	170 lb	|	-	|	-	|
|	<b>Height</b>	|	71 in	|	64 in	|	-	|	-	|	71 in	|	71 in	|	71 in	|	71 in	|	4.5 ft	|	7 ft	|
|	<b>Body Fat Fraction</b>	|	0.21	|	0.28	|	-	|	-	|	0.25	|	0.15	|	0.18	|	0.18	|	0.32	|	0.02	|
|	<b>Heart Rate Baseline</b>	|	72 bpm	|	72 bpm	|	-	|	-	|	72 bpm	|	72 bpm	|	110 bpm	|	50 bpm	|	60 bpm	|	100 bpm	|
|	<b>Systolic Arterial Pressure Baseline</b>	|	114 mmHg	|	114 mmHg	|	-	|	-	|	114 mmHg	|	114 mmHg	|	114 mmHg	|	114 mmHg	|	90 mmHg	|	120 mmHg	|
|	<b>Diastolic Arterial Pressure Baseline</b>	|	73.5 mmHg	|	73.5 mmHg	|	-	|	-	|	73.5 mmHg	|	73.5 mmHg	|	73.5 mmHg	|	73.5 mmHg	|	60 mmHg	|	80 mmHg	|
|	<b>Respiration Rate Baseline</b>	|	16 bpm	|	16 bpm	|	-	|	-	|	16 bpm	|	16 bpm	|	20 bpm	|	12 bpm	|	12 bpm	|	20 bpm	|
|	<b>Right Lung Ratio</b>	|	-	|	-	|	-	|	-	|	-	|	-	|	-	|	-	|	0.5	|	0.6	|

<center><br>
*Table 3. There are several patients we created for combined effects validation and showcase scenarios.*
</center>

|	<b>Name</b>	|	Cynthia	|	Gus	|	Joel	|	Nathan	|	Rick	|	Hassan	|	Soldier	|	Jeff	|	Carol	|	Jane	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	<b>Sex</b>	|	Female	|	Male	|	Male	|	Male	|	Male	|	Male	|	Male	|	Male	|	Female	|	Female	|
|	<b>Age</b>	|	30 yr	|	32 yr	|	44 yr	|	44 yr	|	23 yr	|	28 yr	|	22 yr	|	25 yr	|	40 yr	|	18 yr	|
|	<b>Weight</b>	|	130 lb	|	190 lb	|	170 lb	|	170 lb	|	161 lb	|	185 lb	|	170 lb	|	180 lb	|	160 lb	|	120 lb	|
|	<b>Height</b>	|	64 in	|	70 in	|	71 in	|	71 in	|	68 in	|	72 in	|	71 in	|	72 in	|	66 in	|	64 in	|
|	<b>Body Fat Fraction</b>	|	0.18	|	0.18	|	0.18	|	0.18	|	0.18	|	0.18	|	0.18	|	0.12	|	0.25	|	0.18	|
|	<b>Heart Rate Baseline</b>	|	72 bpm	|	93 bpm	|	110 bpm	|	72 bpm	|	100 bpm	|	110 bpm	|	84 bpm	|	72 bpm	|	72 bpm	|	72 bpm	|
|	<b>Systolic Arterial Pressure Baseline</b>	|	114 mmHg	|	90 mmHg	|	120 mmHg	|	114 mmHg	|	120 mmHg	|	90 mmHg	|	114 mmHg	|	114 mmHg	|	114 mmHg	|	114 mmHg	|
|	<b>Diastolic Arterial Pressure Baseline</b>	|	73.5 mmHg	|	60 mmHg	|	80 mmHg	|	73.5 mmHg	|	80 mmHg	|	60 mmHg	|	73.5 mmHg	|	73.5 mmHg	|	73.5 mmHg	|	73.5 mmHg	|
|	<b>Respiration Rate Baseline</b>	|	18 bpm	|	14 bpm	|	15 bpm	|	16 bpm	|	16 bpm	|	18 bpm	|	16 bpm	|	16 bpm	|	18 bpm	|	18 bpm	|

All of the patients that are included with the %BioGears deployment have been validated to ensure that they reach a homeostatic point that hits all of the defined criteria. The following tables show the state of the patients after all stabilization phases (see @ref SystemMethodology for details) and compare the resulting patient values to those initially set/computed. Values that are explicitly set are expected to have very tight tolerances, while those that are initially calculated may drift due to the complex interaction of all patient parameters and system settings. The most sensitive of these parameters to external effects is tidal volume.

Values that are colored green are within 10% of the original set or calculated/estimated (expected) value, yellow are within 30%, and red are greater than 30% error. Values that are not within 10% are not necessarily a failure.  As previously described, many of these are estimated before the stabilization simulation begins, and are really determined and overwritten after all other factors reach homeostasis. Those values that are explicitly set by the user/patient file should have much tighter tolerances.

<center>
*Table 4. StandardMale patient values with expected and %BioGears Engine output.*
</center><br>

@insert doc/validation/StandardMalePatientValidationTable.md

<center>
*Table 5. StandardFemale patient values with expected and %BioGears Engine output.*
</center><br>

@insert doc/validation/StandardFemalePatientValidationTable.md

<center>
*Table 6. DefaultMale patient values with expected and %BioGears Engine output.*
</center><br>

@insert doc/validation/DefaultMalePatientValidationTable.md

<center>
*Table 7. DefaultFemale patient values with expected and %BioGears Engine output.*
</center><br>

@insert doc/validation/DefaultFemalePatientValidationTable.md

<center>
*Table 8. Overweight patient values with expected and %BioGears Engine output.*
</center><br>

@insert doc/validation/OverweightPatientValidationTable.md

<center>
*Table 9. Underweight patient values with expected and %BioGears Engine output.*
</center><br>

@insert doc/validation/UnderweightPatientValidationTable.md

<center>
*Table 10. Tachycardic patient values with expected and %BioGears Engine output.*
</center><br>

@insert doc/validation/TachycardicPatientValidationTable.md

<center>
*Table 11. Bradycardic patient values with expected and %BioGears Engine output.*
</center><br>

@insert doc/validation/BradycardicPatientValidationTable.md

<center>
*Table 12. Cynthia patient values with expected and %BioGears Engine output.*
</center><br>

@insert doc/validation/CynthiaPatientValidationTable.md

<center>
*Table 13. Gus patient values with expected and %BioGears Engine output.*
</center><br>

@insert doc/validation/GusPatientValidationTable.md

<center>
*Table 14. Joel patient values with expected and %BioGears Engine output.*
</center><br>

@insert doc/validation/JoelPatientValidationTable.md

<center>
*Table 15. Nathan patient values with expected and %BioGears Engine output.*
</center><br>

@insert doc/validation/NathanPatientValidationTable.md

<center>
*Table 16. Hassan patient values with expected and %BioGears Engine output.*
</center><br>

@insert doc/validation/HassanPatientValidationTable.md

<center>
*Table 17. Soldier patient values with expected and %BioGears Engine output.*
</center><br>

@insert doc/validation/SoldierPatientValidationTable.md

<center>
*Table 18. Jeff patient values with expected and %BioGears Engine output.*
</center><br>

@insert doc/validation/JeffPatientValidationTable.md

<center>
*Table 19. Carol patient values with expected and %BioGears Engine output.*
</center><br>

@insert doc/validation/CarolPatientValidationTable.md

<center>
*Table 20. Jane patient values with expected and %BioGears Engine output.*
</center><br>

@insert doc/validation/JanePatientValidationTable.md

<center>
*Table 21. ExtremeFemale patient values with expected and %BioGears Engine output.*
</center><br>

@insert doc/validation/ExtremeFemalePatientValidationTable.md

<center>
*Table 22. ExtremeMale patient values with expected and %BioGears Engine output.*
</center><br>

@insert doc/validation/ExtremeMalePatientValidationTable.md

Conclusions
-----------

The methodology used to define and set patient parameters is successful in allowing for the simulation of unique humans.
@anchor patient-future
Future Work
===========

Coming Soon
-----------

There are no planned near-term additions.

Recommended Improvements
------------------------

The %BioGears physiology engine is providing consistent and accurate physiology simulation, and its ability to simulate physiology for a generic representative of a population will improve as development continues. It is not within the scope of the current project for %BioGears to predictively simulate the physiology of a specific (or average) individual within a population, but the development of technologies that make %BioGears usable and extensible to the community also facilitate multiscale model development and can be leveraged for predictive simulation and thus clinical use.

One way to extend %BioGears for clinical use is to use data from sources such as electronic health records (EHRs), lab reports, fitness evaluations, or wearable sensors to create a personalized physiological profile. A tool could then translate the personal physiological profile into parameters that can inform the %BioGears physiology models within the systems' framework.
@anchor patient-appendices
Appendices
==========

Acronyms
--------

CDM - Common Data Model

Data Model Implementation
-------------------------

BioGears
Blood Chemistry Methodology {#BloodChemistryMethodology}
==========================

@anchor bloodchemistry-overview
Overview
========
Abstract
--------
The purpose of the %BioGears&reg; Blood Chemistry System is primarily to hold the system-level blood substance data. It also uses information about the blood to compute plasma volume and whole-body mass and concentration information. In the future, all blood chemistry computations will take place in the Blood Chemistry system, including the acid-base balance computations of blood gas binding and species distribution that are currently performed by the Saturation class. For this reason, the saturation and acid-base balance models are described in the [approach](@ref bloodchemistry-approach) section of this document.

Introduction
------------
The Blood Chemistry system provides a link by which systems and users can access blood substance data.  The Blood Chemistry system also processes information about the blood to compute plasma volume and the %BioGears [Complete Blood Count and Metabolic Panel Assessments](@ref bloodchemistry-assessments).

Although they currently reside in a different class, the blood gas distribution models in the Saturation class are conceptually a part of the Blood Chemistry system. Therefore, models from the Saturation class are described here.

@anchor bloodchemistry-system-design
System Design
=============
Background and Scope
--------------------
The analysis of the blood constituents is critical for assessing physiological function. %Respiratory function and acid-base status are assessed by analyzing the distribution of gas species in the blood. Because respiratory, renal, and cardiovascular system functions interact to maintain acid-base homeostasis, abnormal function in any of these systems can cause an acid-base disturbance. For example, respiratory conditions which lead to an increase in the amount of dissolved CO2 in the blood, such as bronchitis, are associated with respiratory acidosis, while metabolic acidosis often accompanies renal pathologies. In order to investigate the origin of an acid-base disturbance, a provider may order a metabolic panel or a complete blood count. These two assessments, along with blood gas analysis, provide a detailed account of the composition of a patient's blood and assist with characterization of acid-base disturbances. 

There are three models commonly used for acid-base analysis. One model, often called the Boston model, is relatively simple but relies entirely on the Henderson-Hasselbalch equation.  Bicarbonate equilibrium alone, however, is demonstrably insufficient for predicting carbon dioxide species distribution in human blood @cite wooten2010standard. To compensate, the Boston model includes six bicarbonate-based rules that are used to assess acid-base status. Another model extends the Boston approach by introducing a new parameter called "base excess" to quantify the metabolic component of an acid-base disturbance @cite siggaard1977van. The problem with base-excess is that it is not CO2-invariant in vivo as a result of ionic shifts between the intravascular and interstitial compartments. However, at a constant hemoglobin concentration of 50 g/L, base excess is CO2-invariant when it is computed from measured plasma pH and PCO2. A third model developed by Peter Stewart (and thus called the Stewart approach) describes acid-base balance mechanistically through electrochemical equilibrium dynamics @cite kellum2009stewart.

The Stewart approach has been the subject of criticism. Some researchers have used dilutional acidosis as a model to argue that Stewart's approach does not actually provide any mechanistic insights because "neither the principle of electroneutrality, nor a change in [SID], nor increased H2O dissociation play a mechanistic role" @cite doberer2009critique. However, others have noted that the Stewart method "is physicochemically well grounded" and that a dilution of the [SID] is "inevitably a dilution of the buffer base" @cite lang2010comment. Although the Boston model and the Copenhagen model are used more often for clinical, or bedside, analysis of acid-base status, the mechanistic detail of the Stewart model facilitates the computational modeling of acid-base balance and disturbances. We use the Stewart model as the basis for the %BioGears acid-base balance model.

@anchor bloodchemistry-requirements
### Requirements
- Acid-base status 
- %Respiratory and metabolic acid-base disturbances

### Existing
A simple oxygen-hemoglobin binding curve.

@anchor bloodchemistry-approach
### Models

#### Acid-Base and Blood Gas Distribution
The %BioGears Blood Chemistry system is the link between compartment-level and system-level blood constituent data. It populates system-level concentration and other data from compartment level data. 

Additionally, the blood chemistry computations required to compute acid-base balance and blood gas distribution are performed by the Blood Chemistry system (note that these equations are in the separate Saturation class in the current %BioGears engine). The Blood Chemistry system does not adjust mass or concentration of substances independently, but it does adjust concentrations of substance species. For example, the Blood Chemistry system does not adjust the mass or blood concentration of total carbon dioxide (CO2), but it does compute the concentrations of the constituent CO2 species: dissolved CO2, bound CO2, and bicarbonate. That is, the acid-base and hemoglobin saturation models are used to compute the CO2 species distribution. Similarly, the Blood Chemistry system does not change the total amount of oxygen in a compartment, but it does compute how much oxygen is dissolved and how much is bound to hemoglobin.

The %BioGears acid-base and blood gas distribution model is based on the Stewart model (discussed above) and the hemoglobin binding model described by Dash and Bassingthwaighte @cite dash2010erratum. As outlined by Stewart, the model begins with consideration of electroneutrality. The charge of all strong ions in solution must be zero, as demonstrated by Equation 1.

\f[\ \ \left[Na^+\right]+\left[K^+\right]+\left[I^{n+}\right]+\left[Cl^-\right]+\left[La^-\right]+\left[Ket^-\right]+\left[I^{n-}\right]+{\mathrm{[HCO}}^-_{\mathrm{3}}]-\left[A^-\right]\ =0 \f]
<center>
<i>Equation 1.</i>
</center><br>
Note that *I<sup>n,+/-</sup>* represents the n anions and cations not included in %BioGears (All other symbols are defined in [Table 6](@ref bloodchemistry-symbols)). Some ions, such as carbonate, exist in such minute concentrations that their contribution to electrical neutrality is negligible. These ions are neglected and thus excluded from Equation 1.

It is convenient to define a concentration called strong ion difference [SID], described by Equation 2, which allows Equation 1 to be reduced to Equation 3.

\f[[SID]=\left[Na^+\right]+\left[K^+\right]+\left[I^{n+}\right]-\left(\left[Cl^-\right]+\left[La^-\right]+\left[Ket^-\right]+\left[I^{n-}\right]\right) \f]
<center>
<i>Equation 2.</i>
</center><br>

\f[[SID]-{\mathrm{[HCO}}^-_{\mathrm{3}}]-[A^-]=0 \f]
<center>
<i>Equation 3.</i>
</center><br>

Equation 4 is an empirical formula for the blood protein anion contribution (*A<sup>-</sup>*) developed by Figge et al. @cite figge1992serum.

\f[\left[A^-\right]=\left[Alb\right]\left(0.123pH-0.631\right)+[Pi](0.309pH-0.469) \f]
<center>
<i>Equation 4.</i>
</center><br>

The Henderson-Hasselbalch equation and equations representing Henry's law for oxygen and carbon dioxide are shown as Equations 5, 6, and 7.

\f[pH=6.1+{\mathrm{log} \left(\frac{\left[HCO^-_3\right]}{\alpha P_{CO_2}}\right)\ } \f]
<center>
<i>Equation 5.</i>
</center><br>

\f[\left[O_2\right]={\alpha }_{o_2} P_{O_2} \f]
<center>
<i>Equation 6.</i>
</center><br>

\f[S_{CO_2}=f\left(pH,P_{O_2},P_{CO_2},T,DPG\right) \f]
<center>
<i>Equation 7.</i>
</center><br>

The above equations are combined into a single expression with three unknowns: bicarbonate concentration, the partial pressure of oxygen, and the partial pressure of carbon dioxide.

\f[[SID]-{\mathrm{[HCO}}^-_{\mathrm{3}}]-\left[Alb\right]\left\{0.123 \left(6.1+{\mathrm{log} \left(\frac{\left[HCO^-_3\right]}{\alpha_{O_{2}} P_{CO_2}}\right)\ }\right)-0.631\right\}+[Pi]\left\{0.309 \left(6.1+{\mathrm{log} \left(\frac{\left[HCO^-_3\right]}{\alpha_{CO_{2}} P_{CO_2}}\right)\ }\right)-0.469\right\}=0\f]
<center>
<i>Equation 8.</i>
</center><br>

The conservation of mass requires that the respective amounts of oxygen and carbon dioxide remain constant. Therefore, the sum of dissolved oxygen and hemoglobin-bound oxygen must remain constant. Likewise, the sum of dissolved carbon dioxide, hemoglobin-bound carbon dioxide, and bicarbonate must remain constant. The amount of oxygen or carbon dioxide bound to hemoglobin at a given partial pressure of the gas is defined by the model described in @cite dash2010erratum. The binding model and the conservation equations fully define the system, represented by Equation 8 above and Equations 9, 10, 11, and 12 below.

\f[[{\mathrm{T}}_{{\mathrm{O}}_{\mathrm{2}}}]=\left(\left[O_2\right]+4 S_{O_2}\left[Hgb\right]\right) \f]
<center>
<i>Equation 9.</i>
</center><br>

\f[[{\mathrm{T}}_{{\mathrm{CO}}_{\mathrm{2}}}]=\left(\left[CO_2\right]+\left[HCO^-_3\right]+4{S}_{CO_2}\left[Hgb\right]\right) \f]
<center>
<i>Equation 10.</i>
</center><br>

\f[S_{O_2}=f\left(pH,P_{O_2},P_{CO_2},T,DPG\right) \f]
<center>
<i>Equation 11.</i>
</center><br>

\f[S_{CO_2}=f\left(pH,P_{O_2},P_{CO_2},T,DPG\right) \f]
<center>
<i>Equation 12.</i>
</center><br>

The blood gas distribution model is used to compute the acid-base status and gas saturation in every cardiovascular compartment at every time slice. This model is used only for intravascular fluid, which is a lumped model of the intra and extracellular fluid spaces within the blood vessels. In %BioGears, the extravascular fluid does not contain hemoglobin or any substrate for gases to bind.

#### Inflammation Model

BioGears maintains an inflammation model that dictates the sytemic response to pathogen invasion, burn wounds, and hemorrhage.  The model, which is based on the work of @ref chow2005acute @ref nieman2012two @ref brown2015trauma @ref dominguez2017mathematical, accounts for resting and activated macrophages and neutrophils, pro-inflammatory cytokines (tumor necrosis factor, interleukin-6), anti-inflammatory cytokines (interleukin-10), and free radicals (nitric oxide).  The BioGears team has published two articles describing the application of this model 
to severe burn wounds @ref mcdaniel2019full and [sepsis](https://www.frontiersin.org/articles/10.3389/fphys.2019.01321/full) @ref mcdaniel2019whole.  These sources can be consulted for model equations and parameters, as well as further detail regarding model integration with the BioGears physiology engine.

@anchor bloodchemistry-data-flow
Data Flow
---------
### Initialize
Compartment-level substance data is initialized by the BioGearsSubstances controller class, and whole-body blood substance data is initialized from the compartment data by the Blood Chemistry system. The constant values of blood density, blood specific heat, plasma fraction of neutral lipid and phospholipid, white blood cell count, phosphate, and strong ion difference concentrations are also initialized by the Blood Chemistry system. Note that the strong ion difference will not be constant following a planned future improvement. The blood chemistry system does not require stabilization. Blood properties are initialized and initial system data is set in the initialization method. System data is populated throughout the engine stabilization process. 

### Preprocess
There is no system specific function for Preprocess in Blood Chemistry.

### Process
During Process, the blood concentrations, blood gases, and other blood properties that are calculated or changed by other systems are set on the blood chemistry system data objects.

### Post Process
There is no system specific function for Post Process in Blood Chemistry.

### Assessments
Assessments in %BioGears are data collected and packaged to resemble a report or analysis that might be ordered by a physician. There are two %BioGears assessments in the Blood Chemistry system: a Metabolic panel and a complete blood count (CBC). The metabolic panel is modeled after the Chem-14 blood test. Currently, eight of the fourteen Chem-14 components are included in the %BioGears metabolic panel. 

<br>
<img src="./images/BloodChemistry/BloodChemistryDataFlow.png" width="600">
<center>
*Figure 1. The data flow for the Blood Chemistry system consists of a Reset, Conditions, Preprocess, Process, Post Process, and Assessments. Only the Process and Assessments categories have system-specific functionality for the Blood Chemistry System.*
</center><br>

@anchor bloodchemistry-features
Features and Capabilities
-------------------------
The Blood Chemistry system is a support system primarily for storing and relaying data. Thus, the fidelity of the system is a direct result of the fidelity of the other system models. The concentration of a substance in the blood is computed, typically in a specific compartment or multiple compartments, during the simulation of some process within another system. For example, the blood concentration of sodium is adjusted by the [renal system](@ref RenalMethodology) during filtration and clearance and by the [tissue system](@ref TissueMethodology) in the compartment diffusion models. The whole-body, system-level blood concentration of sodium is set to the concentration in the vena cava compartment by the Blood Chemistry system.

Some calculations are preformed by the Blood Chemistry system. The red blood cell count, hematocrit, and the plasma volume are calculated here. The red blood cell volume is calculated based on the hemoglobin content of the body (a variable) and the average mass of hemoglobin per red blood cell (currently a constant parameter). The average volume of a red blood cell is then used to calculate the total volume of the red blood cells. The hematocrit is calculated as the ratio of red blood cell volume to total blood volume. The plasma volume is calculated as the total blood volume less the red blood cell volume.

@anchor bloodchemistry-variability
### Patient Variability
%BioGears substances are initialized in the body by concentration rather than by mass, so morphological variability in patients will have no effect on initial concentrations. However, patient variability does affect physiology, and homeostatic states will vary with patient parameters. For that reason, post-stabilization variance in blood chemistry, including substance concentration, is possible between two different patients. A detailed discussion of patient variability in %BioGears is available in the @ref PatientMethodology report.

### Outputs
The Blood Chemistry system provides (for circulating substances):
- Blood concentration
- Total mass in blood, tissue, and body
- Pulmonary shunt fraction (fraction of blood bypassing pulmonary oxygenation)

@anchor bloodchemistry-assumptions
Assumptions and Limitations
---------------------------
Limitations on the Blood Chemistry system are imposed by the supported systems and by the active substances in the simulation. For example, ALT, AST, and ALP are not supported substances, and therefore cannot be included in the metabolic panel assessment. 

Additionally, blood cells are not modeled in %BioGears. We attempt to account for the contribution of cellular elements to the blood volume, as described above, but white cells and platelets are not included. 

@anchor bloodchemistry-events
Events
------
### Hemolytic Transfusion Reaction (HTR)
HTR occurs when there is an incompatibility between patient and donor blood blood types @cite davenport2005pathophysiology.  The incompatibility triggers an immune response, resulting in agglutination and red blood cell death in the recipient's body. 

The %BioGears Engine triggers an HTR event when an incompatibility is detected. This is based both on the set patient's blood type as well as any blood/antigens introduced into their system. This is a reversible condition and is considered resolved when the foreign cells are removed (over time). The time it takes the body to remove foreign cells will depend on the amount introduced. Once kicked off, the event triggers a set of predator-prey equations to model agglutination and clustering of red blood cells. Once cells reach a cluster of four cells, we consider it stable enough to remove the cells in the cluster and remove associated substances. An example reaction showing the infusion of incompatible donor cells and the reaction with patient cells can be seen below:

HTR Equations

\f[K\left( {i,j} \right) = \frac{{\left( {{D_i} + {D_j}} \right)\left( {{R_i} + {R_j}} \right)}}{{\left( {4{D_1}{\rm{ }}{R_1}{\rm{ }}} \right)}} \f]
<center>
<i>Equation 13. Probablity of cellular collision between cluster of i cells and cluster of j cells</i>
</center><br>

\f[{C_{\left( {rbc,p} \right)}}\left( {t + 1} \right) = RBC\left( t \right) + \left( {RB{C_{birth}} * dt} \right) - (RB{C_{death}} * dt) - dt(\left( {K\left( {1,1} \right) * RBC\left( t \right) * RB{C_T}\left( t \right)} \right) + \left( {K\left( {2,1} \right) * {C_2}\left( t \right) * RBC\left( t \right)} \right) + (K\left( {3,1} \right) * {C_{3d}}\left( t \right) * RBC\left( t \right))) \f]
<center>
<i>Equation 14. Concentration of patient red blood cells</i>
</center><br>

\f[{C_{\left( {rbc,d} \right)}}\left( {t + 1} \right) = RB{C_T}\left( t \right) - (RB{C_{death}} * dt) - dt(\left( {K\left( {1,1} \right) * RBC\left( t \right) * RB{C_T}\left( t \right)} \right) + \left( {K\left( {2,1} \right) * {C_2}\left( t \right) * RB{C_T}\left( t \right)} \right) + (K\left( {3,1} \right) * {C_{3p}}\left( t \right) * RB{C_T}\left( t \right))) \f]
<center>
<i>Equation 15. Concentration of donor red blood cells</i>
</center><br>

\f[{C_2}\left( {t + 1} \right) = {C_2}\left( t \right) + dt(\left( {K\left( {1,1} \right) * RBC\left( t \right) * RB{C_T}\left( t \right)} \right) - \left( {K\left( {2,1} \right) * {C_2}\left( t \right) * RBC\left( t \right)} \right) - (K\left( {2,1} \right) * {C_2}(t) * RB{C_T}\left( t \right)) + ((K(2,2) * {C_2}\left( t \right) * {C_2}\left( t \right)))) \f]
<center>
<i>Equation 16. Concentration of red blood cell pairings (1 donor/1 patient</i>
</center><br>

\f[{C_{3,p}}\left( {t + 1} \right) = {C_{3p}}\left( t \right) + dt(\left( {K\left( {2,1} \right) * {C_2}\left( t \right) * RBC\left( t \right)} \right) - \left( {K\left( {3,1} \right) * {C_{3p}}\left( t \right) * RB{C_T}\left( t \right)} \right)) \f]
<center>
<i>Equation 17a. Concentration of triads (2 patient/1 donor) of red blood cells</i>
</center><br>

\f[{C_{3,d}}\left( {t + 1} \right) = {C_{3d}}\left( t \right) + dt(\left( {K\left( {2,1} \right) * {C_2}\left( t \right) * RB{C_T}\left( t \right)} \right) - \left( {K\left( {3,1} \right) * {C_{3d}}\left( t \right) * RBC\left( t \right)} \right)) \f]
<center>
<i>Equation 17b. Concentration of triads (1 patient/2 donor) of red blood cells</i>
</center><br>

\f[{C_4}\left( {t + 1} \right) = {C_4}\left( t \right) + dt(\left( {\left( {K\left( {2,2} \right) * {C_2}\left( t \right) * {C_2}\left( t \right)} \right)} \right) + (K\left( {3,1} \right) * {C_{3p}}\left( t \right) * RB{C_T}\left( t \right)) + \left( {K\left( {3,1} \right) * {C_{3p}}\left( t \right) * RBC\left( t \right)} \right)) \f]
<center>
<i>Equation 18. Concentration of red blood cell "stable" agglutinates</i>
</center><br>

In the above, D is difussion coefficient, R is radius, RBC is patient red blood cell count, RB{C_T} is transfused red blood cell count, {RB{C_{birth}} and {RB{C_{death}} are the respective birth and death rates of the cell.

@image html RBC_HTR_2axis.png
<center>
*Figure 2. %An injection of 0.5 L introduces donor blood cells which then cause a physiological response as cells (both patient and donor) are eliminated. The reaction results in such changes as increased core temperature, increased heart rate, and increased respiration.*
</center><br>

### Hypercapnia
Hypercapnia occurs when the partial pressure of carbon dioxide in the blood rises to greater than 60&nbsp;mmHg @cite guyton2006medical.  Patients with respiratory conditions, such as severe airway obstruction, frequently exhibit hypercapnia. This condition manifests as a decrease in blood pH and an increase in cardiac output, blood pressure, and heart rate.

The %BioGears Engine triggers the hypercapnia event when the partial pressure of carbon dioxide in the aorta rises above 60&nbsp;mmHg. This is a reversible condition and is considered resolved when the partial pressure falls below 57&nbsp;mmHg. This 3&nbsp;mmHg window provides a buffer to account for normal fluctuations in the model.

If the partial pressure of carbon dioxide in the blood rises to greater than 80&nbsp;mmHg, an irreversible state is triggered and it will be impossible to regain homeostasis.

### Hypoxia
Hypoxia occurs when the partial pressure of oxygen in the blood falls below 65&nbsp;mmHg @cite Pierson2000Pathophysiology. This can be due to a number of conditions that range from heart failure to poor respiratory function. This condition manifests as a decrease in oxygen saturation, increased heart rate, increased respiratory rate, decreased tidal volume, and pulmonary hypertension.

The %BioGears Engine triggers the hypoxia event when the partial pressure of oxygen in the aorta falls below 65&nbsp;mmHg. This is a reversible condition and is considered resolved when the partial pressure rises above 68&nbsp;mmHg. This 3&nbsp;mmHg window provides a buffer to account for normal fluctuations in the model.

If the partial pressure of oxygen in the blood falls below 15&nbsp;mmHg, an irreversible state is triggered and it will be impossible to regain homeostasis.

### Brain Oxygen Deficit
The brain is unable to complete any significant anaerobic metabolism. Therefore, without oxygen in the brain, unconsciousness results within five to ten seconds, and permanent damage can occur within five to ten minutes @cite guyton2006medical. Additionally, irreversible damage can occur if the oxygen tension in the brain is too low for a prolonged period of time @cite dhawan2011neurointensive. 

There are two events in %BioGears related to an oxygen deficit in the brain: Brain Oxygen Deficit and Critical Brain Oxygen Deficit. %BioGears triggers the brain oxygen deficit event when the partial pressure of oxygen in the brain drops below 21 mmHg. If the brain remains in this deficit state for 30 minutes, an irreversible state is triggered, and it will be impossible to regain homeostasis (akin to death). If the partial pressure of oxygen in the brain drops below 10 mmHg, the patient enters a critical brain oxygen deficit state, which causes an irreversible state after 10 minutes. If the critical deficit event is active, it is removed when the oxygen partial pressure increases above 12 mmHg, and only the less-critical deficit event will be active. Both deficit events are removed when the oxygen partial pressure returns above 25 mmHg, assuming the irreversible state was never reached. The threshold values are chosen based on empirical data reviewed in summary in @cite dhawan2011neurointensive, and from data presented in @cite purins2012brain and @cite doppenberg1998determination.

### Myocardium Oxygen Deficit
Blood flow through the coronary arteries provides required oxygen to the heart muscle. If the oxygen supply to the heart is limited through respiratory distress or some other condition, the demand for oxygen will cause dilation of the coronary arteries. By increasing the flow of blood to the heart muscle, the oxygen supply increases @cite guyton2006medical. This effect is not currently modeled in %BioGears, but the myocardium oxygen level is observed and an event is triggered when it becomes too low. In the future, the resistance to flow in the coronary arteries will be increased to represent blood vessel dilation, and the asystole rhythm change will be tied to the arterial partial pressure of oxygen.

The %BioGears Engine triggers the myocardium oxygen deficit event when the partial pressure of oxygen is less than 5&nbsp;mmHg. If the oxygen level in the heart muscle remains low for more than 40 minutes an irreversible state is triggered and it will be impossible to regain homeostasis. 

### Acid-Base Disturbance Events
Acid-base disturbances can be respiratory or metabolic in origin. When an acid-base disturbance is caused by a change in the total amount of carbon dioxide in the blood, it is considered to be a respiratory disturbance because that system regulates carbon dioxide. If the origin is a change in the strong ion difference, then the disturbance is said to be metabolic in origin. Because the respiratory system responds to a metabolic disturbance and the renal and other systems respond to a respiratory disturbance with compensatory mechanisms, it is difficult to observe a purely metabolic or respiratory disturbance. For that reason, disturbances are sometimes further classified as acute, compensated, and mixed @cite hall2011guyton. There are four acid-base disturbance events observed in %BioGears.

#### Metabolic Acidosis
The metabolic acidosis event is triggered when the blood pH drops below the lower bound of the normal range, which is 7.35 for arterial blood (note that the blood pH is sampled from the aorta compartment) @cite Leeuwen2015laboratory, **and** the bicarbonate concentration is less than 22.0&nbsp;mM. This reversible event is removed when the blood pH increases above 7.38. The small buffer of 0.03 is to allow for numerical fluctuations during transitions. If the blood becomes so acidic that the pH drops below 6.5, an irreversible state is triggered, and it will be impossible to regain homeostasis.

#### Metabolic Alkalosis
The metabolic alkalosis event is triggered when the blood pH increases above the upper bound of the normal range, which is 7.45 for arterial blood (note that the blood pH is sampled from the aorta compartment) @cite Leeuwen2015laboratory, **and** the bicarbonate concentration is greater than 26.0&nbsp;mM. This reversible event is removed when the blood pH decreases below 7.42. The small buffer of 0.04 is to allow for numerical fluctuations during transitions. An irreversible state is triggered when the pH increases above 8.5.

#### %Respiratory Acidosis
The respiratory acidosis event is triggered when the blood pH drops below the lower bound of the normal range, which is 7.35 for arterial blood (note that the blood pH is sampled from the aorta compartment) @cite Leeuwen2015laboratory, **and** the partial pressure of carbon dioxide in the arteries is greater than 47.0&nbsp;mmHg. This reversible event is removed when the blood pH increases above 7.38. The small buffer of 0.03 is to allow for numerical fluctuations during transitions. If the blood becomes so acidic that the pH drops below 6.5, an irreversible state is triggered, and it will be impossible to regain homeostasis.

#### %Respiratory Alkalosis
The respiratory alkalosis event is triggered when the blood pH increases above the upper bound of the normal range, which is 7.45 for arterial blood (note that the blood pH is sampled from the aorta compartment) @cite Leeuwen2015laboratory, **and** the partial pressure of carbon dioxide in the arteries less than 44.0&nbsp;mmHg. This reversible event is removed when the blood pH decreases below 7.42. The small buffer of 0.04 is to allow for numerical fluctuations during transitions. An irreversible state is triggered when the pH increases above 8.5.

@image html AcidBaseEvents.png
<center>
*Figure 3. The acidosis and alkalosis events are broken down into either metabolic or respiratory induced. Metabolic induced acidosis/alkalosis is driven from a shift in bicarbonate
while respiratory induced acidosis/alkalosis is due to a chance in arterial carbon dioxide. The shift between either acidosis or alkalosis occurs at a blood pH of 7.4.*
</center><br>

@anchor bloodchemistry-assessments
Assessments
-----------
### Complete Blood Count

A CBC is a blood test that is used to assess a patient&rsquo;s overall health. It measures several components of the blood including red blood cells, white blood cells, hemoglobin concentration, and hematocrit. When measured, these values can reveal a variety of disorders, including anemia and infection. A complete list of the components of this assessment and the results can be found in the [Validation](@ref bloodchemistry-validation-assessments) section.

### Metabolic Panel

A comprehensive metabolic panel, Chem 14, is a screening assessment used by clinicians to evaluate the overall health of a patient. This assessment can be used to screen for diabetes, liver disease, and kidney disease. This test measures a variety of substance concentrations in the blood, including sodium, protein, calcium, glucose, creatinine, and blood urea nitrogen (BUN). A complete list of the components of this assessment and the results can be found in the [Validation](@ref bloodchemistry-validation-assessments) section.

@anchor bloodchemistry-results
Results and Conclusions
=======================

Validation - Resting Physiologic State
--------------------------------------
Published values from the literature were compared to the %BioGears Engine output to perform a quantitative validation of the blood chemistry resting physiology. Table&nbsp;1 shows the validation results.  The validation is specified with a color coding system, with green indicating a less than 10% error, yellow indicating a less than 30% error, and red indicating a greater than 30% error when comparing the %BioGears output to the published values. All references are noted in the table.

<center>
*Table 1. Results of the resting physiology validation of the %BioGears Blood Chemistry System.*
</center>

@insert doc/validation/BloodChemistryValidationTable.md

Overall, the %BioGears Engine meets validation, with 37 of the 44 validation parameters having less than a 10% error when compared to published values.  An additional two parameters have less than a 30% error. The parameters with greater than 30% error are all related to carbon dioxide saturation and binding in the blood stream.  A detailed explanation of how these values are calculated can be found @ref bloodchemistry-approach "above", with additional information in @ref tissue-diffusion "Tissue". This binding calculation will be reviewed to correct errors in the future.

Validation - Conditions and Actions
-----------------------
## Hemorrhage
The Blood Chemistry system does not have any conditions or actions directly embedded, but the outputs of Blood Chemistry are dependent on the actions and conditions of other systems. An example is the Hemorrhage insult, an action of the [Cardiovascular](@ref CardiovascularMethodology) System. Figure&nbsp;2 shows the effects of hemorrhage and fluid resuscitation with saline on the total blood volume and the total hemoglobin for the patient.

<center>
<img src="./plots/BloodChemistry/Volume.jpg" width="1100">
<img src="./plots/BloodChemistry/Hemoglobin.jpg" width="1100">
<img src="./plots/BloodChemistry/BloodChemistryLegend.jpg" width="500">
</center>
<center>
*Figure 4. The total blood volume decreases with the hemorrhage, then increases as saline is infused. The total hemoglobin also decreases with the hemorrhage but remains reduced after the hemorrhage is stopped.*
</center><br>

@anchor bloodchemistry-validation-assessments
Validation - Assessments
------------------------
The two assessments in the blood chemistry system provide the outputs associated with a Chem-14 blood panel and a CBC in a single output request from the engine. These assessments were validated with published data, as shown in Table&nbsp;2 and Table&nbsp;3. As with the resting physiology, the references for all values are provided and the results are color coded. 

<center>
*Table 2. Results of the metabolic panel as compiled during the healthy resting physiology.*
</center>

@insert doc/validation/CompleteMetabolicPanelValidationTable.md

<center>
*Table 3. Results of the complete blood count as compiled during the healthy resting physiology.*
</center>

@insert doc/validation/CompleteBloodCountValidationTable.md

All of the measures in the two Blood Chemistry assessments meet validation. 

@anchor bloodchemistry-fourCompartment
Four Compartment Test
------------------------
As mentioned above, the Blood Chemistry system serves the primary purpose of storing and relaying information between the other %BioGears systems. This heavy reliance on other systems can make it difficult to test the funcitonality of blood gas balance. To this end, the scalability of %BioGears was leveraged to create a simpler system comprised of only four compartments: Pulmonary, to represent the capillaries in the lungs where oxygen and carbon dioxide exchange occurs; Arteries, representing all of the oxygen-rich vasculature running to the tissues; Capillaries, representing the location of oxygen removal from the vasculature; and Veins, representing the section of the vasulature containing oxygen-poor blood. A diagram of this simplified system can be seen in Figure 4 below.

<img src="./images/BloodChemistry/FourCompartment.PNG" width="600">
<center>
*Figure 4. The simplified Four Compartment test uses only Pulmonary, Arteries, Capillaries, and Veins to hone in on the functionality in the Blood Chemistry system.*
</center><br>

Pressures, volumes, and substances were initialized to good values (see Table 4 below), and then the simplified Four Compartment system was run as %BioGears would run the normal, complete model using the Preprocess, Process, Postprocess paradigm. In the Preprocess step, oxygen is removed and carbon dioxide is added in the Capillaries to simulate metabolism while oxygen is added and carbon dioxide is removed in the Pulmonary compartment to represent respiration. If a tissue compartment was present, diffusion could also occur in this stage. In the Process step, circuit calculation and substance transport are done. Then, the Postprocess step moves the "Next" values to "Current". For more information about this paradigm, see @ref CircuitMethodology.

<center>
*Table 4. Initial values for the Four Compartment test. Variables with an asterisk indicate that the values were pulled from a %BioGears simulation run to a stable point.*
| Variable | Initial Value |
| :---------------- | :---------- |
| Veins Pressure | 4 mmHg @cite Leeuwen2015laboratory |
| Pulmonary Pressure | 10 mmHg @cite Edwards2009pocket |
| Arteries Pressure | 90 mmHg @cite Leeuwen2015laboratory |
| Capillaries Pressure | 30 mmHg |
| Blood Flow | 5600 mL/min @cite guyton2006medical |
| Total Blood Volume | 5 L @cite guyton2006medical |
| Arteries Volume | 23% of Total Blood Volume @cite valentin2002icrp @cite hudsmuth2005heartvolume @cite Edwards2009pocket |
| Veins Volume | 60% of Total Blood Volume @cite valentin2002icrp @cite hudsmuth2005heartvolume @cite Edwards2009pocket @cite effros1967vascular |
| Pulmonary Volume | 12% of Total Blood Volume @cite valentin2002icrp @cite hudsmuth2005heartvolume @cite Edwards2009pocket |
| Capillaries Volume | 5% of Total Blood Volume @cite valentin2002icrp @cite hudsmuth2005heartvolume @cite Edwards2009pocket |
| Albumin Concentration | 45 g/L @cite valtin1995renal |
| Hematocrit | .45 @cite guyton2006medical @cite valtin1995renal |
| Body Temperature | 37 C @cite herman2007physics |
| [SID] | 40.5 mmol/L @cite kellum2009stewart |
| Phosphate | 1.1 mmol/L @cite Leeuwen2015laboratory |
| *Pulmonary and Arteries CO2 Saturation | .0282123  |
| *Pulmonary and Arteries O2 Saturation | .974759  |
| *Pulmonary and Arteries CO2 Concentration | 1.30049 mmol/L  |
| *Pulmonary and Arteries O2 Concentration | .129065 mmol/L  |
| *Pulmonary and Arteries HCO3 Concentration | 25.9377 mmol/L  |
| *Pulmonary and Arteries pH | 7.39982  |
| *Capillaries and Veins CO2 Saturation | .162382  |
| *Capillaries and Veins O2 Saturation | .789701  |
| *Capillaries and Veins CO2 Concentration | 1.40557 mmol/L  |
| *Capillaries and Veins O2 Concentration | .0560892 mmol/L  |
| *Capillaries and Veins HCO3 Concentration | 26.1182 mmol/L  |
| *Capillaries and Veins pH | 7.36909 @cite valentin2002icrp |
| Oxygen Exchange Rate | 5954.2 ug/S @cite guyton2006medical |
| Carbon Dioxide Exchange Rate | 6140.4 ug/S @cite guyton2006medical |
</center>

Because the Four Compartment test is initialized to good values, and because of the design of the %BioGears engine and Blood Chemistry system, output values should be within physiological ranges. Indeed, the outputs shown below reflect the proper function of the Blood Chemistry system.
<center>
*Table 5. Stable results of the Four Compartment test.*
Variable			|	Four Compartment Test Ending Value	|	Valid Value	|
------------------------	------------------------	------------------------	|	------------------------	|	------------------------	|
Arteries HCO3			|	0.0158445	|<span class="success">	[0.134, 0.159] g/dL @cite valtin1995renal	</span>|
Arteries CO2			|	41.9369	|<span class="success">	40 mmHg @cite guyton2006medical	</span>|
Arteries O2			|	108.012	|<span class="success">	95 mmHg @cite guyton2006medical	</span>|
Arteries pH			|	7.39498	|<span class="success">	7.2 @cite valentin2002icrp [7.35, 7.4] @cite guyton2006medical	</span>|
Capillaries HCO3			|	0.0159359	|<span class="success">	[0.134, 0.159] @cite valtin1995renal	</span>|
Capillaries pH			|	7.36949	|<span class="success">	7.2 @cite valentin2002icrp [7.35, 7.4] @cite guyton2006medical	</span>|
Pulmonary HCO3			|	0.0158445	|<span class="success">	[0.134, 0.159] @cite valtin1995renal	</span>|
Pulmonary CO2			|	41.9369	|<span class="success">	40 mmHg @cite guyton2006medical	</span>|
Pulmonary O2			|	108.012	|<span class="success">	104 mmHg @cite guyton2006medical	</span>|
Pulmonary pH			|	7.39498	|<span class="success">	7.2 @cite valentin2002icrp [7.35, 7.4] @cite guyton2006medical	</span>|
Veins HCO3			|	0.0159359	|<span class="success">	[0.134, 0.159] @cite valtin1995renal	</span>|
Veins CO2			|	44.7292	|<span class="success">	45 mmHg @cite guyton2006medical	</span>|
Veins O2			|	39.2399	|<span class="success">	40 mmHg @cite guyton2006medical	</span>|
Veins pH			|	7.36949	|<span class="success">	7.2 @cite valentin2002icrp [7.35, 7.4] @cite guyton2006medical	</span>|
Total Hemoglobin			|	765.0004356	|<span class="success">	[661.5, 835.2] g @cite guyton2006medical @cite onofrio1995sim	</span>|
</center>

@anchor bloodchemistry-conclusions
Conclusions
-----------
The current values output from the Blood Chemistry System provide accurate results for a clinical overview of patient health. Blood gases and hemoglobin (multiple types) are circulated using the transport methodology discussed in the @ref CircuitMethodology and the @ref TissueMethodology. The Blood Chemistry System is a powerful tool for validation of multiple systems within the %BioGears Engine. The system is able to illustrate the dependencies each system has on another, i.e., if the %Respiratory System is unable to provide oxygen, a shortage of oxygen in the blood stream results, reducing both the arterial oxygen partial pressure and the oxygen saturation as displayed by the Blood Chemistry system. Future work will focus on adding additional key substances to provide a more complete assessment of a patient's overall health.

@anchor bloodchemistry-future
Future Work
===========

Coming Soon
-----------
Planned improvements to other systems will have an indirect effect on Blood Chemistry. For example, a planned improvement to the vascular tone model in the [Cardiovascular](@ref CardiovascularMethodology) system will reduce the resistance of the blood vessels supplying the myocardium, simulating dilation. This improvement to homeostatic feedback will make the %BioGears heart muscle more robust to an oxygen deficit, enabling a physiologically accurate threshold for the myocardium hypoxia event. Additionally, major work has already been completed toward a more accurate and complete handling of all substances within the %BioGears body. This work will allow for a direct computation of the ion concentrations and the strong ion difference, enabling mechanistic simulation of acid-base disturbances and compensation.

Recommended Improvements
------------------------
The exclusion of blood cellular elements prohibits the implementation of a variety of mechanistic models of physiologic processes, including coagulation, infectious disease, and hematopoiesis to name a few. 
- Inclusion of RBC, platelets, lymphocytes, monocytes, neutrophils, eosinophils, and basophils as blood constituents.
- Adaptive total hemoglobin (for example, increasing hemoglobin with exposure to low-oxygen pressure environments)
- Myoglobin substance models
	- Including interactions with oxygen and carbon monoxide

@anchor bloodchemistry-appendices
Appendices
==========

@anchor bloodchemistry-symbols
Acronyms and Symbols
--------
<center>
*Table 6. List of acronyms and symbols.*
| Symbol or Acronym | Description |
| :---------------- | :---------- |
| *A<sup>-</sup>* | The sum of all weak anions that are the conjugate base of all non-volatile weak acids |
| *Alb* | Albumin |
| ALP | Alkaline Phosphatase |
| ALT | Alanine Aminotransferase |
| AST | Aspartate Aminotransferase |
| *Ca<sup>2+</sup>* | Calcium (strong cation) |
| CBC | Complete Blood Count |
| *Cl<sup>-</sup>* | Chloride (strong anion) |
| *CO<sub>2</sub>* | Carbon dioxide gas |
| *HCO<sub>3</sub><sup>-</sup>* | Bicarbonate  |
| *Hct* | Hematocrit  |
| *Hgb* | Hemoglobin  |
| *I<sup>n+</sup>* | Represents a lumped set of unnamed cations (e.g. Mg*^{2+}* + Cu*^{2+}* + Fe*^{3+}* + ... ) |
| *I^<sup>n-</sup>* | Represents a lumped set of unnamed anions (e.g. sulfate + beta-hydroxybutyrate + ... )  |
| *K<sup>+</sup>* | Potassium (strong cation)  |
| *Ket<sup>-</sup>* | Ketone body (i.e. acetoacetate, beta-hydroxybutyrate, and acetone)  |
| *La<sup>-</sup>* | Lactate (strong anion)  |
| mmHg | Millimeters Mercury |
| *Na<sup>+</sup>* | Sodium (string cation)  |
| *O<sub>2</sub>* | Oxygen gas  |
| *OH<sup>-</sup>* | Hydroxide ion  |
| *P<sub>i</sub>* | Phosphate  |
| *P<sub>CO<sub>2</sub></sub>* | Partial pressure of carbon dioxide  |
| *P<sub>O<sub>2</sub></sub>* | Partial pressure of oxygen  |
| *pH* | Negative log of the hydrogen ion concentration  |
| *S<sub>CO<sub>2</sub></sub>* | Fraction of hemoglobin binding sites with a bound carbon dioxide molecule  |
| *S<sub>O<sub>2</sub></sub>* | Fraction of hemoglobin binding sites with a bound oxygen molecule  |
| *SID* | Strong Ion Difference,  defined as [SID] = [strong cations] - [strong anions]  |
| *T* | Temperature  |
| *T<sub>CO<sub>2</sub></sub>* | Total carbon dioxide  |
| *T<sub>O<sub>2</sub></sub>* | Total oxygen  |
| <i>&alpha;<sub>CO<sub>2</sub></sub></i> | The plasma *CO_2* solubility coefficient  |
| <i>&alpha;<sub>O<sub>2</sub></sub></i> | The plasma *O_2* solubility coefficient  |
</center>

Data Model Implementation
-------------------------
@ref BloodChemistrySystemTable "BloodChemistry"
Substance Transport Methodology {#SubstanceTransportMethodology}
=====================

@anchor substance-overview
Overview
========

Abstract
--------

The %BioGears&reg; transporter was created to extract the methodology of substance transport from the individual systems. This creates a generic set of transport functions that can be used by any existing or future system by performing calculations on the circuits defined in the Common Data Model (CDM).
At each time step, the mass, concentration, substance volume, volume fraction, and/or partial pressure are calculated based on the assumption that substances travel with the fluid flow.
@anchor substance-intro
Introduction
------------

%Substances are transported through the body within the fluids, such as air in the %Respiratory System and blood in the %Cardiovascular System. To assess the substance levels of a patient, the substance must circulate through the system and distribute into the different nodes. The mass and concentration of a fluid must be calculated for substances in a liquid, and the volume of an individual substance (gas) and the volume fraction must be calculated for a gaseous fluid. After these basic calculations have been completed, calculations for partial pressures (gases), saturation (oxygen and carbon dioxide), and system totals (mass, concentration, volumes) are completed.

The %BioGears modeling approach takes the human body and conceptually divides it into various fluid compartments that represents a real division in terms of how portions of the body's water, solutes, and suspended elements are segregated @cite rhoades2012medical.  Compartments can be further discretized into smaller sub-compartments with a hierarchical relationship as you drill into various systems. In %BioGears, compartments can be defined to encapsulate circuit nodes that allow easy organization, access, and synchronization of all system parts.

Links represent connections between compartments with a directional flow component defining a volume change each time-step.  In %BioGears, links can be assigned a path that provides the instantaneous flow value.  Figure 1 shows the base transport elements definitions in an example graph.

<center><img src="./images/SubstanceTransporter/Components.png" width="400"></center>
<center>
<i>Figure 1. This is a conceptual example graph that describes the lowest level elements used to define properties used by the %BioGears Transporter.  Links provide flow between compartments that store both fluid and substance quantity information.</i>
</center><br>
@anchor substance-design
System Design
=============

Background and Scope
--------------------

### Requirements

Given the overall goals of %BioGears, we set out to create a generic
and reusable substance transporter.  Some high-level requirements include:

-   Generic - All systems should be able to use the same basic transporter engine.  This allows rapid development, and makes engine
    outputs much easier to validate and verify.

-   Computational Speed - %BioGears is required to maintain a transient
    full-body solution faster than real time on typical personal
    computers.

-   Modular - Using the same basis for design and construction will aid
    in keeping the system decoupled.

-   Extensible - We must take future growth into consideration and allow
    users and developers the proper tools and building blocks on which
    to add new functionality.

-   Dynamic - Feedback mechanisms are required for each system.  It is
    beneficial to be able to dynamically change, add, or remove
    compartments and links.

-   Conservation - We must uphold sound scientific principles and conserve mass.

-   Common Data Model - The entire solution must reside within and
    effectively use the @ref CDM.

-   Fluid types - %BioGears will include liquid and gas systems.  It is beneficial to use the same solver for both types.

-   Bifurcations - Each compartment can have an unlimited number of links providing flow/substances both in (up stream) and out (down stream).

-   Large flows - The transporter must be able to handle any size flows, including instances where significantly more volume is moved than exists in a given compartment for a given time-step.

### Approach

The Transporter is implemented generically for both liquid and gas systems using the same high-level definitions for substance properties, which are @cite McNaught2014Compendium :
- <b>Extensive property:</b> additive for independent, non-interacting subsystems - proportional to the amount of material in the system
	- Property changes with amount change
	- Pouring some out will change the value
	- Examples: Mass, Volume, Length, Amount
- <b>Intensive property:</b> a bulk property, meaning that it is a physical property of a system that does not depend on the system size or the amount of material in the system
	- Property doesn't change with amount change
	- Defines and identifies substances
	- Pouring some out will not change the value
	- Examples: Concentration, VolumeFraction, and Temperature

The Transporter assumes that fluid movement (i.e. convection) has already taken place - generally calculated and updated inside the system by the circuit solver (@ref CircuitMethodology).  Once the convective fluid movement properties of compartment volume and link flow are updated for the current time-step being analyzed, the extensive and intensive substance values at each compartment can be determined by using the previous time-step state.  Table 1 shows the parameters needed to calculate the advective transport by bulk flow.
	
<center><br>
*Table 1. The variable definitions as well as the mapped property used to calculate advective transport.*
</center>

| Parameter | Definition | Liquid Variable | Gas Variable |
| --- | --- | --- | --- |
| *I<sub>C</sub>* | Current compartment intensive property | Concentration | Volume Fraction |
| *I<sub>SC</sub>* | Source compartment intensive property | Concentration | Volume Fraction |
| *f<sub>IL</sub>* | Input link flow | Volumetric flow | Volumetric flow | 
| *f<sub>OL</sub>* | Output link flow | Volumetric flow | Volumetric flow | 
| *t* | time-step | Time | Time |
| *V<sub>C</sub>* | Current compartment volume | Volume | Volume |
| *E<sub>o,C</sub>* | Current compartment previous time-step extensive property | Mass | Substance volume |

The instantaneous substance quantity values can be determined in each compartment of a graph by doing a mass balance calculation using Equation 1, where *m* is the mass on the current compartment and both *m<sub>in</sub>* and *m<sub>out</sub>* are provided by links to any number of other compartments.

\f[\sum\limits_{}^{} {{m_{in}} - \sum\limits_{}^{} {{m_{out}} = \Delta m} } \f]

<center>
*Equation 1.*
</center><br> 

Equation 1 can be further broken out using the parameters in Table 1 to give Equation 2. 

\f[\sum\limits_{}^{} {{I_{SC}}{f_{IL}}t - \sum\limits_{}^{} {{I_C}{f_{OL}}t = {I_C}{V_C} - {E_{o,C}}} } \f]

<center>
*Equation 2.*
</center><br> 

Rearranging Equation 2 gives Equation 3.

\f[{I_C}{V_C} - \sum\limits_{}^{} {{I_{SC}}{f_{IL}}t + \sum\limits_{}^{} {{I_C}{f_{OL}}t = {E_{o,C}}} } \f]

<center>
*Equation 3.*
</center><br> 

By simultaneously combining Equation 3 for all compartments in a graph, the linear equations can be written in the form of Equation 4 to solve for the new intensive properties throughout.  *A* is the matrix of constants, *x* is the vector of all intensive properties, and *b* is the right side vector of known previous time-step extensive properties.

\f[Ax = b\f]

<center>
*Equation 4.*
</center><br>
@anchor substance-data
Data Flow
---------

The current implementation of the Circuit Transporter is set up to operate on the circuit nodes using the Preprocess, Process, and Post Process methodology of the other %BioGears systems.

### Preprocess

The Circuit Transporter has no functionality in Preprocess.

### Process

The generic substance methodology developed for the %BioGears Engine is used to solve for the mass, concentration, substance volume, and volume fraction in each compartment each time-step.  The steps used by the transporter to solve a graph using Equation 4 in a given time-step are:

1. Loop over compartments to populate the *A* matrix (one row per compartment) - this is the same for all substances
	1. Handle infinite volume (often the environment) by setting intensive property constant
	2. Handle no volume by using an approximate zero (1e-20) value to prevent a singular matrix
	3. Handle source link flows (out of the compartment)
	4. Handle target link flows (into the compartment)
2. Loop over all substances
	1. Populate *b* vector with previous time-step masses
	2. Solve for *x* vector intensive properties
	3. Parse intensive properties and calculate new extensive properties

Each individual system uses the Calculate Substance Transport function to generically complete transport during the Process step.

### Post Process

The Post Process step moves everything calculated in Process from the next time-step calculation to the current time-step calculation. The Circuit Transporter has no specific Post Process functionality. Each system post-processes their circuit and moves the mass, concentration, substance volume, volume fraction, and partial pressure from "next" to "current".
@anchor substance-results
Results and Conclusions
======================= 

## %Verification

There are three successful unit tests that were created specifically to ensure the transporter is giving results as expected:
- Liquid Transport Test
- Gas Transport Test
- Large Flow Transport Test

## Conclusion

This generic methodology has proven to be a satisfactory algorithm for transport within the human body. The methodology is robust and successfully navigates all of the lumped parameter models in %BioGears. The algorithm is able to run faster than real time, allowing the %BioGears Engine to run faster than real time. Mass is conserved throughout the calculation.
@anchor substance-appendices
Appendices
==========

## Data Model Implementation

SECompartmentTransportGraph

SESubstanceTransportAmount
	
SESubstanceTransportEdge
	
SESubstanceTransporter	

SESubstanceTransportGraph	

SESubstanceTransportVertex

## Acronyms

CDM - Common Data Model%Respiratory Methodology {#RespiratoryMethodology}
========================

@anchor respiratory-overview
Overview
========

@anchor respiratory-abstract
Abstract
--------

The %Respiratory System supplies oxygen and removes waste carbon dioxide from the body through a combination of ventilation and gas exchange across the blood-gas barrier (pulmonary capillary-alveoli interface). The %BioGears %Respiratory System is designed to model the ventilatory behavior (both positive- and negative-pressure) of the patient %Respiratory System using electrical analogue lumped parameter models. The %Respiratory Model employs realistic pressure source signal and chemical stimuli feedback mechanisms as drivers for spontaneous ventilation. The model handles several patient conditions, including tension pneumothorax and airway obstruction. The majority of the lung values investigated for the overall model matched the validation data found in publications. Patient conditions also showed strong agreement with clinically significant output parameters, i.e., respiration rate, oxygen saturation, heart rate, and blood pressure.

@anchor respiratory-intro
## Introduction

@anchor respiratory-physiology
### %Respiratory Physiology

The human %Respiratory System consists of the upper airways (region above the cricoid cartilage), the lower airways, the lungs, and the respiratory muscles. The lower airways begin at the trachea and extend to the bronchi, bronchioles, and the alveoli. At the carina, the trachea divides into two mainstem bronchi, the right and left. The bronchi bifurcate into smaller bronchioles that continue branching for up to 23 generations, forming the tracheobronchial tree that terminates with the alveoli. Alveolar ducts and alveolar sacs are the operating units of the lungs where gas exchange occurs with the pulmonary capillaries. The first several generations of airways, where no gas exchange occurs, constitute the anatomic dead space and are referred to as the conducting zone. In contrast, alveolar ducts and sacs that terminate the tracheobronchial tree are referred to as the respiration zone.

@image html Respiratory_Figure01_SystemDiagram.png
<center>
<i>Figure 1. The %Respiratory System consists of the upper and lower airways. The diaphragm acts as a respiratory muscle taking part in the ventilatory driver mechanics. The trachea branches into the right and left bronchi, each of which further bifurcates into multiple generations of smaller bronchioles. These bronchioles form the tracheobronchial tree, which terminates at the alveoli. @cite LadyofHats2014Respiratory </i>
</center><br>

The alveolar-capillary gas exchange is facilitated by the ventilation process, which is driven by the intercostal muscles, the diaphragm, and the chest wall recoil. These mechanisms work in tandem to actively drive fresh air into the lungs and passively remove gases from the lungs. Attached to the chest wall is a thin layer of membrane (pleura) that folds back onto itself, forming two layers, known as the visceral and parietal pleurals. The pleural cavity is filled with fluid. The pressure in this space, known as the intrapleural pressure, is normally slightly below the atmospheric pressure. Even when no inspiratory muscles are contracting, the mechanical interaction between the lung and the chest wall pulls the two pleural membranes apart, resulting in a slightly decreased intrapleural pressure (-3 cm H<SUB>2</SUB>O to -5 cm H<SUB>2</SUB>O) @cite Levitzky2013pulmonary .

@anchor respiratory-math
### Mathematical Model

Mathematical modeling of the respiratory physiology dates back to the work published by Gray in 1945 @cite gray1951pulmonary . Gray provided the first mathematical description for the chemical control of pulmonary ventilation. Later, Gordins et al. developed the first dynamic model of the respiratory system in 1954 @cite grodins1954respiratory . Several mathematical models followed after that, including the work by Guyton and collaborators in 1965 @cite milhorn1965mathematical and others ( @cite grodins1967mathematical , @cite khoo1982factors , @cite saunders1980breathing , @cite ursino2004interaction ). Many of the published models describe a specific aspect of the respiratory physiology in considerable detail. To name a few, Lorandi et al. @cite lorandi2003parametric employed a mathematical model to describe the mechanical properties of the lungs, Murray et al. @cite murray1977techniques described the gas exchange properties of the lungs, Wiberg et al. @cite wiberg1979modeling and Bache et al. @cite bache1981time described the effect of higher levels of CO<SUB>2</SUB> or anesthetic gases on breathing, and Whipp et al. @cite whipp1995obligatory developed a mathematical model to describe the respiratory anaerobiosis in skeletal muscle.

Many mathematical models of mechanical ventilation employ the lumped parameter model that represents the entire ventilation process with a small number of unknowns. The simplest lumped parameter model of mechanical ventilation assumes the conducting zone can be identified with a pipe that connects a collection of alveoli to the atmosphere and exerts pneumatic resistance to the flow. This type of model can be solved with a low computational cost, which reduces runtime. For whole body models/simulations, this is an important requirement. The disadvantage of lumped parameter models can lie in the large number of parameters that can result from required circuit parameters. It is important to identify the key features and behaviors of any model to intelligently reduce the number of required parameters.

The most important parameters in the lumped parameter model of mechanical
ventilation correspond to the elastic behavior of the lung and the flow
resistance of the airways. The thoracic cage and the lung tissue exhibit an
elastic behavior that can be represented with a single compliance or multiple
compliances. The compliance *C* is calculated by taking the ratio of the volume <i>&delta;V</i>
and the pressure <i>&delta;P</i> variations as:

\f[C=\frac{\delta V}{\delta P} \f] 
<center>
<i>Equation 1.</i>
</center><br> 

As a first-order approximation, the volume of the functional unit can be
approximated as:

\f[V(P+\delta P)=V(P)+C\delta P\f] 
<center>
<i>Equation 2.</i>
</center><br> 

In the %Respiratory System, the main source of flow resistance arises from the
flow of air through the branches in the conducting zone. Mathematical models
using the lumped parameter model select functional units for these regions and
designate the variable <i>R</i> for pneumatic flow resistance. The pressure drop <i>&Delta;P</i> across
the respiratory tree can thus be calculated by using Ohm's law analogue as

\f[\Delta P=RQ\f] 
<center>
<i>Equation 3.</i>
</center><br> 

where <i>Q</i> is the volumetric flow rate. The above relation assumes the flow
is laminar and the gas is incompressible. For laminar, viscous, and
incompressible flow, the Hagen-Poiseuille equation relates the pressure drop <i>&Delta;P</i>
in a fluid flowing through a cylindrical pipe of length <i>l</i> and radius <i>r</i> as

\f[\Delta P=\frac{8\mu l}{\pi r^{4} } Q\f] 
<center>
<i>Equation 4.</i>
</center><br> 

where <i>&Mu;</i> is the dynamic viscosity. By defining the flow resistance <i>R</i> as

\f[R=\frac{8\mu l}{\pi r^{4} } \f] 
<center>
<i>Equation 5.</i>
</center><br> 

a relation analogous to Ohm's law can be derived.

@anchor respiratory-systemdesign
System Design
=============

@anchor respiratory-background
Background and Scope
--------------------

### Existing

The %BioGears %Respiratory Model has its roots in the mathematical model of
Yashuri Fukui and N. Ty Smith @cite FukuiSmith1981hybrid . The researchers
developed a lumped parameter mathematical model to describe the uptake and
distribution of halothane. Their %Respiratory Model consisted of two pulmonary
compartments corresponding to the dead space and the alveoli @cite FukuiSmith1981hybrid. 
The %Respiratory Model in the %BioGears Engine is an
extension of the work by Fukui and Smith. This model was developed and released
by Advanced Simulation Corporation as part of the simulator, Body Simulation for
Anesthesia&trade;. This later formed the backbone of the HumanSim&trade; physiology engine,
which continues to provide realistic physiology for several serious training
games in the HumanSim product line, including HumanSim: Combat Medic and
HumanSim: Sedation and Airway @cite Clipp2012Humansim . The HumanSim
physiology engine is the starting point for the %BioGears Physiology Engine. The
basic elements of the %Respiratory System remain the same as the HumanSim
physiology engine; however, the respiratory circuit has been further developed
to allow realistic mechanical responses to pathological conditions.

### Approach

The current version of the %BioGears %Respiratory Model represents the two lungs
and associated airways as five major functional units, or compartments, that are
designated as the carina, right and left anatomic dead space, and right and
left alveoli. In the model, the carina compartment represents the anatomical region
from the airway at the trachea. The right and left anatomic dead
space compartments represent the bronchi and their branching bronchioles that
are part of the conducting airways below the carina. The right and left alveoli
compartments correspond to the collection of alveoli where gas exchange occurs
between the airways and the %Cardiovascular System. The right and left chest wall
compartments represent the right and left sides of the thoracic wall. The new model additionally accounts for the pleural
cavity through circuit elements that allow flow into the pleural space in the
event of respiratory insults that involve gas leak either from the alveoli or
the thoracic wall. To account for flow through the esophagus, an incidence that
may occur during mechanical ventilation (positive-pressure ventilation), the
model provides subordinate compartments representing the esophageal passage and
the stomach. The model also consists of a pressure signal generator representing
the respiratory muscle pressure source driver.

@anchor respiratory-dataflow
Data Flow
---------

The %Respiratory System determines its state at every time step through a three
step process: Preprocess, Process, and Postprocess. In general, Preprocess
determines the circuit element values based on feedback mechanisms and engine
settings/actions. Process uses the generic circuit calculator to compute the
entire state of the circuit and fill in all pertinent values. Postprocess is
used to advance time.

### Initialize

At the beginning of a simulation, patient parameters are used to modify the muscle (pressure source) driver functionality to achieve the specified values at the end of the resting stabilization period - see the @ref respiratory-variability "Patient Variability" section for more details. After resting stabilization is achieved, any user-selected conditions are implemented to reach a new homeostatic point - see the @ref respiratory-conditions "Conditions" section for more details.

### Preprocess

#### Update Pleural Compliance

The compliances of the left and right pleural space are modified as a function of volume.

#### Process Actions

Process Actions deals with modification of the respiratory parameters and
circuit properties due to actions (insults or interventions) specified by the
user.

#### %Respiratory Driver

The respiratory muscle pressure source that drives spontaneous ventilation is
calculated based on chemical stimuli feedback control mechanism.

### Process

The generic circuit methodology developed for the %BioGears Engine is used to
solve for the pressure, flow, and volume at each node or path. For more details
on this, see the @ref CircuitMethodology. Substance volumes and volume
fractions (concentrations) are also calculated during this step. See the @ref SubstanceTransportMethodology for more details.

The Calculate Vital Signs function uses the circuit pressure, flow, and volume
to calculate important system-level quantities for the current time step.

### Postprocess

The Postprocess step moves values calculated in the Process step from the next
time step calculation to the current time step calculation. This allows all
other systems access to the information when completing their Preprocess
analysis during the next time step.

### Assessments

Assessments are called outside of the system to allow compiling of information from multiple systems.

<center><img src="./images/Respiratory/Respiratory_Figure02.png" width="900"></center>
<center> 
<i>Figure 2. The data flow for the %Respiratory System consists of Preprocess,
Process, and Postprocess. Preprocess sets the circuit element values based on
feedback mechanisms and engine actions. Process utilizes generic circuit
methodology to solve the respiratory circuit for pressures, volumes, and flows
along paths and on nodes. Process also updates system level quantities by
calculating the vital signs. Postprocess updates these quantities to the next
time step and then begins advancing time.</i>
</center><br>

@anchor respiratory-features
Features and Capabilities
-------------------------

### %Respiratory Circuit

The %BioGears %Respiratory System designates a set of functional elements, or
compartments, to model mechanical ventilation. The functional elements are
represented by an electric analogue circuit comprised of resistors, capacitors,
switches, diodes, and power sources. The latter represents the driving pressure
from the respiratory muscles. The resistors and capacitors represent the
resistance to flow through the airways and the elastic nature of the airways,
alveoli, and thoracic walls.

The equivalent of an electric switch is used to transition between different
ventilation conditions or flow pathways. For example, the electric analogue switch
accounts for incidents that permit flow through the esophageal tract while
prohibiting flow through the trachea. Unidirectional flow in the respiratory
system is handled through electric analogue diodes that allow flow in one
direction, preventing flow in the opposite direction. Such functional elements
are employed to represent insults that allow unidirectional gas flow into the
pleural cavity through an opening at the alveoli or the thoracic wall.

<img src="./images/Respiratory/Respiratory_Figure03.png" width="650">
<center>
<i>Figure 3. Circuit diagram of the %Respiratory System. The diagram depicts a
closed circuit of the seven major compartments (trachea, right and left anatomic
dead space, right and left alveoli, and right and left chest wall) and the four
subordinate compartments (right and left pleural, esophagus, and stomach). Each
of the alveoli and anatomic dead space compartments are represented by a
combination of resistor and capacitor. The remaining compartments are
represented by resistors or capacitors. The circuit also depicts the muscle
pressure source that serves as the driver for the %Respiratory System. Unless changed
for insults and interventions, the subordinate compartments have "infinitely"
large resistors and behave as open electrical switches.</i>
</center><br>

In the circuit model, the carina and right and left anatomic dead spaces are
represented by resistors to account for pneumatic resistance that impedes flow
of gas across the conducting zones. A capacitor representing the compliance of a
compartment is added to the nodes designated as anatomic dead spaces to account
for the elastic behavior of the lower airways. Each of the right and left
alveoli compartments are represented by a combination of resistors and
capacitors (compliances) to account for the elastic behavior of the alveoli. The
right and left chest wall compartments are represented by variable compliance
that allows flexibility to mechanical insults. Based on the electrical circuit
analogue, the model predicts the dynamic properties of the %Respiratory System.
Figure 3 depicts the network of respiratory circuit elements and their
interconnections.

The %BioGears respiratory circuit employs circuit nodes and paths to represent
physiological state variables belonging to the %Respiratory System's functional
units. In this representation, the pressures across the compartmental units are
designated to the nodes, while all other variables (flow, volume, hydraulic
resistances, and compliances) are assigned to the paths on the circuit. At any
instant of time, the flow <i>Q</i> on a path across a resistor <i>R</i> can be calculated
using the pressure difference <i>&Delta;P</i> between the nodes across the path as <i>Q=&Delta;P/R</i>.
Similarly, the volume change <i>&Delta;V</i> of a respiratory element with compliance <i>C</i> can
be calculated based on the pressure difference <i>&Delta;P</i> between the nodes connected by
the path as <i>&Delta;V=C&Delta;P</i>. The time evolution of the pressures at each node in the
circuit is solved using the %Circuit Solver as described in the @ref
CircuitMethodology.

@anchor respiratory-variability
### Patient Variability

Several patient parameters are set/calculated outside of the %Respiratory System at the beginning of a simulation (See @ref PatientMethodology).  The patient parameters that are used as inputs to the %Respiratory System are:
- Respiration Rate Baseline: used to set the driver frequency
- Functional Residual Capacity: used to set the driver default pressure
- Total Lung Capacity: used to set the driver maximum allowable pressure
- Right Lung Ratio: used in the scaling equation for inspiratory-expiratory ratio for conditions
- Basal Metabolic Rate: used for metabolic effects
- Vital Capacity: used to determine the tidal volume plateau in the driver piecewise target alveolar ventilation function

The Pulmonary Function Test also pulls all lung volumes and capacities, and conscious respiration uses several of the initial capacity values to calculate the driver pressure needed.

Several patient parameters are updated at the end of each stabilization segment (Resting, Conditions, and Feedback).  This allows the simulation to reach new homeostatic points that take into account the whole-body state based on both internal and external factors.  The patient parameters that are reset by the %Respiratory System are:
- Respiration Rate Baseline: from Respiration Rate system data value
- Tidal Volume Baseline: from Tidal Volume system data value
- Functional Residual Capacity: from calculated instantaneous value
- Vital Capacity: calculated as [TLC - RV]
- Expiratory Reserve Volume: calculated as [FRC - RV]
- Inspiratory Reserve Volume: calculated as [TLC - FRC - TV]
- Inspiratory Capacity: calculated as [TLC - FRC]

The patient Alveoli Surface Area is also modified when COPD and lobar pneumonia effects are applied.

@anchor respiratory-feedback
### Feedback

#### Driver Pressure Source

The %BioGears %Respiratory System interacts with other systems in the engine to
receive feedback and adjust spontaneous breathing for homeostasis. To
accurately model the respiratory response under various physiological and
pathological conditions, a robust %Respiratory Model that responds to mechanical
stresses and chemical stimuli is required. To this end, the %BioGears
%Respiratory System employs a time-dependent pressure source based on a chemical
feedback mechanism that mimics the respiratory response to blood gas levels as
sensed by the central and peripheral chemoreceptors. The pressure source
represents the muscle pressure source signal and serves as an input power source
to drive the inspiration and expiration phases of the breathing cycle. 

During inhalation, the driver pressure source is set to a negative value. The end of the exhalation cycle represents the initial
conditions of free breathing, where the alveolar pressure equals the atmospheric
pressure and no air flows
into the lungs. When the inspiratory muscles are not contracting, the mechanical
interaction between the lungs and the chest wall creates a sub-atmospheric intrapleural
pressure. The value of the
driver pressure in the model is selected to meet the unstressed condition at the
pleural node. In the case of mechanical ventilation, the anesthesia machine
controls the pressure at the airway node for positive pressure ventilation. More
details on positive pressure ventilation can be found in the @ref AnesthesiaMachineMethodology.

For a realistic muscle pressure source signal, the %BioGears %Respiratory System
adopted the mathematical model proposed by Albanese, et al. @cite albanese2016integrated , which is
based on clinical data. Accordingly, the time series of the respiratory muscle
pressure, <i>P<sub>mus</sub></i>, is given by,

\f[P_{mus} =\left\{\begin{array}{l} {-\frac{P_{max}}{I\cdot E}\cdot t^{2} + \frac{P_{max}\cdot T}{I\cdot E}\cdot t ,0<t\le I } \\ {\frac{P_{max}}{1-\exp(\frac{-E}{\tau})}\cdot \left(\exp(\frac{t-I}{\tau})-\exp(\frac{-E}{\tau})\right),I <t\le T } \end{array}\right. \f] 
<center>
<i>Equation 6.</i>
</center><br> 

Where I, E, and T are the inspiratory, expiratory, and total respiration times, respectively.  The value &tau; is a time constant for the expiration period and is estimated as E / 5.  The total breathing cycle time T is obtained from the inverse of the respiration rate determined by the [chemoreceptor model](@ref nervous-features-chemoreceptors).  I and E are calculated using the inspiratory:expiratory ratio from the previous time step, which is modified by irregular physiology like asthma and COPD.  The chemoreceptor model also updates the driver amplitude, P<sub>max</sub>.  The baseline value of P<sub>max</sub> for each virtual patient is determined during engine initialization by modifing the amplitude at the requested patient respiration rate until a stable tidal volume is obtained. 

Figure 6 depicts the time-dependent driver pressure source of the %BioGears %Respiratory System as obtained during simulation of the standard patient model of the engine (77 kg adult male) under normal physiological conditions. For comparison, the driver pressure is plotted with the alveolar, intrapleural, and transpulmonary pressures. The figure shows the pressures for several breathing cycles. As seen in Figure 6, the model driver pressure exhibits distinct waveforms during the inspiration and expiration phases. These patterns represent the active distension and passive relaxation behaviors of the inspiratory muscles. As a result of such input, the model distinguishes between the active inspiratory and passive expiratory phases of the breathing cycle. The time-dependent muscle pressure together with the atmospheric pressure and the compliances act in tandem to generate the pleural and alveolar pressure waveforms shown in the figure.

<center>
<table border="0">
<tr>
    <td><img src="./plots/Respiratory/Driver_Pressure.jpg" width="800"></td>    
</tr>
<tr>
    <td><img src="./plots/Respiratory/Pleural_and_Alveoli_Pressure.jpg" width="800"></td>
</tr>
<tr>
    <td><img src="./plots/Respiratory/Transpulmonary_Pressure.jpg" width="800"></td>
</tr>
</table>
</center>

<center>
<i>Figure 4. The driver pressure, or pressure source, that serves as an electrical analogue voltage source for the respiratory circuit is plotted along with the alveolar, intrapleural, and transpulmonary pressures.  The pressure source generates a sub-atmospheric intrapleural pressure that facilitates the inspiration and expiration phases of spontaneous breathing.</i>
</center><br>

#### Standard Lung Volumes and Capacities

There are a number of standard lung volumes and capacities that are measured
during different stages of normal and deep breathing cycles. The inspiratory
reserve volume (IRV), tidal volume (V<sub>T</sub>), expiratory reserve volume (ERV), and
residual volume (RV) correspond to the four standard lung volumes. The
inspiratory capacity (IC), forced residual capacity (FRC), vital capacity (V<sub>C</sub>),
and total lung capacity (TLC) are the four standard lung capacities that consist
of two or more standard lung volumes. These volumes and capacities are good
diagnostics for lung functionality, and the %BioGears %Respiratory Model reports
their values as outputs. As mentioned above, some of the parameters are obtained
from patient data as input variables. The TLC and FRC are two of these
parameters that are drawn from the patient data. For the standard patient in the
model, TLC and FRC are set to be 6.0 L and 2.4 L, respectively. Using these
parameters as inputs, the engine calculates the other standard lung volumes and
capacities as described below.

##### Residual volume (RV)

The residual volume is the volume of gas remaining in the lungs after maximal
exhalation. As mentioned above, the %BioGears %Respiratory Model approximates the residual volume
based on the patient weight (RV = 16.0 mL/kg) (@cite Corning2007pulmonary , 
@cite Rennie2013pulmonary ). 
For the standard patient in the model 
with 77 kg weight, the residual volume
RV=1.23 L. Typical values of RV vary in the literature. For example, for 70 kg
patients: RV=1.5 L @cite Levitzky2013pulmonary , 1.2 L @cite silverthorn2013human, 
and 1.682 L @cite stocks1995reference . The %BioGears
engine employs weight-based relation and the values used in the engine are close 
to those found in the literature @cite silverthorn2013human .

##### Expiratory Reserve Volume (ERV)

ERV is the maximum volume below the tidal volume that can be expired during
maximal forced expiration. ERV can be calculated as

\f[ERV=FRC-RV\f] 
<center>
<i>Equation 7.</i>
</center><br> 

In this equation, both FRC and RV are input values obtained from weight-based
relation. For the standard patient in the model (77 kg adult male), FRC=2.31 L, and
RV=1.23 L, thus ERV becomes ERV=1.08 L. Typically, reported values for 
ERV are around 1.1 L @cite guyton2006medical .

##### Tidal volume (V<sub>T</sub>)

Tidal volume corresponds to the volume of air inspired or expired in a single
breathing cycle during normal quiet breathing. For a healthy 70 kg adult, the
tidal volume is 540 ml per breath. The tidal volume can be calculated by
numerically integrating the volumetric flow rate of inspired air flowing through
the trachea. The %BioGears %Respiratory Model calculates the tidal volume by
taking the difference between the maximum and minimum total lung volumes during
each breathing cycle.

Figure 5 depicts the typical lung volume waveform for multiple breathing
cycles. The %BioGears %Respiratory Model outputs the value of V<sub>T</sub> for each
breathing cycle. Figure 5 presents the plot of the total lung volume and V<sub>T</sub> as a
function of time.

<center><img src="./plots/Respiratory/TidalVolume_from_TotalLungVolume.jpg" width="800"></center>

<center>
<i>Figure 5. This shows the relationship of the total lung volume with the tidal
volume. The tidal volume for each cycle is determined by taking the difference between the maximum and
minimum values of the total lung volume, and is therefore only updated at the end of each full cycle.</i>
</center><br>

##### Inspiratory Reserve Volume (IRV)

IRV is the additional volume, above the tidal volume, that can be inspired
during maximal forced inspiration. IRV can be calculated from total lung
capacity (TLC) using the relation

\f[IRV=TLC-FRC-V_{T} \f] 
<center>
<i>Equation 8.</i>
</center><br> 

Both TLC and FRC are weight-based inputs to the model, whereas V<sub>T</sub> is calculated
as described above. Using TLC=6.16 L (i.e., 80 mL/kg) and FRC=2.31 L (i.e., 30 mL/kg) of
the standard patient in the model, IRV becomes 3.307 L for an average V<sub>T</sub> of
543 mL, where the average V<sub>T</sub> is calculated by taking the time average of V<sub>T</sub>
described above. Using weight-based tidal volume of V<sub>T</sub> = 7 mL/kg @cite Levitzky2013pulmonary , IRV 
can be shown to be 3.31 L for 77 kg patient, in good agreement with the value from 
the model.

##### Vital Capacity (V<sub>C</sub>)

V<sub>C</sub>  is the volume of air that can be expired after maximal inspiration. V<sub>C</sub> can be
calculated as

\f[V_{C} =IRV+V_{T} +ERV\f] 
<center>
<i>Equation 9.</i>
</center><br> 

V<sub>C</sub>  can also be calculated using TLC as:

\f[V_{C} =TLC-RV\f] 
<center>
<i>Equation 10.</i>
</center><br> 

Again, both TLC and RV are weight-based inputs to the model, and V<sub>T</sub> is calculated
as described above. For the standard patient in the model with TLC=6.16 L (80
L/kg) and RV=1.23 L, V<sub>C</sub>=4.93 L. Typical values of V<sub>C</sub>  reported in 
the literature are around 4.6 L @cite silverthorn2013human . The value in the engine corresponds to 
a weight-based vital capacity.

##### Inspiratory Capacity (IC)

The inspiratory capacity is another standard lung capacity that can be
calculated from TLC and FRC as

\f[IC=TLC-FRC\f] 
<center>
<i>Equation 11.</i>
</center><br> 

In the model, both TLC and FRC are weight-based input variables, and IC can be
calculated using the above equation. From the values in the model, IC can be
determined to be 3.85 L.

#### Ventilation

##### Respiration Rate (RR)

As described above, the %BioGears %Respiratory Model employs chemical feedback
mechanisms to regulate the ventilation frequency that affects the breathing cycle
through the respiratory driver. The breathing frequency is adjusted in
accordance to the arterial O<SUB>2</SUB> and CO<SUB>2</SUB> levels and other modifiers, such as drug and metabolic effects.
The engine switches between the inspiratory and expiratory phases based on the
predicted ventilation frequency. The respiration rate is then calculated by
measuring the time taken for a complete breathing cycle and converting it to the
number of breaths per minute. Typically, the respiration rate of a healthy adult
is 16 breaths/min @cite Levitzky2013pulmonary . A similar value is obtained
for the standard patient under normal tidal breathing.

##### Total Pulmonary Ventilation

The total pulmonary ventilation (or minute ventilation or minute volume) is the
volume of air moved into the lungs per minute. Minute ventilation (V<sup><b>.</b></sup><sub>E</sub>) is the
product of tidal volume (V<sub>T</sub>) and respiration rate (RR), i.e.,

\f[\dot{V}_{E} =V_{T} *RR\f] 
<center>
<i>Equation 12.</i>
</center><br> 

The %BioGears %Respiratory Model calculates both V<sub>T</sub> and RR from the simulation data. 
V<sup><b>.</b></sup><sub>E</sub> can thus be determined from the above equation by using the average values of V<sub>T</sub> and
RR. For the standard patient in the model, under normal physiological conditions,
the average values of V<sub>T</sub> and RR are 0.540 L and 16 breaths/min,
respectively. The total pulmonary ventilation obtained from the model equals
6.53 L/min. Typical weight-based value of minute volume is 84 mL/min/kg @cite Levitzky2013pulmonary . 
Using the mass 
of the standard patient (77 kg), the expected value of V<sup><b>.</b></sup><sub>E</sub> is 6.48 L/min, which is close to 
the value found from the engine.

##### Alveolar Ventilation

The term alveolar ventilation corresponds to the volume of air entering and
leaving the alveoli per minute @cite Levitzky2013pulmonary . The alveolar
ventilation is calculated as V<sup><b>.</b></sup><sub>A</sub>=(V<sub>T</sub>-V<sub>D</sub>)\*RR where V<sup><b>.</b></sup><sub>A</sub> is the alveolar
ventilation in liters per minute and V<sub>D</sub> is the volume of the conducting airways
that is referred to as the anatomical dead space. This is the region of
respiratory tract where no gas exchange takes place. In the %BioGears
%Respiratory Model, the volume of the dead space is calculated from the values
assigned to the right and left anatomic dead space nodes. These nodes have continuously changing volumes due the compliances that are connected to the
nodes. The right and left anatomic dead space volumes when
compared to the right and left alveoli volumes are shown in Figure 6. When the patient weight is factored into the 
calculation, the alveolar ventilation predicted from the model is close to the expected value.

<center><img src="./plots/Respiratory/Alveoli_and_Dead_Space_Volumes.jpg" width="800"></center>
<center>
<i>Figure 6. The right and left anatomic dead space volumes together with the right and left 
alveoli volumes. The difference in the alveoli volumes is due to the difference in 
 the lung ratio of the right and left lungs. The right and left lung ratios of the standard patient 
 in the %BioGears %Respiratory Model are 0.525 and 0.475, respectively. The left and right dead space volumes are equivalent.</i>
</center><br>

##### Tracheal Airflow

Airflow is measured by taking the instantaneous pressure difference across a
fixed resistance. The %BioGears %Respiratory Model measures tracheal airflow
<i>Q<sub>trachea</sub></i> by using the instantaneous pressure difference across the tracheal
resistance <i>R<sub>trachea</sub></i> as:

\f[Q_{trachea} =\frac{P_{mouth} -P_{carina} }{R_{trachea} } \f] 
<center>
<i>Equation 13.</i>
</center><br> 

<i>P<sub>mouth</sub></i> and <i>P<sub>carina</sub></i> are the pressures at the mouth and the carina nodes,
respectively (see circuit diagram, Figure 3). The figure shown below presents the absolute flow rate (no distinction for flow direction) for one breathing cycle.

<center><img src="./plots/Respiratory/Total_Flow_Through_Trachea.jpg" width="800"></center>
<center>
<i>Figure 7. Tracheal airflow and total lung volume during one typical breathing
cycle. At the peak of the inspiration phase, the flow rate goes to zero.</i>
</center><br>

#### Alveolar Pressure

The instantaneous pressures at the nodes of the respiratory circuit are
calculated by solving the circuit matrix equation as described in the @ref
CircuitMethodology. The pressures at the right and left alveoli nodes of the
%Respiratory Model represent the alveolar pressure. Typically, the values of
the alveolar pressure vary in the range from -1.8 cm H<SUB>2</SUB>O to 1.8 cm H<SUB>2</SUB>O (relative
to atmospheric pressure) during the inspiration and expiration phases of the
breathing cycle @cite otis1947measurement . The figure
below depicts the alveolar pressure along with lung
volume for one breathing cycle. The alveolar pressure in the engine is absolute (not relative
to atmospheric pressure), so the relative pressure can be determined by subtracting the standard atmospheric pressure of 1033 cmH2O - giving outputs close to the range of  -1.8 cm H<SUB>2</SUB>O to 1.8 cm H<SUB>2</SUB>O
found in the literature @cite otis1947measurement . 

<center><img src="./plots/Respiratory/Lung_Pressure_And_Volume.jpg" width="800"></center>
<center>
<i>Figure 8. Typical lung pressures. The plot shows the instantaneous pressure of
the left alveoli for one breathing cycle. For comparison, the plot also
shows the total lung volume for the same breathing cycle. As seen in this figure,
the lung volume increases as the alveolar pressure falls below the atmospheric pressure of 1033 cm H<SUB>2</SUB>O.
This creates a pressure differential between the airway node and the alveoli,
allowing air to flow into the lungs. When the alveolar pressure goes above 
1033 cm H<SUB>2</SUB>O, the lung volume decreases from its peak, representing the
expiration phase.</i>
</center><br>

##### Transpulmonary pressure

Transpulmonary pressure is defined as the difference between the alveolar 
 and the intrapleural pressures. The %BioGears %Respiratory System derives the transpulmonary 
 pressure from the calculated values of the alveolar pressure and intrapleural pressures. 
 The alveolar pressure is obtained by taking the average of the left and right 
 alveolar pressures. Similarly, the total intrapleural pressure is obtained by taking the average of the 
 the left and right pleural pressures. The plots shown below compare the transpulmonary pressure from the %BioGears 
 Engine with that found in literature @cite guyton2006medical. Some variations in the waveforms and 
 possibly the average values is a consequence of the specific driver pressure and 
 patient parameters employed in the engine.

<center>
<table border="0">
<tr>
    <td><img src="./plots/Respiratory/BioGears_Pressures.jpg" width="550"></td>
    <td><img src="./plots/Respiratory/Guyton_Lung_Pressures.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/Respiratory/BioGears_Lung_Volume.jpg" width="550"></td>
    <td><img src="./plots/Respiratory/Guyton_Lung_Volume.jpg" width="550"></td>
</tr>
</table>
</center>
<center>
<i>Figure 9. A plot showing the transpulmonary pressure obtained from the %BioGears engine with 
that found and digitized from literature @cite guyton2006medical. The left %BioGears plots use absolute pressure, while the right Guyton plots use the pressure difference from ambient (1033 cmH2O).  For comparison, the plot also shows the lung volumes from the two sources.</i>
</center><br>

#### Pressure-Volume (P-V) curve
One method of characterizing the lungs' elastic behavior is to use a diagram
that relates the lung volume changes to changes in pleural pressure. The pressure-volume 
curve of a healthy person shows hysteresis during the inspiratory and expiratory phases. 
Figure 10 presents the pressure-volume diagram of data extracted from the %BioGears
%Respiratory Model. For comparison, the plot also shows a P-V diagram reproduced from literature
@cite guyton2006medical . The figures show the plot of lung volume changes versus pleural pressure
 for one breathing cycle. The pleural pressure from the model is the average of the right and left pleural 
 pressures. The lung volume change corresponds to the change in the total lung volume  
 during a complete breathing cycle.  
As shown in the figure, the %BioGears %Respiratory Model mimics the expected 
hysteresis of the P-V curve.

<center>
<table border="0">
<tr>
    <td><img src="./plots/Respiratory/BioGears_Pulmonary_Compliance.jpg" width="550"></td>
    <td><img src="./plots/Respiratory/Guyton_Pulmonary_Compliance.jpg" width="550"></td>
</tr>
</table>
</center>
<center>
<i>Figure 10. The pressure-volume curve for the standard patient of the %BioGears %Respiratory Model 
under normal physiological conditions. For comparison, the figure includes plot reproduced from 
literature @cite guyton2006medical . The plot from the model shows the expected hysteresis of the P-V 
diagram observed in a healthy person.</i>
</center><br>

#### Partial Pressures of %Respiratory Gases

For any gas mixture, the partial pressure P<sub>gas</sub> of a particular gas in the
mixture can be calculated based on the total pressure P<sub>total</sub> of all gases in the
mixture and the fractional concentration F<sub>gas</sub> of the gas as

\f[P_{gas} =F_{gas} *P_{total} \f] 
<center>
<i>Equation 14.</i>
</center><br> 

The %Respiratory Model calculates the partial pressure of a gas at any node
based on the total pressure and the volume fraction of the gas at the node. The
node volume fraction of the gas and the node pressure are calculated in a manner
described in the @ref SubstanceTransportMethodology and @ref CircuitMethodology,
respectively. The %Respiratory Model employs the above equation to predict the
partial pressure of gases in the %Respiratory System. The sections below present
the results of alveolar and tracheal partial pressures of O<SUB>2</SUB> and CO<SUB>2</SUB>.

##### Alveolar O<SUB>2</SUB> Partial Pressure

The %BioGears Engine calculates the O<SUB>2</SUB> partial pressure P<sub>Lung<sub>O<SUB>2</SUB></sub/></sub/> at each
alveoli node by using the oxygen volume fraction VF<sub>Lung<sub>O<SUB>2</SUB></sub/></sub/> and the total
pressure P<sub>Lung</sub/> at the alveoli nodes as

\f[P_{LungO_{2} } =VF_{LungO_{2} } *P_{Lung} \f] 
<center>
<i>Equation 15.</i>
</center><br> 

The alveolar O<SUB>2</SUB> partial pressure can thus be determined by taking the average of
O<SUB>2</SUB> partial pressures at the two alveoli nodes. In the model, the alveoli node
pressures are gauge pressures and are expressed relative to the atmospheric
pressure. The model converts these relative pressures to their absolute
pressures when calculating the gas partial pressures. The %BioGears Engine
assumes the inspired air is heated and humidified. Therefore, the water vapor
pressure at normal body temperature (P<sub>H<SUB>2</SUB>O</sub>=47 mm Hg) is subtracted from the
standard atmospheric pressure of P<sub>B</sub>=760 mm Hg when the gas partial pressure is
calculated using the absolute lung pressure, i.e.,

\f[P_{LungO_{2} } =VF_{LungO_{2} } *(P_{B} -P_{H_{2} O} +P_{Lung} )\f] 
<center>
<i>Equation 16.</i>
</center><br> 

Figure 11 depicts the plot of P<sub>LungO<SUB>2</SUB></sub>  for the left and right alveoli
of the standard patient. Typically, the average alveolar partial pressure of oxygen 
 is 104 mmHg @cite Levitzky2013pulmonary . The value from the engine is close to that of the 
 literature.

<center><img src="./plots/Respiratory/Alveolar_Oxygen_Partial_Pressure.jpg" width="800"></center>
<center>
<i>Figure 11. Alveolar O<SUB>2</SUB> partial pressure. The partial pressure of O<SUB>2</SUB> at the two
alveoli nodes is calculated by using the pressure and the O<SUB>2</SUB> volume fraction at
each alveoli node. The plot shows the value of O<SUB>2</SUB> partial pressure as a function
of time over the course of multiple breathing cycles.</i>
</center><br>

##### Alveolar CO<SUB>2</SUB> Partial Pressure

The alveolar CO<SUB>2</SUB> partial pressure is calculated in the same manner as the oxygen
partial pressure. Figure 12 depicts the plot of alveolar CO<SUB>2</SUB> partial
pressure for the left and right alveoli nodes. Typically, the average alveolar CO<SUB>2</SUB> partial pressure is 40 mmHg @cite Levitzky2013pulmonary .
The prediction from the engine is close to the expected
literature value.

<center><img src="./plots/Respiratory/Alveolar_Carbon_Dioxide_Partial_Pressure.jpg" width="800"></center>
<center>
<i>Figure 12. Alveolar CO<SUB>2</SUB> partial pressure. The partial pressure of CO<SUB>2</SUB> at the two
alveoli is calculated by using the pressure and the CO<SUB>2</SUB> volume fraction at each
alveoli node. The plot shows the value of the CO<SUB>2</SUB> partial pressure as a function
of time over the course of multiple breathing cycles.</i>
</center><br>

##### Tracheal CO<SUB>2</SUB> Partial Pressure

Tracheal CO<SUB>2</SUB> partial pressure is calculated in the same manner as the alveolar
CO<SUB>2</SUB> partial pressure. The engine calculates the tracheal partial pressure by
using the pressure at the carina node. Recall that the carina node belongs to
the node where the trachea branches into the mainstem bronchi. The figure below
depicts the time variation of tracheal CO<SUB>2</SUB> partial pressure. The waveform for
CO<SUB>2</SUB> partial pressure appears similar to those found in normal capnograph at the mouth
(Fig.3.9, Ref @cite Levitzky2013pulmonary ). 

<center><img src="./plots/Respiratory/Trachea_Carbon_Dioxide_Partial_Pressure.jpg" width="800"></center>
<center>
<i>Figure 13. Tracheal CO<SUB>2</SUB> partial pressure. The partial pressure of CO<SUB>2</SUB> at the
trachea is calculated by using the pressure and the CO<SUB>2</SUB> volume fraction at the
carina node. The plot shows the value of tracheal CO<SUB>2</SUB> partial pressure over the course of one breathing cycle.</i>
</center><br>

##### Tracheal O<SUB>2</SUB> partial pressure

The O<SUB>2</SUB> partial pressure at the trachea is calculated in the same manner as the
alveolar O<SUB>2</SUB> partial pressure. As mentioned for CO<SUB>2</SUB> partial pressure, the model
calculates the tracheal O<SUB>2</SUB> partial pressure by making use of the pressure at the
carina node. The output of the tracheal O<SUB>2</SUB> partial
pressure calculation is presented in Figure 14.

<center><img src="./plots/Respiratory/Trachea_Oxygen_Partial_Pressure.jpg" width="800"></center>
<center>
<i>Figure 14. Tracheal O<SUB>2</SUB> partial pressure. The partial pressure of O<SUB>2</SUB> at the
trachea is calculated by using the pressure and the O<SUB>2</SUB> volume fraction at the
carina node. The plot shows the value of tracheal O<SUB>2</SUB> partial pressure over the course of one breathing cycle.</i>
</center><br>

@anchor respiratory-dependencies
### Dependencies

The %BioGears %Respiratory System interacts with other physiological
systems either directly or indirectly through processes that involve the
transport and exchange of gases. These interdependencies are discussed below.

#### Gas Transport

The transport of gases in the %BioGears %Respiratory Model is handled through
the Transport functionality of the engine, where mass conservation based
on volume fraction and volumetric flow rate of gases at the nodes and paths of
the respiratory circuit is employed (see the @ref SubstanceTransportMethodology for
details). During free breathing, the ambient atmospheric volume fractions are used as an input to the %Respiratory Model.
 
By using methods described in the @ref SubstanceTransportMethodology, the %BioGears
Engine calculates the volume fraction of gases at the nodes associated with the
respiratory functional units. The %Respiratory System uses the calculated volume
fractions and predicts various physiological parameters. For example, the gas volume
fraction at the end of expiration, referred to as end-tidal gas concentration,
can be monitored based on the results from the @ref SubstanceTransportMethodology
calculation. The end-tidal gas concentration is an important clinical parameter
for monitoring patients and preventing mishaps related to insufficient
ventilation or inappropriate gas concentration during anesthesia and immediate
recovery @cite linko1989monitoring. Monitoring end tidal CO<SUB>2</SUB> (ETCO<SUB>2</SUB>) is a
widely established clinical practice for verification of endotracheal tube
placement and is also one of the standard requirements for monitoring patients
in transport @cite donald2006end. The %Respiratory Model provides the
end-tidal gas concentration based on the expired gas volume fractions at the
airway node.

#### %Environment

The %BioGears %Respiratory System interacts with the %Environment System for
the atmospheric pressure values assigned to the mouth node. Changes to the 
environmental conditions, such as changes in altitude, ambient temperature, humidity, 
and others, can affect the breathing pattern of the patient. The %Respiratory System 
interacts with the %Environment System to utilize the values of 
ambient pressures and gas concentrations that reflect the environmental condition.
The %Respiratory System also interacts with the %Environment System for proper handling 
of inhaled gases that can arise from environmental incidents.

#### Alveolar Gas Exchange

At the alveoli-capillary interface, gas transfer occurs. The primary goal is to
transfer incoming oxygen into the bloodstream for transport to other organs and
to transfer waste carbon dioxide out of the body. This mechanism is also used to
transport inhaled agents into the bloodstream from the
%Respiratory System. For more details on gas transport, see @ref
SubstanceTransportMethodology and @ref BloodChemistryMethodology.

In the %BioGears %Respiratory Model, the spontaneous respiration rate is adjusted based on a chemical feedback mechanism
that depends on the arterial oxygen and carbon dioxide levels as described
above. The arterial oxygen and carbon dioxide levels depend on the level of O<SUB>2</SUB>
consumption and CO<SUB>2</SUB> production in the circulatory system, which in turn affects
the gas exchange at the alveolar-capillary interface. As the arterial O<SUB>2</SUB> and CO<SUB>2</SUB>
levels change, the breathing rate, tidal volume, and alveolar ventilation also change,
which in turn facilitates efficient gas exchange between the atmosphere and the
body. This presents an example of how the %BioGears Engine integrates the
circulatory and %Respiratory Systems to regulate the blood gas levels.

#### Drug Effects

As drugs circulate through the system, they affect the %Respiratory System. The drug effects on respiratory rate and tidal volume are calculated in the %Drugs System. These effects are applied to the respiratory driver by modifying the frequency and amplitude of the breathing cycle. During the respiratory driver calculations, the respiratory rate and tidal volume changes that are calculated in the %Drugs System are applied to the resulting respiratory rate and tidal volume calculated in the driver. Additionally, a neuromuscular block level is applied based on the drug effects.  If the neuromuscular block level is above 0.15 on a scale of 0 to 1.0, then the respiratory rate and tidal volume are set to zero. This represents the paralytic effects of a neuromuscular block agent.  The value of 0.15 was chosen to satisfy the block duration for succinylcholine and rocuronium. The strength of these effects, including the block effect, are calculated based on the plasma concentration and the drug parameters in the substance files in the %Drugs System.  For more information on this calculation see @ref DrugsMethodology.

#### Metabolic Effects

A metabolic modifier is set by the %Energy System (@ref EnergyMethodology) to drive the system to reasonable levels achievable during increased metabolic exertion.  The modifier is tuned to achieve the correct respiratory response for near maximal exercise, and a linear relationship is assumed. This modifier is a direct multiplier to the target alveolar ventilation input into the system driver, and it causes an increase in both tidal volume and respiration rate.

#### Anesthesia Machine Connection

The %Respiratory System can be hooked up to the anesthesia machine for positive-pressure ventilation (see the @ref AnesthesiaMachineMethodology).  This is achieved by connecting the two fluid circuits.  The anesthesia connection node is merely connected to the respiratory mouth node to allow for automatic calculation of the fluid mechanics by the circuit solver and transport by the substance transporter.  The mechanistic cascading effects are automatically achieved, and everything else is modeled exactly the same as when the systems are disconnected.

@anchor respiratory-outputs
###Outputs

At each time iteration, the %Circuit Solver calculates the values of the
state variables for that particular time. Using the calculated state variables,
the model predicts various physiological parameters of mechanical ventilation. Many of the calculated system data outputs are derived from the difference between the minimum and maximum lung volumes and the time between Occurrences - e.g., tidal volume and respiration rate.

Other values, like the pulmonary resistance and compliance, are determined instantaneously.  The pulmonary resistance is calculated by taking the ratio of 
the pressure difference between the mouth <i>P<sub>mouth</sub></i> and the alveoli <i>P<sub>alveoli</sub></i> and 
the flow across the trachea <i>Q<sub>trachea</sub></i> as

\f[R_{pulm} =\frac{P_{mouth} -P_{alveoli} }{Q_{trachea} } \f] 
<center>
<i>Equation 15.</i>
</center><br> 

The %BioGears %Respiratory Model calculates the pulmonary compliance <i>C<sub>pulm</sub></i> by dividing the tidal 
volume <i>V<sub>T</sub></i> by the intrapleural pressure <i>P<sub>pleu</sub></i> difference as

\f[C_{pulm} =\frac{V_{T} }{P_{pleau(\max )} -P_{pleu(\min )} } \f] 
<center>
<i>Equation 16.</i>
</center><br> 
 
Here <i>P<sub>pleu(min)</sub></i> and <i>P<sub>pleu(max)</sub></i> are the minimum and maximum respective pressures at the 
right and left pleural nodes.

@anchor respiratory-assumptions
Assumptions and Limitations
---------------------------

### Assumptions

As in most lumped parameter models of mechanical ventilation, the %BioGears
%Respiratory Model is based on two main parameters: the resistance R and the
compliance C (or elastance E). Similar to their electrical analogues, these
elements form a closed circuit to represent the energy dissipation and storage
properties of the normal tidal ventilation. One extension of the linear model
is the addition of inertance in the lumped parameter model. Inertance is
not included in the model on the assumption that inertia does not play a
significant role under conventional tidal breathing, as opposed to high-frequency
ventilation.

In the model, the %Respiratory System is assumed to behave linearly in that the
hydraulic resistance R is assumed to be independent of the flow rate Q, and the
elastance E or the compliance C of the elastic component is assumed to be
independent of the volume V. Therefore, under normal physiological conditions,
the circuit elements (resistors and capacitors) of the %BioGears %Respiratory
System are treated as constants. However, their values can be adjusted when
addressing pathological conditions.

The %Respiratory Model does not distinguish between different functional units
in the upper airway, the anatomical region where the inspired air is heated,
humidified, and filtered. In this region, the flow of air is turbulent, as
opposed to the laminar flow in lower airways. That means that the Ohm's analogue
pressure-volume relation cannot be applied to this region. However, the model
treats the upper airway as part of the mouth node and assumes laminar,
incompressible flow for the remaining airway.

### Limitations

The current version of the %Respiratory Model does not account for the upper
airways, and the detailed nature of upper airway flows will not be captured in
the model. It is known that flows at the upper airways are turbulent, while flows
in the lower respiratory tract are laminar. The absence of upper airways in
the model can underestimate the airway flow. In this regard, the lumped
parameter model is inherently at a disadvantage in handling complex geometries
and flows that can represent detailed upper airway pathology. However, the model
is sufficiently detailed to reproduce the respiratory responses associated with a
number of airway pathologies at relatively low computational cost.

Airway inertance is assumed to be negligible at normal respiratory frequencies.
However, analysis of high-frequency ventilation (HFV) of up to 40 Hz ventilatory
frequency would require the inclusion of inertance @cite rozanek2008design .
In mechanical ventilation, HFV is a useful procedure that permits significantly
decreased pressure amplitude and tidal volume, preventing trauma in artificial
ventilation. When compared to conventional mechanical ventilation, HFV is
considered to be the preferred method of mechanical ventilation in patients with
acute lung injury @cite Krishnan2000hfv . The small tidal volumes needed in
HFV are thought to offer an advantage over conventional ventilation in limiting
additional lung injury from mechanical ventilation @cite Krishnan2000hfv . The
current version of the %Respiratory Model handles conventional mechanical
ventilation but not HFV.

The model makes no distinction between different generations of bifurcating
airways. Therefore, factors affecting the regional ventilation and perfusion of
the lungs cannot be captured by the model.

@anchor respiratory-conditions
Conditions
----------

#### Chronic Obstructive Pulmonary Disease

Chronic Obstructive Pulmonary Disease (COPD) is an obstructive lung disease characterized by chronically reduced airflow into the lungs. Symptoms typically include increased respiration rate, decreased  tidal volume, decreased expiratory flow, and reduced oxygen saturation @cite man2003contemporary . There are also indications that COPD affects amino acid metabolism, causing muscle wasting and altered blood concentrations of certain amino acids @cite engelen2000factors. Unlike asthma, the reduction in lung airflow does not generally improve with medication. Symptoms typically worsen over time, often leading to death. COPD is caused by damage to the small airways in the lungs and destruction of the alveolar membranes. When tissue damage occurs primarily in the airways, the condition is called chronic bronchitis. When tissue destruction is focused in the alveoli, the condition is called emphysema. In %BioGears, COPD is a manifestation of both conditions, although one condition may dominate. 

%BioGears simulates COPD by modeling damage to the small airways and alveolar membranes. COPD severity is controlled by two severity values, a chronic bronchitis severity value and an emphysema severity value. Chronic bronchitis severity is used to scale the airflow resistance through the lower airways in the circuit model, simulating airway tissue damage and scarring. Increasing chronic bronchitis severity increases the airflow resistance through the lower airways. The function used to determine airflow resistance for COPD is the same as that used for asthma (see Figure 18) and is based on chronic bronchitis severity.

Emphysema severity is used to scale destruction of the alveolar membranes. Destruction of the alveolar membranes decreases the effective surface area for gas exchange, thereby reducing alveolar gas transfer. In cases of severe emphysema, up to 80% of the alveolar membranes are destroyed, with a corresponding reduction in effective gas diffusion surface area @cite guyton2006medical. %BioGears scales the gas diffusion surface area using a multiplier based on emphysema severity.  The function used to determine the gas diffusion surface area multiplier is plotted in Figure 15. 

Changes in amino acid metabolism are shifted by static percentages based on data from @cite mathur1999cerebral and @cite engelen2000enhanced. The protein breakdown rate is increased, as well as the fraction of muscle that is required to produce energy anaerobically regardless of O2 presence. These changes result in the increased protein turnover and elevated lactate concentrations observed in COPD patients.

Additionally, %BioGears models destruction of lung tissue with an increase in alveolar compliance.  The function used to determine the compliance multiplier is plotted in Figure 16. 

The destruction of the alveolar membranes also destroys the pulmonary capillaries embedded in the membranes. To model pulmonary capillary destruction, %BioGears increases the pulmonary capillary flow resistance based on severity. Although increased pulmonary capillary resistance is related to alveolar membrane destruction, and therefore associated with emphysema, %BioGears uses either emphysema severity or chronic bronchitis severity (whichever is higher) to determine the pulmonary capillary resistance multiplier - see Figure 19. This was done in an attempt to model increased blood pressure and elevated heart rate, which are symptoms of both emphysema and chronic bronchitis. Increasing the capillary resistance should increase arterial blood pressure as the heart pumps harder to overcome the increased resistance in the lungs. However, we were unable to successfully model either increased blood pressure or elevated heart rate for COPD, as the validation results show (see @ref respiratory-results "Results and Conclusions"). This will be the focus of later efforts to improve the model.  

Decreased Inspiration-Expiration (IE) ratio is another pathophysiologic feature of COPD.  As with asthma, the normal IE ratio is scaled using a multiplier based on severity. Either chronic bronchitis severity or emphysema severity (whichever is higher) is used to determine the IE ratio scaling multiplier - see Figure 18. 

<table border="0">
<tr>
    <td><img src="./images/Respiratory/Respiratory_Figure17_GasDiffusion.png" width="550"></td>
    <td><img src="./images/Respiratory/Respiratory_Figure18_AlveoliCompliance_COPD.png" width="550"></td>
</tr>
<tr>
    <td><center><i>Figure 15. Gas diffusion surface area multiplier as a function of emphysema severity. Gas diffusion surface area decreases with increasing emphysema severity.</i></center></td>
    <td><center><i>Figure 16. Alveoli compliance multiplier as a function of emphysema severity. Alveoli compliance increases with increasing emphysema severity.</i></center></td>
</tr>
<tr>
    <td><img src="./images/Respiratory/Respiratory_Figure19_PCResistance.png" width="550"></td>
    <td><img src="./images/Respiratory/Respiratory_Figure20_IERatio_COPD.png" width="550"></td>
</tr>
<tr>
    <td><center><i>Figure 17. Pulmonary capillary flow resistance multiplier as a function of COPD severity. Pulmonary capillary flow resistance increases with increasing chronic bronchitis or emphysema severity, whichever is higher.</i></center></td>
    <td><center><i>Figure 18. IE ratio multiplier as a function of COPD severity. IE ratio decreases with increasing chronic bronchitis or emphysema severity, whichever is higher.</i></center></td>
</tr>
</table>

#### Lobar Pneumonia

Lobar pneumonia is a form of pneumonia that affects one or more lobes of the lungs.  Symptoms typically include increased respiration rate, decreased  tidal volume, reduced oxygen saturation, decreased IE ratio, and increased body temperature @cite ebell2006outpatient .  As fluid fills portions of the lung, it becomes more difficult to breathe. Fluid also reduces the effective gas diffusion surface area in the alveoli, reducing alveolar transfer of oxygen and carbon dioxide into and out of the bloodstream @cite guyton2006medical . %BioGears simulates lobar pneumonia by decreasing the alveoli compliance in the respiratory circuit, which models increased breathing difficulty due to fluid congestion in the alveoli. Similarly, gas diffusion surface area is reduced using the same function as for COPD (see Figure 15). Decreased IE ratio is pathophysiologic feature of lobar pneumonia.  Like COPD, the normal IE ratio is scaled using a multiplier based on severity. %BioGears uses the same function as for COPD to determine the IE ratio multiplier (see Figure 18).   

<img src="./images/Respiratory/Respiratory_Figure21_AlveoliCompliance_LP.png" width="550">
<center>
<i>Figure 19. Alveoli compliance multiplier as a function of lobar pneumonia severity. Alveoli compliance decreases with increasing severity.</i>
</center><br>

#### Impaired Alveolar Exchange

The impaired alveolar exchange generically models an unspecified reduction of effective alveolar surface area.  This condition causes less effective gas exchange between the %Respiratory and %Cardiovascular systems.  The user can specify either a fraction or area value of the surface area to remove.

@anchor respiratory-actions
Actions
-------

### Insults

#### Acute Respiratory Distress
Acute Lung Injury (ALI) and Acute Respiratory Distress Syndrome (ARDS) arise when the pulmonary capililary endothelium and the alveolar epithelium incur damage @cite matthay2011acute.  The source of the epithelial damage could be a foreign agent (bacteria, virus) or some type of trauma, like smoke inhalation.  Capillary endothelial damage generally results from an acute, dysregulated inflammatory response that impairs starling forces across the endothelial barrier and allows fluid and inflammatory mediators to pass into the alveolar space @cite matuschak2010acute.  The accumulation of fluid and inflammatory agents (like neutrophils) in the alveoli impairs gas exchange, leading to hypoxemia.  Furthermore, ALI and ARDS increase pulmonary capillary resistance and mean pulmonary arterial pressure @cite ryan2014pulmonary. The severity of the systemic inflammatory response induced by pulmonary injury can cause multiple organ dysfunction syndrome (MODS), which greatly increases mortality.  While estimates of mortality from ARDS uncomplicated by MODS range from 26% and 40% @cite ryan2014pulmonary @cite varisco2011pharmacology, mortality can exceed 50% in cases of MODS and septic shock @cite matuschak2010acute.  Current standard of care for ALI and ARDS emphasizes mechanical ventilation with low target tidal volumes (6 mL/kg) @cite bein2016standard.  Patients must be sedated for mechanical ventilation and in some cases require neuormuscular blockers.  Other interventions involve positioning the patient facedown and administering pulmonary vasodilators @cite bein2016standard.  

%BioGears models Acute Respiratory Distress by increasing the resistance on the pulmonary capillary pathways (see [Respiratory Circuit])(@ref respiratory-features) and by decreasing the diffusing capacity in the [alveolar gas exchange model](@ref tissue-alveolarExchange).  The %BioGears ARDS treatment scenario focuses on sedation and mechanical ventilation.   


#### Airway Obstruction

Airway obstruction refers to a blockage in the airway that partially or wholly
prevents airflow into or out of the lungs. Airway obstructions
can occur at the lower and upper airways. They can
be caused by foreign objects, allergic reactions, infections, inflammations,
toxic gases, and other reasons. The pathological symptoms and physiological
manifestations of airway obstructions are as diverse as the causes, and the severities
of the obstructions follow suit. Dyspnea (breathing difficulty) is the
obvious common symptom of airway obstruction. The complications of breathing
difficulty include hypoxia (low oxygen levels) and hypercapnia (high carbon
dioxide levels), respiratory acidosis, and others.

The current version of the %BioGears %Respiratory Model attempts to reproduce
the physiological responses arising from foreign body upper airway obstruction.
Foreign body airway obstruction is fairly common during anesthesia. One example
is the case reported by @cite Roy2005airwayObstruction where the oxygen
saturation (SPO<SUB>2</SUB>) of a patient dropped to 40 percent due to a blockage by a
parasitic nematode during the use of a laryngeal mask airway. The %Respiratory
Model simulates airway obstruction by increasing the flow resistance across the
path connecting the airway node to the carina node. This path corresponds to the
flow through the trachea. The model then calculates the physiological responses
due to increased airway resistance.

<img src="./images/Respiratory/Respiratory_Figure22.png" width="550">
<center>
<i>Figure 20. Logarithmically increasing resistance values used to increase the airway resistance in relation
to the airway obstruction severity levels.</i>
</center><br>

#### Bronchoconstriction

A patient with bronchospasm experiences a sudden constriction of the muscles in
the walls of the bronchioles, decreasing the airway diameter. Such decrease in
the bronchial airway diameter, or bronchoconstriction, results in the reduction of
gas flow into or out of the lungs and causes difficulty in both spontaneous and
mechanical ventilation. A number of factors can trigger bronchospasms, including
a foreign body in the airway and stimulation of an endotracheal tube in patients
with reactive disease @cite Woodworth2012anesthesia . The %BioGears
%Respiratory Model accounts for this respiratory distress by adjusting the
resistance of the bronchi units. The increase in bronchial airway resistance results in a
decrease in gas flow in and out of the alveoli, which in turn affects the gas
concentration in the circulatory system. The engine then responds to the
respiratory distress in proportion to the level of bronchoconstriction.

#### Intubation

The Intubation action sets the type of intubation present: off, tracheal, esophageal, right mainstem, or left mainstem. Tracheal intubation is the successful state. See @ref AnesthesiaMachineMethodology for more details about intubation and mechanical ventilation.

##### Mainstem intubation

During endotracheal intubation, a flexible tube is inserted orally or nasally
down into the trachea to facilitate ventilation of the lungs during mechanical
ventilation. A potential complication of endotracheal intubation is mainstem
intubation, an incident where the tube is inadvertently placed too deep into one
of the two mainstem bronchi. In such incidents, the unintubated lung does not
contribute to gas exchange, and a right-to-left shunt can occur, resulting in
hypoxia @cite divatia2005complications . Furthermore, the intubated lung can
be hyperinflated, predisposing to overdistension and barotrauma @cite divatia2005complications .

Right mainstem intubation is the common complication of endotracheal intubation
because of the shallower angle that the right mainstem bronchus makes with the
trachea, though %BioGears includes the ability to simulate left mainstem intubation as well. The %BioGears %Respiratory Model attempts to simulate this incidence by
manipulating the resistance representing the opposite bronchus functional unit of
the %Respiratory System. A numerically large resistance is selected for this
circuit element for complete blockage of flow into the lung. The model then
evaluates the physiological responses arising from the incidence.

##### Esophageal Intubation

Another complication of endotracheal intubation is esophageal
intubation, an incident where the endotracheal tube is inadvertently placed in
the esophagus. When such an incidence occurs, the lung does not receive air. The
%BioGears %Respiratory System models esophageal intubation by opening
the airflow path leading to the esophagus and closing the flow path leading to
the trachea compartment. This is accomplished in the model by assigning
a numerically large value for the tracheal resistance and assigning a small
resistance to the esophageal compartment. The lack of airflow into the lungs
affects the amount of alveolar-capillary gas exchange, which in turn affects the
hemodynamic and respiratory responses of the engine.

#### Tension Pneumothorax

Pneumothorax is an abnormal accumulation of air in the pleural cavity, i.e., a
thin space between the visceral and parietal pleura separating lungs from the
chest wall, due to factors that cause air to escape into the pleural space. The
accumulation of air in the pleural space limits the space available for full lung
inflation and increases the intrathoracic pressure, causing a collapse in the ipsilateral lung depending on the type and
severity of the incidence. Pneumothoraces are classified as traumatic and
nontraumatic or spontaneous based on their cause. Pneuomothoraces are 
classified as open (sucking chest wound) or closed (intact thoracic cage)
pnumothoraces @cite sharma2008principles .

The %BioGears %Respiratory Model is capable of simulating traumatic tension
pneumothorax. Tension pneumothorax is a life-threatening condition that occurs
when air escapes into the pleural space on inspiration, but cannot return on
expiration due to a one-way valve effect of the injury. This causes a
progressive build-up of air within the pleural space. The %Respiratory
System implements a model for tension pneumothorax by including subordinate
circuit elements that represent the flow and trapping of air in the pleural
space. The circuit elements represent both the open and closed pneumothorax
condition. In the open tension pneumothorax condition, air is allowed to enter
from the atmosphere into the pleural space as in penetrating chest wound
traumas. In the closed tension pneumothorax case, the model allows air to escape
from the lung into the plural space as in laceration of the lung following blunt
trauma. The model evaluates the physiological responses arising from increased
intrapleural pressure and leakage in the airflow of the respiratory circuit.

<img src="./images/Respiratory/PneumoCirucit.png" width="550">
<center>
<i>Figure 21. Both lungs in the engine have elements to mimic the effects of open and closed tension pneumothorax insults as well as chest occlusive dressing (for open) and needle decompression (for both) interventions. The red boxes denote these additional elements.</i>
</center><br>

<img src="./images/Respiratory/PneumoLeakResistance.png" width="550">
<center>
<i>Figure 22. An exponential function is used to map pnemothorax action severity to leak resistance.  This same function is used for both the open and the closed tension pneumothorax.</i>
</center><br>

#### Acute Asthma

Asthma is a common inflammatory disease of the airways (bronchi and bronchioles) where air flow into the lungs is partially obstructed. Acute asthma is an exacerbation of asthma that does not respond to standard treatments. Symptoms include elevated respiration rate, labored breathing, and a reduction in oxygen saturation, among others.  It is generally considered a life-threatening obstruction of the airway requiring immediate medical attention. During an acute asthma attack, airflow is partially obstructed during exhalation, and flow resistance is 5 to 15 times normal @cite papiris2002asthma . %BioGears simulates this by increasing the airway flow resistance in the circuit model. The function used to determine the airway flow resistance multiplier is shown in the figure below. The asthma attack severity is governed by a specified severity value. The higher the severity, the more severe the asthma attack, and the higher the resistance values are set.

Additionally, the inspiratory/expiratory (IE) ratio decreases during an acute asthma attack. It takes a noticeably longer time to exhale than inhale @cite papiris2002asthma .  In %BioGears, the normal IE ratio is scaled using a multiplier based on severity to model decreased IE ratio during an acute asthma attack.

<table border="0">
<tr>
    <td><img src="./images/Respiratory/Respiratory_Figure25_AirwayResistance.png" width="550"></td>
    <td><img src="./images/Respiratory/Respiratory_Figure26_IERatio_Asthma.png" width="550"></td>
</tr>
<tr>
    <td><center><i>Figure 23. Airway flow resistance multiplier as a function of asthma severity. Airway flow resistance increases (during exhalation) with increasing asthma severity.</i></center></td>
    <td><center><i>Figure 24. IE ratio multiplier as a function of asthma severity. IE ratio decreases with increasing asthma severity.</i></center></td>
</tr>
</table>

@anchor respiratory-interventions
### Interventions

#### Occlusive Dressing

The management of an open tension pneumothorax requires sealing the open chest wound with an occlusive dressing. Such intervention slows the progression of tension pneumothorax and may ensure recovery if timely intervention takes place. The %BioGears %Respiratory Model simulates occlusive dressing by assigning large flow resistance across the element that serves as an electrical analogue open switch for the path linking the pleural cavity to the environment. Based on this implementation, the model calculates the physiological responses arising from numerical equivalent occlusive dressing.

#### Needle Decompression

Tension pneumothorax is a life-threatening condition that requires immediate intervention to relieve the air trapped in the pleural space and reduce the intrapleural pressure to the sub-atmospheric level. Needle decompression is a procedure that allows air to escape from the pleural space into the environment. This procedure relieves the accumulated air, allowing full expansion of the affected lung. The %BioGears %Respiratory Model is capable of simulating needle decompression by incorporating a circuit element that permits airflow based on pressure differential between the pleural space and the environment. The model then evaluates physiological responses to the intervention.

### Conscious Respiration 

Conscious respiration consists of a set of commands that model forced exhalation, forced inhalation, holding breath, and inhaler actuation. Collectively, they allow %BioGears to model coordinated use of an inhaler.  It should be noted that the conscious respiration action begins immediately when called, and will continue until completed while the simulation continues.  Users will likely want to advance time for the duration of the conscious respiration command before attempting other actions or completing a scenario.  The following commands can be used in any order and will wait until the completion of the previous command to begin:

- <b>Forced Exhale</b>: The "depth" or "shallowness" of the exhalation is determined by the expiratory reserve volume fraction, which specifies what portion of the expiratory reserve volume to exhale. If set to 1.0, the patient exhales his or her entire expiratory reserve volume, leaving only the residual volume in the lungs. The time period over which to execute the exhale is determined by the period. 
- <b>Forced Inhale</b>: The "depth" or "shallowness" of the inhalation is determined by the inspiratory capacity fraction, which specifies what portion of the inspiratory capacity to inhale. If set to 1.0, the patient inhales his or her entire inspiratory capacity, completely filling the lungs. The time period over which to execute the inhale is determined by the period. 
- <b>Breath Hold</b>: The time period to hold breath is determined by the period. 
- <b>Use %Inhaler</b>: The %Inhaler command is interpreted by the inhaler equipment (@ref InhalerMethodology).

Conscious respiration has any number of potential applications and is likely to be implemented to attain proper breathing while using an inhaler, generate a spirometry curve, or simulate coughing.  Figure 25 shows the %BioGears results for a cough scenario that leverages the conscious respiration action compared to empirical data.

<center>
<table border="0">
<tr>
    <td><img src="./plots/Respiratory/BioGears_Cough_Flow.jpg" width="550"></td>
    <td><img src="./plots/Respiratory/Experimental_Cough_Flow.jpg" width="550"></td>
</tr>
</table>
</center>
<center>
<i>Figure 25. The airflow curve of a simulated cough in %BioGears generated from a conscious respiration with quick forced inhale and exhale in series. For comparison, the figure includes a plot reproduced from 
literature determined by a voluntary cough immediately following office-based vocal fold medialization injections @cite ruddy2014improved.</i>
</center><br>

### Mechanical ventilation

Mechanical ventilation allows the user to specify an instantaneous pressure and/or flow value at the respiratory connection point (i.e., mouth).  The substance volume fractions at the connection can also be specified.  If no volume fractions are given, the ambient values set in the %Environment system will be used by default (see @ref EnvironmentMethodology).  All settings will remain constant during the simulation, unless removed or modified.  This action is likely to be most beneficial implemented in combination with real-world sensors.  Manikin or task trainer values can be fed into the the engine to synchronize in real-time.

@anchor respiratory-events
Events
------

### Maximum Pulmonary Ventilation Rate

The %BioGears %Respiratory system triggers Maximum Pulmonary Ventilation Rate event to alert the user when the patient pulmonary ventilation exceeds maximum value.

### Bradypnea

Bradypnea is defined as an abnormally low breathing rate. Normal breathing rates for healthy adults are relatively standard, but can vary based upon condition. Low breathing rates can decrease total ventilation if not compensated for, which can lead to hypercapnia or hypoxia. In the %BioGears Engine, bradypnea occurs when the respiratory rate falls below 10 breaths per minute @cite overdyk2007continuous and returns to normal when it rises back above 10.5 breaths per minute.

### Tachypnea

Tachypnea is defined as high breathing rate. Normal respiration rate in an adult ranges 12 - 20 breaths per minute. Tachypnea occurs when breathing rate rises above 20 breaths per minute. The  %BioGears Engine engine outputs a tachypnea event when the patient's respiration rate is above 20 breaths per minute. 

### %Respiratory Acidosis

%Respiratory acidosis is triggered when the blood pH drops below 7.37.  An irreversible state (similar to a death state) is reached when the blood pH drops below 6.5.  See @ref EnergyMethodology for more details about acidosis.

### %Respiratory Alkalosis

%Respiratory alkalosis is triggered when the blood pH rises above 7.45.  An irreversible state (similar to a death state) is reached when the blood pH rises above 8.5.  See @ref EnergyMethodology for more details about alkalosis.

### Acute %Respiratory Distress

There are three levels of Acute %Respiratory Distress (ARD) - mild, moderate, and severe.  These levels are based on Carrico Index thresholds that provide insight into the effectiveness of the alveolar transfer of oxygen. 

@anchor respiratory-assessments
Assessments
-----------

### Pulmonary Function Test

The pulmonary function test allows the user to request a collection of respiratory data at a specified time step. This data is then used to produce a waveform representing normal inspiration and expiration. The waveform period is dictated by the respiration rate at the time of the pulmonary function test. The waveform magnitude is calculated from the tidal volume at the specified time and the functional residual capacity. The waveform is calculated for an artificial time of 60 seconds. This time is neither engine time nor real time, but instead is an assumed time, and the complete 60 second duration is calculated at the current engine time step. This does not indicate the duration of an actual pulmonary function test, but it does allow visualization of many breathing cycles. Within two periods of the 60 second duration, the waveform magnitude is modified to produce the inspiratory and expiratory reserve volume. The inspiratory and expiratory reserve volumes are presented as a forced inspiration and expiration. To model this effect, the period of the breathing cycle is dilated during the magnitude increase. At the conclusion of the imposed 60 second duration, a lung volume plot is produced  (see the figure in @ref respiratory-assessmentvalidation "Validation - Assessments") that displays the generated lung volume waveform.

@anchor respiratory-results
Results and Conclusions
=======================

The %BioGears %Respiratory System was validated quantitatively and qualitatively under resting physiological conditions and transient (action-induced) conditions. The validation is specified with a color coding system, with green indicating good agreement with trends/values, yellow indicating moderate agreement with trends/values, and red indicating poor agreement with trends/values.

@anchor respiratory-validation
Validation - Resting Physiologic State
--------------------------------------

The %BioGears %Respiratory Model outputs a number of system-level and compartment-level resting physiologic parameters. The tables below compare the values of system- and compartment-level parameters obtained from the model with referenced values. The outputs from the model correspond to the system- and compartment-level respiratory related values of the standard patient (77kg adult male)  under resting physiologic conditions. As shown in Table 1, the majority of the physiological variables show a good match with the literature. The outputs for the major respiratory variables, such as respiration rate, tidal volume, and pulmonary ventilation specifically match well with the reference data. 

<center>
<i>Table 1. Validation of the resting physiologic state of the %Respiratory System. The table shows comparison of system-level outputs from %BioGears to referenced values. System-level outputs show favorable agreement with validation data. The deviations in end tidal carbon dioxide fraction and transpulmonary pressure can be attributed to the tuned parameters used in the model that are selected to meet the major system level physiological parameters.</i>
</center><br>

@insert doc/validation/RespiratoryValidationTable.md

<center>
<i>Table 2. Validation table for the resting physiologic states. The table shows comparison of compartment-level outputs from %BioGears to referenced values. The majority of the variables show good match with the validation data. There are significant deviations in the carina flow and the dead space volumes. Future versions will address these issues.</i>
</center><br>

@insert doc/validation/RespiratoryCompartmentsValidationTable.md

@anchor respiratory-scenario
Scenario Validation
-------------------

The actions and interventions associated with the %Respiratory System were validated quantitatively where possible and qualitatively elsewhere by comparing the engine output to expected trends and values. For each scenario, the table shows the total number of results in each category. For many investigated scenarios, the model shows good agreement with the expected trends. For the scenarios that did not match with the expected trends, improvements are planned for future %BioGears Engine releases.

<center><br>
*Table 3. Cumulative validation results for %Respiratory specific conditions and actions scenarios.*
</center>

|	Key	|
|	---	|
|<span class="success">	Good agreement: correct trends or <10% deviation from expected	</span>|
|<span class="warning"> 	Some deviation: correct trend and/or <30% deviation from expected	</span>|
|<span class="danger">	Poor agreement: incorrect trends or >30% deviation from expected	</span>|

|	Scenario 	|	Description	|	Good	|	Decent	|	Bad	|
|	---	|	---	|	---	|	---	|	---	|
|	AcuteRespiratoryDistress	|	Levels of respiratory distress from acute lung injury to severe ARDS	|<span class="success">	28	</span>|<span class="warning">	2	</span>|<span class="danger">	0	</span>|
|	AcuteRespiratoryDistress_Ventilation	|	Treat ARDS by putting patient on ventilator	|<span class="success">	14	</span>|<span class="warning">	1	</span>|<span class="danger">	0	</span>|
|	AsthmaAttackModerateAcute	|	Moderate acute asthma attack	|<span class="success">	18	</span>|<span class="warning">	0	</span>|<span class="danger">	0	</span>|
|	AsthmaAttackSevereAcute	|	Severe acute asthma attack	|<span class="success">	15	</span>|<span class="warning">	3	</span>|<span class="danger">	0	</span>|
|	AsthmaAttackLifeThreateningAcute	|	Life threatening acute asthma attack	|<span class="success">	17	</span>|<span class="warning">	1	</span>|<span class="danger">	0	</span>|
|	COPDSevereEmphysema	|	GOLD Stage III Emphysema	|<span class="success">	4	</span>|<span class="warning">	5	</span>|<span class="danger">	1	</span>|
|	COPDSevereBronchitisLeft	|	Severe Chronic Bronchitis	|<span class="success">	4	</span>|<span class="warning">	5	</span>|<span class="danger">	1	</span>|
|	LobarPneumoniaSevereLeftLobe	|	Severe Lobar Pneumonia in one lobe in the left lung	|<span class="success">	5	</span>|<span class="warning">	2	</span>|<span class="danger">	2	</span>|
|	LobarPneumoniaSevereRightLung	|	Severe Lobar Pneumonia in two lobes of right lung	|<span class="success">	7	</span>|<span class="warning">	0	</span>|<span class="danger">	2	</span>|
|	LobarPneumoniaModerateBothLungs	|	Moderate Lobar Pneumonia in both lungs	|<span class="success">	6	</span>|<span class="warning">	0	</span>|<span class="danger">	3	</span>|
|	TensionPneumothoraxOpenVaried	|	Varied open pneumothorax severities and interventions	|<span class="success">	34	</span>|<span class="warning">	4	</span>|<span class="danger">	4	</span>|
|	TensionPneumothoraxClosedVaried	|	Varied closed pneumothorax severities and interventions	|<span class="success">	26	</span>|<span class="warning">	4	</span>|<span class="danger">	4	</span>|
|	TensionPneumothoraxBilateral	|	Open and closed bilateral pneumothorax and interventions	|<span class="success">	23	</span>|<span class="warning">	3	</span>|<span class="danger">	2	</span>|
|	AirwayObstructionVaried	|	Airway Obstruction with varying severities	|<span class="success">	30	</span>|<span class="warning">	0	</span>|<span class="danger">	0	</span>|
|	Bronchoconstriction	|	Bronchoconstriction with varying severities	|<span class="success">	25	</span>|<span class="warning">	1	</span>|<span class="danger">	4	</span>|
|	MainstemIntubation	|	Right and left mainstem intubation and correction (with Succs)	|<span class="success">	25	</span>|<span class="warning">	0	</span>|<span class="danger">	0	</span>|
|	EsophagealIntubation	|	Esophageal intubation and correction (with Succs)	|<span class="success">	15	</span>|<span class="warning">	0	</span>|<span class="danger">	0	</span>|
|	Apnea	|	Varied severities of respiratory apnea	|<span class="success">	6	</span>|<span class="warning">	0	</span>|<span class="danger">	0	</span>|
|		|	Total	|<span class="success">	302	</span>|<span class="warning">	31	</span>|<span class="danger">	23	</span>|


@anchor respiratory-conditionvalidation
Validation - Conditions
-----------------------

### COPD

The COPD condition was validated against two scenarios. The severe emphysema scenario (Emphysema Severity = 0.7, Bronchitis Severity = 0.5) attempts to model Gold Stage III Emphysema. Here, most metrics show moderate to good agreement with trends, but a few poor results were obtained. We plan to investigate these issues in later versions.

<center><br>
<i>Table 4. Validation matrix for severe emphysema. This condition corresponds to GOLD Stage III COPD  @cite man2003contemporary  @cite perez2009can . The table shows the %BioGears engine output compared to validation data for respiratory and hemodynamic values.</i>
</center>

|	Segment	|	Notes	|	Sampled Scenario Time (s)	|	Trachea Flow - Peak Expiratory Flow  (L/min)	|	Respiration Rate (breaths/min)	|	Tidal Volume (mL)	|	Heart Rate (beats/min)	|	Systolic Pressure (mmHg)	|	Oxygen Saturation	|	PaO2 (mmHg)	|	PaCO2 (mmHg)	|	IERatio	|	Aorta Lactate (mg/dL)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Severe Emphysema: Bronchitis Severity = 0.5 Emphysema Severity = 0.7 Both Lungs 100%	|	GOLD Stage III	|	120	|<span class="success">	< 70% of Normal @cite perez2009can	</span>|<span class="success">	Increase,  Tachypnea, Dypsnea,  > 30 @cite gunning2003pathophysiology	</span>|<span class="success">	< 60% of normal @cite gunning2003pathophysiology	</span>|<span class="warning">	Increase, Tachycardia	</span>|<span class="danger">	Increase,   Pulmonary Hypertension ,  > 140 mm Hg @cite keller2003pathophysiology, @cite scharf2002hemodynamic	</span>|<span class="warning">	< 90%  @cite man2003contemporary; < 89% @cite keller2003pathophysiology	</span>|<span class="warning">	Decrease,  Hypoxemia @cite keller2003pathophysiology; < 55 mm Hg  @cite man2003contemporary, @cite keller2003pathophysiology	</span>|<span class="warning">	Increase, Hypercapnia, > 55 mmHg @cite keller2003pathophysiology	</span>|<span class="success">	Decrease @cite van1991physical	</span>|<span class="warning">	Increase @cite mathur1999cerebral	</span>|

<center><br>
<i>Table 5. Validation matrix for severe chronic bronchitis. The table shows the %BioGears engine output compared to validation data for respiratory and hemodynamic values.</i>
</center>

|	Segment	|	Notes	|	Sampled Scenario Time (s)	|	Trachea Flow - Peak Expiratory Flow  (L/min)	|	Respiration Rate (breaths/min)	|	Tidal Volume (mL)	|	Heart Rate (beats/min)	|	Systolic Pressure (mmHg)	|	Oxygen Saturation	|	PaO2 (mmHg)	|	PaCO2 (mmHg)	|	IERatio	|	Aorta Lactate (mg/dL)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Severe Chronic Bronchitis: Bronchitis Severity = 0.65 Emphysema Severity = 0.5 Both Lungs 100%	|		|	120	|<span class="success">	Decrease @cite bergeronSME	</span>|<span class="success">	> 20 @cite bergeronSME	</span>|<span class="success">	< 60% of normal @cite bergeronSME	</span>|<span class="warning">	Increase,  Tachycardia @cite bergeronSME	</span>|<span class="danger">	Increase,  Pulmonary Hypertension,  > 140 mm Hg @cite bergeronSME	</span>|<span class="warning">	< 90% @cite bergeronSME	</span>|<span class="warning">	Decrease,  Hypoxemia,  < 55 mm Hg  @cite bergeronSME	</span>|<span class="warning">	Increase,   Hypercapnia > 55 mmHg   @cite bergeronSME	</span>|<span class="success">	Decrease @cite van1991physical, @cite bergeronSME	</span>|<span class="warning">	Increase @cite mathur1999cerebral	</span>|

### Lobar Pneumonia

The lobar pneumonia condition was validated against three scenarios; severe pneumonia in the one lobe of the left lung (Severity = 0.7), severe pneumonia in two lobes of the right lung (Severity = 0.7), and moderate pneumonia in both lungs (Severity = 0.2). %BioGears does not model the discrete lobes of each lung, so infected-lobe behavior is modeled by applying pneumonia severity against a fraction of each lung. The fraction approximates the portion of the lung (by volume) that the lobe(s) occupy. In the case of the left lung, there are two lobes. Each lobe constitutes approximately 50% of the whole lung. It is understood that the actual volume distribution is not equal for the two lobes, but we use this as an approximation. The same is true for the right lung, which has three lobes. In all three scenarios, expected general trends are observed for most measures. However, heart rate and blood pressure do not follow expected trends. This is similar to COPD, where heart rate and blood pressure also do not always behave as expected. Additionally, increasing core body temperature is also a symptom of lobar pneumonia that fails to agree with the validation data. However, we expected this disagreement since %BioGears doesn't currently model elevated body temperature in response to infection. This is something we plan to address in subsequent releases. 
<center><br>
<i>Table 6. Validation matrix for severe lobar pneumonia concentrated in one lobe in the left lung. The table shows the %BioGears engine output compared to validation data for respiratory and hemodynamic values.</i>
</center>

|	Condition	|	Notes	|	Sampled Scenario Time (s)	|	Respiration Rate (breaths/min)	|	Tidal Volume (mL)	|	Heart Rate (beats/min)	|	Systolic Pressure (mmHg)	|	Oxygen Saturation	|	PaO2 (mmHg)	|	PaCO2 (mmHg)	|	IERatio	|	Core Body Temperature  (Degrees C)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Severe Lobar Pneumonia:  Severity = 0.70 Left Lung = 50% Right Lung = 0%	|	Severe Lobar Pneumonia in one lobe in the left lung. Current model does not include metabolic effects.	|	120	|<span class="success">	Increase,  Tachypnea, Dypsnea  > 20 @cite ebell2007predicting	</span>|<span class="success">	Decrease @cite bergeronSME	</span>|<span class="warning">	Increase,  Tachycardia, > 100 @cite ebell2007predicting	</span>|<span class="danger">	Decrease, < 90 mm Hg @cite fine1997prediction	</span>|<span class="warning">	Decrease , < 95% @cite majumdar2010oxygen	</span>|<span class="danger">	Decrease,  Hypoxemia @cite fine1997prediction < 60 mm Hg	</span>|<span class="success">	Increase @cite bergeronSME	</span>|<span class="success">	Decrease @cite bergeronSME	</span>|<span class="danger">	Increase,   > 37.8 C (100 F) @cite ebell2007predicting	</span>|

<center><br>
<i>Table 7. Validation matrix for severe lobar pneumonia concentrated in two lobes in the right lung. The table shows the %BioGears engine output compared to validation data for respiratory and hemodynamic values.</i>
</center>

|	Condition	|	Notes	|	Sampled Scenario Time (s)	|	Respiration Rate (breaths/min)	|	Tidal Volume (mL)	|	Heart Rate (beats/min)	|	Systolic Pressure (mmHg)	|	Oxygen Saturation	|	PaO2 (mmHg)	|	PaCO2 (mmHg)	|	IERatio	|	Core Body Temperature  (Degrees C)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Severe Lobar Pneumonia: Right Lung Severity = 0.70 Left Lung = 0% Right Lung  = 67%	|	Severe lobar pneumonia in two lobes of the  right lung. Current model does not include metabolic effects.	|	120	|<span class="success">	Increase,  Tachypnea, Dypsnea  > 20 @cite ebell2007predicting	</span>|<span class="success">	Decrease @cite bergeronSME	</span>|<span class="warning">	Increase,  Tachycardia, > 100 @cite ebell2007predicting	</span>|<span class="danger">	Decrease, < 90 mm Hg @cite fine1997prediction	</span>|<span class="success">	Decrease , < 95% @cite majumdar2010oxygen	</span>|<span class="warning">	Decrease,  Hypoxemia @cite fine1997prediction < 60 mm Hg	</span>|<span class="success">	Increase @cite bergeronSME	</span>|<span class="success">	Decrease @cite bergeronSME	</span>|<span class="danger">	Increase,   > 37.8 C (100 F) @cite ebell2007predicting	</span>|

<center><br>
<i>Table 8. Validation matrix for moderate lobar pneumonia in both lungs. The table shows the %BioGears engine output compared to validation data for respiratory and hemodynamic values.</i>
</center>

|	Condition	|	Notes	|	Sampled Scenario Time (s)	|	Respiration Rate (breaths/min)	|	Tidal Volume (mL)	|	Heart Rate (beats/min)	|	Systolic Pressure (mmHg)	|	Oxygen Saturation	|	PaO2 (mmHg)	|	PaCO2 (mmHg)	|	IERatio	|	Core Body Temperature  (Degrees C)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Moderate Lobar Pneumonia: Severity = 0.2 Left Lung = 100% Right Lung  = 100%	|	Moderate pneumonia in both lungs. Current model does not include metabolic effects.	|	120	|<span class="success">	Increase,  Tachypnea, Dypsnea  > 20 @cite ebell2007predicting	</span>|<span class="success">	Decrease @cite bergeronSME	</span>|<span class="danger">	Increase,  Tachycardia, > 100 @cite ebell2007predicting	</span>|<span class="danger">	Decrease, < 90 mm Hg @cite fine1997prediction	</span>|<span class="warning">	Decrease , < 95% @cite majumdar2010oxygen	</span>|<span class="danger">	Decrease,  Hypoxemia @cite fine1997prediction < 60 mm Hg	</span>|<span class="success">	Increase @cite bergeronSME	</span>|<span class="success">	Decrease @cite bergeronSME	</span>|<span class="danger">	Increase,   > 37.8 C (100 F) @cite ebell2007predicting	</span>|

@anchor respiratory-actionvalidation
Validation - Actions
--------------------
### Acute Respiratory Distress
The respiratory distress action was validated by applying varying severities to target clinical symptoms progressing from acute lung injury (ALI) to acute respiratory distress syndrome (ARDS) to severe ARDS.  It was found that severity of 0.5 induces ALI, with lesser severities only causing minor respiratory dysfunction.  Severities greater than 0.5 cause the desired progression towards severe ARDS.  As expected, the patient carrico index (ratio of PaO2 to FiO2) deceased as the severity of respiratory distress increased.  However, a carrico index < 100 mmHg was not observed because patient oxygen levels declined too steeply, triggering a BioGears irreversible state.  Mean pulmonary arterial pressure increased as a function of severity consistent with diagnosis of ALI and ARDS but, as was the case with the carrico index, did not reach the target value for severe ARDS.  The nervous system responded to worsening hypoxemia by inducing tachypnea and tachydardia.

In the ARDS treatment scenario, mechanical ventilation successfully increased carrico index and oxygen saturation above the threshold for ALI.  The cardiovascular effects of propofol induction were apparent in the decrease in mean arterial pressure and heart rate. Both of these metrics, however, remained above baseline physiological values and would require additional intervention to reverse.

<center>
<table border="0">
<tr>
    <td><img src="./plots/Respiratory/ARDS_CarricoIndex.jpg" width="550"></td>
    <td><img src="./plots/Respiratory/ARDS_RespirationRate.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/Respiratory/ARDS_MeanPulmonaryPressure.jpg" width="550"></td>
    <td><img src="./plots/Respiratory/ARDS_O2Saturation.jpg" width="550"></td>
</tr>
</table>
</center>
<center><i>Figure 26. Select outputs from the acute respiratory distress scenario.</i></center>

<center><br>
<i>Table 9. Validation matrix for physiological responses due to varying severities of respiratory distress.</i>
</center>
|	Segment	|	Notes	|	Action Occurrence Time (min)	|	Carrico Index (mmHg)	|	Mean Pulmonary Arterial Pressure (mmHg)	|	Respiration Rate (/min)	|	Heart Rate (/min)	|	Oxygen Saturation	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Respiratory distress, severity = 0.5	|	Target acute lung injury (ALI)	|	0	|<span class="success">	 < 300  @cite varisco2011pharmacology	</span>|<span class="success">	> 25 @cite ryan2014pulmonary	</span>|<span class="success">	Tachypnea @cite matuschak2010acute	</span>|<span class="success">	Tachycardia @cite matuschak2010acute	</span>|<span class="success">	Decrease @cite matuschak2010acute	</span>|
|	Remove action	|		|	10	|<span class="success">	Normal (~450)	</span>|<span class="success">	<25	</span>|<span class="success">	returns to resting physiology 	</span>|<span class="success">	returns to resting physiology 	</span>|<span class="success">	returns to resting physiology 	</span>|
|	Respiratory distress, severity = 0.75	|	Target acute respiratory distress syndrome	|	30	|<span class="success">	<200 @cite varisco2011pharmacology	</span>|<span class="success">	>30 @cite ryan2014pulmonary	</span>|<span class="success">	Tachypnea @cite matuschak2010acute	</span>|<span class="success">	Tachycardia @cite matuschak2010acute	</span>|<span class="success">	< 0.9  @cite matuschak2010acute	</span>|
|	Remove action	|		|	40	|<span class="success">	Normal (~450)	</span>|<span class="success">	<25	</span>|<span class="success">	returns to resting physiology 	</span>|<span class="success">	returns to resting physiology 	</span>|<span class="success">	returns to resting physiology 	</span>|
|	Respiratory distress, severity = 1.0	|	Target severe acute respiratory distress	|	60	|<span class="warning">	<100 @cite varisco2011pharmacology	</span>|<span class="warning">	>45 @cite ryan2014pulmonary	</span>|<span class="success">	Tachypnea @cite matuschak2010acute	</span>|<span class="success">	Tachycardia @cite matuschak2010acute	</span>|<span class="success">	< 0.9  @cite matuschak2010acute	</span>|
|	Remove action	|		|	70	|<span class="success">	Normal (~450)	</span>|<span class="success">	<25	</span>|<span class="success">	returns to resting physiology 	</span>|<span class="success">	returns to resting physiology 	</span>|<span class="success">	returns to resting physiology 	</span>|

<center>
<table border="0">
<tr>
    <td><img src="./plots/Respiratory/ARDSVent_CarricoIndex.jpg" width="550"></td>
    <td><img src="./plots/Respiratory/ARDSVent_RespirationRate.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/Respiratory/ARDSVent_MeanPulmonaryPressure.jpg" width="550"></td>
    <td><img src="./plots/Respiratory/ARDSVent_O2Saturation.jpg" width="550"></td>
</tr>
</table>
</center>
<center><i>Figure 27. Select outputs from the acute respiratory distress with ventilation.</i></center>


<center><br>
<i>Table 10. Validation matrix for treatment of ARDS with ventilation.</i>
</center>
|	Segment	|	Notes	|	Action Occurrence Time (min)	|	Carrico Index (mmHg)	|	Mean Pulmonary Arterial Pressure (mmHg)	|	Respiration Rate (/min)	|	Heart Rate (/min)	|	Oxygen Saturation	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Respiratory distress, severity = 0.75	|	Target ARDS	|	0	|<span class="success">	 < 200  @cite varisco2011pharmacology	</span>|<span class="success">	> 30 @cite ryan2014pulmonary	</span>|<span class="success">	Tachypnea @cite matuschak2010acute	</span>|<span class="success">	Tachycardia @cite matuschak2010acute	</span>|<span class="success">	Decrease @cite matuschak2010acute	</span>|
|	Anesthesia Induction w/ Propofol	|	1.8 mg/kg loading dose followed by 142 ug/kg/min maintenance	|	30	|<span class="success">	Steady	</span>|<span class="success">	Decrease @cite Morgan2006Clinical	</span>|<span class="success">	Decrease @cite Morgan2006Clinical	</span>|<span class="success">	No change @cite Morgan2006Clinical	</span>|<span class="success">	Decrease	</span>|
|	Connect anesthesia machine	|	InletFlow = 6.0 L/min; FiO2 = 0.5; PEEP = 1.0 cmH2O; PMax = 10.0 cmH2O; IE ratio = 0.5; Respiration Rate = 12	|	33	|<span class="success">	>300 @cite varisco2011pharmacology	</span>|<span class="success">	Minimal change @cite bein2016standard @cite varisco2011pharmacology	</span>|<span class="success">	12 (machine setting)	</span>|<span class="warning">	Decrease	</span>|<span class="success">	> 0.9 @cite bein2016standard	</span>|


### Airway Obstruction

The airway obstruction action was validated with a scenario that applies an airway obstruction action with varying severities, 0.3, 0.6, and 1.0. Varying the severity of the obstruction simulates partial obstructions and a complete obstruction. The severity levels logarithmically increase the airway resistance, as shown in Figure 22. 

The resulting outputs from the %BioGears engine are shown in Table 11 for common clinical parameters, such as heart rate, respiratory rate, and oxygen saturation. In the event of complete airway obstruction, a drop in oxygen saturation and respiration rate are expected. In contrast, hemodynamic variables, such as heart rate, are expected to increase. 

<center>
<table border="0">
<tr>
    <td><img src="./plots/Respiratory/AirwayObstruction_TidalVolume.jpg" width="550"></td>
    <td><img src="./plots/Respiratory/AirwayObstruction_CO2PP.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/Respiratory/AirwayObstruction_O2Sat.jpg" width="550"></td>
    <td><img src="./plots/Respiratory/AirwayObstruction_MAP.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/Respiratory/AirwayObstructionLegend.jpg" width="1100"></td>
</tr>
</table>
</center>
<center><i>Figure 28. Select outputs from the airway obstruction scenario.</i></center>
 
<center><br>
<i>Table 11. Validation matrix for physiological responses due to varying severities of airway obstruction.</i>
</center>

|	Segment	|	Notes	|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Respiration Rate (breaths/min)	|	Heart Rate (beats/min)	|	Systolic Pressure (mmHg)	|	Diastolic Pressure (mmHg)	|	Oxygen Saturation	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Trigger airway obstruction;  severity = 0.3	|	Represents the partial (severity = 0.3) obstruction of airway due to  foreign body (obstructing material). Not severe enough to trigger CV responses.	|	30	|	180	|<span class="success">	 Increase  @cite Morgan2006Clinical	</span>|<span class="success">	Increase or no change @cite bergeronSME	</span>|<span class="success">	Increase or no change @cite bergeronSME(case study of awake patient @cite izakson2013complete)	</span>|<span class="success">	Increase @cite bergeronSME (case study of awake patient @cite izakson2013complete)	</span>|<span class="success">	Decrease @cite Morgan2006Clinical	</span>|
|	Remove airway obstruction	|	Airway obstruction is turned off.	|	210	|	300	|<span class="success">	returns to resting physiology 	</span>|<span class="success">	returns to resting physiology 	</span>|<span class="success">	returns to resting physiology 	</span>|<span class="success">	returns to resting physiology 	</span>|<span class="success">	returns to resting physiology 	</span>|
|	Turn on airway obstruction;  severity = 0.6	|	Represents the partial obstruction (severity = 0.6) of airway due to  foreign body.	|	510	|	180	|<span class="success">	Increase > Severity 0.3  @cite Morgan2006Clinical	</span>|<span class="success">	Increase @cite bergeronSME	</span>|<span class="success">	Increase @cite bergeronSME  (case study of awake patient @cite izakson2013complete)	</span>|<span class="success">	Increase @cite bergeronSME (case study of awake patient @cite izakson2013complete)	</span>|<span class="success">	Decrease < Severity 0.3 @cite Morgan2006Clinical	</span>|
|	Remove airway obstruction	|	Airway obstruction is turned off.	|	690	|	300	|<span class="success">	returns to resting physiology 	</span>|<span class="success">	returns to resting physiology 	</span>|<span class="success">	returns to resting physiology 	</span>|<span class="success">	returns to resting physiology 	</span>|<span class="success">	returns to resting physiology 	</span>|
|	Turn on airway obstruction;  severity = 1.0	|	Represents the complete obstruction (severity = 1.0) of airway due to  foreign body.	|	990	|	180	|<span class="success">	Increase  > Severity 0.6 @cite Morgan2006Clinical	</span>|<span class="success">	Increase @cite bergeronSME	</span>|<span class="success">	Increase @cite bergeronSME  (case study of awake patient @cite izakson2013complete)	</span>|<span class="success">	Increase @cite bergeronSME  (case study of awake patient @cite izakson2013complete)	</span>|<span class="success">	Decrease < Severity 0.6 @cite Morgan2006Clinical	</span>|
|	Remove airway obstruction	|	Airway obstruction is turned off.	|	1170	|	500	|<span class="success">	returns to resting physiology 	</span>|<span class="success">	returns to resting physiology 	</span>|<span class="success">	returns to resting physiology 	</span>|<span class="success">	returns to resting physiology 	</span>|<span class="success">	returns to resting physiology 	</span>|

### Bronchoconstriction

The bronchoconstriction action was validated with a scenario that applies a series of bronchoconstrictions with varying severities, 0.3, 0.6, and 1.0 to the patient. Varying the severity of the obstruction simulates the constriction of the bronchii, leading to obstructed air flow. The severity levels logarithmically increase the airway resistance, as shown in Figure 22.

<center>
<table border="0">
<tr>
    <td><img src="./plots/Respiratory/BronchoConstrictionVaried_TidalVolume.jpg" width="550"></td>
    <td><img src="./plots/Respiratory/BronchoConstrictionVaried_CO2PP.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/Respiratory/BronchoConstrictionVaried_O2Sat.jpg" width="550"></td>
    <td><img src="./plots/Respiratory/BronchoConstrictionVaried_MAP.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/Respiratory/BronchoConstrictionVariedLegend.jpg" width="1100"></td>
</tr>
</table>
</center>
<center><i>Figure 29. Select outputs from the bronchoconstriction scenario.</i></center>

<center><br>
<i>Table 12. Validation matrix for physiological responses due to varying severities of bronchoconstriction.</i>
</center>

|	Segment	|	Notes	|	Action Occurrence Time  (s)	|	Sampled Scenario Time  (s)	|	Respiration Rate (breaths/min)	|	Heart Rate (beats/min)	|	Systolic Pressure (mmHg)	|	Diastolic Pressure (mmHg)	|	Oxygen Saturation	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Trigger Bronchoconstriction; severity 0.3	|	Trigger bronchoconstriction of severity 0.3 by Increasing effective breathing resistances on bronchi. Not severe enough to trigger CV responses.	|	30	|	180	|<span class="warning">	Decrease  @cite Morgan2006Clinical	</span>|<span class="success">	No change	</span>|<span class="success">	No change	</span>|<span class="success">	No change	</span>|<span class="success">	Decrease  @cite Morgan2006Clinical	</span>|
|	Remove bronchoconstriction	|	Bronchoconstriction is turned off.	|	210	|	300	|<span class="success">	returns to resting physiology 	</span>|<span class="success">	returns to resting physiology 	</span>|<span class="success">	returns to resting physiology 	</span>|<span class="success">	returns to resting physiology 	</span>|<span class="success">	returns to resting physiology 	</span>|
|	Trigger Bronchoconstriction;   severity = 0.6	|	Trigger bronchoconstriction of severity 0.6 by Increasing effective breathing resistances on bronchi.	|	510	|	180	|<span class="success">	Increase > Severity 0.3  @cite Morgan2006Clinical	</span>|<span class="success">	Increase  tachycardia  @cite faust2002anesthesiology, @cite Morgan2006Clinical	</span>|<span class="danger">	Decrease hypotension  @cite faust2002anesthesiology	</span>|<span class="danger">	Decrease hypotension  @cite faust2002anesthesiology	</span>|<span class="success">	Decrease < Severity 0.3   @cite Morgan2006Clinical	</span>|
|	Remove bronchoconstriction	|	Bronchoconstriction is turned off.	|	690	|	300	|<span class="success">	returns to resting physiology 	</span>|<span class="success">	returns to resting physiology 	</span>|<span class="success">	returns to resting physiology 	</span>|<span class="success">	returns to resting physiology 	</span>|<span class="success">	returns to resting physiology 	</span>|
|	Trigger Bronchoconstriction;   severity = 1.0	|	Trigger bronchoconstriction of severity 1.0 by Increasing effective breathing resistances on bronchi.	|	990	|	180	|<span class="success">	Increase  > Severity 0.6 @cite Morgan2006Clinical	</span>|<span class="success">	Increase  tachycardia  @cite faust2002anesthesiology, @cite Morgan2006Clinical	</span>|<span class="danger">	Decrease hypotension  @cite faust2002anesthesiology	</span>|<span class="danger">	Decrease hypotension  @cite faust2002anesthesiology	</span>|<span class="success">	Decrease < Severity 0.6   @cite Morgan2006Clinical	</span>|
|	Remove bronchoconstriction	|	Bronchoconstriction is turned off.	|	1170	|	500	|<span class="success">	returns to resting physiology 	</span>|<span class="success">	returns to resting physiology 	</span>|<span class="success">	returns to resting physiology 	</span>|<span class="success">	returns to resting physiology 	</span>|<span class="success">	returns to resting physiology 	</span>|

### Pneumothorax

Three scenarios were used to validate tension pneumothorax and its associated interventions, occlusive dressing, and needle decompression. Table 13 demonstrates the open tension pneumothorax on the left side of the chest at increasing severities treated with an occlusive dressing and needle decompression. The %BioGears engine output for relevant parameters was compared to the validation data. The blood pressures (systolic, diastolic, and mean arterial pressure) do not show hypotension during the open wound pneumothorax incidence. Other respiratory and cardiovascular outputs show good agreement with published data. Vital signs become more dire the longer the pneumothorax is applied without intervention and as the severity (i.e. hole size) is increased. Evidence suggests that patients have progressive respiratory deterioration with final respiratory arrest @cite leigh2005tension. However, the current model does not include a definition for respiratory arrest and will continue to run until another irreversible state is reached.

After an occlusive dressing is added to seal the hole and needle decompression is applied, the patient's condition improves and vitals begin returning toward normal.  This is as expected compared to the validation data. 

It is important to note nervous system responses of a conscious patient due to pain or panicking are not included with the Pneumothorax action. Some of these responses can be attained by including an Acute Stress action (see @ref endocrine-actions "Endocrine Methodology").

<center>
<table border="0">
<tr>
    <td><img src="./plots/Respiratory/TensionPneumothoraxOpenVaried_TotalLungVolume.jpg" width="550"></td>
    <td><img src="./plots/Respiratory/TensionPneumothoraxOpenVaried_LeftLungVolume.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/Respiratory/TensionPneumothoraxOpenVaried_O2Sat.jpg" width="550"></td>
	<td><img src="./plots/Respiratory/TensionPneumothoraxOpenVaried_MAP.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/Respiratory/TensionPneumothoraxOpenVariedLegend.jpg" width="1100"></td>
</tr>
</table>
</center>
<center><i>Figure 30. Select outputs from the open tension pneumothorax scenario.</i></center>

<center><br>
<i>Table 13. Validation matrix for physiological responses due to varying severities of open tension pneumothorax, as well as occlusive dressing and needle decompression interventions.</i>
</center>

|	Segment	|	Notes	|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Respiration Rate (breaths/min)	|	Tidal Volume (mL)	|	Heart Rate (beats/min)	|	Systolic Pressure (mmHg)	|	Diastolic Pressure (mmHg)	|	Mean Arterial Pressure (mmHg)	|	Oxygen Saturation	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Tension Pneumothorax;  severity = 0.0	|	A severity of zero should not change the leak resistance from the default open switch value.	|	0	|	15	|<span class="success">	No Change 	</span>|<span class="success">	No Change 	</span>|<span class="success">	No Change 	</span>|<span class="success">	No Change 	</span>|<span class="success">	No Change 	</span>|<span class="success">	No Change 	</span>|<span class="success">	No Change 	</span>|
|	Tension Pneumothorax;  severity = 0.3	|	Represents a "sucking" chest wound on the left lung side with a mild severity (medium size hole).	|	30	|	300	|<span class="success">	 No change or Increase  @cite bond200968w, @cite leigh2005tension 	</span>|<span class="danger">	 Increase  @cite bond200968w, @cite leigh2005tension 	</span>|<span class="success">	 No change or Increase  @cite bond200968w, @cite leigh2005tension 	</span>|<span class="success">	 No change or Decrease  @cite bond200968w, @cite leigh2005tension 	</span>|<span class="danger">	 No change or Decrease  @cite bond200968w, @cite leigh2005tension 	</span>|<span class="success">	 No change or Decrease  @cite bond200968w, @cite leigh2005tension 	</span>|<span class="success">	 No change or Decrease  @cite bond200968w, @cite leigh2005tension 	</span>|
|	Tension Pneumothorax;  severity = 0.6	|	Represents a "sucking" chest wound on the left lung side with a moderate severity (large size hole).	|	360	|	600	|<span class="success">	 Increase @cite bond200968w, @cite leigh2005tension 	</span>|<span class="success">	Increase @cite leigh2005tension	</span>|<span class="success">	Increase @cite bond200968w, @cite leigh2005tension	</span>|<span class="success">	Decrease @cite bond200968w, @cite leigh2005tension	</span>|<span class="danger">	Decrease @cite bond200968w, @cite leigh2005tension	</span>|<span class="success">	Decrease @cite bond200968w, @cite leigh2005tension	</span>|<span class="success">	Decrease @cite bond200968w, @cite leigh2005tension	</span>|
|	Tension Pneumothorax;  severity = 1.0	|	Represents a "sucking" chest wound on the left lung side with the severe severity (maximum size hole).  Should lead to full collapse.	|	660	|	900	|<span class="success">	 Clinical sign: Tachypnea  @cite bond200968w, @cite leigh2005tension 	</span>|<span class="success">	Increase @cite leigh2005tension	</span>|<span class="success">	Clinical sign: Tachycardia  @cite bond200968w, @cite leigh2005tension	</span>|<span class="warning">	Clinical sign: Hypotension @cite bond200968w, @cite leigh2005tension	</span>|<span class="danger">	Clinical sign: Hypotension @cite bond200968w, @cite leigh2005tension	</span>|<span class="warning">	Clinical sign: Hypotension @cite bond200968w, @cite leigh2005tension	</span>|<span class="success">	Clinical sign: Hypoxia  @cite bond200968w, @cite leigh2005tension	</span>|
|	Apply Chest Occlusive dressing on the left side 	|	Represents the closing of chest wound on the left side	|	960	|	1000	|<span class="success">	No Change or Increased  @cite bergeronSME	</span>|<span class="success">	No Change 	</span>|<span class="success">	Tachycardia > 120 @cite bergeronSME	</span>|<span class="success">	Slightly improved to no change  @cite bergeronSME	</span>|<span class="success">	Slightly improved to no change  @cite bergeronSME	</span>|<span class="success">	Slightly improved to no change  @cite bergeronSME	</span>|<span class="success">	Modest decrease to no change  @cite bergeronSME	</span>|
|	Apply needle decompression	|	Needle decompression is applied on the left side of the chest	|	1020	|	1320	|<span class="success">	Returning toward normal @cite bergeronSME	</span>|<span class="warning">	Returning toward normal @cite bergeronSME	</span>|<span class="success">	Returning toward normal @cite bergeronSME	</span>|<span class="success">	Slightly diminished from normal @cite bergeronSME	</span>|<span class="warning">	Slightly diminished from normal @cite bergeronSME	</span>|<span class="success">	Slightly diminished from normal @cite bergeronSME	</span>|<span class="success">	Normal @cite bergeronSME	</span>|

<center>
<table border="0">
<tr>
    <td><img src="./plots/Respiratory/TensionPneumothoraxClosedVaried_TotalLungVolume.jpg" width="550"></td>
    <td><img src="./plots/Respiratory/TensionPneumothoraxClosedVaried_RightLungVolume.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/Respiratory/TensionPneumothoraxClosedVaried_O2Sat.jpg" width="550"></td>
    <td><img src="./plots/Respiratory/TensionPneumothoraxClosedVaried_MAP.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/Respiratory/TensionPneumothoraxClosedVariedLegend.jpg" width="1100"></td>
</tr>
</table>
</center>
<center><i>Figure 31. Select outputs from the closed tension pneumothorax scenario.</i></center>

<center><br>
<i>Table 14. Validation matrix for physiological responses due to varying severities of closed tension pneumothorax, as well as needle decompression interventions.</i>
</center>

|	Segment	|	Notes	|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Respiration Rate (breaths/min)	|	Tidal Volume (mL)	|	Heart Rate (beats/min)	|	Systolic Pressure (mmHg)	|	Diastolic Pressure (mmHg)	|	Mean Arterial Pressure (mmHg)	|	Oxygen Saturation	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Tension Pneumothorax;  severity = 0.0	|	A severity of zero should not change the leak resistance from the default open switch value.	|	0	|	15	|<span class="success">	No Change 	</span>|<span class="success">	No Change 	</span>|<span class="success">	No Change 	</span>|<span class="success">	No Change 	</span>|<span class="success">	No Change 	</span>|<span class="success">	No Change 	</span>|<span class="success">	No Change 	</span>|
|	Tension Pneumothorax;  severity = 0.3	|	Represents a leak on the right lung side with a mild severity (medium size hole).	|	30	|	300	|<span class="success">	 No change or Increase  @cite bond200968w, @cite leigh2005tension 	</span>|<span class="danger">	 Increase  @cite bond200968w, @cite leigh2005tension 	</span>|<span class="success">	 No change or Increase  @cite bond200968w, @cite leigh2005tension 	</span>|<span class="success">	 No change or Decrease  @cite bond200968w, @cite leigh2005tension 	</span>|<span class="danger">	 No change or Decrease  @cite bond200968w, @cite leigh2005tension 	</span>|<span class="success">	 No change or Decrease  @cite bond200968w, @cite leigh2005tension 	</span>|<span class="success">	 No change or Decrease  @cite bond200968w, @cite leigh2005tension 	</span>|
|	Tension Pneumothorax;  severity = 0.6	|	Represents a leak on the right lung side with a moderate severity (large size hole).	|	360	|	600	|<span class="success">	 Increase @cite bond200968w, @cite leigh2005tension 	</span>|<span class="success">	Increase @cite leigh2005tension	</span>|<span class="success">	Increase  @cite bond200968w, @cite leigh2005tension	</span>|<span class="success">	Decrease  @cite bond200968w, @cite leigh2005tension	</span>|<span class="danger">	Decrease @cite bond200968w, @cite leigh2005tension	</span>|<span class="success">	Decrease @cite bond200968w, @cite leigh2005tension	</span>|<span class="success">	Decrease @cite bond200968w, @cite leigh2005tension	</span>|
|	Tension Pneumothorax;  severity = 1.0	|	Represents a leak on the right lung side with the severe severity (maximum size hole).  Should lead to full collapse.	|	660	|	900	|<span class="success">	 Clinical sign: Tachypnea      @cite bond200968w, @cite leigh2005tension 	</span>|<span class="success">	Increase @cite leigh2005tension	</span>|<span class="success">	Clinical sign: Tachycardia   @cite bond200968w, @cite leigh2005tension	</span>|<span class="warning">	Clinical sign: Hypotension  @cite bond200968w, @cite leigh2005tension	</span>|<span class="danger">	Clinical sign: Hypotension  @cite bond200968w, @cite leigh2005tension	</span>|<span class="warning">	Clinical sign: Hypotension  @cite bond200968w, @cite leigh2005tension	</span>|<span class="success">	Clinical sign: Hypoxia   @cite bond200968w, @cite leigh2005tension	</span>|
|	Apply needle decompression	|	Needle decompression is applied on the right side of the chest	|	960	|	1260	|<span class="success">	Slightly lower tachypnea  @cite bergeronSME	</span>|<span class="warning">	Returning toward normal @cite bergeronSME	</span>|<span class="success">	Tachycardia, but may be lower value  @cite bergeronSME	</span>|<span class="success">	Slightly diminished from normal @cite bergeronSME	</span>|<span class="warning">	Slightly diminished from normal @cite bergeronSME	</span>|<span class="success">	Slightly diminished from normal @cite bergeronSME	</span>|<span class="success">	Normal @cite bergeronSME	</span>|

<center>
<table border="0">
<tr>
    <td><img src="./plots/Respiratory/TensionPneumothoraxBilateral_TotalLungVolume.jpg" width="550"></td>
    <td><img src="./plots/Respiratory/TensionPneumothoraxBilateral_TranspulmonaryPressure.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/Respiratory/TensionPneumothoraxBilateral_O2Sat.jpg" width="550"></td>
    <td><img src="./plots/Respiratory/TensionPneumothoraxBilateral_MAP.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/Respiratory/TensionPneumothoraxBilateralLegend.jpg" width="1100"></td>
</tr>
</table>
</center>
<center><i>Figure 32. Select outputs from the bilateral tension pnemothorax scenario.</i></center>

<center><br>
<i>Table 15. Validation matrix for physiological responses due to open tension pneumothorax on one side in combination with closed tension pneumothorax on the other side, as well as occlusive dressing for the open side and needle decompression interventions for both sides.</i>
</center>

|	Segment	|	Notes	|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Respiration Rate (breaths/min)	|	Tidal Volume (mL)	|	Heart Rate (beats/min)	|	Systolic Pressure (mmHg)	|	Diastolic Pressure (mmHg)	|	Mean Arterial Pressure (mmHg)	|	Oxygen Saturation	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Tension Pneumothorax;  Severity: 0.5 Type: Open Side: Left Tension Pneumothorax;  Severity: 0.5 Type: Closed Side: Right	|	Represents a leak on the right and left lung sides simultaneously with a moderate severity (large size hole).	|	30	|	300	|<span class="success">	 Increase  @cite bond200968w, @cite leigh2005tension 	</span>|<span class="danger">	Increase @cite leigh2005tension	</span>|<span class="success">	Increase  @cite bond200968w, @cite leigh2005tension	</span>|<span class="success">	Decrease  @cite bond200968w, @cite leigh2005tension	</span>|<span class="danger">	Decrease  @cite bond200968w, @cite leigh2005tension	</span>|<span class="success">	Decrease  @cite bond200968w, @cite leigh2005tension	</span>|<span class="success">	Decrease  @cite bond200968w, @cite leigh2005tension	</span>|
|	Apply Chest Occlusive dressing on the left side 	|	Represents the closing of chest wound on the left side	|	330	|	360	|<span class="success">	No Change or Increased  @cite bergeronSME	</span>|<span class="success">	No Change 	</span>|<span class="warning">	Tachycardia > 120 @cite bergeronSME	</span>|<span class="success">	Slightly improved to no change  @cite bergeronSME	</span>|<span class="success">	Slightly improved to no change  @cite bergeronSME	</span>|<span class="success">	Slightly improved to no change  @cite bergeronSME	</span>|<span class="success">	Modest decrease to no change  @cite bergeronSME	</span>|
|	Apply needle decompression on the left side 	|	Needle decompression is applied on the left side of the chest	|	390	|	400	|<span class="success">	Returning toward normal @cite bergeronSME	</span>|<span class="success">	Returning toward normal @cite bergeronSME	</span>|<span class="success">	Returning toward normal @cite bergeronSME	</span>|<span class="success">	Diminished from normal @cite bergeronSME	</span>|<span class="warning">	Diminished from normal @cite bergeronSME	</span>|<span class="success">	Diminished from normal @cite bergeronSME	</span>|<span class="success">	Returning toward normal @cite bergeronSME	</span>|
|	Apply needle decompression on the right side 	|	Needle decompression is applied on the right side of the chest	|	420	|	720	|<span class="success">	Slightly lower tachypnea  @cite bergeronSME	</span>|<span class="success">	Returning toward normal @cite bergeronSME	</span>|<span class="success">	Tachycardia, but may be lower value  @cite bergeronSME	</span>|<span class="success">	Slightly diminished from normal @cite bergeronSME	</span>|<span class="warning">	Slightly diminished from normal @cite bergeronSME	</span>|<span class="success">	Slightly diminished from normal @cite bergeronSME	</span>|<span class="success">	Normal @cite bergeronSME	</span>|


### Mainstem Intubation

The right and left mainstem intubation actions were validated with a scenario that simulates mechanical ventilation with improper tracheal tube placement. In this scenario, the patient is injected with succinylcholine, then mechanically ventilated through the placement of the endotracheal tube. During the intubation action, the tube is incorrectly placed, leading to the right and left mainstem intubation. The mainstem intubations are then set to the trachea for proper function.

Common clinical vital signs were compared to validation data, as shown in Table 16.

<center>
<table border="0">
<tr>
    <td><img src="./plots/Respiratory/MainstemIntubation_LeftLungVolume.jpg" width="550"></td>
    <td><img src="./plots/Respiratory/MainstemIntubation_TidalVolume.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/Respiratory/MainstemIntubation_RightLungVolume.jpg" width="550"></td>
    <td><img src="./plots/Respiratory/MainstemIntubation_O2Sat.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/Respiratory/MainstemIntubationLegend.jpg" width="800"></td>
</tr>
</table>
</center>
<center><i>Figure 33. Select outputs from the right mainstem intubation scenario.</i></center>

<center><br>
<i>Table 16. Validation matrix for physiological responses due to varying severities of right mainstem intubation and correction.</i>
</center>

|	Segment	|	Notes	|	Action Occurrence Time  (s)	|	Sampled Scenario Time  (s)	|	Respiration Rate (breaths/min)	|	Heart Rate (beats/min)	|	Systolic Pressure (mmHg)	|	Diastolic Pressure (mmHg)	|	Oxygen Saturation	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Inject succinylcholine; concentration = 4820 ug/ml; Dose = 20 ml	|	Neuromuscular blocker agent is administered to represent events that facilitate endotracheal intubation.	|	30	|	60	|<span class="success">	Goes to Zero @cite Morgan2006Clinical             @cite dukeSME	</span>|<span class="success">	Mild  Increase @cite Morgan2006Clinical                       5-10% Decrease @cite dukeSME    	</span>|<span class="success">	Mild  Increase @cite Morgan2006Clinical                      5-10% Decrease @cite dukeSME	</span>|<span class="success">	Mild  Increase @cite Morgan2006Clinical                       5-10% Decrease @cite dukeSME	</span>|<span class="success">	Begins to drop according to preoxygenated O2 curve 	</span>|
|	Turn on Anesthesia Machine and insert   Endotracheal (ET) tube in the right mainstem	|	Anesthesia machine is turned on to supply oxygen to the subject. ET tube is inserted, but it is set in the right mainstem.	|	60	|	180	|<span class="success">	16  (set by machine) 	</span>|<span class="success">	No Change @cite dukeSME; Increase above normal due to cardiovascular stress  @cite bergeronSME	</span>|<span class="success">	No Change @cite dukeSME; Increase above normal due to cardiovascular stress  @cite bergeronSME	</span>|<span class="success">	No Change @cite dukeSME; Increase above normal due to cardiovascular stress  @cite bergeronSME	</span>|<span class="success">	Continues to drop; 93-95% @cite bergeronSME	</span>|
|	Reset ET tube in the trachea.	|	 ET tube is adjusted to be set in the trachea (good depth).	|	180	|	480	|<span class="success">	16  (set by machine) 	</span>|<span class="success">	Towards normal @cite bergeronSME	</span>|<span class="success">	Towards normal @cite bergeronSME	</span>|<span class="success">	Towards normal @cite bergeronSME	</span>|<span class="success">	Towards normal @cite bergeronSME	</span>|
|	Turn on Anesthesia Machine and insert   Endotracheal (ET) tube in the left mainstem	|	ET tube is adjusted, but it is set in the left mainstem.	|	480	|	600	|<span class="success">	16  (set by machine) 	</span>|<span class="success">	No Change @cite dukeSME; Increase above normal due to cardiovascular stress  @cite bergeronSME	</span>|<span class="success">	No Change @cite dukeSME; Increase above normal due to cardiovascular stress  @cite bergeronSME	</span>|<span class="success">	No Change @cite dukeSME; Increase above normal due to cardiovascular stress  @cite bergeronSME	</span>|<span class="success">	Continues to drop; 93-95% @cite bergeronSME	</span>|
|	Reset ET tube in the trachea.	|	 ET tube is adjusted to be set in the trachea (good depth).	|	600	|	900	|<span class="success">	16  (set by machine) 	</span>|<span class="success">	Towards normal @cite bergeronSME	</span>|<span class="success">	Towards normal @cite bergeronSME	</span>|<span class="success">	Towards normal @cite bergeronSME	</span>|<span class="success">	Towards normal @cite bergeronSME	</span>|


### Esophageal Intubation

The esophageal intubation action was validated with a scenario with multiple intubation attempts, a failed esophageal intubation following by a successful endotracheal intubation. In this scenario, the patient is injected with succinylcholine, followed by mechanical ventilation via an endotracheal tube. The first attempt at intubation results in an esophageal intubation. A second attempt is successful. 

All clinical outputs validated agreed with expected trends. As previously mentioned, oxygen saturation shows a decrease as expected within the sampled time. A longer run time for the scenario would show a sharper decrease in the oxygen saturation. 

<center>
<table border="0">
<tr>
    <td><img src="./plots/Respiratory/EsophagealIntubation_TotalLungVolume.jpg" width="550"></td>
    <td><img src="./plots/Respiratory/EsophagealIntubation_O2Sat.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/Respiratory/EsophagealIntubation_StomachInflow.jpg" width="550"></td>
    <td><img src="./plots/Respiratory/EsophagealIntubation_MAP.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/Respiratory/EsophagealIntubationLegend.jpg" width="800"></td>
</tr>
</table>
</center>
<center><i>Figure 34. Select outputs from the esophageal intubation scenario.</i></center>

<center><br>
<i>Table 17. Validation matrix for physiological responses due esophageal intubation and correction.</i>
</center>

|	Segment	|	Notes	|	Action Occurrence Time  (s)	|	Sample Scenario Time  (s)	|	Respiration Rate  (breaths/min)	|	Heart Rate  (beats/min)	|	Systolic Pressure  (mmHg)	|	Diastolic Pressure  (mmHg)	|	Oxygen Saturation  (mmHg)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Administer Succinylcholine - 96.4mg	|	Drug Onset < 1 minute	|	30	|	30-50	|<span class="success">	Goes to Zero @cite Morgan2006Clinical             @cite dukeSME	</span>|<span class="success">	Mild  Increase @cite Morgan2006Clinical;                       5-10% Decrease @cite dukeSME    	</span>|<span class="succes">	Mild  Increase @cite Morgan2006Clinical;                       5-10% Decrease @cite dukeSME    	</span>|<span class="success">	Mild  Increase @cite Morgan2006Clinical;                       5-10% Decrease @cite dukeSME    	</span>|<span class="success">	Begins to drop according to preoxygenated O2 curve 	</span>|
|	Esophageal Intubation	|	This represents failed  endotracheal intubation that led to esophageal intubation	|	60	|	120	|<span class="success">	Zero 	</span>|<span class="success">	 If undetected and O2 level drops significantly, Mild Increase @cite dukeSME	</span>|<span class="success">	 If undetected and O2 level drops significantly, Mild Increase @cite dukeSME	</span>|<span class="success">	 If undetected and O2 level drops significantly, Mild Increase @cite dukeSME	</span>|<span class="success">	continue to decrease along O2 curve 	</span>|
|	Successful Intubation	|	Represents successful endotracheal intubation	|	180	|	500	|<span class="success">	16  (set by machine) 	</span>|<span class="success">	Towards normal @cite bergeronSME	</span>|<span class="success">	Towards normal @cite bergeronSME	</span>|<span class="success">	Towards normal @cite bergeronSME	</span>|<span class="success">	Towards normal @cite bergeronSME	</span>|

### Asthma

The acute asthma action was validated against three scenarios: moderate (severity = 0.3), severe (severity = 0.7), and life-threatening (severity = 0.9). In all three cases, general trends for respiration rate, trachea flow, tidal volume, and IE ratio fall in line with expected behavior. CO<SUB>2</SUB> partial pressure and O<SUB>2</SUB> saturation also tend to agree with expected values. However, blood pressure and heart rate are less consistent. 

<center>
<table border="0">
<tr>
    <td><img src="./plots/Respiratory/AsthmaAttackModerateAcute_TidalVolume.jpg" width="550"></td>
    <td><img src="./plots/Respiratory/AsthmaAttackModerateAcute_IERatio.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/Respiratory/AsthmaAttackModerateAcute_O2PP.jpg" width="550"></td>
    <td><img src="./plots/Respiratory/AsthmaAttackModerateAcute_RR.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/Respiratory/AsthmaAttackModerateAcuteLegend.jpg" width="300"></td>
</tr>
</table>
</center>
<center><i>Figure 35. Select outputs from the moderate airway obstruction scenario.</i></center>

<center><br>
<i>Table 18. Validation matrix for physiological responses due to moderate acute asthma and correction.</i>
</center>

|	Segment	|	Notes	|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Trachea Flow - Peak Expiratory Flow  (L/min)	|	Respiration Rate (breaths/min)	|	Tidal Volume (mL)	|	Heart Rate (beats/min)	|	Systolic Pressure (mmHg)	|	Oxygen Saturation	|	PaO2 (mmHg)	|	PaCO2 (mmHg)	|	IERatio	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Moderate acute asthma attack (Severity = 0.3)	|		|	30	|	550	|<span class="success">	75% to 90% of normal @cite britishGuideline2014asthma	</span>|<span class="success">	Slight Increase @cite britishGuideline2014asthma	</span>|<span class="success">	 ~95% of normal @cite limmer2011emergency	</span>|<span class="success">	No Change @cite britishGuideline2014asthma	</span>|<span class="success">	No Change @cite papiris2001clinical	</span>|<span class="success">	No Change @cite britishGuideline2014asthma	</span>|<span class="success">	No Change @cite britishGuideline2014asthma	</span>|<span class="success">	No Change @cite britishGuideline2014asthma	</span>|<span class="success">	Decrease @cite van1991physical	</span>|
|	No asthma (Severity = 0.0)	|		|	580	|	775	|<span class="success">	Normal	</span>|<span class="success">	Normal	</span>|<span class="success">	Normal	</span>|<span class="success">	Normal	</span>|<span class="success">	Normal	</span>|<span class="success">	Normal	</span>|<span class="success">	Normal	</span>|<span class="success">	Normal	</span>|<span class="success">	Normal	</span>|


<center>
<table border="0">
<tr>
    <td><img src="./plots/Respiratory/AsthmaAttackSevereAcute_TidalVolume.jpg" width="550"></td>
    <td><img src="./plots/Respiratory/AsthmaAttackSevereAcute_IERatio.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/Respiratory/AsthmaAttackSevereAcute_O2PP.jpg" width="550"></td>
    <td><img src="./plots/Respiratory/AsthmaAttackSevereAcute_RR.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/Respiratory/AsthmaAttackSevereAcuteLegend.jpg" width="300"></td>
</tr>
</table>
</center>
<center><i>Figure 36. Select outputs from the severe asthma attack scenario.</i></center>

<center><br>
<i>Table 19. Validation matrix for physiological responses due to severe acute asthma and correction.</i>
</center>

|	Segment	|	Notes	|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Trachea Flow - Peak Expiratory Flow  (L/min)	|	Respiration Rate (breaths/min)	|	Tidal Volume (mL)	|	Heart Rate (beats/min)	|	Systolic Pressure (mmHg)	|	Oxygen Saturation	|	PaO2 (mmHg)	|	PaCO2 (mmHg)	|	IERatio	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Severe acute asthma attack (Severity = 0.7)	|		|	30	|	550	|<span class="success">	< 50% of Normal @cite papiris2001clinical, @cite britishGuideline2014asthma	</span>|<span class="success">	> 25 @cite britishGuideline2014asthma	</span>|<span class="success">	< 60% of normal @cite limmer2011emergency	</span>|<span class="success">	> 120 @cite papiris2001clinical; > 110 @cite britishGuideline2014asthma	</span>|<span class="warning">	Decrease Pulsus Paradoxus 85 mm Hg @cite papiris2001clinical	</span>|<span class="warning">	< 90% @cite papiris2001clinical; < 92% @cite britishGuideline2014asthma	</span>|<span class="warning">	Decrease Mild Hypoxemia @cite bergeronSME; < 60 mmHg @cite papiris2001clinical, @cite britishGuideline2014asthma	</span>|<span class="success">	Normal or Slight Increase > 45 mm Hg @cite papiris2001clinical, @cite britishGuideline2014asthma 	</span>|<span class="success">	Decrease @cite van1991physical	</span>|
|	No asthma (Severity = 0.0)	|		|	580	|	775	|<span class="success">	Normal	</span>|<span class="success">	Normal	</span>|<span class="success">	Normal	</span>|<span class="success">	Normal	</span>|<span class="success">	Normal	</span>|<span class="success">	Normal	</span>|<span class="success">	Normal	</span>|<span class="success">	Normal	</span>|<span class="success">	Normal	</span>|


<center>
<table border="0">
<tr>
    <td><img src="./plots/Respiratory/AsthmaAttackLifeThreateningAcute_TidalVolume.jpg" width="550"></td>
    <td><img src="./plots/Respiratory/AsthmaAttackLifeThreateningAcute_IERatio.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/Respiratory/AsthmaAttackLifeThreateningAcute_O2PP.jpg" width="550"></td>
    <td><img src="./plots/Respiratory/AsthmaAttackLifeThreateningAcute_RR.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/Respiratory/AsthmaAttackLifeThreateningAcuteLegend.jpg" width="300"></td>
</tr>
</table>
</center>
<center><i>Figure 37. Select outputs from the life threatening asthma attack scenario.</i></center>

<center><br>
<i>Table 20. Validation matrix for physiological responses due to life threatening acute asthma and correction.</i>
</center>

|	Segment	|	Notes	|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Trachea Flow - Peak Expiratory Flow  (L/min)	|	Respiration Rate (breaths/min)	|	Tidal Volume (mL)	|	Heart Rate (beats/min)	|	Systolic Pressure (mmHg)	|	Oxygen Saturation	|	PaO2 (mmHg)	|	PaCO2 (mmHg)	|	IERatio	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Life threatening acute asthma attack (Severity = 0.9) 	|		|	30	|	550	|<span class="success">	< 33% of Normal @cite britishGuideline2014asthma	</span>|<span class="success">	> 25 @cite britishGuideline2014asthma	</span>|<span class="success">	< 60% of normal @cite limmer2011emergency	</span>|<span class="success">	> 120 @cite papiris2001clinical; > 110 @cite britishGuideline2014asthma Arrhythmia, @cite britishGuideline2014asthma	</span>|<span class="warning">	Decrease NO Pulsus Paradoxus 85 mm Hg @cite papiris2001clinical	</span>|<span class="success">	< 92% @cite britishGuideline2014asthma	</span>|<span class="success">	< 60 mmHg @cite britishGuideline2014asthma	</span>|<span class="success">	Normal or Slight Increase > 45 mm Hg @cite papiris2001clinical, @cite britishGuideline2014asthma 	</span>|<span class="success">	Decrease @cite van1991physical	</span>|
|	No asthma (Severity = 0.0)	|		|	580	|	775	|<span class="success">	Normal	</span>|<span class="success">	Normal	</span>|<span class="success">	Normal	</span>|<span class="success">	Normal	</span>|<span class="success">	Normal	</span>|<span class="success">	Normal	</span>|<span class="success">	Normal	</span>|<span class="success">	Normal	</span>|<span class="success">	Normal	</span>|

### Apnea

Apnea directly effects the achieved respiratory driver (breathing muscles) pressure amplitude.  The more severe the apnea, the lower the tidal volume.

|	Segment	|	Notes	|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Tidal Volume (mL)	|
|	---	|	---	|	---	|	---	|	---	|
|	Apnea: Severity = 0.3	|	Mild	|	30	|	210	|<span class="success">	Decrease to ~70% of healthy	</span>|
|	Apnea: Severity = 0.0	|	Healthy	|	210	|	510	|<span class="success">	Normal	</span>|
|	Apnea: Severity = 0.6	|	Moderate	|	510	|	690	|<span class="success">	Decrease to ~30% of healthy	</span>|
|	Apnea: Severity = 0.0	|	Healthy	|	690	|	990	|<span class="success">	Normal	</span>|
|	Apnea: Severity = 1.0	|	Severe	|	990	|	1170	|<span class="success">	Decrease to 0	</span>|
|	Apnea: Severity = 0.0	|	Healthy	|	1170	|	1650	|<span class="success">	Normal	</span>|

### Conscious Respiration 

Conscious respiration is validated as part of inhaler usage described in the %Inhaler methodology section @ref InhalerMethodology.

@anchor respiratory-assessmentvalidation
Validation - Assessments
------------------------

### Pulmonary Function Test

Validation of the pulmonary function test may be concluded from the validation of the resting physiologic quantities. The collection of values produced from the pulmonary function test were validated above. Additional validation comes by comparing the general shape of the produced waveform to an expected waveform. The plot shown in Figure 38 presents a way of representing the %BioGears pulmonary function test visually. There is excellent agreement with the plot produced from the engine and the plot found in the literature @cite Kapwatt2014Lungvolumes.

<center><img src="./plots/Respiratory/Pulmonary_Function_Test_Results.jpg" width="1000"></center>
<center>
<i>Figure 38. The lung volume plot from the pulmonary function test displays the lung volume waveform. The waveform has a frequency associated with the respiration rate of the patient at the time of the pulmonary function test. The inspiratory and expiratory reserve volumes are shown with a dilated period to represent a forced component of the inspiration and expiration. The lung volume plot shown is presented with 100 data points.</i>
</center><br>

@anchor respiratory-conclusion
Conclusion
----------

Mathematical modeling of the pulmonary physiology dates back to the work of Gray (1945) and others, including Guyton and collaborators @cite milhorn1965mathematical . Since then several mathematical models have been developed that describe aspects of the pulmonary physiology in considerable detail. However, the scope and complexity of the human physiology demands a model that integrates the dynamic respiratory physiology with other systems such as cardiovascular, renal, and others. As a result, there is a need for an integrative modeling environment that serves as a tool for predicting aspects of the human physiology and pathology.

The %BioGears %Respiratory Model provides a system-level computational model that allows for real-time simulation of normal pulmonary physiology and respiratory distresses. The model employs a circuit-based lumped parameter mathematical model that allows for the characterization of patient-specific respiratory mechanics at a relatively low computational burden. The coupling of the %BioGears %Respiratory System with the Circulatory System through gas exchange and respiratory distress modifiers permits robust prediction of respiratory effects on hemodynamic variables. Accordingly, the current version of the %BioGears %Respiratory System is capable of predicting respiratory and hemodynamic responses to various insults like airway obstruction, bronchoconstriction, pneumothorax, and others.

As part of a modeling tool for various clinical settings, the %Respiratory Model can be hooked up to the anesthesia machine for mechanical ventilation. During positive pressure ventilation, the %Respiratory Model permits the transport of inhaled volatile gases and others from the anesthesia machine to the circulatory system. Such an integrative approach allows for the examination of respiratory and cardiovascular responses to various stimuli and stresses as observed in practical application.

The %Respiratory System, along with the other systems in the engine, can serve as a training tool for exploring the dynamic interaction between physiological systems. Researchers can benefit from the integration of higher-fidelity modules into the existing model for more detailed and project-specific analysis. In its current form, the %Respiratory Model is adapted to a specific patient using physiological parameters selected from patient data. This limits the practical application of the model to the tested standard model. However, the common data model platform gives the user the ease and flexibility to select parameters to experiment their own patient setting.

@anchor respiratory-future
Future Work
===========

@anchor respiratory-coming
Coming Soon
-----------

@anchor respiratory-improvements
Recommended Improvements
------------------------

The development of an integrative model like the %BioGears %Respiratory System provides opportunities for exploring various pulmonary diseases and distresses. One common pulmonary distress that can be modeled by the %BioGears %Respiratory Model corresponds to the Acute %Respiratory Distress Syndrome (ARDS). This condition compromises alveolar gas transfer and leads to low levels of oxygen in the blood. Future funding may be sought for incorporating and validating ARDS model in the %BioGears %Respiratory System.

A respiratory control mechanism for rhythmic breathing could be added to the engine. Rhythmic breathing is achieved by the respiratory neurons in the medullary respiratory center that give rise to activity in the efferent nerves ennervating the respiratory muscles. The neural control network that relates the respiratory neuron output to the mechanical airflow is not incorporated in the current version of the %Respiratory Model. The current version of the %BioGears  %Respiratory Model has succeeded in controlling the breathing cycle through parameters that  adjust the spontaneous breathing based on a chemical control feedback mechanism.  While this development allows robust chemical stimuli responses,  incorporation of the neural circuitry model that controls the ventilation process can provide a more accurate implementation of the %Respiratory System that includes behavioral control of breathing.

@anchor respiratory-appendices
Appendices
==========

Acronyms
--------

ARD - Acute %Respiratory Distress

RV - Residual volume

ERV - Expiratory reserve volume

V<sub>T</sub> - Tidal volume

IRV - Inspiratory reserve volume

V<sub>C</sub>  - Vital capacity

IC - Inspiratory capacity

FRC - Functional residual capacity

TLC - Total lung capacity

RR - Respiration rate

V<sup><b>.</b></sup><sub>E</sub> - Minute ventilation

V<sup><b>.</b></sup><sub>A</sub> - Alveolar ventilation

Data Model Implementation
-------------------------

@ref RespiratorySystemTable "Respiratory"

Compartments
------------

* Mouth
* Stomach
* Carina
* Lungs
	* LeftLung
		* LeftDeadSpace
		* LeftAlveoli
	* RightLung, RightLungPulmonary
		* RightDeadSpace
		* RightAlveoli
* PleuralCavity
	* LeftPleuralCavity
	* RightPleuralCavity
* LeftAlveoliLeak
* LeftChestLeak
* RightAlveoliLeak
* RightChestLeak
%Methodology Template {#MethodologyTemplate}
==========================

Overview
========

Abstract
--------

Brief summary of what.

In general, write the abstract as one sentence from each major section.
So, one intro, one methods, one validation/results, and one conclusions.

Introduction
------------

More details about why and how.

System Design
=============

Background and Scope
--------------------

Include where the models originated and why we chose them.

### Requirements

### Existing

### Approach

Data Flow
---------

### Initialize

Initialize, setup, tune, conditions, stabilization, etc.

### Preprocess

### Process

### Post Process

### Assessments

Include a data flow diagram.

Features and Capabilities
-------------------------

Include a discussion about model fidelity.

### Definitions

Table with symbols (if needed).

### Circuit

Include a circuit diagram.

### Patient Variability

How changes in the patient affect the model.

### Feedback

How the homeostatic methodology works.

### Dependencies

How does this system interact with other systems?

### Outputs

Any discussion about system outputs?

Assumptions and Limitations
---------------------------

Conditions
----------

Actions
-------

### Insults

### Interventions

Events
------

Include irreversible states.

Assessments
-----------

Results and Conclusions
=======================

%Verification
-------------

Often through unit tests.

Validation - Resting Physiologic State
--------------------------------------

Use the auto-generated table.

Explain the good and the bad, note why the bad is either being addressed
or okay.

Validation - Conditions
-----------------------

Include validation tables.

Validation - Actions
--------------------

Include validation tables and plots.

Validation - Assessments
------------------------

Conclusions
-----------

1-2 sentences for each of the following: sum up what we have done, have
we advanced the state of the art (hint: yes), note the worst and how it
is being addressed or why it is ok, note the best stuff, why is this is
good step, who will find this useful

Future Work
===========

Coming Soon
-----------

What else will be included under this contract.

Improvements to existing methodology planned. Include
a why statement. Need to know why this improvement is a good thing to
do, what do we gain.

Recommended Improvements
------------------------

Be careful not to give away future funding ideas.

Appendices
==========

Acronyms
--------

Define terms and acronyms.

Data Model Implementation
-------------------------

Link to associated classes.

Compartments
------------

Include a tree showing the compartments.System Methodology {#SystemMethodology}
==================

# Overview

## Abstract

The %BioGears Engine is a fully customizable collection of physiologic
systems. It includes models for anatomy-based biological systems and
medical equipment. We developed an advanced engine through a
top-down approach, with the ability to easily increase fidelity and
manipulate system models. It can be integrated with other software and
hardware platforms and devices.

## Introduction

Computational modeling and analysis of the human physiology is extremely
useful for various applications. There are many approaches for simulating the body's response to various stimuli. Some existing
software uses state-based calculation that can only provide a limited
amount of data, by way of preprogrammed responses. Complex interactive
effects are difficult with this approach because it uses additive or
purely multiplicative responses.

Some existing high fidelity models use more complex computational fluid
dynamics (CFD). One major drawback of this approach is the significant computation resources required. Cellular, molecular, and genetic models are difficult
for whole body physiology. Using this low-level individual cell and
receptor methodology often does not provide the level of data a
clinician requires for informed decision-making. Physiologically-based
pharmacokinetic (PBPK) modeling focuses only on drug interactions, and
is generally designed to deal with specific pathophysiology.

%BioGears is designed to be a deterministic multiscale modeling application that can be
integrated with all other types of physiologic software. Models at
any anatomical level can be implemented within the existing
infrastructure.

@image html SystemFidelity.png
<center>
*Figure 1. %BioGears uses a top-down approach to model development with bottom-up hooks for expansion.*
</center><br>

# System Design

## Background and Scope

### Physiology Engine

The %BioGears Physiology Engine is the combination of lumped parameter
mathematical models that, together, simulate whole body physiology. Each
system contributes to the maintenance of homeostasis. The systems are
designed to be modular and each has the ability to work in
isolation. They are all reliant on the Common Data Model (CDM) connections and leverage
several generic solvers and transport algorithms.

The system as a whole needs to be easy to use and modify, modular, and
extensible to virtually any fidelities. Envisioned end user groups
include game developers, mannequin builders, educators, researchers,
sensor system developers, and trainers. Several of these users implement
%BioGears in a way that requires faster than real time processing without
access to computational resources that exceed a personal computer.

### Sub-Systems

There are two different types of systems defined in %BioGears - physiology-based and equipment-based.
Below is a list of the current %BioGears Systems with links to their methodology reports:

@anchor system-systems
Patient Variablity:
- @ref PatientMethodology

Physiologic:
- @ref BloodChemistryMethodology
- @ref CardiovascularMethodology
- @ref DrugsMethodology
- @ref EndocrineMethodology
- @ref EnergyMethodology
- @ref EnvironmentMethodology
- @ref GastrointestinalMethodology
- @ref NervousMethodology
- @ref RenalMethodology
- @ref RespiratoryMethodology
- @ref TissueMethodology
- @ref HepaticMethodology

Equipment:
- @ref AnesthesiaMachineMethodology
- @ref InhalerMethodology
- Electrocardiogram (%ECG) Methodology (see @ref CardiovascularMethodology)

### Solvers

The %BioGears suite of also comes with verified tools for extensibility and elimination of potential error sources:
- @ref CircuitMethodology
- @ref SubstanceTransportMethodology

## Data Flow

%BioGears uses differential equations with control-based feedback mechanisms to
dynamically respond to parameter changes. It is designed and tuned
around a stable resting physiologic state, with the ability to
model pathophysiology through insults and intervention action calls.
Many of the system circuit elements are modified based on substance
concentrations and external action (i.e., insult and intervention)
modifiers.

The Anesthesia Machine is directly connected to the %Respiratory System
through an inter-circuit connection. The current implementation creates
a combined circuit for analysis; therefore, the combined circuit is calculated as one
large circuit with a single fluid. %Substances can be administered by
inhalation directly to the %Respiratory System or through a connected
Anesthesia Machine through the vaporizer component.

The %Respiratory System transfers substances back and forth to the
%Cardiovascular System using Alveoli Transfer. The only direct connections
between the two systems are simple modifiers to the heart rate and a
cardiovascular resistance value during the tension pneumothorax insult.
The %Cardiovascular System is also tied to the %Endocrine System and %ECG.
%Substances can enter directly into the %Cardiovascular System through
bolus injections or the administration of an IV. 

Interactions between systems, such as alveoli transfer and diffusion between the extravascular and vascular space are modeled in the System Interactions methodology. This ensures each system is responsible for only its own behavior while capturing the behavior that occurs between systems.

<img src="./images/System/SystemFlow.png" width="600">
<center>
*Figure 2. Overall %BioGears Engine data flow diagram showing all of the
systems. Dashed lines are only present for certain actions.
Non-italicized elements are defined as classes in the engine.*
</center><br>

### %Timing

The entire %BioGears Engine works off of a transient analysis time step
of 0.02 s (50 Hz). All system states are recalculated every time step.
The differential equations that determine changes are linearized to
approximate system variables, but the small time step provides a very
accurate solution.

To provide the %BioGears systems with sufficient information, the CDM
maintains three time steps for elements and parameters. These times are:

-   Baseline - The original resting physiology set-point. Baseline
    values are often assigned by the patient parameters. These values
    generally only change hen conditions are applied.

-   Current - The time step that has already been calculated. The
    current values are used to determine the values for the next time
    step elements based on system feedback mechanisms.

-   Next - The next values are what will be used to calculate the system
    state on the next process call.

@anchor system-stabilization	
### Stabilization

The %BioGears Engine stabilizes with a multi-step process. The engine must be initialized and reach a stable state prior to modifying the patient condition. This is completed by using a dynamic stabilization protocol to execute the engine until a specified set of criteria are met, then any patient chronic conditions are applied. These conditions modify patient parameters and model values to represent the new patient state. The engine must restabilize using the dynamic protocol to achieve a stable state. This process is outlined in Figure 3.

<center><img src="./images/System/Stabilization.png" width="550"></center>
<center> 
<i>Figure 3. Overall %BioGears Engine stabilization protocol. This highlights the multi-step process required to initialize either a healthy or chronically ill patient prior to executing a scenario.</i>
</center><br>

#### Dynamic Stabilization

The %BioGears Engine must perform numerous cycles of the calculations to reach a point of
convergence for the output values, such as heart rate, tidal volume, systolic pressure, etc. When the engine has reached this convergence point, it is considered to be 
stable. During the stabilization period, the majority of feedback in the engine is inactive. An exception is the Tune Circuit methodology found in the %Cardiovascular System. This modifies the cardiovascular circuit parameters to achieve the mean arterial pressure specified in each patient file. For more information on this function, see the @ref CardiovascularMethodology. No actions or conditions
can be applied during the stabilization time. Because the patient parameters may vary in the patient file,
the time required for the solution to converge may vary. By using a dynamic stabilization algorithm, the engine
will be fully initialized and the solution will have converged at the conclusion of the stabilization period
for any patient. This stabilization algorithm also prevents an unnecessarily long time for a solution that may be required in the time that were 
statically specified for all scenarios and patients.

To determine the convergence time, %BioGears uses a dynamic stabilization algorithm. The stable outputs
were analyzed to determine the variation present in each individual output over time. This percent
variation was identified for the key outputs in each system. The percent difference between the
current time step and the next time step is calculated for output. The percent difference calculated
must be less than the identified convergence criteria for the output. This must be maintained for
the convergence time. This time is specified at approximately seven respiratory cycles 
(at 16 breath/min -> ~25 seconds). The respiratory cycle was chosen because it is the has the longest period. 
Convergence should occur over the longest periodic cycle in the engine to ensure variation is not at a low frequency.

Currently, the convergence criteria and time are specified in the CDM. In the future, the algorithm
implementation will remain in the CDM, but the convergence criteria (including time) will be moved to an XML file, similar
to patient and substance files, for easy manipulation by the user (see @ref PatientData and @ref SubstanceData for more information). 

#### Conditions

Patient conditions that are persistent or recurring are chronic conditions. The human body responds differently to chronic conditions
than it does to acute conditions. While the body is in an altered state, which may be weaker, the body's response to the condition becomes damped over time.
In short, the body reaches a new level of homeostasis that may not be considered healthy, but does not have the same continual
feedback mechanisms operating that occur when acute conditions cause a deviation from homeostasis. 

To implement these conditions in the %BioGears Engine, a two step stabilization process was implemented.
As discussed above, the dynamic stabilization criteria was required to ensure all outputs converged prior to
performing any actions (insults or intervention) in a given patient. This is true for conditions, as well. However,
after a condition is applied the solution must again converge to represent the body's new homeostatic state, that while different
from the healthy homeostasis, is still a stable patient state. 

The same convergence methodology is used for this second convergence. However, a set of convergence criteria for the outputs
must be specified for each condition. This criteria reflects the new variation in outputs over time that may exist during the condition.
Examples of these conditions are anemia, pericardial effusion, and arrhythmias. More information on these can be found in the
@ref CardiovascularMethodology.

There is no limit on the number of conditions that can be applied at one time. However, testing an validation of outputs when combining conditions has not been performed.  In the future, when a user applies multiple conditions, a merging algorithm will be used to combine the convergence criteria. This will specify the least stringent requirements as the final convergence criteria for the combined conditions.

### Preprocess

Preprocess is called for each system individually and is generally used
to modify the circuit elements based on feedback mechanisms and actions.
This is also where any system drivers are determined for the upcoming
time step, such as the heart contractility or ventilatory drivers.

At the start of every Preprocess step, the next time step circuit
element values are initialized to the stored baseline values. Generally
speaking, elements are updated by getting the next value and using
multipliers before setting it again. This allows for the "stacking" of
modifiers and the ability to alter the same element based on any number
of mechanisms.

### Process

Process generally determines the entire next time step circuit state. The
generic circuit solver is leveraged to automatically calculate unknown
circuit variables. %Substances are also transported throughout and
between systems with general equations - usually based on path flows or
absorption and diffusion coefficients and renal and hepatic clearance factors.

### Postprocess

Postprocess advances time by moving the next time step values to the
current time step values. The next values are then set to the baseline values in
preparation for the upcoming Preprocess call.

@image html SystemDataFlow.png
<center>
*Figure 4. This shows the repetitive three-step process used each time
step to determine the system states. These three processes are mirrored
in each system's code and are sequentially called by the %BioGears
Engine.*
</center><br>

### Interface

Externally available data is defined within the %BioGears Engine in three major ways:
-	System data
	-	Similar to vitals
	-	Set individually within the system classes at each time step
	-	Example: heart rate
-	Compartment data
	-	Anatomical definitions
	-	Mapped/assigned nodes and paths from which to draw information
	-	Theoretically encompasses any number of circuit nodes and paths
	-	Example: trachea oxygen partial pressure
-	Assessments
	-	Data collected and packaged to resemble a report or analysis that might be ordered by a physician
	-	Intended to give general patient overviews
	-	Calculated on demand
	-	Example: pulmonary function test

The %BioGears modeling approach takes the human body and conceptually divides it into various fluid compartments that represents a real division in terms of how portions of the body's water, solutes, and suspended elements are segregated @cite rhoades2012medical.  Compartments can be further discretized into smaller sub-compartments with a hierarchical relationship as you drill into various systems. In %BioGears, compartments can be defined to encapsulate circuit nodes that allow easy organization, access, and synchronization of all system parts.
	
Compartments are implemented in %BioGears as conceptual physical divisions of the body.  Anatomical data can be pulled from each compartment through optional node and path mapping. Compartments can be further discretized into smaller sub-compartments with a hierarchical relationship as you drill into various systems. In %BioGears, compartments can be defined to encapsulate circuit nodes that allow easy organization, access, and synchronization of all system parts. Figure 5 shows an example of how compartments can be defined in the %Cardiovascular System.
	
@image html CompartmentExample.png
<center>
<i>Figure 5. This is an example of possible %Cardiovascular System compartments.  This is for explanation purposes only and not necessarily indicative of how things are really defined.  See the @ref CardiovascularMethodology documentation for how they are really defined.</i>
</center><br>

Users can customize scenarios for %BioGears by modifying a variety of file types. For example, patients, substances, compound substances, and environments can all be modified
via the text files provided. This allows users to customize scenarios by tailoring patients and environmental conditions to the specific circumstances of the user. 
As an example of this flexibility, %Biogears provides a variety of patient files. These parameters in these patient files can be specified to represent different physiologic states for a patient. 
See @ref PatientMethodology for details.

For more details on interfacing with the %BioGears Engine, see the @ref engine documentation.

## Assumptions and Limitations

%BioGears uses a low fidelity clinical approach and is not necessarily
for high fidelity predictive purposes. It is also assumed to be contained
within the CDM.

See the individual system methodology documents for more specifics.

# Results and Conclusions

## Combined Effects Validation

Several scenarios have been developed and validated to ensure proper interaction not only within, but also between the %BioGears physiologic systems. These patients and scenarios were developed as part of the HumanSim: Sedation and Airway project funded by TATRC, contract number W81XWH-11-C-0045. The combined effects scenarios were validated qualitatively from available literature and from the opinions of subject matter experts (SME).  More detail on patient parameters can be found in the @ref PatientMethodology documentation. A summary of the validation is shown in Table 6.  Green indicates good agreement with validation data, 
yellow indicates agreement with a general trend with some minor disagreement, and red indicates a disagreement with the validation data. The number indicates the number of output parameters for each category of validation
success or failure.

<center>
*Table 6. Validation scenarios were completed for five patients subjected to a variety of insults and intervention: Cynthia, Gus, Hassan, Joel, and Nathan. The resulting effects of each individual insult and intervention were validated both qualitatively and from subject matter experts. Green indicates good agreement with validation data, yellow indicates agreement with a general trend with some minor disagreement, and red indicates a disagreement with the validation data. The number indicates the number of output parameters for each category of validation success or failure.*
</center>

|	Key	|
|	---	|
|<span class="success">	Good agreement: correct trends or <10% deviation from expected	</span>|
|<span class="warning"> 	Some deviation: correct trend and/or <30% deviation from expected	</span>|
|<span class="danger">	Poor agreement: incorrect trends or >30% deviation from expected	</span>|

|	Scenario 	|	Description	|	Good	|	Decent	|	Bad	|
|	---	|	---	|	---	|	---	|	---	|
|	Cynthia	|	Midazolam is injected causing an airway obstruction. Oxygen is supplied via a mask. Ketamine is administered and the airway obstruction is cleared. Rocuronium is administered. An endotracheal tube is set.	|<span class="success">	28	</span>|<span class="warning">	2	</span>|<span class="danger">	0	</span>|
|	Hassan	|	Oxygen is administered via a mask. Ketamine and succinylcholine are administered and the mask is removed. Right mainstem intubation occurs. The tube placement is corrected for a successful endotracheal intubation.	|<span class="success">	25	</span>|<span class="warning">	0	</span>|<span class="danger">	0	</span>|
|	Nathan	|	Nathan is injected with Fentanyl. Oxygen is administered via a mask. An oxygen wall port pressure loss occurs. The oxygen bottle is attached to supply oxygen.	|<span class="success">	19	</span>|<span class="warning">	0	</span>|<span class="danger">	1	</span>|
|	Gus	|	A ventilator mask is placed on Gus and oxygen is administered via a mask. Succinylcholine is injected and the mask is removed. A failed endotracheal intubation occurs as an esophageal intubation. Endotracheal intubation then occurs.	|<span class="success">	18	</span>|<span class="warning">	2	</span>|<span class="danger">	0	</span>|
|	Joel	|	Airway obstruction occurs at the beginning of the scenario. the obstruction is removed, and oxygen is supplied via a mask. Etomidate and succinylcholine are administered, and the mask is removed. An endotracheal tube is set.	|<span class="success">	23	</span>|<span class="warning">	2	</span>|<span class="danger">	5	</span>|
|		|	Total	|<span class="success">	113	</span>|<span class="warning">	6	</span>|<span class="danger">	6	</span>|

Although the tidal volume is not shown for these scenarios, it should be noted that mild to moderate discrepancies between the tidal volume set on the anesthesia machine and that observed for the patient were observed in %BioGears. This is also a common observation in clinical practice, because the anesthesia machine may have to overcome the patient condition and airway resistance to provide sufficient tidal volume @cite Morgan2006Clinical .

### Cynthia

The Cynthia scenario begins with the administration of midazolam at 50&nbsp;seconds with a full severity airway obstruction occurring at 110&nbsp;seconds. At 260&nbsp;seconds a ventilator mask is placed on Cynthia. Twenty seconds later, the airway obstruction is removed and 30&nbsp;milligrams of ketamine are administered via a bolus injection. Rocuronium is administered one minute later, and Cynthia is intubated 40&nbsp;seconds after that. The Cynthia scenario shows excellent agreement with the qualitative and expected SME trends. However, there is a minor inconsistency in the systolic and diastolic pressure decrease following the administration of midazolam.  The SME validation predicted a 15-25% decrease in systolic/diastolic pressures following the administration; however, the observed decrease was approximately 10%. This decrease was considered acceptable since other references predicted varying degrees of pressure decrease.

<center>
<table border="0">
<tr>
    <td><img src="./plots/System/Cynthia_HR.jpg" width="550"></td>
    <td><img src="./plots/System/Cynthia_ArterialPressure.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/System/Cynthia_RR.jpg" width="550"></td>
    <td><img src="./plots/System/Cynthia_O2Sat.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/System/CynthiaLegend.jpg" width="1100"></td>
</tr>
</table>
</center>
<center><i>Figure 6. Select outputs from the Cynthia combined effects scenario.</i></center>

<center><br>
*Table 7. The Cynthia scenario displays the effects of sequential midazolam administration, airway obstruction, ventilator mask, ketamine and rocuronium administration, and intubation. This scenario shows some good agreement with the expected qualitative and SME trends.*
</center>

|	Segment	|	Notes	|	Action Occurrence Time  (s)	|	Sample Scenario Time  (s)	|	Heart Rate  (beats/min)	|	Systolic Pressure  (mmHg)	|	Diastolic Pressure  (mmHg)	|	Respiration Rate  (mmHg)	|	Oxygen Saturation  (mmHg)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Administer Midazolam - 3 mg	|	Drug Onset in 1-3 min	|	30	|	90	|<span class="success">	Mild Increase @cite Morgan2006Clinical; 5-10% Decrease @cite dukeSME	</span>|<span class="success">	Moderate Decrease @cite Morgan2006Clinical; 15-25% Decrease @cite dukeSME    	</span>|<span class="success">	Moderate Decrease @cite Morgan2006Clinical; 15-25% Decrease @cite dukeSME    	</span>|<span class="success">	Moderate Decrease @cite Morgan2006Clinical; Marked Decrease to 0 @cite dukeSME	</span>|<span class="success">	As long as some respiration rate, little change. If RR=0, then begins to drop along O2 curve. 	</span>|
|	Airway Obstruction of Severity 1.0	|	Represents an aspiration	|	90	|	240	|<span class="success">	If O2 decreases, Increase @cite dukeSME	</span>|<span class="success">	If O2 decreases, Increase @cite dukeSME	</span>|<span class="success">	If O2 decreases, Increase @cite dukeSME	</span>|<span class="success">	 NC @cite dukeSME	</span>|<span class="success">	Drop according to preoxygenated O2 curve  	</span>|
|	Apply Mask and Turn on Ventilator	|		|	240	|	250	|<span class="danger">	If above increased, Decrease @cite dukeSME	</span>|<span class="warning">	If above increased, Decrease @cite dukeSME	</span>|<span class="warning">	If above increased, Decrease @cite dukeSME	</span>|<span class="success">	Ventilator On; but complete block, so NC  	</span>|<span class="success">	Continues to decrease because still fully blocked  	</span>|
|	Administer Ketamine - 30 mg/ End of Airway Obstruction	|	Drug Onset < 2 minutes	|	250	|	310	|<span class="danger">	Moderate Increase @cite Morgan2006Clinical; 15-25% Increase @cite dukeSME    	</span>|<span class="success">	Moderate Increase @cite Morgan2006Clinical; 15-25% Increase @cite dukeSME    	</span>|<span class="success">	Moderate Increase @cite Morgan2006Clinical; 15-25% Increase @cite dukeSME    	</span>|<span class="warning">	Ventilator On; therefore, 16  	</span>|<span class="success">	Begins to increase because airway obstruction is cleared  	</span>|
|	Administer Rocuronium - 62mg	|	Drug Onset in 60-90 seconds	|	310	|	400	|<span class="success">	NC @cite PaulGBarash2009; NC @cite dukeSME     	</span>|<span class="success">	NC @cite PaulGBarash2009; NC @cite dukeSME     	</span>|<span class="success">	NC @cite PaulGBarash2009; NC @cite dukeSME     	</span>|<span class="success">	Ventilator On; therefore, 16  	</span>|<span class="success">	Continues to increase 	</span>|
|	Successful Intubation	|		|	400	|	900	|<span class="success">	NC @cite dukeSME	</span>|<span class="success">	NC @cite dukeSME	</span>|<span class="success">	NC @cite dukeSME	</span>|<span class="success">	NC 	</span>|<span class="success">	NC 	</span>|


### Gus
A ventilator mask is applied to Gus at 50&nbsp;seconds, and succinylcholine is injected one minute later. After an additional minute, an unsuccessful intubation occurs, leading to the endotracheal tube residing within the esophagus. After two minutes the tube is removed and correctly set in the trachea. The produced results show excellent agreement with the expected validation trends. 

<center>
<table border="0">
<tr>
    <td><img src="./plots/System/Gus_HR.jpg" width="550"></td>
    <td><img src="./plots/System/Gus_ArterialPressure.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/System/Gus_RR.jpg" width="550"></td>
    <td><img src="./plots/System/Gus_O2Sat.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/System/GusLegend.jpg" width="1100"></td>
</tr>
</table>
</center>
<center><i>Figure 7. Select outputs from the Gus combined effects scenario.</i></center>

<center><br>
*Table 8. The Gus scenario displays the effects of sequential ventilator mask application, succinylcholine injection, esophageal intubation and then a successful endotracheal intubation. This scenario shows good agreement with the SME and qualitative validation.*
</center>

|	Segment	|	Notes	|	Action Occurrence Time  (s)	|	Sample Scenario Time  (s)	|	Heart Rate  (beats/min)	|	Systolic Pressure  (mmHg)	|	Diastolic Pressure  (mmHg)	|	Respiration Rate  (mmHg)	|	Oxygen Saturation  (mmHg)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Mask Ventilate with Oxygen	|		|	30	|	90	|<span class="success">	NC @cite dukeSME	</span>|<span class="success">	NC @cite dukeSME	</span>|<span class="success">	NC @cite dukeSME	</span>|<span class="success">	Slight decrease to due anesthesia machine resistance ; NC @cite dukeSME	</span>|<span class="success">	100	</span>|
|	Administer Succinycholine - 100mg	|	Drug Onset < 1 minute	|	90	|	150	|<span class="success">	Mild  Increase @cite Morgan2006Clinical; 5-10% Decrease @cite dukeSME    	</span>|<span class="success">	Mild  Increase @cite Morgan2006Clinical; 5-10% Decrease @cite dukeSME    	</span>|<span class="success">	Mild  Increase @cite Morgan2006Clinical; 5-10% Decrease @cite dukeSME    	</span>|<span class="success">	Goes to Zero @cite Morgan2006Clinical; Goes to Zero @cite dukeSME     	</span>|<span class="success">	NC @cite dukeSME	</span>|
|	Esophageal Intubation	|		|	150	|	270	|<span class="success">	If O2 decreases, Increase @cite dukeSME	</span>|<span class="danger">	If O2 decreases, Increase @cite dukeSME	</span>|<span class="success">	If O2 decreases, Increase @cite dukeSME	</span>|<span class="success">	Ventilator rate is 16  	</span>|<span class="success">	May slowly decrease along O2 curve @cite dukeSME	</span>|
|	Successful Intubation	|		|	270	|	770	|<span class="success">	If above increased, Decrease @cite dukeSME	</span>|<span class="danger">	If above increased, Decrease @cite dukeSME	</span>|<span class="success">	If above increased, Decrease @cite dukeSME	</span>|<span class="success">	16	</span>|<span class="success">	Begins to increase to a normal level @cite dukeSME	</span>|


### Hassan

At the beginning of the scenario, a ventilator mask is applied to Hassan. He then receives bolus injections of ketamine and succinylcholine. This leads to an increase in the heart rate and arterial pressures due to the ketamine injection. The respiration rate begins to decrease as the patient begins to lose consciousness. After the succinylcholine injection, there is a decrease in the heart rate and arterial pressures. In addition, the respiration rate falls to zero due to the neuromuscular block. At 180&nbsp;seconds, an endotracheal tube is set into the right bronchi, leading to only one lung being ventilated. This leads to a minor reduction in the oxygen saturation. Due to the reduced oxygen intake, the heart rate and arterial pressures begin to increase to compensate. The tube is reset into the trachea and the vital signs begin to return to normal. All of these trends follow the expected validation trends.

<center>
<table border="0">
<tr>
    <td><img src="./plots/System/Hassan_HR.jpg" width="550"></td>
    <td><img src="./plots/System/Hassan_ArterialPressure.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/System/Hassan_RR.jpg" width="550"></td>
    <td><img src="./plots/System/Hassan_O2Sat.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/System/HassanLegend.jpg" width="1100"></td>
</tr>
</table>
</center>
<center><i>Figure 8. Select outputs from the Hassan combined effects scenario.</i></center>

<center><br>
*Table 9. Hassan displays the effects of a right mainstem intubation after the patient has been injected with ketamine and succinylcholine. The engine output shows complete agreement with the validation trends.*
</center>

|	Segment	|	Notes	|	Action Occurrence Time  (s)	|	Sample Scenario Time  (s)	|	Heart Rate  (beats/min)	|	Systolic Pressure  (mmHg)	|	Diastolic Pressure  (mmHg)	|	Respiration Rate  (mmHg)	|	Oxygen Saturation  (mmHg)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Mask Ventilate with Oxygen	|		|	30	|	90	|<span class="success">	NC @cite dukeSME	</span>|<span class="success">	NC @cite dukeSME	</span>|<span class="success">	NC @cite dukeSME	</span>|<span class="success">	Slight decrease to due anesthesia machine resistance ; NC @cite dukeSME	</span>|<span class="success">	100 @cite dukeSME	</span>|
|	Administer Ketamine - 168 mg	|	Drug Onset < 1 minute	|	90	|	130	|<span class="success">	Moderate Increase @cite Morgan2006Clinical; 15-25% Increase @cite dukeSME    	</span>|<span class="success">	Moderate Increase @cite Morgan2006Clinical; 15-25% Increase @cite dukeSME    	</span>|<span class="success">	Moderate Increase @cite Morgan2006Clinical; 15-25% Increase @cite dukeSME    	</span>|<span class="success">	Mild Decrease @cite Morgan2006Clinical; 25-50% Decrease @cite dukeSME    	</span>|<span class="success">	NC @cite dukeSME	</span>|
|	Administer Succinycholine - 115mg	|	Drug Onset < 1 minute	|	130	|	160	|<span class="success">	Mild  Increase @cite Morgan2006Clinical; 5-10% Decrease @cite dukeSME    	</span>|<span class="success">	Mild  Increase @cite Morgan2006Clinical; 5-10% Decrease @cite dukeSME    	</span>|<span class="success">	Mild  Increase @cite Morgan2006Clinical; 5-10% Decrease @cite dukeSME    	</span>|<span class="success">	Ventilator is on; therefore, 16    	</span>|<span class="success">	Begins to drop according to preoxygenated O2 curve @cite dukeSME 	</span>|
|	Right Mainstem Intubation	|	Intent is to simulate tube migration	|	160	|	250	|<span class="warning">	NC, If undetected and O2 level drops significantly, Mild Increase @cite dukeSME	</span>|<span class="success">	NC, If undetected and O2 level drops significantly, Mild Increase @cite dukeSME	</span>|<span class="success">	NC, If undetected and O2 level drops significantly, Mild Increase @cite dukeSME	</span>|<span class="success">	16	</span>|<span class="success">	May slowly decrease along O2 curve @cite dukeSME	</span>|
|	Correct tube placement	|	Reset the intubation tube	|	250	|	450	|<span class="success">	NC, If above result in a Mild Increase, will return to normal @cite dukeSME	</span>|<span class="success">	NC, If above result in a Mild Increase, will return to normal @cite dukeSME	</span>|<span class="warning">	NC, If above result in a Mild Increase, will return to normal @cite dukeSME	</span>|<span class="success">	16	</span>|<span class="success">	Begins to increase to a normal level >97% @cite dukeSME	</span>|


### Joel
The Joel scenario begins with a full severity airway obstruction at 50&nbsp;seconds. This leads to decreasing oxygen saturation and increasing heart rate and arterial pressures due to the sympathetic (endocrine) response. At 170&nbsp;seconds, the obstruction is removed and the vital signs begin to return to normal. A ventilator mask is applied to Joel at 230&nbsp;seconds, and he receives a 27&nbsp;mg bolus injection of etomidate at 290&nbsp;seconds. This leads to decreasing arterial pressure. At 310&nbsp;seconds, a bolus injection of succinylcholine occurs before an endoctracheal tube is set 30&nbsp;seconds later. The administration of succinylcholine leads to immediate decreases in heart rate, arterial pressures, and oxygen saturation. After the intubation occurs, the oxygen saturation begins to return to normal due to adequate ventilation. All of these results show strong agreement with the subject matter expert's expected trends.

<center>
<table border="0">
<tr>
    <td><img src="./plots/System/Joel_HR.jpg" width="550"></td>
    <td><img src="./plots/System/Joel_ArterialPressure.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/System/Joel_RR.jpg" width="550"></td>
    <td><img src="./plots/System/Joel_O2Sat.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/System/JoelLegend.jpg" width="1100"></td>
</tr>
</table>
</center>
<center><i>Figure 9. Select outputs from the Joel combined effects scenario.</i></center>

<center><br>
*Table 10. The Joel scenario displays the effects of sequential airway obstruction, ventilator mask application, etomidate administration, succinylcholine administration, and intubation. The %BioGears results show excellent agreement with the validation trends.*
</center>

|	Segment	|	Notes	|	Action Occurrence Time  (s)	|	Sample Scenario Time  (s)	|	Heart Rate  (beats/min)	|	Systolic Pressure  (mmHg)	|	Diastolic Pressure  (mmHg)	|	Respiration Rate  (mmHg)	|	Oxygen Saturation  (mmHg)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Airway Obstruction of Severity 1.0	|	Represents an aspiration	|	30	|	150	|<span class="success">	NC until O2 drops then increase @cite dukeSME	</span>|<span class="warning">	NC until O2 drops then increase @cite dukeSME	</span>|<span class="success">	NC until O2 drops then increase @cite dukeSME	</span>|<span class="warning">	Goes to Zero @cite dukeSME	</span>|<span class="success">	Begins to drop according to O2 curve @cite dukeSME	</span>|
|	End Airway Obstruction	|	Represents suctioning	|	150	|	210	|<span class="success">	If above increased, Decrease @cite dukeSME	</span>|<span class="success">	If above increased, Decrease @cite dukeSME	</span>|<span class="success">	If above increased, Decrease @cite dukeSME	</span>|<span class="danger">	Return to resting physiology @cite dukeSME	</span>|<span class="success">	Begins to increase to a normal level >97% @cite dukeSME	</span>|
|	Ventilate w/ O2 tank	|	O2 Source is set to Tank 1	|	210	|	270	|<span class="success">	NC @cite dukeSME	</span>|<span class="success">	NC @cite dukeSME	</span>|<span class="success">	NC @cite dukeSME	</span>|<span class="success">	NC @cite dukeSME	</span>|<span class="success">	NC @cite dukeSME	</span>|
|	Adminster Etomidate - 27 mg	|	No Etomidate in drug directory, so we use Katamine instead; Drug Onset < 1 minute	|	270	|	290	|<span class="success">	NC @cite Morgan2006Clinical; NC @cite dukeSME     	</span>|<span class="danger">	Mild  Decrease @cite Morgan2006Clinical; 5-10% Decrease @cite dukeSME     	</span>|<span class="danger">	Mild  Decrease @cite Morgan2006Clinical; 5-10% Decrease @cite dukeSME     	</span>|<span class="danger">	Mask Ventilation, so NC  	</span>|<span class="success">	NC @cite dukeSME	</span>|
|	Administer Succinycholine - 115 mg	|	Drug Onset < 1 minute	|	290	|	320	|<span class="success">	Mild  Increase @cite Morgan2006Clinical; 5-10% Decrease @cite dukeSME     	</span>|<span class="success">	Mild  Increase @cite Morgan2006Clinical; 5-10% Decrease @cite dukeSME     	</span>|<span class="success">	Mild  Increase @cite Morgan2006Clinical; 5-10% Decrease @cite dukeSME     	</span>|<span class="danger">	Mask Ventilation, so NC  	</span>|<span class="success">	Begins to drop according to preoxygenated O2 curve @cite dukeSME 	</span>|
|	Intubate	|		|	320	|	520	|<span class="success">	NC @cite dukeSME	</span>|<span class="success">	NC @cite dukeSME	</span>|<span class="success">	NC @cite dukeSME	</span>|<span class="success">	NC @cite dukeSME	</span>|<span class="success">	Begins to increase to a normal level >97% @cite PaulGBarash2009	</span>|



### Nathan
Nathan receives a bolus injection of fentanyl at a dose of 150&nbsp;micrograms at a scenario time of 50&nbsp;seconds, and a ventilator mask is placed at 140&nbsp;seconds. Due to the administration of fentanyl, there is an observed decrease in the heart rate, arterial pressures and respiration rate. There is very good agreement with the expected trends during this time period. The The respiration rate is expected to decrease by 15-25%; however, the observed increase is approximately 10%. This was acceptable due to the trend expressed in other validation resources. The oxygen wall connection loses pressure at 230&nbsp;seconds. This leads to decreasing oxygen saturation and increasing heart rate and arterial pressures due to the epinephrine response. Following this, the connection is reset to a secondary oxygen tank, and the vital signs return to normal. This behavior matches the validation trends.

<center>
<table border="0">
<tr>
    <td><img src="./plots/System/Nathan_HR.jpg" width="550"></td>
    <td><img src="./plots/System/Nathan_ArterialPressure.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/System/Nathan_RR.jpg" width="550"></td>
    <td><img src="./plots/System/Nathan_O2Sat.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/System/NathanLegend.jpg" width="800"></td>
</tr>
</table>
</center>
<center><i>Figure 10. Select outputs from the Nathan combined effects scenario.</i></center>

<center><br>
*Table 8. The Nathan scenario displays the effects of sequential bolus fentanyl injection, ventilator mask application, and oxygen wall pressure loss. The engine results show strong agreement with the validation trends.*
</center>

|	Segment	|	Notes	|	Action Occurrence Time  (s)	|	Sample Scenario Time  (s)	|	Heart Rate  (beats/min)	|	Systolic Pressure  (mmHg)	|	Diastolic Pressure  (mmHg)	|	Respiration Rate  (mmHg)	|	Oxygen Saturation  (mmHg)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Administer Fentanyl - 150ug 	|	Drug Onset < 2 minutes	|	30	|	120	|<span class="success">	Stable @cite Morgan2006Clinical; 5-10% Decrease @cite dukeSME   	</span>|<span class="success">	Stable @cite Morgan2006Clinical; 5-10% Decrease @cite dukeSME   	</span>|<span class="success">	Stable @cite Morgan2006Clinical; 5-10% Decrease @cite dukeSME   	</span>|<span class="success">	Decrease @cite Morgan2006Clinical; 15-25% Decrease @cite dukeSME   	</span>|<span class="success">	NC @cite dukeSME	</span>|
|	Apply Mask and Turn on Ventilator	|		|	120	|	210	|<span class="success">	NC @cite dukeSME	</span>|<span class="success">	NC @cite dukeSME	</span>|<span class="success">	NC @cite dukeSME	</span>|<span class="warning">	Slight decrease to due anesthesia machine resistance    NC @cite dukeSME	</span>|<span class="success">	NC @cite dukeSME	</span>|
|	O2 Wall Pressure Loss	|		|	210	|	340	|<span class="success">	If O2 decreases, Increase @cite dukeSME	</span>|<span class="success">	If O2 decreases, Increase @cite dukeSME	</span>|<span class="success">	If O2 decreases, Increase @cite dukeSME	</span>|<span class="warning">	NC @cite dukeSME	</span>|<span class="success">	If undetected long enough, Decrease @cite dukeSME	</span>|
|	Connect O2 Bottle 1 as O2 Source	|		|	340	|	840	|<span class="success">	If above increased, Decrease @cite dukeSME	</span>|<span class="success">	If above increased, Decrease @cite dukeSME	</span>|<span class="success">	If above increased, Decrease @cite dukeSME	</span>|<span class="success">	NC @cite dukeSME	</span>|<span class="success">	If above decreased, Increase @cite dukeSME	</span>|


## Showcase Scenarios Validation

Four scenarios were created to showcase the ability of the %BioGears open-source physiology engine to simulate complex and combinatory insults and interventions. The %BioGears Showcase Scenarios demonstrate the ability of the engine to fill the physiology simulation needs of the medical simulation community.

### Combat Multitrauma

A team of soldiers is conducting a patrol when an explosive device detonates, injuring one of the soldiers. The squad medic applies direct pressure to a hemorrhaging wound, and, suspecting a tension pneumothorax, performs a needle decompression. Direct pressure is not controlling the bleeding, so a tourniquet is applied. Morphine and IV fluids are administered.

See <a href="CombatMultitraumaValidation.pdf">Combat Multitrauma Validation Matrix</a> for complete details on validation.

<center>
<table border="0">
<tr>
    <td><img src="./plots/System/CombatMultitrauma_TotalLungVolume.jpg" width="550"></td>
    <td><img src="./plots/System/CombatMultitrauma_BloodVolume.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/System/CombatMultitrauma_HR.jpg" width="550"></td>
    <td><img src="./plots/System/CombatMultitrauma_O2Sat.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/System/CombatMultitraumaLegend.jpg" width="1100"></td>
</tr>
</table>
</center>
<center><i>Figure 11. Select outputs from the Combat Multitrauma showcase scenario.</i></center>

### Asthma Attack

A 40 year old female with a history of asthma is having an asthma attack. She arrives at the hospital ten minutes after the beginning of the attack. A doctor administers albuterol and her condition improves.

See <a href="AsthmaAttackValidation.pdf">Asthma Attack Validation Matrix</a> for complete details on validation.

<center>
<table border="0">
<tr>
    <td><img src="./plots/System/AsthmaAttack_TotalLungVolume.jpg" width="550"></td>
    <td><img src="./plots/System/AsthmaAttack_TidalVolume.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/System/AsthmaAttack_HR.jpg" width="550"></td>
    <td><img src="./plots/System/AsthmaAttack_O2Sat.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/System/AsthmaAttackLegend.jpg" width="700"></td>
</tr>
</table>
</center>
<center><i>Figure 12. Select outputs from the Asthma Attack showcase scenario.</i></center>

### Heat Stroke

A 25 year old male is hiking towards a rock formation to begin a recreational free climb. During the hike, he is working at ~10% of his maximum capacity. The man arrives at a rock formation and begins climbing at an intensity of ~1/2 of his maximum capacity. At the top of the rock formation, the man becomes dizzy and passes out. An off-duty medic takes action to treat a heat injury.

See <a href="HeatStrokeValidation.pdf">Heat Stroke Validation Matrix</a> for complete details on validation.

<center>
<table border="0">
<tr>
    <td><img src="./plots/System/HeatStroke_AchievedExerciseLevel.jpg" width="550"></td>
    <td><img src="./plots/System/HeatStroke_CoreTemp.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/System/HeatStroke_HR.jpg" width="550"></td>
    <td><img src="./plots/System/HeatStroke_RR.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/System/HeatStrokeLegend.jpg" width="1100"></td>
</tr>
</table>
</center>
<center><i>Figure 13. Select outputs from the Heat Stroke showcase scenario.</i></center>

### %Environment Exposure

A 17 year old female leaves her Alaskan home in the mid-winter to retrieve a newspaper. The door closes as she exits the house. She is stuck outside for 45 minutes where it is -10<sup>o</sup>C. When the woman's housemates realize that she is outside, they bring her back in and sit her next to a fire.

See <a href="EnvironmentExposureValidation.pdf">Environment Exposure Validation Matrix</a> for complete details on validation.

<center>
<table border="0">
<tr>
    <td><img src="./plots/System/EnvironmentExposure_TotalMetabolicRate.jpg" width="550"></td>
    <td><img src="./plots/System/EnvironmentExposure_O2Consumption.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/System/EnvironmentExposure_HR.jpg" width="550"></td>
    <td><img src="./plots/System/EnvironmentExposure_RR.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/System/EnvironmentExposureLegend.jpg" width="900"></td>
</tr>
</table>
</center>
<center><i>Figure 14. Select outputs from the %Environment Exposure showcase scenario.</i></center>

## Conclusion

The %BioGears Engine has successfully leveraged the CDM to model and
simulate whole body physiology. The interaction of the existing system
models was validated by looking at the combined effects of multiple actions on a variety of patients. Virtually
everything within the engine is modular and extensible. Fidelity can be
changed at any level within the system. %BioGears is a fully standalone
simulator, but can also implement or interface with external software.

# Future Work

We will continue to improve system modularity.  See the system specific methodology documents for more information:

@secreflist
  @refitem anesthesia-improvements "Anesthesia Machine Future Work"
  @refitem bloodchemistry-future "BloodChemistry Future Work"
  @refitem cardiovascular-future "Cardiovascular Future Work"
  @refitem drugs-future "Drugs Future Work"
  @refitem endrocrine-comingsoon "Endocrine Future Work"
  @refitem energy-future "Energy Future Work"
  @refitem environment-future "Environment Future Work"
  @refitem GI-futurework "Gastrointestinal Future Work"
  @refitem hepatic-comingsoon "Hepatic Future Work"
  @refitem inhaler-future "Inhaler Future Work"
  @refitem nervous-future "Nervous Future Work"  
  @refitem renal-futurework "Renal Future Work"
  @refitem respiratory-future "Respiratory Future Work"
  @refitem tissue-future "Tissue Future Work"
@endsecreflist

## Recommended Improvements

An area of potential future advancements for the %BioGears Engine includes the integration of differing time step and un-fixed time
step models. This could be achieved through simple
interpolation and down-sampling, where applicable. Other advanced system
dynamics, adaptive step sizes, and mixed fidelity techniques could also be investigated.

Investigations into the effects of biological variability and parameter (inputs and outputs) error bound and confidence intervals could be applied throughout.

# Appendices

## Data Model Implementation

BioGears

BioGearsConfiguration

BioGearsEngine

## Acronyms

CDM - Common Data Model

%ECG - Electrocardiography

CFD - Computational Fluid Dynamics

PBPK - Physiologically-Based Pharmacokinetic

SME - Subject Matter Expert
%Endocrine Methodology {#EndocrineMethodology}
=====================

Overview
========
@anchor endrocrine-abstract
Abstract
--------

The %BioGears&reg; %Endocrine System is a basic system implementation of endocrine signaling. There are currently three hormones included in the %BioGears %Endocrine system: epinephrine, glucagon, and insulin. The effects of epinephrine are modeled by leveraging the %BioGears [pharmacodynamics](@ref DrugsMethodology) model. The insulin/glucagon model has regulatory effect on the concentration of glucose in the blood, and thus controls many of the nutrient storage/burning/diffusion/synthesis models.

System Design
=============
@anchor endrocrine-background
Background and Scope
--------------------

The %Endocrine System&rsquo;s primary function is to release hormones to maintain homeostasis and regulate body functions, including metabolism, growth and development, reproduction, and behavior. The human body&rsquo;s multiple endocrine glands release a variety of hormones. Different endocrine glands respond to stimuli by stimulating synthesis or inhibiting the release of hormones. In general, hormone release is described by a negative feedback mechanism, meaning the effects of the hormone on the physiology cause a cessation of further hormone release. Positive feedback can occur with some hormone release, which leads to more hormone release as the physiologic effects of the hormone are felt. Positive feedback mechanisms are generally isolated to sexual and reproductive hormones @cite guyton2006medical. The %BioGears %Endocrine System will affect metabolic function, renal function, ion regulation, and sympathetic stimulation. The engine does not attempt to model growth, sexual, or reproductive hormones.

@anchor endrocrine-dataflow
Data Flow
---------

### Reset, Conditions, and Initialization
Hormone values are initialized where necessary during reset. The %Endocrine System contains models for diabetes mellitus, both types 1 and 2. The diabetes type 1 condition is active, while the type 2 condition is planned for a future release.

### Preprocess
Currently, all %Endocrine functionality occurs in Preprocess. Epinephrine, insulin, and glucagon are all released in their own methods here.

### Process
There is no system-specific function for Process in the %Endocrine System.

### Post Process
There is no system-specific function for Post Process in the %Endocrine System.

### Assessments
There are no assessments in the %Endocrine System.

<br>
<img src="./images/Endocrine/EndocrineDataFlow.png" width="800">
<center>
*Figure 1. The data flow for the %Endocrine System consists of a Reset, Conditions, Preprocess, Process, Post Process, and Assessments. Only the Process function has system-specific functionality for the %Endocrine System.*
</center><br>

@anchor endrocrine-features
Features, Capabilities, and Dependencies
----------------------------------------

The %BioGears %Endocrine System currently contains three hormones: epinephrine, glucagon, and insulin. Each of these hormones is released endogenously and is cleared at a specified rate. The release rate of insulin and glucagon is determined by the concentration of glucose in the blood, and the release rate of epinephrine is perturbed from the basal rate by two actions: exercise and acute stress.

@anchor endrocrine-insulin
### Insulin
Insulin is secreted by the pancreatic beta cells in response to increased blood glucose. The model utilized for insulin synthesis was obtained from a full-body insulin and glucose feedback model @cite tolic2000modeling. The synthesis rate is defined by:

\f[f(G)=\frac{R}{1+\exp \left((C-G)\left(\frac{1}{a_{1} } \right)\right)} \f] 
<center>
*Equation 1.*
</center><br>
Where *R* is the tuned basal insulin synthesis rate in microunits per minute. *C* is the upper glucose concentration set-point in grams per liter. *a<sub>1</sub>* is the lower glucose concentration set-point in grams per liter.

The equation gives the insulin synthesis rate in microunits per minute as a function of the blood glucose concentration in grams per liter. The remaining parameters are tuning constants that were adjusted to achieve the correct insulin response. Because the %BioGears %Cardiovascular System does not contain a pancreas compartment, the insulin is synthesized directly in the splanchnic. It utilizes the glucose concentration in the aorta to determine the amount of insulin to be synthesized. Insulin clearance was determined empirically at a value that allows blood concentrations to remain steady when blood glucose remains steady at 90 mg/dL.

Insulin does not have any effects attributable to the %BioGears pharmacodynamic model. Its only purpose in %BioGears is to modulate the usage of nutrients in the body.

@anchor endrocrine-glucagon
### Glucagon
Glucagon is a peptide hormone secreted by the pancreatic alpha cells. It can be considered the "opposite" of insulin in that its secretion is stimulated by low blood glucose. Glucagon synthesis utilizes the same form as Equation 1 above for insulin. Whereas insulin synthesis was directly modeled from literature values and had clearance set empirically, glucagon has a validated clearance rate of 9 mL/min*kg from @cite alford1976glucagon, which was used to empirically set a synthesis rate. The same method was used to get stable glucagon blood concentrations when glucose is held steady at 90 mg/dL.

### Epinephrine
Epinephrine is released by the adrenal medulla at a basal endogenous rate of approximately 0.18 (&mu;g/min) @cite best1982release. To enable patient variability, %BioGears uses a mass-normalized basal epinephrine release rate of about 0.003 (&mu;g/kg-min). Because %BioGears does not have an adrenal gland compartment, epinephrine is released equally from the kidneys' efferent arterioles. The epinephrine clearance rate in %BioGears is constant and will not change with concentration (note that the actual amount cleared is a function of the concentration as well as the hemodynamics). A basal clearance rate of 68.7 (mL/kg-min) was computed from the basal release rate and by assuming a steady-state normal concentration of 0.032 (&mu;g/L) @cite wortsman1984adrenomedullary. The basal clearance rate is a property of the epinephrine substance in %BioGears; therefore, the clearance rate can be adjusted in the epinephrine substance file to achieve a desired steady-state concentration. A wide range of normal basal epinephrine concentrations are reported in literature @cite wortsman1984adrenomedullary @cite stratton1985hemodynamic @cite stein1998basal @cite penesova2008role @cite Zauner2000resting. 

Two stimuli, exercise and acute stress, can modify the epinephrine release rate. 

#### Exercise
The increase in epinephrine release as a function of above-basal exercise was developed using data in @cite stratton1985hemodynamic and @cite tidgren1991renal. We assume that the epinephrine clearance rate is constant; therefore, the fractional increase in epinephrine concentration described in @cite stratton1985hemodynamic and @cite tidgren1991renal can be assumed to be due to a similar fractional increase in release rate. Using that assumption, we fit a logistic function to the basal-normalized epinephrine steady-state concentrations during exercise presented in @cite tidgren1991renal. The release modifier varies from 1 to 19.75, as shown in Figure 2, meaning that the epinephrine release rate will be 19.75 times the basal release rate with maximal exercise. The model is implemented by first computing the above-basal metabolic rate and then using the generic logistic function with the appropriate parameter values to compute the release rate multiplier.

<img src="./plots/Endocrine/EpiExercise.jpg" width="600">
<center>
*Figure 2. The increase in epinephrine release during exercise is computed as a fraction of the basal rate.*
</center><br>


The whole-body physiological response to exercise is compared to empirical data in the [exercise validation](@ref energy-exercise-validation) section of the @ref EnergyMethodology report.

#### Acute Stress
Acute Stress is an action that causes an increase in the basal production of epinephrine. For details see the [Actions](@ref endocrine-actions) section of this document.

The pharmacodynamic effects of epinephrine are applied using the sigmoid 'E<sub>max</sub>' model described in [phamacodynamics](@ref drugs-pharmacodynamics) section of the @ref DrugsMethodology report. Maximum effects for the model were derived from data in @cite clutter1980epinephrine, @cite stratton1985hemodynamic, and @cite tidgren1991renal. Note that these researchers were examining the relationships between epinephrine and exercise, and not the relationships between epinephrine and acute stress. For that reason, the model of epinephrine release during the acute stress action is more phenomenological.

@anchor endocrine-hormone-factor
#### Hormone Factor
The hormone factor is a calculated metric relating the change in insulin and glucagon from their baseline values. Because many processes in the %Hepatic, %Gastrointestinal, and %Tissue Systems are controlled by insulin and glucagon, and because they behave in largely antagonistic ways, a method was needed to quantify the change of one with respect to the other. Thus, the "hormone factor" was introduced. Its caluclation can be seen below in Equation 2.

\f[HF=\frac{I_{t} -I_{0} }{I_{0} } -\frac{G_{t} -G_{0} }{G_{0} } \f] 
<center>
*Equation 2.*
</center><br>

Where *I<sub>t</sub>* and *G<sub>t</sub>* are insulin and glucagon vascular concentrations in a given compartment at a given time, *I<sub>0</sub>* and *G<sub>0</sub>* are post-stabilization set points for concentration in a given compartment, and *HF* is the resultant hormone factor. When there is no perturbance of the system, the hormone factor is expected to stay near 0. When the system is perturbed, either by increased blood glucose when eating a meal or decreased blood glucose in response to metabolism, the hormone factor responds accordingly. For details on the implementation of the hormone factor, see @ref HepaticMethodology and @ref TissueMethodology.

@anchor endocrine-dependencies
### Dependencies

The %Endocrine System is dependent on the key systems of the %BioGears Engine. The %Endocrine System is dependent on the metabolic rate determined in the %Energy System. The insulin and glucagon synthesis rates are dependent on the concentration of glucose in the %Cardiovascular System, which fluctuates depending on changes in the %Gastrointestinal and %Tissue Systems.

@anchor endrocrine-assumptions
Assumptions and Limitations
---------------------------

Epinephrine has a variety of effects in the human body, but all of the effects do not occur at the same epinephrine concentration thresholds. For example, a patient with a slowly increasing blood concentration of epinephrine would experience an elevated heart rate before they experience a decrease in diastolic pressure @cite clutter1980epinephrine. The current state of the %BioGears pharmacodynamic model cannot replicate this behavior, and so a threshold was chosen to minimize overall error. Furthermore, it is known that there is interplay between the blood concentration of epinephrine and that of glucose and insulin, but this interaction is not modeled. Another effect of the current pharmacodynamic model is that a higher concentration of epinephrine is needed in the blood to evoke the desired responses for an acute stress action. Additionally, the clearance rate of epinephrine is constant in %BioGears, whereas there is evidence that the clearance rate of epinephrine changes with blood concentration @cite clutter1980epinephrine. %BioGears does not currently model the physiological effects of epinephrine deficiency, so a diminished blood concentration of epinephrine will have no effect in %BioGears. Finally, the only physiologic trigger for increased epinephrine release is exercise. Many insults modeled in %BioGears are known to cause increased concentrations of epinephrine due to pain or stress, but this release is not directly included in those insult models. Instead, a separate Acute Stress action must be specified to account for this epinephrine release.

The insulin and glucagon response is currently only dependent on the blood glucose level. In reality, synthesis is dependent on both blood protein and blood glucose concentrations. The current dependence on just blood glucose is a reasonable assumption, because the resting synthesis rate has been accurately validated. This assumption holds true unless a meal consisting of only protein occurs.

The design of the endocrine system favors normal resting physiology, and doesn't perfectly replicate agitation. For example, when eating a glucose-heavy meal, the blood glucose level doesn't rise and fall as would be expected in a brief timescale. This is primarily due to two factors: an insulin synthesis process that responds instantly to glucose fluctuations, and a consumption process that uses free glucose as the highest-priority fuel. These factors work to keep the system in homeostasis at baseline values, and they perhaps perform too well, keeping blood glucose very steady. This also affects conditions like diabetes, where large fluctuations due to glucose consumption are expected, but behavior is a bit less extreme and more steady than expected, and changes require larger simulation times. These limitations will be improved as the %Endocrine system is improved in future releases.

@anchor endocrine-actions
Actions
-------
### Insults
#### Acute Stress
The Acute Stress action is used to directly modify the production of epinephrine. In the human body, stress or anxiety causes a sympathetic release of epinephrine and norepinephrine from the adrenal medullae. In %BioGears, an Acute Stress action is rated a severity between 0 and 1, and that severity is used to modify the epinephrine release rate linearly. A severity of 0 will return the rate to the normal, basal level, while a severity of 1 will increase the release rate by a factor of 30. This model was chosen to approximate the physiological responses observed during mild pain, mental stress tests, and panic attacks @cite greisen2001acute @cite herd1991cardiovascular @cite wilkinson1998sympathetic.

Results and Conclusions
=======================

@anchor endrocrine-validation
Validation - Resting Physiologic State
--------------------------------------

In the %BioGears Engine, the hormone epinephrine is the same substance as the synthetic epinephrine available for injection. This generic epinephrine substance was validated as part of the drugs validation detailed in the @ref drugs-validation-pharmacodynamic "Drugs Validation's Table&nbsp;1". At resting physiological levels, epinephrine has no pharmacodynamic effects.

Validation - Actions
--------------------------------------
@anchor endocrine-acute-stress
### Acute Stress
The effects of epinephrine release on the physiology can be clearly seen by triggering an Acute Stress action. The patient in this scenario undergoes three bouts of Acute Stress, with the first representing mild pain, the second representing mental stress, and the third representing a panic attack. The severity levels were chosen by checking the blood concentration of epinephrine to ensure it met published values. However, a known issue with epinephrine modeling in the current %BioGears release is that higher concentrations of epinephrine are needed to produce the effects noted in literature. In order to achieve the physiological effects, the severity was proportionally scaled up. This shortcoming is the reason for the failing epinephrine concentrations, but it allows for the other effects of epinephrine to be modeled, which has beneficial effects for other systems utilizing epinephrine. Heart rate behaves as expected, but blood pressure effects aren't always in line with expectations. This is likely an effect of baroreceptor reflex counterbalancing epinephrine effects. Figure 3 shows the blood concentration and select effects with the acute stress action.

<center>
<table border="0">
<tr>
    <td><img src="./plots/Endocrine/AcuteStressBloodEpi.jpg" width="550"></td>
    <td><img src="./plots/Endocrine/AcuteStressHR.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/Endocrine/AcuteStressSystolic.jpg" width="550"></td>
    <td><img src="./plots/Endocrine/AcuteStressDiastolic.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/Endocrine/AcuteStressLegend.jpg" width="1100"></td>
</tr>
</table>
</center>
<center><i>
*Figure 3. Epinephrine is released in response to Acute Stress actions.*
</i>
</center><br>

<center>
*Table 1. Actions associated with the %Endocrine System were validated by comparing the engine output to expected trends and data. Engine results show favorable agreement (green), some agreement (yellow), or bad agreement (red). Results mostly matched expected trends.*
</center>														
|	Action	|	Notes	|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Arterial Epinephrine (ug/L)	|	Systolic Pressure (mmHg)	|	Diastolic Pressure (mmHg)	|	Heart Rate (1/min)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Acute Stress	|	Severity 0.4; mild pain	|	20	|	210	|<span class="danger">	90% Increase @cite greisen2001acute Increase @cite herd1991cardiovascular	</span>|<span class="danger">	15 mmHg Increase @cite herd1991cardiovascular	</span>|<span class="danger">	17 mmHg Increase @cite herd1991cardiovascular	</span>|<span class="warning">	Increase @cite herd1991cardiovascular	</span>|
|	Acute Stress	|	Severity 0	|	220	|	610	|<span class="success">	.034 ug/L @cite wortsman1984adrenomedullary	</span>|<span class="success">	100-140 mmHg @cite Leeuwen2015laboratory	</span>|<span class="success">	60-90 mmHg @cite Leeuwen2015laboratory	</span>|<span class="success">	72 @cite guyton2006medical	</span>|
|	Acute Stress	|	Severity 0.8; mental stress	|	620	|	810	|<span class="danger">	37-273% Increase @cite herd1991cardiovascular .09 @cite wilkinson1998sympathetic	</span>|<span class="danger">	15 mmHg Increase @cite herd1991cardiovascular 143.8 @cite wilkinson1998sympathetic	</span>|<span class="success">	8 mmHg Increase @cite herd1991cardiovascular	</span>|<span class="success">	Increase @cite herd1991cardiovascular  82 @cite wilkinson1998sympathetic	</span>|
|	Acute Stress	|	Severity 0	|	820	|	1210	|<span class="success">	.034 ug/L @cite wortsman1984adrenomedullary	</span>|<span class="success">	100-140 mmHg @cite Leeuwen2015laboratory	</span>|<span class="success">	60-90 mmHg @cite Leeuwen2015laboratory	</span>|<span class="success">	72 @cite guyton2006medical	</span>|
|	Acute Stress	|	Severity 1; panic attack	|	1220	|	1410	|<span class="danger">	54-199% Increase @cite wilkinson1998sympathetic	</span>|<span class="success">	NC to Slight Increase @cite wilkinson1998sympathetic	</span>|<span class="success">	NC to Slight Increase @cite wilkinson1998sympathetic	</span>|<span class="success">	85-108 @cite wilkinson1998sympathetic	</span>|

Conditions
--------------------------------------
@anchor endocrine-conditions
### Diabetes Type 1
Diabetes mellitus type 1, also known as insulin-dependent diabetes mellitus, affects the pancreas gland's ability to produce insulin. An autoimmune response results in the breakdown of the pancreatic beta cells, where insulin is synthesized. The lack of insulin in the body interrupts the body's metabolic mechanisms, resulting in a rise in blood glucose concentration. Another complication of rapid onset is a rise in ketone production, possibly resulting in diabetec ketoacidosis @cite boron2012medical.

The diabetes type 1 condition in %BioGears works directly on the insulin synthesis in the %Endocrine system. A severity value determines the extent of the reduction in endogenous insulin production, with a value of 1 meaning the body produces no insulin, and a value of 0 meaning there is no reduction in insulin production, in which case the condition would have no effect. This value is directly applied to the amount of insulin synthesized each timestep, meaning a patient with a diabetes type 1 severity of .5 would produce half the insulin of a healthy patient at the same blood glucose level.

Because of the way %BioGears models the @ref endocrine-hormone-factor "hormone factor", a complete elimination of insulin production without a modification to glucagon would see glucagon production also approaching zero as the blood sugar rises, and so metabolic processes would reach a state much like that in a healthy, fed patient. As shown in @cite brown2008too, patients with diabetes type 1 show acute glucagon responses similar to those in healthy patients, with values tending toward 40 ng/L. Thus, glucagon synthesis is also modified in %BioGears when a diabetes type 1 condition is present. Glucagon is allowed to respond normally to blood sugar down to a minimum rate determined by the severity, where production is restricted to an amount that results in a blood concentration of about 40 ng/L in the most severe case, and no impact in a case with 0 severity.

The diabetes type 1 condition exhibits the expected short-term results in %BioGears. As shown in figure 4 below, a patient with a severity 1 diabetes type 1 condition sees an increasing trend in blood sugar over time. During this period, ketone production increases, reaching nearly ketotic levels. Glucagon concentration also drops in response to the high blood sugar, but not lower than around 40 ng/L. When insulin is administered, blood sugar drops, glucagon returns to near normal, and blood ketones dip, as expected.

<center>
<table border="0">
<tr>
    <td><img src="./plots/Endocrine/DiabetesType1Glucose.jpg" width="550"></td>
    <td><img src="./plots/Endocrine/DiabetesType1Insulin.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/Endocrine/DiabetesType1Glucagon.jpg" width="550"></td>
    <td><img src="./plots/Endocrine/DiabetesType1Ketones.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/Endocrine/DiabetesType1Legend.jpg" width="250"></td>
</tr>
</table>
</center>
<center><i>
*Figure 4. An insulin bolus returns blood sugar levels elevated by Diabetes Type 1 to normal.*
</i>
</center><br>

### Diabetes Type 2
Diabetes mellitus type 2, also known as non-insulin-dependent diabetes mellitus, is a long-term disorder that has many of the same symptoms as type 1 diabetes. Rather than resulting completely from impaired insulin synthesis, however, the symptoms of type 2 diabetes also stem from the body's resistance to insulin. That is, the amount of insulin that would perfectly maintain the metabolic state of a healthy patient would have less of an effect on a patient with type 2 diabetes. Both insulin production impairment and insulin resistance combine to result in the behavior commonly classified as diabetes type 2 @cite kahn2003relative. Similarly to type 1 diabetes, type 2 diabetes results in elevated blood glucose, but a differentiating point is the blood insulin, which is elevated in type 2 diabetes @cite guyton2006medical.

The diabetes type 2 condition in %BioGears has two components: an insulin production impairment scale and an insulin resistance scale. The insulin production impairment severity is exactly as in diabetes type 1. The insulin resistance severity is also a 0 to 1 severity value, where 0 means that the patient responds to insulin in a normal fashion, and 1 means that insulin has no effect in the system. Whereas in diabetes type 1, the only model effects were limitations on insulin and glucagon production, in diabetes type 2, the insulin model is modified a bit. The normal insulin production curve in %BioGears is based on @cite tolic2000modeling, but because its range is restricted to relatively normal blood glucose levels, this had to be modified. An insulin production multiplier was introduced that scaled insulin production for very high glucose values, as seen in @cite guyton2006medical. Figure 5 shows how insulin production is modified nonlinearly with increasing blood glucose.

<img src="./plots/Endocrine/DiabetesType2InsulinProduction.jpg" width="600">
<center>
*Figure 5. Insulin production is modified in diabetes type 2.*
</center><br>

This curve accounts for the possibility of increased glucose, but the genesis of increased glucose is caused by the modification to the computed @ref endocrine-hormone-factor "hormone factor". Because the effect of insulin resistance is the body needing more insulin to register an equivalent effect, the insulin resistance severity is used to scale down the insulin concentration used in the hormone factor calculation, lowering its effect. With this, the effects of insulin in %BioGears (storage of glucose, deactivation of gluconeogenesis and ketogenesis, storage of fats and amino acids, etc.) are reduced, causing the elevated blood nutrients observed. Figure 6 shows the results of a glucose tolerance test on a patient with the diabetes type 2 condition. As expected, elevated glucose levels and amino acid levels are observed, and insulin response is much higher than in a non-diabetic patient. The slow return to lower blood glucose levels in diabetes type 2 is a known issue, and will be investigated in future releases.

<center>
<table border="0">
<tr>
    <td><img src="./plots/Endocrine/DiabetesType2StomachContents.jpg" width="550"></td>
    <td><img src="./plots/Endocrine/DiabetesType2Glucose.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/Endocrine/DiabetesType2Insulin.jpg" width="550"></td>
    <td><img src="./plots/Endocrine/DiabetesType2Glucagon.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/Endocrine/DiabetesType2AA.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/Endocrine/DiabetesType2Legend.jpg" width="250"></td>
</tr>
</table>
</center>
<center><i>
*Figure 6. Glucose consumption with the diabetes type 2 condition results in elevated glucose, insulin, and amino acids without depleting glucagon.*
</i>
</center><br>

@anchor endrocrine-conclusions
Conclusions
-----------

The hormones included in the current %Endocrine System provide a validated response to the effects modeled.

Future Work
===========
@anchor endrocrine-comingsoon
Coming Soon
-----------

Recommended Improvements
------------------------
-A larger hormone library, including sex-specific and reproductive hormones

Appendices
==========

Data Model Implementation
-------------------------

@ref EndocrineSystemTable "Endocrine" %Gastrointestinal Methodology {#GastrointestinalMethodology}
===========================

@anchor GI-overview
Overview
========

Abstract
--------

The %BioGears %Gastrointestinal System models the ingestion of macro-nutrients and their subsequent digestion and transportation into the %Cardiovascular system. 
Digestion rates for each macro-nutrient are calculated by their interaction with the corresponding stomach enzyme. Although many enzymes participate in this digestion process, %BioGears assumes lipase, amylase, and trypsin are the enzymes facilitating digestion of fat, carbohydrates, and proteins. Each macro-nutrient, after reacting with its digestive enzyme, is modeled in %BioGears as only one substance. Fat is modeled as triacylglycerol (tripalmitin), carbohydrate is modeled as glucose, and protein is modeled by a lumped amino acids substance that represents alanine. We currently do no support the various other products produced during the digestion and absorption cycle within the GI system. Once digestion takes place, the product is moved into the small intestine chyme compartment, where it is absorbed into the vasculature through the intestinal walls via the major sodium co-transporters.


@anchor GI-introduction
Introduction
------------

### %Gastrointestinal Physiology

The %Gastrointestinal tract is responsible for consuming and digesting food, absorbing nutrients, and expelling waste.
The tract consists of the stomach and intestines breaking down food into usable nutrients for the body. 
After being eaten, food is stored in the stomach, gradually digested, then released into the gut chyme. 
At this point, nutrients are either quickly absorbed into the blood through the intestinal wall or further broken down prior to absorption. 
An overview of the digestive tract is shown in Figure 1.

@image html DigestiveTractDetail.png
<center>
<i>
Figure 1. The human digestive tract is composed of several distinct sections @cite LadyofHats2006Digestive. 
Currently, the %BioGears gastrointestinal model replicates the behavior of the stomach and small intestine.
</i>
</center><br>

@anchor GI-system
System Design
=============

@anchor GI-background
Background and Scope
--------------------
The %Gastorintestinal System was created so that end users could supply and replenish water and substances in the body to counterbalance losses associated with exercise and clearance.
It provides a low-fidelity enzyme kinetic model coupled with gastrointestinal secretions to supply  the major sodium co-transporter SGLUT1. Sodium absorption is directly coupled to movement of amino acids and glucose into the vasculature. Fat is handled separately with direct absorption of triacylglycerol into the blood stream, with future work targeting an expanded lymph system to facilitate more accurate fat absorption modeling and transport. Other major ions and water have constant digestion and absorption rates. Future work to properly model the major ion transport proteins will help enhance this current implementation.

@anchor GI-dataflow
Data Flow
---------

### Reset

The stomach is initialized with configurable amounts of each macronutrient, sodium, calcium, and water.

### Preprocess

#### Gastric Secretion

A static gastric secretion flow rate of 2.46 mL/min is used to manually move water volume from the gut tissue into the stomach.

#### Digestion

Digestion of nutrients in %BioGears is modeled as an enzymatic process that aims to replicate product formation in the human body as nutrients travel through the gastrointestinal system. The enzyme-substrate interaction and product formation is modeled using Michaelis-Menten kinetics describing the following process: 

\f[\left[ E \right] + \left[ S \right] \mathbin{\lower.3ex\hbox{$\buildrel\textstyle\rightarrow\over
{\smash{\leftarrow}}$}} \left[ {ES} \right] \to \left[ E \right] + \left[ P \right].\f]
<center>
*Equation 1*
</center><br>


We model this rate formation of product as a non-linear function of substrate concentration: 

\f[\frac{{dp}}{{dt}} = \frac{{{V_{\max }}\left[ S \right]}}{{{K_m} + \left[ S \right]}}.\f]
<center>
*Equation 2*
</center><br>

Then multiply by our timestep to compute total mass of product moved into the intestinal chyme:

\f[\Delta P = \left( {\frac{{{V_{\max }}\left[ S \right]}}{{{K_m} + \left[ S \right]}}} \right)\Delta t\f]
<center>
*Equation 3*
</center><br>

Where *&Delta;P* is in units of mass and is computed each time step given the current substrate concentration on the stomach. Rate constants were taken from experimental papers, often times from in vitro studies, and can be seen in table 1. Even though these may not be perfect numbers, the mouth to cecum transit times line up well with experimental data.

<br><center>
<i>Table 1. Michaelis-Menten kinetic parameters for nutrient digestion model. Parameters for this model are taken from @cite botham1997lipolysis @cite gangadharan2009biochemical @cite ohta1986purification</i>
</center>

| Nutrient | Enzyme | Product | Vmax (mg/min) | Km (mg/mL) | 
| :---- | :---- | :---- | :---- | :---- | 
| Fat | Lipase | Triacylglycerol | 3.40 | 0.34 | 
| Carbohydrate | Amylase | Glucose | 7.11 | 3.076 | 
| Protein | Trypsin | Amino Acids | 7.1 | 0.71 | 


 
#### Chyme Secretion
The human body uses gastric sections to produce bile, which has the enzymes and pH level necessary to digest the nutrients eaten into absorbable substances that fuel metabolism. This process of secretions is generally performed by the pancreas and is a function of composition and quantity of the food eaten. %BioGears, during digestion, assumes full availability of these digestive enzymes. Substances are then secreted into the small intestine. Bile salts are critical in digestion and absorption of nutrients into the bloodstream and are needed to facilitate our model of absorption so %BioGears mimics this functionality seen in the body.

We model this functionality similarly to an enzyme kinetics model by using a non-linear function of stomach mass to compute sodium secreted. %BioGears models sodium mass transfer as a Hill-type non-linear function of total digested product in the small intestine chyme: 

\f[\Delta N{a_{\sec }} = \left( {\frac{{{V_{\max }}{N_{mass}}}}{{{K_m} + {N_{mass}}}}} \right)\Delta t\f]
<br><center>
*Equation 4 sodium secretion function*
</center>

Where *N<sub>mass</sub>* is the maximum total mass of a nutrient currently in the stomach. Sodium is currently the only supported ion to be included in this process, but the model could be extended to include hydrogen and bicarbonate in order to validate pH levels of the chyme.

#### Absorption
Absorption of digested nutrients is facilitated through sodium uptake. In a similar manner to secretion absorption is modeled as a non-linear function of product mass in the chyme, as in equation 4. Once absorbed sodium mass is computed, co-transport of glucose and amino acids begins. We assume that 1 mg of sodium can transport 2 mg of glucose or 1 mg of amino acids in our simplified model. This allows for total transit times that are in line with validated data @cite ladas1989reproducible

A static absorption flow rate of 3.3 mL/min is used to manually move water volume from the gut chyme into small intestine vasculature.

### Conditions

#### ConsumeMeal

The ConsumeMeal condition provides a nutrition object along with an elapsed time since those nutrients were in the stomach.
The GI System will then zero out the contents of the stomach and gut chyme and add the nutrient data provided in the condition to the stomach.
The digestion algorithm is executed for the provided elapsed time, removing nutrient mass from the stomach and adding it to substance quantities in the gut chyme.
Next, the absorption algorithm will be executed for the provided elapsed time, removing substance mass from the gut chyme and distributing that mass throughout the blood and tissues.

### Actions

#### ConsumeNutrients
Each nutrient is optional in the ConsumeNutrients action. If a nutrient mass is provided with no accompanying digestion rate, the engine will default to the rate specified in the %BioGears configuration. 
If a digestion rate is provided for a macronutrient, it will be combined with the current stomach digestion rate for that macronutrient by volume weighting the two rates together.

### Process
There is no system specific function for Process in the %Gastrointestinal System.

### Post Process
There is no system specific function for Post Process in the %Gastrointestinal System.

### Assessments
There are no system specific assessments in the %Gastrointestinal System.

<br>
<img src="./images/GI/GastrointestinalDataFlow.png" width="600">
<center>
*Figure 2. The data flow for the %Gastrointestinal System consists of a Reset, Conditions, Preprocess, Process, Post Process, and Assessments. 
Only the Reset, Conditions, and Preprocess have system-specific functionality.*
</center><br>

@anchor GI-features
Features, Capabilities, and Dependencies 
----------------------
### Circuit

@anchor GI-circuit
@image html GICircuit.png
<center>
<i>
Figure 3. The %GI circuit is made up of nodes and paths with elements and is connected to extravascular tissue and the cardiovascular system.
</i>
</center><br>

The constant secretion of fluid into the digestive tract allows for macronutrients to be slowly digested and eventually absorbed through the intestinal wall @cite hall2011guyton. In %BioGears, these macronutrients are placed into the CV system via an absorption rate that mirrors that of the human GI tract. The calculated flow rate transports substance from the small intestine C1 node to the CV system's small intestine 1 node @ref GI-circuit. Similar transport connects the tissue and the small intestine and determines water balance and secretion rate.
The digestion rate of macronutrients is heavily dependent on the nature of the source food, as the digestive tract needs to free substances from the rest of the food's molecular structure before it can be absorbed @cite wolever1991glycemic.
Fats spend several hours being emulsified by the small intestine and are then quickly absorbed once exposed to pancreatic enzymes @cite hall2011guyton. 
Many of the small ions are quickly absorbed with active transport in a manner similar to the %Renal system @cite hall2011guyton.
An exception to this is calcium, which is carefully regulated by parathyroid hormone to ensure that the body's requirements are met @cite hall2011guyton.
Water is absorbed via the osmotic gradient that either naturally exists between the gut chyme and the blood or that is created by the active transport of substances across the intestinal wall @cite hall2011guyton.


@anchor GI-dependencies
Dependencies
------------
The %Gastrointestinal System is not dependent on any other system, but other systems are dependent on GI to provided and replenish substances that they use.

@anchor GI-assumptions
Assumptions and Limitations
---------------------------

Currently there is no modeling of defecation or the lower portion of the digestive tract. This means that all food that is ingested will eventually be absorbed.
The current digestion methodology is rudimentary and assumes all digestive rates are constant and move with the water flow after leaving the stomach. Active transport is used across cell walls with no diffusion currently implemented. Contraction of the small intestines to allow for transport and defecation is also not modeled.

@anchor GI-results
Results and Conclusions
=======================
Validation - Resting Physiologic State
--------------------------------------

The %BioGears %Gastrointestinal System was validated qualitatively by observing the relative changes in the substance and nutrient masses between the stomach contents
and the gut contents. The transport to the CV system was then shown by demonstrating an increase in substance concentration in the blood.
Concentrations are not solely a function of GI; other systems are potentially using substances in the blood.

<center>
*Table 2. Macronutrient mass in the stomach, gut, and blood as a function of time.*
</center>

|  Macronutrient/Substance |                           Stomach Mass(g) vs. Time(s)                            |                               Gut Mass(g) vs. Time(s)                                  |                     Blood Concentration (ug/mL) vs. Time (s)                         |
|:---:                     |:---:                                                                             |:---:                                                                                   |:---:                                                                                 |
| Carbohydrate/Glucose     | <img src="./plots/GI/CarbMeal_StomachCarbs.jpg" height="100" width="200"> | <img src="./plots/GI/CarbMeal_IntestineGlucose.jpg" height="100" width="200">    | <img src="./plots/GI/CarbMeal_BloodGlucose.jpg" height="100" width="200">    |
| Fat/Triacylglycerol           | <img src="./plots/GI/FatMeal_StomachFat.jpg" height="100" width="200">          | <img src="./plots/GI/FatMeal_IntestineTriacylglycerol.jpg" height="100" width="200"> | <img src="./plots/GI/FatMeal_BloodTriacylglycerol.jpg" height="100" width="200"> |
| Protein/AminoAcids             | <img src="./plots/GI/ProteinMeal_StomachProtein.jpg" height="100" width="200">      | <img src="./plots/GI/12hr_IntestineUrea.jpg" height="100" width="200">       | <img src="./plots/GI/12hr_BloodUrea.jpg" height="100" width="200">       |
| Calcium                  | <img src="./plots/GI/12hr_StomachCalcium.jpg" height="100" width="200">      | <img src="./plots/GI/CarbMeal_IntestineCalcium.jpg" height="100" width="200">    | <img src="./plots/GI/12hr_BloodCalcium.jpg" height="100" width="200">    |
| Sodium                   | <img src="./plots/GI/12hr_StomachSodium.jpg" height="100" width="200">       | <img src="./plots/GI/12hr_IntestineSodium.jpg" height="100" width="200">     | <img src="./plots/GI/12hr_BloodSodium.jpg" height="100" width="200">     |
| Water                    | <img src="./plots/GI/12hr_StomachWater.jpg" height="100" width="200">        | <img src="./plots/GI/12hr_IntestineVolume.jpg" height="100" width="200">          | <img src="./plots/GI/CarbMeal_BV.jpg" height="100" width="200">             |


As seen in Table 2, the macronutrient masses in the stomach are depleted based on different digestion rates and eventually reach zero. 
The center column of plots shows the associated mass of substances in the gut chyme increase as the nutrients are removed from the stomach.
Concentrations are provided in the far right column demonstrating the effect of the increasing mass on the blood concentrations.
These overall trends meet the expectations of the model performance.

Validation - Actions
--------------------
### Consume Meal Standard Validation 
Mouth to cecum transit times are validated against experimental data @cite ladas1989reproducible. Total transit is on the slow end of reported ranges of patients ingesting a standard meal but still within the deviation found in the study. 

<center>
*Table 3. Validated transit times are in line with experimental data @cite ladas1989reproducible @cite seimon2013gastric.*
</center>

|	Actions	|	Notes	|	Occurrence Time (s)	|	Sampled Scenario Time (hr)	|	Gastric emptying time (carbs) (min)	|	Mouth-to-Cecum Transit Time (min)	|	Blood Glucose Concentration return time (min)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	---	|	Give a meal to the patient: 26g carbohydrate, 6g protein, 30g fat	|	30s	|	4hr	|<span class="success">	40-90 min @cite ladas1989reproducible	</span>|<span class="success">	64-144 min @cite ladas1989reproducible	</span>|<span class="success">	120-180 min @cite seimon2013gastric	</span>|


<center>
<table border="0">
<tr>
    <td><img src="./plots/GI/StandardMeal_IntestineSodium.jpg" width="550"></td>
    <td><img src="./plots/GI/StandardMeal_IntestineGlucose.jpg" width="550"></td>
</tr>
</table>
</center>
<center>
<i>
Figure 4. The plots show sodium secretion in the chyme as a response to digested glucose load after a standard meal. The secretion and absorption functionality work together to absorb the nutrients into the body at the proper time scales. 
</i>

</center><br>

<center>
<table border="0">
<tr>
    <td><img src="./plots/GI/StandardMeal_StomachCarb.jpg" width="550"></td>
    <td><img src="./plots/GI/StandardMeal_BloodGlucose.jpg" width="550"></td>
</tr>
</table>
</center>
<center>
<i>
Figure 5. Times taken for digestion to take place to affect blood glucose levels are seen here. The internal nutrient kinetics models use glucose as it enters the bloodstream or stores excess as glycogen. Times scales are at values seen in experimental data. 
</i>
</center><br>

@anchor GI-conclusions
Conclusions
-----------

The %BioGears %Gastrointestinal System lays a foundation for future efforts in modeling digestive physiology. The flexible definition of food allows for easy extensibility
of the system by creating the ability for users to add to a library of meals. 

@anchor GI-futurework
Future Work
===========

Coming Soon
-----------
The osmolarity of the digested material is not currently measured. Implementing this would allow for the creation of a more realistic osmolar gradient to determine the water absorption rate and gastric secretion. Permeability of the tissue matrix and substance concentration levels will be added to allow for actual diffusion of substances and fluid instead of active transport being implemented currently.

@anchor GI-recommended
Recommended Improvements
------------------------

- More %Substances: By expanding the list of %substances %BioGears works with, the %Gastrointestinal model will inherently gain fidelity as mechanisms
	are constructed to properly handle new inputs.

- Defecation: The %Gastrointestinal system should eventually be able to handle undigestible material and the fluid loss due to defecation. This
	will also allow for a variety of disease states to be implemented.
	
- Vomiting: This can be added either as a symptom or an intervention to aid someone who has ingested a harmful substance.

@anchor GI-appendices
Appendices
==========

Acronyms
--------

GI - %Gastrointestinal

CV - %Cardiovascular


Data Model Implementation
-------------------------

@ref CardiovascularSystemTable "Cardiovascular"

@ref GastrointestinalSystemTable "Gastrointestinal"

Compartments
------------

The GI system is integrated into the %cardiovascular system, please refer to the compartement description associated with @ref cardiovascular-appendices "Cardiovascular Methodology".%Inhaler Methodology {#InhalerMethodology}
==============================

@anchor inhaler-overview
Overview
========
@anchor inhaler-abstract
Abstract
--------

The %BioGears&reg; %Inhaler Model is a generic representation of a pressurized metered dose inhaler (pMDI).  The results show an excellent correlation with the expected trends.  Future work will address the current limitations of the system, including incorrect use of the device and re-inhalation of exhaled drug dose.

<center><img src="./images/Inhaler/Inhaler_Figure01.png" width="400"></center>
<center>
<i>Figure 1. A pressurized metered dose inhaler @cite Wikiphoto2016How.</i>
</center>
@anchor inhaler-intro
Introduction
------------

pMDIs are portable, user-actuated devices used to administer drug doses into the lungs and airways. The drug is either a liquid or a solid in liquid suspension, stored in a small pressurized cylinder with a propellant spray. When the device is actuated, a fixed amount of the pressurized suspension is released through a nozzle. As the suspension reaches atmospheric pressure at the nozzle exit, the nozzle geometry and rapid evaporation of the propellant causes the suspension to be aerosolized into droplets that are sufficiently small to remain in suspension in the air for several seconds. The properties of propellant and nozzle design, strongly dictate the efficiency with which the liquid suspension is aerosolized @cite newman2005principles .  

The user coordinates actuation of the device with a controlled inhalation, so that the aerosolized dose is inhaled deep into the lungs. Immediately prior to actuation, the user exhales as much as possible to their minimum lung volume (residual volume). As the device is actuated, the user begins a steady inhalation until they have reached their maximum inhalation volume (vital capacity).  At the end of the inhalation, the user holds their breath for up to 10 seconds and then exhales. In this way, the user maximizes the amount of inhaled dose that is deposited in the lungs. 

To help poor coordinators, the pMDI is often used in conjunction with a spacer. A spacer is simply a closed volume in front of the nozzle that is held between the pMDI and the mouth. When the pMDI is actuated, the aerosolized dose is initially contained within the spacer. If the user begins inhalation too late or breaths in too shallow, the spacer holds the aerosolized dose and extends the time over which the user can inhale and still maximize lung deposition.  
@anchor inhaler-system
System Design
=============

Background and Scope
--------------------

### Aerosol Deposition 

Most pMDI's deliver 100-200 &mu;g of a substance per actuation, but a large portion of the dose does not actually make it into the lungs. Many studies, including one by Newman @cite newman1991improvement in 1991, showed that as little as 20% of the drugs delivered by the pMDI actually make it into the lung tissue. The actual amount depends strongly on droplet size and the technique of the user. Users categorized as "good coordinators" average 20% lung deposition, but can often get much better depending on how "good" they are.  "Bad coordinators" - those that actuate the inhaler too early, too late, or fail to inhale sufficiently - do much worse. Table 1 summarizes the results from Newman's study.  

<center>
<i>Table 1. Mean percentage or aerosol dose located at various sites after inhalation by Good and Bad coordinators @cite newman1991improvement . </i>
</center>

| Site | Correct: "Good Coordinators" | Incorrect: "Bad Coordinators" |
| --- | --- | --- |
| Lungs | 18.6% | 7.2% |
| Oropharynx | 64.6% | 67.7% |
| Actuator | 16.1% | 23.5% |
| Exhaled | 0.7% | 1.7% |

Regardless of how coordinated the user is with the device, a large portion of the dose is generally deposited in the oropharynx, lost in the actuator and nozzle, or is exhaled. In %BioGears, the portion of aerosolized dose lost to deposition in oropharynx region is estimated using an empirical relation developed for a study conducted by Yeh in 1996 @cite yeh1996comparisons ,

\f[ODF = \frac{1}{{1 + {{\left( {\frac{{\rho {d^2}Q}}{{30000}}} \right)}^{ - 1.37}}}}\f]
<center>
*Equation 1.*
</center><br> 

In this relation, the oral deposition fraction (ODF) is a function of droplet density, <i>&rho;</i>(g/cm<SUP>3</SUP>), droplet diameter, d (&mu;m), and flow rate, Q (cm<SUP>3</SUP>/sec). Nozzle and actuator losses are generally design dependent and will vary depending on the specific pMDI used. In %BioGears, nozzle losses are treated as a property of the pMDI.

### Requirements
The inhaler implementation is used to meet the requirement to administer a beta agonist (e.g., albuterol) as an intervention action for acute asthma.

### Approach
In %BioGears, the pMDI is modeled as a simple circuit (Figure 2) conditionally appended to the existing %Respiratory System circuit model. The inhaler circuit consists of a single inhaler node with a fixed volume connected to the external environment &ldquo;ground.&rdquo; If a spacer is specified in the scenario, the volume of the spacer is added to the inhaler node volume. When the pMDI is actuated during a scenario, the inhaler circuit is connected to the mouth node of the respiratory model, replacing the connection from the mouth to the external environment. Atmospheric air initially fills the inhaler volume and airflow into and out of the respiratory system temporarily passes through the inhaler node. 

@image html Inhaler_Figure02.png
<center>
<i>Figure 2. %Inhaler circuit (red) connected to the %Respiratory System circuit. The inhaler circuit consists of a single inhaler node that is added the respiratory circuit when the pMDI is actuated.</i>
</center><br>

To model coordinated use of the pMDI, conscious breathing actions were implemented.  The end volumes and time lengths for exhalation, inhalation, and holding breath are specified as parameters for these actions, along with pMDI actuation timing.  When the pMDI is actuated, the drug dose is added to the atmospheric air in the inhaler node. Depending on airflow direction, the air/drug mixture in the inhaler node flows into the mouth or out into the external environment. A fraction of the drug that flows through the mouth and trachea is removed from the system per Equation 1. The remaining drug mass flows into and out of the lungs and alveoli. The drug mass in the inhaler node is assessed each time step during the scenario. When the drug mass in the inhaler node drops to approximately zero, the inhaler is disconnected from the respiratory circuit.

Once in the alveoli, the drug diffuses into the blood stream. See the @ref DrugsMethodology for PK and PD effects. 
@anchor inhaler-dataflow
Data Flow
---------

### Initialize
The volume of the inhaler node is filled with atmospheric air at the beginning of the scenario. 

### Preprocess
The system determines if and when an inhaler actuation command is initiated. 

When an actuation command is detected, the following occurs:
- The inhaler circuit is connected to the respiratory circuit.
- If specified in the scenario, the spacer volume is added to the inhaler node volume.
- The drug dose is added to the inhaler node volume and mixed with air. 
- The system begins tracking the mass of drug remaining in the inhaler node.

While the inhaler circuit remains connected to the respiratory circuit, the following occurs:
- A fraction of drug mass in the trachea is removed from the system each time step to model onopharynx deposition.
- If the drug mass in the inhaler node drops to near zero, the inhaler circuit is disconnected.

After the inhaler circuit is disconnected from the respiratory circuit, the following occurs:
- If specified in the scenario, the spacer volume is subtracted from the inhaler node volume.

Coordinated use of the pMDI is dependent on the conscious breathing actions executed by the %Respiratory System during the Preprocess step. Refer to the @ref RespiratoryMethodology for conscious breathing actions.

### Process

The current %BioGears implementation has no specific Process functionality for the inhaler. %Inhaler processing is currently done in the %Respiratory System with the combined circuit methodology.

### Post Process

The Post Process step moves values calculated in the Process step from the next time step calculation to the current time step calculation. The current %BioGears has no specific Post Process functionality for the inhaler. All postprocessing is done in the %Respiratory System with the combined circuit methodology.

The following figure presents the data flow of the %Inhaler data processing steps. 

<center><img src="./images/Inhaler/Inhaler_Figure03.png" width="550"></center>

<center>
<i>Figure 3. All primary inhaler activity occurs during the preprocessing step.</i>
</center>
@anchor inhaler-features
Features and Capabilities
-------------------------

### Features

#### Connecting to the %Respiratory Circuit
When the pMDI is used by a patient, there is a direct connection that allows air to flow freely between the inhaler and respiratory circuits.  Both individually defined circuits are combined into a single circuit that is then used for calculations.

#### pMDI (%Inhaler) Settings
<center>
*Table 2. The table shows the basic settings parameters used in BioGears as inputs to use the pMDI*
</center>

| Preset Parameter | Description |
| --- | --- |
| Substance | The name of the substance being administered.  The substance must be one of the gaseous substances handled by BioGears. |
| Metered Dose (ug) | Dose mass. |
| Nozzle Loss | Fraction of the dose that is lost in actuation nozzle. |
| Spacer Volume (mL) | Volume of the spacer used with the pMDI.  Optional. |

#### Important drug properties
Losses due to drug deposition in the oropharynx region are estimated based on the droplet diameter and density. The droplet density used by the model is the density of the suspension (liquid solution including the drug).  Typically, the density of these suspensions is close to that of water @cite yeh1996comparisons .  Droplet diameters typically range between 0.5 and 6.0 &mu;m. 
@anchor inhaler-dependencies
### Dependencies
The %BioGears pMDI interacts with the %Respiratory System through a connection that delivers atmospheric air and drugs into the %Respiratory System (@ref RespiratoryMethodology). The two systems are connected via a path joining the mouth node of the %Respiratory System to the inhaler node (see Figure 2). Before and after the pMDI is actuated, the mouth node of the %Respiratory System is connected to the atmosphere through the %Environment System that serves as a ground node for the %Respiratory System. 

When the pMDI is actuated, a network of combined circuits that include the elements from both the %Respiratory System and the inhaler is created. When the combined circuit is generated at the run-time, the ground environment node connected to the mouth node of the %Respiratory System is replaced by the inhaler node that represents the nozzle, becoming one combined circuit.

Coordinated use of the pMDI depends on the uses of the conscious breathing action/command set. Refer to the @ref RespiratoryMethodology for details.
@anchor inhaler-assumptions
Assumptions and Limitations
---------------------------
- The modeling approach used by %BioGears assumes a monodisperse aerosol (constant droplet diameter). pMDI&rsquo;s are known to generate a polydisperse aerosol. Moreover, evaporation decreases droplet diameter within a short timescale. Both of these effects are known to be significant but are not currently addressed in the %BioGears inhaler model. 
- The system transport of the administered drug is treated as a gas. As a result, the model does not address drug deposition and absorption through respiratory airway surfaces other than the alveoli. All diffusion into the bloodstream occurs as a result of gas transfer from the alveoli into the pulmonary capillaries. Other transport mechanisms, such as the dissolving of deposited mass into the mucous membranes or the ingestion of orally-deposited drug mass are not addressed. 
- The re-inhalation of exhaled drug mass is not addressed. Drug mass that is exhaled is presumed lost.
- Losses due to drug deposition on airway surfaces during exhalation are not addressed. Although losses during inhalation are handled, the flow geometry is more complex during exhalation and is more difficult to model. 
@anchor inhaler-actions
Actions
-------

### Actuation

pMDI actuation is part of the conscious breathing action/command set described in the @ref RespiratoryMethodology.  
@anchor inhaler-events
Events
------

### Multiple Actuations

If the inhaler is actuated too soon after an initial actuation, %BioGears triggers an event to alert the user.
@anchor inhaler-results
Results and Conclusions
=======================
@anchor inhaler-actionsvalidation
Validation - Actions
--------------------

pMDI actuation and coordinated breathing actions were validated in several scenarios. A summary of this validation is shown in Table 3. For each scenario, the table shows the total number of results in each category. For many investigated scenarios, the model shows good agreement with the expected trends. For the scenarios that did not match with the expected trends, improvements are planned for future %BioGears Engine releases.

<center><br>
*Table 3. Cumulative validation results for %Inhaler specific conditions and actions scenarios.*
</center>

|	Key	|
|	---	|
|<span class="success">	Good agreement: correct trends or <10% deviation from expected	</span>|
|<span class="warning"> 	Some deviation: correct trend and/or <30% deviation from expected	</span>|
|<span class="danger">	Poor agreement: incorrect trends or >30% deviation from expected	</span>|

|	Scenario 	|	Description	|	Good	|	Decent	|	Bad	|
|	---	|	---	|	---	|	---	|	---	|
|	Inhaler_OneActuation	|	Single actuation of pMDI with 90 ug dose of albuterol	|<span class="success">	12	</span>|<span class="warning">	0	</span>|<span class="danger">	0	</span>|
|	Inhaler_OneActuationWithSpacer	|	Single actuation of pMDI with 90 ug dose of albuterol using a 500 mL spacer	|<span class="success">	12	</span>|<span class="warning">	0	</span>|<span class="danger">	0	</span>|
|	Inhaler_TwoActuations	|	Two actuations of pMDI with 90 ug doses of albuterol (180 ug total)	|<span class="success">	24	</span>|<span class="warning">	0	</span>|<span class="danger">	0	</span>|
|	Inhaler_OneActuationIncorrectUse	|	Single actuation of pMDI with 90 ug dose of albuterol, 3 seconds before inhale	|<span class="success">	12	</span>|<span class="warning">	0	</span>|<span class="danger">	0	</span>|
|	Inhaler_OneActuationWithSpacerIncorrectUse	|	Single actuation of pMDI with 90 ug dose of albuterol using a 500 mL spacer, 3 seconds before inhale	|<span class="success">	12	</span>|<span class="warning">	0	</span>|<span class="danger">	0	</span>|
|		|	Total	|<span class="success">	72	</span>|<span class="warning">	0	</span>|<span class="danger">	0	</span>|

### Single Actuation with Correct Use

The single actuation scenarios with and without a spacer use the same setup and actions otherwise.  Therefore, they produce extremely similar results. Validation results for a single pMDI actuation using a metered dose of 90 &mu;g of albuterol with good coordination. Results agree very well with published data.

<center>
<table border="0">
<tr>
    <td><img src="./plots/Inhaler/Inhaler_OneActuation_TotalLungVolume.jpg" width="550"></td>
    <td><img src="./plots/Inhaler/Inhaler_OneActuation_AlbuterolConcentration.jpg" width="550"></td>
</tr>
</table>
</center>
<center><i>Figure 4. Select outputs from the single actuation scenario. With and without a spacer give extremely similar results.</i></center>

<center><br>
<i>Table 4. Validation matrix for physiological responses due to inhaler single actuation with correct use. With and without a spacer give the same results.</i>
</center>

|	Segment	|	Notes	|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Albuterol Mass in Alveoli (ug)	|	Trachea Flow - Peak Flow  (L/min)	|	Total Lung Volume (mL)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Forced Exhale	|	Patient exhales: ERV Fraction = 1.0 (100%), Period = 3 sec	|	30	|	33	|<span class="success">	0.0 ug	</span>|<span class="success">	Negative Value	</span>|<span class="success">	Residual Volume (~1.2L)	</span>|
|	Inhaler Actuation & Forced Inhale	|	Patient activates inhaler at beginning of forced inhale: Substance = Albuterol, dose = 90 ug, Nozzle Loss = 0.04. Patient inhales: IC Fraction = 1.0 (100%), Period = 5 sec	|	33	|	38	|<span class="success">	Peak 47% to 65% of total dose @cite usmani2005regional (42.4 to 59.0 ug)	</span>|<span class="success">	Positive Value	</span>|<span class="success">	Total Lung Capacity (~6.2L)	</span>|
|	Hold Breath	|	Patient holds breath: Period = 10 sec	|	38	|	48	|<span class="success">	< Peak	</span>|<span class="success">	0 L/min	</span>|<span class="success">	Total Lung Capacity (~6.2L)	</span>|
|	Normal Breathing	|	Patient returns to normal breathing	|	48	|	180	|<span class="success">	> 0.0 ug	</span>|<span class="success">	Normal	</span>|<span class="success">	Normal	</span>|

### Single Actuation with Poor Coordination

These scenarios use the inhaler about as incorrectly as possible by activating it at the moment the exhale begins - thereby, blowing the albuterol out of the inhaler into the environment.

#### Without Spacer

The incorrect use, single actuation scenarios with and without a spacer use the same setup and actions otherwise.  Therefore, they produce similar results, except for the amount of alburterol that gets into the body. Validation results for a single pMDI actuation using a metered dose of 90 &mu;g of albuterol with poor coordination agree very well with published data.  Without the spacer present, when the user exhales, the entire dose is lost to the exterior environment. In reality, some of this dose is inhaled from the remaining aerosolized mass in the surrounding air. The inhalation of exhaled albuterol is not currently modeled.

<center>
<table border="0">
<tr>
    <td><img src="./plots/Inhaler/Inhaler_OneActuationIncorrectUse_TotalLungVolume.jpg" width="550"></td>
    <td><img src="./plots/Inhaler/Inhaler_OneActuationIncorrectUse_AlbuterolMass.jpg" width="550"></td>
</tr>
</table>
</center>
<center><i>Figure 5. Select outputs from the single actuation scenario with incorrect use and without the use of a spacer.</i></center>

#### With Spacer

When the spacer is included with the inhaler, a small amount of albuterol still gets into the body.  

<center>
<table border="0">
<tr>
    <td><img src="./plots/Inhaler/Inhaler_OneActuationWithSpacerIncorrectUse_TotalLungVolume.jpg" width="550"></td>
    <td><img src="./plots/Inhaler/Inhaler_OneActuationWithSpacerIncorrectUse_AlbuterolMass.jpg" width="550"></td>
</tr>
</table>
</center>
<center><i>Figure 6. Select outputs from the single actuation scenario with incorrect use and with the use of a spacer.</i></center>

<center><br>
<i>Table 5. Validation matrix for physiological responses due to inhaler single actuation with incorrect use. The same matrix can be used to analyze both with and without the spacer.</i>
</center>

|	Segment	|	Notes	|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Albuterol Mass in Alveoli (ug)	|	Trachea Flow - Peak Flow  (L/min)	|	Total Lung Volume (mL)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Forced Exhale	|	Patient exhales: ERV Fraction = 1.0 (100%), Period = 3 sec	|	30	|	33	|<span class="success">	0.0 ug	</span>|<span class="success">	Negative Value	</span>|<span class="success">	Residual Volume (~1.2L)	</span>|
|	Inhaler Actuation & Forced Inhale	|	Patient activates inhaler at beginning of forced inhale: Substance = Albuterol, dose = 90 ug, Nozzle Loss = 0.04. Patient inhales: IC Fraction = 1.0 (100%), Period = 5 sec	|	33	|	38	|<span class="success">	Peak 0% to 12% of total dose @cite newman1991improvement (0.0 to 10.8 ug)	</span>|<span class="success">	Positive Value	</span>|<span class="success">	Total Lung Capacity (~6.2L)	</span>|
|	Hold Breath	|	Patient holds breath: Period = 10 sec	|	38	|	48	|<span class="success">	< Peak	</span>|<span class="success">	0 L/min	</span>|<span class="success">	Total Lung Capacity (~6.2L)	</span>|
|	Normal Breathing	|	Patient returns to normal breathing	|	48	|	180	|<span class="success">	< Peak	</span>|<span class="success">	Normal	</span>|<span class="success">	Normal	</span>|


### Two Actuations

Below are validation results for two non-concurrent pMDI actuations using a metered dose of 90 &mu;g of albuterol per actuation (180 &mu;g total) with good coordination. Results agree very well with data. 

<center>
<table border="0">
<tr>
    <td><img src="./plots/Inhaler/Inhaler_TwoActuations_TotalLungVolume.jpg" width="550"></td>
    <td><img src="./plots/Inhaler/Inhaler_TwoActuations_AlbuterolConcentration.jpg" width="550"></td>
</tr>
</table>
</center>
<center><i>Figure 7. Select outputs from the double actuation scenario with correct use.</i></center>

<center><br>
<i>Table 6. Validation matrix for physiological responses due to inhaler double actuation with correct use.</i>
</center>

|	Segment	|	Notes	|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Albuterol Mass in Alveoli (ug)	|	Trachea Flow - Peak Flow  (L/min)	|	Total Lung Volume (mL)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Forced Exhale	|	Patient exhales: ERV Fraction = 1.0 (100%), Period = 3 sec	|	30	|	33	|<span class="success">	0.0 ug	</span>|<span class="success">	Negative Value	</span>|<span class="success">	Residual Volume (~1.2L)	</span>|
|	Inhaler Actuation & Forced Inhale	|	Patient activates inhaler at beginning of forced inhale: Substance = Albuterol, dose = 90 ug, Nozzle Loss = 0.04. Patient inhales: IC Fraction = 1.0 (100%), Period = 5 sec	|	33	|	38	|<span class="success">	Peak 47% to 65% of total dose @cite usmani2005regional (42.4 to 59.0 ug)	</span>|<span class="success">	Positive Value	</span>|<span class="success">	Total Lung Capacity (~6.2L)	</span>|
|	Hold Breath	|	Patient holds breath: Period = 10 sec	|	38	|	48	|<span class="success">	< Peak	</span>|<span class="success">	0 L/min	</span>|<span class="success">	Total Lung Capacity (~6.2L)	</span>|
|	Normal Breathing	|	Patient returns to normal breathing	|	48	|	85	|<span class="success">	> 0.0 ug	</span>|<span class="success">	Normal	</span>|<span class="success">	Normal	</span>|
|	Forced Exhale	|	Patient exhales: ERV Fraction = 1.0 (100%), Period = 3 sec	|	85	|	88	|<span class="success">	> 0.0 ug	</span>|<span class="success">	Negative Value	</span>|<span class="success">	Residual Volume (~1.2L)	</span>|
|	Inhaler Actuation & Forced Inhale	|	Patient activates inhaler at beginning of forced inhale: Substance = Albuterol, dose = 90 ug, Nozzle Loss = 0.04. Patient inhales: IC Fraction = 1.0 (100%), Period = 5 sec	|	88	|	93	|<span class="success">	Peak 47% to 65% of total dose @cite usmani2005regional (42.4 to 59.0 ug)	</span>|<span class="success">	Positive Value	</span>|<span class="success">	Total Lung Capacity (~6.2L)	</span>|
|	Hold Breath	|	Patient holds breath: Period = 10 sec	|	93	|	103	|<span class="success">	< Peak	</span>|<span class="success">	0 L/min	</span>|<span class="success">	Total Lung Capacity (~6.2L)	</span>|
|	Normal Breathing	|	Patient returns to normal breathing	|	103	|	235	|<span class="success">	> 0.0 ug	</span>|<span class="success">	Normal	</span>|<span class="success">	Normal	</span>|

@anchor inhaler-conclusions
Conclusions
-----------
The %BioGears inhaler implementation was developed to simulate the administration of inhaled drugs and support the administration of a beta-agonist to intervene in the case of acute asthma.  The results from the validation cases show the successful modeling of albuterol administration using a pMDI. The model also supports multiple use of a pMDI as shown in another validation case. 
@anchor inhaler-future
Future Work
===========

Coming Soon
-----------

There are no planned near term additions.

@anchor inhaler-improvements
Recommended Improvements
------------------------

- The system transport of inhaled drugs is currently treated as gaseous. As a result, the model does not address drug deposition and absorption through respiratory airway surfaces or through oral ingestion. Albuterol is actually a solid in a liquid suspension, but %BioGears treats it as a gas with assumed gas diffusion and transport properties. All diffusion into the bloodstream occurs as a result of gas transfer from the alveoli into the pulmonary capillaries. The development of a model that includes lung and airway particle deposition could improve realism and allow drug transport through other tissues (such as the bronchial tubes or intestine lining), which are known to be a significant transport mechanisms. 
- %BioGears currently assumes that aerosolized drugs are monodisperse. The development of a polydisperse model for %BioGears that includes droplet evaporation could improve realism. 
- In %BioGears, aerosolized drug that is is exhaled after initial inhalation is currently lost. The ability to model the re-inhalation of exhaled drug would improve the simulation of uncoordinated use of the inhaler. 
@anchor inhaler-appendices
Appendices
==========

Data Model Implementation
-------------------------

@ref InhalerTable "Inhaler"

Acronyms
--------

pMDI - Pressurized Metered Dose %Inhaler

Compartments
------------

- Mouthpiece%Renal Methodology {#RenalMethodology}
==========================

Overview
========

@anchor renal-abstract
Abstract
--------

The %Renal system's purpose in the body is to filter blood. Each kidney is modeled in the %BioGears Engine using a single, lumped nephron to represent its behavior. By using this circuit analogue of a lumped nephron, %BioGears accurately simulates the major functions of the kidney: filtration, clearance, secretion, and reabsorption. Through these nephron-level processes, we are able to accurately recreate the mechanism by which urine is produced and blood is filtered. The renal system is complex and able to mechanistically and accurately recreate the clearance of the substances currently in %BioGears. The system passes the majority of validation and demonstrates the correct trends in scenarios that test specific functionality.

@anchor renal-introduction
Introduction
------------
### %Renal Physiology

The renal (or urinary) system's primary job is filtering the blood to remove waste and manage
fluid volume, which helps the body maintain homeostasis.  

Figure 1 shows the entire system and is labeled with the following:

1. Urinary system
2. Kidney
3. %Renal pelvis
4. Ureter
5. Urinary bladder
6. Urethra (Left side with frontal section)
7. Adrenal gland
8. %Renal artery and vein
9. Inferior vena cava
10. Abdominal aorta
11. Common iliac artery and vein
12. Liver
13. Large intestine
14. Pelvis

@image html RenalSystem.png
<center>
<i>
Figure 1. This is an illustration of the human renal system.  The bounds of the system are generally accepted to be the combination of everything encompassing #1-6 (urinary system) and #8 renal artery and vein. @cite Jmarchn2010Urinary
</i>
</center><br>

After the blood enters the kidneys through the renal arteries, it is distributed into the nephrons through the branching afferent arteriole network. From the arterioles, the blood then travels into the glomerular capillaries. The capillary wall's permeability coupled with the pressure difference across the permeable layer allows filtration of about 20 percent @cite hall2011guyton of the incoming blood into the Bowman's space and downstream into the tubules, vas rectum and, eventually, the collecting duct where the fluid travels through the ureters and into the bladder in the form of urine. 

The average glomerular filtration rate (GFR) in a human is around 120 mL/min @cite valtin1995renal between both kidneys. The filtered fluid primarily consists of the smaller particles in the blood, as a general rule less than 10 nm in diameter @cite valtin1995renal. This filtered fluid will eventually become urine; however, most of the fluid will be reabsorbed back into the blood to preserve substance and fluid levels in the body. Some substances, such as creatinine, are not reabsorbed, making these good a clinical markers of GFR. 

Pressure and GFR are tightly regulated through the tubuloglomerular feedback mechanism of the kidney. This mechanism allows for an increase or decrease in the diameter of the arteries on either side of the glomerular capillaries, effectively regulating down stream fluid flow and protecting the fragile capillaries of the glomerulus.

Once filtered, reabsorption and secretion both occur along the length of the tubules into the collecting ducts. In this area, the body reabsorbs many of the substances and, on average, over 99% of the water back into the blood @cite hall2011guyton. Reabsorption is heavily dependent on the sodium concentration of the blood, the downstream pressure at the afferent arteriole, and osmotic and oncotic pressure gradients along the tubular cell walls and intra-vascular space. The kidneys influence the osmolarity of the urine by adjusting the amount of sodium and water excreted to the bladder. Secretion allows for further filtration of the blood by the kidneys via active transport of substances back into the tubules from the peritubular capillaries. All of these processes are tightly regulated and allow for careful management of substances levels within the entire body. 
Any water or substance that is not reabsorbed is moved to the bladder as urine through the collecting ducts.

By regulating the substance content of the blood, the kidneys are responsible for maintaining blood pH and 
nitrogen content over long periods of time. Nitrogen waste from the liver is excreted from the kidneys as urea, one of the primary constituents of urine. Bicarbonate is used by the blood as a buffer against much of the acidic content of the body. The kidneys are able to reabsorb virtually all of the filtered bicarbonate and even reclaim bicarbonate that has been used to buffer acids that are filtered into the renal tubules @cite valtin1995renal. 

### %Renal Modeling

The %BioGears %Renal model lumps each kidney into a single nephron (Figure 2). The  fluid flow through the kidney is modeled with an electrical circuit analogue, and the substance transport is a combination of the generic transporter and active filtration/reabsorption/secretion based on feedback and substance parameters. Alternatively, excretion can be directly applied and is governed via clearance equations - as with the drug model (@ref DrugsMethodology). This is consistent with other %BioGears system models, which also employ lumped parameter modeling.

@image html LumpedNephron.png
<center>
<i>
Figure 2. The basic anatomical structure of the nephron. The blood is filtered into the Bowman's capsule from the Glomerular Capillaries. Most of the fluid is then reabsorbed into the Peritubluar Capillaries.  The remaining fluid and substances continue through the tubules and are excreted into the bladder. @cite Madhero882010Nephron
</i>
</center><br>  
@anchor renal-system
System Design
=============

@anchor renal-background
Background and Scope
--------------------

### Requirements

The %Renal system needs to meet the following requirements in %BioGears:
- General kidney functions @cite guyton2006medical
	- Excretion of metabolic waste products and foreign chemicals
	- Regulation of water and electrolyte balances
	- Regulation of body fluid osmolality and electrolyte concentrations
	- Regulation of arterial pressure
	- Regulation of acid-base balance
	- Secretion, metabolism, and excretion of hormones
	- Gluconeogenesis
- Specific effects required by the Engine
	- Nutritional states based on feeding, starvation, water loads, dehydration, and salt intake
	- %Renal stenosis and hypoperfusion
	- General hematology: electrolytes, ions, blood gases, protein, urea, creatinine, hemoglobin, hematocrit, RBC, WBC, and Acid/Base Balance
	- System parameters: Urine volume, osmolality, GFR, and BUN
	- %Respiratory acidosis/alkalosis, metabolic acidosis
	- %Drugs pharmacokinetics handling
	- Extensible and flexible model
	- Maintain conservation of mass and homeostasis

### Existing

There have been numerous approaches to renal modeling in the past. The most significant results stem from a region-based modeling 
approach that discretizes the depth and allows varying permeability along the length of the nephron @cite layton2011mathematical. 
Focusing on the region encompassing the collecting duct allows for consideration of the flow in the tubules as 
well as interactions with the interstitial space. This space plays a key role in the re-absorption process and is modeled with striking detail.
 This model has been shown to be an extremely accurate representation of the urine concentrating mechanism along the length of the 
 short and long nephron and has been shown to accurately model the two feedback mechanisms that control glomerular filtration and 
 blood flow: the myogenic response and the tubuloglomerular feedback. This model uses a region-based approach to model a single 
 collecting duct, then it scales results up to a system-wide representation. The idea of representing the kidney as a functional
 nephron is something that many other researchers in the past have used to accomplish very promising results regarding kidney functionality 
 @cite moss2009computational, @cite moss2014hormonal, @cite stephenson1976model. 


In each of these efforts the nephron is broken into its distinct sections and the reabsorption of substances and fluid is calculated
at each point. Each calculation is done in series until the filtered fluid reaches the collecting ducts and exits as urine. By assuming 
the results carry over to the other nephrons in the kidney, these models are able to create realistic urine outputs under most conditions.
While these models can show the tubular transport of substances and fluid in great detail, some do not directly show the renal
hemodynamics at work to control glomerular pressure and GFR. In addition, many of these results are steady state rather than time-dependent, which limits the range of applications and outputs.

Many models use the Starling equation (Equation 1) to calculate the fluid flow between the tubules and blood. While this is generally considered to be an accurate representation of the fluid flow, this approach assumes a constant oncotic pressure. This makes the hydrostatic pressures the only dynamic inputs to the Starling equation. The accuracy of this approach can be enhanced by accounting for differences in the substance content in vascular compartments in the glomerular capillaries and deriving the respective oncotic pressures, which allows for a better representation of the pressure differential that drives the flow. 

\f[{J_v} = {K_f}\left( {\left[ {{P_c} - {P_i}} \right] - \sigma \left[ {{\pi _c} - {\pi _i}} \right]} \right)\f]
<center>
*Equation 1*
</center><br>


J<sub>v</sub> represents net fluid flow, P<sub>c</sub> and P<sub>i</sub> represent the capillary and interstitial hydrostatic pressures,
&pi;<sub>c</sub> and &pi;<sub>i</sub> are the capillary and interstitial oncotic pressures, respectively, K<sub>f</sub> is the filtration coefficient, and &sigma; is the reflection coefficient. The equation uses a V=IR approach to derive the flow from the total pressure differential and the resistance, which is represented by the filtration and reflection coefficients.

There are other existing models that show the hemodynamics present within the kidney. The model shown by Moss @cite moss2009computational uses a network of nephrons and a fluid
circuit to represent the kidney. The fluid circuit can affect the delivery of blood to the various glomerular capillaries by changing the resistance that represents the afferent 
arteriole. This allows the model to represent the physiological feedback from changes in sodium reabsorption, which can have notable effects on renal blood flow. The fidelity
of this sort of model is very flexible, as it is entirely dependent on the detail put into the nephron. Unfortunately, as the complexity of the nephron model or the
desired size of the network increases, it can become computationally intensive. This can cause problems when attempting to use this approach as part of a real-time model.

### Approach

The %Renal model includes both the upper and lower urinary tracts.  The urine formation in the kidneys is simulated using a single lumped nephron model, which is the internal functional element shown as #2 in Figure 1. The computational representation of the system includes the renal artery that feeds into the afferent arteriole. Filtration takes place through the glomerular capillaries, and re-absorption occurs through the tubules into the peritubular capillaries. The peritubular capillaries connect into the vena cava, allowing transport back into the cardiovascular system. The ureter and bladder make up the rest of the kidney model in %BioGears. This model of the renal system allows for tubuloglomerular feedback and re-absorption feedback. These mechanisms allow for diuresis under heavy pressure loads and filtrate regulation under variable pressure scenarios. 

The %Renal fluid circuit is inserted in the the circulatory system and replaces a calibrated three-element (two resistances and one compliance) Windkessel model that is used for %Cardiovascular validation.  In this way, the standalone circuit at a resting physiologic state matches from a total fluid mechanics standpoint.

The transport of substances is a combination of the generic methodology and active filtration at specific locations. However, if a substance has a defined renal clearance parameter, as with drugs, that value is directly applied and is only dependent on blood flow changes.

@anchor renal-dataflow
Data Flow
---------

The %Renal System determines its state at every time step through a three-step process: Preprocess, Process, and Postprocess (Figure 3). In general, Preprocess sets the circuit element values based on feedback mechanisms and engine settings/actions. Process uses the generic circuit calculator to compute the entire state of the circuit. Postprocess is used to advance time. More specifics about these steps are detailed below.

### Reset, Conditions, and Initialization

During initialization of the system, the filterability parameter for each defined substance is calculated.  This is only done and set once because it is only dependent on the substance's molecular weight and charge in blood, which do not change.

The two %Renal system conditions are also applied using the engine stabilization methodology.  %Renal stenosis and consume meal are discussed in detail later.

### Preprocess

Preprocess is called by the engine to apply the system-specific feedback mechanisms.  Each of these mechanisms for the %Renal system are described individually and in detail later.

### Process

The generic circuit solving and transporting are done on the combined circulatory system inside the %Cardiovascular system.  After the new state of the circuit is determined, active transport is performed, and the system data is calculated and set.

### Post Process

The Postprocess step moves values calculated in the Process step from the next time step calculation to the current time step calculation. The current implementation has no specific postprocess functionality for the %Renal system. All postprocessing is done on the combined circulatory system inside the %Cardiovascular system.

### Assessments

There is one assessment that is done within the %Renal system - urinalysis.  This is a data call that requires further calculations and analysis to create outputs.  It is discussed in detail later.

<img src="./images/Renal/RenalFlowChart.png" width="1100">
<center>
<i>
Figure 3. The data flow for the %Renal System consists of Preprocess, Process, and Postprocess. Preprocess determines the circuit element values based on feedback mechanisms and engine actions. Process uses generic circuit methodology to solve the circuit for pressures, volumes, and flows along the paths and on nodes. Postprocess updates these quantities to the next time step and then begins advancing time, which causes the loop to repeat.
</i>
</center><br>

@anchor renal-features
Features, Capabilities, and Dependencies
----------------------------------------

### Circuit

The %BioGears %Renal circuit (Figure 4) determines blood and urine pressure, flow, and volume, organized by compartments.  These compartments are comprised of lumped parameter models that use resistors, capacitors, pressure sources, and valves.  The number of lumped parameter models used to represent the %Renal System was chosen to provide a level of fidelity that meets the requirements of the overall project and to provide sufficient system capability.
@anchor renal-circuit
@image html RenalCircuit.png
<center>
<i>
Figure 4. The %Renal circuit is made up of nodes and paths with elements.  The fluid mechanics of the system are calculated assuming no transition between blood and urine.  The substance quantities are what distinguish the two fluid types.  Everything above the dashed line can be considered blood.  Fluid below the dashed line transitions to urine as substance filtration and reabsorption change its make-up. The "T" labeled paths are location where transport of each substance is calculated by the %Renal system directly.  All others are done generically.
</i>
</center><br>

Nodes serve as the connection points for paths and are the locations at which
pressures are measured. Each %Renal node contains a pressure value, which is given
with respect to the atmospheric reference node (indicated in the diagram by the
equipotential symbol). Nodes can also hold the values for volume or the mass and
concentration of substances. Paths contain information about the flow (volume
per time) and volumes of compliances. The %BioGears @ref CircuitMethodology document
contains more information about %BioGears circuit definitions and modeling. The %BioGears @ref SubstanceTransportMethodology
contains more information about the substance transport and mass and concentration calculations.

Dynamic pressure sources are used at the filtration and reabsorption locations to model the colloid osmotic pressures caused by protein in the fluid.  These pressures, in combination with the hydrostatic pressures, give a correct total pressure difference to cause a correct net fluid flow.

The bladder is currently modeled as a constant pressure source and the volume is calculated directly based on the flow through that source.  Future versions will likely replace the source with a compliance and incorporate a peristaltic flow model.

The compliance and volume of the %Renal circuit are determined by validated volume data and model assumptions made regarding the various nodes in the renal system. These compliances allow for some volume variability as the capillaries experience pressure and flow changes.

The volume distribution of the %Renal system is as follows:
- Kidneys = 69.6 mL
	- Right Kidney = 34.8 mL
		- Right Small Vasculature = 17.4 mL
			- Right Glomerular Capillary (compliance) = 3.48 mL
			- Right Peritubular Capillary = 3.48 mL
			- Right Efferent Arteriole = 3.48 mL
			- Right Bowmans Capsule = 3.48
			- Right Afferent Arteriole = 3.48 mL
		- Right Large Vasculature = 17.4 mL
			- Right %Renal Vein (compliance) = 5.8 mL
			- Right %Renal Artery (compliance) = 5.8 mL
			- Right Tubules = 5.8
	- Left Kidney = 34.8 mL
		- Left Small Vasculature = 17.4 mL
			- Left Glomerular Capillary (compliance) = 3.48 mL
			- Left Peritubular Capillary = 3.48 mL
			- Left Efferent Arteriole = 3.48 mL
			- Left Bowmans Capsule = 3.48
			- Left Afferent Arteriole = 3.48 mL
		- Left Large Vasculature = 17.4 mL
			- Left %Renal Vein (compliance) = 5.8 mL
			- Left %Renal Artery (compliance) = 5.8 mL
			- Left Tubules = 5.8

The %Renal circuit contains 11 resistances for each kidney, and a shared urethra resistance - see Figure 5.  The urethra (number 12) is typically set as an open switch (approximately infinite resistance), unless the patient is urinating, when it becomes closed to allow flow out of the bladder.

@image html RenalResistances.png
<center>
<i>
Figure 5. The resistances in the left kidney are numbered for further description.
</i>
</center><br>

Baseline resistances were determined by leveraging reference pressure and flow values from literature.  Each resistance is calculated using Equation 2, where R is the resistance along the path, &Delta;P is the pressure difference between the source and target nodes, and F is the fluid flow through the path.

\f[R = \frac{{\Delta P}}{F}\f]
<center>
*Equation 2*
</center><br>

The untuned baseline resistance values were calculated using the values in Table 1.  These values and the model were verified with unit testing using a static arteriole pressure equal to the normal mean value.

<br><center>
<i>Table 1. The baseline resistances used in the %Renal circuit @cite guyton2006medical. The numbering matches that of the previous figure.</i>
</center>

| Location | Number in Figure | Beginning Pressure (mmHg) | Ending Pressure (mmHg)| Flow (mL/min) | Resulting Baseline Resistance (mmHg/mL/min) |
| :---- | :---- | :---- | :---- | :---- | :---- |
| Arteries | 1 | 100 | 100 | 600 | 0.0 |
| RenalArteries | 2 | 100 | 85 | 600 | 0.0250 |
| AfferentArteriole | 3 | 85 | 60 | 600 | 0.0417 |
| GlomerularCapillaries | 4 | 60 | 59 | 537.5 | 0.0019 |
| EfferentArteriole | 5 | 59 | 18 | 537.5 | 0.0763 |
| PeritubularCapillaries | 6 | 18 | 8 | 600 | 0.0167 |
| RenalVeins | 7 | 8 | 4 | 600 | 0.0066 |
| GlomerularFilter | 8 | 28 | 18 | 62.5 | 0.1600 |
| Tubules | 9 | 18 | 6 | 62.5 | 0.1920 |
| Reabsorption | 10 | -9 | -19 | 62 | 0.1613 |
| Ureter | 11 | 6 | 4 | 0.52 | 3.846 |
| Urethra (during urination) | 12 | 4 | 0 | 1320 | 0.0030 |
@anchor renal-dependencies
### Patient Variability

The total vascular volume of the %renal system in %BioGears is defined to be a fraction of the body total, making kidney volume directly dependent upon the patient data: height, weight, and gender. The volume fraction of both kidneys is set to be 0.0202, or about 1/50th of total blood volume @cite valtin1995renal. This volume is then divided in half to populate the total vascular volume of one kidney. For a further breakdown of the volume of each kidney compartment see @ref renal-features "Features".


### Feedback

#### Colloid Osmotic Pressure

The net filtration and reabsorption pressures are determined by the sum of the hydrostatic and colloid osmotic forces across the membranes.  While the hydrostatic pressure is automatically handled using the generic circuit algorithms, the colloid osmotic pressure needs to be determined specifically.  This is done by using the Landis-Pappenheimer equation (Equation 4) that is dependent on total protein @cite khazaei2008new.  Since the %BioGears engine tracks albumin, a constant relationship between total protein and albumin can be leveraged using Equation 3.  *C<sub>TP</sub>* is the local concentration of total protein, *C<sub>A</sub>* is the local concentration of albmuin, and *P<sub>CO</sub>* is colloid osmotic pressure.

\f[{C_{TP}} = 1.6{C_A}\f]
<center>
*Equation 3*
</center><br>

\f[{P_{CO}} = 2.1{C_{TP}} + 0.16C_{_{TP}}^2 + 0.009C_{_{TP}}^3\f]
<center>
*Equation 4*
</center><br>

The Landis-Pappenheimer equation is applied at the pressure sources numbered in Figure 6.

@image html RenalPressureSources.png
<center>
<i>
Figure 6. The four pressure sources that represent the colloid osmotic pressure are determined via the local albumin concentration.
</i>
</center><br>

Table 2 gives the typical colloid osmotic pressure values.  The tubular osmotic pressure is set constant in our model and does not change because we are not modeling the interstitial space.  The remaining osmotic pressures will increase and decrease in relation to the local albumin concentration variations.  See Table 5 for %BioGears resulting values when albumin concentration feedback is applied.

<br><center>
<i>Table 2. The baseline colloid osmotic pressure values used for both initialization and model validation @cite guyton2006medical. The numbering matches that of the previous figure. </i>
</center>

| Location | Number in Figure | Baseline Colloid Osmotic Pressure (mmHg) |
| :---- | :---- | :---- |
| GlomerularCapillariesOsmoticPressure | 1 | -32 |
| BowmansCapsulesOsmoticPressure | 2 | 0 |
| TubulesOsmoticPressure | 3 | -15 |
| PeritubularCapillariesOsmoticPressure | 4 | -32 |

#### Fluid Permeability

The hydraulic and colloid osmotic pressure gradients across the glomerular and pertitubular capillary membranes cause the movement of fluid. The rate of fluid movement is a function of the fluid permeability of the membranes, the total membrane surface area, and the pressure gradient. Considering a nominal pressure that is equal to the linear integral of the effective pressure over the length of the glomerular capillary (with the effective pressure being the difference between the hydrostatic and colloid osmotic pressure gradients at each point along the length of the capillary), the glomerular filtration rate is proportional to the nominal pressure, as shown in Equation 5.

\f[GFR = {L_p}{S_f}\int {\left[ {\left( {{P_g}\left( x \right) - {P_B}} \right) - \sigma \left( {{\pi _g}\left( x \right) - {\pi _B}} \right)} \right]dx}  = {L_p}{S_f}\Delta P\f]
<center>
*Equation 5*
</center><br>

*L<sub>p</sub>* is the hydraulic conductivity of the membrane and S<sub>f</sub> is the surface area. Relating the equation above to the linear fluid dynamics equations, it is apparent that the resistance to flow can be modeled as 1/(*L<sub>p</sub>* *S<sub>f</sub>*). The product, *L<sub>p</sub>* *S<sub>f</sub>*, has been called the glomerular filtration coefficient @cite tuma2011microcirculation. A similar reabsorption coefficient can be derived from the equations governing the reabsorption of water in the renal tubules @cite tuma2011microcirculation.

The hydraulic permeability is computed for use in the %BioGears model using filtration coefficients reported in literature as well as surface area estimates. The surface area of the glomerular capillaries is estimated to be about 2.0 square meters per human kidney @cite valentin2002icrp. The surface area of the peritubular capillaries is larger than the glomerular capillaries @cite tuma2011microcirculation. However, how much larger is difficult to quantify. For the %BioGears model, it is assumed that the surface area of the peritubular capillaries is 25 percent larger than the surface area of the glomerular capillaries. Using the glomerular filtration coefficient value of 13 mL/min-mmHg for both kidneys as reported in @cite tuma2011microcirculation (6.5 mL/min-mmHg each kidney assuming an equal distribution of surface area), the hydraulic conductivity of the glomerular capillaries is calculated to be 3.67647 mL/m2-min-mmHg. Likewise, using the reabsorption coefficient value of 10  mL/min-mmHg for both kidneys @cite tuma2011microcirculation, the hydraulic conductivity of the peritubular capillaries is computed to be 2.91747 mL/m2-min-mmHg.

These values are applied in the model as the resistances circled in Figure 7 and are determined using Equation 6.  *R<sub>Filt</sub>* is the membrane fluid resistance, *k<sub>p</sub>* is the permeability, and *A* is the total membrane area.

\f[{R_{filt}} = \frac{1}{{{k_p}A}}\f]
<center>
*Equation 6*
</center><br>

@image html Filtration.png
<center>
<i>
Figure 7. The membrane fluid filtration for both glomerular filtration and peritubular reabsorption locations are modeled as the resistances circled in red.
</i>
</center><br>

#### Ultrafiltration

Individual substance transport from the glomerular capillaries to the bowmans capsules is directly proportional to the fluid glomerular filtration rate and filterability value based on molecular size and charge.  The glomerular capillary membrane is thicker than most other capillaries, but it is also much more porous and filters fluid at a high rate @cite guyton2006medical.  Ultrafiltration is a largely passive mechanism.  

Empirical data was used to determine a generic relationship between molecular weight and molecular radius.  Values for water, glucose, inulin, myoglobin, hemoglobin, and albumin were used to determine the best fit shown in Figure 8 @cite rhoades2003medical.

<center><img src="./images/Renal/MolecularRadius.png" width="600"></center>
<center>
<i>
Figure 8. This is the best fit of molecules present in the blood molecular weight to molecular radius.  The resulting equation shows very good fit.
</i>
</center><br>

Electrical charge of each molecule has an effect on the total filterability.  A negative charge restricts filtration, where as a positive charge filters more readily.  Figure 9 shows the relative filterability best fit for positively charged, neutral, and negatively charged molecules as a function of molecular radius (determined using the relationship in Figure 8).

<center><img src="./images/Renal/RelativeFilterability.png" width="600"></center>
<center>
<i>
Figure 9. Effect of size and electrical charge of dextran on its filterability by the glomerular capillaries @cite bohrer1978permselectivity.  A value of 1.0 indicates that the substance is filtered as freely as water.  This relationship can be extrapollated to all substances.  The resulting equations show very good fit.
</i>
</center><br>

#### Reabsorption

Reabsorption works in much the same way as glomerular filtration, with additional active mechanisms that are different for each substance. For the %BioGears %Renal model, all of the active transport mechanisms are lumped into a single reabsorption ratio parameter defined for each substance. This ratio represents the extent to which the substance is reabsorbed from the tubules to the peritubular capillaries in relation to the fluid flow. Therefore, a value of 1.0 indicates that the substance is reabsorbed at the same rate as water, values below 1.0 indicate that it is reabsorbed more avidly than water, and values above 1.0 indicate that it is reabsorbed less avidly than water. Each substance also has a transport maximum value defined that prevents substance mass movement at a specific transient value. A value of infinity is allowed for both reabsorption-specific substance parameters.

##### Reabsorption Feedback

Pressure natriuresis and diuresis are simulated in the %BioGears %Renal model through reabsorption premeability modifiers. In the kidneys, as arterial pressure increases, shear stresses are developed along the cell walls. These stresses induce the release of nitric oxide, which is diffused downstream to the tubules of the nephron. This serves to decrease the sodium reabsorption rate through select entry pathways @cite navar1999kidney. This decrease in sodium resabsorptioin decreases the osmotic gradient along this fluid path, consequently leading to a coupled decrease in fluid reabsorption. %BioGears does not currently support nitric oxide, so we model this mechanism by coupling the fluid permeability characteristics of the tubules as function of arterial pressure. This allows for the connection to downstream pressure changes and sodium/water reabsorption. A second order polynomial (See Equation 7) is fit to the data taken from @cite guyton2006medical that scales the tubules' fluid permeability as a function of arterial pressure. This permeability then directly affects the resistance along the reabsorption pathway, leading to a decrease or increase in water/sodium transport. Diuretic administration can also inhibit the reabsorption by modifying the tubular lumen permeabilty as a function of plasma concentration of Furosemide, @ref DrugsMethodology.

<center><img src="./plots/Renal/MAP_vs_UPR.jpg" width="600"></center>
<center> 
<i>Figure 10 shows the urine production rate as a function of the mean arterial pressure. Blue is experimental data and red is %BioGears simulation data. Diuresis takes place in response to increased blood pressure, in agreement with renal functionality.</i>
</center><br>


\f[{k_p} = \left( {2.01 \times {{10}^{ - 6}}} \right)P_A^2 - \left( {8.1 \times {{10}^{ - 4}}} \right)P_A^{} + 9.4 \times {10^{ - 2}}\f]
<center>
*Equation 7*
</center><br>

##### Diuretic Response
In addition to alterations in permeability due to ion concentration and perfusion pressure, the renal system's reabsorption pathway can also be effected by diuretic concentrations in the blood plasma. These concentrations directly effect the tubular permeability, simulating Furosmide binding to the Na-K-2Cl symporter. This effect is accounted for by calculating the response due to plasma concentrations in the Drug methodology. This permeability modifier is then stored and accessed in the %Renal system each time step to calculate the reabsorption pathway resistance: 

\f[{k_p}_{i+1} =  {k_p}_i * K_{mod}(C_p) \f]
<center>
*Equation 8*
</center><br>
Here *K<sub>mod</sub>(C<sub>p</sub>)* is the modification coefficient calculated as a hill type functional response to Furosemide blood plasma concentration, *C<sub>p</sub>*.


### Secretion 
Flow of substances can be directed from the peritubular capillaries into the tubules through secretion. As fluid travels along the tubules and into the collecting ducts of the nephron, various stages of the journey either reabsorb or secret substances and water back into and out of the fluid. One of the most important tasks of the renal system is handling of potassium levels in the blood stream. Secretion is one of the main ways that the renal system handles this task with active transport mechanisms located along the tubular cell wall. %BioGears simulates this active handling of potassium by allowing for secretion into the ureter from peritubular capillaries. Mass is calculated and tightly regulated by the %BioGears renal system by actively moving any mass above a certain threshold directly into the urine. %BioGears does not currently support disturbances to the permeable layer and so shifts in potassium levels in the body will not currently be possible. 

#### Bladder and Excretion

Excretion of fluid and substances to the bladder is done generically by the circuit solver (@ref CircuitMethodology) and transporter (@ref SubstanceTransportMethodology).

#### Tubuloglomerular Feedback

One of the main mechanisms for kidney regulation of renal blood flow due to changes in the mean arterial pressure is tubuloglomerular feedback. This is caused by a transient increase in GFR that leads to an increased sodium delivery to the macula densa @cite rhoades2003medical. For the %BioGears %Renal model, the macula densa is lumped with all tubules. When the mass flow rate of sodium changes in the tubules, a response is applied.

Within an autoregulatory range of 80 to 180 mmHg for the MAP, the renal blood flow is is maintained by either constricting (increased resistance) or dilating (decreased resistance) the afferent areteriole @cite rhoades2003medical. A higher sodium mass flow rate in the tubules leads to a higher resistance, and a lower mass flow rate leads to a lower resistance (see Figure 11). The minimum and maximum resistances for the afferent arteriole were determined through a unit test to be 1.7 mmHg/mL-s (at MAP of 80 mmHg) and 12.382 mmHg/mL-s (at MAP of 180 mmHg).

@image html TubuloglomerularFeedbackFlow.png
<center>
<i>
Figure 11. A flow diagram showing the tubuloglomerular response to increased and decreased MAP.  @cite rhoades2003medical
</i>
</center><br>

High frequency oscillations of the tubuloglomerular feedback are damped using the tuning constant, alpha, in Equation 8.  *R<sub>Aff</sub>* is the afferent arteriole resistance where i and i+1 signify the current and the next resistance respectively, ||&Delta; U<sub>Na</sub>|| is the normalized change in sodium flow into the tubules. Normalization is used to decrease sensitivity of the response. Large fluctuations are not only physiologically incorrect but also create instabilities during heartbeats. The sodium mass flow setpoint is the expected value with a stable standard patient. Figure 13 indicates fair agreement between the mechanics of %BioGears and experimental data. Autoregulation is handled entirely from the TGF response, with the myogenic response not modeled currently.

\f[R_{Aff}^{i + 1} * = \frac{{R_{Aff}^i}}{{R_{Aff}^{i + 1}}} + \alpha \left\| {\Delta {U_{Na}}} \right\|\f]
<center>
*Equation 9*
</center><br>


@image html TubuloglomerularFeedback.png
<center>
<i>
Figure 12. The circle shows the location used to determine the sodium value, and the resistance where the feedback is applied is denoted with an oval.
</i>
</center><br>

<center>
<table border="0">
<tr>
    <td><img src="./plots/Renal/MAP_vs_GFR.jpg" width="550"></td>
    <td><img src="./plots/Renal/MAP_vs_RBF.jpg" width="550"></td>
</tr>
</table>
</center>
<center>
<i>
Figure 13. The plots show the flow response to variations in mean arterial pressure: blue indicates experimental data and red indicates %BioGears simulation data. Autoregulatory plateau is seen in GFR and RBF. 
</i>
</center><br>


#### Osmoreceptor Feedback

When osmolarity (plasma sodium concentration) increases/decreases above/below normal the osmoreceptor feedback system compensates in the way shown in Figure 14. Increased water permeability in the distal nephron segments causes increased water reabsorption and excretion of a small volume of concentrated urine @cite guyton2006medical. This mechanism tends to keep the plasma sodium concentration stable and was calibrated with extended multi-hour simulations.

@image html OsmoreceptorFeedbackFlow.png
<center>
<i>
Figure 14. A flow diagram showing the osmoreceptor response to increased and decreased plasma sodium concentration.  @cite guyton2006medical
</i>
</center><br>

The sensitivity of the osmoreceptor feedback is calibrated using the tuning constant *x* in Equation 9.  *k<sub>P</sub>* is the reabsorption fluid permeability, *C<sub>SP</sub>* is the sodium plasma concentration in the tubules, and *C<sub>SP,Set</sub>* is the setpoint.  The sodium plasma concentration setpoint is the expected value with a stable standard patient.

\f[{k_p} *  = {\left( {\frac{{{C_{SP}}}}{{{C_{SP,Set}}}}} \right)^x}\f]
<center>
*Equation 10*
</center><br>

#### Gluconeogenesis

Gluconeogenesis is a metabolic pathway that results in the generation of glucose from non-carbohydrate carbon substrates, such as lactate.  Within the %BioGears %Renal system, this process is applied by converting excreted lactate at a one-to-one mass ratio into reabsorbed glucose. This effectively removes all lactate from the urine until the transport maximum is achieved. If the converted mass exceeds the transport maximum, it is capped and the remainder continues to the urine.

#### %Renal Clearance

Some substance in the engine specify a single renal clearance parameter (e.g. drugs) instead of inputs required for the previously described active transport mechanisms.  If so, that value is directly applied and is only dependent on blood flow and plasma concentration changes.  Details on this behavior are in @ref DrugsMethodology.

### Dependencies

The %Renal system is directly connected to the %Cardiovascular system at the aorta and vena cava.  The boundary condition feedback in and out is automatically handled through the node and path interactions.  It also shares a reference node.



@anchor renal-assumptionlimitations
Assumptions and Limitations
---------------------------

There are several assumptions that were used in creating this %Renal model:
- All nephrons in each kidney are lumped together
- Secretion effects are lumped with reabsorption
- No specific substance metabolism is modeled
- Both kidneys are identical
- The difference between blood and urine is substance dependent only; they are the same fluid from a physics standpoint

@anchor renal-conditions
Conditions
----------

### Consume Meal

The consume meal condition takes the resting renal clearance of each substance and extrapolates how much would be removed during a specified fasting period.  The substances are then decremented accordingly.  The loss of fluid is handled by the %Energy system (@ref EnergyMethodology).  In this way, both dehydration and starvation can be modeled.  See the @ref GastrointestinalMethodology document for more information.

@anchor renal-actions
Actions
-------

### Urinate

Urination empties the bladder. This can be called as an action or will automatically occur with the functional incontinence event. This action decreases the urethra resistance to allow the bladder to empty to 1 mL.  The fluid mechanics and transport are done generically by the circuit solver and transporter respectively.

<center><img src="./plots/Renal/12hr_Urination.jpg" width="800"></center>
<center> 
<i>Figure 15. Bladder volume over time is shown during a twelve hour simulation. Once the bladder reaches the maximum allowable volume, 400mL, the patient will experience functional incontinence.</i>
</center><br>

@anchor renal-events
Events
------

### Diuresis

Diuresis is the state of producing urine at a significantly faster rate than normal. In %BioGears, we consider diuresis to occur when the average production rate of urine exceeds 8 mL/min @cite upsdell1988diuretic, and the event is reversed when the average production rate of urine falls below 7.5 mL/min.

### Antidiuresis

Antidiuresis occurs when small quantities of urine are being produced at an osmolarity greater than that of the plasma. For %BioGears, this event is thrown when the average urine production rate is less than 0.5 mL/min and the urine osmolarity is greater than 280 mOsm/L @cite valtin1995renal. The event is reversed when either the urine production rate rises above 0.55 mL/min or the urine osmolarity falls below 275 mOsm/L.

### Natriuresis

Natriuresis is the state of an unusually high rate of sodium excretion into the urine. In %BioGears, the setpoint for natriuresis is an average sodium excretion rate of 14.4 mg/min, or 6 times the resting sodium excretion rate @cite moss2014hormonal. The event is reversed when the average sodium excretion rate falls below 14.0 mg/min.

### Functional Incontinence

Functional incontinence occurs when the bladder becomes too full of urine and automatcially urinates. When the urine volume exceeds the maximum volume of the bladder in %BioGears, 400 mL, the functional incontinence event is thrown and a urinate action is automatically executed.

@anchor renal-assessments
Assessments
-----------

### Urinalysis

Validation of the urine panel is done by analyzing at the resting physiologic quantities. Urine color is determined based on the osmolality of the urine according to Table 3 below. Urine blood content registers as positive if the concentration of hemoglobin in the urine is greater than 0.15 ug/mL @cite walker1990clinical. Currently, the presence of blood does not affect the color of the urine, with urine color being determined entirely by osmolality of the urine, table 3.

<br><center>
<i>Table 3. Healthy urine color is determined by the osmolality of the urine. The colors used in %BioGears were derived from @cite kovacs1999urine . </i>
</center>
|Urine Osmolality                     |Urine Color              | 
|------------------------             |------------------------ |
|UrineOsmolality <= 400 mOsm/kg		  |Pale Yellow	            |
|400 < UrineOsmolality <=750 mOsm/kg  |Yellow		            |
|UrineOsmolality > 750 mOsm/kg        |Light Brown              |

<br><center>
<i>Table 4. Urinalysis is conducted as described by Roxe @cite walker1990clinical . In the current scope of %BioGears, there are no scenarios where blood, glucose, ketones, or 
proteins are expected in the urine in significant quantities, which is confirmed below. Specific gravity and color are each dependent on the osmolarity of the urine and are within
expected values for a healthy individual at rest.</i>
</center>
@insert doc/validation/UrinalysisValidationTable.md

Results and Conclusions
=======================

@anchor renal-restingvalidation
Validation - Resting Physiologic State
--------------------------------------

Validation results for system and compartment quantities for a resting standard patient are listed in Tables 5 and 6. System-level quantities show favorable agreement with validation values. The pressure value discrepancies in urine compartments can be largely explained by the %Renal model's lack of a specific extracellular space that affects both the hydrostatic and osmotic pressures.

<br><center>
*Table 5. Validation of the resting physiologic state system-level outputs from %BioGears compared 
to referenced values.*
</center>

@insert doc/validation/RenalValidationTable.md

<br><center>
*Table 6. Validation of the resting physiologic state comparison of compartment-level outputs from %BioGears
to referenced values.*
</center>

@insert doc/validation/RenalCompartmentsValidationTable.md

Validation results for substance parameters that are determined and/or applied by the %Renal system are shown in Table 7.  These values are highly dependent on substance parameters and show favorable overall agreement for the resting standard patient. Sodium is especially critical for the osmoreceptor and tubuloglomerular feedback mechanisms.

<br><center>
*Table 7. Validation of the resting physiologic state substance parameter outputs from %BioGears compared 
to referenced values.*
</center>

@insert doc/validation/RenalSubstancesValidationTable.md

@anchor renal-actionconditionvalidation
Validation - Actions and Conditions
--------------------

<center>
<i>
Table 8. There are seven scenarios that test renal capabilities, saline ingestion, water ingestion, unilateral
and bilateral stenosis, Two severities of hemorrhage, and high altitude environment change. Each scenario is designed to test the mechanical functionality of the renal system, its feedback mechanisms that control filtration and re-absorption, or the system's substance handling. We have a total of 35 green, 6 yellow, and 0 red parameters among the 3 
scenarios indicating a very good agreement with the validation data.
</i>
</center>
|	Scenario 	|	Description	|	Good	|	Decent	|	Bad	|
|	---	|	---	|	---	|	---	|	---	|
|	SalineIngestion	|	Patient ingests 1L of hypertonic saline	|<span class="success">	2	</span>|<span class="warning">	0	</span>|<span class="danger">	0	</span>|
|	RenalStenosisModerateUnilateral	|	60% occlusion of left kidney	|<span class="success">	6	</span>|<span class="warning">	2	</span>|<span class="danger">	0	</span>|
|	RenalStenosisSevereBilateral	|	90 % occlusion of both kidneys	|<span class="success">	6	</span>|<span class="warning">	2	</span>|<span class="danger">	0	</span>|
|	HemorrhageClass2NoFluid	|	250 mL/min hemorrhage for 240 seconds	|<span class="success">	8	</span>|<span class="warning">	0	</span>|<span class="danger">	0	</span>|
|	HemorrhageClass4NoFluid	|	250 mL/min hemorrhage for 490seconds	|<span class="success">	8	</span>|<span class="warning">	0	</span>|<span class="danger">	0	</span>|
|	HighAltitudeEnvironmentChange	|	High altitude environment change for 900s 	|<span class="success">	1	</span>|<span class="warning">	2	</span>|<span class="danger">	0	</span>|
|	WaterIngestion	|	Patient ingests 1 L of fluid	|<span class="success">	3	</span>|<span class="warning">	1	</span>|<span class="danger">	0	</span>|
|	Starvation	|	Patient has not eaten for 72 hrs	|<span class="success">	1	</span>|<span class="warning">	3	</span>|<span class="danger">	0	</span>|
|		|	Total	|<span class="success">	35	</span>|<span class="warning">	10	</span>|<span class="danger">	0	</span>|


### Hemorrhage - Action

The renal system regulates glomerular filtration and renal blood flow through the tubuloglomerular feedback mechanism. This mechanism is modeled through constriction and dilation of the afferent arteriole as a function of sodium flow rate into the tubules. This mechanism is tested through two different hemorrhage scenarios: HemorrhageClass2NoFluid and HemorrhageClass4NoFluid. In each scenario, the body loses blood at a rate of approximately 250 mL/min with the differences in class being the length of the bleeding. As the patient bleeds, blood volume and mean arterial pressure decrease, causing a reduction in blood flow to the renal system. As this decrease happens, the TGF system tries to compensate by dilating the afferent arteriole, reducing the resistance to flow into the glomerular capillaries. This balances the filtrate up to a point, but as blood volume and pressure continue to decrease, renal blood flow and glomerular filtration rate see reductions in line with past research @cite corcoran1943effects. This is more pronounced as the hemorrhage is allowed to persist. Qualitatively, the renal system in %BioGears displays the mechanical trends seen in past research. 

Table 9 shows the effect seen in the %Renal system. For a complete write-up of the hemorrhage scenarios, see the @ref CardiovascularMethodology.

<center>
<i>
Table 9. In this scenario, the patient is given a class II hemorrhage, which causes changes in MAP and blood volume as the bleeding progresses.
Quantification of the renal system during and after the bleed ends are displayed here. The renal system shows responses in line with actual patients.
</i>
</center>
|	Action	|	Notes	|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Glomerular Filtration Rate (mL/min)	|	Mean Arterial Pressure (mmHg)	|	%Renal Blood Flow(mL/min)	|	Urine Production Rate (mL/min)	|
|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|
|	Initiate 100 mL/min Hemorrhage	|		|	30	|	930	|<span class="success">	Decrease @cite corcoran1943effects	</span>|<span class="success">	Decrease @cite corcoran1943effects	</span>|<span class="success">	decrease @cite corcoran1943effects	</span>|<span class="success">	Decrease @cite corcoran1943effects	</span>|
|	Stop Hemorrhage	|		|	930	|	1230	|<span class="success">	Slight Increase @cite corcoran1943effects	</span>|<span class="success">	Steady @cite corcoran1943effects	</span>|<span class="success">	Slight Increase @cite corcoran1943effects	</span>|<span class="success">	Steady @cite corcoran1943effects	</span>|


The class 4 hemorrhage as the same bleed rate but the bleeding persists for a longer duration: 490 seconds. %Renal function displays the same trends but with a further reduction in glomerular filtration and renal blood flow due to the duration of the bleed action. 

<center>
<i>
Table 10. In this scenario the patient is given a class IV hemorrhage, which causes changes in MAP and blood volume as the bleeding progresses.
Quantification of the renal system during and after the bleed ends are displayed here. The renal system shows responses in line with actual patients.
</i>
</center>
|	Action	|	Notes	|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Glomerular Filtration Rate (mL/min)	|	Mean Arterial Pressure (mmHg)	|	%Renal Blood Flow(mL/min)	|	Urine Production Rate (mL/min)	|
|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|
|	Initiate 150 mL/min Hemorrhage	|		|	30	|	1830	|<span class="success">	Decrease @cite corcoran1943effects	</span>|<span class="success">	Decrease @cite corcoran1943effects	</span>|<span class="success">	decrease @cite corcoran1943effects	</span>|<span class="success">	Decrease @cite corcoran1943effects	</span>|
|	Stop Hemorrhage	|		|	1830	|	2130	|<span class="success">	Slight Increase @cite corcoran1943effects	</span>|<span class="success">	Steady @cite corcoran1943effects	</span>|<span class="success">	Slight Increase @cite corcoran1943effects	</span>|<span class="success">	Steady @cite corcoran1943effects</span>	</span>|


<center>
<table border="0">
<tr>
    <td><img src="./plots/Renal/HemorrhageGFR.jpg" width="550"></td>
    <td><img src="./plots/Renal/HemorrhageMAP.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/Renal/HemorrhageRBF.jpg" width="550"></td>
    <td><img src="./plots/Renal/HemorrhageArterioleResistance.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/Renal/HemorrhageClass4NoFluidLegend.jpg" width="1100"></td>
</tr>
</table>
</center>
<center><i>Figure 16. In this scenario the patient is given a class IV hemorrhage, which causes changes in MAP as the bleeding progresses.
The top left figure shows the mean arterial pressure response to the hemorrhage. The top right and bottom left figures display the glomerular filtration and renal blood flow  
dropping with the pressure drop. This response is mitigated through dilation of the afferent arteriole, seen in the bottom right figure.</i>
</center><br>

### High Altitude - Action 

After a rapid change in altitude, epinephrine is released in response to the O2 deficit in the myocardium. This causes an increase in mean
arterial pressure, which triggers a renal tubuloglomerular feedback response. The afferent arteriole resistance increases in an attempt to maintain
GFR and renal blood flow. The MAP increases quickly and stabilizes at approximately 97 mmHg, which causes initial increases in GFR and %Renal
Blood Flow. These changes are quickly accounted for through the regulation from the tubuloglomerular feedback as expected.

Table 11 shows the effect seen in the %Renal system. For a complete write-up of the High Altitude scenario see the @ref EnergyMethodology.

<center>
<i>
Table 11. In this scenario, the patient environment is changed to simulate the effects of altitude. The acute response increases MAP,
as shown in the top-left plot. The bottom-right plot shows the afferent arteriole resistance increasing to accommodate the changes in input pressure. The other two plots show
the resulting affects on GFR and renal blood flow. Note the increases in renal blood flow and GFR just before 100 seconds, which
are attenuated as the afferent arteriole resistance increases.
</i>
</center>
|	Action	|	Notes	|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Glomerular Filtration Rate (mL/min)	|	%Renal Blood Flow(mL/min)	|	Urine Production Rate (mL/min)	|
|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|
|	Initiate %Environment Change	|		|	50	|	900	|<span class="warning">	Slight Increase @cite guyton2006medical	</span>|<span class="warning">	Slight Increase @cite guyton2006medical	</span>|<span class="success">	Increase @cite guyton2006medical	</span>|


<center>
<table border="0">
<tr>
    <td><img src="./plots/Renal/AltitudeMAP.jpg" width="550"></td>
    <td><img src="./plots/Renal/AltitudeGFR.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/Renal/AltitudeRBF.jpg" width="550"></td>
    <td><img src="./plots/Renal/AltitudeArterioleResistance.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/Renal/HighAltitudeEnvironmentChangeLegend.jpg" width="1100"></td>
</tr>
</table>
</center>
<center><i>
Figure 17. In this scenario, the patient environment is changed to simulate the effects of altitude. The acute response increases MAP,
as shown in the top left figure. The top right and bottom left figures show the mechanical affects of this pressure change in the glomerular filtration rate and the renal blood flow, respectively.  The final figure shows the afferent arteriole resistance increasing by about 30% to accommodate the changes in input pressure. 
</i>
</center><br>

### Salt Ingestion - Action

The salt ingestion scenario simulates the patient drinking 1 L of seawater. Since the osmolarity of seawater exceeds the concentrating capacity of the
human renal system @cite hall2011guyton, this provides a relevant edge case that pushes the kidneys to the limit of their ability to clear the excess
sodium. A 30 minute condition is applied to allow some time to pass for digestion before the effects are analyzed. After the ingestion of seawater, there is a significant change 
in the concentration of sodium in the blood. Since %BioGears does not currently model the fluid shift seen from the intracellular to extracellular space in response 
to changes in osmolarity, this causes a drastic drop in urine production rate as seen in Table 12 @cite hall2011guyton. The resulting urine
becomes hyperosmotic relative to the plasma and contains a high concentration of sodium.

<center>
<i>
Table 12. Urine production rate stays steady. Kidney doesn't quite respond accordingly due to such slight sodium changes. Chronic salt loading causes much more pronounced changed in renal function in patients, something %BioGears isn't meant to handle now.
</i>
</center>
|	Actions	|	Notes	|	Occurrence Time (s)	|	Urine Production Rate (mL/min)	|	Mean Arterial Pressure (mmHg)	|	Urine Chloride Concentration (g/L)	|	Urine Urea Concentration (g/L)	|
|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|
|	After stabilization patient given 1L of water with 100g NaCl 	|		|	0	|<span class="success">	Slight Increase @cite dean1949renal	</span>|<span class="success">	No Change @cite dean1949renal	</span>|<span class="Warning">	Increased Then Normal @cite dean1949renal	</span>|<span class="Warning">	Increase @cite dean1949renal	</span>|

### Water Ingestion - Action

The water ingestion scenario simulates the patient drinking 1 liter of water with no other nutrients included. This causes a lowering of blood plasma concentration. This reaction leads to a decrease in the permeability of the tubules leading to an increase in urine production rate. The renal system in %BioGears shows a marked increase of dilute urine in line with that seen in a patient @cite baldes1934effect. 

<center>
<i>
Table 13. Urine production rate and plasma sodium concentrations behave as expected. Bladder substance handling is still not quite in line with experimental results.
</i>
</center>
|	Action	|	Notes	|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Urine Production Rate (mL/min)	|	Mean Arterial Pressure (mmHg)	|	Plasma Sodium Concentration (mg/L)	|	Urine sodium concentration (mg/L)	|
|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|
|	After stabilization patient given 1L of water	|		|	0	|	14400	|<span class="success">	Increase @cite baldes1934effect	</span>|<span class="success">	No Change @cite baldes1934effect	</span>|<span class="success">	Decrease @cite baldes1934effect	</span>|<span class="warning">	Decrease Then Slight increase @cite baldes1934effect	</span>|



### Dehydration
%BioGears does not currently support this functionality, although scenarios can be made and run for days at a time without fluid intake, simulating dehydration.

Dehydration conditions demonstrate the %Renal system's capability to conserve substances. As the blood volume is lowered, the production rate of urine is decreased
to a minimum by reabsorbing as much of the filtered water as possible. The outputs of the urine content are dependent on what food has been consumed during the dehydration
period. As the osmolarity of the plasma increases due to the drop in water volume, the renal system compensates by conserving as much water as possible @cite hall2011guyton.
This is accompanied by a drop in sodium excretion as it is reabsorbed to create a strong osmotic gradient to reabsorb the water. The renal methodology has no direct input in 
how dehydration is implemented, aside from clearing substances from the blood according to the time dictated by the consume meal condition and their clearance rates.

For other effects of dehydration on physiology see @ref EnergyMethodology.

<br><center>
<i>Table 14. %Renal behavior changes during dehydration. The excretion of substances slows as the water is conserved. There is also a notable
decrease in sodium excretion and urine sodium concentration in keeping with the validation data. </i>
</center>
|Parameter                    |Normal Values             |Dehydration              | 
|------------------------             |------------------------ |------------------------ |
|Urine Production Rate (mL/min))		  |	  1.1          | 0.07                |
|Urine Sodium Excretion Rate (mg/min) |		  2.4          |     0.65                  |
|Urine Sodium Concentration  (g/L)|        0.717      |       0.1           |

### Starvation
When nutrients from food are restricted, the kidneys attempt to conserve substances in the blood, which leads to decreases in urinary concentrations. This is accomplished 
by increasing reabsorptive capacity where it is possible to, namely with sodium. The %BioGears starvation scenario provides the patient with enough water to last the 4 day starvation 
period, which leads to decreasing concentrations of substances in the blood. The %renal methodology has no direct input in how starvation is implemented, aside from clearing substances
from the blood according to the time dictated by the consume meal condition.

For other effects of starvation on physiology see @ref EnergyMethodology .

<center>
<i>
Figure 15. There is inconsistent agreement with the validation detail in the %renal response to starvation. The middling responses in urine sodium,
calcium, and creatinine are due to an unvalidated drop in urine production rate, which skews the concentration values.
</i>
</center>
|	Action	|	Notes	|	Action Occurrence Time (s)	|	Sampled Scenario Time (days)	|	Urine Sodium Concentration (mmol/L)	|	Urine Calcium Concentration (mmol/L)	|	Creatine Concentration (mmol/L)	|	Urine Creatine Clearance (mL/min)	|
|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|
|	Patient is experiencing starvation	|		|	0	|	4	|<span class="warning">	Decrease @cite sapir1975ketones	</span>|<span class="warning">	Decrease @cite sapir1975ketones	</span>|<span class="success">	Increase @cite sapir1975ketones	</span>|<span class="success">	Decrease @cite sapir1975ketones	</span>|

### %Renal Stenosis - Condition

The %Renal Stenosis condition is validated against two different scenarios, testing severity and location of the condition. The moderate unilateral stenosis places a 60 percent severity on the left kidney and none on the right. The direct effect of a stenosis is a seen through a decrease in renal blood flow and glomerular filtration rate. Total vascular resistance and mean arterial pressure is increased because of the increased resistance on the renal artery. This causes a drop in cardiac output via the baroreceptor feedback response to the increased vascular resistance. The blood volume is steady in %BioGears due to a lack of angiotensin 2 and aldesterone hormonal responses.

<center>
<i>
Table 16. Direct affects of blood flow and glomerular filtration are seen in the stenosis condition. All cardiovascular effects meet validation except that blood volume decreases caused by release of aldesterone and angiotensin 2. 
</i>
</center>
|	Condition	|	Notes	|	Sampled Scenario Time (s)	|	Blood Volume (mL)	|	Systemic Vascular Resistance (mmHg/mL/s)	|	Cardiac Output (mL/min)	|	Mean Arterial Pressure (mmHg)	|	Systolic Pressure (mmHg)	|	Diastolic Pressure (mmHg)	|	%Renal Blood Flow (mL/min)	|	GFR (mL/min)	|
|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|
|	%Renal Stenosis 	|	60% unilateral occlusion of kidneys 	|	120	|<span class="warning">	Increase @cite klabunde2011cardiovascular	</span>|<span class="success">	Increase @cite klabunde2011cardiovascular	</span>|<span class="warning">	Mild decrease @cite anderson1990development	</span>|<span class="success">	Increase @cite debrunye2006assessment	</span>|<span class="success">	Increase @cite debrunye2006assessment	</span>|<span class="success">	Increase @cite debrunye2006assessment	</span>|<span class="success">	Decrease @cite textor2011atherosclerotic	</span>|<span class="success">	Decrease @cite textor2011atherosclerotic	</span>|


The severe bilateral stenosis condition tests the effects of kidney function due to a 90 percent increase in renal artery resistance on both kidneys. Like the unilateral condition, the kidney and cardiovascular function meet validation. A more pronounced decrease in renal blood flow and glomerular filtration rate can be seen due to the increased severity of the stenosis. 

<center>
<i>
Table 17. There is good agreement in %BioGears in response to the 90% bilateral stenosis. The effect on pressure is significant enough
to meet validation. The blood volume is unaffected. The systemic vascular resistance increases as expected along with the corresponding 
decrease in cardiac output. The renal blood flow and GFR each decrease as expected.
</i>
</center>
|	Condition	|	Notes	|	Sampled Scenario Time (s)	|	Blood Volume (mL)	|	Systemic Vascular Resistance (mmHg/mL/s)	|	Cardiac Output (mL/min)	|	Mean Arterial Pressure (mmHg)	|	Systolic Pressure (mmHg)	|	Diastolic Pressure (mmHg)	|	%Renal Blood Flow (mL/min)	|	GFR (mL/min)	|
|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|
|	%Renal Stenosis 	|	90% bilateral occlusion of kidneys 	|	120	|<span class="warning">	Increase @cite klabunde2011cardiovascular	</span>|<span class="success">	Increase @cite klabunde2011cardiovascular	</span>|<span class="warning">	Mild decrease @cite anderson1990development	</span>|<span class="success">	Increase @cite debrunye2006assessment	</span>|<span class="success">	Increase @cite debrunye2006assessment	</span>|<span class="success">	Increase @cite debrunye2006assessment	</span>|<span class="success">	Decrease @cite textor2011atherosclerotic	</span>|<span class="success">	Decrease @cite textor2011atherosclerotic	</span>|

@anchor renal-conclusions
Conclusions
-----------

%BioGears has implemented a functional renal model in conjunction with a whole-body physiology engine. Even with the simplified single-node nephron model, the system is able to accurately follow the overall renal transport of fluid and substances. The capabilities of the renal model allow for fluctuations in urine production rate, changes in blood pressure, and variations in clearance rates, as well as allowing for autogregulation with accurate responses to pressure changes and sodium concentration changes in the tubules.

@anchor renal-futurework
Future Work
===========

Coming Soon
-----------

These additions are expected in the near future:
- Acid-base balance reinvestigation
- Improvements to make the %Cardiovascular and %Renal systems more independent and modular
- Better handling of substance clearance

@anchor renal-recommended
Recommended Improvements
------------------------

These are some recommended improvements and additions:
- Peristalsis model
- Insulin as an optional substance for kidney validation/verification
- Remaining urinalysis substances
- Renin and angiotensin control
- Handle drugs similarly to metabolites
- Model kidney failure
- Catheterization
- Kidney stones
- Dialysis machine model

@anchor renal-appendices
Appendices
==========

Acronyms
--------

BUN - Blood Urea Nitrogen
	
GFR - Glomerular Filtration Rate

IV - Intravenous

MAP - Mean Arterial Pressure

RBC - Red Blood Cells

WBC - White Blood Cells

mL - Milliliters

min - Minutes

kg - Kilograms

g - Grams

ug - Micrograms

L - Liters

mol - Moles

mOsm - Milliosmoles

mmHg - Millimeters of Mercury

Data Model Implementation
-------------------------

@ref RenalSystemTable "Renal"

@anchor renal-compartments
Compartments
------------

* LeftKidney
	* LeftRenalArtery
	* LeftNephron
		* LeftAfferentArteriole
		* LeftEfferentArteriole
		* LeftGlomerularCapillaries
		* LeftBowmansCapsules
		* LeftTubules
	* LeftRenalVein
	
* RightKidney
	* RightRenalArtery
	* RightNephron
		* RightAfferentArteriole
		* RightEfferentArteriole
		* RightGlomerularCapillaries
		* RightBowmansCapsules
		* RightTubules
	* RightRenalVein
* Ureters
	* LeftUreter
	* RightUreter
* Bladder
%Hepatic Methodology {#HepaticMethodology}
=====================

Overview
========
@anchor hepatic-abstract
Abstract
--------

The %BioGears&reg; %Hepatic System provides the starting point for modeling the many functions of the liver. The current implementation focuses on nutrient management, including albumin production, glycogenolysis, glycogenesis, lipogenesis, and gluconeogenesis.

System Design
=============
@anchor hepatic-background
Background and Scope
--------------------

The %Hepatic System has many functions in the human body, including detoxification, protein synthesis, the production of digestive enzymes, regulation of metabolism, decomposition of red blood cells, and production of hormones. The liver is a highly specialized organ, mostly consisting of hepatocytes. Many hepatic functions were previously lumped into the %BioGears %Tissue System, but with the expansion of nutrient modeling, a %Hepatic System was called for. The %Hepatic System in %BioGears manages the storage of glucose as glycogen, the production of the protein albumin, the generation of triacylglycerol as tripalmitin in cases of nutrient excess, and the generation of glucose and ketones via gluconeogenesis in times of nutrient shortage.

@anchor heptaic-dataflow
Data Flow
---------

<br>
<img src="./images/Hepatic/HepaticDataFlow.PNG" width="800">
<center>
*Figure 1. The data flow for the %Hepatic System consists of PreProcess, Process, and PostProcess. ProduceAlbumin occurs in the PreProcess step, while all other %Hepatic functionality occurs in the Process step.*
</center><br>

### Reset, Conditions, and Initialization
Glycogen values are initialized to full. This replaces the previous version's intialization with a meal in the stomach. There are currently no conditions in the %Hepatic System.

### Preprocess
#### Produce Albumin
The liver produces albumin as part of normal function. A flat rate of albumin production is estimated as 0.15 mg/s, as found in @cite Jarnum1972plasma. At each time step, the mass is added to the liver extracellular compartment. Improvements to the albumin production and transport model are [planned](@ref hepatic-comingsoon).

### Process
#### Glycogenesis
When the body has high blood glucose, the liver converts blood-borne glucose molecules to the chain-form glycogen, which it then stores for times when blood glucose drops. This process, as well as the following Process methods, is governed by the hormones insulin and glucagon, the modeling of which can be read about in @ref EndocrineMethodology.

#### Glycogenolysis
Glycogenolysis is essentially the reverse of glycogenesis. Glycogen is broken down into glucose and dumped into the blood.

#### Lipogenesis
Triacylglycerol (as tripalmitin) is generated from excess glucose or amino acids (modeled as alanine).

#### Gluconeogenesis
Glucose is generated from three different sources in times of low blood glucose: generation from lactate (produced during anaerobic metabolism), synthesis from amino acids, or conversion of the glycerol backbone of triacylglycerol. Note that ketogenesis (the generation of acetoacetate) happens concurrently with the conversion of the triacylglycerol's glycerol backbone.

### Post Process
There is no system specific function for Post Process in the %Hepatic System.

### Assessments
There are no assessments in the %Hepatic System.

@anchor hepatic-features
Features, Capabilities, and Dependencies
----------------------------------------

The behavior of the %BioGears %Hepatic System is driven by the two blood-borne hormones insulin and glucagon. The ratio of these hormones, the so-called [hormone factor] (@ref endocrine-hormone-factor) determines whether the liver works to pull glucose from the blood (converting it to glycogen or triacylglycerol) or replenish blood glucose (by breaking down glycogen or converting amino acids, lactate, or glycerol). Once the direction of glucose change is determined, the stoichiometric relationships between reactants and products are used to increment substances as necessary. The general approach used is to define an upper and lower rate for the given process, then to use a logistic function to model the transition between rates as a function of hormone factor. The logistic equation is of the form

\f[R=R_{L} +\frac{R_{U} -R_{L} }{1+e^{-S(HF-O)} } \f] 
<center>
*Equation 1.*
</center><br>
Where *R* is the resultant rate of substance conversion, *R<sub>L</sub>* is the lower substance conversion rate, *R<sub>U</sub>* is the upper substance conversion rate, *S* is a shape modifier for the sigmoid, *O* is an offset value, and *HF* is the hormone factor in the blood compartment.

### Glycogenesis
The liver can store up to 5-8% of its weight as glycogen @cite guyton2006medical, and the total mass of the liver in %BioGears varies based on patient characteristics. The %Hepatic System sets the maximum value for liver glycogen to 6.5% of its mass. The muscle tissue also contains glycogen, up to 1-3% of muscle mass, but this glycogen is limited to the muscle, since when it is broken down, it retains phosphorylation and cannot diffuse @cite guyton2006medical. The rate of glycogenesis is estimated from @cite tardie2008glycogen, which states that under normal circumstances, 2% of total body glycogen is regenerated every hour after depletion. Furthermore, with ideal feeding, that rate can rise to 5% of total glycogen replenished each hour. In @cite rothman1991quantitation, the glycogenesis rate after a meal following a 64 hour fast is in agreement with this range. Thus, the glycogenesis rate logistic function is given two clear inputs, and the shape modifier and offset were empirically set to give good response. Note that for liver glycogenesis, the hormone factor is checked in the liver, whereas for muscle glycogenesis, it is checked in the muscle. Figure 1 below shows glycogenesis in action during a %BioGears run.

<img src="./images/Hepatic/Glycogenesis.png" width="800">
<center>
*Figure 1. Liver glycogen increases after ingesting 100g of carbohydrates at 30 minutes. Once the meal is digested and absorbed, the liver glycogen increases within the rates described.*
</center><br>

### Glycogenolysis
Glycogen is broken down in times of low blood glucose, such as during exercise or fasting. The lower glycogenolysis rate was calculated from @cite rothman1991quantitation, which shows a nearly linear rate for the first 22 hours of a fast, with a slowdown as the liver runs out of glycogen in later days of fasting. This resting glycogenolysis rate is echoed in @cite petersen2004regulation, which shows a similar resting rate. As metabolic demands increase under exercise conditions (when the hormone factor would drop), the glycogenolysis rate increases @cite petersen2004regulation. Note that muscle glycogenolysis does not occur in the same manner as hepatic glycogenolysis, and so is unaffected by this rate calculation. For more information on muscle glycogenolysis, see the [Tissue System] (@ref tissue-metabolic-production). Figure 2 below shows glycogenolysis during a 6 hour fast.

<img src="./images/Hepatic/Glycogenolysis.png" width="800">
<center>
*Figure 2. Liver glycogen drops from its initial full value as fasting progresses.*
</center><br>

### Lipogenesis
The liver has the ability to convert glucose or amino acids to lipid in times of excess. De novo hepatic lipogenesis rates are given by @cite mcdevitt2001denovo. Adipose tissue is also shown to contribute to generation of new lipid @cite chascione1987effect, giving a higher whole-body rate of lipogenesis. The two functions are lumped together in this one method, and the rate varies logistically between them. Because the amount of lipid derived from amino acid or the lipid derived from glucose depends on the amount of each precursor available, an assumption is made that the amount of lipid generated from either is proportional to the concentration in the liver. Thus, if there is twice as much amino acid in the liver as glucose, there will be twice as much lipid generated from amino acid. Lipogenesis is an energy-consuming process, but it is assumed that the energy cost is rolled into the basal metabolic rate. Lipogenesis results in carbon dioxide production, and if amino acids are converted, urea is produced as well. Figure 3 below shows instantaneous lipogenesis activity over 12 hours when ingesting 100g of carbohydrates.

<img src="./images/Hepatic/Lipogenesis.png" width="800">
<center>
*Figure 3. Lipogenesis occurs as long as there is a positive hormone factor.*
</center><br>

@anchor hepatic-gluconeogenesis
#### Gluconeogenesis
%Hepatic gluconeogenesis contributes heavily to maintenance of blood glucose. Three primary chemical conversions result in production of new glucose: the recombination of lactate into glucose, the removal and conversion of the glycerol backbone of triacylglycerol, and the conversion of amino acid into glucose precursor and urea @cite boulpaep2009medical. Because of this, gluconeogenesis is an important mechanism during fasting and starvation. In %BioGears, empirical rates are used for most of the gluconeogenic processes, tuned using values from @cite garber1974hepatic. For now, the liver converts any lactate in its extracellular compartment into glucose each timestep. Breakdown of triacylglycerol accomplishes two goals: generation of new glucose from the glycerol backbone and ketogenesis via cleaving of the palmitate chains. Note that %BioGears assumes that all ketones are acetoacetate; acetone and beta-hydroxybutyrate are not considered. Also note that ketogenesis is an oxygen-consuming process that contributes to the respiratory quotient. The breakdown of amino acids occurs at a constant rate, and urea is generated as a byproduct. All processes are stoichiometrically consistent. The total rate of glucose production and ketone production, as well as the proportions of substrates used to generate those substances, are generally in line with those outlined in @cite garber1974hepatic.

@anchor hepatic-dependencies
### Dependencies

The %Hepatic System is dependent on the key systems of the %BioGears Engine. The %Gastrointestinal System determines how much of each nutrient is absorbed into the blood. The %Endocrine System governs the synthesis rate of insulin and glucagon, which regulate %Hepatic processes, and the %Tissue System governs the diffusion of these nutrients between vascular and extravascular compartments.

@anchor hepatic-assumptions
Assumptions and Limitations
---------------------------

One major assumtion in %BioGears' implementation of nutrition is that fat remains in its triacylglycerol form. The formation of chylomicrons, the breakdown and reconsititution from fatty acids and the participation of albumin, or the packaging of LDLs are not considered. Lipogenesis is an energy-consuming process, but we assume that the energy required is rolled up in the basal metabolic rate. Additionally, while the substrates converted to fat in lipogenesis really depend on the presence and concentrations of facilitating enzymes, we assume that the proportion of glucose converted to fat with respect to amino acid is relative to the concentrations of glucose and amino acid in the liver, since the enzymes are not modeled. One limitation of the current implementation is that glucose is the only substance that directly affects the production of insulin and glucagon, when in reality both consumed protein and released epinephrine can affect these hormones. Many assumptions are made in the simple model of lactate generation and reconversion, the primary one being that gluconeogenesis is not rate limited in this respect.

Many of these assumptions are planned to be addressed or reduced in upcoming releases.

Conditions
----------
There are no conditions associated with the %BioGears %Hepatic system.

Actions
-------
There are currently no actions in the %Hepatic system.

Events
------
- LiverGlycogenDepleted: Set when liver glycogen drops below one gram. Removed when liver glycogen returns above 10 grams.
- MuscleGlycogenDepleted: Set when muscle glycogen drops below one gram. Removed when muscle glycogen returns above 10 grams.

Results and Conclusions
=======================

@anchor hepatic-validation
Validation - Resting Physiologic State
--------------------------------------

The %Hepatic System regulates substances that are validated in other systems. For validation of resting physiological state, see the relevant sections in the [Tissue] (@ref tissue-validation) and [Gastrointestinal] (@ref GI-results) Systems.

Validation - Actions
--------------------------------------
There are currently no validated actions or conditions associated with the %Hepatic system. However, some of its funcitonality is validated in the [Energy] (@ref energy-results) System.

@anchor hepatic-conclusions
Conclusions
-----------

The %Hepatic System provides functionality for the storage and conversion of ingested nutrients to maintain physiological homeostasis.

Future Work
===========
@anchor hepatic-comingsoon
Coming Soon
-----------

Recommended Improvements
------------------------
- FFA/Albumin binding and better fat metabolism
- Epinephrine interactions with insulin/glucagon/exercise

Appendices
==========

Data Model Implementation
-------------------------

@ref HepaticSystemTable "Hepatic" %Nervous Methodology {#NervousMethodology}
==========================

@anchor nervous-overview
Overview
========
Abstract
--------
The %BioGears %Nervous System models the central %nervous system (CNS) and peripheral %nervous system (PNS). Currently, the system models afferent signal generation as a result of baroreceptor, chemoreceptor, lung-strech receptor activity.  These signal are processed by a rudimentary central nervous system (CNS) that relays commands to efferent effectors.  Additionally, the %Nervous System provides tracks parameters related to traumatic brain injury events, pain stimulus response, and drug effects.  In the future, other feedback mechanisms will be added for improved modeling of homeostatic and crisis states.

@anchor nervous-introduction
Introduction
------------
The %Nervous System in the human body is comprised of the central nervous system (CNS) and the the peripheral %nervous system (PNS). The CNS contains the brain and the spinal cord, while the PNS is subdivided into the somatic, autonomic, and enteric nervous systems. The somatic system controls voluntary movement. The autonomic nervous system is further subdivided into the parasympathetic and sympathetic nervous system. The sympathetic nervous system is activated in times of stress and crisis, commonly referred to as the "flight or fight" behaviors. The parasympathetic nervous system controls homeostasis functions during the relaxed state. The enteric nervous system controls the gastrointestinal system. Many of these behaviors are tightly connected to the %Endocrine System. In many cases, the %Nervous System receptors trigger hormone releases that cause systemic changes, such as dilation and constriction of vessels, heart contractility changes, and respiration changes @cite guyton2006medical. 

%BioGears will focus on a select few mechanisms within the %Nervous System to provide feedback for both homeostasis and crisis behavior in the body. The list is shown below:
 - Baroreceptors (aortic, carotid, and low-pressure)
 - Chemoreceptors
 - Pulmonary stretch receptors
 - Local autoregulation
 - Temperature Effects
 - Generic parasympathetic and sympathetic effects
 - Pain stimulus
 - Sleep and sleep deprivation

 Additionally, %BioGears will model the following features related to brain function: 
 - Simple model of traumatic brain injury (TBI)
 - Intracranial pressure
 - Pupillary response to drugs and TBI

@anchor nervous-system-design
System Design
=============
Background and Scope
--------------------
The %BioGears Nervous System began as a low-fidelity model that tracked metrics associated with brain injury.  Over time, feedback models local to other %BioGears systems (i.e. baroreceptors, chemoreceptors) were relocated to the %Nervous System with the goal of moving towards a centralized %Nervous model that serves as a global controller.  The primary difficulty with such a model is that most detailed %Nervous models rely on differential equations that output nerve firing rates.  If the frequencies of these firing rates are too high compared to the time-step used to model the equations, the solution will be unstable.  Fortunately, as %BioGears stability has been improved, the simulation time-step has been decreased to its current value of 0.02 s.  This step size allowed us to work toward strengthenging the interactions between the existing models.

Data Flow
---------
An overview of the data flow in the %BioGears %Nervous system is shown in Figure 1.

### Initialization and Stabilization
The %BioGears initialization and stabilization is described in detail in the [stabilization section](@ref system-stabilization) of the @ref SystemMethodology report. Setpoints related to Nervous model are reset once the %Cardiovascular system reaches a homeostatic state.  These include the baseline baroreceptor operating pressure, the baseline oxygen and carbon dioxide blood gas levels, the baseline sympathetic and parasympathetic firing rates, and the baseline cerebral blood flow.

### Preprocess
#### Afferent Signal Generation
This method surveys the current state of the patient (blood pressure, blood-gas levels, lung capacity) and relays information to the central nervous system.
##### Baroreceptor Feedback
The baroreceptor mechanism provides rapid negative feedback control of arterial pressure. A drop in arterial pressure is sensed by the baroreceptors and leads to an increase in sympathetic activity and vagal (parasympathetic) withdrawal. These changes operate with the goal of maintaining arterial pressure at its healthy resting level.  %BioGears distinguishes between aortic, carotid, and low-pressure (cardiopulmonary) receptors.  Aortic and carotid receptors are both sensitive to changes in systolic arterial pressure, but their relative locations in the body affect this response.  Aortic baroreceptors respond to the transmural pressure between the aorta and the intrapleural space.  The carotid baroreceptors, located a distance above the heart, are affected by pressure head (except when an individual is lying down).  Low-pressure receptors are located near the venous return to the heart and are therefore sensitive to the central venous pressure and pleural pressure. %BioGears uses a stress-strain relationship to calculate the signal generated at the aortic and carotid baroreceptors @cite randall2019model and a first-order, low-pass filter to generate the low-pressure receptors signal @cite lim2013cardiovascular.
 
##### Chemoreceptor Feedback
The chemoreceptor mechanism provides feedback control to regulate the partial pressures of oxygen and carbon dioxide in the blood.  The response is divided in to two components:  peripheral and central.  Peripheral chemoreceptors (which reside in the aortic arch and carotid arteries) are sensitive to deviations in both oxygen and carbon dioxide pressures.  The central chemoreptors, so named because of their location in the central nervous system, respond only the carbon dioxide pressure changes.  Assuming that each receptor class operates independently, the feedback from the central and peripheral chemoreceptors is summed to generate changes in respiration frequency and respiratory driver pressure to restore blood gas levels to their set points.  The peripheral chemoreceptor signal is also processed by the central nervous model to account for its contribution to cardiovascular activity.    

##### Pulmonary Stretch Receptors
Stretch receptors respond to changes in lung volume.  They tend to balance the cardiovascular effects induced by peripheral chemoreceptors @cite ursino2000acute.  

#### Central Nervous (CNS) Response
This method consolidates the signals generated by the afferent models and adjusts the relative firing rates of the sympathetic and parasympathetic arms of the nervous system.  The sympathetic signal consists of two outputs:  a sinoatrial portion that affects heart rate and elastance and a peripheral portion that affects systemic resistance and unstressed venous volume.  The parasympathetic signal only acts upon the heart rate.  The threshold for sympthetic firing is lowered under hypoxic conditions to model the ischemic response @cite ursino2000acute.

#### Efferent Response
The efferent response carries out the work of updating the state of the cardiovascular system according to the autonomic response determined by the CNS.  This model determines the normalized changes in heart rate, heart elastance, peripheral vessel resistance, and peripheral venous compliance.  The peripheral resistance response distinguishes between the splanchnic, extrasplanchnic, and skeletomuscular regions of the body to account for preferential blood-flow shunting during certain extreme responses (e.g. maintaining flow to vital organs during hemorrhage).

#### Local Autoregulation
The brain, muscle, and myocardium prioritize blood flow and oxygen delivery.  %BioGears models cerebral autoregulation in which vessels in the brain constrict or dilate to maintain constant cerebral blood flow, perfusion pressure, and oxygen saturation.  %BioGears also models local regulation in the vessels of the muscle and myocardium in which vasodilation occurs in response to oxygen depletion.

#### Check Pain Stimulus
Pain can be initiated in BioGears via a Pain Stimulus action.  The BioGears pain response takes into account both the magnitude of the pain action and the susceptibility of the virtual patient to pain (set in the Patient definition file).  The result is an output  corresponding to the pain Visual Analog Scale (VAS) on a 0-10 interval.  The VAS score dictates the patient cardiovascular and respiratory response to the pain stimulus.  Increased epinephrine production is also stimulated.  

### Calculate Sleep Effects
As a patient becomes more and more sleep deprived many physiological factors begin to become manifest in the body. Notabley, control of hormone function becomes greately deminished as a function of sleep deprivation. %BioGears models this disregulation as the patient is awake for a prolonged period of time. This is of particular importance in prolonge field care scenarios where evacuation to higher levels of care may be denied for up to 72 hours. In these intances, if the patientes sleep isn't managed properly, insuluin synthesis will become greatly dimished, resulting in hyperglycemia. In addition, %BioGears now provides access to the psychomotor vigilance task patient assesment which will allow the care provider to determine sleep debt through analysis of reaction time and lapses in attention while performing a focused task.

### Process
#### Check %Nervous Status
Events associated with nervous system functionality are evaluated.  First, intracranial pressure is checked to determine whether or not to throw the Intracranial Hypertension or Intracranial Hypotension events.  Then, the presence of fasciculation (muscle spasms) is evaluated based on the activity of the compounds Sarin and Succinylcholine.  The latter drug generally produces transient fasciculation which disappear when neuromuscular block is achieved @cite appiah2004pharmacology.  For a description of the neurological effects of Sarin, see the [Nerve Agent: Sarin](@ref sarin-nerve) section in the Drugs documentation.  It should be noted that fasciculation can also be caused by an ion imbalance; however, this functionality is not yet active.  When completed, the criteria for this type of event will be calcium deficiency. If the patient's arterial calcium concentration falls below one milligram per deciliter, then fasciculation will be triggered @cite gropper2013nutrition.

#### SetPupilEffects
Pupil modifiers are retrieved from the %Drugs System and combined with any modifiers from TBI before being applied to the eyes.

### Post Process
There is no system specific functionality for the %Nervous System in Post Process.

@anchor nervous-assessments
### Assessments
Assessments in %BioGears are data collected and packaged to resemble a report or analysis that might be ordered by a physician. No %BioGears assessments are associated with the %Nervous System.

<br>
<img src="./images/Nervous/NervousDataFlow.png" width="800">
<center>
*Figure 1. The data flow for the %Nervous System consists of Preprocess, Process and Postprocess. In Preprocess, the baroreceptor feedback is calculated. In Process, the brain metrics are checked to see if event thresholds have been reached and pupil effects are determined and applied.*
</center><br>

@anchor nervous-features
Features, Capabilities, and Dependencies
----------------------------------------
### Baroreceptors
%BioGears calculates the strain exerted on the aortic and carotid baroreceptors as @cite randall2019model

\f[ \epsilon_{w} = 1 - \sqrt{\frac{1 + \exp(-q_w\left(P_{input} - s_w\right))}{A+\exp(-q_w\left(P_{input}-s_w\right))}} \f]
<center>
*Equation 1.*
</center><br>

The input pressure (P<sub>input</sub>) is the systolic arterial pressure (P<sub>sys</sub>) for the carotid baroreceptors and the difference between systolic pressure and pleural pressure (P<sub>systolic</sub> - P<sub>pleural</sub>) for aortic baroreceptors.  The parameters for Equation 1 are maximum stressed to unstressed vessel cross-sectional area (A), steepness (q<sub>w</sub>), and operating point (s<sub>w</sub>) and the output is wall strain (&epsilon;<sub>w</sub>).  As described in the [cardiovascular](@ref cardiovascular-initialize) methodology report, the%BioGears %cardiovascular system is initialized according to patient definitions and stabilized to a homeostatic state. The baroreceptor operating pressure is the systolic pressure following the engine stabilization period. This operating point is adjusted dynamically with certain actions and insults, as shown in Equation 2.

\f[ s_w = P_{sys, set} + \Delta P_{drugs} + \Delta P_{exercise} + \Delta P_{pain} \f]
<center>
*Equation 2.*
</center><br>

Where &Delta;<b>p</b><sub>drugs</sub>, &Delta;<b>p</b><sub>exercise</sub>, and &Delta;<b>p</b><sub>pain</sub> are changes in apparent operating point due to certain drugs, exercise, and pain response.  It is known that aortic and carotid baroreceptors adapt to sustained changes in arterial pressure.  We model adaptation by adjusting the operating point as a function of systolic arterial pressure @cite pruett2013population

\f[ \frac{ds_w}{dt} = k_{adapt}\left(P_{sys} - s_w\right) \f].
<center>
 *Equation 3.*
</center><br>

The value of k<sub>adapt</sub> reflects a half-time of approximately 16 hours for the baroreceptors to acclimate to a sustained step-change in systolic pressure @cite pruett2013population. 

The baroreceptors are assumed to operate as a spring-dashpot system.  The strain induced by the stress $\epsilon_w $ is determined by the differential equation @cite randall2019model

\f[ \frac{d \epsilon_{b}}{dt} = \frac{-\epsilon_{b} + K_b \epsilon_{w}}{\tau_{b}} \f]
<center>
*Equation 4.*
</center><br>

The strain signal &epsilon;<sub>b</sub> determined for each baroreceptor class (aortic and carotid) is processed in the Central Nervous response.

The low-pressure baroreceptor signal is generated by sampling the transmural pressure between the venous return (CVP) and the pleural space @cite lim2013cardiovascular.

\f[ \frac{d\tilde{P_{cp}}}{dt} = \left(\frac{1}{\tau_{cp}}\right)\left(-\tilde{P_{cp}} + \left(CVP - P_{pleural}\right)\right) \f]
<center>
*Equation 5*
</center><br>

\f[ f_{cp} = \frac{f_{cp,max}}{1.0 + \exp(\left(CVP_{base} - P_{pleural,base}\right) - \tilde{P_{cp}})} \f]
<center>
*Equation 6*
</center><br>

Equation 5 acts as a low-pass filter that reduces signal noise and Equation 6 generates a firing rate (f<sub>cp</sub>) that is processed by the CNS. &tau;<sub>cp</sub> and f<sub>cp,max</sub> are the time constant for the response and the maximum firing rate, while CVP<sub>base</sub> and P<sub>pleural,base</sub> represent the normal central venous and pleural pressures.

@anchor nervous-features-chemoreceptors
### Chemoreceptors
The chemoreceptors are cells that are sensitive to changes in blood gas concentration. Peripheral chemosensitive cells--located in the aortic arch--detect fluctuations in both arterial carbon dioxide and oxygen partial pressures.  The central chemoreceptors, which reside in the central nervous system (CNS), respond to changes in cerebral blood pH caused by irregular carbon dioxide levels.  It is assumed in this model that carbon dioxide transport across the blood-brain barrier and equilibration therein are rapid processes; that is, it assumed that the central chemoreceptor response is proportional to deviations in arterial carbon dioxide.  

Control of respiration is based on the model developed by Albanese, Magosso, and Ursino in @cite ursino2000acute, @cite magosso2001mathematical, @cite ursino2002theoretical, and @cite albanese2016integrated.  It is assumed that the central (c) and peripheral (p) feedback define deviations from baseline patient respiration frequency (f) and peak respiratory driver pressure (P<sub>max</sub>), and that their contributions are independent (and thus additive).  

\f[f_{target} = f_{base} + \Delta f_c + \Delta f_p \f]
<center>
*Equation 7.*
</center><br>

\f[P_{max,target} = P_{max,base} + \Delta P_{max,c} + \Delta P_{max,p} \f]
<center>
*Equation 8.*
</center><br>

Clearly, the blood gas levels are normal, central V<sub>c</sub> and peripheral V<sub>p</sub> changes in ventilation are 0, and target ventilation V<sub>target</sub> equals patient baseline ventilation V<sub>base</sub>.

The evolution of the central chemoreceptor feedback is assumed to be a function of arterial carbon dioxide partial pressure (P<sub>CO<sub>2</sub></sub>) deviation from its normal set point (P<sub>CO<sub>2,set</sub></sub> = 40 mmHg).

\f[ \frac{d\Delta E_{c}}{dt} = \left(\frac{1}{ \tau_{E,c}} \right) \left( -\Delta{E_{c}} + g_{E,c}\left(P_{CO_{2}}(t) - P_{CO_{2,set}}\right)\right) \f]
<center>
*Equation 9.*
</center><br>

The effects (E) are frequency (f) and driver amplitude (P) with associated time constants (&tau;<sub>c</sub>) and controller gains (&tau;g<sub>c</sub>).  Equation 9 is identical to that reported in @cite albanese2016integrated, except that a delay term is omitted.  This omission is a result of being currently unable to model delay differential equations in %BioGears.  

The form of the peripheral feedback in Equation 10 is similar to Equation 8, but the response is dictated by the combined effects of carbon dioxide pressure and oxygen saturation.  Equations 11-12 show how this firing rate (f(t)) is calculated.  As before, this equation is applied twice per effect (E = {f, P}).

\f[ \frac{d\Delta E_{p}}{dt} = \left(\frac{1}{ \tau_{E,p}} \right) \left( -\Delta E_{p} + g_{E,p}\left(f(t) - f_{set}\right)\right) \f]
<center>
*Equation 10.*
</center><br>

\f[ \frac{df}{dt} = \frac{1}{\tau_{f}}\left(-f(t) + \psi\right) \f]
<center>
*Equation 11.*
</center><br>

\f[ \psi = \frac{f_{max} + f_{min}*exp\left(\frac{P_{O_{2}} - P_{O_{2,half}}}{k_{O_{2}}}\right)}{1 + exp\left(\frac{P_{O_{2}} - P_{O_{2,half}}}{k_{O_{2}}}\right)}*\left[K*ln\left(\frac{P_{CO_{2}}}{P_{CO_{2,set}}}\right) + \gamma\right] \f]
<center>
*Equation 12.*
</center><br>

Equation 12 is a sigmoid that outputs a chemoreceptor firing rate (bounded by f<sub>min</sub> and f<sub>max</sub>) based on the current arterial oxygen pressure (P<sub>O<sub>2</sub></sub>).  The variables P<sub>O<sub>2,half</sub></sub> and k<sub>O<sub>2</sub></sub> are the half-max and slope parameters that define the shape of the sigmoid, a shape which is increasingly adjusted as the arterial carbon dioxide pressure (P<sub>CO<sub>2</sub></sub>) deviates from normal (P<sub>CO<sub>2,set</sub></sub>).  K and &gamma; are tuning parameters required to maintain a steady output when oxygen and carbon dioxide pressures are at their normal values.  Values for all parameters in Equations 9-12 are available in @cite magosso2001mathematical.

### Pulmonary Stretch Receptors
Pulmonary stretch receptors are modeled as in @cite ursino2000acute, except that total lung volume (V<sub>lung</sub>) is used as the input rather than tidal volume.  We make this change because the total lung volume in %BioGears is a more stable output than tidal volume.

\f[ \frac{df_{ps}}{dt} = \frac{-f_{ps} + G_{ps}V_{lung}}{\tau_{ps}} \f]
<center>
*Equation 13.*
</center><br>

### Central Nervous Response
The CNS model outputs three signals: a sympathetic efferent signal to the sinoatrial node (f<sub>es,s</sub>), a sympathetic efferent signal to the periphery (f<sub>es,p</sub>), and an efferent vagal (parasympathetic) signal to the sinoatrial node (f<sub>ev</sub>).  These signals are determined by weighting the following inputs from the afferent arm of the nervous system: aortic baroreceptors (f<sub>ab</sub>), carotid baroreceptors (f<sub>cb</sub>), cardiopulmonary baroreceptors (f<sub>cp</sub>), peripheral chemoreceptors (f<sub>pc</sub>), and pulmonary stretch receptors (f<sub>ps</sub>).  The equations for the three efferent signals have been developed in @cite ursino2000acute, @cite magosso2001mathematical, @cite magosso2002theoretical, and @cite albanese2016integrated.  Weighting parameters (W<sub>j</sub>) were established from these sources as well as from @cite lim2013cardiovascular and @cite liang2006simulation.

The sympathetic efferent firing rates are modeled as sigmoids:

\f[ f_{es,s} = f_{es,\infty} + \left(f_{es,0} - f_{es,\infty}\right)exp(k_{es}\left(W_{s,ab}f_{ab} + W_{s,cb}f_{cb} + W_{s,pc}f_{pc} + W_{cp}\left(f_{cp} - f_{cp,0}\right) - \theta_{s}\right)) \f]
<center>
*Equation 14.*
</center><br>

\f[ f_{es,p} = f_{es,\infty} + \left(f_{es,0} - f_{es,\infty}\right)exp(k_{es}\left(W_{p,ab}f_{ab} + W_{p,cb}f_{cb} + W_{p,pc}f_{pc} + W_{p,ps}f_{ps} + W_{cp}\left(f_{cp} - f_{cp,0}\right) - \theta_{p}\right)) \f]
<center>
*Equation 15.*
</center><br>

The parameters &theta;<sub>s</sub> and &theta;<sub>p</sub> refer to the firing thresholds of the sinoatrial and peripheral sympathetic signals.  These thresholds are adjusted under hypoxic and hypercapnic scenarios to model the ischemic response @cite Magosso2001Mathematical.

\f[ \frac{d\Delta\theta_{O_2,j}}{dt} = \frac{1}{\tau_{isc}}\left(-\Delta\theta_{O_2,j} + \frac{\chi_{j}}{1+\exp(\frac{P_{O_2}-h_{O_2,j}}{k_{isc,j}})}\right) \f]
 <center>
*Equation 16.*
</center><br>

\f[ \frac{d\Delta\theta_{CO_2,j}}{dt}= \frac{-\theta_{CO_2,j}+g_{c,j}\left(P_{CO_2}-P_{CO_2,n}\right)}{\tau_c} \f]
 <center>
*Equation 17.*
</center><br>

\f[ \theta_{j} = \theta_{j,n} - \Delta\theta_{O_2,j} - \Delta\theta_{CO_2,j} \f]
 <center>
*Equation 18.*
</center><br>

In Equations 16-18, the subscript "j" refers to either the sinoatrial or peripheral ischemic response.  The parameters &theta;<sub>j,n</sub> refer to the basal values the thresholds take on at normal blood-gas levels.  The maximum slope and half-saturation parameters (h<sub>O2,j</sub> and k<sub>isc</sub>, respectively) determine the overall shape of the response.

The parasympathetic signal is a sigmoidal function of the baroreceptor firing rate and linear with respect to the chemoreceptor and pulmonary stretch signals.

\f[ f_{ev} = \left(\frac{f_{ev,0} + f_{ev,\infty}\exp(\frac{f_{b,avg}-f_{b,avg,0}}{k_{ev}})}{1+\exp(\frac{f_{b,avg}-f_{b,avg,0}}{k_{ev}})}\right) + W_{v,pc}f_{pc}+W_{v,ps}f_{ps} - \theta_{ev} \f]
 <center>
*Equation 19.*
</center><br>

Unlike the sympathetic efferent pathway, the parasympathetic path assumes a constant firing theshold \(\theta_{ev}\).  The value f<sub>b,avg</sub> is the average firing rate of the carotid and aortic baroreceptors.  The low-pressure baroreceptors do not effect the parasympthetic signal.

### Efferent Response    
The three signals relayed by the CNS (f<sub>es,s</sub>, f<sub>es,p</sub>, and f<sub>ev</sub>) are used to update the cardiovascular model using effector equations outlined in @cite ursino2000acute.  These equations have been normalized so that their outputs represent a fractional change from baseline.  As in @cite ursino2000acute, the response to the sympathetic activation is assumed to involve a logarithmic transformation, whereas the vagal response is linear.  Furthermore, the sympathetic signals are bounded at the minimum signal f<sub>es,min</sub> to maintain continuity.  In each of the normalized equations, the values &sigma;<sub>0</sub> correspond to the effector input at the baseline sympathetic and parasympathetic firing rates.  

#### Heart Rate
\f[ \frac{d \sigma_{HR, s}}{dt} = \frac{-\sigma_{HR,s} + G_{HR,s}\ln(f_{es,s}-f_{es,min}+1}{\tau_{HR,s}} \f]
 <center>
*Equation 20.*
</center><br>

\f[ \frac{d \sigma_{HR, v}}{dt} = \frac{-\sigma_{HR,v} + G_{HR,v}f_{ev}}{\tau_{HR,v}} \f]
 <center>
*Equation 21.*
</center><br>

\f[ HR_{norm} = \frac{1 + \sigma_{HR,s,0} + \sigma_{HR,v,0}}{1 + \sigma_{HR,s} + \sigma_{HR,v}} \f]
 <center>
*Equation 22.*
</center><br>

In Equation 22, the initial state of the heart rate appears in the numerator because Equations 21-22 reflect changes in heart period.  Thus, inverting the ratio of next heart period to baseline heart period gives the normalized change in heart rate.

#### Heart Elastance

\f[ \frac{d \sigma_{E}}{dt} = \frac{-\sigma_{E} + G_{E}\ln(f_{es,s}-f_{es,min}+1}{\tau_{E}} \f]
 <center>
*Equation 23.*
</center><br>

\f[ E_{norm} = \frac{1 + \sigma_{E}}{1 + \sigma_{E,0}} \f]
 <center>
*Equation 24.*
</center><br>

#### Vessel Resistance

\f[ \frac{d \sigma_{R,j}}{dt} = \frac{-\sigma_{R,j} + G_{R,j}\ln(f_{es,p}-f_{es,min}+1}{\tau_{R,j}} \f]
 <center>
*Equation 25.*
</center><br>

\f[ R_{j,norm} = \frac{1 + \sigma_{R,j}}{1 + \sigma_{R,j,0}} \f]
 <center>
*Equation 26.*
</center><br>
  
In Equations 25-26, j = splanchnic, extrasplanchnic, or muscle.  Each region has distinct controller gains and time constants associated with it.

#### Veinous Compliance
\f[ \frac{d \sigma_{C}}{dt} = \frac{-\sigma_{C} + G_{C}\ln(f_{es,p}-f_{es,min}+1}{\tau_{C}} \f]
 <center>
*Equation 27.*
</center><br>

\f[ C_{norm} = \frac{1 + \sigma_{C}}{1 + \sigma_{C,0}} \f]
 <center>
*Equation 28.*
</center><br>

The normalized compliance is applied to each fluid capacitor in the systemic portion of the BioGears cardiovascular circuit.

### Local Autoregulation
#### Cerebral Autoregulation
The cerebral autoregulation model is based on that of @cite ursino1998cerebral and @cite ursino2001role.  The model distinguishes between changes in large arteries (upstream resistor in the BioGears cerebral circuit) and small arteries (downstream resistor).  The large arteries respond to changes in cerebral perfusion pressure (CPP), while the small arteries are sensitive to changes in cerebral blood flow (CBF).  Each response is attenuated by cerebral venous oxygen saturaton and cerebral carbon dioxide levels.  In all equations, a subscript "n" refers to the normal, basal value of the parameter.

The response of large arteries to changes in CPP is governed by @cite ursino1998cerebral

\f[\frac{dx_{auto,l}}{dt} = \frac{-x_{auto,l} + G_{auto,l}\left(CPP - CPP_{n}\right)}{\tau_{auto,l}} \f]
<center>
*Equation 29.*
</center><br>  

The response of small arteries to CBF is likewise given by @cite ursino1998cerebral 
\f[\frac{dx_{auto,s}}{dt} = \frac{-x_{auto,s} + G_{auto,l}\left(\frac{CBF-CBF_n}{CBF_n}\right)}{\tau_{auto,s}} \f]
<center>
*Equation 30.*
</center><br>  

The effects of carbon dioxide @cite ursino1998cerebral and oxygen saturation @cite ursino2001role are tracked as follows:

\f[\frac{dx_{CO_2}}{dt} = \frac{-x_{CO_2} + A_{CO_2}G_{CO_2}log_{10}\left(\frac{P_{CO_2}}{P_{CO_2,n}}\right)}{\tau_{CO_2}} \f]
<center>
*Equation 31.*
</center><br> 

\f[ A_{CO_2} = \left(1+\exp(-K_{CO_2}\left(\frac{CBF-CBF_n}{CBF_n}\right) - b_{CO_2})\right)^{-1} \f]
<center>
*Equation 32.*
</center><br> 

\f[\frac{dx_{O_2}}{dt} = \frac{-x_{O_2} + G_{O_2}\left(S_{v,O_2}-S_{v,O_2,n}\right)}{\tau_{O_2}} \f]
<center>
*Equation 33.*
</center><br>

The total autoregulatory effects used to scale the large and small cerebral resistances are:

\f[ x_{large} = x_{auto,l} + x_{CO_2} + x_{O_2} \f]
\f[ x_{small} = x_{auto,s} + x_{CO_2} + x_{O_2} \f]
<center>
*Equation 34.*
</center><br>

#### Coronary and Skeletomuscular Autoregulation
Autoregulation in the heart and muscle is modeled similarly to @cite ursino2000acute.  The autoregulatory parameter for each location is:

\f[ \frac{dx_{O_2,j}}{dt} = \frac{-x_{O_2,j} + G_{O_2,j}\left(C_{v,O_2,j}-C_{v,O_2,jn}\right)}{\tau_{O_2,j}} \f]
<center>
*Equation 35.*
</center><br>

where j is either the muscle or heart, C<sub>v,O2</sub> is the current venous oxygen concentration, and C<sub>v,O2,n</sub> is the baseline venous oxygen concentration.

The change in resistance of the circuit path providing blood flow to the heart is 

\f[ R_{h} = \frac{R_{h,0}}{1 + x_{O_2,h}}\f]
<center>
*Equation 36.*
</center><br>
 
in which R<sub>h,0</sub> is the baseline resistance (constant).  The expression describing the change in muscle resistance is similar to Equation 36.

\f[ R_{m} = \frac{R_{m,next}}{1 + x_{O_2,m}}\f]
<center>
*Equation 37.*
</center><br>

It should be noted, however, that because the muscle resistance is also controlled by the sympathetic response, the resistance against which the autoregulation acts is that determined by Equation 26.
  
### TBI
Traumatic brain injuries are relatively common, affecting millions annually. They are usually defined as a blunt or penetrating injury to the head that disrupts normal brain function. Furthermore, they are classified as either focal (e.g. cerebral contusions, lacerations, and hematomas) or diffuse (e.g. concussions and diffuse axonal injuries) @cite adoni2007pupillary. The scope of the %BioGears TBI model is somewhat limited by the low fidelity of the modeled brain. The brain is represented in the current %BioGears %Cardiovascular circuit with only two resistors and a compliance, all within one compartment. Because the circuit is modeled in this way, TBI is considered as an acute, non-localized, non-penetrating injury from a circuit perspective. Thus, the TBI model can account for intracranial pressure and cerebral blood flow on the basis of the whole brain, but not to specific areas of the brain. However, %BioGears does provide for three injury inputs: diffuse, right focal, and left focal.  Similarly to the @ref RenalMethodology "Renal System", the %BioGears brain circuit could be expanded to accommodate a higher-fidelity brain model.

Three important metrics are used to evaluate patients with traumatic brain injuries: intracranial pressure (ICP), cerebral blood flow (CBF), and cerebral perfusion pressure (CPP). Patients with brain injuries often exhibit increased intracranial pressure, decreased cerebral blood flow, and a cerebral perfusion pressure outside of a normal range @cite steiner2006monitoring. Cerebral perfusion pressure is defined as

\f[ CPP = MAP - ICP \f]
<center>
*Equation 8.*
</center><br>

 Where MAP is the mean arterial pressure. In order to model these behaviors, the Brain Injury action in %BioGears will modify the resistors of the brain circuit, which is shown in Figure 5 below. The brain circuit is a section of the @ref cardiovascular-features "cardiovascular circuit".

<img src="./images/Nervous/BrainCircuit.png" width="550">
<center>
*Figure 5. The %BioGears brain is represented by two resistors and a compliance. The upstream resistor, R1, is connected to the aorta, and the downstream resistor, R2, is connected to the vena cava.*
</center><br>

By increasing R1 and R2, the ICP can be increased while CBF decreases. The resistors are tuned based on the severity (on a scale from 0 to 1) of TBI such that ICP is above 25 mmHg and CBF is near 8 mL per 100 grams of brain tissue per minute for the most severe injury.

### Pupillary Response
The pupil's diameter is controlled by the autonomic nervous system in order to affect the amount of light entering the eye. This diameter control is carried out by two muscles: the pupilloconstrictor, which is parasympathetically controlled (via the neurotransmitter acetylcholine), and the pupillodilator, which is sympathetically controlled (via the neurotransmitter norepinephrine) @cite adoni2007pupillary. The synaptic pathways for these two pupil effects differ, and thus, any deviation from normal pupil behavior can shed light on any interference, whether that be a brain injury causing increased pressure on a nerve or a drug effect interfering with synaptic transmission. Because pupil examination is informative yet minimally invasive, it is an excellent tool for helping to diagnose neurological problems. The PERRLA assessment is commonly used to classify pupillary behavior: Pupils Equal, Round, Reactive to Light, and Accomodating. Pupillary response in %BioGears is modeled by tracking any modification to pupil size and light reactivity for each eye. Both TBI @ref nervous-actions "actions" and @ref drugs-pharmacodynamics "drugs" can apply modifiers to pupil size and reactivity.

Pupil size and reactivity can vary patient-to-patient based on many factors, including age, mental and emotional state, and ambient light conditions @cite winn1994factors. Furthermore, pupillary assessments like PERRLA are often qualitative rather than quantitative. Because of this, %BioGears models pupillary modifiers rather than discrete pupil sizes and size changes over time. In this way, the pupillary effects can be assessed qualitatively without the need for baseline patient values for pupil size and reactivity.

There are two ways to affect the pupils in %BioGears: drug pharmacodynamic effects and TBI. For a discussion of pharmacodynamic effects on pupillary response, see @ref drugs-pharmacodynamics "Drugs Methodology". The other method of influencing pupillary response is TBI. Increasing intracranial pressures constrict ocular nerves, causing pupillary disruptions @cite kingsly2012severe. Because of this relationship between ICP and pupil effects, ICP is the metric on which pupillary modifiers are based. For the pupil size modifier, Equation 9 was developed so that pupil size (m<sub>s</sub>) begins to see effects when ICP increases above around 20 mmHg, the pressure at which recoverable brain damage is often observed @cite steiner2006monitoring, hitting its maximum slope at 22.5 mmHg, and leveling off as ICP approaches 25 mmHg. For pupil light reactivity (m<sub>r</sub>), the curve seen in Equation 10 was developed so that it remains at 0 until ICP approaches 20 mmHg, then drops off sharply towards -1 as ICP approaches 25 mmHg. For both of these modifiers, a 1 represents "maximal" effect, -1 represents "minimal" effect, and 0 represents no effect. So a pupil size modifier of 1 would mean the pupil size is at its maximum possible diameter.

\f[m_s=1/(1+(e^{-2.3\left(ICP-22.5\right)})\f] 
<center>
*Equation 9.*
</center><br>

\f[m_{r} =-.001*10^{.3(ICP-15)} \f] 
<center>
*Equation 10.*
</center><br>

For diffuse injuries, these modifiers are applied to both eyes. For focal injuries, the modifiers are applied to the eye on the same side as the injury, as this reflects the most common observed behavior @cite kingsly2012severe.

@anchor nervous-features-sleep-effects
### Sleep Effects
We adapt the work of @cite schmidt2017state and quantifies the effects of sleep deprivation through evolution of biological debt on the patient. We let biological debt be a function of metabolic demands that change based upon the amount of sleep the patient has had. A normal range of sleep throughout a simulation (6-9 hours a night) will show no noticable physiological signs on the patient. The evolution of the sleep debt equations are characterized by the following equations: 

\f[a(t) = \left( {\frac{{{L_1}}}{{\left( {1 + {e^{ - k({S_r}(t) - s)}}} \right)}}} \right) + {L_0}\f] 
<center>
*Equation 11.*
</center><br>

\f[c(t) = 5.1 - a(t)\sin \left( {\left( {\frac{\pi }{{48}}} \right)t} \right) \f] 
<center>
*Equation 12.*
</center><br>

\f[\frac{{dB}}{{dt}} = {p_w}{r_{wt}} + {p_{b1}}{r_{bt}}B(t) - {r_{bt}}x(t) \f] 
<center>
*Equation 13.*
</center><br>

\f[x(t) = c(t)\left( {\frac{{B(t)}}{{1 + B{{(t)}^2}}}} \right). \f] 
<center>
*Equation 14.*
</center><br>

Here S<sub>r</sub>(t) is the ratio between total sleep and wake times over the simulation period. In general, a patient begins with 8 hours of sleep. The amount of sleep a patient has had over the previous 24 hours can be set in the patient xml. This allows for sleep deprived patients to be loaded into %BioGears without the need to run a simulation for a long period of time. equations 11 and 12 define the circadian rhythm burden on the patient and are scaled as a function of how much sleep the patient has had. equation 13 defines how the biological debt evolves over time. The parameters in the ODE are defined to be: pw = Price of expending energy while awake, rw= rate of energy to waking effort, rb = rates of energy to biological investment, Pb1 = come scaling factor that relates price of using energy while awake.


@anchor nervous-variability
### Patient Variability
The baroreceptor reflex gains and time constants are independent of patient configuration. However, set-points are computed after stabilization, and the baroreceptor reflex drives towards a configuration-specific homeostasis. Because TBI uses multiplicative factors to modify the brain section of the cardiovascular circuit, the patient variability of TBI is the same as that seen in the %Cardiovascular System. There is no patient variability inherent in %BioGears pupillary response model, as it uses modifiers instead of baseline values. A detailed discussion of patient configuration and variability in %BioGears is available in the @ref PatientMethodology report.
@anchor nervous-dependencies
### Dependencies
The baroreceptors interact directly with the %cardiovascular system by modifying %cardiovascular circuit and heart driver parameters. These include heart rate, left and right heart elastance, systemic vascular resistance and systemic vascular compliance. The %cardiovascular parameters are modified by first calculating the normalized change in the parameter. This normalized change is passed into the common data model (CDM) where the %cardiovascular system may extract the change and then apply it locally as a fractional adjustment. The resultant hemodynamic changes feedback directly to the MAP, and the updated MAP is used by the %Nervous system in the next time slice to compute new normalized baroreceptor effects. TBI also directly modifies the cardiovascular circuit, specifically both the upstream and downstream brain resistors. Finally, the %Drugs system can contribute additional pupillary effects, that combine with possible TBI pupil effects via simple summation.

@anchor nervous-assumptions
Assumptions and Limitations
---------------------------
The current implementation of the baroreceptor model does not track adjustments in unstressed volume. Currently %BioGears does not correctly represent the physiologic unstressed volume under resting conditions. Therefore, these changes to the unstressed volume cannot be reflected in the %BioGears engine. This may be addressed in future releases of the engine.

The %BioGears TBI model assumes an acute injury to the brain; chronic effects are not considered, and neither are consciousness assessments. The brain is considered to be one compartment and is not subdivided into discrete functional areas. This means that there is no circuit distinction between a diffuse TBI and a Left Focal TBI. The only difference between the TBI actions is the expected pupil effect. TBI metric validation assumes a supine adult. The current TBI model does not consider specific effects of CO<sub>2</sub> saturation, O<sub>2</sub> saturation, or blood pH on CBF.

@anchor nervous-actions
Actions
-------

### Insults

#### Pain Stimulus
The Pain Stimulus action triggers changes to the cardiovascular, respiratory, and endocrine systems that are associated with sympathetic activation.  Action initiation requires a severity (0-1 scale) and a location.  If multiple locations are specified, the pain severities are summed until the upper boundary of 1 is reached.  The severity is assumed to correspond directly to the Visual Analog Scale (VAS) used to quantify pain in clinical settings (e.g. Pain Severity = 0.5 --> VAS = 5).  To account for the fact that individuals perceive pain differently, a Pain Susceptibility parameter was added to the %BioGears patient definition.  This modifier has a range of -1 to 1, with positive values implying heightened sensitivity to pain and negative values indicating blunted sensitivity.  If not susceptibility is set in the patient file, then a value of 0 is assumed (i.e. "average susceptibility).  The susceptibility parameter is added to the pain severity when a stimulus action is applied.  For example, when exposed to a pain stimulus with a severity of 0.5, two patients with pain susceptibilities of -0.1 and 0.2 would report respective pain levels of 0.4 and 0.7 (4 and 7 on the Visual Analog Scale).

The pain scale calculated from the action severity and patient susceptibility is used to establish changes in physiology.  Heart rate, respiration rate, and blood pressure all increase as the consolidated pain scale increases.  Much of this response is achieved by increasing production of epinephrine in the %BioGears %Endocrine system.  Administration of analgesic drugs (such as Morphine or Fentanyl) lowers the pain score and reverses these physiological trends.   

#### Brain Injury
The Brain Injury action refers to an acute injury to the brain. Whereas in a real patient, a TBI could have manifold effects, both acute and chronic, %BioGears only considers the effects of TBI on brain vascular resistance and pupillary response. The effects of the Brain Injury action depend on the severity value from 0 to 1 and a Type (Diffuse, Left Focal, or Right Focal) assigned to the injury. A severity of 0 will result in a multiplicative factor of 1 being applied to both the upstream and downstream resistors in the brain, which equates to no deviation from the normal, uninjured state. For an injury severity of 1, the upstream resistor is assigned a multiplicative factor of 4.775 and the downstream resistor is assigned a multiplicative factor of 30.409. Severity values between 0 and 1 are converted to multiplicative factors linearly. Increased flow resistance results in increased ICP and decreased CBF. A Diffuse-type injury affects both eyes equally, whereas a Left Focal injury affects only the left eye, and a Right Focal injury affects only the right eye.

@anchor nervous-events
Events
------

### Intracranial Hypertension
Intracranial Hypertension is triggered when the intracranial pressure exceeds 25 mmHg. The event is reversed when the pressure returns below 24 mmHg.

### Intracranial Hypertension
Intracranial Hypotension is triggered when the intracranial pressure drops below 7 mmHg. The event is reversed when the pressure returns above 7.5 mmHg.

@anchor nervous-results
Results and Conclusions
=======================

Validation - Resting Physiologic State
--------------------------------------
No resting state physiology validation was completed, because the baroreceptors do not provide feedback during the initial resting and chronic stabilization phases. For more information on the stabilization phases, see @ref SystemMethodology.

Validation - Actions and Conditions
--------------------
### Baroreceptor Reflex
The baroreceptor reflex is validated through simulation of an acute hemorrhage scenario. This scenario begins with the healthy male patient. After a short time a massive hemorrhage (bleeding rate of ~1000 mL/min) is initiated, and after 30 seconds the bleeding is stopped. The responses for cardiac output, heart rate, systemic vascular resistance and blood volume are shown in Figure 6. The responses shown in the plot are initially driven by decreasing blood volume, resulting in decreased preload and a subsequent reduction in cardiac output. The baroreceptors attempt to maintain the arterial pressure by increasing the heart rate and contractility, increasing the systemic vascular resistance, and reducing vascular compliance. For this specific hemorrhage scenario, there is an expected 30% increase in heart rate, 15-20% decrease in cardiac output, and 10-15% increase in systemic vascular resistance @cite hosomi1979effect. %BioGears baroreceptor feedback does a fair job reaching these values, though it is a bit aggressive in doing so. The heart rate increase observed is much higher than expected, but the changes in cardiac output and systemic vascular resistance are in line with expectations. The validation results are shown in tabular form below. Additional validation related to the baraoreceptor reflex and hemodynamic stability can be found in the @ref CardiovascularMethodology report.

<img src="./plots/Nervous/Baroreceptors_CardiacOutput.jpg" width="1100">
<img src="./plots/Nervous/Baroreceptors_HR.jpg" width="1100">
<img src="./plots/Nervous/Baroreceptors_VascularResistance.jpg" width="1100">
<img src="./plots/Nervous/BaroreceptorsLegend.jpg" width="1100">
<center> *Figure 6. The hemodynamic response to hemorrhage with feedback from the baroreceptor reflex.* </center>

<center><br>
*Table 1. The cardiac output, heart rate, systemic vascular resistance and blood volume are shown as a function of time for the acute hemorrhage scenario. There is a noticeable decrease in blood volume and cardiac output due to the fluid loss. The baroreceptor response in this situation leads to an increase in heart rate and systemic vascular resistance in an attempt to maintain arterial pressure.*
</center>
|	Action	|	Notes	|	Sampled Scenario Time (s)	|	Heart Rate (beats/min)	|	Cardiac Output (mL/min)	|	Systemic Vascular Resistance (mmHg s/mL)	|
|	---	|	---	|	---	|	---	|	---	|	---	|
|	Hemorrhage	|	10% blood loss in 30 s	|	325	|<span class="warning">	Increase ~30% @cite Hosomi1979effect @cite Ottesen2004applied	</span>|<span class="success">	Decrease ~15-20% @cite Hosomi1979effect @cite Ottesen2004applied	</span>|<span class="success">	Increase 10-15% @cite Hosomi1979effect @cite Ottesen2004applied	</span>|

### Chemoreceptor Reflex
The performance of the Chemoreceptor model in %BioGears was validated under both hypercanpic and hypoxic conditions.  Hypercapnic validation was based on the study of Reynolds, Milhorn, and Holloman @cite reynolds1972transient, in which healthy volunteers breathed air with supernormal carbon dioxide levels (3%, 5%, 6%, and 7% CO<sub>2</sub>).  Likewise, hypoxic validation scenarios used data from Reynolds and Milhorn @cite reynolds1973transient.  Volunteers in this study breathed 7%, 8%, and 9% O<sub>2</sub> in nitrogen. 


### Brain Injury
The Brain Injury action is validated through repeated application and removal of increasing severities of TBI. The scenario begins with a healthy male patient. After a short time, a mild brain injury (Severity = 0.2, Type = Diffuse) is applied, and the patient is allowed to stabilize before the injury state is removed (only one TBI action can be in effect at a time, so adding a Diffuse Severity 0 TBI removes all TBI effects). This process is repeated for a more severe injury (Severity = 0.75, Type = Left Focal) and severe (Severity = 1, Type = Right Focal) brain injury. We expect to see increases in ICP, with the most severe injury resulting in an ICP greater than 25 mmHg and decreases in CBF, with the most severe case approaching 8 mL per 100 grams of brain per minute (108 mL per minute for the validated patient) @cite bergeronSME @cite steiner2006monitoring. The scenario shows good agreement for these values, and the patient even dies after some time with the most severe injury. We expect CPP to either increase above its maximum normal value or decrease below its minimum normal value, but, though we see a drop, it isn't quite as pronounced as expected @cite steiner2006monitoring. We can also see that for the low severity injury, ICP doesn't quite reach the threshold to strongly affect the pupils. For the Left Focal injury, only the left pupil is affected, and for the Right Focal injury, only the right pupil is affected.

<table>
<tr>
<td><img src="./plots/Nervous/TBI_ICP.jpg" width="550">
</td>
<td><img src="./plots/Nervous/TBI_CBF.jpg" width="550">
</td>
</tr>
<tr>
<td><img src="./plots/Nervous/TBI_MAP.jpg" width="550">
</td>
<td><img src="./plots/Nervous/TBI_CPP.jpg" width="550">
</td>
</tr>
</table>
<img src="./plots/Nervous/TBILegend.jpg" width="800">
<center>
*Figure 7. Traumatic brain injury response at three different severity levels.*
</center>

<br>
<br>

<center>
<img src="./plots/Nervous/TBI_PupilICP.jpg" width="1200">
<img src="./plots/Nervous/TBI_LeftSize.jpg" width="1200">
<img src="./plots/Nervous/TBI_LeftReactivity.jpg" width="1200">
<img src="./plots/Nervous/TBI_RightSize.jpg" width="1200">
<img src="./plots/Nervous/TBI_RightReactivity.jpg" width="1200">
<img src="./plots/Nervous/PupilLegend.jpg" width="1200">
</center>
<center>
*Figure 8. Pupillary response to the same TBI scenario as shown in Figure 5 where increasing severities are applied first as Diffuse, then as Left Focal, then as Right Focal.*
</center><br>

<center><br>
*Table 2. The validation data for the TBI scenario shows good agreement with expected results.*
</center>
|	Action	|	Notes	|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Intracranial Pressure (mmHg)	|	Cerebral Blood Flow (mL/min)	|	Cerebral Perfusion Pressure (mmHg)	|	Heart Rate (1/min)	|	Respiration Rate (1/min)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Brain Injury	|	Severity 0.2, CPP=MAP-ICP, StandardMale brain mass=1450g	|	20	|	600	|<span class="warning">	10% Increase @cite bergeronSME	</span>|<span class="success">	Decrease @cite steiner2006monitoring	</span>|<span class="success">	Decrease @cite balestreri2006impact	</span>|<span class="warning">	0-10% Decrease @cite bergeronSME	</span>|<span class="warning">	0-10% Decrease @cite bergeronSME	</span>|
|	Brain Injury	|	Severity 0	|	620	|	900	|<span class="success">	7-15 mmHg @cite steiner2006monitoring	</span>|<span class="success">	50-65 mL/100g/min @cite guyton2006medical	</span>|<span class="success">	60-98 mmHg	</span>|<span class="success">	72 @cite guyton2006medical	</span>|<span class="success">	[12.0, 20.0], [13.0, 19.0] @cite silverthorn2013human @cite mantoni2007reduced	</span>|
|	Brain Injury	|	Severity 0.75	|	920	|	1700	|<span class="success">	Increase @cite steiner2006monitoring	</span>|<span class="success">	Decrease @cite steiner2006monitoring	</span>|<span class="success">	Decrease @cite balestreri2006impact	</span>|<span class="warning">	0-15% Decrease @cite bergeronSME	</span>|<span class="warning">	0-20% Decrease @cite bergeronSME	</span>|
|	Brain Injury	|	Severity 0	|	1720	|	2000	|<span class="success">	7-15 mmHg @cite steiner2006monitoring	</span>|<span class="success">	50-65 mL/100g/min @cite guyton2006medical	</span>|<span class="success">	60-98 mmHg	</span>|<span class="success">	72 @cite guyton2006medical	</span>|<span class="success">	[12.0, 20.0], [13.0, 19.0] @cite silverthorn2013human @cite mantoni2007reduced	</span>|
|	Brain Injury	|	Severity 1	|	2020	|	3220	|<span class="success">	>25 mmHg @cite steiner2006monitoring	</span>|<span class="success">	<8 mL/100g/min @cite steiner2006monitoring	</span>|<span class="success">	Decrease @cite balestreri2006impact	</span>|<span class="warning">	Decrease @cite bergeronSME	</span>|<span class="warning">	Decrease @cite bergeronSME	</span>|


###Sleep Deprivation 
The sleep model in BioGears is validated through by running a patient through a sleep deprivation study that quantifies the patients response to meals and measures blood glucose and insulin synthesis @cite spiegel1999impact. We create a study that parallels the methods used in Speigel by creating patients who are restricted in their sleep to just 4 hours a night over a three day period. We measure different physiological markers similar to the ones used in the study, table 3. Vigilance and attention lapse data is validated from @cite mchill2018chronic. The study performed by McHill was similar enough for us to validate all metrics with a single sleep deprivation study.

<img src="./plots/Nervous/SleepBiologicalDebt.jpg" width="1100">
<img src="./plots/Nervous/SleepInsulinConcentration.jpg" width="1100">
<img src="./plots/Nervous/SleepGlucoseConcentration.jpg" width="1100">
<img src="./plots/Nervous/SleepBrainGlucose.jpg" width="1100">
<center> *Figure 9. The metabolic response to sleep deprivation is quantified here through the reduction in insulin synthesis and the increase in blood glucose as the sleep debt increases. Recovery periods can be seen during sleep cycles but becuase the patient isn't receiving enough sleep, the debt is continuing to accrue.* </center>

<center><br>
*Table 3. The validation data for the sleep deprivation study shows good agreement with collected data.*
</center>
	Sleep																			
|	Action	|	Notes	|	Sampled Scenario Time (min)	|	Glucose clearance (%/min)	|	Glucose effectiveness (%/min) clearance without insulin	|	Insulin response (pmol/min) 	|	Stanford sleepyness score (unitless)	|	Cerebral glucose uptake (% from baseline)	|	Reaction Time (ms)	|	Attention Lapses (% from baseline)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Sleep of 4hr per day for 5 days	|	sleep deprivation study taken from spiegle and McHill, given carb rich meal (62%) every 5h 	|	7200	|<span class="success">	40% lower (1.42%/min compared to 2.4%/min @cite spiegel1999impact	</span>|<span class="warning">	Decrease 30% @cite spiegel1999impact	</span>|<span class="success">	Decrease from 432 to 304 pmol/min @cite spiegel1999impact	</span>|<span class="success">	Increase from 2.1 to 4.4  @cite spiegel1999impact	</span>|<span class="success">	Decrease of 7% @cite spiegel1999impact	</span>|<span class="success">	Increase ~1000% @cite mchill2018chronic	</span>|<span class="success">	Increase of 100% @cite mchill2018chronic	</span>|



@anchor nervous-conclusions
Conclusions
-----------
The %Nervous System is currently in a preliminary state that contains only a baroreceptor feedback model, basic TBI, and pupillary response. The baroreceptor feedback is used to control rapid changes in arterial pressure by adjusting heart rate, heart elastance, and vascular resistance and compliance. The baroreceptor model has been validated by comparing the %BioGears outputs to experimental data for hemorrhage. It currently shows good agreement with the expected trends, but the magnitude of the response is not a strong as the validation data. This is due to a large total vascular compliance, which allows for large changes in blood volume with small changes in pressure. Future adjustments to the %cardiovascular circuit would correct the vascular compliance and improve the accuracy of the model. The TBI model shows good agreement for the most prominent TBI metrics, ICP and CBF, for acute brain injuries. Pupillary response behaves as expected and arms %BioGears with yet another tool for matching output with clinical data.

@anchor nervous-future
Future Work
===========
Coming Soon
-----------
- Chemoreceptor modification of heart contractility
- Local autoregulation
- Improved sympathetic and parasympathetic control

Recommended Improvements
------------------------
- Including unstressed volume in the tone model
- Patient variability incorporated into the baroreceptor reflex model constants
- Positional variability
- Higher-fidelity brain model with localized injuries
- Interventions to treat TBI like drainage of cerebrospinal fluid, hypothermic treatment, hyperventilation, and mannitol therapy
- Consciousness model, including reduced level of consciousness with carbon monoxide and other intoxications
- Sleep
- Heart rate and respiration rate effects from TBI

Appendices
==========
Acronyms
--------
CBF - Cerebral Blood Flow

CNS - Central %Nervous System

CPP - Cerebral Perfusion Pressure

ICP - Intracranial Pressure

PERRLA - Pupils Equal, Round, Reactive to Light, and Accommodating

PNS - Peripheral %Nervous System

MAP	- Mean Arterial Pressure

mmHg - Millimeters mercury

Data Model Implementation
-------------------------
@ref NervousSystemTable "Nervous"
Circuit Methodology {#CircuitMethodology}
===================

@anchor circuit-overview
# Overview

## Abstract

The BioGear&reg; Circuit Solver was created to extract the lumped parameter
physics calculations from within the individual systems and centralize the generic calculations.  Closed loop
circuits can be defined easily within the Common Data Model (CDM).  They
are fully dynamic, and can be manipulated and modified at each time
step.  The solver can be used to analyze electrical, fluid, and thermal
circuits using native units.  The solver outputs were validated using the outputs from an existing known third party circuit solver.  All results matched the validation data, confirming this is a sound approach for
the %BioGears Engine. 

@anchor circuit-intro
## Introduction

%BioGears systems use lumped parameter circuits to mimic the physiology
of the human body.  These circuits use fluid or thermal elements that
are analogous to electrical circuit elements.  The
circuits have several types of feedback mechanisms that can be set and
changed at every time step.  Figure&nbsp;1 presents a generic example of very low fidelity lumped
parameter physiology circuits.  Circuits can be thought of as pipe networks for fluid analysis.

@image html CircuitLumpExample.png
<center>
<i>Figure 1.  An example of physiology lumped parameter modeling.  This
example shows very low fidelity models of specific cardiovascular
compartments (left), and a respiratory combined mechanical ventilation
and free breathing model (right) @cite Clipp2012Humansim.</i>
</center><br>

*Note: For simplicity, this document uses electrical components and terminology when discussing the solver functionality.  See Table&nbsp;1 for analogies and details about the mapping of electrical components.*

The %BioGears CDM includes many of the same generic definitions traditionally used to define and analyze circuits.  Paths are ideal conductor branches that may contain elements (i.e., resistors, capacitors, inductors, diodes, etc.).  Nodes are junctions at the intersection of paths.  Figure&nbsp;2 shows these base circuit element definitions.  Paths are each assigned one source and one target node.  We use the convention of positive current from source to target when performing calculations.

@image html CircuitBaseDefinitions.png
<center>
<i>Figure 2. Nodes and paths are the lowest level elements used to define all circuits in %BioGears.  Paths correspond to ideal conductors (i.e., wires).  Nodes are placed at the intersections of paths.  In fluid systems, paths can be thought of as frictionless pipes and nodes as pipe junctions.</i>
</center><br>

@anchor circuit-design
# System Design

## Background and Scope

### Legacy Solution

The physiology engine from which %BioGears was spawned (released by
Advanced Simulation Corporation as Body Simulation for Anesthesia&trade;)
solved each lump parameter circuit using non-closed loop calculations
contained within each system.  The systems were extensively
interconnected and the circuit calculations were unable to be decoupled.
This made it extremely difficult to find sources of errors and inconsistencies in output results.

The circuit math was performed often using different approaches, with
mixed results.  It became apparent that neither energy nor mass was
being truly conserved.  This was true within the individual systems, as
well as within the entire body.  One-way fluid valves in
the %Cardiovascular System (similar to electrical diodes) would
inadvertently allow backflow for a single time step when flows changed
direction.  Little validation was done to provide a sound
fundamental base on which to build physiological systems.

### Requirements

Given the overall goals of %BioGears, we set out to create a generic
circuit solver on which to expand.  The solver is required to efficiently
solve transient linear and non-linear circuits.  Some other high-level
requirements include:

-   Generic - All systems should be able to use the same basic circuit
    physics engine.  This allows rapid development, and makes engine
    outputs much easier to validate and verify.

-   Computational Speed - %BioGears is required to maintain a transient
    full-body solution faster than real time on typical personal
    computers.

-   Modular - Using the same basis for design and construction will aid
    in keeping the system decoupled.

-   Extensible - We must take future growth into consideration and allow
    users and developers the proper tools and building blocks on which
    to add new functionality.

-   Dynamic - Feedback mechanisms are required for each system.  It is
    beneficial to be able to dynamically change, add, or remove
    circuits, sub-circuits, elements, and connections at run time.

-   Conservation - We must uphold sound scientific principles and conserve energy, mass, and momentum.

-   Common Data Model - The entire solution must reside within and
    effectively use the @ref CDM.

-   Circuits types - %BioGears will include fluid and thermal
    circuits.  It is beneficial to use the same solver for both types of
    systems.

-   Elements - The solver must be able to handle all base circuit
    elements (i.e., resistors, capacitors, inductors, etc.), with room
    for expansion.

-   Open source - As a whole, any third-party software must adhere to the same
    license requirements as %BioGears.

### Solutions Investigated

We considered several circuit solver options, including investigating existing
software to determine if any would meet our needs.  The best external
code we found for integration into our C++ open source code was a
variation on SPICE (Berkeley University&rsquo;s Spice3f5), called Ngspice
(http://ngspice.sourceforge.net/).  Ngspice is an open source, general
purpose circuit simulation program for nonlinear and linear analyses.  It
is developed with a C++ wrapper that maintains the original parametric
netlist format (netlists can contain parameters and expressions)
as inputs.

During implementation, we found Ngspice to be a good fit for our
needs to solve closed-loop circuits, yet several negatives forced us to
deviate from that path.  We discovered a small number of bugs, some with
possible workarounds.  It was found to be very slow for our approach of
solving a new (i.e., dynamic) circuit each time step.  It was cumbersome
to parse using the netlist input and output style.  The most significant
downside against implementing Ngspice into %BioGears was the
unavoidable build-up of memory that caused our software to crash.

Due to the aforementioned complications, we ultimately determined that
creating our own generalized circuit analysis code was ideal and less
labor-intensive than implementing existing software.  This has the added
benefit of allowing us to control every aspect of the analysis.

@anchor circuit-data
## Data Flow

Our generic Circuit Solver intentionally mirrors the same data flow
found in each of the systems.  Each time step consists of a Preprocess
call to set up the circuits for analysis, a Process call to do the
analysis, and a Post Process call to advance the time in preparation for the
next time step.

### Preprocess

There is no literal Preprocess call within the circuit solver class (i.e., code).  Each system individually and directly modifies its circuit(s).  This allows our functionality to be entirely dynamic by performing a separate calculation every time step.  See the Data Model Implementation Appendix below for details about %BioGears circuit elements and the data they contain.

### Process

For the circuit-solving software to effectively calculate the nonlinear
circuit states without human intervention, a robust, generic, and
repeatable methodology is required.  The solver performs numerical
integration through linearization (first order approximations) by
assuming a direct current (DC) solution for the given time step.  The Modified Nodal
Analysis (MNA) approach is used to determine the state of every node and path
within the circuit.  The steps for solving a circuit in a given time step
are (also see Figure&nbsp;3):

1.  Perform numerical integration by using linearization (first order
    approximations) through MNA, using the matrix
    algebra equation:

	<center>
    \f[Ax = b\f]
	</center>
	
	<center>
	*Equation 1.*
	</center><br>    

    Where *A* is the matrix of constants, *x* is the vector of unknowns/variables, and *b* is the right side vector of knowns.

    1.  Use Kirkoff&rsquo;s Current Laws (KCL) (sum of the currents is zero at
        each node) to populate the *A* matrix and *b* vector.  Table&nbsp;2 shows the equations used for determining flows, where flows (*F*) are equivalent to currents, and pressures (*P*) are equivalent to voltages.

    2.  Leverage the Eigen templated library (released by Tuxfamily) LU
        decomposition linear solver (FullPivLU) to solve for unknown
        variables (i.e., voltages and currents).  This decomposition
        provides the generic approach to solving systems of linear
        equations​, and represents an LU decomposition of any matrix
        with complete pivoting.  The matrix *A* is decomposed as:

		\f[A=P^{-1} LUQ^{-1} \f] 
		<center>
		*Equation 2.*
		</center><br> 

		Where *L* is unit-lower-triangular, *U* is upper-triangular, and *P* and *Q* are
		permutation matrices.  This is a rank-revealing LU decomposition.  The
		eigenvalues (diagonal coefficients) of *U* are sorted in such a way that
		any zeros are at the end.

2.  Calculate remaining unknown currents using the Trapezoidal Rule,
    where applicable.  For nonlinear elements from time *a* to *b*, this is:

	\f[\int _{a}^{b}f(x)dx\approx (b-a)\left[\frac{f(a)+f(b)}{2} \right] \f] 

	<center>
	*Equation 3.*
	</center><br>

3.  Calculate diode currents using assumed open or closed switch states (cannot be solved directly).
    Iterate Steps 1&ndash;3 until a satisfactory solution is attained.

4.  Calculate the change in charge (*Q*) across capacitors based on the capacitance (*C*) and voltage (*V*) change, and increment the
    total charge.  The charge is conserved on the source and target nodes by incrementing/decrementing this amount.  Selecting which node gains charge and which loses is done by the direction of the current in the path.
	
	\f[Q(t)=C*V(t)\f] 

	<center>
	*Equation 4.*
	</center><br>

5.  Invalidate the current on any path the user has specified to ignore.
    Note that this is to prevent unwanted transport to the reference
    node (i.e., ground).

<center><img src="./images/Circuit/CircuitDataFlow.png" width="400"></center>
<center>
*Figure 3.  Data flow chart showing the steps used at each time step to
determine each circuit state.*
</center><br>

There are several nuances for the handling of certain elements:

-   Paths without elements are calculated as if they are zero voltage
    sources, which allows for direct solving for the current.

-   Switches are modeled as element-less paths when closed and
    essentially infinite resistances (10<sup>100</sup> ohms) when open.

-   Valves (i.e., Diodes) are modeled as closed switches when the path source node voltage is higher than the target node (on) and as open switches when the voltage is greater on the target node (off).

-   Polarized elements are modeled as an open switch when preventing polarity reversal

Because diode and polarized element states cannot be determined
before solving a circuit, we are forced to make a considered guess.
Therefore, at each time step, we begin with an assumed on/off state for
each of these elements in the circuit.  We begin with the previous time step solution
to minimize the number of iterations.  The entire circuit is solved using
this assumed state, and the currents and voltages associated with the
paths are considered.  When a diode/polarized element is found to have an incorrect
current direction or voltage reversed relative to its assumed state, the
state is flipped.  This process is then repeated until the valid solution
is found.  Intelligently choosing which diode states to switch is
important because there are 2^(Number of Diodes/Polarized Elements) possible combinations in each circuit.

### Post Process

Post Process effectively advances time by moving the &ldquo;next&rdquo; time step
parameter values to the &ldquo;current&rdquo; time step parameter values (see @ref SystemMethodology).

@anchor circuit-features
## Features and Capabilities

### Approach and Extensibility

The %BioGears systems use two different lumped parameter system types,
each with analogies to electrical circuits.  Most systems models (%Cardiovascular, %Respiratory, etc.) use fluid units (pressures, flows,
and volumes), while others (%Energy, %Environment, etc.) use thermal units
(temperature, heat transfer rate, and heat).  Both model types use the same
underlying physics to define relationships and solve for unknown values.
We included the ability to calculate in native electrical
units (voltage, current, and charge) to help with validation using known
electrical circuit models and outputs.  Table&nbsp;1 lists system analogies and
how they map to circuit elements.

<center><br>
<i>Table 1.  The various types of circuits that can be solved by the
Circuit Solver.  Some have units already defined and included in the CDM,
while the others can be easily added.  Possible inputs, outputs, elements,
and parameters are all defined.  Valves (diodes) and
switches are not shown @cite riggs1976control.</i>
</center>
@image html CircuitSystemAnalogies.png
<br>

While we designed the generic Circuit Solver to analyze our fluid,
thermal, and electrical models, we can easily extend it to include any
of the model types in Table&nbsp;1.  Further details specific to the
implementation of our model with the hydraulic analogy are shown in Table&nbsp;2.  
A more intuitive pipe analogy is described through images.  The
%BioGears defined fluid model elements are outlined in the first column.
The flow equations are important for our analysis technique outlined earlier.

<center><br>
<i>Table 2.  In-depth description of the hydraulic analogy for electrical
circuits that are used extensively inside %BioGears.  The Elements are
defined by the CDM and used by the solver.  The Flow equations are
important for solving for the unknown parameters.  @cite HydraulicAnalogy2014 </i>
</center>
@image html CircuitHydraulicAnalogyTable.png
<br>

The Circuit Solver is equipped to use all of the element types given in the second column of Table&nbsp;2.  All three passive element types (resistor, capacitor, and inductor) have a polarized element modeling option.  When active, polarized elements will short the circuit when the target node voltage becomes greater than that of the source node.  This allows the user to model electrolytic capacitors and further ensures fluid will not be added to hydraulic systems if compliances switch polarity.

### Data Model Implementation

Our mathematical approach to solving circuits is relatively straightforward, but our CDM implementation and integration with physiological
models is novel.  We implemented the Circuit Solver to use generic terms
that are not specific to any one model type (see Table&nbsp;1).  Conversions to base units for each model are done in the
background using the CDM unit conversion functionality.  These base units
are selected to prevent unnecessary conversion that would use critical
computation resources, while still maintaining a direct mathematical
relationship on both sides of the equation when performing the
calculations.

The CDM allows us to implement the circuit math in a completely modular
fashion.  Every system uses the same physics, with the same CDM-defined
elements.  The burden on system developers is merely in setting up and
manipulating the circuit correctly.  A benefit to using a CDM is a significantly reduced
development time.

The only significant requirement for circuit design is that it must be closed-loop.  Beyond that, the solver can handle any combination of valid nodes,
paths, and elements.  Only one (or zero) element is allowed on each path
and the convention of positive current going from source node to target
node is used.

This sound foundation for defining and calculating circuit parameters, allows
the engine to transport substances in a similarly generic fashion (see @ref DrugsMethodology).

@anchor circuit-assumptions
## Assumptions

### Circuits

There are several assumptions made for the circuit math in general:

-   All circuits must be closed-loop

-   All elements are idealized

-   All circuits must have a reference node defined with a known voltage - unlike other circuit solvers, it does not necessarily have to have a potential value of zero

-   Nonlinear sources need initial values.  Therefore, paths with
    capacitors that do not have an initial voltage on the source or
    target nodes will assign it the reference node voltage value, and
    paths with inductors that do not have an initial current will be
    assigned a value of zero.

### Fluids

There are other assumptions specific to the %BioGears fluid systems:

-   Flows are incompressible

-   Fluids are inviscid

## Limitations

There are frequency response limitations that are directly tied to the
time step set externally to the solver itself.  %BioGears uses a time step
of 1/50 s (~ 2 ms).  Because the solver calculates nonlinear
behavior through linearization using the trapezoid rule, extremely high
frequency signal components will be lost.  However, this far exceeds the Nyquist frequency for all current %BioGears system models.

While it is not necessarily a limitation, users of the Circuit Solver
must be careful to assign current and voltage source elements to set currents and voltages respectively.
Attempting to directly set node voltages or path currents would result
in an over-constrained solution.  In these instances, the solver will
overwrite those values.

@anchor circuit-results
# Results and Conclusions

## Validation

Beyond several numerical (hand) calculations used during implementation, we
performed validation on the generalized circuit math by duplicating
circuits in LTspice (version 4.21s, created by Linear Technology,
Milpitas, CA).  LTspice is based on the well-documented and validated
SPICE simulator.  We created circuits using all %BioGears elements
individually and in combination.  We used several different types of
dynamically-changing drivers to ensure proper transient functionality.
The resulting voltage and current values were interpolated and validated
to match for all 114 circuits.  Table&nbsp;3 shows a summary of the validation
circuits investigated.

<center><br>
*Table 3.  The list of circuits created in %BioGears and validated against LTspice.  Every element is covered in combination with each other.*
</center>

|	Test Name	|	Purpose of test	|	Results Summary	|
|	------------------------	|	------------------------	|	------------------------	|
|	BadDiodeDC	|	Intentionally failing to test Diode implementation	|	<span class="success">LTSpice: Will not run, %BioGears: successfully fails - provides feedback on failure</span>	|
|	BadDiodePULSE	|	Intentionally failing to test Diode implementation	|	<span class="success">LTSpice: Will not run, %BioGears: successfully fails - provides feedback on failure</span>	|
|	BadDiodeSIN	|	Intentionally failing to test Diode implementation	|	<span class="success">LTSpice: Will not run, %BioGears: successfully fails - provides feedback on failure</span>	|
|	BasicCircuit	|	Simple pressure source with resistances in series	|	<span class="success">Match</span>	|
|	BasicDiodeDCCurrent	|	Diode behavior in response to a current source	|	<span class="success">Match</span>	|
|	BasicDiodePULSECurrent	|	Diode behavior in response to a current source	|	<span class="success">Match</span>	|
|	BasicDiodeSINCurrent	|	Diode behavior in response to a current source	|	<span class="success">Match</span>	|
|	Comprehensive1DC	|	Combination of element types with a voltage source and a current source	|	<span class="success">Match</span>	|
|	Comprehensive1PULSE	|	Combination of element types with a voltage source and a current source	|	<span class="success">Match</span>	|
|	Comprehensive1SIN	|	Combination of element types with a voltage source and a current source	|	<span class="success">Match</span>	|
|	CurrentCompDC	|	Combination of resistors and diode in response to a current source	|	<span class="success">Match</span>	|
|	CurrentCompPulse	|	Combination of resistors and diode in response to a current source	|	<span class="success">Match</span>	|
|	CurrentCompSIN	|	Combination of resistors and diode in response to a current source	|	<span class="success">Match</span>	|
|	ParallelCapDC	|	Adding capacitors in parallel with a voltage source	|	<span class="success">Match</span>	|
|	ParallelCapDCCurrent	|	Adding capacitors in parallel with a current source	|	<span class="success">Match</span>	|
|	ParallelCapPulse	|	Adding capacitors in parallel with a voltage source	|	<span class="success">Match</span>	|
|	ParallelCapPULSECurrent	|	Adding capacitors in parallel with a current source	|	<span class="success">Match</span>	|
|	ParallelCapSIN	|	Adding capacitors in parallel with a voltage source	|	<span class="success">Match</span>	|
|	ParallelCapSINCurrent	|	Adding capacitors in parallel with a current source	|	<span class="success">Match</span>	|
|	ParallelCurrentSourceAdditionDC	|	Adding current sources in parallel	|	<span class="success">Match</span>	|
|	ParallelCurrentSourceAdditionPULSE	|	Adding current sources in parallel	|	<span class="success">Match</span>	|
|	ParallelCurrentSourceAdditionSIN	|	Adding current sources in parallel	|	<span class="success">Match</span>	|
|	ParallelIndDC	|	Adding inductors in parallel with a voltage source	|	<span class="success">Match</span>	|
|	ParallelIndDCCurrent	|	Adding inductors in parallel with a current source	|	<span class="success">Match</span>	|
|	ParallelIndPULSE	|	Adding inductors in parallel with a voltage source	|	<span class="success">Match</span>	|
|	ParallelIndPULSECurrent	|	Adding inductors in parallel with a current source	|	<span class="success">Match</span>	|
|	ParallelIndSIN	|	Adding inductors in parallel with a voltage source	|	<span class="success">Match</span>	|
|	ParallelIndSINCurrent	|	Adding inductors in parallel with a current source	|	<span class="success">Match</span>	|
|	ParallelPressureSourceAdditionDC	|	Intentionally failing adding voltage sources in parallel	|	<span class="success">%BioGears: successfully fails - provides feedback on failure</span>	|
|	ParallelPressureSourceAdditionPULSE	|	Intentionally failing adding voltage sources in parallel	|	<span class="success">%BioGears: successfully fails - provides feedback on failure</span>	|
|	ParallelPressureSourceAdditionSIN	|	Intentionally failing adding voltage sources in parallel	|	<span class="success">%BioGears: successfully fails - provides feedback on failure</span>	|
|	ParallelRCDC	|	Parallel resistor and capacitor response to a voltage source	|	<span class="success">Match</span>	|
|	ParallelRCPULSE	|	Parallel resistor and capacitor response to a voltage source	|	<span class="success">Match</span>	|
|	ParallelRCSIN	|	Parallel resistor and capacitor response to a voltage source	|	<span class="success">Match</span>	|
|	ParallelRDC	|	Parallel resistor response to a voltage source	|	<span class="success">Match</span>	|
|	ParallelRDCCurrent	|	Parallel resistor response to a current source	|	<span class="success">Match</span>	|
|	ParallelRLCDC	|	Parallel resistor, inductor, and capacitor response to a voltage source	|	<span class="success">Match</span>	|
|	ParallelRLCDCCurrent	|	Parallel resistor, inductor and capacitor response to a current source	|	<span class="success">Match</span>	|
|	ParallelRLCPULSE	|	Parallel resistor, inductor, and capacitor response to a voltage source	|	<span class="success">Match</span>	|
|	ParallelRLCPULSECurrent	|	Parallel resistor, inductor and capacitor response to a current source	|	<span class="success">Match</span>	|
|	ParallelRLCSIN	|	Parallel resistor, inductor, and capacitor response to a voltage source	|	<span class="success">Match</span>	|
|	ParallelRLCSINCurrent	|	Parallel resistor, inductor and capacitor response to a current source	|	<span class="success">Match</span>	|
|	ParallelRLDC	|	Parallel resistor and inductor response to a voltage source	|	<span class="success">Match</span>	|
|	ParallelRLDCCurrent	|	Parallel resistor and inductor response to a current source	|	<span class="success">Match</span>	|
|	ParallelRLPULSE	|	Parallel resistor and inductor response to a voltage source	|	<span class="success">Match</span>	|
|	ParallelRLPULSECurrent	|	Parallel resistor and inductor response to a current source	|	<span class="success">Match</span>	|
|	ParallelRLSIN	|	Parallel resistor and inductor response to a voltage source	|	<span class="success">Match</span>	|
|	ParallelRLSINCentered	|	Parallel resistor and inductor response to a voltage source	|	<span class="success">Match</span>	|
|	ParallelRLSINCurrent	|	Parallel resistor and inductor response to a current source	|	<span class="success">Match</span>	|
|	ParallelRPULSE	|	Parallel resistor response to a voltage source	|	<span class="success">Match</span>	|
|	ParallelRPulseCurrent	|	Parallel resistor response to a current source	|	<span class="success">Match</span>	|
|	ParallelRSIN	|	Parallel resistor response to a voltage source	|	<span class="success">Match</span>	|
|	ParallelRSINCurrent	|	Parallel resistor response to a current source	|	<span class="success">Match</span>	|
|	SeriesCapDC	|	Adding capacitors in series with a voltage source	|	<span class="success">Match</span>	|
|	SeriesCapDCCurrent	|	Adding capacitors in series with a current source	|	<span class="success">Match</span>	|
|	SeriesCapPulse	|	Adding capacitors in series with a voltage source	|	<span class="success">Match</span>	|
|	SeriesCapPULSECurrent	|	Adding capacitors in series with a current source	|	<span class="success">Match</span>	|
|	SeriesCapSIN	|	Adding capacitors in series with a voltage source	|	<span class="success">Match</span>	|
|	SeriesCapSINCurrent	|	Adding capacitors in series with a current source	|	<span class="success">Match</span>	|
|	SeriesCurrentSourceAdditionDC	|	Intentionally failing adding current sources in series	|	<span class="success">%BioGears: successfully fails - provides feedback on failure</span>	|
|	SeriesCurrentSourceAdditionPULSE	|	Intentionally failing adding current sources in series	|	<span class="success">%BioGears: successfully fails - provides feedback on failure</span>	|
|	SeriesCurrentSourceAdditionSIN	|	Intentionally failing adding current sources in series	|	<span class="success">%BioGears: successfully fails - provides feedback on failure</span>	|
|	SeriesIndDC	|	Adding inductors in series with a voltage source	|	<span class="success">Match</span>	|
|	SeriesIndDCCurrent	|	Adding inductors in series with a current source	|	<span class="success">Match</span>	|
|	SeriesIndPULSE	|	Adding inductors in series with a voltage source	|	<span class="success">Match</span>	|
|	SeriesIndPULSECurrent	|	Adding inductors in series with a current source	|	<span class="success">Match</span>	|
|	SeriesIndSIN	|	Adding inductors in series with a voltage source	|	<span class="success">Match</span>	|
|	SeriesIndSINCurrent	|	Adding inductors in series with a current source	|	<span class="success">Match</span>	|
|	SeriesPressureSourceAdditionDC	|	Adding voltage sources in series	|	<span class="success">Match</span>	|
|	SeriesPressureSourceAdditionPULSE	|	Adding voltage sources in series	|	<span class="success">Match</span>	|
|	SeriesPressureSourceAdditionSIN	|	Adding voltage sources in series	|	<span class="success">Match</span>	|
|	SeriesRCDC	|	Series resistor and capacitor response to a voltage source	|	<span class="success">Match</span>	|
|	SeriesRCPULSE	|	Series resistor and capacitor response to a voltage source	|	<span class="success">Match</span>	|
|	SeriesRCSIN	|	Series resistor and capacitor response to a voltage source	|	<span class="success">Match</span>	|
|	SeriesRDC	|	Series resistors response to a voltage source	|	<span class="success">Match</span>	|
|	SeriesRLCDC	|	Series resistor, inductor and capacitor response to a voltage source	|	<span class="success">Match</span>	|
|	SeriesRLCDCCurrent	|	Series resistor, inductor and capacitor response to a current source	|	<span class="success">Match</span>	|
|	SeriesRLCPULSE	|	Series resistor, inductor and capacitor response to a voltage source	|	<span class="success">Match</span>	|
|	SeriesRLCPULSECurrent	|	Series resistor, inductor and capacitor response to a current source	|	<span class="success">Match</span>	|
|	SeriesRLCSIN	|	Series resistor, inductor and capacitor response to a voltage source	|	<span class="success">Match</span>	|
|	SeriesRLCSINCurrent	|	Series resistor, inductor and capacitor response to a current source	|	<span class="success">Match</span>	|
|	SeriesRLDC	|	Series resistor and inductor response to a voltage source	|	<span class="success">Match</span>	|
|	SeriesRLDCCurrent	|	Series resistor and inductor response to a current source	|	<span class="success">Match</span>	|
|	SeriesRLPULSE	|	Series resistor and inductor response to a voltage source	|	<span class="success">Match</span>	|
|	SeriesRLPULSECurrent	|	Series resistor and inductor response to a current source	|	<span class="success">Match</span>	|
|	SeriesRLSIN	|	Series resistor and inductor response to a voltage source	|	<span class="success">Match</span>	|
|	SeriesRLSINCurrent	|	Series resistor and inductor response to a current source	|	<span class="success">Match</span>	|
|	SeriesRPULSE	|	Series resistors response to a voltage source	|	<span class="success">Match</span>	|
|	SeriesRSIN	|	Series resistors response to a voltage source	|	<span class="success">Match</span>	|
|	SimpleDiodeDC	|	Diode behavior in response to a voltage source	|	<span class="success">Match</span>	|
|	SimpleDiodePULSE	|	Diode behavior in response to a voltage source	|	<span class="success">Match</span>	|
|	SimpleDiodeSIN	|	Diode behavior in response to a voltage source	|	<span class="success">Match</span>	|
|	SwitchRCDC	|	Circuit behavior when a switch is thrown with a voltage source	|	<span class="success">Match</span>	|
|	SwitchRCDCCurrent	|	Circuit behavior when a switch is thrown with a current source	|	<span class="success">Match</span>	|
|	SwitchRCPULSE	|	Circuit behavior when a switch is thrown with a voltage source	|	<span class="success">Match</span>	|
|	SwitchRCPULSECurrent	|	Circuit behavior when a switch is thrown with a current source	|	<span class="success">Match</span>	|
|	SwitchRCSIN	|	Circuit behavior when a switch is thrown with a voltage source	|	<span class="success">Match</span>	|
|	SwitchRCSINCurrent	|	Circuit behavior when a switch is thrown with a current source	|	<span class="success">Match</span>	|

<br>
The most interesting and complex circuits use a combination of all
elements.  Two of these comprehensive circuits are shown below because of
their complexity.  The first uses sine wave sources and the second uses
pulse sources.  There is a strong correlation between the given LTspice
outputs and the calculated %BioGears outputs.  Note that the sign
convention of the current across voltage sources is reversed for LTspice,
because it does not maintain the source-to-target positive current
standard as is done with %BioGears.

@image html Comprehensive2CircuitDiagram.png
<center>
*Figure 4.  The LTspice circuit diagram for the first comprehensive
circuit that exemplifies the validation completed on the solver.  Many
different types of elements and multiple voltage and current sources are
included.  The same circuit was defined using the CDM for comparison.*
</center><br>

@image html ValidationComprehensive2SINCenteredPressure.png
<center>
*Figure 5.  %BioGears node voltage outputs for the first comprehensive
circuit compared to LTspice baseline values, using sinusoid sources.  All
are very tightly correlated.*
</center><br>

@image html ValidationComprehensive2SINCenteredFlow.png
<center>
*Figure 6.  %BioGears path current outputs for the first comprehensive
circuit compared to LTspice baseline values.  All are very tightly
correlated.  The signs of the current through voltage sources are reversed
because of differing conventions for those elements.*
</center><br>

@image html Comprehensive1CircuitDiagram.png
<center>
*Figure 7.  The LTspice circuit diagram for the second comprehensive
circuit that exemplifies the validation completed on the solver.  Many
different types of elements and multiple voltage and current sources are
included.  The same circuit was defined using the CDM for comparison.*
</center><br>

@image html ValidationComprehensive1PulsePressure.png
<center>
*Figure 8.  %BioGears node voltage outputs for the second comprehensive
circuit compared to LTspice baseline values, using pulse sources.  All
are very tightly correlated.*
</center><br>

@image html ValidationComprehensive1PulseFlow.png
<center>
*Figure 9.  %BioGears path current outputs for the second comprehensive
circuit compared to LTspice baseline values.  All are very tightly
correlated.  The sign of the current through voltage sources are reversed
because of differing conventions for those elements.*
</center><br>

%BioGears has been shown to successfully conserve mass, energy, and momentum within all defined closed-loop systems. The successful conservation of mass provided by the solver is shown in Figure 10. The volume (quantity/charge) within cardiovascular circuit nodes through approximately 2.5 full heart beat cycles.  The total volume of all compartments remains at a constant value of 5L throughout the entire process.

@image html CardiovascularCompartmentVolumes.png
<center>
*Figure 10.  The blood volume within each compartment continuously varies in the cardiovascular circuit.   The sum of these volumes does not change at any time step, and system mass is successfully conserved.*
</center><br>

All basic Circuit Solver functionality is further validated and verified with specific unit tests that target individual methods.  The following functionality has been successfully validated by individual tests:
-   Incorrect usage errors
-   Combined circuits
-   Dynamically-changing elements
-   Dynamically-changing circuits
-   Added and removed elements
-   Fluid, thermal, and electrical model types
-   Non-zero reference node voltage
-   Polarized elements
-   Dynamically modifying capacitor charge

@anchor circuit-conclusion
## Conclusion

The %BioGears generalized circuit analysis techniques have successfully
accomplished all of the goals for that area of the code.  The solver
implements the CDM effectively and generically in a way that allows for
infinitely complex lumped parameter models that can also be combined.  We
made it simple to modify existing functionality and add new models.
This has saved, and will continue to save, significant amounts of
development time.  The generic and modular design also allows for much
easier bug finding and fixing, system parameter tuning, and system
validation.  The burden for the modeler is on creating a good model and not
worrying about implementing sound physics math.

The Circuit Solver has also been designed such that it can be extracted
from %BioGears and used for any software that needs to perform transient
circuit analysis.

@anchor circuit-future
# Future Work

## Coming Soon

The list below includes some of the planned functionality additions.

-   Functionality to create port connections between circuits for improved modularity between systems/subsystems

## Recommended Improvements

Other functionality that may be beneficial includes:

-   A sensitivity analysis and error term determination

-   An analysis of frequency constraints

-   More circuit elements (i.e., amplifiers, transformers, transistors,
    non-ideal elements, zener diodes, etc.)

@anchor circuit-appendices
# Appendices

## Data Model Implementation

SECircuit

SECircuitCalculator

SECircuitNode

SECircuitPath
@anchor circuit-acronyms
## Acronyms

CDM - Common Data Model

DC - Direct Current

GUI - Graphical User Interface

KCL - Kirkoff's Current Law

MNA - Modified Nodal Analysis

SPICE - Simulation Program with Integrated Circuit Emphasis


%Tissue Methodology {#TissueMethodology}
==========================

Overview
========

Abstract
--------
This %BioGears %Tissue system manages the extravascular space. It handles substance transport between the organs and the blood vessels, and it computes substance storage, transformation (e.g. chemical conversion), clearance, and excretion. 

Introduction
------------
The %BioGears %Tissue system is a low-resolution, mid-fidelity model of the tissues of the body. The system is mechanically tied to the [Cardiovascular](@ref CardiovascularMethodology) and [Respiratory](@ref RespiratoryMethodology) systems, and it interacts with the [Energy](@ref EnergyMethodology) and [Drugs](@ref DrugsMethodology) systems. The tissue system handles the non-advective transport of substances between the intravascular and extravascular spaces, as well as the conversion of substance (including chemical conversion of species and clearance/excretion). The metabolic production and consumption of substances takes place in the tissue system, and the tissues generate substances that are produced in the organs by any process or mode.

System Design
=============

Background and Scope
--------------------
Groups of cells in the body that share a common embryonic origin can be described collectively as a tissue. In the classical organizational hierarchy of organisms, tissues are at the level directly below organs, meaning that groups of tissues interacting to perform a function are an organ. There are four types of tissue in the human body: epithelial, connective, muscle, and nervous tissue. In %BioGears, the term tissue refers to the extravascular space of an organ. In other words, in %BioGears 'tissue' is a collective term that generally refers to the parenchyma.

@anchor tissue-data-flow
Data Flow
---------
Like the other %BioGears systems, the %Tissue system uses the execution structure described in @ref SystemMethodology. Figure 1 shows the data flow.

<img src="./images/Tissue/TissueDataFlow.png" width="900">
<center>
<i>Figure 1. The %Tissue data flow consists of a Preprocess, Process, and Post Process loop. Conditions are applied during initialization prior to executing the loop. Most of the tissue functionality is executed during the process step, including all modes of mass transport and conversion. Time is advanced in the postprocess step.</i>
</center><br>

### Preprocess
At this time the there are no PreProcess steps in the %Tissue System. 

### Process

#### Calculate Metabolic Consumption and Production
Conversions of nutrients to metabolic energy and byproducts are calculated for each relevant compartment.

#### Calculate Pulmonary Capillary Substance Transport
Gases are transferred from the lungs (alveoli) to the pulmonary capillaries and vice versa during this calculation. This allows for the transport of oxygen into the cardiovascular system from the ambient air, providing the required substances for metabolism. By the same process, carbon dioxide waste is removed from the %Cardiovascular System and moves through the %Respiratory System into the ambient air. 

#### Calculate Diffusion
%Substances move from the vascular space into and out of the extravascular or tissue space for metabolism, waste removal, and/ or clearance. This functionality moves gases across the membrane between the vascular and extravascular spaces using one or more of several diffusion models, discussed below.
- Perfusion Limited Diffusion
- Instant Diffusion
- Simple Diffusion
- Facilitated Diffusion
- Active Transport (Currently for sodium, potassium, and chloride)

#### Calculate Vital Signs
In this method the tissue volumes are summed in order to compute total body water. Body system level data is also set in this method. In the future, this method will compute and set tissue substance concentrations and trigger concentration-based events.

#### Protein Storage and Release
In this method, amino acid is stored or released from storage in muscle as dictated by the local hormone factor.

#### Fat Storage and Release
In this method, triacylglycerol is stored or released from storage in fat as dictated by the local hormone factor.

### Post Process
At this time the there are no PostProcess steps in the %Tissue System. 

### Assessments
Assessments in %BioGears are data collected and packaged to resemble a report or analysis that might be ordered by a physician. No %BioGears assessments are associated with the %Tissue system.

Features, Capabilities, and Dependencies
----------------------------------------
The %BioGears %Tissue system is a low-resolution, mid-fidelity model of the tissues of the body. One of the primary functions of the %BioGears %Tissue system is to control the transport of substances between the tissues and the blood. There are several transport models that help the %Tissue system perform that function. Figure 2 provides an overview of the extravascular space and the various modes of substance transport between the blood and the tissues. The %BioGears %Tissue system also handles the conversion of substance (i.e. metabolic consumption and production).

@anchor tissue-fig2
<img src="./images/Tissue/tissueTransport.png" width="500">
<center>
<i>Figure 2. The tissue compartment is partitioned into two distinct fluid spaces, and the non-fluid volume and mass are accounted for. There are several modes of transport between the spaces.</i>
</center><br>

### Bulk Flow and Advection
The movement of fluid between the intravascular and extravascular space is modeled using the %BioGears @ref CircuitMethodology.  In most cases, each tissue circuit node is connected to one and only one cardiovascular circuit node. However, the gut tissue compartment is a lumped representation of the abdominal viscera organ tissues, and thus the large intestine, small intestine, and splanchnic vascular circuit nodes all connect to the gut tissue circuit node.  Each tissue sub-circuit, a representative sample of which is shown in [Figure 3](@ref tissue-fig3), is modeled after the Starling Equation (Equation 1 below). 

\f[{J_v} = {K_f}\left( {\left[ {{P_c} - {P_t}} \right] - \sigma \left[ {{\pi _c} - {\pi _t}} \right]} \right)\f]
<center>
*Equation 1*
</center><br>

P<sub>c</sub> and P<sub>t</sub> are the hydrostatic pressures stored on the vascular capillary and extracellular tissue nodes.  The parameters K<sub>F</sub> and &sigma; are the filtration and reflection coefficients, respectively.  The filtration coefficient defines the tendency of fluid to pass into the extravasculature, and its inverse has units of resistance (hence the resistor element in Figure 3).  The resistance of each tissue is assumed to be proportionate to its mass.  The reflection coefficient describes the effectiveness of a membrane in maintaining an osmotic gradient on a 0 to 1 scalse.  Finally, &pi;<sub>c</sub> and &pi;<sub>t</sub> capture the oncotic pressures generated by proteins in the plasma and interstitial fluids.  As in the @ref RenalMethodology, we assume that total plasma protein concentration is directly correlated with albumin concentration and use the Landis-Pappenheimer Equation to determine oncotic pressures.  The current implementation assumes that all plasma proteins (including albumin) are too large to diffuse from the vascular space (that is, &sigma; = 1) @cite gyenge1999transport, meaning that &pi;<sub>c</sub> and &pi;<sub>t</sub> maintain constant values.  This decision is due to albumin leak to the interstitium generally being small and the lack of a %BioGears lymph system to return albumin to the bloodstream.  Under regular physiological conditions, this assumption is not unreasonable.  However, in situations of rapid plasma osmolarity changes (as in during fluid administration), this assumption breaks down.  To account for these situations, we calculate the mean concentrations of albumin in the vasculature and interstitial spaces.  We then evaluate the Landis-Pappenheimer Equation in both the vasculature and intersitium using these mean concentrations and adjust the appropriate oncotic pressure source paths.  This approach still neglects the possibility of albumin leak, which can occur in the leadup to hypovolemic shock.  Future work will address the need for an accurate albumin diffusion model. 

The volume in the tissue compartment is partitioned into the extracellular and intracellular space, as shown in [Figure 2](@ref tissue-fig2) and [Figure 3](@ref tissue-fig3).  Fluid exchange between the extracellular and intracellular compartments of each tissue is directly calculated and set using a flow source.  This process of intracellular volume regulation is inextricably linked with maintanance of ionic gradients; thus discussion of it is deferred to the [Active Transport](@ref tissue-activetransport) section.  All substance transport into the tissue fluid space is simulated using one or more of the transport modes described below.

@anchor tissue-fig3
<img src="./images/Tissue/TissueCircuitExample.png" width="500">
<center>
<i>Figure 3. The tissue components of the %BioGears circuit are modeled according to the Starling Equation (Eq. 1).  Fluid exchange between the vascular and extracellular spaces is determined by hydrostatic and oncotic pressure gradients, while intracellular volume regulation and all tissue substance transport are calculated directly.</i>
</center><br>


@anchor tissue-variability
### Patient Variability
The %BioGears %Tissue system is heavily dependent on the patient configuration. Fluid volume distributions and parenchyma masses both depend heavily on the patient sex, height, weight, and body fat fraction. Transport properties are also affected by patient variability. For example, permeability coefficients are computed from membrane permeability and membrane surface area, where the surface area is a function of the tissue mass, which in turn is a function of the patient weight. A detailed discussion of patient configuration and variability in %BioGears is available in the @ref PatientMethodology report.

@anchor tissue-diffusion
##Transport Processes
@anchor tissue-perfusionLimitedDiffusion
### Perfusion-Limited Diffusion
Perfusion-limited diffusion is a technique for describing drug kinetics in physiology-based pharmacokintic models. Partition coefficents are using to compute the amount of a drug crossing a membrane at a given perfusion rate. The partition coefficients are calculated based on the physical chemical properties of the drug, the tissue properties of the organ, and the blood properties. They represent a specific substance&rsquo;s affinity for moving across the blood-tissue partition. %BioGears uses this methodology to simulate drug diffusion, and details of the partition coefficient calculation can be found in the @ref DrugsMethodology. All current drugs in the %BioGears Engine use perfusion-limited diffusion as found in @cite khalil2011physiologically @cite huisinga2012modeling. In the [future](@ref drugs-future), permeability-limited diffusion could be used. Equation 2 shows the calculation used to move mass from the vascular to the tissue and vice versa for perfusion-limited diffusion @cite khalil2011physiologically .

\f[\Delta M = Q_{T} * C_{V} - \frac{Q_{T} * C_{T}}{K_{P}} \f]
<center>
*Equation 2.*
</center><br>

Where <i>&Delta;M</i> is the change in mass due to diffusion, *Q<sub>T</sub>* is the blood flow to the organ, *C<sub>V</sub>* is the concentration of the drug in the organ vasculature, *C<sub>T</sub>* is the concentration of the drug in the organ tissue, and *K<sub>P</sub>* is the partition coefficient for the drug and organ. This calculation is performed for each drug or substance and each tissue organ/compartment.

### Gas Exchange - Alveoli Transfer
At the alveoli-pulmonary capillary interface, oxygen diffuses from the alveoli into the pulmonary capillaries, while carbon dioxide diffuses from the pulmonary capillaries into the alveoli. In reality, gas exchange at the alveoli is a multi-step process in space, where gases dissolve into liquid according to Henry's law and diffuse through liquid and across membranes according to Fick's law. In %BioGears, alveolar gas exchange is driven by the partial pressure differential between the pulmonary capillaries and the alveoli in a one-step process, as shown in Figure 4. The partial pressures of each gas in the capillaries are calculated using Equation&nbsp;3, while the partial pressures of each gas in the alveoli are calculated using Equation&nbsp;4.

<img src="./images/Tissue/AlveolarDiffusion.png" width="700">
<center>
<i>Figure 4. Alveolar gas exchange in %BioGears is a single-step, lumped-diffusion process driven by a partial pressure gradient, where the partial pressures are computed using Equations 2 and 3.</i>
</center><br>

\f[P_{P} = \frac{C}{d * C_{S}} \f]
<center>
*Equation 3.*
</center><br>

\f[P_{P} = P * V_{f} \f]
<center>
*Equation 4.*
</center><br>

Where, *P<sub>p</sub>* is the partial pressure, *C* is the concentration, *d* is the density, *C<sub>s</sub>* is the solubility coefficient, *P* is the total pressure, and *V<sub>f</sub>* is the volume fraction.

The diffusion rate is calculated using Equation&nbsp;5 @cite guyton2006medical . 

\f[\dot{D} = \frac{D_{co} * C_{D} * \Delta P_{P} * SA_{a}}{D_{d}} \f]
<center>
*Equation 5.*
</center><br>

Where *D<sub>co</sub>* is the diffusing capacity of oxygen, *C<sub>D</sub>* is the relative diffusion coefficient, \delta *P<sub>p</sub>* is the partial pressure differential between the alveoli and the capillaries, *SA<sub>a</sub>* is the surface area of the alveoli, and *D<sub>d</sub>* is the diffusion distance. The surface area of the alveoli for an individual patient is related to the standard alveoli surface area and the patient&rsquo;s total lung capacity. This calculation is shown in Equation&nbsp;6.

\f[SA_{a} = \frac{TLC_{p}}{TLC_{s}} * SA_{as} \f]
<center>
*Equation 6.*
</center><br>

Where *TLC<sub>p</sub>* is the total lung capacity of the patient, as specified in the patient file (@ref PatientData). *TLC<sub>s</sub>* is the standard healthy total lung capacity of 5.8&nbsp;L @cite guyton2006medical . The *SA<sub>as</sub>* is standard alveoli surface area of 70&nbsp;square meters @cite guyton2006medical. For more information about patient variability, please see the @ref PatientMethodology report.

The mass diffused at each time step is calculated using Equation&nbsp;7. This mass is either added or removed from the pulmonary capillaries and the corresponding volume is either added or removed from the alveoli.

\f[D_{m} = \dot{D} * \Delta t * d \f]
<center>
*Equation 7.*
</center><br>

### Instant Diffusion
Some substances are able to diffuse across biological membranes at a rate that ensures concentration equilibrium within one %BioGears time step. The instant diffusion model is included in the %BioGears %Tissue system in order to simulate transport processes that fully evolve in a time period much smaller than the %BioGears time step. All of the gases in %BioGears are transported by instant diffusion, and in the current release, sodium is also transported by instant diffusion. Active pumping mechanism for ions are planned for a future release. 

### Simple Diffusion
Simple diffusion is an implementation of Fick's law in one dimension with a known constant distance. In this case, Fick's law can be described by Equation&nbsp;8.

\f[J_{X} = P_{x} * \left([X]_{v} - [X]_{t} \right) \f]
<center>
*Equation 8.*
</center><br>
Where *J<sub>x</sub>* is the mass flux (mass per area-time) of substance *X*, *[X]<sub>v,t</sub>* is the concentration of substance X in compartment v (or t), and *P<sub>x</sub>* is a proportionality constant defining the permeability. The flux is multiplied by an area to obtain a rate of mass transfer. It is incredibly difficult to experimentally determine the capillary surface area for a given tissue, and it may be impossible to experimentally determine the total cellular membrane surface area. Additionally, lumped tissue models can be difficult to delineate. In %BioGears, the capillary and cellular membrane surface areas are assumed to be proportional to the mass of a given organ or tissue group, such that the mass transfered in one time step (*D<sub>m</sub>*) may be computed by Equation&nbsp;9, where *k* is the empirically-determined constant relating the tissue mass (*m<sub>t</sub>*) to the surface area.

\f[ D_{m} = k * m_{t} * J_{X} * \Delta t \f]
<center>
*Equation 9.*
</center><br>

In %BioGears, simple diffusion is limited to substances with a molar mass under 1000 grams per mole.

### Facilitated Diffusion
Facilitated diffusion uses Michaelis-Menten kinetics to model the facilitated transport across a membrane. Note that this type of diffusion does not require energy and it is still a gradient-based transport mode. In contrast to simple diffusion, where substance flux can continue to increase with the concentration gradient, the flux is asymptotic in facilitated diffusion. The flux limit reflects a saturation of the membrane transporter mechanisms. However, at smaller concentration gradients, substance flux is higher in facilitated diffusion than with simple Fick's law diffusion. Figure&nbsp;5 demonstrates the difference in flux between facilitated and simple diffusion. The mass flux given by Michaelis-Menten kinetics is computed using Equation&nbsp 10;, where *J<sub>max</sub>* is the maximum flux and *K<sub>m</sub> is the Michaelis constant.

\f[ J_{X} = \frac{\left([X]_{v} - [X]_{t} \right) * J_{max}}{K_{m} * \left([X]_{v} - [X]_{t} \right)} \f]
<center>
*Equation 10.*
</center><br>

<img src="./plots/Tissue/FluxVsGradient.jpg" width="900">
<center>
<i>Figure 5. In simple diffusion based on Fick's law, the substance flux is directly proportional to the concentration gradient. There is no upper limit to transport. In contrast, the flux is asymptotic in facilitated diffusion, reflecting a saturation of transporters in the membrane. These two transport processes can work in concert.</i>
</center><br>

In %BioGears, glucose is the main substance moved by facillitated diffusion. Triacylglycerol and ketones are given flux values to increase their diffusion quantity as a way to model lipophilicity, though this is planned to be improved in a future release.

@anchor tissue-activetransport
### Active Transport
Active transport is currently utilized to shuttle the ions sodium, potassium, and chloride between the intracellular and extracellular tissue compartments.  In doing so, concentration gradients characteristic of each ion are preserved (see Table 1).  These ions were emphasized because they play diverse roles in facilitated transport of other compounds (i.e. sodium-glucose), cell signaling, and cell volume maintainance @cite guyton2006medical. Volume regulation is currently implemented and discussed below, while the other functions remain areas of future work.

@anchor tissue-tab1
<br><center>
*Table 1. Typical intracellular and extracellular concentrations of ions utilized by %BioGears*
</center>
|Ion               				|Intracellular (mM)        |Extracellular (mM)        |
|------------------------       |------------------------  |------------------------  |
|Sodium 	                    |15                        |145					      |
|Potassium                 		|120                       |4.5                       |
|Chloride                       |4.5                       |102                       |

Actively transported substances are unique in %BioGears in that their movements are coupled.  The motivation for this choice stems from the fact that ions are simultaneously subject to electromotive forces arising from the resting cell membrane potential (generally -60 to -90 mV, depending on the type of cell).    

There are several approaches to modeling electrochemical flux, but the most useful for our purposes assumes that the cell membrane acts as a simple circuit (see Figure 6 below).  The cell is assumed to possess a constant capacitance and selectively permeable ionic channels that serve as branches of current flow.  Each channel has a characteristic conductance (inverse resistance) and a voltage equal to the Nernst potential of the ion in question.  

<img src="./images/Tissue/MembraneCircuit.png" width="700">
<center>
<i>Figure 6. Simple circuit model of a cell membrane @cite looie2010cell.  Membrane channels are represented by circuit paths selectively conductive to a single ion.</i>
</center><br>

The Nernst potential of Ion X, which accounts for the influence of both chemical and electrical driving forces, is given by

\f[V_{x} = \frac{RT}{zF} * ln \left(\frac{C_{e}}{C_{i}} \right) \f]   
<center>
*Equation 11*
</center><br>

In Equation 11 @cite lindblad1996model, R is the Ideal Gas Constant (8.314 J/mol-K), T is body temperature in Kelvin, z is the charge of the Ion x, F is Faraday's Constant (96,485 C/mol), C<sub>e</sub> is the extracellular concentration of Ion x, and C<sub>i</sub> is the intracellular concentration of Ion x.  If one assumes that each channel exhibits linear current-voltage (i.e. ohmic) behavior, the current across each pathway is 

\f[I_{x} = g_{x} \left(V_{m} - V_{x} \right) \f]
<center>
*Equation 12*
</center><br>

where V<sub>m</sub> is the total membrane potential @cite lindblad1996model.  By convention, a positive current is associated with movement directed from the intracellular to the extracellular space.  The conductances (obtained from @cite yi2003mathematical) assume resting membrane permeability unassociated with action potential transmission.  Equation 12 indicates that the larger the difference between an ion's Nernst potential and the membrane potential, the greater the current (flux) that ion will experience.  Intuitively, when the Nernst potential is equal to the membrane potential, the ion will experience no net flux.  These passive currents, if left unchecked, will disrupt the gradients shown in Table 1 and must be opposed by active transport.  While cells employ a large number of transport mechanisms, we model only the Na-K-ATPase in the manner of @cite luo1994dynamic using parameters from @cite yi2003mathematical.

\f[P = P_{max}f_{NaK}\left( \frac{K_{e}}{K_{e}+k_{m,K}} \right) \left(\frac{1}{1 + \left(\frac{k_{m,Na}}{Na_{i}}\right)^{1.5}}\right)\f]
<center>
*Equation 13*
</center><br>

\f[f_{NaK} = \frac{1}{1 + 0.1245exp\left(\frac{-0.1V_{m}F}{RT}\right) + 0.2555exp\left(\frac{-V_{m}F}{RT}\right)\left(exp\left(\frac{Na_{e}}{67.3}\right)-1\right)}\f]
<center>
*Equation 14*
</center><br>

The pump rate yielded by Equations 13-14 is in terms of the net number of cations moved during each pumping instance.  The net movement of 1 cation is the result of pumping three sodium ions out of the cell and two potassium ions in to the cell.  Thus, the transport rates of sodium and potassium (from an intracellular perspective) are 3P and -2P, respectively.  The parameters k<sub>m,K</sub> and k<sub>m,Na</sub> are half-saturation values for the pump.  The factor f<sub>NaK</sub> allows for the possibility of pump inhibition by extracellular Sodium.  P<sub>max</sub> represents the maximum pumping rate and, as before, F is Faraday's constant and V<sub>m</sub> is the membrane potential.  

Combining Equations 12-14 yields the system of ionic fluxes in Equations 15-17.  Chloride is assumed to undergo no active transport; rather, it is subject only to its electrochemical gradient.  This assumption is consistent with a "Pump-Leak Model", in which chloride is allowed to leak across membrane channels to conserve both charge and osmotic pressures in the face of active potassium and sodium transport.  It follows immediately that chloride concentrations must equilibrate in such a way to produce a Nernst potential identical to the membrane potential.

\f[J_{Na} = -\left(\frac{I_{Na}}{F} + 3P \right) \f]
<center>
*Equation 15*
</center><br>

\f[J_{K} = -\left(\frac{I_{K}}{F} - 2P \right) \f]
 <center>
*Equation 16*
</center><br>

\f[J_{Cl} = \frac{I_{Cl}}{F} \f]
 <center>
*Equation 17*
</center><br>

The negative signs in Equations 15 and 16 stem from the previously stated convention of positive current being directed outward.  No negative sign is included in Equation 17 because Chloride has a negative charge and will thus travel in the opposite direction of a postive current.  Division of the ionic currents by Faraday's constant renders them in a molar basis.  

The presence of these ionic fluxes affects the resting membrane potential via the differential equation

\f[-C\frac{dV_{m}}{dt} = \sum{}^{}I = g_{Na}*\left(V_{m} - V_{Na} \right) + g_{K}*\left(V_{m} - V_{K} \right) + g_{Cl}*\left(V_{m} - V_{Cl} \right) + FP\f]
<center>
*Equation 18*
</center><br>

where C is the membrane capacitance.  Equation 18, which implicitly guarantees conservation of charge, is frequently simplified in literature by making the assumption that the change in membrane potential occurs on a much faster time scale than the changes in ionic concentrations, allowing the approximation dV<sub>m</sub>/dt = 0 @cite armstrong2003k.  Equation 18 can then be arraged and solved for V<sub>m</sub>

\f[V_{m} = \frac{-FP + g_{Na}V_{Na} + g_{K}V_{K} + g_{Cl}V_{Cl}}{g_{Na} + g_{K} + g_{Cl}} \f]
<center>
*Equation 19*
</center><br>

As mentioned above, ionic transport plays a prominent role in intracellular volume regulation.  Water flows across the semi-permeable cell membrane in response to changes in osmotic pressure.  We assume that the osmotic pressure on each side of the cell membrane is solely a function of Na, K, and Cl concentrations, with one caveat.  A quick consultation of [Table 1](@ref tissue-tab1) reveals a drastic difference in the target total extracellular and intracellular molarities.  These values would immediately lead to a significant osmotic pressure gradient.  We follow the examples of @cite armstrong2003k and @cite yi2003mathematical and lump the other anionic intracellular species into a single parameter, A<sup>-</sup>.  We assume that the amount (in moles) of A<sup>-</sup> is sufficient to ensure isotonicity at our target concentrations.  With this simplification, we can express the osmotic gradient as 

\f[\Delta \Pi = RT\left(Na_{i} + K_{i} + Cl_{i} + \frac{A^{-}}{W_{i}} - Na_{e}-K_{e}-Cl_{e}\right) \f]
 <center>
*Equation 20*
</center><br>

In Equation 20, W<sub>i</sub> represents the current intracellular volume (W used to avoid confusion with membrane potential).  The degree to which water is able to pass through the cell membrane is characterized by the membrane hydraulic conductivity, denoted L<sub>p</sub>, which leads to the following equation for water flux:

\f[J_{w} = L_{p}\Delta \Pi \f] 
 <center>
*Equation 21*
</center><br>

There are biological limits to the amount of water that can be transported by this method, as cells can only tolerate mild fluctuations in volume (1-5%, depending on the cell @cite hernandez1998modeling).  However, we did not exceed these bounds during our testing due to the robustness of the Pump-Leak model.  There is research to indicate that when cells approach intolerable volumes, rapid changes in ion concentrations are initiated to correct the osmotic gradient.  This feature could be an interesting area of future modeling work, should it be required. 

To summarize, at each time-step in %BioGears, Equations 11-17, 19, and 20-21 are solved in order.  The Nernst potentials of all ions are calculated, followed by the ionic channel currents, the pump current, and molar fluxes.  The next membrane potential is calculated and stored for the subsequent time-step.  The ionic concentrations are then used in Equations 20-21 to calculate water flux into the intracellular compartment.  This flux is set on the flow source representing the cell membrane in [Figure 3](@ref tissue-fig3).


@anchor tissue-metabolic-production
## Metabolic Production and Consumption
In older versions of %BioGears, metabolic production and consumption were calculated based on a respiratory quotient, using the O2 and CO2 values to back-calculate the amount of nutrients that must have been consumed. As of %BioGears version 6.2, this functionality has been re-engineered to work in a more intuitive way. Now, the consumption and production in any given tissue is dependent on the nutrients available in that tissue and the desired metabolic rate.

The first step in the metabolic consumption and production method is to determine the amount of energy requested for the given timestep. This data is pulled from the [Energy System] (@ref EnergyMethodology), and it includes any exercise work. The next step is to portion out the energy demands to individual tissues. The brain requires a more or less constant portion of the body's total energy, about 20% @cite raichle2002appraising, so this portion is set aside. The remaining portion of the energy requested is divided between the tissues of the body according to the amount of blood flow they receive. Though this assumption may not be 100% accurate, it follows that since the blood carries oxygen and nutrients to tissues, tissues that require more blood will also require more nutrients. This method allows us to leverage already validated blood flow rates and have the potential to add new organs without micromanaging consumption for every tissue.

The next step is to determine the stoichiometric relationships between nutrients during consumption and production. Metabolism is, at its most basic level, simply a collection of chemical reactions, a transferrance of energy from ingested food to ATP. But because %BioGears doesn't model each and every substance and enzyme in the human body, the chemical reactions can't be completely characterized. However, the molar ratios between reactants and products can be used to ensure that the both the starting points and ending points of the reactions are captured. Table 1 below shows the chemical relationships used in %BioGears' metabolic production and consumption.

<br><center>
*Table 2. The molar ratios of nutrients in the metabolic production and cosumption method*
</center>
|Chemical Relationship                          |Value                     |Notes                     |
|------------------------                     	|------------------------  |------------------------  |
|ATP produced per Glucose                       |29.85                     |@cite rich2003molecular   |
|CO2 produced per Glucose                  		|6                         |                          |
|O2 consumed per Glucose                        |6                         |                          |
|ATP produced per Ketone                        |24                        |Assumed acetoacetate      |
|CO2 produced per Ketone                  		|6                         |                          |
|O2 consumed per Ketone                         |6                         |                          |
|ATP produced per Amino Acid                    |13                        |Assumed alanine           |
|CO2 produced per Amino Acid                    |1.5                       |                          |
|O2 consumed per Amino Acid                     |1.875                     |                          |
|Urea produced per Amino Acid                   |.5                        |                          |
|ATP produced per Triacylglycerol               |330                       |Assumed tripalmitin @cite boron2012medical |
|CO2 produced per Triacylglycerol               |55                        |@cite meerman2014somebody     |
|O2 consumed per Triacylglycerol                |78                        |                          |
|Aerobic ATP produced per Glycogen              |ATP per Glucose + 1 (30.85)|@cite guyton2006medical  |
|Anaerobic ATP produced per Glycogen       		|3                         |                          |
|Anaerobic Lactate produced per Glycogen        |2                         |                          |
|Anaerobic ATP produced per Glucose       		|2                         |                          |
|Anaerobic Lactate produced per Glucose         |2                         |                          |

One thing to be aware of is that not all of the energy contained in one mole of a given nutrient is converted to cellular work. For example, when burning glucose in a bomb calorimeter, 686 kcal/mol of energy is released @cite boron2012medical. In the body, however, glucose is converted to CO2, H2O, and energy by a series of reactions, none of which are as efficient as a bomb calorimeter. Thus, only a fraction of the energy in a given nutrient is used to perform cellular work; the rest is lost as heat. In order to calculate the efficiency of nutrient consumption, Equation 10 was used. These efficiency values were used for glucose, triacylglycerol, amino acids, and ketones according to @cite boron2012medical, @cite voet2013fundamentals, and @cite livesey1984energy.

\f[Efficiency=\frac{E_{ATP} *n_{ATP} }{E_{free} } \f] 
<center>
*Equation 20.*
</center><br>

Where *E<sub>ATP</sub>* is the energy released under cellular conditions per mole of ATP, *n<sub>ATP</sub>* is the number of moles of ATP generated per mole of nutrient, and *E<sub>free</sub>* is the free energy contained in one mole of nutrient.

In order to capture the appropriate nutrient responses of the body during exertion, nutrients are consumed in a certain order. Because the brain is so important to body function, its nutrient requirements are addressed first. It consumes glucose, and if none is available, it consumes ketones @cite guyton2006medical. The %BioGears brain cannot consume amino acids or triacylglycerol. Next, non-brain tissues are considered. Because the body has an obligatory protein requirement of 30 grams or more @cite guyton2006medical, amino acids are consumed next. Their rate of consumption is such that the body will always consume 30 grams a day, and this is increased under conditions where the hormone factor is low. Amino acid consumption produces urea as a byproduct, which is incremented in the liver, which, when combined with %Hepatic [gluconeogenesis] (@ref hepatic-gluconeogenesis), provides an abbreviated model of the Cori cycle. Next in line for consumption is triacylglycerol, the body's largest aerobic energy source. Consumption of triacylglycerol is rate-limited based on the hormone factor, with low hormone factors allowing for greater fat consumption. Next, aerobic glucose consumption via glycolysis and the citric acid cycle is considered. If the tissue in question is the muscle tissue, muscle glycogen can be used, first aerobically, then, in the absence of blood glucose, anaerobically. This ordering of consumption ensures that, under a resting state, the body tries to maintain its nutrient stores.

If at any time a tissue does not have the necessary oxygen or nutrient required to meet its energy demands, the deficit is logged as fatigue.

## Protein/Fat Storage and Release
As the liver stores glucose in the form of glycogen to serve as an energy source in times of need, it also stores fat in the adipose tissue @cite guyton2006medical. Protein doesn't have explicit storage in the same fashion, but it has been shown that about 1% of total body protein can be quickly broken down and metabolized during times of need @cite institute2005dietary. This is not to be confused with muscle catabolism, which happens more slowly and after a longer period of time, and which is not modeled in %BioGears. The functionality of fat and protein storage and release works very similarly to glycogenesis and glycogenolysis in the liver (see [Hepatic Syetem] (@ref hepatic-features)). First, a local [hormone factor] (@ref endocrine-hormone-factor) is calculated, which is used to determine a rate of storage or release. These rates were determined empirically to give appropriate time courses of nutrient in the blood. Then, the substances are incremented accordingly.

Assumptions and Limitations
---------------------------
Proteins are large molecules that take up space. About 7% of plasma volume is due to proteins. Proteins also have a net negative charge. In reality, diffusion depends on the elctrochemical gradient, not just the chemical gradient. With the exception of the perfusion-limited diffusion model, the %BioGears diffusion models do not account for the entire elecrochemical gradient or the volume of the protein in the plasma (i.e. no "plasma water").

The diffusional exchange of water between the capillaries and extravascular space amounts to as much as 80,000 liters per day. Convective capillary exchange is much less, on the order of 16 liters per day. The diffusional exchange of water is not modeled in %BioGears.

The lipophilicity of molecules like triacylglycerol is not considered in %BioGears' current diffusion method. If triacylglycerol was diffused only based on its considerable molar mass, it would hardly move at all in %BioGears. To remedy this, some triacylglycerol is moved via facilitated diffusion by assigning it diffusion flux parameters.

The brain is currently the only tissue capable of metabolizing ketones, though in reality both myocardium and muscle can also consume ketones for energy. This limitation will be reexamined with upcoming changes to the modeling of starvation.

The muscle is considered as one lumped compartment in %BioGears. For accurate modeling of consumption and production, especially in the case of isolated lactate production, discrete muscle compartments would give more accurate results.

Glucose diffusion is regulated by insulin level in many tissues, though this behavior is not modeled in %BioGears.

Conditions
----------

#Dehydration
The dehydration condition is biogears is initialized as a fraction that corresponds to a decrease in patient weight from baseline. We assume that dehydration and extracellular fluid volume depletion are differing events with the latter corresponding to substance depletion in addition to fluid loss @cite mange1997language. We are careful to not associate extracellular sodium with dehydration to allow for correct bedside assessment of the patient in regards to elevated solute concentrations in the fluid compartments of the body. The dehydration fraction determines how much weight the patient has lost as a result of fluid depletion, for example a 0.05 fraction corresponds to a 5 percent body weight decrease. We also support a patient event that informs the user when the patient has reached over 3 percent decrease in body weight. This is generally the threshold when active performance begins to decrease and obvious signs of fluid imbalance become prominent @cite cheuvront2014dehydration. In addition BioGears will inform the user that the patient is thirsty at an increase of 3 percent plasma osmolarity. 

This fraction is then used to reduce fluid volume levels across all fluid compartments within BioGears, including the vasculature @cite dill1974calculation. We compute the fractional amount of fluid decrease in each compartment by first computing total water lost as a function of patient mass lost: 

\f V = {M_P}\left( {\frac{{{M_{frac}}}}{\rho }} \right). \f] 

Here *M<sub>P</sub>* is the baseline patient weight (generally initialized in the patient file), *M<sub>frac</sub>* is the fractional weight lost due to dehydration, set by the condition, and *\rho* is the standard density of water at internal body temperatures. We then scale to get a fractional fluid lost by dividing through by total body water. This value is then multiplied by the extracellular, intracellular, and vascular fluid compartments to get total fluid lost in each.

Validated data is mostly qualitative for a 5 percent patient weight decrease from baseline due to fluid loss. The condition does well for total body water reductions, patient weight decrease, and plasma osmolarity increases, table 4. This allows for a clinician to properly diagnose the condition in a training scenario.

For significant dehydration (> 5 percent) it is known that performance at submaximal exercise and in variable environmental conditions increases heart rate and decrease stroke volume @cite cheuvront2014dehydration. These cardiovascular deficiencies have not been tested for our virtual patient but could be an interesting lead into future work. Due to the strenuous and variable nature of testing dehydrated patients during exercise, at altitude, and in varying hot/cold environments, having a computational tool and validated model to determine performance to the soldier or athlete would be useful. 

<br><center>
*Table 3. Validated values for patient dehydration at rest.*
</center>
|	Condition	|	Notes	|	Sampled Scenario Time (s)	|	Body Weight	|	Aorta Sodium Concentration	|	Blood Volume	|
|	---	|	---	|	---	|	---	|	---	|	---	|
|	5% body weight dehydration condition applied	|	Dehydration condition is applied as a reduction of patient weight from baseline	|	60	|<span class="success">	Decrease @cite saltin1964circulatory	</span>|<span class="success">	Increase @cite saltin1964circulatory	</span>|<span class="success">	Decrease @cite saltin1964circulatory	</span>|



Actions
-------
### Insults

#### Burn Wound
A burn wound is triggered in BioGears by specifying the total body surface area (TBSA) affected by the burn and any compartments (Trunk, Left Arm, Right Arm, Left Leg, Right Leg) affected. If no compartments are specified, the trunk will be assumed. Additionally, if multiple compartments are identified, then the TBSA is assumed to be split over the affected compartments. Once initiated, the burn model causes a cascade of reactions in the physiology engine through instantiating a a pain response and an inflammatory response through the inflammation model resulting in some hemorrhaging and numerous physiolgoical responses including decreased urine output, cardiovascular distress, respiratory distress, and increased skin temperatures. The changes in urine output are especially important in burn resuscitation protocols. The burn specific model responds by calculating the change in pressure (mmHg) of the muscle tissue to increase the resistance of blood flow into the affected compartments. The result is a decrease in flow and a pressure differential that can contribute to the decrease in cardiac output, renal blood flow, and urine output that is expected during a burn. This compartmental change can also lead to compartment syndrome based on the severity of the burn and the severity of any fluid creep present as a result of burn resuscitation. For additional infortmation on the burn model, please reference the BioGears paper detailing it at (insert link here).

### Interventions

#### Escharotomy
If the compartment syndrome event has been triggered, an escharotomy can be performed on the affected compartment. In BioGears, an escharotomy action currently provides immediate pressure relief to the associated compartment by negating any resistance changes built up as a result of modifications to the tissue pressure and blood flow. The engine does not currently account for the changes in infection risk associated with this procedure. 

Additionally, other system actions can affect the diffusion properties or other transports by modifying the diffusion surface area. An example of this is found in the @ref RespiratoryMethodology for lobar pneumonia. As the alveoli fill with fluid, they are unable to participate in gas exchange. This reduces the alveoli surface area, which leads to a reduction of available oxygen in the @ref CardiovascularSystem and the @ref EnergySystem. 

Events
------

- Compartment Syndrome: Triggered during a burn as a result of the pressure differential in an affected compartment 
- Dehydration: Set when fluid loss exceeds 3% of body mass @cite who2005dehydration
- Thirst: Set when plasma osmolarity increases 2% @cite cheuvront2014dehydration

Results and Conclusions
=======================

%Verification
-------------
%Verification of the diffusion methods is achieved through several units tests. One of the simple diffusion unit tests was used to generate data for Figure 7. The figure shows the time-evolution of the concentrations of four different compartments. Table 3 shows the initial conditions. Note that the units are arbitrary, thus not shown. The red, blue, and green compartment all share a boundary with the yellow compartment, but not with each other.
<br><center>
*Table 4. Initial conditions for a four compartment simple diffusion unit test.*
</center>
| Compartment | Volume | Mass | Concentration |
| ----------- | ------ | ---- | ------------- |
| Red         | 50.0   | 2.0  | 0.04          |
| Blue        | 10.0   | 2.5  | 0.20          |
| Green       | 20.0   | 10   | 0.50          |
| Yellow      | 50.0   | 0.0  | 0.00          |

<img src="./plots/Tissue/ConcentrationEquilibration.jpg" width="900">
<center>
<i>Figure 7. Four compartments start with different concentrations which equilibrate after some time. Initial conditions are shown in Table 3 above.</i>
</center><br>

@anchor tissue-validation
Validation - Resting Physiologic State
--------------------------------------
The tissue system volumes are validated using data from @cite valentin2002icrp.  

<br><center>
*Table 4. Validation of the resting physiologic state comparison of system-level outputs from %BioGears to referenced values. System-level outputs show favorable agreement with validation data.*
</center>
@insert doc/validation/TissueValidationTable.md

<br><center>
*Table 5. Validation of the resting physiologic state comparison of compartment-level outputs from %BioGears to referenced values. The compartments are currently validated on a flow/volume basis. Flows and most of the volumes show good agreement with validation values.*
</center>
@insert doc/validation/TissueCompartmentsValidationTable.md

More validation of this system can be found in the system outputs of all other systems, e.g., the oxygen and carbon dioxide saturation, the blood pH, and the bicarbonate concentration values are found in the @ref BloodChemistryMethodology and the alveoli oxygen and carbon dioxide partial pressures are found in the @ref RespiratoryMethodology.

Validation - Actions and Conditions
-----------------------
There are currently no validated actions or conditions associated with the %Tissue system. However, there will be condition validation after [improvements](@ref energy-future) are made to the dehydration and starvation condition models.


@anchor tissue-future
Future Work
===========
Coming Soon
-----------
- Albumin transport model to affect blood albumin concentration based on hepatic production and to model changes in plasma colloid oncotic pressure.

Recommended Improvements
------------------------
- Permeability-limited diffusion model
- Endogenous carbon monoxide production
- Muscle catabolism
- Insulin effects on glucose diffusion

Appendices
==========

Data Model Implementation
-------------------------
@ref TissueSystemTable "Tissue"

Compartments
------------
- Bone
 - BoneExtracellular
 - BoneIntracellular
- Brain
 - BrainExtracellular
 - BrainIntracellular
- Fat
 - FatExtracellular
 - FatIntracellular
- Gut
 - GutExtracellular
 - GutIntracellular
- LeftKidney
 - LeftKidneyExtracellular
 - LeftKidneyIntracellular
- LeftLung
 - LeftLungExtracellular
 - LeftLungIntracellular
- Liver
 - LiverExtracellular
 - LiverIntracellular
- Muscle
 - MuscleExtracellular
 - MuscleIntracellular
- Myocardium
 - MyocardiumExtracellular
 - MyocardiumIntracellular
- RightKidney
 - RightKidneyExtracellular
 - RightKidneyIntracellular
- RightLung
 - RightLungExtracellular
 - RightLungIntracellular
- Skin
 - SkinExtracellular
 - SkinIntracellular
- Spleen
 - SpleenExtracellular
 - SpleenIntracellular		  
- Lymph
%Cardiovascular Methodology {#CardiovascularMethodology}
==========================

@anchor cardiovascular-overview
Overview
========

Abstract
--------
The %BioGears&reg; %Cardiovascular system is a closed-loop series of compartments that represent the human circulatory system. The cardiovascular system is sometimes described as two separate circulations: the systemic circulation and the pulmonary circulation. Both circulations, as well as the lymph, are represented in the %BioGears %Cardiovascular system. The dynamical state of the %Cardiovascular System is determined at every timestep through a three-step process. First, intra- and inter-system feedback is applied during the preprocess step. Next, the state of the circulatory system is determined by solving an equivalent circuit. Finally, the state of the system is updated in preparation for time advancement. Through system and action validation, we demonstrate the accuracy of the model at resting physiology and a under variety of cardiovascular related insults and interventions with appropriate system feedback. 

Introduction
------------
### %Cardiovascular Physiology
The cardiovascular system is a large organ system comprised of the heart and the blood vessels. It serves as the body's primary transport and distribution system.  As mentioned above, it is frequently subdivided into the the systemic circulation and the pulmonary circulation. In the systemic circulation, oxygenated blood leaves the left side of the heart, travels  through arteries and into the capillaries, and then returns as deoxygenated blood through the veins to the right side of the heart. Upon exiting the right side of the heart, the deoxygenated blood enters the pulmonary circulation through the pulmonary arteries, is re-oxygenated in the pulmonary capillaries, and then returns to the left side of the heart through the pulmonary veins (Figure&nbsp;1). Whether the cardiovascular circulation is thought of as one closed circuit with a single dual-purpose pump or two interconnected circuits with separate synchronous pumps, the combination of systemic and pulmonary vasculature is a critical distribution and exchange network.  It provides vital oxygen to the tissues and removes toxic carbon dioxide while distributing nutrients and other substances necessary for healthy physiologic function.

@image html CardiovascularSystemDiagram.png
<center>
<i>Figure&nbsp;1. The %Cardiovascular System with pulmonary and systemic circulations of the human body @cite KVDP2014Circulatory . 
Both the pulmonary and systemic circulation originate in the heart, which acts as a pump driving the blood through the entire body. Blood enters the heart from the vena cava via the right atrium and transitions to the right ventricle, which sends it to the pulmonary arteries and into the lungs. Gas exchange occurs in the vascular arterioles, creating oxygen-rich blood. This blood then returns to the heart via the pulmonary veins to the left atrium. The oxygen-rich blood enters the systemic circulation through the left ventricle, providing oxygen to the rest of the body.</i>
</center><br>

The cardiovascular system is one of two systems that comprise the circulatory system, the other being the lymphatic system. The %BioGears cardiovascular system includes a low-resolution lymphatic circulation. Because the resolution of the lymph circulation in %BioGears is considerably lower than the resolution of the cardiovascular circulation, we refer to the comprehensive circuit as the circulatory circuit, but we maintain the term *cardiovascular* as system nomenclature to highlight the modeling focus. In other words, the %BioGears lymph circulation is not a physiological model of the human lymphatic system. It only serves as an accessory return for interstitial fluid. More information about the lymph model and the other accessory advection models is available in the @ref TissueMethodology report.

### %Cardiovascular Modeling
#### Brief Introduction to %Cardiovascular Modeling
%Cardiovascular (CV) models range from high-resolution, small-region-of-interest models @cite taylor1999predictive @cite wan2002one @cite formaggia2003one to low-resolution, whole-body simulations @cite heldt2004computational. Increased resolution may provide a more accurate solution in a localized region at the expense of decreasing the size of the modeling region. However, the level of accuracy achieved with such a model is not always necesssary to answer the question posed. Three-dimensional (3D) computational fluid dynamics (CFD) models are generally limited to short segments of the arterial or venous tree due to high computational requirements. These high-resolution models are well-suited for investigating complex geometries following surgical intervention, such as the Fontan Circulation @cite whitehead2007nonlinear @cite de1996use. However, they are not suitable for modeling long-term or systemic effects in the CV System @cite timmons1995cardiovascular because it is only feasible to simulate a few cardiac cycles.

One-dimensional (1D) models can also predict the pressure and flow wave dynamics within a region of interest @cite womersley1955xxiv @cite womersley1957oscillatory and have the advantage of a lower computational cost than full CFD models. Like CFD models, 1D models begin with the Navier-Stokes equations, but they are reduced to a hyperbolic partial differential equation by including an axisymmetric assumption. The resultant equation dictates a constant pressure in the radial direction under the assumption that the radius of the vessel is small relative to a characteristic wave length @cite vcanic2003mathematical. These models are well-suited for investigating specific CV problems @cite spilker2007morphometry @cite clipp2009impedance and have been coupled with dynamic downstream effects to investigate problems over numerous cardiac cycles @cite clipp2012evaluation. However, the computational cost associated with 1D models still renders them unsuitable for modeling the complete closed-circuit dynamics of the circulation over the entire range of vessel diameters and lengths in the human body.

The lumped parameter method of hemodynamic modeling assumes that characteristics of organ vessels or vessel segments can be  &ldquo;lumped&rdquo; into representative parameters. For instance, the viscosity of a fluid can be lumped together with the geometry of a vessel segment to derive an effective resistance to blood flow. If a tube is rigid, the pressure-flow dynamics are completely characterized by the resistance to flow in the tube. If a tube is elastic, an additional compliance parameter can be used to characterize the impedance and calculate the pressure-flow dynamics in the segment. In humans, small arteries can be approximated as rigid @cite olufsen2004deriving, while the corresponding veins are significantly (up to \~24 times) more compliant @cite hall2011guyton.

Otto Frank first formulated the quantitative model (called a Windkessel model after the wave-filtering air chamber in archaic firefighting equipment) relating systemic blood pressure with vessel elasticity @cite frank1899grundform. 
In the analogous electrical circuit (see @ref CircuitMethodology), the Windkessel model includes a capacitor--representing the compliance of the artery, connected in parallel with a resistor, representing the resistance to flow through the vessel. An alternate three-element form of the Windkessel model was developed by Westerhof et al. @cite westerhof1969analog that includes a black-box, two-terminal input circuit (representing an additional resistor), to quantify the characteristic impedance of the arterial tree. Similar applications of the electric analogue circuit have been applied to form CV circuits that can be classified as multi-segment or multi-system models.

#### %BioGears Lumped Parameter Model
Lumped parameter models have proven suitable for larger-scale, longer time-scale simulations due to lower computational requirements and sufficient fidelity. A multi-organ model can run in real time while still producing an accurate blood pressure and flow waveform prediction. The drawback to this type of model is the large number of parameters requiring definition or tuning. However, significant work has been targeted toward determining appropriate parameter values @cite bronzino1999biomedical @cite kokalari2013review @cite milivsic2004analysis @cite olufsen2004deriving @cite rideout1967difference @cite westerhof2009arterial. This body of work, in particular Heldt&rsquo;s open source CV simulator that includes lumped parameter models @cite heldt2004computational @cite hester2011hummod @cite summers2008development, supports the decision to implement these models in %BioGears.

@anchor cardiovascular-system-design
System Design
=============
Background and Scope
--------------------
### History of the %BioGears %Cardiovascular Model
The %BioGears CV Model has its origins in Guyton&apos;s four-compartment (three vascular, plus one heart) model of the CV System designed to analyze the effect of varying circulatory factors on cardiac output @cite guyton1955effect. Rideout, et al. used Guyton&rsquo;s foundation and an electric-hydraulic analogy to streamline the generation of difference-differential equations for modeling fluid flow in distensible vessels @cite rideout1967difference. Yasuhiro Fukui leveraged the previous work to model the CV and respiratory systems and their interactions @cite fukui1972study. The development of a mass and momentum transport model of the CV System allowed for the simulation of interactions between the CV System and angiotensin @cite masuzawa1992cardiovascular. Following the success of the angiotensin-cardiovascular simulator, development of the drug-interaction model continued and eventually led to an anesthesiology simulator that incorporated CV, respiratory, and drug models @cite smith1995pc. This simulator, released by Advanced Simulation Corporation as Body Simulation for Anesthesia&tm;, formed the backbone of the HumanSim&tm; physiology engine, which continues to provide realistic physiology for several serious training games in the HumanSim product line, including HumanSim: Sedation and Airway @cite Clipp2012Humansim. The HumanSim physiology engine is the starting-point for the %BioGears Engine. The basic building blocks of the CV System remain as described in Masuzawa et al. @cite masuzawa1992cardiovascular; however, the CV circuit has been further developed to provide a more accurate CV simulation and drive other %BioGears systems through intersystem dependencies.

Data Flow
---------
The state of the CV System is determined at every time step through a three-step process: Preprocess, Process, and Postprocess (Figure&nbsp;2). In the Preprocess step, feedback from other systems, as well as intrasystem feedback, is processed in preparation for determining the state of the system. Process uses the %BioGears [circuit](@ref CircuitMethodology) calculator to compute the new state of the system. Postprocess is used to prepare the system for the advancement of time. More specifics about these steps are detailed below.

@anchor cardiovascular-initialize
### Initialization and Stabilization
First, the patient-specific homeostatic state of the cardiovascular circuit is computed. Next, all system parameters are initialized. The %Cardiovascular System is then initialized to the resting state. Conditions are then applied by modifying system and patient parameters and restabilizing the engine. The available @ref cardiovascular-conditions "conditions" in the %Cardiovascular System are anemia, heart failure, pericardial effusion, and pulmonary shunt.

#### Tune Circuit
In the tune circuit step, the resistors and capacitors associated with tissue compartments are tuned during stabilization to achieve the mean arterial pressure given in the patient file.

### Preprocess
#### Begin Cardiac Cycle
Cardiac cycle calculations include methodology for updating the driving force (heart contraction and relaxation) of the CV System throughout the duration of a CV cycle (a single heart beat). This includes a set of systolic calculations that updates contractility at the beginning of the cycle to represent a heart contraction. Modifications to heart rate and heart compliance are calculated by BeginCardiacCycle and applied for the remainder of the current cardiac cycle. Changes to things like heart rate and heart contractility can only occur at the top of the current cardiac cycle after the last cardiac cycle has completed. This helps to avoid discontinuous behavior such as the complete cessation of heart function mid-contraction.

#### Calculate Heart Elastance
This method tracks the progress of the current cardiac cycle and modifies the compliance of the left and right heart to drive the cardiovascular circuit. The reduced compliance at the beginning of the cycle acts to increase the pressure, driving flow out of the heart. The compliance is then reduced, allowing flow into the heart. The reduction and increase in compliance represents the systolic and diastolic portions of the cardiac cycle, respectively. The compliance is driven by a Hill-type elastance equation @cite stergiopulos1996elastance.

#### Process Actions
Process Actions modifies the CV parameters and circuit properties due to actions (insults or interventions) specified by the user. The @ref cardiovascular-actions "actions" found in the action process are: CPR, hemorrhage, pericardial effusion, and cardiac arrest.

### Process
The generic circuit methodology developed for the %BioGears Engine is used to solve for the pressure, flow, and volume at each node or path in the equivalent circuit. For more details, see @ref CircuitMethodology.

#### Calculate Vital Signs
This function takes the current timestep&rsquo;s circuit quantities to calculate important system-level quantities for the current time step. The system pressures and flow rates related to shunting are calculated here. In addition, the events hypovolemia, tachycardia, bradycardia, and asystole are triggered in this function.

### Postprocess
The Postprocess step moves everything calculated in Process from the next time step calculation to the current time step calculation. This allows all other systems access to the information when completing their Preprocess analysis for the next time step.

### Assessments
Assessments in %BioGears are data collected and packaged to resemble a report or analysis that might be ordered by a physician. No %BioGears assessments are associated with the %Cardiovascular system.

<br>
<img src="./images/Cardiovascular/CardiovascularDataFlow.png" width="900">
<center>
*Figure 2. The data flow for the CV System consists of Preprocess, Process, and Postprocess. Preprocess determines the circuit element values based on feedback mechanisms and engine actions. Preprocess also updates system-level quantities via the CalculateVitalSigns and cardiac cycle calculations. Process uses generic circuit methodology to solve the CV circuit for pressures, volumes, and flows along the paths and on nodes. Postprocess updates these quantities to the next time step and then begins advancing time, which causes the loop to repeat.*
</center><br>

@anchor cardiovascular-features
Features, Capabilities, and Dependencies
----------------------------------------
### The %BioGears %Cardiovascular Circuit
The %BioGears CV circuit (Figure&nbsp;3) estimates blood pressure, flow, and volume for organs that are represented by several compartments. These compartments are comprised of lumped parameter models that use resistors and capacitors. Inductors may also be used to model inertial effects. The system is discretized into nodes that are connected by paths (see @ref CircuitMethodology). The circuit used to represent the CV System was designed to provide a level of resolution and fidelity that meets the [requirements](@ref MainPageFAQ) of the overall project.

For example, to provide a means for clearing drugs and substances from the bloodstream, the liver and kidneys must have blood flow, pressure, and volume calculations. Another example is the four extremities (right and left arms and legs) that provide extremity hemorrhage capabilities, having been implemented in a previous project (HumanSim: Combat Medic). In this way, the lumped parameter model provides a mechanism for increasing fidelity as required by the anatomic region or physiologic condition being modeled. The large thoracic arteries are lumped together into one &ldquo;Aorta&rdquo; compartment, which is represented by four nodes and three paths. The fidelity of any compartment could be easily improved by increasing the level of discretization. By adding nodes and paths, the %BioGears &ldquo;Aorta&rdquo; could become the &ldquo;Ascending Aorta&rdquo; and &ldquo;Descending Aorta&rdquo; to accommodate the fidelity demands of the other systems. This could provide an opportunity to model more complex geometries and pathologies, such as stenosis. Figure 3 shows the %BioGears cardiovascular circuit. For clarity, the more discretized [renal circuit](@ref renal-circuit) is not shown in this diagram. 
<img src="./images/Cardiovascular/BioGearsCardiovascularCircuit.png" width="900">
<center>
*Figure 3. The cardiovascular circuit consists of nodes that are connected via paths. These segments of nodes and paths are mapped to several compartments which represent the anatomy of the cardiovascular system. The circuit is used to estimate the blood pressure, flow, and volume of these anatomical compartments.*
</center><br>

Nodes serve as the connection points for paths and are the locations at which pressures are measured. Each CV node contains a pressure value, which is given with respect to the atmospheric reference node (indicated in the diagram by the equipotential symbol). Paths contain information about the flow (volume per time). The %BioGears @ref CircuitMethodology document contains more information about %BioGears circuit definitions and modeling. The %BioGears @ref SubstanceTransportMethodology contains more information about the substance transport. In general, nodes contain "across" information and paths contain "through" information.

The elements of the %BioGears CV circuit are used to model the fluid dynamics of the human CV System (hemodynamics). The hemodynamic pressure and flow are calculated from the lumped parameters that are determined by the circuit element. The equations used to calculate pressure and flow are shown below. These equations are automatically generated and solved simultaneously by the %BioGears [Circuit Solver](@ref CircuitMethodology).

Derived values for the hemodynamic parameters are available, particularly for specific vessel segments @cite olufsen2004deriving @cite rideout1972cardiovascular. However, in the %BioGears Engine, these parameters are tuned to provide the physiologic response of a given [patient profile](@ref PatientMethodology). We completed the tuning process by choosing estimates for each parameter value based on the existing system values (Body and HumanSim) or based on physiologic data. We then analyzed the model outputs and adjusted parameters to obtain organ and system-level outputs that satisfied the validation requirements. The @ref cardiovascular-results "Results and Conclusions" section describes validation in more detail.

### The %BioGears Heart
The %BioGears heart generates pressure that drives the hemodynamics through a variable capacitor that simulates the changing elastance of the myocardium throughout the cardiac cycle. The %BioGears heart has two sides, left and right, simulating the two sides of the human heart. The atria are not included in the heart model; only the ventricular behavior is modeled.

#### Heart Elastance and Compliance
The heart compliance is calculated from the inverse of the heart elastance. The heart elastance model used in %BioGears is adapted from the one developed by Stergiopulos&nbsp;et&nbsp;al @cite stergiopulos1996elastance. This model utilizes a double Hill function to represent heart elastance over the cardiac cycle time period. It was chosen due to its ability to scale with increasing or decreasing cardiac cycle times. The functional form for elastance of both left and right ventricles is shown in Equation&nbsp;1 and Equation&nbsp;2.

\f[E_{v} (t)=(E_{\max ,v} -E_{\min ,v} )\left(\frac{f(t)}{f_{\max } } \right)+E_{\min ,v} \f] 
<center>
*Equation 1.*
</center><br>

Where *E<sub>max,v</sub>* is the maximum ventricle elastance in mmHg per mL, *E<sub>min,v</sub>* is the minimum ventricle elastance in mmHg per mL, and *f<sub>max</sub>* is the maximum value of the double Hill over the cardiac cycle length.  The double Hill function *f(t)* is given by

\f[f(t)=\left[\frac{\left(\frac{t}{\alpha _{1} T} \right)^{n_{1} } }{1+\left(\frac{t}{\alpha _{1} T} \right)^{n_{1} } } \right]\left[\frac{1}{1+\left(\frac{t}{\alpha _{2} T} \right)^{n_{2} } } \right]\f] 
<center>
*Equation 2.*
</center><br>

Where &alpha;<sub>1</sub> , &alpha;<sub>2</sub> , *n<sub>1</sub>*, and *n<sub>2</sub>* are shape parameters used to determine the distribution of the double Hill function, *T* is the cardiac cycle time period, and *t* is the current time within the cardiac cycle.

The relationship between the elastance and compliance in %BioGears is shown in Figure&nbsp;4.
<img src="./plots/Cardiovascular/ComplianceVsElastance.jpg" width="1100">
<center>
*Figure 4. The left heart compliance and elastance are shown to be inversely related to each other. The elastance represents the change in pressure per change in volume, while the compliance is the change in volume per change in pressure. These quantities define the contraction of the heart, which drives the pressure and flow of the cardiovascular circuit.*
</center><br>

#### Heart Pressure, Volume, and Flow
The variable compliance, which is used to model heart contraction and relaxation, yields pressure and volume changes that drive the flow through the CV circuit. This variable compliance driver allows the pressures and volumes to be calculated within the heart, as shown in Figure&nbsp;5.

<img src="./plots/Cardiovascular/PressureVsVolume.jpg" width="1100">
<center>
*Figure 5. Relationship between pressure and volume in the left heart throughout the cardiac cycle. The relaxation of the heart muscle is modeled by increasing the compliance, resulting in an increase in left heart volume with a relatively constant left heart pressure. The contraction is represented by a rapid decrease in the compliance, leading to large pressure increases for small volume additions. This large pressure value drives the fluid out of the heart with flow rates calculated based on the circuit solution.*
</center><br>

A pressure-volume curve is used to represent the evolution of the cardiac cycle from the systolic contraction to disatolic relaxation. The pressure-volume curve for the left ventricle is shown in Figure&nbsp;6. Starting from the bottom left and moving clockwise, the curve demonstrates a rapid increase in pressure with no change in volume. This indicates the systolic contraction of the cardiac cycle. Following this, the pressure declines rapidly as the heart expands during diastole. The last portion of the curve shows decreasing volume at constant pressure. Normally, the pressure would decrease slightly due to the imperfect mitral valve, which does not close instantly. %BioGears uses ideal valves, which close instantaneously, causing the pressure to be maintained as volume decreases.

<img src="./plots/Cardiovascular/PVLoop.jpg" width="400">
<center>
*Figure 6. The pressure-volume curve for the left ventricle is represented as a pressure vs. volume plot. It demonstrates the the contracting and relaxing portions of the cardiac cycle. In addition, the curve demonstrates the use of ideal valves in the %BioGears heart due to instantaneous changes in volume at a set pressure.*
</center><br>

### Drug Effects
As drugs circulate through the system, they affect the %Cardiovascular System. The drug effects on heart rate, mean arterial pressure, and pulse pressure are calculated in the %Drugs System. These effects are applied to the heart driver by incrementing the frequency of the heart based on the system parameter calculated in the %Drugs System. Additionally, the mean arterial pressure is modified by incrementing the resistance of the blood vessels. At each timestep, the resistors are incremented on each cardiovascular path. In the future, the pulse pressure will be modified by changing the heart elastance. However, no drugs in the %BioGears library currently cause pulse pressure changes; therefore, this step has not been necessary. The strength of these effects are calculated based on the plasma concentration and the drug parameters in the substance files.  For more information on this calculation see @ref DrugsMethodology.

### Electrocardiogram
The %BioGears electrocardiogram (%ECG) machine outputs an %ECG waveform to represent the currently supported heart rhythms. The waveform is not a calculated value; the current %BioGears %ECG system does not model the electrical activity of the heart. This data is stored in a text file and interpolated to match the %BioGears timestep. Currently, we support tachycardia, bradycardia, asystole, and normal sinus. Figure 7 shows an example of the %ECG normal sinus output. By allowing the points on the waveform to sum together based on the length of the heart beat compared to the length of the %ECG waveform, a tachycardic waveform will have a compressed P and T wave as expected in the validation data. For asystole, %BioGears outputs 0&nbsp;mV.

<img src="./images/Cardiovascular/BasicECGWaveform.png" width="300">
<center>
<i>Figure 7. The %ECG system produces a normal sinus waveform with the expected features.</i>
</center><br>

@anchor cardiovascular-variability
### Patient Variability
The state of the cardiovascular system is heavily dependent on the patient. The cardiovascular circuit parameters are all set based on the patient configuration. For example, the proportion of blood flow to the fat compartment depends on the sex of the patient, and the initial flow proportions dictate initial vascular resistance. Likewise, the volume of the fat compartment depends on the sex, body fat fraction, height, and weight of the patient, and the initial volume and pressure are used to compute the initial vascular compliance for paths associated with a compartment. After initial values are computed, the cardiovascular circuit parameters are adjusted to meet target pressures, which are also specified in the patient file. Because the cardiovascular system is profoundly dependent on the patient configuration, several stabilization iterations are sometimes necessary to achieve a specific desired state. For that reason, we make modifications to the parameters to move the circuit closer to a %BioGears Standard Male patient state to speed stabilization for that patient, and other custom states may require longer stabilization periods. A detailed discussion of patient configuration and variability in %BioGears is available in the @ref PatientMethodology report.
@anchor cardiovascular-dependencies
### Dependencies
The other %BioGears systems depend on the CV System to function accurately and appropriately. The CV System provides the medium (blood) and the energy (pressure differentials and subsequent flow) to transport substances throughout the %BioGears system. A complete list of %BioGears substances and their descriptions can be found in the @ref SubstanceTransportMethodology. Similarly, the %BioGears CV System is dependent on all of the other systems. Feedback from other systems influences the cardiac behavior via heart rate and elastance modifiers and the systemic circulation through resistance modifications to the circuits. For example, as the plasma concentration of a drug increases, the elastance of the heart, heart rate, and systemic resistance are modified through the pharmacodynamic effects (multipliers), resulting in changes to the blood flow, pressure, and volume calculations. Additionally, baroreceptor feedback leads to modification of heart rate, elastance, and vascular resistance and compliance in order to regulate the arterial pressure. For more information on baroreceptor feedback see @ref NervousMethodology.

Gas exchange is one area that closely ties the respiratory system, substances, and the CV system together. Gas exchange is calculated based on the partial pressures of the gases in the lungs and the capillary beds (@ref SubstanceTransportMethodology). If respiration was to stop due to a drug effect or physiologic condition, such as airway obstruction, the gas exchange calculations would be affected. Oxygen would stop diffusing into the CV System from the %Respiratory System, resulting in an oxygen deficit in the cardiac tissue.

@anchor cardiovascular-assumptions
Assumptions and Limitations
---------------------------
The constant-compliance equations in the CV System assume a linear diastolic pressure-volume relationship in the vessels (&Delta;V = C&Delta;P) representative of an elastic vessel wall. This assumption is appropriate for many blood vessels under normal physiologic conditions; however, research shows that many vessel walls are more appropriately modeled as visco-elastic walls @cite alastruey2011visco. While the non-linear pressure-volume relationship within the heart is modeled using a variable capacitor, the visco-elastic nature of the blood vessels is not addressed in this model. Rigid and elastic wall properties have been used to model the hemodynamic behavior in high-fidelity models with success @cite alastruey2011visco. Therefore, the level of fidelity of the lumped parameter model and resulting simulation support the use of elastic walls.

The CV model has a waveform limitation because of our ideal diode model. We assume the valves are perfect Booleans. In reality, there is a transition time between the open and closed states of a valve. During this time, a small amount of retrograde flow may be present @cite halliwill2010retro . In the future, we hope to incorporate retrograde flow through a Zener diode. However, the current methodology does allow retrograde flow at any location without a diode. The Zener diode implementation will provide a mechanism for future modeling efforts, including abnormal heart valves that result in significant retrograde flow.

@anchor cardiovascular-conditions
Conditions
----------
#### Anemia
Anemia conditions reduce the oxygen-carrying capacity of the blood. The %BioGears Engine models iron deficiency anemia as a chronic condition, which is characterized by a decrease in hemoglobin concentration and subsequent decreases in hematocrit and blood viscosity @cite Duke1969Hemodynamic. These factors lead to a decrease in systemic vascular resistance @cite guyton2006medical. The %BioGears Engine currently supports up to a 30% decrease in hemoglobin. After engine stabilization, the chronic condition reduces the hemoglobin throughout the circuit and reduces the systemic vascular resistance to represent the change in viscosity. The engine then re-stabilizes based on the chronic condition criteria. For more information, see @ref SystemMethodology. There is an observable increase in venous return due to the decreased systemic vascular resistance. As validation data supports, there are no observable effects from the decreased oxygen-carrying capacity at rest. These effects will be evident in the future with incorporation of exercise.

#### Arrhythmias
A heart arrhythmia is a deviation from the normal sinus rhythm seen in a healthy heart beat. In %BioGears, we have implemented two chronic arrhythmias, sinus bradycardia and sinus tachycardia. Bradycardia is a low resting heart rate (less than 60&nbsp;beats per minute) @cite mangrum2000evaluation, which is generally considered benign when cardiac output is stable. Tachycardia is a high resting heart rate (greater than 100&nbsp;beats per minute) @cite yusuf2005deciphering. After engine stabilization, the heart rate baseline is modified. After the engine re-stabilizes based on the chronic condition criteria (see @ref SystemMethodology for details), the heart rate value, and corresponding changes in the %ECG waveform, are the only clear sign of the condition. This matches the validation data. A limitation of the current chronic tachycardia model is a lack of changes to the compression force. Currently, only the rate is affected.

#### Heart Failure
Heart failure refers to a malfunctioning of the heart, resulting in inadequate cardiac output. The current mode of heart failure being modeled in %BioGears is chronic left ventricular systolic dysfunction. After engine stabilization, the chronic heart failure is modeled with a 45% reduction in the contractility of the left heart. After the engine has re-stabilized using the chronic condition criteria (@ref SystemMethodology), the heart ejection fraction of 60% is reduced to 31% as noted in the validation data @cite chatterjee2007systolic. The reduced contractility leads to a decrease in heart stroke volume, which causes an immediate drop in the cardiac output and arterial pressures. The baroreceptor reflex responds with an increase in heart rate in an attempt to return mean arterial pressure to its normal value.

#### Pericardial Effusion
A chronic pericardial effusion is an accumulation of fluid in the pericardial sac that surrounds the heart. This accumulation occurs over a long period of time, resulting in the body making some accommodation to these changes. The result is an increased pressure on the heart, which leads to strained performance. This performance strain is less severe than in an acute effusion. Due to this increasing pressure, there is a decrease in stroke volume, which leads to a decrease in cardiac output and arterial pressures. To compensate for reduced stroke volume, the baroreceptor response raises the heart rate in an attempt to restore cardiac output.

@anchor cardiovascular-actions
Actions
-------
### Insults

#### Cardiac Arrest
The cardiac arrest action is used to externally initiate a cardiac arrest event (see @ref cardiovascular-arrest "Cardiac Arrest" event below). The cardiac arrest event is a cessation of all cardiac and respiratory function. The cardiac arrest event can be triggered by systems of the biogears engine when physiological thresholds are reached (physiological cardiac arrest), or it can be triggered by the cardiac arrest action (interface-initiated cardiac arrest). The cardiac arrest insult can be thought of as an idiopathic sudden cardiac arrest.

#### Hemorrhage
A hemorrhage is a substantial blood loss that may lead to inadequate supply of oxygen to tissues and vital organs. Reduced blood volume leads to reduced cardiac output, reduced mean arterial pressure, and an increase in heart rate due to baroreceptor response.  Hemorrhages are initiated in %BioGears by specifying a physiological compartment and a severity (on a scale of 0 to 1).  Table 1 identifies the compartments which currently support hemorrhages.  Looking to the potential inclusion of the Hemorrhage Action in a broader Injury Model, %BioGears Hemorrhage scenarios output an injury code corresponding to the Military Combat Injury Scale (MCIS).  Furthermore, the %BioGears Software Development Kit (SDK) demonstrates a method for initiating a Hemorrhage via the MCIS.  An MCIS code contains five digits, the first of which defines the wound severity (on a scale of 1-5).  The second digit defines the region in which the wound occured (Head, Torso, Arm, Leg, Multiple), while the third through fifth digits provide further specificity (i.e. tissue type, organ, etc).  Wound resolution in %BioGears is not detailed enough to require all five digits; in instances where digits are not required, they are simply set to 0.  Table 1 shows sample injury codes for each compartment that supports hemorrhages, as well as information about how the code is derived.  For a complete description of the development of the MCIS, see @cite lawnick2013combat (list of sample codes in Supplemental Digital Content 3).


<center>
*Table 1 . Valid compartments in which hemorrhages can be defined in %BioGears (Column 1) with sample injury code output (Column 2, note that S = severity)) and derivation of code (Columns 3-5)*
</center>

|%BioGears Compartment		|Injury Code|	Region (Digit 2)	|	Subregion (Digit 3)	|	Specific Site (Digit 4)	|	
|------------------------	|-----------|	--------------------|	--------------------|	-----------------------	|
|	Head					| S1610		|	Head				|	Vessels				|	Cranial					|
|	Major Artery			| S2640		|	Torso				|	Vessels				|	Arterial				|
|	Vena Cava				| S2660		|	Torso				|	Vessels				|	Other vessel (Assign to Vena Cava)	|
|	Lung					| S2710		|	Torso				|	Chest Organ			|	Lung					|
|	Myocardium				| S2720		|	Torso				|	Chest Organ			|	Myocardium				|
|	Liver					| S2810		|	Torso				|	Abdominal Organ		|	Liver					|
|	Spleen					| S2820		|	Torso				|	Abdominal Organ		|	Spleen					|
|	Splanchnic				| S2830		|	Torso				|	Abdominal Organ		|	Pancreas				|
|	Kidney					| S2840		|	Torso				|	Abdominal Organ		|	Kidney					|
|	Small Intestine			| S2850		|	Torso				|	Abdominal Organ		|	GI Tract (Assign to Small Intestine)|
|	Large Intestine			| S2860		|	Torso				|	Abdominal Organ		|	Other (Assign to Large Intestine)	|
|	Arm						| S3000		|	Arm					|	----				|	----					|
|	Leg						| S4000		|	Leg					|	----				|	----					|



During initialization, circuit pathways with open switches connecting each compartment to ground are defined to provide potential bleeding routes.  The compartment provided in the Hemorrhage Action is then mapped to the appropriate pathway, closing the switch and adjusting its resistance to flow as a function of wound severity.  Multiple hemorrhages may exist at any given time. Bleeding rates--from which the total blood volume is calculated--are obtained by sampling the flows from relevant circuit pathways to ground.  Figure 8 highlights the pulsatile nature of the bleeding rate, which trends downward as the total blood volume decreases.  Concurrently, arterial pressures and cardiac output drop. The baroreceptor reflex initiates an increase in heart rate to compensate for the reduced mean arterial pressure.  

Associated with the decrease in total blood volume is a decrease in blood-borne substances that will be linearly proportional to the bleed rate. For more specific information regarding these substances and their loss due to bleeding, see @ref BloodChemistryMethodology and @ref SubstanceTransportMethodology. Figure 8 shows the blood volume and hemoglobin content before, during, and after a Class 2 hemorrhage (15-30% blood loss @cite garrioch2004body) event with no intervention other than the cessation of hemorrhage. Figure 9 shows a hemorrhage event with subsequent saline administration. Note that the hemoglobin content remains diminished as the blood volume recovers with IV saline. By comparison, [Figure 12](@ref cardiovascular-blood-administration) shows a blood-product intervention following a hemorrhage event. In that figure, the hemoglobin increases with the blood infusion.

<table>
<tr>
<td><img src="./plots/Cardiovascular/Class2NoFluid_BloodVolume.jpg" width="550">
</td>
<td><img src="./plots/Cardiovascular/Class2NoFluid_Hb.jpg" width="550">
</td>
</tr>
</table>
<img src="./plots/Cardiovascular/Class2NoFluid_Legend.jpg" width="500">
<center>
*Figure 8. Blood volume and hemoglobin content before, during, and after a Class 2 hemorrhage event with no subsequent intervention.*
</center><br>

<table>
<tr>
<td><img src="./plots/Cardiovascular/Class2Saline_BloodVolume.jpg" width="550">
</td>
<td><img src="./plots/Cardiovascular/Class2Saline_Hb.jpg" width="550">
</td>
</tr>
</table>
<img src="./plots/Cardiovascular/Class2Saline_Legend.jpg" width="500">
<center>
*Figure 9. Blood volume and hemoglobin content before, during, and after a Class 2 hemorrhage event with a subsequent infusion of saline.*
</center><br>

#### Pericardial Effusion
The pericardial effusion action is used to model acute pericardial effusion by adding a flow source on the pericardium. This action leads to a volume accumulation over the course of the simulation. The accumulated volume is used to calculate a pressure source that is applied to the left and right heart. This pressure source is identical to the one used in the pericardial effusion condition. For the pericardial effusion action, the strain-rate dependent compliance of the pericardium is modeled so that the change in intrapericardial pressure is a function of flow rate and the current volume of the pericardium @cite Metoyer2014Modeling.

### Interventions

#### Cardiopulmonary Resuscitation (CPR)
CPR can be preformed to maintain some perfusion through the induced hemodynamic action. The American Heart Association recommends performing chest compressions at a rate of 100&nbsp;per minute with enough force to achieve a chest-compression depth of 5&nbsp;cm @cite berg2010adult. Gruben et al estimated a required force of approximately 400&nbsp;N to achieve the required depth @cite gruben1990system. Kim et al found that stroke volumes of 12&ndash;35&nbsp;mL and an effective cardiac output of 1200&ndash;3500&nbsp;mL/min could be achieved with the proper compression technique. The aforementioned hemodynamics are accompanied by an increase in systolic and diastolic pressures and, by extension, mean arterial pressure @cite kim2008direction.

Chest compressions are simulated in the %BioGears Engine by adjusting the value of the pressure source that is connected between the common equipotential node and the capacitors that represent the right and left cardiac compliances. The pressure source represents the intrathoracic pressure, thus a positive pressure value represents an increase in intrathoracic pressure. The pressure on the source is calculated from the input force and the biometrics of the patient.

The CPR force of compression can be expressed to the %BioGears Engine in one of two ways. A direct application method allows a user to explicitly set the applied force at the time of the CPR compression call. Successive compressions are achieved by calling a CPR compression at a desired force, calling for an advance in simulation time, and then calling the next desired force. In this way, the shape of the time-dependent force curve is explicitly set by the user. The other method of applying a compression force during CPR is to use the CPR force scale. In this method, the user can call for a compression by selecting a scale value between 0 and 1, with 0 being no force and 1 being the maximum possible force. The maximum possible compression force, currently 500&nbsp;N, is based on the work of Arbogast et al @cite arbogast2006anterior. When the force scale is used, the evolution of the force curve in time is set by the %BioGears engine. A bell shaped force curve is used in accordance with @cite stanley2012recreating. Once a force scale compression is called for, another compression will not be possible until the current compression has fully evolved. A new explicit force will also be prohibited until the compression is complete. In the current %BioGears Engine, recovery from cardiac arrest is not possible as the only arrest rhythm is asystole. Consequently, it is not possible to reverse a cardiac arrest event with defibrillation. Of course, it is never possible to convert an ineffective rhythm with CPR alone.

The compression-only CPR in the %BioGears Engine is consistent with AHA guidelines for lay rescuer adult CPR @cite hazinski2010. The addition of rescue breathing for health-care provider CPR is mentioned as a recommended improvement to the %BioGears Engine.

#### Hemorrhage Cessation
A &ldquo;tourniquet&rdquo; may be applied to a bleeding distal portion of the body to reduce blood flow entering that portion and effectively stopping the bleeding. With the bleeding managed, vital signs should return to their normal state after a sufficient period of time. A &ldquo;tourniquet&rdquo; may be simulated in the %BioGears Engine by creating a Hemorrhage action in the bleeding compartment with a decreased severity.  A hemorrhage can be stopped completely by entering a severity of 0.  If the hemorrhage leads to significant blood loss, reducing the severity of that hemorrhage will lead to a return of the arterial pressures as the baroreceptor reflex (which is not affected by blood loss) exerts itself. As the MAP increases, heart rate will begin to decrease in proportion to the error control dictated by the baroreceptor reflex. Lastly, cardiac output will begin to increase to a new, stabilized value associated with the new total blood volume.

#### Tranexamic Acid
Tranexamic Acid (TXA) may be injected into the body via substance infusion to effectively reduce hemorrhage bleeding over time. Pharmacodynamics of TXA show quick dispersion in the blood with very quick effects taking place at the effect site. Depending on the severity and location of the hemorrhage, TXA increases the chance of survival by slowing the bleeding either entirely or enough to get the patient help. With the bleeding managed, vital signs should eventually return to their normal state. If the hemorrhage is stopped completely, there will be a return of the arterial pressures as the baroreceptor reflex (which is not affected by blood loss) exerts itself. As the MAP increases, heart rate will begin to decrease in proportion to the error control dictated by the baroreceptor reflex. Lastly, cardiac output will begin to increase to a new, stabilized value associated with the new total blood volume.

#### Intravenous Fluid Administration
Intravenous fluid administration is a continual injection of a compatible fluid into the veins of a patient that has suffered from fluid loss. Due to increasing volume from intravenous fluid administration, blood volume and arterial pressures will increase. Stroke volume will increase due to increased venous return, which will cause an increase in cardiac output. The baroreceptor response will lead to a decrease in the heart rate.

Intravenous fluid administration is simulated by applying a flow source to the vena cava. The flow source and duration of the administration is dictated by the user in the form of a flow rate and IV bag volume. This results in an increase in total blood volume, heart stroke volume, cardiac output, and arterial pressures. Due to increasing arterial pressures, the baroreceptor reflex will begin to decrease heart rate according to the error between the mean arterial pressure and its set point. Additional effects may occur depending on the type of fluid administered. Currently, the user may administer blood or saline. Each of these fluids is considered a compound substance with multiple substance components. Further information on the effect of substances during fluid administration can be found in @ref SubstanceTransportMethodology.

@anchor cardiovascular-events
##Events

### Asystole
Asystole is defined as a cessation of the heart beat. Asystole is currently triggered by an oxygen deficit in the myocardium (a partial pressure for oxygen less than 25&nbsp;mmHg) or by a heart rate of less than 26&nbsp;beats per minute @cite guinness2005lowest. The %BioGears Engine continues to process during Asystole, which allows for resuscitation efforts, such as CPR, to be attempted. If the Asystole is not addressed, the engine will reach an irreversible state due to a lack of oxygen in the brain. More detail on oxygen deficits can be found in @ref BloodChemistryMethodology.

@anchor cardiovascular-arrest
### Cardiac Arrest
Cardiac Arrest is a condition in which the pumping of the heart is no longer effective @cite nhlbi2011sca. The Cardiac Arrest event can be triggered in the %BioGears engine either by the Cardiac Arrest action or by the evolution of engine physiology. For instance, an oxygen deficit in the heart muscle will trigger both an asystole event (see below) and a cardiac arrest event. In the current version of the %BioGears engine, the only rhythm associated with cardiac arrest is asystole, but the cardiac arrest event is included to facilitate control and to allow the future inclusion of other ineffective rhythms such as ventricular fibrillation. In the current %BioGears engine, it is not possible to recover from cardiac arrest. It is, however, possible to maintain some perfusion by performing chest compressions (see CPR).

### Cardiogenic Shock
In general, the term "shock" refers to inadequate perfusion of the tissues. The several categories of shock serve to signify the origin of the disturbance. Cardiogenic shock is inadequate perfusion due a reduction in the pumping capability of the heart. In %BioGears, the Cardiogenic Shock event is activated when the cardiac index () is below 2.2 (L/min-m^2) *and* the systolic blood pressure is less than 90.0 (mmHg) *and* the pulmonary capillary wedge pressure is greater than 15.0 (mmHg) @cite dhakam2008review.

### Hypovolemia
Hypovolemia is defined as a reduction in total blood volume by 35&nbsp;percent @cite guyton2006medical. Typically, this is classified by elevated heart rate and decreased arterial pressure. In %BioGears, hypovolemia is triggered during a hemorrhage when blood volume has fallen below 65&nbsp;percent of its normal value. 

@anchor cardiovascular-results
Results and Conclusions
=======================
The %BioGears %Cardiovascular System was validated quantitatively and qualitatively under resting physiologic conditions and transient (action-induced) conditions. The validation is specified with a color coding system, with green indicating good agreement with trends/values, yellow indicating moderate agreement with trends/values, and shades red indicating poor agreement with trends/values. 

@anchor cardiovascular-validation-resting
Validation - Resting Physiologic State
--------------------------------------
Validation results for system and compartment quantities are listed in Tables&nbsp;1 and&nbsp;2. System-level quantities show favorable agreement with validation values. Heart rate, arterial pressures, blood volume, heart stroke volume, and cardiac output are the predominant CV System quantities. These values agree, on average, within ~8 percent of the expected values for the healthy standard patient. 

<br><center>
*Table 2. Validation of the resting physiologic state comparison of system-level outputs from %BioGears to referenced values. System-level outputs show favorable agreement with validation data.*
</center>
@insert doc/validation/CardiovascularValidationTable.md

<br><center>
*Table 3. Validation of the resting physiologic state comparison of compartment-level outputs from %BioGears to referenced values. The compartments are currently validated on a flow/volume basis. Flows and most of the volumes show good agreement with validation values.*
</center>
@insert doc/validation/CardiovascularCompartmentsValidationTable.md

Compartment-level quantities show reasonable agreement with the validation values. All of the flows match the reference values within ~10&nbsp;percent. The volumes show some moderate differences for a few specific compartments. The aorta compartment volume is much smaller than the validated value. The compliance on this compartment needed to remain low in order to preserve the arterial pressure waveform, which led to less volume than expected. Similarly, the vena cava compliance was set in order to maintain the correct cardiac output and arterial pressures; therefore, its expected volume was limited. The right heart pressures and volumes show some disagreement with the validation data. The minimum values for right heart pressure and volume are much lower than valid ranges. This is due to restriction of unstressed volume in the right heart, which currently has an unstressed volume of zero. An increase in unstressed volume would shift the pressure volume minimums up, while also preserving the maximum values within their respective ranges. The %Cardiovascular System is tuned to vitals output validation (Table&nbsp;2), as well as good agreement with insults' and interventions&rsquo; expected trends and values (see the following section).  In addition, compartment validation was achieved on a reasonable level.

The arterial pressure waveform was validated according to the plot shown in Figure&nbsp;10. It displays the %BioGears arterial pressure against measured arterial pressure. The diastolic and systolic pressures were validated using data shown in Table&nbsp;1. To validate the waveform shape and demonstrate the overall feature match of the %BioGears pressure waveform with the  validation data, we used a waveform from PhysioNet @cite goldberger2000physiobank . However, the patient heart rate and parameters are slightly different than the %BioGears patient. This led to timing discrepancies and differences in the diastolic and systolic pressures. To demonstrate the waveform feature matching, a separate axis is used for each data set. In the future, we may create a patient more representative of the one used for the PhysioNet waveform. This would show both the ability of the engine to model different patients and the pressure waveform feature matching.  The shapes of both waveforms match well, showing rapid pressure increases as the heart contraction begins to occur. The main difference in the shape of each plot is the small pressure oscillations that occur after the initial pressure drop in the validation data. This is the dicrotic notch,  which occurs from slight flow reversal from the aorta back into the left ventricle before the valves close @cite korakianitis2005notch . This is not currently being modeled in %BioGears, but improvements in the circuit model, including the addition of inertance and diodes that allow retrograde flow, will likely enable the %BioGears waveform to capture more detail.

<center>
<table>
<tr>
<td><img src="./plots/Cardiovascular/BioGearsPressure.jpg" width="400"></td>
<td><img src="./plots/Cardiovascular/PhysioNetPressure.jpg" width="400"></td>
</tr>
</table>
<i>Figure 10. Arterial pressure waveform comparisons. The diastolic and systolic pressures were validated using the data shown in Table&nbsp;1. To validate the waveform shape and demonstrate the overall feature match of the %BioGears pressure waveform with the  validation data, a waveform was found on PhysioNet @cite goldberger2000physiobank . However, the patient heart rate and parameters are slightly different than the %BioGears patient. This led to timing discrepancies and differences in the diastolic and systolic pressures. To demonstrate the waveform feature matching, a separate axis is used for each data set. Both the validation waveform and the %BioGears waveform show sharp increases in pressure during the systolic period. After the contraction occurs, the pressure begins decreasing and that is where the main difference in the engine and the validation data occur. There is a dip and subsequent rise in the arterial pressure that occurs due to the dicrotic notch, which the engine does not capture.</i>
</center><br>

@anchor cardiovascular-validation-conditions
Validation - Actions and Conditions
--------------------
All actions in the CV System were validated. A summary of this validation is shown in Table&nbsp;4. More details on each individual scenario's validation can be found below.

<center>
*Table 4. Actions associated with the %Cardiovascular System were validated qualitatively by comparing the engine output to expected trends. Engine results show favorable agreement (green), some agreement (yellow), or no agreement (red). The table shows the total number of results analyzed for each scenario. Results mostly matched expected trends.*
</center>

|	Scenario 	|	Description	|	Validation Type	|	Good	|	Decent	|	Bad	|
|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|
|	Pericardial Effusion - Chronic	|	Patient has an effused pericardium with an accumulated volume of 500 ml.	|	Qualitative	|<span class="success">	9	</span>|<span class="warning">	0	</span>|<span class="danger">	0	</span>|
|	Ventricular Systolic Failure	|	Chronic heart failure is initiated, which causes a permanent reduction in heart contractility for the scenario duration.	|	Qualitative	|<span class="success">	8	</span>|<span class="warning">	0	</span>|<span class="danger">	1	</span>|
|	Hemorrhage Class 2 - Blood	|	Bleed occurring in an artery, with a blood transfusion occurring afterwards.	|	Qualitative	|<span class="success">	18	</span>|<span class="warning">	0	</span>|<span class="danger">	0	</span>|
|	Hemorrhage Class 2 - Saline	|	Bleed occurring in an artery, with intravenous saline administration occurring afterwards.	|	Qualitative	|<span class="success">	15	</span>|<span class="warning">	0	</span>|<span class="danger">	0	</span>|
|	Hemorrhage Class 2 - No Fluid	|	Bleed occurring in an artery.	|	Qualitative	|<span class="success">	10	</span>|<span class="warning">	0	</span>|<span class="danger">	0	</span>|
|	Hemorrhage Class 4 - No Fluid	|	Bleed occurring in an artery.	|	Qualitative	|<span class="success">	5	</span>|<span class="warning">	0	</span>|<span class="danger">	0	</span>|
|	Hemorrhage - Combat Medic 1	|	Bleeding in the right leg and vena cava. After bleeding has stopped, intravenous saline is administered.	|	Qualitative	|<span class="success">	21	</span>|<span class="warning">	3	</span>|<span class="danger">	1	</span>|
|	Hemorrhage - Combat Medic 2	|	Bleeding in the right leg and vena cava. After bleeding has stopped, a bolus morphine injection is given and intravenous saline is administered.	|	Qualitative	|<span class="success">	21	</span>|<span class="warning">	7	</span>|<span class="danger">	2	</span>|
	Hemorrhage-TXA Intervention	|	Bleeding in the right leg followed by infusion of intravenous tranexamic acid (TXA)	|	Qualitative	|<span class="success">	8	</span>|<span class="warning">	0	</span>|<span class="danger">	0	</span>|
|	Pulmonary Shunt - 10%	|	A pulmonary shunt occurs, which reduces the effective oxygen transfer to the pulmonary capillaries.	|	Qualitative	|<span class="success">	1	</span>|<span class="warning">	2	</span>|<span class="danger">	1	</span>|
|	Pulmonary Shunt - 30%	|	A pulmonary shunt of increasing severity occurs, which reduces the effective oxygen transfer to the pulmonary capillaries.	|	Qualitative	|<span class="success">	1	</span>|<span class="warning">	1	</span>|<span class="danger">	2	</span>|
|	IV Fluids	|	Intravenous saline administration on a healthy patient.	|	Qualitative	|<span class="success">	6	</span>|<span class="warning">	1	</span>|<span class="danger">	0	</span>|
|	Sinus Bradycardia	|	Heart rate set to 50 beats per minute.	|	Qualitative	|<span class="success">	5	</span>|<span class="warning">	2	</span>|<span class="danger">	0	</span>|
|	Sinus Tachycardia	|	Heart rate set to 140 beats per minute.	|	Qualitative	|<span class="success">	5	</span>|<span class="warning">	2	</span>|<span class="danger">	0	</span>|
|	Anemia - 30%	|	Hemoglobin content reduced by 30  percent.	|	Qualitative	|<span class="success">	7	</span>|<span class="warning">	1	</span>|<span class="danger">	0	</span>|
|	CPR 	|	Cardiac arrest is initiated, and CPR is performed.	|	Qualitative	|<span class="success">	16	</span>|<span class="warning">	6	</span>|<span class="danger">	2	</span>|


### Cardiopulmonary Resuscitation (CPR)
There are two CPR scenarios for validation. The first scenario validates the CPR methodology using the explicit force setting functionality, and the second scenario uses the force scale methodology (see CPR intervention documentation above). In both scenarios, cardiac arrest is initiated externally. Ten seconds later, compressions begin. In both scenarios, the compression rate is set to 80 per minute, and the force is set to 70 pounds to match the conditions in @cite redberg1993physiology. Supplemental literature sources were used to validate outputs not available in @cite redberg1993physiology. All of the physiological variables were within validation ranges in both scenarios with the exception of mean arterial pressure and ejection fraction. The mean arterial pressure in the %BioGears engine is slightly higher than expected. This is most likely due to the fact that the intravascular pressures are higher than those reported in @cite redberg1993physiology. However, the %BioGears pressures are within ranges reported in other references @cite kim2008direction , @cite gruben1990system. The ejection fraction is considerably lower in %BioGears during CPR than the value reported in @cite kim2008direction. %BioGears ejection fraction is lower because blood tends to pool in the %BioGears right heart during cardiac arrest. The validation failures that occur right at cardiac arrest are mostly due residual dynamics following asystole in %BioGears. Errors associated with the cessation of heart function in %BioGears are a known issue, and resolving this issue is a part of the cardiac arrest recommended improvements discussed [below](@ref cardiovascular-future).  

<br><center>
*Table 5. Cardiopulmonary resuscitation (CPR) validation results.*
</center>
|	Notes	|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Heart Rate(beats/min)	|	Systolic Pressure (mmHg)	|	Diastolic Pressure (mmHg)	|	Mean Arterial Pressure (mmHg)	|	Cardiac Output (mL/min)	|	Stroke Volume (mL)	|	Carotid Artery (Brain) Flow (mL/min)	|	Ejection Fraction (%)	|
|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|
|	Cardiac Arrest	|	30	|	35	|<span class="success">	0	</span>|<span class="success">	N/A	</span>|<span class="success">	N/A	</span>|<span class="success">	33 +/- 10 @cite kim2008direction %BioGears 96	</span>|<span class="warning">	0	</span>|<span class="warning">	0	</span>|<span class="warning">	Approach 0 	</span>|<span class="warning">	0	</span>|
|	Chest Compressions	|	40	|	165-170	|<span class="success">	80 per minute [Direct calculation] %BioGears 80	</span>|<span class="success">	39.27 @cite redberg1993physiology Approx. 70-80 @cite gruben1990system 105 +/- 41 @cite kim2008direction %BioGears 43.14	</span>|<span class="success">	13.97 @cite redberg2014physiology Approx. 40 @cite gruben1990system 33 +/- 10 @cite kim2008direction %BioGears 30	</span>|<span class="warning">	21.13 @cite redberg1993physiology %BioGears 34.12	</span>|<span class="success">	17-27% Normal @cite kim2008direction %BioGears 27%	</span>|<span class="success">	19.7 @cite redberg2014physiology 25 +/- 8 @cite kim2008direction %BioGears 21	</span>|<span class="success">	> 0 during compression (translated from dog study) @cite rudikoff1980mechanisms %BioGears 3.5 ml/s Average	</span>|<span class="danger">	34 +/- 16 @cite kim2008direction %BioGears 4%	</span>|
|	Chest CompressionsForce Scale	|	40	|	165-170	|<span class="success">	80 per minute [Direct calculation] %BioGears 80	</span>|<span class="success">	39.27 @cite redberg1993physiology Approx. 70-80 @cite gruben1990system 105 +/- 41 @cite kim2008direction %BioGears 40	</span>|<span class="success">	13.97 @cite redberg1993physiology Approx. 40 @cite gruben1990system 33 +/- 10 @cite kim2008direction %BioGears 28	</span>|<span class="warning">	21.13 @cite redberg1993physiology %BioGears 33	</span>|<span class="success">	17-27% Normal @cite kim2008direction %BioGears 25%	</span>|<span class="success">	19.7 @cite redberg2014physiology 25 +/- 8 @cite kim2008direction %BioGears 20.20	</span>|<span class="success">	> 0 during compression (translated from dog study) @cite rudikoff1980mechanisms %BioGears 3.5 ml/s Average	</span>|<span class="danger">	34 +/- 16 @cite kim2008direction %BioGears 4%	</span>|


### Heart Failure
The heart failure scenario has a ventricular systolic dysfunction condition applied to the patient prior to stabilization. The effect is modeled as a chronic condition, meaning there is a permanent change in the heart contractility for the entirety of the&nbsp; scenario duration. After sustaining heart failure for 20 minutes the patient then receives a diuretic to reduce congestion in the body. This follows a standardized treatment of heart failure patients and displays %BioGears' ability to not only simulate the problem but also to provide and investigate the solution. The PD effects of furosemide are detailed in the @ref DrugsMethodology, but the results of increased urine production are seen here on the total blood volume of the patient. The heart failure scenario was validated qualitatively, with only the diastolic pressure failing to match the expected behavior.

<img src="./plots/Cardiovascular/HeartFailure.jpg" width="1100">
<center>
<i>Figure 11. Total blood volume due to the localized effect on the renal tubular lumen can be seen here after administration of the diuretic furosemide.</i>
</center><br>

<br><center>
*Table 6a. Heart failure validation results (Condition).*
</center>
|	Condition	|	Sampled Scenario Time (min)	|	Heart Ejection Fraction	|	Heart Rate (/min)	|	Systolic Pressure (mmHg)	|	Diastolic Pressure (mmHg)	|	Cardiac Output (mL/min)	|	Heart Stroke Volume (mL)	|	Pulmonary Capillaries Wedge Pressure (mmHg)	|	Oxygen Saturation	|	Systemic Vascular Resistance	|
|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|
|	Ventricular Systolic Failure	|	20	|<span class="success">	Decrease @cite klabunde2012cvconcepts 31% @cite chatterjee2007systolic	</span>|<span class="success">	Mild Increase (89 +/- 22) @cite gheorghiade2009acute	</span>|<span class="success">	No change, increase, or decrease @cite gheorghiade2009acute	</span>|<span class="danger">	No change, increase, or decrease @cite gheorghiade2009acute	</span>|<span class="success">	Decrease @cite gheorghiade2009acute	</span>|<span class="success">	Decrease @cite gheorghiade2009acute	</span>|<span class="success">	Increase @cite gheorghiade2009acute	</span>|<span class="success">	NC	</span>|<span class="success">	Increase @cite klabunde2012cvconcepts	</span>|


<br><center>
*Table 6b. Heart failure validation results (furosemide administration).*
</center>
|	Action	|	Sampled Scenario Time (min)	|	Urine Production Rate (mL/min)	|	Chloride Excretion (mmol/min)	|	Sodium Excretion (mol/min)	|	Vascular Volume (mL)	|
|	------------------------	|	------------------------	|	---	|	---	|	---	|	---	|
|	Furosemide Administration	|	120	|<span class="success">	300-400% Increase @cite lahav1992intermittent 	</span>|<span class="success">	300-400% Increase @cite lahav1992intermittent 	</span>|<span class="success">	300-400% Increase @cite lahav1992intermittent 	</span>|<span class="success">	Decrease (around 500-1000 mL) @cite hammarlund1985acute	</span>|


### Hemorrhage
The hemorrhage action is tested using several scenarios. The class 2 hemorrhage scenario with blood intravenous (IV) administration begins with a healthy patient. After a few seconds, a severe arterial (severity value = 0.8) hemorrhage is initiated. The hemorrhage continues for ten minutes before the bleeding is stopped. After two minutes, 500&nbsp;mL of IV blood is administered intravenously over five minutes. The other hemorrhage scenarios are similar but with different subsequent interventions. There are also two multi-compartment hemorrhage scenarios that demonstrate bleeding from other compartments (the spleen and leg). Finally, tranexamic acid is validated as an antifibrinolytic agent based on its ability to reduce bleeding of a hemorrhage. Figure 11 demonstrates the time-evolution of select data, and the validation results are displayed in Tables 7a-h.

The results show decreases in the systolic pressure and minor increases in the diastolic pressure during the course of the hemorrhage. In response to the decreasing arterial pressures, the baroreceptor response raises the heart rate. Note that fluid can shift between the intravascular and extravascular space. This shift is evident in the period between cessation of hemorrhage and the start of the infusion (top-left panel of Figure 11).

Following the completion of the hemorrhage, intravenous blood is administered. The validation of this action can be found in the IV Fluid Administration section, with the exception of hemoglobin content. There will be an increase in hemoglobin content directly proportional to the amount of blood added from the IV.

@anchor cardiovascular-blood-administration
<table>
<tr>
<td><img src="./plots/Cardiovascular/Class2Blood_BloodVolume.jpg" width="550">
</td>
<td><img src="./plots/Cardiovascular/Class2Blood_Diastolic.jpg" width="550">
</td>
</tr>
<tr>
<td><img src="./plots/Cardiovascular/Class2Blood_Hb.jpg" width="550">
</td>
<td><img src="./plots/Cardiovascular/Class2Blood_Systolic.jpg" width="550">
</td>
</tr>
</table>
<img src="./plots/Cardiovascular/Class2Blood_Legend.jpg" width="1100">
<center>
*Figure 12. The class 2 hemorrhage scenario shows the blood volume decreasing after a severe arterial hemorrhage. The blood hemoglobin content follows this exact trend. At the conclusion of the bleed, the blood volume and hemoglobin are at a lower value. Five hundred (500) milliliters of blood is then administered intravenously over the course of 5&nbsp;minutes. Both the blood volume and hemoglobin content increase linearly with this administration.*
</center>

<br><center>
*Table 7a. Class 2 hemorrhage with blood administration validation results.*
</center>
|	Action	|	Notes			|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Heart Rate (/min)	|	Systolic Pressure (mmHg)	|	Diastolic Pressure (mmHg)	|	Cardiac Output (mL/min)	|	Heart Stroke Volume (mL)	|	Respiration Rate (/min)	|
|	------------------------	|	------------------------	------------------------	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|
|	Initiate Hemorrhage from Aorta	|	Injury Code:  42640 (Severity = 4)			|	30	|	900	|<span class="success">	Increase @cite garrioch2004body	</span>|<span class="success">	Slight decrease or no change at all @cite garrioch2004body	</span>|<span class="success">	Increase @cite garrioch2004body	</span>|<span class="success">	Decrease @cite garrioch2004body	</span>|<span class="success">	Decrease @cite garrioch2004body	</span>|<span class="success">	Increase @cite garrioch2004body	</span>|
|	Stop Hemorrhage	|	We see an increase in BV because of pressure differentials allowing fluid transfer from tissues. This is not accounted for in the calculated value.			|	900	|	1020	|<span class="success">	Decrease @cite Chalmers1967	</span>|<span class="success">	Slight increase or no change at all @cite garrioch2004body	</span>|<span class="success">	Decrease @cite Chalmers1967	</span>|<span class="success">	Increase @cite Chalmers1967	</span>|<span class="success">	Increase @cite Chalmers1967	</span>|<span class="success">	Decrease @cite garrioch2004body	</span>|
|	Start IV Fluids: Blood at 100 mL/min with a 500 mL bag	|				|	1020	|	1420	|<span class="success">	Decrease @cite metoyer2016SME	</span>|<span class="success">	Increase @cite metoyer2016SME	</span>|<span class="success">	Increase @cite metoyer2016SME	</span>|<span class="success">	Increase @cite metoyer2016SME	</span>|<span class="success">	Increase @cite metoyer2016SME	</span>|<span class="success">	Increase @cite garrioch2004body	</span>|


<br><center>
*Table 7b. Class 2 hemorrhage with saline administration validation results.*
</center>
|	Action	|	Notes			|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Heart Rate (/min)	|	Systolic Pressure (mmHg)	|	Diastolic Pressure (mmHg)	|	Cardiac Output (mL/min)	|	Heart Stroke Volume (mL)	|
|	------------------------	|	------------------------	------------------------	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|
|	Initiate Hemorrhage from Aorta	|	Injury Code:  42640 (Severity = 4)			|	30	|	900	|<span class="success">	Increase @cite garrioch2004body	</span>|<span class="success">	Slight decrease or no change at all @cite garrioch2004body	</span>|<span class="success">	Increase @cite garrioch2004body	</span>|<span class="success">	Decrease @cite garrioch2004body	</span>|<span class="success">	Decrease @cite garrioch2004body	</span>|
|	Stop Hemorrhage	|	We see an increase in BV because of pressure differentials allowing fluid transfer from tissues. This is not accounted for in the calculated value.			|	900	|	1020	|<span class="success">	Decrease @cite Chalmers1967	</span>|<span class="success">	Slight increase or no change at all @cite garrioch2004body	</span>|<span class="success">	Decrease @cite Chalmers1967	</span>|<span class="success">	Increase @cite Chalmers1967	</span>|<span class="success">	Increase @cite Chalmers1967	</span>|
|	Start IV Fluids: Saline at 100 mL/min with a 500 mL bag	|				|	1020	|	1420	|<span class="success">	Decrease @cite metoyer2016SME	</span>|<span class="success">	Increase @cite metoyer2016SME	</span>|<span class="success">	Increase @cite metoyer2016SME	</span>|<span class="success">	Increase @cite metoyer2016SME	</span>|<span class="success">	Increase @cite metoyer2016SME	</span>|


<br><center>
*Table 7c. Class 2 hemorrhage with no intervention validation results.*
</center>
|	Action	|	Notes			|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Heart Rate (/min)	|	Systolic Pressure (mmHg)	|	Diastolic Pressure (mmHg)	|	Cardiac Output (mL/min)	|	Heart Stroke Volume (mL)	|
|	------------------------	|	------------------------	------------------------	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|
|	Initiate Hemorrhage from Aorta	|	Injury Code:  42640			|	30	|	630	|<span class="success">	Increase @cite garrioch2004body	</span>|<span class="success">	Slight decrease or no change at all @cite garrioch2004body	</span>|<span class="success">	Increase @cite garrioch2004body	</span>|<span class="success">	Decrease @cite garrioch2004body	</span>|<span class="success">	Decrease @cite garrioch2004body	</span>|
|	Stop Hemorrhage	|	We see an increase in BV because of pressure differentials allowing fluid transfer from tissues. This is not accounted for in the calculated value.			|	630	|	1030	|<span class="success">	Decrease @cite Chalmers1967	</span>|<span class="success">	Slight increase or no change at all @cite garrioch2004body	</span>|<span class="success">	Decrease @cite Chalmers1967	</span>|<span class="success">	Increase @cite Chalmers1967	</span>|<span class="success">	Increase @cite Chalmers1967	</span>|


<br><center>
*Table 7d. Class 4 hemorrhage with no intervention validation results.*
</center>
|	Action	|	Notes			|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Heart Rate (/min)	|	Systolic Pressure (mmHg)	|	Diastolic Pressure (mmHg)	|	Cardiac Output (mL/min)	|	Heart Stroke Volume (mL)	|
|	------------------------	|	------------------------	------------------------	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|
|	Initiate Hemorrhage from Aorta	|	Injury Code:  41630			|	30	|	1030	|<span class="success">	Increase @cite garrioch2004body	</span>|<span class="success">	Decrease @cite garrioch2004body	</span>|<span class="success">	Decrease @cite garrioch2004body	</span>|<span class="success">	Decrease @cite garrioch2004body	</span>|<span class="success">	Decrease @cite garrioch2004body	</span>|



<br><center>
*Table 7e. Multi-compartment hemorrhage with saline administration validation results.*
</center>
|	Action	|	Notes			|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Heart Rate (/min)	|	Systolic Pressure (mmHg)	|	Diastolic Pressure (mmHg)	|	Cardiac Output (mL/min)	|	Stroke Volume (mL)	|
|	------------------------	|	------------------------	------------------------	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|
|	Initiate Hemorrhages in Spleen (Severity 4) and Leg (Severity 4)	|	Injury Codes:  42820, 44000			|	30	|	530	|<span class="success">	Increase @cite garrioch2004body	</span>|<span class="success">	Decrease @cite garrioch2004body	</span>|<span class="success">	Increase @cite garrioch2004body	</span>|<span class="success">	Decrease @cite garrioch2004body	</span>|<span class="success">	Decrease @cite garrioch2004body	</span>|
|	Reduce bleeding rate in Spleen (Severity 3)	|	Injury Code:  32820			|	530	|	590	|<span class="success">	Increase @cite garrioch2004body	</span>|<span class="success">	Decrease @cite garrioch2004body	</span>|<span class="success">	Increase @cite garrioch2004body	</span>|<span class="success">	Decrease @cite garrioch2004body	</span>|<span class="success">	Decrease @cite garrioch2004body	</span>|
|	Stop Hemorrhage in Leg	|	Injury Code:  04000			|	590	|	830	|<span class="success">	Increase @cite garrioch2004body	</span>|<span class="success">	Decrease @cite garrioch2004body	</span>|<span class="danger">	Increase @cite garrioch2004body	</span>|<span class="warning">	Decrease @cite garrioch2004body	</span>|<span class="warning">	Decrease @cite garrioch2004body	</span>|
|	Stop Hemorrhage in Spleen	|	Injury Code:  02820			|	830	|	1070	|<span class="success">	Decrease @cite Chalmers1967	</span>|<span class="warning">	Increase @cite Chalmers1967	</span>|<span class="success">	Decrease @cite Chalmers1967	</span>|<span class="success">	Increase @cite Chalmers1967	</span>|<span class="success">	Increase @cite Chalmers1967	</span>|
|	Administer 500 mL of Saline at a rate of 100 mL/min	|				|	1070	|	1570	|<span class="success">	Decrease @cite metoyer2016SME	</span>|<span class="success">	Increase @cite metoyer2016SME	</span>|<span class="success>	Increase @cite metoyer2016SME	</span>|<span class="success">	Increase @cite metoyer2016SME	</span>|<span class="success">	Increase @cite metoyer2016SME	</span>|


<br><center>
*Table 7f. Multi-compartment hemorrhage with saline and morphine administration validation results.*
</center>
|	Action	|	Notes			|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Heart Rate (/min)	|	Systolic Pressure (mmHg)	|	Diastolic Pressure (mmHg)	|	Cardiac Output (mL/min)	|	Heart Stroke Volume (mL)	|
|	------------------------	|	------------------------	------------------------	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|
|	Initiate Hemorrhages in Spleen (Severity 4) and Leg (Severity 4)	|	Injury Codes:  42820, 44000			|	30	|	530	|<span class="success">	Increase @cite garrioch2004body	</span>|<span class="success">	Decrease @cite garrioch2004body	</span>|<span class="success">	Increase @cite garrioch2004body	</span>|<span class="success">	Decrease @cite garrioch2004body	</span>|<span class="success">	Decrease @cite garrioch2004body	</span>|
|	Reduce bleeding rate in Spleen (Severity 3)	|	Injury Code:  32820			|	530	|	590	|<span class="success">	Increase @cite garrioch2004body	</span>|<span class="success">	Decrease @cite garrioch2004body	</span>|<span class="success">	Increase @cite garrioch2004body	</span>|<span class="success">	Decrease @cite garrioch2004body	</span>|<span class="success">	Decrease @cite garrioch2004body	</span>|
|	Stop Hemorrhage in Leg	|	Injury Code:  04000			|	590	|	830	|<span class="success">	Increase @cite garrioch2004body	</span>|<span class="success">	Decrease @cite garrioch2004body	</span>|<span class="danger">	Increase @cite garrioch2004body	</span>|<span class="warning">	Decrease @cite garrioch2004body	</span>|<span class="warning">	Decrease @cite garrioch2004body	</span>|
|	Stop Hemorrhage in Spleen	|	Injury Code:  02820			|	830	|	1070	|<span class="warning">	Decrease @cite Chalmers1967	</span>|<span class="warning">	Increase @cite Chalmers1967	</span>|<span class="warning">	Decrease @cite Chalmers1967	</span>|<span class="warning">	Increase @cite Chalmers1967	</span>|<span class="warning">	Increase @cite Chalmers1967	</span>|
|	Administer bolus of morphine (20 mL) 	|				|	1070	|	1080	|<span class="success">	Mild Decrease @cite Morgan2006Clinical	</span>|<span class="success">	Decrease - dose dependent @cite Morgan2006Clinical	</span>|<span class="success">	Decrease - dose dependent @cite Morgan2006Clinical	</span>|<span class="success">	Decrease @cite Morgan2006Clinical	</span>|<span class="danger">	Decrease @cite Morgan2006Clinical	</span>|
|	Administer 500 mL of Saline at a rate of 100 mL/min	|				|	1080	|	1580	|<span class="success">	Decrease @cite metoyer2016SME	</span>|<span class="success">	Increase @cite metoyer2016SME	</span>|<span class="success">	Increase @cite metoyer2016SME	</span>|<span class="success">	Increase @cite metoyer2016SME	</span>|<span class="success">	Increase @cite metoyer2016SME	</span>|

<br><center>
*Table 7g. Low severity hemorrhage with immediate infusion of tranexamic acid monitored over twelve hours (validation results).*
</center>
|	Action	|	Notes			|	Action Occurrence Time (hr)	|	Sampled Scenario Time (s)	|	Blood Volume-No Intervention	|	Expected Flow Change with TXA	|	Blood Volume-TXA Intervention	|
|	------------------------	|	------------------------	------------------------	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|
|	Hemorrhage in Right Leg (5.0 mL/min), begin TXA Infusion	|	Initiate hemorrhage			|	0	|	0	|	5401.7	|	0	|<span class="success">	5401.7 @cite kushioka2017high	</span>|
|		|				|	--	|	2	|	5013.1	|	29.3	|<span class="success">	5285.6 @cite kushioka2017high	</span>|
|		|				|	--	|	4	|	4528.9	|	48.1	|<span class="success">	5047.2 @cite kushioka2017high	</span>|
|		|				|	--	|	6	|	4107.1	|	52.9	|<span class="success">	4859.6 @cite kushioka2017high	</span>|
|		|				|	--	|	8	|	3709.5	|	55.4	|<span class="success">	4674.2 @cite kushioka2017high	</span>|
|		|				|	--	|	10	|	3355.9	|	55.2	|<span class="success">	4505.1 @cite kushioka2017high	</span>|
|		|				|	--	|	12	|	3047	|	55.6	|<span class="success">	4342.0 @cite kushioka2017high	</span>|

<br><center>
*Table 7h. Chance of survival increase when applying TXA as tested by time of survival during an extreme hemorrhage with and without intervention (validation results).*
</center>
|	Hemorrhage Survival Time (s)	|	TXA Survival Time (s)	|	Percent Change	|	Time To Death	|
|	------------------------	|	------------------------	|	------------------------	|	------------------------	|
|	4691.4	| 5341.2 | +13.9 |<span class="success">	Increase @cite roberts2013crash	</span>|


### Pericardial Effusion
The pericardial effusion scenario has a chronic effusion applied to the patient with a volume accumulation on the pericardium of 500&nbsp;milliliters. There is a decrease in stroke volume, arterial pressures, and cardiac output. This is due to increasing intrapericardial pressure leading to a reduction in end diastolic volume. The validation trends somewhat follow this same behavior. 

<br><center>
*Table 8. Pericardial Effusion validation results.*
</center>
|	Condition	|	Notes	|	Sampled Scenario Time (s)	|	Heart Rate (/min)	|	Systolic Pressure (mmHg)	|	Diastolic Pressure (mmHg)	|	Cardiac Output (mL/min)	|	Heart Stroke Volume (mL)	|	Pericardium Pressure (mmHg)	|	Pericardium Volume (mL)	|	Pulmonary Capillaries Wedge Pressure (mmHg)	|	Oxygen Saturation	|
|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|
|	Pericardial Effusion Condition	|	Volume accumulation on pericardium is set at 500 ml.	|	120	|<span class="success">	Slight increase or no change @cite spodick1998tamp	</span>|<span class="success">	Slight decrease @cite spodick1998tamp	</span>|<span class="success">	Slight decrease @cite spodick1998tamp	</span>|<span class="success">	Decrease @cite spodick1998tamp	</span>|<span class="success">	Decrease @cite spodick1998tamp	</span>|<span class="success">	Increase @cite spodick1998tamp	</span>|<span class="success">	Increase @cite spodick1998tamp	</span>|<span class="success">	Slight increase @cite spodick1998tamp	</span>|<span class="success">	NC 	</span>|

### IV Fluid Administration
IV fluid administration was validated by administering 500&nbsp;mL of saline over the course of five minutes. This action was validated qualitatively by a subject matter expert and through direct calculation. The engine output followed expected trends. The subject matter expert predicted an increasing arterial pressure and blood volume. The expected total blood volume increase was quantified by directly adding the volume of saline to the total blood volume. The actual increase in blood volume is slightly less than direct addition due to fluid shifting to the extravascular space. This shift is slightly exaggerated because the pressure differential caused by the saline administration neglects the osmotic pressure gradient. This is [known](@ref tissue-future) deficiency that will be corrected in a future release. Overall, the IV fluid administration action showed excellent agreement with the subject matter expert's predicted trends.

<br><center>
*Table 9. IV saline administration validation results.*
</center>
|	Action	|	Action Occurrence Time (s)	|	Sample Scenario Time (s)	|	Heart Rate	|	Cardiac Output	|	Diastolic Pressure (mmHg)	|	Systolic Pressure (mmHg)	|	Stroke Volume (mL)	|
|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|
|	Administering 500 mL of Saline at a rate of 100 mL/min	|	30	|	570	|<span class="success">	Minor Decrease  @cite metoyer2016SME	</span>|<span class="success">	Increase  @cite metoyer2016SME	</span>|<span class="success">	Mild Increase @cite metoyer2016SME	</span>|<span class="success">	Mild Increase @cite metoyer2016SME	</span>|<span class="success">	Increase @cite metoyer2016SME	</span>|

### Sinus Bradycardia
Sinus bradycardia was validated by setting the baseline heart rate to 50&nbsp;beats per minute. The condition was allowed to stabilize in keeping with chronic condition  methodology; for more information, see @ref SystemMethodology. The cardiac output remains relatively unchanged due to compensatory increases in the stroke volume, which is expected in asymptomatic sinus bradycardia @cite mangrum2000evaluation. The validation data that was found for  systolic and diastolic pressures was vague, only mentioning that the condition was often asymptomatic, indicating relatively normal pressure values. The systolic and diastolic pressures in %BioGears do change slightly as the heart driver accommodates the increased stroke volume; however, the values remain within normal bounds at 50&nbsp;beats per minute. The %ECG waveform in bradycardia is similar to a normal sinus waveform, with the exception of an extended R-R interval due to a slower heart rate (see Figure 12), which is also seen in the %BioGears output. Other systemic data is not significantly changed.

@image html SinusBradycardia_ECGPedia.png
@image html SinusBradycardia_BioGears.PNG
<center>
<i>Figure 13. The increased R-R interval is evident in both waveforms. This is the primary indication of the low heart rate. Validation image courtesy of @cite vanderBilt2010sinus .</i>
</center><br>

<br><center>
*Table 10. Bradycardia validation results.*
</center>
|	Condition	|	Notes	|	Occurrence Time (s)	|	Sample Scenario Time (s)	|	Heart Rate (beats/min)	|	Mean Arterial Pressure (mmHg)	|	Respiration Rate (breaths/min)	|	Oxygen Saturation (mmHg)	|	Cardiac Output(mL/min)	|	Stroke Volume (mL)	|	ECG Output (mV)	|
|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|
|	Bradycardia	|	Heart Rate set to 50 beats per minute (<60 considered bradycardia)	|	0	|	120	|<span class="success">	50 @cite mangrum2000evaluation	</span>|<span class="warning">	Decrease @cite klabunde2012cvconcepts	</span>|<span class="warning">	No Relevant changes @cite mangrum2000evaluation	</span>|<span class="success">	No Relevant changes @cite mangrum2000evaluation	</span>|<span class="success">	Decrease @cite klabunde2012cvconcepts	</span>|<span class="success">	Increase @cite mangrum2000evaluation	</span>|<span class="success">	Sinus @cite mangrum2000evaluation	</span>|

### Sinus Tachycardia
Sinus Tachycardia was validated by setting the baseline heart rate to 110&nbsp;beats per minute. The condition was allowed to stabilize in keeping with chronic condition  methodology; for more information, see @ref SystemMethodology. Validation data predicted a decrease in stroke volume with the increase in heart rate @cite aroesty1985simultaneous, which was also found in the %BioGears engine. Because of the decrease in stroke volume, the cardiac  output remains relatively unchanged. This is due to the model not currently affecting compression force, only compression rate. The %ECG output in tachycardia is generally similar to normal sinus; however, in some cases, the T wave can experience constructive  interference from the following heart beat&apos;s P wave. This is shown in the %BioGears %ECG output seen in Figure 13.

@image html SinusTachycardia_Physionet.png
@image html SinusTachycardia_BioGears.PNG
<center>
<i>Figure 14. Due to the high heart rate, the %BioGears output is summing together the P and T waves. In the image from PhysioNet, the output is not summed together as dramatically, due to the slight physiological compression of the waveform that the current %ECG system and heart model do not support. @cite healey2005detecting @cite goldberger2000physiobank</i>
</center><br>

<br><center>
*Table 11. Tachycardia validation results.*
</center>
|	Condition	|	Notes	|	Action Occurrence Time (s)	|	Sample Scenario Time (s)	|	Heart Rate(beats/min)	|	Mean Arterial Pressure (mmHg)	|	Respiration Rate (breaths/min)	|	Oxygen Saturation (mmHg)	|	Cardiac Output(mL/min)	|	Stroke Volume (mL)	|	%ECG Output (mV)	|
|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|
|	Tachycardia	|	Variable.Set Heart Rate = 110 beats per minute; in Aroestyl, patients were not ideal, a number had previous heart conditions or coronary artery disease. Tachycardia was achieved via atrial pacemakers	|	0	|	120	|<span class="success">	110 @cite yusuf2005deciphering	</span>|<span class="warning">	13% increase @cite still2005prevalence	</span>|<span class="warning">	NC @cite yusuf2005deciphering	</span>|<span class="success">	NC @cite yusuf2005deciphering	</span>|<span class="success">	Dependent on SV and HR @cite sohn2007hemodynamic	</span>|<span class="success">	Decreases as Heart Rate increases @cite aroesty1985simultaneous	</span>|<span class="success">	Normal sinus/Tachycardia Waveform 	</span>|

### Anemia
The anemia condition reduces the oxygen carrying capacity of the blood. The anemia validation results are shown in Table&nbsp;12, and were in excellent agreement with literature.

<center>
*Table 12. Anemia validation results.*
</center>
|	Condition	|	Notes			|	Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Hemoglobin Concentration (g/dL)	|	Heart Rate (/min)	|	Cardiac Output (L/min)	|	Stroke Volume (mL/beat)	|	Oxygen Saturation %	|	Respiration Rate (/min)	|	Hematocrit %	|	Systemic Vascular Resistance (mmHg*min/L)	|
|	------------------------	|	------------------------	------------------------	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|	------------------------	|
|	Anemia	|	30% severity			|	0	|	120	|<span class="success">	10.3 as determined by initial hemoglobin and severity [set by severity]	</span>|<span class="success">	Approximately no change from normal @cite duke1969hemodynamic	</span>|<span class="success">	Slight increase from normal with increase in stroke volume @cite duke1969hemodynamic	</span>|<span class="success">	Slight increase with stroke volume @cite duke1969hemodynamic	</span>|<span class="success">	94.5 @cite duke1969hemodynamic; Should not affect oxygen saturation, especially at our severities @cite moss2016SME	</span>|<span class="success">	NC until exertion @cite toy2000fatigue	</span>|<span class="warning">	Decrease as determined by severity and initial hemoglobin @cite duke1969hemodynamic	</span>|<span class="success">	Decrease @cite guyton2006medical	</span>|

@anchor cardiovascular-conclusions
Conclusions
-----------
The %BioGears CV System uses circuit methodology to simulate blood flow and physiological connection to other systems within the human body. This system provides accurate results for resting physiology and allows for the simulation of a variety of CV-related insults and interventions with appropriate system feedback. Circuit modeling of the CV System is both quick and effective, and the %BioGears implementation allows the user to easily modify the circuit to vary the model resolution and fidelity or integrate their own models.

@anchor cardiovascular-future
Future Work
===========

Recommended Improvements
------------------------

An area of potential future advancement lies in the inertance of the %Cardiovascular System. The %BioGears @ref CircuitMethodology has the ability to incorporate inertance into the lumped parameter models. In the future, this could be added to the %Cardiovascular Model to provide a more accurate blood pressure waveform.

Another potential area for improvement is simulation of a tourniquet. An intervention could be added to simulate the increased resistance to flow and external pressure application that would be present with the application of a tourniquet.

The cardiac arrest functionality also needs improvement. Like in the human body, most of the %BioGears systems require a beating heart to function properly. Also like in the human body, the %BioGears systems tend to go haywire when the heart stops. However, the ways in which the %BioGears systems go haywire deviate from the human physiological systems' response during cardiac arrest. Improvements to the engine functionality during cardiac arrest would allow for many desirable scenarios, including Advanced Cardiac Life Support (ACLS) scenarios where some recovery of the patient is actually possible. As described in the CPR section, recovery from cardiac arrest is currently impossible in the %BioGears engine. Rescue breathing would also be a valuable improvement.

@anchor cardiovascular-appendices
Appendices
==========

Acronyms
--------
AHA - American Heart Association

CFD - Computational Fluid Dynamics

CPR - Cardiopulmonary Resuscitation

CV - %Cardiovascular

%ECG - Electrocardiogram

MAP - Mean Arterial Pressure

MCIS - Military Injury Combat Scale

mL - Milliliters

mmHg - Millimeters mercury

Data Model Implementation
-------------------------
@ref CardiovascularSystemTable "Cardiovascular"

@ref ElectroCardioGramTable "ElectroCardioGram"

@ref TissueSystemTable "Tissue"

@anchor cardiovascular-compartments
Compartments
------------
- Aorta
- Heart
 - Myocardium
 - LeftHeart
 - RightHeart
 - Pericardium
- VenaCava

- PulmonaryArteries
- PulmonaryCapillaries
- PulmonaryVeins
- Lungs
 - LeftLung
  - LeftPulmonaryArteries
  - LeftPulmonaryCapillaries
  - LeftPulmonaryVeins
 - RightLung
  - RightPulmonaryArteries
  - RightPulmonaryCapillaries
  - RightPulmonaryVeins

- Kidneys
 - LeftKidney
  - LeftRenalArtery
  - LeftNephron
   - LeftAfferentArteriole
   - LeftGlomerularCapillaries
   - LeftEfferentArteriole
   - LeftPeritubularCapillaries
   - LeftBowmansCapsules
   - LeftTubules
  - LeftRenalVein
 - RightKidney
  - RightRenalArtery
  - RightNephron
   - RightAfferentArteriole
   - RightGlomerularCapillaries
   - RightEfferentArteriole
   - RightPeritubularCapillaries
   - RightBowmansCapsules
   - RightTubules
  - RightRenalVein

- Bone
- Brain
- Fat
- Gut
 - Splanchnic
 - SmallIntestine
 - LargeIntestine
- Liver
- Spleen
- Skin
- Muscle
- LeftArm
- LeftLeg
- RightArm
- RightLeg
- Ground
Anesthesia Machine Methodology {#AnesthesiaMachineMethodology}
==========================

@anchor anesthesia-overview
# Overview
@anchor anesthesia-abstract
## Abstract

The %BioGears Anesthesia Model is a generic representation of mechanical ventilation and 
inhaled agent administration. It models a semi-closed circuit breathing system. The 
Anesthesia System employs a pressure-control ventilation mode for positive-pressure 
ventilation and gas delivery. The model uses a constant ventilator pressure source to drive 
inspiration and release pressure during expiration.
This model provides an important system required for the training 
and simulation of
a complex medical field. 
A variety of common equipment failures associated with 
anesthesia machines are available with this system. The results of this system were 
qualitatively evaluated for all common failures
modeled. The results show an excellent correlation with the expected trends. 
Future work will address the current limitations of the system and development 
of a more robust Anesthesia Machine.
@anchor anesthesia-intro
## Introduction

### Anesthesia Machines and Positive-Pressure Ventilation

An anesthesia machine is used to both control a patient's gas exchange through mechanical ventilation, or positive-pressure ventilation, and administer inhaled anesthetic agents. These machines are highly 
sophisticated, and they require training and problem-solving skills for appropriate use and avoidance of user- and equipment-related failures. While there are relatively few companies manufacturing 
anesthesia machines, there is a large variety in models and types of machines. This provides a wide range of complexity and functionality that is dependent on a facility's expectations and requirements. 
The administration of anesthesia also requires the delivery of multiple drugs that are chosen based on the individual patient's condition and the procedure type and duration. The combination of equipment and drug 
administration makes anesthesiology a difficult discipline that requires significant training to avoid failures that could lead to patient injury or death. @cite roth2007anesthesia @cite morgan2006clinical

In general, anesthesia machines contain the following basic components @cite roth2007anesthesia @cite morgan2006clinical :

-   Compressed gas cylinders - These contain oxygen and/or nitrogen for portable use with the machine. Oxygen may also be supplied through a wall port.

-   Vaporizer - Inhaled agents are inserted into the machine in liquid form and must be aerosolized for patient administration.

-   Unidirectional valves - Valves open and close during the inspiratory and expiratory cycles to prevent rebreathing of carbon dioxide.

-   Carbon dioxide absorber - Carbon dioxide is scrubbed from the anesthesia machine circuit using different absorbers.

-   Ventilator - The bellows supplies gas flow to the patient.

-	Tubing - All equipment is connected through tubing that can become tangled or develop leaks if not monitored regularly.

Airway management is also associated with a number of components that provide air to the patient depending on the patient's condition and surgery specifics. These components include the following @cite stackhouse2007airway :

-   Mask - A mask can be placed on the patient's face to supply supplemental oxygen or an inhaled anesthetic. 

-   Endotracheal tube (ETT) - This tube can be placed in the trachea through the intubation process for positive-pressure ventilation for a fully sedated patient. These come in a variety of styles and sizes
    that must be chosen based on the patient.

-   Laryngeal Mask Airway (LMA) - This device has become common for patients with a difficult airway. It combines a silicone rubber mask with a flexible shaft to provide positive-pressure ventilation to a fully
    sedated patient without requiring intubation. 

The large number of components, the sophisticated computer controls, and the high variability in patient needs can lead to user errors with anesthesia equipment. Significant pre-surgery planning is required to ensure an appropriate airway plan is in place. Equipment failures are not uncommon with anesthesia machines. These failures range from minor tube leaks and tangles to a failure of the ventilator to provide adequate pressure
for respiration. @cite morgan2006clinical
	
The %BioGears Anesthesia Machine Model is designed to provide a generic anesthesia machine capable of simulating mechanical ventilation and the administration of inhaled anesthetic agents. This model is capable of powering training simulations regarding the use of anesthesia equipment. In combination with the rest of the %BioGears Engine, the model provides a whole-body approach promoting training on drug administration, airway
management, maintenance of anesthesia, transitions from negative- to positive-pressure ventilation, and diagnosis and problem-solving related to equipment failures. 
@anchor anesthesia-system
# System Design
@anchor anesthesia-background
## Background and Scope

### %BioGears Anesthesia Machine History

The %BioGears Anesthesia Machine has its origin in the work of Dr. N. Ty Smith and colleagues at the University of California at San Diego. Their
work on this subject can be traced back to a series of cardiopulmonary and drug distribution models that they developed @cite Smith1972multiplemodel , @cite Zwart1972analog . For
example, they developed multicompartment physiology and pharmacology models describing the uptake and distribution of halothane
@cite FukuiSmith1981hybrid . The early models of the group employed analog and hybrid computers. Later, they developed models on standard personal computers that formed
a basis for the screen-based simulator originally known as SLEEPER. This program eventually evolved into an advanced anesthesia simulator
called BODY simulation that was distributed by the Advanced Simulation Corporation. The BODY simulation program consisted of cardiovascular, respiratory, 
and drug models that formed the basis for the HumanSim physiology engine, of which the %BioGears Engine is an extension.  The major components that control the gas flow and delivery of the anesthetic drugs are derived from the HumanSim engine. Various modifications, including the integration of the Anesthesia System as part of the closed circuit-based common data model, led to the current version of the %BioGears Anesthesia Delivery System. As it stands currently, the Anesthesia System is under development and its use has some restrictions as described in the Features section below. Future releases will provide a more rigorously tested Anesthesia System that is fully integrated with the respiratory and drug systems.
@anchor anesthesia-flow
## Data Flow

### Preprocess

The ventilator and valves operate via time-based cycles based on the settings. There are four phases to each complete cycle, as shown in Figure 1.

<center><img src="./images/AnesthesiaMachine/Cycles.png" width="550"></center>
<center><i>Figure 1. Flow diagram showing the ventilatory cycles as implemented in the %BioGears %Anesthesia Machine.</i></center> 

#### Process Actions

Process Actions deals with modification of the anesthesia machine
parameters and circuit properties due to actions (insults) as
specified by the user. The insults are described in the Actions section
below.

#### Update Resistances

The breathing circuit of the commonly used Anesthesia Machine consists of unidirectional valves
on the inspiratory and expiratory limbs that control the flow of gases during the inspiration and 
expiration phases. The %BioGears Anesthesia Machine models the unidirectional valves using
resistors and logical on/off conditions that serve as switches. These resistors are updated depending 
on the phase of the respiration cycle. During the inspiratory phase of the breathing cycle, the resistor representing the
unidirectional valve on the inspiratory limb is lowered to a pre-set value (an analogue to a closed electrical circuit), opening it to flow.  At the same time, the 
unidirectional valve at the expiratory limb is closed to flow by employing a very high resistance (an open circuit). The process is reversed during the expiratory phase of 
the breathing cycle.  In addition to being modified during normal breathing cycles, the valve resistances are also adjusted to
model various equipment failures such as inspiratory/expiratory valve leaks and obstructions. 

#### Ventilator Pressure Calculation

In the use of the Anesthesia Machine, there are two basic ventilation modes for precise control of ventilation. These are the pressure controlled and volume-controlled modes.
In the volume-controlled mode, the Anesthesia Machine delivers a preset tidal volume 
to generate ventilation. In this delivery mode, positive-pressure ventilation is delivered at a
predetermined breathing frequency and tidal volume to provide predictable minute
ventilation. In the pressure-controlled ventilation mode, a set plateau (control) pressure  is 
delivered at each breathing 
cycle. The tidal volume delivered to the patient is derived from the control ventilator pressure and the
 lung compliance of 
the patient. The current version of the %BioGears Anesthesia Machine employs the pressure-controlled 
ventilator. During each inspiratory phase, a constant control pressure that
serves as a driving pressure source is applied to the ventilator. This pressure drives flow across 
the inspiratory limb path to supply gas to the patient. During the expiratory phase, the ventilator
driver pressure drops, allowing gas to return to the ventilator. In the
engine, the volume of the ventilator is adjusted based on the gas flow into the ventilator node. 

#### Gas Inlet Flow Calculation

The gas inlet is modeled as a flow source to add substances in the form of air and/or inhaled drugs to the system. The inlet flow is set by the total
flow parameter, and the volume fractions are based on gas source
concentrations.

#### Breathing Cycle Calculation

The %BioGears Engine calculates the inspiratory and expiratory phase times based on 
a preset respiration rate and inspiration-expiration ratio parameters that are selected as 
inputs for the Anesthesia Machine configurations. 

### Process

The current %BioGears implementation has no specific circuit or transport process
functionality for the anesthesia machine. Anesthesia Machine processing
is currently done in the %Respiratory System with the combined circuit
methodology.  However, the substance volume fractions and volumes are updated to remove carbon dioxide in the scrubber during this step.

### Postprocess

The Postprocess step moves values calculated in the Process step from the next
time step calculation to the current time step calculation. The current %BioGears has no 
specific postprocess functionality for the anesthesia machine. All postprocessing is done in the
%Respiratory System with the combined circuit methodology.

The following figure presents the data flow of the %Anesthesia System data processing steps.

<center><img src="./images/AnesthesiaMachine/AnesthesiaMachineDataFlow.png" width="550"></center>
<center><i>Figure 2. The Anesthesia Machine system has three processing steps at the highest level. Each time step, Preprocess prepares the circuit to be calculated; Process performs the calculation to determine the entire circuit state; and Postprocess advances time.</i></center>    
<br>
@anchor anesthesia-features
## Features and Capabilities

### The %BioGears Anesthesia Machine Circuit

The %BioGears %Anesthesia Machine System represents a generic mechanical ventilation system. 
It models a semi-closed anesthesia breathing circuit that employs a pressure-control 
mode of positive-pressure ventilation. The Model is designed to represent the
functions of the ventilator, breathing circuit, and vaporizer typical of an
Anesthesia Delivery System. To incorporate these functions, the
Anesthesia Machine utilizes seven major compartments. These
compartments correspond to the three flow tubes (the Y-piece and the
inspiratory and expiratory tubes), the CO2 absorber,
the fresh gas inlet of the breathing circuit, the ventilator, 
and the mask compartments. The fresh gas inlet allows the flow of 
volatile anesthetic medications and fresh gas into the breathing circuit. 
The vaporizer functionality is incorporated 
by adjusting the concentration of the gas mixture delivered through the 
fresh gas inlet. In anesthesia machine systems, vaporizers are designed to hold 
only one specific volatile anesthetic medication. The %BioGears Anesthesia 
Model reproduces this constraint by employing conditions that represent
left and right chambers, each of which hold
a specific inhaled agent. The model treats the agents in these chambers as vapors and 
mixes them with other gases to be delivered into the breathing circuit, adjusting the total gas mixture accordingly.  In particular, the model deducts the volume
fraction of the nitrogen gas proportional to the oxygen and/or inhaled volatile gases concentration. 

In this way, the %BioGears Anesthesia Machine captures the effect of the vaporizer functionality. 
For supplemental oxygen delivery, the model offers a separate wall port or an oxygen bottle as options. The model
allows for two oxygen bottles that can be used sequentially, depending on the length of 
anesthesia delivery and level of oxygen depletion. In the event of endotracheal
intubation, the mask compartment is replaced by the endotracheal tube
compartment. These two compartments are used interchangeably. The
inspiratory and expiratory flow tube compartments are regulated by
unidirectional valve switches and allow gas flow into and out of the
lung. The ventilator compartment supports positive-pressure ventilation for oxygen delivery and ventilation. Finally, 
the CO2 absorption compartment provides critical functionality by reducing the amount of
inhaled CO2 in re-breathed gas. The diagram in Figure 3 presents the
compartmental view of the Anesthesia Delivery System.

<center><img src="./images/AnesthesiaMachine/AnesthesiaMachineGasFlowDiagram.png" width="650"></center>
<center>
*Figure 3. Compartmental view of the Anesthesia System. The %BioGears Anesthesia 
Machine consists of seven major compartments, as shown in the breathing circuit. 
These are the mask/endotracheal tube compartment, the Y-piece, the inspiratory 
and expiratory limbs, the fresh gas inlet, the CO2 absorber, 
and the ventilator compartments.  The model also includes a relief valve path that emulates 
the exhaust scavenging interface.  
The mask and the endotracheal tube compartments are interchangeable. 
The unidirectional 
switches shown on the inspiratory and expiratory limbs permit unidirectional airflow 
in the patient's breathing cycle. The connecting lines represent the flow of gas between 
compartments and 
the arrows show the flow direction. The red lines and the associated arrows in the figure 
show flow direction during the inspiratory phase.  The %BioGears
Anesthesia Machine also interacts with the %Respiratory System through a circuit 
path that links the two systems at the airway node.*
</center><br>

Compartments in the Anesthesia System are modeled as a network of
resistors to account for the resistance to flow through each compartment. 
The ventilator is designed to accumulate volume on its node based on the flow 
across its path. When mechanical ventilation is not
invoked, the model switches to spontaneous breathing via negative
pressure ventilation that is driven by the respiratory muscle pressure.

The gas flow in the compartments is driven
by pressure gradients across the resistors
in the breathing circuit. The flow resistances in the breathing circuit are primarily due to the the unidirectional valves and other regions of constriction.  Since practically
all of the pressure drops occur across these resistors, the circuit diagram
(Figure 4) representing the %Anesthesia %Machine is a condensed version of the compartmental
diagram (Figure 3). A single node encapsulates both the mask and Y-piece, as well as a current source which can be used to model a leak at the mask. The exhaust node can represent either a leak to the atmosphere or a scavenging system that relieves pressure build up. Normally, 
the exhaust is connected to a breathing bag reservoir; however, the current version
lacks such a reservoir. Instead, the exhaust in the circuit diagram serves as a current 
source that facilitates appropriate pressure maintenance. 

@image html AnesthesiaMachineCircuitDiagram.png
<center>
*Figure 4. Circuit diagram of the Anesthesia Machine. The circuit 
employs the ventilator as a driver pressure source. 
One of the two pressure sources is selected using the switch. 
The fresh gas inlet serves as a current source that drives fresh gas and anesthetic drugs 
into the breathing circuit. When the Anesthesia Machine is turned on, the switch between 
the mask (represented as AnesthesiaConnection) and the mouth node of the 
%Respiratory System becomes closed. The switch between the mask and the mouth node remains 
open when the Anesthesia Machine is turned off. The unidirectional valves of 
the Anesthesia Machine are handled by assigning very high resistance on the 
inspiratory or expiratory limbs depending on the breathing phase.*
</center><br>

### Connecting to the %Respiratory Circuit

When an anesthesia machine is used on a patient, there is a direct
connection that allows air to flow freely between them. In the same
fashion, the Anesthesia Machine and %Respiratory circuits in %BioGears are
directly connected and allowed to share the same fluid. When the machine
is turned on, both individually defined circuits are combined into a
single circuit that is then used for calculations.

### The %BioGears Anesthesia Machine Settings

Anesthesia machines operate in a specific mode for positive-pressure
delivery. As discussed above, there are two modes of ventilation commonly used.
The %BioGears Anesthesia Machine employs a pressure-controlled ventilator
for positive-pressure ventilation. 

To achieve this functionality, a set of parameters are pre-defined to control the Anesthesia Machine configuration. The %BioGears
Anesthesia Machine assigns the ventilator pressure,
respiration rate, inspiration-to-expiration ratio, relief valve
pressure, and positive end expired pressure as input parameters. 
Additionally, the %BioGears model assigns a set of parameters to control the fresh gas
flow and the various gas fractions delivered to the breathing circuit.
Table 1 lists the Anesthesia Machine configuration settings
currently used in the engine. It should be noted that the current
version of the Anesthesia Machine Model is restricted to the values
listed in the table until a wider range of values are tested. Future
versions will allow user-defined Anesthesia Machine settings.

<center>
*Table 1. Inputs to Anesthesia Machine*
</center>
|	Parameter	|	Description 	|
|	---	|	---	|
|	<b>General Anesthesia Machine Settings</b>	|		|
|	State (on/off)	|	Determines if the anesthesia machine is running or not	|
|	VentilatorPressure	|	The maximum ventilator pressure applied each cycle	|
|	RespiratoryRate	|	Frequency of the ventilator	|
|	PositiveEndExpiredPressure	|	The minimum ventilator pressure applied each cycle	|
|	InspiratoryExpiratoryRatio	|	The relative ratio of the period of positive and negative flow in and out of the lungs	|
|	ReliefValvePressure	|	The pressure at which air will be vented out of the machine - pressure will not be able to go above this	|
|	<b>Gas Inlet Settings</b>	|		|
|	InletFlow	|	The quantity of flow from the Oxygen Source	|
|	OxygenFraction	|	Sets the volume fraction of oxygen in delivered mixture	|
|	PrimaryGas (Nitrogen/Air)	|	Sets a selection for primary gas, such as air (ambient) or nitrogen	|
|	OxygenSource (Bottle 1/Bottle 2/Wall)	|	Sets the source of the gas composition to be used	|
|	OxygenBottleOneVolume	|	Sets the volume of the first  oxygen Source tank	|
|	OxygenBottleTwoVolume	|	Sets the volume of the second  oxygen Source tank	|
|	<b>Related Settings</b>	|		|
|	AirwayMode (Free/Mask/Tube)	|	Used to add or remove the mask or tube via Actions	|
|	SubstanceAdministration	|	Substances are added to the vaporizer (i.e. gas source) by specifying a substance and chamber (left or right)	|

@anchor anesthesia-example
#### Example Scenario: Standard Settings
All BioGears anesthesia machine scenarios assume that the patient is fully unconsious and requires total ventilation support (see [Assumptions/Limitations](@ref anesthesia-assumptions).  Full patient sedation is accomplished by administering a sedative from the BioGears Drug library, such as propofol or midazolam.  A neuromuscular blocker such as succinylcholine or rocuronium can be co-administered to simulate situations in which the airway muscles must be forcibly relaxed.  Our demonstration uses an induction dose of 1.8 mg/kg of propofol followed by a maintenance propofol infusion of 140 ug/kg/min (@cite becker2012pharmacodynamic).  When the patient is deeply sedated, we intubate and configure the anesthesia machine with the following settings:
		- Ventilator Pressure = 10 cmH2O
		- Positive End Expired Pressure = 1 cmH2O
		- Inspiratory:Expiratory Ratio = 0.5
		- Respiratory Rate = 12 /min
		- Inlet Flow = 5.0 L/min
		- Oxygen Fraction = 0.25
		- Primary Gas = Nitrogen

We maintain the anesthesia machine for 45 minutes (to simulate some procedure), discontinue the propofol infusion, wait 15 minutes for the sedation to begin to wear off, and then take the patient off the ventilator.  Figure 5 shows the results of this simulation.

<center>
<table border="0">
<tr>
    <td><img src="./plots/AnesthesiaMachine/AnesthesiaDemo_RASS.jpg" width="550"></td>
    <td><img src="./plots/AnesthesiaMachine/AnesthesiaDemo_Propofol.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/AnesthesiaMachine/AnesthesiaDemo_RespirationRate.jpg" width="550"></td>
    <td><img src="./plots/AnesthesiaMachine/AnesthesiaDemo_Ventilation.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/AnesthesiaMachine/InspiratoryValveObstructionLegend.jpg" width="1100"></td>
</tr>
</table>
</center>
<center><i>Figure 12. Select outputs from the inspiratory valve obstruction scenario.</i></center>

@anchor anesthesia-dependencies
### Dependencies

The %BioGears Anesthesia Machine interacts with 
the %Respiratory System, delivering gases and anesthetic drugs into the
%Respiratory System (see @ref RespiratoryMethodology). The two 
systems are connected to each other via a path between the mouth node of the
 %Respiratory System and the mask/endotracheal node (AnesthesiaConnection 
in Figure 4) of the anesthesia
system. During spontaneous ventilation, the mouth node of the
%Respiratory System is connected to the atmosphere via the %Environment System. 
This serves as a ground node for the %Respiratory System. 

When the Anesthesia Machine is turned on, a network of combined circuits that
includes the elements from both the %Respiratory and Anesthesia Systems is
created. When the combined circuit is generated at runtime, the
ground environment node connected to the mouth node of the %Respiratory System is
replaced by the AnesthesiaConnection node that represents the mask/endotracheal node.
As a result, the %BioGears Anesthesia Delivery System interacts with the flow resistances of the %Respiratory System. In
this regard, the ventilator driver pressure serves as a positive-pressure source for 
the combined circuit, allowing the Anesthesia System to manage pressure-controlled ventilation.
To provide air for the breathing circuit, the Anesthesia Machine is linked to the %Environment 
System, which regulates the gas concentration and atmospheric pressure. The %BioGears engine
%Anesthesia System has system data that allows for selection of the primary gas (air, 
nitrogen, etc.). Once the primary gas is selected, the delivery of supplemental 
oxygen and/or anesthetic gases is adjusted according the composition and fraction of gases derived from the 
%Environment System (see @ref EnvironmentMethodology).
@anchor anesthesia-assumptions
## Assumptions and Limitations

The %BioGears Anesthesia Machine is a generic model representing components crucial to the administration of 
inhaled agents and mechanical ventilation. This system is intended to provide
a model that can simulate the response of the body to positive-pressure ventilation, 
drug administration, and equipment failures. The model employs a pressure-controlled, positive-pressure ventilation mode and sets
a constant ventilator pressure to drive inspiration.  

Currently, there are two limitations with the model. First, the model does not allow 
manual breathing using a bag reservoir as a pressure source. Secondly, the current version of 
the model sets a constant respiration rate consistent with pressure-controlled or volume-controlled ventilators. 
This modeling choice is necessary for modeling a fully paralyzed patient. However, for anesthetic 
cases that do not require complete muscle paralysis, a synchronized ventilation that allows for spontaneous breathing would be necessary.
Future efforts may be directed towards rectifying this limitation.
@anchor anesthesia-actions
## Actions

The %BioGears Anesthesia Machine can model a variety of different equipment
failures. Breathing circuit disconnection is a leading cause of critical
incidents in anesthesia @cite roth2007anesthesia @cite morgan2006clinical .
Disconnections, which are effectively leaks, can be complete or partial. Some of these
failures are modeled in %BioGears by changing a resistance value based on
the severity. Figure 6 shows the resistance value that is set based on
the insult severity.  This mapping will be referenced with respect to the actions in this
section. We have chosen a resistance of 1000 cmH2O-s/L to be associated with a fully open 
switch (closed valve) and a resistance of 0.001 cmH2O-s/L to represent a fully 
closed switch (open valve). Intermediate resistances map logarithmically between those two
extents. Several insults modify the same resistance that leads to ground
(atmosphere) because the change in flow produced by the ventilator
reaches the patient in the same way.

@image html AMSeverityMap.png

<center>
*Figure 6. This plot shows the resistance value that is set based on the
leak severity modifier. The severity is mapped logarithmically in a
manner that allows the highest severity to allow virtually all flow through
the path and the lowest severity to allow almost no flow through the
path. Obstructions use the same function with the severity reversed
(i.e., 1-severity).*
</center><br>

### Connection Leaks

The following three types of leaks are all implemented via the same strategy in the Anesthesia Machine code. The mask/tube leaks and the Y-piece disconnect can be called simultaneously, and the severities will be combined.

#### Mask Leak

One of the most common equipment failures in the anesthetic delivery system
is a mask leak. If a mask is being used on an Anesthesia Machine and is improperly
secured or has damage, leaks can occur. The %BioGears Engine models mask
leaks by varying a resistance based on the severity in the manner previously
described and shown in Figure 6.

#### Tube Leak

Endotracheal tube cuff leaks are often caused by structural defects in the endotracheal tube, 
but may also arise from improper placement or a failure to fully inflate the cuff.  Large leaks can lead to inadequate ventilation. The %BioGears Engine models endotracheal tube 
leaks by varying a resistance based on severity
in the manner previously described and shown in Figure 5.

#### Y-Piece Disconnect

The most common disconnection site is at the Y-piece @cite roth2007anesthesia @cite morgan2006clinical .
The %BioGears Engine models endotracheal tube leaks by varying a resistance based on the severity in 
the manner previously described and shown
in Figure 6. In the case of both a tube and Y-piece leak, the severities are combined and bounded by an upper value of one.

### Valve Leaks

Inspiratory and expiratory valves leaks are similar to endotracheal tube cuff leaks. 
The valves are modeled as resistors that alternate
between very high and very low resistance values to mimic a switch opening and closing 
every breath cycle. This prevents rebreathing of carbon 
dioxide during the respiration cycle. If there is a leak present, the valve closed (i.e., high resistance) 
value is modified based on the severity
in the manner previously described and shown in Figure 6.

### Valve Obstructions

Anesthesia Machine tracheal tubes can become kinked, and hoses throughout the breathing circuit are subject to occlusions by external
mechanical forces that can impinge the flow @cite roth2007anesthesia @cite morgan2006clinical .
%BioGears models both inspiratory and expiratory valve obstructions similarly to leaks, by modifying the valve resistance values. If there is an
obstruction present, the valve open (i.e., low resistance) value is modified based on the severity in the manner previously described and shown
in Figure 6.

### Ventilator Pressure Loss

Both piston and bellows ventilators can sometimes experience mechanical failures that prevent them from filling completely. This causes the driving pressure of the 
ventilator to be less than the target. %BioGears uses a ventilator failure severity to linearly reduce the pressure produced by the ventilator, i.e., the pressure source value. The
pressure source can range from the full calculated value to a zero pressure. Failure in ventilator pressure may affect the amount of tidal volume delivered to the patient, which 
in turn reduces the amount of fresh gas delivered to the body, affecting the oxygen saturation level and other physiological parameters.

### Soda Lime Failure

Re-breathing circuits direct expired gases through soda lime granules to remove CO2. To ensure that a soda lime canister (CO2 absorber) is
functioning properly, it must be replaced regularly. If it is not monitored and replaced, the absorber will become less effective or completely unable
to remove CO2 from the breathing circuit. The %BioGears Anesthesia Machine models soda lime failure by reducing the amount of CO2 removed. The severity
linearly decreases the amount of CO2 removed.

### Vaporizer Failure

Vaporizers in anesthesia machines convert volatile anesthetic medications from a liquid state to a vapor at a predetermined concentration for patient delivery. Anesthesia Machine 
vaporizers can fail due to as leaks in seals and O-rings or electronics failures. The %BioGears Engine models vaporizer
failures by reducing the inhaled agent volume fraction at the vaporizer flow source linearly with the severity.

### Oxygen Supply Failures

Supplemental oxygen is supplied to the Anesthesia System either from the wall source or an oxygen tank bottle. 
Many scenarios, such as problems with the oxygen pipeline, can cause disruption of gas delivery. To identify these issues, 
pressures in the gas supplies are often checked. The %BioGears Anesthesia System attempts to capture such 
equipment failures by triggering oxygen wall pressure loss and oxygen tank pressure loss conditions. When the
oxygen pressure loss conditions are triggered, the model
removes the supplemental oxygen delivered to the gas flow inlet by setting the oxygen inlet volume 
fractions to zero.
@anchor anesthesia-events
Events
------

### Oxygen Bottle Exhausted

The %BioGears %Anesthesia system triggers the oxygen bottle tank exhausted  event to alert 
the user of the depletion of each of the oxygen bottles when they are selected as a source of 
supplemental oxygen. 

### Relief Valve Active 

The %BioGears %Anesthesia system triggers the relief valve active event to
alert the user when the relief valve is closed (in the electric analogue sense), 
and a condition the alerts the user when the pressure has reached the threshold value. 
@anchor anesthesia-results
# Results and Conclusions
@anchor anesthesia-settingsvalidation
## Validation - Settings

The Anesthesia Machine Settings are fully dynamic and do not have any bounds enforced.  A scenario that varies these settings in several different combinations is included with the %BioGears deployment and produces the outputs shown in Figure 7.  This scenario also tests the relief valve functionality and causes the active event to be logged when the Ventilator Pressure is set too high.

<center>
<table border="0">
<tr>
    <td><img src="./plots/AnesthesiaMachine/AnesthesiaMachineVariedConfiguration_VentilatorVolume.jpg" width="550"></td>
    <td><img src="./plots/AnesthesiaMachine/AnesthesiaMachineVariedConfiguration_TotalLungVolume.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/AnesthesiaMachine/AnesthesiaMachineVariedConfiguration_RR.jpg" width="550"></td>
    <td><img src="./plots/AnesthesiaMachine/AnesthesiaMachineVariedConfiguration_TidalVolume.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/AnesthesiaMachine/AMVariedLegend.jpg" width="1100"></td>
</tr>
</table>
</center>
<center><i>Figure 7. These plots show the successful implementation of various Anesthesia Machine settings. The first region has the machine connected and provides supplemental Oxygen only. The other three regions follw the administration of Succinylcholine provided to stop unassisted breathing. The upper plot shows compartment data from the ventilator and relief valves. The relief valve is shown to be active by the blue flow plot when the ventilator pressure is set too high. These plots exemplify the interconnectivity of the Anesthesia Machine with the %Respiratory System.</i></center><br>
@anchor anesthesia-actionsvalidation
## Validation - Actions

All equipment failures in the Anesthesia Machine system were validated quantitatively where possible and qualitatively elsewhere by comparing the engine output to expected trends and values. For each scenario, the table shows the total number of results in each category. For many investigated scenarios, the model shows good agreement with the expected trends. For the scenarios that did not match with the expected trends, improvements are planned for future %BioGears Engine releases.

<center><br>
Table 2. Cumulative validation results for Anesthesia Machine specific conditions and actions scenarios.
</center>

|	Key	|
|	---	|
|<span class="success">	Good agreement: correct trends or <10% deviation from expected	</span>|
|<span class="warning"> 	Some deviation: correct trend and/or <30% deviation from expected	</span>|
|<span class="danger">	Poor agreement: incorrect trends or >30% deviation from expected	</span>|

|	Scenario 	|	Description	|	Good	|	Decent	|	Bad	|
|	---	|	---	|	---	|	---	|	---	|
|	EndotrachealTubeLeakVaried	|	Tests endotracheal tube leak incidence; Succinylcholine (Succs) injected and anesthesia machine is turned on. Endotracheal tube is intubated and multiple levels (severity:  0.5, 1.0)  of tube cuff leak is tested.	|<span class="success">	21	</span>|<span class="warning">	5	</span>|<span class="danger">	4	</span>|
|	ExpiratoryValveLeakVaried	|	Tests multiple levels (severity: 0.5, 1.0) of expiratory valve leak event. Succs injected, anesthesia machine is turned on, and mask is placed. 	|<span class="success">	25	</span>|<span class="warning">	5	</span>|<span class="danger">	0	</span>|
|	ExpiratoryValveObstructionVaried	|	Tests multiple levels (severity: 0.5, 1.0) of expiratory valve obstruction event. Succs injected, anesthesia machine is turned on, and mask is placed. 	|<span class="success">	23	</span>|<span class="warning">	4	</span>|<span class="danger">	3	</span>|
|	InspiratoryValveLeakVaried	|	Tests multiple levels (severity: 0.5, 1.0) of inspiratory valve leak event. Succs injected, anesthesia machine is turned on, and mask is placed. 	|<span class="success">	28	</span>|<span class="warning">	1	</span>|<span class="danger">	1	</span>|
|	InspiratoryValveObstructionVaried	|	Tests multiple levels (severity: 0.5, 1.0) of inspiratory valve obstruction event. Succs injected, anesthesia machine is turned on, and mask is placed. 	|<span class="success">	23	</span>|<span class="warning">	3	</span>|<span class="danger">	4	</span>|
|	MaskLeakVaried	|	Tests multiple levels (severity: 0.5, 1.0) of mask leak event. Succs injected, anesthesia machine is turned on, and mask is placed. 	|<span class="success">	21	</span>|<span class="warning">	5	</span>|<span class="danger">	4	</span>|
|	OxygenTankPressureLoss	|	Tests oxygen tank pressure loss event. Succs injected, anesthesia machine with oxygen bottle tank as oxygen source is turned on, and mask is placed. Oxygen bottle tank loss event is triggered followed by a replacement with a second oxygen bottle.	|<span class="success">	28	</span>|<span class="warning">	2	</span>|<span class="danger">	0	</span>|
|	OxygenWallPressureLoss	|	Tests oxygen wall pressure loss event. Succs injected, anesthesia machine turned on, and mask is placed. Oxygen wall port is used as oxygen source. Then oxygen wall pressure loss event is triggered followed by a replacement of the wall port oxygen source by oxygen bottle tank.	|<span class="success">	22	</span>|<span class="warning">	3	</span>|<span class="danger">	0	</span>|
|	SodaLimeFailureVaried	|	Tests multiple severity (0.5, 1.0) soda lime CO2 scrubber failure event. Succs injected, anesthesia machine turned on, and mask is placed. 	|<span class="success">	27	</span>|<span class="warning">	2	</span>|<span class="danger">	1	</span>|
|	VaporizorFailureVaried	|	Tests multiple severity (0.5, 1.0) vaporizer failure event. Succs injected, anesthesia machine turned on, and mask is placed. 	|<span class="success">	8	</span>|<span class="warning">	4	</span>|<span class="danger">	0	</span>|
|	VentilatorPressureLossVaried	|	Tests multiple severity (0.5, 1.0) of ventilator pressure loss event. Succs injected, anesthesia machine turned on, and mask is placed. 	|<span class="success">	26	</span>|<span class="warning">	3	</span>|<span class="danger">	1	</span>|
|	YPieceDisconnectVaried	|	Tests multiple levels  (severity: 0.5, 1.0) of Y-piece disconnection event. Succs injected, anesthesia machine turned on, and mask is placed. 	|<span class="success">	21	</span>|<span class="warning">	5	</span>|<span class="danger">	4	</span>|
|		|	Total	|<span class="success">	252	</span>|<span class="warning">	42	</span>|<span class="danger">	22	</span>|


#### Connection Leaks

##### Mask Leak

The mask leak, endotracheal tube leak, and Y-piece disconnect scenarios use the same severities at the same times.  Since they share a methodology in implementation by modifying a leak resistor, they produce the same results.  Each scenario begins by administering a neuromuscular blocker to cause a cessation of normal 
respiration. Mechanical ventilation then begins using a mask to supply the gas. The severity of 
the mask leak / endotracheal tube leak / Y-piece disconnect was varied. The oxygen supplied to the patient is expected to decrease with a leak, which in turn 
causes the oxygen saturation to drop and the carbon dioxide 
in the bloodstream to increase. This response is more pronounced
with increasing severity.

<center>
<table border="0">
<tr>
    <td><img src="./plots/AnesthesiaMachine/MaskLeakVaried_Inflow.jpg" width="550"></td>
    <td><img src="./plots/AnesthesiaMachine/MaskLeakVaried_TotalLungVolume.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/AnesthesiaMachine/MaskLeakVaried_O2Sat.jpg" width="550"></td>
    <td><img src="./plots/AnesthesiaMachine/MaskLeakVaried_CO2PP.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/AnesthesiaMachine/MaskLeakLegend.jpg" width="1100"></td>
</tr>
</table>
</center>
<center><i>Figure 8. Select outputs from the Mask Leak scenario. MaskLeakVaried, EndotrachealTubeLeakVaried, and YPieceDisconnectVaried all give the same results.</i></center>

<center><br>
<i>Table 3. Validation matrix for physiological responses due to varying severities of leaks. MaskLeakVaried, EndotrachealTubeLeakVaried, and YPieceDisconnectVaried all give the same results.</i>
</center>

|	Segment	|	Notes	|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Respiration Rate (breaths/min)	|	Oxygen Saturation	|	Tidal Volume (mL)	|	Aorta Oxygen Partial Pressure (mmHg)	|	Aorta Carbon Dioxide Partial Pressure (mmHg)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Administer succinylcholine.	|	Neuromuscular blocker agent is injected and patient stops breathing.	|	30	|	90	|<span class="success">	Goes to Zero @cite morgan2006clinical                         Goes to Zero @cite dukeSME    	</span>|<span class="success">	Begins to drop according to desaturation curve 	</span>|<span class="success">	Goes to Zero @cite morgan2006clinical	</span>|<span class="success">	Begins to drop w/o respiration 	</span>|<span class="success">	Begins to increase w/o respiration 	</span>|
|	Turn on Anesthesia Machine with leak severity of 0.0.	|	Place mask / insert tube. Zero severity has no effect.	|	90	|	150	|<span class="success">	16 from machine setting 	</span>|<span class="success">	Begins to increase with return of respiration 	</span>|<span class="warning">	Patient Normal Volume 	</span>|<span class="success">	Begins to increase with respiration 	</span>|<span class="success">	Begins to decrease with respiration 	</span>|
|	Turn on leak with severity 0.5.	|	Leak not severe enough to noticeably drop O2 Sat. Causes pressures below FRC, leading to higher TV.	|	150	|	240	|<span class="success">	No Change 	</span>|<span class="warning">	Begins to drop according to desaturation curve 	</span>|<span class="danger">	Drop in tidal volume @cite el2013endotracheal	</span>|<span class="danger">	Begins to drop w/o full tidal volume 	</span>|<span class="danger">	Begins to increase w/o full tidal volume 	</span>|
|	Remove the leak.	|		|	240	|	330	|<span class="success">	No Change 	</span>|<span class="warning">	Begins to increase with return of respiration 	</span>|<span class="warning">	Patient Normal Volume 	</span>|<span class="success">	Begins to increase with respiration 	</span>|<span class="success">	Begins to decrease with respiration 	</span>|
|	Turn on leak again  with severity 1.0.	|	Total lung volume goes to approx. zero - RR is based on waveform, so still registers.	|	330	|	420	|<span class="danger">	Goes to Zero @cite el2013endotracheal	</span>|<span class="success">	Begins to drop according to desaturation curve 	</span>|<span class="success">	Goes to Zero @cite el2013endotracheal	</span>|<span class="success">	Begins to drop w/o respiration 	</span>|<span class="success">	Begins to increase w/o respiration 	</span>|
|	Remove the leak.	|		|	420	|	510	|<span class="success">	16 from machine setting 	</span>|<span class="success">	Begins to increase with return of full tidal volume 	</span>|<span class="warning">	Patient Normal Volume 	</span>|<span class="success">	Begins to increase with full tidal volume 	</span>|<span class="success">	Begins to decrease with full tidal volume 	</span>|


#### Valve Leaks

The inspiratory and expiratory valve leak scenarios begin by administering a neuromuscular 
blocker to cause a cessation of normal respiration. Mechanical ventilation then begins using a 
mask to supply the gas. The severity of the leaks are varied.

##### Expiratory Valve Leak

<center>
<table border="0">
<tr>
    <td><img src="./plots/AnesthesiaMachine/ExpiratoryValveLeakVaried_Inflow.jpg" width="550"></td>
    <td><img src="./plots/AnesthesiaMachine/ExpiratoryValveLeakVaried_EndTidalCO2.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/AnesthesiaMachine/ExpiratoryValveLeakVaried_CO2PP.jpg" width="550"></td>
    <td><img src="./plots/AnesthesiaMachine/ExpiratoryValveLeakVaried_O2PP.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/AnesthesiaMachine/ExpiratoryValveLeakLegend.jpg" width="1100"></td>
</tr>
</table>
</center>
<center><i>Figure 9. Select outputs from the expiratory valve leak scenario.</i></center>

<center><br>
<i>Table 4. Validation matrix for physiological responses due to varying severities of an expiratory valve leak.</i>
</center>

|	Segment	|	Notes	|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Respiration Rate (breaths/min)	|	Oxygen Saturation	|	Tidal Volume (mL)	|	Aorta Oxygen Partial Pressure (mmHg)	|	Aorta Carbon Dioxide Partial Pressure (mmHg)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Administer succinylcholine.	|	Neuromuscular blocker agent is injected and patient stops breathing.	|	30	|	90	|<span class="success">	Goes to Zero @cite morgan2006clinical                         Goes to Zero @cite dukeSME    	</span>|<span class="success">	Begins to drop according to desaturation curve 	</span>|<span class="success">	Goes to Zero @cite morgan2006clinical	</span>|<span class="success">	Begins to drop w/o respiration 	</span>|<span class="success">	Begins to increase w/o respiration 	</span>|
|	Turn on Anesthesia Machine with expiratory valve leak severity of 0.0.	|	Place mask. Zero severity has no effect.	|	90	|	180	|<span class="success">	16 from machine setting 	</span>|<span class="success">	Begins to increase with return of respiration 	</span>|<span class="warning">	Patient Normal Volume 	</span>|<span class="success">	Begins to increase with respiration 	</span>|<span class="success">	Begins to decrease with respiration 	</span>|
|	Turn on expiratory valve leak with severity 0.5.	|		|	180	|	360	|<span class="success">	No Change @cite morgan2006clinical	</span>|<span class="success">	No Change @cite morgan2006clinical	</span>|<span class="success">	No Change @cite morgan2006clinical	</span>|<span class="success">	No Change @cite morgan2006clinical	</span>|<span class="warning">	Increases leading to hypercapnia if untreated @cite morgan2006clinical	</span>|
|	Turn off expiratory valve leak.	|		|	360	|	540	|<span class="success">	No Change 	</span>|<span class="success">	No Change 	</span>|<span class="success">	No Change 	</span>|<span class="success">	No Change 	</span>|<span class="warning">	Begins to decrease slowly return of functionality 	</span>|
|	Turn on expiratory valve leak with severity 1.0.	|	Aorta CO2 not likely to increase all the way to hypercapnia.	|	540	|	720	|<span class="success">	No Change @cite morgan2006clinical	</span>|<span class="success">	No Change @cite morgan2006clinical	</span>|<span class="success">	No Change @cite morgan2006clinical	</span>|<span class="warning">	No Change @cite morgan2006clinical	</span>|<span class="success">	Increases leading to hypercapnia if untreated @cite morgan2006clinical	</span>|
|	Turn off expiratory valve leak.	|		|	720	|	900	|<span class="success">	No Change 	</span>|<span class="success">	No Change 	</span>|<span class="success">	No Change 	</span>|<span class="warning">	No Change 	</span>|<span class="success">	Begins to decrease slowly return of functionality 	</span>|


##### Inspiratory Valve Leak

<center>
<table border="0">
<tr>
    <td><img src="./plots/AnesthesiaMachine/InspiratoryValveLeakVaried_Inflow.jpg" width="550"></td>
    <td><img src="./plots/AnesthesiaMachine/InspiratoryValveLeakVaried_EndTidalCO2.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/AnesthesiaMachine/InspiratoryValveLeakVaried_CO2PP.jpg" width="550"></td>
    <td><img src="./plots/AnesthesiaMachine/InspiratoryValveLeakVaried_O2PP.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/AnesthesiaMachine/InspiratoryValveLeakLegend.jpg" width="1100"></td>
</tr>
</table>
</center>
<center><i>Figure 10. Select outputs from the inspiratory valve leak scenario.</i></center>

<center><br>
<i>Table 5. Validation matrix for physiological responses due to varying severities of an inspiratory valve leak.</i>
</center>

|	Segment	|	Notes	|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Respiration Rate (breaths/min)	|	Oxygen Saturation	|	Tidal Volume (mL)	|	Aorta Oxygen Partial Pressure (mmHg)	|	Aorta Carbon Dioxide Partial Pressure (mmHg)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Administer succinylcholine.	|	Neuromuscular blocker agent is injected and patient stops breathing.	|	30	|	90	|<span class="success">	Goes to Zero @cite Morgan2006Clinical                         Goes to Zero @cite dukeSME    	</span>|<span class="success">	Begins to drop according to desaturation curve 	</span>|<span class="success">	Goes to Zero @cite Morgan2006Clinical	</span>|<span class="success">	Begins to drop w/o respiration 	</span>|<span class="success">	Begins to increase w/o respiration 	</span>|
|	Turn on Anesthesia Machine with inspiratory valve leak severity of 0.0.	|	Place mask. Zero severity has no effect.	|	90	|	180	|<span class="success">	16 from machine setting 	</span>|<span class="success">	Begins to increase with return of respiration 	</span>|<span class="warning">	Patient Normal Volume 	</span>|<span class="success">	Begins to increase with respiration 	</span>|<span class="success">	Begins to decrease with respiration 	</span>|
|	Turn on inspiratory valve leak with severity 0.5.	|		|	180	|	360	|<span class="success">	No Change @cite Morgan2006Clinical	</span>|<span class="success">	No Change @cite Morgan2006Clinical	</span>|<span class="success">	No Change @cite Morgan2006Clinical	</span>|<span class="success">	No Change @cite Morgan2006Clinical	</span>|<span class="danger">	Increases leading to hypercapnia if untreated @cite Morgan2006Clinical	</span>|
|	Turn off expiratory valve obstruction.	|		|	360	|	540	|<span class="success">	No Change 	</span>|<span class="success">	No Change 	</span>|<span class="success">	No Change 	</span>|<span class="success">	No Change 	</span>|<span class="success">	Begins to decrease slowly return of functionality 	</span>|
|	Turn on inspiratory valve leak with severity 1.0.	|		|	540	|	720	|<span class="success">	No Change @cite Morgan2006Clinical	</span>|<span class="success">	No Change @cite Morgan2006Clinical	</span>|<span class="success">	No Change @cite Morgan2006Clinical	</span>|<span class="success">	No Change @cite Morgan2006Clinical	</span>|<span class="success">	Increases leading to hypercapnia if untreated @cite Morgan2006Clinical	</span>|
|	Turn off expiratory valve obstruction.	|		|	720	|	900	|<span class="success">	No Change 	</span>|<span class="success">	No Change 	</span>|<span class="success">	No Change 	</span>|<span class="success">	No Change 	</span>|<span class="success">	Begins to decrease slowly return of functionality 	</span>|


#### Valve Obstructions

The inspiratory and expiratory valve obstruction scenarios begin by administering a neuromuscular 
blocker to cause a cessation of normal respiration. Mechanical ventilation then begins using a 
mask to supply the gas. 
The severity of the obstruction was varied. A full obstruction has the potential to cause a complete 
cessation of respiration or tidal volume due to flow obstruction. Since the neuromuscular blocker
 freezes the patient's spontaneous breathing, the full obstruction of the valves limits the flow of air
 that serves as positive-pressure ventilation to inflate the lungs. 

##### Expiratory Valve Obstruction

<center>
<table border="0">
<tr>
    <td><img src="./plots/AnesthesiaMachine/ExpiratoryValveObstructionVaried_Inflow.jpg" width="550"></td>
    <td><img src="./plots/AnesthesiaMachine/ExpiratoryValveObstructionVaried_EndTidalCO2.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/AnesthesiaMachine/ExpiratoryValveObstructionVaried_CO2PP.jpg" width="550"></td>
    <td><img src="./plots/AnesthesiaMachine/ExpiratoryValveObstructionVaried_O2PP.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/AnesthesiaMachine/ExpiratoryValveObstructionLegend.jpg" width="1100"></td>
</tr>
</table>
</center>
<center><i>Figure 11. Select outputs from the expiratory valve obstruction scenario.</i></center>

<center><br>
<i>Table 6. Validation matrix for physiological responses due to varying severities of an expiratory valve obstruction.</i>
</center>

|	Segment	|	Notes	|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Respiration Rate (breaths/min)	|	Oxygen Saturation	|	Tidal Volume (mL)	|	Aorta Oxygen Partial Pressure (mmHg)	|	Aorta Carbon Dioxide Partial Pressure (mmHg)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Administer succinylcholine.	|	Neuromuscular blocker agent is injected and patient stops breathing.	|	30	|	90	|<span class="success">	Goes to Zero @cite Morgan2006Clinical                         Goes to Zero @cite dukeSME    	</span>|<span class="success">	Begins to drop according to desaturation curve 	</span>|<span class="success">	Goes to Zero @cite Morgan2006Clinical	</span>|<span class="success">	Begins to drop w/o respiration 	</span>|<span class="success">	Begins to increase w/o respiration 	</span>|
|	Turn on Anesthesia Machine with expiratory valve obstruction severity of 0.0.	|	Place mask. Zero severity has no effect.	|	90	|	180	|<span class="success">	16 from machine setting 	</span>|<span class="success">	Begins to increase with return of respiration 	</span>|<span class="warning">	Patient Normal Volume 	</span>|<span class="success">	Begins to increase with respiration 	</span>|<span class="success">	Begins to decrease with respiration 	</span>|
|	Turn on expiratory valve obstruction with severity 0.5.	|	Leak not severe enough to noticeably increase Aorta CO2.	|	180	|	360	|<span class="success">	No Change @cite Morgan2006Clinical	</span>|<span class="success">	No Change @cite Morgan2006Clinical	</span>|<span class="warning">	No Change @cite Morgan2006Clinical	</span>|<span class="success">	No Change @cite Morgan2006Clinical	</span>|<span class="danger">	Increases leading to hypercapnia if untreated @cite Morgan2006Clinical	</span>|
|	Turn off expiratory valve obstruction.	|		|	360	|	540	|<span class="success">	No Change 	</span>|<span class="warning">	Begins to increase with return of respiration 	</span>|<span class="warning">	No Change 	</span>|<span class="success">	Begins to increase with respiration 	</span>|<span class="success">	Begins to decrease slowly return of functionality 	</span>|
|	Turn on expiratory valve obstruction with severity 1.0.	|	Total lung volume goes to approx. zero - RR is based on waveform, so still registers.	|	540	|	720	|<span class="danger">	Goes to zero @cite Roth2000Anesthesia	</span>|<span class="success">	Begins to drop according to desaturation curve 	</span>|<span class="success">	Goes to zero @cite Roth2000Anesthesia	</span>|<span class="success">	Begins to drop w/o respiration 	</span>|<span class="success">	Increases leading to hypercapnia if untreated @cite Morgan2006Clinical	</span>|
|	Turn off expiratory valve obstruction.	|		|	720	|	900	|<span class="success">	16 from machine setting 	</span>|<span class="success">	Begins to increase with return of full tidal volume 	</span>|<span class="danger">	No Change 	</span>|<span class="success">	Begins to increase with respiration 	</span>|<span class="success">	Begins to decrease slowly return of functionality 	</span>|


##### Inspiratory Valve Obstruction

<center>
<table border="0">
<tr>
    <td><img src="./plots/AnesthesiaMachine/InspiratoryValveObstructionVaried_Inflow.jpg" width="550"></td>
    <td><img src="./plots/AnesthesiaMachine/InspiratoryValveObstructionVaried_EndTidalCO2.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/AnesthesiaMachine/InspiratoryValveObstructionVaried_CO2PP.jpg" width="550"></td>
    <td><img src="./plots/AnesthesiaMachine/InspiratoryValveObstructionVaried_O2PP.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/AnesthesiaMachine/InspiratoryValveObstructionLegend.jpg" width="1100"></td>
</tr>
</table>
</center>
<center><i>Figure 12. Select outputs from the inspiratory valve obstruction scenario.</i></center>

<center><br>
<i>Table 7. Validation matrix for physiological responses due to varying severities of an inspiratory valve obstruction.</i>
</center>

|	Segment	|	Notes	|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Respiration Rate (breaths/min)	|	Oxygen Saturation	|	Tidal Volume (mL)	|	Aorta Oxygen Partial Pressure (mmHg)	|	Aorta Carbon Dioxide Partial Pressure (mmHg)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Administer succinylcholine.	|	Neuromuscular blocker agent is injected and patient stops breathing.	|	30	|	90	|<span class="success">	Goes to Zero @cite Morgan2006Clinical                         Goes to Zero @cite dukeSME    	</span>|<span class="success">	Begins to drop according to desaturation curve 	</span>|<span class="success">	Goes to Zero @cite Morgan2006Clinical	</span>|<span class="success">	Begins to drop w/o respiration 	</span>|<span class="success">	Begins to increase w/o respiration 	</span>|
|	Turn on Anesthesia Machine with inspiratory valve obstruction severity of 0.0.	|	Place mask. Zero severity has no effect.	|	90	|	180	|<span class="success">	16 from machine setting 	</span>|<span class="success">	Begins to increase with return of respiration	</span>|<span class="warning">	Patient Normal Volume 	</span>|<span class="success">	Begins to increase with respiration 	</span>|<span class="success">	Begins to decrease with respiration 	</span>|
|	Turn on inspiratory valve obstruction with severity 0.5.	|		|	180	|	360	|<span class="success">	No Change @cite Morgan2006Clinical	</span>|<span class="success">	No Change @cite Morgan2006Clinical	</span>|<span class="danger">	Reduced tidal volume (minor)@cite Roth2000Anesthesia	</span>|<span class="danger">	Begins to decrease without full tidal volume (very minor if at all) @cite Roth2000Anesthesia	</span>|<span class="danger">	Increases leading to hypercapnia if untreated @cite Morgan2006Clinical	</span>|
|	Turn off inspiratory valve obstruction.	|		|	360	|	540	|<span class="success">	No Change 	</span>|<span class="success">	No Change 	</span>|<span class="warning">	Return to patient normal volume with unobstruction 	</span>|<span class="success">	Begins to increase with full tidal volume 	</span>|<span class="success">	Begins to decrease slowly return of functionality 	</span>|
|	Turn on inspiratory valve obstruction with severity 1.0.	|	Total lung volume goes to approx. zero - RR is based on waveform, so still registers.	|	540	|	720	|<span class="danger">	Goes to zero @cite Roth2000Anesthesia	</span>|<span class="success">	Begins to drop according to desaturation curve 	</span>|<span class="success">	system obstruction reduces tidal volume @cite Roth2000Anesthesia	</span>|<span class="success">	Begins to decrease without full tidal volume @cite Roth2000Anesthesia	</span>|<span class="success">	Increases leading to hypercapnia if untreated @cite Morgan2006Clinical	</span>|
|	Turn off inspiratory valve obstruction.	|		|	720	|	900	|<span class="success">	No Change 	</span>|<span class="success">	Begins to increase with full tidal volume 	</span>|<span class="warning">	Return to patient normal volume with unobstruction 	</span>|<span class="success">	Begins to increase with full tidal volume 	</span>|<span class="success">	Begins to decrease slowly return of functionality 	</span>|


#### Ventilator Pressure Loss

The ventilator failure scenario begins by administering a neuromuscular blocker to cause a 
cessation of normal respiration. Mechanical ventilation then begins using a mask to supply the gas. 
The severity of the failure was varied. At the full failure, a cessation of respiration was expected. 

<center>
<table border="0">
<tr>
    <td><img src="./plots/AnesthesiaMachine/VentilatorPressureLossVaried_VentilatorVolume.jpg" width="550"></td>
    <td><img src="./plots/AnesthesiaMachine/VentilatorPressureLossVaried_TotalLungVolume.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/AnesthesiaMachine/VentilatorPressureLossVaried_CO2PP.jpg" width="550"></td>
    <td><img src="./plots/AnesthesiaMachine/VentilatorPressureLossVaried_O2PP.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/AnesthesiaMachine/VentilatorPressureLossLegend.jpg" width="1100"></td>
</tr>
</table>
</center>
<center><i>Figure 13. Select outputs from the ventilator pressure loss scenario.</i></center>

<center><br>
<i>Table 8. Validation matrix for physiological responses due to varying severities of a ventilator pressure loss.</i>
</center>

|	Segment	|	Notes	|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Respiration Rate (breaths/min)	|	Oxygen Saturation	|	Tidal Volume (mL)	|	Aorta Oxygen Partial Pressure (mmHg)	|	Aorta Carbon Dioxide Partial Pressure (mmHg)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Administer succinylcholine.	|	Neuromuscular blocker agent is injected and patient stops breathing.	|	30	|	90	|<span class="success">	Goes to Zero @cite Morgan2006Clinical                         Goes to Zero @cite dukeSME    	</span>|<span class="success">	Begins to drop according to desaturation curve 	</span>|<span class="success">	Goes to Zero @cite Morgan2006Clinical	</span>|<span class="success">	Begins to drop w/o respiration 	</span>|<span class="success">	Begins to increase w/o respiration 	</span>|
|	Turn on Anesthesia Machine with ventilator pressure loss severity 0.0.	|	Place mask. Zero severity has no effect.	|	90	|	180	|<span class="success">	16 from machine setting 	</span>|<span class="success">	Begins to increase with return of respiration 	</span>|<span class="warning">	Patient Normal Volume 	</span>|<span class="success">	Begins to increase with respiration 	</span>|<span class="success">	Begins to decrease with respiration 	</span>|
|	Turn on ventilator pressure loss with severity 0.5.	|		|	180	|	270	|<span class="success">	May drop or remain at 16 depending on pressure supplied by ventilator @cite Morgan2006Clinical	</span>|<span class="success">	Begins to drop according to desaturation curve @cite Morgan2006Clinical	</span>|<span class="success">	Drops without full ventilator pressure @cite Morgan2006Clinical	</span>|<span class="success">	Begins to decrease with lack of full tidal volume @cite Morgan2006Clinical	</span>|<span class="success">	Begins to increase with lack of full tidal volume@cite Morgan2006Clinical	</span>|
|	Turn off Ventilator pressure loss.	|		|	270	|	360	|<span class="success">	16 from machine setting 	</span>|<span class="success">	Begins to increase with return of respiration machine function @cite Morgan2006Clinical	</span>|<span class="warning">	Patient Normal Volume 	</span>|<span class="success">	Begins to increase with a return of ventilator function @cite Morgan2006Clinical	</span>|<span class="success">	Begins to decrease with a return of ventilator function @cite Morgan2006Clinical	</span>|
|	Turn on ventilator pressure loss with severity 1.0.	|		|	360	|	450	|<span class="danger">	Drops with lack of ventilator pressure @cite Morgan2006Clinical	</span>|<span class="success">	Begins to drop according to desaturation curve @cite Morgan2006Clinical	</span>|<span class="success">	Goes to zero with a lack of ventilator pressure @cite Morgan2006Clinical	</span>|<span class="success">	Begins to decrease with lack of oxygen @cite Morgan2006Clinical	</span>|<span class="success">	Begins to increase with lack of ventilator function @cite Morgan2006Clinical	</span>|
|	Turn off Ventilator pressure loss.	|		|	450	|	540	|<span class="success">	16 from machine setting 	</span>|<span class="success">	Begins to increase with return of respiration machine function @cite Morgan2006Clinical	</span>|<span class="warning">	Patient Normal Volume 	</span>|<span class="success">	Begins to increase with a return of ventilator function @cite Morgan2006Clinical	</span>|<span class="success">	Begins to decrease with a return of ventilator function @cite Morgan2006Clinical	</span>|

#### Soda Lime Failure

The soda lime failure scenario begins by administering a neuromuscular blocker to cause a 
cessation of normal respiration. Mechanical ventilation then begins using a mask to supply the gas. 
The severity of the failure was varied. As expected, respiration rate, tidal volume, and oxygen levels 
remain stable. Carbon dioxide in the blood increases. 

<center>
<table border="0">
<tr>
    <td><img src="./plots/AnesthesiaMachine/SodaLimeFailureVaried_InletCO2.jpg" width="550"></td>
    <td><img src="./plots/AnesthesiaMachine/SodaLimeFailureVaried_EndTidalCO2.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/AnesthesiaMachine/SodaLimeFailureVaried_CO2PP.jpg" width="550"></td>
    <td><img src="./plots/AnesthesiaMachine/SodaLimeFailureVaried_O2PP.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/AnesthesiaMachine/SodaLimeFailureLegend.jpg" width="1100"></td>
</tr>
</table>
</center>
<center><i>Figure 14. Select outputs from the soda lime failure scenario.</i></center>

<center><br>
<i>Table 9. Validation matrix for physiological responses due to varying severities of a soda lime failure.</i>
</center>

|	Segment	|	Notes	|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Respiration Rate (breaths/min)	|	Oxygen Saturation	|	Tidal Volume (mL)	|	Aorta Oxygen Partial Pressure (mmHg)	|	Aorta Carbon Dioxide Partial Pressure (mmHg)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Administer succinylcholine.	|	Neuromuscular blocker agent is injected and patient stops breathing.	|	30	|	90	|<span class="success">	Goes to Zero @cite Morgan2006Clinical                         Goes to Zero @cite dukeSME    	</span>|<span class="success">	Begins to drop according to desaturation curve 	</span>|<span class="success">	Goes to Zero @cite Morgan2006Clinical	</span>|<span class="success">	Begins to drop w/o respiration 	</span>|<span class="success">	Begins to increase w/o respiration 	</span>|
|	Turn on Anesthesia Machine with soda lime failure severity 0.0.	|	Place mask. Zero severity has no effect.	|	90	|	180	|<span class="success">	16 from machine setting 	</span>|<span class="success">	Begins to increase with return of respiration	</span>|<span class="warning">	Patient Normal Volume 	</span>|<span class="success">	Begins to increase with respiration 	</span>|<span class="success">	Begins to decrease with respiration	</span>|
|	Turn on soda lime failure with severity 0.5.	|	CO2 not scrubbed as effectively.	|	180	|	360	|<span class="success">	No Change 	</span>|<span class="success">	No Change 	</span>|<span class="success">	No Change	</span>|<span class="success">	No Change 	</span>|<span class="danger">	Begins to increase w/ failure to absorb CO2 @cite Roth2000Anesthesia	</span>|
|	Turn off soda lime failure.	|		|	360	|	540	|<span class="success">	No Change 	</span>|<span class="success">	No Change 	</span>|<span class="success">	No Change	</span>|<span class="success">	No Change 	</span>|<span class="success">	Begins to decrease with functional equipment @cite Roth2000Anesthesia	</span>|
|	Turn on soda lime failure with severity 1.0.	|	CO2 not scrubbed at all.	|	540	|	720	|<span class="success">	No Change 	</span>|<span class="success">	No Change 	</span>|<span class="success">	No Change	</span>|<span class="warning">	No Change 	</span>|<span class="success">	Begins to increase w/ failure to absorb CO2 @cite Roth2000Anesthesia	</span>|
|	Turn off soda lime failure.	|		|	720	|	810	|<span class="success">	No Change 	</span>|<span class="success">	No Change 	</span>|<span class="success">	No Change	</span>|<span class="success">	No Change 	</span>|<span class="success">	Begins to decrease with functional equipment @cite Roth2000Anesthesia	</span>|


#### Vaporizer Failure

The vaporizer failure scenario begins by administering supplemental oxygen via a mask to 
the patient. As previously noted, this is a limitation of the system. Mechanical ventilation 
supplies gas at a prescribed
respiration rate, regardless of whether the patient is conscious and breathing naturally. 
The Anesthesia Machine is then turned on with an inhaled agent, 
desflurane, being supplied. The severity of the failure 
was varied. At a severity of one, the vaporizer supplied no desflurane to the patient, and no changes 
were observed. This was the expected outcome. The drug amount increases in the system and drug effects are 
observed for all other severity levels.

<center>
<table border="0">
<tr>
    <td><img src="./plots/AnesthesiaMachine/VaporizerFailureVaried_DesfluraneConcentration.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/AnesthesiaMachine/VaporizerFailureLegend.jpg" width="800"></td>
</tr>
</table>
</center>
<center><i>Figure 15. Select output from the vaporizer failure scenario.</i></center>

<center><br>
<i>Table 10. Validation matrix for physiological responses due to varying severities of a vaporizer failure.</i>
</center>

|	Segment	|	Notes	|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Respiration Rate (breaths/min)	|	Tidal Volume (mL)	|	Drug Plasma Concentration (ug/mL)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Connect Anesthesia Machine. 	|	Place mask. 	|	30	|	120	|<span class="success">	No Significant Change 	</span>|<span class="success">	No Significant Change 	</span>|<span class="success">	Zero 	</span>|
|	Administer drug via vaporizer. Turn on vaporizer failure of severity 1.0.	|	Full vaporizer failure - no drug in body.	|	120	|	300	|<span class="warning">	No Change 	</span>|<span class="warning">	No Change 	</span>|<span class="success">	Zero 	</span>|
|	Decrease vaporizer failure severity to 0.5.	|		|	300	|	480	|<span class="warning">	Moderate Increase @cite Morgan2006Clinical	</span>|<span class="warning">	Moderate Increase @cite Morgan2006Clinical	</span>|<span class="success">	Increase 	</span>|
|	Turn off vaporizer failure.	|		|	480	|	660	|<span class="success">	Continued drug response @cite Morgan2006Clinical	</span>|<span class="success">	Continued drug response @cite Morgan2006Clinical	</span>|<span class="success">	Increase at faster rate 	</span>|

#### Oxygen Supply Failures

##### Oxygen Wall Pressure Loss

The oxygen wall pressure loss failure scenario begins by administering a neuromuscular blocker to cause a 
cessation of normal respiration. Mechanical ventilation then begins using a mask to supply the gas. 
Oxygen is supplied from the wall source, and the patient is stabilized. Then, oxygen wall port 
pressure source loss is triggered, and the physiological responses of the patient are observed. The process is 
repeated by turning the oxygen wall pressure loss condition on and off, 
followed by the use of the oxygen bottle tank as a source.
The oxygen supply failure due to O2 wall pressure loss causes a drop in arterial oxygen level, while re-supplying the system with
oxygen by turning off the failure brings the oxygen level to normal.

<center>
<table border="0">
<tr>
    <td><img src="./plots/AnesthesiaMachine/OxygenWallPressureLoss_InletO2.jpg" width="550"></td>
    <td><img src="./plots/AnesthesiaMachine/OxygenWallPressureLoss_O2Sat.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/AnesthesiaMachine/OxygenWallPressureLoss_CO2PP.jpg" width="550"></td>
    <td><img src="./plots/AnesthesiaMachine/OxygenWallPressureLoss_O2PP.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/AnesthesiaMachine/OxygenWallPressureLossLegend.jpg" width="1100"></td>
</tr>
</table>
</center>
<center><i>Figure 16. Select outputs from the oxygen wall pressure loss scenario.</i></center>

<center><br>
<i>Table 11. Validation matrix for physiological responses due to an oxygen wall pressure loss.</i>
</center>

|	Segment	|	Notes	|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Respiration Rate (breaths/min)	|	Oxygen Saturation	|	Tidal Volume (mL)	|	Aorta Oxygen Partial Pressure (mmHg)	|	Aorta Carbon Dioxide Partial Pressure (mmHg)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Administer succinylcholine.	|	Neuromuscular blocker agent is injected and patient stops breathing.	|	30	|	90	|<span class="success">	Goes to Zero @cite Morgan2006Clinical                         Goes to Zero @cite dukeSME    	</span>|<span class="success">	Begins to drop according to desaturation curve 	</span>|<span class="success">	Goes to Zero @cite Morgan2006Clinical	</span>|<span class="success">	Begins to drop w/o respiration 	</span>|<span class="success">	Begins to increase w/o respiration 	</span>|
|	Turn on Anesthesia Machine.	|	Place mask. Use wall oxygen as oxygen source.	|	90	|	210	|<span class="success">	16 from machine setting 	</span>|<span class="success">	Begins to increase with return of respiration 	</span>|<span class="warning">	Patient Normal Volume 	</span>|<span class="success">	Begins to increase with respiration 	</span>|<span class="success">	Begins to decrease with respiration 	</span>|
|	Turn on oxygen wall port pressure loss.	|		|	210	|	270	|<span class="success">	No Change 	</span>|<span class="success">	Begins to drop according to desaturation curve @cite Morgan2006Clinical	</span>|<span class="success">	No Change 	</span>|<span class="success">	Begins to decrease with lack of oxygen @cite Morgan2006Clinical	</span>|<span class="success">	No significant change 	</span>|
|	Turn off oxygen wall port pressure loss.	|		|	270	|	390	|<span class="success">	No Change 	</span>|<span class="success">	Begins to increase with return of oxygen supply @cite Morgan2006Clinical	</span>|<span class="success">	No Change 	</span>|<span class="success">	Begins to increase with return of oxygen supply @cite Morgan2006Clinical	</span>|<span class="success">	No significant change 	</span>|
|	Switch from wall port to oxygen bottle.	|		|	390	|	570	|<span class="success">	No Change 	</span>|<span class="success">	Continues to increase with oxygen supply @cite Morgan2006Clinical	</span>|<span class="success">	No Change 	</span>|<span class="warning">	No Change 	</span>|<span class="warning">	No significant change 	</span>|


##### Oxygen Tank Pressure Loss

The %BioGears Anesthesia System has two oxygen tanks as additional oxygen supply sources. Under normal 
conditions, the model allows 
selection between the two tanks as substitutes for the wall source. To address the effect of 
oxygen tank pressure loss, the model employs a scenario that begins 
by administering a neuromuscular blocker to disrupt normal respiration. Then, the Anesthesia Machine is turned on, a 
mask is placed (similarly to the wall pressure loss scenario), and an oxygen bottle tank is set as the
oxygen source. The patient is stabilized first for some time. Then, the oxygen tank pressure loss event is triggered and the 
physiological responses of the patient are examined. The process is repeated by switching  
the oxygen tank pressure loss state on and off.  Finally, the original oxygen bottle tank 
is replaced with the second tank.
The model responded well and produced expected behavior, showing 
a decline in oxygen saturation and arterial oxygen partial pressure.

<center>
<table border="0">
<tr>
    <td><img src="./plots/AnesthesiaMachine/OxygenTankPressureLoss_InletO2.jpg" width="550"></td>
    <td><img src="./plots/AnesthesiaMachine/OxygenTankPressureLoss_O2Sat.jpg" width="550"></td>
</tr>
<tr>
    <td><img src="./plots/AnesthesiaMachine/OxygenTankPressureLoss_CO2PP.jpg" width="550"></td>
    <td><img src="./plots/AnesthesiaMachine/OxygenTankPressureLoss_O2PP.jpg" width="550"></td>
</tr>
<tr>
    <td colspan="2"><img src="./plots/AnesthesiaMachine/OxygenTankPressureLossLegend.jpg" width="1100"></td>
</tr>
</table>
</center>
<center><i>Figure 17. Select outputs from the oxygen tank pressure loss scenario.</i></center>

<center><br>
<i>Table 12. Validation matrix for physiological responses due to an oxygen tank pressure loss.</i>
</center>

|	Segment	|	Notes	|	Action Occurrence Time (s)	|	Sampled Scenario Time (s)	|	Respiration Rate (breaths/min)	|	Oxygen Saturation	|	Tidal Volume (mL)	|	Aorta Oxygen Partial Pressure (mmHg)	|	Aorta Carbon Dioxide Partial Pressure (mmHg)	|
|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|	---	|
|	Administer succinylcholine.	|	Neuromuscular blocker agent is injected and patient stops breathing.	|	30	|	60	|<span class="success">	Goes to Zero @cite Morgan2006Clinical                         Goes to Zero @cite dukeSME    	</span>|<span class="success">	Begins to drop according to desaturation curve 	</span>|<span class="success">	Goes to Zero @cite Morgan2006Clinical	</span>|<span class="success">	Begins to drop w/o respiration 	</span>|<span class="success">	Begins to increase w/o respiration 	</span>|
|	Turn on Anesthesia Machine.	|	Place mask. Use oxygen bottle tank as oxygen source.	|	60	|	180	|<span class="success">	16 from machine setting 	</span>|<span class="success">	Begins to increase with return of respiration 	</span>|<span class="warning">	Patient Normal Volume 	</span>|<span class="success">	Begins to increase with respiration 	</span>|<span class="success">	Begins to decrease with respiration 	</span>|
|	Turn on oxygen tank pressure loss.	|		|	180	|	210	|<span class="success">	No Change 	</span>|<span class="success">	Begins to drop according to desaturation curve @cite Morgan2006Clinical	</span>|<span class="success">	No Change 	</span>|<span class="success">	Begins to decrease with lack of oxygen @cite Morgan2006Clinical	</span>|<span class="success">	No significant change 	</span>|
|	Turn off oxygen tank pressure loss.	|		|	210	|	246	|<span class="success">	No Change 	</span>|<span class="success">	Begins to increase with return of oxygen supply @cite Morgan2006Clinical	</span>|<span class="success">	No Change 	</span>|<span class="success">	Begins to increase with return of oxygen supply @cite Morgan2006Clinical	</span>|<span class="success">	No significant change 	</span>|
|	Oxygen take runs out.	|		|	246	|	270	|<span class="success">	No Change 	</span>|<span class="success">	Begins to drop according to desaturation curve @cite Morgan2006Clinical	</span>|<span class="success">	No Change 	</span>|<span class="success">	Begins to decrease with lack of oxygen @cite Morgan2006Clinical	</span>|<span class="success">	No significant change 	</span>|
|	Replace oxygen tank.	|	Use the second bottle.	|	270	|	390	|<span class="success">	No Change 	</span>|<span class="success">	Continues to increase with oxygen supply @cite Morgan2006Clinical	</span>|<span class="success">	No Change 	</span>|<span class="success">	Begins to increase with return of oxygen supply @cite Morgan2006Clinical	</span>|<span class="warning">	No significant change 	</span>|


@anchor anesthesia-conclusion
## Conclusion

The %BioGears Anesthesia Machine provides a generic representation of a complex and specialized system.  The model predicts physiological response to mechanical ventilation, inhaled agent administration, and common equipment 
failures.  This system is a strong addition to the %BioGears Engine with the potential for 
future development. This system has previously been used to power the training application 
HumanSim: Sedation and Airway, which
was funded by TATRC under contract number W81XWH-11-C-0045.

@anchor anesthesia-improvements
# Future Work

## Coming Soon

There are no planned near term additions.

## Recommended Improvements

The %BioGears Engine modularity could be taken advantage of to add parameters and elements 
for specific equipment models and manufacturers. 

Including  bag squeeze capability to
allow the use of manual ventilation as an additional option will expand 
the functionality of the %BioGears Anesthesia Machine system.

Further functionality to the %BioGears Anesthesia Machine system includes the 
implementation of  synchronized ventilation
that allows spontaneous breathing along with positive-pressure ventilation.
@anchor anesthesia-appendices
# Appendices

## Data Model Implementation

@ref AnesthesiaMachineTable "Anesthesia Machine"

## Acronyms

ETT - Endotracheal tube

## Compartments

* AnesthesiaConnection
* ExpiratoryLimb
* GasInlet
* GasSource
* InspiratoryLimb
* ReliefValve
* Scrubber
* Selector
* Squeeze
* Ventilator
* VentilatorConnection
* YPiece
---
title: 'BioGears: A C++ library for whole body physiology simulations'
tags:
  - C++
  - physiology
  - medicine
  - biology
  - pharmacology
authors:
  - name: Austin Baird^[Corresponding author.]
    orcid: 0000-0002-4711-3016
    affiliation: 1
  - name: Matthew McDaniel
    affiliation: 1
  - name: Steven A. White
    affiliation: 1
  - name: Nathan Tatum
    affiliation: 1
  - name: Lucas Marin
    affiliation: 1
affiliations:
 - name: Applied Research Associates, Inc. Advanced Modeling and Simulation Systems Directorate, Raleigh, NC, USA
   index: 1
date: 27 August 2020
bibliography: paper.bib
---

# Summary

BioGears is an open source, extensible human physiology computational engine that is designed to enhance medical education, research, and training technologies. BioGears is primarily written in C++ and uses an electric circuit analog to characterize the fluid dynamics of the cardiopulmonary system. As medical training requirements become more complex, there is a need to supplement traditional simulators with physiology simulations. To this end, BioGears provides an extensive number of validated injury models and related interventions that may be applied to the simulated patient. In addition, BioGears compiled libraries may be used for computational medical research to construct *in-silico* clinical trials related to patient treatment and outcomes. Variable patient inputs support diversity and specification in a given application. The engine can be used standalone or integrated with simulators, sensor interfaces, and models of all fidelities. The Library, and all associated projects, are published under the Apache 2.0 license and are made available through the public GitHub repository. BioGears aims to lower the barrier to create complex physiological simulations for a variety of uses and requirements.

Physiological models have been used for healthcare simulation for many years but generally, complex models of the human physiology are not implemented and the patient state is preprogrammed by the instructor. The most prominent commercial implementation of similar software is Maestro, developed by Canadian Aviation Electronics, Inc. This product is proprietary and it is not clear to the authors how models are developed, implemented, and validated. Other similar open source projects include: [pk-sim](https://github.com/Open-Systems-Pharmacology/PK-Sim) the pharmacological modeling framework [@willmann2003pk], CellML a generic biological modeling markup language with applications spanning biological applications [@lloyd2004cellml], and [Pulse](https://gitlab.kitware.com/physiology/engine) a maintained BioGears fork lacking some of the recent models but instead focusing on Unity VR integration. Each of these either has submodels that are integrated into BioGears (like a pkpd pharmacological model), or is a more generic implementation of biological modeling, like CellML. BioGears is unique in the depth of integrated models, while being free and open-source for the research community.

# Statement of need 

The fields of Medical simulation and computational medicine are growing in application diversity and complexity [@sweet2017crest]. Simple CPR manikins are now being replaced with complex robotic systems that can simulate breathing and react to the performance of the trainee. As these systems use cases grow, there is a requirement that they be supplemented with physiology modeling. BioGears fills this need by providing a free computational framework to use as a backbone to many of these robotic training manikins and may support other computational medicine research applications. The BioGears project aims to better democratize the construction of high-fidelity medical training by providing a sophisticated, complex physiology engine to developers that is easy to integrate and free to use.

BioGears uses a lumped circuit model to describe the circulatory and respiratory systems. This approximation of the simulated cardiopulmonary system has been studied in the past and shown to accurately represent the hemodynamics of the arterial system by using resistance and compliance elements [@otto1899grundform; @westerhof2009arterial]. This approximation creates a system that can be solved for rapidly, decreasing the simulation run-time and computational requirements. In addition, BioGears implements models of diffusion and substance transport to properly simulate the gas/blood interface in the lungs. To handle more complex models of physiology, such as pharmacological models, BioGears constructs a set of hierarchal compartments built on top of the circuit analogs. Top-most compartments represents the system level data, such as the liver, with sub-compartments representing more granular biology of the patient such as the nephron, extravascular tissue, and even intracellular spaces. A generic data request framework, leveraging XML, is used to access various substance, fluid, thermal, and gas information for a specific compartment of the body. 

The BioGears engine has been used in numerous computational medical research applications including: models of sepsis [@mcdaniel2019whole], burn [@mcdaniel2019full], surgical planning [@potter2017physiology], and pharmacological kinetics and clearance [@mcdaniel2019open]. For each application, the patient physiology and traditional interventions used to treat each injury are validated. Full documentation and validation for every action available to the user is provided through our [website](https://www.biogearsengine.com/).

New drugs can be implemented in BioGears by filling in the appropriate physiochemical properties in the provided eXtensible Markup Language (XML) format. The BioGears engine handles computation of clearance, tissue diffusion, and patient responses based on this file and does not require additional C++ programming by the user. The software architecture of BioGears is implemented in three layers of abstraction to encourage easy integration and provide a robust application programming interface (API), see \autoref{fig:example}.

![Overview of the BioGears engine software structure. The SE layer provides a generic physiology API and may be leveraged for other physiology engine implementations and/or integration with other computational biology applications.\label{fig:example}](Fig1.png)




# Features

The BioGears engine, once compiled, provides a set of libraries that may be included in any application that wishes to leverage a physiological simulation of a patient. In addition, BioGears provides build support and testing for all major user platforms (MacOS, Windows, Linux, and ARM). An instance of a BioGears engine models a single patient's physiology and can be edited at the start of runtime or during the simulation, in the following ways: 

- The patient is defined by parameters, such as height, weight, systolic and diastolic pressure.
- Initialize the patient with specific chronic and/or disease states via conditions.
- Modify the patients external environmental conditions (weather, submerge in water, etc.)
- Apply various actions: acute insults/injuries, interventions, conscious breathing, exercise, etc.
- Interact with equipment models, such as an Anesthesia and/or an ECG Machine as well as an Inhaler via the action framework.

Constructing a pointer to an engine, loading a patient, creating data requests (published to a given .csv file), and applying actions to the patient is easy and can be done in only a few lines of code:

```C++
using namespace biogears;
void HowToFasciculation()
{
  // Create the engine and load the patient
  std::unique_ptr<PhysiologyEngine> bg = 
  CreateBioGearsEngine("HowToFasciculation.log");
  bg->GetLogger()->Info("HowToFasciculation");

  if (!bg->LoadState("./states/StandardMale@0s.xml")) {
    bg->GetLogger()->Error("Could not load state, check the error");
    return;
  }
  
  // Set data requests and create output file
  bg->GetEngineTrack()->GetDataRequestManager().
  CreateLiquidCompartmentDataRequest().
  Set("VenaCava", *Na, "Molarity", AmountPerVolumeUnit::mmol_Per_L);
  bg->GetEngineTrack()->GetDataRequestManager().
  SetResultsFilename("HowToFasciculation.csv");

  // Advance Simulation time by 1 minute 
  bg->AdvanceModelTime(1.0, TimeUnit::min); 

  // Create an AirwayObstruction action with a severity of .6
  obstruction.GetSeverity().SetValue(0.6);
  bg->ProcessAction(obstruction);
  
  bg->GetLogger()->Info("Giving the patient an airway obstruction.");
  for( auto i = 0; i < 60*60; --i){ 
     //Advance time for 1 hour and log EngineTracks to ouput
     bg->AdvanceModelTime(1.0, TimeUnit::s);  
     m_Engine.GetEngineTrack()->
     TrackData(m_Engine.GetSimulationTime(TimeUnit::s),m_append_data);
  }
``` 

The Biogears project includes 36 complete howto's which demonstrate various uses of the API and a versatile command line utility for running XML defined Scenarios using the CDM. Additional support can be found by going to <https://biogearsengine.com/>, <https://www.biogears.dev/>, or by joining our IRC [slack channel](https://github.com/BioGearsEngine/core/wiki/Getting-access-to-our-IRC-development-channel).


# Acknowledgments

This work was made possible by cooperative agreements that were awarded and administered by the US Army Medical Research & Materiel Command (USAMRMC) and the Telemedicine and Advanced Technology Research Center (TATRC), at Fort Detrick, MD, under Contract Number W81XWH‐13‐2‐0068 and W81XWH-17-C-0172. We'd like to acknowledge the support and guidance of Hugh Connacher, Harvey Magee, and Geoff Miller.

# References
## HowTo Examples
This directory is a list of howto's for the BioGears API. Several of the examples are short examples of common actions in BioGears. A few are more complete integration examples for the API and threading control.  

TODO: Add README.MD to each howto

## Help
The BioGears build system provides a helpful STAGE target to run cmakes fixup_bundle utility for consolidating all external DLLs to a single runtime directory. This is inteneded to assist users in avoiding putting all of their third party development DLLs in their system path. 

After BioGears has completed building run the STAGE target one time to ensure all the third party DLLS will be moved in to the output directory before attempting to launch the application. This functionality should only needed to be run once per build configuration.
# Patient Geneation
This HowTo Demonstrates how to batch run multiple physiology-engines within a thread-pool. The HowTo takes in
a series of command line arguments to determine the paramaters of multiple trials which will use the trails individual parmaters
to determine the individual behavior of the trial.

## Motivation
Combining Physics simulations results with ML systems as training data is a way to push back a common lack of common data reguarding 
specific human adversities due to the lack of ethical ways of collecting controll data.  In this scenario we generate a series of patients intended to train
a Hypvolemic Shock detector. A given trail will create a virtual patient with an initial infection then beign applying one of several treatment plans after a specific interval of time

### Treatments
 * NONE - No Treatment or Interventions Applied 
 * STANDARD - An infusion of Antibiotics wil start after a given interval of X hours and reapplied every Y minutes
 * REFRESH - If the patient  - REstricted Fluid REsuscitation in Sepsis-associated Hypotension[1]
 * EGDT - EGDT - Early Goal Directed Therapy
 * RANDOM - Select a Random choice of the above.

## Trial Treatment
In this HowTo a veradic scenario is run based on the command line arguments. The scenario is the following

1. Patient Infection of [severity] with a MIC [strength] occurs at 5 minutes after start
2. Patient receives antibiotic treatment at [X] minutes and every [interval] minutes there after
3. Scenario runs for a total of [maximum] hours 
4. Patient is feed every 8 hours.
5. TODO: Manual Patient Urination

You can run up [N] scenarios over [J] threads each thread will need its own set of arguents.  The device can read a csv file to simplify the input

Example -j 5 --duration 5 --trials "1,1,60,60" "1,2,60,60" "1,1,30,30" "1,2,120,120" "3,3,120,120"
## Arguments
| Name     |  Syntax | Purpose | Example | Default | 
|----------|---------|---------|---------|---------|
| Thread Count | -j[N] | Set the maximum number of threads | -j 4 | Total - 1 |
| Duration | --duration [N] | Set the Initial Infection level | --duration 5 | 6 horus |  
| INPUTS| --trials [ [PARAMS]... ] | Inline trial paramaters for multiple trials  | See (#Trial Parameters) | |
| CONFIG | --config [file] | Configure trial inputs with a csv file | --config trials.csv |

## Trial Parameters


| NAME | Optional | Type    |  Values  | Purpose |  Default    |
|------|----------|---------|----------|--------|-------|
| Severity | Required | Enum  | [None, Low, Moderate, Severe] | Sets the everity of the initial infection ||
| Treatment Plan | Required | Enum  | [None, Standard, REFRESH, EGDT, RANDOM] | Sets the everity of the initial infection ||
| MIC | Required | Real | mic_g_per_l | Minimal inhebitory concentration of the bacterial infection ||
| First treatment occurence | Required |  Integral | minutes | Delay Until first antibiotics is applied allowing time for infection to grow ||
| Interval between treatments | Required |  Integral | minutes | Number of minutes between antibiotic treatments|
| Patient files  | Optional |  Integral | hours | Specifcy which patient to load  | patient/StandardMale@0s.xml |
| Total Duration | Optional | Integral | hours | Length of full trial | (6) or --duration N |

 `Example --trials "SEVERE,STANDARD,1,1,60, States/StandardMale@0s, 60" "MODERATE,REFRESH,1,2,60"`

## CSV FORMAT

Simple CSV FILE
```
Duration, [Thread Pool]
Default Patient
Severity, PLAN, MIC, First Treatment, Interval for treatments, [Patient, Duration]
Severity, PLAN, MIC, First Treatment, Interval for treatments, [Patient, Duration]
Severity, PLAN, MIC, First Treatment, Interval for treatments, [Patient, Duration]
Severity, PLAN, MIC, First Treatment, Interval for treatments, [Patient, Duration]
Severity, PLAN, MIC, First Treatment, Interval for treatments, [Patient, Duration]
```

## Citations
 * McDaniel, Matthew, et al. "A Whole-Body Mathematical Model of Sepsis Progression and Treatment Designed in the BioGears Physiology Engine." Frontiers in physiology 10 (2019): 1321. https://www.frontiersin.org/articles/10.3389/fphys.2019.01321/full# minizip 2.9.0

minizip is a zip manipulation library written in C that is supported on Windows, macOS, and Linux.

[![Master Branch Status](https://github.com/nmoinvaz/minizip/workflows/CI/badge.svg)](https://github.com/nmoinvaz/minizip/actions)
[![Fuzzing Status](https://oss-fuzz-build-logs.storage.googleapis.com/badges/minizip.svg)](https://bugs.chromium.org/p/oss-fuzz/issues/list?sort=-opened&can=1&q=proj:minizip)
[![CodeFactor](https://www.codefactor.io/repository/github/nmoinvaz/minizip/badge)](https://www.codefactor.io/repository/github/nmoinvaz/minizip)
[![License: Zlib](https://img.shields.io/badge/license-zlib-lightgrey.svg)](https://github.com/nmoinvaz/minizip/blob/master/LICENSE)

Maintained by Nathan Moinvaziri.

## Branches

| Name | Description |
|:- |:-|
|[master](https://github.com/nmoinvaz/minizip/tree/master)|Modern rewrite that includes more advanced features, improvements in code maintainability and readability, and the reduction of duplicate code. Compatibility layer provided for older versions.|
|[dev](https://github.com/nmoinvaz/minizip/tree/dev)|Latest development code|
|[1.2](https://github.com/nmoinvaz/minizip/tree/1.2)|Drop-in replacement for zlib's minizip that includes WinZip AES encryption, disk splitting, I/O buffering and some additional fixes.|
|[1.1](https://github.com/nmoinvaz/minizip/tree/1.1)|Original minizip as of zlib 1.2.11.|

## History

Minizip was originally developed by [Gilles Vollant](https://www.winimage.com/zLibDll/minizip.html) in 1998. It was first included in the zlib distribution as an additional code contribution starting in zlib 1.1.2. Since that time, it has been continually improved upon and contributed to by many people. The original [project](https://github.com/madler/zlib/tree/master/contrib/minizip) can still be found in the zlib distribution that is maintained by Mark Adler.

My work with the minizip library started in 2006 when I fixed a few bugs I found and submitted them to
Gilles Vollant. In 2010, I implemented WinZip AES encryption, disk splitting, and
I/O buffering that were necessary for another project I was working on. Shortly after, I created this public repository
so I could share my improvements with the community. In early 2017, I began the work to refactor and rewrite
the library as version 2 because it had become difficult to maintain and code readability suffered over the years.

## Features

+ Creating and extracting zip archives.
+ Adding and removing entries from zip archives.
+ Read and write raw zip entry data.
+ Reading and writing zip archives from memory.
+ Zlib, BZIP2, and LZMA compression methods.
+ Password protection through Traditional PKWARE and [WinZIP AES](https://www.winzip.com/aes_info.htm) encryption.
+ Buffered streaming for improved I/O performance.
+ NTFS timestamp support for UTC last modified, last accessed, and creation dates.
+ Disk split support for splitting zip archives into multiple files.
+ Preservation of file attributes across file systems.
+ Follow and store symbolic links.
+ Unicode filename support through UTF-8 encoding.
+ Legacy character encoding support CP437, CP932, CP936, CP950.
+ Turn off compilation of compression, decompression, or encryption.
+ Windows (Win32 & WinRT), macOS and Linux platform support.
+ Streaming interface for easy implementation of additional platforms.
+ Support for Apple's compression library ZLIB implementation.
+ Zero out local file header information.
+ Zip/unzip of central directory to reduce size.
+ Ability to generate and verify CMS signature for each entry.
+ Recover the central directory if it is corrupt or missing.
+ Example minizip command line tool.

## Build

To generate project files for your platform:

1. [Download and install](https://cmake.org/install/) cmake.
2. Run cmake in the minizip directory.

```
cmake . -DMZ_BUILD_TEST=ON
cmake --build .
```

## Build Options

| Name | Description | Default Value |
|:- |:-|:-:|
| MZ_COMPAT | Enables compatibility layer | ON |
| MZ_ZLIB | Enables ZLIB compression | ON |
| MZ_BZIP2 | Enables BZIP2 compression | ON |
| MZ_LZMA | Enables LZMA compression | ON |
| MZ_PKCRYPT | Enables PKWARE traditional encryption | ON |
| MZ_WZAES | Enables WinZIP AES encryption | ON |
| MZ_LIBCOMP | Enables Apple compression | OFF |
| MZ_OPENSSL | Enables OpenSSL encryption | OFF |
| MZ_BRG | Enables Brian Gladman's library | OFF |
| MZ_SIGNING | Enables zip signing support | ON |
| MZ_COMPRESS_ONLY | Only support compression | OFF |
| MZ_DECOMPRESS_ONLY | Only support decompression | OFF |
| MZ_BUILD_TEST | Builds minizip test executable | OFF |
| MZ_BUILD_UNIT_TEST | Builds minizip unit test project | OFF |
| MZ_BUILD_FUZZ_TEST | Builds minizip fuzz executables | OFF |

## Contents

| File(s) | Description |
|:- |:-|
| minizip.c | Sample application |
| mz_compat.\* | Minizip 1.x compatibility layer |
| mz.h | Error codes and flags |
| mz_os\* | Platform specific file/utility functions |
| mz_crypt\* | Configuration specific crypto/hashing functions |
| mz_strm.\* | Stream interface |
| mz_strm_buf.\* | Buffered stream |
| mz_strm_bzip.\* | BZIP2 stream using libbzip2 |
| mz_strm_libcomp.\* | Apple compression stream |
| mz_strm_lzma.\* | LZMA stream using liblzma |
| mz_strm_mem.\* | Memory stream |
| mz_strm_split.\* | Disk splitting stream |
| mz_strm_pkcrypt.\* | PKWARE traditional encryption stream |
| mz_strm_os\* | Platform specific file stream |
| mz_strm_wzaes.\* | WinZIP AES stream |
| mz_strm_zlib.\* | Deflate stream using zlib |
| mz_zip.\* | Zip format |
| mz_zip_rw.\* | Zip reader/writer |

## Third-Party Libraries

+ [zlib](https://zlib.net/) written by Mark Adler and Jean-loup Gailly.
  + Not included in this repository
  + Or alternatively, [zlib-ng](https://github.com/Dead2/zlib-ng) by Hans Kristian Rosbach
+ [BZIP2](https://www.sourceware.org/bzip2/) written by Julian Seward.
+ [liblzma](https://tukaani.org/xz/) written by Lasse Collin.
  + Modifications were made to support the ZIP file format specification
+ [AES](https://github.com/BrianGladman/aes) and [SHA](https://github.com/BrianGladman/sha) libraries of Brian Gladman.

## Acknowledgments

Thanks go out to all the people who have taken the time to contribute code reviews, testing and/or patches. This project would not have been as good without you.

Thanks to [Gilles Vollant](https://www.winimage.com/zLibDll/minizip.html) on which this work is originally based on.

The [ZIP format](https://github.com/nmoinvaz/minizip/blob/master/doc/appnote.txt) was defined by Phil Katz of PKWARE.
