[![Build Status](https://travis-ci.org/ropensci/treeStartR.svg?branch=master)](https://travis-ci.org/wrightaprilm/treeStartR)
[![codecov](https://codecov.io/gh/ropensci/treeStartR/branch/master/graph/badge.svg)](https://codecov.io/gh/ropensci/treeStartR)
[![](https://badges.ropensci.org/239_status.svg)](https://github.com/ropensci/onboarding/issues/239)

## treeStartR

treeStartR is an R package for auomated generation of starting trees for total-evidence phylogenetic analyses. 

Phylogenetic trees, and particularly time-scaled phylogenetic trees, are 
increasingly estimated using parameter-rich models of evolution [@catmodel; @wright2016] and models incorporating macroevoloutionary processes [@fbd]. Finding a starting tree with a computable likelihood to perform a Bayesian MCMC under these complex models can be a challenge, particularly when when estimation involves many
taxa, large datasets, and missing data.

Different phylogenetic estimation packages allow users to find starting trees in different ways, such as estimating a tree under parsimony (@raxml8) or neighbor-joining [@beast2], randomly adding taxa to the tree [@beast2; @raxml8], or 
allowing the user to specify the tree. Addition of taxa is usually performed based on data - that is, using an algorithm (such as parsimony or neighbor-joining) to add the tips to the tree. However, in analyses of the fossil record, specimens may be included for which there are no molecular or morphological characters available, but the taxonomy of the specimen is known via expert opinion. This is the case with many specimens harvested via repositories such as the Paleobiology Database. 

The purpose of this package is to allow users to efficiently add taxa to a given tree to generate a reasonable starting tree. Functions in this package allow taxa to be added to a tree according to either their taxonomy (if other specimens from the same genus are present) in function `present_tippr`, at random (`rand_absent_tippr`), or via other user-specified groupings (`text_placr`). The package uses functions from 
phytools [@phytools] and ape [@ape].

## Installation

Install the development version directly from Github

```{r eval=FALSE}
library(devtools)
install_github("wrightaprilm/treeStartR")
```

## Usage Example

```{r}
library(treestartr)
data(bears)
```

First, we need to load a list of the total set of taxa present in the tree. The "total set" refers to any taxa that will be included in your analysis. This can be either a CSV or a TSV file. A sample list, bears, has been provided as part of the bears data object, but you can also generate one using the function dataf_parser. Note that if you have higher order taxa, for example, if you are placing a fossil that you have identified to family level, but not species level, these should be included as "family_spp." treeStartR relies on the underscore to separate between the higher order and lower order taxa. Use consistent formatting between the tree and taxon list - i.e, do not call a taxon "Ge_spp" on the tree and "Genus_spp" in the taxon list.

```{r}
tax_frame <- dataf_parsr(system.file("extdata/bears_taxa.tsv", package = "treestartr"))
```

Next, we find out which of the taxa from our total set are not represented on the tree already. This function takes as input a tree, with or without branch lengths, but without annotations (such as 95% HPDs). It also takes the total set of taxa generated by `dataf_parser`:

```{r warning=FALSE}
absent_list <- genera_strippr(tree, tax_frame)
```

## Adding tips with congeners

Finally, we add the tips that are not present to the tree. If there are other representatives of the same genera as an absent taxon (for example, adding an additional "Ursus" species to the example tree), those taxa will be used to place the tip. If there are multiple species of the genera, the new tip will subtend the most recent common ancestor of the tips already on the tree. If there is only one representative, the tip will subtend the parent node of that taxon.

```{r warning=FALSE}
new_tree <- present_tippr(tree, absent_list)
plot(new_tree)
```

After we have added our tips, we can check how many tips remain to be added, and which they are:

```{r warning=FALSE}
genera_strippr(new_tree, tax_frame)
```

## Adding tips manually

We can also add the tips that have no congeners. This function will ask for input. A pop-up will be produced, showing node labels. When the program asks for input, you will tell it what node you would like the tip to subtend. Alternatively, if you would like to place the tip as sister to a tip on the tree, enter the number of the tip. Tip numbers are highlighted in yellow.

```{r eval=FALSE}
new_tree <- absent_tippr(tree, absent_list)
```

## Adding tips at random
Or, if there are no congeners, you may choose to add tips at random:

```{r warning=FALSE}
new_tree <- rand_absent_tippr(tree, absent_list)
plot(new_tree)
```

Running the command again to generate a second random addition tree, and comparing it to the first using the RF distance [@Robinson1981], shows us that the tree topologies truly are different. 

```{r warning=FALSE}
second_random <- rand_absent_tippr(tree, absent_list)
phangorn::RF.dist(new_tree, second_random)
```

## Adding tips via CSV

Lastly, you may have a TSV file that specifies the tips to be added, and a taxon set. treeStartR will locate the MRCA of the taxon set, and add the tips subtending that node.

```{r warning=FALSE}
new_tree <- text_placr(tree, mrca_df)
plot(new_tree)
```

## Resolving polytomies

By default, treeStartR outputs trees with polytomies. This is because the placement of tips is not being estimated from data, but rather placed arbitrarily. RevBayes and BEAST2 read non-bifurcating trees. If you are working with analytical software that does not, you may want to resolve polytomies before export, such as with ape's `multi2di` [function](https://www.rdocumentation.org/packages/ape/versions/5.1/topics/multi2di):

```{r}
ape::multi2di(new_tree)
```

## Subtrees and Clade Constraints

All tip addition functions, except `absent_tippr` have the capability to write out [RevBayes-formatted](https://revbayes.github.io/tutorials/fbd/#Exercise-ModelFBD) clade constraint strings. The `echo_revbayes` argument allows this.

```{r warning=FALSE}
text_placr(tree, mrca_df, echo_revbayes = TRUE)
```

Likewise, subtrees with tips added can be printed to the screen.

```{r warning=FALSE}
present_tippr(tree, absent_list, echo_subtrees = TRUE)
```


## Outputting results

The final tree can be output using standard functions in ape, such as the  `write.nexus()` [function](https://www.rdocumentation.org/packages/ape/versions/5.2/topics/write.tree):

```{r eval=FALSE}
ape::write.nexus(new_tree, file = "data/export_example.tre")
```

## Flowchart

Here is the flowchart of the functions in `treeStartR`, and when to use them.

![](images/flowchart/flowchart.001.png)

[![ropensci_footer](https://ropensci.org/public_images/ropensci_footer.png)](https://ropensci.org)

## References
As contributors and maintainers of this project, we pledge to respect all people who contribute through reporting issues, 
posting feature requests, updating documentation, submitting pull requests or patches, and other activities.

treestartr is dedicated to providing a harassment-free community experience for all participants 
regardless of gender, gender identity and expression, sexual orientation, disability, 
physical appearance, body size, race, age or religion, familial status, or socioeconomic status. 
The developer, contributors, and users do not tolerate harassment in any form in communications
related to this project, whether those communications are public (comments on issues or pull requests,
discussions at conferences), or private (e-mails or direct messages)

Examples of unacceptable behavior by participants include the use of sexual language or 
imagery, derogatory comments or personal attacks, trolling, public or private harassment, 
insults, or other unprofessional conduct. 

Project maintainers have the right and responsibility to remove, edit, or reject comments, 
commits, code, wiki edits, issues, and other contributions that are not aligned to this 
Code of Conduct. Project maintainers also reserve the right to block users who refuse to
follow the policies of this document. Project maintainers who do not follow the Code of 
Conduct may be removed from the project team.

Instances of abusive, harassing, or otherwise unacceptable behavior may be reported by 
opening an issue or contacting maintainer [April Wright](wright.aprilm@gmail.com) directly.
treeStartR is an [ROpenSci package](https://ropensci.org/), and thus covered by its [code of conduct](https://ropensci.org/coc/).
If, for any reason, you feel uncomfortable reporting violations to treeStartR personnel,
they can be reported to the [ROpenSci conduct team](conduct@ropensci.org).  


This Code of Conduct is adapted from the Contributor Covenant, version 1.0.0, 
available at https://www.contributor-covenant.org/version/1/0/0/code-of-conduct.html
# CONTRIBUTING #

### Fixing typos

Small typos or grammatical errors in documentation may be edited directly using
the GitHub web interface, so long as the changes are made in the _source_ file.

*  YES: you edit a roxygen comment in a `.R` file below `R/`.
*  NO: you edit an `.Rd` file below `man/`.

### Prerequisites

Before you make a substantial pull request, you should always file an issue and
make sure someone from the team agrees that it’s a problem. If you’ve found a
bug, create an associated issue and illustrate the bug with a minimal 
[reprex](https://www.tidyverse.org/help/#reprex).

### Pull request process

*  We recommend that you create a Git branch for each pull request (PR).  
*  Look at the Travis build status before and after making changes.
The `README` should contain badges for any continuous integration services used
by the package.  
*  We recommend the tidyverse [style guide](http://style.tidyverse.org).
You can use the [styler](https://CRAN.R-project.org/package=styler) package to
apply these styles, but please don't restyle code that has nothing to do with 
your PR.  
*  We use [roxygen2](https://cran.r-project.org/package=roxygen2).  
*  We use [testthat](https://cran.r-project.org/package=testthat). Contributions
with test cases included are easier to accept.  
*  For user-facing changes, add a bullet to the top of `NEWS.md` below the
current development version header describing the changes made followed by your
GitHub username, and links to relevant issue(s)/PR(s).

### Code of Conduct

Please note that the {{{ package }}} project is released with a
[Contributor Code of Conduct](CODE_OF_CONDUCT.md). By contributing to this
project you agree to abide by its terms.

### See rOpenSci [contributing guide](https://ropensci.github.io/dev_guide/contributingguide.html)
for further details.

### Discussion forum

Check out our [discussion forum](https://discuss.ropensci.org) if you think your issue requires a longer form discussion.

### Prefer to Email? 

Email the person listed as maintainer in the `DESCRIPTION` file of this repo.

Though note that private discussions over email don't help others - of course email is totally warranted if it's a sensitive problem of any kind.

### Thanks for contributing!

This contributing guide is adapted from the tidyverse contributing guide available at https://raw.githubusercontent.com/r-lib/usethis/master/inst/templates/tidy-contributing.md 
