If you use GRNsight in your work, please cite:

Dahlquist, K.D., Dionisio, J.D.N., Fitzpatrick, B.G., Anguiano, N.A., Varshneya, A., Southwick, B.J., Samdarshi, M. (2016) GRNsight: a web application and service for visualizing models of small- to medium-scale gene regulatory networks. _PeerJ Computer Science_ 2:e85. [DOI: 10.7717/peerj-cs.85](https://doi.org/10.7717/peerj-cs.85).

The DOI for the latest release is:

[![DOI](https://zenodo.org/badge/16195791.svg)](https://zenodo.org/badge/latestdoi/16195791)
GRNsight
========
[![DOI](https://zenodo.org/badge/16195791.svg)](https://zenodo.org/badge/latestdoi/16195791)
[![Build Status](https://app.travis-ci.com/dondi/GRNsight.svg?branch=master)](https://app.travis-ci.com/dondi/GRNsight)
[![Coverage Status](https://coveralls.io/repos/github/dondi/GRNsight/badge.svg?branch=master)](https://coveralls.io/github/dondi/GRNsight?branch=master)

http://dondi.github.io/GRNsight/

GRNsight is an open source web application and service for visualizing models of small- to medium-scale gene regulatory networks. GRNsight is a joint project of the [Loyola Marymount University](http://www.lmu.edu) Bioinformatics and Biomathematics Groups, headed by [Dr. Kam Dahlquist](http://myweb.lmu.edu/kdahlqui/index.htm), [Dr. John David N. Dionisio](http://myweb.lmu.edu/dondi/), and [Dr. Ben G. Fitzpatrick](http://myweb.lmu.edu/bfitzpatrick/). Undergraduate students initiated the development of GRNsight in Spring 2014, including Britain Southwick (Computer Science, ’14) and Nicole Anguiano (Computer Science, ’16), with consultation from Katrina Sherbina (Biomathematics, ’14). For current contributors, please see our [People](http://dondi.github.io/GRNsight/people.html) page.

A gene regulatory network (GRN) consists of genes, transcription factors, and the regulatory connections between them, which govern the level of expression of mRNA and protein from those genes. GRNs can be mathematically modeled and simulated by applications such as [GRNmap](http://kdahlquist.github.io/GRNmap/), a MATLAB program that estimates the parameters and performs forward simulations of a differential equations model of a GRN. Computer representations of GRNs, such as the models output by GRNmap, are in the form of a tabular spreadsheet (adjacency matrix) that is not easily interpretable. Ideally, GRNs should be displayed as diagrams (graphs) detailing the regulatory relationships (edges) between each gene (node) in the network. To address this need, we developed GRNsight.

GRNsight allows users to upload Excel workbooks generated by GRNmap (as well as Simple Interaction Format text files and GraphML XML files) and uses the network information to automatically create and display a graph of the GRN model. The application colors the edges and adjusts their thickness based on the sign (activation or repression) and the strength (magnitude) of the regulatory relationship, respectively. Finally, GRNsight then allows the user to modify the graph in order to define the best visual layout for the network. Most of GRNsight is written in JavaScript. HTTP requests are handled using Node.js and the Express framework. Graphs are generated through D3.js, a JavaScript data visualization library.

Although originally designed for gene regulatory networks, we believe that GRNsight has general applicability for displaying any small, unweighted or weighted network with directed edges for systems biology or other application domains.

Most users will want to access GRNsight through the web application at http://dondi.github.io/GRNsight/.  The source code is available for developers who wish to run their own instance of the GRNsight web service and/or web client.

Documentation on how to use GRNsight is found at https://github.com/dondi/GRNsight, with additional information on the wiki here: https://github.com/dondi/GRNsight/wiki.

If you use GRNsight in your work, please cite:

Dahlquist, K.D., Dionisio, J.D.N., Fitzpatrick, B.G., Anguiano, N.A., Varshneya, A., Southwick, B.J., Samdarshi, M. (2016) GRNsight: a web application and service for visualizing models of small- to medium-scale gene regulatory networks. _PeerJ Computer Science_ 2:e85. [DOI: 10.7717/peerj-cs.85](https://doi.org/10.7717/peerj-cs.85).
Please note that this project is released with a Contributor Code of Conduct. By participating in this project you agree to abide by its terms.

# Contributor Covenant Code of Conduct

## Our Pledge

In the interest of fostering an open and welcoming environment, we as
contributors and maintainers pledge to making participation in our project and
our community a harassment-free experience for everyone, regardless of age, body
size, disability, ethnicity, gender identity and expression, level of experience,
nationality, personal appearance, race, religion, or sexual identity and
orientation.

## Our Standards

Examples of behavior that contributes to creating a positive environment
include:

* Using welcoming and inclusive language
* Being respectful of differing viewpoints and experiences
* Gracefully accepting constructive criticism
* Focusing on what is best for the community
* Showing empathy towards other community members

Examples of unacceptable behavior by participants include:

* The use of sexualized language or imagery and unwelcome sexual attention or
advances
* Trolling, insulting/derogatory comments, and personal or political attacks
* Public or private harassment
* Publishing others' private information, such as a physical or electronic
  address, without explicit permission
* Other conduct which could reasonably be considered inappropriate in a
  professional setting

### Loyola Marymount University Policies

Loyola Marymount University student participants in this project are also bound by the Lion's Code and the LMU Student Conduct Code found in the  and the [LMU Community Standards Publication](http://studentaffairs.lmu.edu/media/lmustudentaffairs/administration/judicialaffairs/Community%20Standards%20Booklet.pdf) and the [LMU Academic Honesty Policy](http://academics.lmu.edu/honesty). 

## Our Responsibilities

Project maintainers are responsible for clarifying the standards of acceptable
behavior and are expected to take appropriate and fair corrective action in
response to any instances of unacceptable behavior.

Project maintainers have the right and responsibility to remove, edit, or
reject comments, commits, code, wiki edits, issues, and other contributions
that are not aligned to this Code of Conduct, or to ban temporarily or
permanently any contributor for other behaviors that they deem inappropriate,
threatening, offensive, or harmful.

## Scope

This Code of Conduct applies both within project spaces and in public spaces
when an individual is representing the project or its community. Examples of
representing a project or community include using an official project e-mail
address, posting via an official social media account, or acting as an appointed
representative at an online or offline event. Representation of a project may be
further defined and clarified by project maintainers.

## Enforcement

Instances of abusive, harassing, or otherwise unacceptable behavior may be
reported by contacting the Principal Investigators, Kam D. Dahlquist ([kdahlquist@lmu.edu](mailto:kdahlquist@lmu.edu)) or John David N. Dionsio ([dondi@lmu.edu](mailto:dondi@lmu.edu)). All
complaints will be reviewed and investigated and will result in a response that
is deemed necessary and appropriate to the circumstances. The project team is
obligated to maintain confidentiality with regard to the reporter of an incident.
Further details of specific enforcement policies may be posted separately.

Project maintainers who do not follow or enforce the Code of Conduct in good
faith may face temporary or permanent repercussions as determined by other
members of the project's leadership.

## Attribution

This Code of Conduct is adapted from the [Contributor Covenant][homepage], version 1.4,
available at [http://contributor-covenant.org/version/1/4][version]

[homepage]: http://contributor-covenant.org
[version]: http://contributor-covenant.org/version/1/4/

### License

This Code of Conduct is used under the The MIT License (MIT), which requires the inclusion of the following statement below:

Copyright (c) 2014 Coraline Ada Ehmke

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
the Software, and to permit persons to whom the Software is furnished to do so,
subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
The GRNsight server is a node.js based service. More information to follow.The web client for GRNsightOur Team was tasked with designing the api framework for retrieving the information from the 5 databases through the getGeneInformation function.
This function takes in a gene symbol as a parameter, and returns a deferred jQuery object that ultimately resolves to an object containing the values found from the search. If there is an error in contacting the databases, the function alerts to the main page that an error occurred.

In the function, there were some pieces of data that were requested but ultimately unable to be retrieved through the API endpoints we were tasked with retrieving. These pieces of data are noted in the function with a comment: // Information unavailable via regular API.
Our Team was Tasked with Designing the Page that Hosts all the Data that is pulled from the various biological databases used. There are not many specific features that are unique to the page design team, but we utilized the bootstrap front-end framework to display all the information in the most easily understandable and aesthetically pleasing fashion.  We used, description lists, tables, and some style tools such as color and "text-muted".
**Blair Hamilton and Zach Van Ysseldyk’s README File**

# General Information

The Integration and Interaction team has implemented a right-click function thats purpose is to open a new tab and produce a gene page with data

brought in from multiple credible databases. The primary goal of our info.js, info.html and api.js files is to bring in the work of the gene database

API team, JASPAR team and the page design team, and produce a working demo of the GRNsight right click function.

# Right-Click Functionality

First, the right click function bridges the original GRNSight existing gene page with the new gene page. The GRNSight gene regulatory map depicts a

relationship between each gene. If a user would like to know more information on one of those genes depicted in the regulatory map, they may do so by

right clicking on that gene.

The following section will delve into the actual code:

-First, the graph.js file contains a function called “dbclick” which refers to the right click functionality.

-Next, the “_ blank” syntax opens up into a new tab.

-Unfortunately, due to data security concerns, the syntax for opening up a link in a new tab has been taken down from the D3 help page. Because of this, the integration team had to circumnavigate this problem by adding a fake anchor.

-The fake anchor tag in the beginning of the function conceptually creates a fake, temporary landing page for the URL. Once the URL has landed on the anchor, the anchor is taken away via the tempLink.remove() function. Once the landing page has been taken away, the new link can now live on its own.

The link itself is broken up into two parts.

1.The first part is a template URL.

2.The parameter portion refers to the gene name being pulled from the selected right-click node.

# Variable Links

In collaboration with the Page-Design team a layout for specific data points was given to the Interaction team in order to put the correct variable

link with data point in divided tabs.

-For this portion a hreftemplate link was made in the info.js file for each of the five databases in which a generic gene page URL from the respective database can be called upon.

-Next variable links were made between the info.html page classes and api.js API calls in order to produce data on the gene page.
  -This can be seen in the info.js file which contains comments representing the different tabs on the gene page.

-Four individual for-loops were created for the frequency matrix in order to account for the difference in data per gene.

	-For example, some yeast genes contain anywhere from 0-15 data points for the frequency matrix involving: A,T,G, and C dna points

-Lastly, the dnaSequence and Gene Map sections are currently empty due to API difficulties in the data retrieval, but currently have a placeholder for when the API call is working.

-All desired data points have received a variable link and a spot in the gene page template. If no data currently shows then it is left blank, but otherwise is functional.
