<p><img src="./figures/vulcanspot.png" height="70" align="middle"> a tool to prioritize therapeutic vulnerabilities in cancer.</p>

> The tool could be accessed at http://www.vulcanspot.org/

## Graphical abstract
![Alt text](./figures/vulcanspot_workflow.png)

## Abstract
Genetic alterations lead to tumour progression and survival in cancer, but also uncover gene dependencies such as oncogenic addictions and synthetic lethals, which could be exploited to extend the current catalog of molecularly matched treatments for precision medicine. VulcanSpot is a novel computational approach that exploits the notion of collateral therapeutic vulnerabilities caused by the acquisition of cancer mutations. To this aim, our method mines genomic profiles from cancer cell lines and genome-wide gene loss-of-function screenings to identify potential vulnerabilities in cancer. Then, vulcanSpot prioritizes drugs to target genotype-selective gene dependencies using a weighted scoring system that integrates two complementary strategies for computational drug prescription: i) Pandrugs - a comprehensive database of known gene-drug relationships, and ii) a novel drug repositioning method that matches drugs whose transcriptional signature mimics the functional depletion of the target gene.

## Requirements
* bash (UNIX)
* sed (UNIX)
* find (UNIX)
* python2.7 (UNIX)
* R (UNIX)
- - - 
Vulcanspot's pipeline is based on R scripts that process, transform and compare data from different biological sources. The following R packages are required (R version 3.4.4):

| package name | source | Recommended version |
| ---------- | ------- | --------- |
| `biomaRt` | Bioconductor | 2.34.2 |
| `GenomicFeatures` | Bioconductor | 1.30.3 |
| `SCAN.UPC` | Bioconductor | 2.20.0 |
| `fgsea` | Bioconductor | 1.4.1 |
| `igraph` | Bioconductor | 1.2.2 |
| `reshape2` | CRAN | 1.4.3 |
| `mygene` | Bioconductor | 1.14.0 |

- - - 
<img src="./figures/BU.png" height="100" align="left">

![Alt text](./figures/cnio_stopcancer.png)
The .architect file can be edited with [SQL Power Architect](http://www.bestofbi.com/page/architect_download_os)
# Drug-Protein-Protein Interaction Network
Herein, universal (pan-cancer) and contextualized (only proteins that are expressed in cell lines from a given cell lineage) are presented. These networks are generated by aggregating the drug-target interactions described in the CLUE database (clue.io/api) with a Protein-Protein Interaction network (provided by PINA, only physical interactions described in more than one resource). The scripts that generate these networks are coordinated in `./src/04_drug_prescription.sh`:

*  `./src/__NET_create_DrugProteinProtein_interaction_usingPINA.R`
* `./src/__NET_contextualize_DrugPPIN.R`

The original file (`PINA_Homosapiens-20140521.tsv`) from PINA is downloaded automatically, but not stored in the repository because of large file size (144M).
# CMap-L1000 data for KD-CP perturbations
The file sizes are very large to be included in the repository. The data can be downloaded from https://figshare.com/s/b34fc8e7f4d77ca6d580

The files contains the TAU scores from the signature maching between knock-down and compound cellular perturbations in cancer cell lines. In brief, transcriptional signatures of gene knock-down (KD) and compound (CP) perturbations in cancer cell lines that belong to the same context was obtained using limma (https://github.com/jperales/CLUE). Then, a Total Enrichment Score (TES) was calculated by each pair of KD-CP (Iorio et al. PNAS 2010). Finally, the TES were scaled using the TAU approach proposed in Subramanian et al 2017.

Files:
* `BREAST.TAU_matrix.rds`
* `CMap_contexts.txt`
* `KIDNEY.TAU_matrix.rds`
* `LARGE_INTESTINE.TAU_matrix.rds`
* `LIVER.TAU_matrix.rds`
* `LUNG.TAU_matrix.rds`
* `PANCANCER.TAU_matrix.rds`
* `PROSTATE.TAU_matrix.rds`
* `SKIN.TAU_matrix.rds`
# Molecular Profiles of cancer cell lines
This folder contains the molecular profiles of cancer cell lines in the raw and processed forms. File sizes are very large, so an ad-hoc script (`02_CCL_profiles.sh`) downloads and processed the data from the original data portals.

| Filename | Description |
| ----- | ----- |
| `CCLE_DepMap_18Q2_maf_20180502.txt` | Merged mutation calls (coding region, germline filtered) |
| `CCLE_copynumber_byGene_2013-12-03.txt` | Copy-number values per gene. Affymetrix SNP6.0 arrays |
| `CCLE_DepMap_18Q2_RNAseq_reads_20180502.gct` | CCLE RNAseq gene expression data (read count) |
| `portal-Avana-2018-05-10.csv` | CERES inferred gene effect matrix (Meyers et al. 2017) |
| `portal-RNAi_merged-2018-05-10.csv` | DEMETER2 inferred gene dependency scores (McFarland et al. 2018) |

The latest two files data access are now depreated. You can find a mirror of the original data at https://figshare.com/s/1c2f1319b7c95067fad6
